# Comparing `tmp/geonomics-1.3.9.tar.gz` & `tmp/geonomics-1.4.0.tar.gz`

## filetype from file(1)

```diff
@@ -1 +1 @@
-gzip compressed data, was "geonomics-1.3.9.tar", last modified: Thu Apr 13 16:57:12 2023, max compression
+gzip compressed data, was "geonomics-1.4.0.tar", last modified: Mon Apr  1 13:05:52 2024, max compression
```

## Comparing `geonomics-1.3.9.tar` & `geonomics-1.4.0.tar`

### file list

```diff
@@ -1,130 +1,130 @@
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.940418 geonomics-1.3.9/
--rwxrwxr-x   0 deth      (1000) deth      (1000)     1112 2021-12-01 11:42:45.000000 geonomics-1.3.9/LICENSE.txt
--rwxrwxr-x   0 deth      (1000) deth      (1000)       74 2021-03-24 04:15:31.000000 geonomics-1.3.9/MANIFEST.in
--rw-rw-r--   0 deth      (1000) deth      (1000)     9440 2023-04-13 16:57:12.940418 geonomics-1.3.9/PKG-INFO
--rw-rw-r--   0 deth      (1000) deth      (1000)     8326 2022-01-01 18:34:30.000000 geonomics-1.3.9/README.rst
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.896418 geonomics-1.3.9/geonomics/
--rwxrwxr-x   0 deth      (1000) deth      (1000)      499 2022-01-01 17:50:26.000000 geonomics-1.3.9/geonomics/__init__.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.896418 geonomics-1.3.9/geonomics/data/
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.904418 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/
--rwxrwxr-x   0 deth      (1000) deth      (1000)     1573 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/MMRR.R
--rwxrwxr-x   0 deth      (1000) deth      (1000)     3514 2022-09-11 23:56:38.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/MMRR.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.908418 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/__pycache__/
--rwxrwxr-x   0 deth      (1000) deth      (1000)     3091 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-38.pyc
--rwxrwxr-x   0 deth      (1000) deth      (1000)     3067 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-39.pyc
--rwxrwxr-x   0 deth      (1000) deth      (1000)  1392400 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/env.csv
--rwxrwxr-x   0 deth      (1000) deth      (1000)  1392400 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/gen.csv
--rwxrwxr-x   0 deth      (1000) deth      (1000)  1392400 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/geo.csv
--rwxrwxr-x   0 deth      (1000) deth      (1000)      898 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/IBD_IBE_demo/run_mantel.R
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.908418 geonomics-1.3.9/geonomics/data/default_models/
--rw-rw-r--   0 deth      (1000) deth      (1000)    14817 2021-10-09 01:45:22.000000 geonomics-1.3.9/geonomics/data/default_models/selection_params.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.908418 geonomics-1.3.9/geonomics/data/yosemite_demo/
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35632 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_DEM_90x90.tif
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.908418 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.916418 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34429 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/509_ppt_2015_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34416 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/514_ppt_2020_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34424 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/519_ppt_2025_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34399 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/524_ppt_2030_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34379 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/529_ppt_2035_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34386 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/534_ppt_2040_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34604 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/539_ppt_2045_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34326 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/544_ppt_2050_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34361 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/549_ppt_2055_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34407 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/554_ppt_2060_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34382 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/559_ppt_2065_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34292 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/564_ppt_2070_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34435 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/569_ppt_2075_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34386 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/574_ppt_2080_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34416 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/579_ppt_2085_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34351 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/584_ppt_2090_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34256 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/589_ppt_2095_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    34516 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/594_ppt_2100_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    29050 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt_1980-2010_90x90.tif
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.920418 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35877 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/509_sdm_2015_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36099 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/514_sdm_2020_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35864 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/519_sdm_2025_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35990 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/524_sdm_2030_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35953 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/529_sdm_2035_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35962 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/534_sdm_2040_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36177 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/539_sdm_2045_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35922 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/544_sdm_2050_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35910 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/549_sdm_2055_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35932 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/554_sdm_2060_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35896 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/559_sdm_2065_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35910 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/564_sdm_2070_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36275 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/569_sdm_2075_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36218 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/574_sdm_2080_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36091 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/579_sdm_2085_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36181 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/584_sdm_2090_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    35874 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/589_sdm_2095_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36217 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/594_sdm_2100_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36109 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm_1980-2010_90x90.tif
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.928418 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31648 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/509_tmp_2015_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31717 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/514_tmp_2020_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31807 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/519_tmp_2025_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31507 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/524_tmp_2030_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31710 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/529_tmp_2035_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31645 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/534_tmp_2040_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31584 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/539_tmp_2045_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31437 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/544_tmp_2050_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31454 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/549_tmp_2055_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31455 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/554_tmp_2060_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31446 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/559_tmp_2065_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31556 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/564_tmp_2070_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31499 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/569_tmp_2075_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31484 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/574_tmp_2080_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31434 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/579_tmp_2085_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31516 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/584_tmp_2090_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31416 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/589_tmp_2095_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    31225 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/594_tmp_2100_90x90.tif
--rwxrwxr-x   0 deth      (1000) deth      (1000)    20497 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp_1980-2010_90x90.tif
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.928418 geonomics-1.3.9/geonomics/demos/
--rwxrwxr-x   0 deth      (1000) deth      (1000)    40012 2021-08-30 15:43:03.000000 geonomics-1.3.9/geonomics/demos/_IBD_IBE.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)       76 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/demos/__init__.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    23746 2021-08-30 20:54:50.000000 geonomics-1.3.9/geonomics/demos/_simult_select.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    40559 2021-08-30 20:54:24.000000 geonomics-1.3.9/geonomics/demos/_yosemite.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.932418 geonomics-1.3.9/geonomics/help/
--rwxrwxr-x   0 deth      (1000) deth      (1000)      728 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/help/__init__.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     2260 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/help/_memory_help.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    19956 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/help/_param_help.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    38124 2022-01-01 17:50:11.000000 geonomics-1.3.9/geonomics/main.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.936418 geonomics-1.3.9/geonomics/ops/
--rwxrwxr-x   0 deth      (1000) deth      (1000)        0 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/ops/__init__.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    36149 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/ops/change.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    13079 2023-03-19 21:57:12.000000 geonomics-1.3.9/geonomics/ops/demography.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    10192 2021-09-21 22:53:47.000000 geonomics-1.3.9/geonomics/ops/mating.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     6555 2021-03-24 04:15:27.000000 geonomics-1.3.9/geonomics/ops/movement.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     7761 2021-07-09 03:11:40.000000 geonomics-1.3.9/geonomics/ops/mutation.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     5094 2021-07-09 03:11:40.000000 geonomics-1.3.9/geonomics/ops/selection.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.936418 geonomics-1.3.9/geonomics/sim/
--rwxrwxr-x   0 deth      (1000) deth      (1000)        0 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/sim/__init__.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     3405 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/sim/burnin.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    23562 2022-12-26 20:26:33.000000 geonomics-1.3.9/geonomics/sim/data.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)   132048 2023-02-15 16:08:40.000000 geonomics-1.3.9/geonomics/sim/model.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    45641 2022-12-26 20:55:09.000000 geonomics-1.3.9/geonomics/sim/params.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    19972 2023-04-13 16:46:44.000000 geonomics-1.3.9/geonomics/sim/stats.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.940418 geonomics-1.3.9/geonomics/structs/
--rwxrwxr-x   0 deth      (1000) deth      (1000)      116 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/structs/__init__.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     5433 2021-07-09 03:11:40.000000 geonomics-1.3.9/geonomics/structs/community.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    63099 2022-09-04 20:07:18.000000 geonomics-1.3.9/geonomics/structs/genome.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)     8672 2021-07-09 03:11:40.000000 geonomics-1.3.9/geonomics/structs/individual.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    32817 2021-07-09 03:11:40.000000 geonomics-1.3.9/geonomics/structs/landscape.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)   120572 2023-02-15 14:34:17.000000 geonomics-1.3.9/geonomics/structs/species.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.940418 geonomics-1.3.9/geonomics/utils/
--rwxrwxr-x   0 deth      (1000) deth      (1000)       82 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/utils/__init__.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)      607 2021-03-24 04:15:28.000000 geonomics-1.3.9/geonomics/utils/_str_repr_.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    10807 2023-04-13 16:46:40.000000 geonomics-1.3.9/geonomics/utils/io.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    21410 2022-12-26 20:39:40.000000 geonomics-1.3.9/geonomics/utils/spatial.py
--rwxrwxr-x   0 deth      (1000) deth      (1000)    16746 2021-03-24 04:15:29.000000 geonomics-1.3.9/geonomics/utils/viz.py
--rw-rw-r--   0 deth      (1000) deth      (1000)       22 2023-04-13 16:54:23.000000 geonomics-1.3.9/geonomics/version.py
-drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2023-04-13 16:57:12.896418 geonomics-1.3.9/geonomics.egg-info/
--rwxrwxr-x   0 deth      (1000) deth      (1000)     9440 2023-04-13 16:57:12.000000 geonomics-1.3.9/geonomics.egg-info/PKG-INFO
--rwxrwxr-x   0 deth      (1000) deth      (1000)     5542 2023-04-13 16:57:12.000000 geonomics-1.3.9/geonomics.egg-info/SOURCES.txt
--rwxrwxr-x   0 deth      (1000) deth      (1000)        1 2023-04-13 16:57:12.000000 geonomics-1.3.9/geonomics.egg-info/dependency_links.txt
--rwxrwxr-x   0 deth      (1000) deth      (1000)       86 2023-04-13 16:57:12.000000 geonomics-1.3.9/geonomics.egg-info/requires.txt
--rwxrwxr-x   0 deth      (1000) deth      (1000)       10 2023-04-13 16:57:12.000000 geonomics-1.3.9/geonomics.egg-info/top_level.txt
--rwxrwxr-x   0 deth      (1000) deth      (1000)      107 2023-04-13 16:57:12.940418 geonomics-1.3.9/setup.cfg
--rwxrwxr-x   0 deth      (1000) deth      (1000)     2680 2022-01-01 17:57:19.000000 geonomics-1.3.9/setup.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.293627 geonomics-1.4.0/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     1112 2024-01-03 01:50:01.000000 geonomics-1.4.0/LICENSE.txt
+-rwxrwxr-x   0 deth      (1000) deth      (1000)       74 2021-03-24 04:15:31.000000 geonomics-1.4.0/MANIFEST.in
+-rw-rw-r--   0 deth      (1000) deth      (1000)     9440 2024-04-01 13:05:52.293627 geonomics-1.4.0/PKG-INFO
+-rw-rw-r--   0 deth      (1000) deth      (1000)     8326 2022-01-01 18:34:30.000000 geonomics-1.4.0/README.rst
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.241626 geonomics-1.4.0/geonomics/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)      499 2022-01-01 17:50:26.000000 geonomics-1.4.0/geonomics/__init__.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.241626 geonomics-1.4.0/geonomics/data/
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.253626 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     1573 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/MMRR.R
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     2807 2023-06-14 19:20:14.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/MMRR.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.257627 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/__pycache__/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     3091 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-38.pyc
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     3067 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-39.pyc
+-rwxrwxr-x   0 deth      (1000) deth      (1000)  1392400 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/env.csv
+-rwxrwxr-x   0 deth      (1000) deth      (1000)  1392400 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/gen.csv
+-rwxrwxr-x   0 deth      (1000) deth      (1000)  1392400 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/geo.csv
+-rwxrwxr-x   0 deth      (1000) deth      (1000)      898 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/IBD_IBE_demo/run_mantel.R
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.257627 geonomics-1.4.0/geonomics/data/default_models/
+-rw-rw-r--   0 deth      (1000) deth      (1000)    14817 2021-10-09 01:45:22.000000 geonomics-1.4.0/geonomics/data/default_models/selection_params.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.257627 geonomics-1.4.0/geonomics/data/yosemite_demo/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35632 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_DEM_90x90.tif
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.257627 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.265627 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34429 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/509_ppt_2015_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34416 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/514_ppt_2020_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34424 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/519_ppt_2025_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34399 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/524_ppt_2030_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34379 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/529_ppt_2035_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34386 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/534_ppt_2040_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34604 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/539_ppt_2045_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34326 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/544_ppt_2050_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34361 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/549_ppt_2055_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34407 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/554_ppt_2060_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34382 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/559_ppt_2065_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34292 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/564_ppt_2070_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34435 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/569_ppt_2075_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34386 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/574_ppt_2080_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34416 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/579_ppt_2085_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34351 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/584_ppt_2090_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34256 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/589_ppt_2095_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    34516 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/594_ppt_2100_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    29050 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt_1980-2010_90x90.tif
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.277627 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35877 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/509_sdm_2015_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36099 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/514_sdm_2020_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35864 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/519_sdm_2025_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35990 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/524_sdm_2030_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35953 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/529_sdm_2035_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35962 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/534_sdm_2040_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36177 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/539_sdm_2045_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35922 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/544_sdm_2050_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35910 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/549_sdm_2055_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35932 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/554_sdm_2060_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35896 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/559_sdm_2065_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35910 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/564_sdm_2070_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36275 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/569_sdm_2075_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36218 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/574_sdm_2080_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36091 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/579_sdm_2085_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36181 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/584_sdm_2090_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    35874 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/589_sdm_2095_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36217 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/594_sdm_2100_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36109 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm_1980-2010_90x90.tif
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.285627 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31648 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/509_tmp_2015_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31717 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/514_tmp_2020_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31807 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/519_tmp_2025_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31507 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/524_tmp_2030_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31710 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/529_tmp_2035_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31645 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/534_tmp_2040_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31584 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/539_tmp_2045_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31437 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/544_tmp_2050_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31454 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/549_tmp_2055_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31455 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/554_tmp_2060_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31446 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/559_tmp_2065_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31556 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/564_tmp_2070_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31499 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/569_tmp_2075_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31484 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/574_tmp_2080_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31434 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/579_tmp_2085_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31516 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/584_tmp_2090_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31416 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/589_tmp_2095_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    31225 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/594_tmp_2100_90x90.tif
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    20497 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp_1980-2010_90x90.tif
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.285627 geonomics-1.4.0/geonomics/demos/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    40012 2021-08-30 15:43:03.000000 geonomics-1.4.0/geonomics/demos/_IBD_IBE.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)       76 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/demos/__init__.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    23746 2021-08-30 20:54:50.000000 geonomics-1.4.0/geonomics/demos/_simult_select.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    40559 2021-08-30 20:54:24.000000 geonomics-1.4.0/geonomics/demos/_yosemite.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.285627 geonomics-1.4.0/geonomics/help/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)      728 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/help/__init__.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     2260 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/help/_memory_help.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    19956 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/help/_param_help.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    42853 2024-04-01 12:51:51.000000 geonomics-1.4.0/geonomics/main.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.289627 geonomics-1.4.0/geonomics/ops/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)        0 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/ops/__init__.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    36149 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/ops/change.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    11576 2024-03-26 13:35:21.000000 geonomics-1.4.0/geonomics/ops/demography.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    10192 2021-09-21 22:53:47.000000 geonomics-1.4.0/geonomics/ops/mating.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     6555 2021-03-24 04:15:27.000000 geonomics-1.4.0/geonomics/ops/movement.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     8345 2024-03-29 11:47:20.000000 geonomics-1.4.0/geonomics/ops/mutation.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     5094 2021-07-09 03:11:40.000000 geonomics-1.4.0/geonomics/ops/selection.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.289627 geonomics-1.4.0/geonomics/sim/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)        0 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/sim/__init__.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     3405 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/sim/burnin.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    23562 2022-12-26 20:26:33.000000 geonomics-1.4.0/geonomics/sim/data.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)   142781 2024-04-01 03:38:11.000000 geonomics-1.4.0/geonomics/sim/model.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    49119 2024-04-01 12:20:05.000000 geonomics-1.4.0/geonomics/sim/params.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    19972 2023-04-13 16:46:44.000000 geonomics-1.4.0/geonomics/sim/stats.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.289627 geonomics-1.4.0/geonomics/structs/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)      116 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/structs/__init__.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     5433 2021-07-09 03:11:40.000000 geonomics-1.4.0/geonomics/structs/community.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    87852 2024-03-31 14:11:08.000000 geonomics-1.4.0/geonomics/structs/genome.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     8672 2021-07-09 03:11:40.000000 geonomics-1.4.0/geonomics/structs/individual.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    32817 2021-07-09 03:11:40.000000 geonomics-1.4.0/geonomics/structs/landscape.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)   153475 2024-04-01 03:44:04.000000 geonomics-1.4.0/geonomics/structs/species.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.293627 geonomics-1.4.0/geonomics/utils/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)       82 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/utils/__init__.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)      607 2021-03-24 04:15:28.000000 geonomics-1.4.0/geonomics/utils/_str_repr_.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    10991 2023-04-13 17:58:34.000000 geonomics-1.4.0/geonomics/utils/io.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    21410 2022-12-26 20:39:40.000000 geonomics-1.4.0/geonomics/utils/spatial.py
+-rwxrwxr-x   0 deth      (1000) deth      (1000)    16746 2021-03-24 04:15:29.000000 geonomics-1.4.0/geonomics/utils/viz.py
+-rw-rw-r--   0 deth      (1000) deth      (1000)       22 2024-04-01 03:57:06.000000 geonomics-1.4.0/geonomics/version.py
+drwxrwxr-x   0 deth      (1000) deth      (1000)        0 2024-04-01 13:05:52.245627 geonomics-1.4.0/geonomics.egg-info/
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     9440 2024-04-01 13:05:51.000000 geonomics-1.4.0/geonomics.egg-info/PKG-INFO
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     5542 2024-04-01 13:05:52.000000 geonomics-1.4.0/geonomics.egg-info/SOURCES.txt
+-rwxrwxr-x   0 deth      (1000) deth      (1000)        1 2024-04-01 13:05:51.000000 geonomics-1.4.0/geonomics.egg-info/dependency_links.txt
+-rwxrwxr-x   0 deth      (1000) deth      (1000)       86 2024-04-01 13:05:51.000000 geonomics-1.4.0/geonomics.egg-info/requires.txt
+-rwxrwxr-x   0 deth      (1000) deth      (1000)       10 2024-04-01 13:05:51.000000 geonomics-1.4.0/geonomics.egg-info/top_level.txt
+-rwxrwxr-x   0 deth      (1000) deth      (1000)      107 2024-04-01 13:05:52.293627 geonomics-1.4.0/setup.cfg
+-rwxrwxr-x   0 deth      (1000) deth      (1000)     2680 2022-01-01 17:57:19.000000 geonomics-1.4.0/setup.py
```

### Comparing `geonomics-1.3.9/LICENSE.txt` & `geonomics-1.4.0/LICENSE.txt`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/PKG-INFO` & `geonomics-1.4.0/PKG-INFO`

 * *Files 1% similar despite different names*

```diff
@@ -1,13 +1,13 @@
 Metadata-Version: 2.1
 Name: geonomics
-Version: 1.3.9
+Version: 1.4.0
 Summary: A package for landscape genomic simulation
 Home-page: https://github.com/drewhart/geonomics
-Download-URL: https://github.com/drewhart/geonomics/archive/1.3.9.tar.gz
+Download-URL: https://github.com/drewhart/geonomics/archive/1.4.0.tar.gz
 Author: Drew Ellison Hart
 Author-email: drew.ellison.hart@gmail.com
 Project-URL: Documentation, https://htmlpreview.github.io/?https://github.com/drewhart/geonomics/blob/master/doc/built/doc.html
 Project-URL: Source, https://github.com/drewhart/geonomics
 Keywords: landscape genomics genetics ecology evolution simulation model environmental model agent-based
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: MIT License
```

### Comparing `geonomics-1.3.9/README.rst` & `geonomics-1.4.0/README.rst`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/MMRR.R` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/MMRR.R`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/MMRR.py` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/MMRR.py`

 * *Files 14% similar despite different names*

```diff
@@ -46,44 +46,33 @@
     Fprob = 1
     # get the row numbers
     rownums = [*range(nrowsY)]
     # perform permutations
     for i in range(nperm):
         # shuffle the row numbers
         np.random.shuffle(rownums)
-        # NOTE: I thought I had triple-checked everything in this script,
-        #       but the following line was now just returning the 0s from the
-        #       diagonal, rather than a permuted symmetric matrix,
-        #       so I corrected; cannot determine whether this was an oversight
-        #       of numpy indexing behavior shifted...
         Yperm = Y[rownums,:][:, rownums]
         yperm = _unfold_tril(Yperm)
         permmod_y, permmod_x = _prep_mod_data(yperm, xs)
         permmod = sm.OLS(permmod_y, permmod_x).fit()
-        # NOTE: previous results were probably overly optimistic
-        #       because I forgot the first np.abs
-        #       in the following line, so negative permutation t-values
-        #       larger in absolute value than a positive main t-value
-        #       would not have been factored into the t-tests'
-        #       empirical p-values!
         tprob += (np.abs(permmod.tvalues) >= np.abs(tstat))
         Fprob += (permmod.fvalue >= Fstat)
 
     # calculate the empirical p-values
     tp = tprob/(nperm+1)
     Fp = Fprob/(nperm+1)
 
     # return values
     coeff_names = ["Intercept"] + Xnames
     output = OD()
     output["R^2"] = r2
     output.update({c: cval for c, cval in zip(coeff_names, coeffs)})
     output.update({c+ "(t)": tval for c, tval in zip(coeff_names, tstat)})
     output.update({c+ "(p)": pval for c, pval in zip(coeff_names, tp)})
-    output["F-statistc"] = Fstat
+    output["F-statistic"] = Fstat
     output["F p-value"] = Fp
 
     return output
 
 
 def _prep_mod_data(y, xs, intercept=True):
     """
```

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-38.pyc` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-38.pyc`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-39.pyc` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/__pycache__/MMRR.cpython-39.pyc`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/env.csv` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/env.csv`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/gen.csv` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/gen.csv`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/geo.csv` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/geo.csv`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/IBD_IBE_demo/run_mantel.R` & `geonomics-1.4.0/geonomics/data/IBD_IBE_demo/run_mantel.R`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/default_models/selection_params.py` & `geonomics-1.4.0/geonomics/data/default_models/selection_params.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_DEM_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_DEM_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/509_ppt_2015_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/509_ppt_2015_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/514_ppt_2020_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/514_ppt_2020_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/519_ppt_2025_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/519_ppt_2025_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/524_ppt_2030_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/524_ppt_2030_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/529_ppt_2035_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/529_ppt_2035_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/534_ppt_2040_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/534_ppt_2040_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/539_ppt_2045_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/539_ppt_2045_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/544_ppt_2050_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/544_ppt_2050_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/549_ppt_2055_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/549_ppt_2055_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/554_ppt_2060_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/554_ppt_2060_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/559_ppt_2065_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/559_ppt_2065_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/564_ppt_2070_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/564_ppt_2070_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/569_ppt_2075_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/569_ppt_2075_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/574_ppt_2080_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/574_ppt_2080_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/579_ppt_2085_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/579_ppt_2085_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/584_ppt_2090_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/584_ppt_2090_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/589_ppt_2095_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/589_ppt_2095_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/594_ppt_2100_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt/594_ppt_2100_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/ppt_1980-2010_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/ppt_1980-2010_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/509_sdm_2015_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/509_sdm_2015_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/514_sdm_2020_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/514_sdm_2020_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/519_sdm_2025_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/519_sdm_2025_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/524_sdm_2030_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/524_sdm_2030_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/529_sdm_2035_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/529_sdm_2035_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/534_sdm_2040_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/534_sdm_2040_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/539_sdm_2045_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/539_sdm_2045_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/544_sdm_2050_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/544_sdm_2050_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/549_sdm_2055_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/549_sdm_2055_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/554_sdm_2060_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/554_sdm_2060_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/559_sdm_2065_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/559_sdm_2065_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/564_sdm_2070_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/564_sdm_2070_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/569_sdm_2075_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/569_sdm_2075_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/574_sdm_2080_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/574_sdm_2080_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/579_sdm_2085_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/579_sdm_2085_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/584_sdm_2090_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/584_sdm_2090_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/589_sdm_2095_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/589_sdm_2095_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/594_sdm_2100_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm/594_sdm_2100_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/sdm_1980-2010_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/sdm_1980-2010_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/509_tmp_2015_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/509_tmp_2015_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/514_tmp_2020_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/514_tmp_2020_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/519_tmp_2025_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/519_tmp_2025_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/524_tmp_2030_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/524_tmp_2030_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/529_tmp_2035_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/529_tmp_2035_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/534_tmp_2040_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/534_tmp_2040_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/539_tmp_2045_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/539_tmp_2045_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/544_tmp_2050_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/544_tmp_2050_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/549_tmp_2055_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/549_tmp_2055_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/554_tmp_2060_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/554_tmp_2060_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/559_tmp_2065_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/559_tmp_2065_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/564_tmp_2070_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/564_tmp_2070_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/569_tmp_2075_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/569_tmp_2075_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/574_tmp_2080_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/574_tmp_2080_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/579_tmp_2085_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/579_tmp_2085_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/584_tmp_2090_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/584_tmp_2090_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/589_tmp_2095_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/589_tmp_2095_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/594_tmp_2100_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp/594_tmp_2100_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/data/yosemite_demo/yosemite_lyrs/tmp_1980-2010_90x90.tif` & `geonomics-1.4.0/geonomics/data/yosemite_demo/yosemite_lyrs/tmp_1980-2010_90x90.tif`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/demos/_IBD_IBE.py` & `geonomics-1.4.0/geonomics/demos/_IBD_IBE.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/demos/_simult_select.py` & `geonomics-1.4.0/geonomics/demos/_simult_select.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/demos/_yosemite.py` & `geonomics-1.4.0/geonomics/demos/_yosemite.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/help/__init__.py` & `geonomics-1.4.0/geonomics/help/__init__.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/help/_memory_help.py` & `geonomics-1.4.0/geonomics/help/_memory_help.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/help/_param_help.py` & `geonomics-1.4.0/geonomics/help/_param_help.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/main.py` & `geonomics-1.4.0/geonomics/main.py`

 * *Files 6% similar despite different names*

```diff
@@ -38,15 +38,16 @@
     >>> geonomics.make_parameters_file?
 """
 #geonomics imports
 from geonomics.sim.model import Model
 from geonomics.sim.params import (_read_params_file, _make_params_file,
                                   ParametersDict)
 from geonomics.structs.landscape import _make_landscape
-from geonomics.structs.genome import _make_genomic_architecture
+from geonomics.structs.genome import (_make_genomic_architecture,
+                                      _sim_msprime_individuals)
 from geonomics.structs.individual import _make_individual
 from geonomics.structs.species import _make_species
 from geonomics.structs.community import _make_community
 from geonomics.structs import landscape, genome, individual, species, community
 from geonomics import demos
 
 #other imports
@@ -73,15 +74,15 @@
 
 #wrapper around params.make_parameters_file
 def make_parameters_file(filepath=None, layers=1, species=1, data=False,
         stats=False):
     """
     Create a new parameters file.
 
-    Write to disk a new, template parameters file. The file will contain the
+    Write to disk a new template parameters file. The file will contain the
     numbers and types of sections indicated by the parameters fed to this
     function. The new file can then be used 'out of the box'
     to make a new Model object, but typically it will be
     edited by the user to stipulate the scenario being simulated,
     then used to instantiate a Model.
 
     Parameters
@@ -92,107 +93,75 @@
         "GNX_params_<datetime>.py" will be written to the working
         directory.
 
     layers : {int, list of dicts}, optional, default: 1
         Number (and optionally, types) of Layer-parameter sections to include
         in the parameters file that is generated. Defaults to 1. Valid values
         and their associated behaviors are:
-
-        *If an int is passed*:
-            Add sections for the stipulated number of Layers, each with default
-            settings:
-                - parameters for creating Layers of type 'random' (i.e.
-                  Layers that will be generated by interpolation from
-                  randomly valued random points)
-                - no LandscapeChanger parameters
-
-        *If [dict, ..., dict] is passed*:
-            Each dict in this list should have the following key-value pairs:
-                KEY                    VALUE
-                'type':                {'random', 'defined', 'file', 'nlmpy'}
-
-                'change':               bool
-
-            This will add one section of Layer parameters, with the
-            contents indicated, for each dict in this list.
+        1.) *If an int is passed*, add sections for the stipulated number of
+        Layers, each with default settings, including parameters for creating
+        Layers of type 'random' (i.e. Layers that will be generated by
+        interpolation from randomly valued random points),
+        and no LandscapeChanger parameters.
+        2.) *If [dict, ..., dict] is passed*, each dict in the list
+        should have the following key-value pairs:
+        'type': {'random', 'defined', 'file', 'nlmpy'}, 'change': bool.
+        This will add one section of Layer parameters, with the
+        contents indicated, for each dict in this list.
 
     species : {int, list of dicts}, optional, default: 1
         Number (and optionally, types) of Species-parameter sections to
         include in the parameters file that is generated. Defaults to 1. Valid
         values and their associated behaviors are:
-
-        *If an int is passed*:
-            Add sections for the stipulated number of Species, each with
-            default settings:
-                - parameters for movement and dispersal without
-                  _ConductanceSurfaces
-                - parameters for a GenomicArchitecture with 0 Traits (i.e. with
-                  only neutral loci)
-                - no _SpeciesChanger parameters
-
-        *If [dict, ..., dict] is passed*:
-            Each dict in this list can contain any of the following
-            key-value pairs:
-                KEY                     VALUE
-
-                'movement':             bool,
-
-                'movement_surface':     bool,
-
-                'dispersal_surface':    bool,
-
-                'genomes':              {bool, 'custom'},
-
-                'n_traits':             int,
-
-                'demographic_change':   int,
-
-                'parameter_change':     bool
-
-            This will add one section of Species parameters, customized
-            as indicated, for each dict in the list. (Note that if the
-            'genomes' argument is True or 'custom', a section for
-            parameterization of the genomic architecture will be added,
-            but if it is 'custom' then a template custom genomic architecture
-            file will also be created (a CSV file), which can be filled in
-            to stipulate the locus-wise values for starting allele frequency,
-            recombination rate, dominance, associated traits, and effect
-            sizes.)
+        1.) *If an int is passed*, add sections for the stipulated number
+        of Species, each with default settings, including parameters for
+        movement and dispersal without _ConductanceSurfaces, parameters for
+        a GenomicArchitecture with 0 Traits (i.e. with only neutral loci),
+        and no _SpeciesChanger parameters.
+        2.) *If [dict, ..., dict] is passed*, each dict in the list
+        can contain any of the following key-value pairs:
+        'msprime_source_pops': int, 'movement': bool, 'movement_surface': bool,
+        'dispersal_surface': bool, 'genomes': {bool, 'custom'}, 'n_traits': int,
+        'demographic_change': int, 'parameter_change': bool.
+        This will add one section of Species parameters,
+        customized as indicated, for each dict in the list.
+        (Note that if the 'genomes' argument is True or 'custom',
+        a section for parameterization of the genomic architecture will be added,
+        but if it is 'custom' then a template custom genomic architecture
+        CSV file will also be created, which can be filled in to stipulate
+        the locus-wise values for starting allele frequency,
+        recombination rate, dominance, associated traits, and effect sizes.)
 
     data : bool, optional, default: False
         Whether to include a Data-parameter section in the parameters file that
         is generated. Defaults to None. Valid values and their associated
         behaviors are:
-
-        None, False:
-            Will not add a section for parameterizing data to be collected.
-            No _DataCollector will be created for the Model object made from
-            the resulting parameters file, and no data will be collected
-            during the model runs.
-        True:
-            Will add a section that can be used to parameterize which
-            data will be collected during the model runs, when, and what
-            file formats will be used to write it to disk.
-            (This which will be managed by the model's _DataCollector
-            object.)
+        1.) None, False: Will not add a section for parameterizing
+        data to be collected. No _DataCollector will be created
+        for the Model object made from the resulting parameters file,
+        and no data will be collected during the model runs.
+        2.) True: Will add a section that can be used to parameterize which
+        data will be collected during the model runs, when, and what
+        file formats will be used to write it to disk.
+        (This which will be managed by the model's _DataCollector
+        object.)
 
     stats : bool, optional, default: False
-        Whether to include a Stats-parameter section in the parameters file that
-        is generated. Defaults to None. Valid values and their associated
-        behaviors are:
-
-        None, False:
-            Will not add a section for parameterizing the statistics to be
-            calculated. No _StatsCollector will be created for the Model
-            object made from the resulting parameters file, and no
-            statistics will be calculated during the model runs.
-        True:
-            Will add a section that can be used to parameterize which
-            statistics will be calculated during the model runs, and when.
-            (This will be managed by the model's _StatsCollector object.)
+        Whether to include a Stats-parameter section in the
+        parameters file that is generated.  Defaults to None.
+        Valid values and their associated behaviors are:
+        1.) None, False: Will not add a section for parameterizing
+        the statistics to be calculated. No _StatsCollector
+        will be created for the Model object made from
+        the resulting parameters file, and no statistics
+        will be calculated during the model runs.
+        2.) True: Will add a section that can be used
+        to parameterize which statistics will be calculated
+        during the model runs, and when. (This will be
+        managed by the model's _StatsCollector object.)
 
     Returns
     -------
     None
         Returns no output. Resulting parameters file will be written to the
         location and filename indicated (or by default, will be written to a
         file named "GNX_params_<datetime>.py" in the working directory).
@@ -218,15 +187,14 @@
     current working directory, so that it can be unambiguously identified) we
     can call the gnx.make_model function to create a Model object from that
     file, and then call the Model.run method to run the model (setting the
     'verbose' parameter to True, so that we can observe model output).
 
     >>> gnx.make_parameters_file()
     >>> mod = gnx.make_model()
-
     NOTE: Using the following file, in the current working directory to create the Model object:
             GNX_params_21-01-2020_17:22:08.py
 
     >>> mod.run(verbose = True)
     ###############################################################################################
     Running model "GNX_params_21-01-2020_17:22:08"...
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@@ -253,24 +221,24 @@
     components for their sections of the parameters file) and with a section
     for parameterizing data collection.
 
     >>> gnx.make_parameters_file(layers = 3, data = True)
 
     As a more complex example that is likely to be similar to most use cases,
     we can create a parameters file for a model scenario with:
-        - 2 Layers (one being an nlmpy Layer that will not change over model
-          time, the other being a raster read in from a GIS file and being
-          subject to change over model time);
-        - 2 Species (the first having genomes, 2 Traits, and movement
-          that is dictated by a _ConductanceSurface; the second not having
-          genomes but having dispersal determined by a
-          _ConductanceSurface, and undergoing
-          demographic change)
-        - data-collection;
-        - stats-collection;
+    - 2 Layers (one being an nlmpy Layer that will not change over model
+    time, the other being a raster read in from a GIS file and being
+    subject to change over model time);
+    - 2 Species (the first having genomes, 2 Traits, and movement
+    that is dictated by a _ConductanceSurface; the second not having
+    genomes but having dispersal determined by a
+    _ConductanceSurface, and undergoing
+    demographic change)
+    - data-collection;
+    - stats-collection;
     We can save this to a file named "2-spp_2-trait_model.py" in our current
     working directory.
 
     >>> gnx.make_parameters_file(
     >>>     #list of 2 dicts, each containing the values for each Layer's
     >>>     #parameters section
     >>>     layers = [
@@ -295,18 +263,27 @@
     >>>     #destination to which to write the resulting parameter file
     >>>     filepath = '2-spp_2-trait_model.py')
 
     """
 
     #check if any keys in the layers or species dicts are abnormal, and
     #provide warning if so
-    valid_l_keys = ['type', 'change']
-    valid_p_keys = ['movement', 'movement_surface', 'dispersal_surface',
-        'genomes', 'n_traits', 'custom_genomic_architecture',
-        'demographic_change', 'parameter_change']
+    valid_l_keys = ['type',
+                    'change',
+                   ]
+    valid_p_keys = ['msprime_source_pops',
+                    'movement',
+                    'movement_surface',
+                    'dispersal_surface',
+                    'genomes',
+                    'n_traits',
+                    'custom_genomic_architecture',
+                    'demographic_change',
+                    'parameter_change',
+                   ]
     if isinstance(layers, list):
         for n, lyr_dict in enumerate(layers):
             if False in [k in valid_l_keys for k in lyr_dict.keys()]:
                 invalid_keys = ', '.join([k for k in lyr_dict.keys(
                     ) if k not in valid_l_keys])
                 err_msg = ("One or more of the keys in dict number %i of "
                     "the 'layers' argument is not valid. Invalid values: "
@@ -468,15 +445,15 @@
 
     Use either a ParametersDict object or the path to a valid Geonomics
     parameters file (whichever is provided to the 'parameters' argument) to
     create a new Model object.
 
     Parameters
     ----------
-    parameters : {ParametersDict, dict, str}, optional, default: None
+    parameters: {ParametersDict, dict, str}, optional, default: None
         The parameters to be used to make the Model object.
 
         If `parameters` is a ParametersDict object, the object will be used to
         make the Model.
 
         If `parameters` is a dict object, Geonomics will attempt to convert it
         to a ParametersDict object, then use that object to make the Model.
@@ -926,15 +903,15 @@
     :class:`geonomics.structs.species.Species`
         A Species instance.
     """
     spp = _make_species(landscape, spp_params)
     return(spp)
 
 
-#wrapper around community.make_comunity
+#wrapper around community.make_community
     #burn can be True (i.e. then the individuals will have a [[0,0]] genome)
 def make_community(landscape, params):
     """
     Make a Community object.
 
     Use the parameter values stored in the input params argument,
     and the Landscape object input as the landscape argument,
@@ -953,7 +930,109 @@
     Returns
     -------
     :class:`geonomics.structs.community.Community`
         A Community instance.
     """
     comm = _make_community(landscape, params)
     return comm
+
+
+# wrapper around genome._sim_msprime_individuals
+def sim_msprime_individuals(n,
+                            L,
+                            recomb_rate,
+                            mut_rate,
+                            demography=None,
+                            population_size=None,
+                            ancestry_model=None,
+                            random_seed=None,
+                            use_tskit=True,
+                           ):
+    '''
+    Use an msprime simulation to generate new Geonomics Individuals.
+
+    Parameters
+    ----------
+
+    n: int
+        The number of Geonomics Individuals to be simulated
+
+    L: int
+        The length of the simulated genome.
+
+    recomb_rate: {float, list, numpy.ndarray}
+        Either a scalar float, indicating the genome-wide recombination rate;
+        or a 1d numpy.ndarray or 1d-array-like of length L-1, indicating the
+        inter-locus recombination rates along the length-L simulated genome.
+        (Note: This parameter is expressed according to the convention
+        used by Geonomics, as explained above. This differs from the convention
+        used by msprime, the translation to which will be handled internally by
+        Geonomics.)
+
+    mut_rate: {float, list, numpy.ndarray}
+        Either a scalar float, indicating the genome-wide neutral mutation rate;
+        or a 1d numpy.ndarray or 1d-array-like of length L, indicating the
+        mutation rates of all loci in the the length-L simulated genome.
+        (Note: This parameter is expressed according to the convention
+        used by Geonomics, as explained above. This differs from the convention
+        used by msprime, the translation to which will be handled internally by
+        Geonomics.)
+
+    demography: msprime.Demography, optional, default: None
+        An `msprime.Demography` object to be used by `msprime.sim_ancestry`.
+        `msprime` allows specification of complex demographic histories,
+        including multiple extant and ancestral populations with fixed
+        or variable sizes, growth rates, or migration rates, as well as
+        timed demographic events. Defaults to None, in which case
+        `msprime.sim_ancestry` will simulate a single population
+        whose size will be determined by the `population_size` parameter
+        (or will default to 1, if `population_size` is also None).
+        Cannot be specified along with `population_size`.
+        (See the `msprime.Demography` documentation and references therein
+        for more details.)
+
+    population_size: int, optional, default: None
+        The population size of the default single population to be simulated
+        by `msprime.sim_ancestry`. That default single population will only be
+        simulated if a more complex demographic scenario is not specified using
+        the `demography` parameter (i.e., if `demography` is None).
+        Thus, `population_size` cannot be specified along with `demography`.
+        If both are None then `msprime` simulates a single population of size 1.
+        (See the `msprime.sim_ancestry` documentation for more details.)
+
+    ancestry_model: {`msprime.AncestryModel`, list}, optional, default: None
+        The `msprime.AncestryModel` instance (or list of them) to be used by
+        `msprime.sim_ancestry`. See `msprime.sim_ancestry` documentation and
+        references therein for more details.
+
+    random_seed: int, optional, default: None
+        The random seed value to be provided to `msprime.sim_ancestry` and
+        `msprime.sim_mutations`. Use if reproducibility is desirable.
+        See `msprime.sim_ancestry` and `msprime.sim_mutations` documentation
+        for detail.
+
+    use_tskit: bool, optional, default: True
+        Whether or not the Geonomics `Species` for which the `Individual`s are
+        intended uses tskit (i.e., has Species.gen_arch.use_tskit as True).
+        If not, `Individual`s will each store a copy of their Lx2 genotype
+        array in the 'g' attribute; otherwise, this attribute will be None.
+
+    Returns
+    -------
+    :class:`list`
+        A list of `geonomics.structs.individual.Individual` objects.
+        Individuals are assigned sequential index integers starting from 0,
+        placed at x,y coordinates (0, 0), and otherwise given default attribute
+        values. Attributes should be updated as necessary in downstream code.
+
+    '''
+    individs = _sim_msprime_individuals(n=n,
+                                        L=L,
+                                        recomb_rate=recomb_rate,
+                                        mut_rate=mut_rate,
+                                        population_size=population_size,
+                                        demography=demography,
+                                        ancestry_model=ancestry_model,
+                                        random_seed=random_seed,
+                                        use_tskit=use_tskit,
+                                       )
+    return individs
```

### Comparing `geonomics-1.3.9/geonomics/ops/change.py` & `geonomics-1.4.0/geonomics/ops/change.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/ops/demography.py` & `geonomics-1.4.0/geonomics/ops/demography.py`

 * *Files 7% similar despite different names*

```diff
@@ -177,16 +177,16 @@
     if len(deaths) > 0:
         ig = itemgetter(*deaths)
         [spp.pop(ind) for ind in deaths];
     return len(deaths)
 
 
 def _do_pop_dynamics(spp, land, with_selection = True, burn = False,
-    births_before_deaths = False, asserts = True, debug = False):
-    '''Generalized function for implementation population dynamics.
+                     asserts = True, debug = False):
+    '''Generalized function for implementing population dynamics.
     Will carry out one round of mating and death, according to parameterization
     laid out in params dict (which were grabbed as Species attributes).
 
        If with_selection == False, only density-dependent death will occur.
 
        If burn == True, selection will be coerced to False, the Species.K
        attribute will be updated, stationarity metrics will be assessed, and
@@ -215,20 +215,17 @@
         assert n_pairs.min() >= 0, 'n_pairs.min() == %0.2f' %(n_pairs.min())
         assert not np.any(np.isnan(n_pairs))
         assert not np.any(np.isinf(n_pairs))
     #add debug plot
     if debug:
         dp._next_plot('n_pairs', n_pairs)
 
-    #if births should happen before (and thus be included in the calculation of)
-    #deaths, then mate and disperse babies now
-    if births_before_deaths:
-        #Feed the land and mating pairs to spp.do_mating, to produce and
-        #disperse zygotes
-        spp._do_mating(land, pairs, burn)
+    #Feed the land and mating pairs to spp.do_mating, to produce and
+    #disperse zygotes
+    spp._do_mating(land, pairs, burn)
 
     #calc N raster, set it as spp.N, then get it  
     spp._calc_density(set_N = True)
     N = spp.N
     #run checks on N
     if asserts:
         assert N.min() >= 0
@@ -294,36 +291,16 @@
         assert d.max() <= 1, 'd.max() is %0.2f' % d.max()
         assert not np.any(np.isnan(d))
         assert not np.any(np.isinf(d))
     #add debug plot
     if debug:
         dp._next_plot('d', d)
 
-    #If births should happen after (and thus not be included in the 
-    #calculation of) deaths, then instead of having mated and dispersed
-    #babies up above, do it now (i.e. now that the d raster has been 
-    #calculated)
-    #NOTE: The way I'm looking at it now, calculating likely births from number
-    #of existing pairs, then calculating death raster, then having births take
-    #place, then having per-individual death probabilies determined and having
-    #deaths take place (such that newborns could also immediately die after
-    #birth, based on the death raster calculated before they were born, and
-    #even based on individs who weren't their parents, assuming they dispersed
-    #across cell lines), which is what I'm doing currently, is the only
-    #obvious/reasonable way to do this. Of course there will always be an
-    #inherent rub between reality and discrete-time simulations because of
-    #the continuous/discrete time conflict. But in this case, is there any
-    #strong reason to CHANGE THIS? Or are there any foreseeable and undesirable
-    #results/effects?... NEED TO PUT MORE THOUGHT INTO THIS LATER.
-    if not births_before_deaths:
-        #Feed the land and mating pairs to the mating functions, to produce
-        #and disperse zygotes
-        spp._do_mating(land, pairs, burn)
     #Get death probabilities
-    death_probs = d[spp.cells[:,1], spp.cells[:,0]]
+    death_probs = d[spp._cells[:,1], spp._cells[:,0]]
     #If with_selection (i.e. if death probs should account for fitness),
     #then use the d raster and individuals' fitnesses to calculate
     #per-individual probabilities of death
     if with_selection:
         death_probs = _calc_prob_death(spp, death_probs)
     #run checks on death_probs
     if asserts:
```

### Comparing `geonomics-1.3.9/geonomics/ops/mating.py` & `geonomics-1.4.0/geonomics/ops/mating.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/ops/movement.py` & `geonomics-1.4.0/geonomics/ops/movement.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/ops/mutation.py` & `geonomics-1.4.0/geonomics/ops/mutation.py`

 * *Files 2% similar despite different names*

```diff
@@ -30,16 +30,25 @@
     est = int(2.5 * est)
     return est
 
 # add a row to the tskit.TableCollection.mutations table for a new mutation
 def _do_add_row_muts_table(spp, individ, homol, locus):
     # update the tskit.TableCollection.mutations table
     node_id = spp[individ]._nodes_tab_ids[homol]
-    mut_id = spp._tc.mutations.add_row(site=locus, node=node_id,
-                                       derived_state='1')
+    mut_id = spp._tc.mutations.add_row(site=locus,
+                                       node=node_id,
+                                       derived_state='1',
+                                       # NOTE: DETH: 29-03-24: gnx will now add
+                                       #       the time to the mutations table
+                                       #       (not required and not
+                                       #       influential, so no harm in
+                                       #       adding and I see no reason it
+                                       #       should break older code)
+                                       time=-1*spp.t,
+                                      )
     return mut_id
 
 
 # do a neutral mutation for a single Individual, chosen from the offspring
 def _do_neutral_mutation(spp, offspring, locus=None, individ=None):
     #randomly choose an individual to mutate, if not provided
     if individ is None:
```

### Comparing `geonomics-1.3.9/geonomics/ops/selection.py` & `geonomics-1.4.0/geonomics/ops/selection.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/sim/burnin.py` & `geonomics-1.4.0/geonomics/sim/burnin.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/sim/data.py` & `geonomics-1.4.0/geonomics/sim/data.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/sim/model.py` & `geonomics-1.4.0/geonomics/sim/model.py`

 * *Files 2% similar despite different names*

```diff
@@ -20,14 +20,15 @@
 from geonomics.utils._str_repr_ import _get_str_spacing
 from geonomics.utils.viz import _check_display
 
 #other imports
 import numpy as np
 import numpy.random as r
 import random
+import warnings
 from copy import deepcopy
 import sys, os, traceback, psutil
 import matplotlib as mpl
 _check_display()
 import matplotlib.pyplot as plt
 from sklearn.decomposition import PCA
 
@@ -266,22 +267,25 @@
                 "its key in the Landscape dict). Instead, a %s was "
                 "provided.") % str(type(lyr_id)))
 
 
     #method for getting a species' number, given a name (str) or number
     def _get_spp_num(self, spp_id):
         if isinstance(spp_id, int):
+            assert spp_id in self.comm.keys(), (f"A Species with numeric "
+                                                "spp_id {spp_id} does "
+                                                 "not exist.")
             return(spp_id)
         elif isinstance(spp_id, str):
             #get nums for all spps with  matching names
             spp_nums = [k for k, spp in self.comm.items(
                                                     ) if spp.name == spp_id]
             assert len(spp_nums) == 1, ("Expected to find a single Species "
                 "with a name matching the name provided (%s). Instead "
-                "found %i.") % (spp_id, len(nums))
+                "found %i.") % (spp_id, len(spp_nums))
             spp_num = spp_nums[0]
             return spp_num
         else:
             raise ValueError(("The Species identifier must be either a str "
                 "(indicating the Species' name) or an int (indicating "
                 "its key in the Community dict). Instead, a %s was "
                 "provided.") % str(type(spp_id)))
@@ -477,14 +481,23 @@
                         if not self.repeat_burn:
                             #verbose output
                             if self._verbose:
                                 print(('\nAssigning genomes for '
                                        'species "%s"...\n\n') % spp.name,
                                       flush=True)
                             spp._set_genomes_and_tables(self.burn_T, self.T)
+                            # replace gnx-simulated Individuals with
+                            # msprime-simulated Individuals, if required
+                            if ('msprime' in self.params['comm']['species'][
+                                                            spp.name]['init']):
+                                msp = self.params['comm']['species'][
+                                                spp.name]['init']['msprime']
+                                spp._init_msprime_pop(land=self.land,
+                                                      msprime_init_params=msp,
+                                                     )
 
         else:
             #verbose ouput
             if self._verbose:
                 print(('Creating new community for '
                                     'iteration %i...\n\n') % self.it,
                       flush=True)
@@ -591,28 +604,30 @@
             queue.append(self._set_comm_t)
             for spp in self.comm.values():
                 queue.append(lambda: self._set_spp_t(spp.idx))
 
         #append the set_age_stage methods to the queue
         for spp in self.comm.values():
             queue.append(lambda: self._set_age_stage(spp.idx))
-        #append the set_Nt methods
-        for spp in self.comm.values():
-            queue.append(lambda: self._set_Nt(spp.idx))
         #append the do_movement_methods, if spp._move
         for spp in self.comm.values():
             if spp._move:
                 queue.append(lambda: self._do_movement(spp.idx))
         #append the do_pop_dynamics methods
         #FIXME: Consider whether the order of these needs to be specified, or
         #randomized, should people want to eventually simulate
         #multiple, interacting species
         for spp in self.comm.values():
             queue.append(lambda: self._do_pop_dynamics(spp.idx))
 
+        #append the set_Nt methods
+        # (now that this time step's population sizes are set)
+        for spp in self.comm.values():
+            queue.append(lambda: self._set_Nt(spp.idx))
+
         #add the Changer.make_change, data._DataCollector._write_data, and 
         #stats._StatsCollector._write_stats methods, if this is not the burn-in
         #and if spp and/or land have Changer objects, or if the model has 
         #_DataCollector or _StatsCollector objects (in the self._data_collector 
         #and self._stats_collector attributes)
         if not burn:
             #add land._make_change method
@@ -676,14 +691,23 @@
                         if spp.gen_arch is not None:
                             #verbose output
                             if self._verbose:
                                 print(('\nAssigning genomes for '
                                        'species "%s"...\n\n') % spp.name,
                                       flush=True)
                             spp._set_genomes_and_tables(self.burn_T, self.T)
+                            # replace gnx-simulated Individuals with
+                            # msprime-simulated Individuals, if required
+                            if ('msprime' in self.params['comm']['species'][
+                                                            spp.name]['init']):
+                                msp = self.params['comm']['species'][
+                                                spp.name]['init']['msprime']
+                                spp._init_msprime_pop(land=self.land,
+                                                      msprime_init_params=msp,
+                                                     )
                     #and then set the reassign_genomes attribute to False, so
                     #that they won't get reassigned again during this iteration
                     self.reassign_genomes = False
                 #mark the community as burned in
                 self.comm.burned = True
                 #verbose output
                 if self._verbose:
@@ -714,15 +738,15 @@
                         if self._verbose:
                             print(('\n\nnow sorting and simplifying '
                                    "tskit tables for Species: '%s'") % spp.name)
                             print("\tNUMBER EDGES BEFORE SIMPLIFICATION:",
                                   self.comm[0]._tc.edges.num_rows)
                             print("\tNUMBER INDIVIDS BEFORE SIMPLIFICATION:",
                                   self.comm[0]._tc.individuals.num_rows)
-                        spp._sort_simplify_table_collection()
+                        spp._sort_and_simplify_table_collection()
                         if self._verbose:
                             print("\n\tNUMBER EDGES AFTER SIMPLIFICATION: ",
                                   self.comm[0]._tc.edges.num_rows)
                             print("\tNUMBER INDIVIDS AFTER SIMPLIFICATION: ",
                                   self.comm[0]._tc.individuals.num_rows, '\n',
                                   flush=True)
 
@@ -1111,14 +1135,34 @@
             # end the iteration early if any species is extinct
             if extinct:
                 break
         # reset self._verbose to False
         self._verbose = old_verbose
 
 
+    # helper method to burn a model in if it hasn't been burned in already
+    # to succinctly differentiate it from 'run'
+    def burn(self):
+        """
+        Helper function to burn a model in, if it hasn't been burned in already.
+
+        Returns
+        -------
+        out : None
+            Returns no output. Alters Model in place by running burn-in phase.
+
+        """
+        if self.comm.burned:
+            print("\nModel has already been burned in.\n")
+        else:
+            while not self.comm.burned:
+                self.walk(10, 'burn')
+            print("\nModel has now been burned in.\n")
+
+
     # method to use the self._data_collector object to sample and write data
     def _write_data(self):
         self._data_collector._write_data(self.comm, self.land, self.it)
 
 
     #method to use the self._stats_collector object to sample and write data
     def calc_stats(self):
@@ -1842,15 +1886,15 @@
                           fitness_cbar=fitness_cbar, alpha=alpha,
                           zoom_width=zoom_width, x=x, y=y, ticks=ticks,
                           mask_rast=mask_rast)
         #add spp name
         #plt.suptitle(spp.name)
 
 
-    def plot_genetic_PCA(self, spp=0, lyr=0,
+    def plot_genetic_PCA(self, n_pcs=3, spp=0, lyr=0,
                         individs=None, text=False,
                         edge_color='black', text_color='black', cbar=True,
                         size=25, text_size=9,
                         alpha=False, zoom_width=None, x=None, y=None,
                         ticks=False, hide_land=False, mask_rast=None):
         """
         Create a scatter plot of the Individuals in a Species,
@@ -1862,14 +1906,23 @@
         spp : {int, str}, optional, default: 0
             A reference to the Species whose Individuals should be scattered
             on the plot. Can be either the Species' index number (i.e. its
             integer key in the Community dict), or its name (as a character
             string). If None, will cause only the Landscape
             to be plotted.
 
+        n_pcs : int, optional, default: 3
+            The number of top PCs to use for the plot. If 1 is used, all three
+            red, green, and blue (RGB) channels will be scaled the same,
+            creating a grayscale mapping of the top PC. If 2 are used, the red
+            and blue channels only will be used, creating a visualization of
+            the top 2 PCs that is colorblind-friendly. If 3 are used, the top 3
+            PCs will be fed into all RGB channels, creating a visualization
+            with the most information but which is not colorblind-friendly.
+
         lyr : {int, str}, optional, default: None
             A reference to the Layer whose raster should be plotted. Can be
             either the Layer's index number (i.e. its integer key in the
             Landscape dict), or its name (as a character string). Defaults
             to None, which will either cause only the Layer associated with the
             chosen Trait to be plotted, if *trt* is not None; or else
             will cause all Layers to be plotted as an overlay of transparent
@@ -1956,25 +2009,37 @@
         # get array of resulting genomic data (i.e. 'speciome'),
         # genotypes meaned by individual
         if spp.gen_arch.use_tskit:
             speciome = spp._get_genotypes(biallelic=False)
         else:
             speciome = np.mean(np.stack([i.g for i in spp.values()]), axis=2)
         # run PCA on speciome
-        pca = PCA(n_components=3)
+        pca = PCA(n_components=n_pcs)
         PCs = pca.fit_transform(speciome)
         # normalize the PC results
         norm_PCs = (PCs - np.min(PCs,
                                  axis=0)) / (np.max(PCs,
                                                     axis=0) - np.min(PCs,
                                                                      axis=0))
-        # use first 3 PCs to get normalized values for R, G, & B colors
-        PC_colors = norm_PCs * 255
-        # scatter all individuals on top of landscape, colored by the
-        # RBG colors developed from the first 3 geonmic PCs
+        # use top PCs to get normalized values for R, G, & B colors
+        # grayscale if n_pcs==1
+        # red-blue if n_pcs==2
+        # RGB if n_pcs==3
+        if n_pcs == 1:
+            PC_colors = np.hstack([norm_PCs * 255 for _ in range(3)])
+        if n_pcs == 2:
+            PC_colors = np.vstack([norm_PCs[:, 0] * 255,
+                                   np.zeros(norm_PCs.shape[0]),
+                                   norm_PCs[:, 1] * 255,
+                                  ]).T
+        if n_pcs == 3:
+            PC_colors = norm_PCs * 255
+
+        # scatter all individuals on top of landscape, colored by top PCs
+        # (number determined by n_pcs)
         xs = self.comm[0]._get_x()
         ys = self.comm[0]._get_y()
 
         # subset xs, ys, and PC_colors, if individs requested
         if individs is not None:
             indices = [i in individs for i in spp]
             PC_colors = PC_colors[indices, :]
@@ -1987,39 +2052,42 @@
                   edge_color=edge_color, text_color=text_color, cbar=cbar,
                   size=size, text_size=text_size, alpha=alpha,
                   zoom_width=zoom_width, x=x, y=y, ticks=ticks,
                   hide_land=hide_land, mask_rast=mask_rast)
 
 
     #wrapper around Species._plot_allele_frequencies
-    def plot_allele_frequencies(self, spp=0):
+    def plot_allele_frequencies(self, spp=0, color='red'):
         """
         Plot a Species' current allele frequencies for all alleles
 
         For the Species indicated, create a plot showing, across all loci,
-        the current allele frequencies (in solid blue) versus the starting
-        frequencies (in dashed red).
+        the current allele frequencies (in the indicated color) versus the starting
+        frequencies (in dashed black).
 
         Parameters
         ----------
         spp : {int, str}, optional, default: 0
             A reference to the Species whose allele frequencies will be
             plotted. Can be either the Species' index number (i.e. its
             integer key in the Community dict), or its name (as a character
             string).
 
+        color: str, optional, default: 'red'
+            The color to use for plotting the current allele frequencies.
+
         Returns
         -------
         None
             Returns no output.
         """
         #get the spp
         spp = self.comm[self._get_spp_num(spp)]
         #call the fn
-        spp._plot_allele_frequencies()
+        spp._plot_allele_frequencies(color=color)
 
 
     #wrapper around Species._plot_hist_fitness
     def plot_hist_fitness(self, spp=0):
         """
         Plot a Species
 
@@ -2513,15 +2581,15 @@
         -------
         np.array
             Returns an nx2 numpy array, where n is the number of Individuals
             whose coordinates were requested, and the 2 columns correspond to
             the x then y (i.e. lon then lat) coordinates.
 
         Note: Rows are sorted by index number of the Individuals,
-            regardless of input index order.
+              regardless of input index order.
         """
         spp = self.comm[self._get_spp_num(spp)]
         if individs is not None:
             individs = np.sort(individs)
         coords = spp._get_coords(individs=individs)
         return coords
 
@@ -2871,19 +2939,178 @@
             raise Exception(("TreeSequence object cannot be generated "
                              "for a Species whose pedigree is not being "
                              "tracked by tskit."))
         assert spp.gen_arch is not None, ("Species without Genomes "
                                           "have no tskit data structures "
                                           "to be returned.")
         # prep the TableCollection, then get its TreeSequence
-        spp._sort_simplify_table_collection()
+        spp._sort_and_simplify_table_collection()
         ts = spp._tc.tree_sequence()
         return ts
 
 
+    ###############
+    #altering data#
+    ###############
+
+    def remove_individuals(self,
+                           spp=0,
+                           n=None,
+                           n_left=None,
+                           individs=None,
+                          ):
+        """
+        Remove Individuals from a Species.
+
+        Parameters
+        ----------
+        spp : {int, str}, optional, default: 0
+            A reference to the Species for which values should be returned.
+            Can be either the Species' index number (i.e. its
+            integer key in the Community dict), or its name (as a character
+            string).
+
+        n: {int}, optional, default: None
+            The number of random Individuals to be removed.
+            If None then `n_out` or `individs` must be provided.
+
+        n_left: {int}, optional, default: None
+            The number of Individuals left in the Species
+            after a random selection of Individuals have been removed.
+            If None then `n` or `individs` must be provided.
+
+        individs: {list, tuple, numpy.ndarray}, optional, default: None
+            An iterable containing the integer IDs (i.e., the Species dict's keys)
+            of the Individuals to be removed. If None then `n` or `n_out` must be provided.
+
+        Returns
+        -------
+        None
+            Alters the specified Species in place by removing all specified
+            (if `individs` is not None) or randomly drawn (if `n` is not None)
+            Individuals.
+
+        """
+        # get the desired species
+        spp = self.comm[self._get_spp_num(spp)]
+
+        # call that species' corresponding method
+        spp._remove_individuals(individs=individs,
+                                n=n,
+                                n_left=n_left,
+                               )
+
+
+    def add_individuals(self,
+                        n,
+                        coords,
+                        recip_spp=0,
+                        source_spp=None,
+                        source_msprime_params=None,
+                        individs=None,
+                       ):
+        """
+        Add individuals to a recipient Species object, either using a second
+        Species object as the source population or feeding a dict of parameters
+        into msprime to simulate the source population.
+
+        Parameters
+        ----------
+        n: int
+            The number of Individuals to be added.
+
+        coords: {tuple, list, numpy.ndarray}
+            A tuple, list of numpy.ndarray indicating the x,y coordinates where the
+            new Individuals should be added to the recipient Species.
+            If `coords` contains 2 values then all added Individuals will be placed
+            at the x,y coordinate pair indicated by those 2 values.
+            Otherwise, `coords` may be of size `n` x 2, indicating the x,y
+            coordinate pair at which each of the `n` Individuals will be placed.
+
+        recip_spp: {gnx.structs.species.Species, int, str}, optional, default: None
+            An int or str referring to the Species in this Model to which
+            Individuals should be added.
+
+        source_spp: {gnx.structs.species.Species, int, str}, optional, default: None
+            A Geonomics.structs.species.Species object to act as the source
+            population from which individuals will be taken
+            (or an int or str referring to that Species within this Model).
+
+        source_msprime_params: dict, optional, default: None
+            A dict of keyword args to be fed to `gnx.sim_msprime_individs()`,
+            to parameterize the msprime source population model from which
+            new Individuals will be drawn. These include:
+                recomb_rate         (required)
+                mut_rate            (required)
+                demography          (optional; default: None)
+                population_size     (optional; default: None)
+                ancestry_model      (optional; default: None)
+                random_seed         (optional; default: None)
+            See `gnx.sim_msprime_individs` for details.
+
+        individs: {tuple, list, numpy.ndarray}, optional, default: None
+            A 1d numpy.ndarray or equivalent array-like containing the int indices
+            of the Individuals to be taken from `source_spp` and added to
+            `recip_spp`. If None
+
+        Returns
+        -------
+        None
+            Alters `recip_spp` in place by adding the provided
+            (if `source_spp` is not None) or simulated (if `source_msprime_params`
+            is not None) Individuals.
+
+        """
+        warnings.warn(("This function is new and has been checked but not "
+                       "extensively tested. User beware. Please immediately report "
+                       "any suspected bugs at "
+                       "https://github.com/erthward/geonomics/issues "
+                       "or by email to drew.terasaki.hart@gmail.com"))
+
+        # get the recipient species
+        if isinstance(recip_spp, int) or isinstance(recip_spp, str):
+            recip_spp = self.comm[self._get_spp_num(recip_spp)]
+
+        # and make sure it has already been burned in
+        assert recip_spp.burned, ("Individuals cannot be manually added to a "
+                                  "Species that hasn't been burned in yet.")
+
+        # make sure either a Species object is given or a set of valid msprime args
+        assert ((source_spp is not None and source_msprime_params is None) or
+                (source_spp is None and source_msprime_params is not None)), (
+                        "Source population must be provided either as "
+                        "a gnx Species object ('source_spp') "
+                        "or as a set of valid args for gnx.run_msprime_sim() "
+                        "('source_msprime_params'), but not both.")
+        if source_spp is not None:
+            assert (isinstance(source_spp, gnx.structs.species.Species) or
+                    isinstance(source_spp, int) or
+                    isinstnace(source_spp, str)), ("The "
+                        "value given to 'source_spp' identifies a gnx "
+                        "Species object that will serve as the source "
+                        "population, so must be either a gnx Species "
+                        "instance (i.e., an object of the class "
+                        "gnx.structs.species.Species) whose Individuals will "
+                        "be added to the recipient Species, or a valid int or str "
+                        "identifier of another Species in this model whose"
+                        "Individuals will be added to the recipeint Species.")
+            # get the source_spp from this Model, if necessary
+            if isinstance(source_spp, int) or isinstance(source_spp, str):
+                source_spp = self.comm[self._get_spp_num(source_spp)]
+
+        # call the Species method
+        recip_spp._add_individuals(n=n,
+                                   coords=coords,
+                                   land=self.land,
+                                   source_spp=source_spp,
+                                   source_msprime_params=source_msprime_params,
+                                   individs=individs,
+                                  )
+
+
     ##############
     #writing data#
     ##############
 
     def write_gendata(self, filepath, spp=0, n=None, include_fixed_sites=True):
         """
         A convenience function, designed for interactive use, that writes
@@ -3031,7 +3258,8 @@
                                 v) in spp._tc.asdict().items() if k in needed}
         # write out each table
         for k, table in tables.items():
             filepath = file_basename + '_' + k.upper() + '.csv'
             table.to_csv(filepath, index=False, sep=sep)
 
 
+
```

### Comparing `geonomics-1.3.9/geonomics/sim/params.py` & `geonomics-1.4.0/geonomics/sim/params.py`

 * *Files 2% similar despite different names*

```diff
@@ -256,14 +256,15 @@
                         'n_steps':          5,
                         }, # <END> event %i'''
 
 #block of species params
 #STRING SLOTS:
     #%s = spp_name,
     #%i = spp_num,
+    #%s = msprime_init_params,
     #%i = spp_num,
     #%i = spp_num,
     #%i = spp_num,
     #%s = move_surf_params,
     #%s = disp_surf_params,
     #%s = genome_params,
     #%s = change_params,
@@ -278,15 +279,15 @@
 
                 'init': {
                     #starting number of individs
                     'N':                250,
                     #carrying-capacity Layer name
                     'K_layer':          'lyr_0',
                     #multiplicative factor for carrying-capacity layer
-                    'K_factor':         1,
+                    'K_factor':         1,%s
                     }, # <END> 'init'
 
             #-------------------------------------#
             #--- spp num. %i: mating parameters ---#
             #-------------------------------------#
 
                 'mating'    : {
@@ -353,14 +354,42 @@
                     },    # <END> 'movement'
 
 %s
 %s
                 }, # <END> spp num. %i
 '''
 
+# block for params to initiate model with
+# individuals drawn from msprime simulations
+MSPRIME_INIT_PARAMS = '''
+                    # params for 1+ msprime pops for sampling starting individs
+                    'msprime':        {%s
+                        },
+'''
+
+MSPRIME_POP_PARAMS = '''
+                        # index of msprime source pop
+                        %i:  {
+                            # number of individs to sample from pop
+                            100: {
+                                 # 1x2 coord pair or nx2 coord pairs of individs
+                                 'coords':              [0, 0],
+                                 # valid kwargs dict for Model.add_individuals
+                                 # 'source_msprime_params' argument...
+                                 'recomb_rate':         0.5,
+                                 'mut_rate':            0.0001,
+                                 'demography':          None,
+                                 'population_size':     None,
+                                 'ancestry_model':      None,
+                                 'random_seed':         None,
+                                 }
+                            },
+
+'''
+
 #block for movement-surface params
 MOVE_SURF_PARAMS = '''
                     'move_surf'     : {
                         #move-surf Layer name
                         'layer':                'lyr_0',
                         #whether to use mixture distrs
                         'mixture':              True,
@@ -812,14 +841,23 @@
     #join the whole list into one str
     lyrs_params_str = '\n'.join(lyrs_params_list)
     return lyrs_params_str
 
 
 #function to create the spps-params section of a params file
 def _make_species_params_str(species=1):
+    assert (isinstance(species, int) or
+            (isinstance(species, list) and
+             np.all([isinstance(item, dict) for item in species]))), (
+                 "The 'species' param must be either an int (the number "
+                 "of species parameters sections to include in the file) "
+                 "or a list of dicts (one dict per species parameters "
+                 "section to be included in the file, with keys and "
+                 "values stipulating the aspects of the species to be "
+                 "parameterized).")
     #create an empty list, to be filled with one params string per spp
     spps_params_list = []
     #if species is an integer, create a string of identical parameter sections
     if type(species) is int:
         #assert that it's an integer greater than 0
         assert species > 0, ("The number of Species to be created "
         "must be a positive integer.")
@@ -827,45 +865,57 @@
         for i in range(species):
             #use genome params, but with no traits (i.e. string-format with a
             #zero-length str)
             genome_params = GENOME_PARAMS % (i, 'None', '')
             #add no change params
             change_params = ''
             #create the spp_params str, with no move_surf or disp_surf params
-            spp_params_str = SPP_PARAMS % ("'spp_%i'" % i, i, i, i, i,
-                '', '', genome_params, change_params, i)
+            spp_params_str = SPP_PARAMS % ("'spp_%i'" % i,
+                                           i,
+                                           '',
+                                           i,
+                                           i,
+                                           i,
+                                           '',
+                                           '',
+                                           genome_params,
+                                           change_params,
+                                           i,
+                                          )
             #append to the spps_params_list
             spps_params_list.append(spp_params_str)
 
     #or if species is a list of dicts, then create individually customized
     #params sections for each Species
     elif type(species) is list:
         #create an empty list, to which any species who should have custom
         #gen-arch files created will be appended
         #assert that each item in the list is a dict
         assert False not in [type(item) is dict for item in species], ("If"
             " the 'species' argument is a list then it must contain only "
             "objects of type dict.")
         #create a lookup dict for the params strings for different spp params
         #sections
-        params_str_dict = {'move_surf': {True: MOVE_SURF_PARAMS},
+        params_str_dict = {
+                        'msprime': {True: MSPRIME_INIT_PARAMS},
+                        'move_surf': {True: MOVE_SURF_PARAMS},
                         'disp_surf': {True: DISP_SURF_PARAMS},
                         'genome': {True: GENOME_PARAMS,
                                    'custom': GENOME_PARAMS},
                         'dem_change': {True: SPP_DEM_CHANGE_EVENT_PARAMS},
                         'dem_events': {True: SPP_DEM_CHANGE_EVENTS_PARAMS},
                         'param_change': {True: SPP_PARAM_CHANGE_PARAMS},
 
                             }
         [v.update({False: ''}) for v in params_str_dict.values()]
         #for each spp
         for i, spp_dict in enumerate(species):
             #assert that the argument values are valid
             bool_args= ['movement_surface', 'genomes', 'parameter_change']
-            int_args = ['n_traits', 'demographic_change']
+            int_args = ['n_traits', 'demographic_change', 'msprime_source_pops']
             for arg in bool_args:
                 if arg in [*spp_dict]:
                     assert (type(spp_dict[arg]) is bool
                         or arg == 'genomes' and spp_dict[arg] == 'custom'), (
                         "The '%s' key in "
                         "each Species' dictionary must contain a "
                         "boolean value. But dict number %i in the "
@@ -876,18 +926,30 @@
                     assert type(spp_dict[arg]) is int, ("The '%s' "
                         "key in each Species' dictionary must contain an "
                         "integer value. But dict number %i in the "
                         "'species' argument contains a non-integer "
                         "value:\n\n\t" "'%s': %s ") % (arg, i, arg,
                                                             str(spp_dict[arg]))
                     int_arg_str_fmt_dict = {'n_traits':'Traits',
-                            'demographic_change': 'demographic change events'}
+                            'demographic_change': 'demographic change events',
+                            'msprime_source_pops': 'msprime source populations',
+                                           }
                     assert spp_dict[arg] >= 0, ("The number of %s to "
                                     "be created must be 0 or a positive "
                                     "integer.") % (int_arg_str_fmt_dict[arg])
+            # get the msprime source pops params, if required
+            if 'msprime_source_pops' in [*spp_dict]:
+                print('making msprime params')
+                n_pops = spp_dict['msprime_source_pops']
+                msprime_params = params_str_dict['msprime'][n_pops>0]
+                if n_pops > 0:
+                    msprime_params = msprime_params % ('\n'.join([
+                        MSPRIME_POP_PARAMS % i for i in range(n_pops)]))
+            else:
+                msprime_params = ''
             #get the movement surf and dispersal surf params, if required
             if 'movement_surface' in [*spp_dict]:
                 ms_arg = spp_dict['movement_surface']
                 move_surf_params = params_str_dict['move_surf'][ms_arg]
             else:
                 move_surf_params = ''
             if 'dispersal_surface' in [*spp_dict]:
@@ -960,17 +1022,26 @@
                     events_params_str=events_params_str+param_change_params_str
                 change_params = SPP_CHANGE_PARAMS
                 change_params = change_params % (i, events_params_str)
             #or get empty str
             else:
                 change_params = ''
             #get the overall spp params str for this spp
-            spp_params_str = SPP_PARAMS % ("'spp_%i'" % i, i, i, i, i,
-                move_surf_params, disp_surf_params, genome_params,
-                change_params, i)
+            spp_params_str = SPP_PARAMS % ("'spp_%i'" % i,
+                                           i,
+                                           msprime_params,
+                                           i,
+                                           i,
+                                           i,
+                                           move_surf_params,
+                                           disp_surf_params,
+                                           genome_params,
+                                           change_params,
+                                           i,
+                                          )
             #append to the spps_params_list
             spps_params_list.append(spp_params_str)
     #join the whole list into one str
     spps_params_str = ''.join(spps_params_list)
     return spps_params_str
```

### Comparing `geonomics-1.3.9/geonomics/sim/stats.py` & `geonomics-1.4.0/geonomics/sim/stats.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/structs/community.py` & `geonomics-1.4.0/geonomics/structs/community.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/structs/genome.py` & `geonomics-1.4.0/geonomics/structs/genome.py`

 * *Files 15% similar despite different names*

```diff
@@ -7,30 +7,33 @@
 '''
 Classes, associated methods, and supporting functions for all genomic
 components
 '''
 
 # geonomics imports
 from geonomics.ops import mutation
+from geonomics.structs.individual import Individual
 from geonomics.utils.viz import _check_display
 
 # other imports
 import numpy as np
 import pandas as pd
 from numpy import random as r
 import matplotlib as mpl
 _check_display()
 import matplotlib.pyplot as plt
 from collections import OrderedDict as OD
 import warnings
 import random
 import bisect
 import bitarray
+import msprime
 import tskit
 
+
 ######################################
 # -----------------------------------#
 # CLASSES ---------------------------#
 # -----------------------------------#
 ######################################
 
 
@@ -192,15 +195,15 @@
         breakpoints = [self._positions[np.where(
                                         recomb)] for recomb in recombinations]
         for bp in breakpoints:
             if len(bp) > 0:
                 assert 0 not in bp, '%s' % str(bp)
         # recast it as an int-keyed dict, so that when I use the recombination
         # events to simulate recombination I don't have to pass around the long
-        # arrays, but instead can just random keys to index them out on the fly
+        # arrays, but instead just random keys to index them out on the fly
         breakpoints = dict(zip(range(len(breakpoints)), breakpoints))
 
         if use_subsetters:
             # get the recombination paths used to make the bitarray subsetters
             if ((use_tskit and len(nonneutral_loci) > 0)
                 or not use_tskit):
                 recomb_paths = [np.cumsum(
@@ -340,15 +343,15 @@
             Maximum absolute value that can be drawn for a locus
             effect size (i.e. alpha). Effect sizes are clipped
             to the closed interval [-max_alpha_mag, max_alpha_mag]
             when the model is parameterized.
 
         mu:
             Mutation rate for this trait (expressed in mutations per locus per
-            timestep)
+            time step)
 
         name:
             The string name of the trait
 
         n_loci:
             Number of loci subtending this trait
 
@@ -385,15 +388,15 @@
         self.loci_idxs = np.int64([])
         self.alpha = np.array([])
 
     def _get_phi(self, spp):
         if type(self.phi) in (float, int):
             phi = np.array([self.phi]*len(spp))
         else:
-            phi = self.phi[spp.cells[:, 1], spp.cells[:, 0]]
+            phi = self.phi[spp._cells[:, 1], spp._cells[:, 0]]
         return(phi)
 
     def _set_loci(self, loci):
         # set the loci
         self.loci = np.hstack((self.loci, np.array([*loci])))
         self.loci.sort()
         # set the number of loci
@@ -486,21 +489,21 @@
             and 1 indicates a locus for which the '1' allele is dominant)
 
         L:
             Length of the simulated genome
 
         mu_delet:
             Genome-wide deleterious mutation rate (expressed in mutations
-            per locus per timestep). Deleterious mutations do not influence
+            per locus per time step). Deleterious mutations do not influence
             trait phenotypes, but instead are univerally
             deleterious (to an extent determined by their effect sizes).
 
         mu_neut:
             Genome-wide neutral mutation rate (expressed in mutations
-            per locus per timestep).
+            per locus per time step).
 
         neut_loci:
             A 1d numpy array that tracks all loci that do not influence the
             phenotypes of any traits.
 
         nonneut_loci:
             A 1d numpy array that tracks all loci that either influence the
@@ -782,19 +785,23 @@
             self.delet_loci_s = np.hstack((self.delet_loci_s[:del_idx],
                                          delet_s,
                                          self.delet_loci_s[del_idx:]))
         return idx
 
 
     # method for plotting all allele frequencies for the species
-    def _plot_allele_frequencies(self, spp):
-        speciome = np.stack([ind.g for ind in spp.values()])
-        freqs = speciome.sum(axis=2).sum(axis=0) / (2*speciome.shape[0])
-        plt.plot(range(self.L), self.p, ':r', label='start freq.')
-        plt.plot(range(self.L), freqs, '-b', label='curr. freq.')
+    def _plot_allele_frequencies(self, spp, color='red'):
+        if spp.gen_arch.use_tskit:
+            speciome = spp._get_genotypes(biallelic=False)
+        else:
+            speciome = np.mean(np.stack([ind.g for ind in spp.values()]), axis=2)
+        freqs = np.mean(speciome, axis=0)
+        assert len(freqs) == spp.gen_arch.L
+        plt.plot(range(self.L), self.p, ':k', label='start freq.', alpha=0.5)
+        plt.plot(range(self.L), freqs, '-', color=color, label='curr. freq.')
         plt.xlabel('locus')
         plt.ylabel('frequency')
         plt.legend()
         plt.show()
 
     # method for pickling a genomic architecture
     def _write_pickle(self, filename):
@@ -1090,18 +1097,20 @@
         mutables = [*set(range(gen_arch.L)).difference(
                                             set(gen_arch.nonneut_loci))]
         r.shuffle(mutables)
         gen_arch._mutables = [*mutables]
     return
 
 
-# function to generate mutations, after burn-in,
-# and to assign them to a species' TableCollection's  current nodes,
-# to produce the starting 1-allele frequencies parameterized for the species
 def _make_starting_mutations(spp, tables):
+    '''
+    function to generate mutations, after burn-in,
+    and to assign them to a species' TableCollection's current nodes,
+    to produce the starting 1-allele frequencies parameterized for the species
+    '''
     # get the starting frequencies for each site
     start_freqs = spp.gen_arch.p
 
     # get a set of all homologues, as tuples of (individual id, homologue idx)
     homologues = [*zip(np.repeat([*spp], 2),
                          [*range(spp.gen_arch.x)] * len(spp))]
 
@@ -1148,17 +1157,483 @@
 def read_pickled_genomic_architecture(filename):
     import cPickle
     with open(filename, 'rb') as f:
         gen_arch = cPickle.load(f)
     return gen_arch
 
 
-##########################################################################
-# FUNCTIONS FOR WORKING WITH THE SPATIAL PEDGREE SAVED IN tskit STRUCTURES
-##########################################################################
+################################################################
+# FUNCTIONS FOR WORKING WITH msprime TO SIMULATE GNX INDIVIDUALS
+################################################################
+
+def _format_msprime_rate(rate, L):
+    """
+    Either return a simple float, if the given rate
+    is homogeneous, or else return an msprime.RateMap object.
+    Used for msprime recombination and mutation rates.
+    """
+    assert (type(rate) in [float, list, np.ndarray] or rate == 0 or rate == 1)
+    if (isinstance(rate, float) or rate == 0 or rate == 1):
+        return rate
+    elif len(np.unique(rate)) == 1:
+        return rate[0]
+    else:
+        assert len(rate) == L
+        # NOTE: position array has to be one longer than the sequence length,
+        #       so that rightmost position brackets a stretch of genomic
+        #       intervals equal to the sequence_length
+        positions = np.linspace(0, L, L+1)
+        ratemap = msprime.RateMap(position=positions, rate=rate)
+        return ratemap
+
+
+def _sim_msprime_individuals(n,
+                             L,
+                             recomb_rate,
+                             mut_rate,
+                             demography=None,
+                             population_size=None,
+                             ancestry_model=None,
+                             random_seed=None,
+                             use_tskit=True,
+                            ):
+    '''
+    Use an msprime simulation to generate new Geonomics Individuals.
+    (docs in the main.py wrapper)
+    '''
+    # handle the recombination and mutation rates
+    if (not isinstance(recomb_rate, float) and
+        not (recomb_rate == 0)):
+        # NOTE: tack a 0.5 onto the righthand side of the recomb_rate vector,
+        #       to satisfy the difference between msprime (for which L
+        #       expresses length in intervals, each of which needs a recomb
+        #       rate specified, hence the recomb_rate vector needing to have a
+        #       number of rates equal to L) and gnx (for which L expressed
+        #       length in loci, between which recomb rates are needed, hence
+        #       needing L-1 recomb rates);
+        #       TODO: DOUBLE-CHECK THE FOLLOWING LOGIC
+        #       this 0.5 value will not affect gnx data because it effectively
+        #       sits between sites L-1 and L, but a simulated gnx genome of
+        #       length L only includes sites [0, L-1]
+        recomb_rate = [*recomb_rate] + [0.5]
+    recomb_rate = _format_msprime_rate(recomb_rate, L)
+    mut_rate = _format_msprime_rate(mut_rate, L)
+
+    # simulate ancestry and mutations
+    # TODO: look back at how gnx uses genome positions, because 
+    #       it looks like we need to provide discrete_genome=false
+    #       to sim_ancestry and true to sim_mutations;
+    #       or perhaps instead just fine to use false for both
+    #       and then subtract 0.5 from all non-zero 'left' values
+    #       from all non-L 'right' values when handles the edges table?
+    ts = msprime.sim_ancestry(samples=n,
+                              sequence_length=L,
+                              demography=demography,
+                              population_size=population_size,
+                              recombination_rate=recomb_rate,
+                              ploidy=2,
+                              model=ancestry_model,
+                              random_seed=random_seed,
+                              discrete_genome=True,
+                             )
+    # NOTE: mutations use a simple binary 0|1 model, with 0 always the
+    #       ancestral state, and with mutations being state-independent
+    #       (i.e., always placing a 1, even if this results in a silent
+    #       mutation of 11);
+    #       this model is the best aligned to gnx data
+    mut_model = msprime.BinaryMutationModel(state_independent=True)
+    mts = msprime.sim_mutations(tree_sequence=ts,
+                                rate=mut_rate,
+                                model=mut_model,
+                               )
+    # get the mutated tskit.TableCollection object
+    mtc = mts.dump_tables()
+
+    # double-check the number of Individuals
+    assert np.sum(mtc.nodes.asdict()['individual']>-1) == 2*n
+
+    # create the Individuals, assigning placeholder values for the attributes
+    # (to be later updated by downstream code)
+    individs = [Individual(idx=0, x=0, y=0) for i in range(n)]
+    # get the information linking Individuals to rows in TableCollection tables
+    nodes_tab_individs_col = mtc.nodes.asdict()['individual']
+    for i, ind in enumerate(individs):
+        # get the nodes table rows corresponding to this Individual
+        nodes_tab_ids = np.where(nodes_tab_individs_col==i)[0]
+        assert len(nodes_tab_ids) == 2
+        # set hidden attributes tracking the individs' rows in the 'nodes' table
+        # and id in the 'individuals' table
+        ind._set_nodes_tab_ids(nodes_tab_ids)
+        ind._individuals_tab_id = i
+        # if the gnx Model these Individuals are intended for is not using tskit
+        # then walk the variants generator and compile the individ's
+        # genotype array, then store it in the attribute 'g'
+        if not use_tskit:
+            var_gen = mts.variants(samples=nodes_tab_ids)
+            gt = np.stack([next(var_gen).genotypes for _ in range(L)])
+            ind.g = gt
+
+    return individs, mtc
+
+
+###############################################################################
+# FUNCTIONS FOR WORKING WITH THE SPATIAL PEDGREE SAVED IN tskit DATA STRUCTURES
+###############################################################################
+
+def _prep_tskit_tabcoll_for_gnx_spp_union(tc,
+                                          recip_spp,
+                                          coords,
+                                          source_software,
+                                         ):
+    '''
+    Reformat a tskit.TableCollection, produced by either a gnx simulation
+    or an msprime simulation, so that it conforms to Geonomics' expected format
+    and has the other data it needs to be fed into the .union() method
+    of the given Geonomics Species' current TableCollection.
+
+    The code and comments walk through each table, column by column,
+    and specify if and how the contents need to be reformatted or updated
+    in order to match Geonomics conventions.
+
+    Returns the list of gnx Individual idx numbers that were set in the
+    TableCollection, for downstream use and/or validation.
+    '''
+    # check source_software is valid
+    assert source_software in ['gnx', 'msprime'], ('This function can only '
+                                                   'prep TableCollections '
+                                                   'that originate from '
+                                                   'gnx Models or msprime '
+                                                   'simulations.')
+
+    # use the coords array to get the number of new individuals being unioned
+    # into the recipient population
+    assert (isinstance(coords, np.ndarray) and
+            len(coords.shape)==2 and
+            coords.shape[1] == 2), ("The coords param must be a 2d "
+                                    "numpy.ndarray with the x,y coords in "
+                                    "the 2 columns.")
+    n_source_individs = coords.shape[0]
+    if source_software == 'msprime':
+        assert tc.individuals.num_rows == n_source_individs
+
+    # reformat nodes table content:
+    #-----------------------------
+
+    if source_software == 'msprime':
+        # gnx uses the time column as follows:
+            # - smallest values are the most recent individuals
+            # - nodes with val 1 (i.e. -1 * t=-1 before main phase begins) are
+            #   nodes that originated at the end of the burn-in phase
+            # - nodes with val 0, -1, -2, ... (i.e., -1 * t =0, 1, 2,...)
+            #   originated during those time steps
+        # msprime times are expressed as t before present (i.e. samples are 0s).
+        # thus, convert msprime to gnx convention by subtracting from the whole
+        # column the value of spp.t at the time of their introduction
+        # (i.e., the time of their simulation and sampling)
+        time = tc.nodes.time - recip_spp.t
+        assert np.all([t>=(-1*recip_spp.t) for t in time]), ("Values "
+                            "attempting to be set on 'time' column of nodes "
+                            "table are invalid (i.e., are more negative than "
+                            "-1 * recipient Species' current model time, and "
+                            "thus would refer to future time steps not yet "
+                            "simulated.")
+        tc.nodes.time = time
+
+        # msprime uses the flags column to indicate extant nodes
+        # (i.e., docs say "NODE_IS_SAMPLE = 1"); gnx uses flags to distinguish
+        # between nodes in 'real' gnx Individuals and msprime nodes;
+        # thus, we actually just want to keep this column as is, because the
+        # samples are only nodes that will be assigned to new gnx Individuals,
+        # and thus are the only ones we would otherwise coerce to 1s anyhow!
+
+        # nothing to do the population column; TableCollection.union()
+        # will be asked to use the next available int in the recipient Species'
+        # populations table to automatically record the
+        # population these nodes came from
+
+        # the individual column needn't be touched; TableCollection.union()
+        # will automatically increment this col by using the next individual id
+        # available in the recipient Species' individuals table;
+        # we will just need to later track what the next id was and how many
+        # individuals were added, so that TableCollection.individuals ids can
+        # be matched to gnx Individuals via Individuals._individuals_tab_id attr
+
+        # metadata, metadata_offset, and metadata_schema columns are currently
+        # unused by both msprime and gnx
+
+    elif source_software == 'gnx':
+        # no need to alter any of this data if it's already from a gnx model
+        pass
+
+
+
+    # reformat individuals table content:
+    #-----------------------------------
+
+    if source_software == 'msprime':
+        # all individuals in the individuals table are extant and will be turned
+        # into gnx Individual objects, and gnx uses a 1 in the 'flags' column to
+        # indicated 'real' Individuals (i.e., individuals that were actual
+        # gnx.Individuals at some point), so set the entire 'flags' column to 1s
+        flags = tc.individuals.flags*0+1
+    elif source_software == 'gnx':
+        # flags should already be correct for a gnx model,
+        flags = tc.individuals.flags
+
+    # gnx stores individs' starting x,y coordinates in the 'location' column
+    # (NOTE: also the phenotype and fitness values, but those are not relevant
+    # in the neutral-loci-only msprime simulations permitted by gnx);
+    # TODO: think about if/how sweeps would be allowed and
+    #       how they would be represented/matched up to genarch
+    # location_offset tells the TableCollection to jump 2 values (x, y) per row
+    # NOTE: 'ragged columns' actually consist of the column's contents in a
+    #       flattened array format and a column of n+1 index values that can be
+    #       offset and used to index the start and end positions of each row's
+    #       contents within that array
+    if source_software == 'msprime':
+        location_offset = [int(n) for n in np.linspace(0,
+                                                   2*tc.individuals.num_rows,
+                                                   tc.individuals.num_rows+1)]
+    elif source_software == 'gnx':
+        # in a TableCollection from a gnx Model, locations will need to be
+        # blanked out for all Individuals, current or past (so that fns that
+        # later depend on the stored locations information don't accidentally
+        # use data conveying locations on a different landscape from before the
+        # introduction of these source Individuals to their new population),
+        # and then will need to be reset to the new coords values for only the
+        # bottom n rows; thus, need to tack leading zeros onto the
+        # location_offset values just generated, making the final vector
+        # num_rows+1 long
+        location_offset = [int(n) for n in np.linspace(0,
+                                                   2*n_source_individs,
+                                                   n_source_individs + 1)]
+        location_offset = [0] * (tc.individuals.num_rows -
+                                 n_source_individs) + location_offset
+        assert len(location_offset) == (tc.individuals.num_rows + 1), ("The "
+                                "location_offset vector must be one longer "
+                                "than the number of rows in the individuals "
+                                "table on which it will be set.")
+
+    # parents column is unused by geonomics
+    # (wasn't implemented until 21/01/21, after gnx was developed;
+    # tskit-dev commit 8ebfd8b2900217dab2b82d6f530c830238037fc8).
+    # useful but not necessary to implement;
+    # this column is also empty in msprime
+
+    # gnx uses the metadata column to store each individual's
+    # gnx Individual idx, for later matching to rows in the individuals table
+    # when the table is updated after tskit's simplify algo filters individuals;
+    # determine the next n indx values using the Species.max_ind_idx attribute
+    gnx_individ_idxs = [*range(recip_spp.max_ind_idx+1,
+                       recip_spp.max_ind_idx + n_source_individs + 1)]
+    metadata=[idx.to_bytes(length=4,
+                           byteorder='little') for idx in gnx_individ_idxs]
+    # for msprime, make location_offset vals uniform across new Individs' rows
+    if source_software == 'msprime':
+        metadata_offset = [int(n) for n in np.linspace(0,
+                                                   4*tc.individuals.num_rows,
+                                                   tc.individuals.num_rows+1)]
+    elif source_software == 'gnx':
+        # same as for location_offset above, need to tack leading zeros onto
+        # the metadata vector if setting columns in a gnx-generated individuals
+        # table
+        metadata_offset = [int(n) for n in np.linspace(0,
+                                                   4*n_source_individs,
+                                                   n_source_individs+1)]
+        metadata_offset = [0] * (tc.individuals.num_rows -
+                                 n_source_individs) + metadata_offset
+        assert len(metadata_offset) == (tc.individuals.num_rows + 1), ("The "
+                                "metadata_offset vector must be one longer "
+                                "than the number of rows in the individuals "
+                                "table on which it will be set.")
+
+    # NOTE the set_columns() method fails because the metadata column, although
+    #      supposed to be a binary data type according to the docs, is actually
+    #      coded as np.int8 and so fails to convert the 4-byte binary data we
+    #      are trying to provide; never ran into this problem in tha past
+    #      because the .add_row() method allows the binary data for some reason
+    #      and that's the only method I've used before (just adding 1
+    #      individual at a time as they're born);
+    #      however, somewhat conveniently, we were already planning to change
+    #      every column already containing data in this table anyhow, so for
+    #      now we are just:
+    #           1.) double-checking the length of the original table
+    #           2.) clearing the table
+    #           3.) manually looping and adding new rows
+    #           4.) confirming identical resulting length
+    if source_software == 'msprime':
+        assert (len(flags) ==
+            coords.shape[0] ==
+            (len(location_offset)-1) ==
+            len(metadata) ==
+            (len(metadata_offset)-1)), ("The 'flags', 'location', and "
+                                        "'metadata' columns for the "
+                                        "individuals table must all "
+                                        "be the same length as the coords "
+                                        "array's 0th dimension and the "
+                                        "'location_offset' and "
+                                        "'metadata_offset' columns must be "
+                                        "one longer.")
+    elif source_software == 'gnx':
+        assert (len(flags) ==
+                (len(location_offset)-1) ==
+                (len(metadata_offset)-1)), ("The 'flags' column must be "
+                            "one shorter than the 'location_offset' and "
+                            "'metadata_offset' columns.")
+        assert len(metadata) == coords.shape[0] == n_source_individs, ("The "
+                            "'metadata' column must have one value "
+                            "per coordinate pair provided.")
+    original_inds_tab_len = tc.individuals.num_rows
+    if source_software == 'msprime':
+        assert original_inds_tab_len == coords.shape[0], ("Length "
+                                             "of individuals table "
+                                             "generated by msprime "
+                                             "differs from length of coords "
+                                             "attempting to be added to it.")
+    tc.individuals.clear()
+    for i, flag in enumerate(flags):
+        # for gnx, add original flags and otherwise empty rows for all but the
+        # final rows corresponding to the Individuals being unioned into a
+        # recipient Species
+        if source_software == 'gnx':
+            if i < (len(flags) - n_source_individs):
+                row_coords = None
+                row_metadata = None
+            else:
+                i_adjusted = i - (len(flags) - n_source_individs)
+                row_coords = coords[i_adjusted, :]
+                row_metadata = metadata[i_adjusted]
+        elif source_software == 'msprime':
+            row_coords = coords[i, :]
+            row_metadata = metadata[i]
+        tc.individuals.add_row(flags=flag,
+                               location=row_coords,
+                               parents=None,
+                               metadata=row_metadata,
+                              )
+    assert tc.individuals.num_rows == original_inds_tab_len, ("Reconstructed "
+                                            "individuals table is not the same "
+                                            "length as the original table "
+                                            "generated by msprime.")
+
+
+    # reformat edges table content:
+    #-----------------------------
+
+    # geonomics models all edges as being delineated by positions belonging
+    # to the set [0, 0.5, 1.5, ..., L-2.5, L-1.5, L];
+    # this is more complicated than the saner msprime
+    # convention of using the set [0, 1, 2, ..., L-2, L-1, L],
+    # but this was a decision I made based on modeling recombination
+    # breakpoints halfway between loci and it is equally functional,
+    # so for msprime-derived tables we need to adjust the positions
+    if source_software == 'msprime':
+        L = recip_spp.gen_arch.L
+        left = tc.edges.left
+        right = tc.edges.right
+        left = np.clip(left - 0.5, a_min=0, a_max=L)
+        right = np.clip([r-0.5 if r != L else r for r in right], a_min=0, a_max=L)
+        tc.edges.left = left
+        tc.edges.right = right
+
+    # parent and child columns are both fine as is
+    # (have double-checked that node ids are all automatically and correctly
+    # updated by the tskit.TableCollection.union() method)
+
+    # metadata column is currently unused by both msprime and gnx
+
+
+    # reformat sites table content:
+    #-----------------------------
+
+    # double-check that position and ancestral state columns are identical in
+    # the recipient Species' TableCollection;
+    # if so, then just duplicate that table here
+    # (because the only difference between the two in the metadata column,
+    # which in gnx stores whether a site started as a neutral ('n')
+    # or non-neutral ('t', for trait-influencing) site
+    # NOTE: not running these checks for msprime-derived tables because
+    #       they can lack some sites, depending on recomb rate, Ne, etc
+    if source_software == 'gnx':
+        for col in ['position', 'ancestral_state']:
+            assert np.all(getattr(tc.sites, col) ==
+                          getattr(recip_spp._tc.sites, col)), (
+                                f"The {col} column in the sites table "
+                                f"generated by msprime disagrees with the {col} "
+                                 "column in the recipient Species' sites table.")
+    tc.sites.replace_with(recip_spp._tc.sites)
+
+
+    # reformat mutations table content:
+    #---------------------------------
+
+    # site, node, and parent columns are the core pieces needed to recover the
+    # genotypes produced by the mutations, and we want whatever genotypes
+    # msprime has simulated, so these columns should all remain untouched
+    # site vals are all in interval [0, L-1]
+    assert (np.all(0 <= tc.mutations.site) and
+            np.all(tc.mutations.site <= (recip_spp.gen_arch.L-1))), ("The "
+                        "site column in the mutations table contains values "
+                        "not in the interval [0, recip_spp.gen_arch.L-1].")
+
+    # Times for mutations created at the start of a gnx simulation, to
+    # match the starting allele frequency array ('p') provided to a gnx
+    # model, are artificial as they are all forcibly mutated at at once.
+    # Thus, those times are NaN. Unfortunately, tskit does not all
+    # mutation times to be mixed known and unknwon, so that means we must
+    # set all mutation times to NaN
+    time = [tskit.UNKNOWN_TIME] * len(tc.mutations.time)
+    tc.mutations.time = time
+
+    # double-check that the derived_state_offset column is just serial
+    # integers from 0 to the table length, inclusive
+    assert np.all(tc.mutations.derived_state_offset ==
+                  np.linspace(0,
+                              tc.mutations.num_rows,
+                              tc.mutations.num_rows + 1)), ("The "
+                "derived_state_offset column of the mutations table should "
+                "contain serial integers from 0 to the table length, "
+                "inclusive, but does not.")
+
+    # metadata, metadata_offset, and metadata_schema are all currently unused
+    # in both msprime and gnx
+
+
+    # reformat migrations table content:
+    #-----------------------------------
+
+    # this table is not used in gnx, but if it was used in a complex msprime
+    # Demography that simulated individuals for introduction into gnx then it
+    # will not be empty in this TableCollection; in that case, its contents
+    # will just be unioned into the otherwise empty migrations table of the 
+    # recipient Species and then left as is
+
+
+    # reformat populations table content:
+    #-----------------------------------
+
+    # This table is unused by gnx, but the msprime table has
+    # basic population names (e.g., 'pop_0') and descriptions ('' by default,
+    # but a user could specify).
+    # We won't touch or do anything with this data, and the
+    # TableCollection.union() method will pull it along without editing,
+    # but the rows in this table will be appended to existing rows in the gnx
+    # model and the nodes table will be constructed so as to match those indices
+
+    # reformat provenances table content:
+    #-----------------------------------
+
+    # no need to do anything here; all provenances info will be retained and
+    # combined when the union() method is called on the recipient Species'
+    # TableCollection
+
+    # return the gnx idxs that were set in the individuals table, for
+    # downstream use
+    return gnx_individ_idxs
+
 
 def _get_lineage_dicts(spp, nodes, loci, t_curr, drop_before_sim=True,
                        time_before_present=True, max_time_ago=None,
                        min_time_ago=None):
     '''
     For a sample of nodes and a sample of loci, get the
     loci-sequentially ordered dict of lineage dicts
@@ -1386,7 +1861,28 @@
     if len(lin_dict.values()) < 2:
         return
     else:
         dist = _calc_lineage_distance(lin_dict)
         time = _calc_lineage_time(lin_dict)
         speed = dist/time
         return speed
+
+
+
+
+#######################################################
+#######################################################
+#######################################################
+#######################################################
+#######################################################
+#######################################################
+#######################################################
+
+
+import numpy as np
+import geonomics as gnx
+import msprime
+import tskit
+import warnings
+from copy import deepcopy
+from collections import Counter as C
+
```

### Comparing `geonomics-1.3.9/geonomics/structs/individual.py` & `geonomics-1.4.0/geonomics/structs/individual.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/structs/landscape.py` & `geonomics-1.4.0/geonomics/structs/landscape.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/structs/species.py` & `geonomics-1.4.0/geonomics/structs/species.py`

 * *Files 13% similar despite different names*

```diff
@@ -21,7516 +21,9573 @@
 00000140: 696f 6e5f 7261 7465 732c 0a20 2020 2020  ion_rates,.     
 00000150: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00000160: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00000170: 205f 6d61 6b65 5f73 7461 7274 696e 675f   _make_starting_
 00000180: 6d75 7461 7469 6f6e 732c 0a20 2020 2020  mutations,.     
 00000190: 2020 2020 2020 2020 2020 2020 2020 2020                  
 000001a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000001b0: 205f 6765 745f 6c69 6e65 6167 655f 6469   _get_lineage_di
-000001c0: 6374 732c 0a20 2020 2020 2020 2020 2020  cts,.           
-000001d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000001e0: 2020 2020 2020 2020 2020 205f 6765 745f             _get_
-000001f0: 6c69 6e65 6167 655f 6469 6374 735f 6f6e  lineage_dicts_on
-00000200: 655f 7472 6565 2c0a 2020 2020 2020 2020  e_tree,.        
-00000210: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000220: 2020 2020 2020 2020 2020 2020 2020 5f67                _g
-00000230: 6574 5f74 7265 656e 756d 732c 0a20 2020  et_treenums,.   
-00000240: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000250: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000260: 2020 205f 6361 6c63 5f6c 696e 6561 6765     _calc_lineage
-00000270: 5f73 7461 7429 0a66 726f 6d20 6765 6f6e  _stat).from geon
-00000280: 6f6d 6963 732e 7374 7275 6374 732e 6c61  omics.structs.la
-00000290: 6e64 7363 6170 6520 696d 706f 7274 204c  ndscape import L
-000002a0: 6179 6572 0a66 726f 6d20 6765 6f6e 6f6d  ayer.from geonom
-000002b0: 6963 732e 7374 7275 6374 732e 696e 6469  ics.structs.indi
-000002c0: 7669 6475 616c 2069 6d70 6f72 7420 496e  vidual import In
-000002d0: 6469 7669 6475 616c 2c20 5f6d 616b 655f  dividual, _make_
-000002e0: 696e 6469 7669 6475 616c 0a66 726f 6d20  individual.from 
-000002f0: 6765 6f6e 6f6d 6963 732e 6f70 732e 6d6f  geonomics.ops.mo
-00000300: 7665 6d65 6e74 2069 6d70 6f72 7420 5f64  vement import _d
-00000310: 6f5f 6d6f 7665 6d65 6e74 2c20 5f64 6f5f  o_movement, _do_
-00000320: 6469 7370 6572 7361 6c0a 6672 6f6d 2067  dispersal.from g
-00000330: 656f 6e6f 6d69 6373 2e6f 7073 2e6d 6174  eonomics.ops.mat
-00000340: 696e 6720 696d 706f 7274 205f 6669 6e64  ing import _find
-00000350: 5f6d 6174 6573 2c20 5f64 7261 775f 6e5f  _mates, _draw_n_
-00000360: 6269 7274 6873 2c20 5f64 6f5f 6d61 7469  births, _do_mati
-00000370: 6e67 0a66 726f 6d20 6765 6f6e 6f6d 6963  ng.from geonomic
-00000380: 732e 6f70 732e 7365 6c65 6374 696f 6e20  s.ops.selection 
-00000390: 696d 706f 7274 205f 6361 6c63 5f66 6974  import _calc_fit
-000003a0: 6e65 7373 0a66 726f 6d20 6765 6f6e 6f6d  ness.from geonom
-000003b0: 6963 732e 6f70 732e 6d75 7461 7469 6f6e  ics.ops.mutation
-000003c0: 2069 6d70 6f72 7420 285f 646f 5f6d 7574   import (_do_mut
-000003d0: 6174 696f 6e2c 0a20 2020 2020 2020 2020  ation,.         
-000003e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000003f0: 2020 2020 2020 2020 2020 205f 6361 6c63             _calc
-00000400: 5f65 7374 696d 6174 6564 5f74 6f74 616c  _estimated_total
-00000410: 5f6d 7574 6174 696f 6e73 290a 6672 6f6d  _mutations).from
-00000420: 2067 656f 6e6f 6d69 6373 2e6f 7073 2e64   geonomics.ops.d
-00000430: 656d 6f67 7261 7068 7920 696d 706f 7274  emography import
-00000440: 205f 646f 5f70 6f70 5f64 796e 616d 6963   _do_pop_dynamic
-00000450: 732c 205f 6361 6c63 5f6c 6f67 6973 7469  s, _calc_logisti
-00000460: 635f 736f 6c6e 0a66 726f 6d20 6765 6f6e  c_soln.from geon
-00000470: 6f6d 6963 732e 6f70 732e 6368 616e 6765  omics.ops.change
-00000480: 2069 6d70 6f72 7420 5f53 7065 6369 6573   import _Species
-00000490: 4368 616e 6765 720a 6672 6f6d 2067 656f  Changer.from geo
-000004a0: 6e6f 6d69 6373 2e73 696d 2069 6d70 6f72  nomics.sim impor
-000004b0: 7420 6275 726e 696e 0a0a 236f 7468 6572  t burnin..#other
-000004c0: 2069 6d70 6f72 7473 0a69 6d70 6f72 7420   imports.import 
-000004d0: 6e75 6d70 7920 6173 206e 700a 6672 6f6d  numpy as np.from
-000004e0: 206e 756d 7079 2069 6d70 6f72 7420 7261   numpy import ra
-000004f0: 6e64 6f6d 2061 7320 720a 696d 706f 7274  ndom as r.import
-00000500: 2070 616e 6461 7320 6173 2070 640a 696d   pandas as pd.im
-00000510: 706f 7274 2072 616e 646f 6d0a 696d 706f  port random.impo
-00000520: 7274 206d 6174 706c 6f74 6c69 6220 6173  rt matplotlib as
-00000530: 206d 706c 0a76 697a 2e5f 6368 6563 6b5f   mpl.viz._check_
-00000540: 6469 7370 6c61 7928 290a 696d 706f 7274  display().import
-00000550: 206d 6174 706c 6f74 6c69 622e 7079 706c   matplotlib.pypl
-00000560: 6f74 2061 7320 706c 740a 6672 6f6d 2073  ot as plt.from s
-00000570: 6b6c 6561 726e 2e63 726f 7373 5f64 6563  klearn.cross_dec
-00000580: 6f6d 706f 7369 7469 6f6e 2069 6d70 6f72  omposition impor
-00000590: 7420 4343 410a 6672 6f6d 2073 6369 7079  t CCA.from scipy
-000005a0: 2e73 7461 7473 2e64 6973 7472 6962 7574  .stats.distribut
-000005b0: 696f 6e73 2069 6d70 6f72 7420 6e6f 726d  ions import norm
-000005c0: 0a66 726f 6d20 6d61 7468 2069 6d70 6f72  .from math impor
-000005d0: 7420 7069 2061 7320 6d61 7468 5f70 690a  t pi as math_pi.
-000005e0: 6672 6f6d 2063 6f6c 6c65 6374 696f 6e73  from collections
-000005f0: 2069 6d70 6f72 7420 436f 756e 7465 7220   import Counter 
-00000600: 6173 2043 0a66 726f 6d20 636f 6c6c 6563  as C.from collec
-00000610: 7469 6f6e 7320 696d 706f 7274 204f 7264  tions import Ord
-00000620: 6572 6564 4469 6374 2061 7320 4f44 0a69  eredDict as OD.i
-00000630: 6d70 6f72 7420 7473 6b69 740a 696d 706f  mport tskit.impo
-00000640: 7274 206d 7370 7269 6d65 0a66 726f 6d20  rt msprime.from 
-00000650: 636f 7079 2069 6d70 6f72 7420 6465 6570  copy import deep
-00000660: 636f 7079 0a66 726f 6d20 6f70 6572 6174  copy.from operat
-00000670: 6f72 2069 6d70 6f72 7420 6974 656d 6765  or import itemge
-00000680: 7474 6572 0a66 726f 6d20 6f70 6572 6174  tter.from operat
-00000690: 6f72 2069 6d70 6f72 7420 6174 7472 6765  or import attrge
-000006a0: 7474 6572 0a69 6d70 6f72 7420 7379 730a  tter.import sys.
-000006b0: 0a0a 2323 2323 2323 2323 2323 2323 2323  ..##############
-000006c0: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-000006d0: 2323 2323 2323 2323 0a23 202d 2d2d 2d2d  ########.# -----
-000006e0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-000006f0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 230a  --------------#.
-00000700: 2320 434c 4153 5345 5320 2d2d 2d2d 2d2d  # CLASSES ------
-00000710: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-00000720: 2d2d 2d2d 2d23 0a23 202d 2d2d 2d2d 2d2d  -----#.# -------
-00000730: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-00000740: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 230a 2323  ------------#.##
+000001b0: 205f 7072 6570 5f74 736b 6974 5f74 6162   _prep_tskit_tab
+000001c0: 636f 6c6c 5f66 6f72 5f67 6e78 5f73 7070  coll_for_gnx_spp
+000001d0: 5f75 6e69 6f6e 2c0a 2020 2020 2020 2020  _union,.        
+000001e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000001f0: 2020 2020 2020 2020 2020 2020 2020 5f73                _s
+00000200: 696d 5f6d 7370 7269 6d65 5f69 6e64 6976  im_msprime_indiv
+00000210: 6964 7561 6c73 2c0a 2020 2020 2020 2020  iduals,.        
+00000220: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000230: 2020 2020 2020 2020 2020 2020 2020 5f67                _g
+00000240: 6574 5f6c 696e 6561 6765 5f64 6963 7473  et_lineage_dicts
+00000250: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00000260: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000270: 2020 2020 2020 2020 5f67 6574 5f6c 696e          _get_lin
+00000280: 6561 6765 5f64 6963 7473 5f6f 6e65 5f74  eage_dicts_one_t
+00000290: 7265 652c 0a20 2020 2020 2020 2020 2020  ree,.           
+000002a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000002b0: 2020 2020 2020 2020 2020 205f 6765 745f             _get_
+000002c0: 7472 6565 6e75 6d73 2c0a 2020 2020 2020  treenums,.      
+000002d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000002e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000002f0: 5f63 616c 635f 6c69 6e65 6167 655f 7374  _calc_lineage_st
+00000300: 6174 290a 6672 6f6d 2067 656f 6e6f 6d69  at).from geonomi
+00000310: 6373 2e73 7472 7563 7473 2e6c 616e 6473  cs.structs.lands
+00000320: 6361 7065 2069 6d70 6f72 7420 4c61 7965  cape import Laye
+00000330: 720a 6672 6f6d 2067 656f 6e6f 6d69 6373  r.from geonomics
+00000340: 2e73 7472 7563 7473 2e69 6e64 6976 6964  .structs.individ
+00000350: 7561 6c20 696d 706f 7274 2049 6e64 6976  ual import Indiv
+00000360: 6964 7561 6c2c 205f 6d61 6b65 5f69 6e64  idual, _make_ind
+00000370: 6976 6964 7561 6c0a 6672 6f6d 2067 656f  ividual.from geo
+00000380: 6e6f 6d69 6373 2e6f 7073 2e6d 6f76 656d  nomics.ops.movem
+00000390: 656e 7420 696d 706f 7274 205f 646f 5f6d  ent import _do_m
+000003a0: 6f76 656d 656e 742c 205f 646f 5f64 6973  ovement, _do_dis
+000003b0: 7065 7273 616c 0a66 726f 6d20 6765 6f6e  persal.from geon
+000003c0: 6f6d 6963 732e 6f70 732e 6d61 7469 6e67  omics.ops.mating
+000003d0: 2069 6d70 6f72 7420 5f66 696e 645f 6d61   import _find_ma
+000003e0: 7465 732c 205f 6472 6177 5f6e 5f62 6972  tes, _draw_n_bir
+000003f0: 7468 732c 205f 646f 5f6d 6174 696e 670a  ths, _do_mating.
+00000400: 6672 6f6d 2067 656f 6e6f 6d69 6373 2e6f  from geonomics.o
+00000410: 7073 2e73 656c 6563 7469 6f6e 2069 6d70  ps.selection imp
+00000420: 6f72 7420 5f63 616c 635f 6669 746e 6573  ort _calc_fitnes
+00000430: 730a 6672 6f6d 2067 656f 6e6f 6d69 6373  s.from geonomics
+00000440: 2e6f 7073 2e6d 7574 6174 696f 6e20 696d  .ops.mutation im
+00000450: 706f 7274 2028 5f64 6f5f 6d75 7461 7469  port (_do_mutati
+00000460: 6f6e 2c0a 2020 2020 2020 2020 2020 2020  on,.            
+00000470: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000480: 2020 2020 2020 2020 5f63 616c 635f 6573          _calc_es
+00000490: 7469 6d61 7465 645f 746f 7461 6c5f 6d75  timated_total_mu
+000004a0: 7461 7469 6f6e 7329 0a66 726f 6d20 6765  tations).from ge
+000004b0: 6f6e 6f6d 6963 732e 6f70 732e 6465 6d6f  onomics.ops.demo
+000004c0: 6772 6170 6879 2069 6d70 6f72 7420 5f64  graphy import _d
+000004d0: 6f5f 706f 705f 6479 6e61 6d69 6373 2c20  o_pop_dynamics, 
+000004e0: 5f63 616c 635f 6c6f 6769 7374 6963 5f73  _calc_logistic_s
+000004f0: 6f6c 6e0a 6672 6f6d 2067 656f 6e6f 6d69  oln.from geonomi
+00000500: 6373 2e6f 7073 2e63 6861 6e67 6520 696d  cs.ops.change im
+00000510: 706f 7274 205f 5370 6563 6965 7343 6861  port _SpeciesCha
+00000520: 6e67 6572 0a66 726f 6d20 6765 6f6e 6f6d  nger.from geonom
+00000530: 6963 732e 7369 6d20 696d 706f 7274 2062  ics.sim import b
+00000540: 7572 6e69 6e0a 0a23 6f74 6865 7220 696d  urnin..#other im
+00000550: 706f 7274 730a 696d 706f 7274 206e 756d  ports.import num
+00000560: 7079 2061 7320 6e70 0a66 726f 6d20 6e75  py as np.from nu
+00000570: 6d70 7920 696d 706f 7274 2072 616e 646f  mpy import rando
+00000580: 6d20 6173 2072 0a69 6d70 6f72 7420 7061  m as r.import pa
+00000590: 6e64 6173 2061 7320 7064 0a69 6d70 6f72  ndas as pd.impor
+000005a0: 7420 7261 6e64 6f6d 0a69 6d70 6f72 7420  t random.import 
+000005b0: 6d61 7470 6c6f 746c 6962 2061 7320 6d70  matplotlib as mp
+000005c0: 6c0a 7669 7a2e 5f63 6865 636b 5f64 6973  l.viz._check_dis
+000005d0: 706c 6179 2829 0a69 6d70 6f72 7420 6d61  play().import ma
+000005e0: 7470 6c6f 746c 6962 2e70 7970 6c6f 7420  tplotlib.pyplot 
+000005f0: 6173 2070 6c74 0a66 726f 6d20 736b 6c65  as plt.from skle
+00000600: 6172 6e2e 6372 6f73 735f 6465 636f 6d70  arn.cross_decomp
+00000610: 6f73 6974 696f 6e20 696d 706f 7274 2043  osition import C
+00000620: 4341 0a66 726f 6d20 7363 6970 792e 7374  CA.from scipy.st
+00000630: 6174 732e 6469 7374 7269 6275 7469 6f6e  ats.distribution
+00000640: 7320 696d 706f 7274 206e 6f72 6d0a 6672  s import norm.fr
+00000650: 6f6d 206d 6174 6820 696d 706f 7274 2070  om math import p
+00000660: 6920 6173 206d 6174 685f 7069 0a66 726f  i as math_pi.fro
+00000670: 6d20 636f 6c6c 6563 7469 6f6e 7320 696d  m collections im
+00000680: 706f 7274 2043 6f75 6e74 6572 2061 7320  port Counter as 
+00000690: 430a 6672 6f6d 2063 6f6c 6c65 6374 696f  C.from collectio
+000006a0: 6e73 2069 6d70 6f72 7420 4f72 6465 7265  ns import Ordere
+000006b0: 6444 6963 7420 6173 204f 440a 696d 706f  dDict as OD.impo
+000006c0: 7274 2074 736b 6974 0a69 6d70 6f72 7420  rt tskit.import 
+000006d0: 6d73 7072 696d 650a 6672 6f6d 2063 6f70  msprime.from cop
+000006e0: 7920 696d 706f 7274 2064 6565 7063 6f70  y import deepcop
+000006f0: 790a 6672 6f6d 206f 7065 7261 746f 7220  y.from operator 
+00000700: 696d 706f 7274 2069 7465 6d67 6574 7465  import itemgette
+00000710: 720a 6672 6f6d 206f 7065 7261 746f 7220  r.from operator 
+00000720: 696d 706f 7274 2061 7474 7267 6574 7465  import attrgette
+00000730: 720a 696d 706f 7274 2077 6172 6e69 6e67  r.import warning
+00000740: 730a 696d 706f 7274 2073 7973 0a0a 0a23  s.import sys...#
 00000750: 2323 2323 2323 2323 2323 2323 2323 2323  ################
 00000760: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-00000770: 2323 2323 0a0a 0a23 6120 7369 6d70 6c65  ####...#a simple
-00000780: 206c 6974 746c 6520 636c 6173 7320 696e   little class in
-00000790: 746f 2077 6869 6368 2070 6172 616d 732d  to which params-
-000007a0: 6669 6c65 2070 6172 616d 6574 6572 7320  file parameters 
-000007b0: 616e 6420 7468 6569 7220 7661 6c75 6573  and their values
-000007c0: 2077 696c 6c0a 2362 6520 6475 6d70 6564   will.#be dumped
-000007d0: 2028 6173 2061 7474 7269 6275 7465 7329   (as attributes)
-000007e0: 2e20 5468 6973 2063 6c61 7373 2077 696c  . This class wil
-000007f0: 6c20 6265 2061 6464 6564 2061 7320 6120  l be added as a 
-00000800: 6869 6464 656e 2061 7474 7269 6275 7465  hidden attribute
-00000810: 2074 6f0a 2374 6865 2053 7065 6369 6573   to.#the Species
-00000820: 2e20 4361 6e20 646f 2074 6869 7320 736f  . Can do this so
-00000830: 2074 6861 7420 7468 6573 6520 7061 7261   that these para
-00000840: 6d65 7465 7273 2720 7661 6c75 6573 2063  meters' values c
-00000850: 616e 2073 7469 6c6c 2062 6520 0a23 6163  an still be .#ac
-00000860: 6365 7373 6564 2062 7920 4765 6f6e 6f6d  cessed by Geonom
-00000870: 6963 7320 6675 6e63 7469 6f6e 7320 6561  ics functions ea
-00000880: 7369 6c79 2028 696e 2066 6163 742c 2074  sily (in fact, t
-00000890: 6865 2073 616d 6520 6173 2069 6620 7468  he same as if th
-000008a0: 6579 2077 6572 650a 2353 7065 6369 6573  ey were.#Species
-000008b0: 2061 7474 7269 6275 7465 7320 6265 6361   attributes beca
-000008c0: 7573 6520 6f66 2068 6f77 2074 6865 2053  use of how the S
-000008d0: 7065 6369 6573 2720 5f5f 6765 7461 7474  pecies' __getatt
-000008e0: 725f 5f20 6d65 7468 6f64 2069 730a 2361  r__ method is.#a
-000008f0: 6c74 6572 6564 2062 656c 6f77 292c 2062  ltered below), b
-00000900: 7574 2074 6865 7920 7769 6c6c 2062 6520  ut they will be 
-00000910: 6869 6464 656e 2074 6f20 7468 6520 7265  hidden to the re
-00000920: 6775 6c61 7220 7573 6572 2028 736f 2074  gular user (so t
-00000930: 6861 7420 7468 6579 0a23 646f 6e27 7420  hat they.#don't 
-00000940: 6170 7065 6172 2061 7320 7468 6f75 6768  appear as though
-00000950: 2074 6865 7920 636f 756c 6420 6265 206a   they could be j
-00000960: 7573 7420 6265 2063 6861 6e67 6564 2069  ust be changed i
-00000970: 6e20 6120 6375 7374 6f6d 206d 6f64 656c  n a custom model
-00000980: 2073 6372 6970 740a 2374 6f20 6368 616e   script.#to chan
-00000990: 6765 206d 6f64 656c 2062 6568 6176 696f  ge model behavio
-000009a0: 7220 2873 696e 6365 2073 6f6d 6520 6f66  r (since some of
-000009b0: 2074 6865 6d20 6172 6520 7573 6564 2074   them are used t
-000009c0: 6f20 696e 7374 616e 7469 6174 6520 6f62  o instantiate ob
-000009d0: 6a65 6374 730a 2361 7420 7468 6520 7469  jects.#at the ti
-000009e0: 6d65 2074 6865 206d 6f64 656c 2069 7320  me the model is 
-000009f0: 6d61 6465 2c20 736f 2061 7265 2027 6261  made, so are 'ba
-00000a00: 6b65 6420 696e 272c 2073 7563 6820 7468  ked in', such th
-00000a10: 6174 2074 686f 7365 206f 626a 6563 7473  at those objects
-00000a20: 0a23 776f 756c 6420 6e65 6564 2074 6f20  .#would need to 
-00000a30: 6265 2072 6569 6e73 7461 6e74 6961 7465  be reinstantiate
-00000a40: 6420 6561 6368 2074 696d 6520 7468 6520  d each time the 
-00000a50: 7061 7261 6d73 2063 6f6e 6365 726e 6564  params concerned
-00000a60: 2077 6572 6520 6368 616e 6765 6429 2e0a   were changed)..
-00000a70: 636c 6173 7320 5f50 6172 616d 7356 616c  class _ParamsVal
-00000a80: 733a 0a20 2020 2064 6566 205f 5f69 6e69  s:.    def __ini
-00000a90: 745f 5f28 7365 6c66 2c20 7370 705f 6e61  t__(self, spp_na
-00000aa0: 6d65 293a 0a20 2020 2020 2020 2073 656c  me):.        sel
-00000ab0: 662e 7370 705f 6e61 6d65 203d 2073 7070  f.spp_name = spp
-00000ac0: 5f6e 616d 650a 0a0a 2374 6865 2053 7065  _name...#the Spe
-00000ad0: 6369 6573 2063 6c61 7373 0a63 6c61 7373  cies class.class
-00000ae0: 2053 7065 6369 6573 284f 4429 3a0a 2020   Species(OD):.  
-00000af0: 2020 2222 220a 2020 2020 5265 7072 6573    """.    Repres
-00000b00: 656e 7461 7469 6f6e 206f 6620 7468 6520  entation of the 
-00000b10: 746f 7461 6c20 7369 6d75 6c61 7465 6420  total simulated 
-00000b20: 706f 7075 6c61 7469 6f6e 206f 6620 6120  population of a 
-00000b30: 7370 6563 6965 732e 0a0a 2020 2020 4f72  species...    Or
-00000b40: 6761 6e69 7a65 6420 6173 2061 2063 6f6c  ganized as a col
-00000b50: 6c65 6374 696f 6e73 2e4f 7264 6572 6564  lections.Ordered
-00000b60: 4469 6374 206f 6620 7365 7269 616c 2069  Dict of serial i
-00000b70: 6e74 6567 6572 2d6b 6579 6564 2049 6e64  nteger-keyed Ind
-00000b80: 6976 6964 7561 6c73 2e0a 2020 2020 5468  ividuals..    Th
-00000b90: 6f73 6520 7365 7269 616c 2069 6e64 6963  ose serial indic
-00000ba0: 6573 2063 6f6e 7469 6e75 616c 6c79 2069  es continually i
-00000bb0: 6e63 7265 6d65 6e74 2074 6872 6f75 6768  ncrement through
-00000bc0: 206d 6f64 656c 2074 696d 6520 6173 206e   model time as n
-00000bd0: 6577 0a20 2020 2049 6e64 6976 6964 7561  ew.    Individua
-00000be0: 6c73 2028 692e 652e 206f 6666 7370 7269  ls (i.e. offspri
-00000bf0: 6e67 2920 6173 2070 726f 6475 6365 642c  ng) as produced,
-00000c00: 2073 7563 6820 7468 6174 206e 6f20 7477   such that no tw
-00000c10: 6f20 496e 6469 7669 6475 616c 730a 2020  o Individuals.  
-00000c20: 2020 7769 7468 696e 2074 6865 2066 756c    within the ful
-00000c30: 6c20 6869 7374 6f72 7920 6f66 2061 2073  l history of a s
-00000c40: 696d 756c 6174 696f 6e20 7769 6c6c 2065  imulation will e
-00000c50: 7665 7220 6861 7665 2074 6865 2073 616d  ver have the sam
-00000c60: 6520 696e 6465 780a 2020 2020 6e75 6d62  e index.    numb
-00000c70: 6572 2e0a 0a20 2020 2042 6563 6175 7365  er...    Because
-00000c80: 2074 6865 2053 7065 6369 6573 2063 6c61   the Species cla
-00000c90: 7373 2069 6e68 6572 6974 7320 6672 6f6d  ss inherits from
-00000ca0: 2060 636f 6c6c 6563 7469 6f6e 732e 4f72   `collections.Or
-00000cb0: 6465 7265 6444 6963 7460 2c0a 2020 2020  deredDict`,.    
-00000cc0: 496e 6469 7669 6475 616c 7320 6361 6e20  Individuals can 
-00000cd0: 6265 2069 6e64 6578 6564 206f 7574 2075  be indexed out u
-00000ce0: 7369 6e67 2074 6865 6972 2069 6e64 6578  sing their index
-00000cf0: 2d6e 756d 6265 7220 6b65 7973 0a20 2020  -number keys.   
-00000d00: 2028 652e 672e 2053 7065 6369 6573 5b3c   (e.g. Species[<
-00000d10: 6964 783e 5d29 2e0a 0a20 2020 2041 6c6c  idx>])...    All
-00000d20: 2053 7065 6369 6573 2077 6974 6869 6e20   Species within 
-00000d30: 6120 7369 6d75 6c61 7469 6f6e 2061 7265  a simulation are
-00000d40: 2073 746f 7265 6420 6173 2073 6572 6961   stored as seria
-00000d50: 6c20 696e 7465 6765 722d 6b65 7965 6420  l integer-keyed 
-00000d60: 7661 6c75 6573 0a20 2020 2077 6974 6869  values.    withi
-00000d70: 6e20 7468 6520 436f 6d6d 756e 6974 7920  n the Community 
-00000d80: 6469 6374 2028 7768 6963 6820 696e 2074  dict (which in t
-00000d90: 7572 6e20 6973 2073 746f 7265 6420 6174  urn is stored at
-00000da0: 2074 6865 204d 6f64 656c 2773 0a20 2020   the Model's.   
-00000db0: 2027 6d6f 642e 636f 6d6d 2720 6174 7472   'mod.comm' attr
-00000dc0: 6962 7574 6529 2e0a 0a20 2020 2041 7474  ibute)...    Att
-00000dd0: 7269 6275 7465 730a 2020 2020 2d2d 2d2d  ributes.    ----
-00000de0: 2d2d 2d2d 2d2d 0a0a 2020 2020 2020 2020  ------..        
-00000df0: 4e4f 5445 3a20 466f 7220 6d6f 7265 2064  NOTE: For more d
-00000e00: 6574 6169 6c2c 2073 6565 2074 6865 2064  etail, see the d
-00000e10: 6f63 756d 656e 7461 7469 6f6e 2066 6f72  ocumentation for
-00000e20: 2074 6865 2070 6172 616d 6574 6572 7320   the parameters 
-00000e30: 7468 6174 0a20 2020 2020 2020 2020 2020  that.           
-00000e40: 2020 2063 6f72 7265 7370 6f6e 6420 746f     correspond to
-00000e50: 206d 616e 7920 6f66 2074 6865 2066 6f6c   many of the fol
-00000e60: 6c6f 7769 6e67 2061 7474 7269 6275 7465  lowing attribute
-00000e70: 732e 0a0a 2020 2020 2020 2020 623a 0a20  s...        b:. 
-00000e80: 2020 2020 2020 2020 2020 2054 6865 2053             The S
-00000e90: 7065 6369 6573 2720 696e 7472 696e 7369  pecies' intrinsi
-00000ea0: 6320 6269 7274 6820 7261 7465 2c20 6578  c birth rate, ex
-00000eb0: 7072 6573 7365 6420 6173 2074 6865 2070  pressed as the p
-00000ec0: 726f 6261 6269 6c69 7479 2028 696e 0a20  robability (in. 
-00000ed0: 2020 2020 2020 2020 2020 2061 2042 6572             a Ber
-00000ee0: 6e6f 756c 6c69 2072 616e 646f 6d20 6472  noulli random dr
-00000ef0: 6177 2920 7468 6174 2061 6e20 6f74 6865  aw) that an othe
-00000f00: 7277 6973 6520 656c 6967 6962 6c65 2070  rwise eligible p
-00000f10: 6f74 656e 7469 616c 0a20 2020 2020 2020  otential.       
-00000f20: 2020 2020 206d 6174 696e 6720 7061 6972       mating pair
-00000f30: 2077 696c 6c20 7375 6363 6573 7366 756c   will successful
-00000f40: 6c79 2072 6570 726f 6475 6365 0a0a 2020  ly reproduce..  
-00000f50: 2020 2020 2020 6275 726e 6564 3a0a 2020        burned:.  
-00000f60: 2020 2020 2020 2020 2020 4120 626f 6f6c            A bool
-00000f70: 2066 6c61 6720 696e 6469 6361 7469 6e67   flag indicating
-00000f80: 2077 6865 7468 6572 206f 7220 6e6f 7420   whether or not 
-00000f90: 7468 6520 5370 6563 6965 7320 6861 7320  the Species has 
-00000fa0: 6265 656e 0a20 2020 2020 2020 2020 2020  been.           
-00000fb0: 2062 7572 6e65 6420 696e 2e0a 0a20 2020   burned in...   
-00000fc0: 2020 2020 2063 6f6f 7264 733a 0a20 2020       coords:.   
-00000fd0: 2020 2020 2020 2020 2041 2032 6420 6e75           A 2d nu
-00000fe0: 6d70 7920 6172 7261 792c 206c 656e 6774  mpy array, lengt
-00000ff0: 6828 5370 6563 6965 7329 2078 2032 2c20  h(Species) x 2, 
-00001000: 636f 6d70 6f73 6564 206f 6620 636f 6c75  composed of colu
-00001010: 6d6e 730a 2020 2020 2020 2020 2020 2020  mns.            
-00001020: 636f 6e74 6169 6e69 6e67 2074 6865 2063  containing the c
-00001030: 7572 7265 6e74 2078 2061 6e64 2079 206c  urrent x and y l
-00001040: 6f63 6174 696f 6e73 206f 6620 616c 6c20  ocations of all 
-00001050: 496e 6469 7669 6475 616c 732e 0a20 2020  Individuals..   
-00001060: 2020 2020 2020 2020 2052 6f77 206f 7264           Row ord
-00001070: 6572 2069 7320 6571 7569 7661 6c65 6e74  er is equivalent
-00001080: 2074 6f20 7468 6520 6f72 6465 7220 6f66   to the order of
-00001090: 2069 6e64 6976 6964 7561 6c73 2069 6e20   individuals in 
-000010a0: 7468 6520 5370 6563 6965 7327 0a20 2020  the Species'.   
-000010b0: 2020 2020 2020 2020 204f 7264 6572 6564           Ordered
-000010c0: 4469 6374 2e0a 0a20 2020 2020 2020 2063  Dict...        c
-000010d0: 656c 6c73 3a0a 2020 2020 2020 2020 2020  ells:.          
-000010e0: 2020 4120 3264 206e 756d 7079 2061 7272    A 2d numpy arr
-000010f0: 6179 2c20 6c65 6e67 7468 2853 7065 6369  ay, length(Speci
-00001100: 6573 2920 7820 322c 2063 6f6d 706f 7365  es) x 2, compose
-00001110: 6420 6f66 2063 6f6c 756d 6e73 0a20 2020  d of columns.   
-00001120: 2020 2020 2020 2020 2063 6f6e 7461 696e           contain
-00001130: 696e 6720 7468 6520 6375 7272 656e 7420  ing the current 
-00001140: 6a20 616e 6420 6920 6365 6c6c 206e 756d  j and i cell num
-00001150: 6265 7273 206f 6620 616c 6c20 496e 6469  bers of all Indi
-00001160: 7669 6475 616c 732e 0a20 2020 2020 2020  viduals..       
-00001170: 2020 2020 2052 6f77 206f 7264 6572 2069       Row order i
-00001180: 7320 6571 7569 7661 6c65 6e74 2074 6f20  s equivalent to 
-00001190: 7468 6520 6f72 6465 7220 6f66 2069 6e64  the order of ind
-000011a0: 6976 6964 7561 6c73 2069 6e20 7468 6520  ividuals in the 
-000011b0: 5370 6563 6965 7327 0a20 2020 2020 2020  Species'.       
-000011c0: 2020 2020 204f 7264 6572 6564 4469 6374       OrderedDict
-000011d0: 2e20 284e 6f74 6520 7468 6520 6f72 6465  . (Note the orde
-000011e0: 7220 6f66 2074 6865 2063 6f6c 756d 6e73  r of the columns
-000011f0: 2120 5468 6573 6520 6172 6520 6e6f 7420  ! These are not 
-00001200: 696e 2074 6865 0a20 2020 2020 2020 2020  in the.         
-00001210: 2020 2069 2c6a 206f 7264 6572 2074 6861     i,j order tha
-00001220: 7420 776f 756c 6420 6265 206e 6563 6573  t would be neces
-00001230: 7361 7279 2069 6e20 6f72 6465 7220 7573  sary in order us
-00001240: 6520 7468 656d 2074 6f20 7375 626a 6563  e them to subjec
-00001250: 7420 610a 2020 2020 2020 2020 2020 2020  t a.            
-00001260: 4c61 7965 722e 7261 7374 206f 626a 6563  Layer.rast objec
-00001270: 7421 290a 0a20 2020 2020 2020 2063 686f  t!)..        cho
-00001280: 6f73 655f 6e65 6172 6573 745f 6d61 7465  ose_nearest_mate
-00001290: 3a0a 2020 2020 2020 2020 2020 2020 4120  :.            A 
-000012a0: 626f 6f6c 2066 6c61 6720 696e 6469 6361  bool flag indica
-000012b0: 7469 6e67 2077 6865 7468 6572 206f 7220  ting whether or 
-000012c0: 6e6f 7420 496e 6469 7669 6475 616c 7320  not Individuals 
-000012d0: 7368 6f75 6c64 200a 2020 2020 2020 2020  should .        
-000012e0: 2020 2020 616c 7761 7973 2063 686f 6f73      always choos
-000012f0: 6520 7468 6569 7220 6e65 6172 6573 7420  e their nearest 
-00001300: 6e65 6967 6862 6f72 7320 6173 2074 6865  neighbors as the
-00001310: 6972 2070 6f74 656e 7469 616c 206d 6174  ir potential mat
-00001320: 6573 0a0a 2020 2020 2020 2020 645f 6d61  es..        d_ma
-00001330: 783a 0a20 2020 2020 2020 2020 2020 2054  x:.            T
-00001340: 6865 206d 6178 696d 756d 2070 726f 6261  he maximum proba
-00001350: 6269 6c69 7479 206f 6620 6465 6174 6820  bility of death 
-00001360: 7468 6174 2063 616e 2062 6520 6173 7369  that can be assi
-00001370: 676e 6564 2074 6f20 616e 790a 2020 2020  gned to any.    
-00001380: 2020 2020 2020 2020 496e 6469 7669 6475          Individu
-00001390: 616c 0a0a 2020 2020 2020 2020 645f 6d69  al..        d_mi
-000013a0: 6e3a 0a20 2020 2020 2020 2020 2020 2054  n:.            T
-000013b0: 6865 206d 696e 696d 756d 2070 726f 6261  he minimum proba
-000013c0: 6269 6c69 7479 206f 6620 6465 6174 6820  bility of death 
-000013d0: 7468 6174 2063 616e 2062 6520 6173 7369  that can be assi
-000013e0: 676e 6564 2074 6f20 616e 790a 2020 2020  gned to any.    
-000013f0: 2020 2020 2020 2020 496e 6469 7669 6475          Individu
-00001400: 616c 0a0a 2020 2020 2020 2020 6465 6e73  al..        dens
-00001410: 6974 795f 6772 6964 5f77 696e 646f 775f  ity_grid_window_
-00001420: 7769 6474 683a 0a20 2020 2020 2020 2020  width:.         
-00001430: 2020 2054 6865 2077 6964 7468 206f 6620     The width of 
-00001440: 7468 6520 736c 6964 696e 6720 7769 6e64  the sliding wind
-00001450: 6f77 2074 6861 7420 6973 2075 7365 6420  ow that is used 
-00001460: 6279 2074 6865 2053 7065 6369 6573 270a  by the Species'.
-00001470: 2020 2020 2020 2020 2020 2020 5f44 656e              _Den
-00001480: 7369 7479 4772 6964 5374 6163 6b20 6f62  sityGridStack ob
-00001490: 6a65 6374 2077 6865 6e20 6573 7469 6d61  ject when estima
-000014a0: 7469 6e67 2074 6865 2053 7065 6369 6573  ting the Species
-000014b0: 270a 2020 2020 2020 2020 2020 2020 706f  '.            po
-000014c0: 7075 6c61 7469 6f6e 2d64 656e 7369 7479  pulation-density
-000014d0: 2061 7272 6179 2e0a 0a20 2020 2020 2020   array...       
-000014e0: 2064 6972 6563 7469 6f6e 5f64 6973 7472   direction_distr
-000014f0: 5f5b 6d75 2f6b 6170 7061 5d3a 0a20 2020  _[mu/kappa]:.   
-00001500: 2020 2020 2020 2020 2054 6865 2024 5c6d           The $\m
-00001510: 7524 2061 6e64 2024 5c6b 6170 7061 2420  u$ and $\kappa$ 
-00001520: 7061 7261 6d65 7465 7273 206f 6620 7468  parameters of th
-00001530: 6520 766f 6e20 4d69 7365 7320 6469 7374  e von Mises dist
-00001540: 7269 6275 7469 6f6e 0a20 2020 2020 2020  ribution.       
-00001550: 2020 2020 2066 726f 6d20 7768 6963 6820       from which 
-00001560: 7468 6520 6469 7265 6374 696f 6e20 636f  the direction co
-00001570: 6d70 6f6e 656e 7473 206f 6620 496e 6469  mponents of Indi
-00001580: 7669 6475 616c 7327 206d 6f76 656d 656e  viduals' movemen
-00001590: 740a 2020 2020 2020 2020 2020 2020 7665  t.            ve
-000015a0: 6374 6f72 7320 7769 6c6c 2062 6520 6472  ctors will be dr
-000015b0: 6177 6e2e 2028 5468 6573 6520 7061 7261  awn. (These para
-000015c0: 6d65 7465 7273 2061 7265 206f 6e6c 7920  meters are only 
-000015d0: 7574 696c 697a 6564 2069 6620 7468 650a  utilized if the.
-000015e0: 2020 2020 2020 2020 2020 2020 5370 6563              Spec
-000015f0: 6965 7320 6973 2070 6172 616d 6574 6572  ies is parameter
-00001600: 697a 6564 2077 6974 686f 7574 2061 205f  ized without a _
-00001610: 4d6f 7665 6d65 6e74 5375 7266 6163 652e  MovementSurface.
-00001620: 290a 0a20 2020 2020 2020 2064 6973 7065  )..        dispe
-00001630: 7273 616c 5f64 6973 7461 6e63 655f 6469  rsal_distance_di
-00001640: 7374 723a 0a20 2020 2020 2020 2020 2020  str:.           
-00001650: 2041 2073 7472 696e 6720 696e 6469 6361   A string indica
-00001660: 7469 6e67 2077 6869 6368 2064 6973 7472  ting which distr
-00001670: 6962 7574 696f 6e20 746f 2075 7365 2061  ibution to use a
-00001680: 730a 2020 2020 2020 2020 2020 2020 7468  s.            th
-00001690: 6520 5370 6563 6965 7327 2064 6973 7065  e Species' dispe
-000016a0: 7273 616c 2d64 6973 7461 6e63 6520 6469  rsal-distance di
-000016b0: 7374 7269 6275 7469 6f6e 2028 276c 6f67  stribution ('log
-000016c0: 6e6f 726d 616c 272c 0a20 2020 2020 2020  normal',.       
-000016d0: 2020 2020 2027 7761 6c64 272c 206f 7220       'wald', or 
-000016e0: 276c 6576 7927 290a 0a20 2020 2020 2020  'levy')..       
-000016f0: 2064 6973 7065 7273 616c 5f64 6973 7461   dispersal_dista
-00001700: 6e63 655f 6469 7374 725f 7061 7261 6d5b  nce_distr_param[
-00001710: 312f 325d 3a0a 2020 2020 2020 2020 2020  1/2]:.          
-00001720: 2020 5468 6520 7661 6c75 6573 206f 6620    The values of 
-00001730: 7468 6520 6669 7273 7420 616e 6420 7365  the first and se
-00001740: 636f 6e64 2070 6172 616d 6574 6572 7320  cond parameters 
-00001750: 6f66 2074 6865 2053 7065 6369 6573 270a  of the Species'.
-00001760: 2020 2020 2020 2020 2020 2020 6469 7370              disp
-00001770: 6572 7361 6c2d 6469 7374 616e 6365 2064  ersal-distance d
-00001780: 6973 7472 6962 7574 696f 6e20 286e 616d  istribution (nam
-00001790: 6564 2061 6273 7472 6163 746c 7920 6c69  ed abstractly li
-000017a0: 6b65 2074 6869 730a 2020 2020 2020 2020  ke this.        
-000017b0: 2020 2020 6265 6361 7573 6520 7468 6579      because they
-000017c0: 2063 6f75 6c64 2070 6172 616d 6574 6572   could parameter
-000017d0: 697a 6520 6120 4c6f 676e 6f72 6d61 6c2c  ize a Lognormal,
-000017e0: 2057 616c 642c 0a20 2020 2020 2020 2020   Wald,.         
-000017f0: 2020 206f 7220 4c65 7679 2064 6973 7472     or Levy distr
-00001800: 6962 7574 696f 6e2c 2064 6570 656e 6469  ibution, dependi
-00001810: 6e67 206f 6e20 7468 6520 7661 6c75 6520  ng on the value 
-00001820: 6f66 0a20 2020 2020 2020 2020 2020 2053  of.            S
-00001830: 7065 6369 6573 2e64 6973 7065 7273 616c  pecies.dispersal
-00001840: 5f64 6973 7461 6e63 655f 6469 7374 7229  _distance_distr)
-00001850: 0a0a 2020 2020 2020 2020 6578 7469 6e63  ..        extinc
-00001860: 743a 0a20 2020 2020 2020 2020 2020 2041  t:.            A
-00001870: 2062 6f6f 6c20 666c 6167 2069 6e64 6963   bool flag indic
-00001880: 6174 696e 6720 7768 6574 6865 7220 6f72  ating whether or
-00001890: 206e 6f74 2074 6865 2053 7065 6369 6573   not the Species
-000018a0: 2068 6173 2067 6f6e 6520 6578 7469 6e63   has gone extinc
-000018b0: 742e 0a0a 2020 2020 2020 2020 6964 783a  t...        idx:
-000018c0: 0a20 2020 2020 2020 2020 2020 2049 6e64  .            Ind
-000018d0: 6578 206e 756d 6265 7220 6f66 2074 6865  ex number of the
-000018e0: 2053 7065 6369 6573 2028 692e 652e 2069   Species (i.e. i
-000018f0: 7473 206b 6579 2077 6974 6869 6e20 7468  ts key within th
-00001900: 6520 436f 6d6d 756e 6974 7920 6469 6374  e Community dict
-00001910: 290a 0a20 2020 2020 2020 2069 6e76 6572  )..        inver
-00001920: 7365 5f64 6973 745f 6d61 7469 6e67 3a0a  se_dist_mating:.
-00001930: 2020 2020 2020 2020 2020 2020 4120 626f              A bo
-00001940: 6f6c 2066 6c61 6720 696e 6469 6361 7469  ol flag indicati
-00001950: 6e67 2077 6865 7468 6572 206f 7220 6e6f  ng whether or no
-00001960: 7420 7468 6520 7072 6f62 6162 696c 6974  t the probabilit
-00001970: 6965 7320 7468 6174 2061 6e0a 2020 2020  ies that an.    
-00001980: 2020 2020 2020 2020 496e 6469 7669 6475          Individu
-00001990: 616c 2063 686f 6f73 6573 2061 6e79 206f  al chooses any o
-000019a0: 6620 6974 7320 6e65 6967 6862 6f72 7320  f its neighbors 
-000019b0: 2869 2e65 2e20 6f74 6865 7220 496e 6469  (i.e. other Indi
-000019c0: 7669 6475 616c 730a 2020 2020 2020 2020  viduals.        
-000019d0: 2020 2020 7769 7468 696e 2060 5370 6563      within `Spec
-000019e0: 6965 732e 6d61 7469 6e67 5f72 6164 6975  ies.mating_radiu
-000019f0: 7360 2064 6973 7461 6e63 6520 6f66 2074  s` distance of t
-00001a00: 6865 6d29 2061 7320 7468 6569 720a 2020  hem) as their.  
-00001a10: 2020 2020 2020 2020 2020 706f 7465 6e74            potent
-00001a20: 6961 6c20 6d61 7465 7320 7368 6f75 6c64  ial mates should
-00001a30: 2076 6172 7920 696e 7665 7273 656c 7920   vary inversely 
-00001a40: 7769 7468 2074 6865 2064 6973 7461 6e63  with the distanc
-00001a50: 6573 2062 6574 7765 656e 0a20 2020 2020  es between.     
-00001a60: 2020 2020 2020 2074 6865 2049 6e64 6976         the Indiv
-00001a70: 6964 7561 6c20 616e 6420 6561 6368 206f  idual and each o
-00001a80: 6620 7468 6f73 6520 6e65 6967 6862 6f72  f those neighbor
-00001a90: 732e 2049 6620 4661 6c73 652c 2061 6e79  s. If False, any
-00001aa0: 206e 6569 6768 626f 720a 2020 2020 2020   neighbor.      
-00001ab0: 2020 2020 2020 636f 756c 6420 6265 2063        could be c
-00001ac0: 686f 7365 6e20 7769 7468 2065 7175 616c  hosen with equal
-00001ad0: 2070 726f 6261 6269 6c69 7479 2e20 0a0a   probability. ..
-00001ae0: 2020 2020 2020 2020 6765 6e5f 6172 6368          gen_arch
-00001af0: 3a0a 2020 2020 2020 2020 2020 2020 5468  :.            Th
-00001b00: 6520 5370 6563 6965 7327 2067 656e 6f6d  e Species' genom
-00001b10: 6963 2061 7263 6869 7465 6374 7572 6520  ic architecture 
-00001b20: 2861 7320 6120 4765 6e6f 6d69 6341 7263  (as a GenomicArc
-00001b30: 6869 7465 6374 7572 6520 6f62 6a65 6374  hitecture object
-00001b40: 290a 0a20 2020 2020 2020 204b 3a0a 2020  )..        K:.  
-00001b50: 2020 2020 2020 2020 2020 5468 6520 5370            The Sp
-00001b60: 6563 6965 7327 2063 7572 7265 6e74 2063  ecies' current c
-00001b70: 6172 7279 696e 672d 6361 7061 6369 7479  arrying-capacity
-00001b80: 2072 6173 7465 722e 2054 6869 7320 6973   raster. This is
-00001b90: 2061 206e 756d 7079 2061 7272 6179 0a20   a numpy array. 
-00001ba0: 2020 2020 2020 2020 2020 206f 6620 7368             of sh
-00001bb0: 6170 6520 604c 616e 6473 6361 7065 2e64  ape `Landscape.d
-00001bc0: 696d 602c 2077 6974 6820 6561 6368 2076  im`, with each v
-00001bd0: 616c 7565 200a 2020 2020 2020 2020 2020  alue .          
-00001be0: 2020 696e 6469 6361 7469 6e67 2074 6865    indicating the
-00001bf0: 2063 6172 7279 696e 6720 6361 7061 6369   carrying capaci
-00001c00: 7479 206f 6620 7468 6174 2063 656c 6c20  ty of that cell 
-00001c10: 696e 2074 6865 204c 616e 6473 6361 7065  in the Landscape
-00001c20: 2e0a 2020 2020 2020 2020 2020 2020 4974  ..            It
-00001c30: 2069 7320 6765 6e65 7261 7465 6420 6173   is generated as
-00001c40: 2074 6865 2070 726f 6475 6374 206f 6620   the product of 
-00001c50: 7468 6520 5370 6563 6965 7327 204b 5f66  the Species' K_f
-00001c60: 6163 746f 720a 2020 2020 2020 2020 2020  actor.          
-00001c70: 2020 616e 6420 7468 6520 7261 7374 6572    and the raster
-00001c80: 2061 7272 6179 206f 6620 7468 6520 4c61   array of the La
-00001c90: 7965 7220 7072 6f76 6964 6564 2061 7320  yer provided as 
-00001ca0: 7468 6520 6261 7369 730a 2020 2020 2020  the basis.      
-00001cb0: 2020 2020 2020 6f66 2074 6865 2053 7065        of the Spe
-00001cc0: 6369 6573 2720 6361 7272 7969 6e67 2063  cies' carrying c
-00001cd0: 6170 6163 6974 790a 2020 2020 2020 2020  apacity.        
-00001ce0: 2020 2020 2869 2e65 2e20 6073 7070 2e4b      (i.e. `spp.K
-00001cf0: 5f66 6163 746f 7220 2a20 6d6f 642e 6c61  _factor * mod.la
-00001d00: 6e64 5b73 7070 2e4b 5f6c 6179 6572 5d2e  nd[spp.K_layer].
-00001d10: 7261 7374 6029 2e0a 2020 2020 2020 2020  rast`)..        
-00001d20: 2020 2020 5468 6973 2061 7272 6179 2077      This array w
-00001d30: 696c 6c20 6368 616e 6765 2064 7572 696e  ill change durin
-00001d40: 6720 6120 7369 6d75 6c61 7469 6f6e 2069  g a simulation i
-00001d50: 6620 7468 6520 636f 7272 6573 706f 6e64  f the correspond
-00001d60: 696e 670a 2020 2020 2020 2020 2020 2020  ing.            
-00001d70: 4c61 7965 7220 6973 2070 6172 616d 6574  Layer is paramet
-00001d80: 6572 697a 6564 2074 6f20 756e 6465 7267  erized to underg
-00001d90: 6f20 656e 7669 726f 6e6d 656e 7461 6c20  o environmental 
-00001da0: 6368 616e 6765 2e0a 0a20 2020 2020 2020  change...       
-00001db0: 204b 5f66 6163 746f 723a 0a20 2020 2020   K_factor:.     
-00001dc0: 2020 2020 2020 2054 6865 2066 6163 746f         The facto
-00001dd0: 7220 6279 2077 6869 6368 2074 6f20 6d75  r by which to mu
-00001de0: 6c74 6970 6c79 2074 6865 2053 7065 6369  ltiply the Speci
-00001df0: 6573 2720 6368 6f73 656e 0a20 2020 2020  es' chosen.     
-00001e00: 2020 2020 2020 2063 6172 7279 696e 6720         carrying 
-00001e10: 6361 7061 6369 7479 204c 6179 6572 2028  capacity Layer (
-00001e20: 6d6f 642e 6c61 6e64 5b73 7070 2e4b 5f6c  mod.land[spp.K_l
-00001e30: 6179 6572 5d29 2069 6e20 6f72 6465 7220  ayer]) in order 
-00001e40: 746f 0a20 2020 2020 2020 2020 2020 2067  to.            g
-00001e50: 656e 6572 6174 6520 7468 6520 5370 6563  enerate the Spec
-00001e60: 6965 7327 2063 6172 7279 696e 672d 6361  ies' carrying-ca
-00001e70: 7061 6369 7479 2072 6173 7465 7220 2873  pacity raster (s
-00001e80: 7070 2e4b 292e 0a0a 2020 2020 2020 2020  pp.K)...        
-00001e90: 4b5f 6c61 7965 723a 0a20 2020 2020 2020  K_layer:.       
-00001ea0: 2020 2020 2054 6865 2069 6e64 6578 206e       The index n
-00001eb0: 756d 6265 7220 6f66 2074 6865 204c 6179  umber of the Lay
-00001ec0: 6572 2077 686f 7365 2072 6173 7465 7220  er whose raster 
-00001ed0: 7769 6c6c 2062 6520 6d75 6c74 6970 6c69  will be multipli
-00001ee0: 6564 2062 790a 2020 2020 2020 2020 2020  ed by.          
-00001ef0: 2020 7370 702e 4b5f 6661 6374 6f72 2069    spp.K_factor i
-00001f00: 6e20 6f72 6465 7220 746f 2067 656e 6572  n order to gener
-00001f10: 6174 6520 7468 6520 5370 6563 6965 7327  ate the Species'
-00001f20: 2063 6172 7279 696e 672d 6361 7061 6369   carrying-capaci
-00001f30: 7479 0a20 2020 2020 2020 2020 2020 2072  ty.            r
-00001f40: 6173 7465 7220 2873 7070 2e4b 292e 0a0a  aster (spp.K)...
-00001f50: 2020 2020 2020 2020 6d61 7469 6e67 5f72          mating_r
-00001f60: 6164 6975 733a 0a20 2020 2020 2020 2020  adius:.         
-00001f70: 2020 2054 6865 206d 6174 696e 6720 7261     The mating ra
-00001f80: 6469 7573 2c20 6578 7072 6573 7365 6420  dius, expressed 
-00001f90: 696e 204c 616e 6473 6361 7065 2063 656c  in Landscape cel
-00001fa0: 6c20 7769 6474 6873 2c20 6672 6f6d 2077  l widths, from w
-00001fb0: 6974 6869 6e0a 2020 2020 2020 2020 2020  ithin.          
-00001fc0: 2020 7768 6963 6820 616e 2049 6e64 6976    which an Indiv
-00001fd0: 6964 7561 6c20 6361 6e20 6368 6f6f 7365  idual can choose
-00001fe0: 2061 2070 6f74 656e 7469 616c 206d 6174   a potential mat
-00001ff0: 652e 2054 6869 7320 7769 6c6c 206f 6e6c  e. This will onl
-00002000: 7920 6265 0a20 2020 2020 2020 2020 2020  y be.           
-00002010: 2075 7365 6420 6966 2060 5370 6563 6965   used if `Specie
-00002020: 732e 6368 6f6f 7365 5f6e 6561 7265 7374  s.choose_nearest
-00002030: 5f6d 6174 6560 2069 7320 4661 6c73 652e  _mate` is False.
-00002040: 0a0a 2020 2020 2020 2020 6d61 785f 6167  ..        max_ag
-00002050: 653a 0a20 2020 2020 2020 2020 2020 2054  e:.            T
-00002060: 6865 206d 6178 696d 756d 2061 6765 2c20  he maximum age, 
-00002070: 696e 2074 696d 6520 7374 6570 732c 2074  in time steps, t
-00002080: 6861 7420 616e 7920 496e 6469 7669 6475  hat any Individu
-00002090: 616c 206f 6620 7468 6520 5370 6563 6965  al of the Specie
-000020a0: 730a 2020 2020 2020 2020 2020 2020 6d61  s.            ma
-000020b0: 7920 7265 6163 6820 2861 6674 6572 2077  y reach (after w
-000020c0: 6869 6368 2070 6f69 6e74 2074 6865 7920  hich point they 
-000020d0: 7769 6c6c 2062 6520 6375 6c6c 6564 2066  will be culled f
-000020e0: 726f 6d20 7468 650a 2020 2020 2020 2020  rom the.        
-000020f0: 2020 2020 706f 7075 6c61 7469 6f6e 2062      population b
-00002100: 7920 696d 706f 7365 6420 6d6f 7274 616c  y imposed mortal
-00002110: 6974 7929 0a0a 2020 2020 2020 2020 6d61  ity)..        ma
-00002120: 785f 696e 645f 6964 783a 0a20 2020 2020  x_ind_idx:.     
-00002130: 2020 2020 2020 2054 6865 206d 6178 696d         The maxim
-00002140: 756d 2049 6e64 6976 6964 7561 6c20 696e  um Individual in
-00002150: 6465 7820 6e75 6d62 6572 2074 6861 7420  dex number that 
-00002160: 6861 7320 6375 7272 656e 746c 7920 6265  has currently be
-00002170: 656e 2075 7365 640a 0a20 2020 2020 2020  en used..       
-00002180: 206d 6f76 653a 0a20 2020 2020 2020 2020   move:.         
-00002190: 2020 2041 2062 6f6f 6c20 666c 6167 2069     A bool flag i
-000021a0: 6e64 6963 6174 696e 6720 7768 6574 6865  ndicating whethe
-000021b0: 7220 6f72 206e 6f74 2049 6e64 6976 6964  r or not Individ
-000021c0: 7561 6c73 206f 6620 7468 6520 5370 6563  uals of the Spec
-000021d0: 6965 730a 2020 2020 2020 2020 2020 2020  ies.            
-000021e0: 6361 6e20 6d6f 7665 2061 6674 6572 2074  can move after t
-000021f0: 6865 6972 2069 6e69 7469 616c 2064 6973  heir initial dis
-00002200: 7065 7273 616c 2061 7320 6f66 6673 7072  persal as offspr
-00002210: 696e 672e 0a0a 2020 2020 2020 2020 6d6f  ing...        mo
-00002220: 7665 6d65 6e74 5f64 6973 7461 6e63 655f  vement_distance_
-00002230: 6469 7374 723a 0a20 2020 2020 2020 2020  distr:.         
-00002240: 2020 2041 2073 7472 696e 6720 696e 6469     A string indi
-00002250: 6361 7469 6e67 2077 6869 6368 2064 6973  cating which dis
-00002260: 7472 6962 7574 696f 6e20 746f 2075 7365  tribution to use
-00002270: 2061 730a 2020 2020 2020 2020 2020 2020   as.            
-00002280: 7468 6520 5370 6563 6965 7327 206d 6f76  the Species' mov
-00002290: 656d 656e 742d 6469 7374 616e 6365 2064  ement-distance d
-000022a0: 6973 7472 6962 7574 696f 6e20 2827 6c6f  istribution ('lo
-000022b0: 676e 6f72 6d61 6c27 2c0a 2020 2020 2020  gnormal',.      
-000022c0: 2020 2020 2020 2777 616c 6427 2c20 6f72        'wald', or
-000022d0: 2027 6c65 7679 2729 0a0a 2020 2020 2020   'levy')..      
-000022e0: 2020 6d6f 7665 6d65 6e74 5f64 6973 7461    movement_dista
-000022f0: 6e63 655f 6469 7374 725f 7061 7261 6d5b  nce_distr_param[
-00002300: 312f 325d 3a0a 2020 2020 2020 2020 2020  1/2]:.          
-00002310: 2020 5468 6520 7661 6c75 6573 206f 6620    The values of 
-00002320: 7468 6520 6669 7273 7420 616e 6420 7365  the first and se
-00002330: 636f 6e64 2070 6172 616d 6574 6572 7320  cond parameters 
-00002340: 6f66 2074 6865 2053 7065 6369 6573 270a  of the Species'.
-00002350: 2020 2020 2020 2020 2020 2020 6d6f 7665              move
-00002360: 6d65 6e74 2d64 6973 7461 6e63 6520 6469  ment-distance di
-00002370: 7374 7269 6275 7469 6f6e 2028 6e61 6d65  stribution (name
-00002380: 6420 6162 7374 7261 6374 6c79 206c 696b  d abstractly lik
-00002390: 6520 7468 6973 0a20 2020 2020 2020 2020  e this.         
-000023a0: 2020 2062 6563 6175 7365 2074 6865 7920     because they 
-000023b0: 636f 756c 6420 7061 7261 6d65 7465 7269  could parameteri
-000023c0: 7a65 2061 204c 6f67 6e6f 726d 616c 2c20  ze a Lognormal, 
-000023d0: 5761 6c64 2c0a 2020 2020 2020 2020 2020  Wald,.          
-000023e0: 2020 6f72 204c 6576 7920 6469 7374 7269    or Levy distri
-000023f0: 6275 7469 6f6e 2c20 6465 7065 6e64 696e  bution, dependin
-00002400: 6720 6f6e 2074 6865 2076 616c 7565 206f  g on the value o
-00002410: 660a 2020 2020 2020 2020 2020 2020 5370  f.            Sp
-00002420: 6563 6965 732e 6d6f 7665 6d65 6e74 5f64  ecies.movement_d
-00002430: 6973 7461 6e63 655f 6469 7374 7229 0a0a  istance_distr)..
-00002440: 2020 2020 2020 2020 6d75 7461 7465 3a0a          mutate:.
-00002450: 2020 2020 2020 2020 2020 2020 4120 626f              A bo
-00002460: 6f6c 2066 6c61 6720 696e 6469 6361 7469  ol flag indicati
-00002470: 6e67 2077 6865 7468 6572 206f 7220 6e6f  ng whether or no
-00002480: 7420 7468 6520 5370 6563 6965 7327 2073  t the Species' s
-00002490: 696d 756c 6174 696f 6e0a 2020 2020 2020  imulation.      
-000024a0: 2020 2020 2020 696e 766f 6c76 6573 206d        involves m
-000024b0: 7574 6174 696f 6e2e 0a0a 2020 2020 2020  utation...      
-000024c0: 2020 6d75 745f 6c6f 673a 0a20 2020 2020    mut_log:.     
-000024d0: 2020 2020 2020 2041 2062 6f6f 6c20 666c         A bool fl
-000024e0: 6167 2069 6e64 6963 6174 696e 6720 7768  ag indicating wh
-000024f0: 6574 6865 7220 6f72 206e 6f74 2061 206d  ether or not a m
-00002500: 7574 6174 696f 6e20 6c6f 6720 7368 6f75  utation log shou
-00002510: 6c64 2062 650a 2020 2020 2020 2020 2020  ld be.          
-00002520: 2020 7772 6974 7465 6e20 666f 7220 7468    written for th
-00002530: 6520 5370 6563 6965 732e 0a0a 2020 2020  e Species...    
-00002540: 2020 2020 4e3a 0a20 2020 2020 2020 2020      N:.         
-00002550: 2020 2054 6865 2053 7065 6369 6573 2720     The Species' 
-00002560: 6375 7272 656e 7420 706f 7075 6c61 7469  current populati
-00002570: 6f6e 2d64 656e 7369 7479 2072 6173 7465  on-density raste
-00002580: 722e 0a20 2020 2020 2020 2020 2020 2054  r..            T
-00002590: 6869 7320 6973 2061 206e 756d 7079 2061  his is a numpy a
-000025a0: 7272 6179 206f 6620 7368 6170 6520 604c  rray of shape `L
-000025b0: 616e 6473 6361 7065 2e64 696d 602c 2077  andscape.dim`, w
-000025c0: 6974 6820 6561 6368 2076 616c 7565 200a  ith each value .
-000025d0: 2020 2020 2020 2020 2020 2020 696e 6469              indi
-000025e0: 6361 7469 6e67 2074 6865 2070 6f70 756c  cating the popul
-000025f0: 6174 696f 6e20 6465 6e73 6974 7920 6f66  ation density of
-00002600: 2074 6861 7420 6365 6c6c 2069 6e20 7468   that cell in th
-00002610: 6520 4c61 6e64 7363 6170 652c 0a20 2020  e Landscape,.   
-00002620: 2020 2020 2020 2020 2061 7320 6573 7469           as esti
-00002630: 6d61 7465 6420 6279 2074 6865 2053 7065  mated by the Spe
-00002640: 6369 6573 2720 5f44 656e 7369 7479 4772  cies' _DensityGr
-00002650: 6964 5374 6163 6b20 696e 7374 616e 6365  idStack instance
-00002660: 2e20 0a0a 2020 2020 2020 2020 4e74 3a0a  . ..        Nt:.
-00002670: 2020 2020 2020 2020 2020 2020 4120 6368              A ch
-00002680: 726f 6e6f 6c6f 6769 6361 6c6c 7920 6f72  ronologically or
-00002690: 6465 7265 6420 6c69 7374 2c20 7374 6172  dered list, star
-000026a0: 7469 6e67 2066 726f 6d20 7468 6520 6669  ting from the fi
-000026b0: 7273 7420 7469 6d65 7374 6570 206f 660a  rst timestep of.
-000026c0: 2020 2020 2020 2020 2020 2020 7468 6520              the 
-000026d0: 6275 726e 2d69 6e2c 2063 6f6e 7461 696e  burn-in, contain
-000026e0: 696e 6720 7468 6520 5370 6563 6965 7327  ing the Species'
-000026f0: 2074 6f74 616c 2070 6f70 756c 6174 696f   total populatio
-00002700: 6e20 7369 7a65 2061 7420 6561 6368 0a20  n size at each. 
-00002710: 2020 2020 2020 2020 2020 2074 696d 6520             time 
-00002720: 7374 6570 2e0a 0a20 2020 2020 2020 206e  step...        n
-00002730: 5f62 6972 7468 733a 0a20 2020 2020 2020  _births:.       
-00002740: 2020 2020 2041 2063 6872 6f6e 6f6c 6f67       A chronolog
-00002750: 6963 616c 6c79 206f 7264 6572 6564 206c  ically ordered l
-00002760: 6973 742c 2073 7461 7274 696e 6720 6672  ist, starting fr
-00002770: 6f6d 2074 6865 2066 6972 7374 2074 696d  om the first tim
-00002780: 6573 7465 7020 6f66 0a20 2020 2020 2020  estep of.       
-00002790: 2020 2020 2074 6865 2062 7572 6e2d 696e       the burn-in
-000027a0: 2c20 636f 6e74 6169 6e69 6e67 2074 6865  , containing the
-000027b0: 2053 7065 6369 6573 2720 6e75 6d62 6572   Species' number
-000027c0: 206f 6620 6269 7274 6873 2061 7420 6561   of births at ea
-000027d0: 6368 0a20 2020 2020 2020 2020 2020 2074  ch.            t
-000027e0: 696d 6520 7374 6570 2e0a 0a20 2020 2020  ime step...     
-000027f0: 2020 206e 5f62 6972 7468 735f 6469 7374     n_births_dist
-00002800: 725f 6c61 6d62 6461 3a0a 2020 2020 2020  r_lambda:.      
-00002810: 2020 2020 2020 4966 2060 5370 6563 6965        If `Specie
-00002820: 732e 6e5f 6269 7274 6873 5f66 6978 6564  s.n_births_fixed
-00002830: 6020 6973 2046 616c 7365 2c20 7468 656e  ` is False, then
-00002840: 2074 6869 7320 7661 6c75 6520 7365 7276   this value serv
-00002850: 6573 2061 7320 7468 6520 0a20 2020 2020  es as the .     
-00002860: 2020 2020 2020 2024 5c6c 616d 6264 6124         $\lambda$
-00002870: 2070 6172 616d 6574 6572 206f 6620 7468   parameter of th
-00002880: 6520 506f 6973 736f 6e20 6469 7374 7269  e Poisson distri
-00002890: 6275 7469 6f6e 2066 726f 6d20 7768 6963  bution from whic
-000028a0: 6820 6973 2064 7261 776e 0a20 2020 2020  h is drawn.     
-000028b0: 2020 2020 2020 2074 6865 206e 756d 6265         the numbe
-000028c0: 7220 6f66 206f 6666 7370 7269 6e67 2061  r of offspring a
-000028d0: 206d 6174 696e 6720 7061 6972 2077 696c   mating pair wil
-000028e0: 6c20 7072 6f64 7563 652e 0a20 2020 2020  l produce..     
-000028f0: 2020 2020 2020 2049 6620 6053 7065 6369         If `Speci
-00002900: 6573 2e6e 5f62 6972 7468 735f 6669 7865  es.n_births_fixe
-00002910: 6460 2069 7320 5472 7565 2c20 7468 656e  d` is True, then
-00002920: 2074 6869 7320 6973 2074 6865 2066 6978   this is the fix
-00002930: 6564 206e 756d 6265 720a 2020 2020 2020  ed number.      
-00002940: 2020 2020 2020 6f66 206f 6666 7370 7269        of offspri
-00002950: 6e67 2074 6861 7420 6561 6368 2073 7563  ng that each suc
-00002960: 6365 7373 6675 6c20 6d61 7469 6e67 2070  cessful mating p
-00002970: 6169 7220 7769 6c6c 2070 726f 6475 6365  air will produce
-00002980: 2e0a 0a20 2020 2020 2020 206e 5f62 6972  ...        n_bir
-00002990: 7468 735f 6669 7865 643a 0a20 2020 2020  ths_fixed:.     
-000029a0: 2020 2020 2020 2041 2062 6f6f 6c20 666c         A bool fl
-000029b0: 6167 2069 6e64 6963 6174 696e 6720 7768  ag indicating wh
-000029c0: 6574 6865 7220 6f72 206e 6f74 2074 6865  ether or not the
-000029d0: 206e 756d 6265 7220 6f66 2062 6972 7468   number of birth
-000029e0: 7320 7368 6f75 6c64 0a20 2020 2020 2020  s should.       
-000029f0: 2020 2020 2062 6520 6669 7865 6420 6174       be fixed at
-00002a00: 2060 5370 6563 6965 732e 6e5f 6269 7274   `Species.n_birt
-00002a10: 6873 5f64 6973 7472 5f6c 616d 6264 6160  hs_distr_lambda`
-00002a20: 0a0a 2020 2020 2020 2020 6e5f 6465 6174  ..        n_deat
-00002a30: 6873 3a0a 2020 2020 2020 2020 2020 2020  hs:.            
-00002a40: 4120 6368 726f 6e6f 6c6f 6769 6361 6c6c  A chronologicall
-00002a50: 7920 6f72 6465 7265 6420 6c69 7374 2c20  y ordered list, 
-00002a60: 7374 6172 7469 6e67 2066 726f 6d20 7468  starting from th
-00002a70: 6520 6669 7273 7420 7469 6d65 7374 6570  e first timestep
-00002a80: 206f 660a 2020 2020 2020 2020 2020 2020   of.            
-00002a90: 7468 6520 6275 726e 2d69 6e2c 2063 6f6e  the burn-in, con
-00002aa0: 7461 696e 696e 6720 7468 6520 5370 6563  taining the Spec
-00002ab0: 6965 7327 206e 756d 6265 7220 6f66 2064  ies' number of d
-00002ac0: 6561 7468 7320 6174 2065 6163 680a 2020  eaths at each.  
-00002ad0: 2020 2020 2020 2020 2020 7469 6d65 2073            time s
-00002ae0: 7465 702e 0a0a 2020 2020 2020 2020 6e61  tep...        na
-00002af0: 6d65 3a0a 2020 2020 2020 2020 2020 2020  me:.            
-00002b00: 5468 6520 7374 7269 6e67 206e 616d 6520  The string name 
-00002b10: 6f66 2074 6865 2053 7065 6369 6573 0a0a  of the Species..
-00002b20: 2020 2020 2020 2020 523a 0a20 2020 2020          R:.     
-00002b30: 2020 2020 2020 2054 6865 2069 6e74 7269         The intri
-00002b40: 6e73 6963 2067 726f 7774 6820 7261 7465  nsic growth rate
-00002b50: 206f 6620 7468 6520 5370 6563 6965 732c   of the Species,
-00002b60: 2074 6f20 6265 2075 7365 6420 6173 2027   to be used as '
-00002b70: 5227 2069 6e20 7468 650a 2020 2020 2020  R' in the.      
-00002b80: 2020 2020 2020 7374 6f63 6861 7374 6963        stochastic
-00002b90: 2c20 7370 6174 6961 6c69 7a65 6420 6c6f  , spatialized lo
-00002ba0: 6769 7374 6963 2067 726f 7774 6820 6d6f  gistic growth mo
-00002bb0: 6465 6c20 7468 6174 2063 6f6e 7472 6f6c  del that control
-00002bc0: 730a 2020 2020 2020 2020 2020 2020 706f  s.            po
-00002bd0: 7075 6c61 7469 6f6e 2064 796e 616d 6963  pulation dynamic
-00002be0: 730a 0a20 2020 2020 2020 2072 6570 726f  s..        repro
-00002bf0: 5f61 6765 3a0a 2020 2020 2020 2020 2020  _age:.          
-00002c00: 2020 5468 6520 6167 652c 2069 6e20 7469    The age, in ti
-00002c10: 6d65 2073 7465 7073 2c20 6f66 2066 6972  me steps, of fir
-00002c20: 7374 2072 6570 726f 6475 6374 696f 6e0a  st reproduction.
-00002c30: 0a20 2020 2020 2020 2073 656c 6563 7469  .        selecti
-00002c40: 6f6e 3a0a 2020 2020 2020 2020 2020 2020  on:.            
-00002c50: 4120 626f 6f6c 2066 6c61 6720 696e 6469  A bool flag indi
-00002c60: 6361 7469 6e67 2077 6865 7468 6572 206f  cating whether o
-00002c70: 7220 6e6f 7420 7468 6520 5370 6563 6965  r not the Specie
-00002c80: 7327 2073 696d 756c 6174 696f 6e0a 2020  s' simulation.  
-00002c90: 2020 2020 2020 2020 2020 696e 766f 6c76            involv
-00002ca0: 6573 2073 656c 6563 7469 6f6e 2e0a 0a20  es selection... 
-00002cb0: 2020 2020 2020 2073 6578 3a0a 2020 2020         sex:.    
-00002cc0: 2020 2020 2020 2020 4120 626f 6f6c 2066          A bool f
-00002cd0: 6c61 6720 696e 6469 6361 7469 6e67 2077  lag indicating w
-00002ce0: 6865 7468 6572 206f 7220 6e6f 7420 7468  hether or not th
-00002cf0: 6973 2053 7065 6369 6573 2069 7320 7365  is Species is se
-00002d00: 7865 640a 2020 2020 2020 2020 2020 2020  xed.            
-00002d10: 2869 2e65 2e20 7768 6574 6865 7220 6f72  (i.e. whether or
-00002d20: 206e 6f74 2074 6865 2065 6c69 6769 6269   not the eligibi
-00002d30: 6c69 7479 206f 6620 706f 7465 6e74 6961  lity of potentia
-00002d40: 6c20 6d61 7469 6e67 2070 6169 7273 0a20  l mating pairs. 
-00002d50: 2020 2020 2020 2020 2020 2073 686f 756c             shoul
-00002d60: 6420 6265 2062 6173 6420 6f6e 2077 6865  d be basd on whe
-00002d70: 7468 6572 206f 7220 6e6f 7420 7468 6520  ther or not the 
-00002d80: 496e 6469 7669 6475 616c 7320 696e 2074  Individuals in t
-00002d90: 6865 2070 6f74 656e 7469 616c 0a20 2020  he potential.   
-00002da0: 2020 2020 2020 2020 2070 6169 7220 6172           pair ar
-00002db0: 6520 6f66 206f 7070 6f73 6974 6520 7365  e of opposite se
-00002dc0: 7865 7329 0a0a 2020 2020 2020 2020 7365  xes)..        se
-00002dd0: 785f 7261 7469 6f3a 0a20 2020 2020 2020  x_ratio:.       
-00002de0: 2020 2020 2054 6865 2073 6578 2072 6174       The sex rat
-00002df0: 696f 206f 6620 7468 6520 5370 6563 6965  io of the Specie
-00002e00: 732e 2054 6869 7320 7661 6c75 6520 6973  s. This value is
-00002e10: 2065 7870 7265 7373 6564 2061 730a 2020   expressed as.  
-00002e20: 2020 2020 2020 2020 2020 7468 6520 7072            the pr
-00002e30: 6f70 6f72 7469 6f6e 206f 6620 616c 6c20  oportion of all 
-00002e40: 6f66 6673 7072 696e 6720 7468 6174 2061  offspring that a
-00002e50: 7265 206d 616c 6573 2c20 7375 6368 2074  re males, such t
-00002e60: 6861 7420 6974 2063 616e 2062 650a 2020  hat it can be.  
-00002e70: 2020 2020 2020 2020 2020 6561 7369 6c79            easily
-00002e80: 2075 7365 6420 6173 2074 6865 2070 726f   used as the pro
-00002e90: 6261 6269 6c69 7479 2070 6172 616d 6574  bability paramet
-00002ea0: 6572 2066 6f72 2074 6865 2042 6572 6e6f  er for the Berno
-00002eb0: 756c 6c69 2064 7261 770a 2020 2020 2020  ulli draw.      
-00002ec0: 2020 2020 2020 6f66 2061 6e20 6f66 6673        of an offs
-00002ed0: 7072 696e 6727 7320 7365 782e 2028 4e6f  pring's sex. (No
-00002ee0: 7465 2074 6861 7420 7468 6973 2076 616c  te that this val
-00002ef0: 7565 2069 7320 6465 7269 7665 6420 6672  ue is derived fr
-00002f00: 6f6d 0a20 2020 2020 2020 2020 2020 2074  om.            t
-00002f10: 6865 2027 7365 785f 7261 7469 6f27 2070  he 'sex_ratio' p
-00002f20: 6172 616d 6574 6572 2070 726f 7669 6465  arameter provide
-00002f30: 6420 696e 2074 6865 204d 6f64 656c 2773  d in the Model's
-00002f40: 2070 6172 616d 6574 6572 7320 6669 6c65   parameters file
-00002f50: 2c20 0a20 2020 2020 2020 2020 2020 2062  , .            b
-00002f60: 7574 2075 6e6c 696b 6520 696e 2074 6865  ut unlike in the
-00002f70: 2070 6172 616d 6574 6572 7320 6669 6c65   parameters file
-00002f80: 2074 6869 7320 7661 6c75 6520 6973 0a20   this value is. 
-00002f90: 2020 2020 2020 2020 2020 206e 6f74 2065             not e
-00002fa0: 7870 7265 7373 6564 2069 6e20 6173 2061  xpressed in as a
-00002fb0: 2072 6174 696f 206f 6620 6d61 6c65 7320   ratio of males 
-00002fc0: 746f 2066 656d 616c 6573 2e29 0a0a 2020  to females.)..  
-00002fd0: 2020 2020 2020 7374 6172 745f 4e3a 0a20        start_N:. 
-00002fe0: 2020 2020 2020 2020 2020 2054 6865 206e             The n
-00002ff0: 756d 6265 7220 6f66 2049 6e64 6976 6964  umber of Individ
-00003000: 7561 6c73 2074 6f20 6265 2067 656e 6572  uals to be gener
-00003010: 6174 6564 2066 6f72 2074 6865 2073 7461  ated for the sta
-00003020: 7274 696e 670a 2020 2020 2020 2020 2020  rting.          
-00003030: 2020 706f 7075 6c61 7469 6f6e 206f 6620    population of 
-00003040: 7468 6520 5370 6563 6965 7320 2869 2e65  the Species (i.e
-00003050: 2e20 7468 6520 706f 7075 6c61 7469 6f6e  . the population
-00003060: 2073 697a 6520 6174 2074 6865 0a20 2020   size at the.   
-00003070: 2020 2020 2020 2020 2062 6567 696e 6e69           beginni
-00003080: 6e67 206f 6620 7468 6520 6275 726e 2d69  ng of the burn-i
-00003090: 6e29 0a0a 2020 2020 2020 2020 743a 0a20  n)..        t:. 
-000030a0: 2020 2020 2020 2020 2020 2054 6865 2069             The i
-000030b0: 6e64 6578 206f 6620 7468 6520 6c61 7374  ndex of the last
-000030c0: 2063 6f6d 706c 6574 6564 2074 696d 6520   completed time 
-000030d0: 7374 6570 2072 756e 2066 6f72 2074 6869  step run for thi
-000030e0: 7320 5370 6563 6965 732e 0a20 2020 2020  s Species..     
-000030f0: 2020 2020 2020 2054 6869 7320 6174 7472         This attr
-00003100: 6962 7574 6520 686f 6c64 7320 7468 6520  ibute holds the 
-00003110: 7661 6c75 6520 2d31 2075 6e74 696c 2074  value -1 until t
-00003120: 6865 2053 7065 6369 6573 2068 6173 2062  he Species has b
-00003130: 6565 6e20 6275 726e 6564 0a20 2020 2020  een burned.     
-00003140: 2020 2020 2020 2069 6e2c 2061 7420 7768         in, at wh
-00003150: 6963 6820 706f 696e 7420 6974 2062 6567  ich point it beg
-00003160: 696e 7320 696e 6372 656d 656e 7469 6e67  ins incrementing
-00003170: 2c20 5079 7468 6f6e 6963 616c 6c79 2c20  , Pythonically, 
-00003180: 6672 6f6d 2030 0a20 2020 2020 2020 2020  from 0.         
-00003190: 2020 2028 7375 6368 2074 6861 7420 6120     (such that a 
-000031a0: 5370 6563 6965 7320 7769 7468 2074 203d  Species with t =
-000031b0: 3d20 3939 3920 6861 7320 6265 656e 2072  = 999 has been r
-000031c0: 756e 2066 6f72 2031 3030 3020 7469 6d65  un for 1000 time
-000031d0: 0a20 2020 2020 2020 2020 2020 2073 7465  .            ste
-000031e0: 7073 292e 0a0a 2020 2020 2222 220a 2020  ps)...    """.  
-000031f0: 2020 2323 2323 2323 2323 2323 2323 2323    ##############
-00003200: 2323 2323 2323 2323 230a 2020 2020 2323  #########.    ##
-00003210: 2320 5350 4543 4941 4c20 4d45 5448 4f44  # SPECIAL METHOD
-00003220: 5320 2323 230a 2020 2020 2323 2323 2323  S ###.    ######
-00003230: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-00003240: 230a 0a20 2020 2064 6566 205f 5f69 6e69  #..    def __ini
-00003250: 745f 5f28 7365 6c66 2c20 6e61 6d65 2c20  t__(self, name, 
-00003260: 6964 782c 2069 6e64 732c 206c 616e 642c  idx, inds, land,
-00003270: 2073 7070 5f70 6172 616d 732c 0a20 2020   spp_params,.   
-00003280: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003290: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000032a0: 2020 2020 2020 2020 2020 2020 2067 656e               gen
-000032b0: 6f6d 6963 5f61 7263 6869 7465 6374 7572  omic_architectur
-000032c0: 653d 4e6f 6e65 293a 0a0a 2020 2020 2020  e=None):..      
-000032d0: 2020 2363 6865 636b 2074 6865 2069 6e64    #check the ind
-000032e0: 7320 6f62 6a65 6374 2069 7320 636f 7272  s object is corr
-000032f0: 6563 740a 2020 2020 2020 2020 6173 7365  ect.        asse
-00003300: 7274 2074 7970 6528 696e 6473 2920 696e  rt type(inds) in
-00003310: 2028 4f44 2c20 6469 6374 292c 2028 2241   (OD, dict), ("A
-00003320: 7267 756d 656e 7420 696e 6473 206d 7573  rgument inds mus
-00003330: 7420 6265 206f 6620 220a 2020 2020 2020  t be of ".      
-00003340: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003350: 2020 2020 2020 2020 2020 2274 7970 6520            "type 
-00003360: 6469 6374 206f 7220 7479 7065 2063 6f6c  dict or type col
-00003370: 6c65 6374 696f 6e73 2e4f 7264 6572 6564  lections.Ordered
-00003380: 5f44 6963 7422 290a 2020 2020 2020 2020  _Dict").        
-00003390: 6173 7365 7274 206c 6973 7428 7365 7428  assert list(set(
-000033a0: 5b69 2e5f 5f63 6c61 7373 5f5f 2e5f 5f6e  [i.__class__.__n
-000033b0: 616d 655f 5f20 666f 7220 6920 696e 206c  ame__ for i in l
-000033c0: 6973 7428 696e 6473 2e76 616c 7565 7328  ist(inds.values(
-000033d0: 0a20 2020 2020 2020 2020 2020 2029 295d  .            ))]
-000033e0: 2929 203d 3d20 5b20 2749 6e64 6976 6964  )) == [ 'Individ
-000033f0: 7561 6c27 5d2c 2028 2241 7267 756d 656e  ual'], ("Argumen
-00003400: 7420 696e 6473 206d 7573 7420 6265 2061  t inds must be a
-00003410: 2022 0a20 2020 2020 2020 2020 2020 2022   ".            "
-00003420: 6469 6374 696f 6e61 7279 2063 6f6e 7461  dictionary conta
-00003430: 696e 696e 6720 6f6e 6c79 2069 6e73 7461  ining only insta
-00003440: 6e63 6573 206f 6620 7468 6520 696e 6469  nces of the indi
-00003450: 7669 6475 616c 2e22 0a20 2020 2020 2020  vidual.".       
-00003460: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003470: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003480: 2020 2020 2020 2020 2020 2020 2022 496e               "In
-00003490: 6469 7669 6475 616c 2063 6c61 7373 2229  dividual class")
-000034a0: 0a0a 2020 2020 2020 2020 2361 7474 7269  ..        #attri
-000034b0: 6275 7465 2074 6f20 686f 6c64 2074 6865  bute to hold the
-000034c0: 2053 7065 6369 6573 2720 6964 7820 696e   Species' idx in
-000034d0: 2074 6865 2043 6f6d 6d75 6e69 7479 2064   the Community d
-000034e0: 6963 7469 6f6e 6172 790a 2020 2020 2020  ictionary.      
-000034f0: 2020 7365 6c66 2e69 6478 203d 2020 6964    self.idx =  id
-00003500: 780a 2020 2020 2020 2020 2320 7570 6461  x.        # upda
-00003510: 7465 2077 6974 6820 7468 6520 696e 7075  te with the inpu
-00003520: 7420 6469 6374 206f 6620 616c 6c20 696e  t dict of all in
-00003530: 6469 7669 6475 616c 732c 0a20 2020 2020  dividuals,.     
-00003540: 2020 2023 6173 2069 6e73 7461 6e63 6573     #as instances
-00003550: 206f 6620 696e 6469 7669 6475 616c 2e49   of individual.I
-00003560: 6e64 6976 6964 7561 6c20 636c 6173 730a  ndividual class.
-00003570: 2020 2020 2020 2020 7365 6c66 2e75 7064          self.upd
-00003580: 6174 6528 696e 6473 290a 0a20 2020 2020  ate(inds)..     
-00003590: 2020 2023 7365 7420 6f74 6865 7220 6174     #set other at
-000035a0: 7472 6962 7574 6573 0a20 2020 2020 2020  tributes.       
-000035b0: 2073 656c 662e 6e61 6d65 203d 2073 7472   self.name = str
-000035c0: 286e 616d 6529 0a20 2020 2020 2020 2073  (name).        s
-000035d0: 656c 662e 5f6c 616e 645f 6469 6d20 3d20  elf._land_dim = 
-000035e0: 6c61 6e64 2e64 696d 0a20 2020 2020 2020  land.dim.       
-000035f0: 2073 656c 662e 5f6c 616e 645f 7265 7320   self._land_res 
-00003600: 3d20 6c61 6e64 2e72 6573 0a20 2020 2020  = land.res.     
-00003610: 2020 2073 656c 662e 5f6c 616e 645f 7265     self._land_re
-00003620: 735f 7261 7469 6f20 3d20 6c61 6e64 2e5f  s_ratio = land._
-00003630: 7265 735f 7261 7469 6f0a 2020 2020 2020  res_ratio.      
-00003640: 2020 7365 6c66 2e5f 6c61 6e64 5f75 6c63    self._land_ulc
-00003650: 203d 206c 616e 642e 756c 630a 2020 2020   = land.ulc.    
-00003660: 2020 2020 7365 6c66 2e5f 6c61 6e64 5f70      self._land_p
-00003670: 726a 203d 206c 616e 642e 7072 6a0a 2020  rj = land.prj.  
-00003680: 2020 2020 2020 2361 7474 7269 6275 7465        #attribute
-00003690: 2074 6f20 6b65 6570 2074 7261 636b 206f   to keep track o
-000036a0: 6620 6974 6572 6174 696f 6e20 6e75 6d62  f iteration numb
-000036b0: 6572 2074 6869 7320 7370 7020 6973 2062  er this spp is b
-000036c0: 6569 6e67 2075 7365 6420 666f 720a 2020  eing used for.  
-000036d0: 2020 2020 2020 2328 6f70 7469 6f6e 616c        #(optional
-000036e0: 3b20 7769 6c6c 2062 6520 7365 7420 6279  ; will be set by
-000036f0: 2074 6865 2069 7465 7261 7469 6f6e 2e70   the iteration.p
-00003700: 7920 6d6f 6475 6c65 2069 6620 6361 6c6c  y module if call
-00003710: 6564 290a 2020 2020 2020 2020 7365 6c66  ed).        self
-00003720: 2e5f 6974 203d 204e 6f6e 650a 2020 2020  ._it = None.    
-00003730: 2020 2020 2361 7474 7269 6275 7465 2074      #attribute t
-00003740: 6f20 7472 6163 6b20 746f 7461 6c20 6e75  o track total nu
-00003750: 6d62 6572 206f 6620 7469 6d65 7374 6570  mber of timestep
-00003760: 7320 746f 2062 6520 7275 6e0a 2020 2020  s to be run.    
-00003770: 2020 2020 2328 7769 6c6c 2062 6520 7365      #(will be se
-00003780: 7420 6279 2074 6865 204d 6f64 656c 206f  t by the Model o
-00003790: 626a 6563 7429 0a20 2020 2020 2020 2023  bject).        #
-000037a0: 7365 6c66 2e54 203d 204e 6f6e 6520 0a20  self.T = None . 
-000037b0: 2020 2020 2020 2023 2061 7474 7269 6275         # attribu
-000037c0: 7465 2074 6f20 6b65 6570 206f 6620 7472  te to keep of tr
-000037d0: 6163 6b20 6f66 206e 756d 6265 7220 6f66  ack of number of
-000037e0: 2074 696d 6573 7465 7073 0a20 2020 2020   timesteps.     
-000037f0: 2020 2023 7468 6520 7370 6563 6965 7320     #the species 
-00003800: 6861 7320 6576 6f6c 7665 6420 2873 7461  has evolved (sta
-00003810: 7274 7320 6174 202d 312c 200a 2020 2020  rts at -1, .    
-00003820: 2020 2020 2374 6f20 696e 6469 6361 7465      #to indicate
-00003830: 2075 6e72 756e 2c20 616e 6420 736f 2074   unrun, and so t
-00003840: 6861 7420 6669 7273 7420 7469 6d65 7374  hat first timest
-00003850: 6570 2077 696c 6c0a 2020 2020 2020 2020  ep will.        
-00003860: 2362 6520 7365 7420 746f 2030 2061 7420  #be set to 0 at 
-00003870: 6265 6769 6e6e 696e 6720 6f66 2074 696d  beginning of tim
-00003880: 6573 7465 7029 0a20 2020 2020 2020 2023  estep).        #
-00003890: 204e 4f54 453a 2054 6869 7320 7761 792c   NOTE: This way,
-000038a0: 2069 6620 6d6f 6465 6c20 6973 2073 746f   if model is sto
-000038b0: 7070 6564 2c20 6368 616e 6765 7320 0a20  pped, changes . 
-000038c0: 2020 2020 2020 2023 6172 6520 6d61 6465         #are made
-000038d0: 2c20 7468 656e 2069 7420 6973 2072 756e  , then it is run
-000038e0: 2066 7572 7468 6572 2c0a 2020 2020 2020   further,.      
-000038f0: 2020 2320 7468 6973 206d 616b 6573 2069    # this makes i
-00003900: 7420 6561 7379 2074 6f20 636f 6e74 696e  t easy to contin
-00003910: 7565 2074 7261 636b 696e 6720 6672 6f6d  ue tracking from
-00003920: 2074 6865 2062 6567 696e 6e69 6e67 0a20   the beginning. 
-00003930: 2020 2020 2020 2073 656c 662e 7420 3d20         self.t = 
-00003940: 2d31 0a20 2020 2020 2020 2023 7769 6c6c  -1.        #will
-00003950: 2062 6520 7377 6974 6368 6564 2074 6f20   be switched to 
-00003960: 5472 7565 2077 6865 6e20 7468 6520 7370  True when the sp
-00003970: 6563 6965 7320 7061 7373 6573 2074 6865  ecies passes the
-00003980: 2062 7572 6e2d 696e 2074 6573 7473 0a20   burn-in tests. 
-00003990: 2020 2020 2020 2073 656c 662e 6275 726e         self.burn
-000039a0: 6564 203d 2046 616c 7365 0a20 2020 2020  ed = False.     
-000039b0: 2020 2023 7769 6c6c 2062 6520 7377 6974     #will be swit
-000039c0: 6368 6564 2074 6f20 5472 7565 2069 6620  ched to True if 
-000039d0: 7468 6520 7370 6563 6965 7320 676f 6573  the species goes
-000039e0: 2065 7874 696e 6374 0a20 2020 2020 2020   extinct.       
-000039f0: 2073 656c 662e 6578 7469 6e63 7420 3d20   self.extinct = 
-00003a00: 4661 6c73 650a 2020 2020 2020 2020 2373  False.        #s
-00003a10: 7461 7274 696e 6720 706f 7020 7369 7a65  tarting pop size
-00003a20: 0a20 2020 2020 2020 2073 656c 662e 7374  .        self.st
-00003a30: 6172 745f 4e20 3d20 6c65 6e28 7365 6c66  art_N = len(self
-00003a40: 290a 2020 2020 2020 2020 2363 7265 6174  ).        #creat
-00003a50: 6520 6120 7472 6163 6b65 7220 666f 7220  e a tracker for 
-00003a60: 7468 6520 6d61 7869 6d75 6d20 496e 6469  the maximum Indi
-00003a70: 7669 6475 616c 2069 6e64 6578 2061 6c72  vidual index alr
-00003a80: 6561 6479 2075 7365 640a 2020 2020 2020  eady used.      
-00003a90: 2020 2328 7468 6973 2077 696c 6c20 6265    #(this will be
-00003aa0: 2075 7064 6174 6564 2065 6163 6820 7469   updated each ti
-00003ab0: 6d65 206f 6666 7370 7269 6e67 2061 7265  me offspring are
-00003ac0: 2063 7265 6174 6564 2c20 736f 2074 6861   created, so tha
-00003ad0: 740a 2020 2020 2020 2020 2369 6e64 6963  t.        #indic
-00003ae0: 6573 2061 7265 206e 6576 6572 2072 6570  es are never rep
-00003af0: 6561 7465 642c 206f 7665 7772 7269 7469  eated, ovewrriti
-00003b00: 6e67 2065 7869 7374 696e 6720 496e 6469  ng existing Indi
-00003b10: 7669 6475 616c 7329 0a20 2020 2020 2020  viduals).       
-00003b20: 2073 656c 662e 6d61 785f 696e 645f 6964   self.max_ind_id
-00003b30: 7820 3d20 6d61 7828 5b2a 696e 6473 5d29  x = max([*inds])
-00003b40: 0a20 2020 2020 2020 2023 6174 7472 6962  .        #attrib
-00003b50: 7574 6520 746f 2068 6f6c 6420 6120 6c61  ute to hold a la
-00003b60: 6e64 7363 6170 652e 4c61 7965 7220 6f62  ndscape.Layer ob
-00003b70: 6a65 6374 206f 660a 2020 2020 2020 2020  ject of.        
-00003b80: 2374 6865 2063 7572 7265 6e74 2073 7065  #the current spe
-00003b90: 6369 6573 2064 656e 7369 7479 0a20 2020  cies density.   
-00003ba0: 2020 2020 2073 656c 662e 4e20 3d20 4e6f       self.N = No
-00003bb0: 6e65 0a20 2020 2020 2020 2023 2061 7474  ne.        # att
-00003bc0: 7269 6275 7465 2074 6f20 686f 6c64 2061  ribute to hold a
-00003bd0: 6e20 6c61 6e64 7363 6170 652e 4c61 7965  n landscape.Laye
-00003be0: 7220 6f62 6a65 6374 206f 660a 2020 2020  r object of.    
-00003bf0: 2020 2020 2374 6865 206c 6f63 616c 2063      #the local c
-00003c00: 6172 7279 696e 6720 6361 7061 6369 7479  arrying capacity
-00003c10: 2028 692e 652e 0a20 2020 2020 2020 2023   (i.e..        #
-00003c20: 2774 6172 6765 7427 2064 796e 616d 6963  'target' dynamic
-00003c30: 2065 7175 696c 6962 7269 756d 2073 7065   equilibrium spe
-00003c40: 6369 6573 2064 656e 7369 7479 290a 2020  cies density).  
-00003c50: 2020 2020 2020 7365 6c66 2e4b 203d 204e        self.K = N
-00003c60: 6f6e 650a 2020 2020 2020 2020 2374 6865  one.        #the
-00003c70: 2069 6e64 6578 206e 756d 6265 7220 6f66   index number of
-00003c80: 2074 6865 204c 6179 6572 2074 6f20 6265   the Layer to be
-00003c90: 2075 7365 6420 6173 2074 6869 7320 5370   used as this Sp
-00003ca0: 6563 6965 7327 204b 2d4c 6179 6572 0a20  ecies' K-Layer. 
-00003cb0: 2020 2020 2020 2073 656c 662e 4b5f 6c61         self.K_la
-00003cc0: 7965 7220 3d20 4e6f 6e65 0a20 2020 2020  yer = None.     
-00003cd0: 2020 2023 7468 6520 6661 6374 6f72 2062     #the factor b
-00003ce0: 7920 7768 6963 6820 7468 6973 2053 7065  y which this Spe
-00003cf0: 6369 6573 2720 4b2d 4c61 7965 7220 7368  cies' K-Layer sh
-00003d00: 6f75 6c64 2062 650a 2020 2020 2020 2020  ould be.        
-00003d10: 236d 756c 7469 706c 6965 6420 746f 2067  #multiplied to g
-00003d20: 6574 2069 7473 204b 2072 6173 7465 720a  et its K raster.
-00003d30: 2020 2020 2020 2020 7365 6c66 2e4b 5f66          self.K_f
-00003d40: 6163 746f 7220 3d20 4e6f 6e65 0a20 2020  actor = None.   
-00003d50: 2020 2020 2023 206c 6973 7420 746f 2072       # list to r
-00003d60: 6563 6f72 6420 7370 6563 6965 7320 7369  ecord species si
-00003d70: 7a65 2028 6170 7065 6e64 6564 2065 6163  ze (appended eac
-00003d80: 6820 7469 6d65 0a20 2020 2020 2020 2073  h time.        s
-00003d90: 656c 662e 4e74 203d 205b 5d0a 2020 2020  elf.Nt = [].    
-00003da0: 2020 2020 2320 7370 702e 696e 6372 656d      # spp.increm
-00003db0: 656e 745f 6167 655f 7374 6167 6528 2920  ent_age_stage() 
-00003dc0: 6973 2063 616c 6c65 6429 0a20 2020 2020  is called).     
-00003dd0: 2020 2023 2074 7261 636b 6572 206f 6620     # tracker of 
-00003de0: 6e75 6d62 6572 206f 6620 6269 7274 6873  number of births
-00003df0: 2065 6163 6820 7469 6d65 2073 7070 2e64   each time spp.d
-00003e00: 6f5f 6d61 7469 6e67 2069 7320 6361 6c6c  o_mating is call
-00003e10: 6564 0a20 2020 2020 2020 2073 656c 662e  ed.        self.
-00003e20: 6e5f 6269 7274 6873 203d 205b 5d0a 2020  n_births = [].  
-00003e30: 2020 2020 2020 2320 7472 6163 6b65 7220        # tracker 
-00003e40: 6f66 206e 756d 6265 7220 6f66 2064 6561  of number of dea
-00003e50: 7468 7320 6561 6368 2074 696d 650a 2020  ths each time.  
-00003e60: 2020 2020 2020 2364 656d 6f67 7261 7068        #demograph
-00003e70: 792e 7370 705f 6479 6e61 6d69 6373 2069  y.spp_dynamics i
-00003e80: 7320 6361 6c6c 6564 0a20 2020 2020 2020  s called.       
-00003e90: 2073 656c 662e 6e5f 6465 6174 6873 203d   self.n_deaths =
-00003ea0: 205b 5d0a 2020 2020 2020 2020 2361 7474   [].        #att
-00003eb0: 7269 6275 7465 7320 666f 7220 7374 6f72  ributes for stor
-00003ec0: 696e 6720 6e75 6d70 7920 6172 7261 7973  ing numpy arrays
-00003ed0: 206f 6620 616c 6c20 696e 6469 7669 6475   of all individu
-00003ee0: 616c 7327 0a20 2020 2020 2020 2023 636f  als'.        #co
-00003ef0: 6f72 6469 6e61 7465 7320 616e 6420 6365  ordinates and ce
-00003f00: 6c6c 732c 2074 6f20 6176 6f69 6420 7265  lls, to avoid re
-00003f10: 7065 6174 2063 6f6d 7075 7465 2074 696d  peat compute tim
-00003f20: 6520 6561 6368 2074 7572 6e0a 2020 2020  e each turn.    
-00003f30: 2020 2020 7365 6c66 2e63 6f6f 7264 7320      self.coords 
-00003f40: 3d20 4e6f 6e65 0a20 2020 2020 2020 2073  = None.        s
-00003f50: 656c 662e 6365 6c6c 7320 3d20 4e6f 6e65  elf.cells = None
-00003f60: 0a20 2020 2020 2020 2023 6372 6561 7465  .        #create
-00003f70: 2065 6d70 7479 2061 7474 7269 6275 7465   empty attribute
-00003f80: 7320 746f 2068 6f6c 6420 7370 6174 6961  s to hold spatia
-00003f90: 6c20 6f62 6a65 6374 7320 7468 6174 2077  l objects that w
-00003fa0: 696c 6c0a 2020 2020 2020 2020 2362 6520  ill.        #be 
-00003fb0: 6372 6561 7465 6420 6166 7465 7220 7468  created after th
-00003fc0: 6520 7370 6563 6965 7320 6973 2069 6e73  e species is ins
-00003fd0: 7461 6e74 6961 7465 640a 2020 2020 2020  tantiated.      
-00003fe0: 2020 7365 6c66 2e5f 6b64 5f74 7265 6520    self._kd_tree 
-00003ff0: 3d20 4e6f 6e65 0a20 2020 2020 2020 2023  = None.        #
-00004000: 6372 6561 7465 2065 6d70 7479 2061 7474  create empty att
-00004010: 7269 6275 7465 2074 6f20 686f 6c64 2074  ribute to hold t
-00004020: 6865 205f 4465 6e73 6974 7947 7269 6453  he _DensityGridS
-00004030: 7461 636b 0a20 2020 2020 2020 2073 656c  tack.        sel
-00004040: 662e 5f64 656e 735f 6772 6964 7320 3d20  f._dens_grids = 
-00004050: 4e6f 6e65 0a20 2020 2020 2020 2023 6372  None.        #cr
-00004060: 6561 7465 2061 6e20 6174 7472 6962 7574  eate an attribut
-00004070: 6520 746f 2069 6e64 6963 6174 6520 7768  e to indicate wh
-00004080: 6574 6865 7220 7468 6973 2073 7065 6369  ether this speci
-00004090: 6573 0a20 2020 2020 2020 2023 7769 6c6c  es.        #will
-000040a0: 2068 6176 6520 6d6f 7665 6d65 6e74 3b20   have movement; 
-000040b0: 7365 7420 746f 2046 616c 7365 206e 6f77  set to False now
-000040c0: 2062 7574 2077 696c 6c20 6265 2075 7064   but will be upd
-000040d0: 6174 6564 0a20 2020 2020 2020 2023 6265  ated.        #be
-000040e0: 6c6f 7720 6966 2061 2027 6d6f 7665 6d65  low if a 'moveme
-000040f0: 6e74 2720 7365 6374 696f 6e20 6973 2065  nt' section is e
-00004100: 6e63 6f75 6e74 6572 6564 2069 6e20 7468  ncountered in th
-00004110: 6520 7061 7261 6d73 0a20 2020 2020 2020  e params.       
-00004120: 2073 656c 662e 5f6d 6f76 6520 3d20 4661   self._move = Fa
-00004130: 6c73 650a 2020 2020 2020 2020 2363 7265  lse.        #cre
-00004140: 6174 6520 656d 7074 7920 6174 7472 6962  ate empty attrib
-00004150: 7574 6573 2066 6f72 2073 7061 7469 616c  utes for spatial
-00004160: 2e5f 436f 6e64 7563 7461 6e63 6553 7572  ._ConductanceSur
-00004170: 6661 6365 206f 626a 6563 7473 0a20 2020  face objects.   
-00004180: 2020 2020 2023 7468 6174 2063 6f75 6c64       #that could
-00004190: 2062 6520 7573 6564 2066 6f72 206d 6f76   be used for mov
-000041a0: 656d 656e 7420 616e 642f 6f72 2064 6973  ement and/or dis
-000041b0: 7065 7273 616c 0a20 2020 2020 2020 2023  persal.        #
-000041c0: 286d 6179 2062 6520 6372 6561 7465 642c  (may be created,
-000041d0: 2064 6570 656e 6469 6e67 206f 6e20 7061   depending on pa
-000041e0: 7261 6d74 6572 7329 0a20 2020 2020 2020  ramters).       
-000041f0: 2073 656c 662e 5f6d 6f76 655f 7375 7266   self._move_surf
-00004200: 203d 204e 6f6e 650a 2020 2020 2020 2020   = None.        
-00004210: 7365 6c66 2e5f 6469 7370 5f73 7572 6620  self._disp_surf 
-00004220: 3d20 4e6f 6e65 0a20 2020 2020 2020 2023  = None.        #
-00004230: 6372 6561 7465 2061 6e20 656d 7074 7920  create an empty 
-00004240: 6368 616e 6765 7220 6174 7472 6962 7574  changer attribut
-00004250: 652c 2077 6869 6368 2077 696c 6c0a 2020  e, which will.  
-00004260: 2020 2020 2020 2362 6520 7265 7365 7420        #be reset 
-00004270: 6966 2074 6865 2070 6172 616d 6574 6572  if the parameter
-00004280: 7320 6465 6669 6e65 2063 6861 6e67 6573  s define changes
-00004290: 2066 6f72 2074 6869 7320 7370 700a 2020   for this spp.  
-000042a0: 2020 2020 2020 7365 6c66 2e5f 6368 616e        self._chan
-000042b0: 6765 7220 3d20 4e6f 6e65 0a20 2020 2020  ger = None.     
-000042c0: 2020 2023 7365 7420 7468 6520 7365 785f     #set the sex_
-000042d0: 7261 7469 6f20 746f 2030 2e35 2064 6566  ratio to 0.5 def
-000042e0: 6175 6c74 2028 6275 7420 7468 6973 2077  ault (but this w
-000042f0: 696c 6c20 6265 2063 6861 6e67 6564 2069  ill be changed i
-00004300: 6620 610a 2020 2020 2020 2020 236e 6f6e  f a.        #non
-00004310: 2d31 2f31 2073 6578 2072 6174 696f 2069  -1/1 sex ratio i
-00004320: 7320 7072 6f76 6964 6564 2069 6e20 7468  s provided in th
-00004330: 6520 7061 7261 6d73 290a 2020 2020 2020  e params).      
-00004340: 2020 7365 6c66 2e73 6578 5f72 6174 696f    self.sex_ratio
-00004350: 203d 2030 2e35 0a20 2020 2020 2020 2023   = 0.5.        #
-00004360: 6372 6561 7465 2061 2070 7269 7661 7465  create a private
-00004370: 205f 5061 7261 6d73 5661 6c73 206f 626a   _ParamsVals obj
-00004380: 6563 7420 6173 2074 6865 205f 7076 2061  ect as the _pv a
-00004390: 7474 7269 6275 7465 0a20 2020 2020 2020  ttribute.       
-000043a0: 2073 656c 662e 5f70 7620 3d20 5f50 6172   self._pv = _Par
-000043b0: 616d 7356 616c 7328 7365 6c66 2e6e 616d  amsVals(self.nam
-000043c0: 6529 0a20 2020 2020 2020 2023 7468 656e  e).        #then
-000043d0: 2067 7261 6220 616c 6c20 6f66 2074 6865   grab all of the
-000043e0: 206d 6174 696e 672c 206d 6f72 7461 6c69   mating, mortali
-000043f0: 7479 2c20 616e 6420 6d6f 7665 6d65 6e74  ty, and movement
-00004400: 0a20 2020 2020 2020 2023 7061 7261 6d65  .        #parame
-00004410: 7465 7273 2061 7320 6174 7472 6962 7574  ters as attribut
-00004420: 6573 206f 6620 7468 6174 205f 5061 7261  es of that _Para
-00004430: 6d73 5661 6c73 206f 626a 6563 740a 2020  msVals object.  
-00004440: 2020 2020 2020 666f 7220 7365 6374 696f        for sectio
-00004450: 6e20 696e 205b 276d 6174 696e 6727 2c20  n in ['mating', 
-00004460: 276d 6f72 7461 6c69 7479 272c 2027 6d6f  'mortality', 'mo
-00004470: 7665 6d65 6e74 275d 3a0a 2020 2020 2020  vement']:.      
-00004480: 2020 2020 2020 6966 2073 6563 7469 6f6e        if section
-00004490: 2069 6e20 5b2a 7370 705f 7061 7261 6d73   in [*spp_params
-000044a0: 5d3a 0a20 2020 2020 2020 2020 2020 2020  ]:.             
-000044b0: 2020 2066 6f72 2061 7474 2c76 616c 2069     for att,val i
-000044c0: 6e20 7370 705f 7061 7261 6d73 5b73 6563  n spp_params[sec
-000044d0: 7469 6f6e 5d2e 6974 656d 7328 293a 0a20  tion].items():. 
-000044e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000044f0: 2020 2023 6c65 6176 6520 6f75 7420 7468     #leave out th
-00004500: 6520 6d6f 7665 5f73 7572 6620 616e 6420  e move_surf and 
-00004510: 6469 7370 5f73 7572 6620 636f 6d70 6f6e  disp_surf compon
-00004520: 656e 7473 2c0a 2020 2020 2020 2020 2020  ents,.          
-00004530: 2020 2020 2020 2020 2020 2377 6869 6368            #which
-00004540: 2077 696c 6c20 6265 2068 616e 646c 6564   will be handled
-00004550: 2073 6570 6172 6174 656c 790a 2020 2020   separately.    
-00004560: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004570: 6966 206e 6f74 2069 7369 6e73 7461 6e63  if not isinstanc
-00004580: 6528 7661 6c2c 2064 6963 7429 3a0a 2020  e(val, dict):.  
-00004590: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000045a0: 2020 2020 2020 2363 6f6e 7665 7274 2073        #convert s
-000045b0: 6578 2072 6174 696f 2074 6f20 7468 6520  ex ratio to the 
-000045c0: 7072 6f62 6162 696c 7479 206f 6620 6472  probabilty of dr
-000045d0: 6177 696e 6720 6120 6d61 6c65 0a20 2020  awing a male.   
-000045e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000045f0: 2020 2020 2069 6620 6174 7420 3d3d 2027       if att == '
-00004600: 7365 785f 7261 7469 6f27 3a0a 2020 2020  sex_ratio':.    
-00004610: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004620: 2020 2020 2020 2020 7661 6c20 3d20 7661          val = va
-00004630: 6c20 2f20 2876 616c 202b 2031 290a 2020  l / (val + 1).  
-00004640: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004650: 2020 2020 2020 2361 6464 2061 7320 616e        #add as an
-00004660: 2061 7474 7269 6275 7465 206f 6620 7468   attribute of th
-00004670: 6520 5f50 6172 616d 7356 616c 7320 6f62  e _ParamsVals ob
-00004680: 6a65 6374 2028 5f70 7629 0a20 2020 2020  ject (_pv).     
-00004690: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000046a0: 2020 2073 6574 6174 7472 2873 656c 662e     setattr(self.
-000046b0: 5f70 762c 2061 7474 2c20 7661 6c29 0a20  _pv, att, val). 
-000046c0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
-000046d0: 6966 2074 6865 2073 6563 7469 6f6e 2069  if the section i
-000046e0: 7320 276d 6f76 656d 656e 7427 2c20 616e  s 'movement', an
-000046f0: 6420 6974 2773 2069 6e20 7468 6520 7061  d it's in the pa
-00004700: 7261 6d73 2c0a 2020 2020 2020 2020 2020  rams,.          
-00004710: 2020 2020 2020 2374 6869 7320 6d65 616e        #this mean
-00004720: 7320 7468 6520 5370 6563 6965 7320 7368  s the Species sh
-00004730: 6f75 6c64 2068 6176 6520 6d6f 7665 6d65  ould have moveme
-00004740: 6e74 2c0a 2020 2020 2020 2020 2020 2020  nt,.            
-00004750: 2020 2020 2373 6f20 7570 6461 7465 2074      #so update t
-00004760: 6865 2073 656c 662e 5f6d 6f76 6520 6174  he self._move at
-00004770: 7472 6962 7574 650a 2020 2020 2020 2020  tribute.        
-00004780: 2020 2020 2020 2020 6966 2073 6563 7469          if secti
-00004790: 6f6e 203d 3d20 276d 6f76 656d 656e 7427  on == 'movement'
-000047a0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-000047b0: 2020 2020 2020 6966 2073 7070 5f70 6172        if spp_par
-000047c0: 616d 735b 7365 6374 696f 6e5d 2e6d 6f76  ams[section].mov
-000047d0: 653a 0a20 2020 2020 2020 2020 2020 2020  e:.             
-000047e0: 2020 2020 2020 2020 2020 2073 656c 662e             self.
-000047f0: 5f6d 6f76 6520 3d20 5472 7565 0a0a 2020  _move = True..  
-00004800: 2020 2020 2020 2369 6620 7365 7820 6973        #if sex is
-00004810: 2054 7275 6520 616e 6420 7265 7072 6f5f   True and repro_
-00004820: 6167 6520 6973 2061 6e20 696e 7420 6f72  age is an int or
-00004830: 2066 6c6f 6174 2c20 636f 6572 6365 2074   float, coerce t
-00004840: 6f20 6120 7475 706c 650a 2020 2020 2020  o a tuple.      
-00004850: 2020 2328 6f6e 6520 7661 6c20 666f 7220    #(one val for 
-00004860: 6561 6368 2073 6578 290a 2020 2020 2020  each sex).      
-00004870: 2020 6966 2073 656c 662e 7365 783a 0a20    if self.sex:. 
-00004880: 2020 2020 2020 2020 2020 2069 6620 7479             if ty
-00004890: 7065 2873 656c 662e 7265 7072 6f5f 6167  pe(self.repro_ag
-000048a0: 6529 2069 6e20 5b66 6c6f 6174 2c20 696e  e) in [float, in
-000048b0: 745d 3a0a 2020 2020 2020 2020 2020 2020  t]:.            
-000048c0: 2020 2020 7365 6c66 2e72 6570 726f 5f61      self.repro_a
-000048d0: 6765 203d 2028 7365 6c66 2e72 6570 726f  ge = (self.repro
-000048e0: 5f61 6765 2c20 7365 6c66 2e72 6570 726f  _age, self.repro
-000048f0: 5f61 6765 290a 0a20 2020 2020 2020 2023  _age)..        #
-00004900: 7365 7420 7468 6520 4765 6e6f 6d69 6341  set the GenomicA
-00004910: 7263 6869 7465 6374 7572 6520 6f62 6a65  rchitecture obje
-00004920: 6374 0a20 2020 2020 2020 2073 656c 662e  ct.        self.
-00004930: 6765 6e5f 6172 6368 203d 2067 656e 6f6d  gen_arch = genom
-00004940: 6963 5f61 7263 6869 7465 6374 7572 650a  ic_architecture.
-00004950: 2020 2020 2020 2020 6173 7365 7274 2028          assert (
-00004960: 7365 6c66 2e67 656e 5f61 7263 682e 5f5f  self.gen_arch.__
-00004970: 636c 6173 735f 5f2e 5f5f 6e61 6d65 5f5f  class__.__name__
-00004980: 203d 3d20 2747 656e 6f6d 6963 4172 6368   == 'GenomicArch
-00004990: 6974 6563 7475 7265 270a 2020 2020 2020  itecture'.      
-000049a0: 2020 2020 2020 6f72 2073 656c 662e 6765        or self.ge
-000049b0: 6e5f 6172 6368 2069 7320 4e6f 6e65 292c  n_arch is None),
-000049c0: 2028 2254 6865 2053 7065 6369 6573 2e67   ("The Species.g
-000049d0: 656e 5f61 7263 6820 6174 7472 6962 7574  en_arch attribut
-000049e0: 6520 220a 2020 2020 2020 2020 2020 2020  e ".            
-000049f0: 226d 7573 7420 6265 2061 6e20 696e 7374  "must be an inst
-00004a00: 616e 6365 206f 6620 7468 6520 6765 6e6f  ance of the geno
-00004a10: 6d65 2e47 656e 6f6d 6963 4172 6368 6974  me.GenomicArchit
-00004a20: 6563 7475 7265 2063 6c61 7373 2022 0a20  ecture class ". 
-00004a30: 2020 2020 2020 2020 2020 2022 6f72 2065             "or e
-00004a40: 6c73 6520 4e6f 6e65 2e22 290a 0a20 2020  lse None.")..   
-00004a50: 2020 2020 2023 2063 7265 6174 6520 6120       # create a 
-00004a60: 7473 6b69 742e 5461 626c 6543 6f6c 6c65  tskit.TableColle
-00004a70: 6374 696f 6e20 6f62 6a65 6374 2c20 6966  ction object, if
-00004a80: 2074 6865 2073 7065 6369 6573 2075 7365   the species use
-00004a90: 7320 6765 6e6f 6d65 730a 2020 2020 2020  s genomes.      
-00004aa0: 2020 6966 2073 656c 662e 6765 6e5f 6172    if self.gen_ar
-00004ab0: 6368 2069 7320 6e6f 7420 4e6f 6e65 2061  ch is not None a
-00004ac0: 6e64 2073 656c 662e 6765 6e5f 6172 6368  nd self.gen_arch
-00004ad0: 2e75 7365 5f74 736b 6974 3a0a 2020 2020  .use_tskit:.    
-00004ae0: 2020 2020 2020 2020 7365 6c66 2e5f 7463          self._tc
-00004af0: 203d 2074 736b 6974 2e54 6162 6c65 436f   = tskit.TableCo
-00004b00: 6c6c 6563 7469 6f6e 2873 6571 7565 6e63  llection(sequenc
-00004b10: 655f 6c65 6e67 7468 3d73 656c 662e 6765  e_length=self.ge
-00004b20: 6e5f 6172 6368 2e4c 290a 2020 2020 2020  n_arch.L).      
-00004b30: 2020 2020 2020 7365 6c66 2e5f 7463 5f73        self._tc_s
-00004b40: 6f72 7465 645f 616e 645f 7369 6d70 6c69  orted_and_simpli
-00004b50: 6669 6564 203d 2046 616c 7365 0a20 2020  fied = False.   
-00004b60: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
-00004b70: 2020 2020 2020 2073 656c 662e 5f74 6320         self._tc 
-00004b80: 3d20 4e6f 6e65 0a20 2020 2020 2020 2020  = None.         
-00004b90: 2020 2073 656c 662e 5f74 635f 736f 7274     self._tc_sort
-00004ba0: 6564 5f61 6e64 5f73 696d 706c 6966 6965  ed_and_simplifie
-00004bb0: 6420 3d20 4e6f 6e65 0a0a 2020 2020 2020  d = None..      
-00004bc0: 2020 2373 6574 2074 6865 2073 656c 6563    #set the selec
-00004bd0: 7469 6f6e 2061 7474 7269 6275 7465 2c20  tion attribute, 
-00004be0: 746f 2069 6e64 6963 6174 6520 7768 6574  to indicate whet
-00004bf0: 6865 7220 6f72 206e 6f74 0a20 2020 2020  her or not.     
-00004c00: 2020 2023 6e61 7475 7261 6c20 7365 6c65     #natural sele
-00004c10: 6374 696f 6e20 7368 6f75 6c64 2062 6520  ction should be 
-00004c20: 696d 706c 656d 656e 7465 6420 666f 7220  implemented for 
-00004c30: 7468 6520 7370 6563 6965 730a 2020 2020  the species.    
-00004c40: 2020 2020 7365 6c66 2e73 656c 6563 7469      self.selecti
-00004c50: 6f6e 203d 2028 7365 6c66 2e67 656e 5f61  on = (self.gen_a
-00004c60: 7263 6820 6973 206e 6f74 204e 6f6e 6520  rch is not None 
-00004c70: 616e 640a 2020 2020 2020 2020 2020 2020  and.            
-00004c80: 2873 656c 662e 6765 6e5f 6172 6368 2e6d  (self.gen_arch.m
-00004c90: 755f 6465 6c65 7420 3e20 3020 6f72 2073  u_delet > 0 or s
-00004ca0: 656c 662e 6765 6e5f 6172 6368 2e74 7261  elf.gen_arch.tra
-00004cb0: 6974 7320 6973 206e 6f74 204e 6f6e 6529  its is not None)
-00004cc0: 290a 0a20 2020 2020 2020 2023 7365 7420  )..        #set 
-00004cd0: 7468 6520 7365 6c66 2e6d 7574 6174 6520  the self.mutate 
-00004ce0: 6174 7472 6962 7574 6520 2861 2062 6f6f  attribute (a boo
-00004cf0: 6c65 616e 2069 6e64 6963 6174 696e 6720  lean indicating 
-00004d00: 7768 6574 6865 720a 2020 2020 2020 2020  whether.        
-00004d10: 236f 7220 6e6f 7420 746f 2065 6e61 6374  #or not to enact
-00004d20: 206d 7574 6174 696f 6e2c 2077 6869 6368   mutation, which
-00004d30: 2069 7320 5472 7565 2069 6620 6765 6e5f   is True if gen_
-00004d40: 6172 6368 2e5f 6d75 5f74 6f74 203e 2030  arch._mu_tot > 0
-00004d50: 0a20 2020 2020 2020 2073 656c 662e 6d75  .        self.mu
-00004d60: 7461 7465 203d 2028 7365 6c66 2e67 656e  tate = (self.gen
-00004d70: 5f61 7263 6820 6973 206e 6f74 204e 6f6e  _arch is not Non
-00004d80: 650a 2020 2020 2020 2020 2020 2020 2020  e.              
-00004d90: 2020 2020 2020 2020 2061 6e64 2073 656c           and sel
-00004da0: 662e 6765 6e5f 6172 6368 2e5f 6d75 5f74  f.gen_arch._mu_t
-00004db0: 6f74 2069 7320 6e6f 7420 4e6f 6e65 0a20  ot is not None. 
-00004dc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004dd0: 2020 2020 2020 616e 6420 7365 6c66 2e67        and self.g
-00004de0: 656e 5f61 7263 682e 5f6d 755f 746f 7420  en_arch._mu_tot 
-00004df0: 3e20 3029 0a0a 2020 2020 2020 2020 2373  > 0)..        #s
-00004e00: 6574 2074 6865 2073 656c 662e 6d75 745f  et the self.mut_
-00004e10: 6c6f 6720 6174 7472 6962 7574 652c 2077  log attribute, w
-00004e20: 6869 6368 2064 6963 7461 7465 7320 7768  hich dictates wh
-00004e30: 6574 6865 720a 2020 2020 2020 2020 236f  ether.        #o
-00004e40: 7220 6e6f 7420 6120 6d75 7461 7469 6f6e  r not a mutation
-00004e50: 206c 6f67 2073 686f 756c 6420 6265 2077   log should be w
-00004e60: 7269 7474 656e 2066 6f72 2074 6869 7320  ritten for this 
-00004e70: 7370 700a 2020 2020 2020 2020 7365 6c66  spp.        self
-00004e80: 2e6d 7574 5f6c 6f67 203d 204e 6f6e 650a  .mut_log = None.
-00004e90: 2020 2020 2020 2020 6966 2027 6765 6e5f          if 'gen_
-00004ea0: 6172 6368 2720 696e 205b 2a73 7070 5f70  arch' in [*spp_p
-00004eb0: 6172 616d 735d 3a0a 2020 2020 2020 2020  arams]:.        
-00004ec0: 2020 2020 7365 6c66 2e6d 7574 5f6c 6f67      self.mut_log
-00004ed0: 203d 2073 7070 5f70 6172 616d 732e 6765   = spp_params.ge
-00004ee0: 6e5f 6172 6368 2e6d 7574 5f6c 6f67 0a0a  n_arch.mut_log..
-00004ef0: 2020 2020 2020 2020 2363 7265 6174 6520          #create 
-00004f00: 6120 636f 6f72 6420 6174 7472 6765 7474  a coord attrgett
-00004f10: 6572 2066 756e 6374 696f 6e2c 0a20 2020  er function,.   
-00004f20: 2020 2020 2023 666f 7220 7573 6520 696e       #for use in
-00004f30: 2067 6574 7469 6e67 2061 6c6c 2069 6e64   getting all ind
-00004f40: 6976 6964 7327 2063 6f6f 7264 696e 6174  ivids' coordinat
-00004f50: 6573 0a20 2020 2020 2020 2073 656c 662e  es.        self.
-00004f60: 5f63 6f6f 7264 5f61 7474 7267 6574 7465  _coord_attrgette
-00004f70: 7220 3d20 6174 7472 6765 7474 6572 2827  r = attrgetter('
-00004f80: 7827 2c20 2779 2729 0a0a 2020 2020 2020  x', 'y')..      
-00004f90: 2020 2320 6372 6561 7465 2074 6865 2062    # create the b
-00004fa0: 7572 6e2d 696e 2073 7061 7469 616c 2063  urn-in spatial c
-00004fb0: 6f75 6e74 6572 0a20 2020 2020 2020 2073  ounter.        s
-00004fc0: 656c 662e 5f62 7572 6e69 6e5f 7370 6174  elf._burnin_spat
-00004fd0: 5f74 6573 7465 7220 3d20 6275 726e 696e  _tester = burnin
-00004fe0: 2e53 7061 7469 616c 5465 7374 6572 2873  .SpatialTester(s
-00004ff0: 656c 6629 0a0a 0a20 2020 2023 6f76 6572  elf)...    #over
-00005000: 7269 6465 2074 6865 205f 5f64 6565 7063  ride the __deepc
-00005010: 6f70 795f 5f20 6d65 7468 6f64 2c20 736f  opy__ method, so
-00005020: 2074 6861 7420 7468 6520 7370 6563 6965   that the specie
-00005030: 7320 6361 6e0a 2020 2020 2362 6520 636f  s can.    #be co
-00005040: 7079 2e64 6565 7063 6f70 7927 6420 2862  py.deepcopy'd (b
-00005050: 6563 6175 7365 206f 7468 6572 7769 7365  ecause otherwise
-00005060: 2074 6869 7320 646f 6573 6e27 7420 776f   this doesn't wo
-00005070: 726b 2066 6f72 0a20 2020 2023 636c 6173  rk for.    #clas
-00005080: 7365 7320 7468 6174 2069 6e68 6572 6974  ses that inherit
-00005090: 2066 726f 6d20 2363 6f6c 6c65 6374 696f   from #collectio
-000050a0: 6e73 2e4f 7264 6572 6564 4469 6374 290a  ns.OrderedDict).
-000050b0: 2020 2020 6465 6620 5f5f 6465 6570 636f      def __deepco
-000050c0: 7079 5f5f 2873 656c 662c 206d 656d 6f29  py__(self, memo)
-000050d0: 3a0a 2020 2020 2020 2020 636c 7320 3d20  :.        cls = 
-000050e0: 7365 6c66 2e5f 5f63 6c61 7373 5f5f 0a20  self.__class__. 
-000050f0: 2020 2020 2020 2072 6573 756c 7420 3d20         result = 
-00005100: 636c 732e 5f5f 6e65 775f 5f28 636c 7329  cls.__new__(cls)
-00005110: 0a20 2020 2020 2020 206d 656d 6f5b 6964  .        memo[id
-00005120: 2873 656c 6629 5d20 3d20 7265 7375 6c74  (self)] = result
-00005130: 0a20 2020 2020 2020 2066 6f72 206b 2c20  .        for k, 
-00005140: 7620 696e 2073 656c 662e 5f5f 6469 6374  v in self.__dict
-00005150: 5f5f 2e69 7465 6d73 2829 3a0a 2020 2020  __.items():.    
-00005160: 2020 2020 2020 2020 7365 7461 7474 7228          setattr(
-00005170: 7265 7375 6c74 2c20 6b2c 2064 6565 7063  result, k, deepc
-00005180: 6f70 7928 762c 206d 656d 6f29 290a 2020  opy(v, memo)).  
-00005190: 2020 2020 2020 666f 7220 6b2c 2076 2069        for k, v i
-000051a0: 6e20 7365 6c66 2e69 7465 6d73 2829 3a0a  n self.items():.
-000051b0: 2020 2020 2020 2020 2020 2020 7265 7375              resu
-000051c0: 6c74 5b64 6565 7063 6f70 7928 6b2c 206d  lt[deepcopy(k, m
-000051d0: 656d 6f29 5d20 3d20 6465 6570 636f 7079  emo)] = deepcopy
-000051e0: 2876 2c20 6d65 6d6f 290a 2020 2020 2020  (v, memo).      
-000051f0: 2020 7265 7475 726e 2072 6573 756c 740a    return result.
-00005200: 0a0a 2020 2020 2364 6566 696e 6520 7468  ..    #define th
-00005210: 6520 5f5f 7374 725f 5f20 616e 6420 5f5f  e __str__ and __
-00005220: 7265 7072 5f5f 2073 7065 6369 616c 206d  repr__ special m
-00005230: 6574 686f 6473 0a20 2020 2023 4e4f 5445  ethods.    #NOTE
-00005240: 3a20 7468 6973 2069 7320 6e6f 7420 7265  : this is not re
-00005250: 616c 6c79 2061 2067 7265 6174 2072 6570  ally a great rep
-00005260: 7265 7365 6e74 6174 696f 6e3b 2074 6865  resentation; the
-00005270: 2050 7974 686f 6e0a 2020 2020 2364 6f63   Python.    #doc
-00005280: 7320 696e 6469 6361 7465 2074 6861 7420  s indicate that 
-00005290: 5f5f 7265 7072 5f5f 2073 686f 756c 6420  __repr__ should 
-000052a0: 6964 6561 6c6c 790a 2020 2020 2370 726f  ideally.    #pro
-000052b0: 7669 6465 2061 2072 6570 7265 7365 6e74  vide a represent
-000052c0: 6174 696f 6e20 636f 756c 6420 6265 2075  ation could be u
-000052d0: 7365 6420 746f 2072 6563 7265 6174 6520  sed to recreate 
-000052e0: 7468 650a 2020 2020 236f 626a 6563 742c  the.    #object,
-000052f0: 2062 7574 2069 6620 7468 6174 2069 7320   but if that is 
-00005300: 6e6f 7420 706f 7373 6962 6c65 2074 6865  not possible the
-00005310: 6e20 6974 200a 2020 2020 2373 686f 756c  n it .    #shoul
-00005320: 6420 6174 206c 6561 7374 2070 726f 7669  d at least provi
-00005330: 6465 2061 2073 7472 696e 6720 6f66 2074  de a string of t
-00005340: 6865 2066 6f72 6d20 273c 2e2e 2e20 736f  he form '<... so
-00005350: 6d65 0a20 2020 2023 7573 6566 756c 2064  me.    #useful d
-00005360: 6573 6372 6970 7469 6f6e 202e 2e2e 3e3b  escription ...>;
-00005370: 2049 2776 6520 6174 7465 6d70 7465 6420   I've attempted 
-00005380: 746f 2064 6f0a 2020 2020 2374 6865 206c  to do.    #the l
-00005390: 6174 7465 722c 2069 6e73 7069 7265 6420  atter, inspired 
-000053a0: 6279 2061 2063 6f6d 626f 206f 6620 7768  by a combo of wh
-000053b0: 6174 2049 2776 6520 7365 656e 2069 6e20  at I've seen in 
-000053c0: 610a 2020 2020 2366 6577 206f 7468 6572  a.    #few other
-000053d0: 2070 6163 6b61 6765 7320 2870 616e 6461   packages (panda
-000053e0: 732c 206e 6574 4344 4634 2c20 6f73 6765  s, netCDF4, osge
-000053f0: 6f29 200a 2020 2020 2328 7468 6f75 6768  o) .    #(though
-00005400: 2070 6572 6861 7073 2049 2073 686f 756c   perhaps I shoul
-00005410: 6420 6d6f 7265 2063 6c6f 7365 6c79 2063  d more closely c
-00005420: 6f6e 7369 6465 7220 686f 7720 746f 2068  onsider how to h
-00005430: 616e 646c 6520 0a20 2020 2023 7468 6520  andle .    #the 
-00005440: 7061 7261 6d73 2069 6e20 6120 6d6f 7265  params in a more
-00005450: 206e 7561 6e63 6564 2c20 7072 6563 6973   nuanced, precis
-00005460: 6520 7761 7920 6f6e 6365 2049 276d 0a20  e way once I'm. 
-00005470: 2020 2023 646f 6e65 2077 7269 7469 6e67     #done writing
-00005480: 2074 6865 2063 6f64 6562 6173 652c 2062   the codebase, b
-00005490: 6563 6175 7365 2072 6967 6874 206e 6f77  ecause right now
-000054a0: 2074 6869 7320 6c65 6176 6573 206f 7574   this leaves out
-000054b0: 0a20 2020 2023 7061 7261 6d73 2074 6861  .    #params tha
-000054c0: 7420 6861 7665 206c 6973 7473 2061 6e64  t have lists and
-000054d0: 2064 6963 7469 6f6e 6172 6972 6573 2061   dictionarires a
-000054e0: 740a 2020 2020 2376 616c 7565 732c 2061  t.    #values, a
-000054f0: 6e64 2061 6c73 6f20 696e 636c 7564 6573  nd also includes
-00005500: 2073 6f6d 6520 7061 7261 6d65 7465 7273   some parameters
-00005510: 2074 6861 7420 6172 6520 696e 7075 7420   that are input 
-00005520: 696e 0a20 2020 2023 7061 7261 6d73 2e70  in.    #params.p
-00005530: 7920 6275 7420 6e6f 7420 616c 6c2c 2061  y but not all, a
-00005540: 6e64 2073 6f6d 6520 7061 7261 6d73 2074  nd some params t
-00005550: 6861 740a 2020 2020 2361 7265 2063 7265  hat.    #are cre
-00005560: 6174 6564 2069 6e74 6572 6e61 6c6c 7920  ated internally 
-00005570: 6275 7420 6e6f 7420 616c 6c0a 2020 2020  but not all.    
-00005580: 6465 6620 5f5f 7374 725f 5f28 7365 6c66  def __str__(self
-00005590: 293a 0a20 2020 2020 2020 2023 6765 7420  ):.        #get 
-000055a0: 6120 7374 7269 6e67 2072 6570 7265 7365  a string represe
-000055b0: 6e74 6174 696f 6e20 6f66 2074 6865 2063  ntation of the c
-000055c0: 6c61 7373 0a20 2020 2020 2020 2074 7970  lass.        typ
-000055d0: 655f 7374 7220 3d20 7374 7228 7479 7065  e_str = str(type
-000055e0: 2873 656c 6629 290a 2020 2020 2020 2020  (self)).        
-000055f0: 2367 6574 2061 2073 7472 696e 6720 7265  #get a string re
-00005600: 7072 6573 656e 7461 7469 6f6e 206f 6620  presentation of 
-00005610: 7468 6520 6669 7273 7420 616e 6420 6c61  the first and la
-00005620: 7374 2069 6e64 6976 6964 7561 6c73 0a20  st individuals. 
-00005630: 2020 2020 2020 2069 6e64 735f 7374 7220         inds_str 
-00005640: 3d20 2725 6920 496e 6469 7669 6475 616c  = '%i Individual
-00005650: 733a 5c6e 5c74 2720 2520 6c65 6e28 7365  s:\n\t' % len(se
-00005660: 6c66 290a 2020 2020 2020 2020 6669 7273  lf).        firs
-00005670: 745f 696e 645f 7374 7220 3d20 4f44 2873  t_ind_str = OD(s
-00005680: 656c 662e 6974 656d 7328 2929 2e5f 5f73  elf.items()).__s
-00005690: 7472 5f5f 2829 2e73 706c 6974 280a 2020  tr__().split(.  
-000056a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000056b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000056c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000056d0: 2020 2729 2c20 2729 5b30 5d20 2b20 2729    '), ')[0] + ')
-000056e0: 5c6e 5c74 2e2e 2e5c 6e5c 7427 0a20 2020  \n\t...\n\t'.   
-000056f0: 2020 2020 206c 6173 745f 696e 645f 7374       last_ind_st
-00005700: 7220 3d20 272c 2027 2e6a 6f69 6e28 7365  r = ', '.join(se
-00005710: 6c66 2e69 7465 6d73 2829 2e5f 5f73 7472  lf.items().__str
-00005720: 5f5f 2829 2e73 706c 6974 2827 2c20 2729  __().split(', ')
-00005730: 5b2d 323a 5d29 0a20 2020 2020 2020 2069  [-2:]).        i
-00005740: 6e64 735f 7374 7220 3d20 696e 6473 5f73  nds_str = inds_s
-00005750: 7472 202b 2066 6972 7374 5f69 6e64 5f73  tr + first_ind_s
-00005760: 7472 202b 206c 6173 745f 696e 645f 7374  tr + last_ind_st
-00005770: 7220 2b20 275c 6e27 0a20 2020 2020 2020  r + '\n'.       
-00005780: 2023 6765 7420 6120 7374 7269 6e67 2072   #get a string r
-00005790: 6570 7265 7365 6e74 6174 696f 6e20 6f66  epresentation of
-000057a0: 2074 6865 2066 6972 7374 2074 776f 2061   the first two a
-000057b0: 6e64 206c 6173 7420 7477 6f20 7061 7261  nd last two para
-000057c0: 6d65 7465 7273 0a20 2020 2020 2020 2023  meters.        #
-000057d0: 7061 7261 6d73 203d 2073 6f72 7465 6428  params = sorted(
-000057e0: 5b73 7472 286b 2920 2b20 273a 2027 202b  [str(k) + ': ' +
-000057f0: 7374 7228 7629 2066 6f72 206b 2c76 2069  str(v) for k,v i
-00005800: 6e20 7661 7273 280a 2020 2020 2020 2020  n vars(.        
-00005810: 2320 2020 7365 6c66 292e 6974 656d 7328  #   self).items(
-00005820: 2920 6966 2074 7970 6528 7629 2069 6e20  ) if type(v) in 
-00005830: 2873 7472 2c20 696e 742c 2062 6f6f 6c2c  (str, int, bool,
-00005840: 0a20 2020 2020 2020 2023 2020 2020 2020  .        #      
-00005850: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005860: 2020 2020 2020 2020 2066 6c6f 6174 295d           float)]
-00005870: 2c20 6964 7820 3d20 6c61 6d62 6461 2078  , idx = lambda x
-00005880: 3a20 782e 6c6f 7765 7228 2929 0a20 2020  : x.lower()).   
-00005890: 2020 2020 2023 7061 7261 6d73 5f73 7472       #params_str
-000058a0: 203d 2022 5061 7261 6d65 7465 7273 3a5c   = "Parameters:\
-000058b0: 6e5c 7422 202b 2027 2c5c 6e5c 7427 2e6a  n\t" + ',\n\t'.j
-000058c0: 6f69 6e28 7061 7261 6d73 5b3a 325d 2920  oin(params[:2]) 
-000058d0: 2b20 272c 270a 2020 2020 2020 2020 2370  + ','.        #p
-000058e0: 6172 616d 735f 7374 7220 3d20 7061 7261  arams_str = para
-000058f0: 6d73 5f73 7472 202b 2027 5c6e 5c74 2e2e  ms_str + '\n\t..
-00005900: 2e5c 6e5c 7427 0a20 2020 2020 2020 2023  .\n\t'.        #
-00005910: 7061 7261 6d73 5f73 7472 203d 2070 6172  params_str = par
-00005920: 616d 735f 7374 7220 2b20 272c 5c6e 5c74  ams_str + ',\n\t
-00005930: 272e 6a6f 696e 2870 6172 616d 735b 2d32  '.join(params[-2
-00005940: 3a5d 290a 2020 2020 2020 2020 2372 6574  :]).        #ret
-00005950: 7572 6e20 275c 6e27 2e6a 6f69 6e28 5b74  urn '\n'.join([t
-00005960: 7970 655f 7374 722c 2069 6e64 735f 7374  ype_str, inds_st
-00005970: 722c 2070 6172 616d 735f 7374 725d 290a  r, params_str]).
-00005980: 2020 2020 2020 2020 7265 7475 726e 2027          return '
-00005990: 5c6e 272e 6a6f 696e 285b 7479 7065 5f73  \n'.join([type_s
-000059a0: 7472 2c20 696e 6473 5f73 7472 5d29 0a0a  tr, inds_str])..
-000059b0: 2020 2020 6465 6620 5f5f 7265 7072 5f5f      def __repr__
-000059c0: 2873 656c 6629 3a0a 2020 2020 2020 2020  (self):.        
-000059d0: 7265 7072 5f73 7472 203d 2073 656c 662e  repr_str = self.
-000059e0: 5f5f 7374 725f 5f28 290a 2020 2020 2020  __str__().      
-000059f0: 2020 7265 7475 726e 2072 6570 725f 7374    return repr_st
-00005a00: 720a 0a20 2020 2023 6375 7374 6f6d 697a  r..    #customiz
-00005a10: 6520 7468 6520 5f5f 6765 7461 7474 725f  e the __getattr_
-00005a20: 5f20 7370 6563 6961 6c20 6d65 7468 6f64  _ special method
-00005a30: 2c20 736f 2074 6861 7420 6174 7472 6962  , so that attrib
-00005a40: 7574 6573 2069 6e73 6964 6520 7468 650a  utes inside the.
-00005a50: 2020 2020 2370 7269 7661 7465 205f 5061      #private _Pa
-00005a60: 7261 6d73 5661 6c73 2061 7474 7269 6275  ramsVals attribu
-00005a70: 7465 2028 5f70 7629 2061 6c73 6f20 6265  te (_pv) also be
-00005a80: 6861 7665 2061 7320 7468 6f75 6768 2074  have as though t
-00005a90: 6865 6972 2061 7474 7269 6275 7465 730a  heir attributes.
-00005aa0: 2020 2020 236f 6620 7468 6520 5370 6563      #of the Spec
-00005ab0: 6965 730a 2020 2020 6465 6620 5f5f 6765  ies.    def __ge
-00005ac0: 7461 7474 725f 5f28 7365 6c66 2c20 6174  tattr__(self, at
-00005ad0: 7472 293a 0a20 2020 2020 2020 2074 7279  tr):.        try
-00005ae0: 3a0a 2020 2020 2020 2020 2020 2020 7661  :.            va
-00005af0: 6c20 3d20 7365 6c66 2e5f 7076 2e5f 5f67  l = self._pv.__g
-00005b00: 6574 6174 7472 6962 7574 655f 5f28 6174  etattribute__(at
-00005b10: 7472 290a 2020 2020 2020 2020 2020 2020  tr).            
-00005b20: 7265 7475 726e 2076 616c 0a20 2020 2020  return val.     
-00005b30: 2020 2065 7863 6570 7420 4578 6365 7074     except Except
-00005b40: 696f 6e3a 0a20 2020 2020 2020 2020 2020  ion:.           
-00005b50: 2072 6169 7365 2041 7474 7269 6275 7465   raise Attribute
-00005b60: 4572 726f 7228 2254 6865 2053 7065 6369  Error("The Speci
-00005b70: 6573 2068 6173 206e 6f20 6174 7472 6962  es has no attrib
-00005b80: 7574 6520 2573 2220 2520 6174 7472 290a  ute %s" % attr).
-00005b90: 0a0a 2020 2020 2323 2323 2323 2323 2323  ..    ##########
-00005ba0: 2323 2323 2323 2323 2323 230a 2020 2020  ###########.    
-00005bb0: 2323 2320 4f54 4845 5220 4d45 5448 4f44  ### OTHER METHOD
-00005bc0: 5320 2323 230a 2020 2020 2323 2323 2323  S ###.    ######
-00005bd0: 2323 2323 2323 2323 2323 2323 2323 230a  ###############.
-00005be0: 0a20 2020 2020 2020 2023 2323 2323 2323  .        #######
-00005bf0: 2323 2323 2323 2323 2323 0a20 2020 2020  ##########.     
-00005c00: 2020 2023 7072 6976 6174 6520 6d65 7468     #private meth
-00005c10: 6f64 7323 0a20 2020 2020 2020 2023 2323  ods#.        ###
-00005c20: 2323 2323 2323 2323 2323 2323 2323 0a0a  ##############..
-00005c30: 2020 2020 236d 6574 686f 6420 746f 2063      #method to c
-00005c40: 616c 6375 6c61 7465 2061 6e64 2073 6574  alculate and set
-00005c50: 2073 656c 662e 4b0a 2020 2020 6465 6620   self.K.    def 
-00005c60: 5f73 6574 5f4b 2873 656c 662c 206c 616e  _set_K(self, lan
-00005c70: 6429 3a0a 2020 2020 2020 2020 7365 6c66  d):.        self
-00005c80: 2e4b 203d 206c 616e 645b 7365 6c66 2e4b  .K = land[self.K
-00005c90: 5f6c 6179 6572 5d2e 7261 7374 202a 2073  _layer].rast * s
-00005ca0: 656c 662e 4b5f 6661 6374 6f72 0a0a 2020  elf.K_factor..  
-00005cb0: 2020 236d 6574 686f 6420 746f 2073 6574    #method to set
-00005cc0: 2073 656c 662e 4e0a 2020 2020 6465 6620   self.N.    def 
-00005cd0: 5f73 6574 5f4e 2873 656c 662c 204e 293a  _set_N(self, N):
-00005ce0: 2020 234e 4f54 453a 2052 6571 7569 7265    #NOTE: Require
-00005cf0: 7320 6120 6c61 6e64 7363 6170 652e 4c61  s a landscape.La
-00005d00: 7965 7220 696e 7374 616e 6365 0a20 2020  yer instance.   
-00005d10: 2020 2020 2073 656c 662e 4e20 3d20 4e0a       self.N = N.
-00005d20: 0a20 2020 2023 6d65 7468 6f64 2074 6f20  .    #method to 
-00005d30: 6170 7065 6e64 2063 7572 7265 6e74 2073  append current s
-00005d40: 7070 2073 697a 6520 746f 2074 6865 2073  pp size to the s
-00005d50: 7070 2e4e 7420 6c69 7374 0a20 2020 2064  pp.Nt list.    d
-00005d60: 6566 205f 7365 745f 4e74 2873 656c 6629  ef _set_Nt(self)
-00005d70: 3a0a 2020 2020 2020 2020 7365 6c66 2e4e  :.        self.N
-00005d80: 742e 6170 7065 6e64 286c 656e 2873 656c  t.append(len(sel
-00005d90: 6629 290a 0a20 2020 2023 6d65 7468 6f64  f))..    #method
-00005da0: 2074 6f20 696e 6372 656d 656e 7420 7468   to increment th
-00005db0: 6520 7365 6c66 2e74 2061 7474 7269 6275  e self.t attribu
-00005dc0: 7465 2028 692e 652e 2074 6865 2074 696d  te (i.e. the tim
-00005dd0: 6573 7465 7020 636f 756e 7465 7229 0a20  estep counter). 
-00005de0: 2020 2064 6566 205f 7365 745f 7428 7365     def _set_t(se
-00005df0: 6c66 293a 0a20 2020 2020 2020 2073 656c  lf):.        sel
-00005e00: 662e 7420 2b3d 2031 0a0a 2020 2020 236d  f.t += 1..    #m
-00005e10: 6574 686f 6420 746f 2072 6573 6574 2074  ethod to reset t
-00005e20: 6865 2073 656c 662e 7420 6174 7472 6962  he self.t attrib
-00005e30: 7574 6520 2869 2e65 2e20 7468 6520 7469  ute (i.e. the ti
-00005e40: 6d65 7374 6570 2063 6f75 6e74 6572 290a  mestep counter).
-00005e50: 2020 2020 6465 6620 5f72 6573 6574 5f74      def _reset_t
-00005e60: 2873 656c 6629 3a0a 2020 2020 2020 2020  (self):.        
-00005e70: 7365 6c66 2e74 203d 202d 310a 0a20 2020  self.t = -1..   
-00005e80: 2023 6d65 7468 6f64 2074 6f20 696e 6372   #method to incr
-00005e90: 656d 656e 7420 616c 6c20 7370 6563 6965  ement all specie
-00005ea0: 7327 2061 6765 2062 7920 6f6e 650a 2020  s' age by one.  
-00005eb0: 2020 2328 616c 736f 2061 6464 7320 6375    #(also adds cu
-00005ec0: 7272 656e 7420 7370 7020 7369 7a65 2074  rrent spp size t
-00005ed0: 6f20 7472 6163 6b69 6e67 2061 7272 6179  o tracking array
-00005ee0: 290a 2020 2020 6465 6620 5f73 6574 5f61  ).    def _set_a
-00005ef0: 6765 5f73 7461 6765 2873 656c 6629 3a0a  ge_stage(self):.
-00005f00: 2020 2020 2020 2020 2320 696e 6372 656d          # increm
-00005f10: 656e 7420 6167 6520 6f66 2061 6c6c 2069  ent age of all i
-00005f20: 6e64 6976 6964 7561 6c73 0a20 2020 2020  ndividuals.     
-00005f30: 2020 205b 696e 642e 5f73 6574 5f61 6765     [ind._set_age
-00005f40: 5f73 7461 6765 2829 2066 6f72 2069 6e64  _stage() for ind
-00005f50: 2069 6e20 7365 6c66 2e76 616c 7565 7328   in self.values(
-00005f60: 295d 3b0a 0a20 2020 2023 206d 6574 686f  )];..    # metho
-00005f70: 6420 666f 7220 7275 6e6e 696e 6720 7468  d for running th
-00005f80: 6520 7370 6174 6961 6c20 6275 726e 2d69  e spatial burn-i
-00005f90: 6e20 7465 7374 0a20 2020 2064 6566 205f  n test.    def _
-00005fa0: 646f 5f73 7061 7469 616c 5f62 7572 6e69  do_spatial_burni
-00005fb0: 6e5f 7465 7374 2873 656c 662c 206e 756d  n_test(self, num
-00005fc0: 5f74 696d 6573 7465 7073 5f62 6163 6b29  _timesteps_back)
-00005fd0: 3a0a 2020 2020 2020 2020 2320 6669 7273  :.        # firs
-00005fe0: 7420 7570 6461 7465 2074 6865 2063 6f75  t update the cou
-00005ff0: 6e74 7320 616e 6420 6469 6666 7320 6172  nts and diffs ar
-00006000: 7261 7973 0a20 2020 2020 2020 2023 2028  rays.        # (
-00006010: 7768 6963 6820 636f 756e 7420 7468 6520  which count the 
-00006020: 6e75 6d62 6572 206f 6620 696e 6469 7669  number of indivi
-00006030: 6475 616c 7320 7769 7468 696e 2065 6163  duals within eac
-00006040: 6820 6c61 6e64 7363 6170 6520 6365 6c6c  h landscape cell
-00006050: 2c0a 2020 2020 2020 2020 2320 616e 6420  ,.        # and 
-00006060: 7468 6520 6469 6666 2069 6e20 7468 6174  the diff in that
-00006070: 206e 756d 6265 7220 6265 7477 6565 6e20   number between 
-00006080: 6c61 7374 2063 6865 636b 2061 6e64 2074  last check and t
-00006090: 6869 7320 6f6e 6529 0a20 2020 2020 2020  his one).       
-000060a0: 2073 656c 662e 5f62 7572 6e69 6e5f 7370   self._burnin_sp
-000060b0: 6174 5f74 6573 7465 722e 7570 6461 7465  at_tester.update
-000060c0: 2873 656c 6629 0a20 2020 2020 2020 2072  (self).        r
-000060d0: 6573 756c 7420 3d20 7365 6c66 2e5f 6275  esult = self._bu
-000060e0: 726e 696e 5f73 7061 745f 7465 7374 6572  rnin_spat_tester
-000060f0: 2e72 756e 5f74 6573 7428 6e75 6d5f 7469  .run_test(num_ti
-00006100: 6d65 7374 6570 735f 6261 636b 290a 2020  mesteps_back).  
-00006110: 2020 2020 2020 7265 7475 726e 2072 6573        return res
-00006120: 756c 740a 0a20 2020 2023 6d65 7468 6f64  ult..    #method
-00006130: 2074 6f20 6d6f 7665 2061 6c6c 2069 6e64   to move all ind
-00006140: 6976 6964 7561 6c73 2073 696d 756c 7461  ividuals simulta
-00006150: 6e65 6f75 736c 792c 2061 6e64 2073 616d  neously, and sam
-00006160: 706c 650a 2020 2020 2374 6865 6972 206e  ple.    #their n
-00006170: 6577 206c 6f63 6174 696f 6e73 2720 656e  ew locations' en
-00006180: 7669 726f 6e6d 656e 740a 2020 2020 6465  vironment.    de
-00006190: 6620 5f64 6f5f 6d6f 7665 6d65 6e74 2873  f _do_movement(s
-000061a0: 656c 662c 206c 616e 6429 3a0a 2020 2020  elf, land):.    
-000061b0: 2020 2020 5f64 6f5f 6d6f 7665 6d65 6e74      _do_movement
-000061c0: 2873 656c 6629 0a20 2020 2020 2020 2073  (self).        s
-000061d0: 656c 662e 5f73 6574 5f65 286c 616e 6429  elf._set_e(land)
-000061e0: 0a20 2020 2020 2020 2073 656c 662e 5f73  .        self._s
-000061f0: 6574 5f63 6f6f 7264 735f 616e 645f 6365  et_coords_and_ce
-00006200: 6c6c 7328 290a 0a20 2020 2023 6675 6e63  lls()..    #func
-00006210: 7469 6f6e 2066 6f72 2066 696e 6469 6e67  tion for finding
-00006220: 2061 6c6c 2074 6865 206d 6174 696e 6720   all the mating 
-00006230: 7061 6972 7320 696e 2061 2073 7065 6369  pairs in a speci
-00006240: 6573 0a20 2020 2064 6566 205f 6669 6e64  es.    def _find
-00006250: 5f6d 6174 696e 675f 7061 6972 7328 7365  _mating_pairs(se
-00006260: 6c66 293a 0a20 2020 2020 2020 206d 6174  lf):.        mat
-00006270: 696e 675f 7061 6972 7320 3d20 5f66 696e  ing_pairs = _fin
-00006280: 645f 6d61 7465 7328 7365 6c66 2c20 7365  d_mates(self, se
-00006290: 783d 7365 6c66 2e73 6578 2c0a 2020 2020  x=self.sex,.    
-000062a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000062b0: 2020 2020 2020 2020 2020 2020 2020 2063                 c
-000062c0: 686f 6f73 655f 6e65 6172 6573 743d 7365  hoose_nearest=se
-000062d0: 6c66 2e63 686f 6f73 655f 6e65 6172 6573  lf.choose_neares
-000062e0: 745f 6d61 7465 2c0a 2020 2020 2020 2020  t_mate,.        
-000062f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006300: 2020 2020 2020 2020 2020 2069 6e76 6572             inver
-00006310: 7365 5f64 6973 745f 6d61 7469 6e67 3d73  se_dist_mating=s
-00006320: 656c 662e 696e 7665 7273 655f 6469 7374  elf.inverse_dist
-00006330: 5f6d 6174 696e 6729 0a20 2020 2020 2020  _mating).       
-00006340: 2072 6574 7572 6e20 6d61 7469 6e67 5f70   return mating_p
-00006350: 6169 7273 0a0a 2020 2020 2366 756e 6374  airs..    #funct
-00006360: 696f 6e20 666f 7220 6578 6563 7574 696e  ion for executin
-00006370: 6720 6d61 7469 6e67 2066 6f72 2061 2073  g mating for a s
-00006380: 7065 6369 6573 0a20 2020 2064 6566 205f  pecies.    def _
-00006390: 646f 5f6d 6174 696e 6728 7365 6c66 2c20  do_mating(self, 
-000063a0: 6c61 6e64 2c20 6d61 7469 6e67 5f70 6169  land, mating_pai
-000063b0: 7273 2c20 6275 726e 3d46 616c 7365 2c0a  rs, burn=False,.
-000063c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000063d0: 2020 2063 6865 636b 5f68 6170 733d 4661     check_haps=Fa
-000063e0: 6c73 652c 2063 6865 636b 5f67 656e 6f74  lse, check_genot
-000063f0: 7970 6573 3d46 616c 7365 293a 0a0a 2020  ypes=False):..  
-00006400: 2020 2020 2020 2320 7365 7420 7468 6520        # set the 
-00006410: 6765 6e6f 7479 7065 2d63 6865 636b 2063  genotype-check c
-00006420: 6f75 6e74 6572 2c20 6966 206e 6563 6573  ounter, if neces
-00006430: 7361 7279 0a20 2020 2020 2020 2069 6620  sary.        if 
-00006440: 6368 6563 6b5f 6765 6e6f 7479 7065 733a  check_genotypes:
-00006450: 0a20 2020 2020 2020 2020 2020 2067 745f  .            gt_
-00006460: 6368 6563 6b5f 6e75 6d20 3d20 300a 0a20  check_num = 0.. 
-00006470: 2020 2020 2020 2023 6472 6177 2074 6865         #draw the
-00006480: 206e 756d 6265 7220 6f66 2062 6972 7468   number of birth
-00006490: 7320 666f 7220 6561 6368 2070 6169 722c  s for each pair,
-000064a0: 2061 6e64 2061 7070 656e 640a 2020 2020   and append.    
-000064b0: 2020 2020 2374 6f74 616c 2062 6972 7468      #total birth
-000064c0: 7320 746f 2073 656c 662e 6e5f 6269 7274  s to self.n_birt
-000064d0: 6873 206c 6973 740a 2020 2020 2020 2020  hs list.        
-000064e0: 6966 2073 656c 662e 6e5f 6269 7274 6873  if self.n_births
-000064f0: 5f66 6978 6564 3a0a 2020 2020 2020 2020  _fixed:.        
-00006500: 2020 2020 6e5f 6269 7274 6873 203d 206e      n_births = n
-00006510: 702e 6172 7261 7928 0a20 2020 2020 2020  p.array(.       
-00006520: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006530: 2020 2020 205b 7365 6c66 2e6e 5f62 6972       [self.n_bir
-00006540: 7468 735f 6469 7374 725f 6c61 6d62 6461  ths_distr_lambda
-00006550: 5d20 2a20 6c65 6e28 6d61 7469 6e67 5f70  ] * len(mating_p
-00006560: 6169 7273 2929 0a20 2020 2020 2020 2065  airs)).        e
-00006570: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
-00006580: 206e 5f62 6972 7468 7320 3d20 5f64 7261   n_births = _dra
-00006590: 775f 6e5f 6269 7274 6873 286c 656e 280a  w_n_births(len(.
-000065a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000065b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000065c0: 2020 2020 6d61 7469 6e67 5f70 6169 7273      mating_pairs
-000065d0: 292c 2073 656c 662e 6e5f 6269 7274 6873  ), self.n_births
-000065e0: 5f64 6973 7472 5f6c 616d 6264 6129 0a20  _distr_lambda). 
-000065f0: 2020 2020 2020 2074 6f74 616c 5f62 6972         total_bir
-00006600: 7468 7320 3d20 7375 6d28 6e5f 6269 7274  ths = sum(n_birt
-00006610: 6873 290a 2020 2020 2020 2020 7365 6c66  hs).        self
-00006620: 2e6e 5f62 6972 7468 732e 6170 7065 6e64  .n_births.append
-00006630: 2874 6f74 616c 5f62 6972 7468 7329 0a0a  (total_births)..
-00006640: 2020 2020 2020 2020 2363 7265 6174 6520          #create 
-00006650: 7468 6520 6f66 6673 7072 696e 675f 6964  the offspring_id
-00006660: 730a 2020 2020 2020 2020 6e65 7874 5f6f  s.        next_o
-00006670: 6666 7370 7269 6e67 5f6b 6579 203d 2073  ffspring_key = s
-00006680: 656c 662e 6d61 785f 696e 645f 6964 7820  elf.max_ind_idx 
-00006690: 2b20 310a 2020 2020 2020 2020 6f66 6673  + 1.        offs
-000066a0: 7072 696e 675f 6b65 7973 203d 206c 6973  pring_keys = lis
-000066b0: 7428 7261 6e67 6528 6e65 7874 5f6f 6666  t(range(next_off
-000066c0: 7370 7269 6e67 5f6b 6579 2c0a 2020 2020  spring_key,.    
-000066d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000066e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000066f0: 6e65 7874 5f6f 6666 7370 7269 6e67 5f6b  next_offspring_k
-00006700: 6579 202b 2074 6f74 616c 5f62 6972 7468  ey + total_birth
-00006710: 7329 295b 3a3a 2d31 5d0a 2020 2020 2020  s))[::-1].      
-00006720: 2020 2375 7064 6174 6520 7365 6c66 2e6d    #update self.m
-00006730: 6178 5f69 6e64 5f69 6478 0a20 2020 2020  ax_ind_idx.     
-00006740: 2020 2069 6620 6c65 6e28 6f66 6673 7072     if len(offspr
-00006750: 696e 675f 6b65 7973 2920 3e20 303a 0a20  ing_keys) > 0:. 
-00006760: 2020 2020 2020 2020 2020 2073 656c 662e             self.
-00006770: 6d61 785f 696e 645f 6964 7820 3d20 6f66  max_ind_idx = of
-00006780: 6673 7072 696e 675f 6b65 7973 5b30 5d0a  fspring_keys[0].
-00006790: 0a20 2020 2020 2020 2023 636f 7079 2074  .        #copy t
-000067a0: 6865 206b 6579 732c 2066 6f72 2075 7365  he keys, for use
-000067b0: 2069 6e20 6d75 7461 7469 6f6e 2e64 6f5f   in mutation.do_
-000067c0: 6d75 7461 7469 6f6e 2829 0a20 2020 2020  mutation().     
-000067d0: 2020 206b 6579 735f 6c69 7374 203d 205b     keys_list = [
-000067e0: 2a6f 6666 7370 7269 6e67 5f6b 6579 735d  *offspring_keys]
-000067f0: 0a0a 2020 2020 2020 2020 6966 206e 6f74  ..        if not
-00006800: 2062 7572 6e20 616e 6420 7365 6c66 2e67   burn and self.g
-00006810: 656e 5f61 7263 6820 6973 206e 6f74 204e  en_arch is not N
-00006820: 6f6e 653a 0a20 2020 2020 2020 2020 2020  one:.           
-00006830: 2072 6563 6f6d 625f 6b65 7973 203d 205b   recomb_keys = [
-00006840: 2a72 2e72 616e 6469 6e74 286c 6f77 3d30  *r.randint(low=0
-00006850: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00006860: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006870: 2020 2020 2020 2020 6869 6768 3d73 656c          high=sel
-00006880: 662e 6765 6e5f 6172 6368 2e72 6563 6f6d  f.gen_arch.recom
-00006890: 6269 6e61 7469 6f6e 732e 5f6e 2c0a 2020  binations._n,.  
-000068a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000068b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000068c0: 2020 2020 7369 7a65 3d74 6f74 616c 5f62      size=total_b
-000068d0: 6972 7468 732a 3229 5d0a 2020 2020 2020  irths*2)].      
-000068e0: 2020 2020 2020 2320 4e4f 5445 3a20 7468        # NOTE: th
-000068f0: 6973 2067 6976 6573 2075 7320 6120 6869  is gives us a hi
-00006900: 6572 6172 6368 6963 616c 6c79 206f 7267  erarchically org
-00006910: 616e 697a 6564 206c 6973 742c 0a20 2020  anized list,.   
-00006920: 2020 2020 2020 2020 2023 2063 6f6e 7461           # conta
-00006930: 696e 696e 6720 626f 7468 2074 6865 206f  ining both the o
-00006940: 6666 7370 7269 6e67 2773 206e 6577 2067  ffspring's new g
-00006950: 656e 6f6d 6573 2061 6e64 2074 6865 6972  enomes and their
-00006960: 0a20 2020 2020 2020 2020 2020 2023 2073  .            # s
-00006970: 6567 6d65 6e74 2069 6e66 6f72 6d61 7469  egment informati
-00006980: 6f6e 2028 746f 2062 6520 6164 6465 6420  on (to be added 
-00006990: 746f 2074 6865 0a20 2020 2020 2020 2020  to the.         
-000069a0: 2020 2023 2074 736b 6974 2e54 6162 6c65     # tskit.Table
-000069b0: 436f 6c6c 6563 7469 6f6e 2e65 6467 6573  Collection.edges
-000069c0: 2074 6162 6c65 292c 206f 7267 616e 697a   table), organiz
-000069d0: 6564 2061 7320 6120 7375 626c 6973 7420  ed as a sublist 
-000069e0: 666f 720a 2020 2020 2020 2020 2020 2020  for.            
-000069f0: 2320 6561 6368 2070 6169 722c 2074 6865  # each pair, the
-00006a00: 6e20 6120 7375 6274 7570 6c65 2077 6974  n a subtuple wit
-00006a10: 6869 6e20 7468 6174 2c20 636f 6e74 6169  hin that, contai
-00006a20: 6e69 6e67 2061 206e 6577 2067 656e 6f6d  ning a new genom
-00006a30: 650a 2020 2020 2020 2020 2020 2020 2320  e.            # 
-00006a40: 616e 6420 7468 656e 2061 2032 2d74 7570  and then a 2-tup
-00006a50: 6c65 206f 6620 7365 676d 656e 7473 2069  le of segments i
-00006a60: 6e66 6f20 666f 7220 6561 6368 206f 6620  nfo for each of 
-00006a70: 7468 6520 6e65 770a 2020 2020 2020 2020  the new.        
-00006a80: 2020 2020 2320 6f66 6673 7072 696e 6727      # offspring'
-00006a90: 7320 686f 6d6f 6c6f 6775 6573 0a20 2020  s homologues.   
-00006aa0: 2020 2020 2020 2020 2067 656e 6f6d 6573           genomes
-00006ab0: 5f61 6e64 5f73 6567 7320 3d20 5f64 6f5f  _and_segs = _do_
-00006ac0: 6d61 7469 6e67 2873 656c 662c 206d 6174  mating(self, mat
-00006ad0: 696e 675f 7061 6972 732c 0a20 2020 2020  ing_pairs,.     
-00006ae0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006af0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006b00: 2020 2020 206e 5f62 6972 7468 732c 2072       n_births, r
-00006b10: 6563 6f6d 625f 6b65 7973 290a 0a20 2020  ecomb_keys)..   
-00006b20: 2020 2020 2066 6f72 206e 5f70 6169 722c       for n_pair,
-00006b30: 2070 6169 7220 696e 2065 6e75 6d65 7261   pair in enumera
-00006b40: 7465 286d 6174 696e 675f 7061 6972 7329  te(mating_pairs)
-00006b50: 3a0a 0a20 2020 2020 2020 2020 2020 2070  :..            p
-00006b60: 6172 656e 745f 6d69 6470 6f69 6e74 5f78  arent_midpoint_x
-00006b70: 203d 2028 7365 6c66 5b70 6169 725b 305d   = (self[pair[0]
-00006b80: 5d2e 7820 2b20 7365 6c66 5b70 6169 725b  ].x + self[pair[
-00006b90: 315d 5d2e 7829 2f32 0a20 2020 2020 2020  1]].x)/2.       
-00006ba0: 2020 2020 2070 6172 656e 745f 6d69 6470       parent_midp
-00006bb0: 6f69 6e74 5f79 203d 2028 7365 6c66 5b70  oint_y = (self[p
-00006bc0: 6169 725b 305d 5d2e 7920 2b20 7365 6c66  air[0]].y + self
-00006bd0: 5b70 6169 725b 315d 5d2e 7929 2f32 0a0a  [pair[1]].y)/2..
-00006be0: 2020 2020 2020 2020 2020 2020 6e5f 6f66              n_of
-00006bf0: 6673 7072 696e 6720 3d20 6e5f 6269 7274  fspring = n_birt
-00006c00: 6873 5b6e 5f70 6169 725d 0a0a 2020 2020  hs[n_pair]..    
-00006c10: 2020 2020 2020 2020 666f 7220 6f66 6673          for offs
-00006c20: 7072 696e 675f 6920 696e 2072 616e 6765  pring_i in range
-00006c30: 286e 5f6f 6666 7370 7269 6e67 293a 0a0a  (n_offspring):..
-00006c40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006c50: 2367 6574 2074 6865 206e 6578 7420 6f66  #get the next of
-00006c60: 6673 7072 696e 675f 6b65 790a 2020 2020  fspring_key.    
-00006c70: 2020 2020 2020 2020 2020 2020 6f66 6673              offs
-00006c80: 7072 696e 675f 6b65 7920 3d20 6f66 6673  pring_key = offs
-00006c90: 7072 696e 675f 6b65 7973 2e70 6f70 2829  pring_keys.pop()
-00006ca0: 0a0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00006cb0: 2020 6f66 6673 7072 696e 675f 782c 206f    offspring_x, o
-00006cc0: 6666 7370 7269 6e67 5f79 203d 205f 646f  ffspring_y = _do
-00006cd0: 5f64 6973 7065 7273 616c 280a 2020 2020  _dispersal(.    
-00006ce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006cf0: 7365 6c66 2c20 7061 7265 6e74 5f6d 6964  self, parent_mid
-00006d00: 706f 696e 745f 782c 2070 6172 656e 745f  point_x, parent_
-00006d10: 6d69 6470 6f69 6e74 5f79 2c0a 2020 2020  midpoint_y,.    
-00006d20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006d30: 7365 6c66 2e64 6973 7065 7273 616c 5f64  self.dispersal_d
-00006d40: 6973 7461 6e63 655f 6469 7374 725f 7061  istance_distr_pa
-00006d50: 7261 6d31 2c0a 2020 2020 2020 2020 2020  ram1,.          
-00006d60: 2020 2020 2020 2020 2020 7365 6c66 2e64            self.d
-00006d70: 6973 7065 7273 616c 5f64 6973 7461 6e63  ispersal_distanc
-00006d80: 655f 6469 7374 725f 7061 7261 6d32 290a  e_distr_param2).
-00006d90: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00006da0: 2023 7365 7420 7468 6520 6167 6520 746f   #set the age to
-00006db0: 2030 0a20 2020 2020 2020 2020 2020 2020   0.             
-00006dc0: 2020 2061 6765 203d 2030 0a0a 2020 2020     age = 0..    
-00006dd0: 2020 2020 2020 2020 2020 2020 2373 6574              #set
-00006de0: 2074 6865 2073 6578 2063 6f72 7265 6374   the sex correct
-00006df0: 6c79 0a20 2020 2020 2020 2020 2020 2020  ly.             
-00006e00: 2020 2069 6620 7365 6c66 2e73 6578 3a0a     if self.sex:.
-00006e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006e20: 2020 2020 7365 7820 3d20 722e 6269 6e6f      sex = r.bino
-00006e30: 6d69 616c 2831 2c20 7365 6c66 2e73 6578  mial(1, self.sex
-00006e40: 5f72 6174 696f 290a 2020 2020 2020 2020  _ratio).        
-00006e50: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
-00006e60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006e70: 2020 7365 7820 3d20 4e6f 6e65 0a0a 2020    sex = None..  
-00006e80: 2020 2020 2020 2020 2020 2020 2020 2373                #s
-00006e90: 6574 2074 6865 206e 6577 5f67 656e 6f6d  et the new_genom
-00006ea0: 6520 636f 7272 6563 746c 790a 2020 2020  e correctly.    
-00006eb0: 2020 2020 2020 2020 2020 2020 6966 2073              if s
-00006ec0: 656c 662e 6765 6e5f 6172 6368 2069 7320  elf.gen_arch is 
-00006ed0: 6e6f 7420 4e6f 6e65 3a0a 2020 2020 2020  not None:.      
-00006ee0: 2020 2020 2020 2020 2020 2020 2020 6966                if
-00006ef0: 206e 6f74 2062 7572 6e3a 0a20 2020 2020   not burn:.     
+00000770: 2323 2323 230a 2320 2d2d 2d2d 2d2d 2d2d  #####.# --------
+00000780: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+00000790: 2d2d 2d2d 2d2d 2d2d 2d2d 2d23 0a23 2043  -----------#.# C
+000007a0: 4c41 5353 4553 202d 2d2d 2d2d 2d2d 2d2d  LASSES ---------
+000007b0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+000007c0: 2d2d 230a 2320 2d2d 2d2d 2d2d 2d2d 2d2d  --#.# ----------
+000007d0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+000007e0: 2d2d 2d2d 2d2d 2d2d 2d23 0a23 2323 2323  ---------#.#####
+000007f0: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+00000800: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+00000810: 230a 0a0a 2361 2073 696d 706c 6520 6c69  #...#a simple li
+00000820: 7474 6c65 2063 6c61 7373 2069 6e74 6f20  ttle class into 
+00000830: 7768 6963 6820 7061 7261 6d73 2d66 696c  which params-fil
+00000840: 6520 7061 7261 6d65 7465 7273 2061 6e64  e parameters and
+00000850: 2074 6865 6972 2076 616c 7565 7320 7769   their values wi
+00000860: 6c6c 0a23 6265 2064 756d 7065 6420 2861  ll.#be dumped (a
+00000870: 7320 6174 7472 6962 7574 6573 292e 2054  s attributes). T
+00000880: 6869 7320 636c 6173 7320 7769 6c6c 2062  his class will b
+00000890: 6520 6164 6465 6420 6173 2061 2068 6964  e added as a hid
+000008a0: 6465 6e20 6174 7472 6962 7574 6520 746f  den attribute to
+000008b0: 0a23 7468 6520 5370 6563 6965 732e 2043  .#the Species. C
+000008c0: 616e 2064 6f20 7468 6973 2073 6f20 7468  an do this so th
+000008d0: 6174 2074 6865 7365 2070 6172 616d 6574  at these paramet
+000008e0: 6572 7327 2076 616c 7565 7320 6361 6e20  ers' values can 
+000008f0: 7374 696c 6c20 6265 200a 2361 6363 6573  still be .#acces
+00000900: 7365 6420 6279 2047 656f 6e6f 6d69 6373  sed by Geonomics
+00000910: 2066 756e 6374 696f 6e73 2065 6173 696c   functions easil
+00000920: 7920 2869 6e20 6661 6374 2c20 7468 6520  y (in fact, the 
+00000930: 7361 6d65 2061 7320 6966 2074 6865 7920  same as if they 
+00000940: 7765 7265 0a23 5370 6563 6965 7320 6174  were.#Species at
+00000950: 7472 6962 7574 6573 2062 6563 6175 7365  tributes because
+00000960: 206f 6620 686f 7720 7468 6520 5370 6563   of how the Spec
+00000970: 6965 7327 205f 5f67 6574 6174 7472 5f5f  ies' __getattr__
+00000980: 206d 6574 686f 6420 6973 0a23 616c 7465   method is.#alte
+00000990: 7265 6420 6265 6c6f 7729 2c20 6275 7420  red below), but 
+000009a0: 7468 6579 2077 696c 6c20 6265 2068 6964  they will be hid
+000009b0: 6465 6e20 746f 2074 6865 2072 6567 756c  den to the regul
+000009c0: 6172 2075 7365 7220 2873 6f20 7468 6174  ar user (so that
+000009d0: 2074 6865 790a 2364 6f6e 2774 2061 7070   they.#don't app
+000009e0: 6561 7220 6173 2074 686f 7567 6820 7468  ear as though th
+000009f0: 6579 2063 6f75 6c64 2062 6520 6a75 7374  ey could be just
+00000a00: 2062 6520 6368 616e 6765 6420 696e 2061   be changed in a
+00000a10: 2063 7573 746f 6d20 6d6f 6465 6c20 7363   custom model sc
+00000a20: 7269 7074 0a23 746f 2063 6861 6e67 6520  ript.#to change 
+00000a30: 6d6f 6465 6c20 6265 6861 7669 6f72 2028  model behavior (
+00000a40: 7369 6e63 6520 736f 6d65 206f 6620 7468  since some of th
+00000a50: 656d 2061 7265 2075 7365 6420 746f 2069  em are used to i
+00000a60: 6e73 7461 6e74 6961 7465 206f 626a 6563  nstantiate objec
+00000a70: 7473 0a23 6174 2074 6865 2074 696d 6520  ts.#at the time 
+00000a80: 7468 6520 6d6f 6465 6c20 6973 206d 6164  the model is mad
+00000a90: 652c 2073 6f20 6172 6520 2762 616b 6564  e, so are 'baked
+00000aa0: 2069 6e27 2c20 7375 6368 2074 6861 7420   in', such that 
+00000ab0: 7468 6f73 6520 6f62 6a65 6374 730a 2377  those objects.#w
+00000ac0: 6f75 6c64 206e 6565 6420 746f 2062 6520  ould need to be 
+00000ad0: 7265 696e 7374 616e 7469 6174 6564 2065  reinstantiated e
+00000ae0: 6163 6820 7469 6d65 2074 6865 2070 6172  ach time the par
+00000af0: 616d 7320 636f 6e63 6572 6e65 6420 7765  ams concerned we
+00000b00: 7265 2063 6861 6e67 6564 292e 0a63 6c61  re changed)..cla
+00000b10: 7373 205f 5061 7261 6d73 5661 6c73 3a0a  ss _ParamsVals:.
+00000b20: 2020 2020 6465 6620 5f5f 696e 6974 5f5f      def __init__
+00000b30: 2873 656c 662c 2073 7070 5f6e 616d 6529  (self, spp_name)
+00000b40: 3a0a 2020 2020 2020 2020 7365 6c66 2e73  :.        self.s
+00000b50: 7070 5f6e 616d 6520 3d20 7370 705f 6e61  pp_name = spp_na
+00000b60: 6d65 0a0a 0a23 7468 6520 5370 6563 6965  me...#the Specie
+00000b70: 7320 636c 6173 730a 636c 6173 7320 5370  s class.class Sp
+00000b80: 6563 6965 7328 4f44 293a 0a20 2020 2022  ecies(OD):.    "
+00000b90: 2222 0a20 2020 2052 6570 7265 7365 6e74  "".    Represent
+00000ba0: 6174 696f 6e20 6f66 2074 6865 2074 6f74  ation of the tot
+00000bb0: 616c 2073 696d 756c 6174 6564 2070 6f70  al simulated pop
+00000bc0: 756c 6174 696f 6e20 6f66 2061 2073 7065  ulation of a spe
+00000bd0: 6369 6573 2e0a 0a20 2020 204f 7267 616e  cies...    Organ
+00000be0: 697a 6564 2061 7320 6120 636f 6c6c 6563  ized as a collec
+00000bf0: 7469 6f6e 732e 4f72 6465 7265 6444 6963  tions.OrderedDic
+00000c00: 7420 6f66 2073 6572 6961 6c20 696e 7465  t of serial inte
+00000c10: 6765 722d 6b65 7965 6420 496e 6469 7669  ger-keyed Indivi
+00000c20: 6475 616c 732e 0a20 2020 2054 686f 7365  duals..    Those
+00000c30: 2073 6572 6961 6c20 696e 6469 6365 7320   serial indices 
+00000c40: 636f 6e74 696e 7561 6c6c 7920 696e 6372  continually incr
+00000c50: 656d 656e 7420 7468 726f 7567 6820 6d6f  ement through mo
+00000c60: 6465 6c20 7469 6d65 2061 7320 6e65 770a  del time as new.
+00000c70: 2020 2020 496e 6469 7669 6475 616c 7320      Individuals 
+00000c80: 2869 2e65 2e20 6f66 6673 7072 696e 6729  (i.e. offspring)
+00000c90: 2061 7320 7072 6f64 7563 6564 2c20 7375   as produced, su
+00000ca0: 6368 2074 6861 7420 6e6f 2074 776f 2049  ch that no two I
+00000cb0: 6e64 6976 6964 7561 6c73 0a20 2020 2077  ndividuals.    w
+00000cc0: 6974 6869 6e20 7468 6520 6675 6c6c 2068  ithin the full h
+00000cd0: 6973 746f 7279 206f 6620 6120 7369 6d75  istory of a simu
+00000ce0: 6c61 7469 6f6e 2077 696c 6c20 6576 6572  lation will ever
+00000cf0: 2068 6176 6520 7468 6520 7361 6d65 2069   have the same i
+00000d00: 6e64 6578 0a20 2020 206e 756d 6265 722e  ndex.    number.
+00000d10: 0a0a 2020 2020 4265 6361 7573 6520 7468  ..    Because th
+00000d20: 6520 5370 6563 6965 7320 636c 6173 7320  e Species class 
+00000d30: 696e 6865 7269 7473 2066 726f 6d20 6063  inherits from `c
+00000d40: 6f6c 6c65 6374 696f 6e73 2e4f 7264 6572  ollections.Order
+00000d50: 6564 4469 6374 602c 0a20 2020 2049 6e64  edDict`,.    Ind
+00000d60: 6976 6964 7561 6c73 2063 616e 2062 6520  ividuals can be 
+00000d70: 696e 6465 7865 6420 6f75 7420 7573 696e  indexed out usin
+00000d80: 6720 7468 6569 7220 696e 6465 782d 6e75  g their index-nu
+00000d90: 6d62 6572 206b 6579 730a 2020 2020 2865  mber keys.    (e
+00000da0: 2e67 2e20 5370 6563 6965 735b 3c69 6478  .g. Species[<idx
+00000db0: 3e5d 292e 0a0a 2020 2020 416c 6c20 5370  >])...    All Sp
+00000dc0: 6563 6965 7320 7769 7468 696e 2061 2073  ecies within a s
+00000dd0: 696d 756c 6174 696f 6e20 6172 6520 7374  imulation are st
+00000de0: 6f72 6564 2061 7320 7365 7269 616c 2069  ored as serial i
+00000df0: 6e74 6567 6572 2d6b 6579 6564 2076 616c  nteger-keyed val
+00000e00: 7565 730a 2020 2020 7769 7468 696e 2074  ues.    within t
+00000e10: 6865 2043 6f6d 6d75 6e69 7479 2064 6963  he Community dic
+00000e20: 7420 2877 6869 6368 2069 6e20 7475 726e  t (which in turn
+00000e30: 2069 7320 7374 6f72 6564 2061 7420 7468   is stored at th
+00000e40: 6520 4d6f 6465 6c27 730a 2020 2020 276d  e Model's.    'm
+00000e50: 6f64 2e63 6f6d 6d27 2061 7474 7269 6275  od.comm' attribu
+00000e60: 7465 292e 0a0a 2020 2020 4174 7472 6962  te)...    Attrib
+00000e70: 7574 6573 0a20 2020 202d 2d2d 2d2d 2d2d  utes.    -------
+00000e80: 2d2d 2d0a 0a20 2020 2020 2020 204e 4f54  ---..        NOT
+00000e90: 453a 2046 6f72 206d 6f72 6520 6465 7461  E: For more deta
+00000ea0: 696c 2c20 7365 6520 7468 6520 646f 6375  il, see the docu
+00000eb0: 6d65 6e74 6174 696f 6e20 666f 7220 7468  mentation for th
+00000ec0: 6520 7061 7261 6d65 7465 7273 2074 6861  e parameters tha
+00000ed0: 740a 2020 2020 2020 2020 2020 2020 2020  t.              
+00000ee0: 636f 7272 6573 706f 6e64 2074 6f20 6d61  correspond to ma
+00000ef0: 6e79 206f 6620 7468 6520 666f 6c6c 6f77  ny of the follow
+00000f00: 696e 6720 6174 7472 6962 7574 6573 2e0a  ing attributes..
+00000f10: 0a20 2020 2020 2020 2062 3a0a 2020 2020  .        b:.    
+00000f20: 2020 2020 2020 2020 5468 6520 5370 6563          The Spec
+00000f30: 6965 7327 2069 6e74 7269 6e73 6963 2062  ies' intrinsic b
+00000f40: 6972 7468 2072 6174 652c 2065 7870 7265  irth rate, expre
+00000f50: 7373 6564 2061 7320 7468 6520 7072 6f62  ssed as the prob
+00000f60: 6162 696c 6974 7920 2869 6e0a 2020 2020  ability (in.    
+00000f70: 2020 2020 2020 2020 6120 4265 726e 6f75          a Bernou
+00000f80: 6c6c 6920 7261 6e64 6f6d 2064 7261 7729  lli random draw)
+00000f90: 2074 6861 7420 616e 206f 7468 6572 7769   that an otherwi
+00000fa0: 7365 2065 6c69 6769 626c 6520 706f 7465  se eligible pote
+00000fb0: 6e74 6961 6c0a 2020 2020 2020 2020 2020  ntial.          
+00000fc0: 2020 6d61 7469 6e67 2070 6169 7220 7769    mating pair wi
+00000fd0: 6c6c 2073 7563 6365 7373 6675 6c6c 7920  ll successfully 
+00000fe0: 7265 7072 6f64 7563 650a 0a20 2020 2020  reproduce..     
+00000ff0: 2020 2062 7572 6e65 643a 0a20 2020 2020     burned:.     
+00001000: 2020 2020 2020 2041 2062 6f6f 6c20 666c         A bool fl
+00001010: 6167 2069 6e64 6963 6174 696e 6720 7768  ag indicating wh
+00001020: 6574 6865 7220 6f72 206e 6f74 2074 6865  ether or not the
+00001030: 2053 7065 6369 6573 2068 6173 2062 6565   Species has bee
+00001040: 6e0a 2020 2020 2020 2020 2020 2020 6275  n.            bu
+00001050: 726e 6564 2069 6e2e 0a0a 2020 2020 2020  rned in...      
+00001060: 2020 636f 6f72 6473 3a0a 2020 2020 2020    coords:.      
+00001070: 2020 2020 2020 4120 3264 206e 756d 7079        A 2d numpy
+00001080: 2061 7272 6179 2c20 6c65 6e67 7468 2853   array, length(S
+00001090: 7065 6369 6573 2920 7820 322c 2063 6f6d  pecies) x 2, com
+000010a0: 706f 7365 6420 6f66 2063 6f6c 756d 6e73  posed of columns
+000010b0: 0a20 2020 2020 2020 2020 2020 2063 6f6e  .            con
+000010c0: 7461 696e 696e 6720 7468 6520 6375 7272  taining the curr
+000010d0: 656e 7420 7820 616e 6420 7920 6c6f 6361  ent x and y loca
+000010e0: 7469 6f6e 7320 6f66 2061 6c6c 2049 6e64  tions of all Ind
+000010f0: 6976 6964 7561 6c73 2e0a 2020 2020 2020  ividuals..      
+00001100: 2020 2020 2020 526f 7720 6f72 6465 7220        Row order 
+00001110: 6973 2065 7175 6976 616c 656e 7420 746f  is equivalent to
+00001120: 2074 6865 206f 7264 6572 206f 6620 696e   the order of in
+00001130: 6469 7669 6475 616c 7320 696e 2074 6865  dividuals in the
+00001140: 2053 7065 6369 6573 270a 2020 2020 2020   Species'.      
+00001150: 2020 2020 2020 4f72 6465 7265 6444 6963        OrderedDic
+00001160: 742e 0a0a 2020 2020 2020 2020 6365 6c6c  t...        cell
+00001170: 733a 0a20 2020 2020 2020 2020 2020 2041  s:.            A
+00001180: 2032 6420 6e75 6d70 7920 6172 7261 792c   2d numpy array,
+00001190: 206c 656e 6774 6828 5370 6563 6965 7329   length(Species)
+000011a0: 2078 2032 2c20 636f 6d70 6f73 6564 206f   x 2, composed o
+000011b0: 6620 636f 6c75 6d6e 730a 2020 2020 2020  f columns.      
+000011c0: 2020 2020 2020 636f 6e74 6169 6e69 6e67        containing
+000011d0: 2074 6865 2063 7572 7265 6e74 206a 2061   the current j a
+000011e0: 6e64 2069 2063 656c 6c20 6e75 6d62 6572  nd i cell number
+000011f0: 7320 6f66 2061 6c6c 2049 6e64 6976 6964  s of all Individ
+00001200: 7561 6c73 2e0a 2020 2020 2020 2020 2020  uals..          
+00001210: 2020 526f 7720 6f72 6465 7220 6973 2065    Row order is e
+00001220: 7175 6976 616c 656e 7420 746f 2074 6865  quivalent to the
+00001230: 206f 7264 6572 206f 6620 696e 6469 7669   order of indivi
+00001240: 6475 616c 7320 696e 2074 6865 2053 7065  duals in the Spe
+00001250: 6369 6573 270a 2020 2020 2020 2020 2020  cies'.          
+00001260: 2020 4f72 6465 7265 6444 6963 742e 2028    OrderedDict. (
+00001270: 4e6f 7465 2074 6865 206f 7264 6572 206f  Note the order o
+00001280: 6620 7468 6520 636f 6c75 6d6e 7321 2054  f the columns! T
+00001290: 6865 7365 2061 7265 206e 6f74 2069 6e20  hese are not in 
+000012a0: 7468 650a 2020 2020 2020 2020 2020 2020  the.            
+000012b0: 692c 6a20 6f72 6465 7220 7468 6174 2077  i,j order that w
+000012c0: 6f75 6c64 2062 6520 6e65 6365 7373 6172  ould be necessar
+000012d0: 7920 696e 206f 7264 6572 2075 7365 2074  y in order use t
+000012e0: 6865 6d20 746f 2073 7562 6a65 6374 2061  hem to subject a
+000012f0: 0a20 2020 2020 2020 2020 2020 204c 6179  .            Lay
+00001300: 6572 2e72 6173 7420 6f62 6a65 6374 2129  er.rast object!)
+00001310: 0a0a 2020 2020 2020 2020 6368 6f6f 7365  ..        choose
+00001320: 5f6e 6561 7265 7374 5f6d 6174 653a 0a20  _nearest_mate:. 
+00001330: 2020 2020 2020 2020 2020 2041 2062 6f6f             A boo
+00001340: 6c20 666c 6167 2069 6e64 6963 6174 696e  l flag indicatin
+00001350: 6720 7768 6574 6865 7220 6f72 206e 6f74  g whether or not
+00001360: 2049 6e64 6976 6964 7561 6c73 2073 686f   Individuals sho
+00001370: 756c 6420 0a20 2020 2020 2020 2020 2020  uld .           
+00001380: 2061 6c77 6179 7320 6368 6f6f 7365 2074   always choose t
+00001390: 6865 6972 206e 6561 7265 7374 206e 6569  heir nearest nei
+000013a0: 6768 626f 7273 2061 7320 7468 6569 7220  ghbors as their 
+000013b0: 706f 7465 6e74 6961 6c20 6d61 7465 730a  potential mates.
+000013c0: 0a20 2020 2020 2020 2064 5f6d 6178 3a0a  .        d_max:.
+000013d0: 2020 2020 2020 2020 2020 2020 5468 6520              The 
+000013e0: 6d61 7869 6d75 6d20 7072 6f62 6162 696c  maximum probabil
+000013f0: 6974 7920 6f66 2064 6561 7468 2074 6861  ity of death tha
+00001400: 7420 6361 6e20 6265 2061 7373 6967 6e65  t can be assigne
+00001410: 6420 746f 2061 6e79 0a20 2020 2020 2020  d to any.       
+00001420: 2020 2020 2049 6e64 6976 6964 7561 6c0a       Individual.
+00001430: 0a20 2020 2020 2020 2064 5f6d 696e 3a0a  .        d_min:.
+00001440: 2020 2020 2020 2020 2020 2020 5468 6520              The 
+00001450: 6d69 6e69 6d75 6d20 7072 6f62 6162 696c  minimum probabil
+00001460: 6974 7920 6f66 2064 6561 7468 2074 6861  ity of death tha
+00001470: 7420 6361 6e20 6265 2061 7373 6967 6e65  t can be assigne
+00001480: 6420 746f 2061 6e79 0a20 2020 2020 2020  d to any.       
+00001490: 2020 2020 2049 6e64 6976 6964 7561 6c0a       Individual.
+000014a0: 0a20 2020 2020 2020 2064 656e 7369 7479  .        density
+000014b0: 5f67 7269 645f 7769 6e64 6f77 5f77 6964  _grid_window_wid
+000014c0: 7468 3a0a 2020 2020 2020 2020 2020 2020  th:.            
+000014d0: 5468 6520 7769 6474 6820 6f66 2074 6865  The width of the
+000014e0: 2073 6c69 6469 6e67 2077 696e 646f 7720   sliding window 
+000014f0: 7468 6174 2069 7320 7573 6564 2062 7920  that is used by 
+00001500: 7468 6520 5370 6563 6965 7327 0a20 2020  the Species'.   
+00001510: 2020 2020 2020 2020 205f 4465 6e73 6974           _Densit
+00001520: 7947 7269 6453 7461 636b 206f 626a 6563  yGridStack objec
+00001530: 7420 7768 656e 2065 7374 696d 6174 696e  t when estimatin
+00001540: 6720 7468 6520 5370 6563 6965 7327 0a20  g the Species'. 
+00001550: 2020 2020 2020 2020 2020 2070 6f70 756c             popul
+00001560: 6174 696f 6e2d 6465 6e73 6974 7920 6172  ation-density ar
+00001570: 7261 792e 0a0a 2020 2020 2020 2020 6469  ray...        di
+00001580: 7265 6374 696f 6e5f 6469 7374 725f 5b6d  rection_distr_[m
+00001590: 752f 6b61 7070 615d 3a0a 2020 2020 2020  u/kappa]:.      
+000015a0: 2020 2020 2020 5468 6520 245c 6d75 2420        The $\mu$ 
+000015b0: 616e 6420 245c 6b61 7070 6124 2070 6172  and $\kappa$ par
+000015c0: 616d 6574 6572 7320 6f66 2074 6865 2076  ameters of the v
+000015d0: 6f6e 204d 6973 6573 2064 6973 7472 6962  on Mises distrib
+000015e0: 7574 696f 6e0a 2020 2020 2020 2020 2020  ution.          
+000015f0: 2020 6672 6f6d 2077 6869 6368 2074 6865    from which the
+00001600: 2064 6972 6563 7469 6f6e 2063 6f6d 706f   direction compo
+00001610: 6e65 6e74 7320 6f66 2049 6e64 6976 6964  nents of Individ
+00001620: 7561 6c73 2720 6d6f 7665 6d65 6e74 0a20  uals' movement. 
+00001630: 2020 2020 2020 2020 2020 2076 6563 746f             vecto
+00001640: 7273 2077 696c 6c20 6265 2064 7261 776e  rs will be drawn
+00001650: 2e20 2854 6865 7365 2070 6172 616d 6574  . (These paramet
+00001660: 6572 7320 6172 6520 6f6e 6c79 2075 7469  ers are only uti
+00001670: 6c69 7a65 6420 6966 2074 6865 0a20 2020  lized if the.   
+00001680: 2020 2020 2020 2020 2053 7065 6369 6573           Species
+00001690: 2069 7320 7061 7261 6d65 7465 7269 7a65   is parameterize
+000016a0: 6420 7769 7468 6f75 7420 6120 5f4d 6f76  d without a _Mov
+000016b0: 656d 656e 7453 7572 6661 6365 2e29 0a0a  ementSurface.)..
+000016c0: 2020 2020 2020 2020 6469 7370 6572 7361          dispersa
+000016d0: 6c5f 6469 7374 616e 6365 5f64 6973 7472  l_distance_distr
+000016e0: 3a0a 2020 2020 2020 2020 2020 2020 4120  :.            A 
+000016f0: 7374 7269 6e67 2069 6e64 6963 6174 696e  string indicatin
+00001700: 6720 7768 6963 6820 6469 7374 7269 6275  g which distribu
+00001710: 7469 6f6e 2074 6f20 7573 6520 6173 0a20  tion to use as. 
+00001720: 2020 2020 2020 2020 2020 2074 6865 2053             the S
+00001730: 7065 6369 6573 2720 6469 7370 6572 7361  pecies' dispersa
+00001740: 6c2d 6469 7374 616e 6365 2064 6973 7472  l-distance distr
+00001750: 6962 7574 696f 6e20 2827 6c6f 676e 6f72  ibution ('lognor
+00001760: 6d61 6c27 2c0a 2020 2020 2020 2020 2020  mal',.          
+00001770: 2020 2777 616c 6427 2c20 6f72 2027 6c65    'wald', or 'le
+00001780: 7679 2729 0a0a 2020 2020 2020 2020 6469  vy')..        di
+00001790: 7370 6572 7361 6c5f 6469 7374 616e 6365  spersal_distance
+000017a0: 5f64 6973 7472 5f70 6172 616d 5b31 2f32  _distr_param[1/2
+000017b0: 5d3a 0a20 2020 2020 2020 2020 2020 2054  ]:.            T
+000017c0: 6865 2076 616c 7565 7320 6f66 2074 6865  he values of the
+000017d0: 2066 6972 7374 2061 6e64 2073 6563 6f6e   first and secon
+000017e0: 6420 7061 7261 6d65 7465 7273 206f 6620  d parameters of 
+000017f0: 7468 6520 5370 6563 6965 7327 0a20 2020  the Species'.   
+00001800: 2020 2020 2020 2020 2064 6973 7065 7273           dispers
+00001810: 616c 2d64 6973 7461 6e63 6520 6469 7374  al-distance dist
+00001820: 7269 6275 7469 6f6e 2028 6e61 6d65 6420  ribution (named 
+00001830: 6162 7374 7261 6374 6c79 206c 696b 6520  abstractly like 
+00001840: 7468 6973 0a20 2020 2020 2020 2020 2020  this.           
+00001850: 2062 6563 6175 7365 2074 6865 7920 636f   because they co
+00001860: 756c 6420 7061 7261 6d65 7465 7269 7a65  uld parameterize
+00001870: 2061 204c 6f67 6e6f 726d 616c 2c20 5761   a Lognormal, Wa
+00001880: 6c64 2c0a 2020 2020 2020 2020 2020 2020  ld,.            
+00001890: 6f72 204c 6576 7920 6469 7374 7269 6275  or Levy distribu
+000018a0: 7469 6f6e 2c20 6465 7065 6e64 696e 6720  tion, depending 
+000018b0: 6f6e 2074 6865 2076 616c 7565 206f 660a  on the value of.
+000018c0: 2020 2020 2020 2020 2020 2020 5370 6563              Spec
+000018d0: 6965 732e 6469 7370 6572 7361 6c5f 6469  ies.dispersal_di
+000018e0: 7374 616e 6365 5f64 6973 7472 290a 0a20  stance_distr).. 
+000018f0: 2020 2020 2020 2065 7874 696e 6374 3a0a         extinct:.
+00001900: 2020 2020 2020 2020 2020 2020 4120 626f              A bo
+00001910: 6f6c 2066 6c61 6720 696e 6469 6361 7469  ol flag indicati
+00001920: 6e67 2077 6865 7468 6572 206f 7220 6e6f  ng whether or no
+00001930: 7420 7468 6520 5370 6563 6965 7320 6861  t the Species ha
+00001940: 7320 676f 6e65 2065 7874 696e 6374 2e0a  s gone extinct..
+00001950: 0a20 2020 2020 2020 2069 6478 3a0a 2020  .        idx:.  
+00001960: 2020 2020 2020 2020 2020 496e 6465 7820            Index 
+00001970: 6e75 6d62 6572 206f 6620 7468 6520 5370  number of the Sp
+00001980: 6563 6965 7320 2869 2e65 2e20 6974 7320  ecies (i.e. its 
+00001990: 6b65 7920 7769 7468 696e 2074 6865 2043  key within the C
+000019a0: 6f6d 6d75 6e69 7479 2064 6963 7429 0a0a  ommunity dict)..
+000019b0: 2020 2020 2020 2020 696e 7665 7273 655f          inverse_
+000019c0: 6469 7374 5f6d 6174 696e 673a 0a20 2020  dist_mating:.   
+000019d0: 2020 2020 2020 2020 2041 2062 6f6f 6c20           A bool 
+000019e0: 666c 6167 2069 6e64 6963 6174 696e 6720  flag indicating 
+000019f0: 7768 6574 6865 7220 6f72 206e 6f74 2074  whether or not t
+00001a00: 6865 2070 726f 6261 6269 6c69 7469 6573  he probabilities
+00001a10: 2074 6861 7420 616e 0a20 2020 2020 2020   that an.       
+00001a20: 2020 2020 2049 6e64 6976 6964 7561 6c20       Individual 
+00001a30: 6368 6f6f 7365 7320 616e 7920 6f66 2069  chooses any of i
+00001a40: 7473 206e 6569 6768 626f 7273 2028 692e  ts neighbors (i.
+00001a50: 652e 206f 7468 6572 2049 6e64 6976 6964  e. other Individ
+00001a60: 7561 6c73 0a20 2020 2020 2020 2020 2020  uals.           
+00001a70: 2077 6974 6869 6e20 6053 7065 6369 6573   within `Species
+00001a80: 2e6d 6174 696e 675f 7261 6469 7573 6020  .mating_radius` 
+00001a90: 6469 7374 616e 6365 206f 6620 7468 656d  distance of them
+00001aa0: 2920 6173 2074 6865 6972 0a20 2020 2020  ) as their.     
+00001ab0: 2020 2020 2020 2070 6f74 656e 7469 616c         potential
+00001ac0: 206d 6174 6573 2073 686f 756c 6420 7661   mates should va
+00001ad0: 7279 2069 6e76 6572 7365 6c79 2077 6974  ry inversely wit
+00001ae0: 6820 7468 6520 6469 7374 616e 6365 7320  h the distances 
+00001af0: 6265 7477 6565 6e0a 2020 2020 2020 2020  between.        
+00001b00: 2020 2020 7468 6520 496e 6469 7669 6475      the Individu
+00001b10: 616c 2061 6e64 2065 6163 6820 6f66 2074  al and each of t
+00001b20: 686f 7365 206e 6569 6768 626f 7273 2e20  hose neighbors. 
+00001b30: 4966 2046 616c 7365 2c20 616e 7920 6e65  If False, any ne
+00001b40: 6967 6862 6f72 0a20 2020 2020 2020 2020  ighbor.         
+00001b50: 2020 2063 6f75 6c64 2062 6520 6368 6f73     could be chos
+00001b60: 656e 2077 6974 6820 6571 7561 6c20 7072  en with equal pr
+00001b70: 6f62 6162 696c 6974 792e 200a 0a20 2020  obability. ..   
+00001b80: 2020 2020 2067 656e 5f61 7263 683a 0a20       gen_arch:. 
+00001b90: 2020 2020 2020 2020 2020 2054 6865 2053             The S
+00001ba0: 7065 6369 6573 2720 6765 6e6f 6d69 6320  pecies' genomic 
+00001bb0: 6172 6368 6974 6563 7475 7265 2028 6173  architecture (as
+00001bc0: 2061 2047 656e 6f6d 6963 4172 6368 6974   a GenomicArchit
+00001bd0: 6563 7475 7265 206f 626a 6563 7429 0a0a  ecture object)..
+00001be0: 2020 2020 2020 2020 4b3a 0a20 2020 2020          K:.     
+00001bf0: 2020 2020 2020 2054 6865 2053 7065 6369         The Speci
+00001c00: 6573 2720 6375 7272 656e 7420 6361 7272  es' current carr
+00001c10: 7969 6e67 2d63 6170 6163 6974 7920 7261  ying-capacity ra
+00001c20: 7374 6572 2e20 5468 6973 2069 7320 6120  ster. This is a 
+00001c30: 6e75 6d70 7920 6172 7261 790a 2020 2020  numpy array.    
+00001c40: 2020 2020 2020 2020 6f66 2073 6861 7065          of shape
+00001c50: 2060 4c61 6e64 7363 6170 652e 6469 6d60   `Landscape.dim`
+00001c60: 2c20 7769 7468 2065 6163 6820 7661 6c75  , with each valu
+00001c70: 6520 0a20 2020 2020 2020 2020 2020 2069  e .            i
+00001c80: 6e64 6963 6174 696e 6720 7468 6520 6361  ndicating the ca
+00001c90: 7272 7969 6e67 2063 6170 6163 6974 7920  rrying capacity 
+00001ca0: 6f66 2074 6861 7420 6365 6c6c 2069 6e20  of that cell in 
+00001cb0: 7468 6520 4c61 6e64 7363 6170 652e 0a20  the Landscape.. 
+00001cc0: 2020 2020 2020 2020 2020 2049 7420 6973             It is
+00001cd0: 2067 656e 6572 6174 6564 2061 7320 7468   generated as th
+00001ce0: 6520 7072 6f64 7563 7420 6f66 2074 6865  e product of the
+00001cf0: 2053 7065 6369 6573 2720 4b5f 6661 6374   Species' K_fact
+00001d00: 6f72 0a20 2020 2020 2020 2020 2020 2061  or.            a
+00001d10: 6e64 2074 6865 2072 6173 7465 7220 6172  nd the raster ar
+00001d20: 7261 7920 6f66 2074 6865 204c 6179 6572  ray of the Layer
+00001d30: 2070 726f 7669 6465 6420 6173 2074 6865   provided as the
+00001d40: 2062 6173 6973 0a20 2020 2020 2020 2020   basis.         
+00001d50: 2020 206f 6620 7468 6520 5370 6563 6965     of the Specie
+00001d60: 7327 2063 6172 7279 696e 6720 6361 7061  s' carrying capa
+00001d70: 6369 7479 0a20 2020 2020 2020 2020 2020  city.           
+00001d80: 2028 692e 652e 2060 7370 702e 4b5f 6661   (i.e. `spp.K_fa
+00001d90: 6374 6f72 202a 206d 6f64 2e6c 616e 645b  ctor * mod.land[
+00001da0: 7370 702e 4b5f 6c61 7965 725d 2e72 6173  spp.K_layer].ras
+00001db0: 7460 292e 0a20 2020 2020 2020 2020 2020  t`)..           
+00001dc0: 2054 6869 7320 6172 7261 7920 7769 6c6c   This array will
+00001dd0: 2063 6861 6e67 6520 6475 7269 6e67 2061   change during a
+00001de0: 2073 696d 756c 6174 696f 6e20 6966 2074   simulation if t
+00001df0: 6865 2063 6f72 7265 7370 6f6e 6469 6e67  he corresponding
+00001e00: 0a20 2020 2020 2020 2020 2020 204c 6179  .            Lay
+00001e10: 6572 2069 7320 7061 7261 6d65 7465 7269  er is parameteri
+00001e20: 7a65 6420 746f 2075 6e64 6572 676f 2065  zed to undergo e
+00001e30: 6e76 6972 6f6e 6d65 6e74 616c 2063 6861  nvironmental cha
+00001e40: 6e67 652e 0a0a 2020 2020 2020 2020 4b5f  nge...        K_
+00001e50: 6661 6374 6f72 3a0a 2020 2020 2020 2020  factor:.        
+00001e60: 2020 2020 5468 6520 6661 6374 6f72 2062      The factor b
+00001e70: 7920 7768 6963 6820 746f 206d 756c 7469  y which to multi
+00001e80: 706c 7920 7468 6520 5370 6563 6965 7327  ply the Species'
+00001e90: 2063 686f 7365 6e0a 2020 2020 2020 2020   chosen.        
+00001ea0: 2020 2020 6361 7272 7969 6e67 2063 6170      carrying cap
+00001eb0: 6163 6974 7920 4c61 7965 7220 286d 6f64  acity Layer (mod
+00001ec0: 2e6c 616e 645b 7370 702e 4b5f 6c61 7965  .land[spp.K_laye
+00001ed0: 725d 2920 696e 206f 7264 6572 2074 6f0a  r]) in order to.
+00001ee0: 2020 2020 2020 2020 2020 2020 6765 6e65              gene
+00001ef0: 7261 7465 2074 6865 2053 7065 6369 6573  rate the Species
+00001f00: 2720 6361 7272 7969 6e67 2d63 6170 6163  ' carrying-capac
+00001f10: 6974 7920 7261 7374 6572 2028 7370 702e  ity raster (spp.
+00001f20: 4b29 2e0a 0a20 2020 2020 2020 204b 5f6c  K)...        K_l
+00001f30: 6179 6572 3a0a 2020 2020 2020 2020 2020  ayer:.          
+00001f40: 2020 5468 6520 696e 6465 7820 6e75 6d62    The index numb
+00001f50: 6572 206f 6620 7468 6520 4c61 7965 7220  er of the Layer 
+00001f60: 7768 6f73 6520 7261 7374 6572 2077 696c  whose raster wil
+00001f70: 6c20 6265 206d 756c 7469 706c 6965 6420  l be multiplied 
+00001f80: 6279 0a20 2020 2020 2020 2020 2020 2073  by.            s
+00001f90: 7070 2e4b 5f66 6163 746f 7220 696e 206f  pp.K_factor in o
+00001fa0: 7264 6572 2074 6f20 6765 6e65 7261 7465  rder to generate
+00001fb0: 2074 6865 2053 7065 6369 6573 2720 6361   the Species' ca
+00001fc0: 7272 7969 6e67 2d63 6170 6163 6974 790a  rrying-capacity.
+00001fd0: 2020 2020 2020 2020 2020 2020 7261 7374              rast
+00001fe0: 6572 2028 7370 702e 4b29 2e0a 0a20 2020  er (spp.K)...   
+00001ff0: 2020 2020 206d 6174 696e 675f 7261 6469       mating_radi
+00002000: 7573 3a0a 2020 2020 2020 2020 2020 2020  us:.            
+00002010: 5468 6520 6d61 7469 6e67 2072 6164 6975  The mating radiu
+00002020: 732c 2065 7870 7265 7373 6564 2069 6e20  s, expressed in 
+00002030: 4c61 6e64 7363 6170 6520 6365 6c6c 2077  Landscape cell w
+00002040: 6964 7468 732c 2066 726f 6d20 7769 7468  idths, from with
+00002050: 696e 0a20 2020 2020 2020 2020 2020 2077  in.            w
+00002060: 6869 6368 2061 6e20 496e 6469 7669 6475  hich an Individu
+00002070: 616c 2063 616e 2063 686f 6f73 6520 6120  al can choose a 
+00002080: 706f 7465 6e74 6961 6c20 6d61 7465 2e20  potential mate. 
+00002090: 5468 6973 2077 696c 6c20 6f6e 6c79 2062  This will only b
+000020a0: 650a 2020 2020 2020 2020 2020 2020 7573  e.            us
+000020b0: 6564 2069 6620 6053 7065 6369 6573 2e63  ed if `Species.c
+000020c0: 686f 6f73 655f 6e65 6172 6573 745f 6d61  hoose_nearest_ma
+000020d0: 7465 6020 6973 2046 616c 7365 2e0a 0a20  te` is False... 
+000020e0: 2020 2020 2020 206d 6178 5f61 6765 3a0a         max_age:.
+000020f0: 2020 2020 2020 2020 2020 2020 5468 6520              The 
+00002100: 6d61 7869 6d75 6d20 6167 652c 2069 6e20  maximum age, in 
+00002110: 7469 6d65 2073 7465 7073 2c20 7468 6174  time steps, that
+00002120: 2061 6e79 2049 6e64 6976 6964 7561 6c20   any Individual 
+00002130: 6f66 2074 6865 2053 7065 6369 6573 0a20  of the Species. 
+00002140: 2020 2020 2020 2020 2020 206d 6179 2072             may r
+00002150: 6561 6368 2028 6166 7465 7220 7768 6963  each (after whic
+00002160: 6820 706f 696e 7420 7468 6579 2077 696c  h point they wil
+00002170: 6c20 6265 2063 756c 6c65 6420 6672 6f6d  l be culled from
+00002180: 2074 6865 0a20 2020 2020 2020 2020 2020   the.           
+00002190: 2070 6f70 756c 6174 696f 6e20 6279 2069   population by i
+000021a0: 6d70 6f73 6564 206d 6f72 7461 6c69 7479  mposed mortality
+000021b0: 290a 0a20 2020 2020 2020 206d 6178 5f69  )..        max_i
+000021c0: 6e64 5f69 6478 3a0a 2020 2020 2020 2020  nd_idx:.        
+000021d0: 2020 2020 5468 6520 6d61 7869 6d75 6d20      The maximum 
+000021e0: 496e 6469 7669 6475 616c 2069 6e64 6578  Individual index
+000021f0: 206e 756d 6265 7220 7468 6174 2068 6173   number that has
+00002200: 2063 7572 7265 6e74 6c79 2062 6565 6e20   currently been 
+00002210: 7573 6564 0a0a 2020 2020 2020 2020 6d6f  used..        mo
+00002220: 7665 3a0a 2020 2020 2020 2020 2020 2020  ve:.            
+00002230: 4120 626f 6f6c 2066 6c61 6720 696e 6469  A bool flag indi
+00002240: 6361 7469 6e67 2077 6865 7468 6572 206f  cating whether o
+00002250: 7220 6e6f 7420 496e 6469 7669 6475 616c  r not Individual
+00002260: 7320 6f66 2074 6865 2053 7065 6369 6573  s of the Species
+00002270: 0a20 2020 2020 2020 2020 2020 2063 616e  .            can
+00002280: 206d 6f76 6520 6166 7465 7220 7468 6569   move after thei
+00002290: 7220 696e 6974 6961 6c20 6469 7370 6572  r initial disper
+000022a0: 7361 6c20 6173 206f 6666 7370 7269 6e67  sal as offspring
+000022b0: 2e0a 0a20 2020 2020 2020 206d 6f76 656d  ...        movem
+000022c0: 656e 745f 6469 7374 616e 6365 5f64 6973  ent_distance_dis
+000022d0: 7472 3a0a 2020 2020 2020 2020 2020 2020  tr:.            
+000022e0: 4120 7374 7269 6e67 2069 6e64 6963 6174  A string indicat
+000022f0: 696e 6720 7768 6963 6820 6469 7374 7269  ing which distri
+00002300: 6275 7469 6f6e 2074 6f20 7573 6520 6173  bution to use as
+00002310: 0a20 2020 2020 2020 2020 2020 2074 6865  .            the
+00002320: 2053 7065 6369 6573 2720 6d6f 7665 6d65   Species' moveme
+00002330: 6e74 2d64 6973 7461 6e63 6520 6469 7374  nt-distance dist
+00002340: 7269 6275 7469 6f6e 2028 276c 6f67 6e6f  ribution ('logno
+00002350: 726d 616c 272c 0a20 2020 2020 2020 2020  rmal',.         
+00002360: 2020 2027 7761 6c64 272c 206f 7220 276c     'wald', or 'l
+00002370: 6576 7927 290a 0a20 2020 2020 2020 206d  evy')..        m
+00002380: 6f76 656d 656e 745f 6469 7374 616e 6365  ovement_distance
+00002390: 5f64 6973 7472 5f70 6172 616d 5b31 2f32  _distr_param[1/2
+000023a0: 5d3a 0a20 2020 2020 2020 2020 2020 2054  ]:.            T
+000023b0: 6865 2076 616c 7565 7320 6f66 2074 6865  he values of the
+000023c0: 2066 6972 7374 2061 6e64 2073 6563 6f6e   first and secon
+000023d0: 6420 7061 7261 6d65 7465 7273 206f 6620  d parameters of 
+000023e0: 7468 6520 5370 6563 6965 7327 0a20 2020  the Species'.   
+000023f0: 2020 2020 2020 2020 206d 6f76 656d 656e           movemen
+00002400: 742d 6469 7374 616e 6365 2064 6973 7472  t-distance distr
+00002410: 6962 7574 696f 6e20 286e 616d 6564 2061  ibution (named a
+00002420: 6273 7472 6163 746c 7920 6c69 6b65 2074  bstractly like t
+00002430: 6869 730a 2020 2020 2020 2020 2020 2020  his.            
+00002440: 6265 6361 7573 6520 7468 6579 2063 6f75  because they cou
+00002450: 6c64 2070 6172 616d 6574 6572 697a 6520  ld parameterize 
+00002460: 6120 4c6f 676e 6f72 6d61 6c2c 2057 616c  a Lognormal, Wal
+00002470: 642c 0a20 2020 2020 2020 2020 2020 206f  d,.            o
+00002480: 7220 4c65 7679 2064 6973 7472 6962 7574  r Levy distribut
+00002490: 696f 6e2c 2064 6570 656e 6469 6e67 206f  ion, depending o
+000024a0: 6e20 7468 6520 7661 6c75 6520 6f66 0a20  n the value of. 
+000024b0: 2020 2020 2020 2020 2020 2053 7065 6369             Speci
+000024c0: 6573 2e6d 6f76 656d 656e 745f 6469 7374  es.movement_dist
+000024d0: 616e 6365 5f64 6973 7472 290a 0a20 2020  ance_distr)..   
+000024e0: 2020 2020 206d 7574 6174 653a 0a20 2020       mutate:.   
+000024f0: 2020 2020 2020 2020 2041 2062 6f6f 6c20           A bool 
+00002500: 666c 6167 2069 6e64 6963 6174 696e 6720  flag indicating 
+00002510: 7768 6574 6865 7220 6f72 206e 6f74 2074  whether or not t
+00002520: 6865 2053 7065 6369 6573 2720 7369 6d75  he Species' simu
+00002530: 6c61 7469 6f6e 0a20 2020 2020 2020 2020  lation.         
+00002540: 2020 2069 6e76 6f6c 7665 7320 6d75 7461     involves muta
+00002550: 7469 6f6e 2e0a 0a20 2020 2020 2020 206d  tion...        m
+00002560: 7574 5f6c 6f67 3a0a 2020 2020 2020 2020  ut_log:.        
+00002570: 2020 2020 4120 626f 6f6c 2066 6c61 6720      A bool flag 
+00002580: 696e 6469 6361 7469 6e67 2077 6865 7468  indicating wheth
+00002590: 6572 206f 7220 6e6f 7420 6120 6d75 7461  er or not a muta
+000025a0: 7469 6f6e 206c 6f67 2073 686f 756c 6420  tion log should 
+000025b0: 6265 0a20 2020 2020 2020 2020 2020 2077  be.            w
+000025c0: 7269 7474 656e 2066 6f72 2074 6865 2053  ritten for the S
+000025d0: 7065 6369 6573 2e0a 0a20 2020 2020 2020  pecies...       
+000025e0: 204e 3a0a 2020 2020 2020 2020 2020 2020   N:.            
+000025f0: 5468 6520 5370 6563 6965 7327 2063 7572  The Species' cur
+00002600: 7265 6e74 2070 6f70 756c 6174 696f 6e2d  rent population-
+00002610: 6465 6e73 6974 7920 7261 7374 6572 2e0a  density raster..
+00002620: 2020 2020 2020 2020 2020 2020 5468 6973              This
+00002630: 2069 7320 6120 6e75 6d70 7920 6172 7261   is a numpy arra
+00002640: 7920 6f66 2073 6861 7065 2060 4c61 6e64  y of shape `Land
+00002650: 7363 6170 652e 6469 6d60 2c20 7769 7468  scape.dim`, with
+00002660: 2065 6163 6820 7661 6c75 650a 2020 2020   each value.    
+00002670: 2020 2020 2020 2020 696e 6469 6361 7469          indicati
+00002680: 6e67 2074 6865 2070 6f70 756c 6174 696f  ng the populatio
+00002690: 6e20 6465 6e73 6974 7920 6f66 2074 6861  n density of tha
+000026a0: 7420 6365 6c6c 2069 6e20 7468 6520 4c61  t cell in the La
+000026b0: 6e64 7363 6170 652c 0a20 2020 2020 2020  ndscape,.       
+000026c0: 2020 2020 2061 7320 6573 7469 6d61 7465       as estimate
+000026d0: 6420 6279 2074 6865 2053 7065 6369 6573  d by the Species
+000026e0: 2720 5f44 656e 7369 7479 4772 6964 5374  ' _DensityGridSt
+000026f0: 6163 6b20 696e 7374 616e 6365 2e0a 0a20  ack instance... 
+00002700: 2020 2020 2020 204e 743a 0a20 2020 2020         Nt:.     
+00002710: 2020 2020 2020 2041 2063 6872 6f6e 6f6c         A chronol
+00002720: 6f67 6963 616c 6c79 206f 7264 6572 6564  ogically ordered
+00002730: 206c 6973 742c 2073 7461 7274 696e 6720   list, starting 
+00002740: 6672 6f6d 2074 6865 2066 6972 7374 2074  from the first t
+00002750: 696d 6520 7374 6570 206f 660a 2020 2020  ime step of.    
+00002760: 2020 2020 2020 2020 7468 6520 6275 726e          the burn
+00002770: 2d69 6e2c 2063 6f6e 7461 696e 696e 6720  -in, containing 
+00002780: 7468 6520 5370 6563 6965 7327 2074 6f74  the Species' tot
+00002790: 616c 2070 6f70 756c 6174 696f 6e20 7369  al population si
+000027a0: 7a65 2061 7420 6561 6368 0a20 2020 2020  ze at each.     
+000027b0: 2020 2020 2020 2074 696d 6520 7374 6570         time step
+000027c0: 2e0a 0a20 2020 2020 2020 206e 5f62 6972  ...        n_bir
+000027d0: 7468 733a 0a20 2020 2020 2020 2020 2020  ths:.           
+000027e0: 2041 2063 6872 6f6e 6f6c 6f67 6963 616c   A chronological
+000027f0: 6c79 206f 7264 6572 6564 206c 6973 742c  ly ordered list,
+00002800: 2073 7461 7274 696e 6720 6672 6f6d 2074   starting from t
+00002810: 6865 2066 6972 7374 2074 696d 6520 7374  he first time st
+00002820: 6570 206f 660a 2020 2020 2020 2020 2020  ep of.          
+00002830: 2020 7468 6520 6275 726e 2d69 6e2c 2063    the burn-in, c
+00002840: 6f6e 7461 696e 696e 6720 7468 6520 5370  ontaining the Sp
+00002850: 6563 6965 7327 206e 756d 6265 7220 6f66  ecies' number of
+00002860: 2062 6972 7468 7320 6174 2065 6163 680a   births at each.
+00002870: 2020 2020 2020 2020 2020 2020 7469 6d65              time
+00002880: 2073 7465 702e 0a0a 2020 2020 2020 2020   step...        
+00002890: 6e5f 6269 7274 6873 5f64 6973 7472 5f6c  n_births_distr_l
+000028a0: 616d 6264 613a 0a20 2020 2020 2020 2020  ambda:.         
+000028b0: 2020 2049 6620 6053 7065 6369 6573 2e6e     If `Species.n
+000028c0: 5f62 6972 7468 735f 6669 7865 6460 2069  _births_fixed` i
+000028d0: 7320 4661 6c73 652c 2074 6865 6e20 7468  s False, then th
+000028e0: 6973 2076 616c 7565 2073 6572 7665 7320  is value serves 
+000028f0: 6173 2074 6865 0a20 2020 2020 2020 2020  as the.         
+00002900: 2020 2024 5c6c 616d 6264 6124 2070 6172     $\lambda$ par
+00002910: 616d 6574 6572 206f 6620 7468 6520 506f  ameter of the Po
+00002920: 6973 736f 6e20 6469 7374 7269 6275 7469  isson distributi
+00002930: 6f6e 2066 726f 6d20 7768 6963 6820 6973  on from which is
+00002940: 2064 7261 776e 0a20 2020 2020 2020 2020   drawn.         
+00002950: 2020 2074 6865 206e 756d 6265 7220 6f66     the number of
+00002960: 206f 6666 7370 7269 6e67 2061 206d 6174   offspring a mat
+00002970: 696e 6720 7061 6972 2077 696c 6c20 7072  ing pair will pr
+00002980: 6f64 7563 652e 0a20 2020 2020 2020 2020  oduce..         
+00002990: 2020 2049 6620 6053 7065 6369 6573 2e6e     If `Species.n
+000029a0: 5f62 6972 7468 735f 6669 7865 6460 2069  _births_fixed` i
+000029b0: 7320 5472 7565 2c20 7468 656e 2074 6869  s True, then thi
+000029c0: 7320 6973 2074 6865 2066 6978 6564 206e  s is the fixed n
+000029d0: 756d 6265 720a 2020 2020 2020 2020 2020  umber.          
+000029e0: 2020 6f66 206f 6666 7370 7269 6e67 2074    of offspring t
+000029f0: 6861 7420 6561 6368 2073 7563 6365 7373  hat each success
+00002a00: 6675 6c20 6d61 7469 6e67 2070 6169 7220  ful mating pair 
+00002a10: 7769 6c6c 2070 726f 6475 6365 2e0a 0a20  will produce... 
+00002a20: 2020 2020 2020 206e 5f62 6972 7468 735f         n_births_
+00002a30: 6669 7865 643a 0a20 2020 2020 2020 2020  fixed:.         
+00002a40: 2020 2041 2062 6f6f 6c20 666c 6167 2069     A bool flag i
+00002a50: 6e64 6963 6174 696e 6720 7768 6574 6865  ndicating whethe
+00002a60: 7220 6f72 206e 6f74 2074 6865 206e 756d  r or not the num
+00002a70: 6265 7220 6f66 2062 6972 7468 7320 7368  ber of births sh
+00002a80: 6f75 6c64 0a20 2020 2020 2020 2020 2020  ould.           
+00002a90: 2062 6520 6669 7865 6420 6174 2060 5370   be fixed at `Sp
+00002aa0: 6563 6965 732e 6e5f 6269 7274 6873 5f64  ecies.n_births_d
+00002ab0: 6973 7472 5f6c 616d 6264 6160 0a0a 2020  istr_lambda`..  
+00002ac0: 2020 2020 2020 6e5f 6465 6174 6873 3a0a        n_deaths:.
+00002ad0: 2020 2020 2020 2020 2020 2020 4120 6368              A ch
+00002ae0: 726f 6e6f 6c6f 6769 6361 6c6c 7920 6f72  ronologically or
+00002af0: 6465 7265 6420 6c69 7374 2c20 7374 6172  dered list, star
+00002b00: 7469 6e67 2066 726f 6d20 7468 6520 6669  ting from the fi
+00002b10: 7273 7420 7469 6d65 2073 7465 7020 6f66  rst time step of
+00002b20: 0a20 2020 2020 2020 2020 2020 2074 6865  .            the
+00002b30: 2062 7572 6e2d 696e 2c20 636f 6e74 6169   burn-in, contai
+00002b40: 6e69 6e67 2074 6865 2053 7065 6369 6573  ning the Species
+00002b50: 2720 6e75 6d62 6572 206f 6620 6465 6174  ' number of deat
+00002b60: 6873 2061 7420 6561 6368 0a20 2020 2020  hs at each.     
+00002b70: 2020 2020 2020 2074 696d 6520 7374 6570         time step
+00002b80: 2e0a 0a20 2020 2020 2020 206e 616d 653a  ...        name:
+00002b90: 0a20 2020 2020 2020 2020 2020 2054 6865  .            The
+00002ba0: 2073 7472 696e 6720 6e61 6d65 206f 6620   string name of 
+00002bb0: 7468 6520 5370 6563 6965 730a 0a20 2020  the Species..   
+00002bc0: 2020 2020 2052 3a0a 2020 2020 2020 2020       R:.        
+00002bd0: 2020 2020 5468 6520 696e 7472 696e 7369      The intrinsi
+00002be0: 6320 6772 6f77 7468 2072 6174 6520 6f66  c growth rate of
+00002bf0: 2074 6865 2053 7065 6369 6573 2c20 746f   the Species, to
+00002c00: 2062 6520 7573 6564 2061 7320 2752 2720   be used as 'R' 
+00002c10: 696e 2074 6865 0a20 2020 2020 2020 2020  in the.         
+00002c20: 2020 2073 746f 6368 6173 7469 632c 2073     stochastic, s
+00002c30: 7061 7469 616c 697a 6564 206c 6f67 6973  patialized logis
+00002c40: 7469 6320 6772 6f77 7468 206d 6f64 656c  tic growth model
+00002c50: 2074 6861 7420 636f 6e74 726f 6c73 0a20   that controls. 
+00002c60: 2020 2020 2020 2020 2020 2070 6f70 756c             popul
+00002c70: 6174 696f 6e20 6479 6e61 6d69 6373 0a0a  ation dynamics..
+00002c80: 2020 2020 2020 2020 7265 7072 6f5f 6167          repro_ag
+00002c90: 653a 0a20 2020 2020 2020 2020 2020 2054  e:.            T
+00002ca0: 6865 2061 6765 2c20 696e 2074 696d 6520  he age, in time 
+00002cb0: 7374 6570 732c 206f 6620 6669 7273 7420  steps, of first 
+00002cc0: 7265 7072 6f64 7563 7469 6f6e 0a0a 2020  reproduction..  
+00002cd0: 2020 2020 2020 7365 6c65 6374 696f 6e3a        selection:
+00002ce0: 0a20 2020 2020 2020 2020 2020 2041 2062  .            A b
+00002cf0: 6f6f 6c20 666c 6167 2069 6e64 6963 6174  ool flag indicat
+00002d00: 696e 6720 7768 6574 6865 7220 6f72 206e  ing whether or n
+00002d10: 6f74 2074 6865 2053 7065 6369 6573 2720  ot the Species' 
+00002d20: 7369 6d75 6c61 7469 6f6e 0a20 2020 2020  simulation.     
+00002d30: 2020 2020 2020 2069 6e76 6f6c 7665 7320         involves 
+00002d40: 7365 6c65 6374 696f 6e2e 0a0a 2020 2020  selection...    
+00002d50: 2020 2020 7365 783a 0a20 2020 2020 2020      sex:.       
+00002d60: 2020 2020 2041 2062 6f6f 6c20 666c 6167       A bool flag
+00002d70: 2069 6e64 6963 6174 696e 6720 7768 6574   indicating whet
+00002d80: 6865 7220 6f72 206e 6f74 2074 6869 7320  her or not this 
+00002d90: 5370 6563 6965 7320 6973 2073 6578 6564  Species is sexed
+00002da0: 0a20 2020 2020 2020 2020 2020 2028 692e  .            (i.
+00002db0: 652e 2077 6865 7468 6572 206f 7220 6e6f  e. whether or no
+00002dc0: 7420 7468 6520 656c 6967 6962 696c 6974  t the eligibilit
+00002dd0: 7920 6f66 2070 6f74 656e 7469 616c 206d  y of potential m
+00002de0: 6174 696e 6720 7061 6972 730a 2020 2020  ating pairs.    
+00002df0: 2020 2020 2020 2020 7368 6f75 6c64 2062          should b
+00002e00: 6520 6261 7364 206f 6e20 7768 6574 6865  e basd on whethe
+00002e10: 7220 6f72 206e 6f74 2074 6865 2049 6e64  r or not the Ind
+00002e20: 6976 6964 7561 6c73 2069 6e20 7468 6520  ividuals in the 
+00002e30: 706f 7465 6e74 6961 6c0a 2020 2020 2020  potential.      
+00002e40: 2020 2020 2020 7061 6972 2061 7265 206f        pair are o
+00002e50: 6620 6f70 706f 7369 7465 2073 6578 6573  f opposite sexes
+00002e60: 290a 0a20 2020 2020 2020 2073 6578 5f72  )..        sex_r
+00002e70: 6174 696f 3a0a 2020 2020 2020 2020 2020  atio:.          
+00002e80: 2020 5468 6520 7365 7820 7261 7469 6f20    The sex ratio 
+00002e90: 6f66 2074 6865 2053 7065 6369 6573 2e20  of the Species. 
+00002ea0: 5468 6973 2076 616c 7565 2069 7320 6578  This value is ex
+00002eb0: 7072 6573 7365 6420 6173 0a20 2020 2020  pressed as.     
+00002ec0: 2020 2020 2020 2074 6865 2070 726f 706f         the propo
+00002ed0: 7274 696f 6e20 6f66 2061 6c6c 206f 6666  rtion of all off
+00002ee0: 7370 7269 6e67 2074 6861 7420 6172 6520  spring that are 
+00002ef0: 6d61 6c65 732c 2073 7563 6820 7468 6174  males, such that
+00002f00: 2069 7420 6361 6e20 6265 0a20 2020 2020   it can be.     
+00002f10: 2020 2020 2020 2065 6173 696c 7920 7573         easily us
+00002f20: 6564 2061 7320 7468 6520 7072 6f62 6162  ed as the probab
+00002f30: 696c 6974 7920 7061 7261 6d65 7465 7220  ility parameter 
+00002f40: 666f 7220 7468 6520 4265 726e 6f75 6c6c  for the Bernoull
+00002f50: 6920 6472 6177 0a20 2020 2020 2020 2020  i draw.         
+00002f60: 2020 206f 6620 616e 206f 6666 7370 7269     of an offspri
+00002f70: 6e67 2773 2073 6578 2e20 284e 6f74 6520  ng's sex. (Note 
+00002f80: 7468 6174 2074 6869 7320 7661 6c75 6520  that this value 
+00002f90: 6973 2064 6572 6976 6564 2066 726f 6d0a  is derived from.
+00002fa0: 2020 2020 2020 2020 2020 2020 7468 6520              the 
+00002fb0: 2773 6578 5f72 6174 696f 2720 7061 7261  'sex_ratio' para
+00002fc0: 6d65 7465 7220 7072 6f76 6964 6564 2069  meter provided i
+00002fd0: 6e20 7468 6520 4d6f 6465 6c27 7320 7061  n the Model's pa
+00002fe0: 7261 6d65 7465 7273 2066 696c 652c 0a20  rameters file,. 
+00002ff0: 2020 2020 2020 2020 2020 2062 7574 2075             but u
+00003000: 6e6c 696b 6520 696e 2074 6865 2070 6172  nlike in the par
+00003010: 616d 6574 6572 7320 6669 6c65 2074 6869  ameters file thi
+00003020: 7320 7661 6c75 6520 6973 0a20 2020 2020  s value is.     
+00003030: 2020 2020 2020 206e 6f74 2065 7870 7265         not expre
+00003040: 7373 6564 2069 6e20 6173 2061 2072 6174  ssed in as a rat
+00003050: 696f 206f 6620 6d61 6c65 7320 746f 2066  io of males to f
+00003060: 656d 616c 6573 2e29 0a0a 2020 2020 2020  emales.)..      
+00003070: 2020 7374 6172 745f 4e3a 0a20 2020 2020    start_N:.     
+00003080: 2020 2020 2020 2054 6865 206e 756d 6265         The numbe
+00003090: 7220 6f66 2049 6e64 6976 6964 7561 6c73  r of Individuals
+000030a0: 2074 6f20 6265 2067 656e 6572 6174 6564   to be generated
+000030b0: 2066 6f72 2074 6865 2073 7461 7274 696e   for the startin
+000030c0: 670a 2020 2020 2020 2020 2020 2020 706f  g.            po
+000030d0: 7075 6c61 7469 6f6e 206f 6620 7468 6520  pulation of the 
+000030e0: 5370 6563 6965 7320 2869 2e65 2e20 7468  Species (i.e. th
+000030f0: 6520 706f 7075 6c61 7469 6f6e 2073 697a  e population siz
+00003100: 6520 6174 2074 6865 0a20 2020 2020 2020  e at the.       
+00003110: 2020 2020 2062 6567 696e 6e69 6e67 206f       beginning o
+00003120: 6620 7468 6520 6275 726e 2d69 6e29 0a0a  f the burn-in)..
+00003130: 2020 2020 2020 2020 743a 0a20 2020 2020          t:.     
+00003140: 2020 2020 2020 2054 6865 2069 6e64 6578         The index
+00003150: 206f 6620 7468 6520 6c61 7374 2063 6f6d   of the last com
+00003160: 706c 6574 6564 2074 696d 6520 7374 6570  pleted time step
+00003170: 2072 756e 2066 6f72 2074 6869 7320 5370   run for this Sp
+00003180: 6563 6965 732e 0a20 2020 2020 2020 2020  ecies..         
+00003190: 2020 2054 6869 7320 6174 7472 6962 7574     This attribut
+000031a0: 6520 686f 6c64 7320 7468 6520 7661 6c75  e holds the valu
+000031b0: 6520 2d31 2075 6e74 696c 2074 6865 2053  e -1 until the S
+000031c0: 7065 6369 6573 2068 6173 2062 6565 6e20  pecies has been 
+000031d0: 6275 726e 6564 0a20 2020 2020 2020 2020  burned.         
+000031e0: 2020 2069 6e2c 2061 7420 7768 6963 6820     in, at which 
+000031f0: 706f 696e 7420 6974 2062 6567 696e 7320  point it begins 
+00003200: 696e 6372 656d 656e 7469 6e67 2c20 5079  incrementing, Py
+00003210: 7468 6f6e 6963 616c 6c79 2c20 6672 6f6d  thonically, from
+00003220: 2030 0a20 2020 2020 2020 2020 2020 2028   0.            (
+00003230: 7375 6368 2074 6861 7420 6120 5370 6563  such that a Spec
+00003240: 6965 7320 7769 7468 2074 203d 3d20 3939  ies with t == 99
+00003250: 3920 6861 7320 6265 656e 2072 756e 2066  9 has been run f
+00003260: 6f72 2031 3030 3020 7469 6d65 0a20 2020  or 1000 time.   
+00003270: 2020 2020 2020 2020 2073 7465 7073 292e           steps).
+00003280: 0a0a 2020 2020 2222 220a 2020 2020 2323  ..    """.    ##
+00003290: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+000032a0: 2323 2323 230a 2020 2020 2323 2320 5350  #####.    ### SP
+000032b0: 4543 4941 4c20 4d45 5448 4f44 5320 2323  ECIAL METHODS ##
+000032c0: 230a 2020 2020 2323 2323 2323 2323 2323  #.    ##########
+000032d0: 2323 2323 2323 2323 2323 2323 230a 0a20  #############.. 
+000032e0: 2020 2064 6566 205f 5f69 6e69 745f 5f28     def __init__(
+000032f0: 7365 6c66 2c20 6e61 6d65 2c20 6964 782c  self, name, idx,
+00003300: 2069 6e64 732c 206c 616e 642c 2073 7070   inds, land, spp
+00003310: 5f70 6172 616d 732c 0a20 2020 2020 2020  _params,.       
+00003320: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003330: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003340: 2020 2020 2020 2020 2067 656e 6f6d 6963           genomic
+00003350: 5f61 7263 6869 7465 6374 7572 653d 4e6f  _architecture=No
+00003360: 6e65 293a 0a0a 2020 2020 2020 2020 2363  ne):..        #c
+00003370: 6865 636b 2074 6865 2069 6e64 7320 6f62  heck the inds ob
+00003380: 6a65 6374 2069 7320 636f 7272 6563 740a  ject is correct.
+00003390: 2020 2020 2020 2020 6173 7365 7274 2074          assert t
+000033a0: 7970 6528 696e 6473 2920 696e 2028 4f44  ype(inds) in (OD
+000033b0: 2c20 6469 6374 292c 2028 2241 7267 756d  , dict), ("Argum
+000033c0: 656e 7420 696e 6473 206d 7573 7420 6265  ent inds must be
+000033d0: 206f 6620 220a 2020 2020 2020 2020 2020   of ".          
+000033e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000033f0: 2020 2020 2020 2274 7970 6520 6469 6374        "type dict
+00003400: 206f 7220 7479 7065 2063 6f6c 6c65 6374   or type collect
+00003410: 696f 6e73 2e4f 7264 6572 6564 5f44 6963  ions.Ordered_Dic
+00003420: 7422 290a 2020 2020 2020 2020 6173 7365  t").        asse
+00003430: 7274 206c 6973 7428 7365 7428 5b69 2e5f  rt list(set([i._
+00003440: 5f63 6c61 7373 5f5f 2e5f 5f6e 616d 655f  _class__.__name_
+00003450: 5f20 666f 7220 6920 696e 206c 6973 7428  _ for i in list(
+00003460: 696e 6473 2e76 616c 7565 7328 0a20 2020  inds.values(.   
+00003470: 2020 2020 2020 2020 2029 295d 2929 203d           ))])) =
+00003480: 3d20 5b20 2749 6e64 6976 6964 7561 6c27  = [ 'Individual'
+00003490: 5d2c 2028 2241 7267 756d 656e 7420 696e  ], ("Argument in
+000034a0: 6473 206d 7573 7420 6265 2061 2022 0a20  ds must be a ". 
+000034b0: 2020 2020 2020 2020 2020 2022 6469 6374             "dict
+000034c0: 696f 6e61 7279 2063 6f6e 7461 696e 696e  ionary containin
+000034d0: 6720 6f6e 6c79 2069 6e73 7461 6e63 6573  g only instances
+000034e0: 206f 6620 7468 6520 696e 6469 7669 6475   of the individu
+000034f0: 616c 2e22 0a20 2020 2020 2020 2020 2020  al.".           
+00003500: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003510: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003520: 2020 2020 2020 2020 2022 496e 6469 7669           "Indivi
+00003530: 6475 616c 2063 6c61 7373 2229 0a0a 2020  dual class")..  
+00003540: 2020 2020 2020 2361 7474 7269 6275 7465        #attribute
+00003550: 2074 6f20 686f 6c64 2074 6865 2053 7065   to hold the Spe
+00003560: 6369 6573 2720 6964 7820 696e 2074 6865  cies' idx in the
+00003570: 2043 6f6d 6d75 6e69 7479 2064 6963 7469   Community dicti
+00003580: 6f6e 6172 790a 2020 2020 2020 2020 7365  onary.        se
+00003590: 6c66 2e69 6478 203d 2020 6964 780a 2020  lf.idx =  idx.  
+000035a0: 2020 2020 2020 2320 7570 6461 7465 2077        # update w
+000035b0: 6974 6820 7468 6520 696e 7075 7420 6469  ith the input di
+000035c0: 6374 206f 6620 616c 6c20 696e 6469 7669  ct of all indivi
+000035d0: 6475 616c 732c 0a20 2020 2020 2020 2023  duals,.        #
+000035e0: 6173 2069 6e73 7461 6e63 6573 206f 6620  as instances of 
+000035f0: 696e 6469 7669 6475 616c 2e49 6e64 6976  individual.Indiv
+00003600: 6964 7561 6c20 636c 6173 730a 2020 2020  idual class.    
+00003610: 2020 2020 7365 6c66 2e75 7064 6174 6528      self.update(
+00003620: 696e 6473 290a 0a20 2020 2020 2020 2023  inds)..        #
+00003630: 7365 7420 6f74 6865 7220 6174 7472 6962  set other attrib
+00003640: 7574 6573 0a20 2020 2020 2020 2073 656c  utes.        sel
+00003650: 662e 6e61 6d65 203d 2073 7472 286e 616d  f.name = str(nam
+00003660: 6529 0a20 2020 2020 2020 2073 656c 662e  e).        self.
+00003670: 5f6c 616e 645f 6469 6d20 3d20 6c61 6e64  _land_dim = land
+00003680: 2e64 696d 0a20 2020 2020 2020 2073 656c  .dim.        sel
+00003690: 662e 5f6c 616e 645f 7265 7320 3d20 6c61  f._land_res = la
+000036a0: 6e64 2e72 6573 0a20 2020 2020 2020 2073  nd.res.        s
+000036b0: 656c 662e 5f6c 616e 645f 7265 735f 7261  elf._land_res_ra
+000036c0: 7469 6f20 3d20 6c61 6e64 2e5f 7265 735f  tio = land._res_
+000036d0: 7261 7469 6f0a 2020 2020 2020 2020 7365  ratio.        se
+000036e0: 6c66 2e5f 6c61 6e64 5f75 6c63 203d 206c  lf._land_ulc = l
+000036f0: 616e 642e 756c 630a 2020 2020 2020 2020  and.ulc.        
+00003700: 7365 6c66 2e5f 6c61 6e64 5f70 726a 203d  self._land_prj =
+00003710: 206c 616e 642e 7072 6a0a 2020 2020 2020   land.prj.      
+00003720: 2020 2361 7474 7269 6275 7465 2074 6f20    #attribute to 
+00003730: 6b65 6570 2074 7261 636b 206f 6620 6974  keep track of it
+00003740: 6572 6174 696f 6e20 6e75 6d62 6572 2074  eration number t
+00003750: 6869 7320 7370 7020 6973 2062 6569 6e67  his spp is being
+00003760: 2075 7365 6420 666f 720a 2020 2020 2020   used for.      
+00003770: 2020 2328 6f70 7469 6f6e 616c 3b20 7769    #(optional; wi
+00003780: 6c6c 2062 6520 7365 7420 6279 2074 6865  ll be set by the
+00003790: 2069 7465 7261 7469 6f6e 2e70 7920 6d6f   iteration.py mo
+000037a0: 6475 6c65 2069 6620 6361 6c6c 6564 290a  dule if called).
+000037b0: 2020 2020 2020 2020 7365 6c66 2e5f 6974          self._it
+000037c0: 203d 204e 6f6e 650a 2020 2020 2020 2020   = None.        
+000037d0: 2361 7474 7269 6275 7465 2074 6f20 7472  #attribute to tr
+000037e0: 6163 6b20 746f 7461 6c20 6e75 6d62 6572  ack total number
+000037f0: 206f 6620 7469 6d65 7374 6570 7320 746f   of timesteps to
+00003800: 2062 6520 7275 6e0a 2020 2020 2020 2020   be run.        
+00003810: 2328 7769 6c6c 2062 6520 7365 7420 6279  #(will be set by
+00003820: 2074 6865 204d 6f64 656c 206f 626a 6563   the Model objec
+00003830: 7429 0a20 2020 2020 2020 2023 7365 6c66  t).        #self
+00003840: 2e54 203d 204e 6f6e 6520 0a20 2020 2020  .T = None .     
+00003850: 2020 2023 2061 7474 7269 6275 7465 2074     # attribute t
+00003860: 6f20 6b65 6570 206f 6620 7472 6163 6b20  o keep of track 
+00003870: 6f66 206e 756d 6265 7220 6f66 2074 696d  of number of tim
+00003880: 6573 7465 7073 0a20 2020 2020 2020 2023  esteps.        #
+00003890: 7468 6520 7370 6563 6965 7320 6861 7320  the species has 
+000038a0: 6576 6f6c 7665 6420 2873 7461 7274 7320  evolved (starts 
+000038b0: 6174 202d 312c 200a 2020 2020 2020 2020  at -1, .        
+000038c0: 2374 6f20 696e 6469 6361 7465 2075 6e72  #to indicate unr
+000038d0: 756e 2c20 616e 6420 736f 2074 6861 7420  un, and so that 
+000038e0: 6669 7273 7420 7469 6d65 7374 6570 2077  first timestep w
+000038f0: 696c 6c0a 2020 2020 2020 2020 2362 6520  ill.        #be 
+00003900: 7365 7420 746f 2030 2061 7420 6265 6769  set to 0 at begi
+00003910: 6e6e 696e 6720 6f66 2074 696d 6573 7465  nning of timeste
+00003920: 7029 0a20 2020 2020 2020 2023 204e 4f54  p).        # NOT
+00003930: 453a 2054 6869 7320 7761 792c 2069 6620  E: This way, if 
+00003940: 6d6f 6465 6c20 6973 2073 746f 7070 6564  model is stopped
+00003950: 2c20 6368 616e 6765 7320 0a20 2020 2020  , changes .     
+00003960: 2020 2023 6172 6520 6d61 6465 2c20 7468     #are made, th
+00003970: 656e 2069 7420 6973 2072 756e 2066 7572  en it is run fur
+00003980: 7468 6572 2c0a 2020 2020 2020 2020 2320  ther,.        # 
+00003990: 7468 6973 206d 616b 6573 2069 7420 6561  this makes it ea
+000039a0: 7379 2074 6f20 636f 6e74 696e 7565 2074  sy to continue t
+000039b0: 7261 636b 696e 6720 6672 6f6d 2074 6865  racking from the
+000039c0: 2062 6567 696e 6e69 6e67 0a20 2020 2020   beginning.     
+000039d0: 2020 2073 656c 662e 7420 3d20 2d31 0a20     self.t = -1. 
+000039e0: 2020 2020 2020 2023 7769 6c6c 2062 6520         #will be 
+000039f0: 7377 6974 6368 6564 2074 6f20 5472 7565  switched to True
+00003a00: 2077 6865 6e20 7468 6520 7370 6563 6965   when the specie
+00003a10: 7320 7061 7373 6573 2074 6865 2062 7572  s passes the bur
+00003a20: 6e2d 696e 2074 6573 7473 0a20 2020 2020  n-in tests.     
+00003a30: 2020 2073 656c 662e 6275 726e 6564 203d     self.burned =
+00003a40: 2046 616c 7365 0a20 2020 2020 2020 2023   False.        #
+00003a50: 7769 6c6c 2062 6520 7377 6974 6368 6564  will be switched
+00003a60: 2074 6f20 5472 7565 2069 6620 7468 6520   to True if the 
+00003a70: 7370 6563 6965 7320 676f 6573 2065 7874  species goes ext
+00003a80: 696e 6374 0a20 2020 2020 2020 2073 656c  inct.        sel
+00003a90: 662e 6578 7469 6e63 7420 3d20 4661 6c73  f.extinct = Fals
+00003aa0: 650a 2020 2020 2020 2020 2373 7461 7274  e.        #start
+00003ab0: 696e 6720 706f 7020 7369 7a65 0a20 2020  ing pop size.   
+00003ac0: 2020 2020 2073 656c 662e 7374 6172 745f       self.start_
+00003ad0: 4e20 3d20 6c65 6e28 7365 6c66 290a 2020  N = len(self).  
+00003ae0: 2020 2020 2020 2363 7265 6174 6520 6120        #create a 
+00003af0: 7472 6163 6b65 7220 666f 7220 7468 6520  tracker for the 
+00003b00: 6d61 7869 6d75 6d20 496e 6469 7669 6475  maximum Individu
+00003b10: 616c 2069 6e64 6578 2061 6c72 6561 6479  al index already
+00003b20: 2075 7365 640a 2020 2020 2020 2020 2328   used.        #(
+00003b30: 7468 6973 2077 696c 6c20 6265 2075 7064  this will be upd
+00003b40: 6174 6564 2065 6163 6820 7469 6d65 206f  ated each time o
+00003b50: 6666 7370 7269 6e67 2061 7265 2063 7265  ffspring are cre
+00003b60: 6174 6564 2c20 736f 2074 6861 740a 2020  ated, so that.  
+00003b70: 2020 2020 2020 2369 6e64 6963 6573 2061        #indices a
+00003b80: 7265 206e 6576 6572 2072 6570 6561 7465  re never repeate
+00003b90: 642c 206f 7665 7772 7269 7469 6e67 2065  d, ovewrriting e
+00003ba0: 7869 7374 696e 6720 496e 6469 7669 6475  xisting Individu
+00003bb0: 616c 7329 0a20 2020 2020 2020 2073 656c  als).        sel
+00003bc0: 662e 6d61 785f 696e 645f 6964 7820 3d20  f.max_ind_idx = 
+00003bd0: 6d61 7828 5b2a 696e 6473 5d29 0a20 2020  max([*inds]).   
+00003be0: 2020 2020 2023 6174 7472 6962 7574 6520       #attribute 
+00003bf0: 746f 2068 6f6c 6420 6120 6c61 6e64 7363  to hold a landsc
+00003c00: 6170 652e 4c61 7965 7220 6f62 6a65 6374  ape.Layer object
+00003c10: 206f 660a 2020 2020 2020 2020 2374 6865   of.        #the
+00003c20: 2063 7572 7265 6e74 2073 7065 6369 6573   current species
+00003c30: 2064 656e 7369 7479 0a20 2020 2020 2020   density.       
+00003c40: 2073 656c 662e 4e20 3d20 4e6f 6e65 0a20   self.N = None. 
+00003c50: 2020 2020 2020 2023 2061 7474 7269 6275         # attribu
+00003c60: 7465 2074 6f20 686f 6c64 2061 6e20 6c61  te to hold an la
+00003c70: 6e64 7363 6170 652e 4c61 7965 7220 6f62  ndscape.Layer ob
+00003c80: 6a65 6374 206f 660a 2020 2020 2020 2020  ject of.        
+00003c90: 2374 6865 206c 6f63 616c 2063 6172 7279  #the local carry
+00003ca0: 696e 6720 6361 7061 6369 7479 2028 692e  ing capacity (i.
+00003cb0: 652e 0a20 2020 2020 2020 2023 2774 6172  e..        #'tar
+00003cc0: 6765 7427 2064 796e 616d 6963 2065 7175  get' dynamic equ
+00003cd0: 696c 6962 7269 756d 2073 7065 6369 6573  ilibrium species
+00003ce0: 2064 656e 7369 7479 290a 2020 2020 2020   density).      
+00003cf0: 2020 7365 6c66 2e4b 203d 204e 6f6e 650a    self.K = None.
+00003d00: 2020 2020 2020 2020 2374 6865 2069 6e64          #the ind
+00003d10: 6578 206e 756d 6265 7220 6f66 2074 6865  ex number of the
+00003d20: 204c 6179 6572 2074 6f20 6265 2075 7365   Layer to be use
+00003d30: 6420 6173 2074 6869 7320 5370 6563 6965  d as this Specie
+00003d40: 7327 204b 2d4c 6179 6572 0a20 2020 2020  s' K-Layer.     
+00003d50: 2020 2073 656c 662e 4b5f 6c61 7965 7220     self.K_layer 
+00003d60: 3d20 4e6f 6e65 0a20 2020 2020 2020 2023  = None.        #
+00003d70: 7468 6520 6661 6374 6f72 2062 7920 7768  the factor by wh
+00003d80: 6963 6820 7468 6973 2053 7065 6369 6573  ich this Species
+00003d90: 2720 4b2d 4c61 7965 7220 7368 6f75 6c64  ' K-Layer should
+00003da0: 2062 650a 2020 2020 2020 2020 236d 756c   be.        #mul
+00003db0: 7469 706c 6965 6420 746f 2067 6574 2069  tiplied to get i
+00003dc0: 7473 204b 2072 6173 7465 720a 2020 2020  ts K raster.    
+00003dd0: 2020 2020 7365 6c66 2e4b 5f66 6163 746f      self.K_facto
+00003de0: 7220 3d20 4e6f 6e65 0a20 2020 2020 2020  r = None.       
+00003df0: 2023 206c 6973 7420 746f 2072 6563 6f72   # list to recor
+00003e00: 6420 7370 6563 6965 7320 7369 7a65 2028  d species size (
+00003e10: 6170 7065 6e64 6564 2065 6163 6820 7469  appended each ti
+00003e20: 6d65 0a20 2020 2020 2020 2073 656c 662e  me.        self.
+00003e30: 4e74 203d 205b 5d0a 2020 2020 2020 2020  Nt = [].        
+00003e40: 2320 7370 702e 696e 6372 656d 656e 745f  # spp.increment_
+00003e50: 6167 655f 7374 6167 6528 2920 6973 2063  age_stage() is c
+00003e60: 616c 6c65 6429 0a20 2020 2020 2020 2023  alled).        #
+00003e70: 2074 7261 636b 6572 206f 6620 6e75 6d62   tracker of numb
+00003e80: 6572 206f 6620 6269 7274 6873 2065 6163  er of births eac
+00003e90: 6820 7469 6d65 2073 7070 2e64 6f5f 6d61  h time spp.do_ma
+00003ea0: 7469 6e67 2069 7320 6361 6c6c 6564 0a20  ting is called. 
+00003eb0: 2020 2020 2020 2073 656c 662e 6e5f 6269         self.n_bi
+00003ec0: 7274 6873 203d 205b 5d0a 2020 2020 2020  rths = [].      
+00003ed0: 2020 2320 7472 6163 6b65 7220 6f66 206e    # tracker of n
+00003ee0: 756d 6265 7220 6f66 2064 6561 7468 7320  umber of deaths 
+00003ef0: 6561 6368 2074 696d 650a 2020 2020 2020  each time.      
+00003f00: 2020 2364 656d 6f67 7261 7068 792e 7370    #demography.sp
+00003f10: 705f 6479 6e61 6d69 6373 2069 7320 6361  p_dynamics is ca
+00003f20: 6c6c 6564 0a20 2020 2020 2020 2073 656c  lled.        sel
+00003f30: 662e 6e5f 6465 6174 6873 203d 205b 5d0a  f.n_deaths = [].
+00003f40: 2020 2020 2020 2020 2361 7474 7269 6275          #attribu
+00003f50: 7465 7320 666f 7220 7374 6f72 696e 6720  tes for storing 
+00003f60: 6e75 6d70 7920 6172 7261 7973 206f 6620  numpy arrays of 
+00003f70: 616c 6c20 696e 6469 7669 6475 616c 7327  all individuals'
+00003f80: 0a20 2020 2020 2020 2023 636f 6f72 6469  .        #coordi
+00003f90: 6e61 7465 7320 616e 6420 6365 6c6c 732c  nates and cells,
+00003fa0: 2074 6f20 6176 6f69 6420 7265 7065 6174   to avoid repeat
+00003fb0: 2063 6f6d 7075 7465 2074 696d 6520 6561   compute time ea
+00003fc0: 6368 2074 7572 6e0a 2020 2020 2020 2020  ch turn.        
+00003fd0: 7365 6c66 2e5f 636f 6f72 6473 203d 204e  self._coords = N
+00003fe0: 6f6e 650a 2020 2020 2020 2020 7365 6c66  one.        self
+00003ff0: 2e5f 6365 6c6c 7320 3d20 4e6f 6e65 0a20  ._cells = None. 
+00004000: 2020 2020 2020 2023 6372 6561 7465 2065         #create e
+00004010: 6d70 7479 2061 7474 7269 6275 7465 7320  mpty attributes 
+00004020: 746f 2068 6f6c 6420 7370 6174 6961 6c20  to hold spatial 
+00004030: 6f62 6a65 6374 7320 7468 6174 2077 696c  objects that wil
+00004040: 6c0a 2020 2020 2020 2020 2362 6520 6372  l.        #be cr
+00004050: 6561 7465 6420 6166 7465 7220 7468 6520  eated after the 
+00004060: 7370 6563 6965 7320 6973 2069 6e73 7461  species is insta
+00004070: 6e74 6961 7465 640a 2020 2020 2020 2020  ntiated.        
+00004080: 7365 6c66 2e5f 6b64 5f74 7265 6520 3d20  self._kd_tree = 
+00004090: 4e6f 6e65 0a20 2020 2020 2020 2023 6372  None.        #cr
+000040a0: 6561 7465 2065 6d70 7479 2061 7474 7269  eate empty attri
+000040b0: 6275 7465 2074 6f20 686f 6c64 2074 6865  bute to hold the
+000040c0: 205f 4465 6e73 6974 7947 7269 6453 7461   _DensityGridSta
+000040d0: 636b 0a20 2020 2020 2020 2073 656c 662e  ck.        self.
+000040e0: 5f64 656e 735f 6772 6964 7320 3d20 4e6f  _dens_grids = No
+000040f0: 6e65 0a20 2020 2020 2020 2023 6372 6561  ne.        #crea
+00004100: 7465 2061 6e20 6174 7472 6962 7574 6520  te an attribute 
+00004110: 746f 2069 6e64 6963 6174 6520 7768 6574  to indicate whet
+00004120: 6865 7220 7468 6973 2073 7065 6369 6573  her this species
+00004130: 0a20 2020 2020 2020 2023 7769 6c6c 2068  .        #will h
+00004140: 6176 6520 6d6f 7665 6d65 6e74 3b20 7365  ave movement; se
+00004150: 7420 746f 2046 616c 7365 206e 6f77 2062  t to False now b
+00004160: 7574 2077 696c 6c20 6265 2075 7064 6174  ut will be updat
+00004170: 6564 0a20 2020 2020 2020 2023 6265 6c6f  ed.        #belo
+00004180: 7720 6966 2061 2027 6d6f 7665 6d65 6e74  w if a 'movement
+00004190: 2720 7365 6374 696f 6e20 6973 2065 6e63  ' section is enc
+000041a0: 6f75 6e74 6572 6564 2069 6e20 7468 6520  ountered in the 
+000041b0: 7061 7261 6d73 0a20 2020 2020 2020 2073  params.        s
+000041c0: 656c 662e 5f6d 6f76 6520 3d20 4661 6c73  elf._move = Fals
+000041d0: 650a 2020 2020 2020 2020 2363 7265 6174  e.        #creat
+000041e0: 6520 656d 7074 7920 6174 7472 6962 7574  e empty attribut
+000041f0: 6573 2066 6f72 2073 7061 7469 616c 2e5f  es for spatial._
+00004200: 436f 6e64 7563 7461 6e63 6553 7572 6661  ConductanceSurfa
+00004210: 6365 206f 626a 6563 7473 0a20 2020 2020  ce objects.     
+00004220: 2020 2023 7468 6174 2063 6f75 6c64 2062     #that could b
+00004230: 6520 7573 6564 2066 6f72 206d 6f76 656d  e used for movem
+00004240: 656e 7420 616e 642f 6f72 2064 6973 7065  ent and/or dispe
+00004250: 7273 616c 0a20 2020 2020 2020 2023 286d  rsal.        #(m
+00004260: 6179 2062 6520 6372 6561 7465 642c 2064  ay be created, d
+00004270: 6570 656e 6469 6e67 206f 6e20 7061 7261  epending on para
+00004280: 6d74 6572 7329 0a20 2020 2020 2020 2073  mters).        s
+00004290: 656c 662e 5f6d 6f76 655f 7375 7266 203d  elf._move_surf =
+000042a0: 204e 6f6e 650a 2020 2020 2020 2020 7365   None.        se
+000042b0: 6c66 2e5f 6469 7370 5f73 7572 6620 3d20  lf._disp_surf = 
+000042c0: 4e6f 6e65 0a20 2020 2020 2020 2023 6372  None.        #cr
+000042d0: 6561 7465 2061 6e20 656d 7074 7920 6368  eate an empty ch
+000042e0: 616e 6765 7220 6174 7472 6962 7574 652c  anger attribute,
+000042f0: 2077 6869 6368 2077 696c 6c0a 2020 2020   which will.    
+00004300: 2020 2020 2362 6520 7265 7365 7420 6966      #be reset if
+00004310: 2074 6865 2070 6172 616d 6574 6572 7320   the parameters 
+00004320: 6465 6669 6e65 2063 6861 6e67 6573 2066  define changes f
+00004330: 6f72 2074 6869 7320 7370 700a 2020 2020  or this spp.    
+00004340: 2020 2020 7365 6c66 2e5f 6368 616e 6765      self._change
+00004350: 7220 3d20 4e6f 6e65 0a20 2020 2020 2020  r = None.       
+00004360: 2023 7365 7420 7468 6520 7365 785f 7261   #set the sex_ra
+00004370: 7469 6f20 746f 2030 2e35 2064 6566 6175  tio to 0.5 defau
+00004380: 6c74 2028 6275 7420 7468 6973 2077 696c  lt (but this wil
+00004390: 6c20 6265 2063 6861 6e67 6564 2069 6620  l be changed if 
+000043a0: 610a 2020 2020 2020 2020 236e 6f6e 2d31  a.        #non-1
+000043b0: 2f31 2073 6578 2072 6174 696f 2069 7320  /1 sex ratio is 
+000043c0: 7072 6f76 6964 6564 2069 6e20 7468 6520  provided in the 
+000043d0: 7061 7261 6d73 290a 2020 2020 2020 2020  params).        
+000043e0: 7365 6c66 2e73 6578 5f72 6174 696f 203d  self.sex_ratio =
+000043f0: 2030 2e35 0a20 2020 2020 2020 2023 6372   0.5.        #cr
+00004400: 6561 7465 2061 2070 7269 7661 7465 205f  eate a private _
+00004410: 5061 7261 6d73 5661 6c73 206f 626a 6563  ParamsVals objec
+00004420: 7420 6173 2074 6865 205f 7076 2061 7474  t as the _pv att
+00004430: 7269 6275 7465 0a20 2020 2020 2020 2073  ribute.        s
+00004440: 656c 662e 5f70 7620 3d20 5f50 6172 616d  elf._pv = _Param
+00004450: 7356 616c 7328 7365 6c66 2e6e 616d 6529  sVals(self.name)
+00004460: 0a20 2020 2020 2020 2023 7468 656e 2067  .        #then g
+00004470: 7261 6220 616c 6c20 6f66 2074 6865 206d  rab all of the m
+00004480: 6174 696e 672c 206d 6f72 7461 6c69 7479  ating, mortality
+00004490: 2c20 616e 6420 6d6f 7665 6d65 6e74 0a20  , and movement. 
+000044a0: 2020 2020 2020 2023 7061 7261 6d65 7465         #paramete
+000044b0: 7273 2061 7320 6174 7472 6962 7574 6573  rs as attributes
+000044c0: 206f 6620 7468 6174 205f 5061 7261 6d73   of that _Params
+000044d0: 5661 6c73 206f 626a 6563 740a 2020 2020  Vals object.    
+000044e0: 2020 2020 666f 7220 7365 6374 696f 6e20      for section 
+000044f0: 696e 205b 276d 6174 696e 6727 2c20 276d  in ['mating', 'm
+00004500: 6f72 7461 6c69 7479 272c 2027 6d6f 7665  ortality', 'move
+00004510: 6d65 6e74 275d 3a0a 2020 2020 2020 2020  ment']:.        
+00004520: 2020 2020 6966 2073 6563 7469 6f6e 2069      if section i
+00004530: 6e20 5b2a 7370 705f 7061 7261 6d73 5d3a  n [*spp_params]:
+00004540: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00004550: 2066 6f72 2061 7474 2c76 616c 2069 6e20   for att,val in 
+00004560: 7370 705f 7061 7261 6d73 5b73 6563 7469  spp_params[secti
+00004570: 6f6e 5d2e 6974 656d 7328 293a 0a20 2020  on].items():.   
+00004580: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004590: 2023 6c65 6176 6520 6f75 7420 7468 6520   #leave out the 
+000045a0: 6d6f 7665 5f73 7572 6620 616e 6420 6469  move_surf and di
+000045b0: 7370 5f73 7572 6620 636f 6d70 6f6e 656e  sp_surf componen
+000045c0: 7473 2c0a 2020 2020 2020 2020 2020 2020  ts,.            
+000045d0: 2020 2020 2020 2020 2377 6869 6368 2077          #which w
+000045e0: 696c 6c20 6265 2068 616e 646c 6564 2073  ill be handled s
+000045f0: 6570 6172 6174 656c 790a 2020 2020 2020  eparately.      
+00004600: 2020 2020 2020 2020 2020 2020 2020 6966                if
+00004610: 206e 6f74 2069 7369 6e73 7461 6e63 6528   not isinstance(
+00004620: 7661 6c2c 2064 6963 7429 3a0a 2020 2020  val, dict):.    
+00004630: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004640: 2020 2020 2363 6f6e 7665 7274 2073 6578      #convert sex
+00004650: 2072 6174 696f 2074 6f20 7468 6520 7072   ratio to the pr
+00004660: 6f62 6162 696c 7479 206f 6620 6472 6177  obabilty of draw
+00004670: 696e 6720 6120 6d61 6c65 0a20 2020 2020  ing a male.     
+00004680: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004690: 2020 2069 6620 6174 7420 3d3d 2027 7365     if att == 'se
+000046a0: 785f 7261 7469 6f27 3a0a 2020 2020 2020  x_ratio':.      
+000046b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000046c0: 2020 2020 2020 7661 6c20 3d20 7661 6c20        val = val 
+000046d0: 2f20 2876 616c 202b 2031 290a 2020 2020  / (val + 1).    
+000046e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000046f0: 2020 2020 2361 6464 2061 7320 616e 2061      #add as an a
+00004700: 7474 7269 6275 7465 206f 6620 7468 6520  ttribute of the 
+00004710: 5f50 6172 616d 7356 616c 7320 6f62 6a65  _ParamsVals obje
+00004720: 6374 2028 5f70 7629 0a20 2020 2020 2020  ct (_pv).       
+00004730: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004740: 2073 6574 6174 7472 2873 656c 662e 5f70   setattr(self._p
+00004750: 762c 2061 7474 2c20 7661 6c29 0a20 2020  v, att, val).   
+00004760: 2020 2020 2020 2020 2020 2020 2023 6966               #if
+00004770: 2074 6865 2073 6563 7469 6f6e 2069 7320   the section is 
+00004780: 276d 6f76 656d 656e 7427 2c20 616e 6420  'movement', and 
+00004790: 6974 2773 2069 6e20 7468 6520 7061 7261  it's in the para
+000047a0: 6d73 2c0a 2020 2020 2020 2020 2020 2020  ms,.            
+000047b0: 2020 2020 2374 6869 7320 6d65 616e 7320      #this means 
+000047c0: 7468 6520 5370 6563 6965 7320 7368 6f75  the Species shou
+000047d0: 6c64 2068 6176 6520 6d6f 7665 6d65 6e74  ld have movement
+000047e0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+000047f0: 2020 2373 6f20 7570 6461 7465 2074 6865    #so update the
+00004800: 2073 656c 662e 5f6d 6f76 6520 6174 7472   self._move attr
+00004810: 6962 7574 650a 2020 2020 2020 2020 2020  ibute.          
+00004820: 2020 2020 2020 6966 2073 6563 7469 6f6e        if section
+00004830: 203d 3d20 276d 6f76 656d 656e 7427 3a0a   == 'movement':.
+00004840: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004850: 2020 2020 6966 2073 7070 5f70 6172 616d      if spp_param
+00004860: 735b 7365 6374 696f 6e5d 2e6d 6f76 653a  s[section].move:
+00004870: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00004880: 2020 2020 2020 2020 2073 656c 662e 5f6d           self._m
+00004890: 6f76 6520 3d20 5472 7565 0a0a 2020 2020  ove = True..    
+000048a0: 2020 2020 2369 6620 7365 7820 6973 2054      #if sex is T
+000048b0: 7275 6520 616e 6420 7265 7072 6f5f 6167  rue and repro_ag
+000048c0: 6520 6973 2061 6e20 696e 7420 6f72 2066  e is an int or f
+000048d0: 6c6f 6174 2c20 636f 6572 6365 2074 6f20  loat, coerce to 
+000048e0: 6120 7475 706c 650a 2020 2020 2020 2020  a tuple.        
+000048f0: 2328 6f6e 6520 7661 6c20 666f 7220 6561  #(one val for ea
+00004900: 6368 2073 6578 290a 2020 2020 2020 2020  ch sex).        
+00004910: 6966 2073 656c 662e 7365 783a 0a20 2020  if self.sex:.   
+00004920: 2020 2020 2020 2020 2069 6620 7479 7065           if type
+00004930: 2873 656c 662e 7265 7072 6f5f 6167 6529  (self.repro_age)
+00004940: 2069 6e20 5b66 6c6f 6174 2c20 696e 745d   in [float, int]
+00004950: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+00004960: 2020 7365 6c66 2e72 6570 726f 5f61 6765    self.repro_age
+00004970: 203d 2028 7365 6c66 2e72 6570 726f 5f61   = (self.repro_a
+00004980: 6765 2c20 7365 6c66 2e72 6570 726f 5f61  ge, self.repro_a
+00004990: 6765 290a 0a20 2020 2020 2020 2023 7365  ge)..        #se
+000049a0: 7420 7468 6520 4765 6e6f 6d69 6341 7263  t the GenomicArc
+000049b0: 6869 7465 6374 7572 6520 6f62 6a65 6374  hitecture object
+000049c0: 0a20 2020 2020 2020 2073 656c 662e 6765  .        self.ge
+000049d0: 6e5f 6172 6368 203d 2067 656e 6f6d 6963  n_arch = genomic
+000049e0: 5f61 7263 6869 7465 6374 7572 650a 2020  _architecture.  
+000049f0: 2020 2020 2020 6173 7365 7274 2028 7365        assert (se
+00004a00: 6c66 2e67 656e 5f61 7263 682e 5f5f 636c  lf.gen_arch.__cl
+00004a10: 6173 735f 5f2e 5f5f 6e61 6d65 5f5f 203d  ass__.__name__ =
+00004a20: 3d20 2747 656e 6f6d 6963 4172 6368 6974  = 'GenomicArchit
+00004a30: 6563 7475 7265 270a 2020 2020 2020 2020  ecture'.        
+00004a40: 2020 2020 6f72 2073 656c 662e 6765 6e5f      or self.gen_
+00004a50: 6172 6368 2069 7320 4e6f 6e65 292c 2028  arch is None), (
+00004a60: 2254 6865 2053 7065 6369 6573 2e67 656e  "The Species.gen
+00004a70: 5f61 7263 6820 6174 7472 6962 7574 6520  _arch attribute 
+00004a80: 220a 2020 2020 2020 2020 2020 2020 226d  ".            "m
+00004a90: 7573 7420 6265 2061 6e20 696e 7374 616e  ust be an instan
+00004aa0: 6365 206f 6620 7468 6520 6765 6e6f 6d65  ce of the genome
+00004ab0: 2e47 656e 6f6d 6963 4172 6368 6974 6563  .GenomicArchitec
+00004ac0: 7475 7265 2063 6c61 7373 2022 0a20 2020  ture class ".   
+00004ad0: 2020 2020 2020 2020 2022 6f72 2065 6c73           "or els
+00004ae0: 6520 4e6f 6e65 2e22 290a 0a20 2020 2020  e None.")..     
+00004af0: 2020 2023 2063 7265 6174 6520 6120 7473     # create a ts
+00004b00: 6b69 742e 5461 626c 6543 6f6c 6c65 6374  kit.TableCollect
+00004b10: 696f 6e20 6f62 6a65 6374 2c20 6966 2074  ion object, if t
+00004b20: 6865 2073 7065 6369 6573 2075 7365 7320  he species uses 
+00004b30: 6765 6e6f 6d65 730a 2020 2020 2020 2020  genomes.        
+00004b40: 6966 2073 656c 662e 6765 6e5f 6172 6368  if self.gen_arch
+00004b50: 2069 7320 6e6f 7420 4e6f 6e65 2061 6e64   is not None and
+00004b60: 2073 656c 662e 6765 6e5f 6172 6368 2e75   self.gen_arch.u
+00004b70: 7365 5f74 736b 6974 3a0a 2020 2020 2020  se_tskit:.      
+00004b80: 2020 2020 2020 7365 6c66 2e5f 7463 203d        self._tc =
+00004b90: 2074 736b 6974 2e54 6162 6c65 436f 6c6c   tskit.TableColl
+00004ba0: 6563 7469 6f6e 2873 6571 7565 6e63 655f  ection(sequence_
+00004bb0: 6c65 6e67 7468 3d73 656c 662e 6765 6e5f  length=self.gen_
+00004bc0: 6172 6368 2e4c 290a 2020 2020 2020 2020  arch.L).        
+00004bd0: 2020 2020 7365 6c66 2e5f 7463 5f73 6f72      self._tc_sor
+00004be0: 7465 645f 616e 645f 7369 6d70 6c69 6669  ted_and_simplifi
+00004bf0: 6564 203d 2046 616c 7365 0a20 2020 2020  ed = False.     
+00004c00: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
+00004c10: 2020 2020 2073 656c 662e 5f74 6320 3d20       self._tc = 
+00004c20: 4e6f 6e65 0a20 2020 2020 2020 2020 2020  None.           
+00004c30: 2073 656c 662e 5f74 635f 736f 7274 6564   self._tc_sorted
+00004c40: 5f61 6e64 5f73 696d 706c 6966 6965 6420  _and_simplified 
+00004c50: 3d20 4e6f 6e65 0a0a 2020 2020 2020 2020  = None..        
+00004c60: 2373 6574 2074 6865 2073 656c 6563 7469  #set the selecti
+00004c70: 6f6e 2061 7474 7269 6275 7465 2c20 746f  on attribute, to
+00004c80: 2069 6e64 6963 6174 6520 7768 6574 6865   indicate whethe
+00004c90: 7220 6f72 206e 6f74 0a20 2020 2020 2020  r or not.       
+00004ca0: 2023 6e61 7475 7261 6c20 7365 6c65 6374   #natural select
+00004cb0: 696f 6e20 7368 6f75 6c64 2062 6520 696d  ion should be im
+00004cc0: 706c 656d 656e 7465 6420 666f 7220 7468  plemented for th
+00004cd0: 6520 7370 6563 6965 730a 2020 2020 2020  e species.      
+00004ce0: 2020 7365 6c66 2e73 656c 6563 7469 6f6e    self.selection
+00004cf0: 203d 2028 7365 6c66 2e67 656e 5f61 7263   = (self.gen_arc
+00004d00: 6820 6973 206e 6f74 204e 6f6e 6520 616e  h is not None an
+00004d10: 640a 2020 2020 2020 2020 2020 2020 2873  d.            (s
+00004d20: 656c 662e 6765 6e5f 6172 6368 2e6d 755f  elf.gen_arch.mu_
+00004d30: 6465 6c65 7420 3e20 3020 6f72 2073 656c  delet > 0 or sel
+00004d40: 662e 6765 6e5f 6172 6368 2e74 7261 6974  f.gen_arch.trait
+00004d50: 7320 6973 206e 6f74 204e 6f6e 6529 290a  s is not None)).
+00004d60: 0a20 2020 2020 2020 2023 7365 7420 7468  .        #set th
+00004d70: 6520 7365 6c66 2e6d 7574 6174 6520 6174  e self.mutate at
+00004d80: 7472 6962 7574 6520 2861 2062 6f6f 6c65  tribute (a boole
+00004d90: 616e 2069 6e64 6963 6174 696e 6720 7768  an indicating wh
+00004da0: 6574 6865 720a 2020 2020 2020 2020 236f  ether.        #o
+00004db0: 7220 6e6f 7420 746f 2065 6e61 6374 206d  r not to enact m
+00004dc0: 7574 6174 696f 6e2c 2077 6869 6368 2069  utation, which i
+00004dd0: 7320 5472 7565 2069 6620 6765 6e5f 6172  s True if gen_ar
+00004de0: 6368 2e5f 6d75 5f74 6f74 203e 2030 0a20  ch._mu_tot > 0. 
+00004df0: 2020 2020 2020 2073 656c 662e 6d75 7461         self.muta
+00004e00: 7465 203d 2028 7365 6c66 2e67 656e 5f61  te = (self.gen_a
+00004e10: 7263 6820 6973 206e 6f74 204e 6f6e 650a  rch is not None.
+00004e20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004e30: 2020 2020 2020 2061 6e64 2073 656c 662e         and self.
+00004e40: 6765 6e5f 6172 6368 2e5f 6d75 5f74 6f74  gen_arch._mu_tot
+00004e50: 2069 7320 6e6f 7420 4e6f 6e65 0a20 2020   is not None.   
+00004e60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004e70: 2020 2020 616e 6420 7365 6c66 2e67 656e      and self.gen
+00004e80: 5f61 7263 682e 5f6d 755f 746f 7420 3e20  _arch._mu_tot > 
+00004e90: 3029 0a0a 2020 2020 2020 2020 2373 6574  0)..        #set
+00004ea0: 2074 6865 2073 656c 662e 6d75 745f 6c6f   the self.mut_lo
+00004eb0: 6720 6174 7472 6962 7574 652c 2077 6869  g attribute, whi
+00004ec0: 6368 2064 6963 7461 7465 7320 7768 6574  ch dictates whet
+00004ed0: 6865 720a 2020 2020 2020 2020 236f 7220  her.        #or 
+00004ee0: 6e6f 7420 6120 6d75 7461 7469 6f6e 206c  not a mutation l
+00004ef0: 6f67 2073 686f 756c 6420 6265 2077 7269  og should be wri
+00004f00: 7474 656e 2066 6f72 2074 6869 7320 7370  tten for this sp
+00004f10: 700a 2020 2020 2020 2020 7365 6c66 2e6d  p.        self.m
+00004f20: 7574 5f6c 6f67 203d 204e 6f6e 650a 2020  ut_log = None.  
+00004f30: 2020 2020 2020 6966 2027 6765 6e5f 6172        if 'gen_ar
+00004f40: 6368 2720 696e 205b 2a73 7070 5f70 6172  ch' in [*spp_par
+00004f50: 616d 735d 3a0a 2020 2020 2020 2020 2020  ams]:.          
+00004f60: 2020 7365 6c66 2e6d 7574 5f6c 6f67 203d    self.mut_log =
+00004f70: 2073 7070 5f70 6172 616d 732e 6765 6e5f   spp_params.gen_
+00004f80: 6172 6368 2e6d 7574 5f6c 6f67 0a0a 2020  arch.mut_log..  
+00004f90: 2020 2020 2020 2363 7265 6174 6520 6120        #create a 
+00004fa0: 636f 6f72 6420 6174 7472 6765 7474 6572  coord attrgetter
+00004fb0: 2066 756e 6374 696f 6e2c 0a20 2020 2020   function,.     
+00004fc0: 2020 2023 666f 7220 7573 6520 696e 2067     #for use in g
+00004fd0: 6574 7469 6e67 2061 6c6c 2069 6e64 6976  etting all indiv
+00004fe0: 6964 7327 2063 6f6f 7264 696e 6174 6573  ids' coordinates
+00004ff0: 0a20 2020 2020 2020 2073 656c 662e 5f63  .        self._c
+00005000: 6f6f 7264 5f61 7474 7267 6574 7465 7220  oord_attrgetter 
+00005010: 3d20 6174 7472 6765 7474 6572 2827 7827  = attrgetter('x'
+00005020: 2c20 2779 2729 0a0a 2020 2020 2020 2020  , 'y')..        
+00005030: 2320 6372 6561 7465 2074 6865 2062 7572  # create the bur
+00005040: 6e2d 696e 2073 7061 7469 616c 2063 6f75  n-in spatial cou
+00005050: 6e74 6572 0a20 2020 2020 2020 2073 656c  nter.        sel
+00005060: 662e 5f62 7572 6e69 6e5f 7370 6174 5f74  f._burnin_spat_t
+00005070: 6573 7465 7220 3d20 6275 726e 696e 2e53  ester = burnin.S
+00005080: 7061 7469 616c 5465 7374 6572 2873 656c  patialTester(sel
+00005090: 6629 0a0a 0a20 2020 2023 6f76 6572 7269  f)...    #overri
+000050a0: 6465 2074 6865 205f 5f64 6565 7063 6f70  de the __deepcop
+000050b0: 795f 5f20 6d65 7468 6f64 2c20 736f 2074  y__ method, so t
+000050c0: 6861 7420 7468 6520 7370 6563 6965 7320  hat the species 
+000050d0: 6361 6e0a 2020 2020 2362 6520 636f 7079  can.    #be copy
+000050e0: 2e64 6565 7063 6f70 7927 6420 2862 6563  .deepcopy'd (bec
+000050f0: 6175 7365 206f 7468 6572 7769 7365 2074  ause otherwise t
+00005100: 6869 7320 646f 6573 6e27 7420 776f 726b  his doesn't work
+00005110: 2066 6f72 0a20 2020 2023 636c 6173 7365   for.    #classe
+00005120: 7320 7468 6174 2069 6e68 6572 6974 2066  s that inherit f
+00005130: 726f 6d20 2363 6f6c 6c65 6374 696f 6e73  rom #collections
+00005140: 2e4f 7264 6572 6564 4469 6374 290a 2020  .OrderedDict).  
+00005150: 2020 6465 6620 5f5f 6465 6570 636f 7079    def __deepcopy
+00005160: 5f5f 2873 656c 662c 206d 656d 6f29 3a0a  __(self, memo):.
+00005170: 2020 2020 2020 2020 636c 7320 3d20 7365          cls = se
+00005180: 6c66 2e5f 5f63 6c61 7373 5f5f 0a20 2020  lf.__class__.   
+00005190: 2020 2020 2072 6573 756c 7420 3d20 636c       result = cl
+000051a0: 732e 5f5f 6e65 775f 5f28 636c 7329 0a20  s.__new__(cls). 
+000051b0: 2020 2020 2020 206d 656d 6f5b 6964 2873         memo[id(s
+000051c0: 656c 6629 5d20 3d20 7265 7375 6c74 0a20  elf)] = result. 
+000051d0: 2020 2020 2020 2066 6f72 206b 2c20 7620         for k, v 
+000051e0: 696e 2073 656c 662e 5f5f 6469 6374 5f5f  in self.__dict__
+000051f0: 2e69 7465 6d73 2829 3a0a 2020 2020 2020  .items():.      
+00005200: 2020 2020 2020 7365 7461 7474 7228 7265        setattr(re
+00005210: 7375 6c74 2c20 6b2c 2064 6565 7063 6f70  sult, k, deepcop
+00005220: 7928 762c 206d 656d 6f29 290a 2020 2020  y(v, memo)).    
+00005230: 2020 2020 666f 7220 6b2c 2076 2069 6e20      for k, v in 
+00005240: 7365 6c66 2e69 7465 6d73 2829 3a0a 2020  self.items():.  
+00005250: 2020 2020 2020 2020 2020 7265 7375 6c74            result
+00005260: 5b64 6565 7063 6f70 7928 6b2c 206d 656d  [deepcopy(k, mem
+00005270: 6f29 5d20 3d20 6465 6570 636f 7079 2876  o)] = deepcopy(v
+00005280: 2c20 6d65 6d6f 290a 2020 2020 2020 2020  , memo).        
+00005290: 7265 7475 726e 2072 6573 756c 740a 0a0a  return result...
+000052a0: 2020 2020 2364 6566 696e 6520 7468 6520      #define the 
+000052b0: 5f5f 7374 725f 5f20 616e 6420 5f5f 7265  __str__ and __re
+000052c0: 7072 5f5f 2073 7065 6369 616c 206d 6574  pr__ special met
+000052d0: 686f 6473 0a20 2020 2023 4e4f 5445 3a20  hods.    #NOTE: 
+000052e0: 7468 6973 2069 7320 6e6f 7420 7265 616c  this is not real
+000052f0: 6c79 2061 2067 7265 6174 2072 6570 7265  ly a great repre
+00005300: 7365 6e74 6174 696f 6e3b 2074 6865 2050  sentation; the P
+00005310: 7974 686f 6e0a 2020 2020 2364 6f63 7320  ython.    #docs 
+00005320: 696e 6469 6361 7465 2074 6861 7420 5f5f  indicate that __
+00005330: 7265 7072 5f5f 2073 686f 756c 6420 6964  repr__ should id
+00005340: 6561 6c6c 790a 2020 2020 2370 726f 7669  eally.    #provi
+00005350: 6465 2061 2072 6570 7265 7365 6e74 6174  de a representat
+00005360: 696f 6e20 636f 756c 6420 6265 2075 7365  ion could be use
+00005370: 6420 746f 2072 6563 7265 6174 6520 7468  d to recreate th
+00005380: 650a 2020 2020 236f 626a 6563 742c 2062  e.    #object, b
+00005390: 7574 2069 6620 7468 6174 2069 7320 6e6f  ut if that is no
+000053a0: 7420 706f 7373 6962 6c65 2074 6865 6e20  t possible then 
+000053b0: 6974 200a 2020 2020 2373 686f 756c 6420  it .    #should 
+000053c0: 6174 206c 6561 7374 2070 726f 7669 6465  at least provide
+000053d0: 2061 2073 7472 696e 6720 6f66 2074 6865   a string of the
+000053e0: 2066 6f72 6d20 273c 2e2e 2e20 736f 6d65   form '<... some
+000053f0: 0a20 2020 2023 7573 6566 756c 2064 6573  .    #useful des
+00005400: 6372 6970 7469 6f6e 202e 2e2e 3e3b 2049  cription ...>; I
+00005410: 2776 6520 6174 7465 6d70 7465 6420 746f  've attempted to
+00005420: 2064 6f0a 2020 2020 2374 6865 206c 6174   do.    #the lat
+00005430: 7465 722c 2069 6e73 7069 7265 6420 6279  ter, inspired by
+00005440: 2061 2063 6f6d 626f 206f 6620 7768 6174   a combo of what
+00005450: 2049 2776 6520 7365 656e 2069 6e20 610a   I've seen in a.
+00005460: 2020 2020 2366 6577 206f 7468 6572 2070      #few other p
+00005470: 6163 6b61 6765 7320 2870 616e 6461 732c  ackages (pandas,
+00005480: 206e 6574 4344 4634 2c20 6f73 6765 6f29   netCDF4, osgeo)
+00005490: 200a 2020 2020 2328 7468 6f75 6768 2070   .    #(though p
+000054a0: 6572 6861 7073 2049 2073 686f 756c 6420  erhaps I should 
+000054b0: 6d6f 7265 2063 6c6f 7365 6c79 2063 6f6e  more closely con
+000054c0: 7369 6465 7220 686f 7720 746f 2068 616e  sider how to han
+000054d0: 646c 6520 0a20 2020 2023 7468 6520 7061  dle .    #the pa
+000054e0: 7261 6d73 2069 6e20 6120 6d6f 7265 206e  rams in a more n
+000054f0: 7561 6e63 6564 2c20 7072 6563 6973 6520  uanced, precise 
+00005500: 7761 7920 6f6e 6365 2049 276d 0a20 2020  way once I'm.   
+00005510: 2023 646f 6e65 2077 7269 7469 6e67 2074   #done writing t
+00005520: 6865 2063 6f64 6562 6173 652c 2062 6563  he codebase, bec
+00005530: 6175 7365 2072 6967 6874 206e 6f77 2074  ause right now t
+00005540: 6869 7320 6c65 6176 6573 206f 7574 0a20  his leaves out. 
+00005550: 2020 2023 7061 7261 6d73 2074 6861 7420     #params that 
+00005560: 6861 7665 206c 6973 7473 2061 6e64 2064  have lists and d
+00005570: 6963 7469 6f6e 6172 6972 6573 2061 740a  ictionarires at.
+00005580: 2020 2020 2376 616c 7565 732c 2061 6e64      #values, and
+00005590: 2061 6c73 6f20 696e 636c 7564 6573 2073   also includes s
+000055a0: 6f6d 6520 7061 7261 6d65 7465 7273 2074  ome parameters t
+000055b0: 6861 7420 6172 6520 696e 7075 7420 696e  hat are input in
+000055c0: 0a20 2020 2023 7061 7261 6d73 2e70 7920  .    #params.py 
+000055d0: 6275 7420 6e6f 7420 616c 6c2c 2061 6e64  but not all, and
+000055e0: 2073 6f6d 6520 7061 7261 6d73 2074 6861   some params tha
+000055f0: 740a 2020 2020 2361 7265 2063 7265 6174  t.    #are creat
+00005600: 6564 2069 6e74 6572 6e61 6c6c 7920 6275  ed internally bu
+00005610: 7420 6e6f 7420 616c 6c0a 2020 2020 6465  t not all.    de
+00005620: 6620 5f5f 7374 725f 5f28 7365 6c66 293a  f __str__(self):
+00005630: 0a20 2020 2020 2020 2023 6765 7420 6120  .        #get a 
+00005640: 7374 7269 6e67 2072 6570 7265 7365 6e74  string represent
+00005650: 6174 696f 6e20 6f66 2074 6865 2063 6c61  ation of the cla
+00005660: 7373 0a20 2020 2020 2020 2074 7970 655f  ss.        type_
+00005670: 7374 7220 3d20 7374 7228 7479 7065 2873  str = str(type(s
+00005680: 656c 6629 290a 2020 2020 2020 2020 2367  elf)).        #g
+00005690: 6574 2061 2073 7472 696e 6720 7265 7072  et a string repr
+000056a0: 6573 656e 7461 7469 6f6e 206f 6620 7468  esentation of th
+000056b0: 6520 6669 7273 7420 616e 6420 6c61 7374  e first and last
+000056c0: 2069 6e64 6976 6964 7561 6c73 0a20 2020   individuals.   
+000056d0: 2020 2020 2069 6e64 735f 7374 7220 3d20       inds_str = 
+000056e0: 2725 6920 496e 6469 7669 6475 616c 733a  '%i Individuals:
+000056f0: 5c6e 5c74 2720 2520 6c65 6e28 7365 6c66  \n\t' % len(self
+00005700: 290a 2020 2020 2020 2020 6669 7273 745f  ).        first_
+00005710: 696e 645f 7374 7220 3d20 4f44 2873 656c  ind_str = OD(sel
+00005720: 662e 6974 656d 7328 2929 2e5f 5f73 7472  f.items()).__str
+00005730: 5f5f 2829 2e73 706c 6974 280a 2020 2020  __().split(.    
+00005740: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00005750: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00005760: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00005770: 2729 2c20 2729 5b30 5d20 2b20 2729 5c6e  '), ')[0] + ')\n
+00005780: 5c74 2e2e 2e5c 6e5c 7427 0a20 2020 2020  \t...\n\t'.     
+00005790: 2020 206c 6173 745f 696e 645f 7374 7220     last_ind_str 
+000057a0: 3d20 272c 2027 2e6a 6f69 6e28 7365 6c66  = ', '.join(self
+000057b0: 2e69 7465 6d73 2829 2e5f 5f73 7472 5f5f  .items().__str__
+000057c0: 2829 2e73 706c 6974 2827 2c20 2729 5b2d  ().split(', ')[-
+000057d0: 323a 5d29 0a20 2020 2020 2020 2069 6e64  2:]).        ind
+000057e0: 735f 7374 7220 3d20 696e 6473 5f73 7472  s_str = inds_str
+000057f0: 202b 2066 6972 7374 5f69 6e64 5f73 7472   + first_ind_str
+00005800: 202b 206c 6173 745f 696e 645f 7374 7220   + last_ind_str 
+00005810: 2b20 275c 6e27 0a20 2020 2020 2020 2023  + '\n'.        #
+00005820: 6765 7420 6120 7374 7269 6e67 2072 6570  get a string rep
+00005830: 7265 7365 6e74 6174 696f 6e20 6f66 2074  resentation of t
+00005840: 6865 2066 6972 7374 2074 776f 2061 6e64  he first two and
+00005850: 206c 6173 7420 7477 6f20 7061 7261 6d65   last two parame
+00005860: 7465 7273 0a20 2020 2020 2020 2023 7061  ters.        #pa
+00005870: 7261 6d73 203d 2073 6f72 7465 6428 5b73  rams = sorted([s
+00005880: 7472 286b 2920 2b20 273a 2027 202b 7374  tr(k) + ': ' +st
+00005890: 7228 7629 2066 6f72 206b 2c76 2069 6e20  r(v) for k,v in 
+000058a0: 7661 7273 280a 2020 2020 2020 2020 2320  vars(.        # 
+000058b0: 2020 7365 6c66 292e 6974 656d 7328 2920    self).items() 
+000058c0: 6966 2074 7970 6528 7629 2069 6e20 2873  if type(v) in (s
+000058d0: 7472 2c20 696e 742c 2062 6f6f 6c2c 0a20  tr, int, bool,. 
+000058e0: 2020 2020 2020 2023 2020 2020 2020 2020         #        
+000058f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00005900: 2020 2020 2020 2066 6c6f 6174 295d 2c20         float)], 
+00005910: 6964 7820 3d20 6c61 6d62 6461 2078 3a20  idx = lambda x: 
+00005920: 782e 6c6f 7765 7228 2929 0a20 2020 2020  x.lower()).     
+00005930: 2020 2023 7061 7261 6d73 5f73 7472 203d     #params_str =
+00005940: 2022 5061 7261 6d65 7465 7273 3a5c 6e5c   "Parameters:\n\
+00005950: 7422 202b 2027 2c5c 6e5c 7427 2e6a 6f69  t" + ',\n\t'.joi
+00005960: 6e28 7061 7261 6d73 5b3a 325d 2920 2b20  n(params[:2]) + 
+00005970: 272c 270a 2020 2020 2020 2020 2370 6172  ','.        #par
+00005980: 616d 735f 7374 7220 3d20 7061 7261 6d73  ams_str = params
+00005990: 5f73 7472 202b 2027 5c6e 5c74 2e2e 2e5c  _str + '\n\t...\
+000059a0: 6e5c 7427 0a20 2020 2020 2020 2023 7061  n\t'.        #pa
+000059b0: 7261 6d73 5f73 7472 203d 2070 6172 616d  rams_str = param
+000059c0: 735f 7374 7220 2b20 272c 5c6e 5c74 272e  s_str + ',\n\t'.
+000059d0: 6a6f 696e 2870 6172 616d 735b 2d32 3a5d  join(params[-2:]
+000059e0: 290a 2020 2020 2020 2020 2372 6574 7572  ).        #retur
+000059f0: 6e20 275c 6e27 2e6a 6f69 6e28 5b74 7970  n '\n'.join([typ
+00005a00: 655f 7374 722c 2069 6e64 735f 7374 722c  e_str, inds_str,
+00005a10: 2070 6172 616d 735f 7374 725d 290a 2020   params_str]).  
+00005a20: 2020 2020 2020 7265 7475 726e 2027 5c6e        return '\n
+00005a30: 272e 6a6f 696e 285b 7479 7065 5f73 7472  '.join([type_str
+00005a40: 2c20 696e 6473 5f73 7472 5d29 0a0a 2020  , inds_str])..  
+00005a50: 2020 6465 6620 5f5f 7265 7072 5f5f 2873    def __repr__(s
+00005a60: 656c 6629 3a0a 2020 2020 2020 2020 7265  elf):.        re
+00005a70: 7072 5f73 7472 203d 2073 656c 662e 5f5f  pr_str = self.__
+00005a80: 7374 725f 5f28 290a 2020 2020 2020 2020  str__().        
+00005a90: 7265 7475 726e 2072 6570 725f 7374 720a  return repr_str.
+00005aa0: 0a20 2020 2023 6375 7374 6f6d 697a 6520  .    #customize 
+00005ab0: 7468 6520 5f5f 6765 7461 7474 725f 5f20  the __getattr__ 
+00005ac0: 7370 6563 6961 6c20 6d65 7468 6f64 2c20  special method, 
+00005ad0: 736f 2074 6861 7420 6174 7472 6962 7574  so that attribut
+00005ae0: 6573 2069 6e73 6964 6520 7468 650a 2020  es inside the.  
+00005af0: 2020 2370 7269 7661 7465 205f 5061 7261    #private _Para
+00005b00: 6d73 5661 6c73 2061 7474 7269 6275 7465  msVals attribute
+00005b10: 2028 5f70 7629 2061 6c73 6f20 6265 6861   (_pv) also beha
+00005b20: 7665 2061 7320 7468 6f75 6768 2074 6865  ve as though the
+00005b30: 6972 2061 7474 7269 6275 7465 730a 2020  ir attributes.  
+00005b40: 2020 236f 6620 7468 6520 5370 6563 6965    #of the Specie
+00005b50: 730a 2020 2020 6465 6620 5f5f 6765 7461  s.    def __geta
+00005b60: 7474 725f 5f28 7365 6c66 2c20 6174 7472  ttr__(self, attr
+00005b70: 293a 0a20 2020 2020 2020 2074 7279 3a0a  ):.        try:.
+00005b80: 2020 2020 2020 2020 2020 2020 7661 6c20              val 
+00005b90: 3d20 7365 6c66 2e5f 7076 2e5f 5f67 6574  = self._pv.__get
+00005ba0: 6174 7472 6962 7574 655f 5f28 6174 7472  attribute__(attr
+00005bb0: 290a 2020 2020 2020 2020 2020 2020 7265  ).            re
+00005bc0: 7475 726e 2076 616c 0a20 2020 2020 2020  turn val.       
+00005bd0: 2065 7863 6570 7420 4578 6365 7074 696f   except Exceptio
+00005be0: 6e3a 0a20 2020 2020 2020 2020 2020 2072  n:.            r
+00005bf0: 6169 7365 2041 7474 7269 6275 7465 4572  aise AttributeEr
+00005c00: 726f 7228 2254 6865 2053 7065 6369 6573  ror("The Species
+00005c10: 2068 6173 206e 6f20 6174 7472 6962 7574   has no attribut
+00005c20: 6520 2573 2220 2520 6174 7472 290a 0a0a  e %s" % attr)...
+00005c30: 2020 2020 2323 2323 2323 2323 2323 2323      ############
+00005c40: 2323 2323 2323 2323 230a 2020 2020 2323  #########.    ##
+00005c50: 2320 4f54 4845 5220 4d45 5448 4f44 5320  # OTHER METHODS 
+00005c60: 2323 230a 2020 2020 2323 2323 2323 2323  ###.    ########
+00005c70: 2323 2323 2323 2323 2323 2323 230a 0a20  #############.. 
+00005c80: 2020 2020 2020 2023 2323 2323 2323 2323         #########
+00005c90: 2323 2323 2323 2323 0a20 2020 2020 2020  ########.       
+00005ca0: 2023 7072 6976 6174 6520 6d65 7468 6f64   #private method
+00005cb0: 7323 0a20 2020 2020 2020 2023 2323 2323  s#.        #####
+00005cc0: 2323 2323 2323 2323 2323 2323 0a0a 2020  ############..  
+00005cd0: 2020 236d 6574 686f 6420 746f 2063 616c    #method to cal
+00005ce0: 6375 6c61 7465 2061 6e64 2073 6574 2073  culate and set s
+00005cf0: 656c 662e 4b0a 2020 2020 6465 6620 5f73  elf.K.    def _s
+00005d00: 6574 5f4b 2873 656c 662c 206c 616e 6429  et_K(self, land)
+00005d10: 3a0a 2020 2020 2020 2020 7365 6c66 2e4b  :.        self.K
+00005d20: 203d 206c 616e 645b 7365 6c66 2e4b 5f6c   = land[self.K_l
+00005d30: 6179 6572 5d2e 7261 7374 202a 2073 656c  ayer].rast * sel
+00005d40: 662e 4b5f 6661 6374 6f72 0a0a 2020 2020  f.K_factor..    
+00005d50: 236d 6574 686f 6420 746f 2073 6574 2073  #method to set s
+00005d60: 656c 662e 4e0a 2020 2020 6465 6620 5f73  elf.N.    def _s
+00005d70: 6574 5f4e 2873 656c 662c 204e 293a 2020  et_N(self, N):  
+00005d80: 234e 4f54 453a 2052 6571 7569 7265 7320  #NOTE: Requires 
+00005d90: 6120 6c61 6e64 7363 6170 652e 4c61 7965  a landscape.Laye
+00005da0: 7220 696e 7374 616e 6365 0a20 2020 2020  r instance.     
+00005db0: 2020 2073 656c 662e 4e20 3d20 4e0a 0a20     self.N = N.. 
+00005dc0: 2020 2023 6d65 7468 6f64 2074 6f20 6170     #method to ap
+00005dd0: 7065 6e64 2063 7572 7265 6e74 2073 7070  pend current spp
+00005de0: 2073 697a 6520 746f 2074 6865 2073 7070   size to the spp
+00005df0: 2e4e 7420 6c69 7374 0a20 2020 2064 6566  .Nt list.    def
+00005e00: 205f 7365 745f 4e74 2873 656c 6629 3a0a   _set_Nt(self):.
+00005e10: 2020 2020 2020 2020 7365 6c66 2e4e 742e          self.Nt.
+00005e20: 6170 7065 6e64 286c 656e 2873 656c 6629  append(len(self)
+00005e30: 290a 0a20 2020 2023 6d65 7468 6f64 2074  )..    #method t
+00005e40: 6f20 696e 6372 656d 656e 7420 7468 6520  o increment the 
+00005e50: 7365 6c66 2e74 2061 7474 7269 6275 7465  self.t attribute
+00005e60: 2028 692e 652e 2074 6865 2074 696d 6573   (i.e. the times
+00005e70: 7465 7020 636f 756e 7465 7229 0a20 2020  tep counter).   
+00005e80: 2064 6566 205f 7365 745f 7428 7365 6c66   def _set_t(self
+00005e90: 293a 0a20 2020 2020 2020 2073 656c 662e  ):.        self.
+00005ea0: 7420 2b3d 2031 0a0a 2020 2020 236d 6574  t += 1..    #met
+00005eb0: 686f 6420 746f 2072 6573 6574 2074 6865  hod to reset the
+00005ec0: 2073 656c 662e 7420 6174 7472 6962 7574   self.t attribut
+00005ed0: 6520 2869 2e65 2e20 7468 6520 7469 6d65  e (i.e. the time
+00005ee0: 7374 6570 2063 6f75 6e74 6572 290a 2020  step counter).  
+00005ef0: 2020 6465 6620 5f72 6573 6574 5f74 2873    def _reset_t(s
+00005f00: 656c 6629 3a0a 2020 2020 2020 2020 7365  elf):.        se
+00005f10: 6c66 2e74 203d 202d 310a 0a20 2020 2023  lf.t = -1..    #
+00005f20: 6d65 7468 6f64 2074 6f20 696e 6372 656d  method to increm
+00005f30: 656e 7420 616c 6c20 7370 6563 6965 7327  ent all species'
+00005f40: 2061 6765 2062 7920 6f6e 650a 2020 2020   age by one.    
+00005f50: 2328 616c 736f 2061 6464 7320 6375 7272  #(also adds curr
+00005f60: 656e 7420 7370 7020 7369 7a65 2074 6f20  ent spp size to 
+00005f70: 7472 6163 6b69 6e67 2061 7272 6179 290a  tracking array).
+00005f80: 2020 2020 6465 6620 5f73 6574 5f61 6765      def _set_age
+00005f90: 5f73 7461 6765 2873 656c 6629 3a0a 2020  _stage(self):.  
+00005fa0: 2020 2020 2020 2320 696e 6372 656d 656e        # incremen
+00005fb0: 7420 6167 6520 6f66 2061 6c6c 2069 6e64  t age of all ind
+00005fc0: 6976 6964 7561 6c73 0a20 2020 2020 2020  ividuals.       
+00005fd0: 205b 696e 642e 5f73 6574 5f61 6765 5f73   [ind._set_age_s
+00005fe0: 7461 6765 2829 2066 6f72 2069 6e64 2069  tage() for ind i
+00005ff0: 6e20 7365 6c66 2e76 616c 7565 7328 295d  n self.values()]
+00006000: 3b0a 0a20 2020 2023 206d 6574 686f 6420  ;..    # method 
+00006010: 666f 7220 7275 6e6e 696e 6720 7468 6520  for running the 
+00006020: 7370 6174 6961 6c20 6275 726e 2d69 6e20  spatial burn-in 
+00006030: 7465 7374 0a20 2020 2064 6566 205f 646f  test.    def _do
+00006040: 5f73 7061 7469 616c 5f62 7572 6e69 6e5f  _spatial_burnin_
+00006050: 7465 7374 2873 656c 662c 206e 756d 5f74  test(self, num_t
+00006060: 696d 6573 7465 7073 5f62 6163 6b29 3a0a  imesteps_back):.
+00006070: 2020 2020 2020 2020 2320 6669 7273 7420          # first 
+00006080: 7570 6461 7465 2074 6865 2063 6f75 6e74  update the count
+00006090: 7320 616e 6420 6469 6666 7320 6172 7261  s and diffs arra
+000060a0: 7973 0a20 2020 2020 2020 2023 2028 7768  ys.        # (wh
+000060b0: 6963 6820 636f 756e 7420 7468 6520 6e75  ich count the nu
+000060c0: 6d62 6572 206f 6620 696e 6469 7669 6475  mber of individu
+000060d0: 616c 7320 7769 7468 696e 2065 6163 6820  als within each 
+000060e0: 6c61 6e64 7363 6170 6520 6365 6c6c 2c0a  landscape cell,.
+000060f0: 2020 2020 2020 2020 2320 616e 6420 7468          # and th
+00006100: 6520 6469 6666 2069 6e20 7468 6174 206e  e diff in that n
+00006110: 756d 6265 7220 6265 7477 6565 6e20 6c61  umber between la
+00006120: 7374 2063 6865 636b 2061 6e64 2074 6869  st check and thi
+00006130: 7320 6f6e 6529 0a20 2020 2020 2020 2073  s one).        s
+00006140: 656c 662e 5f62 7572 6e69 6e5f 7370 6174  elf._burnin_spat
+00006150: 5f74 6573 7465 722e 7570 6461 7465 2873  _tester.update(s
+00006160: 656c 6629 0a20 2020 2020 2020 2072 6573  elf).        res
+00006170: 756c 7420 3d20 7365 6c66 2e5f 6275 726e  ult = self._burn
+00006180: 696e 5f73 7061 745f 7465 7374 6572 2e72  in_spat_tester.r
+00006190: 756e 5f74 6573 7428 6e75 6d5f 7469 6d65  un_test(num_time
+000061a0: 7374 6570 735f 6261 636b 290a 2020 2020  steps_back).    
+000061b0: 2020 2020 7265 7475 726e 2072 6573 756c      return resul
+000061c0: 740a 0a20 2020 2023 6d65 7468 6f64 2074  t..    #method t
+000061d0: 6f20 6d6f 7665 2061 6c6c 2069 6e64 6976  o move all indiv
+000061e0: 6964 7561 6c73 2073 696d 756c 7461 6e65  iduals simultane
+000061f0: 6f75 736c 792c 2061 6e64 2073 616d 706c  ously, and sampl
+00006200: 650a 2020 2020 2374 6865 6972 206e 6577  e.    #their new
+00006210: 206c 6f63 6174 696f 6e73 2720 656e 7669   locations' envi
+00006220: 726f 6e6d 656e 740a 2020 2020 6465 6620  ronment.    def 
+00006230: 5f64 6f5f 6d6f 7665 6d65 6e74 2873 656c  _do_movement(sel
+00006240: 662c 206c 616e 6429 3a0a 2020 2020 2020  f, land):.      
+00006250: 2020 5f64 6f5f 6d6f 7665 6d65 6e74 2873    _do_movement(s
+00006260: 656c 6629 0a20 2020 2020 2020 2073 656c  elf).        sel
+00006270: 662e 5f73 6574 5f65 286c 616e 6429 0a20  f._set_e(land). 
+00006280: 2020 2020 2020 2073 656c 662e 5f73 6574         self._set
+00006290: 5f63 6f6f 7264 735f 616e 645f 6365 6c6c  _coords_and_cell
+000062a0: 7328 290a 0a20 2020 2023 6675 6e63 7469  s()..    #functi
+000062b0: 6f6e 2066 6f72 2066 696e 6469 6e67 2061  on for finding a
+000062c0: 6c6c 2074 6865 206d 6174 696e 6720 7061  ll the mating pa
+000062d0: 6972 7320 696e 2061 2073 7065 6369 6573  irs in a species
+000062e0: 0a20 2020 2064 6566 205f 6669 6e64 5f6d  .    def _find_m
+000062f0: 6174 696e 675f 7061 6972 7328 7365 6c66  ating_pairs(self
+00006300: 293a 0a20 2020 2020 2020 206d 6174 696e  ):.        matin
+00006310: 675f 7061 6972 7320 3d20 5f66 696e 645f  g_pairs = _find_
+00006320: 6d61 7465 7328 7365 6c66 2c20 7365 783d  mates(self, sex=
+00006330: 7365 6c66 2e73 6578 2c0a 2020 2020 2020  self.sex,.      
+00006340: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006350: 2020 2020 2020 2020 2020 2020 2063 686f               cho
+00006360: 6f73 655f 6e65 6172 6573 743d 7365 6c66  ose_nearest=self
+00006370: 2e63 686f 6f73 655f 6e65 6172 6573 745f  .choose_nearest_
+00006380: 6d61 7465 2c0a 2020 2020 2020 2020 2020  mate,.          
+00006390: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000063a0: 2020 2020 2020 2020 2069 6e76 6572 7365           inverse
+000063b0: 5f64 6973 745f 6d61 7469 6e67 3d73 656c  _dist_mating=sel
+000063c0: 662e 696e 7665 7273 655f 6469 7374 5f6d  f.inverse_dist_m
+000063d0: 6174 696e 6729 0a20 2020 2020 2020 2072  ating).        r
+000063e0: 6574 7572 6e20 6d61 7469 6e67 5f70 6169  eturn mating_pai
+000063f0: 7273 0a0a 2020 2020 2366 756e 6374 696f  rs..    #functio
+00006400: 6e20 666f 7220 6578 6563 7574 696e 6720  n for executing 
+00006410: 6d61 7469 6e67 2066 6f72 2061 2073 7065  mating for a spe
+00006420: 6369 6573 0a20 2020 2064 6566 205f 646f  cies.    def _do
+00006430: 5f6d 6174 696e 6728 7365 6c66 2c20 6c61  _mating(self, la
+00006440: 6e64 2c20 6d61 7469 6e67 5f70 6169 7273  nd, mating_pairs
+00006450: 2c20 6275 726e 3d46 616c 7365 2c0a 2020  , burn=False,.  
+00006460: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006470: 2063 6865 636b 5f68 6170 733d 4661 6c73   check_haps=Fals
+00006480: 652c 2063 6865 636b 5f67 656e 6f74 7970  e, check_genotyp
+00006490: 6573 3d46 616c 7365 293a 0a0a 2020 2020  es=False):..    
+000064a0: 2020 2020 2320 7365 7420 7468 6520 6765      # set the ge
+000064b0: 6e6f 7479 7065 2d63 6865 636b 2063 6f75  notype-check cou
+000064c0: 6e74 6572 2c20 6966 206e 6563 6573 7361  nter, if necessa
+000064d0: 7279 0a20 2020 2020 2020 2069 6620 6368  ry.        if ch
+000064e0: 6563 6b5f 6765 6e6f 7479 7065 733a 0a20  eck_genotypes:. 
+000064f0: 2020 2020 2020 2020 2020 2067 745f 6368             gt_ch
+00006500: 6563 6b5f 6e75 6d20 3d20 300a 0a20 2020  eck_num = 0..   
+00006510: 2020 2020 2023 6472 6177 2074 6865 206e       #draw the n
+00006520: 756d 6265 7220 6f66 2062 6972 7468 7320  umber of births 
+00006530: 666f 7220 6561 6368 2070 6169 722c 2061  for each pair, a
+00006540: 6e64 2061 7070 656e 640a 2020 2020 2020  nd append.      
+00006550: 2020 2374 6f74 616c 2062 6972 7468 7320    #total births 
+00006560: 746f 2073 656c 662e 6e5f 6269 7274 6873  to self.n_births
+00006570: 206c 6973 740a 2020 2020 2020 2020 6966   list.        if
+00006580: 2073 656c 662e 6e5f 6269 7274 6873 5f66   self.n_births_f
+00006590: 6978 6564 3a0a 2020 2020 2020 2020 2020  ixed:.          
+000065a0: 2020 6e5f 6269 7274 6873 203d 206e 702e    n_births = np.
+000065b0: 6172 7261 7928 0a20 2020 2020 2020 2020  array(.         
+000065c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000065d0: 2020 205b 7365 6c66 2e6e 5f62 6972 7468     [self.n_birth
+000065e0: 735f 6469 7374 725f 6c61 6d62 6461 5d20  s_distr_lambda] 
+000065f0: 2a20 6c65 6e28 6d61 7469 6e67 5f70 6169  * len(mating_pai
+00006600: 7273 2929 0a20 2020 2020 2020 2065 6c73  rs)).        els
+00006610: 653a 0a20 2020 2020 2020 2020 2020 206e  e:.            n
+00006620: 5f62 6972 7468 7320 3d20 5f64 7261 775f  _births = _draw_
+00006630: 6e5f 6269 7274 6873 286c 656e 280a 2020  n_births(len(.  
+00006640: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006650: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006660: 2020 6d61 7469 6e67 5f70 6169 7273 292c    mating_pairs),
+00006670: 2073 656c 662e 6e5f 6269 7274 6873 5f64   self.n_births_d
+00006680: 6973 7472 5f6c 616d 6264 6129 0a20 2020  istr_lambda).   
+00006690: 2020 2020 2074 6f74 616c 5f62 6972 7468       total_birth
+000066a0: 7320 3d20 7375 6d28 6e5f 6269 7274 6873  s = sum(n_births
+000066b0: 290a 2020 2020 2020 2020 7365 6c66 2e6e  ).        self.n
+000066c0: 5f62 6972 7468 732e 6170 7065 6e64 2874  _births.append(t
+000066d0: 6f74 616c 5f62 6972 7468 7329 0a0a 2020  otal_births)..  
+000066e0: 2020 2020 2020 2363 7265 6174 6520 7468        #create th
+000066f0: 6520 6f66 6673 7072 696e 675f 6964 730a  e offspring_ids.
+00006700: 2020 2020 2020 2020 6e65 7874 5f6f 6666          next_off
+00006710: 7370 7269 6e67 5f6b 6579 203d 2073 656c  spring_key = sel
+00006720: 662e 6d61 785f 696e 645f 6964 7820 2b20  f.max_ind_idx + 
+00006730: 310a 2020 2020 2020 2020 6f66 6673 7072  1.        offspr
+00006740: 696e 675f 6b65 7973 203d 206c 6973 7428  ing_keys = list(
+00006750: 7261 6e67 6528 6e65 7874 5f6f 6666 7370  range(next_offsp
+00006760: 7269 6e67 5f6b 6579 2c0a 2020 2020 2020  ring_key,.      
+00006770: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006780: 2020 2020 2020 2020 2020 2020 2020 6e65                ne
+00006790: 7874 5f6f 6666 7370 7269 6e67 5f6b 6579  xt_offspring_key
+000067a0: 202b 2074 6f74 616c 5f62 6972 7468 7329   + total_births)
+000067b0: 295b 3a3a 2d31 5d0a 2020 2020 2020 2020  )[::-1].        
+000067c0: 2375 7064 6174 6520 7365 6c66 2e6d 6178  #update self.max
+000067d0: 5f69 6e64 5f69 6478 0a20 2020 2020 2020  _ind_idx.       
+000067e0: 2069 6620 6c65 6e28 6f66 6673 7072 696e   if len(offsprin
+000067f0: 675f 6b65 7973 2920 3e20 303a 0a20 2020  g_keys) > 0:.   
+00006800: 2020 2020 2020 2020 2073 656c 662e 6d61           self.ma
+00006810: 785f 696e 645f 6964 7820 3d20 6f66 6673  x_ind_idx = offs
+00006820: 7072 696e 675f 6b65 7973 5b30 5d0a 0a20  pring_keys[0].. 
+00006830: 2020 2020 2020 2023 636f 7079 2074 6865         #copy the
+00006840: 206b 6579 732c 2066 6f72 2075 7365 2069   keys, for use i
+00006850: 6e20 6d75 7461 7469 6f6e 2e64 6f5f 6d75  n mutation.do_mu
+00006860: 7461 7469 6f6e 2829 0a20 2020 2020 2020  tation().       
+00006870: 206b 6579 735f 6c69 7374 203d 205b 2a6f   keys_list = [*o
+00006880: 6666 7370 7269 6e67 5f6b 6579 735d 0a0a  ffspring_keys]..
+00006890: 2020 2020 2020 2020 6966 206e 6f74 2062          if not b
+000068a0: 7572 6e20 616e 6420 7365 6c66 2e67 656e  urn and self.gen
+000068b0: 5f61 7263 6820 6973 206e 6f74 204e 6f6e  _arch is not Non
+000068c0: 653a 0a20 2020 2020 2020 2020 2020 2072  e:.            r
+000068d0: 6563 6f6d 625f 6b65 7973 203d 205b 2a72  ecomb_keys = [*r
+000068e0: 2e72 616e 6469 6e74 286c 6f77 3d30 2c0a  .randint(low=0,.
+000068f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006900: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006910: 2020 2020 2020 6869 6768 3d73 656c 662e        high=self.
+00006920: 6765 6e5f 6172 6368 2e72 6563 6f6d 6269  gen_arch.recombi
+00006930: 6e61 7469 6f6e 732e 5f6e 2c0a 2020 2020  nations._n,.    
+00006940: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006950: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006960: 2020 7369 7a65 3d74 6f74 616c 5f62 6972    size=total_bir
+00006970: 7468 732a 3229 5d0a 2020 2020 2020 2020  ths*2)].        
+00006980: 2020 2020 2320 4e4f 5445 3a20 7468 6973      # NOTE: this
+00006990: 2067 6976 6573 2075 7320 6120 6869 6572   gives us a hier
+000069a0: 6172 6368 6963 616c 6c79 206f 7267 616e  archically organ
+000069b0: 697a 6564 206c 6973 742c 0a20 2020 2020  ized list,.     
+000069c0: 2020 2020 2020 2023 2063 6f6e 7461 696e         # contain
+000069d0: 696e 6720 626f 7468 2074 6865 206f 6666  ing both the off
+000069e0: 7370 7269 6e67 2773 206e 6577 2067 656e  spring's new gen
+000069f0: 6f6d 6573 2061 6e64 2074 6865 6972 0a20  omes and their. 
+00006a00: 2020 2020 2020 2020 2020 2023 2073 6567             # seg
+00006a10: 6d65 6e74 2069 6e66 6f72 6d61 7469 6f6e  ment information
+00006a20: 2028 746f 2062 6520 6164 6465 6420 746f   (to be added to
+00006a30: 2074 6865 0a20 2020 2020 2020 2020 2020   the.           
+00006a40: 2023 2074 736b 6974 2e54 6162 6c65 436f   # tskit.TableCo
+00006a50: 6c6c 6563 7469 6f6e 2e65 6467 6573 2074  llection.edges t
+00006a60: 6162 6c65 292c 206f 7267 616e 697a 6564  able), organized
+00006a70: 2061 7320 6120 7375 626c 6973 7420 666f   as a sublist fo
+00006a80: 720a 2020 2020 2020 2020 2020 2020 2320  r.            # 
+00006a90: 6561 6368 2070 6169 722c 2074 6865 6e20  each pair, then 
+00006aa0: 6120 7375 6274 7570 6c65 2077 6974 6869  a subtuple withi
+00006ab0: 6e20 7468 6174 2c20 636f 6e74 6169 6e69  n that, containi
+00006ac0: 6e67 2061 206e 6577 2067 656e 6f6d 650a  ng a new genome.
+00006ad0: 2020 2020 2020 2020 2020 2020 2320 616e              # an
+00006ae0: 6420 7468 656e 2061 2032 2d74 7570 6c65  d then a 2-tuple
+00006af0: 206f 6620 7365 676d 656e 7473 2069 6e66   of segments inf
+00006b00: 6f20 666f 7220 6561 6368 206f 6620 7468  o for each of th
+00006b10: 6520 6e65 770a 2020 2020 2020 2020 2020  e new.          
+00006b20: 2020 2320 6f66 6673 7072 696e 6727 7320    # offspring's 
+00006b30: 686f 6d6f 6c6f 6775 6573 0a20 2020 2020  homologues.     
+00006b40: 2020 2020 2020 2067 656e 6f6d 6573 5f61         genomes_a
+00006b50: 6e64 5f73 6567 7320 3d20 5f64 6f5f 6d61  nd_segs = _do_ma
+00006b60: 7469 6e67 2873 656c 662c 206d 6174 696e  ting(self, matin
+00006b70: 675f 7061 6972 732c 0a20 2020 2020 2020  g_pairs,.       
+00006b80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006b90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006ba0: 2020 206e 5f62 6972 7468 732c 2072 6563     n_births, rec
+00006bb0: 6f6d 625f 6b65 7973 290a 0a20 2020 2020  omb_keys)..     
+00006bc0: 2020 2066 6f72 206e 5f70 6169 722c 2070     for n_pair, p
+00006bd0: 6169 7220 696e 2065 6e75 6d65 7261 7465  air in enumerate
+00006be0: 286d 6174 696e 675f 7061 6972 7329 3a0a  (mating_pairs):.
+00006bf0: 0a20 2020 2020 2020 2020 2020 2070 6172  .            par
+00006c00: 656e 745f 6d69 6470 6f69 6e74 5f78 203d  ent_midpoint_x =
+00006c10: 2028 7365 6c66 5b70 6169 725b 305d 5d2e   (self[pair[0]].
+00006c20: 7820 2b20 7365 6c66 5b70 6169 725b 315d  x + self[pair[1]
+00006c30: 5d2e 7829 2f32 0a20 2020 2020 2020 2020  ].x)/2.         
+00006c40: 2020 2070 6172 656e 745f 6d69 6470 6f69     parent_midpoi
+00006c50: 6e74 5f79 203d 2028 7365 6c66 5b70 6169  nt_y = (self[pai
+00006c60: 725b 305d 5d2e 7920 2b20 7365 6c66 5b70  r[0]].y + self[p
+00006c70: 6169 725b 315d 5d2e 7929 2f32 0a0a 2020  air[1]].y)/2..  
+00006c80: 2020 2020 2020 2020 2020 6e5f 6f66 6673            n_offs
+00006c90: 7072 696e 6720 3d20 6e5f 6269 7274 6873  pring = n_births
+00006ca0: 5b6e 5f70 6169 725d 0a0a 2020 2020 2020  [n_pair]..      
+00006cb0: 2020 2020 2020 666f 7220 6f66 6673 7072        for offspr
+00006cc0: 696e 675f 6920 696e 2072 616e 6765 286e  ing_i in range(n
+00006cd0: 5f6f 6666 7370 7269 6e67 293a 0a0a 2020  _offspring):..  
+00006ce0: 2020 2020 2020 2020 2020 2020 2020 2367                #g
+00006cf0: 6574 2074 6865 206e 6578 7420 6f66 6673  et the next offs
+00006d00: 7072 696e 675f 6b65 790a 2020 2020 2020  pring_key.      
+00006d10: 2020 2020 2020 2020 2020 6f66 6673 7072            offspr
+00006d20: 696e 675f 6b65 7920 3d20 6f66 6673 7072  ing_key = offspr
+00006d30: 696e 675f 6b65 7973 2e70 6f70 2829 0a0a  ing_keys.pop()..
+00006d40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006d50: 6f66 6673 7072 696e 675f 782c 206f 6666  offspring_x, off
+00006d60: 7370 7269 6e67 5f79 203d 205f 646f 5f64  spring_y = _do_d
+00006d70: 6973 7065 7273 616c 280a 2020 2020 2020  ispersal(.      
+00006d80: 2020 2020 2020 2020 2020 2020 2020 7365                se
+00006d90: 6c66 2c20 7061 7265 6e74 5f6d 6964 706f  lf, parent_midpo
+00006da0: 696e 745f 782c 2070 6172 656e 745f 6d69  int_x, parent_mi
+00006db0: 6470 6f69 6e74 5f79 2c0a 2020 2020 2020  dpoint_y,.      
+00006dc0: 2020 2020 2020 2020 2020 2020 2020 7365                se
+00006dd0: 6c66 2e64 6973 7065 7273 616c 5f64 6973  lf.dispersal_dis
+00006de0: 7461 6e63 655f 6469 7374 725f 7061 7261  tance_distr_para
+00006df0: 6d31 2c0a 2020 2020 2020 2020 2020 2020  m1,.            
+00006e00: 2020 2020 2020 2020 7365 6c66 2e64 6973          self.dis
+00006e10: 7065 7273 616c 5f64 6973 7461 6e63 655f  persal_distance_
+00006e20: 6469 7374 725f 7061 7261 6d32 290a 0a20  distr_param2).. 
+00006e30: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00006e40: 7365 7420 7468 6520 6167 6520 746f 2030  set the age to 0
+00006e50: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00006e60: 2061 6765 203d 2030 0a0a 2020 2020 2020   age = 0..      
+00006e70: 2020 2020 2020 2020 2020 2373 6574 2074            #set t
+00006e80: 6865 2073 6578 2063 6f72 7265 6374 6c79  he sex correctly
+00006e90: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00006ea0: 2069 6620 7365 6c66 2e73 6578 3a0a 2020   if self.sex:.  
+00006eb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006ec0: 2020 7365 7820 3d20 722e 6269 6e6f 6d69    sex = r.binomi
+00006ed0: 616c 2831 2c20 7365 6c66 2e73 6578 5f72  al(1, self.sex_r
+00006ee0: 6174 696f 290a 2020 2020 2020 2020 2020  atio).          
+00006ef0: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
 00006f00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006f10: 2020 206e 6577 5f67 656e 6f6d 6520 3d20     new_genome = 
-00006f20: 6765 6e6f 6d65 735f 616e 645f 7365 6773  genomes_and_segs
-00006f30: 5b6e 5f70 6169 725d 5b6f 6666 7370 7269  [n_pair][offspri
-00006f40: 6e67 5f69 5d5b 305d 0a20 2020 2020 2020  ng_i][0].       
-00006f50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006f60: 2023 204e 4f54 453a 2067 656e 6f6d 655f   # NOTE: genome_
-00006f70: 7365 6773 2077 696c 6c20 6a75 7374 2062  segs will just b
-00006f80: 6520 616c 6c20 4e6f 6e65 730a 2020 2020  e all Nones.    
-00006f90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006fa0: 2020 2020 2320 2020 2020 2020 6966 2073      #       if s
-00006fb0: 656c 662e 7573 655f 7473 6b69 7420 3d3d  elf.use_tskit ==
-00006fc0: 2046 616c 7365 0a20 2020 2020 2020 2020   False.         
-00006fd0: 2020 2020 2020 2020 2020 2020 2020 2067                 g
-00006fe0: 656e 6f6d 655f 7365 6773 203d 2067 656e  enome_segs = gen
-00006ff0: 6f6d 6573 5f61 6e64 5f73 6567 735b 6e5f  omes_and_segs[n_
-00007000: 7061 6972 5d5b 6f66 6673 7072 696e 675f  pair][offspring_
-00007010: 695d 5b31 5d0a 2020 2020 2020 2020 2020  i][1].          
-00007020: 2020 2020 2020 2020 2020 656c 7365 3a0a            else:.
+00006f10: 7365 7820 3d20 4e6f 6e65 0a0a 2020 2020  sex = None..    
+00006f20: 2020 2020 2020 2020 2020 2020 2373 6574              #set
+00006f30: 2074 6865 206e 6577 5f67 656e 6f6d 6520   the new_genome 
+00006f40: 636f 7272 6563 746c 790a 2020 2020 2020  correctly.      
+00006f50: 2020 2020 2020 2020 2020 6966 2073 656c            if sel
+00006f60: 662e 6765 6e5f 6172 6368 2069 7320 6e6f  f.gen_arch is no
+00006f70: 7420 4e6f 6e65 3a0a 2020 2020 2020 2020  t None:.        
+00006f80: 2020 2020 2020 2020 2020 2020 6966 206e              if n
+00006f90: 6f74 2062 7572 6e3a 0a20 2020 2020 2020  ot burn:.       
+00006fa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00006fb0: 206e 6577 5f67 656e 6f6d 6520 3d20 6765   new_genome = ge
+00006fc0: 6e6f 6d65 735f 616e 645f 7365 6773 5b6e  nomes_and_segs[n
+00006fd0: 5f70 6169 725d 5b6f 6666 7370 7269 6e67  _pair][offspring
+00006fe0: 5f69 5d5b 305d 0a20 2020 2020 2020 2020  _i][0].         
+00006ff0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00007000: 204e 4f54 453a 2067 656e 6f6d 655f 7365   NOTE: genome_se
+00007010: 6773 2077 696c 6c20 6a75 7374 2062 6520  gs will just be 
+00007020: 616c 6c20 4e6f 6e65 730a 2020 2020 2020  all Nones.      
 00007030: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007040: 2020 2020 2020 2020 6e65 775f 6765 6e6f          new_geno
-00007050: 6d65 3d4e 6f6e 650a 2020 2020 2020 2020  me=None.        
-00007060: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
-00007070: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007080: 2020 6e65 775f 6765 6e6f 6d65 203d 204e    new_genome = N
-00007090: 6f6e 650a 0a20 2020 2020 2020 2020 2020  one..           
-000070a0: 2020 2020 2023 6372 6561 7465 2074 6865       #create the
-000070b0: 206e 6577 2069 6e64 6976 6964 7561 6c0a   new individual.
-000070c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000070d0: 7365 6c66 5b6f 6666 7370 7269 6e67 5f6b  self[offspring_k
-000070e0: 6579 5d20 3d20 496e 6469 7669 6475 616c  ey] = Individual
-000070f0: 2869 6478 3d6f 6666 7370 7269 6e67 5f6b  (idx=offspring_k
-00007100: 6579 2c20 6167 653d 6167 652c 0a20 2020  ey, age=age,.   
+00007040: 2020 2320 2020 2020 2020 6966 2073 656c    #       if sel
+00007050: 662e 7573 655f 7473 6b69 7420 3d3d 2046  f.use_tskit == F
+00007060: 616c 7365 0a20 2020 2020 2020 2020 2020  alse.           
+00007070: 2020 2020 2020 2020 2020 2020 2067 656e               gen
+00007080: 6f6d 655f 7365 6773 203d 2067 656e 6f6d  ome_segs = genom
+00007090: 6573 5f61 6e64 5f73 6567 735b 6e5f 7061  es_and_segs[n_pa
+000070a0: 6972 5d5b 6f66 6673 7072 696e 675f 695d  ir][offspring_i]
+000070b0: 5b31 5d0a 2020 2020 2020 2020 2020 2020  [1].            
+000070c0: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+000070d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000070e0: 2020 2020 2020 6e65 775f 6765 6e6f 6d65        new_genome
+000070f0: 3d4e 6f6e 650a 2020 2020 2020 2020 2020  =None.          
+00007100: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
 00007110: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007120: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007130: 2020 2020 2020 2020 2020 2020 2020 6e65                ne
-00007140: 775f 6765 6e6f 6d65 3d6e 6577 5f67 656e  w_genome=new_gen
-00007150: 6f6d 652c 0a20 2020 2020 2020 2020 2020  ome,.           
-00007160: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007170: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007180: 2020 2020 2020 783d 6f66 6673 7072 696e        x=offsprin
-00007190: 675f 782c 2079 3d6f 6666 7370 7269 6e67  g_x, y=offspring
-000071a0: 5f79 2c0a 2020 2020 2020 2020 2020 2020  _y,.            
+00007120: 6e65 775f 6765 6e6f 6d65 203d 204e 6f6e  new_genome = Non
+00007130: 650a 0a20 2020 2020 2020 2020 2020 2020  e..             
+00007140: 2020 2023 6372 6561 7465 2074 6865 206e     #create the n
+00007150: 6577 2069 6e64 6976 6964 7561 6c0a 2020  ew individual.  
+00007160: 2020 2020 2020 2020 2020 2020 2020 7365                se
+00007170: 6c66 5b6f 6666 7370 7269 6e67 5f6b 6579  lf[offspring_key
+00007180: 5d20 3d20 496e 6469 7669 6475 616c 2869  ] = Individual(i
+00007190: 6478 3d6f 6666 7370 7269 6e67 5f6b 6579  dx=offspring_key
+000071a0: 2c20 6167 653d 6167 652c 0a20 2020 2020  , age=age,.     
 000071b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 000071c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000071d0: 2020 2020 2073 6578 3d73 6578 290a 0a20       sex=sex).. 
-000071e0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
-000071f0: 7365 7420 6e65 7720 696e 6469 7669 6475  set new individu
-00007200: 616c 2773 2070 6865 6e6f 7479 7065 2028  al's phenotype (
-00007210: 776f 6e27 7420 6265 2073 6574 0a20 2020  won't be set.   
-00007220: 2020 2020 2020 2020 2020 2020 2023 6475               #du
-00007230: 7269 6e67 2062 7572 6e2d 696e 2c20 6265  ring burn-in, be
-00007240: 6361 7573 6520 6e6f 2067 656e 6f6d 6573  cause no genomes
-00007250: 2061 7373 6967 6e65 643b 0a20 2020 2020   assigned;.     
-00007260: 2020 2020 2020 2020 2020 2023 776f 6e27             #won'
-00007270: 7420 6265 2073 6574 2069 6620 7468 6520  t be set if the 
-00007280: 7370 6563 6965 7320 6861 7320 6e6f 2067  species has no g
-00007290: 656e 5f61 7263 6829 0a20 2020 2020 2020  en_arch).       
-000072a0: 2020 2020 2020 2020 2069 6620 2873 656c           if (sel
-000072b0: 662e 6765 6e5f 6172 6368 2069 7320 6e6f  f.gen_arch is no
-000072c0: 7420 4e6f 6e65 0a20 2020 2020 2020 2020  t None.         
-000072d0: 2020 2020 2020 2020 2020 2061 6e64 2073             and s
-000072e0: 656c 662e 6765 6e5f 6172 6368 2e74 7261  elf.gen_arch.tra
-000072f0: 6974 7320 6973 206e 6f74 204e 6f6e 650a  its is not None.
-00007300: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007310: 2020 2020 616e 6420 6e6f 7420 6275 726e      and not burn
-00007320: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-00007330: 2020 2020 2020 2073 656c 665b 6f66 6673         self[offs
-00007340: 7072 696e 675f 6b65 795d 2e5f 7365 745f  pring_key]._set_
-00007350: 7a28 7365 6c66 2e67 656e 5f61 7263 6829  z(self.gen_arch)
-00007360: 0a0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00007370: 2020 2320 6475 7269 6e67 2074 6865 206d    # during the m
-00007380: 6169 6e20 7068 6173 652c 2066 6f72 2073  ain phase, for s
-00007390: 7065 6369 6573 2077 6974 6820 6765 6e6f  pecies with geno
-000073a0: 6d65 732c 0a20 2020 2020 2020 2020 2020  mes,.           
-000073b0: 2020 2020 2023 2075 7064 6174 6520 7468       # update th
-000073c0: 6520 7473 6b69 7420 7461 626c 6573 2c20  e tskit tables, 
-000073d0: 6966 206e 6563 6573 7361 7279 0a20 2020  if necessary.   
-000073e0: 2020 2020 2020 2020 2020 2020 2069 6620               if 
-000073f0: 2873 656c 662e 6765 6e5f 6172 6368 2069  (self.gen_arch i
-00007400: 7320 6e6f 7420 4e6f 6e65 0a20 2020 2020  s not None.     
-00007410: 2020 2020 2020 2020 2020 2020 2020 2061                 a
-00007420: 6e64 2073 656c 662e 6765 6e5f 6172 6368  nd self.gen_arch
-00007430: 2e75 7365 5f74 736b 6974 0a20 2020 2020  .use_tskit.     
-00007440: 2020 2020 2020 2020 2020 2020 2020 2061                 a
-00007450: 6e64 206e 6f74 2062 7572 6e29 3a0a 2020  nd not burn):.  
-00007460: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007470: 2020 6c6f 6320 3d20 5b6f 6666 7370 7269    loc = [offspri
-00007480: 6e67 5f78 2c20 6f66 6673 7072 696e 675f  ng_x, offspring_
-00007490: 795d 0a20 2020 2020 2020 2020 2020 2020  y].             
-000074a0: 2020 2020 2020 2069 6620 7365 6c66 2e67         if self.g
-000074b0: 656e 5f61 7263 682e 7472 6169 7473 2069  en_arch.traits i
-000074c0: 7320 6e6f 7420 4e6f 6e65 3a0a 2020 2020  s not None:.    
-000074d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000074e0: 2020 2020 6c6f 6320 3d20 286c 6f63 202b      loc = (loc +
-000074f0: 2073 656c 665b 6f66 6673 7072 696e 675f   self[offspring_
-00007500: 6b65 795d 2e7a 202b 0a20 2020 2020 2020  key].z +.       
-00007510: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007520: 2020 2020 205b 7365 6c66 5b6f 6666 7370       [self[offsp
-00007530: 7269 6e67 5f6b 6579 5d2e 6669 745d 290a  ring_key].fit]).
-00007540: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007550: 2020 2020 6f66 6673 7072 696e 675f 696e      offspring_in
-00007560: 645f 6964 203d 2073 656c 662e 5f74 632e  d_id = self._tc.
-00007570: 696e 6469 7669 6475 616c 732e 6164 645f  individuals.add_
-00007580: 726f 7728 0a20 2020 2020 2020 2020 2020  row(.           
-00007590: 2020 2020 2020 2020 2020 2020 206c 6f63               loc
-000075a0: 6174 696f 6e3d 6c6f 632c 0a20 2020 2020  ation=loc,.     
+000071d0: 2020 2020 2020 2020 2020 2020 6e65 775f              new_
+000071e0: 6765 6e6f 6d65 3d6e 6577 5f67 656e 6f6d  genome=new_genom
+000071f0: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+00007200: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007210: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007220: 2020 2020 783d 6f66 6673 7072 696e 675f      x=offspring_
+00007230: 782c 2079 3d6f 6666 7370 7269 6e67 5f79  x, y=offspring_y
+00007240: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00007250: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007260: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007270: 2020 2073 6578 3d73 6578 290a 0a20 2020     sex=sex)..   
+00007280: 2020 2020 2020 2020 2020 2020 2023 7365               #se
+00007290: 7420 6e65 7720 696e 6469 7669 6475 616c  t new individual
+000072a0: 2773 2070 6865 6e6f 7479 7065 2028 776f  's phenotype (wo
+000072b0: 6e27 7420 6265 2073 6574 0a20 2020 2020  n't be set.     
+000072c0: 2020 2020 2020 2020 2020 2023 6475 7269             #duri
+000072d0: 6e67 2062 7572 6e2d 696e 2c20 6265 6361  ng burn-in, beca
+000072e0: 7573 6520 6e6f 2067 656e 6f6d 6573 2061  use no genomes a
+000072f0: 7373 6967 6e65 643b 0a20 2020 2020 2020  ssigned;.       
+00007300: 2020 2020 2020 2020 2023 776f 6e27 7420           #won't 
+00007310: 6265 2073 6574 2069 6620 7468 6520 7370  be set if the sp
+00007320: 6563 6965 7320 6861 7320 6e6f 2067 656e  ecies has no gen
+00007330: 5f61 7263 6829 0a20 2020 2020 2020 2020  _arch).         
+00007340: 2020 2020 2020 2069 6620 2873 656c 662e         if (self.
+00007350: 6765 6e5f 6172 6368 2069 7320 6e6f 7420  gen_arch is not 
+00007360: 4e6f 6e65 0a20 2020 2020 2020 2020 2020  None.           
+00007370: 2020 2020 2020 2020 2061 6e64 2073 656c           and sel
+00007380: 662e 6765 6e5f 6172 6368 2e74 7261 6974  f.gen_arch.trait
+00007390: 7320 6973 206e 6f74 204e 6f6e 650a 2020  s is not None.  
+000073a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000073b0: 2020 616e 6420 6e6f 7420 6275 726e 293a    and not burn):
+000073c0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000073d0: 2020 2020 2073 656c 665b 6f66 6673 7072       self[offspr
+000073e0: 696e 675f 6b65 795d 2e5f 7365 745f 7a28  ing_key]._set_z(
+000073f0: 7365 6c66 2e67 656e 5f61 7263 6829 0a0a  self.gen_arch)..
+00007400: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007410: 2320 6475 7269 6e67 2074 6865 206d 6169  # during the mai
+00007420: 6e20 7068 6173 652c 2066 6f72 2073 7065  n phase, for spe
+00007430: 6369 6573 2077 6974 6820 6765 6e6f 6d65  cies with genome
+00007440: 732c 0a20 2020 2020 2020 2020 2020 2020  s,.             
+00007450: 2020 2023 2075 7064 6174 6520 7468 6520     # update the 
+00007460: 7473 6b69 7420 7461 626c 6573 2c20 6966  tskit tables, if
+00007470: 206e 6563 6573 7361 7279 0a20 2020 2020   necessary.     
+00007480: 2020 2020 2020 2020 2020 2069 6620 2873             if (s
+00007490: 656c 662e 6765 6e5f 6172 6368 2069 7320  elf.gen_arch is 
+000074a0: 6e6f 7420 4e6f 6e65 0a20 2020 2020 2020  not None.       
+000074b0: 2020 2020 2020 2020 2020 2020 2061 6e64               and
+000074c0: 2073 656c 662e 6765 6e5f 6172 6368 2e75   self.gen_arch.u
+000074d0: 7365 5f74 736b 6974 0a20 2020 2020 2020  se_tskit.       
+000074e0: 2020 2020 2020 2020 2020 2020 2061 6e64               and
+000074f0: 206e 6f74 2062 7572 6e29 3a0a 2020 2020   not burn):.    
+00007500: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007510: 6c6f 6320 3d20 5b6f 6666 7370 7269 6e67  loc = [offspring
+00007520: 5f78 2c20 6f66 6673 7072 696e 675f 795d  _x, offspring_y]
+00007530: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00007540: 2020 2020 2069 6620 7365 6c66 2e67 656e       if self.gen
+00007550: 5f61 7263 682e 7472 6169 7473 2069 7320  _arch.traits is 
+00007560: 6e6f 7420 4e6f 6e65 3a0a 2020 2020 2020  not None:.      
+00007570: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007580: 2020 6c6f 6320 3d20 286c 6f63 202b 2073    loc = (loc + s
+00007590: 656c 665b 6f66 6673 7072 696e 675f 6b65  elf[offspring_ke
+000075a0: 795d 2e7a 202b 0a20 2020 2020 2020 2020  y].z +.         
 000075b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000075c0: 2020 2023 204e 4f54 453a 2075 7369 6e67     # NOTE: using
-000075d0: 2074 6865 206d 6574 6164 6174 6120 636f   the metadata co
-000075e0: 6c75 6d6e 2074 6f20 7374 6f72 6520 746f  lumn to store to
-000075f0: 2067 6e78 0a20 2020 2020 2020 2020 2020   gnx.           
-00007600: 2020 2020 2020 2020 2020 2020 2023 2069               # i
-00007610: 6e64 6976 6964 7561 6c20 6964 782c 2066  ndividual idx, f
-00007620: 6f72 206c 6174 6572 206d 6174 6368 696e  or later matchin
-00007630: 6720 746f 2075 7064 6174 650a 2020 2020  g to update.    
-00007640: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007650: 2020 2020 2320 496e 6469 7669 6475 616c      # Individual
-00007660: 2e5f 696e 6469 7669 6475 616c 735f 7461  ._individuals_ta
-00007670: 625f 6964 2061 6674 6572 2074 736b 6974  b_id after tskit
-00007680: 2773 2073 696d 706c 6966 790a 2020 2020  's simplify.    
-00007690: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000076a0: 2020 2020 2320 616c 676f 7269 7468 6d20      # algorithm 
-000076b0: 6669 6c74 6572 7320 696e 6469 7669 6475  filters individu
-000076c0: 616c 730a 2020 2020 2020 2020 2020 2020  als.            
-000076d0: 2020 2020 2020 2020 2020 2020 6d65 7461              meta
-000076e0: 6461 7461 3d6f 6666 7370 7269 6e67 5f6b  data=offspring_k
-000076f0: 6579 2e74 6f5f 6279 7465 7328 6c65 6e67  ey.to_bytes(leng
-00007700: 7468 3d34 2c0a 2020 2020 2020 2020 2020  th=4,.          
-00007710: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007720: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007730: 2020 2020 2020 2020 2020 2020 2020 6279                by
-00007740: 7465 6f72 6465 723d 276c 6974 746c 6527  teorder='little'
-00007750: 2929 0a20 2020 2020 2020 2020 2020 2020  )).             
-00007760: 2020 2020 2020 2073 656c 665b 6f66 6673         self[offs
-00007770: 7072 696e 675f 6b65 795d 2e5f 696e 6469  pring_key]._indi
-00007780: 7669 6475 616c 735f 7461 625f 6964 203d  viduals_tab_id =
-00007790: 206f 6666 7370 7269 6e67 5f69 6e64 5f69   offspring_ind_i
-000077a0: 640a 0a20 2020 2020 2020 2020 2020 2020  d..             
-000077b0: 2020 2020 2020 2023 2061 6464 2072 6f77         # add row
-000077c0: 7320 746f 2074 6865 206e 6f64 6573 2074  s to the nodes t
-000077d0: 6162 6c65 2c20 7365 7474 696e 670a 2020  able, setting.  
-000077e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000077f0: 2020 2320 7468 6520 2766 6c61 6773 2720    # the 'flags' 
-00007800: 636f 6c75 6d6e 2076 616c 7320 746f 2031  column vals to 1
-00007810: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00007820: 2020 2020 2023 2028 746f 2069 6e64 6963       # (to indic
-00007830: 6174 6520 7468 6579 2772 6520 7265 616c  ate they're real
-00007840: 2069 6e64 6976 6964 732c 206e 6f74 206d   individs, not m
-00007850: 7370 7269 6d65 2d64 6572 6976 6564 290a  sprime-derived).
-00007860: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007870: 2020 2020 2320 616e 6420 7365 7474 696e      # and settin
-00007880: 6720 7468 6520 2769 6e64 6976 6964 7561  g the 'individua
-00007890: 6c27 2063 6f6c 756d 6e20 7661 6c73 2074  l' column vals t
-000078a0: 6f20 6964 730a 2020 2020 2020 2020 2020  o ids.          
-000078b0: 2020 2020 2020 2020 2020 2320 7265 7475            # retu
-000078c0: 726e 6564 2066 726f 6d20 7463 2e69 6e64  rned from tc.ind
-000078d0: 6976 6964 7561 6c73 2e61 6464 5f72 6f77  ividuals.add_row
-000078e0: 2829 2c20 7468 656e 2061 6464 696e 670a  (), then adding.
-000078f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007900: 2020 2020 2320 7468 6520 7265 7475 726e      # the return
-00007910: 6564 2074 736b 6974 204e 6f64 6520 6964  ed tskit Node id
-00007920: 7320 746f 2049 6e64 6976 6964 7561 6c5f  s to Individual_
-00007930: 6e6f 6465 735f 7461 625f 6964 730a 2020  nodes_tab_ids.  
-00007940: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007950: 2020 2320 6174 7472 6962 7574 6520 2877    # attribute (w
-00007960: 6869 6368 2069 7320 6120 6c69 7374 290a  hich is a list).
+000075c0: 2020 205b 7365 6c66 5b6f 6666 7370 7269     [self[offspri
+000075d0: 6e67 5f6b 6579 5d2e 6669 745d 290a 2020  ng_key].fit]).  
+000075e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000075f0: 2020 6f66 6673 7072 696e 675f 696e 645f    offspring_ind_
+00007600: 6964 203d 2073 656c 662e 5f74 632e 696e  id = self._tc.in
+00007610: 6469 7669 6475 616c 732e 6164 645f 726f  dividuals.add_ro
+00007620: 7728 0a20 2020 2020 2020 2020 2020 2020  w(.             
+00007630: 2020 2020 2020 2020 2020 206c 6f63 6174             locat
+00007640: 696f 6e3d 6c6f 632c 0a20 2020 2020 2020  ion=loc,.       
+00007650: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007660: 2023 204e 4f54 453a 2075 7369 6e67 2074   # NOTE: using t
+00007670: 6865 206d 6574 6164 6174 6120 636f 6c75  he metadata colu
+00007680: 6d6e 2074 6f20 7374 6f72 6520 746f 2067  mn to store to g
+00007690: 6e78 0a20 2020 2020 2020 2020 2020 2020  nx.             
+000076a0: 2020 2020 2020 2020 2020 2023 2069 6e64             # ind
+000076b0: 6976 6964 7561 6c20 6964 782c 2066 6f72  ividual idx, for
+000076c0: 206c 6174 6572 206d 6174 6368 696e 6720   later matching 
+000076d0: 746f 2075 7064 6174 650a 2020 2020 2020  to update.      
+000076e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000076f0: 2020 2320 496e 6469 7669 6475 616c 2e5f    # Individual._
+00007700: 696e 6469 7669 6475 616c 735f 7461 625f  individuals_tab_
+00007710: 6964 2061 6674 6572 2074 736b 6974 2773  id after tskit's
+00007720: 2073 696d 706c 6966 790a 2020 2020 2020   simplify.      
+00007730: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007740: 2020 2320 616c 676f 7269 7468 6d20 6669    # algorithm fi
+00007750: 6c74 6572 7320 696e 6469 7669 6475 616c  lters individual
+00007760: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+00007770: 2020 2020 2020 2020 2020 6d65 7461 6461            metada
+00007780: 7461 3d6f 6666 7370 7269 6e67 5f6b 6579  ta=offspring_key
+00007790: 2e74 6f5f 6279 7465 7328 6c65 6e67 7468  .to_bytes(length
+000077a0: 3d34 2c0a 2020 2020 2020 2020 2020 2020  =4,.            
+000077b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000077c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000077d0: 2020 2020 2020 2020 2020 2020 6279 7465              byte
+000077e0: 6f72 6465 723d 276c 6974 746c 6527 2929  order='little'))
+000077f0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00007800: 2020 2020 2073 656c 665b 6f66 6673 7072       self[offspr
+00007810: 696e 675f 6b65 795d 2e5f 696e 6469 7669  ing_key]._indivi
+00007820: 6475 616c 735f 7461 625f 6964 203d 206f  duals_tab_id = o
+00007830: 6666 7370 7269 6e67 5f69 6e64 5f69 640a  ffspring_ind_id.
+00007840: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00007850: 2020 2020 2023 2061 6464 2072 6f77 7320       # add rows 
+00007860: 746f 2074 6865 206e 6f64 6573 2074 6162  to the nodes tab
+00007870: 6c65 2c20 7365 7474 696e 670a 2020 2020  le, setting.    
+00007880: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007890: 2320 7468 6520 2766 6c61 6773 2720 636f  # the 'flags' co
+000078a0: 6c75 6d6e 2076 616c 7320 746f 2031 0a20  lumn vals to 1. 
+000078b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000078c0: 2020 2023 2028 746f 2069 6e64 6963 6174     # (to indicat
+000078d0: 6520 7468 6579 2772 6520 6173 736f 6369  e they're associ
+000078e0: 6174 6564 2077 6974 6820 7265 616c 2069  ated with real i
+000078f0: 6e64 6976 6964 732c 0a20 2020 2020 2020  ndivids,.       
+00007900: 2020 2020 2020 2020 2020 2020 2023 206e               # n
+00007910: 6f74 206d 7370 7269 6d65 2d64 6572 6976  ot msprime-deriv
+00007920: 6564 206f 6e65 7329 0a20 2020 2020 2020  ed ones).       
+00007930: 2020 2020 2020 2020 2020 2020 2023 2061               # a
+00007940: 6e64 2073 6574 7469 6e67 2074 6865 2027  nd setting the '
+00007950: 696e 6469 7669 6475 616c 2720 636f 6c75  individual' colu
+00007960: 6d6e 2076 616c 7320 746f 2069 6473 0a20  mn vals to ids. 
 00007970: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007980: 2020 2020 2320 4e4f 5445 3a20 6d61 6b65      # NOTE: make
-00007990: 2074 696d 6520 6e65 6761 7469 7665 2073   time negative s
-000079a0: 6f20 7468 6174 2070 6172 656e 7420 7469  o that parent ti
-000079b0: 6d65 2069 7320 616c 7761 7973 0a20 2020  me is always.   
-000079c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000079d0: 2023 2067 7265 6174 6572 2074 6861 6e20   # greater than 
-000079e0: 6368 696c 6420 7469 6d65 2028 6173 2069  child time (as i
-000079f0: 7420 776f 756c 6420 6265 2065 7870 7265  t would be expre
-00007a00: 7373 6564 2069 6e20 7468 650a 2020 2020  ssed in the.    
-00007a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007a20: 2320 636f 616c 6573 6365 6e74 2c20 6578  # coalescent, ex
-00007a30: 6365 7074 2074 6861 7420 7765 2063 616e  cept that we can
-00007a40: 2774 2075 7365 2070 6f73 6974 6976 6520  't use positive 
-00007a50: 6e75 6d62 6572 730a 2020 2020 2020 2020  numbers.        
-00007a60: 2020 2020 2020 2020 2020 2020 2320 6865              # he
-00007a70: 7265 2062 6563 6175 7365 2077 6520 7761  re because we wa
-00007a80: 6e74 2074 6f20 616c 6c6f 7720 666f 7220  nt to allow for 
-00007a90: 7468 6520 706f 7373 6962 696c 6974 790a  the possibility.
-00007aa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007ab0: 2020 2020 2320 7468 6174 2061 206d 6f64      # that a mod
-00007ac0: 656c 2063 6f75 6c64 2062 6520 7761 6c6b  el could be walk
-00007ad0: 6564 2066 6f72 2061 6e79 2061 7262 6974  ed for any arbit
-00007ae0: 7261 7279 0a20 2020 2020 2020 2020 2020  rary.           
-00007af0: 2020 2020 2020 2020 2023 206e 756d 6265           # numbe
-00007b00: 7220 6f66 2074 696d 6520 7374 6570 7329  r of time steps)
-00007b10: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00007b20: 2020 2020 2073 656c 665b 6f66 6673 7072       self[offspr
-00007b30: 696e 675f 6b65 795d 2e5f 7365 745f 6e6f  ing_key]._set_no
-00007b40: 6465 735f 7461 625f 6964 7328 0a20 2020  des_tab_ids(.   
-00007b50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007b60: 2020 2020 202a 5b73 656c 662e 5f74 632e       *[self._tc.
-00007b70: 6e6f 6465 732e 6164 645f 726f 7728 666c  nodes.add_row(fl
-00007b80: 6167 733d 312c 2074 696d 653d 2d73 656c  ags=1, time=-sel
-00007b90: 662e 742c 0a20 2020 2020 2020 2020 2020  f.t,.           
-00007ba0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007980: 2020 2023 2072 6574 7572 6e65 6420 6672     # returned fr
+00007990: 6f6d 2074 632e 696e 6469 7669 6475 616c  om tc.individual
+000079a0: 732e 6164 645f 726f 7728 292c 2074 6865  s.add_row(), the
+000079b0: 6e20 6164 6469 6e67 0a20 2020 2020 2020  n adding.       
+000079c0: 2020 2020 2020 2020 2020 2020 2023 2074               # t
+000079d0: 6865 2072 6574 7572 6e65 6420 7473 6b69  he returned tski
+000079e0: 7420 4e6f 6465 2069 6473 2074 6f20 496e  t Node ids to In
+000079f0: 6469 7669 6475 616c 5f6e 6f64 6573 5f74  dividual_nodes_t
+00007a00: 6162 5f69 6473 0a20 2020 2020 2020 2020  ab_ids.         
+00007a10: 2020 2020 2020 2020 2020 2023 2061 7474             # att
+00007a20: 7269 6275 7465 2028 7768 6963 6820 6973  ribute (which is
+00007a30: 2061 206c 6973 7429 0a20 2020 2020 2020   a list).       
+00007a40: 2020 2020 2020 2020 2020 2020 2023 204e               # N
+00007a50: 4f54 453a 206d 616b 6520 7469 6d65 206e  OTE: make time n
+00007a60: 6567 6174 6976 6520 736f 2074 6861 7420  egative so that 
+00007a70: 7061 7265 6e74 2074 696d 6520 6973 2061  parent time is a
+00007a80: 6c77 6179 730a 2020 2020 2020 2020 2020  lways.          
+00007a90: 2020 2020 2020 2020 2020 2320 6772 6561            # grea
+00007aa0: 7465 7220 7468 616e 2063 6869 6c64 2074  ter than child t
+00007ab0: 696d 6520 2861 7320 6974 2077 6f75 6c64  ime (as it would
+00007ac0: 2062 6520 6578 7072 6573 7365 6420 696e   be expressed in
+00007ad0: 2074 6865 0a20 2020 2020 2020 2020 2020   the.           
+00007ae0: 2020 2020 2020 2020 2023 2063 6f61 6c65           # coale
+00007af0: 7363 656e 742c 2065 7863 6570 7420 7468  scent, except th
+00007b00: 6174 2077 6520 6361 6e27 7420 7573 6520  at we can't use 
+00007b10: 706f 7369 7469 7665 206e 756d 6265 7273  positive numbers
+00007b20: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00007b30: 2020 2020 2023 2068 6572 6520 6265 6361       # here beca
+00007b40: 7573 6520 7765 2077 616e 7420 746f 2061  use we want to a
+00007b50: 6c6c 6f77 2066 6f72 2074 6865 2070 6f73  llow for the pos
+00007b60: 7369 6269 6c69 7479 0a20 2020 2020 2020  sibility.       
+00007b70: 2020 2020 2020 2020 2020 2020 2023 2074               # t
+00007b80: 6861 7420 6120 6d6f 6465 6c20 636f 756c  hat a model coul
+00007b90: 6420 6265 2077 616c 6b65 6420 666f 7220  d be walked for 
+00007ba0: 616e 7920 6172 6269 7472 6172 790a 2020  any arbitrary.  
 00007bb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007bc0: 2020 2020 2020 706f 7075 6c61 7469 6f6e        population
-00007bd0: 3d30 2c0a 2020 2020 2020 2020 2020 2020  =0,.            
-00007be0: 2020 2020 2020 2020 2020 2020 696e 6469              indi
-00007bf0: 7669 6475 616c 3d6f 6666 7370 7269 6e67  vidual=offspring
-00007c00: 5f69 6e64 5f69 6429 2066 6f72 205f 2069  _ind_id) for _ i
-00007c10: 6e20 7261 6e67 6528 0a20 2020 2020 2020  n range(.       
-00007c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007c30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007c40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007c50: 2020 2020 2073 656c 662e 6765 6e5f 6172       self.gen_ar
-00007c60: 6368 2e78 295d 290a 0a20 2020 2020 2020  ch.x)])..       
-00007c70: 2020 2020 2020 2020 2020 2020 2023 2061               # a
-00007c80: 6464 2065 6467 6573 2074 6f20 7468 6520  dd edges to the 
-00007c90: 7473 6b69 7420 6564 6765 7320 7461 626c  tskit edges tabl
-00007ca0: 650a 2020 2020 2020 2020 2020 2020 2020  e.              
-00007cb0: 2020 2020 2020 2320 4e4f 5445 3a20 605b        # NOTE: `[
-00007cc0: 2a73 6567 5f73 6574 5d60 2069 7320 6e65  *seg_set]` is ne
-00007cd0: 6365 7373 6172 7920 746f 2073 7461 722d  cessary to star-
-00007ce0: 756e 7061 636b 0a20 2020 2020 2020 2020  unpack.         
-00007cf0: 2020 2020 2020 2020 2020 2023 2074 6865             # the
-00007d00: 207a 6970 206f 626a 6563 740a 2020 2020   zip object.    
-00007d10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007d20: 6564 6765 5f69 6420 3d20 5b73 656c 662e  edge_id = [self.
-00007d30: 5f74 632e 6564 6765 732e 6164 645f 726f  _tc.edges.add_ro
-00007d40: 7728 0a20 2020 2020 2020 2020 2020 2020  w(.             
-00007d50: 2020 2020 2020 2020 2020 2070 6172 656e             paren
-00007d60: 743d 7365 675b 305d 2c20 6c65 6674 3d73  t=seg[0], left=s
-00007d70: 6567 5b31 5d2c 2072 6967 6874 3d73 6567  eg[1], right=seg
-00007d80: 5b32 5d2c 0a20 2020 2020 2020 2020 2020  [2],.           
-00007d90: 2020 2020 2020 2020 2020 2020 2063 6869               chi
-00007da0: 6c64 3d73 656c 665b 6f66 6673 7072 696e  ld=self[offsprin
-00007db0: 675f 6b65 795d 2e5f 6e6f 6465 735f 7461  g_key]._nodes_ta
-00007dc0: 625f 6964 735b 686f 6d6f 6c5d 0a20 2020  b_ids[homol].   
-00007dd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007de0: 2020 2020 2029 2066 6f72 2068 6f6d 6f6c       ) for homol
-00007df0: 2c20 7365 675f 7365 7420 696e 2065 6e75  , seg_set in enu
-00007e00: 6d65 7261 7465 280a 2020 2020 2020 2020  merate(.        
+00007bc0: 2020 2320 6e75 6d62 6572 206f 6620 7469    # number of ti
+00007bd0: 6d65 2073 7465 7073 290a 2020 2020 2020  me steps).      
+00007be0: 2020 2020 2020 2020 2020 2020 2020 7365                se
+00007bf0: 6c66 5b6f 6666 7370 7269 6e67 5f6b 6579  lf[offspring_key
+00007c00: 5d2e 5f73 6574 5f6e 6f64 6573 5f74 6162  ]._set_nodes_tab
+00007c10: 5f69 6473 280a 2020 2020 2020 2020 2020  _ids(.          
+00007c20: 2020 2020 2020 2020 2020 2020 2020 2a5b                *[
+00007c30: 7365 6c66 2e5f 7463 2e6e 6f64 6573 2e61  self._tc.nodes.a
+00007c40: 6464 5f72 6f77 2866 6c61 6773 3d31 2c20  dd_row(flags=1, 
+00007c50: 7469 6d65 3d2d 7365 6c66 2e74 2c0a 2020  time=-self.t,.  
+00007c60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007c70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007c80: 2020 2020 2020 2020 2020 2020 2020 2070                 p
+00007c90: 6f70 756c 6174 696f 6e3d 302c 0a20 2020  opulation=0,.   
+00007ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007cb0: 2020 2020 2069 6e64 6976 6964 7561 6c3d       individual=
+00007cc0: 6f66 6673 7072 696e 675f 696e 645f 6964  offspring_ind_id
+00007cd0: 2920 666f 7220 5f20 696e 2072 616e 6765  ) for _ in range
+00007ce0: 280a 2020 2020 2020 2020 2020 2020 2020  (.              
+00007cf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007d00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007d10: 2020 2020 2020 2020 2020 2020 2020 7365                se
+00007d20: 6c66 2e67 656e 5f61 7263 682e 7829 5d29  lf.gen_arch.x)])
+00007d30: 0a0a 2020 2020 2020 2020 2020 2020 2020  ..              
+00007d40: 2020 2020 2020 2320 6164 6420 6564 6765        # add edge
+00007d50: 7320 746f 2074 6865 2074 736b 6974 2065  s to the tskit e
+00007d60: 6467 6573 2074 6162 6c65 0a20 2020 2020  dges table.     
+00007d70: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00007d80: 204e 4f54 453a 2060 5b2a 7365 675f 7365   NOTE: `[*seg_se
+00007d90: 745d 6020 6973 206e 6563 6573 7361 7279  t]` is necessary
+00007da0: 2074 6f20 7374 6172 2d75 6e70 6163 6b0a   to star-unpack.
+00007db0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007dc0: 2020 2020 2320 7468 6520 7a69 7020 6f62      # the zip ob
+00007dd0: 6a65 6374 0a20 2020 2020 2020 2020 2020  ject.           
+00007de0: 2020 2020 2020 2020 2065 6467 655f 6964           edge_id
+00007df0: 203d 205b 7365 6c66 2e5f 7463 2e65 6467   = [self._tc.edg
+00007e00: 6573 2e61 6464 5f72 6f77 280a 2020 2020  es.add_row(.    
 00007e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007e20: 6765 6e6f 6d65 5f73 6567 7329 2066 6f72  genome_segs) for
-00007e30: 2073 6567 2069 6e20 5b2a 7365 675f 7365   seg in [*seg_se
-00007e40: 745d 5d0a 0a20 2020 2020 2020 2020 2020  t]]..           
-00007e50: 2020 2020 2020 2020 2069 6620 6368 6563           if chec
-00007e60: 6b5f 6861 7073 3a0a 2020 2020 2020 2020  k_haps:.        
-00007e70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007e80: 2320 6368 6563 6b20 7468 6520 696e 6469  # check the indi
-00007e90: 7669 6427 7320 6564 6765 2d6c 656e 6774  vid's edge-lengt
-00007ea0: 6820 7375 6d0a 2020 2020 2020 2020 2020  h sum.          
-00007eb0: 2020 2020 2020 2020 2020 2020 2020 7072                pr
-00007ec0: 696e 7428 2827 4348 4543 4b49 4e47 2045  int(('CHECKING E
-00007ed0: 4447 452d 4c45 4e47 5448 2053 554d 2046  DGE-LENGTH SUM F
-00007ee0: 4f52 2049 4e44 4956 4944 5541 4c20 270a  OR INDIVIDUAL '.
-00007ef0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007f00: 2020 2020 2020 2020 2020 2020 2020 2027                 '
-00007f10: 2569 2729 2025 206f 6666 7370 7269 6e67  %i') % offspring
-00007f20: 5f6b 6579 290a 0a20 2020 2020 2020 2020  _key)..         
-00007f30: 2020 2020 2020 2020 2020 2020 2020 2064                 d
-00007f40: 6566 2067 6574 5f65 6467 6573 5f64 6628  ef get_edges_df(
-00007f50: 7463 293a 0a20 2020 2020 2020 2020 2020  tc):.           
-00007f60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007f70: 2022 2222 0a20 2020 2020 2020 2020 2020   """.           
-00007f80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007f90: 2066 756e 6374 696f 6e20 746f 2067 6574   function to get
-00007fa0: 2070 616e 6461 7320 6466 2066 726f 6d20   pandas df from 
-00007fb0: 6669 7273 740a 2020 2020 2020 2020 2020  first.          
+00007e20: 2020 2020 7061 7265 6e74 3d73 6567 5b30      parent=seg[0
+00007e30: 5d2c 206c 6566 743d 7365 675b 315d 2c20  ], left=seg[1], 
+00007e40: 7269 6768 743d 7365 675b 325d 2c0a 2020  right=seg[2],.  
+00007e50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007e60: 2020 2020 2020 6368 696c 643d 7365 6c66        child=self
+00007e70: 5b6f 6666 7370 7269 6e67 5f6b 6579 5d2e  [offspring_key].
+00007e80: 5f6e 6f64 6573 5f74 6162 5f69 6473 5b68  _nodes_tab_ids[h
+00007e90: 6f6d 6f6c 5d0a 2020 2020 2020 2020 2020  omol].          
+00007ea0: 2020 2020 2020 2020 2020 2020 2020 2920                ) 
+00007eb0: 666f 7220 686f 6d6f 6c2c 2073 6567 5f73  for homol, seg_s
+00007ec0: 6574 2069 6e20 656e 756d 6572 6174 6528  et in enumerate(
+00007ed0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00007ee0: 2020 2020 2020 2020 2067 656e 6f6d 655f           genome_
+00007ef0: 7365 6773 2920 666f 7220 7365 6720 696e  segs) for seg in
+00007f00: 205b 2a73 6567 5f73 6574 5d5d 0a0a 2020   [*seg_set]]..  
+00007f10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007f20: 2020 6966 2063 6865 636b 5f68 6170 733a    if check_haps:
+00007f30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00007f40: 2020 2020 2020 2020 2023 2063 6865 636b           # check
+00007f50: 2074 6865 2069 6e64 6976 6964 2773 2065   the individ's e
+00007f60: 6467 652d 6c65 6e67 7468 2073 756d 0a20  dge-length sum. 
+00007f70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007f80: 2020 2020 2020 2070 7269 6e74 2828 2743         print(('C
+00007f90: 4845 434b 494e 4720 4544 4745 2d4c 454e  HECKING EDGE-LEN
+00007fa0: 4754 4820 5355 4d20 464f 5220 494e 4449  GTH SUM FOR INDI
+00007fb0: 5649 4455 414c 2027 0a20 2020 2020 2020  VIDUAL '.       
 00007fc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007fd0: 2020 3420 636f 6c73 206f 6620 6564 6765    4 cols of edge
-00007fe0: 7320 7461 626c 650a 2020 2020 2020 2020  s table.        
+00007fd0: 2020 2020 2020 2020 2725 6927 2920 2520          '%i') % 
+00007fe0: 6f66 6673 7072 696e 675f 6b65 7929 0a0a  offspring_key)..
 00007ff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008000: 2020 2020 2222 220a 2020 2020 2020 2020      """.        
-00008010: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008020: 2020 2020 6c20 3d20 7463 2e65 6467 6573      l = tc.edges
-00008030: 2e6c 6566 740a 2020 2020 2020 2020 2020  .left.          
+00008000: 2020 2020 2020 2020 6465 6620 6765 745f          def get_
+00008010: 6564 6765 735f 6466 2874 6329 3a0a 2020  edges_df(tc):.  
+00008020: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008030: 2020 2020 2020 2020 2020 2222 220a 2020            """.  
 00008040: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008050: 2020 7220 3d20 7463 2e65 6467 6573 2e72    r = tc.edges.r
-00008060: 6967 6874 0a20 2020 2020 2020 2020 2020  ight.           
-00008070: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008080: 2070 203d 2074 632e 6564 6765 732e 7061   p = tc.edges.pa
-00008090: 7265 6e74 0a20 2020 2020 2020 2020 2020  rent.           
-000080a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000080b0: 2063 203d 2074 632e 6564 6765 732e 6368   c = tc.edges.ch
-000080c0: 696c 640a 2020 2020 2020 2020 2020 2020  ild.            
-000080d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000080e0: 6173 7365 7274 206c 656e 286c 2920 3d3d  assert len(l) ==
-000080f0: 206c 656e 2872 2920 3d3d 206c 656e 2870   len(r) == len(p
-00008100: 2920 3d3d 206c 656e 2863 290a 2020 2020  ) == len(c).    
-00008110: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008120: 2020 2020 2020 2020 6466 203d 2070 642e          df = pd.
-00008130: 4461 7461 4672 616d 652e 6672 6f6d 5f64  DataFrame.from_d
-00008140: 6963 7428 7b27 6c65 6674 273a 206c 2c0a  ict({'left': l,.
-00008150: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008050: 2020 2020 2020 2020 2020 6675 6e63 7469            functi
+00008060: 6f6e 2074 6f20 6765 7420 7061 6e64 6173  on to get pandas
+00008070: 2064 6620 6672 6f6d 2066 6972 7374 0a20   df from first. 
+00008080: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008090: 2020 2020 2020 2020 2020 2034 2063 6f6c             4 col
+000080a0: 7320 6f66 2065 6467 6573 2074 6162 6c65  s of edges table
+000080b0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000080c0: 2020 2020 2020 2020 2020 2020 2022 2222               """
+000080d0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000080e0: 2020 2020 2020 2020 2020 2020 206c 203d               l =
+000080f0: 2074 632e 6564 6765 732e 6c65 6674 0a20   tc.edges.left. 
+00008100: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008110: 2020 2020 2020 2020 2020 2072 203d 2074             r = t
+00008120: 632e 6564 6765 732e 7269 6768 740a 2020  c.edges.right.  
+00008130: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008140: 2020 2020 2020 2020 2020 7020 3d20 7463            p = tc
+00008150: 2e65 6467 6573 2e70 6172 656e 740a 2020  .edges.parent.  
 00008160: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008170: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008180: 2020 2020 2020 2020 2027 7269 6768 7427           'right'
-00008190: 3a20 722c 0a20 2020 2020 2020 2020 2020  : r,.           
-000081a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000081b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000081c0: 2020 2020 2020 2020 2020 2020 2020 2770                'p
-000081d0: 6172 656e 7427 3a20 702c 0a20 2020 2020  arent': p,.     
+00008170: 2020 2020 2020 2020 2020 6320 3d20 7463            c = tc
+00008180: 2e65 6467 6573 2e63 6869 6c64 0a20 2020  .edges.child.   
+00008190: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000081a0: 2020 2020 2020 2020 2061 7373 6572 7420           assert 
+000081b0: 6c65 6e28 6c29 203d 3d20 6c65 6e28 7229  len(l) == len(r)
+000081c0: 203d 3d20 6c65 6e28 7029 203d 3d20 6c65   == len(p) == le
+000081d0: 6e28 6329 0a20 2020 2020 2020 2020 2020  n(c).           
 000081e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000081f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008200: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008210: 2020 2020 2763 6869 6c64 273a 2063 7d29      'child': c})
-00008220: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00008230: 2020 2020 2020 2020 2020 2020 2072 6574               ret
-00008240: 7572 6e20 6466 0a0a 2020 2020 2020 2020  urn df..        
-00008250: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008260: 6e6f 6465 735f 7461 625f 6964 7320 3d20  nodes_tab_ids = 
-00008270: 7365 6c66 5b6f 6666 7370 7269 6e67 5f6b  self[offspring_k
-00008280: 6579 5d2e 5f6e 6f64 6573 5f74 6162 5f69  ey]._nodes_tab_i
-00008290: 6473 0a20 2020 2020 2020 2020 2020 2020  ds.             
-000082a0: 2020 2020 2020 2020 2020 2066 6f72 2069             for i
-000082b0: 2c20 6e6f 6465 735f 7461 625f 6964 2069  , nodes_tab_id i
-000082c0: 6e20 6e6f 6465 735f 7461 625f 6964 732e  n nodes_tab_ids.
-000082d0: 6974 656d 7328 293a 0a20 2020 2020 2020  items():.       
-000082e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000082f0: 2020 2020 2065 6466 203d 2067 6574 5f65       edf = get_e
-00008300: 6467 6573 5f64 6628 7365 6c66 2e5f 7463  dges_df(self._tc
-00008310: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-00008320: 2020 2020 2020 2020 2020 2020 2020 6368                ch
-00008330: 696c 645f 6564 6620 3d20 6564 665b 6564  ild_edf = edf[ed
-00008340: 662e 6368 696c 6420 3d3d 206e 6f64 6573  f.child == nodes
-00008350: 5f74 6162 5f69 645d 0a20 2020 2020 2020  _tab_id].       
+000081f0: 2064 6620 3d20 7064 2e44 6174 6146 7261   df = pd.DataFra
+00008200: 6d65 2e66 726f 6d5f 6469 6374 287b 276c  me.from_dict({'l
+00008210: 6566 7427 3a20 6c2c 0a20 2020 2020 2020  eft': l,.       
+00008220: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008230: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008240: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008250: 2020 2772 6967 6874 273a 2072 2c0a 2020    'right': r,.  
+00008260: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008270: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008280: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008290: 2020 2020 2020 2027 7061 7265 6e74 273a         'parent':
+000082a0: 2070 2c0a 2020 2020 2020 2020 2020 2020   p,.            
+000082b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000082c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000082d0: 2020 2020 2020 2020 2020 2020 2027 6368               'ch
+000082e0: 696c 6427 3a20 637d 290a 2020 2020 2020  ild': c}).      
+000082f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008300: 2020 2020 2020 7265 7475 726e 2064 660a        return df.
+00008310: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00008320: 2020 2020 2020 2020 206e 6f64 6573 5f74           nodes_t
+00008330: 6162 5f69 6473 203d 2073 656c 665b 6f66  ab_ids = self[of
+00008340: 6673 7072 696e 675f 6b65 795d 2e5f 6e6f  fspring_key]._no
+00008350: 6465 735f 7461 625f 6964 730a 2020 2020  des_tab_ids.    
 00008360: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008370: 2020 2020 2061 7373 6572 7420 6e70 2e61       assert np.a
-00008380: 6c6c 636c 6f73 6528 7375 6d28 6368 696c  llclose(sum(chil
-00008390: 645f 6564 662e 7269 6768 7420 2d0a 2020  d_edf.right -.  
-000083a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000083b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000083c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000083d0: 2063 6869 6c64 5f65 6466 2e6c 6566 7429   child_edf.left)
-000083e0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-000083f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008400: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008410: 2073 656c 662e 6765 6e5f 6172 6368 2e4c   self.gen_arch.L
-00008420: 292c 2028 7375 6d28 0a20 2020 2020 2020  ), (sum(.       
-00008430: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008440: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008450: 2020 2020 2063 6869 6c64 5f65 6466 2e72       child_edf.r
-00008460: 6967 6874 202d 2063 6869 6c64 5f65 6466  ight - child_edf
-00008470: 2e6c 6566 7429 2c0a 2020 2020 2020 2020  .left),.        
+00008370: 2020 2020 666f 7220 692c 206e 6f64 6573      for i, nodes
+00008380: 5f74 6162 5f69 6420 696e 206e 6f64 6573  _tab_id in nodes
+00008390: 5f74 6162 5f69 6473 2e69 7465 6d73 2829  _tab_ids.items()
+000083a0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+000083b0: 2020 2020 2020 2020 2020 2020 2020 6564                ed
+000083c0: 6620 3d20 6765 745f 6564 6765 735f 6466  f = get_edges_df
+000083d0: 2873 656c 662e 5f74 6329 0a20 2020 2020  (self._tc).     
+000083e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000083f0: 2020 2020 2020 2063 6869 6c64 5f65 6466         child_edf
+00008400: 203d 2065 6466 5b65 6466 2e63 6869 6c64   = edf[edf.child
+00008410: 203d 3d20 6e6f 6465 735f 7461 625f 6964   == nodes_tab_id
+00008420: 5d0a 2020 2020 2020 2020 2020 2020 2020  ].              
+00008430: 2020 2020 2020 2020 2020 2020 2020 6173                as
+00008440: 7365 7274 206e 702e 616c 6c63 6c6f 7365  sert np.allclose
+00008450: 2873 756d 2863 6869 6c64 5f65 6466 2e72  (sum(child_edf.r
+00008460: 6967 6874 202d 0a20 2020 2020 2020 2020  ight -.         
+00008470: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00008480: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008490: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000084a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000084b0: 2020 2020 2020 2020 2020 6368 696c 645f            child_
-000084c0: 6564 6629 0a0a 2020 2020 2020 2020 2020  edf)..          
-000084d0: 2020 2020 2020 2020 2020 6966 2063 6865            if che
-000084e0: 636b 5f67 656e 6f74 7970 6573 3a0a 2020  ck_genotypes:.  
-000084f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008500: 2020 2020 2020 7365 6c66 2e5f 736f 7274        self._sort
-00008510: 5f73 696d 706c 6966 795f 7461 626c 655f  _simplify_table_
-00008520: 636f 6c6c 6563 7469 6f6e 2829 0a20 2020  collection().   
-00008530: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008540: 2020 2020 2074 7320 3d20 7365 6c66 2e5f       ts = self._
-00008550: 7463 2e74 7265 655f 7365 7175 656e 6365  tc.tree_sequence
-00008560: 2829 0a20 2020 2020 2020 2020 2020 2020  ().             
-00008570: 2020 2020 2020 2020 2020 206d 616e 7561             manua
-00008580: 6c5f 6774 203d 2073 656c 665b 6f66 6673  l_gt = self[offs
-00008590: 7072 696e 675f 6b65 795d 2e67 0a20 2020  pring_key].g.   
-000085a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000085b0: 2020 2020 2074 736b 6974 5f67 7473 5f64       tskit_gts_d
-000085c0: 6963 7420 3d20 6469 6374 287a 6970 2874  ict = dict(zip(t
-000085d0: 732e 7361 6d70 6c65 7328 292c 0a20 2020  s.samples(),.   
-000085e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000085f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008600: 2020 2020 2020 2020 2020 2020 2020 2074                 t
-00008610: 732e 6861 706c 6f74 7970 6573 2829 2929  s.haplotypes()))
-00008620: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00008630: 2020 2020 2020 2020 2074 736b 6974 5f67           tskit_g
-00008640: 7473 203d 205b 7473 6b69 745f 6774 735f  ts = [tskit_gts_
-00008650: 6469 6374 5b0a 2020 2020 2020 2020 2020  dict[.          
-00008660: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008670: 2020 6e6f 6465 5f69 645d 2066 6f72 206e    node_id] for n
-00008680: 6f64 655f 6964 2069 6e20 7365 6c66 5b0a  ode_id in self[.
-00008690: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000086a0: 2020 2020 2020 2020 2020 2020 6f66 6673              offs
-000086b0: 7072 696e 675f 6b65 795d 2e5f 6e6f 6465  pring_key]._node
-000086c0: 735f 7461 625f 6964 732e 7661 6c75 6573  s_tab_ids.values
-000086d0: 2829 5d0a 2020 2020 2020 2020 2020 2020  ()].            
-000086e0: 2020 2020 2020 2020 2020 2020 7473 6b69              tski
-000086f0: 745f 6774 203d 206e 702e 7374 6163 6b28  t_gt = np.stack(
-00008700: 5b5b 696e 7428 6e29 2066 6f72 206e 2069  [[int(n) for n i
-00008710: 6e20 6774 5d20 666f 7220 6774 2069 6e0a  n gt] for gt in.
-00008720: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008730: 2020 2020 2020 2020 2020 2020 7473 6b69              tski
-00008740: 745f 6774 735d 292e 545b 7365 6c66 2e67  t_gts]).T[self.g
-00008750: 656e 5f61 7263 682e 6e6f 6e6e 6575 745f  en_arch.nonneut_
-00008760: 6c6f 6369 2c20 3a5d 0a20 2020 2020 2020  loci, :].       
-00008770: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008780: 2070 7269 6e74 2827 4f46 4653 5052 494e   print('OFFSPRIN
-00008790: 4720 2569 5c6e 5c6e 2720 2520 6f66 6673  G %i\n\n' % offs
-000087a0: 7072 696e 675f 6b65 7929 0a20 2020 2020  pring_key).     
-000087b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000087c0: 2020 2073 6567 735f 3020 3d20 5b2a 2073     segs_0 = [* s
-000087d0: 656c 662e 6765 6e5f 6172 6368 2e72 6563  elf.gen_arch.rec
-000087e0: 6f6d 6269 6e61 7469 6f6e 732e 5f67 6574  ombinations._get
-000087f0: 5f73 6567 5f69 6e66 6f28 0a20 2020 2020  _seg_info(.     
-00008800: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008810: 2020 2020 2020 2020 2020 2020 2073 7461               sta
-00008820: 7274 5f68 6f6d 6f6c 6f67 7565 3d30 2c0a  rt_homologue=0,.
-00008830: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008840: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008850: 2020 6576 656e 745f 6b65 793d 7265 636f    event_key=reco
-00008860: 6d62 5f6b 6579 735b 6774 5f63 6865 636b  mb_keys[gt_check
-00008870: 5f6e 756d 5d2c 0a20 2020 2020 2020 2020  _num],.         
+00008490: 2020 2020 2020 2020 2020 6368 696c 645f            child_
+000084a0: 6564 662e 6c65 6674 292c 0a20 2020 2020  edf.left),.     
+000084b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000084c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000084d0: 2020 2020 2020 2020 2020 7365 6c66 2e67            self.g
+000084e0: 656e 5f61 7263 682e 4c29 2c20 2873 756d  en_arch.L), (sum
+000084f0: 280a 2020 2020 2020 2020 2020 2020 2020  (.              
+00008500: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008510: 2020 2020 2020 2020 2020 2020 2020 6368                ch
+00008520: 696c 645f 6564 662e 7269 6768 7420 2d20  ild_edf.right - 
+00008530: 6368 696c 645f 6564 662e 6c65 6674 292c  child_edf.left),
+00008540: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00008550: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008560: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008570: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008580: 2020 2063 6869 6c64 5f65 6466 290a 0a20     child_edf).. 
+00008590: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000085a0: 2020 2069 6620 6368 6563 6b5f 6765 6e6f     if check_geno
+000085b0: 7479 7065 733a 0a20 2020 2020 2020 2020  types:.         
+000085c0: 2020 2020 2020 2020 2020 2020 2020 2073                 s
+000085d0: 656c 662e 5f73 6f72 745f 616e 645f 7369  elf._sort_and_si
+000085e0: 6d70 6c69 6679 5f74 6162 6c65 5f63 6f6c  mplify_table_col
+000085f0: 6c65 6374 696f 6e28 290a 2020 2020 2020  lection().      
+00008600: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008610: 2020 7473 203d 2073 656c 662e 5f74 632e    ts = self._tc.
+00008620: 7472 6565 5f73 6571 7565 6e63 6528 290a  tree_sequence().
+00008630: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008640: 2020 2020 2020 2020 6d61 6e75 616c 5f67          manual_g
+00008650: 7420 3d20 7365 6c66 5b6f 6666 7370 7269  t = self[offspri
+00008660: 6e67 5f6b 6579 5d2e 670a 2020 2020 2020  ng_key].g.      
+00008670: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008680: 2020 7473 6b69 745f 6774 735f 6469 6374    tskit_gts_dict
+00008690: 203d 2064 6963 7428 7a69 7028 7473 2e73   = dict(zip(ts.s
+000086a0: 616d 706c 6573 2829 2c0a 2020 2020 2020  amples(),.      
+000086b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000086c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000086d0: 2020 2020 2020 2020 2020 2020 7473 2e68              ts.h
+000086e0: 6170 6c6f 7479 7065 7328 2929 290a 2020  aplotypes())).  
+000086f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008700: 2020 2020 2020 7473 6b69 745f 6774 7320        tskit_gts 
+00008710: 3d20 5b74 736b 6974 5f67 7473 5f64 6963  = [tskit_gts_dic
+00008720: 745b 0a20 2020 2020 2020 2020 2020 2020  t[.             
+00008730: 2020 2020 2020 2020 2020 2020 2020 206e                 n
+00008740: 6f64 655f 6964 5d20 666f 7220 6e6f 6465  ode_id] for node
+00008750: 5f69 6420 696e 2073 656c 665b 0a20 2020  _id in self[.   
+00008760: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008770: 2020 2020 2020 2020 206f 6666 7370 7269           offspri
+00008780: 6e67 5f6b 6579 5d2e 5f6e 6f64 6573 5f74  ng_key]._nodes_t
+00008790: 6162 5f69 6473 2e76 616c 7565 7328 295d  ab_ids.values()]
+000087a0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000087b0: 2020 2020 2020 2020 2074 736b 6974 5f67           tskit_g
+000087c0: 7420 3d20 6e70 2e73 7461 636b 285b 5b69  t = np.stack([[i
+000087d0: 6e74 286e 2920 666f 7220 6e20 696e 2067  nt(n) for n in g
+000087e0: 745d 2066 6f72 2067 7420 696e 0a20 2020  t] for gt in.   
+000087f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008800: 2020 2020 2020 2020 2074 736b 6974 5f67           tskit_g
+00008810: 7473 5d29 2e54 5b73 656c 662e 6765 6e5f  ts]).T[self.gen_
+00008820: 6172 6368 2e6e 6f6e 6e65 7574 5f6c 6f63  arch.nonneut_loc
+00008830: 692c 203a 5d0a 2020 2020 2020 2020 2020  i, :].          
+00008840: 2020 2020 2020 2020 2020 2020 2020 7072                pr
+00008850: 696e 7428 274f 4646 5350 5249 4e47 2025  int('OFFSPRING %
+00008860: 695c 6e5c 6e27 2025 206f 6666 7370 7269  i\n\n' % offspri
+00008870: 6e67 5f6b 6579 290a 2020 2020 2020 2020  ng_key).        
 00008880: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008890: 2020 2020 2020 2020 206e 6f64 655f 6964           node_id
-000088a0: 733d 6e70 2e61 7272 6179 285b 302c 2031  s=np.array([0, 1
-000088b0: 5d29 295d 0a20 2020 2020 2020 2020 2020  ]))].           
-000088c0: 2020 2020 2020 2020 2020 2020 2070 7269               pri
-000088d0: 6e74 2827 5c74 6765 6e6f 6d65 5f73 6567  nt('\tgenome_seg
-000088e0: 7320 303a 272c 2073 6567 735f 3029 0a20  s 0:', segs_0). 
-000088f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008900: 2020 2020 2020 2073 6567 735f 3120 3d20         segs_1 = 
-00008910: 5b2a 2073 656c 662e 6765 6e5f 6172 6368  [* self.gen_arch
-00008920: 2e72 6563 6f6d 6269 6e61 7469 6f6e 732e  .recombinations.
-00008930: 5f67 6574 5f73 6567 5f69 6e66 6f28 0a20  _get_seg_info(. 
-00008940: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008890: 7365 6773 5f30 203d 205b 2a20 7365 6c66  segs_0 = [* self
+000088a0: 2e67 656e 5f61 7263 682e 7265 636f 6d62  .gen_arch.recomb
+000088b0: 696e 6174 696f 6e73 2e5f 6765 745f 7365  inations._get_se
+000088c0: 675f 696e 666f 280a 2020 2020 2020 2020  g_info(.        
+000088d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000088e0: 2020 2020 2020 2020 2020 7374 6172 745f            start_
+000088f0: 686f 6d6f 6c6f 6775 653d 302c 0a20 2020  homologue=0,.   
+00008900: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008910: 2020 2020 2020 2020 2020 2020 2020 2065                 e
+00008920: 7665 6e74 5f6b 6579 3d72 6563 6f6d 625f  vent_key=recomb_
+00008930: 6b65 7973 5b67 745f 6368 6563 6b5f 6e75  keys[gt_check_nu
+00008940: 6d5d 2c0a 2020 2020 2020 2020 2020 2020  m],.            
 00008950: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008960: 2073 7461 7274 5f68 6f6d 6f6c 6f67 7565   start_homologue
-00008970: 3d30 2c0a 2020 2020 2020 2020 2020 2020  =0,.            
-00008980: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008990: 2020 2020 2020 6576 656e 745f 6b65 793d        event_key=
-000089a0: 7265 636f 6d62 5f6b 6579 735b 6774 5f63  recomb_keys[gt_c
-000089b0: 6865 636b 5f6e 756d 2b31 5d2c 0a20 2020  heck_num+1],.   
+00008960: 2020 2020 2020 6e6f 6465 5f69 6473 3d6e        node_ids=n
+00008970: 702e 6172 7261 7928 5b30 2c20 315d 2929  p.array([0, 1]))
+00008980: 5d0a 2020 2020 2020 2020 2020 2020 2020  ].              
+00008990: 2020 2020 2020 2020 2020 7072 696e 7428            print(
+000089a0: 275c 7467 656e 6f6d 655f 7365 6773 2030  '\tgenome_segs 0
+000089b0: 3a27 2c20 7365 6773 5f30 290a 2020 2020  :', segs_0).    
 000089c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000089d0: 2020 2020 2020 2020 2020 2020 2020 206e                 n
-000089e0: 6f64 655f 6964 733d 6e70 2e61 7272 6179  ode_ids=np.array
-000089f0: 285b 302c 2031 5d29 295d 0a20 2020 2020  ([0, 1]))].     
-00008a00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008a10: 2020 2070 7269 6e74 2827 5c74 6765 6e6f     print('\tgeno
-00008a20: 6d65 5f73 6567 7320 313a 272c 2073 6567  me_segs 1:', seg
-00008a30: 735f 3129 0a20 2020 2020 2020 2020 2020  s_1).           
-00008a40: 2020 2020 2020 2020 2020 2020 2070 7269               pri
-00008a50: 6e74 2827 5c74 6e6f 6e6e 6575 7420 6c6f  nt('\tnonneut lo
-00008a60: 6369 3a27 2c20 7365 6c66 2e67 656e 5f61  ci:', self.gen_a
-00008a70: 7263 682e 6e6f 6e6e 6575 745f 6c6f 6369  rch.nonneut_loci
-00008a80: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-00008a90: 2020 2020 2020 2020 2020 7072 696e 7428            print(
-00008aa0: 2772 6563 6f6d 6220 6174 206e 6f6e 6e65  'recomb at nonne
-00008ab0: 7574 272c 0a20 2020 2020 2020 2020 2020  ut',.           
-00008ac0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008ad0: 206e 702e 616e 7928 5b6e 702e 666c 6f6f   np.any([np.floo
-00008ae0: 7228 695b 325d 0a20 2020 2020 2020 2020  r(i[2].         
-00008af0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008b00: 2020 2029 2069 6e20 7365 6c66 2e67 656e     ) in self.gen
-00008b10: 5f61 7263 682e 6e6f 6e6e 6575 745f 6c6f  _arch.nonneut_lo
-00008b20: 6369 2066 6f72 2069 2069 6e20 2873 6567  ci for i in (seg
-00008b30: 735f 3020 2b0a 2020 2020 2020 2020 2020  s_0 +.          
-00008b40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000089d0: 2020 2020 7365 6773 5f31 203d 205b 2a20      segs_1 = [* 
+000089e0: 7365 6c66 2e67 656e 5f61 7263 682e 7265  self.gen_arch.re
+000089f0: 636f 6d62 696e 6174 696f 6e73 2e5f 6765  combinations._ge
+00008a00: 745f 7365 675f 696e 666f 280a 2020 2020  t_seg_info(.    
+00008a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008a20: 2020 2020 2020 2020 2020 2020 2020 7374                st
+00008a30: 6172 745f 686f 6d6f 6c6f 6775 653d 302c  art_homologue=0,
+00008a40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00008a50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008a60: 2020 2065 7665 6e74 5f6b 6579 3d72 6563     event_key=rec
+00008a70: 6f6d 625f 6b65 7973 5b67 745f 6368 6563  omb_keys[gt_chec
+00008a80: 6b5f 6e75 6d2b 315d 2c0a 2020 2020 2020  k_num+1],.      
+00008a90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008aa0: 2020 2020 2020 2020 2020 2020 6e6f 6465              node
+00008ab0: 5f69 6473 3d6e 702e 6172 7261 7928 5b30  _ids=np.array([0
+00008ac0: 2c20 315d 2929 5d0a 2020 2020 2020 2020  , 1]))].        
+00008ad0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008ae0: 7072 696e 7428 275c 7467 656e 6f6d 655f  print('\tgenome_
+00008af0: 7365 6773 2031 3a27 2c20 7365 6773 5f31  segs 1:', segs_1
+00008b00: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00008b10: 2020 2020 2020 2020 2020 7072 696e 7428            print(
+00008b20: 275c 746e 6f6e 6e65 7574 206c 6f63 693a  '\tnonneut loci:
+00008b30: 272c 2073 656c 662e 6765 6e5f 6172 6368  ', self.gen_arch
+00008b40: 2e6e 6f6e 6e65 7574 5f6c 6f63 6929 0a20  .nonneut_loci). 
 00008b50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008b60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008b70: 2020 2020 2020 2020 2020 2020 7365 6773              segs
-00008b80: 5f31 295d 2929 0a20 2020 2020 2020 2020  _1)])).         
-00008b90: 2020 2020 2020 2020 2020 2020 2020 2061                 a
-00008ba0: 7373 6572 7420 6e70 2e61 6c6c 286d 616e  ssert np.all(man
-00008bb0: 7561 6c5f 6774 203d 3d20 7473 6b69 745f  ual_gt == tskit_
-00008bc0: 6774 292c 2028 226d 616e 7561 6c6c 7920  gt), ("manually 
-00008bd0: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
-00008be0: 2020 2020 2020 2020 2020 2020 2020 2274                "t
-00008bf0: 7261 636b 6564 2067 656e 6f74 7970 6520  racked genotype 
-00008c00: 6469 6666 6572 7320 6672 6f6d 2067 656e  differs from gen
-00008c10: 6f74 7970 6520 7265 636f 7264 6564 2022  otype recorded "
-00008c20: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00008c30: 2020 2020 2020 2020 2020 2020 2022 696e               "in
-00008c40: 2074 736b 6974 2054 6162 6c65 436f 6c6c   tskit TableColl
-00008c50: 6563 7469 6f6e 3a5c 6e5c 6e4d 414e 5541  ection:\n\nMANUA
-00008c60: 4c5c 6e5c 7425 7322 0a20 2020 2020 2020  L\n\t%s".       
-00008c70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008c80: 2020 2020 2022 5c6e 5c6e 5453 4b49 545c       "\n\nTSKIT\
-00008c90: 6e5c 7425 7322 2920 2520 2873 7472 286d  n\t%s") % (str(m
-00008ca0: 616e 7561 6c5f 6774 292c 0a20 2020 2020  anual_gt),.     
-00008cb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008cd0: 2020 2020 2020 2020 2020 2020 2073 7472               str
-00008ce0: 2874 736b 6974 5f67 7429 290a 2020 2020  (tskit_gt)).    
+00008b60: 2020 2020 2020 2070 7269 6e74 2827 7265         print('re
+00008b70: 636f 6d62 2061 7420 6e6f 6e6e 6575 7427  comb at nonneut'
+00008b80: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00008b90: 2020 2020 2020 2020 2020 2020 2020 6e70                np
+00008ba0: 2e61 6e79 285b 6e70 2e66 6c6f 6f72 2869  .any([np.floor(i
+00008bb0: 5b32 5d0a 2020 2020 2020 2020 2020 2020  [2].            
+00008bc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008bd0: 2920 696e 2073 656c 662e 6765 6e5f 6172  ) in self.gen_ar
+00008be0: 6368 2e6e 6f6e 6e65 7574 5f6c 6f63 6920  ch.nonneut_loci 
+00008bf0: 666f 7220 6920 696e 2028 7365 6773 5f30  for i in (segs_0
+00008c00: 202b 0a20 2020 2020 2020 2020 2020 2020   +.             
+00008c10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008c30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008c40: 2020 2020 2020 2020 2073 6567 735f 3129           segs_1)
+00008c50: 5d29 290a 2020 2020 2020 2020 2020 2020  ])).            
+00008c60: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+00008c70: 7274 206e 702e 616c 6c28 6d61 6e75 616c  rt np.all(manual
+00008c80: 5f67 7420 3d3d 2074 736b 6974 5f67 7429  _gt == tskit_gt)
+00008c90: 2c20 2822 6d61 6e75 616c 6c79 2022 0a20  , ("manually ". 
+00008ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008cb0: 2020 2020 2020 2020 2020 2022 7472 6163             "trac
+00008cc0: 6b65 6420 6765 6e6f 7479 7065 2064 6966  ked genotype dif
+00008cd0: 6665 7273 2066 726f 6d20 6765 6e6f 7479  fers from genoty
+00008ce0: 7065 2072 6563 6f72 6465 6420 220a 2020  pe recorded ".  
 00008cf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008d00: 2020 2020 6774 5f63 6865 636b 5f6e 756d      gt_check_num
-00008d10: 202b 3d20 320a 0a20 2020 2020 2020 2023   += 2..        #
-00008d20: 2073 616d 706c 6520 616c 6c20 696e 6469   sample all indi
-00008d30: 7669 6475 616c 7327 2065 6e76 6972 6f6e  viduals' environ
-00008d40: 6d65 6e74 2076 616c 7565 732c 2074 6f20  ment values, to 
-00008d50: 696e 6974 6961 7465 2066 6f72 206f 6666  initiate for off
-00008d60: 7370 7269 6e67 0a20 2020 2020 2020 2073  spring.        s
-00008d70: 656c 662e 5f73 6574 5f65 286c 616e 6429  elf._set_e(land)
-00008d80: 0a20 2020 2020 2020 2073 656c 662e 5f73  .        self._s
-00008d90: 6574 5f63 6f6f 7264 735f 616e 645f 6365  et_coords_and_ce
-00008da0: 6c6c 7328 290a 0a20 2020 2020 2020 2023  lls()..        #
-00008db0: 2064 6f20 6d75 7461 7469 6f6e 2069 6620   do mutation if 
-00008dc0: 6e65 6365 7373 6172 790a 2020 2020 2020  necessary.      
-00008dd0: 2020 6966 2073 656c 662e 6d75 7461 7465    if self.mutate
-00008de0: 2061 6e64 206e 6f74 2062 7572 6e3a 0a20   and not burn:. 
-00008df0: 2020 2020 2020 2020 2020 2020 5f64 6f5f              _do_
-00008e00: 6d75 7461 7469 6f6e 286b 6579 735f 6c69  mutation(keys_li
-00008e10: 7374 2c20 7365 6c66 2c20 6c6f 6720 3d20  st, self, log = 
-00008e20: 7365 6c66 2e6d 7574 5f6c 6f67 290a 0a20  self.mut_log).. 
-00008e30: 2020 2020 2020 2023 2063 6865 636b 2068         # check h
-00008e40: 6170 6c6f 7974 7065 7320 6861 7665 206e  aploytpes have n
-00008e50: 6f20 6d69 7373 696e 6720 6461 7461 2c20  o missing data, 
-00008e60: 6966 206e 6563 6573 7361 7279 0a20 2020  if necessary.   
-00008e70: 2020 2020 2069 6620 6368 6563 6b5f 6861       if check_ha
-00008e80: 7073 2061 6e64 2073 656c 662e 6275 726e  ps and self.burn
-00008e90: 6564 2061 6e64 2073 656c 662e 6765 6e5f  ed and self.gen_
-00008ea0: 6172 6368 2e75 7365 5f74 736b 6974 3a0a  arch.use_tskit:.
-00008eb0: 2020 2020 2020 2020 2020 2020 7072 696e              prin
-00008ec0: 7428 275c 6e5c 6e43 4845 434b 494e 4720  t('\n\nCHECKING 
-00008ed0: 5448 4154 2048 4150 4c4f 5459 5045 5320  THAT HAPLOTYPES 
-00008ee0: 4841 5645 204e 4f20 4d49 5353 494e 4720  HAVE NO MISSING 
-00008ef0: 4441 5441 5c6e 5c6e 2729 0a20 2020 2020  DATA\n\n').     
-00008f00: 2020 2020 2020 2073 656c 662e 5f73 6f72         self._sor
-00008f10: 745f 7369 6d70 6c69 6679 5f74 6162 6c65  t_simplify_table
-00008f20: 5f63 6f6c 6c65 6374 696f 6e28 290a 2020  _collection().  
-00008f30: 2020 2020 2020 2020 2020 7473 203d 2073            ts = s
-00008f40: 656c 662e 5f74 632e 7472 6565 5f73 6571  elf._tc.tree_seq
-00008f50: 7565 6e63 6528 290a 2020 2020 2020 2020  uence().        
-00008f60: 2020 2020 6861 7073 203d 205b 2a74 732e      haps = [*ts.
-00008f70: 6861 706c 6f74 7970 6573 2829 5d0a 2020  haplotypes()].  
-00008f80: 2020 2020 2020 2020 2020 6d69 7373 696e            missin
-00008f90: 6720 3d20 5b27 2d27 2069 6e20 6861 7020  g = ['-' in hap 
-00008fa0: 666f 7220 6861 7020 696e 2068 6170 735d  for hap in haps]
-00008fb0: 0a20 2020 2020 2020 2020 2020 2061 7373  .            ass
-00008fc0: 6572 7420 7375 6d28 6d69 7373 696e 6729  ert sum(missing)
-00008fd0: 203d 3d20 302c 2027 736f 6d65 2068 6170   == 0, 'some hap
-00008fe0: 6c6f 7479 7065 7320 6861 7665 206d 6973  lotypes have mis
-00008ff0: 7369 6e67 2064 6174 6121 270a 0a0a 2020  sing data!'...  
-00009000: 2020 236d 6574 686f 6420 746f 2064 6f20    #method to do 
-00009010: 7370 6563 6965 7320 6479 6e61 6d69 6373  species dynamics
-00009020: 0a20 2020 2064 6566 205f 646f 5f70 6f70  .    def _do_pop
-00009030: 5f64 796e 616d 6963 7328 7365 6c66 2c20  _dynamics(self, 
-00009040: 6c61 6e64 293a 0a20 2020 2020 2020 2023  land):.        #
-00009050: 696d 706c 656d 656e 7420 7365 6c65 6374  implement select
-00009060: 696f 6e2c 2069 6666 2073 656c 662e 7365  ion, iff self.se
-00009070: 6c65 6374 696f 6e20 6973 2054 7275 6520  lection is True 
-00009080: 616e 6420 7468 6520 7370 7020 6861 730a  and the spp has.
-00009090: 2020 2020 2020 2020 2361 6c72 6561 6479          #already
-000090a0: 2062 6565 6e20 6275 726e 6564 2069 6e0a   been burned in.
-000090b0: 2020 2020 2020 2020 7769 7468 5f73 656c          with_sel
-000090c0: 6563 7469 6f6e 203d 2073 656c 662e 7365  ection = self.se
-000090d0: 6c65 6374 696f 6e20 616e 6420 7365 6c66  lection and self
-000090e0: 2e62 7572 6e65 640a 2020 2020 2020 2020  .burned.        
-000090f0: 6275 726e 203d 206e 6f74 2073 656c 662e  burn = not self.
-00009100: 6275 726e 6564 0a20 2020 2020 2020 2023  burned.        #
-00009110: 7468 656e 2063 6172 7279 206f 7574 2074  then carry out t
-00009120: 6865 2070 6f70 2d64 796e 616d 6963 732c  he pop-dynamics,
-00009130: 2077 6974 6820 7365 6c65 6374 696f 6e20   with selection 
-00009140: 6173 2073 6574 2061 626f 7665 2c20 616e  as set above, an
-00009150: 6420 7361 7665 0a20 2020 2020 2020 2023  d save.        #
-00009160: 7265 7375 6c74 2c20 7768 6963 6820 7769  result, which wi
-00009170: 6c6c 2062 6520 5472 7565 2069 6666 2073  ll be True iff s
-00009180: 7070 2068 6173 2067 6f6e 6520 6578 7469  pp has gone exti
-00009190: 6e63 740a 2020 2020 2020 2020 6578 7469  nct.        exti
-000091a0: 6e63 7420 3d20 5f64 6f5f 706f 705f 6479  nct = _do_pop_dy
-000091b0: 6e61 6d69 6373 2873 656c 662c 206c 616e  namics(self, lan
-000091c0: 642c 0a20 2020 2020 2020 2020 2020 2077  d,.            w
-000091d0: 6974 685f 7365 6c65 6374 696f 6e20 3d20  ith_selection = 
-000091e0: 7769 7468 5f73 656c 6563 7469 6f6e 2c20  with_selection, 
-000091f0: 6275 726e 203d 2062 7572 6e29 0a20 2020  burn = burn).   
-00009200: 2020 2020 2069 6620 6578 7469 6e63 743a       if extinct:
-00009210: 0a20 2020 2020 2020 2020 2020 2023 7365  .            #se
-00009220: 7420 7365 6c66 2e65 7874 696e 6374 2065  t self.extinct e
-00009230: 7175 616c 2074 6f20 5472 7565 2c20 736f  qual to True, so
-00009240: 2074 6861 7420 7468 6520 6974 6572 6174   that the iterat
-00009250: 696f 6e20 7769 6c6c 2065 6e64 0a20 2020  ion will end.   
-00009260: 2020 2020 2020 2020 2073 656c 662e 6578           self.ex
-00009270: 7469 6e63 7420 3d20 6578 7469 6e63 740a  tinct = extinct.
-00009280: 0a20 2020 2023 6d65 7468 6f64 2074 6f20  .    #method to 
-00009290: 6d61 6b65 2073 7065 6369 6573 2063 6861  make species cha
-000092a0: 6e67 6573 0a20 2020 2064 6566 205f 6d61  nges.    def _ma
-000092b0: 6b65 5f63 6861 6e67 6528 7365 6c66 2c20  ke_change(self, 
-000092c0: 7665 7262 6f73 653d 4661 6c73 6529 3a0a  verbose=False):.
-000092d0: 2020 2020 2020 2020 7365 6c66 2e5f 6368          self._ch
-000092e0: 616e 6765 722e 5f6d 616b 655f 6368 616e  anger._make_chan
-000092f0: 6765 2874 203d 2073 656c 662e 742c 2061  ge(t = self.t, a
-00009300: 6464 6974 696f 6e61 6c5f 6172 6773 203d  dditional_args =
-00009310: 207b 2773 7070 273a 2073 656c 667d 2c0a   {'spp': self},.
-00009320: 2020 2020 2020 2020 2020 2020 7665 7262              verb
-00009330: 6f73 6520 3d20 7665 7262 6f73 6529 0a0a  ose = verbose)..
-00009340: 2020 2020 236d 6574 686f 6420 746f 2063      #method to c
-00009350: 6865 636b 2069 6620 7468 6520 7370 6563  heck if the spec
-00009360: 6965 7320 6861 7320 676f 6e65 2065 7874  ies has gone ext
-00009370: 696e 6374 0a20 2020 2064 6566 205f 6368  inct.    def _ch
-00009380: 6563 6b5f 6578 7469 6e63 7428 7365 6c66  eck_extinct(self
-00009390: 293a 0a20 2020 2020 2020 2072 6574 7572  ):.        retur
-000093a0: 6e20 6c65 6e28 7365 6c66 2920 3d3d 2030  n len(self) == 0
-000093b0: 0a0a 2020 2020 236d 6574 686f 6420 746f  ..    #method to
-000093c0: 2063 616c 6375 6c61 7465 2073 7065 6369   calculate speci
-000093d0: 6573 2064 656e 7369 7479 0a20 2020 2064  es density.    d
-000093e0: 6566 205f 6361 6c63 5f64 656e 7369 7479  ef _calc_density
-000093f0: 2873 656c 662c 206e 6f72 6d61 6c69 7a65  (self, normalize
-00009400: 203d 2046 616c 7365 2c20 6173 5f6c 6179   = False, as_lay
-00009410: 6572 203d 2046 616c 7365 2c20 7365 745f  er = False, set_
-00009420: 4e3d 4661 6c73 6529 3a0a 0a20 2020 2020  N=False):..     
-00009430: 2020 2027 2727 0a20 2020 2020 2020 2043     '''.        C
-00009440: 616c 6375 6c61 7465 2061 6e20 696e 7465  alculate an inte
-00009450: 7270 6f6c 6174 6564 2072 6173 7465 7220  rpolated raster 
-00009460: 6f66 206c 6f63 616c 2073 7065 6369 6573  of local species
-00009470: 2064 656e 7369 7479 2c20 7573 696e 670a   density, using.
-00009480: 2020 2020 2020 2020 7468 6520 7370 6174          the spat
-00009490: 6961 6c2e 5f44 656e 7369 7479 4772 6964  ial._DensityGrid
-000094a0: 5374 6163 6b20 6f62 6a65 6374 2073 746f  Stack object sto
-000094b0: 7265 6420 696e 2074 6865 0a20 2020 2020  red in the.     
-000094c0: 2020 2053 7065 6369 6573 2e5f 6465 6e73     Species._dens
-000094d0: 5f67 7269 6473 2061 7474 7269 6275 7465  _grids attribute
-000094e0: 2e0a 0a20 2020 2020 2020 2049 6620 6e6f  ...        If no
-000094f0: 726d 616c 697a 6520 6973 2054 7275 652c  rmalize is True,
-00009500: 2074 6865 2064 656e 7369 7479 2072 6173   the density ras
-00009510: 7465 7220 7769 6c6c 2076 6172 7920 6265  ter will vary be
-00009520: 7477 6565 6e20 3020 616e 6420 310a 2020  tween 0 and 1.  
-00009530: 2020 2020 2020 2861 6674 6572 2062 6569        (after bei
-00009540: 6e67 206e 6f72 6d61 6c69 7a65 6420 6279  ng normalized by
-00009550: 2069 7473 206f 776e 206d 6178 696d 756d   its own maximum
-00009560: 2076 616c 7565 292e 2049 6620 6661 6c73   value). If fals
-00009570: 652c 2069 7420 7769 6c6c 0a20 2020 2020  e, it will.     
-00009580: 2020 2076 6172 7920 6265 7477 6565 6e20     vary between 
-00009590: 3020 616e 6420 6974 7320 6d61 7869 6d75  0 and its maximu
-000095a0: 6d20 6573 7469 6d61 7465 6420 6465 6e73  m estimated dens
-000095b0: 6974 7920 7661 6c75 652e 0a20 2020 2020  ity value..     
-000095c0: 2020 2027 2727 0a20 2020 2020 2020 2023     '''.        #
-000095d0: 6368 6563 6b20 7661 6c69 6469 7479 206f  check validity o
-000095e0: 6620 6e6f 726d 616c 697a 6520 6172 6775  f normalize argu
-000095f0: 6d65 6e74 0a20 2020 2020 2020 2061 7373  ment.        ass
-00009600: 6572 7420 7479 7065 286e 6f72 6d61 6c69  ert type(normali
-00009610: 7a65 2920 6973 2062 6f6f 6c2c 2028 2254  ze) is bool, ("T
-00009620: 6865 2027 6e6f 726d 616c 697a 6527 2061  he 'normalize' a
-00009630: 7267 756d 656e 7420 7461 6b65 7320 220a  rgument takes ".
-00009640: 2020 2020 2020 2020 2020 2020 2261 2062              "a b
-00009650: 6f6f 6c65 616e 2076 616c 7565 2e5c 6e22  oolean value.\n"
-00009660: 290a 2020 2020 2020 2020 2367 6574 2073  ).        #get s
-00009670: 7065 6369 6573 2063 6f6f 7264 696e 6174  pecies coordinat
-00009680: 6573 0a20 2020 2020 2020 2078 203d 2073  es.        x = s
-00009690: 656c 662e 5f67 6574 5f78 2829 0a20 2020  elf._get_x().   
-000096a0: 2020 2020 2079 203d 2073 656c 662e 5f67       y = self._g
-000096b0: 6574 5f79 2829 0a20 2020 2020 2020 2023  et_y().        #
-000096c0: 6361 6c63 756c 6174 6520 7468 6520 6465  calculate the de
-000096d0: 6e73 6974 7920 6172 7261 790a 2020 2020  nsity array.    
-000096e0: 2020 2020 6465 6e73 203d 2073 656c 662e      dens = self.
-000096f0: 5f64 656e 735f 6772 6964 732e 5f63 616c  _dens_grids._cal
-00009700: 635f 6465 6e73 6974 7928 782c 2079 290a  c_density(x, y).
-00009710: 2020 2020 2020 2020 2373 6574 206d 696e          #set min
-00009720: 2076 616c 7565 2074 6f20 300a 2020 2020   value to 0.    
-00009730: 2020 2020 6465 6e73 203d 206e 702e 636c      dens = np.cl
-00009740: 6970 2864 656e 732c 2061 5f6d 696e 203d  ip(dens, a_min =
-00009750: 2030 2c20 615f 6d61 7820 3d20 4e6f 6e65   0, a_max = None
-00009760: 290a 2020 2020 2020 2020 236e 6f72 6d61  ).        #norma
-00009770: 6c69 7a65 2c20 6966 206e 6563 6573 7361  lize, if necessa
-00009780: 7279 200a 2020 2020 2020 2020 6966 206e  ry .        if n
-00009790: 6f72 6d61 6c69 7a65 3a0a 2020 2020 2020  ormalize:.      
-000097a0: 2020 2020 2020 2320 5573 6520 6d61 785f        # Use max_
-000097b0: 7661 6c20 746f 206e 6f72 6d61 6c69 7a65  val to normalize
-000097c0: 2074 6865 2064 656e 7369 7479 2072 6173   the density ras
-000097d0: 7465 7220 746f 2065 6974 6865 720a 2020  ter to either.  
-000097e0: 2020 2020 2020 2020 2020 2330 2074 6f20            #0 to 
-000097f0: 6974 7320 6375 7272 656e 7420 6d61 7820  its current max 
-00009800: 7661 6c20 6f72 2030 2074 6f20 312c 2074  val or 0 to 1, t
-00009810: 6f20 6d61 6b65 2073 7572 6520 0a20 2020  o make sure .   
-00009820: 2020 2020 2020 2020 2023 7468 6520 696e           #the in
-00009830: 7465 7270 6f6c 6174 696f 6e20 6469 646e  terpolation didn
-00009840: 2774 2067 656e 6572 6174 6520 616e 7920  't generate any 
-00009850: 7661 6c75 6573 2073 6c69 6768 746c 790a  values slightly.
-00009860: 2020 2020 2020 2020 2020 2020 236f 7574              #out
-00009870: 7369 6465 2074 6869 7320 7261 6e67 650a  side this range.
-00009880: 2020 2020 2020 2020 2020 2020 6e6f 726d              norm
-00009890: 5f66 6163 746f 7220 3d20 6465 6e73 2e6d  _factor = dens.m
-000098a0: 6178 2829 202d 2064 656e 732e 6d69 6e28  ax() - dens.min(
-000098b0: 290a 2020 2020 2020 2020 2020 2020 6465  ).            de
-000098c0: 6e73 203d 2028 6465 6e73 202d 2064 656e  ns = (dens - den
-000098d0: 732e 6d69 6e28 2929 202f 206e 6f72 6d5f  s.min()) / norm_
-000098e0: 6661 6374 6f72 0a20 2020 2020 2020 2023  factor.        #
-000098f0: 7265 7475 726e 2061 7320 6c61 7965 722c  return as layer,
-00009900: 2069 6620 6e65 6365 7373 6172 790a 2020   if necessary.  
-00009910: 2020 2020 2020 6966 2061 735f 6c61 7965        if as_laye
-00009920: 7220 3d3d 2054 7275 653a 0a20 2020 2020  r == True:.     
-00009930: 2020 2020 2020 2064 656e 7320 3d20 4c61         dens = La
-00009940: 7965 7228 6465 6e73 2c20 2764 656e 7369  yer(dens, 'densi
-00009950: 7479 272c 2027 6465 6e73 6974 7927 2c20  ty', 'density', 
-00009960: 7365 6c66 2e5f 6c61 6e64 5f64 696d 290a  self._land_dim).
-00009970: 2020 2020 2020 2020 2373 6574 2073 656c          #set sel
-00009980: 662e 4e20 6174 7472 6962 7574 652c 2069  f.N attribute, i
-00009990: 6620 6e65 6365 7373 6172 790a 2020 2020  f necessary.    
-000099a0: 2020 2020 6966 2073 6574 5f4e 3a0a 2020      if set_N:.  
-000099b0: 2020 2020 2020 2020 2020 7365 6c66 2e5f            self._
-000099c0: 7365 745f 4e28 6465 6e73 290a 2020 2020  set_N(dens).    
-000099d0: 2020 2020 236f 7220 656c 7365 2072 6574      #or else ret
-000099e0: 7572 6e20 7468 6520 6465 6e73 6974 7920  urn the density 
-000099f0: 6172 7261 790a 2020 2020 2020 2020 656c  array.        el
-00009a00: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-00009a10: 7265 7475 726e 2064 656e 730a 0a20 2020  return dens..   
-00009a20: 2023 206d 6574 686f 6420 746f 2073 6574   # method to set
-00009a30: 2074 6865 2073 7065 6369 6573 2720 6765   the species' ge
-00009a40: 6e6f 6d65 7320 746f 2061 6c6c 207a 6572  nomes to all zer
-00009a50: 6f73 2028 6166 7465 7220 6275 726e 2d69  os (after burn-i
-00009a60: 6e29 2c0a 2020 2020 2320 4e4f 5445 3a0a  n),.    # NOTE:.
-00009a70: 2020 2020 2320 4946 2055 5349 4e47 2054      # IF USING T
-00009a80: 534b 4954 3a0a 2020 2020 2020 2020 2320  SKIT:.        # 
-00009a90: 646f 6573 2074 6869 7320 6f6e 6c79 2069  does this only i
-00009aa0: 6620 7468 6520 7370 6563 6965 7320 6861  f the species ha
-00009ab0: 7320 616e 7920 6e6f 6e6e 6575 7472 616c  s any nonneutral
-00009ac0: 206c 6f63 690a 2020 2020 2020 2020 2320   loci.        # 
-00009ad0: 6f72 2068 6173 206e 6f6e 2d7a 6572 6f20  or has non-zero 
-00009ae0: 6e6f 6e6e 6575 7472 616c 206d 7574 6174  nonneutral mutat
-00009af0: 696f 6e20 7261 7465 730a 2020 2020 2320  ion rates.    # 
-00009b00: 4946 204e 4f54 2055 5349 4e47 2054 534b  IF NOT USING TSK
-00009b10: 4954 3a0a 2020 2020 2020 2020 2320 646f  IT:.        # do
-00009b20: 6573 2074 6869 7320 666f 7220 616c 6c20  es this for all 
-00009b30: 6c6f 6369 2069 6e20 7468 6520 6675 6c6c  loci in the full
-00009b40: 2d6c 656e 6774 6820 7369 6d75 6c61 7465  -length simulate
-00009b50: 6420 6765 6e6f 6d65 0a20 2020 2064 6566  d genome.    def
-00009b60: 205f 7365 745f 6e75 6c6c 5f67 656e 6f6d   _set_null_genom
-00009b70: 6573 2873 656c 6629 3a0a 2020 2020 2020  es(self):.      
-00009b80: 2020 2320 7365 7420 6e75 6c6c 2067 656e    # set null gen
-00009b90: 6f6d 6520 6172 7261 7973 2074 6f20 7472  ome arrays to tr
-00009ba0: 6163 6b20 6a75 7374 206e 6f6e 6e65 7574  ack just nonneut
-00009bb0: 5f6c 6f63 2c20 6966 2075 7369 6e67 2074  _loc, if using t
-00009bc0: 736b 6974 0a20 2020 2020 2020 2069 6620  skit.        if 
-00009bd0: 7365 6c66 2e67 656e 5f61 7263 682e 7573  self.gen_arch.us
-00009be0: 655f 7473 6b69 743a 0a20 2020 2020 2020  e_tskit:.       
-00009bf0: 2020 2020 2069 6620 286c 656e 2873 656c       if (len(sel
-00009c00: 662e 6765 6e5f 6172 6368 2e6e 6f6e 6e65  f.gen_arch.nonne
-00009c10: 7574 5f6c 6f63 6929 203e 2030 206f 720a  ut_loci) > 0 or.
-00009c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00009c30: 2873 656c 662e 6765 6e5f 6172 6368 2e74  (self.gen_arch.t
-00009c40: 7261 6974 7320 6973 206e 6f74 204e 6f6e  raits is not Non
-00009c50: 6520 616e 640a 2020 2020 2020 2020 2020  e and.          
-00009c60: 2020 2020 2020 206c 656e 285b 7472 742e         len([trt.
-00009c70: 6d75 203e 2030 2066 6f72 2074 7274 2069  mu > 0 for trt i
-00009c80: 6e20 7365 6c66 2e67 656e 5f61 7263 682e  n self.gen_arch.
-00009c90: 7472 6169 7473 5d29 203e 2030 2920 6f72  traits]) > 0) or
-00009ca0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00009cb0: 2073 656c 662e 6765 6e5f 6172 6368 2e6d   self.gen_arch.m
-00009cc0: 755f 6465 6c65 7420 3e20 3029 3a0a 2020  u_delet > 0):.  
-00009cd0: 2020 2020 2020 2020 2020 2020 2020 5b69                [i
-00009ce0: 6e64 2e5f 7365 745f 6728 6e70 2e7a 6572  nd._set_g(np.zer
-00009cf0: 6f73 280a 2020 2020 2020 2020 2020 2020  os(.            
-00009d00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00009d10: 286c 656e 2873 656c 662e 6765 6e5f 6172  (len(self.gen_ar
-00009d20: 6368 2e6e 6f6e 6e65 7574 5f6c 6f63 6929  ch.nonneut_loci)
-00009d30: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00009d40: 2020 2020 2020 2020 2020 2020 2020 2073                 s
-00009d50: 656c 662e 6765 6e5f 6172 6368 2e78 2929  elf.gen_arch.x))
-00009d60: 2920 666f 7220 696e 6420 696e 2073 656c  ) for ind in sel
-00009d70: 662e 7661 6c75 6573 2829 5d0a 2020 2020  f.values()].    
-00009d80: 2020 2020 2320 6f72 2073 6574 206e 756c      # or set nul
-00009d90: 6c20 6765 6e6f 6d65 2061 7272 6179 7320  l genome arrays 
-00009da0: 746f 2074 7261 636b 2061 6c6c 206c 6f63  to track all loc
-00009db0: 692c 2069 6620 6e6f 7420 7573 696e 6720  i, if not using 
-00009dc0: 7473 6b69 740a 2020 2020 2020 2020 656c  tskit.        el
-00009dd0: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-00009de0: 5b69 6e64 2e5f 7365 745f 6728 6e70 2e7a  [ind._set_g(np.z
-00009df0: 6572 6f73 280a 2020 2020 2020 2020 2020  eros(.          
-00009e00: 2020 2020 2020 2020 2020 2020 2020 2873                (s
-00009e10: 656c 662e 6765 6e5f 6172 6368 2e4c 2c0a  elf.gen_arch.L,.
-00009e20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00009e30: 2020 2020 2020 2020 2073 656c 662e 6765           self.ge
-00009e40: 6e5f 6172 6368 2e78 2929 2920 666f 7220  n_arch.x))) for 
-00009e50: 696e 6420 696e 2073 656c 662e 7661 6c75  ind in self.valu
-00009e60: 6573 2829 5d0a 0a20 2020 2023 2061 6464  es()]..    # add
-00009e70: 206e 6577 2072 6f77 2074 6f20 7468 6520   new row to the 
-00009e80: 696e 6469 7669 6475 616c 7327 206e 756d  individuals' num
-00009e90: 7079 2061 7272 6179 732c 2066 6f72 2061  py arrays, for a
-00009ea0: 2067 6976 656e 206d 7574 6174 696f 6e20   given mutation 
-00009eb0: 6c6f 6375 730a 2020 2020 6465 6620 5f61  locus.    def _a
-00009ec0: 6464 5f6e 6577 5f6c 6f63 7573 2873 656c  dd_new_locus(sel
-00009ed0: 662c 2069 6478 2c20 6c6f 6375 7329 3a0a  f, idx, locus):.
-00009ee0: 2020 2020 2020 2020 2320 696e 7365 7274          # insert
-00009ef0: 2061 2072 6f77 206f 6620 6f6e 6573 2069   a row of ones i
-00009f00: 6e20 6561 6368 2069 6e64 6976 6964 7561  n each individua
-00009f10: 6c27 7320 6765 6e6f 7479 7065 2061 7272  l's genotype arr
-00009f20: 6179 200a 2020 2020 2020 2020 5b69 6e64  ay .        [ind
-00009f30: 2e5f 6164 645f 6e65 775f 6c6f 6375 7328  ._add_new_locus(
-00009f40: 6964 7829 2066 6f72 2069 6e64 2069 6e20  idx) for ind in 
-00009f50: 7365 6c66 2e76 616c 7565 7328 295d 0a0a  self.values()]..
-00009f60: 2020 2020 236d 6574 686f 6420 746f 2073      #method to s
-00009f70: 6574 2074 6865 2069 6e64 6976 6964 7561  et the individua
-00009f80: 6c73 2720 656e 7669 726f 6e6d 656e 7420  ls' environment 
-00009f90: 7661 6c75 6573 0a20 2020 2064 6566 205f  values.    def _
-00009fa0: 7365 745f 6528 7365 6c66 2c20 6c61 6e64  set_e(self, land
-00009fb0: 2c20 696e 6469 7669 6473 203d 204e 6f6e  , individs = Non
-00009fc0: 6529 3a0a 2020 2020 2020 2020 6966 2069  e):.        if i
-00009fd0: 6e64 6976 6964 7320 6973 204e 6f6e 653a  ndivids is None:
-00009fe0: 0a20 2020 2020 2020 2020 2020 2069 6e64  .            ind
-00009ff0: 735f 746f 5f73 6574 203d 2073 656c 662e  s_to_set = self.
-0000a000: 7661 6c75 6573 2829 0a20 2020 2020 2020  values().       
-0000a010: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
-0000a020: 2020 2069 6720 3d20 6974 656d 6765 7474     ig = itemgett
-0000a030: 6572 282a 696e 6469 7669 6473 290a 2020  er(*individs).  
-0000a040: 2020 2020 2020 2020 2020 696e 6473 5f74            inds_t
-0000a050: 6f5f 7365 7420 3d20 6967 2873 656c 6629  o_set = ig(self)
-0000a060: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
-0000a070: 6973 696e 7374 616e 6365 2869 6e64 735f  isinstance(inds_
-0000a080: 746f 5f73 6574 2c20 696e 6469 7669 6475  to_set, individu
-0000a090: 616c 2e49 6e64 6976 6964 7561 6c29 3a0a  al.Individual):.
-0000a0a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a0b0: 696e 6473 5f74 6f5f 7365 7420 3d20 2869  inds_to_set = (i
-0000a0c0: 6e64 735f 746f 5f73 6574 2c29 0a20 2020  nds_to_set,).   
-0000a0d0: 2020 2020 2068 6162 203d 205b 696e 642e       hab = [ind.
-0000a0e0: 5f73 6574 5f65 285b 6c79 722e 7261 7374  _set_e([lyr.rast
-0000a0f0: 5b69 6e74 2869 6e64 2e79 292c 2069 6e74  [int(ind.y), int
-0000a100: 280a 2020 2020 2020 2020 2020 2020 696e  (.            in
-0000a110: 642e 7829 5d20 666f 7220 6c79 7220 696e  d.x)] for lyr in
-0000a120: 206c 616e 642e 7661 6c75 6573 2829 5d29   land.values()])
-0000a130: 2066 6f72 2069 6e64 2069 6e20 696e 6473   for ind in inds
-0000a140: 5f74 6f5f 7365 745d 0a0a 2020 2020 236d  _to_set]..    #m
-0000a150: 6574 686f 6420 746f 2073 6574 2074 6865  ethod to set the
-0000a160: 2069 6e64 6976 6964 7561 6c73 2720 7068   individuals' ph
-0000a170: 656e 6f74 7970 6520 6174 7472 6962 7574  enotype attribut
-0000a180: 6573 200a 2020 2020 6465 6620 5f73 6574  es .    def _set
-0000a190: 5f7a 2873 656c 6629 3a0a 2020 2020 2020  _z(self):.      
-0000a1a0: 2020 5b69 6e64 2e5f 7365 745f 7a28 7365    [ind._set_z(se
-0000a1b0: 6c66 2e67 656e 5f61 7263 6829 2066 6f72  lf.gen_arch) for
-0000a1c0: 2069 6e64 2069 6e20 7365 6c66 2e76 616c   ind in self.val
-0000a1d0: 7565 7328 295d 3b0a 0a20 2020 2023 6d65  ues()];..    #me
-0000a1e0: 7468 6f64 2066 6f72 2073 6574 7469 6e67  thod for setting
-0000a1f0: 2061 6e20 696e 6469 7669 6475 616c 2773   an individual's
-0000a200: 2070 6865 6e6f 7479 7065 0a20 2020 2064   phenotype.    d
-0000a210: 6566 205f 7365 745f 7a5f 696e 6469 7669  ef _set_z_indivi
-0000a220: 6428 7365 6c66 2c20 696e 6469 7669 6429  d(self, individ)
-0000a230: 3a0a 2020 2020 2020 2020 7365 6c66 5b69  :.        self[i
-0000a240: 6e64 6976 6964 5d2e 5f73 6574 5f7a 2873  ndivid]._set_z(s
-0000a250: 656c 662e 6765 6e5f 6172 6368 290a 0a20  elf.gen_arch).. 
-0000a260: 2020 2023 6d65 7468 6f64 2074 6f20 7365     #method to se
-0000a270: 7420 7468 6520 696e 6469 7669 6475 616c  t the individual
-0000a280: 7327 2066 6974 6e65 7373 2061 7474 7269  s' fitness attri
-0000a290: 6275 7465 730a 2020 2020 6465 6620 5f73  butes.    def _s
-0000a2a0: 6574 5f66 6974 2873 656c 662c 2066 6974  et_fit(self, fit
-0000a2b0: 293a 0a20 2020 2020 2020 205b 696e 642e  ):.        [ind.
-0000a2c0: 5f73 6574 5f66 6974 2866 2920 666f 7220  _set_fit(f) for 
-0000a2d0: 696e 642c 2066 2069 6e20 7a69 7028 7365  ind, f in zip(se
-0000a2e0: 6c66 2e76 616c 7565 7328 292c 2066 6974  lf.values(), fit
-0000a2f0: 295d 3b0a 0a20 2020 2023 6d65 7468 6f64  )];..    #method
-0000a300: 2074 6f20 7365 7420 7370 6563 6965 7327   to set species'
-0000a310: 2063 6f6f 7264 7320 616e 6420 6365 6c6c   coords and cell
-0000a320: 7320 6172 7261 7973 0a20 2020 2064 6566  s arrays.    def
-0000a330: 205f 7365 745f 636f 6f72 6473 5f61 6e64   _set_coords_and
-0000a340: 5f63 656c 6c73 2873 656c 6629 3a0a 2020  _cells(self):.  
-0000a350: 2020 2020 2020 7365 6c66 2e63 6f6f 7264        self.coord
-0000a360: 7320 3d20 7365 6c66 2e5f 6765 745f 636f  s = self._get_co
-0000a370: 6f72 6473 2829 0a20 2020 2020 2020 2073  ords().        s
-0000a380: 656c 662e 6365 6c6c 7320 3d20 6e70 2e69  elf.cells = np.i
-0000a390: 6e74 3332 286e 702e 666c 6f6f 7228 7365  nt32(np.floor(se
-0000a3a0: 6c66 2e63 6f6f 7264 7329 290a 0a20 2020  lf.coords))..   
-0000a3b0: 2023 6d65 7468 6f64 2074 6f20 7365 7420   #method to set 
-0000a3c0: 7468 6520 7370 6563 6965 7327 206b 645f  the species' kd_
-0000a3d0: 7472 6565 2061 7474 7269 6275 7465 2028  tree attribute (
-0000a3e0: 6120 7370 6174 6961 6c2e 5f4b 4454 7265  a spatial._KDTre
-0000a3f0: 6529 0a20 2020 2064 6566 205f 7365 745f  e).    def _set_
-0000a400: 6b64 5f74 7265 6528 7365 6c66 2c20 6c65  kd_tree(self, le
-0000a410: 6166 7369 7a65 203d 2031 3030 293a 0a20  afsize = 100):. 
-0000a420: 2020 2020 2020 2073 656c 662e 5f6b 645f         self._kd_
-0000a430: 7472 6565 203d 2073 7074 2e5f 4b44 5472  tree = spt._KDTr
-0000a440: 6565 2863 6f6f 7264 7320 3d20 7365 6c66  ee(coords = self
-0000a450: 2e63 6f6f 7264 732c 206c 6561 6673 697a  .coords, leafsiz
-0000a460: 6520 3d20 6c65 6166 7369 7a65 290a 0a0a  e = leafsize)...
-0000a470: 2020 2020 236d 6574 686f 6420 746f 2073      #method to s
-0000a480: 6574 2074 6865 2073 7065 6369 6573 2720  et the species' 
-0000a490: 7370 6174 6961 6c2e 5f44 656e 7369 7479  spatial._Density
-0000a4a0: 4772 6964 5374 6163 6b20 6174 7472 6962  GridStack attrib
-0000a4b0: 7574 650a 2020 2020 6465 6620 5f73 6574  ute.    def _set
-0000a4c0: 5f64 656e 735f 6772 6964 7328 7365 6c66  _dens_grids(self
-0000a4d0: 2c20 6c61 6e64 2c20 7769 646f 775f 7769  , land, widow_wi
-0000a4e0: 6474 6820 3d20 4e6f 6e65 293a 0a20 2020  dth = None):.   
-0000a4f0: 2020 2020 2073 656c 662e 5f64 656e 735f       self._dens_
-0000a500: 6772 6964 7320 3d20 7370 742e 5f44 656e  grids = spt._Den
-0000a510: 7369 7479 4772 6964 5374 6163 6b28 6c61  sityGridStack(la
-0000a520: 6e64 203d 206c 616e 642c 0a20 2020 2020  nd = land,.     
+00008d00: 2020 2020 2020 2020 2020 2269 6e20 7473            "in ts
+00008d10: 6b69 7420 5461 626c 6543 6f6c 6c65 6374  kit TableCollect
+00008d20: 696f 6e3a 5c6e 5c6e 4d41 4e55 414c 5c6e  ion:\n\nMANUAL\n
+00008d30: 5c74 2573 220a 2020 2020 2020 2020 2020  \t%s".          
+00008d40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008d50: 2020 225c 6e5c 6e54 534b 4954 5c6e 5c74    "\n\nTSKIT\n\t
+00008d60: 2573 2229 2025 2028 7374 7228 6d61 6e75  %s") % (str(manu
+00008d70: 616c 5f67 7429 2c0a 2020 2020 2020 2020  al_gt),.        
+00008d80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008d90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008da0: 2020 2020 2020 2020 2020 7374 7228 7473            str(ts
+00008db0: 6b69 745f 6774 2929 0a20 2020 2020 2020  kit_gt)).       
+00008dc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008dd0: 2067 745f 6368 6563 6b5f 6e75 6d20 2b3d   gt_check_num +=
+00008de0: 2032 0a0a 2020 2020 2020 2020 2320 7361   2..        # sa
+00008df0: 6d70 6c65 2061 6c6c 2069 6e64 6976 6964  mple all individ
+00008e00: 7561 6c73 2720 656e 7669 726f 6e6d 656e  uals' environmen
+00008e10: 7420 7661 6c75 6573 2c20 746f 2069 6e69  t values, to ini
+00008e20: 7469 6174 6520 666f 7220 6f66 6673 7072  tiate for offspr
+00008e30: 696e 670a 2020 2020 2020 2020 7365 6c66  ing.        self
+00008e40: 2e5f 7365 745f 6528 6c61 6e64 290a 2020  ._set_e(land).  
+00008e50: 2020 2020 2020 7365 6c66 2e5f 7365 745f        self._set_
+00008e60: 636f 6f72 6473 5f61 6e64 5f63 656c 6c73  coords_and_cells
+00008e70: 2829 0a0a 2020 2020 2020 2020 2320 646f  ()..        # do
+00008e80: 206d 7574 6174 696f 6e20 6966 206e 6563   mutation if nec
+00008e90: 6573 7361 7279 0a20 2020 2020 2020 2069  essary.        i
+00008ea0: 6620 7365 6c66 2e6d 7574 6174 6520 616e  f self.mutate an
+00008eb0: 6420 6e6f 7420 6275 726e 3a0a 2020 2020  d not burn:.    
+00008ec0: 2020 2020 2020 2020 205f 646f 5f6d 7574           _do_mut
+00008ed0: 6174 696f 6e28 6b65 7973 5f6c 6973 742c  ation(keys_list,
+00008ee0: 2073 656c 662c 206c 6f67 203d 2073 656c   self, log = sel
+00008ef0: 662e 6d75 745f 6c6f 6729 0a0a 2020 2020  f.mut_log)..    
+00008f00: 2020 2020 2320 6368 6563 6b20 6861 706c      # check hapl
+00008f10: 6f79 7470 6573 2068 6176 6520 6e6f 206d  oytpes have no m
+00008f20: 6973 7369 6e67 2064 6174 612c 2069 6620  issing data, if 
+00008f30: 6e65 6365 7373 6172 790a 2020 2020 2020  necessary.      
+00008f40: 2020 6966 2063 6865 636b 5f68 6170 7320    if check_haps 
+00008f50: 616e 6420 7365 6c66 2e62 7572 6e65 6420  and self.burned 
+00008f60: 616e 6420 7365 6c66 2e67 656e 5f61 7263  and self.gen_arc
+00008f70: 682e 7573 655f 7473 6b69 743a 0a20 2020  h.use_tskit:.   
+00008f80: 2020 2020 2020 2020 2070 7269 6e74 2827           print('
+00008f90: 5c6e 5c6e 4348 4543 4b49 4e47 2054 4841  \n\nCHECKING THA
+00008fa0: 5420 4841 504c 4f54 5950 4553 2048 4156  T HAPLOTYPES HAV
+00008fb0: 4520 4e4f 204d 4953 5349 4e47 2044 4154  E NO MISSING DAT
+00008fc0: 415c 6e5c 6e27 290a 2020 2020 2020 2020  A\n\n').        
+00008fd0: 2020 2020 7365 6c66 2e5f 736f 7274 5f61      self._sort_a
+00008fe0: 6e64 5f73 696d 706c 6966 795f 7461 626c  nd_simplify_tabl
+00008ff0: 655f 636f 6c6c 6563 7469 6f6e 2829 0a20  e_collection(). 
+00009000: 2020 2020 2020 2020 2020 2074 7320 3d20             ts = 
+00009010: 7365 6c66 2e5f 7463 2e74 7265 655f 7365  self._tc.tree_se
+00009020: 7175 656e 6365 2829 0a20 2020 2020 2020  quence().       
+00009030: 2020 2020 2068 6170 7320 3d20 5b2a 7473       haps = [*ts
+00009040: 2e68 6170 6c6f 7479 7065 7328 295d 0a20  .haplotypes()]. 
+00009050: 2020 2020 2020 2020 2020 206d 6973 7369             missi
+00009060: 6e67 203d 205b 272d 2720 696e 2068 6170  ng = ['-' in hap
+00009070: 2066 6f72 2068 6170 2069 6e20 6861 7073   for hap in haps
+00009080: 5d0a 2020 2020 2020 2020 2020 2020 6173  ].            as
+00009090: 7365 7274 2073 756d 286d 6973 7369 6e67  sert sum(missing
+000090a0: 2920 3d3d 2030 2c20 2773 6f6d 6520 6861  ) == 0, 'some ha
+000090b0: 706c 6f74 7970 6573 2068 6176 6520 6d69  plotypes have mi
+000090c0: 7373 696e 6720 6461 7461 2127 0a0a 0a20  ssing data!'... 
+000090d0: 2020 2023 6d65 7468 6f64 2074 6f20 646f     #method to do
+000090e0: 2073 7065 6369 6573 2064 796e 616d 6963   species dynamic
+000090f0: 730a 2020 2020 6465 6620 5f64 6f5f 706f  s.    def _do_po
+00009100: 705f 6479 6e61 6d69 6373 2873 656c 662c  p_dynamics(self,
+00009110: 206c 616e 6429 3a0a 2020 2020 2020 2020   land):.        
+00009120: 2369 6d70 6c65 6d65 6e74 2073 656c 6563  #implement selec
+00009130: 7469 6f6e 2c20 6966 6620 7365 6c66 2e73  tion, iff self.s
+00009140: 656c 6563 7469 6f6e 2069 7320 5472 7565  election is True
+00009150: 2061 6e64 2074 6865 2073 7070 2068 6173   and the spp has
+00009160: 0a20 2020 2020 2020 2023 616c 7265 6164  .        #alread
+00009170: 7920 6265 656e 2062 7572 6e65 6420 696e  y been burned in
+00009180: 0a20 2020 2020 2020 2077 6974 685f 7365  .        with_se
+00009190: 6c65 6374 696f 6e20 3d20 7365 6c66 2e73  lection = self.s
+000091a0: 656c 6563 7469 6f6e 2061 6e64 2073 656c  election and sel
+000091b0: 662e 6275 726e 6564 0a20 2020 2020 2020  f.burned.       
+000091c0: 2062 7572 6e20 3d20 6e6f 7420 7365 6c66   burn = not self
+000091d0: 2e62 7572 6e65 640a 2020 2020 2020 2020  .burned.        
+000091e0: 2374 6865 6e20 6361 7272 7920 6f75 7420  #then carry out 
+000091f0: 7468 6520 706f 702d 6479 6e61 6d69 6373  the pop-dynamics
+00009200: 2c20 7769 7468 2073 656c 6563 7469 6f6e  , with selection
+00009210: 2061 7320 7365 7420 6162 6f76 652c 2061   as set above, a
+00009220: 6e64 2073 6176 650a 2020 2020 2020 2020  nd save.        
+00009230: 2372 6573 756c 742c 2077 6869 6368 2077  #result, which w
+00009240: 696c 6c20 6265 2054 7275 6520 6966 6620  ill be True iff 
+00009250: 7370 7020 6861 7320 676f 6e65 2065 7874  spp has gone ext
+00009260: 696e 6374 0a20 2020 2020 2020 2065 7874  inct.        ext
+00009270: 696e 6374 203d 205f 646f 5f70 6f70 5f64  inct = _do_pop_d
+00009280: 796e 616d 6963 7328 7365 6c66 2c20 6c61  ynamics(self, la
+00009290: 6e64 2c0a 2020 2020 2020 2020 2020 2020  nd,.            
+000092a0: 7769 7468 5f73 656c 6563 7469 6f6e 203d  with_selection =
+000092b0: 2077 6974 685f 7365 6c65 6374 696f 6e2c   with_selection,
+000092c0: 2062 7572 6e20 3d20 6275 726e 290a 2020   burn = burn).  
+000092d0: 2020 2020 2020 6966 2065 7874 696e 6374        if extinct
+000092e0: 3a0a 2020 2020 2020 2020 2020 2020 2373  :.            #s
+000092f0: 6574 2073 656c 662e 6578 7469 6e63 7420  et self.extinct 
+00009300: 6571 7561 6c20 746f 2054 7275 652c 2073  equal to True, s
+00009310: 6f20 7468 6174 2074 6865 2069 7465 7261  o that the itera
+00009320: 7469 6f6e 2077 696c 6c20 656e 640a 2020  tion will end.  
+00009330: 2020 2020 2020 2020 2020 7365 6c66 2e65            self.e
+00009340: 7874 696e 6374 203d 2065 7874 696e 6374  xtinct = extinct
+00009350: 0a0a 2020 2020 236d 6574 686f 6420 746f  ..    #method to
+00009360: 206d 616b 6520 7370 6563 6965 7320 6368   make species ch
+00009370: 616e 6765 730a 2020 2020 6465 6620 5f6d  anges.    def _m
+00009380: 616b 655f 6368 616e 6765 2873 656c 662c  ake_change(self,
+00009390: 2076 6572 626f 7365 3d46 616c 7365 293a   verbose=False):
+000093a0: 0a20 2020 2020 2020 2073 656c 662e 5f63  .        self._c
+000093b0: 6861 6e67 6572 2e5f 6d61 6b65 5f63 6861  hanger._make_cha
+000093c0: 6e67 6528 7420 3d20 7365 6c66 2e74 2c20  nge(t = self.t, 
+000093d0: 6164 6469 7469 6f6e 616c 5f61 7267 7320  additional_args 
+000093e0: 3d20 7b27 7370 7027 3a20 7365 6c66 7d2c  = {'spp': self},
+000093f0: 0a20 2020 2020 2020 2020 2020 2076 6572  .            ver
+00009400: 626f 7365 203d 2076 6572 626f 7365 290a  bose = verbose).
+00009410: 0a20 2020 2023 6d65 7468 6f64 2074 6f20  .    #method to 
+00009420: 6368 6563 6b20 6966 2074 6865 2073 7065  check if the spe
+00009430: 6369 6573 2068 6173 2067 6f6e 6520 6578  cies has gone ex
+00009440: 7469 6e63 740a 2020 2020 6465 6620 5f63  tinct.    def _c
+00009450: 6865 636b 5f65 7874 696e 6374 2873 656c  heck_extinct(sel
+00009460: 6629 3a0a 2020 2020 2020 2020 7265 7475  f):.        retu
+00009470: 726e 206c 656e 2873 656c 6629 203d 3d20  rn len(self) == 
+00009480: 300a 0a20 2020 2023 6d65 7468 6f64 2074  0..    #method t
+00009490: 6f20 6361 6c63 756c 6174 6520 7370 6563  o calculate spec
+000094a0: 6965 7320 6465 6e73 6974 790a 2020 2020  ies density.    
+000094b0: 6465 6620 5f63 616c 635f 6465 6e73 6974  def _calc_densit
+000094c0: 7928 7365 6c66 2c20 6e6f 726d 616c 697a  y(self, normaliz
+000094d0: 6520 3d20 4661 6c73 652c 2061 735f 6c61  e = False, as_la
+000094e0: 7965 7220 3d20 4661 6c73 652c 2073 6574  yer = False, set
+000094f0: 5f4e 3d46 616c 7365 293a 0a0a 2020 2020  _N=False):..    
+00009500: 2020 2020 2727 270a 2020 2020 2020 2020      '''.        
+00009510: 4361 6c63 756c 6174 6520 616e 2069 6e74  Calculate an int
+00009520: 6572 706f 6c61 7465 6420 7261 7374 6572  erpolated raster
+00009530: 206f 6620 6c6f 6361 6c20 7370 6563 6965   of local specie
+00009540: 7320 6465 6e73 6974 792c 2075 7369 6e67  s density, using
+00009550: 0a20 2020 2020 2020 2074 6865 2073 7061  .        the spa
+00009560: 7469 616c 2e5f 4465 6e73 6974 7947 7269  tial._DensityGri
+00009570: 6453 7461 636b 206f 626a 6563 7420 7374  dStack object st
+00009580: 6f72 6564 2069 6e20 7468 650a 2020 2020  ored in the.    
+00009590: 2020 2020 5370 6563 6965 732e 5f64 656e      Species._den
+000095a0: 735f 6772 6964 7320 6174 7472 6962 7574  s_grids attribut
+000095b0: 652e 0a0a 2020 2020 2020 2020 4966 206e  e...        If n
+000095c0: 6f72 6d61 6c69 7a65 2069 7320 5472 7565  ormalize is True
+000095d0: 2c20 7468 6520 6465 6e73 6974 7920 7261  , the density ra
+000095e0: 7374 6572 2077 696c 6c20 7661 7279 2062  ster will vary b
+000095f0: 6574 7765 656e 2030 2061 6e64 2031 0a20  etween 0 and 1. 
+00009600: 2020 2020 2020 2028 6166 7465 7220 6265         (after be
+00009610: 696e 6720 6e6f 726d 616c 697a 6564 2062  ing normalized b
+00009620: 7920 6974 7320 6f77 6e20 6d61 7869 6d75  y its own maximu
+00009630: 6d20 7661 6c75 6529 2e20 4966 2066 616c  m value). If fal
+00009640: 7365 2c20 6974 2077 696c 6c0a 2020 2020  se, it will.    
+00009650: 2020 2020 7661 7279 2062 6574 7765 656e      vary between
+00009660: 2030 2061 6e64 2069 7473 206d 6178 696d   0 and its maxim
+00009670: 756d 2065 7374 696d 6174 6564 2064 656e  um estimated den
+00009680: 7369 7479 2076 616c 7565 2e0a 2020 2020  sity value..    
+00009690: 2020 2020 2727 270a 2020 2020 2020 2020      '''.        
+000096a0: 2363 6865 636b 2076 616c 6964 6974 7920  #check validity 
+000096b0: 6f66 206e 6f72 6d61 6c69 7a65 2061 7267  of normalize arg
+000096c0: 756d 656e 740a 2020 2020 2020 2020 6173  ument.        as
+000096d0: 7365 7274 2074 7970 6528 6e6f 726d 616c  sert type(normal
+000096e0: 697a 6529 2069 7320 626f 6f6c 2c20 2822  ize) is bool, ("
+000096f0: 5468 6520 276e 6f72 6d61 6c69 7a65 2720  The 'normalize' 
+00009700: 6172 6775 6d65 6e74 2074 616b 6573 2022  argument takes "
+00009710: 0a20 2020 2020 2020 2020 2020 2022 6120  .            "a 
+00009720: 626f 6f6c 6561 6e20 7661 6c75 652e 5c6e  boolean value.\n
+00009730: 2229 0a20 2020 2020 2020 2023 6765 7420  ").        #get 
+00009740: 7370 6563 6965 7320 636f 6f72 6469 6e61  species coordina
+00009750: 7465 730a 2020 2020 2020 2020 7820 3d20  tes.        x = 
+00009760: 7365 6c66 2e5f 6765 745f 7828 290a 2020  self._get_x().  
+00009770: 2020 2020 2020 7920 3d20 7365 6c66 2e5f        y = self._
+00009780: 6765 745f 7928 290a 2020 2020 2020 2020  get_y().        
+00009790: 2363 616c 6375 6c61 7465 2074 6865 2064  #calculate the d
+000097a0: 656e 7369 7479 2061 7272 6179 0a20 2020  ensity array.   
+000097b0: 2020 2020 2064 656e 7320 3d20 7365 6c66       dens = self
+000097c0: 2e5f 6465 6e73 5f67 7269 6473 2e5f 6361  ._dens_grids._ca
+000097d0: 6c63 5f64 656e 7369 7479 2878 2c20 7929  lc_density(x, y)
+000097e0: 0a20 2020 2020 2020 2023 7365 7420 6d69  .        #set mi
+000097f0: 6e20 7661 6c75 6520 746f 2030 0a20 2020  n value to 0.   
+00009800: 2020 2020 2064 656e 7320 3d20 6e70 2e63       dens = np.c
+00009810: 6c69 7028 6465 6e73 2c20 615f 6d69 6e20  lip(dens, a_min 
+00009820: 3d20 302c 2061 5f6d 6178 203d 204e 6f6e  = 0, a_max = Non
+00009830: 6529 0a20 2020 2020 2020 2023 6e6f 726d  e).        #norm
+00009840: 616c 697a 652c 2069 6620 6e65 6365 7373  alize, if necess
+00009850: 6172 7920 0a20 2020 2020 2020 2069 6620  ary .        if 
+00009860: 6e6f 726d 616c 697a 653a 0a20 2020 2020  normalize:.     
+00009870: 2020 2020 2020 2023 2055 7365 206d 6178         # Use max
+00009880: 5f76 616c 2074 6f20 6e6f 726d 616c 697a  _val to normaliz
+00009890: 6520 7468 6520 6465 6e73 6974 7920 7261  e the density ra
+000098a0: 7374 6572 2074 6f20 6569 7468 6572 0a20  ster to either. 
+000098b0: 2020 2020 2020 2020 2020 2023 3020 746f             #0 to
+000098c0: 2069 7473 2063 7572 7265 6e74 206d 6178   its current max
+000098d0: 2076 616c 206f 7220 3020 746f 2031 2c20   val or 0 to 1, 
+000098e0: 746f 206d 616b 6520 7375 7265 200a 2020  to make sure .  
+000098f0: 2020 2020 2020 2020 2020 2374 6865 2069            #the i
+00009900: 6e74 6572 706f 6c61 7469 6f6e 2064 6964  nterpolation did
+00009910: 6e27 7420 6765 6e65 7261 7465 2061 6e79  n't generate any
+00009920: 2076 616c 7565 7320 736c 6967 6874 6c79   values slightly
+00009930: 0a20 2020 2020 2020 2020 2020 2023 6f75  .            #ou
+00009940: 7473 6964 6520 7468 6973 2072 616e 6765  tside this range
+00009950: 0a20 2020 2020 2020 2020 2020 206e 6f72  .            nor
+00009960: 6d5f 6661 6374 6f72 203d 2064 656e 732e  m_factor = dens.
+00009970: 6d61 7828 2920 2d20 6465 6e73 2e6d 696e  max() - dens.min
+00009980: 2829 0a20 2020 2020 2020 2020 2020 2064  ().            d
+00009990: 656e 7320 3d20 2864 656e 7320 2d20 6465  ens = (dens - de
+000099a0: 6e73 2e6d 696e 2829 2920 2f20 6e6f 726d  ns.min()) / norm
+000099b0: 5f66 6163 746f 720a 2020 2020 2020 2020  _factor.        
+000099c0: 2372 6574 7572 6e20 6173 206c 6179 6572  #return as layer
+000099d0: 2c20 6966 206e 6563 6573 7361 7279 0a20  , if necessary. 
+000099e0: 2020 2020 2020 2069 6620 6173 5f6c 6179         if as_lay
+000099f0: 6572 203d 3d20 5472 7565 3a0a 2020 2020  er == True:.    
+00009a00: 2020 2020 2020 2020 6465 6e73 203d 204c          dens = L
+00009a10: 6179 6572 2864 656e 732c 2027 6465 6e73  ayer(dens, 'dens
+00009a20: 6974 7927 2c20 2764 656e 7369 7479 272c  ity', 'density',
+00009a30: 2073 656c 662e 5f6c 616e 645f 6469 6d29   self._land_dim)
+00009a40: 0a20 2020 2020 2020 2023 7365 7420 7365  .        #set se
+00009a50: 6c66 2e4e 2061 7474 7269 6275 7465 2c20  lf.N attribute, 
+00009a60: 6966 206e 6563 6573 7361 7279 0a20 2020  if necessary.   
+00009a70: 2020 2020 2069 6620 7365 745f 4e3a 0a20       if set_N:. 
+00009a80: 2020 2020 2020 2020 2020 2073 656c 662e             self.
+00009a90: 5f73 6574 5f4e 2864 656e 7329 0a20 2020  _set_N(dens).   
+00009aa0: 2020 2020 2023 6f72 2065 6c73 6520 7265       #or else re
+00009ab0: 7475 726e 2074 6865 2064 656e 7369 7479  turn the density
+00009ac0: 2061 7272 6179 0a20 2020 2020 2020 2065   array.        e
+00009ad0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
+00009ae0: 2072 6574 7572 6e20 6465 6e73 0a0a 2020   return dens..  
+00009af0: 2020 2320 6d65 7468 6f64 2074 6f20 7365    # method to se
+00009b00: 7420 7468 6520 7370 6563 6965 7327 2067  t the species' g
+00009b10: 656e 6f6d 6573 2074 6f20 616c 6c20 7a65  enomes to all ze
+00009b20: 726f 7320 2861 6674 6572 2062 7572 6e2d  ros (after burn-
+00009b30: 696e 292c 0a20 2020 2023 204e 4f54 453a  in),.    # NOTE:
+00009b40: 0a20 2020 2023 2049 4620 5553 494e 4720  .    # IF USING 
+00009b50: 5453 4b49 543a 0a20 2020 2020 2020 2023  TSKIT:.        #
+00009b60: 2064 6f65 7320 7468 6973 206f 6e6c 7920   does this only 
+00009b70: 6966 2074 6865 2073 7065 6369 6573 2068  if the species h
+00009b80: 6173 2061 6e79 206e 6f6e 6e65 7574 7261  as any nonneutra
+00009b90: 6c20 6c6f 6369 0a20 2020 2020 2020 2023  l loci.        #
+00009ba0: 206f 7220 6861 7320 6e6f 6e2d 7a65 726f   or has non-zero
+00009bb0: 206e 6f6e 6e65 7574 7261 6c20 6d75 7461   nonneutral muta
+00009bc0: 7469 6f6e 2072 6174 6573 0a20 2020 2023  tion rates.    #
+00009bd0: 2049 4620 4e4f 5420 5553 494e 4720 5453   IF NOT USING TS
+00009be0: 4b49 543a 0a20 2020 2020 2020 2023 2064  KIT:.        # d
+00009bf0: 6f65 7320 7468 6973 2066 6f72 2061 6c6c  oes this for all
+00009c00: 206c 6f63 6920 696e 2074 6865 2066 756c   loci in the ful
+00009c10: 6c2d 6c65 6e67 7468 2073 696d 756c 6174  l-length simulat
+00009c20: 6564 2067 656e 6f6d 650a 2020 2020 6465  ed genome.    de
+00009c30: 6620 5f73 6574 5f6e 756c 6c5f 6765 6e6f  f _set_null_geno
+00009c40: 6d65 7328 7365 6c66 293a 0a20 2020 2020  mes(self):.     
+00009c50: 2020 2023 2073 6574 206e 756c 6c20 6765     # set null ge
+00009c60: 6e6f 6d65 2061 7272 6179 7320 746f 2074  nome arrays to t
+00009c70: 7261 636b 206a 7573 7420 6e6f 6e6e 6575  rack just nonneu
+00009c80: 745f 6c6f 632c 2069 6620 7573 696e 6720  t_loc, if using 
+00009c90: 7473 6b69 740a 2020 2020 2020 2020 6966  tskit.        if
+00009ca0: 2073 656c 662e 6765 6e5f 6172 6368 2e75   self.gen_arch.u
+00009cb0: 7365 5f74 736b 6974 3a0a 2020 2020 2020  se_tskit:.      
+00009cc0: 2020 2020 2020 6966 2028 6c65 6e28 7365        if (len(se
+00009cd0: 6c66 2e67 656e 5f61 7263 682e 6e6f 6e6e  lf.gen_arch.nonn
+00009ce0: 6575 745f 6c6f 6369 2920 3e20 3020 6f72  eut_loci) > 0 or
+00009cf0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00009d00: 2028 7365 6c66 2e67 656e 5f61 7263 682e   (self.gen_arch.
+00009d10: 7472 6169 7473 2069 7320 6e6f 7420 4e6f  traits is not No
+00009d20: 6e65 2061 6e64 0a20 2020 2020 2020 2020  ne and.         
+00009d30: 2020 2020 2020 2020 6c65 6e28 5b74 7274          len([trt
+00009d40: 2e6d 7520 3e20 3020 666f 7220 7472 7420  .mu > 0 for trt 
+00009d50: 696e 2073 656c 662e 6765 6e5f 6172 6368  in self.gen_arch
+00009d60: 2e74 7261 6974 735d 2920 3e20 3029 206f  .traits]) > 0) o
+00009d70: 720a 2020 2020 2020 2020 2020 2020 2020  r.              
+00009d80: 2020 7365 6c66 2e67 656e 5f61 7263 682e    self.gen_arch.
+00009d90: 6d75 5f64 656c 6574 203e 2030 293a 0a20  mu_delet > 0):. 
+00009da0: 2020 2020 2020 2020 2020 2020 2020 205b                 [
+00009db0: 696e 642e 5f73 6574 5f67 286e 702e 7a65  ind._set_g(np.ze
+00009dc0: 726f 7328 0a20 2020 2020 2020 2020 2020  ros(.           
+00009dd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00009de0: 2028 6c65 6e28 7365 6c66 2e67 656e 5f61   (len(self.gen_a
+00009df0: 7263 682e 6e6f 6e6e 6575 745f 6c6f 6369  rch.nonneut_loci
+00009e00: 292c 0a20 2020 2020 2020 2020 2020 2020  ),.             
+00009e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00009e20: 7365 6c66 2e67 656e 5f61 7263 682e 7829  self.gen_arch.x)
+00009e30: 2929 2066 6f72 2069 6e64 2069 6e20 7365  )) for ind in se
+00009e40: 6c66 2e76 616c 7565 7328 295d 0a20 2020  lf.values()].   
+00009e50: 2020 2020 2023 206f 7220 7365 7420 6e75       # or set nu
+00009e60: 6c6c 2067 656e 6f6d 6520 6172 7261 7973  ll genome arrays
+00009e70: 2074 6f20 7472 6163 6b20 616c 6c20 6c6f   to track all lo
+00009e80: 6369 2c20 6966 206e 6f74 2075 7369 6e67  ci, if not using
+00009e90: 2074 736b 6974 0a20 2020 2020 2020 2065   tskit.        e
+00009ea0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
+00009eb0: 205b 696e 642e 5f73 6574 5f67 286e 702e   [ind._set_g(np.
+00009ec0: 7a65 726f 7328 0a20 2020 2020 2020 2020  zeros(.         
+00009ed0: 2020 2020 2020 2020 2020 2020 2020 2028                 (
+00009ee0: 7365 6c66 2e67 656e 5f61 7263 682e 4c2c  self.gen_arch.L,
+00009ef0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00009f00: 2020 2020 2020 2020 2020 7365 6c66 2e67            self.g
+00009f10: 656e 5f61 7263 682e 7829 2929 2066 6f72  en_arch.x))) for
+00009f20: 2069 6e64 2069 6e20 7365 6c66 2e76 616c   ind in self.val
+00009f30: 7565 7328 295d 0a0a 2020 2020 2320 6164  ues()]..    # ad
+00009f40: 6420 6e65 7720 726f 7720 746f 2074 6865  d new row to the
+00009f50: 2069 6e64 6976 6964 7561 6c73 2720 6e75   individuals' nu
+00009f60: 6d70 7920 6172 7261 7973 2c20 666f 7220  mpy arrays, for 
+00009f70: 6120 6769 7665 6e20 6d75 7461 7469 6f6e  a given mutation
+00009f80: 206c 6f63 7573 0a20 2020 2064 6566 205f   locus.    def _
+00009f90: 6164 645f 6e65 775f 6c6f 6375 7328 7365  add_new_locus(se
+00009fa0: 6c66 2c20 6964 782c 206c 6f63 7573 293a  lf, idx, locus):
+00009fb0: 0a20 2020 2020 2020 2023 2069 6e73 6572  .        # inser
+00009fc0: 7420 6120 726f 7720 6f66 206f 6e65 7320  t a row of ones 
+00009fd0: 696e 2065 6163 6820 696e 6469 7669 6475  in each individu
+00009fe0: 616c 2773 2067 656e 6f74 7970 6520 6172  al's genotype ar
+00009ff0: 7261 7920 0a20 2020 2020 2020 205b 696e  ray .        [in
+0000a000: 642e 5f61 6464 5f6e 6577 5f6c 6f63 7573  d._add_new_locus
+0000a010: 2869 6478 2920 666f 7220 696e 6420 696e  (idx) for ind in
+0000a020: 2073 656c 662e 7661 6c75 6573 2829 5d0a   self.values()].
+0000a030: 0a20 2020 2023 6d65 7468 6f64 2074 6f20  .    #method to 
+0000a040: 7365 7420 7468 6520 696e 6469 7669 6475  set the individu
+0000a050: 616c 7327 2065 6e76 6972 6f6e 6d65 6e74  als' environment
+0000a060: 2076 616c 7565 730a 2020 2020 6465 6620   values.    def 
+0000a070: 5f73 6574 5f65 2873 656c 662c 206c 616e  _set_e(self, lan
+0000a080: 642c 2069 6e64 6976 6964 7320 3d20 4e6f  d, individs = No
+0000a090: 6e65 293a 0a20 2020 2020 2020 2069 6620  ne):.        if 
+0000a0a0: 696e 6469 7669 6473 2069 7320 4e6f 6e65  individs is None
+0000a0b0: 3a0a 2020 2020 2020 2020 2020 2020 696e  :.            in
+0000a0c0: 6473 5f74 6f5f 7365 7420 3d20 7365 6c66  ds_to_set = self
+0000a0d0: 2e76 616c 7565 7328 290a 2020 2020 2020  .values().      
+0000a0e0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+0000a0f0: 2020 2020 6967 203d 2069 7465 6d67 6574      ig = itemget
+0000a100: 7465 7228 2a69 6e64 6976 6964 7329 0a20  ter(*individs). 
+0000a110: 2020 2020 2020 2020 2020 2069 6e64 735f             inds_
+0000a120: 746f 5f73 6574 203d 2069 6728 7365 6c66  to_set = ig(self
+0000a130: 290a 2020 2020 2020 2020 2020 2020 6966  ).            if
+0000a140: 2069 7369 6e73 7461 6e63 6528 696e 6473   isinstance(inds
+0000a150: 5f74 6f5f 7365 742c 2069 6e64 6976 6964  _to_set, individ
+0000a160: 7561 6c2e 496e 6469 7669 6475 616c 293a  ual.Individual):
+0000a170: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000a180: 2069 6e64 735f 746f 5f73 6574 203d 2028   inds_to_set = (
+0000a190: 696e 6473 5f74 6f5f 7365 742c 290a 2020  inds_to_set,).  
+0000a1a0: 2020 2020 2020 6861 6220 3d20 5b69 6e64        hab = [ind
+0000a1b0: 2e5f 7365 745f 6528 5b6c 7972 2e72 6173  ._set_e([lyr.ras
+0000a1c0: 745b 696e 7428 696e 642e 7929 2c20 696e  t[int(ind.y), in
+0000a1d0: 7428 0a20 2020 2020 2020 2020 2020 2069  t(.            i
+0000a1e0: 6e64 2e78 295d 2066 6f72 206c 7972 2069  nd.x)] for lyr i
+0000a1f0: 6e20 6c61 6e64 2e76 616c 7565 7328 295d  n land.values()]
+0000a200: 2920 666f 7220 696e 6420 696e 2069 6e64  ) for ind in ind
+0000a210: 735f 746f 5f73 6574 5d0a 0a20 2020 2023  s_to_set]..    #
+0000a220: 6d65 7468 6f64 2074 6f20 7365 7420 7468  method to set th
+0000a230: 6520 696e 6469 7669 6475 616c 7327 2070  e individuals' p
+0000a240: 6865 6e6f 7479 7065 2061 7474 7269 6275  henotype attribu
+0000a250: 7465 7320 0a20 2020 2064 6566 205f 7365  tes .    def _se
+0000a260: 745f 7a28 7365 6c66 293a 0a20 2020 2020  t_z(self):.     
+0000a270: 2020 205b 696e 642e 5f73 6574 5f7a 2873     [ind._set_z(s
+0000a280: 656c 662e 6765 6e5f 6172 6368 2920 666f  elf.gen_arch) fo
+0000a290: 7220 696e 6420 696e 2073 656c 662e 7661  r ind in self.va
+0000a2a0: 6c75 6573 2829 5d3b 0a0a 2020 2020 236d  lues()];..    #m
+0000a2b0: 6574 686f 6420 666f 7220 7365 7474 696e  ethod for settin
+0000a2c0: 6720 616e 2069 6e64 6976 6964 7561 6c27  g an individual'
+0000a2d0: 7320 7068 656e 6f74 7970 650a 2020 2020  s phenotype.    
+0000a2e0: 6465 6620 5f73 6574 5f7a 5f69 6e64 6976  def _set_z_indiv
+0000a2f0: 6964 2873 656c 662c 2069 6e64 6976 6964  id(self, individ
+0000a300: 293a 0a20 2020 2020 2020 2073 656c 665b  ):.        self[
+0000a310: 696e 6469 7669 645d 2e5f 7365 745f 7a28  individ]._set_z(
+0000a320: 7365 6c66 2e67 656e 5f61 7263 6829 0a0a  self.gen_arch)..
+0000a330: 2020 2020 236d 6574 686f 6420 746f 2073      #method to s
+0000a340: 6574 2074 6865 2069 6e64 6976 6964 7561  et the individua
+0000a350: 6c73 2720 6669 746e 6573 7320 6174 7472  ls' fitness attr
+0000a360: 6962 7574 6573 0a20 2020 2064 6566 205f  ibutes.    def _
+0000a370: 7365 745f 6669 7428 7365 6c66 2c20 6669  set_fit(self, fi
+0000a380: 7429 3a0a 2020 2020 2020 2020 5b69 6e64  t):.        [ind
+0000a390: 2e5f 7365 745f 6669 7428 6629 2066 6f72  ._set_fit(f) for
+0000a3a0: 2069 6e64 2c20 6620 696e 207a 6970 2873   ind, f in zip(s
+0000a3b0: 656c 662e 7661 6c75 6573 2829 2c20 6669  elf.values(), fi
+0000a3c0: 7429 5d3b 0a0a 2020 2020 236d 6574 686f  t)];..    #metho
+0000a3d0: 6420 746f 2073 6574 2073 7065 6369 6573  d to set species
+0000a3e0: 2720 636f 6f72 6473 2061 6e64 2063 656c  ' coords and cel
+0000a3f0: 6c73 2061 7272 6179 730a 2020 2020 6465  ls arrays.    de
+0000a400: 6620 5f73 6574 5f63 6f6f 7264 735f 616e  f _set_coords_an
+0000a410: 645f 6365 6c6c 7328 7365 6c66 293a 0a20  d_cells(self):. 
+0000a420: 2020 2020 2020 2073 656c 662e 5f63 6f6f         self._coo
+0000a430: 7264 7320 3d20 7365 6c66 2e5f 6765 745f  rds = self._get_
+0000a440: 636f 6f72 6473 2829 0a20 2020 2020 2020  coords().       
+0000a450: 2073 656c 662e 5f63 656c 6c73 203d 206e   self._cells = n
+0000a460: 702e 696e 7433 3228 6e70 2e66 6c6f 6f72  p.int32(np.floor
+0000a470: 2873 656c 662e 5f63 6f6f 7264 7329 290a  (self._coords)).
+0000a480: 0a20 2020 2023 6d65 7468 6f64 2074 6f20  .    #method to 
+0000a490: 7365 7420 7468 6520 7370 6563 6965 7327  set the species'
+0000a4a0: 206b 645f 7472 6565 2061 7474 7269 6275   kd_tree attribu
+0000a4b0: 7465 2028 6120 7370 6174 6961 6c2e 5f4b  te (a spatial._K
+0000a4c0: 4454 7265 6529 0a20 2020 2064 6566 205f  DTree).    def _
+0000a4d0: 7365 745f 6b64 5f74 7265 6528 7365 6c66  set_kd_tree(self
+0000a4e0: 2c20 6c65 6166 7369 7a65 3d31 3030 293a  , leafsize=100):
+0000a4f0: 0a20 2020 2020 2020 2073 656c 662e 5f6b  .        self._k
+0000a500: 645f 7472 6565 203d 2073 7074 2e5f 4b44  d_tree = spt._KD
+0000a510: 5472 6565 2863 6f6f 7264 733d 7365 6c66  Tree(coords=self
+0000a520: 2e5f 636f 6f72 6473 2c0a 2020 2020 2020  ._coords,.      
 0000a530: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a540: 2020 2020 2020 2020 2020 2077 696e 646f             windo
-0000a550: 775f 7769 6474 6820 3d20 7365 6c66 2e64  w_width = self.d
-0000a560: 656e 7369 7479 5f67 7269 645f 7769 6e64  ensity_grid_wind
-0000a570: 6f77 5f77 6964 7468 290a 0a0a 2020 2020  ow_width)...    
-0000a580: 2320 6d65 7468 6f64 2074 6f20 7365 7420  # method to set 
-0000a590: 7468 6520 7370 6563 6965 7327 2067 656e  the species' gen
-0000a5a0: 6f6d 6573 2c20 6368 6563 6b20 666f 7220  omes, check for 
-0000a5b0: 6164 6571 7561 7465 206d 7574 6162 6c65  adequate mutable
-0000a5c0: 206c 6f63 692c 0a20 2020 2023 2061 6e64   loci,.    # and
-0000a5d0: 2066 696c 6c20 7468 6520 7473 6b69 742e   fill the tskit.
-0000a5e0: 5461 626c 6543 6f6c 6c65 6374 696f 6e27  TableCollection'
-0000a5f0: 7320 7461 626c 6573 0a20 2020 2023 2028  s tables.    # (
-0000a600: 746f 2062 6520 6361 6c6c 6564 2061 6674  to be called aft
-0000a610: 6572 2074 6865 206d 6f64 656c 2068 6173  er the model has
-0000a620: 2062 7572 6e65 6420 696e 290a 2020 2020   burned in).    
-0000a630: 6465 6620 5f73 6574 5f67 656e 6f6d 6573  def _set_genomes
-0000a640: 5f61 6e64 5f74 6162 6c65 7328 7365 6c66  _and_tables(self
-0000a650: 2c20 6275 726e 5f54 2c20 5429 3a0a 2020  , burn_T, T):.  
-0000a660: 2020 2020 2020 2320 7365 7420 7374 6172        # set star
-0000a670: 7469 6e67 2067 656e 6f6d 6520 6172 7261  ting genome arra
-0000a680: 7973 2074 6f20 616c 6c20 7a65 726f 730a  ys to all zeros.
-0000a690: 2020 2020 2020 2020 7365 6c66 2e5f 7365          self._se
-0000a6a0: 745f 6e75 6c6c 5f67 656e 6f6d 6573 2829  t_null_genomes()
-0000a6b0: 0a0a 2020 2020 2020 2020 2320 7573 6520  ..        # use 
-0000a6c0: 6d65 616e 206e 5f62 6972 7468 7320 6174  mean n_births at
-0000a6d0: 2074 6169 6c20 656e 6420 6f66 2062 7572   tail end of bur
-0000a6e0: 6e2d 696e 2074 6f20 6573 7469 6d61 7465  n-in to estimate
-0000a6f0: 206e 756d 6265 7220 6f66 0a20 2020 2020   number of.     
-0000a700: 2020 2023 206d 7574 6174 696f 6e73 2066     # mutations f
-0000a710: 6f72 2074 6869 7320 7370 6563 6965 732c  or this species,
-0000a720: 2074 6865 6e20 7761 726e 2074 6865 2075   then warn the u
-0000a730: 7365 7220 6966 2074 6865 7265 2069 7320  ser if there is 
-0000a740: 6f72 206d 6179 2062 650a 2020 2020 2020  or may be.      
-0000a750: 2020 2320 696e 6164 6571 7561 7465 2073    # inadequate s
-0000a760: 7061 6365 2066 6f72 2074 6865 2070 6172  pace for the par
-0000a770: 616d 6574 6572 697a 6564 206d 7574 6174  ameterized mutat
-0000a780: 696f 6e20 7261 7465 2028 6265 6361 7573  ion rate (becaus
-0000a790: 6520 6f6e 6c79 0a20 2020 2020 2020 2023  e only.        #
-0000a7a0: 2075 7369 6e67 2069 6e66 696e 6974 652d   using infinite-
-0000a7b0: 7369 7465 7320 6d75 7461 7469 6f6e 290a  sites mutation).
-0000a7c0: 2020 2020 2020 2020 6573 745f 746f 745f          est_tot_
-0000a7d0: 6d75 7473 203d 205f 6361 6c63 5f65 7374  muts = _calc_est
-0000a7e0: 696d 6174 6564 5f74 6f74 616c 5f6d 7574  imated_total_mut
-0000a7f0: 6174 696f 6e73 2873 656c 662c 2062 7572  ations(self, bur
-0000a800: 6e5f 542c 2054 290a 0a20 2020 2020 2020  n_T, T)..       
-0000a810: 2023 2063 6865 636b 2077 6865 7468 6572   # check whether
-0000a820: 2074 6865 7265 2061 7265 2061 6465 7175   there are adequ
-0000a830: 6174 6520 6d75 7461 626c 6520 6c6f 6369  ate mutable loci
-0000a840: 2066 6f72 2074 6869 7320 7370 6563 6965   for this specie
-0000a850: 730a 2020 2020 2020 2020 5f63 6865 636b  s.        _check
-0000a860: 5f6d 7574 6174 696f 6e5f 7261 7465 7328  _mutation_rates(
-0000a870: 7365 6c66 2e67 656e 5f61 7263 682c 2065  self.gen_arch, e
-0000a880: 7374 5f74 6f74 5f6d 7574 732c 2062 7572  st_tot_muts, bur
-0000a890: 6e5f 542c 2054 290a 0a20 2020 2020 2020  n_T, T)..       
-0000a8a0: 2023 2063 7265 6174 6520 616e 6420 636f   # create and co
-0000a8b0: 6e66 6967 7572 6520 7473 6b69 7420 7461  nfigure tskit ta
-0000a8c0: 626c 6573 0a20 2020 2020 2020 2069 6620  bles.        if 
-0000a8d0: 7365 6c66 2e67 656e 5f61 7263 682e 7573  self.gen_arch.us
-0000a8e0: 655f 7473 6b69 743a 0a20 2020 2020 2020  e_tskit:.       
-0000a8f0: 2020 2020 2023 2073 696d 756c 6174 6520       # simulate 
-0000a900: 6120 636f 616c 6573 6365 6e74 2061 6e63  a coalescent anc
-0000a910: 6573 7472 7920 7769 7468 206e 756d 6265  estry with numbe
-0000a920: 7220 6f66 2073 616d 706c 6573 2065 7175  r of samples equ
-0000a930: 616c 2074 6f20 6f75 720a 2020 2020 2020  al to our.      
-0000a940: 2020 2020 2020 2320 7370 6563 6965 7327        # species'
-0000a950: 206e 756d 6265 7220 6f66 2068 6170 6c6f   number of haplo
-0000a960: 6964 2067 656e 6f6d 6573 2028 692e 652e  id genomes (i.e.
-0000a970: 2032 2a4e 5f30 290a 2020 2020 2020 2020   2*N_0).        
-0000a980: 2020 2020 7473 203d 206d 7370 7269 6d65      ts = msprime
-0000a990: 2e73 696d 756c 6174 6528 6c65 6e28 7365  .simulate(len(se
-0000a9a0: 6c66 2920 2a20 322c 0a20 2020 2020 2020  lf) * 2,.       
-0000a9b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a9c0: 2020 2020 2020 2020 2020 204e 653d 3130             Ne=10
-0000a9d0: 3030 2c0a 2020 2020 2020 2020 2020 2020  00,.            
-0000a9e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a9f0: 2020 2020 2020 6c65 6e67 7468 3d73 656c        length=sel
-0000aa00: 662e 6765 6e5f 6172 6368 2e4c 290a 0a20  f.gen_arch.L).. 
-0000aa10: 2020 2020 2020 2020 2020 2023 2074 6865             # the
-0000aa20: 6e20 6772 6162 2074 6865 2073 696d 756c  n grab the simul
-0000aa30: 6174 696f 6e27 7320 7472 6565 2c20 616e  ation's tree, an
-0000aa40: 6420 7468 6520 7472 6565 2773 2054 6162  d the tree's Tab
-0000aa50: 6c65 436f 6c6c 6563 7469 6f6e 0a20 2020  leCollection.   
-0000aa60: 2020 2020 2020 2020 2023 204e 4f54 453a           # NOTE:
-0000aa70: 2074 6865 2054 7265 6553 6571 7565 6e63   the TreeSequenc
-0000aa80: 6520 6f62 6a65 6374 206f 6e6c 7920 6861  e object only ha
-0000aa90: 7320 6f6e 6520 7472 6565 2c0a 2020 2020  s one tree,.    
-0000aaa0: 2020 2020 2020 2020 2320 6265 6361 7573          # becaus
-0000aab0: 6520 6e6f 2072 6563 6f6d 6269 6e61 7469  e no recombinati
-0000aac0: 6f6e 2077 6173 2075 7365 6420 696e 2074  on was used in t
-0000aad0: 6865 2073 696d 0a20 2020 2020 2020 2020  he sim.         
-0000aae0: 2020 2074 7265 6520 3d20 7473 2e66 6972     tree = ts.fir
-0000aaf0: 7374 2829 0a20 2020 2020 2020 2020 2020  st().           
-0000ab00: 2074 6162 6c65 7320 3d20 7473 2e64 756d   tables = ts.dum
-0000ab10: 705f 7461 626c 6573 2829 0a0a 2020 2020  p_tables()..    
-0000ab20: 2020 2020 2020 2020 2320 7365 7420 7468          # set th
-0000ab30: 6520 7365 7175 656e 6365 206c 656e 6774  e sequence lengt
-0000ab40: 680a 2020 2020 2020 2020 2020 2020 2374  h.            #t
-0000ab50: 6162 6c65 732e 7365 7175 656e 6365 5f6c  ables.sequence_l
-0000ab60: 656e 6774 6820 3d20 7365 6c66 2e67 656e  ength = self.gen
-0000ab70: 5f61 7263 682e 4c0a 2020 2020 2020 2020  _arch.L.        
-0000ab80: 2020 2020 2320 636c 6561 7220 7468 6520      # clear the 
-0000ab90: 6d75 7461 7469 6f6e 7320 7461 626c 650a  mutations table.
-0000aba0: 2020 2020 2020 2020 2020 2020 7461 626c              tabl
-0000abb0: 6573 2e6d 7574 6174 696f 6e73 2e63 6c65  es.mutations.cle
-0000abc0: 6172 2829 0a0a 2020 2020 2020 2020 2020  ar()..          
-0000abd0: 2020 2320 6c6f 6f70 206f 7665 7220 616c    # loop over al
-0000abe0: 6c20 7369 7465 732c 2073 6f20 7468 6174  l sites, so that
-0000abf0: 2065 6163 6820 7369 7465 2773 2072 6f77   each site's row
-0000ac00: 2067 6f65 7320 696e 746f 0a20 2020 2020   goes into.     
-0000ac10: 2020 2020 2020 2023 2074 6865 2073 6974         # the sit
-0000ac20: 6573 2074 6162 6c65 2069 6e20 6f72 6465  es table in orde
-0000ac30: 7220 2873 7563 6820 7468 6174 2031 2e29  r (such that 1.)
-0000ac40: 2074 6865 7265 2773 206e 6f20 6e65 6564   there's no need
-0000ac50: 0a20 2020 2020 2020 2020 2020 2023 2074  .            # t
-0000ac60: 6f20 7472 6163 6b20 686f 7720 676e 7820  o track how gnx 
-0000ac70: 7369 7465 7320 6d61 7020 6f6e 746f 2073  sites map onto s
-0000ac80: 6974 6573 2d74 6162 6c65 2072 6f77 2069  ites-table row i
-0000ac90: 6473 2c0a 2020 2020 2020 2020 2020 2020  ds,.            
-0000aca0: 2320 616e 6420 322e 2920 7468 6572 6520  # and 2.) there 
-0000acb0: 7769 6c6c 2062 6520 6e6f 206e 6565 6420  will be no need 
-0000acc0: 746f 2064 6564 7570 6c69 6361 7465 2073  to deduplicate s
-0000acd0: 6974 6573 206c 6174 6572 206f 6e29 0a20  ites later on). 
-0000ace0: 2020 2020 2020 2020 2020 2066 6f72 2073             for s
-0000acf0: 6974 6520 696e 2072 616e 6765 2873 656c  ite in range(sel
-0000ad00: 662e 6765 6e5f 6172 6368 2e4c 293a 0a20  f.gen_arch.L):. 
-0000ad10: 2020 2020 2020 2020 2020 2020 2020 2023                 #
-0000ad20: 6465 7465 726d 696e 6520 7768 6574 6865  determine whethe
-0000ad30: 7220 7468 6973 2069 7320 6120 6e65 7574  r this is a neut
-0000ad40: 7261 6c20 6f72 206e 6f6e 2d6e 6575 7472  ral or non-neutr
-0000ad50: 616c 2073 6974 650a 2020 2020 2020 2020  al site.        
-0000ad60: 2020 2020 2020 2020 6966 2073 6974 6520          if site 
-0000ad70: 696e 2073 656c 662e 6765 6e5f 6172 6368  in self.gen_arch
-0000ad80: 2e6e 6575 745f 6c6f 6369 3a0a 2020 2020  .neut_loci:.    
-0000ad90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000ada0: 6d65 7461 6461 7461 3d27 6e27 2e65 6e63  metadata='n'.enc
-0000adb0: 6f64 6528 2761 7363 6969 2729 0a20 2020  ode('ascii').   
-0000adc0: 2020 2020 2020 2020 2020 2020 2065 6c69               eli
-0000add0: 6620 7369 7465 2069 6e20 7365 6c66 2e67  f site in self.g
-0000ade0: 656e 5f61 7263 682e 6e6f 6e6e 6575 745f  en_arch.nonneut_
-0000adf0: 6c6f 6369 3a0a 2020 2020 2020 2020 2020  loci:.          
-0000ae00: 2020 2020 2020 2020 2020 6d65 7461 6461            metada
-0000ae10: 7461 3d27 7427 2e65 6e63 6f64 6528 2761  ta='t'.encode('a
-0000ae20: 7363 6969 2729 0a20 2020 2020 2020 2020  scii').         
-0000ae30: 2020 2020 2020 2023 2061 6464 2074 6865         # add the
-0000ae40: 2076 6172 6961 6e74 2773 2073 6974 6520   variant's site 
-0000ae50: 746f 2074 6865 2073 6974 6573 2074 6162  to the sites tab
-0000ae60: 6c65 0a20 2020 2020 2020 2020 2020 2020  le.             
-0000ae70: 2020 2074 6162 6c65 732e 7369 7465 732e     tables.sites.
-0000ae80: 6164 645f 726f 7728 706f 7369 7469 6f6e  add_row(position
-0000ae90: 3d73 6974 652c 2061 6e63 6573 7472 616c  =site, ancestral
-0000aea0: 5f73 7461 7465 3d27 3027 2c0a 2020 2020  _state='0',.    
-0000aeb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000aec0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000aed0: 206d 6574 6164 6174 613d 6d65 7461 6461   metadata=metada
-0000aee0: 7461 290a 0a20 2020 2020 2020 2020 2020  ta)..           
-0000aef0: 2023 2067 7261 6220 7468 6520 6e6f 6465   # grab the node
-0000af00: 7320 666c 6167 732c 2077 6869 6368 2061  s flags, which a
-0000af10: 7265 2031 2066 6f72 2063 7572 7265 6e74  re 1 for current
-0000af20: 206e 6f64 6573 2c0a 2020 2020 2020 2020   nodes,.        
-0000af30: 2020 2020 2320 3020 666f 7220 7061 7374      # 0 for past
-0000af40: 206e 6f64 6573 2c20 696e 746f 2074 776f   nodes, into two
-0000af50: 2073 6570 6172 6174 6520 6f62 6a65 6374   separate object
-0000af60: 730a 2020 2020 2020 2020 2020 2020 6375  s.            cu
-0000af70: 7272 656e 745f 6e6f 6465 7320 3d20 5b2a  rrent_nodes = [*
-0000af80: 6e70 2e77 6865 7265 2874 6162 6c65 732e  np.where(tables.
-0000af90: 6e6f 6465 732e 666c 6167 7320 3d3d 2031  nodes.flags == 1
-0000afa0: 295b 305d 5d0a 2020 2020 2020 2020 2020  )[0]].          
-0000afb0: 2020 2320 7265 7665 7273 652c 2073 6f20    # reverse, so 
-0000afc0: 7468 6174 2049 2063 616e 2070 6f70 206e  that I can pop n
-0000afd0: 6f64 6573 206f 6666 2074 6865 2027 6672  odes off the 'fr
-0000afe0: 6f6e 7427 0a20 2020 2020 2020 2020 2020  ont'.           
-0000aff0: 2063 7572 7265 6e74 5f6e 6f64 6573 203d   current_nodes =
-0000b000: 2063 7572 7265 6e74 5f6e 6f64 6573 5b3a   current_nodes[:
-0000b010: 3a2d 315d 0a20 2020 2020 2020 2020 2020  :-1].           
-0000b020: 2070 6173 745f 6e6f 6465 7320 3d20 5b2a   past_nodes = [*
-0000b030: 6e70 2e77 6865 7265 2874 6162 6c65 732e  np.where(tables.
-0000b040: 6e6f 6465 732e 666c 6167 7320 213d 2031  nodes.flags != 1
-0000b050: 295b 305d 5d0a 2020 2020 2020 2020 2020  )[0]].          
-0000b060: 2020 2320 6372 6561 7465 2061 6e20 656d    # create an em
-0000b070: 7074 7920 6c69 7374 2c20 746f 2066 696c  pty list, to fil
-0000b080: 6c20 7570 2074 6865 2069 6e64 6976 6964  l up the individ
-0000b090: 7561 6c20 6964 7320 666f 7220 6561 6368  ual ids for each
-0000b0a0: 206e 6f64 650a 2020 2020 2020 2020 2020   node.          
-0000b0b0: 2020 2320 4e4f 5445 3a20 7365 7474 696e    # NOTE: settin
-0000b0c0: 6720 746f 2061 2076 6563 746f 7220 6f66  g to a vector of
-0000b0d0: 202d 3120 696e 6974 6961 6c6c 792c 2074   -1 initially, t
-0000b0e0: 6f20 6561 7369 6c79 2063 6865 636b 2074  o easily check t
-0000b0f0: 6861 740a 2020 2020 2020 2020 2020 2020  hat.            
-0000b100: 2320 616c 6c20 7661 6c75 6573 2068 6176  # all values hav
-0000b110: 6520 6265 656e 2061 7373 6967 6e65 6420  e been assigned 
-0000b120: 6174 2074 6865 2065 6e64 2062 7920 6173  at the end by as
-0000b130: 6b69 6e67 2069 6620 616c 6c20 3e3d 2030  king if all >= 0
-0000b140: 200a 2020 2020 2020 2020 2020 2020 6e6f   .            no
-0000b150: 6465 735f 7461 625f 696e 6469 7669 6475  des_tab_individu
-0000b160: 616c 5f63 6f6c 203d 206e 702e 696e 7433  al_col = np.int3
-0000b170: 3228 6e70 2e6f 6e65 7328 6c65 6e28 0a20  2(np.ones(len(. 
-0000b180: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b190: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b1a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b1b0: 2020 2020 2020 2074 6162 6c65 732e 6e6f         tables.no
-0000b1c0: 6465 732e 666c 6167 7329 292a 2d31 290a  des.flags))*-1).
-0000b1d0: 0a20 2020 2020 2020 2020 2020 2023 204e  .            # N
-0000b1e0: 4f54 453a 2074 6865 7265 2061 7265 206e  OTE: there are n
-0000b1f0: 6f20 7265 7175 6972 656d 656e 7473 0a20  o requirements. 
-0000b200: 2020 2020 2020 2020 2020 2023 2020 2020             #    
-0000b210: 2020 206f 7220 7265 7374 7269 6374 696f     or restrictio
-0000b220: 6e73 2066 6f72 2074 6865 2069 6e64 6976  ns for the indiv
-0000b230: 6964 7561 6c73 0a20 2020 2020 2020 2020  iduals.         
-0000b240: 2020 2023 2061 6e64 206e 6f64 6573 2074     # and nodes t
-0000b250: 6162 6c65 7320 2865 2e67 2e20 6f72 6465  ables (e.g. orde
-0000b260: 722c 2065 7463 2e29 2c20 616e 6420 7468  r, etc.), and th
-0000b270: 7573 2074 6865 2074 6162 6c65 7327 206f  us the tables' o
-0000b280: 7264 6572 0a20 2020 2020 2020 2020 2020  rder.           
-0000b290: 2023 2069 7320 6e6f 7420 6166 6665 6374   # is not affect
-0000b2a0: 6564 2062 7920 7468 6520 5461 626c 6543  ed by the TableC
-0000b2b0: 6f6c 6c65 6374 696f 6e2e 7369 6d70 6c69  ollection.simpli
-0000b2c0: 6679 2061 6c67 6f72 6974 686d 2e0a 2020  fy algorithm..  
-0000b2d0: 2020 2020 2020 2020 2020 2320 536f 2c20            # So, 
-0000b2e0: 4920 636f 756c 6420 6164 6420 6569 7468  I could add eith
-0000b2f0: 6572 2063 7572 7265 6e74 0a20 2020 2020  er current.     
-0000b300: 2020 2020 2020 2023 206f 7220 7061 7374         # or past
-0000b310: 2069 6e64 6976 6964 7561 6c73 2061 6e64   individuals and
-0000b320: 206e 6f64 6573 2066 6972 7374 3b0a 2020   nodes first;.  
-0000b330: 2020 2020 2020 2020 2020 2320 6368 6f6f            # choo
-0000b340: 7369 6e67 2074 6f20 6164 6420 7061 7374  sing to add past
-0000b350: 2066 6972 7374 2c20 736f 2074 6861 7420   first, so that 
-0000b360: 616c 6c20 2772 6561 6c27 2069 6e64 6976  all 'real' indiv
-0000b370: 6964 7561 6c73 0a20 2020 2020 2020 2020  iduals.         
-0000b380: 2020 2023 2066 726f 6d20 7468 6520 7374     # from the st
-0000b390: 6172 7420 6f66 2074 6865 2067 656f 6e6f  art of the geono
-0000b3a0: 6d69 6373 2073 696d 756c 6174 696f 6e20  mics simulation 
-0000b3b0: 666f 7277 6172 6420 7769 6c6c 0a20 2020  forward will.   
-0000b3c0: 2020 2020 2020 2020 2023 2077 696e 6420           # wind 
-0000b3d0: 7570 2068 6176 696e 6720 7468 6569 7220  up having their 
-0000b3e0: 696e 6469 7669 6475 616c 7320 616e 6420  individuals and 
-0000b3f0: 6e6f 6465 7320 726f 7773 2069 6e20 6120  nodes rows in a 
-0000b400: 7369 6e67 6c65 0a20 2020 2020 2020 2020  single.         
-0000b410: 2020 2023 2062 6c6f 636b 2061 7420 7468     # block at th
-0000b420: 6520 7461 626c 6573 2720 626f 7474 6f6d  e tables' bottom
-0000b430: 730a 0a20 2020 2020 2020 2020 2020 2023  s..            #
-0000b440: 2061 6464 2061 6e20 696e 6469 7669 6475   add an individu
-0000b450: 616c 2074 6f20 7468 6520 696e 6469 7669  al to the indivi
-0000b460: 6475 616c 7320 7461 626c 6520 666f 720a  duals table for.
-0000b470: 2020 2020 2020 2020 2020 2020 2320 6561              # ea
-0000b480: 6368 2063 6f61 6c65 7363 656e 742d 7369  ch coalescent-si
-0000b490: 6d75 6c61 7465 6420 6e6f 6465 2062 6566  mulated node bef
-0000b4a0: 6f72 6520 7468 6520 6375 7272 656e 7420  ore the current 
-0000b4b0: 7469 6d65 0a20 2020 2020 2020 2020 2020  time.           
-0000b4c0: 2023 204e 4f54 453a 2061 6464 696e 6720   # NOTE: adding 
-0000b4d0: 6e6f 206d 6574 6164 6174 612c 2061 6e64  no metadata, and
-0000b4e0: 206e 6f20 6c6f 6361 7469 6f6e 2c0a 2020   no location,.  
-0000b4f0: 2020 2020 2020 2020 2020 2320 2020 2020            #     
-0000b500: 2020 746f 2069 6e64 6963 6174 6520 7468    to indicate th
-0000b510: 6174 2074 6869 7320 6973 2061 2027 6661  at this is a 'fa
-0000b520: 6b65 2720 696e 6469 7669 6475 616c 2c0a  ke' individual,.
-0000b530: 2020 2020 2020 2020 2020 2020 2320 2020              #   
-0000b540: 2020 2020 696e 7665 6e74 6564 206a 7573      invented jus
-0000b550: 7420 746f 206d 6174 6368 2075 7020 746f  t to match up to
-0000b560: 2074 6865 206e 6f64 6573 0a20 2020 2020   the nodes.     
-0000b570: 2020 2020 2020 2023 2020 2020 2020 2073         #       s
-0000b580: 696d 756c 6174 6564 2066 6f72 2074 6865  imulated for the
-0000b590: 2073 7461 7274 696e 6720 706f 7075 6c61   starting popula
-0000b5a0: 7469 6f6e 0a20 2020 2020 2020 2020 2020  tion.           
-0000b5b0: 2066 6f72 206e 6f64 6520 696e 2070 6173   for node in pas
-0000b5c0: 745f 6e6f 6465 733a 0a20 2020 2020 2020  t_nodes:.       
-0000b5d0: 2020 2020 2020 2020 2069 6e64 5f69 6420           ind_id 
-0000b5e0: 3d20 7461 626c 6573 2e69 6e64 6976 6964  = tables.individ
-0000b5f0: 7561 6c73 2e61 6464 5f72 6f77 2866 6c61  uals.add_row(fla
-0000b600: 6773 3d30 290a 2020 2020 2020 2020 2020  gs=0).          
-0000b610: 2020 2020 2020 2320 7374 6f72 6520 6974        # store it
-0000b620: 7320 696e 6469 7669 6420 6964 2069 6e20  s individ id in 
-0000b630: 7468 6520 6e6f 6465 7320 7461 626c 6527  the nodes table'
-0000b640: 7320 696e 6469 7669 6475 616c 7320 636f  s individuals co
-0000b650: 6c75 6d6e 0a20 2020 2020 2020 2020 2020  lumn.           
-0000b660: 2020 2020 206e 6f64 6573 5f74 6162 5f69       nodes_tab_i
-0000b670: 6e64 6976 6964 7561 6c5f 636f 6c5b 6e6f  ndividual_col[no
-0000b680: 6465 5d20 3d20 696e 645f 6964 0a0a 2020  de] = ind_id..  
-0000b690: 2020 2020 2020 2020 2020 2320 6372 6561            # crea
-0000b6a0: 7465 2061 6e64 2061 6464 2074 6f20 7468  te and add to th
-0000b6b0: 6520 696e 6469 7669 6475 616c 7320 7461  e individuals ta
-0000b6c0: 626c 6520 6120 6e65 7720 726f 7720 666f  ble a new row fo
-0000b6d0: 720a 2020 2020 2020 2020 2020 2020 2320  r.            # 
-0000b6e0: 6561 6368 2072 6561 6c20 696e 6469 7669  each real indivi
-0000b6f0: 6475 616c 0a20 2020 2020 2020 2020 2020  dual.           
-0000b700: 2066 6f72 2069 6e64 2069 6e20 7365 6c66   for ind in self
-0000b710: 2e76 616c 7565 7328 293a 0a20 2020 2020  .values():.     
-0000b720: 2020 2020 2020 2020 2020 2023 2067 6574             # get
-0000b730: 2074 6865 2027 6c6f 6361 7469 6f6e 2720   the 'location' 
-0000b740: 636f 6c75 6d6e 2069 6e66 6f2c 2077 6869  column info, whi
-0000b750: 6368 2077 696c 6c20 696e 636c 7564 6520  ch will include 
-0000b760: 7468 6520 7820 616e 6420 790a 2020 2020  the x and y.    
-0000b770: 2020 2020 2020 2020 2020 2020 2320 706f              # po
-0000b780: 7369 7469 6f6e 7320 6f66 2061 6e20 696e  sitions of an in
-0000b790: 6469 7669 6475 616c 2c20 6173 2077 656c  dividual, as wel
-0000b7a0: 6c20 6173 2074 6865 2069 6e64 6976 6964  l as the individ
-0000b7b0: 7561 6c27 730a 2020 2020 2020 2020 2020  ual's.          
-0000b7c0: 2020 2020 2020 2320 7068 656e 6f74 7970        # phenotyp
-0000b7d0: 6573 2061 6e64 2066 6974 6e65 7373 2c20  es and fitness, 
-0000b7e0: 6966 2074 7261 6974 7320 6172 6520 6265  if traits are be
-0000b7f0: 696e 6720 7573 6564 0a20 2020 2020 2020  ing used.       
-0000b800: 2020 2020 2020 2020 206c 6f63 203d 205b           loc = [
-0000b810: 696e 642e 782c 2069 6e64 2e79 5d0a 2020  ind.x, ind.y].  
-0000b820: 2020 2020 2020 2020 2020 2020 2020 6966                if
-0000b830: 2073 656c 662e 6765 6e5f 6172 6368 2e74   self.gen_arch.t
-0000b840: 7261 6974 7320 6973 206e 6f74 204e 6f6e  raits is not Non
-0000b850: 653a 0a20 2020 2020 2020 2020 2020 2020  e:.             
-0000b860: 2020 2020 2020 206c 6f63 203d 206c 6f63         loc = loc
-0000b870: 202b 2069 6e64 2e7a 202b 205b 696e 642e   + ind.z + [ind.
-0000b880: 6669 745d 0a20 2020 2020 2020 2020 2020  fit].           
-0000b890: 2020 2020 2023 2061 6464 2061 206e 6577       # add a new
-0000b8a0: 2072 6f77 2074 6f20 7468 6520 696e 6469   row to the indi
-0000b8b0: 7669 6475 616c 7320 7461 626c 652c 2073  viduals table, s
-0000b8c0: 6574 7469 6e67 2074 6865 206c 6f63 6174  etting the locat
-0000b8d0: 696f 6e0a 2020 2020 2020 2020 2020 2020  ion.            
-0000b8e0: 2020 2020 2320 636f 6c75 6d6e 2773 2076      # column's v
-0000b8f0: 616c 7565 2074 6f20 6c6f 630a 2020 2020  alue to loc.    
-0000b900: 2020 2020 2020 2020 2020 2020 2320 4e4f              # NO
-0000b910: 5445 3a20 7573 696e 6720 7468 6520 6d65  TE: using the me
-0000b920: 7461 6461 7461 2063 6f6c 756d 6e20 746f  tadata column to
-0000b930: 2073 746f 7265 2074 6f20 7468 6520 676e   store to the gn
-0000b940: 780a 2020 2020 2020 2020 2020 2020 2020  x.              
-0000b950: 2020 2320 696e 6469 7669 6475 616c 2069    # individual i
-0000b960: 6478 2c20 666f 7220 6c61 7465 7220 6d61  dx, for later ma
-0000b970: 7463 6869 6e67 2074 6f20 7570 6461 7465  tching to update
-0000b980: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000b990: 2023 2049 6e64 6976 6964 7561 6c2e 5f69   # Individual._i
-0000b9a0: 6e64 6976 6964 7561 6c73 5f74 6162 5f69  ndividuals_tab_i
-0000b9b0: 6420 6166 7465 7220 7473 6b69 7427 7320  d after tskit's 
-0000b9c0: 7369 6d70 6c69 6679 0a20 2020 2020 2020  simplify.       
-0000b9d0: 2020 2020 2020 2020 2023 2061 6c67 6f72           # algor
-0000b9e0: 6974 686d 2066 696c 7465 7273 2069 6e64  ithm filters ind
-0000b9f0: 6976 6964 7561 6c73 0a20 2020 2020 2020  ividuals.       
-0000ba00: 2020 2020 2020 2020 2069 6e64 5f69 6420           ind_id 
-0000ba10: 3d20 7461 626c 6573 2e69 6e64 6976 6964  = tables.individ
-0000ba20: 7561 6c73 2e61 6464 5f72 6f77 2866 6c61  uals.add_row(fla
-0000ba30: 6773 3d31 2c20 6c6f 6361 7469 6f6e 3d6c  gs=1, location=l
-0000ba40: 6f63 2c0a 2020 2020 2020 2020 2020 2020  oc,.            
-0000ba50: 2020 2020 2020 2020 6d65 7461 6461 7461          metadata
-0000ba60: 3d69 6e64 2e69 6478 2e74 6f5f 6279 7465  =ind.idx.to_byte
-0000ba70: 7328 6c65 6e67 7468 3d34 2c20 6279 7465  s(length=4, byte
-0000ba80: 6f72 6465 723d 276c 6974 746c 6527 2929  order='little'))
-0000ba90: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000baa0: 2069 6e64 2e5f 696e 6469 7669 6475 616c   ind._individual
-0000bab0: 735f 7461 625f 6964 203d 2069 6e64 5f69  s_tab_id = ind_i
-0000bac0: 640a 0a20 2020 2020 2020 2020 2020 2020  d..             
-0000bad0: 2020 2023 2061 7373 6967 6e20 7468 6520     # assign the 
-0000bae0: 696e 6469 7669 6475 616c 2032 2072 616e  individual 2 ran
-0000baf0: 646f 6d6c 7920 6368 6f73 656e 206e 6f64  domly chosen nod
-0000bb00: 6573 2066 726f 6d0a 2020 2020 2020 2020  es from.        
-0000bb10: 2020 2020 2020 2020 2320 7468 6520 6375          # the cu
-0000bb20: 7272 656e 7420 7469 6d65 2073 7465 702c  rrent time step,
-0000bb30: 2061 6e64 2061 7373 6f63 6961 7465 2074   and associate t
-0000bb40: 6865 2069 6e64 6976 6964 7561 6c27 730a  he individual's.
-0000bb50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000bb60: 2320 5f69 6e64 6976 6964 7561 6c73 5f74  # _individuals_t
-0000bb70: 6162 5f69 6420 7769 7468 2074 686f 7365  ab_id with those
-0000bb80: 2032 206e 6f64 6573 2069 6e20 736f 6d65   2 nodes in some
-0000bb90: 2064 6174 610a 2020 2020 2020 2020 2020   data.          
-0000bba0: 2020 2020 2020 2320 7374 7275 6374 7572        # structur
-0000bbb0: 6520 746f 2063 6f6c 6c65 6374 2074 6869  e to collect thi
-0000bbc0: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
-0000bbd0: 2020 696e 645f 6e6f 6465 5f69 6473 203d    ind_node_ids =
-0000bbe0: 5b63 7572 7265 6e74 5f6e 6f64 6573 2e70  [current_nodes.p
-0000bbf0: 6f70 2829 2066 6f72 205f 2069 6e20 7261  op() for _ in ra
-0000bc00: 6e67 6528 0a20 2020 2020 2020 2020 2020  nge(.           
-0000bc10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000bc20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000bc30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000bc40: 2073 656c 662e 6765 6e5f 6172 6368 2e78   self.gen_arch.x
-0000bc50: 295d 0a20 2020 2020 2020 2020 2020 2020  )].             
-0000bc60: 2020 2069 6e64 2e5f 7365 745f 6e6f 6465     ind._set_node
-0000bc70: 735f 7461 625f 6964 7328 2a69 6e64 5f6e  s_tab_ids(*ind_n
-0000bc80: 6f64 655f 6964 7329 0a20 2020 2020 2020  ode_ids).       
-0000bc90: 2020 2020 2020 2020 2023 2061 6464 2074           # add t
-0000bca0: 6869 7320 696e 6469 7669 6475 616c 2773  his individual's
-0000bcb0: 2069 6e64 5f69 6420 746f 2074 6865 0a20   ind_id to the. 
-0000bcc0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
-0000bcd0: 206e 6f64 6573 5f74 6162 5f69 6e64 6976   nodes_tab_indiv
-0000bce0: 6964 7561 6c5f 636f 6c20 6c69 7374 2c20  idual_col list, 
-0000bcf0: 6f6e 6365 2066 6f72 2065 6163 6820 6e6f  once for each no
-0000bd00: 6465 0a20 2020 2020 2020 2020 2020 2020  de.             
-0000bd10: 2020 206e 6f64 6573 5f74 6162 5f69 6e64     nodes_tab_ind
-0000bd20: 6976 6964 7561 6c5f 636f 6c5b 696e 645f  ividual_col[ind_
-0000bd30: 6e6f 6465 5f69 6473 5d20 3d20 696e 645f  node_ids] = ind_
-0000bd40: 6964 0a20 2020 2020 2020 2020 2020 2023  id.            #
-0000bd50: 206d 616b 6520 7375 7265 2074 6861 7420   make sure that 
-0000bd60: 616c 6c20 6e6f 6465 7320 7765 7265 2061  all nodes were a
-0000bd70: 7373 6967 6e65 6420 746f 2069 6e64 6976  ssigned to indiv
-0000bd80: 6964 7561 6c73 0a20 2020 2020 2020 2020  iduals.         
-0000bd90: 2020 2061 7373 6572 7420 6e70 2e61 6c6c     assert np.all
-0000bda0: 286e 6f64 6573 5f74 6162 5f69 6e64 6976  (nodes_tab_indiv
-0000bdb0: 6964 7561 6c5f 636f 6c20 3e3d 2030 292c  idual_col >= 0),
-0000bdc0: 2028 2753 6f6d 6520 6e6f 6465 7320 6e6f   ('Some nodes no
-0000bdd0: 7420 270a 2020 2020 2020 2020 2020 2020  t '.            
+0000a540: 2020 2020 2020 2020 2020 2020 2020 6c65                le
+0000a550: 6166 7369 7a65 3d6c 6561 6673 697a 652c  afsize=leafsize,
+0000a560: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000a570: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000a580: 2020 2020 290a 0a20 2020 2023 6d65 7468      )..    #meth
+0000a590: 6f64 2074 6f20 7365 7420 7468 6520 7370  od to set the sp
+0000a5a0: 6563 6965 7327 2073 7061 7469 616c 2e5f  ecies' spatial._
+0000a5b0: 4465 6e73 6974 7947 7269 6453 7461 636b  DensityGridStack
+0000a5c0: 2061 7474 7269 6275 7465 0a20 2020 2064   attribute.    d
+0000a5d0: 6566 205f 7365 745f 6465 6e73 5f67 7269  ef _set_dens_gri
+0000a5e0: 6473 2873 656c 662c 206c 616e 642c 2077  ds(self, land, w
+0000a5f0: 6964 6f77 5f77 6964 7468 203d 204e 6f6e  idow_width = Non
+0000a600: 6529 3a0a 2020 2020 2020 2020 7365 6c66  e):.        self
+0000a610: 2e5f 6465 6e73 5f67 7269 6473 203d 2073  ._dens_grids = s
+0000a620: 7074 2e5f 4465 6e73 6974 7947 7269 6453  pt._DensityGridS
+0000a630: 7461 636b 286c 616e 6420 3d20 6c61 6e64  tack(land = land
+0000a640: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0000a650: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000a660: 2020 7769 6e64 6f77 5f77 6964 7468 203d    window_width =
+0000a670: 2073 656c 662e 6465 6e73 6974 795f 6772   self.density_gr
+0000a680: 6964 5f77 696e 646f 775f 7769 6474 6829  id_window_width)
+0000a690: 0a0a 0a20 2020 2023 206d 6574 686f 6420  ...    # method 
+0000a6a0: 746f 2073 6574 2074 6865 2073 7065 6369  to set the speci
+0000a6b0: 6573 2720 6765 6e6f 6d65 732c 2063 6865  es' genomes, che
+0000a6c0: 636b 2066 6f72 2061 6465 7175 6174 6520  ck for adequate 
+0000a6d0: 6d75 7461 626c 6520 6c6f 6369 2c0a 2020  mutable loci,.  
+0000a6e0: 2020 2320 616e 6420 6669 6c6c 2074 6865    # and fill the
+0000a6f0: 2074 736b 6974 2e54 6162 6c65 436f 6c6c   tskit.TableColl
+0000a700: 6563 7469 6f6e 2773 2074 6162 6c65 730a  ection's tables.
+0000a710: 2020 2020 2320 2874 6f20 6265 2063 616c      # (to be cal
+0000a720: 6c65 6420 6166 7465 7220 7468 6520 6d6f  led after the mo
+0000a730: 6465 6c20 6861 7320 6275 726e 6564 2069  del has burned i
+0000a740: 6e29 0a20 2020 2064 6566 205f 7365 745f  n).    def _set_
+0000a750: 6765 6e6f 6d65 735f 616e 645f 7461 626c  genomes_and_tabl
+0000a760: 6573 2873 656c 662c 2062 7572 6e5f 542c  es(self, burn_T,
+0000a770: 2054 293a 0a20 2020 2020 2020 2023 2073   T):.        # s
+0000a780: 6574 2073 7461 7274 696e 6720 6765 6e6f  et starting geno
+0000a790: 6d65 2061 7272 6179 7320 746f 2061 6c6c  me arrays to all
+0000a7a0: 207a 6572 6f73 0a20 2020 2020 2020 2073   zeros.        s
+0000a7b0: 656c 662e 5f73 6574 5f6e 756c 6c5f 6765  elf._set_null_ge
+0000a7c0: 6e6f 6d65 7328 290a 0a20 2020 2020 2020  nomes()..       
+0000a7d0: 2023 2075 7365 206d 6561 6e20 6e5f 6269   # use mean n_bi
+0000a7e0: 7274 6873 2061 7420 7461 696c 2065 6e64  rths at tail end
+0000a7f0: 206f 6620 6275 726e 2d69 6e20 746f 2065   of burn-in to e
+0000a800: 7374 696d 6174 6520 6e75 6d62 6572 206f  stimate number o
+0000a810: 660a 2020 2020 2020 2020 2320 6d75 7461  f.        # muta
+0000a820: 7469 6f6e 7320 666f 7220 7468 6973 2073  tions for this s
+0000a830: 7065 6369 6573 2c20 7468 656e 2077 6172  pecies, then war
+0000a840: 6e20 7468 6520 7573 6572 2069 6620 7468  n the user if th
+0000a850: 6572 6520 6973 206f 7220 6d61 7920 6265  ere is or may be
+0000a860: 0a20 2020 2020 2020 2023 2069 6e61 6465  .        # inade
+0000a870: 7175 6174 6520 7370 6163 6520 666f 7220  quate space for 
+0000a880: 7468 6520 7061 7261 6d65 7465 7269 7a65  the parameterize
+0000a890: 6420 6d75 7461 7469 6f6e 2072 6174 6520  d mutation rate 
+0000a8a0: 2862 6563 6175 7365 206f 6e6c 790a 2020  (because only.  
+0000a8b0: 2020 2020 2020 2320 7573 696e 6720 696e        # using in
+0000a8c0: 6669 6e69 7465 2d73 6974 6573 206d 7574  finite-sites mut
+0000a8d0: 6174 696f 6e29 0a20 2020 2020 2020 2065  ation).        e
+0000a8e0: 7374 5f74 6f74 5f6d 7574 7320 3d20 5f63  st_tot_muts = _c
+0000a8f0: 616c 635f 6573 7469 6d61 7465 645f 746f  alc_estimated_to
+0000a900: 7461 6c5f 6d75 7461 7469 6f6e 7328 7365  tal_mutations(se
+0000a910: 6c66 2c20 6275 726e 5f54 2c20 5429 0a0a  lf, burn_T, T)..
+0000a920: 2020 2020 2020 2020 2320 6368 6563 6b20          # check 
+0000a930: 7768 6574 6865 7220 7468 6572 6520 6172  whether there ar
+0000a940: 6520 6164 6571 7561 7465 206d 7574 6162  e adequate mutab
+0000a950: 6c65 206c 6f63 6920 666f 7220 7468 6973  le loci for this
+0000a960: 2073 7065 6369 6573 0a20 2020 2020 2020   species.       
+0000a970: 205f 6368 6563 6b5f 6d75 7461 7469 6f6e   _check_mutation
+0000a980: 5f72 6174 6573 2873 656c 662e 6765 6e5f  _rates(self.gen_
+0000a990: 6172 6368 2c20 6573 745f 746f 745f 6d75  arch, est_tot_mu
+0000a9a0: 7473 2c20 6275 726e 5f54 2c20 5429 0a0a  ts, burn_T, T)..
+0000a9b0: 2020 2020 2020 2020 2320 6372 6561 7465          # create
+0000a9c0: 2061 6e64 2063 6f6e 6669 6775 7265 2074   and configure t
+0000a9d0: 736b 6974 2074 6162 6c65 730a 2020 2020  skit tables.    
+0000a9e0: 2020 2020 6966 2073 656c 662e 6765 6e5f      if self.gen_
+0000a9f0: 6172 6368 2e75 7365 5f74 736b 6974 3a0a  arch.use_tskit:.
+0000aa00: 2020 2020 2020 2020 2020 2020 2320 7369              # si
+0000aa10: 6d75 6c61 7465 2061 2063 6f61 6c65 7363  mulate a coalesc
+0000aa20: 656e 7420 616e 6365 7374 7279 2077 6974  ent ancestry wit
+0000aa30: 6820 6e75 6d62 6572 206f 6620 7361 6d70  h number of samp
+0000aa40: 6c65 7320 6571 7561 6c20 746f 206f 7572  les equal to our
+0000aa50: 0a20 2020 2020 2020 2020 2020 2023 2073  .            # s
+0000aa60: 7065 6369 6573 2720 6e75 6d62 6572 206f  pecies' number o
+0000aa70: 6620 6861 706c 6f69 6420 6765 6e6f 6d65  f haploid genome
+0000aa80: 7320 2869 2e65 2e20 322a 4e5f 3029 0a20  s (i.e. 2*N_0). 
+0000aa90: 2020 2020 2020 2020 2020 2023 204e 4f54             # NOT
+0000aaa0: 453a 2044 4554 483a 2032 352d 3033 2d32  E: DETH: 25-03-2
+0000aab0: 343a 206d 7370 7269 6d65 2e73 696d 756c  4: msprime.simul
+0000aac0: 6174 6520 6973 206e 6f77 0a20 2020 2020  ate is now.     
+0000aad0: 2020 2020 2020 2023 2020 2020 2020 2020         #        
+0000aae0: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+0000aaf0: 2064 6570 7265 6361 7465 6420 6c65 6761   deprecated lega
+0000ab00: 6379 2066 6e20 6672 6f6d 2076 302e 782c  cy fn from v0.x,
+0000ab10: 0a20 2020 2020 2020 2020 2020 2023 2020  .            #  
+0000ab20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000ab30: 2020 2020 2069 6e20 6661 766f 7220 6f66       in favor of
+0000ab40: 206d 7370 7269 6d65 2e73 696d 5f61 6e63   msprime.sim_anc
+0000ab50: 6573 7472 792c 0a20 2020 2020 2020 2020  estry,.         
+0000ab60: 2020 2023 2020 2020 2020 2020 2020 2020     #            
+0000ab70: 2020 2020 2020 2020 2020 2062 7574 2069             but i
+0000ab80: 7420 6973 2069 6e64 6566 696e 6974 656c  t is indefinitel
+0000ab90: 7920 7375 7070 6f72 7465 642c 0a20 2020  y supported,.   
+0000aba0: 2020 2020 2020 2020 2023 2020 2020 2020           #      
+0000abb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000abc0: 2073 6f20 6e6f 2069 6d6d 6564 6961 7465   so no immediate
+0000abd0: 206e 6565 6420 746f 2073 7769 7463 680a   need to switch.
+0000abe0: 2020 2020 2020 2020 2020 2020 7473 203d              ts =
+0000abf0: 206d 7370 7269 6d65 2e73 696d 756c 6174   msprime.simulat
+0000ac00: 6528 6c65 6e28 7365 6c66 2920 2a20 322c  e(len(self) * 2,
+0000ac10: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000ac20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000ac30: 2020 204e 653d 3130 3030 2c0a 2020 2020     Ne=1000,.    
+0000ac40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000ac50: 2020 2020 2020 2020 2020 2020 2020 6c65                le
+0000ac60: 6e67 7468 3d73 656c 662e 6765 6e5f 6172  ngth=self.gen_ar
+0000ac70: 6368 2e4c 290a 0a20 2020 2020 2020 2020  ch.L)..         
+0000ac80: 2020 2023 2074 6865 6e20 6772 6162 2074     # then grab t
+0000ac90: 6865 2073 696d 756c 6174 696f 6e27 7320  he simulation's 
+0000aca0: 5461 626c 6543 6f6c 6c65 6374 696f 6e0a  TableCollection.
+0000acb0: 2020 2020 2020 2020 2020 2020 7461 626c              tabl
+0000acc0: 6573 203d 2074 732e 6475 6d70 5f74 6162  es = ts.dump_tab
+0000acd0: 6c65 7328 290a 0a20 2020 2020 2020 2020  les()..         
+0000ace0: 2020 2023 2073 6574 2074 6865 2073 6571     # set the seq
+0000acf0: 7565 6e63 6520 6c65 6e67 7468 0a20 2020  uence length.   
+0000ad00: 2020 2020 2020 2020 2023 7461 626c 6573           #tables
+0000ad10: 2e73 6571 7565 6e63 655f 6c65 6e67 7468  .sequence_length
+0000ad20: 203d 2073 656c 662e 6765 6e5f 6172 6368   = self.gen_arch
+0000ad30: 2e4c 0a20 2020 2020 2020 2020 2020 2023  .L.            #
+0000ad40: 2063 6c65 6172 2074 6865 206d 7574 6174   clear the mutat
+0000ad50: 696f 6e73 2074 6162 6c65 0a20 2020 2020  ions table.     
+0000ad60: 2020 2020 2020 2074 6162 6c65 732e 6d75         tables.mu
+0000ad70: 7461 7469 6f6e 732e 636c 6561 7228 290a  tations.clear().
+0000ad80: 0a20 2020 2020 2020 2020 2020 2023 206c  .            # l
+0000ad90: 6f6f 7020 6f76 6572 2061 6c6c 2073 6974  oop over all sit
+0000ada0: 6573 2c20 736f 2074 6861 7420 6561 6368  es, so that each
+0000adb0: 2073 6974 6527 7320 726f 7720 676f 6573   site's row goes
+0000adc0: 2069 6e74 6f0a 2020 2020 2020 2020 2020   into.          
+0000add0: 2020 2320 7468 6520 7369 7465 7320 7461    # the sites ta
+0000ade0: 626c 6520 696e 206f 7264 6572 2028 7375  ble in order (su
+0000adf0: 6368 2074 6861 7420 312e 2920 7468 6572  ch that 1.) ther
+0000ae00: 6527 7320 6e6f 206e 6565 640a 2020 2020  e's no need.    
+0000ae10: 2020 2020 2020 2020 2320 746f 2074 7261          # to tra
+0000ae20: 636b 2068 6f77 2067 6e78 2073 6974 6573  ck how gnx sites
+0000ae30: 206d 6170 206f 6e74 6f20 7369 7465 732d   map onto sites-
+0000ae40: 7461 626c 6520 726f 7720 6964 732c 0a20  table row ids,. 
+0000ae50: 2020 2020 2020 2020 2020 2023 2061 6e64             # and
+0000ae60: 2032 2e29 2074 6865 7265 2077 696c 6c20   2.) there will 
+0000ae70: 6265 206e 6f20 6e65 6564 2074 6f20 6465  be no need to de
+0000ae80: 6475 706c 6963 6174 6520 7369 7465 7320  duplicate sites 
+0000ae90: 6c61 7465 7220 6f6e 290a 2020 2020 2020  later on).      
+0000aea0: 2020 2020 2020 666f 7220 7369 7465 2069        for site i
+0000aeb0: 6e20 7261 6e67 6528 7365 6c66 2e67 656e  n range(self.gen
+0000aec0: 5f61 7263 682e 4c29 3a0a 2020 2020 2020  _arch.L):.      
+0000aed0: 2020 2020 2020 2020 2020 2364 6574 6572            #deter
+0000aee0: 6d69 6e65 2077 6865 7468 6572 2074 6869  mine whether thi
+0000aef0: 7320 7374 6172 7465 6420 6173 2061 206e  s started as a n
+0000af00: 6575 7472 616c 206f 7220 6e6f 6e2d 6e65  eutral or non-ne
+0000af10: 7574 7261 6c20 7369 7465 0a20 2020 2020  utral site.     
+0000af20: 2020 2020 2020 2020 2020 2069 6620 7369             if si
+0000af30: 7465 2069 6e20 7365 6c66 2e67 656e 5f61  te in self.gen_a
+0000af40: 7263 682e 6e65 7574 5f6c 6f63 693a 0a20  rch.neut_loci:. 
+0000af50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000af60: 2020 206d 6574 6164 6174 613d 276e 272e     metadata='n'.
+0000af70: 656e 636f 6465 2827 6173 6369 6927 290a  encode('ascii').
+0000af80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000af90: 656c 6966 2073 6974 6520 696e 2073 656c  elif site in sel
+0000afa0: 662e 6765 6e5f 6172 6368 2e6e 6f6e 6e65  f.gen_arch.nonne
+0000afb0: 7574 5f6c 6f63 693a 0a20 2020 2020 2020  ut_loci:.       
+0000afc0: 2020 2020 2020 2020 2020 2020 206d 6574               met
+0000afd0: 6164 6174 613d 2774 272e 656e 636f 6465  adata='t'.encode
+0000afe0: 2827 6173 6369 6927 290a 2020 2020 2020  ('ascii').      
+0000aff0: 2020 2020 2020 2020 2020 2320 6164 6420            # add 
+0000b000: 7468 6520 7661 7269 616e 7427 7320 7369  the variant's si
+0000b010: 7465 2074 6f20 7468 6520 7369 7465 7320  te to the sites 
+0000b020: 7461 626c 650a 2020 2020 2020 2020 2020  table.          
+0000b030: 2020 2020 2020 7461 626c 6573 2e73 6974        tables.sit
+0000b040: 6573 2e61 6464 5f72 6f77 2870 6f73 6974  es.add_row(posit
+0000b050: 696f 6e3d 7369 7465 2c20 616e 6365 7374  ion=site, ancest
+0000b060: 7261 6c5f 7374 6174 653d 2730 272c 0a20  ral_state='0',. 
+0000b070: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000b080: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000b090: 2020 2020 6d65 7461 6461 7461 3d6d 6574      metadata=met
+0000b0a0: 6164 6174 6129 0a0a 2020 2020 2020 2020  adata)..        
+0000b0b0: 2020 2020 2320 6772 6162 2074 6865 206e      # grab the n
+0000b0c0: 6f64 6573 2066 6c61 6773 2c20 7768 6963  odes flags, whic
+0000b0d0: 6820 6172 6520 3120 666f 7220 6375 7272  h are 1 for curr
+0000b0e0: 656e 7420 6e6f 6465 732c 0a20 2020 2020  ent nodes,.     
+0000b0f0: 2020 2020 2020 2023 2030 2066 6f72 2070         # 0 for p
+0000b100: 6173 7420 6e6f 6465 732c 2069 6e74 6f20  ast nodes, into 
+0000b110: 7477 6f20 7365 7061 7261 7465 206f 626a  two separate obj
+0000b120: 6563 7473 0a20 2020 2020 2020 2020 2020  ects.           
+0000b130: 2063 7572 7265 6e74 5f6e 6f64 6573 203d   current_nodes =
+0000b140: 205b 2a6e 702e 7768 6572 6528 7461 626c   [*np.where(tabl
+0000b150: 6573 2e6e 6f64 6573 2e66 6c61 6773 203d  es.nodes.flags =
+0000b160: 3d20 3129 5b30 5d5d 0a20 2020 2020 2020  = 1)[0]].       
+0000b170: 2020 2020 2023 2072 6576 6572 7365 2c20       # reverse, 
+0000b180: 736f 2074 6861 7420 4920 6361 6e20 706f  so that I can po
+0000b190: 7020 6e6f 6465 7320 6f66 6620 7468 6520  p nodes off the 
+0000b1a0: 2766 726f 6e74 270a 2020 2020 2020 2020  'front'.        
+0000b1b0: 2020 2020 6375 7272 656e 745f 6e6f 6465      current_node
+0000b1c0: 7320 3d20 6375 7272 656e 745f 6e6f 6465  s = current_node
+0000b1d0: 735b 3a3a 2d31 5d0a 2020 2020 2020 2020  s[::-1].        
+0000b1e0: 2020 2020 7061 7374 5f6e 6f64 6573 203d      past_nodes =
+0000b1f0: 205b 2a6e 702e 7768 6572 6528 7461 626c   [*np.where(tabl
+0000b200: 6573 2e6e 6f64 6573 2e66 6c61 6773 2021  es.nodes.flags !
+0000b210: 3d20 3129 5b30 5d5d 0a20 2020 2020 2020  = 1)[0]].       
+0000b220: 2020 2020 2023 2063 7265 6174 6520 616e       # create an
+0000b230: 2065 6d70 7479 206c 6973 742c 2074 6f20   empty list, to 
+0000b240: 6669 6c6c 2075 7020 7468 6520 696e 6469  fill up the indi
+0000b250: 7669 6475 616c 2069 6473 2066 6f72 2065  vidual ids for e
+0000b260: 6163 6820 6e6f 6465 0a20 2020 2020 2020  ach node.       
+0000b270: 2020 2020 2023 204e 4f54 453a 2073 6574       # NOTE: set
+0000b280: 7469 6e67 2074 6f20 6120 7665 6374 6f72  ting to a vector
+0000b290: 206f 6620 2d31 2069 6e69 7469 616c 6c79   of -1 initially
+0000b2a0: 2c20 746f 2065 6173 696c 7920 6368 6563  , to easily chec
+0000b2b0: 6b20 7468 6174 0a20 2020 2020 2020 2020  k that.         
+0000b2c0: 2020 2023 2061 6c6c 2076 616c 7565 7320     # all values 
+0000b2d0: 6861 7665 2062 6565 6e20 6173 7369 676e  have been assign
+0000b2e0: 6564 2061 7420 7468 6520 656e 6420 6279  ed at the end by
+0000b2f0: 2061 736b 696e 6720 6966 2061 6c6c 203e   asking if all >
+0000b300: 3d20 3020 0a20 2020 2020 2020 2020 2020  = 0 .           
+0000b310: 206e 6f64 6573 5f74 6162 5f69 6e64 6976   nodes_tab_indiv
+0000b320: 6964 7561 6c5f 636f 6c20 3d20 6e70 2e69  idual_col = np.i
+0000b330: 6e74 3332 286e 702e 6f6e 6573 286c 656e  nt32(np.ones(len
+0000b340: 280a 2020 2020 2020 2020 2020 2020 2020  (.              
+0000b350: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000b360: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000b370: 2020 2020 2020 2020 2020 7461 626c 6573            tables
+0000b380: 2e6e 6f64 6573 2e66 6c61 6773 2929 2a2d  .nodes.flags))*-
+0000b390: 3129 0a0a 2020 2020 2020 2020 2020 2020  1)..            
+0000b3a0: 2320 4e4f 5445 3a20 7468 6572 6520 6172  # NOTE: there ar
+0000b3b0: 6520 6e6f 2072 6571 7569 7265 6d65 6e74  e no requirement
+0000b3c0: 730a 2020 2020 2020 2020 2020 2020 2320  s.            # 
+0000b3d0: 2020 2020 2020 6f72 2072 6573 7472 6963        or restric
+0000b3e0: 7469 6f6e 7320 666f 7220 7468 6520 696e  tions for the in
+0000b3f0: 6469 7669 6475 616c 730a 2020 2020 2020  dividuals.      
+0000b400: 2020 2020 2020 2320 616e 6420 6e6f 6465        # and node
+0000b410: 7320 7461 626c 6573 2028 652e 672e 206f  s tables (e.g. o
+0000b420: 7264 6572 2c20 6574 632e 292c 2061 6e64  rder, etc.), and
+0000b430: 2074 6875 7320 7468 6520 7461 626c 6573   thus the tables
+0000b440: 2720 6f72 6465 720a 2020 2020 2020 2020  ' order.        
+0000b450: 2020 2020 2320 6973 206e 6f74 2061 6666      # is not aff
+0000b460: 6563 7465 6420 6279 2074 6865 2054 6162  ected by the Tab
+0000b470: 6c65 436f 6c6c 6563 7469 6f6e 2e73 696d  leCollection.sim
+0000b480: 706c 6966 7920 616c 676f 7269 7468 6d2e  plify algorithm.
+0000b490: 0a20 2020 2020 2020 2020 2020 2023 2053  .            # S
+0000b4a0: 6f2c 2049 2063 6f75 6c64 2061 6464 2065  o, I could add e
+0000b4b0: 6974 6865 7220 6375 7272 656e 740a 2020  ither current.  
+0000b4c0: 2020 2020 2020 2020 2020 2320 6f72 2070            # or p
+0000b4d0: 6173 7420 696e 6469 7669 6475 616c 7320  ast individuals 
+0000b4e0: 616e 6420 6e6f 6465 7320 6669 7273 743b  and nodes first;
+0000b4f0: 0a20 2020 2020 2020 2020 2020 2023 2063  .            # c
+0000b500: 686f 6f73 696e 6720 746f 2061 6464 2070  hoosing to add p
+0000b510: 6173 7420 6669 7273 742c 2073 6f20 7468  ast first, so th
+0000b520: 6174 2061 6c6c 2027 7265 616c 2720 696e  at all 'real' in
+0000b530: 6469 7669 6475 616c 730a 2020 2020 2020  dividuals.      
+0000b540: 2020 2020 2020 2320 6672 6f6d 2074 6865        # from the
+0000b550: 2073 7461 7274 206f 6620 7468 6520 6765   start of the ge
+0000b560: 6f6e 6f6d 6963 7320 7369 6d75 6c61 7469  onomics simulati
+0000b570: 6f6e 2066 6f72 7761 7264 2077 696c 6c0a  on forward will.
+0000b580: 2020 2020 2020 2020 2020 2020 2320 7769              # wi
+0000b590: 6e64 2075 7020 6861 7669 6e67 2074 6865  nd up having the
+0000b5a0: 6972 2069 6e64 6976 6964 7561 6c73 2061  ir individuals a
+0000b5b0: 6e64 206e 6f64 6573 2072 6f77 7320 696e  nd nodes rows in
+0000b5c0: 2061 2073 696e 676c 650a 2020 2020 2020   a single.      
+0000b5d0: 2020 2020 2020 2320 626c 6f63 6b20 6174        # block at
+0000b5e0: 2074 6865 2074 6162 6c65 7327 2062 6f74   the tables' bot
+0000b5f0: 746f 6d73 0a0a 2020 2020 2020 2020 2020  toms..          
+0000b600: 2020 2320 6164 6420 616e 2069 6e64 6976    # add an indiv
+0000b610: 6964 7561 6c20 746f 2074 6865 2069 6e64  idual to the ind
+0000b620: 6976 6964 7561 6c73 2074 6162 6c65 2066  ividuals table f
+0000b630: 6f72 0a20 2020 2020 2020 2020 2020 2023  or.            #
+0000b640: 2065 6163 6820 636f 616c 6573 6365 6e74   each coalescent
+0000b650: 2d73 696d 756c 6174 6564 206e 6f64 6520  -simulated node 
+0000b660: 6265 666f 7265 2074 6865 2063 7572 7265  before the curre
+0000b670: 6e74 2074 696d 650a 2020 2020 2020 2020  nt time.        
+0000b680: 2020 2020 2320 4e4f 5445 3a20 6164 6469      # NOTE: addi
+0000b690: 6e67 206e 6f20 6d65 7461 6461 7461 2c20  ng no metadata, 
+0000b6a0: 616e 6420 6e6f 206c 6f63 6174 696f 6e2c  and no location,
+0000b6b0: 0a20 2020 2020 2020 2020 2020 2023 2020  .            #  
+0000b6c0: 2020 2020 2074 6f20 696e 6469 6361 7465       to indicate
+0000b6d0: 2074 6861 7420 7468 6973 2069 7320 6120   that this is a 
+0000b6e0: 2766 616b 6527 2069 6e64 6976 6964 7561  'fake' individua
+0000b6f0: 6c2c 0a20 2020 2020 2020 2020 2020 2023  l,.            #
+0000b700: 2020 2020 2020 2069 6e76 656e 7465 6420         invented 
+0000b710: 6a75 7374 2074 6f20 6d61 7463 6820 7570  just to match up
+0000b720: 2074 6f20 7468 6520 6e6f 6465 730a 2020   to the nodes.  
+0000b730: 2020 2020 2020 2020 2020 2320 2020 2020            #     
+0000b740: 2020 7369 6d75 6c61 7465 6420 666f 7220    simulated for 
+0000b750: 7468 6520 7374 6172 7469 6e67 2070 6f70  the starting pop
+0000b760: 756c 6174 696f 6e0a 2020 2020 2020 2020  ulation.        
+0000b770: 2020 2020 666f 7220 6e6f 6465 2069 6e20      for node in 
+0000b780: 7061 7374 5f6e 6f64 6573 3a0a 2020 2020  past_nodes:.    
+0000b790: 2020 2020 2020 2020 2020 2020 696e 645f              ind_
+0000b7a0: 6964 203d 2074 6162 6c65 732e 696e 6469  id = tables.indi
+0000b7b0: 7669 6475 616c 732e 6164 645f 726f 7728  viduals.add_row(
+0000b7c0: 666c 6167 733d 3029 0a20 2020 2020 2020  flags=0).       
+0000b7d0: 2020 2020 2020 2020 2023 2073 746f 7265           # store
+0000b7e0: 2069 7473 2069 6e64 6976 6964 2069 6420   its individ id 
+0000b7f0: 696e 2074 6865 206e 6f64 6573 2074 6162  in the nodes tab
+0000b800: 6c65 2773 2069 6e64 6976 6964 7561 6c73  le's individuals
+0000b810: 2063 6f6c 756d 6e0a 2020 2020 2020 2020   column.        
+0000b820: 2020 2020 2020 2020 6e6f 6465 735f 7461          nodes_ta
+0000b830: 625f 696e 6469 7669 6475 616c 5f63 6f6c  b_individual_col
+0000b840: 5b6e 6f64 655d 203d 2069 6e64 5f69 640a  [node] = ind_id.
+0000b850: 0a20 2020 2020 2020 2020 2020 2023 2063  .            # c
+0000b860: 7265 6174 6520 616e 6420 6164 6420 746f  reate and add to
+0000b870: 2074 6865 2069 6e64 6976 6964 7561 6c73   the individuals
+0000b880: 2074 6162 6c65 2061 206e 6577 2072 6f77   table a new row
+0000b890: 2066 6f72 0a20 2020 2020 2020 2020 2020   for.           
+0000b8a0: 2023 2065 6163 6820 7265 616c 2069 6e64   # each real ind
+0000b8b0: 6976 6964 7561 6c0a 2020 2020 2020 2020  ividual.        
+0000b8c0: 2020 2020 666f 7220 696e 6420 696e 2073      for ind in s
+0000b8d0: 656c 662e 7661 6c75 6573 2829 3a0a 2020  elf.values():.  
+0000b8e0: 2020 2020 2020 2020 2020 2020 2020 2320                # 
+0000b8f0: 6765 7420 7468 6520 276c 6f63 6174 696f  get the 'locatio
+0000b900: 6e27 2063 6f6c 756d 6e20 696e 666f 2c20  n' column info, 
+0000b910: 7768 6963 6820 7769 6c6c 2069 6e63 6c75  which will inclu
+0000b920: 6465 2074 6865 2078 2061 6e64 2079 0a20  de the x and y. 
+0000b930: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+0000b940: 2070 6f73 6974 696f 6e73 206f 6620 616e   positions of an
+0000b950: 2069 6e64 6976 6964 7561 6c2c 2061 7320   individual, as 
+0000b960: 7765 6c6c 2061 7320 7468 6520 696e 6469  well as the indi
+0000b970: 7669 6475 616c 2773 0a20 2020 2020 2020  vidual's.       
+0000b980: 2020 2020 2020 2020 2023 2070 6865 6e6f           # pheno
+0000b990: 7479 7065 7320 616e 6420 6669 746e 6573  types and fitnes
+0000b9a0: 732c 2069 6620 7472 6169 7473 2061 7265  s, if traits are
+0000b9b0: 2062 6569 6e67 2075 7365 640a 2020 2020   being used.    
+0000b9c0: 2020 2020 2020 2020 2020 2020 6c6f 6320              loc 
+0000b9d0: 3d20 5b69 6e64 2e78 2c20 696e 642e 795d  = [ind.x, ind.y]
+0000b9e0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000b9f0: 2069 6620 7365 6c66 2e67 656e 5f61 7263   if self.gen_arc
+0000ba00: 682e 7472 6169 7473 2069 7320 6e6f 7420  h.traits is not 
+0000ba10: 4e6f 6e65 3a0a 2020 2020 2020 2020 2020  None:.          
+0000ba20: 2020 2020 2020 2020 2020 6c6f 6320 3d20            loc = 
+0000ba30: 6c6f 6320 2b20 696e 642e 7a20 2b20 5b69  loc + ind.z + [i
+0000ba40: 6e64 2e66 6974 5d0a 2020 2020 2020 2020  nd.fit].        
+0000ba50: 2020 2020 2020 2020 2320 6164 6420 6120          # add a 
+0000ba60: 6e65 7720 726f 7720 746f 2074 6865 2069  new row to the i
+0000ba70: 6e64 6976 6964 7561 6c73 2074 6162 6c65  ndividuals table
+0000ba80: 2c20 7365 7474 696e 6720 7468 6520 6c6f  , setting the lo
+0000ba90: 6361 7469 6f6e 0a20 2020 2020 2020 2020  cation.         
+0000baa0: 2020 2020 2020 2023 2063 6f6c 756d 6e27         # column'
+0000bab0: 7320 7661 6c75 6520 746f 206c 6f63 0a20  s value to loc. 
+0000bac0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+0000bad0: 204e 4f54 453a 2075 7369 6e67 2074 6865   NOTE: using the
+0000bae0: 206d 6574 6164 6174 6120 636f 6c75 6d6e   metadata column
+0000baf0: 2074 6f20 7374 6f72 6520 7468 6520 676e   to store the gn
+0000bb00: 780a 2020 2020 2020 2020 2020 2020 2020  x.              
+0000bb10: 2020 2320 696e 6469 7669 6475 616c 2069    # individual i
+0000bb20: 6478 2c20 666f 7220 6c61 7465 7220 6d61  dx, for later ma
+0000bb30: 7463 6869 6e67 2074 6f20 7570 6461 7465  tching to update
+0000bb40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000bb50: 2023 2049 6e64 6976 6964 7561 6c2e 5f69   # Individual._i
+0000bb60: 6e64 6976 6964 7561 6c73 5f74 6162 5f69  ndividuals_tab_i
+0000bb70: 6420 6166 7465 7220 7473 6b69 7427 7320  d after tskit's 
+0000bb80: 7369 6d70 6c69 6679 0a20 2020 2020 2020  simplify.       
+0000bb90: 2020 2020 2020 2020 2023 2061 6c67 6f72           # algor
+0000bba0: 6974 686d 2066 696c 7465 7273 2069 6e64  ithm filters ind
+0000bbb0: 6976 6964 7561 6c73 0a20 2020 2020 2020  ividuals.       
+0000bbc0: 2020 2020 2020 2020 2069 6e64 5f69 6420           ind_id 
+0000bbd0: 3d20 7461 626c 6573 2e69 6e64 6976 6964  = tables.individ
+0000bbe0: 7561 6c73 2e61 6464 5f72 6f77 2866 6c61  uals.add_row(fla
+0000bbf0: 6773 3d31 2c20 6c6f 6361 7469 6f6e 3d6c  gs=1, location=l
+0000bc00: 6f63 2c0a 2020 2020 2020 2020 2020 2020  oc,.            
+0000bc10: 2020 2020 2020 2020 6d65 7461 6461 7461          metadata
+0000bc20: 3d69 6e64 2e69 6478 2e74 6f5f 6279 7465  =ind.idx.to_byte
+0000bc30: 7328 6c65 6e67 7468 3d34 2c20 6279 7465  s(length=4, byte
+0000bc40: 6f72 6465 723d 276c 6974 746c 6527 2929  order='little'))
+0000bc50: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000bc60: 2069 6e64 2e5f 696e 6469 7669 6475 616c   ind._individual
+0000bc70: 735f 7461 625f 6964 203d 2069 6e64 5f69  s_tab_id = ind_i
+0000bc80: 640a 0a20 2020 2020 2020 2020 2020 2020  d..             
+0000bc90: 2020 2023 2061 7373 6967 6e20 7468 6520     # assign the 
+0000bca0: 696e 6469 7669 6475 616c 2032 2072 616e  individual 2 ran
+0000bcb0: 646f 6d6c 7920 6368 6f73 656e 206e 6f64  domly chosen nod
+0000bcc0: 6573 2066 726f 6d0a 2020 2020 2020 2020  es from.        
+0000bcd0: 2020 2020 2020 2020 2320 7468 6520 6375          # the cu
+0000bce0: 7272 656e 7420 7469 6d65 2073 7465 702c  rrent time step,
+0000bcf0: 2061 6e64 2061 7373 6f63 6961 7465 2074   and associate t
+0000bd00: 6865 2069 6e64 6976 6964 7561 6c27 730a  he individual's.
+0000bd10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000bd20: 2320 5f69 6e64 6976 6964 7561 6c73 5f74  # _individuals_t
+0000bd30: 6162 5f69 6420 7769 7468 2074 686f 7365  ab_id with those
+0000bd40: 2032 206e 6f64 6573 2069 6e20 736f 6d65   2 nodes in some
+0000bd50: 2064 6174 610a 2020 2020 2020 2020 2020   data.          
+0000bd60: 2020 2020 2020 2320 7374 7275 6374 7572        # structur
+0000bd70: 6520 746f 2063 6f6c 6c65 6374 2074 6869  e to collect thi
+0000bd80: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+0000bd90: 2020 696e 645f 6e6f 6465 5f69 6473 203d    ind_node_ids =
+0000bda0: 5b63 7572 7265 6e74 5f6e 6f64 6573 2e70  [current_nodes.p
+0000bdb0: 6f70 2829 2066 6f72 205f 2069 6e20 7261  op() for _ in ra
+0000bdc0: 6e67 6528 0a20 2020 2020 2020 2020 2020  nge(.           
+0000bdd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 0000bde0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 0000bdf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000be00: 2020 2020 2020 2020 2020 2020 2020 2027                 '
-0000be10: 6769 7665 6e20 696e 6469 7669 6473 2729  given individs')
-0000be20: 0a0a 2020 2020 2020 2020 2020 2020 6e6f  ..            no
-0000be30: 6465 735f 636f 6c73 203d 2074 6162 6c65  des_cols = table
-0000be40: 732e 6e6f 6465 732e 6173 6469 6374 2829  s.nodes.asdict()
-0000be50: 0a20 2020 2020 2020 2020 2020 2023 2075  .            # u
-0000be60: 7365 2074 6865 206e 6f64 655f 7461 625f  se the node_tab_
-0000be70: 696e 6469 7669 6475 616c 5f63 6f6c 2064  individual_col d
-0000be80: 6174 6120 7374 7275 6374 7572 6520 746f  ata structure to
-0000be90: 2072 6573 6574 0a20 2020 2020 2020 2020   reset.         
-0000bea0: 2020 2023 2074 6865 2069 6e64 6976 6964     # the individ
-0000beb0: 7561 6c73 2063 6f6c 756d 6e20 696e 2074  uals column in t
-0000bec0: 6865 206e 6f64 6573 2074 6162 6c65 0a20  he nodes table. 
-0000bed0: 2020 2020 2020 2020 2020 206e 6f64 6573             nodes
-0000bee0: 5f63 6f6c 735b 2769 6e64 6976 6964 7561  _cols['individua
-0000bef0: 6c27 5d5b 3a5d 203d 206e 6f64 6573 5f74  l'][:] = nodes_t
-0000bf00: 6162 5f69 6e64 6976 6964 7561 6c5f 636f  ab_individual_co
-0000bf10: 6c0a 2020 2020 2020 2020 2020 2020 2320  l.            # 
-0000bf20: 696e 6372 656d 656e 7420 616c 6c20 6269  increment all bi
-0000bf30: 7274 6820 7469 6d65 7320 6279 2031 2c20  rth times by 1, 
-0000bf40: 736f 2074 6861 7420 616c 6c20 696e 6469  so that all indi
-0000bf50: 7669 6473 2061 7265 206d 6172 6b65 640a  vids are marked.
-0000bf60: 2020 2020 2020 2020 2020 2020 2320 6173              # as
-0000bf70: 2068 6176 696e 6720 6265 656e 2062 6f72   having been bor
-0000bf80: 6e20 6265 666f 7265 2074 6865 2073 7461  n before the sta
-0000bf90: 7274 206f 6620 7468 6520 6d6f 6465 6c27  rt of the model'
-0000bfa0: 7320 6d61 696e 2070 6861 7365 0a20 2020  s main phase.   
-0000bfb0: 2020 2020 2020 2020 206e 6f64 6573 5f63           nodes_c
-0000bfc0: 6f6c 735b 2774 696d 6527 5d20 2b3d 2031  ols['time'] += 1
-0000bfd0: 0a20 2020 2020 2020 2020 2020 2074 6162  .            tab
-0000bfe0: 6c65 732e 6e6f 6465 732e 7365 745f 636f  les.nodes.set_co
-0000bff0: 6c75 6d6e 7328 2a2a 6e6f 6465 735f 636f  lumns(**nodes_co
-0000c000: 6c73 290a 0a20 2020 2020 2020 2023 206f  ls)..        # o
-0000c010: 7468 6572 7769 7365 2c20 6966 2074 736b  therwise, if tsk
-0000c020: 6974 2069 736e 2774 2074 6f20 6265 2075  it isn't to be u
-0000c030: 7365 642c 2074 6865 6e20 7365 7420 7461  sed, then set ta
-0000c040: 626c 6573 2074 6f20 4e6f 6e65 0a20 2020  bles to None.   
-0000c050: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
-0000c060: 2020 2020 2020 2074 6162 6c65 7320 3d20         tables = 
-0000c070: 4e6f 6e65 0a0a 2020 2020 2020 2020 2320  None..        # 
-0000c080: 6164 6420 7375 6666 6963 6965 6e74 206d  add sufficient m
-0000c090: 7574 6174 696f 6e73 2c20 2861 7420 6f6e  utations, (at on
-0000c0a0: 6c79 2063 7572 7265 6e74 206e 6f64 6573  ly current nodes
-0000c0b0: 2c20 6973 2075 7369 6e67 2074 736b 6974  , is using tskit
-0000c0c0: 292c 0a20 2020 2020 2020 2023 2074 6f20  ),.        # to 
-0000c0d0: 7072 6f64 7563 6520 7468 6520 7374 6172  produce the star
-0000c0e0: 7469 6e67 2031 2d61 6c6c 656c 6520 6672  ting 1-allele fr
-0000c0f0: 6571 7565 6e63 6965 730a 2020 2020 2020  equencies.      
-0000c100: 2020 2320 7061 7261 6d65 7465 7269 7a65    # parameterize
-0000c110: 6420 666f 7220 7468 6973 2073 7065 6369  d for this speci
-0000c120: 6573 0a20 2020 2020 2020 205f 6d61 6b65  es.        _make
-0000c130: 5f73 7461 7274 696e 675f 6d75 7461 7469  _starting_mutati
-0000c140: 6f6e 7328 7365 6c66 2c20 7461 626c 6573  ons(self, tables
-0000c150: 290a 0a20 2020 2020 2020 2023 2054 4f44  )..        # TOD
-0000c160: 4f3a 2041 4444 2050 524f 5645 4e41 4e43  O: ADD PROVENANC
-0000c170: 4553 2052 4f57 210a 0a20 2020 2020 2020  ES ROW!..       
-0000c180: 2023 2061 7373 6967 6e20 6173 2074 6865   # assign as the
-0000c190: 2073 7065 6369 6573 2720 5461 626c 6543   species' TableC
-0000c1a0: 6f6c 6c65 6374 696f 6e0a 2020 2020 2020  ollection.      
-0000c1b0: 2020 7365 6c66 2e5f 7463 203d 2074 6162    self._tc = tab
-0000c1c0: 6c65 730a 0a20 2020 2020 2020 2072 6574  les..        ret
-0000c1d0: 7572 6e0a 0a0a 2020 2020 2320 6d65 7468  urn...    # meth
-0000c1e0: 6f64 2074 6f20 736f 7274 2061 6e64 2073  od to sort and s
-0000c1f0: 696d 706c 6966 7920 7468 6520 5461 626c  implify the Tabl
-0000c200: 6543 6f6c 6c65 6374 696f 6e2c 0a20 2020  eCollection,.   
-0000c210: 2023 2061 6e64 2075 7064 6174 6520 496e   # and update In
-0000c220: 6469 7669 6475 616c 7327 206e 6f64 6520  dividuals' node 
-0000c230: 4944 730a 2020 2020 2320 4e4f 5445 3a20  IDs.    # NOTE: 
-0000c240: 7261 7468 6572 2074 6861 6e20 6465 6475  rather than dedu
-0000c250: 706c 6963 6174 6520 7369 7465 7320 2861  plicate sites (a
-0000c260: 6e64 2073 6f6d 6574 696d 6573 0a20 2020  nd sometimes.   
-0000c270: 2023 2063 6f6d 7075 7465 7220 7061 7265   # computer pare
-0000c280: 6e74 2073 6974 6573 292c 2061 7320 7265  nt sites), as re
-0000c290: 636f 6d6d 656e 6465 6420 6279 2074 736b  commended by tsk
-0000c2a0: 6974 2773 2064 6f63 732c 0a20 2020 2023  it's docs,.    #
-0000c2b0: 2077 6520 6f6e 6c79 2068 6176 6520 746f   we only have to
-0000c2c0: 2073 6f72 7420 616e 6420 7369 6d70 6c69   sort and simpli
-0000c2d0: 6679 2068 6572 652c 2062 6563 6175 7365  fy here, because
-0000c2e0: 2077 6520 6861 7665 0a20 2020 2023 2063   we have.    # c
-0000c2f0: 686f 7365 6e20 746f 206a 7573 7420 6164  hosen to just ad
-0000c300: 6420 616c 6c20 7369 6d75 6c61 7465 6420  d all simulated 
-0000c310: 7369 7465 7320 746f 2074 6865 2073 6974  sites to the sit
-0000c320: 6573 2074 6162 6c65 2028 696e 0a20 2020  es table (in.   
-0000c330: 2023 2053 7065 6369 6573 2e5f 7365 745f   # Species._set_
-0000c340: 7461 626c 655f 636f 6c6c 6563 7469 6f6e  table_collection
-0000c350: 2920 6174 2074 6865 2073 696d 756c 6174  ) at the simulat
-0000c360: 696f 6e27 7320 6f75 7473 6574 2c0a 2020  ion's outset,.  
-0000c370: 2020 2320 616e 6420 616c 736f 2074 6f20    # and also to 
-0000c380: 6f6e 6c79 2061 6c6c 6f77 2061 6e20 696e  only allow an in
-0000c390: 6669 6e69 7465 2d73 6974 6573 206d 6f64  finite-sites mod
-0000c3a0: 656c 0a20 2020 2023 2028 464f 5220 4e4f  el.    # (FOR NO
-0000c3b0: 572c 2061 6e79 686f 772e 2e2e 2063 6f75  W, anyhow... cou
-0000c3c0: 6c64 2072 656c 6178 2074 6865 7365 2063  ld relax these c
-0000c3d0: 6f6e 7374 7261 696e 7473 206c 6174 6572  onstraints later
-0000c3e0: 2c20 6966 2064 6573 6972 6564 2c0a 2020  , if desired,.  
-0000c3f0: 2020 2320 696e 2077 6869 6368 2063 6173    # in which cas
-0000c400: 6520 776f 756c 6420 6e65 6564 2074 6f20  e would need to 
-0000c410: 7265 7661 6d70 2074 6869 7320 6170 7072  revamp this appr
-0000c420: 6f61 6368 290a 2020 2020 6465 6620 5f73  oach).    def _s
-0000c430: 6f72 745f 7369 6d70 6c69 6679 5f74 6162  ort_simplify_tab
-0000c440: 6c65 5f63 6f6c 6c65 6374 696f 6e28 7365  le_collection(se
-0000c450: 6c66 2c20 6368 6563 6b5f 6e6f 6465 733d  lf, check_nodes=
-0000c460: 4661 6c73 652c 0a20 2020 2020 2020 2020  False,.         
-0000c470: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000c480: 2020 2020 2020 2020 2020 2020 2020 2063                 c
-0000c490: 6865 636b 5f69 6e64 6976 6964 7561 6c73  heck_individuals
-0000c4a0: 3d46 616c 7365 293a 0a20 2020 2020 2020  =False):.       
-0000c4b0: 2023 2073 6f72 7420 7468 6520 5461 626c   # sort the Tabl
-0000c4c0: 6543 6f6c 6c65 6374 696f 6e0a 2020 2020  eCollection.    
-0000c4d0: 2020 2020 7365 6c66 2e5f 7463 2e73 6f72      self._tc.sor
-0000c4e0: 7428 290a 2020 2020 2020 2020 2320 6765  t().        # ge
-0000c4f0: 7420 616e 2061 7272 6179 206f 6620 616c  t an array of al
-0000c500: 6c20 7468 6520 6375 7272 656e 7420 696e  l the current in
-0000c510: 6469 7669 6475 616c 7327 206e 6f64 6573  dividuals' nodes
-0000c520: 2c0a 2020 2020 2020 2020 2320 7468 6520  ,.        # the 
-0000c530: 6e6f 6465 7320 666f 7220 7768 6963 6820  nodes for which 
-0000c540: 7468 6520 7461 626c 6573 2077 696c 6c20  the tables will 
-0000c550: 6265 2073 696d 706c 6966 6965 640a 2020  be simplified.  
-0000c560: 2020 2020 2020 6375 7272 5f6e 6f64 6573        curr_nodes
-0000c570: 203d 2073 656c 662e 5f67 6574 5f6e 6f64   = self._get_nod
-0000c580: 6573 2829 0a0a 2020 2020 2020 2020 2320  es()..        # 
-0000c590: 7275 6e20 636f 6465 206e 6563 6573 7361  run code necessa
-0000c5a0: 7279 2066 6f72 2063 6865 636b 696e 6720  ry for checking 
-0000c5b0: 7468 6174 2069 6e64 6976 6964 7561 6c73  that individuals
-0000c5c0: 2720 7461 626c 6520 6964 7320 6172 650a  ' table ids are.
-0000c5d0: 2020 2020 2020 2020 2320 636f 7272 6563          # correc
-0000c5e0: 746c 7920 6173 7369 676e 6564 2c20 6966  tly assigned, if
-0000c5f0: 2072 6571 7565 7374 6564 0a20 2020 2020   requested.     
-0000c600: 2020 2069 6620 6368 6563 6b5f 696e 6469     if check_indi
-0000c610: 7669 6475 616c 733a 0a20 2020 2020 2020  viduals:.       
-0000c620: 2023 204e 4f54 453a 2074 6869 7320 6368   # NOTE: this ch
-0000c630: 6563 6b20 7769 6c6c 206f 6e6c 7920 776f  eck will only wo
-0000c640: 726b 2066 6f72 206d 6f64 656c 7320 7468  rk for models th
-0000c650: 6174 2068 6176 6520 6174 206c 6561 7374  at have at least
-0000c660: 206f 6e65 0a20 2020 2020 2020 2023 2074   one.        # t
-0000c670: 7261 6974 2c20 6265 6361 7573 6520 6f6e  rait, because on
-0000c680: 6c79 2074 686f 7365 206d 6f64 656c 7320  ly those models 
-0000c690: 7769 6c6c 2068 6176 6520 6174 206c 6561  will have at lea
-0000c6a0: 7374 2033 2076 616c 7565 7320 696e 2074  st 3 values in t
-0000c6b0: 6865 200a 2020 2020 2020 2020 2320 6c6f  he .        # lo
-0000c6c0: 6361 7469 6f6e 2063 6f6c 756d 6e27 7320  cation column's 
-0000c6d0: 7661 6c75 6520 2861 6e64 206f 6e6c 7920  value (and only 
-0000c6e0: 7468 6520 7661 6c75 6573 2061 6674 6572  the values after
-0000c6f0: 2074 6865 2066 6972 7374 2032 2077 696c   the first 2 wil
-0000c700: 6c0a 2020 2020 2020 2020 2320 6265 2074  l.        # be t
-0000c710: 6865 2073 616d 6520 6163 726f 7373 2061  he same across a
-0000c720: 6e20 696e 6469 7669 6475 616c 2773 206c  n individual's l
-0000c730: 6966 6574 696d 652c 2073 696e 6365 2074  ifetime, since t
-0000c740: 6865 7920 6d6f 7665 2070 6f73 6974 696f  hey move positio
-0000c750: 6e0a 2020 2020 2020 2020 2320 6672 6571  n.        # freq
-0000c760: 7565 6e74 6c79 2062 7574 2064 6f20 6e6f  uently but do no
-0000c770: 7420 6361 686e 6765 2070 6865 6e6f 7479  t cahnge phenoty
-0000c780: 7065 2061 7420 616c 6c29 0a0a 2020 2020  pe at all)..    
-0000c790: 2020 2020 2020 2020 6d65 7461 5f62 3420          meta_b4 
-0000c7a0: 3d20 7365 6c66 2e5f 7463 2e69 6e64 6976  = self._tc.indiv
-0000c7b0: 6964 7561 6c73 2e6d 6574 6164 6174 610a  iduals.metadata.
-0000c7c0: 2020 2020 2020 2020 2020 2020 6c6f 6320              loc 
-0000c7d0: 3d20 7365 6c66 2e5f 7463 2e69 6e64 6976  = self._tc.indiv
-0000c7e0: 6964 7561 6c73 2e6c 6f63 6174 696f 6e0a  iduals.location.
-0000c7f0: 2020 2020 2020 2020 2020 2020 6c6f 636f              loco
-0000c800: 6666 203d 2073 656c 662e 5f74 632e 696e  ff = self._tc.in
-0000c810: 6469 7669 6475 616c 732e 6c6f 6361 7469  dividuals.locati
-0000c820: 6f6e 5f6f 6666 7365 740a 2020 2020 2020  on_offset.      
-0000c830: 2020 2020 2020 6c6f 635f 6234 5f62 795f        loc_b4_by_
-0000c840: 7461 626c 655f 6964 7320 3d20 7b0a 2020  table_ids = {.  
-0000c850: 2020 2020 2020 2020 2020 2020 2020 7461                ta
-0000c860: 626c 655f 6964 3a20 286c 6f63 5b6c 6f63  ble_id: (loc[loc
-0000c870: 6f66 665b 7461 626c 655f 6964 5d3a 206c  off[table_id]: l
-0000c880: 6f63 6f66 665b 0a20 2020 2020 2020 2020  ocoff[.         
-0000c890: 2020 2020 2020 2020 2020 2074 6162 6c65             table
-0000c8a0: 5f69 642b 315d 5d29 2066 6f72 2074 6162  _id+1]]) for tab
-0000c8b0: 6c65 5f69 6420 696e 2072 616e 6765 286c  le_id in range(l
-0000c8c0: 656e 280a 2020 2020 2020 2020 2020 2020  en(.            
-0000c8d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000c8e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000c8f0: 2020 2020 7365 6c66 2e5f 7463 2e69 6e64      self._tc.ind
-0000c900: 6976 6964 7561 6c73 2e66 6c61 6773 2929  ividuals.flags))
-0000c910: 7d0a 2020 2020 2020 2020 2020 2020 7a30  }.            z0
-0000c920: 5f62 3420 3d20 7b69 6e64 2e69 6478 3a20  _b4 = {ind.idx: 
-0000c930: 6c6f 635f 6234 5f62 795f 7461 626c 655f  loc_b4_by_table_
-0000c940: 6964 735b 0a20 2020 2020 2020 2020 2020  ids[.           
-0000c950: 2020 2020 2020 2020 2020 2020 2069 6e64               ind
-0000c960: 2e5f 696e 6469 7669 6475 616c 735f 7461  ._individuals_ta
-0000c970: 625f 6964 5d5b 325d 2066 6f72 2069 6e64  b_id][2] for ind
-0000c980: 2069 6e20 7365 6c66 2e76 616c 7565 7328   in self.values(
-0000c990: 297d 0a0a 2020 2020 2020 2020 2320 6e6f  )}..        # no
-0000c9a0: 7720 7369 6d70 6c69 6679 2074 6865 2074  w simplify the t
-0000c9b0: 6162 6c65 7320 616e 6420 6765 7420 7468  ables and get th
-0000c9c0: 6520 6e65 7720 6964 7320 6f75 7470 7574  e new ids output
-0000c9d0: 200a 2020 2020 2020 2020 2320 6672 6f6d   .        # from
-0000c9e0: 2074 6865 2074 632e 7369 6d70 6c69 6679   the tc.simplify
-0000c9f0: 206d 6574 686f 640a 2020 2020 2020 2020   method.        
-0000ca00: 2320 4e4f 5445 3a20 7765 2061 7265 206e  # NOTE: we are n
-0000ca10: 6f74 2075 7369 6e67 2074 6865 206f 7570  ot using the oup
-0000ca20: 7574 2c20 6275 7420 6974 2069 7320 616e  ut, but it is an
-0000ca30: 2061 7272 6179 2077 6865 7265 0a20 2020   array where.   
-0000ca40: 2020 2020 2023 2065 6163 6820 7661 6c75       # each valu
-0000ca50: 6520 6973 2074 6865 206e 6577 206e 6f64  e is the new nod
-0000ca60: 6520 4944 206f 6620 7468 6520 6e6f 6465  e ID of the node
-0000ca70: 2074 6861 7420 7761 7320 696e 2074 6861   that was in tha
-0000ca80: 7420 696e 6465 7827 730a 2020 2020 2020  t index's.      
-0000ca90: 2020 2320 726f 7720 706f 7369 7469 6f6e    # row position
-0000caa0: 2069 6e20 7468 6520 6f6c 6420 6e6f 6465   in the old node
-0000cab0: 7320 7461 626c 6520 2861 6e64 202d 3120  s table (and -1 
-0000cac0: 6966 2074 6861 7420 6e6f 6465 2077 6173  if that node was
-0000cad0: 0a20 2020 2020 2020 2023 2064 726f 7070  .        # dropp
-0000cae0: 6564 2064 7572 696e 6720 7369 6d70 6c69  ed during simpli
-0000caf0: 6361 7469 6f6e 293b 2073 6565 2074 736b  cation); see tsk
-0000cb00: 6974 2064 6f63 7320 666f 7220 6465 7461  it docs for deta
-0000cb10: 696c 730a 2020 2020 2020 2020 6f75 7470  ils.        outp
-0000cb20: 7574 203d 2073 656c 662e 5f74 632e 7369  ut = self._tc.si
-0000cb30: 6d70 6c69 6679 2863 7572 725f 6e6f 6465  mplify(curr_node
-0000cb40: 732c 0a20 2020 2020 2020 2020 2020 2020  s,.             
-0000cb50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000cb60: 2020 2020 2020 6669 6c74 6572 5f69 6e64        filter_ind
-0000cb70: 6976 6964 7561 6c73 3d54 7275 652c 0a20  ividuals=True,. 
-0000cb80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000cb90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000cba0: 2020 6669 6c74 6572 5f73 6974 6573 3d46    filter_sites=F
-0000cbb0: 616c 7365 290a 0a20 2020 2020 2020 2023  alse)..        #
-0000cbc0: 206d 616b 6520 616e 204e 7833 206e 7020   make an Nx3 np 
-0000cbd0: 6172 7261 7920 636f 6e74 6169 6e69 6e67  array containing
-0000cbe0: 2031 2e29 2067 6e78 2069 6473 2c20 322e   1.) gnx ids, 2.
-0000cbf0: 2920 7468 6520 6e65 770a 2020 2020 2020  ) the new.      
-0000cc00: 2020 2320 686f 6d6f 6c6f 6775 6520 3020    # homologue 0 
-0000cc10: 6e6f 6465 2069 6473 2c20 616e 6420 332e  node ids, and 3.
-0000cc20: 2920 7468 6520 6e65 7720 686f 6d6f 6c6f  ) the new homolo
-0000cc30: 6775 6520 3120 6e6f 6465 2069 6473 2c0a  gue 1 node ids,.
-0000cc40: 2020 2020 2020 2020 2320 696e 2065 6163          # in eac
-0000cc50: 6820 6974 7320 3320 636f 6c73 3b20 7468  h its 3 cols; th
-0000cc60: 656e 206c 6f6f 7020 6f76 6572 2069 7473  en loop over its
-0000cc70: 2072 6f77 7320 746f 2075 7064 6174 650a   rows to update.
-0000cc80: 2020 2020 2020 2020 2320 6561 6368 2069          # each i
-0000cc90: 6e64 6976 6964 7561 6c27 7320 5f6e 6f64  ndividual's _nod
-0000cca0: 6573 5f74 6162 5f69 6473 2061 7474 7269  es_tab_ids attri
-0000ccb0: 6275 7465 0a20 2020 2020 2020 2069 6e64  bute.        ind
-0000ccc0: 735f 676e 785f 6964 7320 3d20 6e70 2e61  s_gnx_ids = np.a
-0000ccd0: 7272 6179 285b 2a73 656c 665d 292e 7265  rray([*self]).re
-0000cce0: 7368 6170 6528 286c 656e 2873 656c 6629  shape((len(self)
-0000ccf0: 2c20 3129 290a 2020 2020 2020 2020 6e65  , 1)).        ne
-0000cd00: 775f 6964 7320 3d20 6e70 2e68 7374 6163  w_ids = np.hstac
-0000cd10: 6b28 2869 6e64 735f 676e 785f 6964 732c  k((inds_gnx_ids,
-0000cd20: 206e 702e 6172 7261 7928 5b2a 7261 6e67   np.array([*rang
-0000cd30: 6528 6c65 6e28 0a20 2020 2020 2020 2020  e(len(.         
-0000cd40: 2020 2069 6e64 735f 676e 785f 6964 7329     inds_gnx_ids)
-0000cd50: 202a 2032 295d 292e 7265 7368 6170 6528   * 2)]).reshape(
-0000cd60: 286c 656e 2869 6e64 735f 676e 785f 6964  (len(inds_gnx_id
-0000cd70: 7329 2c20 3229 2929 290a 2020 2020 2020  s), 2)))).      
-0000cd80: 2020 666f 7220 6964 2c20 6e30 2c20 6e31    for id, n0, n1
-0000cd90: 2069 6e20 6e65 775f 6964 733a 0a20 2020   in new_ids:.   
-0000cda0: 2020 2020 2020 2020 2073 656c 665b 6964           self[id
-0000cdb0: 5d2e 5f6e 6f64 6573 5f74 6162 5f69 6473  ]._nodes_tab_ids
-0000cdc0: 2e75 7064 6174 6528 7b30 3a6e 302c 2031  .update({0:n0, 1
-0000cdd0: 3a6e 317d 290a 0a20 2020 2020 2020 2023  :n1})..        #
-0000cde0: 2075 7064 6174 6520 496e 6469 7669 6475   update Individu
-0000cdf0: 616c 7327 2074 6162 6c65 2069 6473 0a20  als' table ids. 
-0000ce00: 2020 2020 2020 2023 2028 692e 652e 2074         # (i.e. t
-0000ce10: 6865 6972 2049 6e64 6976 6964 7561 6c2e  heir Individual.
-0000ce20: 5f69 6e64 6976 6964 7561 6c73 5f74 6162  _individuals_tab
-0000ce30: 5f69 6420 6174 7472 6962 7574 6573 290a  _id attributes).
-0000ce40: 2020 2020 2020 2020 696e 645f 6d65 7461          ind_meta
-0000ce50: 203d 2073 656c 662e 5f74 632e 696e 6469   = self._tc.indi
-0000ce60: 7669 6475 616c 732e 6d65 7461 6461 7461  viduals.metadata
-0000ce70: 0a20 2020 2020 2020 2069 6620 6368 6563  .        if chec
-0000ce80: 6b5f 696e 6469 7669 6475 616c 733a 0a20  k_individuals:. 
-0000ce90: 2020 2020 2020 2020 2020 2070 7269 6e74             print
-0000cea0: 2827 6234 272c 206c 656e 286d 6574 615f  ('b4', len(meta_
-0000ceb0: 6234 292c 2027 6166 272c 206c 656e 2869  b4), 'af', len(i
-0000cec0: 6e64 5f6d 6574 6129 290a 2020 2020 2020  nd_meta)).      
-0000ced0: 2020 696e 645f 6f66 6620 3d20 7365 6c66    ind_off = self
-0000cee0: 2e5f 7463 2e69 6e64 6976 6964 7561 6c73  ._tc.individuals
-0000cef0: 2e6d 6574 6164 6174 615f 6f66 6673 6574  .metadata_offset
-0000cf00: 0a20 2020 2020 2020 206e 6577 5f69 6e64  .        new_ind
-0000cf10: 6976 6964 7561 6c73 5f74 6162 5f69 6420  ividuals_tab_id 
-0000cf20: 3d20 7b69 6e74 2e66 726f 6d5f 6279 7465  = {int.from_byte
-0000cf30: 7328 0a20 2020 2020 2020 2020 2020 2069  s(.            i
-0000cf40: 6e64 5f6d 6574 615b 696e 645f 6f66 665b  nd_meta[ind_off[
-0000cf50: 695d 3a20 696e 645f 6f66 665b 692b 315d  i]: ind_off[i+1]
-0000cf60: 5d2e 746f 6279 7465 7328 292c 0a20 2020  ].tobytes(),.   
-0000cf70: 2020 2020 2020 2020 2027 6c69 7474 6c65           'little
-0000cf80: 2729 3a20 6920 666f 7220 6920 696e 2072  '): i for i in r
-0000cf90: 616e 6765 286c 656e 2869 6e64 5f6f 6666  ange(len(ind_off
-0000cfa0: 2920 2d20 3129 7d0a 2020 2020 2020 2020  ) - 1)}.        
-0000cfb0: 5b73 6574 6174 7472 2869 6e64 2c20 275f  [setattr(ind, '_
-0000cfc0: 696e 6469 7669 6475 616c 735f 7461 625f  individuals_tab_
-0000cfd0: 6964 272c 0a20 2020 2020 2020 2020 2020  id',.           
-0000cfe0: 2020 2020 2020 6e65 775f 696e 6469 7669        new_indivi
-0000cff0: 6475 616c 735f 7461 625f 6964 5b69 645d  duals_tab_id[id]
-0000d000: 2920 666f 7220 6964 2c20 696e 6420 696e  ) for id, ind in
-0000d010: 2073 656c 662e 6974 656d 7328 295d 0a0a   self.items()]..
-0000d020: 2020 2020 2020 2020 2320 6368 6563 6b20          # check 
-0000d030: 7468 6174 2069 6e64 6976 6964 7561 6c73  that individuals
-0000d040: 2720 6e6f 6465 732d 7461 626c 6520 6964  ' nodes-table id
-0000d050: 7320 7765 7265 2063 6f72 7265 636c 7479  s were correclty
-0000d060: 2075 7064 6174 6564 2c0a 2020 2020 2020   updated,.      
-0000d070: 2020 2320 6966 2074 6865 2063 6865 636b    # if the check
-0000d080: 2069 7320 7265 7175 6573 7465 640a 2020   is requested.  
-0000d090: 2020 2020 2020 6966 2063 6865 636b 5f6e        if check_n
-0000d0a0: 6f64 6573 3a0a 2020 2020 2020 2020 2020  odes:.          
-0000d0b0: 2020 2363 7265 6174 6520 616e 6f74 6865    #create anothe
-0000d0c0: 7220 6964 656e 7469 6361 6c6c 7920 7374  r identically st
-0000d0d0: 7275 6374 7572 6564 204e 7833 206e 7020  ructured Nx3 np 
-0000d0e0: 6172 7261 792c 2074 6f20 686f 6c64 2074  array, to hold t
-0000d0f0: 6865 0a20 2020 2020 2020 2020 2020 2023  he.            #
-0000d100: 696e 6469 7669 6475 616c 7327 2067 6e78  individuals' gnx
-0000d110: 2069 6473 2061 6e64 2074 6865 6972 206e   ids and their n
-0000d120: 6f64 6520 6964 7320 6163 636f 7264 696e  ode ids accordin
-0000d130: 6720 746f 2074 6865 2074 736b 6974 0a20  g to the tskit. 
-0000d140: 2020 2020 2020 2020 2020 2023 6e6f 6465             #node
-0000d150: 7320 7461 626c 6520 2866 6f72 2063 726f  s table (for cro
-0000d160: 7373 2d63 6865 636b 696e 6729 0a20 2020  ss-checking).   
-0000d170: 2020 2020 2020 2020 206e 6577 5f69 6473           new_ids
-0000d180: 5f66 726f 6d5f 7461 626c 6573 203d 206e  _from_tables = n
-0000d190: 702e 6f6e 6573 2828 6c65 6e28 696e 6473  p.ones((len(inds
-0000d1a0: 5f67 6e78 5f69 6473 292c 2033 2929 202a  _gnx_ids), 3)) *
-0000d1b0: 206e 702e 6e61 6e0a 2020 2020 2020 2020   np.nan.        
-0000d1c0: 2020 2020 6e65 775f 6964 735f 6672 6f6d      new_ids_from
-0000d1d0: 5f74 6162 6c65 735b 3a2c 305d 203d 2069  _tables[:,0] = i
-0000d1e0: 6e64 735f 676e 785f 6964 735b 3a2c 305d  nds_gnx_ids[:,0]
-0000d1f0: 0a20 2020 2020 2020 2020 2020 206e 6f64  .            nod
-0000d200: 6564 6620 3d20 7064 2e44 6174 6146 7261  edf = pd.DataFra
-0000d210: 6d65 287b 6b3a 7620 666f 7220 6b2c 2076  me({k:v for k, v
-0000d220: 2069 6e20 7365 6c66 2e5f 7463 2e6e 6f64   in self._tc.nod
-0000d230: 6573 2e61 7364 6963 7428 0a20 2020 2020  es.asdict(.     
-0000d240: 2020 2020 2020 2020 2020 2029 2e69 7465             ).ite
-0000d250: 6d73 2829 2069 6620 6b20 696e 205b 2774  ms() if k in ['t
-0000d260: 696d 6527 2c20 2769 6e64 6976 6964 7561  ime', 'individua
-0000d270: 6c27 5d7d 290a 2020 2020 2020 2020 2020  l']}).          
-0000d280: 2020 696e 645f 7461 626c 655f 6964 7320    ind_table_ids 
-0000d290: 3d20 5b73 656c 665b 6964 785b 0a20 2020  = [self[idx[.   
-0000d2a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d2b0: 2020 2020 2020 2020 2030 5d5d 2e5f 696e           0]]._in
-0000d2c0: 6469 7669 6475 616c 735f 7461 625f 6964  dividuals_tab_id
-0000d2d0: 2066 6f72 2069 6478 2069 6e20 696e 6473   for idx in inds
-0000d2e0: 5f67 6e78 5f69 6473 5d0a 2020 2020 2020  _gnx_ids].      
-0000d2f0: 2020 2020 2020 2320 6c6f 6f70 206f 7665        # loop ove
-0000d300: 7220 696e 6469 7669 6475 616c 7327 2067  r individuals' g
-0000d310: 6e78 2069 6473 2061 6e64 2063 6f72 7265  nx ids and corre
-0000d320: 7370 6f6e 6469 6e67 2074 6162 6c65 2072  sponding table r
-0000d330: 6f77 2069 6473 2c0a 2020 2020 2020 2020  ow ids,.        
-0000d340: 2020 2020 2320 6669 6c6c 696e 6720 7570      # filling up
-0000d350: 2074 6865 2063 6f6d 7061 7269 736f 6e20   the comparison 
-0000d360: 7461 626c 650a 2020 2020 2020 2020 2020  table.          
-0000d370: 2020 666f 7220 696e 642c 2069 6478 2069    for ind, idx i
-0000d380: 6e20 7a69 7028 696e 6473 5f67 6e78 5f69  n zip(inds_gnx_i
-0000d390: 6473 5b3a 2c30 5d2c 2069 6e64 5f74 6162  ds[:,0], ind_tab
-0000d3a0: 6c65 5f69 6473 293a 0a20 2020 2020 2020  le_ids):.       
-0000d3b0: 2020 2020 2020 2020 2073 7562 6466 203d           subdf =
-0000d3c0: 206e 6f64 6564 665b 6e6f 6465 6466 5b27   nodedf[nodedf['
-0000d3d0: 696e 6469 7669 6475 616c 275d 203d 3d20  individual'] == 
-0000d3e0: 6964 785d 0a20 2020 2020 2020 2020 2020  idx].           
-0000d3f0: 2020 2020 2070 7269 6e74 2827 2e2e 2e5c       print('...\
-0000d400: 6e67 6e78 2069 643a 272c 2069 6e64 2c20  ngnx id:', ind, 
-0000d410: 275c 7474 6162 6c65 2069 643a 272c 2069  '\ttable id:', i
-0000d420: 6478 290a 2020 2020 2020 2020 2020 2020  dx).            
-0000d430: 2020 2020 7072 696e 7428 2767 6e78 206e      print('gnx n
-0000d440: 6f64 6520 6964 733a 2020 272c 205b 2a73  ode ids:  ', [*s
-0000d450: 656c 665b 696e 645d 2e5f 6e6f 6465 735f  elf[ind]._nodes_
-0000d460: 7461 625f 6964 732e 7661 6c75 6573 2829  tab_ids.values()
-0000d470: 5d29 0a20 2020 2020 2020 2020 2020 2020  ]).             
-0000d480: 2020 2070 7269 6e74 2827 7461 626c 6520     print('table 
-0000d490: 6e6f 6465 2069 6473 3a27 2c20 5b2a 7375  node ids:', [*su
-0000d4a0: 6264 662e 696e 6465 785d 290a 2020 2020  bdf.index]).    
-0000d4b0: 2020 2020 2020 2020 2020 2020 6e6f 6465              node
-0000d4c0: 5f69 645f 7661 6c73 203d 205b 2a73 7562  _id_vals = [*sub
-0000d4d0: 6466 2e69 6e64 6578 5d0a 2020 2020 2020  df.index].      
-0000d4e0: 2020 2020 2020 2020 2020 726f 775f 6964            row_id
-0000d4f0: 7820 3d20 6e70 2e77 6865 7265 286e 6577  x = np.where(new
-0000d500: 5f69 6473 5f66 726f 6d5f 7461 626c 6573  _ids_from_tables
-0000d510: 5b3a 2c30 5d20 3d3d 2069 6e64 295b 305d  [:,0] == ind)[0]
-0000d520: 5b30 5d0a 2020 2020 2020 2020 2020 2020  [0].            
-0000d530: 2020 2020 6e65 775f 6964 735f 6672 6f6d      new_ids_from
-0000d540: 5f74 6162 6c65 735b 726f 775f 6964 782c  _tables[row_idx,
-0000d550: 2031 3a5d 203d 206e 6f64 655f 6964 5f76   1:] = node_id_v
-0000d560: 616c 730a 2020 2020 2020 2020 2020 2020  als.            
-0000d570: 2320 6e6f 7720 6d61 6b65 2073 7572 6520  # now make sure 
-0000d580: 7461 626c 7320 6172 6520 6964 656e 7469  tabls are identi
-0000d590: 6361 6c0a 2020 2020 2020 2020 2020 2020  cal.            
-0000d5a0: 6173 7365 7274 206e 702e 616c 6c28 6e65  assert np.all(ne
-0000d5b0: 775f 6964 735f 6672 6f6d 5f74 6162 6c65  w_ids_from_table
-0000d5c0: 7320 3d3d 206e 6577 5f69 6473 292c 2028  s == new_ids), (
-0000d5d0: 2241 6c6c 2074 6865 206e 6f64 6520 220a  "All the node ".
-0000d5e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d5f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d600: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d610: 2020 2020 2020 2020 2020 2020 2269 6473              "ids
-0000d620: 2064 6f6e 2774 206d 6174 6368 2122 290a   don't match!").
-0000d630: 2020 2020 2020 2020 2020 2020 7072 696e              prin
-0000d640: 7428 2741 4c4c 204e 4f44 4520 4944 5320  t('ALL NODE IDS 
-0000d650: 4150 5045 4152 2043 4f52 5245 4354 4c59  APPEAR CORRECTLY
-0000d660: 2052 4541 5353 4947 4e45 445c 6e27 290a   REASSIGNED\n').
-0000d670: 0a20 2020 2020 2020 2023 2063 6865 636b  .        # check
-0000d680: 2074 6861 7420 696e 6469 7669 6475 616c   that individual
-0000d690: 7327 2069 6e64 6976 6964 7561 6c73 2d74  s' individuals-t
-0000d6a0: 6162 6c65 2069 6473 2077 6572 6520 636f  able ids were co
-0000d6b0: 7272 6563 6c74 7920 7570 6461 7465 642c  rreclty updated,
-0000d6c0: 0a20 2020 2020 2020 2023 2069 6620 7468  .        # if th
-0000d6d0: 6520 6368 6563 6b20 6973 2072 6571 7565  e check is reque
-0000d6e0: 7374 6564 0a20 2020 2020 2020 2069 6620  sted.        if 
-0000d6f0: 6368 6563 6b5f 696e 6469 7669 6475 616c  check_individual
-0000d700: 733a 0a20 2020 2020 2020 2020 2020 2023  s:.            #
-0000d710: 2063 6865 636b 2074 6861 7420 696e 6469   check that indi
-0000d720: 7669 6475 616c 7327 2074 6162 6c65 2069  viduals' table i
-0000d730: 6473 2077 6572 6520 636f 7272 6563 746c  ds were correctl
-0000d740: 7920 7265 6173 7369 676e 6564 2c20 6279  y reassigned, by
-0000d750: 0a20 2020 2020 2020 2020 2020 2023 2077  .            # w
-0000d760: 6179 206f 6620 6173 7375 7269 6e67 2074  ay of assuring t
-0000d770: 6861 7420 696e 6469 7669 6475 616c 732c  hat individuals,
-0000d780: 2061 7320 696e 6465 7865 6420 6279 2062   as indexed by b
-0000d790: 6f74 6820 7468 6569 7220 6f6c 640a 2020  oth their old.  
-0000d7a0: 2020 2020 2020 2020 2020 2320 616e 6420            # and 
-0000d7b0: 7468 6569 7220 6e65 7720 696e 6469 7669  their new indivi
-0000d7c0: 6475 616c 732d 7461 626c 6520 6964 732c  duals-table ids,
-0000d7d0: 2068 6176 6520 6964 656e 7469 6361 6c20   have identical 
-0000d7e0: 7068 656e 6f74 7970 6573 0a20 2020 2020  phenotypes.     
-0000d7f0: 2020 2020 2020 206c 6f63 203d 2073 656c         loc = sel
-0000d800: 662e 5f74 632e 696e 6469 7669 6475 616c  f._tc.individual
-0000d810: 732e 6c6f 6361 7469 6f6e 0a20 2020 2020  s.location.     
-0000d820: 2020 2020 2020 206c 6f63 6f66 6620 3d20         locoff = 
-0000d830: 7365 6c66 2e5f 7463 2e69 6e64 6976 6964  self._tc.individ
-0000d840: 7561 6c73 2e6c 6f63 6174 696f 6e5f 6f66  uals.location_of
-0000d850: 6673 6574 0a20 2020 2020 2020 2020 2020  fset.           
-0000d860: 206c 6f63 5f61 665f 6279 5f74 6162 6c65   loc_af_by_table
-0000d870: 5f69 6473 203d 207b 0a20 2020 2020 2020  _ids = {.       
-0000d880: 2020 2020 2020 2020 2074 6162 6c65 5f69           table_i
-0000d890: 643a 2028 6c6f 635b 6c6f 636f 6666 5b74  d: (loc[locoff[t
-0000d8a0: 6162 6c65 5f69 645d 3a20 6c6f 636f 6666  able_id]: locoff
-0000d8b0: 5b0a 2020 2020 2020 2020 2020 2020 2020  [.              
-0000d8c0: 2020 2020 2020 7461 626c 655f 6964 2b31        table_id+1
-0000d8d0: 5d5d 2920 666f 7220 7461 626c 655f 6964  ]]) for table_id
-0000d8e0: 2069 6e20 7261 6e67 6528 6c65 6e28 0a20   in range(len(. 
-0000d8f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d900: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d910: 2020 2020 2020 2020 2020 2020 2020 2073                 s
-0000d920: 656c 662e 5f74 632e 696e 6469 7669 6475  elf._tc.individu
-0000d930: 616c 732e 666c 6167 7329 297d 0a20 2020  als.flags))}.   
-0000d940: 2020 2020 2020 2020 207a 305f 6166 203d           z0_af =
-0000d950: 207b 696e 642e 6964 783a 206c 6f63 5f61   {ind.idx: loc_a
-0000d960: 665f 6279 5f74 6162 6c65 5f69 6473 5b0a  f_by_table_ids[.
-0000d970: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000d980: 2020 2020 2020 2020 696e 642e 5f69 6e64          ind._ind
-0000d990: 6976 6964 7561 6c73 5f74 6162 5f69 645d  ividuals_tab_id]
-0000d9a0: 5b32 5d20 666f 7220 696e 6420 696e 2073  [2] for ind in s
-0000d9b0: 656c 662e 7661 6c75 6573 2829 7d0a 2020  elf.values()}.  
-0000d9c0: 2020 2020 2020 2020 2020 7072 696e 7428            print(
-0000d9d0: 7a30 5f61 6629 0a20 2020 2020 2020 2020  z0_af).         
-0000d9e0: 2020 207a 305f 6368 6563 6b20 3d20 5b7a     z0_check = [z
-0000d9f0: 3020 3d3d 207a 305f 6234 5b69 6478 5d20  0 == z0_b4[idx] 
-0000da00: 666f 7220 6964 782c 207a 3020 696e 207a  for idx, z0 in z
-0000da10: 305f 6166 2e69 7465 6d73 2829 5d0a 2020  0_af.items()].  
-0000da20: 2020 2020 2020 2020 2020 666f 7220 6964            for id
-0000da30: 7820 696e 205b 2a73 656c 665d 3a0a 2020  x in [*self]:.  
-0000da40: 2020 2020 2020 2020 2020 2020 2020 7072                pr
-0000da50: 696e 7428 6964 782c 2027 2062 3420 272c  int(idx, ' b4 ',
-0000da60: 207a 305f 6234 5b69 6478 5d29 0a20 2020   z0_b4[idx]).   
-0000da70: 2020 2020 2020 2020 2020 2020 2070 7269               pri
-0000da80: 6e74 2869 6478 2c20 2720 6166 2027 2c20  nt(idx, ' af ', 
-0000da90: 7a30 5f61 665b 6964 785d 290a 2020 2020  z0_af[idx]).    
-0000daa0: 2020 2020 2020 2020 2020 2020 7072 696e              prin
-0000dab0: 7428 272e 2e2e 2e2e 2e2e 2e2e 2e2e 2e2e  t('.............
-0000dac0: 2e2e 2e2e 2e27 290a 2020 2020 2020 2020  .....').        
-0000dad0: 2020 2020 6173 7365 7274 206e 702e 616c      assert np.al
-0000dae0: 6c28 7a30 5f63 6865 636b 292c 2028 2770  l(z0_check), ('p
-0000daf0: 6865 6e6f 7479 7065 7320 6172 6520 6e6f  henotypes are no
-0000db00: 7420 7468 6520 7361 6d65 2120 270a 2020  t the same! '.  
-0000db10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000db20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000db30: 2020 2020 2720 284d 414b 4520 5355 5245      ' (MAKE SURE
-0000db40: 2059 4f55 2055 5345 4420 4120 4d4f 4445   YOU USED A MODE
-0000db50: 4c20 5749 5448 2027 0a20 2020 2020 2020  L WITH '.       
-0000db60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000db70: 2020 2020 2020 2020 2020 2020 2020 2027                 '
-0000db80: 2041 5420 4c45 4153 5420 4f4e 4520 5452   AT LEAST ONE TR
-0000db90: 4149 5421 2927 290a 2020 2020 2020 2020  AIT!)').        
-0000dba0: 2020 2020 7072 696e 7428 2741 4c4c 2049      print('ALL I
-0000dbb0: 4e44 4956 4944 5541 4c20 4944 5320 4150  NDIVIDUAL IDS AP
-0000dbc0: 5045 4152 2043 4f52 5245 4354 4c59 2052  PEAR CORRECTLY R
-0000dbd0: 4541 5353 4947 4e45 445c 6e27 290a 0a20  EASSIGNED\n').. 
-0000dbe0: 2020 2020 2020 2023 2073 6574 2074 6865         # set the
-0000dbf0: 2073 6f72 7465 645f 616e 645f 7369 6d70   sorted_and_simp
-0000dc00: 6c69 6669 6564 2066 6c61 6720 746f 2054  lified flag to T
-0000dc10: 7275 650a 2020 2020 2020 2020 7365 6c66  rue.        self
-0000dc20: 2e5f 7463 5f73 6f72 7465 645f 616e 645f  ._tc_sorted_and_
-0000dc30: 7369 6d70 6c69 6669 6564 203d 2054 7275  simplified = Tru
-0000dc40: 650a 0a0a 2020 2020 2320 6765 7420 7468  e...    # get th
-0000dc50: 6520 6e6f 6465 732d 7461 626c 6520 4944  e nodes-table ID
-0000dc60: 7320 666f 7220 616c 6c20 696e 6469 7669  s for all indivi
-0000dc70: 6475 616c 730a 2020 2020 6465 6620 5f67  duals.    def _g
-0000dc80: 6574 5f6e 6f64 6573 2873 656c 662c 2069  et_nodes(self, i
-0000dc90: 6e64 6976 6964 733d 4e6f 6e65 293a 0a20  ndivids=None):. 
-0000dca0: 2020 2020 2020 2069 6620 696e 6469 7669         if indivi
-0000dcb0: 6473 2069 7320 4e6f 6e65 3a0a 2020 2020  ds is None:.    
-0000dcc0: 2020 2020 2020 2020 696e 6469 7669 6473          individs
-0000dcd0: 203d 205b 2a73 656c 665d 0a20 2020 2020   = [*self].     
-0000dce0: 2020 206e 6f64 6573 203d 206e 702e 6873     nodes = np.hs
-0000dcf0: 7461 636b 285b 5b2a 7365 6c66 5b69 5d2e  tack([[*self[i].
-0000dd00: 5f6e 6f64 6573 5f74 6162 5f69 6473 2e76  _nodes_tab_ids.v
-0000dd10: 616c 7565 7328 0a20 2020 2020 2020 2020  alues(.         
+0000be00: 2073 656c 662e 6765 6e5f 6172 6368 2e78   self.gen_arch.x
+0000be10: 295d 0a20 2020 2020 2020 2020 2020 2020  )].             
+0000be20: 2020 2069 6e64 2e5f 7365 745f 6e6f 6465     ind._set_node
+0000be30: 735f 7461 625f 6964 7328 2a69 6e64 5f6e  s_tab_ids(*ind_n
+0000be40: 6f64 655f 6964 7329 0a20 2020 2020 2020  ode_ids).       
+0000be50: 2020 2020 2020 2020 2023 2061 6464 2074           # add t
+0000be60: 6869 7320 696e 6469 7669 6475 616c 2773  his individual's
+0000be70: 2069 6e64 5f69 6420 746f 2074 6865 0a20   ind_id to the. 
+0000be80: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+0000be90: 206e 6f64 6573 5f74 6162 5f69 6e64 6976   nodes_tab_indiv
+0000bea0: 6964 7561 6c5f 636f 6c20 6c69 7374 2c20  idual_col list, 
+0000beb0: 6f6e 6365 2066 6f72 2065 6163 6820 6e6f  once for each no
+0000bec0: 6465 0a20 2020 2020 2020 2020 2020 2020  de.             
+0000bed0: 2020 206e 6f64 6573 5f74 6162 5f69 6e64     nodes_tab_ind
+0000bee0: 6976 6964 7561 6c5f 636f 6c5b 696e 645f  ividual_col[ind_
+0000bef0: 6e6f 6465 5f69 6473 5d20 3d20 696e 645f  node_ids] = ind_
+0000bf00: 6964 0a20 2020 2020 2020 2020 2020 2023  id.            #
+0000bf10: 206d 616b 6520 7375 7265 2074 6861 7420   make sure that 
+0000bf20: 616c 6c20 6e6f 6465 7320 7765 7265 2061  all nodes were a
+0000bf30: 7373 6967 6e65 6420 746f 2069 6e64 6976  ssigned to indiv
+0000bf40: 6964 7561 6c73 0a20 2020 2020 2020 2020  iduals.         
+0000bf50: 2020 2061 7373 6572 7420 6e70 2e61 6c6c     assert np.all
+0000bf60: 286e 6f64 6573 5f74 6162 5f69 6e64 6976  (nodes_tab_indiv
+0000bf70: 6964 7561 6c5f 636f 6c20 3e3d 2030 292c  idual_col >= 0),
+0000bf80: 2028 2753 6f6d 6520 6e6f 6465 7320 6e6f   ('Some nodes no
+0000bf90: 7420 270a 2020 2020 2020 2020 2020 2020  t '.            
+0000bfa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000bfb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000bfc0: 2020 2020 2020 2020 2020 2020 2020 2027                 '
+0000bfd0: 6769 7665 6e20 696e 6469 7669 6473 2729  given individs')
+0000bfe0: 0a0a 2020 2020 2020 2020 2020 2020 6e6f  ..            no
+0000bff0: 6465 735f 636f 6c73 203d 2074 6162 6c65  des_cols = table
+0000c000: 732e 6e6f 6465 732e 6173 6469 6374 2829  s.nodes.asdict()
+0000c010: 0a20 2020 2020 2020 2020 2020 2023 2075  .            # u
+0000c020: 7365 2074 6865 206e 6f64 655f 7461 625f  se the node_tab_
+0000c030: 696e 6469 7669 6475 616c 5f63 6f6c 2064  individual_col d
+0000c040: 6174 6120 7374 7275 6374 7572 6520 746f  ata structure to
+0000c050: 2072 6573 6574 0a20 2020 2020 2020 2020   reset.         
+0000c060: 2020 2023 2074 6865 2069 6e64 6976 6964     # the individ
+0000c070: 7561 6c73 2063 6f6c 756d 6e20 696e 2074  uals column in t
+0000c080: 6865 206e 6f64 6573 2074 6162 6c65 0a20  he nodes table. 
+0000c090: 2020 2020 2020 2020 2020 206e 6f64 6573             nodes
+0000c0a0: 5f63 6f6c 735b 2769 6e64 6976 6964 7561  _cols['individua
+0000c0b0: 6c27 5d5b 3a5d 203d 206e 6f64 6573 5f74  l'][:] = nodes_t
+0000c0c0: 6162 5f69 6e64 6976 6964 7561 6c5f 636f  ab_individual_co
+0000c0d0: 6c0a 2020 2020 2020 2020 2020 2020 2320  l.            # 
+0000c0e0: 696e 6372 656d 656e 7420 616c 6c20 6269  increment all bi
+0000c0f0: 7274 6820 7469 6d65 7320 6279 2031 2c20  rth times by 1, 
+0000c100: 736f 2074 6861 7420 616c 6c20 696e 6469  so that all indi
+0000c110: 7669 6473 2061 7265 206d 6172 6b65 640a  vids are marked.
+0000c120: 2020 2020 2020 2020 2020 2020 2320 6173              # as
+0000c130: 2068 6176 696e 6720 6265 656e 2062 6f72   having been bor
+0000c140: 6e20 6265 666f 7265 2074 6865 2073 7461  n before the sta
+0000c150: 7274 206f 6620 7468 6520 6d6f 6465 6c27  rt of the model'
+0000c160: 7320 6d61 696e 2070 6861 7365 0a20 2020  s main phase.   
+0000c170: 2020 2020 2020 2020 206e 6f64 6573 5f63           nodes_c
+0000c180: 6f6c 735b 2774 696d 6527 5d20 2b3d 2031  ols['time'] += 1
+0000c190: 0a20 2020 2020 2020 2020 2020 2074 6162  .            tab
+0000c1a0: 6c65 732e 6e6f 6465 732e 7365 745f 636f  les.nodes.set_co
+0000c1b0: 6c75 6d6e 7328 2a2a 6e6f 6465 735f 636f  lumns(**nodes_co
+0000c1c0: 6c73 290a 0a20 2020 2020 2020 2023 206f  ls)..        # o
+0000c1d0: 7468 6572 7769 7365 2c20 6966 2074 736b  therwise, if tsk
+0000c1e0: 6974 2069 736e 2774 2074 6f20 6265 2075  it isn't to be u
+0000c1f0: 7365 642c 2074 6865 6e20 7365 7420 7461  sed, then set ta
+0000c200: 626c 6573 2074 6f20 4e6f 6e65 0a20 2020  bles to None.   
+0000c210: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
+0000c220: 2020 2020 2020 2074 6162 6c65 7320 3d20         tables = 
+0000c230: 4e6f 6e65 0a0a 2020 2020 2020 2020 2320  None..        # 
+0000c240: 6164 6420 7375 6666 6963 6965 6e74 206d  add sufficient m
+0000c250: 7574 6174 696f 6e73 2c20 2861 7420 6f6e  utations, (at on
+0000c260: 6c79 2063 7572 7265 6e74 206e 6f64 6573  ly current nodes
+0000c270: 2c20 6966 2075 7369 6e67 2074 736b 6974  , if using tskit
+0000c280: 292c 0a20 2020 2020 2020 2023 2074 6f20  ),.        # to 
+0000c290: 7072 6f64 7563 6520 7468 6520 7374 6172  produce the star
+0000c2a0: 7469 6e67 2031 2d61 6c6c 656c 6520 6672  ting 1-allele fr
+0000c2b0: 6571 7565 6e63 6965 730a 2020 2020 2020  equencies.      
+0000c2c0: 2020 2320 7061 7261 6d65 7465 7269 7a65    # parameterize
+0000c2d0: 6420 666f 7220 7468 6973 2073 7065 6369  d for this speci
+0000c2e0: 6573 0a20 2020 2020 2020 205f 6d61 6b65  es.        _make
+0000c2f0: 5f73 7461 7274 696e 675f 6d75 7461 7469  _starting_mutati
+0000c300: 6f6e 7328 7365 6c66 2c20 7461 626c 6573  ons(self, tables
+0000c310: 290a 0a20 2020 2020 2020 2023 2054 4f44  )..        # TOD
+0000c320: 4f3a 2041 4444 2050 524f 5645 4e41 4e43  O: ADD PROVENANC
+0000c330: 4553 2052 4f57 210a 0a20 2020 2020 2020  ES ROW!..       
+0000c340: 2023 2061 7373 6967 6e20 6173 2074 6865   # assign as the
+0000c350: 2073 7065 6369 6573 2720 5461 626c 6543   species' TableC
+0000c360: 6f6c 6c65 6374 696f 6e0a 2020 2020 2020  ollection.      
+0000c370: 2020 7365 6c66 2e5f 7463 203d 2074 6162    self._tc = tab
+0000c380: 6c65 730a 0a20 2020 2020 2020 2072 6574  les..        ret
+0000c390: 7572 6e0a 0a0a 2020 2020 2320 6d65 7468  urn...    # meth
+0000c3a0: 6f64 2074 6f20 736f 7274 2061 6e64 2073  od to sort and s
+0000c3b0: 696d 706c 6966 7920 7468 6520 5461 626c  implify the Tabl
+0000c3c0: 6543 6f6c 6c65 6374 696f 6e2c 0a20 2020  eCollection,.   
+0000c3d0: 2023 2061 6e64 2075 7064 6174 6520 496e   # and update In
+0000c3e0: 6469 7669 6475 616c 7327 206e 6f64 6520  dividuals' node 
+0000c3f0: 4944 730a 2020 2020 2320 4e4f 5445 3a20  IDs.    # NOTE: 
+0000c400: 7261 7468 6572 2074 6861 6e20 6465 6475  rather than dedu
+0000c410: 706c 6963 6174 6520 7369 7465 7320 2861  plicate sites (a
+0000c420: 6e64 2073 6f6d 6574 696d 6573 0a20 2020  nd sometimes.   
+0000c430: 2023 2063 6f6d 7075 7465 2070 6172 656e   # compute paren
+0000c440: 7420 7369 7465 7329 2c20 6173 2072 6563  t sites), as rec
+0000c450: 6f6d 6d65 6e64 6564 2062 7920 7473 6b69  ommended by tski
+0000c460: 7427 7320 646f 6373 2c0a 2020 2020 2320  t's docs,.    # 
+0000c470: 7765 206f 6e6c 7920 6861 7665 2074 6f20  we only have to 
+0000c480: 736f 7274 2061 6e64 2073 696d 706c 6966  sort and simplif
+0000c490: 7920 6865 7265 2c20 6265 6361 7573 6520  y here, because 
+0000c4a0: 7765 2068 6176 650a 2020 2020 2320 6368  we have.    # ch
+0000c4b0: 6f73 656e 2074 6f20 6a75 7374 2061 6464  osen to just add
+0000c4c0: 2061 6c6c 2073 696d 756c 6174 6564 2073   all simulated s
+0000c4d0: 6974 6573 2074 6f20 7468 6520 7369 7465  ites to the site
+0000c4e0: 7320 7461 626c 6520 2869 6e0a 2020 2020  s table (in.    
+0000c4f0: 2320 5370 6563 6965 732e 5f73 6574 5f74  # Species._set_t
+0000c500: 6162 6c65 5f63 6f6c 6c65 6374 696f 6e29  able_collection)
+0000c510: 2061 7420 7468 6520 7369 6d75 6c61 7469   at the simulati
+0000c520: 6f6e 2773 206f 7574 7365 742c 0a20 2020  on's outset,.   
+0000c530: 2023 2061 6e64 2061 6c73 6f20 746f 206f   # and also to o
+0000c540: 6e6c 7920 616c 6c6f 7720 616e 2069 6e66  nly allow an inf
+0000c550: 696e 6974 652d 7369 7465 7320 6d6f 6465  inite-sites mode
+0000c560: 6c0a 2020 2020 2320 2846 4f52 204e 4f57  l.    # (FOR NOW
+0000c570: 2c20 616e 7968 6f77 2e2e 2e20 636f 756c  , anyhow... coul
+0000c580: 6420 7265 6c61 7820 7468 6573 6520 636f  d relax these co
+0000c590: 6e73 7472 6169 6e74 7320 6c61 7465 722c  nstraints later,
+0000c5a0: 2069 6620 6465 7369 7265 642c 0a20 2020   if desired,.   
+0000c5b0: 2023 2069 6e20 7768 6963 6820 6361 7365   # in which case
+0000c5c0: 2077 6f75 6c64 206e 6565 6420 746f 2072   would need to r
+0000c5d0: 6576 616d 7020 7468 6973 2061 7070 726f  evamp this appro
+0000c5e0: 6163 6829 0a20 2020 2064 6566 205f 736f  ach).    def _so
+0000c5f0: 7274 5f61 6e64 5f73 696d 706c 6966 795f  rt_and_simplify_
+0000c600: 7461 626c 655f 636f 6c6c 6563 7469 6f6e  table_collection
+0000c610: 2873 656c 662c 2063 6865 636b 5f6e 6f64  (self, check_nod
+0000c620: 6573 3d46 616c 7365 2c0a 2020 2020 2020  es=False,.      
+0000c630: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000c640: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000c650: 2020 6368 6563 6b5f 696e 6469 7669 6475    check_individu
+0000c660: 616c 733d 4661 6c73 6529 3a0a 2020 2020  als=False):.    
+0000c670: 2020 2020 2320 736f 7274 2074 6865 2054      # sort the T
+0000c680: 6162 6c65 436f 6c6c 6563 7469 6f6e 0a20  ableCollection. 
+0000c690: 2020 2020 2020 2073 656c 662e 5f74 632e         self._tc.
+0000c6a0: 736f 7274 2829 0a20 2020 2020 2020 2023  sort().        #
+0000c6b0: 2067 6574 2061 6e20 6172 7261 7920 6f66   get an array of
+0000c6c0: 2061 6c6c 2074 6865 2063 7572 7265 6e74   all the current
+0000c6d0: 2069 6e64 6976 6964 7561 6c73 2720 6e6f   individuals' no
+0000c6e0: 6465 732c 0a20 2020 2020 2020 2023 2074  des,.        # t
+0000c6f0: 6865 206e 6f64 6573 2066 6f72 2077 6869  he nodes for whi
+0000c700: 6368 2074 6865 2074 6162 6c65 7320 7769  ch the tables wi
+0000c710: 6c6c 2062 6520 7369 6d70 6c69 6669 6564  ll be simplified
+0000c720: 0a20 2020 2020 2020 2063 7572 725f 6e6f  .        curr_no
+0000c730: 6465 7320 3d20 7365 6c66 2e5f 6765 745f  des = self._get_
+0000c740: 6e6f 6465 7328 290a 0a20 2020 2020 2020  nodes()..       
+0000c750: 2023 2072 756e 2063 6f64 6520 6e65 6365   # run code nece
+0000c760: 7373 6172 7920 666f 7220 6368 6563 6b69  ssary for checki
+0000c770: 6e67 2074 6861 7420 696e 6469 7669 6475  ng that individu
+0000c780: 616c 7327 2074 6162 6c65 2069 6473 2061  als' table ids a
+0000c790: 7265 0a20 2020 2020 2020 2023 2063 6f72  re.        # cor
+0000c7a0: 7265 6374 6c79 2061 7373 6967 6e65 642c  rectly assigned,
+0000c7b0: 2069 6620 7265 7175 6573 7465 640a 2020   if requested.  
+0000c7c0: 2020 2020 2020 6966 2063 6865 636b 5f69        if check_i
+0000c7d0: 6e64 6976 6964 7561 6c73 3a0a 2020 2020  ndividuals:.    
+0000c7e0: 2020 2020 2320 4e4f 5445 3a20 7468 6973      # NOTE: this
+0000c7f0: 2063 6865 636b 2077 696c 6c20 6f6e 6c79   check will only
+0000c800: 2077 6f72 6b20 666f 7220 6d6f 6465 6c73   work for models
+0000c810: 2074 6861 7420 6861 7665 2061 7420 6c65   that have at le
+0000c820: 6173 7420 6f6e 650a 2020 2020 2020 2020  ast one.        
+0000c830: 2320 7472 6169 742c 2062 6563 6175 7365  # trait, because
+0000c840: 206f 6e6c 7920 7468 6f73 6520 6d6f 6465   only those mode
+0000c850: 6c73 2077 696c 6c20 6861 7665 2061 7420  ls will have at 
+0000c860: 6c65 6173 7420 3320 7661 6c75 6573 2069  least 3 values i
+0000c870: 6e20 7468 6520 0a20 2020 2020 2020 2023  n the .        #
+0000c880: 206c 6f63 6174 696f 6e20 636f 6c75 6d6e   location column
+0000c890: 2773 2076 616c 7565 2028 616e 6420 6f6e  's value (and on
+0000c8a0: 6c79 2074 6865 2076 616c 7565 7320 6166  ly the values af
+0000c8b0: 7465 7220 7468 6520 6669 7273 7420 3220  ter the first 2 
+0000c8c0: 7769 6c6c 0a20 2020 2020 2020 2023 2062  will.        # b
+0000c8d0: 6520 7468 6520 7361 6d65 2061 6372 6f73  e the same acros
+0000c8e0: 7320 616e 2069 6e64 6976 6964 7561 6c27  s an individual'
+0000c8f0: 7320 6c69 6665 7469 6d65 2c20 7369 6e63  s lifetime, sinc
+0000c900: 6520 7468 6579 206d 6f76 6520 706f 7369  e they move posi
+0000c910: 7469 6f6e 0a20 2020 2020 2020 2023 2066  tion.        # f
+0000c920: 7265 7175 656e 746c 7920 6275 7420 646f  requently but do
+0000c930: 206e 6f74 2063 6168 6e67 6520 7068 656e   not cahnge phen
+0000c940: 6f74 7970 6520 6174 2061 6c6c 290a 0a20  otype at all).. 
+0000c950: 2020 2020 2020 2020 2020 206d 6574 615f             meta_
+0000c960: 6234 203d 2073 656c 662e 5f74 632e 696e  b4 = self._tc.in
+0000c970: 6469 7669 6475 616c 732e 6d65 7461 6461  dividuals.metada
+0000c980: 7461 0a20 2020 2020 2020 2020 2020 206c  ta.            l
+0000c990: 6f63 203d 2073 656c 662e 5f74 632e 696e  oc = self._tc.in
+0000c9a0: 6469 7669 6475 616c 732e 6c6f 6361 7469  dividuals.locati
+0000c9b0: 6f6e 0a20 2020 2020 2020 2020 2020 206c  on.            l
+0000c9c0: 6f63 6f66 6620 3d20 7365 6c66 2e5f 7463  ocoff = self._tc
+0000c9d0: 2e69 6e64 6976 6964 7561 6c73 2e6c 6f63  .individuals.loc
+0000c9e0: 6174 696f 6e5f 6f66 6673 6574 0a20 2020  ation_offset.   
+0000c9f0: 2020 2020 2020 2020 206c 6f63 5f62 345f           loc_b4_
+0000ca00: 6279 5f74 6162 6c65 5f69 6473 203d 207b  by_table_ids = {
+0000ca10: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000ca20: 2074 6162 6c65 5f69 643a 2028 6c6f 635b   table_id: (loc[
+0000ca30: 6c6f 636f 6666 5b74 6162 6c65 5f69 645d  locoff[table_id]
+0000ca40: 3a20 6c6f 636f 6666 5b0a 2020 2020 2020  : locoff[.      
+0000ca50: 2020 2020 2020 2020 2020 2020 2020 7461                ta
+0000ca60: 626c 655f 6964 2b31 5d5d 2920 666f 7220  ble_id+1]]) for 
+0000ca70: 7461 626c 655f 6964 2069 6e20 7261 6e67  table_id in rang
+0000ca80: 6528 6c65 6e28 0a20 2020 2020 2020 2020  e(len(.         
+0000ca90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000caa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000cab0: 2020 2020 2020 2073 656c 662e 5f74 632e         self._tc.
+0000cac0: 696e 6469 7669 6475 616c 732e 666c 6167  individuals.flag
+0000cad0: 7329 297d 0a20 2020 2020 2020 2020 2020  s))}.           
+0000cae0: 207a 305f 6234 203d 207b 696e 642e 6964   z0_b4 = {ind.id
+0000caf0: 783a 206c 6f63 5f62 345f 6279 5f74 6162  x: loc_b4_by_tab
+0000cb00: 6c65 5f69 6473 5b0a 2020 2020 2020 2020  le_ids[.        
+0000cb10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000cb20: 696e 642e 5f69 6e64 6976 6964 7561 6c73  ind._individuals
+0000cb30: 5f74 6162 5f69 645d 5b32 5d20 666f 7220  _tab_id][2] for 
+0000cb40: 696e 6420 696e 2073 656c 662e 7661 6c75  ind in self.valu
+0000cb50: 6573 2829 7d0a 0a20 2020 2020 2020 2023  es()}..        #
+0000cb60: 206e 6f77 2073 696d 706c 6966 7920 7468   now simplify th
+0000cb70: 6520 7461 626c 6573 2061 6e64 2067 6574  e tables and get
+0000cb80: 2074 6865 206e 6577 2069 6473 206f 7574   the new ids out
+0000cb90: 7075 7420 0a20 2020 2020 2020 2023 2066  put .        # f
+0000cba0: 726f 6d20 7468 6520 7463 2e73 696d 706c  rom the tc.simpl
+0000cbb0: 6966 7920 6d65 7468 6f64 0a20 2020 2020  ify method.     
+0000cbc0: 2020 2023 204e 4f54 453a 2077 6520 6172     # NOTE: we ar
+0000cbd0: 6520 6e6f 7420 7573 696e 6720 7468 6520  e not using the 
+0000cbe0: 6f75 7470 7574 2c20 6275 7420 6974 2069  output, but it i
+0000cbf0: 7320 616e 2061 7272 6179 2077 6865 7265  s an array where
+0000cc00: 0a20 2020 2020 2020 2023 2065 6163 6820  .        # each 
+0000cc10: 7661 6c75 6520 6973 2074 6865 206e 6577  value is the new
+0000cc20: 206e 6f64 6520 4944 206f 6620 7468 6520   node ID of the 
+0000cc30: 6e6f 6465 2074 6861 7420 7761 7320 696e  node that was in
+0000cc40: 2074 6861 7420 696e 6465 7827 730a 2020   that index's.  
+0000cc50: 2020 2020 2020 2320 726f 7720 706f 7369        # row posi
+0000cc60: 7469 6f6e 2069 6e20 7468 6520 6f6c 6420  tion in the old 
+0000cc70: 6e6f 6465 7320 7461 626c 6520 2861 6e64  nodes table (and
+0000cc80: 202d 3120 6966 2074 6861 7420 6e6f 6465   -1 if that node
+0000cc90: 2077 6173 0a20 2020 2020 2020 2023 2064   was.        # d
+0000cca0: 726f 7070 6564 2064 7572 696e 6720 7369  ropped during si
+0000ccb0: 6d70 6c69 6361 7469 6f6e 293b 2073 6565  mplication); see
+0000ccc0: 2074 736b 6974 2064 6f63 7320 666f 7220   tskit docs for 
+0000ccd0: 6465 7461 696c 730a 2020 2020 2020 2020  details.        
+0000cce0: 6f75 7470 7574 203d 2073 656c 662e 5f74  output = self._t
+0000ccf0: 632e 7369 6d70 6c69 6679 2863 7572 725f  c.simplify(curr_
+0000cd00: 6e6f 6465 732c 0a20 2020 2020 2020 2020  nodes,.         
+0000cd10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000cd20: 2020 2020 2020 2020 2020 6669 6c74 6572            filter
+0000cd30: 5f69 6e64 6976 6964 7561 6c73 3d54 7275  _individuals=Tru
+0000cd40: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+0000cd50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000cd60: 2020 2020 2020 6669 6c74 6572 5f73 6974        filter_sit
+0000cd70: 6573 3d46 616c 7365 290a 0a20 2020 2020  es=False)..     
+0000cd80: 2020 2023 206d 616b 6520 616e 204e 7833     # make an Nx3
+0000cd90: 206e 7020 6172 7261 7920 636f 6e74 6169   np array contai
+0000cda0: 6e69 6e67 2031 2e29 2067 6e78 2069 6473  ning 1.) gnx ids
+0000cdb0: 2c20 322e 2920 7468 6520 6e65 770a 2020  , 2.) the new.  
+0000cdc0: 2020 2020 2020 2320 686f 6d6f 6c6f 6775        # homologu
+0000cdd0: 6520 3020 6e6f 6465 2069 6473 2c20 616e  e 0 node ids, an
+0000cde0: 6420 332e 2920 7468 6520 6e65 7720 686f  d 3.) the new ho
+0000cdf0: 6d6f 6c6f 6775 6520 3120 6e6f 6465 2069  mologue 1 node i
+0000ce00: 6473 2c0a 2020 2020 2020 2020 2320 696e  ds,.        # in
+0000ce10: 2065 6163 6820 6974 7320 3320 636f 6c73   each its 3 cols
+0000ce20: 3b20 7468 656e 206c 6f6f 7020 6f76 6572  ; then loop over
+0000ce30: 2069 7473 2072 6f77 7320 746f 2075 7064   its rows to upd
+0000ce40: 6174 650a 2020 2020 2020 2020 2320 6561  ate.        # ea
+0000ce50: 6368 2069 6e64 6976 6964 7561 6c27 7320  ch individual's 
+0000ce60: 5f6e 6f64 6573 5f74 6162 5f69 6473 2061  _nodes_tab_ids a
+0000ce70: 7474 7269 6275 7465 0a20 2020 2020 2020  ttribute.       
+0000ce80: 2069 6e64 735f 676e 785f 6964 7320 3d20   inds_gnx_ids = 
+0000ce90: 6e70 2e61 7272 6179 285b 2a73 656c 665d  np.array([*self]
+0000cea0: 292e 7265 7368 6170 6528 286c 656e 2873  ).reshape((len(s
+0000ceb0: 656c 6629 2c20 3129 290a 2020 2020 2020  elf), 1)).      
+0000cec0: 2020 6e65 775f 6964 7320 3d20 6e70 2e68    new_ids = np.h
+0000ced0: 7374 6163 6b28 2869 6e64 735f 676e 785f  stack((inds_gnx_
+0000cee0: 6964 732c 206e 702e 6172 7261 7928 5b2a  ids, np.array([*
+0000cef0: 7261 6e67 6528 6c65 6e28 0a20 2020 2020  range(len(.     
+0000cf00: 2020 2020 2020 2069 6e64 735f 676e 785f         inds_gnx_
+0000cf10: 6964 7329 202a 2032 295d 292e 7265 7368  ids) * 2)]).resh
+0000cf20: 6170 6528 286c 656e 2869 6e64 735f 676e  ape((len(inds_gn
+0000cf30: 785f 6964 7329 2c20 3229 2929 290a 2020  x_ids), 2)))).  
+0000cf40: 2020 2020 2020 666f 7220 6964 2c20 6e30        for id, n0
+0000cf50: 2c20 6e31 2069 6e20 6e65 775f 6964 733a  , n1 in new_ids:
+0000cf60: 0a20 2020 2020 2020 2020 2020 2073 656c  .            sel
+0000cf70: 665b 6964 5d2e 5f6e 6f64 6573 5f74 6162  f[id]._nodes_tab
+0000cf80: 5f69 6473 2e75 7064 6174 6528 7b30 3a6e  _ids.update({0:n
+0000cf90: 302c 2031 3a6e 317d 290a 0a20 2020 2020  0, 1:n1})..     
+0000cfa0: 2020 2023 2075 7064 6174 6520 496e 6469     # update Indi
+0000cfb0: 7669 6475 616c 7327 2074 6162 6c65 2069  viduals' table i
+0000cfc0: 6473 0a20 2020 2020 2020 2023 2028 692e  ds.        # (i.
+0000cfd0: 652e 2074 6865 6972 2049 6e64 6976 6964  e. their Individ
+0000cfe0: 7561 6c2e 5f69 6e64 6976 6964 7561 6c73  ual._individuals
+0000cff0: 5f74 6162 5f69 6420 6174 7472 6962 7574  _tab_id attribut
+0000d000: 6573 290a 2020 2020 2020 2020 696e 645f  es).        ind_
+0000d010: 6d65 7461 203d 2073 656c 662e 5f74 632e  meta = self._tc.
+0000d020: 696e 6469 7669 6475 616c 732e 6d65 7461  individuals.meta
+0000d030: 6461 7461 0a20 2020 2020 2020 2069 6620  data.        if 
+0000d040: 6368 6563 6b5f 696e 6469 7669 6475 616c  check_individual
+0000d050: 733a 0a20 2020 2020 2020 2020 2020 2070  s:.            p
+0000d060: 7269 6e74 2827 6234 272c 206c 656e 286d  rint('b4', len(m
+0000d070: 6574 615f 6234 292c 2027 6166 272c 206c  eta_b4), 'af', l
+0000d080: 656e 2869 6e64 5f6d 6574 6129 290a 2020  en(ind_meta)).  
+0000d090: 2020 2020 2020 696e 645f 6f66 6620 3d20        ind_off = 
+0000d0a0: 7365 6c66 2e5f 7463 2e69 6e64 6976 6964  self._tc.individ
+0000d0b0: 7561 6c73 2e6d 6574 6164 6174 615f 6f66  uals.metadata_of
+0000d0c0: 6673 6574 0a20 2020 2020 2020 206e 6577  fset.        new
+0000d0d0: 5f69 6e64 6976 6964 7561 6c73 5f74 6162  _individuals_tab
+0000d0e0: 5f69 6420 3d20 7b69 6e74 2e66 726f 6d5f  _id = {int.from_
+0000d0f0: 6279 7465 7328 0a20 2020 2020 2020 2020  bytes(.         
+0000d100: 2020 2069 6e64 5f6d 6574 615b 696e 645f     ind_meta[ind_
+0000d110: 6f66 665b 695d 3a20 696e 645f 6f66 665b  off[i]: ind_off[
+0000d120: 692b 315d 5d2e 746f 6279 7465 7328 292c  i+1]].tobytes(),
+0000d130: 0a20 2020 2020 2020 2020 2020 2027 6c69  .            'li
+0000d140: 7474 6c65 2729 3a20 6920 666f 7220 6920  ttle'): i for i 
+0000d150: 696e 2072 616e 6765 286c 656e 2869 6e64  in range(len(ind
+0000d160: 5f6f 6666 2920 2d20 3129 7d0a 2020 2020  _off) - 1)}.    
+0000d170: 2020 2020 5b73 6574 6174 7472 2869 6e64      [setattr(ind
+0000d180: 2c20 275f 696e 6469 7669 6475 616c 735f  , '_individuals_
+0000d190: 7461 625f 6964 272c 0a20 2020 2020 2020  tab_id',.       
+0000d1a0: 2020 2020 2020 2020 2020 6e65 775f 696e            new_in
+0000d1b0: 6469 7669 6475 616c 735f 7461 625f 6964  dividuals_tab_id
+0000d1c0: 5b69 645d 2920 666f 7220 6964 2c20 696e  [id]) for id, in
+0000d1d0: 6420 696e 2073 656c 662e 6974 656d 7328  d in self.items(
+0000d1e0: 295d 0a0a 2020 2020 2020 2020 2320 6368  )]..        # ch
+0000d1f0: 6563 6b20 7468 6174 2069 6e64 6976 6964  eck that individ
+0000d200: 7561 6c73 2720 6e6f 6465 732d 7461 626c  uals' nodes-tabl
+0000d210: 6520 6964 7320 7765 7265 2063 6f72 7265  e ids were corre
+0000d220: 636c 7479 2075 7064 6174 6564 2c0a 2020  clty updated,.  
+0000d230: 2020 2020 2020 2320 6966 2074 6865 2063        # if the c
+0000d240: 6865 636b 2069 7320 7265 7175 6573 7465  heck is requeste
+0000d250: 640a 2020 2020 2020 2020 6966 2063 6865  d.        if che
+0000d260: 636b 5f6e 6f64 6573 3a0a 2020 2020 2020  ck_nodes:.      
+0000d270: 2020 2020 2020 2363 7265 6174 6520 616e        #create an
+0000d280: 6f74 6865 7220 6964 656e 7469 6361 6c6c  other identicall
+0000d290: 7920 7374 7275 6374 7572 6564 204e 7833  y structured Nx3
+0000d2a0: 206e 7020 6172 7261 792c 2074 6f20 686f   np array, to ho
+0000d2b0: 6c64 2074 6865 0a20 2020 2020 2020 2020  ld the.         
+0000d2c0: 2020 2023 696e 6469 7669 6475 616c 7327     #individuals'
+0000d2d0: 2067 6e78 2069 6473 2061 6e64 2074 6865   gnx ids and the
+0000d2e0: 6972 206e 6f64 6520 6964 7320 6163 636f  ir node ids acco
+0000d2f0: 7264 696e 6720 746f 2074 6865 2074 736b  rding to the tsk
+0000d300: 6974 0a20 2020 2020 2020 2020 2020 2023  it.            #
+0000d310: 6e6f 6465 7320 7461 626c 6520 2866 6f72  nodes table (for
+0000d320: 2063 726f 7373 2d63 6865 636b 696e 6729   cross-checking)
+0000d330: 0a20 2020 2020 2020 2020 2020 206e 6577  .            new
+0000d340: 5f69 6473 5f66 726f 6d5f 7461 626c 6573  _ids_from_tables
+0000d350: 203d 206e 702e 6f6e 6573 2828 6c65 6e28   = np.ones((len(
+0000d360: 696e 6473 5f67 6e78 5f69 6473 292c 2033  inds_gnx_ids), 3
+0000d370: 2929 202a 206e 702e 6e61 6e0a 2020 2020  )) * np.nan.    
+0000d380: 2020 2020 2020 2020 6e65 775f 6964 735f          new_ids_
+0000d390: 6672 6f6d 5f74 6162 6c65 735b 3a2c 305d  from_tables[:,0]
+0000d3a0: 203d 2069 6e64 735f 676e 785f 6964 735b   = inds_gnx_ids[
+0000d3b0: 3a2c 305d 0a20 2020 2020 2020 2020 2020  :,0].           
+0000d3c0: 206e 6f64 6564 6620 3d20 7064 2e44 6174   nodedf = pd.Dat
+0000d3d0: 6146 7261 6d65 287b 6b3a 7620 666f 7220  aFrame({k:v for 
+0000d3e0: 6b2c 2076 2069 6e20 7365 6c66 2e5f 7463  k, v in self._tc
+0000d3f0: 2e6e 6f64 6573 2e61 7364 6963 7428 0a20  .nodes.asdict(. 
+0000d400: 2020 2020 2020 2020 2020 2020 2020 2029                 )
+0000d410: 2e69 7465 6d73 2829 2069 6620 6b20 696e  .items() if k in
+0000d420: 205b 2774 696d 6527 2c20 2769 6e64 6976   ['time', 'indiv
+0000d430: 6964 7561 6c27 5d7d 290a 2020 2020 2020  idual']}).      
+0000d440: 2020 2020 2020 696e 645f 7461 626c 655f        ind_table_
+0000d450: 6964 7320 3d20 5b73 656c 665b 6964 785b  ids = [self[idx[
+0000d460: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000d470: 2020 2020 2020 2020 2020 2020 2030 5d5d               0]]
+0000d480: 2e5f 696e 6469 7669 6475 616c 735f 7461  ._individuals_ta
+0000d490: 625f 6964 2066 6f72 2069 6478 2069 6e20  b_id for idx in 
+0000d4a0: 696e 6473 5f67 6e78 5f69 6473 5d0a 2020  inds_gnx_ids].  
+0000d4b0: 2020 2020 2020 2020 2020 2320 6c6f 6f70            # loop
+0000d4c0: 206f 7665 7220 696e 6469 7669 6475 616c   over individual
+0000d4d0: 7327 2067 6e78 2069 6473 2061 6e64 2063  s' gnx ids and c
+0000d4e0: 6f72 7265 7370 6f6e 6469 6e67 2074 6162  orresponding tab
+0000d4f0: 6c65 2072 6f77 2069 6473 2c0a 2020 2020  le row ids,.    
+0000d500: 2020 2020 2020 2020 2320 6669 6c6c 696e          # fillin
+0000d510: 6720 7570 2074 6865 2063 6f6d 7061 7269  g up the compari
+0000d520: 736f 6e20 7461 626c 650a 2020 2020 2020  son table.      
+0000d530: 2020 2020 2020 666f 7220 696e 642c 2069        for ind, i
+0000d540: 6478 2069 6e20 7a69 7028 696e 6473 5f67  dx in zip(inds_g
+0000d550: 6e78 5f69 6473 5b3a 2c30 5d2c 2069 6e64  nx_ids[:,0], ind
+0000d560: 5f74 6162 6c65 5f69 6473 293a 0a20 2020  _table_ids):.   
+0000d570: 2020 2020 2020 2020 2020 2020 2073 7562               sub
+0000d580: 6466 203d 206e 6f64 6564 665b 6e6f 6465  df = nodedf[node
+0000d590: 6466 5b27 696e 6469 7669 6475 616c 275d  df['individual']
+0000d5a0: 203d 3d20 6964 785d 0a20 2020 2020 2020   == idx].       
+0000d5b0: 2020 2020 2020 2020 2070 7269 6e74 2827           print('
+0000d5c0: 2e2e 2e5c 6e67 6e78 2069 643a 272c 2069  ...\ngnx id:', i
+0000d5d0: 6e64 2c20 275c 7474 6162 6c65 2069 643a  nd, '\ttable id:
+0000d5e0: 272c 2069 6478 290a 2020 2020 2020 2020  ', idx).        
+0000d5f0: 2020 2020 2020 2020 7072 696e 7428 2767          print('g
+0000d600: 6e78 206e 6f64 6520 6964 733a 2020 272c  nx node ids:  ',
+0000d610: 205b 2a73 656c 665b 696e 645d 2e5f 6e6f   [*self[ind]._no
+0000d620: 6465 735f 7461 625f 6964 732e 7661 6c75  des_tab_ids.valu
+0000d630: 6573 2829 5d29 0a20 2020 2020 2020 2020  es()]).         
+0000d640: 2020 2020 2020 2070 7269 6e74 2827 7461         print('ta
+0000d650: 626c 6520 6e6f 6465 2069 6473 3a27 2c20  ble node ids:', 
+0000d660: 5b2a 7375 6264 662e 696e 6465 785d 290a  [*subdf.index]).
+0000d670: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000d680: 6e6f 6465 5f69 645f 7661 6c73 203d 205b  node_id_vals = [
+0000d690: 2a73 7562 6466 2e69 6e64 6578 5d0a 2020  *subdf.index].  
+0000d6a0: 2020 2020 2020 2020 2020 2020 2020 726f                ro
+0000d6b0: 775f 6964 7820 3d20 6e70 2e77 6865 7265  w_idx = np.where
+0000d6c0: 286e 6577 5f69 6473 5f66 726f 6d5f 7461  (new_ids_from_ta
+0000d6d0: 626c 6573 5b3a 2c30 5d20 3d3d 2069 6e64  bles[:,0] == ind
+0000d6e0: 295b 305d 5b30 5d0a 2020 2020 2020 2020  )[0][0].        
+0000d6f0: 2020 2020 2020 2020 6e65 775f 6964 735f          new_ids_
+0000d700: 6672 6f6d 5f74 6162 6c65 735b 726f 775f  from_tables[row_
+0000d710: 6964 782c 2031 3a5d 203d 206e 6f64 655f  idx, 1:] = node_
+0000d720: 6964 5f76 616c 730a 2020 2020 2020 2020  id_vals.        
+0000d730: 2020 2020 2320 6e6f 7720 6d61 6b65 2073      # now make s
+0000d740: 7572 6520 7461 626c 7320 6172 6520 6964  ure tabls are id
+0000d750: 656e 7469 6361 6c0a 2020 2020 2020 2020  entical.        
+0000d760: 2020 2020 6173 7365 7274 206e 702e 616c      assert np.al
+0000d770: 6c28 6e65 775f 6964 735f 6672 6f6d 5f74  l(new_ids_from_t
+0000d780: 6162 6c65 7320 3d3d 206e 6577 5f69 6473  ables == new_ids
+0000d790: 292c 2028 2241 6c6c 2074 6865 206e 6f64  ), ("All the nod
+0000d7a0: 6520 220a 2020 2020 2020 2020 2020 2020  e ".            
+0000d7b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000d7c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000d7d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000d7e0: 2269 6473 2064 6f6e 2774 206d 6174 6368  "ids don't match
+0000d7f0: 2122 290a 2020 2020 2020 2020 2020 2020  !").            
+0000d800: 7072 696e 7428 2741 4c4c 204e 4f44 4520  print('ALL NODE 
+0000d810: 4944 5320 4150 5045 4152 2043 4f52 5245  IDS APPEAR CORRE
+0000d820: 4354 4c59 2052 4541 5353 4947 4e45 445c  CTLY REASSIGNED\
+0000d830: 6e27 290a 0a20 2020 2020 2020 2023 2063  n')..        # c
+0000d840: 6865 636b 2074 6861 7420 696e 6469 7669  heck that indivi
+0000d850: 6475 616c 7327 2069 6e64 6976 6964 7561  duals' individua
+0000d860: 6c73 2d74 6162 6c65 2069 6473 2077 6572  ls-table ids wer
+0000d870: 6520 636f 7272 6563 6c74 7920 7570 6461  e correclty upda
+0000d880: 7465 642c 0a20 2020 2020 2020 2023 2069  ted,.        # i
+0000d890: 6620 7468 6520 6368 6563 6b20 6973 2072  f the check is r
+0000d8a0: 6571 7565 7374 6564 0a20 2020 2020 2020  equested.       
+0000d8b0: 2069 6620 6368 6563 6b5f 696e 6469 7669   if check_indivi
+0000d8c0: 6475 616c 733a 0a20 2020 2020 2020 2020  duals:.         
+0000d8d0: 2020 2023 2063 6865 636b 2074 6861 7420     # check that 
+0000d8e0: 696e 6469 7669 6475 616c 7327 2074 6162  individuals' tab
+0000d8f0: 6c65 2069 6473 2077 6572 6520 636f 7272  le ids were corr
+0000d900: 6563 746c 7920 7265 6173 7369 676e 6564  ectly reassigned
+0000d910: 2c20 6279 0a20 2020 2020 2020 2020 2020  , by.           
+0000d920: 2023 2077 6179 206f 6620 6173 7375 7269   # way of assuri
+0000d930: 6e67 2074 6861 7420 696e 6469 7669 6475  ng that individu
+0000d940: 616c 732c 2061 7320 696e 6465 7865 6420  als, as indexed 
+0000d950: 6279 2062 6f74 6820 7468 6569 7220 6f6c  by both their ol
+0000d960: 640a 2020 2020 2020 2020 2020 2020 2320  d.            # 
+0000d970: 616e 6420 7468 6569 7220 6e65 7720 696e  and their new in
+0000d980: 6469 7669 6475 616c 732d 7461 626c 6520  dividuals-table 
+0000d990: 6964 732c 2068 6176 6520 6964 656e 7469  ids, have identi
+0000d9a0: 6361 6c20 7068 656e 6f74 7970 6573 0a20  cal phenotypes. 
+0000d9b0: 2020 2020 2020 2020 2020 206c 6f63 203d             loc =
+0000d9c0: 2073 656c 662e 5f74 632e 696e 6469 7669   self._tc.indivi
+0000d9d0: 6475 616c 732e 6c6f 6361 7469 6f6e 0a20  duals.location. 
+0000d9e0: 2020 2020 2020 2020 2020 206c 6f63 6f66             locof
+0000d9f0: 6620 3d20 7365 6c66 2e5f 7463 2e69 6e64  f = self._tc.ind
+0000da00: 6976 6964 7561 6c73 2e6c 6f63 6174 696f  ividuals.locatio
+0000da10: 6e5f 6f66 6673 6574 0a20 2020 2020 2020  n_offset.       
+0000da20: 2020 2020 206c 6f63 5f61 665f 6279 5f74       loc_af_by_t
+0000da30: 6162 6c65 5f69 6473 203d 207b 0a20 2020  able_ids = {.   
+0000da40: 2020 2020 2020 2020 2020 2020 2074 6162               tab
+0000da50: 6c65 5f69 643a 2028 6c6f 635b 6c6f 636f  le_id: (loc[loco
+0000da60: 6666 5b74 6162 6c65 5f69 645d 3a20 6c6f  ff[table_id]: lo
+0000da70: 636f 6666 5b0a 2020 2020 2020 2020 2020  coff[.          
+0000da80: 2020 2020 2020 2020 2020 7461 626c 655f            table_
+0000da90: 6964 2b31 5d5d 2920 666f 7220 7461 626c  id+1]]) for tabl
+0000daa0: 655f 6964 2069 6e20 7261 6e67 6528 6c65  e_id in range(le
+0000dab0: 6e28 0a20 2020 2020 2020 2020 2020 2020  n(.             
+0000dac0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000dad0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000dae0: 2020 2073 656c 662e 5f74 632e 696e 6469     self._tc.indi
+0000daf0: 7669 6475 616c 732e 666c 6167 7329 297d  viduals.flags))}
+0000db00: 0a20 2020 2020 2020 2020 2020 207a 305f  .            z0_
+0000db10: 6166 203d 207b 696e 642e 6964 783a 206c  af = {ind.idx: l
+0000db20: 6f63 5f61 665f 6279 5f74 6162 6c65 5f69  oc_af_by_table_i
+0000db30: 6473 5b0a 2020 2020 2020 2020 2020 2020  ds[.            
+0000db40: 2020 2020 2020 2020 2020 2020 696e 642e              ind.
+0000db50: 5f69 6e64 6976 6964 7561 6c73 5f74 6162  _individuals_tab
+0000db60: 5f69 645d 5b32 5d20 666f 7220 696e 6420  _id][2] for ind 
+0000db70: 696e 2073 656c 662e 7661 6c75 6573 2829  in self.values()
+0000db80: 7d0a 2020 2020 2020 2020 2020 2020 7072  }.            pr
+0000db90: 696e 7428 7a30 5f61 6629 0a20 2020 2020  int(z0_af).     
+0000dba0: 2020 2020 2020 207a 305f 6368 6563 6b20         z0_check 
+0000dbb0: 3d20 5b7a 3020 3d3d 207a 305f 6234 5b69  = [z0 == z0_b4[i
+0000dbc0: 6478 5d20 666f 7220 6964 782c 207a 3020  dx] for idx, z0 
+0000dbd0: 696e 207a 305f 6166 2e69 7465 6d73 2829  in z0_af.items()
+0000dbe0: 5d0a 2020 2020 2020 2020 2020 2020 666f  ].            fo
+0000dbf0: 7220 6964 7820 696e 205b 2a73 656c 665d  r idx in [*self]
+0000dc00: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+0000dc10: 2020 7072 696e 7428 6964 782c 2027 2062    print(idx, ' b
+0000dc20: 3420 272c 207a 305f 6234 5b69 6478 5d29  4 ', z0_b4[idx])
+0000dc30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000dc40: 2070 7269 6e74 2869 6478 2c20 2720 6166   print(idx, ' af
+0000dc50: 2027 2c20 7a30 5f61 665b 6964 785d 290a   ', z0_af[idx]).
+0000dc60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000dc70: 7072 696e 7428 272e 2e2e 2e2e 2e2e 2e2e  print('.........
+0000dc80: 2e2e 2e2e 2e2e 2e2e 2e27 290a 2020 2020  .........').    
+0000dc90: 2020 2020 2020 2020 6173 7365 7274 206e          assert n
+0000dca0: 702e 616c 6c28 7a30 5f63 6865 636b 292c  p.all(z0_check),
+0000dcb0: 2028 2770 6865 6e6f 7479 7065 7320 6172   ('phenotypes ar
+0000dcc0: 6520 6e6f 7420 7468 6520 7361 6d65 2120  e not the same! 
+0000dcd0: 270a 2020 2020 2020 2020 2020 2020 2020  '.              
+0000dce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000dcf0: 2020 2020 2020 2020 2720 284d 414b 4520          ' (MAKE 
+0000dd00: 5355 5245 2059 4f55 2055 5345 4420 4120  SURE YOU USED A 
+0000dd10: 4d4f 4445 4c20 5749 5448 2027 0a20 2020  MODEL WITH '.   
 0000dd20: 2020 2020 2020 2020 2020 2020 2020 2020                  
 0000dd30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000dd40: 2020 2020 2020 2020 2020 2020 2020 2029                 )
-0000dd50: 5d20 666f 7220 6920 696e 2069 6e64 6976  ] for i in indiv
-0000dd60: 6964 735d 290a 2020 2020 2020 2020 7265  ids]).        re
-0000dd70: 7475 726e 206e 6f64 6573 0a0a 0a20 2020  turn nodes...   
-0000dd80: 2023 2323 2323 2323 2323 2323 2323 2323   ###############
-0000dd90: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000dda0: 2323 2323 2323 2323 2323 230a 2020 2020  ###########.    
-0000ddb0: 2320 4655 4e43 5449 4f4e 5320 4652 4f4d  # FUNCTIONS FROM
-0000ddc0: 2054 5241 434b 5f53 5041 5449 414c 5f50   TRACK_SPATIAL_P
-0000ddd0: 4544 4947 5245 452e 5059 0a20 2020 2023  EDIGREE.PY.    #
-0000dde0: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000ddf0: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000de00: 2323 2323 2323 2323 230a 0a20 2020 2023  #########..    #
-0000de10: 2066 6f72 2061 2073 616d 706c 6520 6f66   for a sample of
-0000de20: 206e 6f64 6573 2061 6e64 2061 2073 616d   nodes and a sam
-0000de30: 706c 6520 6f66 206c 6f63 692c 2067 6574  ple of loci, get
-0000de40: 2074 6865 0a20 2020 2023 206c 6f63 692d   the.    # loci-
-0000de50: 7365 7175 656e 7469 616c 6c79 206f 7264  sequentially ord
-0000de60: 6572 6564 2064 6963 7420 6f66 206c 696e  ered dict of lin
-0000de70: 6561 6765 2064 6963 7473 0a20 2020 2023  eage dicts.    #
-0000de80: 2028 636f 6e74 6169 6e69 6e67 2065 6163   (containing eac
-0000de90: 6820 6e6f 6465 2069 6e20 6120 6368 696c  h node in a chil
-0000dea0: 6420 6e6f 6465 2773 206c 696e 6561 6765  d node's lineage
-0000deb0: 2061 7320 7468 650a 2020 2020 2320 6b65   as the.    # ke
-0000dec0: 7920 616e 6420 6974 7320 6269 7274 682d  y and its birth-
-0000ded0: 7469 6d65 2061 6e64 2062 6972 7468 2d6c  time and birth-l
-0000dee0: 6f63 6174 696f 6e20 6173 2061 206c 656e  ocation as a len
-0000def0: 6774 682d 3220 6c69 7374 206f 6620 7661  gth-2 list of va
-0000df00: 6c75 6573 0a20 2020 2064 6566 205f 6765  lues.    def _ge
-0000df10: 745f 6c69 6e65 6167 655f 6469 6374 7328  t_lineage_dicts(
-0000df20: 7365 6c66 2c20 6c6f 6369 2c20 6e6f 6465  self, loci, node
-0000df30: 733d 4e6f 6e65 2c20 6472 6f70 5f62 6566  s=None, drop_bef
-0000df40: 6f72 655f 7369 6d3d 5472 7565 2c0a 2020  ore_sim=True,.  
-0000df50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000df60: 2020 2020 2020 2020 2074 696d 655f 6265           time_be
-0000df70: 666f 7265 5f70 7265 7365 6e74 3d54 7275  fore_present=Tru
-0000df80: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-0000df90: 2020 2020 2020 2020 2020 2020 2020 7573                us
-0000dfa0: 655f 696e 6469 7669 6473 5f63 7572 725f  e_individs_curr_
-0000dfb0: 706f 733d 5472 7565 2c0a 2020 2020 2020  pos=True,.      
-0000dfc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000dfd0: 2020 2020 206d 6178 5f74 696d 655f 6167       max_time_ag
-0000dfe0: 6f3d 4e6f 6e65 2c0a 2020 2020 2020 2020  o=None,.        
-0000dff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e000: 2020 206d 696e 5f74 696d 655f 6167 6f3d     min_time_ago=
-0000e010: 4e6f 6e65 293a 0a20 2020 2020 2020 2069  None):.        i
-0000e020: 6620 6e6f 6465 7320 6973 204e 6f6e 653a  f nodes is None:
-0000e030: 0a20 2020 2020 2020 2020 2020 2023 736f  .            #so
-0000e040: 7274 2061 6e64 2073 696d 706c 6966 7920  rt and simplify 
-0000e050: 7468 6520 5461 626c 6543 6f6c 6c65 6374  the TableCollect
-0000e060: 696f 6e2c 2069 6620 6e65 6564 6564 0a20  ion, if needed. 
-0000e070: 2020 2020 2020 2020 2020 2069 6620 6e6f             if no
-0000e080: 7420 7365 6c66 2e5f 7463 5f73 6f72 7465  t self._tc_sorte
-0000e090: 645f 616e 645f 7369 6d70 6c69 6669 6564  d_and_simplified
-0000e0a0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-0000e0b0: 2020 7365 6c66 2e5f 736f 7274 5f73 696d    self._sort_sim
-0000e0c0: 706c 6966 795f 7461 626c 655f 636f 6c6c  plify_table_coll
-0000e0d0: 6563 7469 6f6e 2829 0a20 2020 2020 2020  ection().       
-0000e0e0: 2020 2020 206e 6f64 6573 203d 2073 656c       nodes = sel
-0000e0f0: 662e 5f67 6574 5f6e 6f64 6573 2829 0a20  f._get_nodes(). 
-0000e100: 2020 2020 2020 206c 696e 5f64 6963 7473         lin_dicts
-0000e110: 203d 205f 6765 745f 6c69 6e65 6167 655f   = _get_lineage_
-0000e120: 6469 6374 7328 7365 6c66 2c20 6e6f 6465  dicts(self, node
-0000e130: 732c 206c 6f63 692c 2074 5f63 7572 723d  s, loci, t_curr=
-0000e140: 7365 6c66 2e74 2c0a 2020 2020 2020 2020  self.t,.        
-0000e150: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e160: 2020 2020 2020 2020 2020 2020 2020 2064                 d
-0000e170: 726f 705f 6265 666f 7265 5f73 696d 3d64  rop_before_sim=d
-0000e180: 726f 705f 6265 666f 7265 5f73 696d 2c0a  rop_before_sim,.
-0000e190: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e1a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e1b0: 2020 2020 2020 2074 696d 655f 6265 666f         time_befo
-0000e1c0: 7265 5f70 7265 7365 6e74 3d74 696d 655f  re_present=time_
-0000e1d0: 6265 666f 7265 5f70 7265 7365 6e74 2c0a  before_present,.
-0000e1e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e1f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e200: 2020 2020 2020 206d 6178 5f74 696d 655f         max_time_
-0000e210: 6167 6f3d 6d61 785f 7469 6d65 5f61 676f  ago=max_time_ago
-0000e220: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-0000e230: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e240: 2020 2020 2020 2020 206d 696e 5f74 696d           min_tim
-0000e250: 655f 6167 6f3d 6d69 6e5f 7469 6d65 5f61  e_ago=min_time_a
-0000e260: 676f 290a 2020 2020 2020 2020 2320 6966  go).        # if
-0000e270: 2072 6571 7565 7374 6564 2c20 7075 7420   requested, put 
-0000e280: 7468 6520 6375 7272 656e 7420 696e 6469  the current indi
-0000e290: 7669 6475 616c 7327 2063 7572 7265 6e74  viduals' current
-0000e2a0: 200a 2020 2020 2020 2020 2320 706f 7369   .        # posi
-0000e2b0: 7469 6f6e 7320 696e 2074 6865 206c 696e  tions in the lin
-0000e2c0: 6561 6765 2064 6963 7473 2c20 7261 7468  eage dicts, rath
-0000e2d0: 6572 2074 6861 6e20 7573 696e 6720 7468  er than using th
-0000e2e0: 6520 6269 7274 680a 2020 2020 2020 2020  e birth.        
-0000e2f0: 2320 706f 7369 7469 6f6e 7320 2877 6869  # positions (whi
-0000e300: 6368 2061 7265 2069 6e20 7468 6572 6520  ch are in there 
-0000e310: 6279 2064 6566 6175 6c74 2062 6563 6175  by default becau
-0000e320: 7365 2074 686f 7365 0a20 2020 2020 2020  se those.       
-0000e330: 2023 2061 7265 2074 6865 2070 6f73 6974   # are the posit
-0000e340: 696f 6e73 2074 6861 7420 6172 6520 7374  ions that are st
-0000e350: 6f72 6564 2077 6974 6869 6e20 7468 6520  ored within the 
-0000e360: 7473 6b69 7420 7461 626c 6573 290a 2020  tskit tables).  
-0000e370: 2020 2020 2020 6966 2075 7365 5f69 6e64        if use_ind
-0000e380: 6976 6964 735f 6375 7272 5f70 6f73 3a0a  ivids_curr_pos:.
-0000e390: 2020 2020 2020 2020 2020 2020 666f 7220              for 
-0000e3a0: 6c6f 6375 7320 696e 206c 696e 5f64 6963  locus in lin_dic
-0000e3b0: 7473 2e6b 6579 7328 293a 0a20 2020 2020  ts.keys():.     
-0000e3c0: 2020 2020 2020 2020 2020 2066 6f72 2063             for c
-0000e3d0: 7572 725f 6e6f 6465 5f69 6420 696e 206c  urr_node_id in l
-0000e3e0: 696e 5f64 6963 7473 5b6c 6f63 7573 5d2e  in_dicts[locus].
-0000e3f0: 6b65 7973 2829 3a0a 2020 2020 2020 2020  keys():.        
-0000e400: 2020 2020 2020 2020 2020 2020 6966 206c              if l
-0000e410: 656e 286c 696e 5f64 6963 7473 5b6c 6f63  en(lin_dicts[loc
-0000e420: 7573 5d5b 6375 7272 5f6e 6f64 655f 6964  us][curr_node_id
-0000e430: 5d29 203e 2030 3a0a 2020 2020 2020 2020  ]) > 0:.        
-0000e440: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e450: 7463 5f69 6e64 6976 6964 735f 6964 203d  tc_individs_id =
-0000e460: 2073 656c 662e 5f74 632e 6e6f 6465 735b   self._tc.nodes[
-0000e470: 6375 7272 5f6e 6f64 655f 6964 5d2e 696e  curr_node_id].in
-0000e480: 6469 7669 6475 616c 0a20 2020 2020 2020  dividual.       
-0000e490: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e4a0: 2069 6e64 6976 6964 203d 205b 696e 6420   individ = [ind 
-0000e4b0: 666f 7220 696e 6420 696e 2073 656c 662e  for ind in self.
-0000e4c0: 7661 6c75 6573 2829 2069 660a 2020 2020  values() if.    
-0000e4d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e4e0: 2020 2020 2020 2020 2020 2020 2020 2069                 i
-0000e4f0: 6e64 2e5f 696e 6469 7669 6475 616c 735f  nd._individuals_
-0000e500: 7461 625f 6964 203d 3d20 7463 5f69 6e64  tab_id == tc_ind
-0000e510: 6976 6964 735f 6964 5d0a 2020 2020 2020  ivids_id].      
-0000e520: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e530: 2020 6173 7365 7274 206c 656e 2869 6e64    assert len(ind
-0000e540: 6976 6964 2920 3d3d 2031 2c20 2822 466f  ivid) == 1, ("Fo
-0000e550: 756e 6420 6d75 6c74 6970 6c65 2069 6e64  und multiple ind
-0000e560: 6976 6964 7561 6c73 2022 0a20 2020 2020  ividuals ".     
-0000e570: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e580: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e590: 2020 2020 2020 2020 2020 2020 2020 2277                "w
-0000e5a0: 6974 6820 7468 6520 7361 6d65 2069 6e64  ith the same ind
-0000e5b0: 6976 6964 7561 6c73 2022 0a20 2020 2020  ividuals ".     
-0000e5c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e5d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e5e0: 2020 2020 2020 2020 2020 2020 2020 2269                "i
-0000e5f0: 6420 6e20 7468 6520 7473 6b69 7420 7461  d n the tskit ta
-0000e600: 626c 6573 2e22 290a 2020 2020 2020 2020  bles.").        
-0000e610: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e620: 696e 6469 7669 6420 3d20 696e 6469 7669  individ = indivi
-0000e630: 645b 305d 0a20 2020 2020 2020 2020 2020  d[0].           
-0000e640: 2020 2020 2020 2020 2020 2020 2063 7572               cur
-0000e650: 725f 706f 7320 3d20 6e70 2e61 7272 6179  r_pos = np.array
-0000e660: 285b 696e 6469 7669 642e 782c 2069 6e64  ([individ.x, ind
-0000e670: 6976 6964 2e79 5d29 0a20 2020 2020 2020  ivid.y]).       
-0000e680: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e690: 2062 6972 7468 5f74 203d 6c69 6e5f 6469   birth_t =lin_di
-0000e6a0: 6374 735b 6c6f 6375 735d 5b63 7572 725f  cts[locus][curr_
-0000e6b0: 6e6f 6465 5f69 645d 5b63 7572 725f 6e6f  node_id][curr_no
-0000e6c0: 6465 5f69 645d 5b30 5d0a 2020 2020 2020  de_id][0].      
-0000e6d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e6e0: 2020 6c69 6e5f 6469 6374 735b 6c6f 6375    lin_dicts[locu
-0000e6f0: 735d 5b63 7572 725f 6e6f 6465 5f69 645d  s][curr_node_id]
-0000e700: 5b63 7572 725f 6e6f 6465 5f69 645d 203d  [curr_node_id] =
-0000e710: 2028 6269 7274 685f 742c 0a20 2020 2020   (birth_t,.     
-0000e720: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e730: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e740: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000dd40: 2020 2027 2041 5420 4c45 4153 5420 4f4e     ' AT LEAST ON
+0000dd50: 4520 5452 4149 5421 2927 290a 2020 2020  E TRAIT!)').    
+0000dd60: 2020 2020 2020 2020 7072 696e 7428 2741          print('A
+0000dd70: 4c4c 2049 4e44 4956 4944 5541 4c20 4944  LL INDIVIDUAL ID
+0000dd80: 5320 4150 5045 4152 2043 4f52 5245 4354  S APPEAR CORRECT
+0000dd90: 4c59 2052 4541 5353 4947 4e45 445c 6e27  LY REASSIGNED\n'
+0000dda0: 290a 0a20 2020 2020 2020 2023 2073 6574  )..        # set
+0000ddb0: 2074 6865 2073 6f72 7465 645f 616e 645f   the sorted_and_
+0000ddc0: 7369 6d70 6c69 6669 6564 2066 6c61 6720  simplified flag 
+0000ddd0: 746f 2054 7275 650a 2020 2020 2020 2020  to True.        
+0000dde0: 7365 6c66 2e5f 7463 5f73 6f72 7465 645f  self._tc_sorted_
+0000ddf0: 616e 645f 7369 6d70 6c69 6669 6564 203d  and_simplified =
+0000de00: 2054 7275 650a 0a0a 2020 2020 2320 6765   True...    # ge
+0000de10: 7420 7468 6520 6e6f 6465 732d 7461 626c  t the nodes-tabl
+0000de20: 6520 4944 7320 666f 7220 616c 6c20 696e  e IDs for all in
+0000de30: 6469 7669 6475 616c 730a 2020 2020 6465  dividuals.    de
+0000de40: 6620 5f67 6574 5f6e 6f64 6573 2873 656c  f _get_nodes(sel
+0000de50: 662c 2069 6e64 6976 6964 733d 4e6f 6e65  f, individs=None
+0000de60: 293a 0a20 2020 2020 2020 2023 2068 616e  ):.        # han
+0000de70: 646c 6520 7468 6520 696e 7374 616e 6365  dle the instance
+0000de80: 2077 6865 7265 2074 6865 2070 6f70 756c   where the popul
+0000de90: 6174 696f 6e20 6861 7320 7369 7a65 2030  ation has size 0
+0000dea0: 0a20 2020 2020 2020 2023 2028 652e 672e  .        # (e.g.
+0000deb0: 2c20 6166 7465 7220 7275 6e6e 696e 6720  , after running 
+0000dec0: 7370 702e 5f72 656d 6f76 655f 696e 6469  spp._remove_indi
+0000ded0: 7669 6475 616c 7328 2920 746f 2067 6574  viduals() to get
+0000dee0: 2072 6964 206f 6620 616c 6c0a 2020 2020   rid of all.    
+0000def0: 2020 2020 2320 696e 6469 7669 6475 616c      # individual
+0000df00: 7320 6265 666f 7265 2074 6865 6e20 7265  s before then re
+0000df10: 706c 6163 696e 6720 7468 656d 2077 6974  placing them wit
+0000df20: 6820 696e 6469 7669 6475 616c 7320 6672  h individuals fr
+0000df30: 6f6d 2061 2073 6f75 7263 650a 2020 2020  om a source.    
+0000df40: 2020 2020 2320 706f 7075 6c61 7469 6f6e      # population
+0000df50: 290a 2020 2020 2020 2020 6966 206c 656e  ).        if len
+0000df60: 2873 656c 6629 203d 3d20 303a 0a20 2020  (self) == 0:.   
+0000df70: 2020 2020 2020 2020 2072 6574 7572 6e20           return 
+0000df80: 6e70 2e61 7272 6179 285b 5d29 0a20 2020  np.array([]).   
+0000df90: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
+0000dfa0: 2020 2020 2020 2069 6620 696e 6469 7669         if indivi
+0000dfb0: 6473 2069 7320 4e6f 6e65 3a0a 2020 2020  ds is None:.    
+0000dfc0: 2020 2020 2020 2020 2020 2020 696e 6469              indi
+0000dfd0: 7669 6473 203d 205b 2a73 656c 665d 0a20  vids = [*self]. 
+0000dfe0: 2020 2020 2020 2020 2020 206e 6f64 6573             nodes
+0000dff0: 203d 206e 702e 6873 7461 636b 285b 5b2a   = np.hstack([[*
+0000e000: 7365 6c66 5b69 5d2e 5f6e 6f64 6573 5f74  self[i]._nodes_t
+0000e010: 6162 5f69 6473 2e76 616c 7565 7328 0a20  ab_ids.values(. 
+0000e020: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e030: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e040: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e050: 2020 2020 2020 2029 5d20 666f 7220 6920         )] for i 
+0000e060: 696e 2069 6e64 6976 6964 735d 290a 2020  in individs]).  
+0000e070: 2020 2020 2020 2020 2020 7265 7475 726e            return
+0000e080: 206e 6f64 6573 0a0a 0a20 2020 2023 2066   nodes...    # f
+0000e090: 6f72 2061 2073 616d 706c 6520 6f66 206e  or a sample of n
+0000e0a0: 6f64 6573 2061 6e64 2061 2073 616d 706c  odes and a sampl
+0000e0b0: 6520 6f66 206c 6f63 692c 2067 6574 2074  e of loci, get t
+0000e0c0: 6865 0a20 2020 2023 206c 6f63 692d 7365  he.    # loci-se
+0000e0d0: 7175 656e 7469 616c 6c79 206f 7264 6572  quentially order
+0000e0e0: 6564 2064 6963 7420 6f66 206c 696e 6561  ed dict of linea
+0000e0f0: 6765 2064 6963 7473 0a20 2020 2023 2028  ge dicts.    # (
+0000e100: 636f 6e74 6169 6e69 6e67 2065 6163 6820  containing each 
+0000e110: 6e6f 6465 2069 6e20 6120 6368 696c 6420  node in a child 
+0000e120: 6e6f 6465 2773 206c 696e 6561 6765 2061  node's lineage a
+0000e130: 7320 7468 650a 2020 2020 2320 6b65 7920  s the.    # key 
+0000e140: 616e 6420 6974 7320 6269 7274 682d 7469  and its birth-ti
+0000e150: 6d65 2061 6e64 2062 6972 7468 2d6c 6f63  me and birth-loc
+0000e160: 6174 696f 6e20 6173 2061 206c 656e 6774  ation as a lengt
+0000e170: 682d 3220 6c69 7374 206f 6620 7661 6c75  h-2 list of valu
+0000e180: 6573 0a20 2020 2064 6566 205f 6765 745f  es.    def _get_
+0000e190: 6c69 6e65 6167 655f 6469 6374 7328 7365  lineage_dicts(se
+0000e1a0: 6c66 2c20 6c6f 6369 2c20 6e6f 6465 733d  lf, loci, nodes=
+0000e1b0: 4e6f 6e65 2c20 6472 6f70 5f62 6566 6f72  None, drop_befor
+0000e1c0: 655f 7369 6d3d 5472 7565 2c0a 2020 2020  e_sim=True,.    
+0000e1d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e1e0: 2020 2020 2020 2074 696d 655f 6265 666f         time_befo
+0000e1f0: 7265 5f70 7265 7365 6e74 3d54 7275 652c  re_present=True,
+0000e200: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000e210: 2020 2020 2020 2020 2020 2020 7573 655f              use_
+0000e220: 696e 6469 7669 6473 5f63 7572 725f 706f  individs_curr_po
+0000e230: 733d 5472 7565 2c0a 2020 2020 2020 2020  s=True,.        
+0000e240: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e250: 2020 206d 6178 5f74 696d 655f 6167 6f3d     max_time_ago=
+0000e260: 4e6f 6e65 2c0a 2020 2020 2020 2020 2020  None,.          
+0000e270: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e280: 206d 696e 5f74 696d 655f 6167 6f3d 4e6f   min_time_ago=No
+0000e290: 6e65 293a 0a20 2020 2020 2020 2069 6620  ne):.        if 
+0000e2a0: 6e6f 6465 7320 6973 204e 6f6e 653a 0a20  nodes is None:. 
+0000e2b0: 2020 2020 2020 2020 2020 2023 736f 7274             #sort
+0000e2c0: 2061 6e64 2073 696d 706c 6966 7920 7468   and simplify th
+0000e2d0: 6520 5461 626c 6543 6f6c 6c65 6374 696f  e TableCollectio
+0000e2e0: 6e2c 2069 6620 6e65 6564 6564 0a20 2020  n, if needed.   
+0000e2f0: 2020 2020 2020 2020 2069 6620 6e6f 7420           if not 
+0000e300: 7365 6c66 2e5f 7463 5f73 6f72 7465 645f  self._tc_sorted_
+0000e310: 616e 645f 7369 6d70 6c69 6669 6564 3a0a  and_simplified:.
+0000e320: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e330: 7365 6c66 2e5f 736f 7274 5f61 6e64 5f73  self._sort_and_s
+0000e340: 696d 706c 6966 795f 7461 626c 655f 636f  implify_table_co
+0000e350: 6c6c 6563 7469 6f6e 2829 0a20 2020 2020  llection().     
+0000e360: 2020 2020 2020 206e 6f64 6573 203d 2073         nodes = s
+0000e370: 656c 662e 5f67 6574 5f6e 6f64 6573 2829  elf._get_nodes()
+0000e380: 0a20 2020 2020 2020 206c 696e 5f64 6963  .        lin_dic
+0000e390: 7473 203d 205f 6765 745f 6c69 6e65 6167  ts = _get_lineag
+0000e3a0: 655f 6469 6374 7328 7365 6c66 2c20 6e6f  e_dicts(self, no
+0000e3b0: 6465 732c 206c 6f63 692c 2074 5f63 7572  des, loci, t_cur
+0000e3c0: 723d 7365 6c66 2e74 2c0a 2020 2020 2020  r=self.t,.      
+0000e3d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e3e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e3f0: 2064 726f 705f 6265 666f 7265 5f73 696d   drop_before_sim
+0000e400: 3d64 726f 705f 6265 666f 7265 5f73 696d  =drop_before_sim
+0000e410: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0000e420: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e430: 2020 2020 2020 2020 2074 696d 655f 6265           time_be
+0000e440: 666f 7265 5f70 7265 7365 6e74 3d74 696d  fore_present=tim
+0000e450: 655f 6265 666f 7265 5f70 7265 7365 6e74  e_before_present
+0000e460: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0000e470: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e480: 2020 2020 2020 2020 206d 6178 5f74 696d           max_tim
+0000e490: 655f 6167 6f3d 6d61 785f 7469 6d65 5f61  e_ago=max_time_a
+0000e4a0: 676f 2c0a 2020 2020 2020 2020 2020 2020  go,.            
+0000e4b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e4c0: 2020 2020 2020 2020 2020 206d 696e 5f74             min_t
+0000e4d0: 696d 655f 6167 6f3d 6d69 6e5f 7469 6d65  ime_ago=min_time
+0000e4e0: 5f61 676f 290a 2020 2020 2020 2020 2320  _ago).        # 
+0000e4f0: 6966 2072 6571 7565 7374 6564 2c20 7075  if requested, pu
+0000e500: 7420 7468 6520 6375 7272 656e 7420 696e  t the current in
+0000e510: 6469 7669 6475 616c 7327 2063 7572 7265  dividuals' curre
+0000e520: 6e74 200a 2020 2020 2020 2020 2320 706f  nt .        # po
+0000e530: 7369 7469 6f6e 7320 696e 2074 6865 206c  sitions in the l
+0000e540: 696e 6561 6765 2064 6963 7473 2c20 7261  ineage dicts, ra
+0000e550: 7468 6572 2074 6861 6e20 7573 696e 6720  ther than using 
+0000e560: 7468 6520 6269 7274 680a 2020 2020 2020  the birth.      
+0000e570: 2020 2320 706f 7369 7469 6f6e 7320 2877    # positions (w
+0000e580: 6869 6368 2061 7265 2069 6e20 7468 6572  hich are in ther
+0000e590: 6520 6279 2064 6566 6175 6c74 2062 6563  e by default bec
+0000e5a0: 6175 7365 2074 686f 7365 0a20 2020 2020  ause those.     
+0000e5b0: 2020 2023 2061 7265 2074 6865 2070 6f73     # are the pos
+0000e5c0: 6974 696f 6e73 2074 6861 7420 6172 6520  itions that are 
+0000e5d0: 7374 6f72 6564 2077 6974 6869 6e20 7468  stored within th
+0000e5e0: 6520 7473 6b69 7420 7461 626c 6573 290a  e tskit tables).
+0000e5f0: 2020 2020 2020 2020 6966 2075 7365 5f69          if use_i
+0000e600: 6e64 6976 6964 735f 6375 7272 5f70 6f73  ndivids_curr_pos
+0000e610: 3a0a 2020 2020 2020 2020 2020 2020 666f  :.            fo
+0000e620: 7220 6c6f 6375 7320 696e 206c 696e 5f64  r locus in lin_d
+0000e630: 6963 7473 2e6b 6579 7328 293a 0a20 2020  icts.keys():.   
+0000e640: 2020 2020 2020 2020 2020 2020 2066 6f72               for
+0000e650: 2063 7572 725f 6e6f 6465 5f69 6420 696e   curr_node_id in
+0000e660: 206c 696e 5f64 6963 7473 5b6c 6f63 7573   lin_dicts[locus
+0000e670: 5d2e 6b65 7973 2829 3a0a 2020 2020 2020  ].keys():.      
+0000e680: 2020 2020 2020 2020 2020 2020 2020 6966                if
+0000e690: 206c 656e 286c 696e 5f64 6963 7473 5b6c   len(lin_dicts[l
+0000e6a0: 6f63 7573 5d5b 6375 7272 5f6e 6f64 655f  ocus][curr_node_
+0000e6b0: 6964 5d29 203e 2030 3a0a 2020 2020 2020  id]) > 0:.      
+0000e6c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e6d0: 2020 7463 5f69 6e64 6976 6964 735f 6964    tc_individs_id
+0000e6e0: 203d 2073 656c 662e 5f74 632e 6e6f 6465   = self._tc.node
+0000e6f0: 735b 6375 7272 5f6e 6f64 655f 6964 5d2e  s[curr_node_id].
+0000e700: 696e 6469 7669 6475 616c 0a20 2020 2020  individual.     
+0000e710: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e720: 2020 2069 6e64 6976 6964 203d 205b 696e     individ = [in
+0000e730: 6420 666f 7220 696e 6420 696e 2073 656c  d for ind in sel
+0000e740: 662e 7661 6c75 6573 2829 2069 660a 2020  f.values() if.  
 0000e750: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000e760: 2020 6375 7272 5f70 6f73 290a 2020 2020    curr_pos).    
-0000e770: 2020 2020 7265 7475 726e 206c 696e 5f64      return lin_d
-0000e780: 6963 7473 0a0a 0a20 2020 2023 2063 6865  icts...    # che
-0000e790: 636b 2077 6865 7468 6572 2073 7065 6369  ck whether speci
-0000e7a0: 6669 6564 2069 6e64 6976 6964 7561 6c73  fied individuals
-0000e7b0: 2068 6176 6520 636f 616c 6573 6365 6420   have coalesced 
-0000e7c0: 6174 2073 7065 6369 6669 6564 206c 6f63  at specified loc
-0000e7d0: 690a 2020 2020 2320 2877 6974 6869 6e20  i.    # (within 
-0000e7e0: 7468 6520 666f 7277 6172 642d 7469 6d65  the forward-time
-0000e7f0: 2047 656f 6e6f 6d69 6373 2072 756e 2c20   Geonomics run, 
-0000e800: 6f66 2063 6f75 7273 652c 2073 696e 6365  of course, since
-0000e810: 2062 7920 6465 7369 676e 0a20 2020 2023   by design.    #
-0000e820: 2065 7665 7279 7468 696e 6720 636f 616c   everything coal
-0000e830: 6573 6365 7320 696e 2074 6865 206d 732d  esces in the ms-
-0000e840: 6765 6e65 7261 7465 6420 7072 652d 7369  generated pre-si
-0000e850: 6d75 6c61 7469 6f6e 2070 6564 6967 7265  mulation pedigre
-0000e860: 6529 0a20 2020 2064 6566 205f 6368 6563  e).    def _chec
-0000e870: 6b5f 636f 616c 6573 6365 6e63 6528 7365  k_coalescence(se
-0000e880: 6c66 2c20 696e 6469 7669 6473 3d4e 6f6e  lf, individs=Non
-0000e890: 652c 206c 6f63 693d 4e6f 6e65 2c20 616c  e, loci=None, al
-0000e8a0: 6c5f 6c6f 6369 3d46 616c 7365 293a 0a20  l_loci=False):. 
-0000e8b0: 2020 2020 2020 2023 736f 7274 2061 6e64         #sort and
-0000e8c0: 2073 696d 706c 6966 7920 7468 6520 5461   simplify the Ta
-0000e8d0: 626c 6543 6f6c 6c65 6374 696f 6e2c 2069  bleCollection, i
-0000e8e0: 6620 6e65 6564 6564 0a20 2020 2020 2020  f needed.       
-0000e8f0: 2069 6620 6e6f 7420 7365 6c66 2e5f 7463   if not self._tc
-0000e900: 5f73 6f72 7465 645f 616e 645f 7369 6d70  _sorted_and_simp
-0000e910: 6c69 6669 6564 3a0a 2020 2020 2020 2020  lified:.        
-0000e920: 2020 2020 7365 6c66 2e5f 736f 7274 5f73      self._sort_s
-0000e930: 696d 706c 6966 795f 7461 626c 655f 636f  implify_table_co
-0000e940: 6c6c 6563 7469 6f6e 2829 0a20 2020 2020  llection().     
-0000e950: 2020 2069 6620 696e 6469 7669 6473 2069     if individs i
-0000e960: 7320 4e6f 6e65 3a0a 2020 2020 2020 2020  s None:.        
-0000e970: 2020 2020 696e 6469 7669 6473 203d 205b      individs = [
-0000e980: 2a73 656c 665d 0a20 2020 2020 2020 2023  *self].        #
-0000e990: 2067 6574 2074 6865 206e 6f64 6520 4944   get the node ID
-0000e9a0: 730a 2020 2020 2020 2020 6e6f 6465 7320  s.        nodes 
-0000e9b0: 3d20 7365 6c66 2e5f 6765 745f 6e6f 6465  = self._get_node
-0000e9c0: 7328 696e 6469 7669 6473 3d69 6e64 6976  s(individs=indiv
-0000e9d0: 6964 7329 0a20 2020 2020 2020 2069 6620  ids).        if 
-0000e9e0: 6c6f 6369 2069 7320 4e6f 6e65 3a0a 2020  loci is None:.  
-0000e9f0: 2020 2020 2020 2020 2020 6c6f 6369 203d            loci =
-0000ea00: 205b 2a72 616e 6765 2873 656c 662e 6765   [*range(self.ge
-0000ea10: 6e5f 6172 6368 2e4c 295d 0a20 2020 2020  n_arch.L)].     
-0000ea20: 2020 2023 204e 4f54 453a 2069 6620 7275     # NOTE: if ru
-0000ea30: 6e6e 696e 6720 696e 746f 206d 656d 6f72  nning into memor
-0000ea40: 7920 6973 7375 6573 2077 6865 6e20 6765  y issues when ge
-0000ea50: 7474 696e 6720 7468 6520 6c69 6e65 6167  tting the lineag
-0000ea60: 6520 6469 6374 730a 2020 2020 2020 2020  e dicts.        
-0000ea70: 2320 616c 6c20 6174 206f 6e63 6520 666f  # all at once fo
-0000ea80: 7220 7369 6d75 6c61 7469 6f6e 7320 7769  r simulations wi
-0000ea90: 7468 206c 6172 6765 2070 6f70 2073 697a  th large pop siz
-0000eaa0: 6573 2061 6e64 206c 6172 6765 206e 756d  es and large num
-0000eab0: 7320 6f66 0a20 2020 2020 2020 2023 206c  s of.        # l
-0000eac0: 6f63 6920 7468 656e 2073 686f 756c 6420  oci then should 
-0000ead0: 7472 7920 696e 7374 6561 6420 6765 7474  try instead gett
-0000eae0: 696e 6720 7468 656d 2031 2062 7920 3120  ing them 1 by 1 
-0000eaf0: 696e 2074 6865 206c 6f6f 7020 6265 6c6f  in the loop belo
-0000eb00: 770a 2020 2020 2020 2020 6c69 6e5f 6469  w.        lin_di
-0000eb10: 6374 7320 3d20 7365 6c66 2e5f 6765 745f  cts = self._get_
-0000eb20: 6c69 6e65 6167 655f 6469 6374 7328 6c6f  lineage_dicts(lo
-0000eb30: 6369 2c20 6e6f 6465 733d 6e6f 6465 7329  ci, nodes=nodes)
-0000eb40: 0a20 2020 2020 2020 2072 6573 756c 7420  .        result 
-0000eb50: 3d20 7b7d 0a20 2020 2020 2020 2066 6f72  = {}.        for
-0000eb60: 206c 6f63 2c20 6c69 6e5f 6469 6374 2069   loc, lin_dict i
-0000eb70: 6e20 6c69 6e5f 6469 6374 732e 6974 656d  n lin_dicts.item
-0000eb80: 7328 293a 0a20 2020 2020 2020 2020 2020  s():.           
-0000eb90: 2023 2067 6574 2074 6865 206e 756d 6265   # get the numbe
-0000eba0: 7220 6f66 2075 6e69 7175 6520 6f6c 6465  r of unique olde
-0000ebb0: 7374 206e 6f64 6573 2061 6372 6f73 7320  st nodes across 
-0000ebc0: 7468 6520 6c69 6e65 6167 6520 6469 6374  the lineage dict
-0000ebd0: 0a20 2020 2020 2020 2020 2020 206f 6c64  .            old
-0000ebe0: 6573 745f 6e6f 6465 7320 3d20 5b5b 2a6c  est_nodes = [[*l
-0000ebf0: 696e 5f64 6963 745b 6e6f 6465 5d2e 6b65  in_dict[node].ke
-0000ec00: 7973 280a 2020 2020 2020 2020 2020 2020  ys(.            
-0000ec10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000ec20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000ec30: 295d 5b2d 315d 2066 6f72 206e 6f64 6520  )][-1] for node 
-0000ec40: 696e 206c 696e 5f64 6963 742e 6b65 7973  in lin_dict.keys
-0000ec50: 2829 5d0a 2020 2020 2020 2020 2020 2020  ()].            
-0000ec60: 6e5f 756e 6971 7565 5f6e 6f64 6573 203d  n_unique_nodes =
-0000ec70: 206c 656e 286e 702e 756e 6971 7565 286f   len(np.unique(o
-0000ec80: 6c64 6573 745f 6e6f 6465 7329 290a 2020  ldest_nodes)).  
-0000ec90: 2020 2020 2020 2020 2020 7265 7375 6c74            result
-0000eca0: 5b6c 6f63 5d20 3d20 6e5f 756e 6971 7565  [loc] = n_unique
-0000ecb0: 5f6e 6f64 6573 203d 3d20 310a 2020 2020  _nodes == 1.    
-0000ecc0: 2020 2020 2320 6368 6563 6b20 636f 616c      # check coal
-0000ecd0: 6573 6365 6e63 6520 6163 726f 7373 2061  escence across a
-0000ece0: 6c6c 206c 6f63 692c 2069 6620 7265 7175  ll loci, if requ
-0000ecf0: 6573 7465 640a 2020 2020 2020 2020 6966  ested.        if
-0000ed00: 2061 6c6c 5f6c 6f63 693a 0a20 2020 2020   all_loci:.     
-0000ed10: 2020 2020 2020 2072 6573 756c 7420 3d20         result = 
-0000ed20: 6e70 2e61 6c6c 285b 2a72 6573 756c 742e  np.all([*result.
-0000ed30: 7661 6c75 6573 2829 5d29 0a20 2020 2020  values()]).     
-0000ed40: 2020 2072 6574 7572 6e20 7265 7375 6c74     return result
-0000ed50: 0a0a 0a20 2020 2064 6566 205f 6361 6c63  ...    def _calc
-0000ed60: 5f6c 696e 6561 6765 5f73 7461 7473 2873  _lineage_stats(s
-0000ed70: 656c 662c 2069 6e64 6976 6964 733d 4e6f  elf, individs=No
-0000ed80: 6e65 2c20 6e6f 6465 733d 4e6f 6e65 2c20  ne, nodes=None, 
-0000ed90: 6c6f 6369 3d4e 6f6e 652c 0a20 2020 2020  loci=None,.     
-0000eda0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000edb0: 2020 2020 2020 2073 7461 7473 3d5b 2764         stats=['d
-0000edc0: 6972 272c 2027 6469 7374 272c 2027 7469  ir', 'dist', 'ti
-0000edd0: 6d65 272c 2027 7370 6565 6427 5d2c 0a20  me', 'speed'],. 
-0000ede0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000edf0: 2020 2020 2020 2020 2020 2075 7365 5f69             use_i
-0000ee00: 6e64 6976 6964 735f 6375 7272 5f70 6f73  ndivids_curr_pos
-0000ee10: 3d54 7275 652c 0a20 2020 2020 2020 2020  =True,.         
-0000ee20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000ee30: 2020 206d 6178 5f74 696d 655f 6167 6f3d     max_time_ago=
-0000ee40: 4e6f 6e65 2c20 6d69 6e5f 7469 6d65 5f61  None, min_time_a
-0000ee50: 676f 3d4e 6f6e 6529 3a0a 2020 2020 2020  go=None):.      
-0000ee60: 2020 2727 270a 2020 2020 2020 2020 4361    '''.        Ca
-0000ee70: 6c63 756c 6174 6520 7374 6174 7320 666f  lculate stats fo
-0000ee80: 7220 7468 6520 6c69 6e65 6167 6573 206f  r the lineages o
-0000ee90: 6620 6120 6769 7665 6e20 7365 7420 6f66  f a given set of
-0000eea0: 206e 6f64 6573 2061 6e64 206c 6f63 693b   nodes and loci;
-0000eeb0: 0a20 2020 2020 2020 2072 6574 7572 6e73  .        returns
-0000eec0: 2064 6963 7420 6f66 2073 7472 7563 743a   dict of struct:
-0000eed0: 0a20 2020 2020 2020 2020 2020 207b 6b3d  .            {k=
-0000eee0: 7374 6174 2c20 763d 7b6b 3d6c 6f63 2c20  stat, v={k=loc, 
-0000eef0: 763d 5b76 616c 5f6e 6f64 6531 202e 2e2e  v=[val_node1 ...
-0000ef00: 2076 616c 5f6e 6f64 655f 4e5d 7d7d 0a0a   val_node_N]}}..
-0000ef10: 2020 2020 2020 2020 4e4f 5445 3a20 6469          NOTE: di
-0000ef20: 7265 6374 696f 6e73 2061 7265 2065 7870  rections are exp
-0000ef30: 7265 7373 6564 2061 7320 2763 6f6d 7061  ressed as 'compa
-0000ef40: 7373 2720 6469 7265 6374 696f 6e73 0a20  ss' directions. 
-0000ef50: 2020 2020 2020 2020 2020 2020 2028 692e               (i.
-0000ef60: 652e 2c20 6578 7072 6573 7365 6420 696e  e., expressed in
-0000ef70: 2064 6567 7265 6573 2c20 636c 6f63 6b77   degrees, clockw
-0000ef80: 6973 6520 6672 6f6d 2030 2064 6567 7265  ise from 0 degre
-0000ef90: 6573 2061 7420 6e6f 7274 6829 0a20 2020  es at north).   
-0000efa0: 2020 2020 2027 2727 0a20 2020 2020 2020       '''.       
-0000efb0: 2023 2067 6574 2061 6c6c 206e 6f64 6573   # get all nodes
-0000efc0: 2066 6f72 2074 6865 2070 726f 7669 6465   for the provide
-0000efd0: 6420 696e 6469 7669 6475 616c 732c 206f  d individuals, o
-0000efe0: 7220 666f 7220 616c 6c20 696e 6469 7669  r for all indivi
-0000eff0: 6475 616c 732c 0a20 2020 2020 2020 2023  duals,.        #
-0000f000: 2069 6620 6e6f 6465 7320 4944 7320 6e6f   if nodes IDs no
-0000f010: 7420 7072 6f76 6964 6564 0a20 2020 2020  t provided.     
-0000f020: 2020 2069 6620 6e6f 6465 7320 6973 204e     if nodes is N
-0000f030: 6f6e 653a 0a20 2020 2020 2020 2020 2020  one:.           
-0000f040: 2023 736f 7274 2061 6e64 2073 696d 706c   #sort and simpl
-0000f050: 6966 7920 7468 6520 5461 626c 6543 6f6c  ify the TableCol
-0000f060: 6c65 6374 696f 6e2c 2069 6620 6e65 6564  lection, if need
-0000f070: 6564 0a20 2020 2020 2020 2020 2020 2069  ed.            i
-0000f080: 6620 6e6f 7420 7365 6c66 2e5f 7463 5f73  f not self._tc_s
-0000f090: 6f72 7465 645f 616e 645f 7369 6d70 6c69  orted_and_simpli
-0000f0a0: 6669 6564 3a0a 2020 2020 2020 2020 2020  fied:.          
-0000f0b0: 2020 2020 2020 7365 6c66 2e5f 736f 7274        self._sort
-0000f0c0: 5f73 696d 706c 6966 795f 7461 626c 655f  _simplify_table_
-0000f0d0: 636f 6c6c 6563 7469 6f6e 2829 0a20 2020  collection().   
-0000f0e0: 2020 2020 2020 2020 206e 6f64 6573 203d           nodes =
-0000f0f0: 2073 656c 662e 5f67 6574 5f6e 6f64 6573   self._get_nodes
-0000f100: 2869 6e64 6976 6964 7329 0a20 2020 2020  (individs).     
-0000f110: 2020 2023 2067 6574 2061 6c6c 206c 6f63     # get all loc
-0000f120: 692c 2069 6620 6c6f 6369 206e 6f74 2070  i, if loci not p
-0000f130: 726f 7669 6465 640a 2020 2020 2020 2020  rovided.        
-0000f140: 6966 206c 6f63 6920 6973 204e 6f6e 653a  if loci is None:
-0000f150: 0a20 2020 2020 2020 2020 2020 206c 6f63  .            loc
-0000f160: 6920 3d20 5b2a 7261 6e67 6528 7365 6c66  i = [*range(self
-0000f170: 2e67 656e 5f61 7263 682e 4c29 5d0a 2020  .gen_arch.L)].  
-0000f180: 2020 2020 2020 6c69 6e5f 6469 6374 7320        lin_dicts 
-0000f190: 3d20 7365 6c66 2e5f 6765 745f 6c69 6e65  = self._get_line
-0000f1a0: 6167 655f 6469 6374 7328 6c6f 6369 2c20  age_dicts(loci, 
-0000f1b0: 6e6f 6465 733d 6e6f 6465 732c 0a20 2020  nodes=nodes,.   
-0000f1c0: 2020 2020 2020 2020 2020 2020 2075 7365               use
-0000f1d0: 5f69 6e64 6976 6964 735f 6375 7272 5f70  _individs_curr_p
-0000f1e0: 6f73 3d75 7365 5f69 6e64 6976 6964 735f  os=use_individs_
-0000f1f0: 6375 7272 5f70 6f73 2c0a 2020 2020 2020  curr_pos,.      
-0000f200: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f210: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f220: 2020 2020 2020 6d61 785f 7469 6d65 5f61        max_time_a
-0000f230: 676f 3d6d 6178 5f74 696d 655f 6167 6f2c  go=max_time_ago,
-0000f240: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000f250: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f260: 2020 2020 2020 2020 2020 2020 206d 696e               min
-0000f270: 5f74 696d 655f 6167 6f3d 6d69 6e5f 7469  _time_ago=min_ti
-0000f280: 6d65 5f61 676f 290a 2020 2020 2020 2020  me_ago).        
-0000f290: 7374 6174 7320 3d20 7b73 7461 743a 207b  stats = {stat: {
-0000f2a0: 7d20 666f 7220 7374 6174 2069 6e20 7374  } for stat in st
-0000f2b0: 6174 737d 0a20 2020 2020 2020 2066 6f72  ats}.        for
-0000f2c0: 2073 7461 7420 696e 2073 7461 7473 3a0a   stat in stats:.
-0000f2d0: 2020 2020 2020 2020 2020 2020 666f 7220              for 
-0000f2e0: 6c6f 6320 696e 206c 6f63 693a 0a20 2020  loc in loci:.   
-0000f2f0: 2020 2020 2020 2020 2020 2020 206c 6f63               loc
-0000f300: 5f73 7461 745f 6c69 7374 203d 205b 5d0a  _stat_list = [].
-0000f310: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f320: 666f 7220 6e6f 6465 2069 6e20 6e6f 6465  for node in node
-0000f330: 733a 0a20 2020 2020 2020 2020 2020 2020  s:.             
-0000f340: 2020 2020 2020 206c 6f63 5f73 7461 745f         loc_stat_
-0000f350: 6c69 7374 2e61 7070 656e 6428 5f63 616c  list.append(_cal
-0000f360: 635f 6c69 6e65 6167 655f 7374 6174 280a  c_lineage_stat(.
-0000f370: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f380: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f390: 2020 2020 2020 2020 2020 2020 6c69 6e5f              lin_
-0000f3a0: 6469 6374 735b 6c6f 635d 5b6e 6f64 655d  dicts[loc][node]
-0000f3b0: 2c20 7374 6174 3d73 7461 7429 290a 2020  , stat=stat)).  
-0000f3c0: 2020 2020 2020 2020 2020 2020 2020 7374                st
-0000f3d0: 6174 735b 7374 6174 5d5b 6c6f 635d 203d  ats[stat][loc] =
-0000f3e0: 206c 6f63 5f73 7461 745f 6c69 7374 0a20   loc_stat_list. 
-0000f3f0: 2020 2020 2020 2072 6574 7572 6e20 7374         return st
-0000f400: 6174 730a 0a20 2020 2023 2323 2323 2323  ats..    #######
-0000f410: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000f420: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000f430: 2323 2323 2323 230a 2020 2020 2320 454e  #######.    # EN
-0000f440: 4420 4655 4e43 5449 4f4e 5320 4652 4f4d  D FUNCTIONS FROM
-0000f450: 2054 5241 434b 5f53 5041 5449 414c 5f50   TRACK_SPATIAL_P
-0000f460: 4544 4947 5245 452e 5059 0a20 2020 2023  EDIGREE.PY.    #
-0000f470: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000f480: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0000f490: 2323 2323 2323 2323 2323 2323 230a 0a0a  #############...
-0000f4a0: 2020 2020 2320 6d65 7468 6f64 2074 6f20      # method to 
-0000f4b0: 6765 7420 696e 6469 7669 6473 2720 656e  get individs' en
-0000f4c0: 7669 726f 6e6d 656e 7420 7661 6c75 6573  vironment values
-0000f4d0: 0a20 2020 2064 6566 205f 6765 745f 6528  .    def _get_e(
-0000f4e0: 7365 6c66 2c20 6c79 725f 6e75 6d3d 4e6f  self, lyr_num=No
-0000f4f0: 6e65 2c20 696e 6469 7669 6473 3d4e 6f6e  ne, individs=Non
-0000f500: 6529 3a0a 2020 2020 2020 2020 6966 2069  e):.        if i
-0000f510: 6e64 6976 6964 7320 6973 204e 6f6e 653a  ndivids is None:
-0000f520: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
-0000f530: 6c79 725f 6e75 6d20 6973 204e 6f6e 653a  lyr_num is None:
-0000f540: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000f550: 2065 203d 206e 702e 6172 7261 7928 5b69   e = np.array([i
-0000f560: 6e64 2e65 2066 6f72 2069 6e64 2069 6e20  nd.e for ind in 
-0000f570: 7365 6c66 2e76 616c 7565 7328 295d 290a  self.values()]).
-0000f580: 2020 2020 2020 2020 2020 2020 656c 7365              else
-0000f590: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-0000f5a0: 2020 6520 3d20 6e70 2e61 7272 6179 285b    e = np.array([
-0000f5b0: 696e 642e 655b 6c79 725f 6e75 6d5d 2066  ind.e[lyr_num] f
-0000f5c0: 6f72 2069 6e64 2069 6e20 7365 6c66 2e76  or ind in self.v
-0000f5d0: 616c 7565 7328 295d 290a 2020 2020 2020  alues()]).      
-0000f5e0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
-0000f5f0: 2020 2020 6967 203d 2069 7465 6d67 6574      ig = itemget
-0000f600: 7465 7228 2a69 6e64 6976 6964 7329 0a20  ter(*individs). 
-0000f610: 2020 2020 2020 2020 2020 2069 6620 6c79             if ly
-0000f620: 725f 6e75 6d20 6973 204e 6f6e 653a 0a20  r_num is None:. 
-0000f630: 2020 2020 2020 2020 2020 2020 2020 2065                 e
-0000f640: 203d 207b 693a 696e 642e 6520 666f 7220   = {i:ind.e for 
-0000f650: 692c 2069 6e64 2069 6e20 7365 6c66 2e69  i, ind in self.i
-0000f660: 7465 6d73 2829 7d0a 2020 2020 2020 2020  tems()}.        
-0000f670: 2020 2020 2020 2020 6520 3d20 6e70 2e61          e = np.a
-0000f680: 7272 6179 2869 6728 6529 290a 2020 2020  rray(ig(e)).    
-0000f690: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
-0000f6a0: 2020 2020 2020 2020 2020 2020 2020 6520                e 
-0000f6b0: 3d20 7b69 3a69 6e64 2e65 5b6c 7972 5f6e  = {i:ind.e[lyr_n
-0000f6c0: 756d 5d20 666f 7220 692c 2069 6e64 2069  um] for i, ind i
-0000f6d0: 6e20 7365 6c66 2e69 7465 6d73 2829 7d0a  n self.items()}.
-0000f6e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f6f0: 6520 3d20 6e70 2e61 7272 6179 2869 6728  e = np.array(ig(
-0000f700: 6529 290a 2020 2020 2020 2020 7265 7475  e)).        retu
-0000f710: 726e 2065 0a0a 0a20 2020 2064 6566 205f  rn e...    def _
-0000f720: 6765 745f 6765 6e6f 7479 7065 7328 7365  get_genotypes(se
-0000f730: 6c66 2c20 6c6f 6369 3d4e 6f6e 652c 2069  lf, loci=None, i
-0000f740: 6e64 6976 6964 733d 4e6f 6e65 2c20 6269  ndivids=None, bi
-0000f750: 616c 6c65 6c69 633d 5472 7565 2c0a 2020  allelic=True,.  
-0000f760: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f770: 2020 2020 2061 735f 6469 6374 3d46 616c       as_dict=Fal
-0000f780: 7365 293a 0a20 2020 2020 2020 2023 206d  se):.        # m
-0000f790: 616b 6520 7375 7265 2061 735f 6469 6374  ake sure as_dict
-0000f7a0: 2061 6e64 2062 6961 6c6c 656c 6963 2061   and biallelic a
-0000f7b0: 7265 2054 7275 6520 6f72 2046 616c 7365  re True or False
-0000f7c0: 0a20 2020 2020 2020 2061 7373 6572 7420  .        assert 
-0000f7d0: 6173 5f64 6963 7420 696e 205b 5472 7565  as_dict in [True
-0000f7e0: 2c20 4661 6c73 655d 2c20 2822 5468 6520  , False], ("The 
-0000f7f0: 2761 735f 6469 6374 2720 6172 6775 6d65  'as_dict' argume
-0000f800: 6e74 2022 0a20 2020 2020 2020 2020 2020  nt ".           
-0000f810: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f820: 2020 2020 2020 2020 2020 2020 2020 2022                 "
-0000f830: 6d75 7374 2062 6520 6569 7468 6572 2022  must be either "
-0000f840: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000f850: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f860: 2020 2020 2020 2020 2020 2022 5472 7565             "True
-0000f870: 206f 7220 4661 6c73 652e 2229 0a0a 2020   or False.")..  
-0000f880: 2020 2020 2020 6173 7365 7274 2062 6961        assert bia
-0000f890: 6c6c 656c 6963 2069 6e20 5b54 7275 652c  llelic in [True,
-0000f8a0: 2046 616c 7365 5d2c 2028 2254 6865 2027   False], ("The '
-0000f8b0: 6269 616c 6c65 6c69 6327 2061 7267 756d  biallelic' argum
-0000f8c0: 656e 7420 220a 2020 2020 2020 2020 2020  ent ".          
+0000e760: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e770: 2069 6e64 2e5f 696e 6469 7669 6475 616c   ind._individual
+0000e780: 735f 7461 625f 6964 203d 3d20 7463 5f69  s_tab_id == tc_i
+0000e790: 6e64 6976 6964 735f 6964 5d0a 2020 2020  ndivids_id].    
+0000e7a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e7b0: 2020 2020 6173 7365 7274 206c 656e 2869      assert len(i
+0000e7c0: 6e64 6976 6964 2920 3d3d 2031 2c20 2822  ndivid) == 1, ("
+0000e7d0: 466f 756e 6420 6d75 6c74 6970 6c65 2069  Found multiple i
+0000e7e0: 6e64 6976 6964 7561 6c73 2022 0a20 2020  ndividuals ".   
+0000e7f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e800: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e810: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e820: 2277 6974 6820 7468 6520 7361 6d65 2069  "with the same i
+0000e830: 6e64 6976 6964 7561 6c73 2022 0a20 2020  ndividuals ".   
+0000e840: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e850: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e860: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e870: 2269 6420 6e20 7468 6520 7473 6b69 7420  "id n the tskit 
+0000e880: 7461 626c 6573 2e22 290a 2020 2020 2020  tables.").      
+0000e890: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e8a0: 2020 696e 6469 7669 6420 3d20 696e 6469    individ = indi
+0000e8b0: 7669 645b 305d 0a20 2020 2020 2020 2020  vid[0].         
+0000e8c0: 2020 2020 2020 2020 2020 2020 2020 2063                 c
+0000e8d0: 7572 725f 706f 7320 3d20 6e70 2e61 7272  urr_pos = np.arr
+0000e8e0: 6179 285b 696e 6469 7669 642e 782c 2069  ay([individ.x, i
+0000e8f0: 6e64 6976 6964 2e79 5d29 0a20 2020 2020  ndivid.y]).     
+0000e900: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e910: 2020 2062 6972 7468 5f74 203d 6c69 6e5f     birth_t =lin_
+0000e920: 6469 6374 735b 6c6f 6375 735d 5b63 7572  dicts[locus][cur
+0000e930: 725f 6e6f 6465 5f69 645d 5b63 7572 725f  r_node_id][curr_
+0000e940: 6e6f 6465 5f69 645d 5b30 5d0a 2020 2020  node_id][0].    
+0000e950: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e960: 2020 2020 6c69 6e5f 6469 6374 735b 6c6f      lin_dicts[lo
+0000e970: 6375 735d 5b63 7572 725f 6e6f 6465 5f69  cus][curr_node_i
+0000e980: 645d 5b63 7572 725f 6e6f 6465 5f69 645d  d][curr_node_id]
+0000e990: 203d 2028 6269 7274 685f 742c 0a20 2020   = (birth_t,.   
+0000e9a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e9b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e9c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e9d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000e9e0: 2020 2020 6375 7272 5f70 6f73 290a 2020      curr_pos).  
+0000e9f0: 2020 2020 2020 7265 7475 726e 206c 696e        return lin
+0000ea00: 5f64 6963 7473 0a0a 0a20 2020 2023 2063  _dicts...    # c
+0000ea10: 6865 636b 2077 6865 7468 6572 2073 7065  heck whether spe
+0000ea20: 6369 6669 6564 2069 6e64 6976 6964 7561  cified individua
+0000ea30: 6c73 2068 6176 6520 636f 616c 6573 6365  ls have coalesce
+0000ea40: 6420 6174 2073 7065 6369 6669 6564 206c  d at specified l
+0000ea50: 6f63 690a 2020 2020 2320 2877 6974 6869  oci.    # (withi
+0000ea60: 6e20 7468 6520 666f 7277 6172 642d 7469  n the forward-ti
+0000ea70: 6d65 2047 656f 6e6f 6d69 6373 2072 756e  me Geonomics run
+0000ea80: 2c20 6f66 2063 6f75 7273 652c 2073 696e  , of course, sin
+0000ea90: 6365 2062 7920 6465 7369 676e 0a20 2020  ce by design.   
+0000eaa0: 2023 2065 7665 7279 7468 696e 6720 636f   # everything co
+0000eab0: 616c 6573 6365 7320 696e 2074 6865 206d  alesces in the m
+0000eac0: 732d 6765 6e65 7261 7465 6420 7072 652d  s-generated pre-
+0000ead0: 7369 6d75 6c61 7469 6f6e 2070 6564 6967  simulation pedig
+0000eae0: 7265 6529 0a20 2020 2064 6566 205f 6368  ree).    def _ch
+0000eaf0: 6563 6b5f 636f 616c 6573 6365 6e63 6528  eck_coalescence(
+0000eb00: 7365 6c66 2c20 696e 6469 7669 6473 3d4e  self, individs=N
+0000eb10: 6f6e 652c 206c 6f63 693d 4e6f 6e65 2c20  one, loci=None, 
+0000eb20: 616c 6c5f 6c6f 6369 3d46 616c 7365 293a  all_loci=False):
+0000eb30: 0a20 2020 2020 2020 2023 736f 7274 2061  .        #sort a
+0000eb40: 6e64 2073 696d 706c 6966 7920 7468 6520  nd simplify the 
+0000eb50: 5461 626c 6543 6f6c 6c65 6374 696f 6e2c  TableCollection,
+0000eb60: 2069 6620 6e65 6564 6564 0a20 2020 2020   if needed.     
+0000eb70: 2020 2069 6620 6e6f 7420 7365 6c66 2e5f     if not self._
+0000eb80: 7463 5f73 6f72 7465 645f 616e 645f 7369  tc_sorted_and_si
+0000eb90: 6d70 6c69 6669 6564 3a0a 2020 2020 2020  mplified:.      
+0000eba0: 2020 2020 2020 7365 6c66 2e5f 736f 7274        self._sort
+0000ebb0: 5f61 6e64 5f73 696d 706c 6966 795f 7461  _and_simplify_ta
+0000ebc0: 626c 655f 636f 6c6c 6563 7469 6f6e 2829  ble_collection()
+0000ebd0: 0a20 2020 2020 2020 2069 6620 696e 6469  .        if indi
+0000ebe0: 7669 6473 2069 7320 4e6f 6e65 3a0a 2020  vids is None:.  
+0000ebf0: 2020 2020 2020 2020 2020 696e 6469 7669            indivi
+0000ec00: 6473 203d 205b 2a73 656c 665d 0a20 2020  ds = [*self].   
+0000ec10: 2020 2020 2023 2067 6574 2074 6865 206e       # get the n
+0000ec20: 6f64 6520 4944 730a 2020 2020 2020 2020  ode IDs.        
+0000ec30: 6e6f 6465 7320 3d20 7365 6c66 2e5f 6765  nodes = self._ge
+0000ec40: 745f 6e6f 6465 7328 696e 6469 7669 6473  t_nodes(individs
+0000ec50: 3d69 6e64 6976 6964 7329 0a20 2020 2020  =individs).     
+0000ec60: 2020 2069 6620 6c6f 6369 2069 7320 4e6f     if loci is No
+0000ec70: 6e65 3a0a 2020 2020 2020 2020 2020 2020  ne:.            
+0000ec80: 6c6f 6369 203d 205b 2a72 616e 6765 2873  loci = [*range(s
+0000ec90: 656c 662e 6765 6e5f 6172 6368 2e4c 295d  elf.gen_arch.L)]
+0000eca0: 0a20 2020 2020 2020 2023 204e 4f54 453a  .        # NOTE:
+0000ecb0: 2069 6620 7275 6e6e 696e 6720 696e 746f   if running into
+0000ecc0: 206d 656d 6f72 7920 6973 7375 6573 2077   memory issues w
+0000ecd0: 6865 6e20 6765 7474 696e 6720 7468 6520  hen getting the 
+0000ece0: 6c69 6e65 6167 6520 6469 6374 730a 2020  lineage dicts.  
+0000ecf0: 2020 2020 2020 2320 616c 6c20 6174 206f        # all at o
+0000ed00: 6e63 6520 666f 7220 7369 6d75 6c61 7469  nce for simulati
+0000ed10: 6f6e 7320 7769 7468 206c 6172 6765 2070  ons with large p
+0000ed20: 6f70 2073 697a 6573 2061 6e64 206c 6172  op sizes and lar
+0000ed30: 6765 206e 756d 7320 6f66 0a20 2020 2020  ge nums of.     
+0000ed40: 2020 2023 206c 6f63 6920 7468 656e 2073     # loci then s
+0000ed50: 686f 756c 6420 7472 7920 696e 7374 6561  hould try instea
+0000ed60: 6420 6765 7474 696e 6720 7468 656d 2031  d getting them 1
+0000ed70: 2062 7920 3120 696e 2074 6865 206c 6f6f   by 1 in the loo
+0000ed80: 7020 6265 6c6f 770a 2020 2020 2020 2020  p below.        
+0000ed90: 6c69 6e5f 6469 6374 7320 3d20 7365 6c66  lin_dicts = self
+0000eda0: 2e5f 6765 745f 6c69 6e65 6167 655f 6469  ._get_lineage_di
+0000edb0: 6374 7328 6c6f 6369 2c20 6e6f 6465 733d  cts(loci, nodes=
+0000edc0: 6e6f 6465 7329 0a20 2020 2020 2020 2072  nodes).        r
+0000edd0: 6573 756c 7420 3d20 7b7d 0a20 2020 2020  esult = {}.     
+0000ede0: 2020 2066 6f72 206c 6f63 2c20 6c69 6e5f     for loc, lin_
+0000edf0: 6469 6374 2069 6e20 6c69 6e5f 6469 6374  dict in lin_dict
+0000ee00: 732e 6974 656d 7328 293a 0a20 2020 2020  s.items():.     
+0000ee10: 2020 2020 2020 2023 2067 6574 2074 6865         # get the
+0000ee20: 206e 756d 6265 7220 6f66 2075 6e69 7175   number of uniqu
+0000ee30: 6520 6f6c 6465 7374 206e 6f64 6573 2061  e oldest nodes a
+0000ee40: 6372 6f73 7320 7468 6520 6c69 6e65 6167  cross the lineag
+0000ee50: 6520 6469 6374 0a20 2020 2020 2020 2020  e dict.         
+0000ee60: 2020 206f 6c64 6573 745f 6e6f 6465 7320     oldest_nodes 
+0000ee70: 3d20 5b5b 2a6c 696e 5f64 6963 745b 6e6f  = [[*lin_dict[no
+0000ee80: 6465 5d2e 6b65 7973 280a 2020 2020 2020  de].keys(.      
+0000ee90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000eea0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000eeb0: 2020 2020 2020 295d 5b2d 315d 2066 6f72        )][-1] for
+0000eec0: 206e 6f64 6520 696e 206c 696e 5f64 6963   node in lin_dic
+0000eed0: 742e 6b65 7973 2829 5d0a 2020 2020 2020  t.keys()].      
+0000eee0: 2020 2020 2020 6e5f 756e 6971 7565 5f6e        n_unique_n
+0000eef0: 6f64 6573 203d 206c 656e 286e 702e 756e  odes = len(np.un
+0000ef00: 6971 7565 286f 6c64 6573 745f 6e6f 6465  ique(oldest_node
+0000ef10: 7329 290a 2020 2020 2020 2020 2020 2020  s)).            
+0000ef20: 7265 7375 6c74 5b6c 6f63 5d20 3d20 6e5f  result[loc] = n_
+0000ef30: 756e 6971 7565 5f6e 6f64 6573 203d 3d20  unique_nodes == 
+0000ef40: 310a 2020 2020 2020 2020 2320 6368 6563  1.        # chec
+0000ef50: 6b20 636f 616c 6573 6365 6e63 6520 6163  k coalescence ac
+0000ef60: 726f 7373 2061 6c6c 206c 6f63 692c 2069  ross all loci, i
+0000ef70: 6620 7265 7175 6573 7465 640a 2020 2020  f requested.    
+0000ef80: 2020 2020 6966 2061 6c6c 5f6c 6f63 693a      if all_loci:
+0000ef90: 0a20 2020 2020 2020 2020 2020 2072 6573  .            res
+0000efa0: 756c 7420 3d20 6e70 2e61 6c6c 285b 2a72  ult = np.all([*r
+0000efb0: 6573 756c 742e 7661 6c75 6573 2829 5d29  esult.values()])
+0000efc0: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
+0000efd0: 7265 7375 6c74 0a0a 0a20 2020 2064 6566  result...    def
+0000efe0: 205f 6361 6c63 5f6c 696e 6561 6765 5f73   _calc_lineage_s
+0000eff0: 7461 7473 2873 656c 662c 2069 6e64 6976  tats(self, indiv
+0000f000: 6964 733d 4e6f 6e65 2c20 6e6f 6465 733d  ids=None, nodes=
+0000f010: 4e6f 6e65 2c20 6c6f 6369 3d4e 6f6e 652c  None, loci=None,
+0000f020: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000f030: 2020 2020 2020 2020 2020 2020 2073 7461               sta
+0000f040: 7473 3d5b 2764 6972 272c 2027 6469 7374  ts=['dir', 'dist
+0000f050: 272c 2027 7469 6d65 272c 2027 7370 6565  ', 'time', 'spee
+0000f060: 6427 5d2c 0a20 2020 2020 2020 2020 2020  d'],.           
+0000f070: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f080: 2075 7365 5f69 6e64 6976 6964 735f 6375   use_individs_cu
+0000f090: 7272 5f70 6f73 3d54 7275 652c 0a20 2020  rr_pos=True,.   
+0000f0a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f0b0: 2020 2020 2020 2020 206d 6178 5f74 696d           max_tim
+0000f0c0: 655f 6167 6f3d 4e6f 6e65 2c20 6d69 6e5f  e_ago=None, min_
+0000f0d0: 7469 6d65 5f61 676f 3d4e 6f6e 6529 3a0a  time_ago=None):.
+0000f0e0: 2020 2020 2020 2020 2727 270a 2020 2020          '''.    
+0000f0f0: 2020 2020 4361 6c63 756c 6174 6520 7374      Calculate st
+0000f100: 6174 7320 666f 7220 7468 6520 6c69 6e65  ats for the line
+0000f110: 6167 6573 206f 6620 6120 6769 7665 6e20  ages of a given 
+0000f120: 7365 7420 6f66 206e 6f64 6573 2061 6e64  set of nodes and
+0000f130: 206c 6f63 693b 0a20 2020 2020 2020 2072   loci;.        r
+0000f140: 6574 7572 6e73 2064 6963 7420 6f66 2073  eturns dict of s
+0000f150: 7472 7563 743a 0a20 2020 2020 2020 2020  truct:.         
+0000f160: 2020 207b 6b3d 7374 6174 2c20 763d 7b6b     {k=stat, v={k
+0000f170: 3d6c 6f63 2c20 763d 5b76 616c 5f6e 6f64  =loc, v=[val_nod
+0000f180: 6531 202e 2e2e 2076 616c 5f6e 6f64 655f  e1 ... val_node_
+0000f190: 4e5d 7d7d 0a0a 2020 2020 2020 2020 4e4f  N]}}..        NO
+0000f1a0: 5445 3a20 6469 7265 6374 696f 6e73 2061  TE: directions a
+0000f1b0: 7265 2065 7870 7265 7373 6564 2061 7320  re expressed as 
+0000f1c0: 2763 6f6d 7061 7373 2720 6469 7265 6374  'compass' direct
+0000f1d0: 696f 6e73 0a20 2020 2020 2020 2020 2020  ions.           
+0000f1e0: 2020 2028 692e 652e 2c20 6578 7072 6573     (i.e., expres
+0000f1f0: 7365 6420 696e 2064 6567 7265 6573 2c20  sed in degrees, 
+0000f200: 636c 6f63 6b77 6973 6520 6672 6f6d 2030  clockwise from 0
+0000f210: 2064 6567 7265 6573 2061 7420 6e6f 7274   degrees at nort
+0000f220: 6829 0a20 2020 2020 2020 2027 2727 0a20  h).        '''. 
+0000f230: 2020 2020 2020 2023 2067 6574 2061 6c6c         # get all
+0000f240: 206e 6f64 6573 2066 6f72 2074 6865 2070   nodes for the p
+0000f250: 726f 7669 6465 6420 696e 6469 7669 6475  rovided individu
+0000f260: 616c 732c 206f 7220 666f 7220 616c 6c20  als, or for all 
+0000f270: 696e 6469 7669 6475 616c 732c 0a20 2020  individuals,.   
+0000f280: 2020 2020 2023 2069 6620 6e6f 6465 7320       # if nodes 
+0000f290: 4944 7320 6e6f 7420 7072 6f76 6964 6564  IDs not provided
+0000f2a0: 0a20 2020 2020 2020 2069 6620 6e6f 6465  .        if node
+0000f2b0: 7320 6973 204e 6f6e 653a 0a20 2020 2020  s is None:.     
+0000f2c0: 2020 2020 2020 2023 736f 7274 2061 6e64         #sort and
+0000f2d0: 2073 696d 706c 6966 7920 7468 6520 5461   simplify the Ta
+0000f2e0: 626c 6543 6f6c 6c65 6374 696f 6e2c 2069  bleCollection, i
+0000f2f0: 6620 6e65 6564 6564 0a20 2020 2020 2020  f needed.       
+0000f300: 2020 2020 2069 6620 6e6f 7420 7365 6c66       if not self
+0000f310: 2e5f 7463 5f73 6f72 7465 645f 616e 645f  ._tc_sorted_and_
+0000f320: 7369 6d70 6c69 6669 6564 3a0a 2020 2020  simplified:.    
+0000f330: 2020 2020 2020 2020 2020 2020 7365 6c66              self
+0000f340: 2e5f 736f 7274 5f61 6e64 5f73 696d 706c  ._sort_and_simpl
+0000f350: 6966 795f 7461 626c 655f 636f 6c6c 6563  ify_table_collec
+0000f360: 7469 6f6e 2829 0a20 2020 2020 2020 2020  tion().         
+0000f370: 2020 206e 6f64 6573 203d 2073 656c 662e     nodes = self.
+0000f380: 5f67 6574 5f6e 6f64 6573 2869 6e64 6976  _get_nodes(indiv
+0000f390: 6964 7329 0a20 2020 2020 2020 2023 2067  ids).        # g
+0000f3a0: 6574 2061 6c6c 206c 6f63 692c 2069 6620  et all loci, if 
+0000f3b0: 6c6f 6369 206e 6f74 2070 726f 7669 6465  loci not provide
+0000f3c0: 640a 2020 2020 2020 2020 6966 206c 6f63  d.        if loc
+0000f3d0: 6920 6973 204e 6f6e 653a 0a20 2020 2020  i is None:.     
+0000f3e0: 2020 2020 2020 206c 6f63 6920 3d20 5b2a         loci = [*
+0000f3f0: 7261 6e67 6528 7365 6c66 2e67 656e 5f61  range(self.gen_a
+0000f400: 7263 682e 4c29 5d0a 2020 2020 2020 2020  rch.L)].        
+0000f410: 6c69 6e5f 6469 6374 7320 3d20 7365 6c66  lin_dicts = self
+0000f420: 2e5f 6765 745f 6c69 6e65 6167 655f 6469  ._get_lineage_di
+0000f430: 6374 7328 6c6f 6369 2c20 6e6f 6465 733d  cts(loci, nodes=
+0000f440: 6e6f 6465 732c 0a20 2020 2020 2020 2020  nodes,.         
+0000f450: 2020 2020 2020 2075 7365 5f69 6e64 6976         use_indiv
+0000f460: 6964 735f 6375 7272 5f70 6f73 3d75 7365  ids_curr_pos=use
+0000f470: 5f69 6e64 6976 6964 735f 6375 7272 5f70  _individs_curr_p
+0000f480: 6f73 2c0a 2020 2020 2020 2020 2020 2020  os,.            
+0000f490: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f4a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f4b0: 6d61 785f 7469 6d65 5f61 676f 3d6d 6178  max_time_ago=max
+0000f4c0: 5f74 696d 655f 6167 6f2c 0a20 2020 2020  _time_ago,.     
+0000f4d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f4e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f4f0: 2020 2020 2020 206d 696e 5f74 696d 655f         min_time_
+0000f500: 6167 6f3d 6d69 6e5f 7469 6d65 5f61 676f  ago=min_time_ago
+0000f510: 290a 2020 2020 2020 2020 7374 6174 7320  ).        stats 
+0000f520: 3d20 7b73 7461 743a 207b 7d20 666f 7220  = {stat: {} for 
+0000f530: 7374 6174 2069 6e20 7374 6174 737d 0a20  stat in stats}. 
+0000f540: 2020 2020 2020 2066 6f72 2073 7461 7420         for stat 
+0000f550: 696e 2073 7461 7473 3a0a 2020 2020 2020  in stats:.      
+0000f560: 2020 2020 2020 666f 7220 6c6f 6320 696e        for loc in
+0000f570: 206c 6f63 693a 0a20 2020 2020 2020 2020   loci:.         
+0000f580: 2020 2020 2020 206c 6f63 5f73 7461 745f         loc_stat_
+0000f590: 6c69 7374 203d 205b 5d0a 2020 2020 2020  list = [].      
+0000f5a0: 2020 2020 2020 2020 2020 666f 7220 6e6f            for no
+0000f5b0: 6465 2069 6e20 6e6f 6465 733a 0a20 2020  de in nodes:.   
+0000f5c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f5d0: 206c 6f63 5f73 7461 745f 6c69 7374 2e61   loc_stat_list.a
+0000f5e0: 7070 656e 6428 5f63 616c 635f 6c69 6e65  ppend(_calc_line
+0000f5f0: 6167 655f 7374 6174 280a 2020 2020 2020  age_stat(.      
+0000f600: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f610: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f620: 2020 2020 2020 6c69 6e5f 6469 6374 735b        lin_dicts[
+0000f630: 6c6f 635d 5b6e 6f64 655d 2c20 7374 6174  loc][node], stat
+0000f640: 3d73 7461 7429 290a 2020 2020 2020 2020  =stat)).        
+0000f650: 2020 2020 2020 2020 7374 6174 735b 7374          stats[st
+0000f660: 6174 5d5b 6c6f 635d 203d 206c 6f63 5f73  at][loc] = loc_s
+0000f670: 7461 745f 6c69 7374 0a20 2020 2020 2020  tat_list.       
+0000f680: 2072 6574 7572 6e20 7374 6174 730a 0a0a   return stats...
+0000f690: 2020 2020 2320 6d65 7468 6f64 2074 6f20      # method to 
+0000f6a0: 6765 7420 696e 6469 7669 6473 2720 656e  get individs' en
+0000f6b0: 7669 726f 6e6d 656e 7420 7661 6c75 6573  vironment values
+0000f6c0: 0a20 2020 2064 6566 205f 6765 745f 6528  .    def _get_e(
+0000f6d0: 7365 6c66 2c20 6c79 725f 6e75 6d3d 4e6f  self, lyr_num=No
+0000f6e0: 6e65 2c20 696e 6469 7669 6473 3d4e 6f6e  ne, individs=Non
+0000f6f0: 6529 3a0a 2020 2020 2020 2020 6966 2069  e):.        if i
+0000f700: 6e64 6976 6964 7320 6973 204e 6f6e 653a  ndivids is None:
+0000f710: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
+0000f720: 6c79 725f 6e75 6d20 6973 204e 6f6e 653a  lyr_num is None:
+0000f730: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000f740: 2065 203d 206e 702e 6172 7261 7928 5b69   e = np.array([i
+0000f750: 6e64 2e65 2066 6f72 2069 6e64 2069 6e20  nd.e for ind in 
+0000f760: 7365 6c66 2e76 616c 7565 7328 295d 290a  self.values()]).
+0000f770: 2020 2020 2020 2020 2020 2020 656c 7365              else
+0000f780: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+0000f790: 2020 6520 3d20 6e70 2e61 7272 6179 285b    e = np.array([
+0000f7a0: 696e 642e 655b 6c79 725f 6e75 6d5d 2066  ind.e[lyr_num] f
+0000f7b0: 6f72 2069 6e64 2069 6e20 7365 6c66 2e76  or ind in self.v
+0000f7c0: 616c 7565 7328 295d 290a 2020 2020 2020  alues()]).      
+0000f7d0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+0000f7e0: 2020 2020 6967 203d 2069 7465 6d67 6574      ig = itemget
+0000f7f0: 7465 7228 2a69 6e64 6976 6964 7329 0a20  ter(*individs). 
+0000f800: 2020 2020 2020 2020 2020 2069 6620 6c79             if ly
+0000f810: 725f 6e75 6d20 6973 204e 6f6e 653a 0a20  r_num is None:. 
+0000f820: 2020 2020 2020 2020 2020 2020 2020 2065                 e
+0000f830: 203d 207b 693a 696e 642e 6520 666f 7220   = {i:ind.e for 
+0000f840: 692c 2069 6e64 2069 6e20 7365 6c66 2e69  i, ind in self.i
+0000f850: 7465 6d73 2829 7d0a 2020 2020 2020 2020  tems()}.        
+0000f860: 2020 2020 2020 2020 6520 3d20 6e70 2e61          e = np.a
+0000f870: 7272 6179 2869 6728 6529 290a 2020 2020  rray(ig(e)).    
+0000f880: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+0000f890: 2020 2020 2020 2020 2020 2020 2020 6520                e 
+0000f8a0: 3d20 7b69 3a69 6e64 2e65 5b6c 7972 5f6e  = {i:ind.e[lyr_n
+0000f8b0: 756d 5d20 666f 7220 692c 2069 6e64 2069  um] for i, ind i
+0000f8c0: 6e20 7365 6c66 2e69 7465 6d73 2829 7d0a  n self.items()}.
 0000f8d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f8e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f8f0: 226d 7573 7420 6265 2065 6974 6865 7220  "must be either 
-0000f900: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
-0000f910: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f920: 2020 2020 2020 2020 2020 2020 2254 7275              "Tru
-0000f930: 6520 6f72 2046 616c 7365 2e22 290a 0a20  e or False.").. 
-0000f940: 2020 2020 2020 2023 2067 6574 2074 6865         # get the
-0000f950: 206c 6973 7420 6f66 206c 6f63 690a 2020   list of loci.  
-0000f960: 2020 2020 2020 6966 206c 6f63 6920 6973        if loci is
-0000f970: 206e 6f74 204e 6f6e 653a 0a20 2020 2020   not None:.     
-0000f980: 2020 2020 2020 2061 7373 6572 7420 6e70         assert np
-0000f990: 2e69 7465 7261 626c 6528 6c6f 6369 292c  .iterable(loci),
-0000f9a0: 2028 2254 6865 2027 6c6f 6369 2720 6172   ("The 'loci' ar
-0000f9b0: 6775 6d65 6e74 206e 6565 6473 2022 0a20  gument needs ". 
-0000f9c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f9d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000f9e0: 2020 2020 2020 2265 6974 6865 7220 4e6f        "either No
-0000f9f0: 6e65 206f 7220 220a 2020 2020 2020 2020  ne or ".        
+0000f8e0: 6520 3d20 6e70 2e61 7272 6179 2869 6728  e = np.array(ig(
+0000f8f0: 6529 290a 2020 2020 2020 2020 7265 7475  e)).        retu
+0000f900: 726e 2065 0a0a 0a20 2020 2064 6566 205f  rn e...    def _
+0000f910: 6765 745f 6765 6e6f 7479 7065 7328 7365  get_genotypes(se
+0000f920: 6c66 2c20 6c6f 6369 3d4e 6f6e 652c 2069  lf, loci=None, i
+0000f930: 6e64 6976 6964 733d 4e6f 6e65 2c20 6269  ndivids=None, bi
+0000f940: 616c 6c65 6c69 633d 5472 7565 2c0a 2020  allelic=True,.  
+0000f950: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f960: 2020 2020 2061 735f 6469 6374 3d46 616c       as_dict=Fal
+0000f970: 7365 293a 0a20 2020 2020 2020 2023 206d  se):.        # m
+0000f980: 616b 6520 7375 7265 2061 735f 6469 6374  ake sure as_dict
+0000f990: 2061 6e64 2062 6961 6c6c 656c 6963 2061   and biallelic a
+0000f9a0: 7265 2054 7275 6520 6f72 2046 616c 7365  re True or False
+0000f9b0: 0a20 2020 2020 2020 2061 7373 6572 7420  .        assert 
+0000f9c0: 6173 5f64 6963 7420 696e 205b 5472 7565  as_dict in [True
+0000f9d0: 2c20 4661 6c73 655d 2c20 2822 5468 6520  , False], ("The 
+0000f9e0: 2761 735f 6469 6374 2720 6172 6775 6d65  'as_dict' argume
+0000f9f0: 6e74 2022 0a20 2020 2020 2020 2020 2020  nt ".           
 0000fa00: 2020 2020 2020 2020 2020 2020 2020 2020                  
 0000fa10: 2020 2020 2020 2020 2020 2020 2020 2022                 "
-0000fa20: 616e 2069 7465 7261 626c 6520 6f66 206c  an iterable of l
-0000fa30: 6f63 7573 2069 6e64 6963 6573 2e22 290a  ocus indices.").
-0000fa40: 2020 2020 2020 2020 2320 6765 7420 7468          # get th
-0000fa50: 6520 6c69 7374 206f 6620 696e 6469 7669  e list of indivi
-0000fa60: 6475 616c 730a 2020 2020 2020 2020 6966  duals.        if
-0000fa70: 2069 6e64 6976 6964 7320 6973 204e 6f6e   individs is Non
-0000fa80: 653a 0a20 2020 2020 2020 2020 2020 2069  e:.            i
-0000fa90: 6e64 6976 6964 7320 3d20 5b2a 7365 6c66  ndivids = [*self
-0000faa0: 5d0a 2020 2020 2020 2020 656c 7365 3a0a  ].        else:.
-0000fab0: 2020 2020 2020 2020 2020 2020 6173 7365              asse
-0000fac0: 7274 206e 702e 6974 6572 6162 6c65 2869  rt np.iterable(i
-0000fad0: 6e64 6976 6964 7329 2c20 2822 5468 6520  ndivids), ("The 
-0000fae0: 2769 6e64 6976 6964 7327 2061 7267 756d  'individs' argum
-0000faf0: 656e 7420 6e65 6564 7320 220a 2020 2020  ent needs ".    
+0000fa20: 6d75 7374 2062 6520 6569 7468 6572 2022  must be either "
+0000fa30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000fa40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fa50: 2020 2020 2020 2020 2020 2022 5472 7565             "True
+0000fa60: 206f 7220 4661 6c73 652e 2229 0a0a 2020   or False.")..  
+0000fa70: 2020 2020 2020 6173 7365 7274 2062 6961        assert bia
+0000fa80: 6c6c 656c 6963 2069 6e20 5b54 7275 652c  llelic in [True,
+0000fa90: 2046 616c 7365 5d2c 2028 2254 6865 2027   False], ("The '
+0000faa0: 6269 616c 6c65 6c69 6327 2061 7267 756d  biallelic' argum
+0000fab0: 656e 7420 220a 2020 2020 2020 2020 2020  ent ".          
+0000fac0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fad0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fae0: 226d 7573 7420 6265 2065 6974 6865 7220  "must be either 
+0000faf0: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
 0000fb00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000fb10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000fb20: 2020 2022 616e 2069 7465 7261 626c 6520     "an iterable 
-0000fb30: 6f66 2069 6e64 6976 6964 7561 6c20 4944  of individual ID
-0000fb40: 732e 2229 0a20 2020 2020 2020 2023 2073  s.").        # s
-0000fb50: 6f72 7420 7468 6520 696e 6469 7669 6475  ort the individu
-0000fb60: 616c 730a 2020 2020 2020 2020 2320 4e4f  als.        # NO
-0000fb70: 5445 3a20 6d61 6b65 7320 6f75 7470 7574  TE: makes output
-0000fb80: 2064 6174 6120 7374 7275 6375 7465 7320   data strucutes 
-0000fb90: 6e69 6365 7220 2873 696e 6365 2069 6e64  nicer (since ind
-0000fba0: 6976 6964 7561 6c73 2720 6964 7320 6172  ividuals' ids ar
-0000fbb0: 650a 2020 2020 2020 2020 2320 616c 7761  e.        # alwa
-0000fbc0: 7973 2069 6e20 6e75 6d65 7269 6361 6c20  ys in numerical 
-0000fbd0: 6f72 6465 7229 2c20 616e 6420 616c 736f  order), and also
-0000fbe0: 206d 616b 6573 2069 7420 7369 6d70 6c65   makes it simple
-0000fbf0: 7220 746f 2065 6e73 7572 6520 7468 6174  r to ensure that
-0000fc00: 0a20 2020 2020 2020 2023 2074 6865 7265  .        # there
-0000fc10: 2061 7265 206e 6f20 6d69 736d 6174 6368   are no mismatch
-0000fc20: 696e 6720 6964 7320 616e 6420 6765 6e6f  ing ids and geno
-0000fc30: 7479 7065 7320 696e 2064 6f77 6e73 7472  types in downstr
-0000fc40: 6561 6d20 6461 7461 0a20 2020 2020 2020  eam data.       
-0000fc50: 2023 2073 7472 7563 7475 7265 730a 2020   # structures.  
-0000fc60: 2020 2020 2020 696e 6469 7669 6473 203d        individs =
-0000fc70: 206e 702e 736f 7274 2869 6e64 6976 6964   np.sort(individ
-0000fc80: 7329 0a0a 2020 2020 2020 2020 2320 6966  s)..        # if
-0000fc90: 2075 7369 6e67 2074 736b 6974 2c20 6765   using tskit, ge
-0000fca0: 7420 6765 6e6f 7479 7065 7320 6672 6f6d  t genotypes from
-0000fcb0: 2074 6865 7265 0a20 2020 2020 2020 2069   there.        i
-0000fcc0: 6620 7365 6c66 2e67 656e 5f61 7263 682e  f self.gen_arch.
-0000fcd0: 7573 655f 7473 6b69 743a 0a20 2020 2020  use_tskit:.     
-0000fce0: 2020 2020 2020 2023 2073 6f72 7420 616e         # sort an
-0000fcf0: 6420 7369 6d70 6c69 6679 2074 6865 2074  d simplify the t
-0000fd00: 6162 6c65 2063 6f6c 6c65 6374 696f 6e0a  able collection.
-0000fd10: 2020 2020 2020 2020 2020 2020 2320 2874              # (t
-0000fd20: 6875 7320 6472 6f70 7069 6e67 2061 6e79  hus dropping any
-0000fd30: 2075 6e6e 6563 6573 7361 7279 2069 6e64   unnecessary ind
-0000fd40: 6976 6964 7561 6c73 2720 6461 7461 2069  ividuals' data i
-0000fd50: 6e20 7468 6572 6520 616e 6420 616c 736f  n there and also
-0000fd60: 0a20 2020 2020 2020 2020 2020 2023 206d  .            # m
-0000fd70: 616b 696e 6720 7468 6520 7461 626c 6573  aking the tables
-0000fd80: 2720 7374 7275 6374 7572 6520 7369 6d6c  ' structure siml
-0000fd90: 6572 2061 6e64 206d 6f72 6520 7072 6564  er and more pred
-0000fda0: 6963 7461 626c 6529 2c0a 2020 2020 2020  ictable),.      
-0000fdb0: 2020 2020 2020 7365 6c66 2e5f 736f 7274        self._sort
-0000fdc0: 5f73 696d 706c 6966 795f 7461 626c 655f  _simplify_table_
-0000fdd0: 636f 6c6c 6563 7469 6f6e 2829 0a20 2020  collection().   
-0000fde0: 2020 2020 2020 2020 2023 2074 6865 6e20           # then 
-0000fdf0: 6765 7420 7468 6520 5472 6565 5365 7175  get the TreeSequ
-0000fe00: 656e 6365 0a20 2020 2020 2020 2020 2020  ence.           
-0000fe10: 2074 7320 3d20 7365 6c66 2e5f 7463 2e74   ts = self._tc.t
-0000fe20: 7265 655f 7365 7175 656e 6365 2829 0a0a  ree_sequence()..
-0000fe30: 2020 2020 2020 2020 2020 2020 2320 6765              # ge
-0000fe40: 7420 7468 6520 6c69 7374 206f 6620 7468  t the list of th
-0000fe50: 6520 696e 6469 7669 6475 616c 7327 206e  e individuals' n
-0000fe60: 6f64 6573 0a20 2020 2020 2020 2020 2020  odes.           
-0000fe70: 2073 616d 706c 6573 5f74 6f5f 6b65 6570   samples_to_keep
-0000fe80: 203d 206e 702e 696e 7436 3428 6e70 2e68   = np.int64(np.h
-0000fe90: 7374 6163 6b28 5b0a 2020 2020 2020 2020  stack([.        
-0000fea0: 2020 2020 2020 2020 5b2a 7365 6c66 5b69          [*self[i
-0000feb0: 6e64 5d2e 5f6e 6f64 6573 5f74 6162 5f69  nd]._nodes_tab_i
-0000fec0: 6473 2e76 616c 7565 7328 295d 2066 6f72  ds.values()] for
-0000fed0: 2069 6e64 2069 6e20 696e 6469 7669 6473   ind in individs
-0000fee0: 5d29 290a 2020 2020 2020 2020 2020 2020  ])).            
-0000fef0: 6173 7365 7274 206c 656e 2873 616d 706c  assert len(sampl
-0000ff00: 6573 5f74 6f5f 6b65 6570 2920 3d3d 2073  es_to_keep) == s
-0000ff10: 656c 662e 6765 6e5f 6172 6368 2e78 202a  elf.gen_arch.x *
-0000ff20: 206c 656e 2869 6e64 6976 6964 7329 2c20   len(individs), 
-0000ff30: 2827 270a 2020 2020 2020 2020 2020 2020  (''.            
-0000ff40: 2020 2020 2020 2020 2020 2020 274e 756d              'Num
-0000ff50: 6265 7220 6f66 206e 6f64 6573 2064 6f65  ber of nodes doe
-0000ff60: 7320 6e6f 7420 6d61 7463 6820 6e75 6d62  s not match numb
-0000ff70: 6572 206f 6620 696e 6469 7669 6473 2127  er of individs!'
-0000ff80: 290a 0a20 2020 2020 2020 2020 2020 2023  )..            #
-0000ff90: 2067 6574 2068 6170 6c6f 7479 7065 7320   get haplotypes 
-0000ffa0: 666f 7220 616c 6c20 7361 6d70 6c65 730a  for all samples.
-0000ffb0: 2020 2020 2020 2020 2020 2020 2320 544f              # TO
-0000ffc0: 444f 3a20 6e70 2e77 6865 7265 2873 656c  DO: np.where(sel
-0000ffd0: 662e 5f74 632e 6e6f 6465 732e 666c 6167  f._tc.nodes.flag
-0000ffe0: 7329 5b31 5d20 6765 7473 206f 6e6c 7920  s)[1] gets only 
-0000fff0: 7468 6520 6e6f 6465 7320 7769 7468 0a20  the nodes with. 
-00010000: 2020 2020 2020 2020 2020 2023 2020 2020             #    
-00010010: 2020 2066 6c61 6773 206f 6620 312c 2069     flags of 1, i
-00010020: 2e65 2e20 6f6e 6c79 2027 7265 616c 2720  .e. only 'real' 
-00010030: 696e 6469 7669 6475 616c 730a 2020 2020  individuals.    
-00010040: 2020 2020 2020 2020 2320 2020 2020 2020          #       
-00010050: 7768 6f20 7765 7265 206e 6f74 2070 6172  who were not par
-00010060: 7420 6f66 2074 6865 206d 7370 7269 6d65  t of the msprime
-00010070: 2d73 6f75 7263 6564 0a20 2020 2020 2020  -sourced.       
-00010080: 2020 2020 2023 2020 2020 2020 2070 7265       #       pre
-00010090: 2d73 696d 756c 6174 696f 6e20 6661 6b65  -simulation fake
-000100a0: 2067 656e 6561 6c6f 6779 0a20 2020 2020   genealogy.     
-000100b0: 2020 2020 2020 2023 2054 4f44 4f3a 2068         # TODO: h
-000100c0: 6170 6c6f 7479 7065 7320 7265 7475 726e  aplotypes return
-000100d0: 6564 2069 6e20 6f72 6465 7220 6f66 2073  ed in order of s
-000100e0: 616d 706c 6573 3f3f 3f0a 2020 2020 2020  amples???.      
-000100f0: 2020 2020 2020 2368 6170 7320 3d20 5b6e        #haps = [n
-00010100: 702e 696e 7438 285b 2a68 6170 5d29 2066  p.int8([*hap]) f
-00010110: 6f72 206e 2c20 6861 7020 696e 207a 6970  or n, hap in zip
-00010120: 280a 2020 2020 2020 2020 2020 2020 2320  (.            # 
+0000fb10: 2020 2020 2020 2020 2020 2020 2254 7275              "Tru
+0000fb20: 6520 6f72 2046 616c 7365 2e22 290a 0a20  e or False.").. 
+0000fb30: 2020 2020 2020 2023 2067 6574 2074 6865         # get the
+0000fb40: 206c 6973 7420 6f66 206c 6f63 690a 2020   list of loci.  
+0000fb50: 2020 2020 2020 6966 206c 6f63 6920 6973        if loci is
+0000fb60: 206e 6f74 204e 6f6e 653a 0a20 2020 2020   not None:.     
+0000fb70: 2020 2020 2020 2061 7373 6572 7420 6e70         assert np
+0000fb80: 2e69 7465 7261 626c 6528 6c6f 6369 292c  .iterable(loci),
+0000fb90: 2028 2254 6865 2027 6c6f 6369 2720 6172   ("The 'loci' ar
+0000fba0: 6775 6d65 6e74 206e 6565 6473 2022 0a20  gument needs ". 
+0000fbb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fbc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fbd0: 2020 2020 2020 2265 6974 6865 7220 4e6f        "either No
+0000fbe0: 6e65 206f 7220 220a 2020 2020 2020 2020  ne or ".        
+0000fbf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fc00: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+0000fc10: 616e 2069 7465 7261 626c 6520 6f66 206c  an iterable of l
+0000fc20: 6f63 7573 2069 6e64 6963 6573 2e22 290a  ocus indices.").
+0000fc30: 2020 2020 2020 2020 2320 6765 7420 7468          # get th
+0000fc40: 6520 6c69 7374 206f 6620 696e 6469 7669  e list of indivi
+0000fc50: 6475 616c 730a 2020 2020 2020 2020 6966  duals.        if
+0000fc60: 2069 6e64 6976 6964 7320 6973 204e 6f6e   individs is Non
+0000fc70: 653a 0a20 2020 2020 2020 2020 2020 2069  e:.            i
+0000fc80: 6e64 6976 6964 7320 3d20 5b2a 7365 6c66  ndivids = [*self
+0000fc90: 5d0a 2020 2020 2020 2020 656c 7365 3a0a  ].        else:.
+0000fca0: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+0000fcb0: 7274 206e 702e 6974 6572 6162 6c65 2869  rt np.iterable(i
+0000fcc0: 6e64 6976 6964 7329 2c20 2822 5468 6520  ndivids), ("The 
+0000fcd0: 2769 6e64 6976 6964 7327 2061 7267 756d  'individs' argum
+0000fce0: 656e 7420 6e65 6564 7320 220a 2020 2020  ent needs ".    
+0000fcf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fd00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000fd10: 2020 2022 616e 2069 7465 7261 626c 6520     "an iterable 
+0000fd20: 6f66 2069 6e64 6976 6964 7561 6c20 4944  of individual ID
+0000fd30: 732e 2229 0a20 2020 2020 2020 2023 2073  s.").        # s
+0000fd40: 6f72 7420 7468 6520 696e 6469 7669 6475  ort the individu
+0000fd50: 616c 730a 2020 2020 2020 2020 2320 4e4f  als.        # NO
+0000fd60: 5445 3a20 6d61 6b65 7320 6f75 7470 7574  TE: makes output
+0000fd70: 2064 6174 6120 7374 7275 6375 7465 7320   data strucutes 
+0000fd80: 6e69 6365 7220 2873 696e 6365 2069 6e64  nicer (since ind
+0000fd90: 6976 6964 7561 6c73 2720 6964 7320 6172  ividuals' ids ar
+0000fda0: 650a 2020 2020 2020 2020 2320 616c 7761  e.        # alwa
+0000fdb0: 7973 2069 6e20 6e75 6d65 7269 6361 6c20  ys in numerical 
+0000fdc0: 6f72 6465 7229 2c20 616e 6420 616c 736f  order), and also
+0000fdd0: 206d 616b 6573 2069 7420 7369 6d70 6c65   makes it simple
+0000fde0: 7220 746f 2065 6e73 7572 6520 7468 6174  r to ensure that
+0000fdf0: 0a20 2020 2020 2020 2023 2074 6865 7265  .        # there
+0000fe00: 2061 7265 206e 6f20 6d69 736d 6174 6368   are no mismatch
+0000fe10: 696e 6720 6964 7320 616e 6420 6765 6e6f  ing ids and geno
+0000fe20: 7479 7065 7320 696e 2064 6f77 6e73 7472  types in downstr
+0000fe30: 6561 6d20 6461 7461 0a20 2020 2020 2020  eam data.       
+0000fe40: 2023 2073 7472 7563 7475 7265 730a 2020   # structures.  
+0000fe50: 2020 2020 2020 696e 6469 7669 6473 203d        individs =
+0000fe60: 206e 702e 736f 7274 2869 6e64 6976 6964   np.sort(individ
+0000fe70: 7329 0a0a 2020 2020 2020 2020 2320 6966  s)..        # if
+0000fe80: 2075 7369 6e67 2074 736b 6974 2c20 6765   using tskit, ge
+0000fe90: 7420 6765 6e6f 7479 7065 7320 6672 6f6d  t genotypes from
+0000fea0: 2074 6865 7265 0a20 2020 2020 2020 2069   there.        i
+0000feb0: 6620 7365 6c66 2e67 656e 5f61 7263 682e  f self.gen_arch.
+0000fec0: 7573 655f 7473 6b69 743a 0a20 2020 2020  use_tskit:.     
+0000fed0: 2020 2020 2020 2023 2073 6f72 7420 616e         # sort an
+0000fee0: 6420 7369 6d70 6c69 6679 2074 6865 2074  d simplify the t
+0000fef0: 6162 6c65 2063 6f6c 6c65 6374 696f 6e0a  able collection.
+0000ff00: 2020 2020 2020 2020 2020 2020 2320 2874              # (t
+0000ff10: 6875 7320 6472 6f70 7069 6e67 2061 6e79  hus dropping any
+0000ff20: 2075 6e6e 6563 6573 7361 7279 2069 6e64   unnecessary ind
+0000ff30: 6976 6964 7561 6c73 2720 6461 7461 2069  ividuals' data i
+0000ff40: 6e20 7468 6572 6520 616e 6420 616c 736f  n there and also
+0000ff50: 0a20 2020 2020 2020 2020 2020 2023 206d  .            # m
+0000ff60: 616b 696e 6720 7468 6520 7461 626c 6573  aking the tables
+0000ff70: 2720 7374 7275 6374 7572 6520 7369 6d6c  ' structure siml
+0000ff80: 6572 2061 6e64 206d 6f72 6520 7072 6564  er and more pred
+0000ff90: 6963 7461 626c 6529 2c0a 2020 2020 2020  ictable),.      
+0000ffa0: 2020 2020 2020 7365 6c66 2e5f 736f 7274        self._sort
+0000ffb0: 5f61 6e64 5f73 696d 706c 6966 795f 7461  _and_simplify_ta
+0000ffc0: 626c 655f 636f 6c6c 6563 7469 6f6e 2829  ble_collection()
+0000ffd0: 0a20 2020 2020 2020 2020 2020 2023 2074  .            # t
+0000ffe0: 6865 6e20 6765 7420 7468 6520 5472 6565  hen get the Tree
+0000fff0: 5365 7175 656e 6365 0a20 2020 2020 2020  Sequence.       
+00010000: 2020 2020 2074 7320 3d20 7365 6c66 2e5f       ts = self._
+00010010: 7463 2e74 7265 655f 7365 7175 656e 6365  tc.tree_sequence
+00010020: 2829 0a0a 2020 2020 2020 2020 2020 2020  ()..            
+00010030: 2320 6765 7420 7468 6520 6c69 7374 206f  # get the list o
+00010040: 6620 7468 6520 696e 6469 7669 6475 616c  f the individual
+00010050: 7327 206e 6f64 6573 0a20 2020 2020 2020  s' nodes.       
+00010060: 2020 2020 2073 616d 706c 6573 5f74 6f5f       samples_to_
+00010070: 6b65 6570 203d 206e 702e 696e 7436 3428  keep = np.int64(
+00010080: 6e70 2e68 7374 6163 6b28 5b0a 2020 2020  np.hstack([.    
+00010090: 2020 2020 2020 2020 2020 2020 5b2a 7365              [*se
+000100a0: 6c66 5b69 6e64 5d2e 5f6e 6f64 6573 5f74  lf[ind]._nodes_t
+000100b0: 6162 5f69 6473 2e76 616c 7565 7328 295d  ab_ids.values()]
+000100c0: 2066 6f72 2069 6e64 2069 6e20 696e 6469   for ind in indi
+000100d0: 7669 6473 5d29 290a 2020 2020 2020 2020  vids])).        
+000100e0: 2020 2020 6173 7365 7274 206c 656e 2873      assert len(s
+000100f0: 616d 706c 6573 5f74 6f5f 6b65 6570 2920  amples_to_keep) 
+00010100: 3d3d 2073 656c 662e 6765 6e5f 6172 6368  == self.gen_arch
+00010110: 2e78 202a 206c 656e 2869 6e64 6976 6964  .x * len(individ
+00010120: 7329 2c20 2827 270a 2020 2020 2020 2020  s), (''.        
 00010130: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00010140: 2020 2020 2020 2020 2020 2020 2020 6e70                np
-00010150: 2e77 6865 7265 2873 656c 662e 5f74 632e  .where(self._tc.
-00010160: 6e6f 6465 732e 666c 6167 7329 5b30 5d2c  nodes.flags)[0],
-00010170: 0a20 2020 2020 2020 2020 2020 2023 2020  .            #  
-00010180: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00010190: 2020 2020 2020 2020 2020 2020 2074 732e               ts.
-000101a0: 6861 706c 6f74 7970 6573 2829 2920 6966  haplotypes()) if
-000101b0: 206e 2069 6e20 7361 6d70 6c65 735d 0a20   n in samples]. 
-000101c0: 2020 2020 2020 2020 2020 2068 6170 735f             haps_
-000101d0: 6469 6374 203d 207b 733a 6820 666f 7220  dict = {s:h for 
-000101e0: 732c 2068 2069 6e20 7a69 7028 7473 2e73  s, h in zip(ts.s
-000101f0: 616d 706c 6573 2829 2c20 7473 2e68 6170  amples(), ts.hap
-00010200: 6c6f 7479 7065 7328 2929 7d0a 2020 2020  lotypes())}.    
-00010210: 2020 2020 2020 2020 6861 7073 203d 205b          haps = [
-00010220: 6e70 2e69 6e74 3828 5b2a 6861 7073 5f64  np.int8([*haps_d
-00010230: 6963 745b 735d 5d29 2066 6f72 2073 2069  ict[s]]) for s i
-00010240: 6e20 7361 6d70 6c65 735f 746f 5f6b 6565  n samples_to_kee
-00010250: 705d 0a0a 2020 2020 2020 2020 2020 2020  p]..            
-00010260: 2320 6765 7420 7468 6520 6765 6e6f 7479  # get the genoty
-00010270: 7065 7320 6279 2063 6f6d 6269 6e69 6e67  pes by combining
-00010280: 2065 6163 6820 636f 6e73 6563 7574 6976   each consecutiv
-00010290: 6520 6772 6f75 700a 2020 2020 2020 2020  e group.        
-000102a0: 2020 2020 2320 6f66 2078 2068 6170 6c6f      # of x haplo
-000102b0: 7479 7065 732c 2077 6865 7265 2078 2069  types, where x i
-000102c0: 7320 7468 6520 706c 6f69 6479 2c20 7468  s the ploidy, th
-000102d0: 656e 2076 7374 6163 6b69 6e67 2074 6865  en vstacking the
-000102e0: 6d0a 2020 2020 2020 2020 2020 2020 6772  m.            gr
-000102f0: 6f75 7065 645f 6861 7073 203d 207a 6970  ouped_haps = zip
-00010300: 282a 5b68 6170 735b 693a 3a73 656c 662e  (*[haps[i::self.
-00010310: 6765 6e5f 6172 6368 2e78 5d20 666f 7220  gen_arch.x] for 
-00010320: 6920 696e 2072 616e 6765 280a 2020 2020  i in range(.    
+00010140: 274e 756d 6265 7220 6f66 206e 6f64 6573  'Number of nodes
+00010150: 2064 6f65 7320 6e6f 7420 6d61 7463 6820   does not match 
+00010160: 6e75 6d62 6572 206f 6620 696e 6469 7669  number of indivi
+00010170: 6473 2127 290a 0a20 2020 2020 2020 2020  ds!')..         
+00010180: 2020 2023 2067 6574 2068 6170 6c6f 7479     # get haploty
+00010190: 7065 7320 666f 7220 616c 6c20 7361 6d70  pes for all samp
+000101a0: 6c65 730a 2020 2020 2020 2020 2020 2020  les.            
+000101b0: 2320 544f 444f 3a20 6e70 2e77 6865 7265  # TODO: np.where
+000101c0: 2873 656c 662e 5f74 632e 6e6f 6465 732e  (self._tc.nodes.
+000101d0: 666c 6167 7329 5b31 5d20 6765 7473 206f  flags)[1] gets o
+000101e0: 6e6c 7920 7468 6520 6e6f 6465 7320 7769  nly the nodes wi
+000101f0: 7468 0a20 2020 2020 2020 2020 2020 2023  th.            #
+00010200: 2020 2020 2020 2066 6c61 6773 206f 6620         flags of 
+00010210: 312c 2069 2e65 2e20 6f6e 6c79 2027 7265  1, i.e. only 're
+00010220: 616c 2720 696e 6469 7669 6475 616c 730a  al' individuals.
+00010230: 2020 2020 2020 2020 2020 2020 2320 2020              #   
+00010240: 2020 2020 7768 6f20 7765 7265 206e 6f74      who were not
+00010250: 2070 6172 7420 6f66 2074 6865 206d 7370   part of the msp
+00010260: 7269 6d65 2d73 6f75 7263 6564 0a20 2020  rime-sourced.   
+00010270: 2020 2020 2020 2020 2023 2020 2020 2020           #      
+00010280: 2070 7265 2d73 696d 756c 6174 696f 6e20   pre-simulation 
+00010290: 6661 6b65 2067 656e 6561 6c6f 6779 0a20  fake genealogy. 
+000102a0: 2020 2020 2020 2020 2020 2023 2054 4f44             # TOD
+000102b0: 4f3a 2068 6170 6c6f 7479 7065 7320 7265  O: haplotypes re
+000102c0: 7475 726e 6564 2069 6e20 6f72 6465 7220  turned in order 
+000102d0: 6f66 2073 616d 706c 6573 3f3f 3f0a 2020  of samples???.  
+000102e0: 2020 2020 2020 2020 2020 2368 6170 7320            #haps 
+000102f0: 3d20 5b6e 702e 696e 7438 285b 2a68 6170  = [np.int8([*hap
+00010300: 5d29 2066 6f72 206e 2c20 6861 7020 696e  ]) for n, hap in
+00010310: 207a 6970 280a 2020 2020 2020 2020 2020   zip(.          
+00010320: 2020 2320 2020 2020 2020 2020 2020 2020    #             
 00010330: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00010340: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00010350: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00010360: 2020 2020 2020 2020 7365 6c66 2e67 656e          self.gen
-00010370: 5f61 7263 682e 7829 5d29 0a20 2020 2020  _arch.x)]).     
-00010380: 2020 2020 2020 2067 7473 203d 205b 6e70         gts = [np
-00010390: 2e76 7374 6163 6b28 6829 2e54 2066 6f72  .vstack(h).T for
-000103a0: 2068 2069 6e20 6772 6f75 7065 645f 6861   h in grouped_ha
-000103b0: 7073 5d0a 0a20 2020 2020 2020 2023 206f  ps]..        # o
-000103c0: 7468 6572 7769 7365 2c20 6765 7420 7468  therwise, get th
-000103d0: 6520 6765 6e6f 7479 7065 7320 6672 6f6d  e genotypes from
-000103e0: 2074 6865 2069 6e64 6976 6964 7561 6c73   the individuals
-000103f0: 2064 6972 6563 746c 790a 2020 2020 2020   directly.      
-00010400: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
-00010410: 2020 2020 6774 7320 3d20 5b69 6e64 2e67      gts = [ind.g
-00010420: 2066 6f72 2069 6478 2c20 696e 6420 696e   for idx, ind in
-00010430: 2073 656c 662e 6974 656d 7328 2920 6966   self.items() if
-00010440: 2069 6478 2069 6e20 696e 6469 7669 6473   idx in individs
-00010450: 5d0a 0a20 2020 2020 2020 2023 2073 7562  ]..        # sub
-00010460: 7365 7420 6c6f 6369 2c20 6966 206e 6565  set loci, if nee
-00010470: 6465 640a 2020 2020 2020 2020 6966 206c  ded.        if l
-00010480: 6f63 6920 6973 206e 6f74 204e 6f6e 653a  oci is not None:
-00010490: 0a20 2020 2020 2020 2020 2020 2067 7473  .            gts
-000104a0: 203d 205b 6774 5b6c 6f63 692c 203a 5d20   = [gt[loci, :] 
-000104b0: 666f 7220 6774 2069 6e20 6774 735d 0a0a  for gt in gts]..
-000104c0: 2020 2020 2020 2020 2320 6765 7420 6d65          # get me
-000104d0: 616e 2067 656e 6f74 7970 6520 666f 7220  an genotype for 
-000104e0: 6561 6368 2069 6e64 6976 6964 7561 6c2c  each individual,
-000104f0: 2069 6620 6e65 6365 7373 6172 790a 2020   if necessary.  
-00010500: 2020 2020 2020 6966 206e 6f74 2062 6961        if not bia
-00010510: 6c6c 656c 6963 3a0a 2020 2020 2020 2020  llelic:.        
-00010520: 2020 2020 6774 7320 3d20 5b6e 702e 6d65      gts = [np.me
-00010530: 616e 2867 742c 2061 7869 733d 3129 2066  an(gt, axis=1) f
-00010540: 6f72 2067 7420 696e 2067 7473 5d0a 0a20  or gt in gts].. 
-00010550: 2020 2020 2020 2023 2073 7461 636b 2069         # stack i
-00010560: 6e74 6f20 7468 6520 7370 6563 696f 6d65  nto the speciome
-00010570: 2c20 6966 2064 6963 7420 6e6f 7420 7265  , if dict not re
-00010580: 7175 6573 7465 640a 2020 2020 2020 2020  quested.        
-00010590: 2320 2864 696d 7320 6172 6520 4e20 7820  # (dims are N x 
-000105a0: 4c20 7820 582c 2077 6865 7265 204e 3d6e  L x X, where N=n
-000105b0: 756d 2069 6e64 6976 6964 732c 204c 3d6e  um individs, L=n
-000105c0: 756d 206c 6f63 692c 0a20 2020 2020 2020  um loci,.       
-000105d0: 2023 2020 616e 6420 583d 706c 6f69 6479   #  and X=ploidy
-000105e0: 2069 6620 6269 616c 6c65 6c69 633d 5472   if biallelic=Tr
-000105f0: 7565 2c20 656c 7365 2058 3d31 290a 2020  ue, else X=1).  
-00010600: 2020 2020 2020 6966 206e 6f74 2061 735f        if not as_
-00010610: 6469 6374 3a0a 2020 2020 2020 2020 2020  dict:.          
-00010620: 2020 6774 7320 3d20 6e70 2e73 7461 636b    gts = np.stack
-00010630: 2867 7473 290a 0a20 2020 2020 2020 2023  (gts)..        #
-00010640: 206f 7220 656c 7365 2063 6173 7420 6173   or else cast as
-00010650: 2064 6963 740a 2020 2020 2020 2020 656c   dict.        el
-00010660: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-00010670: 6774 7320 3d20 7b69 6e64 6976 6964 3a20  gts = {individ: 
-00010680: 6774 2066 6f72 2069 6e64 6976 6964 2c20  gt for individ, 
-00010690: 6774 2069 6e20 7a69 7028 696e 6469 7669  gt in zip(indivi
-000106a0: 6473 2c20 6774 7329 7d0a 0a20 2020 2020  ds, gts)}..     
-000106b0: 2020 2072 6574 7572 6e20 6774 730a 0a0a     return gts...
-000106c0: 2020 2020 2363 6f6e 7665 6e69 656e 6365      #convenience
-000106d0: 206d 6574 686f 6420 666f 7220 6765 7474   method for gett
-000106e0: 696e 6720 6120 7363 616c 6172 2061 7474  ing a scalar att
-000106f0: 7269 6275 7465 2066 6f72 2073 6f6d 6520  ribute for some 
-00010700: 6f72 2061 6c6c 2069 6e64 6976 6964 730a  or all individs.
-00010710: 2020 2020 6465 6620 5f67 6574 5f73 6361      def _get_sca
-00010720: 6c61 725f 6174 7472 2873 656c 662c 2061  lar_attr(self, a
-00010730: 7474 725f 6e61 6d65 2c20 696e 6469 7669  ttr_name, indivi
-00010740: 6473 3d4e 6f6e 6529 3a0a 2020 2020 2020  ds=None):.      
-00010750: 2020 6966 2069 6e64 6976 6964 7320 6973    if individs is
-00010760: 204e 6f6e 653a 0a20 2020 2020 2020 2020   None:.         
-00010770: 2020 2076 616c 7320 3d20 6e70 2e61 7272     vals = np.arr
-00010780: 6179 285b 6765 7461 7474 7228 696e 642c  ay([getattr(ind,
-00010790: 2061 7474 725f 6e61 6d65 2920 666f 7220   attr_name) for 
-000107a0: 696e 6420 696e 2073 656c 662e 7661 6c75  ind in self.valu
-000107b0: 6573 2829 5d29 0a20 2020 2020 2020 2065  es()]).        e
-000107c0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
-000107d0: 2069 6720 3d20 6974 656d 6765 7474 6572   ig = itemgetter
-000107e0: 282a 696e 6469 7669 6473 290a 2020 2020  (*individs).    
-000107f0: 2020 2020 2020 2020 7661 6c73 203d 207b          vals = {
-00010800: 693a 2067 6574 6174 7472 2869 6e64 2c20  i: getattr(ind, 
-00010810: 6174 7472 5f6e 616d 6529 2066 6f72 2069  attr_name) for i
-00010820: 2c20 696e 6420 696e 2073 656c 662e 6974  , ind in self.it
-00010830: 656d 7328 297d 0a20 2020 2020 2020 2020  ems()}.         
-00010840: 2020 2076 616c 7320 3d20 6e70 2e61 7272     vals = np.arr
-00010850: 6179 2869 6728 7661 6c73 2929 0a20 2020  ay(ig(vals)).   
-00010860: 2020 2020 2072 6574 7572 6e20 7661 6c73       return vals
-00010870: 0a0a 2020 2020 2363 6f6e 7665 6e69 656e  ..    #convenien
-00010880: 6365 206d 6574 686f 6420 666f 7220 6765  ce method for ge
-00010890: 7474 696e 6720 6167 6520 6f66 2077 686f  tting age of who
-000108a0: 6c65 2073 7065 6369 6573 0a20 2020 2064  le species.    d
-000108b0: 6566 205f 6765 745f 6167 6528 7365 6c66  ef _get_age(self
-000108c0: 2c20 696e 6469 7669 6473 3d4e 6f6e 6529  , individs=None)
-000108d0: 3a0a 2020 2020 2020 2020 6167 6573 203d  :.        ages =
-000108e0: 2073 656c 662e 5f67 6574 5f73 6361 6c61   self._get_scala
-000108f0: 725f 6174 7472 2827 6167 6527 2c20 696e  r_attr('age', in
-00010900: 6469 7669 6473 3d69 6e64 6976 6964 7329  divids=individs)
-00010910: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
-00010920: 6167 6573 0a0a 2020 2020 2320 636f 6e76  ages..    # conv
-00010930: 656e 6965 6e63 6520 6d65 7468 6f64 2066  enience method f
-00010940: 6f72 2067 6574 7469 6e67 2077 686f 6c65  or getting whole
-00010950: 2073 7065 6369 6573 2720 7068 656e 6f74   species' phenot
-00010960: 7970 650a 2020 2020 6465 6620 5f67 6574  ype.    def _get
-00010970: 5f7a 2873 656c 662c 2074 7261 6974 5f6e  _z(self, trait_n
-00010980: 756d 3d4e 6f6e 652c 2069 6e64 6976 6964  um=None, individ
-00010990: 733d 4e6f 6e65 293a 0a20 2020 2020 2020  s=None):.       
-000109a0: 207a 7320 3d20 7365 6c66 2e5f 6765 745f   zs = self._get_
-000109b0: 7363 616c 6172 5f61 7474 7228 277a 272c  scalar_attr('z',
-000109c0: 2069 6e64 6976 6964 733d 696e 6469 7669   individs=indivi
-000109d0: 6473 290a 2020 2020 2020 2020 6966 2074  ds).        if t
-000109e0: 7261 6974 5f6e 756d 2069 7320 6e6f 7420  rait_num is not 
-000109f0: 4e6f 6e65 3a0a 2020 2020 2020 2020 2020  None:.          
-00010a00: 2020 7a73 203d 206e 702e 6174 6c65 6173    zs = np.atleas
-00010a10: 745f 3264 287a 7329 5b3a 2c74 7261 6974  t_2d(zs)[:,trait
-00010a20: 5f6e 756d 5d0a 2020 2020 2020 2020 7265  _num].        re
-00010a30: 7475 726e 207a 730a 0a20 2020 2023 636f  turn zs..    #co
-00010a40: 6e76 656e 6965 6e63 6520 6d65 7468 6f64  nvenience method
-00010a50: 2066 6f72 2067 6574 7469 6e67 2077 686f   for getting who
-00010a60: 6c65 2073 7065 6369 6573 2720 6669 746e  le species' fitn
-00010a70: 6573 7365 730a 2020 2020 6465 6620 5f67  esses.    def _g
-00010a80: 6574 5f66 6974 2873 656c 662c 2069 6e64  et_fit(self, ind
-00010a90: 6976 6964 7320 3d20 4e6f 6e65 293a 0a20  ivids = None):. 
-00010aa0: 2020 2020 2020 2066 6974 7320 3d20 7365         fits = se
-00010ab0: 6c66 2e5f 6765 745f 7363 616c 6172 5f61  lf._get_scalar_a
-00010ac0: 7474 7228 2766 6974 272c 2069 6e64 6976  ttr('fit', indiv
-00010ad0: 6964 733d 696e 6469 7669 6473 290a 2020  ids=individs).  
-00010ae0: 2020 2020 2020 7265 7475 726e 2066 6974        return fit
-00010af0: 730a 0a20 2020 2064 6566 205f 6361 6c63  s..    def _calc
-00010b00: 5f66 6974 6e65 7373 2873 656c 662c 2074  _fitness(self, t
-00010b10: 7261 6974 5f6e 756d 203d 204e 6f6e 652c  rait_num = None,
-00010b20: 2073 6574 5f66 6974 203d 2054 7275 6529   set_fit = True)
-00010b30: 3a0a 2020 2020 2020 2020 6669 7420 3d20  :.        fit = 
-00010b40: 5f63 616c 635f 6669 746e 6573 7328 7365  _calc_fitness(se
-00010b50: 6c66 2c20 7472 6169 745f 6e75 6d20 3d20  lf, trait_num = 
-00010b60: 7472 6169 745f 6e75 6d29 0a20 2020 2020  trait_num).     
-00010b70: 2020 2023 7365 7420 696e 6469 7669 6475     #set individu
-00010b80: 616c 7327 2066 6974 6e65 7373 2061 7474  als' fitness att
-00010b90: 7269 6275 7465 732c 2069 6620 696e 6469  ributes, if indi
-00010ba0: 6361 7465 640a 2020 2020 2020 2020 6966  cated.        if
-00010bb0: 2073 6574 5f66 6974 3a0a 2020 2020 2020   set_fit:.      
-00010bc0: 2020 2020 2020 7365 6c66 2e5f 7365 745f        self._set_
-00010bd0: 6669 7428 6669 7429 0a20 2020 2020 2020  fit(fit).       
-00010be0: 2072 6574 7572 6e20 6669 740a 0a20 2020   return fit..   
-00010bf0: 2064 6566 205f 6765 745f 646f 6d28 7365   def _get_dom(se
-00010c00: 6c66 2c20 6c6f 6375 7329 3a0a 2020 2020  lf, locus):.    
-00010c10: 2020 2020 7265 7475 726e 207b 6c6f 6375      return {locu
-00010c20: 733a 2073 656c 662e 6765 6e5f 6172 6368  s: self.gen_arch
-00010c30: 2e68 5b6c 6f63 7573 5d7d 0a0a 2020 2020  .h[locus]}..    
-00010c40: 6465 6620 5f67 6574 5f63 6f6f 7264 7328  def _get_coords(
-00010c50: 7365 6c66 2c20 696e 6469 7669 6473 3d4e  self, individs=N
-00010c60: 6f6e 652c 2061 735f 666c 6f61 743d 5472  one, as_float=Tr
-00010c70: 7565 293a 0a20 2020 2020 2020 2063 6f6f  ue):.        coo
-00010c80: 7264 7320 3d20 6c69 7374 286d 6170 2873  rds = list(map(s
-00010c90: 656c 662e 5f63 6f6f 7264 5f61 7474 7267  elf._coord_attrg
-00010ca0: 6574 7465 722c 2073 656c 662e 7661 6c75  etter, self.valu
-00010cb0: 6573 2829 2929 0a20 2020 2020 2020 2069  es())).        i
-00010cc0: 6620 696e 6469 7669 6473 2069 7320 6e6f  f individs is no
-00010cd0: 7420 4e6f 6e65 3a0a 2020 2020 2020 2020  t None:.        
-00010ce0: 2020 2020 6967 203d 2069 7465 6d67 6574      ig = itemget
-00010cf0: 7465 7228 2a69 6e64 6976 6964 7329 0a20  ter(*individs). 
-00010d00: 2020 2020 2020 2020 2020 2063 6f6f 7264             coord
-00010d10: 7320 3d20 6967 2864 6963 7428 7a69 7028  s = ig(dict(zip(
-00010d20: 5b2a 7365 6c66 5d2c 2063 6f6f 7264 7329  [*self], coords)
-00010d30: 2929 0a20 2020 2020 2020 2069 6620 6173  )).        if as
-00010d40: 5f66 6c6f 6174 3a0a 2020 2020 2020 2020  _float:.        
-00010d50: 2020 2020 636f 6f72 6473 203d 206e 702e      coords = np.
-00010d60: 666c 6f61 7436 3428 636f 6f72 6473 290a  float64(coords).
-00010d70: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
-00010d80: 2020 2020 2020 2020 2020 636f 6f72 6473            coords
-00010d90: 203d 206e 702e 696e 7433 3228 6e70 2e66   = np.int32(np.f
-00010da0: 6c6f 6f72 2863 6f6f 7264 7329 290a 2020  loor(coords)).  
-00010db0: 2020 2020 2020 2320 6d61 6b65 2073 7572        # make sur
-00010dc0: 6520 6974 2773 2061 7420 6c65 6173 7420  e it's at least 
-00010dd0: 3264 2028 696e 2063 6173 6520 6120 7369  2d (in case a si
-00010de0: 6e67 6c65 2069 6e64 6976 6964 7561 6c20  ngle individual 
-00010df0: 6973 2072 6571 7565 7374 6564 290a 2020  is requested).  
-00010e00: 2020 2020 2020 636f 6f72 6473 203d 206e        coords = n
-00010e10: 702e 6174 6c65 6173 745f 3264 2863 6f6f  p.atleast_2d(coo
-00010e20: 7264 7329 0a20 2020 2020 2020 2072 6574  rds).        ret
-00010e30: 7572 6e20 636f 6f72 6473 0a0a 2020 2020  urn coords..    
-00010e40: 6465 6620 5f67 6574 5f70 6c6f 745f 636f  def _get_plot_co
-00010e50: 6f72 6473 2873 656c 662c 2069 6e64 6976  ords(self, indiv
-00010e60: 6964 733d 4e6f 6e65 2c20 6365 6c6c 5f63  ids=None, cell_c
-00010e70: 6f6f 7264 733d 4661 6c73 6529 3a0a 2020  oords=False):.  
-00010e80: 2020 2020 2020 636f 6f72 6473 203d 2073        coords = s
-00010e90: 656c 662e 5f67 6574 5f63 6f6f 7264 7328  elf._get_coords(
-00010ea0: 696e 6469 7669 6473 3d69 6e64 6976 6964  individs=individ
-00010eb0: 7329 0a20 2020 2020 2020 2069 6620 6e6f  s).        if no
-00010ec0: 7420 6365 6c6c 5f63 6f6f 7264 733a 0a20  t cell_coords:. 
-00010ed0: 2020 2020 2020 2020 2020 2063 6f6f 7264             coord
-00010ee0: 735b 3a2c 2030 5d20 3d20 636f 6f72 6473  s[:, 0] = coords
-00010ef0: 5b3a 2c20 305d 202a 2073 656c 662e 5f6c  [:, 0] * self._l
-00010f00: 616e 645f 7265 735b 305d 202b 2073 656c  and_res[0] + sel
-00010f10: 662e 5f6c 616e 645f 756c 635b 305d 0a20  f._land_ulc[0]. 
-00010f20: 2020 2020 2020 2020 2020 2063 6f6f 7264             coord
-00010f30: 735b 3a2c 2031 5d20 3d20 636f 6f72 6473  s[:, 1] = coords
-00010f40: 5b3a 2c20 315d 202a 2073 656c 662e 5f6c  [:, 1] * self._l
-00010f50: 616e 645f 7265 735b 315d 202b 2073 656c  and_res[1] + sel
-00010f60: 662e 5f6c 616e 645f 756c 635b 315d 0a20  f._land_ulc[1]. 
-00010f70: 2020 2020 2020 2072 6574 7572 6e20 636f         return co
-00010f80: 6f72 6473 0a0a 2020 2020 6465 6620 5f67  ords..    def _g
-00010f90: 6574 5f63 656c 6c73 2873 656c 662c 2069  et_cells(self, i
-00010fa0: 6e64 6976 6964 733d 4e6f 6e65 293a 0a20  ndivids=None):. 
-00010fb0: 2020 2020 2020 2063 656c 6c73 203d 2073         cells = s
-00010fc0: 656c 662e 5f67 6574 5f63 6f6f 7264 7328  elf._get_coords(
-00010fd0: 696e 6469 7669 6473 3d69 6e64 6976 6964  individs=individ
-00010fe0: 732c 2061 735f 666c 6f61 743d 4661 6c73  s, as_float=Fals
-00010ff0: 6529 0a20 2020 2020 2020 2072 6574 7572  e).        retur
-00011000: 6e20 6365 6c6c 730a 0a20 2020 2064 6566  n cells..    def
-00011010: 205f 6765 745f 7828 7365 6c66 2c20 696e   _get_x(self, in
-00011020: 6469 7669 6473 3d4e 6f6e 6529 3a0a 2020  divids=None):.  
-00011030: 2020 2020 2020 636f 6f72 6473 203d 2073        coords = s
-00011040: 656c 662e 5f67 6574 5f63 6f6f 7264 7328  elf._get_coords(
-00011050: 696e 6469 7669 6473 3d69 6e64 6976 6964  individs=individ
-00011060: 7329 0a20 2020 2020 2020 2072 6574 7572  s).        retur
-00011070: 6e20 636f 6f72 6473 5b3a 2c20 305d 0a0a  n coords[:, 0]..
-00011080: 2020 2020 6465 6620 5f67 6574 5f79 2873      def _get_y(s
-00011090: 656c 662c 2069 6e64 6976 6964 733d 4e6f  elf, individs=No
-000110a0: 6e65 293a 0a20 2020 2020 2020 2063 6f6f  ne):.        coo
-000110b0: 7264 7320 3d20 7365 6c66 2e5f 6765 745f  rds = self._get_
-000110c0: 636f 6f72 6473 2869 6e64 6976 6964 733d  coords(individs=
-000110d0: 696e 6469 7669 6473 290a 2020 2020 2020  individs).      
-000110e0: 2020 7265 7475 726e 2063 6f6f 7264 735b    return coords[
-000110f0: 3a2c 2031 5d0a 0a20 2020 2023 206d 6574  :, 1]..    # met
-00011100: 686f 6420 746f 2072 6574 7572 6e20 616e  hod to return an
-00011110: 206e 2d6c 656e 6774 6820 6c69 7374 206f   n-length list o
-00011120: 6620 7261 6e64 6f6d 2069 6e64 6976 6964  f random individ
-00011130: 733b 0a20 2020 2023 2072 6574 7572 6e20  s;.    # return 
-00011140: 696e 6469 7669 6475 616c 732c 206f 7220  individuals, or 
-00011150: 696e 6469 6365 732c 2061 7320 696e 6469  indices, as indi
-00011160: 6361 7465 640a 2020 2020 6465 6620 5f67  cated.    def _g
-00011170: 6574 5f72 616e 646f 6d5f 696e 6469 7669  et_random_indivi
-00011180: 6475 616c 7328 7365 6c66 2c20 6e2c 2072  duals(self, n, r
-00011190: 6574 7572 6e5f 666f 726d 6174 3d27 696e  eturn_format='in
-000111a0: 6465 7827 293a 0a20 2020 2020 2020 2061  dex'):.        a
-000111b0: 7373 6572 7420 7265 7475 726e 5f66 6f72  ssert return_for
-000111c0: 6d61 7420 696e 205b 0a20 2020 2020 2020  mat in [.       
-000111d0: 2020 2020 2027 696e 6465 7827 2c20 2769       'index', 'i
-000111e0: 6e64 6976 6964 7561 6c27 5d2c 2028 2241  ndividual'], ("A
-000111f0: 7267 756d 656e 7420 7265 7475 726e 5f66  rgument return_f
-00011200: 6f72 6d61 7420 6361 6e20 7461 6b65 206f  ormat can take o
-00011210: 6e6c 7920 220a 2020 2020 2020 2020 2020  nly ".          
-00011220: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011230: 2020 2020 2020 2020 2020 2022 2769 6e64             "'ind
-00011240: 6578 2720 6f72 2027 696e 6469 7669 6475  ex' or 'individu
-00011250: 616c 2720 6173 2076 616c 7565 7320 220a  al' as values ".
-00011260: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011270: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011280: 2020 2020 2022 2864 6566 6175 6c74 7320       "(defaults 
-00011290: 746f 2027 696e 6465 7827 292e 2229 0a20  to 'index')."). 
-000112a0: 2020 2020 2020 2063 686f 6963 6573 203d         choices =
-000112b0: 2063 686f 6963 6573 203d 2072 2e63 686f   choices = r.cho
-000112c0: 6963 6528 6c69 7374 2872 616e 6765 286c  ice(list(range(l
-000112d0: 656e 2873 656c 6629 2929 2c20 6e2c 2072  en(self))), n, r
-000112e0: 6570 6c61 6365 3d46 616c 7365 290a 2020  eplace=False).  
-000112f0: 2020 2020 2020 696e 6473 203d 206e 702e        inds = np.
-00011300: 6172 7261 7928 6c69 7374 2873 656c 662e  array(list(self.
-00011310: 6b65 7973 2829 2929 5b63 686f 6963 6573  keys()))[choices
-00011320: 5d0a 2020 2020 2020 2020 6966 2072 6574  ].        if ret
-00011330: 7572 6e5f 666f 726d 6174 3d3d 2769 6e64  urn_format=='ind
-00011340: 6976 6964 7561 6c27 3a0a 2020 2020 2020  ividual':.      
-00011350: 2020 2020 2020 696e 6473 203d 205b 7365        inds = [se
-00011360: 6c66 5b69 6e64 5d20 666f 7220 696e 6420  lf[ind] for ind 
-00011370: 696e 2069 6e64 735d 0a20 2020 2020 2020  in inds].       
-00011380: 2072 6574 7572 6e28 696e 6473 290a 0a20   return(inds).. 
-00011390: 2020 2023 206d 6574 686f 6420 746f 2072     # method to r
-000113a0: 6564 7563 6520 6120 7370 7020 746f 2073  educe a spp to s
-000113b0: 6f6d 6520 6368 6f73 656e 206e 756d 6265  ome chosen numbe
-000113c0: 722c 2062 7920 7261 6e64 6f6d 6c79 2072  r, by randomly r
-000113d0: 656d 6f76 696e 670a 2020 2020 2320 4e5f  emoving.    # N_
-000113e0: 6375 7272 5f74 202d 206e 2069 6e64 6976  curr_t - n indiv
-000113f0: 6964 7561 6c73 0a20 2020 2064 6566 205f  iduals.    def _
-00011400: 7265 6475 6365 2873 656c 662c 206e 293a  reduce(self, n):
-00011410: 0a20 2020 2020 2020 2069 6e64 7320 3d20  .        inds = 
-00011420: 5b2a 7365 6c66 5d0a 2020 2020 2020 2020  [*self].        
-00011430: 6b65 6570 203d 206e 702e 7261 6e64 6f6d  keep = np.random
-00011440: 2e63 686f 6963 6528 696e 6473 2c20 6e2c  .choice(inds, n,
-00011450: 2072 6570 6c61 6365 3d46 616c 7365 290a   replace=False).
-00011460: 2020 2020 2020 2020 666f 7220 696e 6420          for ind 
-00011470: 696e 2069 6e64 733a 0a20 2020 2020 2020  in inds:.       
-00011480: 2020 2020 2069 6620 696e 6420 6e6f 7420       if ind not 
-00011490: 696e 206b 6565 703a 0a20 2020 2020 2020  in keep:.       
-000114a0: 2020 2020 2020 2020 2073 656c 662e 706f           self.po
-000114b0: 7028 696e 6429 0a0a 2020 2020 2375 7365  p(ind)..    #use
-000114c0: 2074 6865 206b 645f 7472 6565 2061 7474   the kd_tree att
-000114d0: 7269 6275 7465 2074 6f20 6669 6e64 206d  ribute to find m
-000114e0: 6174 696e 6720 7061 6972 7320 6569 7468  ating pairs eith
-000114f0: 6572 0a20 2020 2023 7769 7468 696e 2074  er.    #within t
-00011500: 6865 2073 7065 6369 6573 2c20 6966 2077  he species, if w
-00011510: 6974 6869 6e20 3d3d 2054 7275 652c 206f  ithin == True, o
-00011520: 7220 6265 7477 6565 6e20 7468 6520 7370  r between the sp
-00011530: 6563 6965 730a 2020 2020 2361 6e64 2074  ecies.    #and t
-00011540: 6865 2070 6f69 6e74 7320 7072 6f76 6964  he points provid
-00011550: 6564 2c20 6966 2077 6974 6869 6e20 3d3d  ed, if within ==
-00011560: 2046 616c 7365 2061 6e64 2070 6f69 6e74   False and point
-00011570: 7320 6973 206e 6f74 204e 6f6e 650a 2020  s is not None.  
-00011580: 2020 6465 6620 5f67 6574 5f6d 6174 696e    def _get_matin
-00011590: 675f 7061 6972 7328 7365 6c66 2c20 7769  g_pairs(self, wi
-000115a0: 7468 696e 3d54 7275 652c 2063 6f6f 7264  thin=True, coord
-000115b0: 733d 4e6f 6e65 2c0a 2020 2020 2020 2020  s=None,.        
-000115c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000115d0: 2020 6368 6f6f 7365 5f6e 6561 7265 7374    choose_nearest
-000115e0: 3d46 616c 7365 2c20 696e 7665 7273 655f  =False, inverse_
-000115f0: 6469 7374 5f6d 6174 696e 673d 4661 6c73  dist_mating=Fals
-00011600: 6529 3a0a 2020 2020 2020 2020 234e 4f54  e):.        #NOT
-00011610: 453a 2049 6e20 6c69 6575 206f 6620 6120  E: In lieu of a 
-00011620: 6d6f 7265 2073 6f70 6869 7374 6963 6174  more sophisticat
-00011630: 6564 2077 6179 206f 660a 2020 2020 2020  ed way of.      
-00011640: 2020 2364 6574 6572 6d69 6e69 6e67 2077    #determining w
-00011650: 6865 7468 6572 2074 6865 206b 645f 7472  hether the kd_tr
-00011660: 6565 206e 6565 6473 2074 6f20 6265 2075  ee needs to be u
-00011670: 7064 6174 6564 200a 2020 2020 2020 2020  pdated .        
-00011680: 2328 692e 652e 2069 6620 7468 6520 7370  #(i.e. if the sp
-00011690: 6563 6965 7320 6861 7665 2075 6e64 6572  ecies have under
-000116a0: 676f 6e65 206d 6f76 656d 656e 742c 206d  gone movement, m
-000116b0: 6174 696e 672c 0a20 2020 2020 2020 2023  ating,.        #
-000116c0: 6f72 206d 6f72 7461 6c69 7479 2073 696e  or mortality sin
-000116d0: 6365 2074 6865 206c 6173 7420 7469 6d65  ce the last time
-000116e0: 2069 7420 7761 7320 0a20 2020 2020 2020   it was .       
-000116f0: 2023 636f 6e73 7472 7563 7465 6429 2c20   #constructed), 
-00011700: 616e 6420 696e 2061 6e20 6566 666f 7274  and in an effort
-00011710: 2074 6f20 6d69 6e69 6d69 7a65 2074 6865   to minimize the
-00011720: 206e 756d 6265 720a 2020 2020 2020 2020   number.        
-00011730: 236f 6620 7469 6d65 7320 6974 2069 7320  #of times it is 
-00011740: 636f 6e73 7472 7563 7465 6420 6561 6368  constructed each
-00011750: 2074 696d 6520 2873 696e 6365 0a20 2020   time (since.   
-00011760: 2020 2020 2023 6974 2773 206e 6f74 2061       #it's not a
-00011770: 6e20 696e 636f 6e73 6571 7565 6e74 6961  n inconsequentia
-00011780: 6c20 7275 6e74 696d 652c 2073 6f20 7465  l runtime, so te
-00011790: 6c6c 696e 6720 7468 650a 2020 2020 2020  lling the.      
-000117a0: 2020 236d 6f64 656c 2074 6f20 7265 6275    #model to rebu
-000117b0: 696c 6420 6974 2061 6674 6572 2065 6163  ild it after eac
-000117c0: 6820 6d6f 7665 2c20 6269 7274 682c 206f  h move, birth, o
-000117d0: 720a 2020 2020 2020 2020 2364 6561 7468  r.        #death
-000117e0: 2073 7465 7020 636f 756c 6420 6265 2075   step could be u
-000117f0: 6e6e 6365 7373 6172 696c 7920 636f 7374  nncessarily cost
-00011800: 6c79 292c 2066 6f72 206e 6f77 2049 0a20  ly), for now I. 
-00011810: 2020 2020 2020 2023 616d 206a 7573 7420         #am just 
-00011820: 7465 6c6c 696e 6720 7468 6520 7472 6565  telling the tree
-00011830: 2074 6f20 6265 2072 6562 7569 6c74 2065   to be rebuilt e
-00011840: 6163 6820 7469 6d65 200a 2020 2020 2020  ach time .      
-00011850: 2020 2374 6865 2073 7070 2e5f 6765 745f    #the spp._get_
-00011860: 6d61 7469 6e67 5f70 6169 7273 2829 206d  mating_pairs() m
-00011870: 6574 686f 6420 6973 2063 616c 6c65 6421  ethod is called!
-00011880: 0a20 2020 2020 2020 2073 656c 662e 5f73  .        self._s
-00011890: 6574 5f6b 645f 7472 6565 2829 0a0a 2020  et_kd_tree()..  
-000118a0: 2020 2020 2020 2320 6966 206d 6174 696e        # if matin
-000118b0: 675f 7261 6469 7573 2069 7320 4e6f 6e65  g_radius is None
-000118c0: 2c20 7468 656e 206a 7573 7420 7573 6520  , then just use 
-000118d0: 5772 6967 6874 2d46 6973 6865 720a 2020  Wright-Fisher.  
-000118e0: 2020 2020 2020 2320 7374 796c 6520 7061        # style pa
-000118f0: 6e6d 6978 6961 2028 6472 6177 2077 6974  nmixia (draw wit
-00011900: 6820 7265 706c 6163 656d 656e 7420 6120  h replacement a 
-00011910: 7361 6d70 6c65 0a20 2020 2020 2020 2023  sample.        #
-00011920: 206f 6620 7369 7a65 203d 204e 742a 622c   of size = Nt*b,
-00011930: 2077 6865 7265 204e 7420 6973 2074 6865   where Nt is the
-00011940: 200a 2020 2020 2020 2020 2320 6375 7272   .        # curr
-00011950: 656e 7420 706f 7020 7369 7a65 2061 6e64  ent pop size and
-00011960: 2062 2069 7320 7468 6520 6269 7274 6820   b is the birth 
-00011970: 7261 7465 2028 692e 652e 206d 6174 696e  rate (i.e. matin
-00011980: 6720 7072 6f62 6162 696c 6974 7929 3b0a  g probability);.
-00011990: 2020 2020 2020 2020 2320 7468 6174 2073          # that s
-000119a0: 616d 706c 6520 7265 7072 6573 656e 7473  ample represents
-000119b0: 2061 6c6c 206d 6174 696e 6720 696e 6469   all mating indi
-000119c0: 7669 6475 616c 732c 2061 6e64 2065 6163  viduals, and eac
-000119d0: 6820 6f66 2074 686f 7365 0a20 2020 2020  h of those.     
-000119e0: 2020 2023 2069 6e64 6976 6964 7561 6c73     # individuals
-000119f0: 2072 616e 646f 6d6c 7920 6368 6f6f 7365   randomly choose
-00011a00: 7320 6974 7320 6d61 7465 0a20 2020 2020  s its mate.     
-00011a10: 2020 2069 6620 7365 6c66 2e6d 6174 696e     if self.matin
-00011a20: 675f 7261 6469 7573 2069 7320 4e6f 6e65  g_radius is None
-00011a30: 3a0a 2020 2020 2020 2020 2020 2020 2320  :.            # 
-00011a40: 6472 6177 2061 206e 756d 6265 7220 6f66  draw a number of
-00011a50: 206d 6174 696e 6720 696e 6469 7669 6475   mating individu
-00011a60: 616c 7320 6173 2061 2062 696e 6f6d 6961  als as a binomia
-00011a70: 6c20 7276 0a20 2020 2020 2020 2020 2020  l rv.           
-00011a80: 2023 2077 6974 6820 6e75 6d20 7472 6961   # with num tria
-00011a90: 6c73 2065 7175 616c 2074 6f20 706f 7020  ls equal to pop 
-00011aa0: 7369 7a65 2061 6e64 2070 726f 6261 6269  size and probabi
-00011ab0: 6c69 7479 2065 7175 616c 2074 6f0a 2020  lity equal to.  
-00011ac0: 2020 2020 2020 2020 2020 2320 7468 6520            # the 
-00011ad0: 7370 6563 6965 7327 2062 6972 7468 2072  species' birth r
-00011ae0: 6174 650a 2020 2020 2020 2020 2020 2020  ate.            
-00011af0: 6966 2073 656c 662e 6220 3c20 313a 0a20  if self.b < 1:. 
-00011b00: 2020 2020 2020 2020 2020 2020 2020 206e                 n
-00011b10: 5f6d 6174 6573 203d 206e 702e 7261 6e64  _mates = np.rand
-00011b20: 6f6d 2e62 696e 6f6d 6961 6c28 6e3d 6c65  om.binomial(n=le
-00011b30: 6e28 7365 6c66 292c 2070 3d73 656c 662e  n(self), p=self.
-00011b40: 6229 0a20 2020 2020 2020 2020 2020 2065  b).            e
-00011b50: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
-00011b60: 2020 2020 206e 5f6d 6174 6573 203d 206c       n_mates = l
-00011b70: 656e 2873 656c 6629 0a20 2020 2020 2020  en(self).       
-00011b80: 2020 2020 2074 7265 655f 696e 6473 203d       tree_inds =
-00011b90: 2073 656c 662e 5f6b 645f 7472 6565 2e74   self._kd_tree.t
-00011ba0: 7265 652e 696e 6469 6365 730a 2020 2020  ree.indices.    
-00011bb0: 2020 2020 2020 2020 2320 6472 6177 2032          # draw 2
-00011bc0: 2a6e 5f6d 6174 6573 206d 6174 696e 6720  *n_mates mating 
-00011bd0: 696e 6469 7669 6475 616c 732c 2077 6974  individuals, wit
-00011be0: 6820 7265 706c 6163 656d 656e 7420 2861  h replacement (a
-00011bf0: 7320 696e 2057 460a 2020 2020 2020 2020  s in WF.        
-00011c00: 2020 2020 2320 6d6f 6465 6c29 2c20 7468      # model), th
-00011c10: 656e 2066 6f6c 6420 696e 746f 2061 6e20  en fold into an 
-00011c20: 6e5f 6d61 7465 7320 7820 3220 6d61 7465  n_mates x 2 mate
-00011c30: 2d70 6169 7273 2061 7272 6179 0a20 2020  -pairs array.   
-00011c40: 2020 2020 2020 2020 2070 6169 7273 203d           pairs =
-00011c50: 206e 702e 7261 6e64 6f6d 2e63 686f 6963   np.random.choic
-00011c60: 6528 7472 6565 5f69 6e64 732c 2072 6570  e(tree_inds, rep
-00011c70: 6c61 6365 3d54 7275 652c 0a20 2020 2020  lace=True,.     
-00011c80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011c90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011ca0: 7369 7a65 3d6e 5f6d 6174 6573 2a32 292e  size=n_mates*2).
-00011cb0: 7265 7368 6170 6528 286e 5f6d 6174 6573  reshape((n_mates
-00011cc0: 2c20 3229 290a 2020 2020 2020 2020 2020  , 2)).          
-00011cd0: 2020 2320 6765 7420 7269 6420 6f66 2073    # get rid of s
-00011ce0: 656c 6669 6e67 2070 6169 7273 0a20 2020  elfing pairs.   
-00011cf0: 2020 2020 2020 2020 2070 6169 7273 203d           pairs =
-00011d00: 206e 702e 6172 7261 7928 5b6c 6973 7428   np.array([list(
-00011d10: 7061 6972 2920 666f 7220 7061 6972 2069  pair) for pair i
-00011d20: 6e20 6c69 7374 280a 2020 2020 2020 2020  n list(.        
-00011d30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011d40: 2020 2020 2020 2020 2020 2020 6d61 7028              map(
-00011d50: 7365 742c 2070 6169 7273 2929 2069 6620  set, pairs)) if 
-00011d60: 6c65 6e28 7061 6972 2920 3d3d 2032 5d29  len(pair) == 2])
-00011d70: 0a20 2020 2020 2020 2020 2020 2070 6169  .            pai
-00011d80: 7273 203d 206e 702e 6172 7261 7928 7061  rs = np.array(pa
-00011d90: 6972 7329 0a0a 2020 2020 2020 2020 2320  irs)..        # 
-00011da0: 6f74 6865 7277 6973 652c 2063 686f 6f73  otherwise, choos
-00011db0: 6520 6d61 7465 7320 7573 696e 6720 6d61  e mates using ma
-00011dc0: 7469 6e67 2072 6164 6975 730a 2020 2020  ting radius.    
-00011dd0: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
-00011de0: 2020 2020 2020 2369 6620 6e65 6967 6862        #if neighb
-00011df0: 6f72 7320 6172 6520 746f 2062 6520 666f  ors are to be fo
-00011e00: 756e 6420 7769 7468 696e 2074 6865 2073  und within the s
-00011e10: 7065 6369 6573 2c0a 2020 2020 2020 2020  pecies,.        
-00011e20: 2020 2020 2373 6574 2063 6f6f 7264 7320      #set coords 
-00011e30: 746f 2073 656c 662e 636f 6f72 6473 2028  to self.coords (
-00011e40: 6f74 6865 7277 6973 652c 2074 6865 2063  otherwise, the c
-00011e50: 6f6f 7264 7320 746f 0a20 2020 2020 2020  oords to.       
-00011e60: 2020 2020 2023 6669 6e64 206e 6561 7265       #find neare
-00011e70: 7374 206e 6569 6768 626f 7273 2077 6974  st neighbors wit
-00011e80: 6820 7368 6f75 6c64 2068 6176 6520 6265  h should have be
-00011e90: 656e 2070 726f 7669 6465 6429 0a20 2020  en provided).   
-00011ea0: 2020 2020 2020 2020 2069 6620 7769 7468           if with
-00011eb0: 696e 3a0a 2020 2020 2020 2020 2020 2020  in:.            
-00011ec0: 2020 2020 636f 6f72 6473 203d 2073 656c      coords = sel
-00011ed0: 662e 636f 6f72 6473 0a0a 2020 2020 2020  f.coords..      
-00011ee0: 2020 2020 2020 2371 7565 7279 2074 6865        #query the
-00011ef0: 2074 7265 6520 746f 2067 6574 206d 6174   tree to get mat
-00011f00: 696e 6720 7061 6972 730a 2020 2020 2020  ing pairs.      
-00011f10: 2020 2020 2020 7061 6972 7320 3d20 7365        pairs = se
-00011f20: 6c66 2e5f 6b64 5f74 7265 652e 5f67 6574  lf._kd_tree._get
-00011f30: 5f6d 6174 696e 675f 7061 6972 7328 636f  _mating_pairs(co
-00011f40: 6f72 6473 3d63 6f6f 7264 732c 0a20 2020  ords=coords,.   
-00011f50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011f60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011f70: 2020 2020 2020 2020 2020 2020 2064 6973               dis
-00011f80: 743d 7365 6c66 2e6d 6174 696e 675f 7261  t=self.mating_ra
-00011f90: 6469 7573 2c0a 2020 2020 2020 2020 2020  dius,.          
-00011fa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011fb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011fc0: 2020 2020 2020 6368 6f6f 7365 5f6e 6561        choose_nea
-00011fd0: 7265 7374 3d63 686f 6f73 655f 6e65 6172  rest=choose_near
-00011fe0: 6573 742c 0a20 2020 2020 2020 2020 2020  est,.           
-00011ff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012000: 2020 2020 2020 2020 2020 2020 2069 6e76               inv
-00012010: 6572 7365 5f64 6973 745f 6d61 7469 6e67  erse_dist_mating
-00012020: 3d69 6e76 6572 7365 5f64 6973 745f 6d61  =inverse_dist_ma
-00012030: 7469 6e67 290a 0a20 2020 2020 2020 2020  ting)..         
-00012040: 2020 2023 2075 7365 2074 6865 2073 7065     # use the spe
-00012050: 6369 6573 2720 6269 7274 6820 7261 7465  cies' birth rate
-00012060: 2074 6f20 6465 6369 6465 2028 6173 2062   to decide (as b
-00012070: 6572 6e6f 756c 6c69 2064 7261 7773 290a  ernoulli draws).
-00012080: 2020 2020 2020 2020 2020 2020 2320 7768              # wh
-00012090: 6574 6865 7220 6561 6368 2070 6169 7220  ether each pair 
-000120a0: 6361 6e20 6d61 7465 0a20 2020 2020 2020  can mate.       
-000120b0: 2020 2020 2063 616e 5f6d 6174 6520 3d20       can_mate = 
-000120c0: 6e70 2e62 6f6f 6c38 286e 702e 7261 6e64  np.bool8(np.rand
-000120d0: 6f6d 2e62 696e 6f6d 6961 6c28 6e3d 312c  om.binomial(n=1,
-000120e0: 2070 3d73 656c 662e 622c 0a20 2020 2020   p=self.b,.     
-000120f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012100: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012110: 2020 2020 2020 2020 2020 2020 2020 7369                si
-00012120: 7a65 3d70 6169 7273 2e73 6861 7065 5b30  ze=pairs.shape[0
-00012130: 5d29 290a 2020 2020 2020 2020 2020 2020  ])).            
-00012140: 7061 6972 7320 3d20 7061 6972 735b 6361  pairs = pairs[ca
-00012150: 6e5f 6d61 7465 2c20 3a5d 0a20 2020 2020  n_mate, :].     
-00012160: 2020 2072 6574 7572 6e20 7061 6972 730a     return pairs.
-00012170: 0a0a 2020 2020 6465 6620 5f6d 616b 655f  ..    def _make_
-00012180: 6765 615f 6466 2873 656c 662c 206c 7972  gea_df(self, lyr
-00012190: 5f6e 756d 3d31 293a 0a20 2020 2020 2020  _num=1):.       
-000121a0: 2022 2222 0a0a 2020 2020 2020 2020 4d61   """..        Ma
-000121b0: 6b65 7320 6120 6461 7461 6672 616d 6520  kes a dataframe 
-000121c0: 666f 7220 4745 4120 616e 616c 7973 6973  for GEA analysis
-000121d0: 2063 6f6e 7461 696e 696e 6720 6765 6e6f   containing geno
-000121e0: 7479 7065 732c 2063 6f6f 7264 696e 6174  types, coordinat
-000121f0: 6573 2c0a 2020 2020 2020 2020 616e 6420  es,.        and 
-00012200: 656e 7620 7661 6c75 6573 2066 6f72 2061  env values for a
-00012210: 6c6c 2069 6e64 6976 6964 7561 6c73 2e0a  ll individuals..
-00012220: 0a20 2020 2020 2020 204e 4f54 453a 2043  .        NOTE: C
-00012230: 7572 7265 6e74 6c79 206f 6e6c 7920 6163  urrently only ac
-00012240: 6365 7074 7320 6f6e 6520 4c61 6e64 7363  cepts one Landsc
-00012250: 6170 6520 4c61 7965 7220 2869 2e65 2e20  ape Layer (i.e. 
-00012260: 656e 7669 726f 6e6d 656e 7461 6c0a 2020  environmental.  
-00012270: 2020 2020 2020 2020 2020 2020 7661 7269              vari
-00012280: 6162 6c65 290a 0a20 2020 2020 2020 2050  able)..        P
-00012290: 6172 616d 6574 6572 730a 2020 2020 2020  arameters.      
-000122a0: 2020 2d2d 2d2d 2d2d 2d2d 2d2d 0a20 2020    ----------.   
-000122b0: 2020 2020 206c 7972 5f6e 756d 203a 2069       lyr_num : i
-000122c0: 6e74 0a20 2020 2020 2020 2020 2020 2054  nt.            T
-000122d0: 6865 206e 756d 6265 7220 6f66 2074 6865  he number of the
-000122e0: 204c 616e 6473 6361 7065 204c 6179 6572   Landscape Layer
-000122f0: 2066 726f 6d20 7768 6963 6820 746f 2065   from which to e
-00012300: 7874 7261 6374 0a20 2020 2020 2020 2020  xtract.         
-00012310: 2020 2065 6e76 6972 6f6e 6d65 6e74 616c     environmental
-00012320: 2076 616c 7565 732e 2044 6566 6175 6c74   values. Default
-00012330: 7320 746f 2031 2e0a 0a20 2020 2020 2020  s to 1...       
-00012340: 2052 6574 7572 6e73 0a20 2020 2020 2020   Returns.       
-00012350: 202d 2d2d 2d2d 2d2d 0a20 2020 2020 2020   -------.       
-00012360: 206f 7574 203a 2070 616e 6461 732e 4461   out : pandas.Da
-00012370: 7461 4672 616d 650a 2020 2020 2020 2020  taFrame.        
-00012380: 2020 2020 4120 4461 7461 4672 616d 6520      A DataFrame 
-00012390: 696e 2077 6869 6368 2063 6f6c 756d 6e73  in which columns
-000123a0: 2061 7265 206c 6f63 6920 616e 6420 726f   are loci and ro
-000123b0: 7773 2061 7265 2069 6e64 6976 6964 7561  ws are individua
-000123c0: 6c73 0a20 2020 2020 2020 2022 2222 0a20  ls.        """. 
-000123d0: 2020 2020 2020 2023 6765 7420 6172 7261         #get arra
-000123e0: 7920 6f66 205b 307c 302e 357c 315d 2067  y of [0|0.5|1] g
-000123f0: 656e 6f74 7970 6573 0a20 2020 2020 2020  enotypes.       
-00012400: 2022 2222 0a20 2020 2020 2020 2067 656e   """.        gen
-00012410: 203d 2073 656c 662e 5f67 6574 5f67 656e   = self._get_gen
-00012420: 6f74 7970 6573 2829 0a0a 2020 2020 2020  otypes()..      
-00012430: 2020 236c 6f6f 7020 746f 2063 6f6e 7665    #loop to conve
-00012440: 7274 2062 696e 6172 7920 616c 6c65 6c65  rt binary allele
-00012450: 2067 656e 6f74 7970 6573 2028 307c 3129   genotypes (0|1)
-00012460: 0a20 2020 2020 2020 2023 696e 746f 2073  .        #into s
-00012470: 696e 676c 6520 6469 6769 7420 6765 6e6f  ingle digit geno
-00012480: 7479 7065 7320 2830 2c20 302e 352c 2031  types (0, 0.5, 1
-00012490: 290a 2020 2020 2020 2020 6765 6e6f 5f69  ).        geno_i
-000124a0: 6e64 203d 205b 5d0a 2020 2020 2020 2020  nd = [].        
-000124b0: 666f 7220 696e 6420 696e 2072 616e 6765  for ind in range
-000124c0: 286c 656e 2867 656e 2929 3a0a 2020 2020  (len(gen)):.    
-000124d0: 2020 2020 2020 2020 6765 6e6f 5f6c 6f63          geno_loc
-000124e0: 203d 205b 5d0a 2020 2020 2020 2020 2020   = [].          
-000124f0: 2020 666f 7220 6c6f 6369 2069 6e20 7261    for loci in ra
-00012500: 6e67 6528 6c65 6e28 6765 6e5b 696e 645d  nge(len(gen[ind]
-00012510: 2929 3a0a 2020 2020 2020 2020 2020 2020  )):.            
-00012520: 2020 2020 6765 6e6f 7479 7065 203d 2067      genotype = g
-00012530: 656e 5b69 6e64 5d5b 6c6f 6369 5d2e 6d65  en[ind][loci].me
-00012540: 616e 2829 2023 2063 6f64 6564 2061 7320  an() # coded as 
-00012550: 302f 302e 352f 310a 2020 2020 2020 2020  0/0.5/1.        
-00012560: 2020 2020 2020 2020 6765 6e6f 5f6c 6f63          geno_loc
-00012570: 2e61 7070 656e 6428 6765 6e6f 7479 7065  .append(genotype
-00012580: 290a 2020 2020 2020 2020 2020 2020 6765  ).            ge
-00012590: 6e6f 5f69 6e64 2e61 7070 656e 6428 6765  no_ind.append(ge
-000125a0: 6e6f 5f6c 6f63 290a 2020 2020 2020 2020  no_loc).        
-000125b0: 2222 220a 2020 2020 2020 2020 6774 7320  """.        gts 
-000125c0: 3d20 7365 6c66 2e5f 6765 745f 6765 6e6f  = self._get_geno
-000125d0: 7479 7065 7328 6269 616c 6c65 6c69 633d  types(biallelic=
-000125e0: 4661 6c73 6529 0a0a 2020 2020 2020 2020  False)..        
-000125f0: 2363 6f6e 7665 7274 2074 6f20 6461 7461  #convert to data
-00012600: 6672 616d 650a 2020 2020 2020 2020 6765  frame.        ge
-00012610: 615f 6466 203d 2070 642e 4461 7461 4672  a_df = pd.DataFr
-00012620: 616d 6528 6774 7329 0a0a 2020 2020 2020  ame(gts)..      
-00012630: 2020 2367 6574 2065 6e76 6972 6f6e 6d65    #get environme
-00012640: 6e74 616c 2064 6174 610a 2020 2020 2020  ntal data.      
-00012650: 2020 656e 7620 3d20 7365 6c66 2e5f 6765    env = self._ge
-00012660: 745f 6528 290a 2020 2020 2020 2020 6765  t_e().        ge
-00012670: 615f 6466 5b27 656e 7627 5d20 3d20 6c69  a_df['env'] = li
-00012680: 7374 2865 6e76 5b3a 2c6c 7972 5f6e 756d  st(env[:,lyr_num
-00012690: 5d29 0a0a 2020 2020 2020 2020 2367 6574  ])..        #get
-000126a0: 2063 6f6f 7264 730a 2020 2020 2020 2020   coords.        
-000126b0: 636f 6f72 6420 3d20 7365 6c66 2e5f 6765  coord = self._ge
-000126c0: 745f 636f 6f72 6473 2829 0a20 2020 2020  t_coords().     
-000126d0: 2020 2067 6561 5f64 665b 276c 6174 275d     gea_df['lat']
-000126e0: 203d 206c 6973 7428 636f 6f72 645b 3a2c   = list(coord[:,
-000126f0: 305d 290a 2020 2020 2020 2020 6765 615f  0]).        gea_
-00012700: 6466 5b27 6c6f 6e67 275d 203d 206c 6973  df['long'] = lis
-00012710: 7428 636f 6f72 645b 3a2c 315d 290a 0a20  t(coord[:,1]).. 
-00012720: 2020 2020 2020 2072 6574 7572 6e20 6765         return ge
-00012730: 615f 6466 0a0a 0a20 2020 2064 6566 205f  a_df...    def _
-00012740: 7275 6e5f 6363 6128 7365 6c66 2c20 7472  run_cca(self, tr
-00012750: 745f 6e75 6d3d 302c 2073 6361 6c65 3d54  t_num=0, scale=T
-00012760: 7275 652c 2070 6c6f 743d 5472 7565 2c20  rue, plot=True, 
-00012770: 706c 6f74 5f73 643d 5472 7565 2c0a 2020  plot_sd=True,.  
-00012780: 2020 2020 2020 2020 2020 2020 2020 2073                 s
-00012790: 643d 3329 3a0a 2020 2020 2020 2020 2222  d=3):.        ""
-000127a0: 220a 2020 2020 2020 2020 5275 6e73 2061  ".        Runs a
-000127b0: 2063 616e 6f6e 6963 616c 2063 6f72 7265   canonical corre
-000127c0: 6c61 7469 6f6e 2061 6e61 6c79 7369 7320  lation analysis 
-000127d0: 2843 4341 2920 6f6e 2074 6865 2063 7572  (CCA) on the cur
-000127e0: 7265 6e74 2067 656e 6574 6963 2061 6e64  rent genetic and
-000127f0: 0a20 2020 2020 2020 2065 6e76 6972 6f6e  .        environ
-00012800: 6d65 6e74 616c 2064 6174 6120 666f 7220  mental data for 
-00012810: 7468 6520 5370 6563 6965 7327 2063 7572  the Species' cur
-00012820: 7265 6e74 2070 6f70 756c 6174 696f 6e2c  rent population,
-00012830: 2075 7369 6e67 2074 6865 0a20 2020 2020   using the.     
-00012840: 2020 2069 6e64 6963 6174 6564 2054 7261     indicated Tra
-00012850: 6974 2e20 506c 6f74 7320 7468 6520 7265  it. Plots the re
-00012860: 7375 6c74 7320 616e 6420 7265 7475 726e  sults and return
-00012870: 7320 7468 656d 2069 6e20 6120 6469 6374  s them in a dict
-00012880: 2e0a 0a20 2020 2020 2020 2043 4341 206d  ...        CCA m
-00012890: 6f64 656c 2066 6f72 6d75 6c61 3a20 6765  odel formula: ge
-000128a0: 6e6f 7479 7065 207e 2065 6e76 202b 206c  notype ~ env + l
-000128b0: 6174 202b 206c 6f6e 670a 0a20 2020 2020  at + long..     
-000128c0: 2020 204e 4f54 453a 2063 7572 7265 6e74     NOTE: current
-000128d0: 6c79 206f 6e6c 7920 706f 7373 6962 6c65  ly only possible
-000128e0: 2074 6f20 646f 2033 2063 6f6d 706f 6e65   to do 3 compone
-000128f0: 6e74 732f 3320 7661 7269 6162 6c65 733b  nts/3 variables;
-00012900: 0a20 2020 2020 2020 2020 2020 2020 2077  .              w
-00012910: 696c 6c20 6164 6420 6772 6561 7465 7220  ill add greater 
-00012920: 6675 6e63 7469 6f6e 616c 6974 7920 696e  functionality in
-00012930: 2074 6865 2066 7574 7572 650a 0a20 2020   the future..   
-00012940: 2020 2020 2050 6172 616d 6574 6572 730a       Parameters.
-00012950: 2020 2020 2020 2020 2d2d 2d2d 2d2d 2d2d          --------
-00012960: 2d2d 0a20 2020 2020 2020 2074 7274 5f6e  --.        trt_n
-00012970: 756d 203a 2069 6e74 0a20 2020 2020 2020  um : int.       
-00012980: 2020 2020 2054 6865 206e 756d 6265 7220       The number 
-00012990: 6f66 2074 6865 2054 7261 6974 2074 6f20  of the Trait to 
-000129a0: 7275 6e20 7468 6520 4745 4120 6f6e 2e20  run the GEA on. 
-000129b0: 4465 6661 756c 7473 2074 6f20 302e 0a20  Defaults to 0.. 
-000129c0: 2020 2020 2020 2073 6361 6c65 203a 2062         scale : b
-000129d0: 6f6f 6c0a 2020 2020 2020 2020 2020 2020  ool.            
-000129e0: 4966 2054 7275 652c 2073 6361 6c65 7320  If True, scales 
-000129f0: 7468 6520 7661 7269 6162 6c65 2c20 696e  the variable, in
-00012a00: 6469 7669 6475 616c 2c0a 2020 2020 2020  dividual,.      
-00012a10: 2020 2020 2020 616e 6420 6c6f 6375 7320        and locus 
-00012a20: 6c6f 6164 696e 6773 2066 726f 6d20 2d31  loadings from -1
-00012a30: 2074 6f20 3120 746f 206d 616b 6520 7468   to 1 to make th
-00012a40: 656d 2065 6173 6965 720a 2020 2020 2020  em easier.      
-00012a50: 2020 2020 2020 746f 2076 6973 7561 6c69        to visuali
-00012a60: 7a65 2e20 4465 6661 756c 7473 2074 6f20  ze. Defaults to 
-00012a70: 5472 7565 2e0a 2020 2020 2020 2020 706c  True..        pl
-00012a80: 6f74 203a 2062 6f6f 6c0a 2020 2020 2020  ot : bool.      
-00012a90: 2020 2020 2020 5768 6574 6865 7220 6f72        Whether or
-00012aa0: 206e 6f74 2074 6f20 706c 6f74 2074 6865   not to plot the
-00012ab0: 206d 6f64 656c 2e20 4465 6661 756c 7473   model. Defaults
-00012ac0: 2074 6f20 5472 7565 2e0a 2020 2020 2020   to True..      
-00012ad0: 2020 706c 6f74 5f73 6420 3a20 626f 6f6c    plot_sd : bool
-00012ae0: 0a20 2020 2020 2020 2020 2020 2049 6620  .            If 
-00012af0: 5472 7565 2c20 6120 7374 616e 6461 7264  True, a standard
-00012b00: 2064 6576 6961 7469 6f6e 2065 6c6c 6970   deviation ellip
-00012b10: 7365 2069 7320 706c 6f74 7465 642c 2075  se is plotted, u
-00012b20: 7369 6e67 2074 6865 206e 756d 6265 720a  sing the number.
-00012b30: 2020 2020 2020 2020 2020 2020 6f66 2073              of s
-00012b40: 7461 6e64 6172 6420 6465 7669 6174 696f  tandard deviatio
-00012b50: 6e73 2069 6e64 6963 6174 6564 2062 7920  ns indicated by 
-00012b60: 7468 6520 6172 6775 6d65 6e74 2027 7364  the argument 'sd
-00012b70: 272e 2044 6566 6175 6c74 7320 746f 0a20  '. Defaults to. 
-00012b80: 2020 2020 2020 2020 2020 2054 7275 652e             True.
-00012b90: 0a20 2020 2020 2020 2073 6420 3a20 7b69  .        sd : {i
-00012ba0: 6e74 2c20 666c 6f61 747d 0a20 2020 2020  nt, float}.     
-00012bb0: 2020 2020 2020 204e 756d 6265 7220 6f66         Number of
-00012bc0: 2073 7461 6e64 6172 6420 6465 7669 6174   standard deviat
-00012bd0: 696f 6e73 2074 6f20 7573 6520 666f 7220  ions to use for 
-00012be0: 706c 6f74 7469 6e67 2074 6865 2073 7461  plotting the sta
-00012bf0: 6e64 6172 640a 2020 2020 2020 2020 2020  ndard.          
-00012c00: 2020 6465 7669 6174 696f 6e20 656c 6c69    deviation elli
-00012c10: 7073 652e 2044 6566 6175 6c74 7320 746f  pse. Defaults to
-00012c20: 2033 2e0a 0a20 2020 2020 2020 2052 6574   3...        Ret
-00012c30: 7572 6e73 0a20 2020 2020 2020 202d 2d2d  urns.        ---
-00012c40: 2d2d 2d2d 0a20 2020 2020 2020 206f 7574  ----.        out
-00012c50: 203a 2064 6963 740a 2020 2020 2020 2020   : dict.        
-00012c60: 2020 2020 4120 6469 6374 206f 6620 7468      A dict of th
-00012c70: 6520 666f 6c6c 6f77 696e 6720 6b65 792d  e following key-
-00012c80: 7661 6c75 6520 7061 6972 733a 0a20 2020  value pairs:.   
-00012c90: 2020 2020 2020 2020 2020 2020 2022 6765               "ge
-00012ca0: 615f 6466 223a 2061 2044 6174 6146 7261  a_df": a DataFra
-00012cb0: 6d65 206f 6620 696e 6469 7669 6475 616c  me of individual
-00012cc0: 7327 2067 656e 6f74 7970 6573 2c20 656e  s' genotypes, en
-00012cd0: 7669 726f 6e6d 656e 7461 6c0a 2020 2020  vironmental.    
-00012ce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012cf0: 2020 2020 2020 7661 6c75 6573 2c20 616e        values, an
-00012d00: 6420 636f 6f72 6469 6e61 7465 730a 2020  d coordinates.  
-00012d10: 2020 2020 2020 2020 2020 2020 2020 2269                "i
-00012d20: 6e64 5f43 4341 6466 223a 2069 6e64 6976  nd_CCAdf": indiv
-00012d30: 6964 7561 6c20 6c6f 6164 696e 6773 2028  idual loadings (
-00012d40: 636f 6c75 6d6e 7320 3d20 6178 6573 2c0a  columns = axes,.
-00012d50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012d60: 2020 2020 2020 2020 2020 2020 2072 6f77               row
-00012d70: 7320 3d20 696e 6469 7669 6475 616c 7329  s = individuals)
-00012d80: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00012d90: 2022 6c6f 6369 5f43 4341 6466 223a 206c   "loci_CCAdf": l
-00012da0: 6f63 7573 206c 6f61 6469 6e67 7320 2863  ocus loadings (c
-00012db0: 6f6c 756d 6e73 203d 2061 7865 732c 2072  olumns = axes, r
-00012dc0: 6f77 7320 3d20 6c6f 6369 290a 2020 2020  ows = loci).    
-00012dd0: 2020 2020 2020 2020 2020 2020 2276 6172              "var
-00012de0: 5f43 4341 6466 223a 2076 6172 6961 626c  _CCAdf": variabl
-00012df0: 6520 6c6f 6164 696e 6773 2028 636f 6c75  e loadings (colu
-00012e00: 6d6e 7320 3d20 6178 6573 2c0a 2020 2020  mns = axes,.    
-00012e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012e20: 2020 2020 2020 2020 2072 6f77 7320 3d20           rows = 
-00012e30: 7661 7269 6162 6c65 2028 656e 762c 206c  variable (env, l
-00012e40: 6174 2c20 6c6f 6e67 2929 0a20 2020 2020  at, long)).     
-00012e50: 2020 2020 2020 2020 2020 2022 7472 6169             "trai
-00012e60: 745f 6c6f 6369 223a 206c 6973 7420 6f66  t_loci": list of
-00012e70: 2069 6e64 6578 6573 2066 6f72 206c 6f63   indexes for loc
-00012e80: 6920 7468 6174 2075 6e64 6572 6c69 6520  i that underlie 
-00012e90: 7468 6520 7472 6169 740a 2020 2020 2020  the trait.      
-00012ea0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012eb0: 2020 2020 2020 2020 666f 7220 7768 6963          for whic
-00012ec0: 6820 7468 6520 4343 4120 7761 7320 7275  h the CCA was ru
-00012ed0: 6e0a 2020 2020 2020 2020 2222 220a 2020  n.        """.  
-00012ee0: 2020 2020 2020 236d 616b 6520 4446 0a20        #make DF. 
-00012ef0: 2020 2020 2020 2067 6561 5f64 6620 3d73         gea_df =s
-00012f00: 656c 662e 5f6d 616b 655f 6765 615f 6466  elf._make_gea_df
-00012f10: 286c 7972 5f6e 756d 3d73 656c 662e 6765  (lyr_num=self.ge
-00012f20: 6e5f 6172 6368 2e74 7261 6974 735b 7472  n_arch.traits[tr
-00012f30: 745f 6e75 6d5d 2e6c 7972 5f6e 756d 290a  t_num].lyr_num).
-00012f40: 0a20 2020 2020 2020 2023 6765 7420 6164  .        #get ad
-00012f50: 6170 7469 7665 206c 6f63 6920 2875 7365  aptive loci (use
-00012f60: 6420 6c61 7465 7220 696e 2070 6c6f 7474  d later in plott
-00012f70: 696e 6729 0a20 2020 2020 2020 2074 7261  ing).        tra
-00012f80: 6974 5f6c 6f63 6920 3d20 7365 6c66 2e67  it_loci = self.g
-00012f90: 656e 5f61 7263 682e 7472 6169 7473 5b74  en_arch.traits[t
-00012fa0: 7274 5f6e 756d 5d2e 6c6f 6369 2023 6765  rt_num].loci #ge
-00012fb0: 7473 2074 7261 6974 206c 6f63 6920 696e  ts trait loci in
-00012fc0: 6465 7865 730a 0a20 2020 2020 2020 2023  dexes..        #
-00012fd0: 6765 7420 6e75 6d62 6572 206f 6620 6c6f  get number of lo
-00012fe0: 6369 0a20 2020 2020 2020 204c 203d 2073  ci.        L = s
-00012ff0: 656c 662e 6765 6e5f 6172 6368 2e4c 0a20  elf.gen_arch.L. 
-00013000: 2020 2020 2020 2023 6465 6669 6e65 2078         #define x
-00013010: 2061 6e64 2079 3a0a 2020 2020 2020 2020   and y:.        
-00013020: 2320 7075 6c6c 206f 7574 2067 656e 6f74  # pull out genot
-00013030: 7970 6573 0a20 2020 2020 2020 2059 203d  ypes.        Y =
-00013040: 2067 6561 5f64 662e 696c 6f63 5b3a 2c72   gea_df.iloc[:,r
-00013050: 616e 6765 284c 295d 0a20 2020 2020 2020  ange(L)].       
-00013060: 2023 2070 756c 6c20 6f75 7420 7072 6564   # pull out pred
-00013070: 6963 746f 7220 7661 7269 6162 6c65 730a  ictor variables.
-00013080: 2020 2020 2020 2020 2320 544f 444f 3a20          # TODO: 
-00013090: 6d61 6b65 2069 7420 706f 7373 6962 6c65  make it possible
-000130a0: 2074 6f20 6368 6f6f 7365 206d 6f72 6520   to choose more 
-000130b0: 7661 7273 2069 6e20 6675 7475 7265 0a20  vars in future. 
-000130c0: 2020 2020 2020 2058 203d 2067 6561 5f64         X = gea_d
-000130d0: 665b 5b27 656e 7627 2c27 6c61 7427 2c27  f[['env','lat','
-000130e0: 6c6f 6e67 275d 5d0a 0a20 2020 2020 2020  long']]..       
-000130f0: 2023 6465 6669 6e65 206e 5f63 6f6d 706f   #define n_compo
-00013100: 6e65 6e74 730a 2020 2020 2020 2020 6e5f  nents.        n_
-00013110: 636f 6d70 6f6e 656e 7473 203d 2033 0a20  components = 3. 
-00013120: 2020 2020 2020 2023 6372 6561 7465 2074         #create t
-00013130: 6865 2043 4341 206d 6f64 656c 2061 6e64  he CCA model and
-00013140: 2066 6974 2069 7420 746f 2074 6865 2064   fit it to the d
-00013150: 6174 610a 2020 2020 2020 2020 6363 6120  ata.        cca 
-00013160: 3d20 4343 4128 6e5f 636f 6d70 6f6e 656e  = CCA(n_componen
-00013170: 7473 203d 206e 5f63 6f6d 706f 6e65 6e74  ts = n_component
-00013180: 7329 0a20 2020 2020 2020 2063 6361 2e66  s).        cca.f
-00013190: 6974 2859 2c20 5829 0a0a 2020 2020 2020  it(Y, X)..      
-000131a0: 2020 2374 7261 6e73 666f 726d 2064 6174    #transform dat
-000131b0: 610a 2020 2020 2020 2020 595f 6320 3d20  a.        Y_c = 
-000131c0: 6363 612e 7472 616e 7366 6f72 6d28 5929  cca.transform(Y)
-000131d0: 0a0a 2020 2020 2020 2020 2363 7265 6174  ..        #creat
-000131e0: 6520 6466 2066 6f72 2069 6e64 6976 6964  e df for individ
-000131f0: 7561 6c73 0a20 2020 2020 2020 2069 6e64  uals.        ind
-00013200: 5f64 6620 3d20 7064 2e44 6174 6146 7261  _df = pd.DataFra
-00013210: 6d65 2859 5f63 290a 2020 2020 2020 2020  me(Y_c).        
-00013220: 236e 616d 696e 6720 4343 4120 636f 6c75  #naming CCA colu
-00013230: 6d6e 7320 7374 6172 7469 6e67 2061 7420  mns starting at 
-00013240: 310a 2020 2020 2020 2020 696e 645f 6466  1.        ind_df
-00013250: 2e63 6f6c 756d 6e73 203d 205b 7374 7228  .columns = [str(
-00013260: 7229 2066 6f72 2072 2069 6e20 7261 6e67  r) for r in rang
-00013270: 6528 312c 206e 5f63 6f6d 706f 6e65 6e74  e(1, n_component
-00013280: 7320 2b20 3129 5d0a 0a20 2020 2020 2020  s + 1)]..       
-00013290: 2023 6372 6561 7465 2064 6620 666f 7220   #create df for 
-000132a0: 6c6f 6369 0a20 2020 2020 2020 206c 6f63  loci.        loc
-000132b0: 695f 6466 203d 2070 642e 4461 7461 4672  i_df = pd.DataFr
-000132c0: 616d 6528 6363 612e 785f 6c6f 6164 696e  ame(cca.x_loadin
-000132d0: 6773 5f29 0a20 2020 2020 2020 206c 6f63  gs_).        loc
-000132e0: 695f 6466 2e63 6f6c 756d 6e73 203d 2069  i_df.columns = i
-000132f0: 6e64 5f64 662e 636f 6c75 6d6e 7320 2373  nd_df.columns #s
-00013300: 616d 6520 636f 6c75 6d6e 206e 616d 6573  ame column names
-00013310: 0a0a 2020 2020 2020 2020 2363 7265 6174  ..        #creat
-00013320: 6520 6466 2066 6f72 2076 6172 6961 626c  e df for variabl
-00013330: 6573 0a20 2020 2020 2020 2076 6172 5f64  es.        var_d
-00013340: 6620 3d20 7064 2e44 6174 6146 7261 6d65  f = pd.DataFrame
-00013350: 2863 6361 2e79 5f6c 6f61 6469 6e67 735f  (cca.y_loadings_
-00013360: 290a 2020 2020 2020 2020 7661 725f 6466  ).        var_df
-00013370: 2e63 6f6c 756d 6e73 203d 2069 6e64 5f64  .columns = ind_d
-00013380: 662e 636f 6c75 6d6e 7320 2373 616d 6520  f.columns #same 
-00013390: 636f 6c75 6d6e 206e 616d 6573 0a0a 2020  column names..  
-000133a0: 2020 2020 2020 236d 616b 6520 6469 6374        #make dict
-000133b0: 696f 6e61 7279 206f 6620 6461 7461 6672  ionary of datafr
-000133c0: 616d 6573 2074 6f20 7265 7475 726e 0a20  ames to return. 
-000133d0: 2020 2020 2020 2063 6361 5f64 6963 7420         cca_dict 
-000133e0: 3d20 7b27 6765 615f 6466 273a 6765 615f  = {'gea_df':gea_
-000133f0: 6466 2c0a 2020 2020 2020 2020 2020 2020  df,.            
-00013400: 2020 2020 2020 2020 2769 6e64 5f64 6627          'ind_df'
-00013410: 3a69 6e64 5f64 662c 0a20 2020 2020 2020  :ind_df,.       
-00013420: 2020 2020 2020 2020 2020 2020 2027 6c6f               'lo
-00013430: 6369 5f64 6627 3a6c 6f63 695f 6466 2c0a  ci_df':loci_df,.
+00010340: 2020 6e70 2e77 6865 7265 2873 656c 662e    np.where(self.
+00010350: 5f74 632e 6e6f 6465 732e 666c 6167 7329  _tc.nodes.flags)
+00010360: 5b30 5d2c 0a20 2020 2020 2020 2020 2020  [0],.           
+00010370: 2023 2020 2020 2020 2020 2020 2020 2020   #              
+00010380: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00010390: 2074 732e 6861 706c 6f74 7970 6573 2829   ts.haplotypes()
+000103a0: 2920 6966 206e 2069 6e20 7361 6d70 6c65  ) if n in sample
+000103b0: 735d 0a20 2020 2020 2020 2020 2020 2068  s].            h
+000103c0: 6170 735f 6469 6374 203d 207b 733a 6820  aps_dict = {s:h 
+000103d0: 666f 7220 732c 2068 2069 6e20 7a69 7028  for s, h in zip(
+000103e0: 7473 2e73 616d 706c 6573 2829 2c20 7473  ts.samples(), ts
+000103f0: 2e68 6170 6c6f 7479 7065 7328 2929 7d0a  .haplotypes())}.
+00010400: 2020 2020 2020 2020 2020 2020 6861 7073              haps
+00010410: 203d 205b 6e70 2e69 6e74 3828 5b2a 6861   = [np.int8([*ha
+00010420: 7073 5f64 6963 745b 735d 5d29 2066 6f72  ps_dict[s]]) for
+00010430: 2073 2069 6e20 7361 6d70 6c65 735f 746f   s in samples_to
+00010440: 5f6b 6565 705d 0a0a 2020 2020 2020 2020  _keep]..        
+00010450: 2020 2020 2320 6765 7420 7468 6520 6765      # get the ge
+00010460: 6e6f 7479 7065 7320 6279 2063 6f6d 6269  notypes by combi
+00010470: 6e69 6e67 2065 6163 6820 636f 6e73 6563  ning each consec
+00010480: 7574 6976 6520 6772 6f75 700a 2020 2020  utive group.    
+00010490: 2020 2020 2020 2020 2320 6f66 2078 2068          # of x h
+000104a0: 6170 6c6f 7479 7065 732c 2077 6865 7265  aplotypes, where
+000104b0: 2078 2069 7320 7468 6520 706c 6f69 6479   x is the ploidy
+000104c0: 2c20 7468 656e 2076 7374 6163 6b69 6e67  , then vstacking
+000104d0: 2074 6865 6d0a 2020 2020 2020 2020 2020   them.          
+000104e0: 2020 6772 6f75 7065 645f 6861 7073 203d    grouped_haps =
+000104f0: 207a 6970 282a 5b68 6170 735b 693a 3a73   zip(*[haps[i::s
+00010500: 656c 662e 6765 6e5f 6172 6368 2e78 5d20  elf.gen_arch.x] 
+00010510: 666f 7220 6920 696e 2072 616e 6765 280a  for i in range(.
+00010520: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00010530: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00010540: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00010550: 2020 2020 2020 2020 2020 2020 7365 6c66              self
+00010560: 2e67 656e 5f61 7263 682e 7829 5d29 0a20  .gen_arch.x)]). 
+00010570: 2020 2020 2020 2020 2020 2067 7473 203d             gts =
+00010580: 205b 6e70 2e76 7374 6163 6b28 6829 2e54   [np.vstack(h).T
+00010590: 2066 6f72 2068 2069 6e20 6772 6f75 7065   for h in groupe
+000105a0: 645f 6861 7073 5d0a 0a20 2020 2020 2020  d_haps]..       
+000105b0: 2023 206f 7468 6572 7769 7365 2c20 6765   # otherwise, ge
+000105c0: 7420 7468 6520 6765 6e6f 7479 7065 7320  t the genotypes 
+000105d0: 6672 6f6d 2074 6865 2069 6e64 6976 6964  from the individ
+000105e0: 7561 6c73 2064 6972 6563 746c 790a 2020  uals directly.  
+000105f0: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
+00010600: 2020 2020 2020 2020 6774 7320 3d20 5b69          gts = [i
+00010610: 6e64 2e67 2066 6f72 2069 6478 2c20 696e  nd.g for idx, in
+00010620: 6420 696e 2073 656c 662e 6974 656d 7328  d in self.items(
+00010630: 2920 6966 2069 6478 2069 6e20 696e 6469  ) if idx in indi
+00010640: 7669 6473 5d0a 0a20 2020 2020 2020 2023  vids]..        #
+00010650: 2073 7562 7365 7420 6c6f 6369 2c20 6966   subset loci, if
+00010660: 206e 6565 6465 640a 2020 2020 2020 2020   needed.        
+00010670: 6966 206c 6f63 6920 6973 206e 6f74 204e  if loci is not N
+00010680: 6f6e 653a 0a20 2020 2020 2020 2020 2020  one:.           
+00010690: 2067 7473 203d 205b 6774 5b6c 6f63 692c   gts = [gt[loci,
+000106a0: 203a 5d20 666f 7220 6774 2069 6e20 6774   :] for gt in gt
+000106b0: 735d 0a0a 2020 2020 2020 2020 2320 6765  s]..        # ge
+000106c0: 7420 6d65 616e 2067 656e 6f74 7970 6520  t mean genotype 
+000106d0: 666f 7220 6561 6368 2069 6e64 6976 6964  for each individ
+000106e0: 7561 6c2c 2069 6620 6e65 6365 7373 6172  ual, if necessar
+000106f0: 790a 2020 2020 2020 2020 6966 206e 6f74  y.        if not
+00010700: 2062 6961 6c6c 656c 6963 3a0a 2020 2020   biallelic:.    
+00010710: 2020 2020 2020 2020 6774 7320 3d20 5b6e          gts = [n
+00010720: 702e 6d65 616e 2867 742c 2061 7869 733d  p.mean(gt, axis=
+00010730: 3129 2066 6f72 2067 7420 696e 2067 7473  1) for gt in gts
+00010740: 5d0a 0a20 2020 2020 2020 2023 2073 7461  ]..        # sta
+00010750: 636b 2069 6e74 6f20 7468 6520 7370 6563  ck into the spec
+00010760: 696f 6d65 2c20 6966 2064 6963 7420 6e6f  iome, if dict no
+00010770: 7420 7265 7175 6573 7465 640a 2020 2020  t requested.    
+00010780: 2020 2020 2320 2864 696d 7320 6172 6520      # (dims are 
+00010790: 4e20 7820 4c20 7820 582c 2077 6865 7265  N x L x X, where
+000107a0: 204e 3d6e 756d 2069 6e64 6976 6964 732c   N=num individs,
+000107b0: 204c 3d6e 756d 206c 6f63 692c 0a20 2020   L=num loci,.   
+000107c0: 2020 2020 2023 2020 616e 6420 583d 706c       #  and X=pl
+000107d0: 6f69 6479 2069 6620 6269 616c 6c65 6c69  oidy if bialleli
+000107e0: 633d 5472 7565 2c20 656c 7365 2058 3d31  c=True, else X=1
+000107f0: 290a 2020 2020 2020 2020 6966 206e 6f74  ).        if not
+00010800: 2061 735f 6469 6374 3a0a 2020 2020 2020   as_dict:.      
+00010810: 2020 2020 2020 6774 7320 3d20 6e70 2e73        gts = np.s
+00010820: 7461 636b 2867 7473 290a 0a20 2020 2020  tack(gts)..     
+00010830: 2020 2023 206f 7220 656c 7365 2063 6173     # or else cas
+00010840: 7420 6173 2064 6963 740a 2020 2020 2020  t as dict.      
+00010850: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+00010860: 2020 2020 6774 7320 3d20 7b69 6e64 6976      gts = {indiv
+00010870: 6964 3a20 6774 2066 6f72 2069 6e64 6976  id: gt for indiv
+00010880: 6964 2c20 6774 2069 6e20 7a69 7028 696e  id, gt in zip(in
+00010890: 6469 7669 6473 2c20 6774 7329 7d0a 0a20  divids, gts)}.. 
+000108a0: 2020 2020 2020 2072 6574 7572 6e20 6774         return gt
+000108b0: 730a 0a0a 2020 2020 2363 6f6e 7665 6e69  s...    #conveni
+000108c0: 656e 6365 206d 6574 686f 6420 666f 7220  ence method for 
+000108d0: 6765 7474 696e 6720 6120 7363 616c 6172  getting a scalar
+000108e0: 2061 7474 7269 6275 7465 2066 6f72 2073   attribute for s
+000108f0: 6f6d 6520 6f72 2061 6c6c 2069 6e64 6976  ome or all indiv
+00010900: 6964 730a 2020 2020 6465 6620 5f67 6574  ids.    def _get
+00010910: 5f73 6361 6c61 725f 6174 7472 2873 656c  _scalar_attr(sel
+00010920: 662c 2061 7474 725f 6e61 6d65 2c20 696e  f, attr_name, in
+00010930: 6469 7669 6473 3d4e 6f6e 6529 3a0a 2020  divids=None):.  
+00010940: 2020 2020 2020 6966 2069 6e64 6976 6964        if individ
+00010950: 7320 6973 204e 6f6e 653a 0a20 2020 2020  s is None:.     
+00010960: 2020 2020 2020 2076 616c 7320 3d20 6e70         vals = np
+00010970: 2e61 7272 6179 285b 6765 7461 7474 7228  .array([getattr(
+00010980: 696e 642c 2061 7474 725f 6e61 6d65 2920  ind, attr_name) 
+00010990: 666f 7220 696e 6420 696e 2073 656c 662e  for ind in self.
+000109a0: 7661 6c75 6573 2829 5d29 0a20 2020 2020  values()]).     
+000109b0: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
+000109c0: 2020 2020 2069 6720 3d20 6974 656d 6765       ig = itemge
+000109d0: 7474 6572 282a 696e 6469 7669 6473 290a  tter(*individs).
+000109e0: 2020 2020 2020 2020 2020 2020 7661 6c73              vals
+000109f0: 203d 207b 693a 2067 6574 6174 7472 2869   = {i: getattr(i
+00010a00: 6e64 2c20 6174 7472 5f6e 616d 6529 2066  nd, attr_name) f
+00010a10: 6f72 2069 2c20 696e 6420 696e 2073 656c  or i, ind in sel
+00010a20: 662e 6974 656d 7328 297d 0a20 2020 2020  f.items()}.     
+00010a30: 2020 2020 2020 2076 616c 7320 3d20 6e70         vals = np
+00010a40: 2e61 7272 6179 2869 6728 7661 6c73 2929  .array(ig(vals))
+00010a50: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
+00010a60: 7661 6c73 0a0a 2020 2020 2363 6f6e 7665  vals..    #conve
+00010a70: 6e69 656e 6365 206d 6574 686f 6420 666f  nience method fo
+00010a80: 7220 6765 7474 696e 6720 6167 6520 6f66  r getting age of
+00010a90: 2077 686f 6c65 2073 7065 6369 6573 0a20   whole species. 
+00010aa0: 2020 2064 6566 205f 6765 745f 6167 6528     def _get_age(
+00010ab0: 7365 6c66 2c20 696e 6469 7669 6473 3d4e  self, individs=N
+00010ac0: 6f6e 6529 3a0a 2020 2020 2020 2020 6167  one):.        ag
+00010ad0: 6573 203d 2073 656c 662e 5f67 6574 5f73  es = self._get_s
+00010ae0: 6361 6c61 725f 6174 7472 2827 6167 6527  calar_attr('age'
+00010af0: 2c20 696e 6469 7669 6473 3d69 6e64 6976  , individs=indiv
+00010b00: 6964 7329 0a20 2020 2020 2020 2072 6574  ids).        ret
+00010b10: 7572 6e20 6167 6573 0a0a 2020 2020 2320  urn ages..    # 
+00010b20: 636f 6e76 656e 6965 6e63 6520 6d65 7468  convenience meth
+00010b30: 6f64 2066 6f72 2067 6574 7469 6e67 2077  od for getting w
+00010b40: 686f 6c65 2073 7065 6369 6573 2720 7068  hole species' ph
+00010b50: 656e 6f74 7970 650a 2020 2020 6465 6620  enotype.    def 
+00010b60: 5f67 6574 5f7a 2873 656c 662c 2074 7261  _get_z(self, tra
+00010b70: 6974 5f6e 756d 3d4e 6f6e 652c 2069 6e64  it_num=None, ind
+00010b80: 6976 6964 733d 4e6f 6e65 293a 0a20 2020  ivids=None):.   
+00010b90: 2020 2020 207a 7320 3d20 7365 6c66 2e5f       zs = self._
+00010ba0: 6765 745f 7363 616c 6172 5f61 7474 7228  get_scalar_attr(
+00010bb0: 277a 272c 2069 6e64 6976 6964 733d 696e  'z', individs=in
+00010bc0: 6469 7669 6473 290a 2020 2020 2020 2020  divids).        
+00010bd0: 6966 2074 7261 6974 5f6e 756d 2069 7320  if trait_num is 
+00010be0: 6e6f 7420 4e6f 6e65 3a0a 2020 2020 2020  not None:.      
+00010bf0: 2020 2020 2020 7a73 203d 206e 702e 6174        zs = np.at
+00010c00: 6c65 6173 745f 3264 287a 7329 5b3a 2c74  least_2d(zs)[:,t
+00010c10: 7261 6974 5f6e 756d 5d0a 2020 2020 2020  rait_num].      
+00010c20: 2020 7265 7475 726e 207a 730a 0a20 2020    return zs..   
+00010c30: 2023 636f 6e76 656e 6965 6e63 6520 6d65   #convenience me
+00010c40: 7468 6f64 2066 6f72 2067 6574 7469 6e67  thod for getting
+00010c50: 2077 686f 6c65 2073 7065 6369 6573 2720   whole species' 
+00010c60: 6669 746e 6573 7365 730a 2020 2020 6465  fitnesses.    de
+00010c70: 6620 5f67 6574 5f66 6974 2873 656c 662c  f _get_fit(self,
+00010c80: 2069 6e64 6976 6964 7320 3d20 4e6f 6e65   individs = None
+00010c90: 293a 0a20 2020 2020 2020 2066 6974 7320  ):.        fits 
+00010ca0: 3d20 7365 6c66 2e5f 6765 745f 7363 616c  = self._get_scal
+00010cb0: 6172 5f61 7474 7228 2766 6974 272c 2069  ar_attr('fit', i
+00010cc0: 6e64 6976 6964 733d 696e 6469 7669 6473  ndivids=individs
+00010cd0: 290a 2020 2020 2020 2020 7265 7475 726e  ).        return
+00010ce0: 2066 6974 730a 0a20 2020 2064 6566 205f   fits..    def _
+00010cf0: 6361 6c63 5f66 6974 6e65 7373 2873 656c  calc_fitness(sel
+00010d00: 662c 2074 7261 6974 5f6e 756d 203d 204e  f, trait_num = N
+00010d10: 6f6e 652c 2073 6574 5f66 6974 203d 2054  one, set_fit = T
+00010d20: 7275 6529 3a0a 2020 2020 2020 2020 6669  rue):.        fi
+00010d30: 7420 3d20 5f63 616c 635f 6669 746e 6573  t = _calc_fitnes
+00010d40: 7328 7365 6c66 2c20 7472 6169 745f 6e75  s(self, trait_nu
+00010d50: 6d20 3d20 7472 6169 745f 6e75 6d29 0a20  m = trait_num). 
+00010d60: 2020 2020 2020 2023 7365 7420 696e 6469         #set indi
+00010d70: 7669 6475 616c 7327 2066 6974 6e65 7373  viduals' fitness
+00010d80: 2061 7474 7269 6275 7465 732c 2069 6620   attributes, if 
+00010d90: 696e 6469 6361 7465 640a 2020 2020 2020  indicated.      
+00010da0: 2020 6966 2073 6574 5f66 6974 3a0a 2020    if set_fit:.  
+00010db0: 2020 2020 2020 2020 2020 7365 6c66 2e5f            self._
+00010dc0: 7365 745f 6669 7428 6669 7429 0a20 2020  set_fit(fit).   
+00010dd0: 2020 2020 2072 6574 7572 6e20 6669 740a       return fit.
+00010de0: 0a20 2020 2064 6566 205f 6765 745f 646f  .    def _get_do
+00010df0: 6d28 7365 6c66 2c20 6c6f 6375 7329 3a0a  m(self, locus):.
+00010e00: 2020 2020 2020 2020 7265 7475 726e 207b          return {
+00010e10: 6c6f 6375 733a 2073 656c 662e 6765 6e5f  locus: self.gen_
+00010e20: 6172 6368 2e68 5b6c 6f63 7573 5d7d 0a0a  arch.h[locus]}..
+00010e30: 2020 2020 6465 6620 5f67 6574 5f63 6f6f      def _get_coo
+00010e40: 7264 7328 7365 6c66 2c20 696e 6469 7669  rds(self, indivi
+00010e50: 6473 3d4e 6f6e 652c 2061 735f 666c 6f61  ds=None, as_floa
+00010e60: 743d 5472 7565 293a 0a20 2020 2020 2020  t=True):.       
+00010e70: 2063 6f6f 7264 7320 3d20 6c69 7374 286d   coords = list(m
+00010e80: 6170 2873 656c 662e 5f63 6f6f 7264 5f61  ap(self._coord_a
+00010e90: 7474 7267 6574 7465 722c 2073 656c 662e  ttrgetter, self.
+00010ea0: 7661 6c75 6573 2829 2929 0a20 2020 2020  values())).     
+00010eb0: 2020 2069 6620 696e 6469 7669 6473 2069     if individs i
+00010ec0: 7320 6e6f 7420 4e6f 6e65 3a0a 2020 2020  s not None:.    
+00010ed0: 2020 2020 2020 2020 6967 203d 2069 7465          ig = ite
+00010ee0: 6d67 6574 7465 7228 2a69 6e64 6976 6964  mgetter(*individ
+00010ef0: 7329 0a20 2020 2020 2020 2020 2020 2063  s).            c
+00010f00: 6f6f 7264 7320 3d20 6967 2864 6963 7428  oords = ig(dict(
+00010f10: 7a69 7028 5b2a 7365 6c66 5d2c 2063 6f6f  zip([*self], coo
+00010f20: 7264 7329 2929 0a20 2020 2020 2020 2069  rds))).        i
+00010f30: 6620 6173 5f66 6c6f 6174 3a0a 2020 2020  f as_float:.    
+00010f40: 2020 2020 2020 2020 636f 6f72 6473 203d          coords =
+00010f50: 206e 702e 666c 6f61 7436 3428 636f 6f72   np.float64(coor
+00010f60: 6473 290a 2020 2020 2020 2020 656c 7365  ds).        else
+00010f70: 3a0a 2020 2020 2020 2020 2020 2020 636f  :.            co
+00010f80: 6f72 6473 203d 206e 702e 696e 7433 3228  ords = np.int32(
+00010f90: 6e70 2e66 6c6f 6f72 2863 6f6f 7264 7329  np.floor(coords)
+00010fa0: 290a 2020 2020 2020 2020 2320 6d61 6b65  ).        # make
+00010fb0: 2073 7572 6520 6974 2773 2061 7420 6c65   sure it's at le
+00010fc0: 6173 7420 3264 2028 696e 2063 6173 6520  ast 2d (in case 
+00010fd0: 6120 7369 6e67 6c65 2069 6e64 6976 6964  a single individ
+00010fe0: 7561 6c20 6973 2072 6571 7565 7374 6564  ual is requested
+00010ff0: 290a 2020 2020 2020 2020 636f 6f72 6473  ).        coords
+00011000: 203d 206e 702e 6174 6c65 6173 745f 3264   = np.atleast_2d
+00011010: 2863 6f6f 7264 7329 0a20 2020 2020 2020  (coords).       
+00011020: 2072 6574 7572 6e20 636f 6f72 6473 0a0a   return coords..
+00011030: 2020 2020 6465 6620 5f67 6574 5f70 6c6f      def _get_plo
+00011040: 745f 636f 6f72 6473 2873 656c 662c 2069  t_coords(self, i
+00011050: 6e64 6976 6964 733d 4e6f 6e65 2c20 6365  ndivids=None, ce
+00011060: 6c6c 5f63 6f6f 7264 733d 4661 6c73 6529  ll_coords=False)
+00011070: 3a0a 2020 2020 2020 2020 636f 6f72 6473  :.        coords
+00011080: 203d 2073 656c 662e 5f67 6574 5f63 6f6f   = self._get_coo
+00011090: 7264 7328 696e 6469 7669 6473 3d69 6e64  rds(individs=ind
+000110a0: 6976 6964 7329 0a20 2020 2020 2020 2069  ivids).        i
+000110b0: 6620 6e6f 7420 6365 6c6c 5f63 6f6f 7264  f not cell_coord
+000110c0: 733a 0a20 2020 2020 2020 2020 2020 2063  s:.            c
+000110d0: 6f6f 7264 735b 3a2c 2030 5d20 3d20 636f  oords[:, 0] = co
+000110e0: 6f72 6473 5b3a 2c20 305d 202a 2073 656c  ords[:, 0] * sel
+000110f0: 662e 5f6c 616e 645f 7265 735b 305d 202b  f._land_res[0] +
+00011100: 2073 656c 662e 5f6c 616e 645f 756c 635b   self._land_ulc[
+00011110: 305d 0a20 2020 2020 2020 2020 2020 2063  0].            c
+00011120: 6f6f 7264 735b 3a2c 2031 5d20 3d20 636f  oords[:, 1] = co
+00011130: 6f72 6473 5b3a 2c20 315d 202a 2073 656c  ords[:, 1] * sel
+00011140: 662e 5f6c 616e 645f 7265 735b 315d 202b  f._land_res[1] +
+00011150: 2073 656c 662e 5f6c 616e 645f 756c 635b   self._land_ulc[
+00011160: 315d 0a20 2020 2020 2020 2072 6574 7572  1].        retur
+00011170: 6e20 636f 6f72 6473 0a0a 2020 2020 6465  n coords..    de
+00011180: 6620 5f67 6574 5f63 656c 6c73 2873 656c  f _get_cells(sel
+00011190: 662c 2069 6e64 6976 6964 733d 4e6f 6e65  f, individs=None
+000111a0: 293a 0a20 2020 2020 2020 2063 656c 6c73  ):.        cells
+000111b0: 203d 2073 656c 662e 5f67 6574 5f63 6f6f   = self._get_coo
+000111c0: 7264 7328 696e 6469 7669 6473 3d69 6e64  rds(individs=ind
+000111d0: 6976 6964 732c 2061 735f 666c 6f61 743d  ivids, as_float=
+000111e0: 4661 6c73 6529 0a20 2020 2020 2020 2072  False).        r
+000111f0: 6574 7572 6e20 6365 6c6c 730a 0a20 2020  eturn cells..   
+00011200: 2064 6566 205f 6765 745f 7828 7365 6c66   def _get_x(self
+00011210: 2c20 696e 6469 7669 6473 3d4e 6f6e 6529  , individs=None)
+00011220: 3a0a 2020 2020 2020 2020 636f 6f72 6473  :.        coords
+00011230: 203d 2073 656c 662e 5f67 6574 5f63 6f6f   = self._get_coo
+00011240: 7264 7328 696e 6469 7669 6473 3d69 6e64  rds(individs=ind
+00011250: 6976 6964 7329 0a20 2020 2020 2020 2072  ivids).        r
+00011260: 6574 7572 6e20 636f 6f72 6473 5b3a 2c20  eturn coords[:, 
+00011270: 305d 0a0a 2020 2020 6465 6620 5f67 6574  0]..    def _get
+00011280: 5f79 2873 656c 662c 2069 6e64 6976 6964  _y(self, individ
+00011290: 733d 4e6f 6e65 293a 0a20 2020 2020 2020  s=None):.       
+000112a0: 2063 6f6f 7264 7320 3d20 7365 6c66 2e5f   coords = self._
+000112b0: 6765 745f 636f 6f72 6473 2869 6e64 6976  get_coords(indiv
+000112c0: 6964 733d 696e 6469 7669 6473 290a 2020  ids=individs).  
+000112d0: 2020 2020 2020 7265 7475 726e 2063 6f6f        return coo
+000112e0: 7264 735b 3a2c 2031 5d0a 0a20 2020 2023  rds[:, 1]..    #
+000112f0: 206d 6574 686f 6420 746f 2072 6574 7572   method to retur
+00011300: 6e20 616e 206e 2d6c 656e 6774 6820 6c69  n an n-length li
+00011310: 7374 206f 6620 7261 6e64 6f6d 2069 6e64  st of random ind
+00011320: 6976 6964 733b 0a20 2020 2023 2072 6574  ivids;.    # ret
+00011330: 7572 6e20 696e 6469 7669 6475 616c 732c  urn individuals,
+00011340: 206f 7220 696e 6469 6365 732c 2061 7320   or indices, as 
+00011350: 696e 6469 6361 7465 640a 2020 2020 6465  indicated.    de
+00011360: 6620 5f67 6574 5f72 616e 646f 6d5f 696e  f _get_random_in
+00011370: 6469 7669 6475 616c 7328 7365 6c66 2c20  dividuals(self, 
+00011380: 6e2c 2072 6574 7572 6e5f 666f 726d 6174  n, return_format
+00011390: 3d27 696e 6465 7827 293a 0a20 2020 2020  ='index'):.     
+000113a0: 2020 2061 7373 6572 7420 7265 7475 726e     assert return
+000113b0: 5f66 6f72 6d61 7420 696e 205b 0a20 2020  _format in [.   
+000113c0: 2020 2020 2020 2020 2027 696e 6465 7827           'index'
+000113d0: 2c20 2769 6e64 6976 6964 7561 6c27 5d2c  , 'individual'],
+000113e0: 2028 2241 7267 756d 656e 7420 7265 7475   ("Argument retu
+000113f0: 726e 5f66 6f72 6d61 7420 6361 6e20 7461  rn_format can ta
+00011400: 6b65 206f 6e6c 7920 220a 2020 2020 2020  ke only ".      
+00011410: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011420: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+00011430: 2769 6e64 6578 2720 6f72 2027 696e 6469  'index' or 'indi
+00011440: 7669 6475 616c 2720 6173 2076 616c 7565  vidual' as value
+00011450: 7320 220a 2020 2020 2020 2020 2020 2020  s ".            
+00011460: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011470: 2020 2020 2020 2020 2022 2864 6566 6175           "(defau
+00011480: 6c74 7320 746f 2027 696e 6465 7827 292e  lts to 'index').
+00011490: 2229 0a20 2020 2020 2020 2063 686f 6963  ").        choic
+000114a0: 6573 203d 2063 686f 6963 6573 203d 2072  es = choices = r
+000114b0: 2e63 686f 6963 6528 6c69 7374 2872 616e  .choice(list(ran
+000114c0: 6765 286c 656e 2873 656c 6629 2929 2c20  ge(len(self))), 
+000114d0: 6e2c 2072 6570 6c61 6365 3d46 616c 7365  n, replace=False
+000114e0: 290a 2020 2020 2020 2020 696e 6473 203d  ).        inds =
+000114f0: 206e 702e 6172 7261 7928 6c69 7374 2873   np.array(list(s
+00011500: 656c 662e 6b65 7973 2829 2929 5b63 686f  elf.keys()))[cho
+00011510: 6963 6573 5d0a 2020 2020 2020 2020 6966  ices].        if
+00011520: 2072 6574 7572 6e5f 666f 726d 6174 3d3d   return_format==
+00011530: 2769 6e64 6976 6964 7561 6c27 3a0a 2020  'individual':.  
+00011540: 2020 2020 2020 2020 2020 696e 6473 203d            inds =
+00011550: 205b 7365 6c66 5b69 6e64 5d20 666f 7220   [self[ind] for 
+00011560: 696e 6420 696e 2069 6e64 735d 0a20 2020  ind in inds].   
+00011570: 2020 2020 2072 6574 7572 6e28 696e 6473       return(inds
+00011580: 290a 0a20 2020 2023 2072 656d 6f76 6520  )..    # remove 
+00011590: 6120 7072 6564 6574 6572 6d69 6e65 6420  a predetermined 
+000115a0: 6f72 2072 616e 646f 6d20 7365 7420 6f66  or random set of
+000115b0: 2049 6e64 6976 6964 7561 6c73 2066 726f   Individuals fro
+000115c0: 6d20 7468 6520 5370 6563 6965 730a 2020  m the Species.  
+000115d0: 2020 6465 6620 5f72 656d 6f76 655f 696e    def _remove_in
+000115e0: 6469 7669 6475 616c 7328 7365 6c66 2c0a  dividuals(self,.
+000115f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011600: 2020 2020 2020 2020 2020 2020 696e 6469              indi
+00011610: 7669 6473 3d4e 6f6e 652c 0a20 2020 2020  vids=None,.     
+00011620: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011630: 2020 2020 2020 206e 3d4e 6f6e 652c 0a20         n=None,. 
+00011640: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011650: 2020 2020 2020 2020 2020 206e 5f6c 6566             n_lef
+00011660: 743d 4e6f 6e65 2c0a 2020 2020 2020 2020  t=None,.        
+00011670: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011680: 2020 2020 6b65 6570 5f73 6974 6573 5f74      keep_sites_t
+00011690: 6162 3d46 616c 7365 2c0a 2020 2020 2020  ab=False,.      
+000116a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000116b0: 2020 2020 2020 6368 6563 6b5f 6578 7469        check_exti
+000116c0: 6e63 743d 4661 6c73 652c 0a20 2020 2020  nct=False,.     
+000116d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000116e0: 2020 2020 2020 293a 0a0a 2020 2020 2020        ):..      
+000116f0: 2020 2320 7661 6c69 6461 7465 2061 7267    # validate arg
+00011700: 730a 2020 2020 2020 2020 6173 7365 7274  s.        assert
+00011710: 2028 6e70 2e73 756d 285b 7020 6973 204e   (np.sum([p is N
+00011720: 6f6e 6520 666f 7220 7020 696e 205b 696e  one for p in [in
+00011730: 6469 7669 6473 2c20 6e2c 206e 5f6c 6566  divids, n, n_lef
+00011740: 745d 5d29 203d 3d20 3220 616e 640a 2020  t]]) == 2 and.  
+00011750: 2020 2020 2020 2020 2020 2020 2020 6e70                np
+00011760: 2e73 756d 285b 7020 6973 206e 6f74 204e  .sum([p is not N
+00011770: 6f6e 6520 666f 7220 7020 696e 205b 696e  one for p in [in
+00011780: 6469 7669 6473 2c20 6e2c 206e 5f6c 6566  divids, n, n_lef
+00011790: 745d 5d29 203d 3d20 3129 2c20 280a 2020  t]]) == 1), (.  
+000117a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000117b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000117c0: 2020 224f 6e65 206f 6620 2769 6e64 6976    "One of 'indiv
+000117d0: 6964 7327 2c20 276e 272c 2061 6e64 2027  ids', 'n', and '
+000117e0: 6e5f 6c65 6674 2720 6d75 7374 220a 2020  n_left' must".  
+000117f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011800: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011810: 2020 2262 6520 7072 6f76 6964 6564 2c20    "be provided, 
+00011820: 7468 6520 6f74 6865 7220 7477 6f20 6d75  the other two mu
+00011830: 7374 2062 6520 4e6f 6e65 2e22 290a 2020  st be None.").  
+00011840: 2020 2020 2020 6966 2069 6e64 6976 6964        if individ
+00011850: 7320 6973 206e 6f74 204e 6f6e 653a 0a20  s is not None:. 
+00011860: 2020 2020 2020 2020 2020 2066 6f72 2069             for i
+00011870: 6420 696e 2069 6e64 6976 6964 733a 0a20  d in individs:. 
+00011880: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+00011890: 7373 6572 7420 6964 2069 6e20 7365 6c66  ssert id in self
+000118a0: 2e6b 6579 7328 292c 2066 2249 6e64 6976  .keys(), f"Indiv
+000118b0: 6964 7561 6c20 7b69 647d 2064 6f65 7320  idual {id} does 
+000118c0: 6e6f 7420 6578 6973 742e 220a 2020 2020  not exist.".    
+000118d0: 2020 2020 6966 2069 6e64 6976 6964 7320      if individs 
+000118e0: 6973 204e 6f6e 653a 0a20 2020 2020 2020  is None:.       
+000118f0: 2020 2020 2069 6620 6e5f 6c65 6674 2069       if n_left i
+00011900: 7320 6e6f 7420 4e6f 6e65 3a0a 2020 2020  s not None:.    
+00011910: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+00011920: 7274 206e 5f6c 6566 7420 3e3d 2030 2061  rt n_left >= 0 a
+00011930: 6e64 206e 5f6c 6566 743c 3d20 6c65 6e28  nd n_left<= len(
+00011940: 7365 6c66 292c 2028 2227 6e5f 6c65 6674  self), ("'n_left
+00011950: 2720 6d75 7374 2062 6520 220a 2020 2020  ' must be ".    
+00011960: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011970: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011980: 2020 2020 2020 2020 2020 2020 2261 206e              "a n
+00011990: 756d 6265 7220 6574 7765 656e 2030 2061  umber etween 0 a
+000119a0: 6e64 2022 0a20 2020 2020 2020 2020 2020  nd ".           
+000119b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000119c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000119d0: 2020 2020 2022 7468 6520 6375 7272 656e       "the curren
+000119e0: 7420 7369 7a65 206f 6620 7468 6520 220a  t size of the ".
+000119f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011a00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011a20: 6622 706f 7075 6c61 7469 6f6e 2028 7b6c  f"population ({l
+00011a30: 656e 2873 656c 6629 7d29 2e22 290a 2020  en(self)}).").  
+00011a40: 2020 2020 2020 2020 2020 2020 2020 6e20                n 
+00011a50: 3d20 6c65 6e28 7365 6c66 2920 2d20 6e5f  = len(self) - n_
+00011a60: 6c65 6674 0a20 2020 2020 2020 2020 2020  left.           
+00011a70: 2061 7373 6572 7420 6973 696e 7374 616e   assert isinstan
+00011a80: 6365 286e 2c20 696e 7429 2061 6e64 206e  ce(n, int) and n
+00011a90: 203e 3d20 302c 2028 2227 6e27 206d 7573   >= 0, ("'n' mus
+00011aa0: 7420 6569 7468 6572 2062 6520 220a 2020  t either be ".  
+00011ab0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011ac0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011ad0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011ae0: 2022 6120 6e6f 6e2d 6e65 6761 7469 7665   "a non-negative
+00011af0: 2069 6e74 2022 0a20 2020 2020 2020 2020   int ".         
+00011b00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011b10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011b20: 2020 2020 2020 2020 2020 226f 7220 4e6f            "or No
+00011b30: 6e65 2e22 290a 2020 2020 2020 2020 2020  ne.").          
+00011b40: 2020 6173 7365 7274 206e 203c 3d20 6c65    assert n <= le
+00011b50: 6e28 7365 6c66 292c 2028 2243 616e 6e6f  n(self), ("Canno
+00011b60: 7420 7265 6d6f 7665 206d 6f72 6520 496e  t remove more In
+00011b70: 6469 7669 6475 616c 7320 7468 616e 2022  dividuals than "
+00011b80: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00011b90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011ba0: 2020 2020 2263 7572 7265 6e74 6c79 2065      "currently e
+00011bb0: 7869 7374 2028 6375 7272 656e 7420 706f  xist (current po
+00011bc0: 7075 6c61 7469 6f6e 2073 697a 6520 220a  pulation size ".
+00011bd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011be0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011bf0: 2020 2066 2269 7320 7b6c 656e 2873 656c     f"is {len(sel
+00011c00: 6629 7d29 2e22 290a 0a20 2020 2020 2020  f)}).")..       
+00011c10: 2023 2064 7261 7720 7261 6e64 6f6d 2049   # draw random I
+00011c20: 6e64 6976 6964 7561 6c73 2c20 6966 206e  ndividuals, if n
+00011c30: 6563 6573 7361 7279 0a20 2020 2020 2020  ecessary.       
+00011c40: 2069 6620 696e 6469 7669 6473 2069 7320   if individs is 
+00011c50: 4e6f 6e65 3a0a 2020 2020 2020 2020 2020  None:.          
+00011c60: 2020 6964 7320 3d20 5b2a 7365 6c66 5d0a    ids = [*self].
+00011c70: 2020 2020 2020 2020 2020 2020 696e 6469              indi
+00011c80: 7669 6473 203d 206e 702e 7261 6e64 6f6d  vids = np.random
+00011c90: 2e63 686f 6963 6528 6964 732c 206e 2c20  .choice(ids, n, 
+00011ca0: 7265 706c 6163 653d 4661 6c73 6529 0a0a  replace=False)..
+00011cb0: 2020 2020 2020 2020 2320 6765 7420 6e75          # get nu
+00011cc0: 6d62 6572 206f 6620 496e 6469 7669 6475  mber of Individu
+00011cd0: 616c 7320 6265 666f 7265 206f 7065 7261  als before opera
+00011ce0: 7469 6f6e 0a20 2020 2020 2020 204e 5f62  tion.        N_b
+00011cf0: 3420 3d20 6c65 6e28 7365 6c66 290a 0a20  4 = len(self).. 
+00011d00: 2020 2020 2020 2023 2072 656d 6f76 6520         # remove 
+00011d10: 7468 6520 496e 6469 7669 6475 616c 730a  the Individuals.
+00011d20: 2020 2020 2020 2020 666f 7220 6964 2069          for id i
+00011d30: 6e20 696e 6469 7669 6473 3a0a 2020 2020  n individs:.    
+00011d40: 2020 2020 2020 2020 676f 6f64 6279 6520          goodbye 
+00011d50: 3d20 7365 6c66 2e70 6f70 2869 6429 0a0a  = self.pop(id)..
+00011d60: 2020 2020 2020 2020 2320 6173 7365 7274          # assert
+00011d70: 2074 6861 7420 706f 7075 6c61 7469 6f6e   that population
+00011d80: 2073 697a 6520 6861 7320 6368 616e 6765   size has change
+00011d90: 6420 7468 6520 636f 7272 6563 7420 616d  d the correct am
+00011da0: 6f75 6e74 0a20 2020 2020 2020 204e 5f61  ount.        N_a
+00011db0: 6620 3d20 6c65 6e28 7365 6c66 290a 2020  f = len(self).  
+00011dc0: 2020 2020 2020 6173 7365 7274 2028 4e5f        assert (N_
+00011dd0: 6234 202d 204e 5f61 6629 203d 3d20 6c65  b4 - N_af) == le
+00011de0: 6e28 696e 6469 7669 6473 292c 2028 2249  n(individs), ("I
+00011df0: 6e63 6f72 7265 6374 206e 756d 6265 7220  ncorrect number 
+00011e00: 6f66 2022 0a20 2020 2020 2020 2020 2020  of ".           
+00011e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011e20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00011e30: 2020 2020 2022 496e 6469 7669 6475 6c73       "Individuls
+00011e40: 2072 656d 6f76 6564 2122 290a 0a20 2020   removed!")..   
+00011e50: 2020 2020 2023 2075 7064 6174 6520 7468       # update th
+00011e60: 6520 5461 626c 6543 6f6c 6c65 6374 696f  e TableCollectio
+00011e70: 6e2c 2069 6620 6e65 6365 7373 6172 790a  n, if necessary.
+00011e80: 2020 2020 2020 2020 2320 2862 7574 206b          # (but k
+00011e90: 6565 7020 616e 6420 636f 7079 2062 6163  eep and copy bac
+00011ea0: 6b20 696e 2074 6865 2073 6974 6573 2074  k in the sites t
+00011eb0: 6162 6c65 2c20 6966 206e 6563 6573 7361  able, if necessa
+00011ec0: 7279 2c0a 2020 2020 2020 2020 2320 652e  ry,.        # e.
+00011ed0: 672e 2c20 7768 656e 2069 6e69 7420 7061  g., when init pa
+00011ee0: 7261 6d73 2061 7265 2072 656d 6f76 696e  rams are removin
+00011ef0: 6720 616e 6420 7265 706c 6163 696e 6720  g and replacing 
+00011f00: 616c 6c20 676e 782d 7369 6d75 6c61 7465  all gnx-simulate
+00011f10: 640a 2020 2020 2020 2020 2320 496e 6469  d.        # Indi
+00011f20: 7669 6475 616c 7329 0a20 2020 2020 2020  viduals).       
+00011f30: 2069 6620 7365 6c66 2e67 656e 5f61 7263   if self.gen_arc
+00011f40: 682e 7573 655f 7473 6b69 743a 0a20 2020  h.use_tskit:.   
+00011f50: 2020 2020 2020 2020 2069 6620 6b65 6570           if keep
+00011f60: 5f73 6974 6573 5f74 6162 3a0a 2020 2020  _sites_tab:.    
+00011f70: 2020 2020 2020 2020 2020 2020 7369 7465              site
+00011f80: 735f 7461 6220 3d20 7365 6c66 2e5f 7463  s_tab = self._tc
+00011f90: 2e73 6974 6573 0a20 2020 2020 2020 2020  .sites.         
+00011fa0: 2020 2073 656c 662e 5f73 6f72 745f 616e     self._sort_an
+00011fb0: 645f 7369 6d70 6c69 6679 5f74 6162 6c65  d_simplify_table
+00011fc0: 5f63 6f6c 6c65 6374 696f 6e28 290a 2020  _collection().  
+00011fd0: 2020 2020 2020 2020 2020 6966 206b 6565            if kee
+00011fe0: 705f 7369 7465 735f 7461 623a 0a20 2020  p_sites_tab:.   
+00011ff0: 2020 2020 2020 2020 2020 2020 2073 656c               sel
+00012000: 662e 5f74 632e 7369 7465 732e 7265 706c  f._tc.sites.repl
+00012010: 6163 655f 7769 7468 2873 6974 6573 5f74  ace_with(sites_t
+00012020: 6162 290a 0a20 2020 2020 2020 2023 2075  ab)..        # u
+00012030: 7064 6174 6520 7365 6c66 2e63 6f6f 7264  pdate self.coord
+00012040: 7320 616e 6420 7365 6c66 2e63 656c 6c73  s and self.cells
+00012050: 2061 7474 7269 6275 7465 730a 2020 2020   attributes.    
+00012060: 2020 2020 7365 6c66 2e5f 7365 745f 636f      self._set_co
+00012070: 6f72 6473 5f61 6e64 5f63 656c 6c73 2829  ords_and_cells()
+00012080: 0a0a 2020 2020 2020 2020 2320 616e 6420  ..        # and 
+00012090: 7570 6461 7465 2065 7874 696e 6374 2066  update extinct f
+000120a0: 6c61 670a 2020 2020 2020 2020 6966 2063  lag.        if c
+000120b0: 6865 636b 5f65 7874 696e 6374 3a0a 2020  heck_extinct:.  
+000120c0: 2020 2020 2020 2020 2020 7365 6c66 2e65            self.e
+000120d0: 7874 696e 6374 203d 2073 656c 662e 5f63  xtinct = self._c
+000120e0: 6865 636b 5f65 7874 696e 6374 2829 0a0a  heck_extinct()..
+000120f0: 2020 2020 2020 2020 2320 7072 696e 7420          # print 
+00012100: 696e 666f 726d 6174 6976 6520 6f75 7470  informative outp
+00012110: 7574 0a20 2020 2020 2020 2070 7269 6e74  ut.        print
+00012120: 2866 225c 6e7b 6c65 6e28 696e 6469 7669  (f"\n{len(indivi
+00012130: 6473 297d 2049 6e64 6976 6964 7561 6c73  ds)} Individuals
+00012140: 2073 7563 6365 7373 6675 6c6c 7920 7265   successfully re
+00012150: 6d6f 7665 642e 5c6e 2229 0a20 2020 2020  moved.\n").     
+00012160: 2020 2072 6574 7572 6e0a 0a0a 2020 2020     return...    
+00012170: 6465 6620 5f61 6464 5f69 6e64 6976 6964  def _add_individ
+00012180: 7561 6c73 2873 656c 662c 0a20 2020 2020  uals(self,.     
+00012190: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000121a0: 2020 2020 6e2c 0a20 2020 2020 2020 2020      n,.         
+000121b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000121c0: 636f 6f72 6473 2c0a 2020 2020 2020 2020  coords,.        
+000121d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000121e0: 206c 616e 642c 0a20 2020 2020 2020 2020   land,.         
+000121f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012200: 736f 7572 6365 5f73 7070 3d4e 6f6e 652c  source_spp=None,
+00012210: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00012220: 2020 2020 2020 2020 2020 736f 7572 6365            source
+00012230: 5f6d 7370 7269 6d65 5f70 6172 616d 733d  _msprime_params=
+00012240: 4e6f 6e65 2c0a 2020 2020 2020 2020 2020  None,.          
+00012250: 2020 2020 2020 2020 2020 2020 2020 2069                 i
+00012260: 6e64 6976 6964 733d 4e6f 6e65 2c0a 2020  ndivids=None,.  
+00012270: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012280: 2020 2020 2020 293a 0a20 2020 2020 2020        ):.       
+00012290: 2022 2222 0a20 2020 2020 2020 2041 6464   """.        Add
+000122a0: 2069 6e64 6976 6964 7561 6c73 2074 6f20   individuals to 
+000122b0: 7468 6973 2072 6563 6970 6965 6e74 2053  this recipient S
+000122c0: 7065 6369 6573 206f 626a 6563 742c 2065  pecies object, e
+000122d0: 6974 6865 7220 7573 696e 6720 6120 7365  ither using a se
+000122e0: 636f 6e64 0a20 2020 2020 2020 2053 7065  cond.        Spe
+000122f0: 6369 6573 206f 626a 6563 7420 6173 2074  cies object as t
+00012300: 6865 2073 6f75 7263 6520 706f 7075 6c61  he source popula
+00012310: 7469 6f6e 206f 7220 6665 6564 696e 6720  tion or feeding 
+00012320: 6120 6469 6374 206f 6620 7061 7261 6d65  a dict of parame
+00012330: 7465 7273 0a20 2020 2020 2020 2069 6e74  ters.        int
+00012340: 6f20 6d73 7072 696d 6520 746f 2073 696d  o msprime to sim
+00012350: 756c 6174 6520 7468 6520 736f 7572 6365  ulate the source
+00012360: 2070 6f70 756c 6174 696f 6e2e 0a20 2020   population..   
+00012370: 2020 2020 2028 646f 6373 2069 6e20 7468       (docs in th
+00012380: 6520 6d6f 6465 6c2e 7079 2077 7261 7070  e model.py wrapp
+00012390: 6572 206d 6574 686f 6429 0a20 2020 2020  er method).     
+000123a0: 2020 2022 2222 0a20 2020 2020 2020 2023     """.        #
+000123b0: 206d 616b 6520 7375 7265 2073 6f75 7263   make sure sourc
+000123c0: 655f 7370 7020 6973 2065 6974 6865 7220  e_spp is either 
+000123d0: 6120 5370 6563 6965 7320 6f62 6a65 6374  a Species object
+000123e0: 206f 7220 4e6f 6e65 0a20 2020 2020 2020   or None.       
+000123f0: 2061 7373 6572 7420 2869 7369 6e73 7461   assert (isinsta
+00012400: 6e63 6528 736f 7572 6365 5f73 7070 2c20  nce(source_spp, 
+00012410: 5370 6563 6965 7329 206f 7220 736f 7572  Species) or sour
+00012420: 6365 5f73 7070 2069 7320 4e6f 6e65 292c  ce_spp is None),
+00012430: 2028 2254 6865 2022 0a20 2020 2020 2020   ("The ".       
+00012440: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012450: 2022 736f 7572 6365 5f73 7070 2070 6172   "source_spp par
+00012460: 616d 206d 7573 7420 6265 2065 6974 6865  am must be eithe
+00012470: 7220 4e6f 6e65 206f 7220 616e 6f74 6865  r None or anothe
+00012480: 7220 220a 2020 2020 2020 2020 2020 2020  r ".            
+00012490: 2020 2020 2020 2020 2020 2020 2253 7065              "Spe
+000124a0: 6369 6573 206f 626a 6563 742e 2229 0a0a  cies object.")..
+000124b0: 2020 2020 2020 2020 2320 6765 7420 7468          # get th
+000124c0: 6520 6e65 7874 206e 2067 6e78 2049 6e64  e next n gnx Ind
+000124d0: 6976 6964 7561 6c73 2720 6964 730a 2020  ividuals' ids.  
+000124e0: 2020 2020 2020 6e65 7874 5f6e 5f69 6478        next_n_idx
+000124f0: 7320 3d20 5b2a 7261 6e67 6528 7365 6c66  s = [*range(self
+00012500: 2e6d 6178 5f69 6e64 5f69 6478 202b 2031  .max_ind_idx + 1
+00012510: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00012520: 2020 2020 2020 2020 2020 2020 2020 7365                se
+00012530: 6c66 2e6d 6178 5f69 6e64 5f69 6478 202b  lf.max_ind_idx +
+00012540: 206e 202b 2031 295d 0a0a 2020 2020 2020   n + 1)]..      
+00012550: 2020 2320 6368 6563 6b20 616e 6420 6861    # check and ha
+00012560: 6e64 6c65 2074 6865 2063 6f6f 7264 7320  ndle the coords 
+00012570: 7061 7261 6d0a 2020 2020 2020 2020 6173  param.        as
+00012580: 7365 7274 206e 702e 6174 6c65 6173 745f  sert np.atleast_
+00012590: 3264 2863 6f6f 7264 7329 2e73 6861 7065  2d(coords).shape
+000125a0: 2069 6e20 5b28 312c 3229 2c20 286e 2c32   in [(1,2), (n,2
+000125b0: 295d 2c20 280a 2020 2020 2020 2020 2020  )], (.          
+000125c0: 2020 2254 6865 2027 636f 6f72 6473 2720    "The 'coords' 
+000125d0: 6172 6720 6d75 7374 2062 6520 7072 6f76  arg must be prov
+000125e0: 6964 6564 2065 6974 6865 7220 220a 2020  ided either ".  
+000125f0: 2020 2020 2020 2020 2020 2261 2073 696e            "a sin
+00012600: 676c 6520 782c 7920 636f 6f72 6469 6e61  gle x,y coordina
+00012610: 7465 2070 6169 7220 2861 7420 7768 6963  te pair (at whic
+00012620: 6820 616c 6c20 496e 6469 7669 6475 616c  h all Individual
+00012630: 7320 220a 2020 2020 2020 2020 2020 2020  s ".            
+00012640: 2277 696c 6c20 6265 2069 6e74 726f 6475  "will be introdu
+00012650: 6365 6429 206f 7220 616e 206e 7832 2061  ced) or an nx2 a
+00012660: 7272 6179 206f 6620 782c 7920 636f 6f72  rray of x,y coor
+00012670: 6469 6e61 7465 2070 6169 7273 2022 0a20  dinate pairs ". 
+00012680: 2020 2020 2020 2020 2020 2022 2869 6e64             "(ind
+00012690: 6963 6174 696e 6720 7468 6520 696e 7472  icating the intr
+000126a0: 6f64 7563 7469 6f6e 206c 6f63 6174 696f  oduction locatio
+000126b0: 6e73 2066 6f72 2022 0a20 2020 2020 2020  ns for ".       
+000126c0: 2020 2020 2022 6561 6368 206f 6620 7468       "each of th
+000126d0: 6520 6e20 496e 6469 7669 6475 616c 7320  e n Individuals 
+000126e0: 6265 696e 6720 696e 7472 6f64 7563 6564  being introduced
+000126f0: 2e22 290a 2020 2020 2020 2020 6966 206e  .").        if n
+00012700: 702e 6174 6c65 6173 745f 3264 2863 6f6f  p.atleast_2d(coo
+00012710: 7264 7329 2e73 6861 7065 203d 3d20 2831  rds).shape == (1
+00012720: 2c20 3229 3a0a 2020 2020 2020 2020 2020  , 2):.          
+00012730: 2020 636f 6f72 6473 203d 206e 702e 7374    coords = np.st
+00012740: 6163 6b28 5b63 6f6f 7264 7320 666f 7220  ack([coords for 
+00012750: 5f20 696e 2072 616e 6765 286e 295d 290a  _ in range(n)]).
+00012760: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+00012770: 7274 2063 6f6f 7264 732e 7368 6170 6520  rt coords.shape 
+00012780: 3d3d 2028 6e2c 2032 290a 2020 2020 2020  == (n, 2).      
+00012790: 2020 2320 6d61 6b65 2073 7572 6520 636f    # make sure co
+000127a0: 6f72 6473 2061 7265 2061 6c6c 2077 6974  ords are all wit
+000127b0: 6869 6e20 7468 6520 6c61 6e64 7363 6170  hin the landscap
+000127c0: 650a 2020 2020 2020 2020 666f 7220 636f  e.        for co
+000127d0: 6f72 6420 696e 2063 6f6f 7264 733a 0a20  ord in coords:. 
+000127e0: 2020 2020 2020 2020 2020 2023 204e 4f54             # NOT
+000127f0: 453a 2073 7562 7472 6163 7469 6e67 2030  E: subtracting 0
+00012800: 2e30 3031 2066 726f 6d20 6469 6d73 2c20  .001 from dims, 
+00012810: 746f 206d 6174 6368 2068 6f77 2064 696d  to match how dim
+00012820: 7320 6172 6520 6861 6e64 6c65 640a 2020  s are handled.  
+00012830: 2020 2020 2020 2020 2020 2320 2020 2020            #     
+00012840: 2020 666f 7220 7374 6172 7469 6e67 2063    for starting c
+00012850: 6f6f 7264 696e 6174 6520 6173 7369 676e  oordinate assign
+00012860: 6d65 6e74 2069 6e0a 2020 2020 2020 2020  ment in.        
+00012870: 2020 2020 2320 2020 2020 2020 696e 6469      #       indi
+00012880: 7669 6475 616c 732e 5f6d 616b 655f 696e  viduals._make_in
+00012890: 6469 7669 6475 616c 0a20 2020 2020 2020  dividual.       
+000128a0: 2020 2020 2061 7373 6572 7420 2830 203c       assert (0 <
+000128b0: 3d20 636f 6f72 645b 305d 203c 3d20 7365  = coord[0] <= se
+000128c0: 6c66 2e5f 6c61 6e64 5f64 696d 5b30 5d20  lf._land_dim[0] 
+000128d0: 2d20 302e 3030 3120 616e 640a 2020 2020  - 0.001 and.    
+000128e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000128f0: 3020 3c3d 2063 6f6f 7264 5b31 5d20 3c3d  0 <= coord[1] <=
+00012900: 2073 656c 662e 5f6c 616e 645f 6469 6d5b   self._land_dim[
+00012910: 315d 202d 2030 2e30 3031 292c 2028 0a20  1] - 0.001), (. 
+00012920: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012930: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012940: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012950: 2020 2020 2243 6f6f 7264 7320 6d75 7374      "Coords must
+00012960: 2062 6520 220a 2020 2020 2020 2020 2020   be ".          
+00012970: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012980: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012990: 2020 2020 2020 2020 2020 2022 7661 6c69             "vali
+000129a0: 6420 666f 7220 7468 6520 7265 6369 7069  d for the recipi
+000129b0: 656e 7420 220a 2020 2020 2020 2020 2020  ent ".          
+000129c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000129d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000129e0: 2020 2020 2020 2020 2020 2022 5370 6563             "Spec
+000129f0: 6965 7327 204c 616e 6473 6361 7065 2022  ies' Landscape "
+00012a00: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00012a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012a20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012a30: 2020 2020 2020 2228 692e 652e 2c20 7820        "(i.e., x 
+00012a40: 616e 6420 7920 6d75 7374 2062 6520 220a  and y must be ".
+00012a50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012a60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012a70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012a80: 2020 2020 2022 696e 2074 6865 2069 6e63       "in the inc
+00012a90: 6c75 7369 7665 2022 0a20 2020 2020 2020  lusive ".       
+00012aa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012ab0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012ac0: 2020 2020 2020 2020 2020 2020 2020 2220                " 
+00012ad0: 696e 7465 7276 616c 7320 220a 2020 2020  intervals ".    
+00012ae0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012af0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012b00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012b10: 2022 5b30 2c20 6c61 6e64 5f64 696d 5f78   "[0, land_dim_x
+00012b20: 5d20 616e 6422 0a20 2020 2020 2020 2020  ] and".         
+00012b30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012b40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012b50: 2020 2020 2020 2020 2020 2020 225b 302c              "[0,
+00012b60: 206c 616e 645f 6469 6d5f 795d 2c20 220a   land_dim_y], ".
+00012b70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012b80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012b90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012ba0: 2020 2020 2022 7265 7370 6563 7469 7665       "respective
+00012bb0: 6c79 2e22 290a 0a20 2020 2020 2020 2023  ly.")..        #
+00012bc0: 2065 6974 6865 7220 6765 7420 496e 6469   either get Indi
+00012bd0: 7669 6475 616c 7320 616e 6420 7468 6569  viduals and thei
+00012be0: 7220 7473 6b69 742e 5461 626c 6543 6f6c  r tskit.TableCol
+00012bf0: 6c65 6374 696f 6e0a 2020 2020 2020 2020  lection.        
+00012c00: 2320 6672 6f6d 2074 6865 2067 6976 656e  # from the given
+00012c10: 2067 6e78 2053 7065 6369 6573 206f 626a   gnx Species obj
+00012c20: 6563 742e 2e2e 0a20 2020 2020 2020 2069  ect....        i
+00012c30: 6620 736f 7572 6365 5f73 7070 2069 7320  f source_spp is 
+00012c40: 6e6f 7420 4e6f 6e65 3a0a 2020 2020 2020  not None:.      
+00012c50: 2020 2020 2020 2320 7261 6973 6520 7761        # raise wa
+00012c60: 726e 696e 6720 6162 6f75 7420 706f 7465  rning about pote
+00012c70: 6e74 6961 6c20 756e 666f 7265 7365 656e  ntial unforeseen
+00012c80: 2065 6666 6563 7473 206f 6620 696e 7472   effects of intr
+00012c90: 6f64 7563 696e 670a 2020 2020 2020 2020  oducing.        
+00012ca0: 2020 2020 2320 496e 6469 7669 6475 616c      # Individual
+00012cb0: 7320 6672 6f6d 2061 6e20 696e 636f 7272  s from an incorr
+00012cc0: 6563 746c 7920 7061 7261 6d65 7465 7269  ectly parameteri
+00012cd0: 7a65 6420 736f 7572 6365 2053 7065 6369  zed source Speci
+00012ce0: 6573 0a20 2020 2020 2020 2020 2020 2077  es.            w
+00012cf0: 6172 6e69 6e67 732e 7761 726e 2828 2249  arnings.warn(("I
+00012d00: 6e74 726f 6475 6369 6e67 2049 6e64 6976  ntroducing Indiv
+00012d10: 6964 7561 6c73 2066 726f 6d20 6f6e 6520  iduals from one 
+00012d20: 5370 6563 6965 7320 6f62 6a65 6374 2022  Species object "
+00012d30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00012d40: 2020 2020 2022 696e 746f 2061 6e6f 7468       "into anoth
+00012d50: 6572 2063 6f75 6c64 2063 6175 7365 2075  er could cause u
+00012d60: 6e6e 6f74 6963 6564 2069 7373 7565 732e  nnoticed issues.
+00012d70: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00012d80: 2020 2020 2020 2022 4765 6f6e 6f6d 6963         "Geonomic
+00012d90: 7320 7769 6c6c 2063 6865 636b 2061 6e64  s will check and
+00012da0: 2070 7265 7665 6e74 2063 6f6e 666c 6963   prevent conflic
+00012db0: 7420 6265 7477 6565 6e20 220a 2020 2020  t between ".    
+00012dc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012dd0: 2273 7472 6963 746c 7920 696e 636f 6d70  "strictly incomp
+00012de0: 6174 6962 6c65 2061 7474 7269 6275 7465  atible attribute
+00012df0: 7320 6f66 2074 6865 2053 7065 6369 6573  s of the Species
+00012e00: 2c20 220a 2020 2020 2020 2020 2020 2020  , ".            
+00012e10: 2020 2020 2020 2020 2247 656e 6f6d 6963          "Genomic
+00012e20: 4172 6368 6974 6563 7475 7265 732c 2061  Architectures, a
+00012e30: 6e64 2054 7261 6974 732e 2048 6f77 6576  nd Traits. Howev
+00012e40: 6572 2c20 6174 7472 6962 7574 6573 2022  er, attributes "
+00012e50: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00012e60: 2020 2020 2022 7768 6f73 6520 636f 6d70       "whose comp
+00012e70: 6174 6962 696c 6974 7920 6465 7065 6e64  atibility depend
+00012e80: 7320 6f6e 2074 6865 2073 6369 656e 7469  s on the scienti
+00012e90: 6669 6320 220a 2020 2020 2020 2020 2020  fic ".          
+00012ea0: 2020 2020 2020 2020 2020 2272 6561 736f            "reaso
+00012eb0: 6e69 6e67 2062 6568 696e 6420 7468 6520  ning behind the 
+00012ec0: 7363 656e 6172 696f 2062 6569 6e67 2073  scenario being s
+00012ed0: 696d 756c 6174 6564 2022 0a20 2020 2020  imulated ".     
+00012ee0: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+00012ef0: 2865 2e67 2e2c 206d 7574 6174 696f 6e61  (e.g., mutationa
+00012f00: 6c20 7061 7261 6d65 7465 7273 2c20 220a  l parameters, ".
+00012f10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00012f20: 2020 2020 2263 6172 7279 696e 6720 6361      "carrying ca
+00012f30: 7061 6369 7479 2070 6172 616d 6574 6572  pacity parameter
+00012f40: 732c 2065 7463 2e29 2022 0a20 2020 2020  s, etc.) ".     
+00012f50: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+00012f60: 7769 6c6c 2062 6520 616c 6c6f 7765 6420  will be allowed 
+00012f70: 746f 2064 6966 6665 7220 6265 7477 6565  to differ betwee
+00012f80: 6e20 7468 6520 736f 7572 6365 2061 6e64  n the source and
+00012f90: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00012fa0: 2020 2020 2020 2022 7265 6369 7069 656e         "recipien
+00012fb0: 7420 5370 6563 6965 7320 6f62 6a65 6374  t Species object
+00012fc0: 7320 616e 642c 2022 0a20 2020 2020 2020  s and, ".       
+00012fd0: 2020 2020 2020 2020 2020 2020 2022 6966               "if
+00012fe0: 2074 6865 7365 2064 6966 6665 7265 6e63   these differenc
+00012ff0: 6573 2077 6572 6520 756e 696e 7465 6e64  es were unintend
+00013000: 6564 2062 7574 2067 6f20 220a 2020 2020  ed but go ".    
+00013010: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013020: 2275 6e64 6574 6563 7465 642c 2063 6f75  "undetected, cou
+00013030: 6c64 206c 6561 6420 746f 2069 6e63 6f72  ld lead to incor
+00013040: 7265 6374 206f 7220 6d69 736c 6561 6469  rect or misleadi
+00013050: 6e67 2064 6174 6120 220a 2020 2020 2020  ng data ".      
+00013060: 2020 2020 2020 2020 2020 2020 2020 2261                "a
+00013070: 6e64 2072 6573 756c 7473 2e20 506c 6561  nd results. Plea
+00013080: 7365 2075 7365 2063 6172 6520 7768 656e  se use care when
+00013090: 2070 6c61 6e6e 696e 672c 2022 0a20 2020   planning, ".   
+000130a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000130b0: 2022 696d 706c 656d 656e 7469 6e67 2c20   "implementing, 
+000130c0: 616e 6420 6368 6563 6b69 6e67 2079 6f75  and checking you
+000130d0: 7220 636f 6465 2e22 2929 0a0a 2020 2020  r code."))..    
+000130e0: 2020 2020 2020 2020 2320 6765 7420 7468          # get th
+000130f0: 6520 736f 7572 6365 2053 7065 6369 6573  e source Species
+00013100: 2066 726f 6d20 7769 7468 696e 2074 6869   from within thi
+00013110: 7320 6d6f 6465 6c2c 2069 6620 6e65 6365  s model, if nece
+00013120: 7373 6172 790a 2020 2020 2020 2020 2020  ssary.          
+00013130: 2020 6966 2069 7369 6e73 7461 6e63 6528    if isinstance(
+00013140: 736f 7572 6365 5f73 7070 2c20 696e 7429  source_spp, int)
+00013150: 206f 7220 6973 696e 7374 616e 6365 2873   or isinstance(s
+00013160: 6f75 7263 655f 7370 702c 2073 7472 293a  ource_spp, str):
+00013170: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00013180: 2073 6f75 7263 655f 7370 7020 3d20 7365   source_spp = se
+00013190: 6c66 2e63 6f6d 6d5b 7365 6c66 2e5f 6765  lf.comm[self._ge
+000131a0: 745f 7370 705f 6e75 6d28 736f 7572 6365  t_spp_num(source
+000131b0: 5f73 7070 295d 0a0a 2020 2020 2020 2020  _spp)]..        
+000131c0: 2020 2020 2320 6173 7365 7274 2074 6861      # assert tha
+000131d0: 7420 6578 6163 746c 7920 6f6e 6520 6f66  t exactly one of
+000131e0: 2074 6865 2027 6e27 2061 6e64 2027 696e   the 'n' and 'in
+000131f0: 6469 7669 6473 270a 2020 2020 2020 2020  divids'.        
+00013200: 2020 2020 2320 7061 7261 6d73 2069 7320      # params is 
+00013210: 7072 6f76 6964 6564 2c20 616e 6420 7468  provided, and th
+00013220: 6174 2074 6865 2073 697a 6520 6f66 2027  at the size of '
+00013230: 6e27 0a20 2020 2020 2020 2020 2020 2023  n'.            #
+00013240: 206f 7220 7369 7a65 2061 6e64 2063 6f6e   or size and con
+00013250: 7465 6e74 7320 6f66 2027 696e 6469 7669  tents of 'indivi
+00013260: 6473 2720 6172 6520 7661 6c69 640a 2020  ds' are valid.  
+00013270: 2020 2020 2020 2020 2020 6173 7365 7274            assert
+00013280: 2028 286e 2069 7320 6e6f 7420 4e6f 6e65   ((n is not None
+00013290: 2061 6e64 2069 6e64 6976 6964 7320 6973   and individs is
+000132a0: 204e 6f6e 6529 206f 720a 2020 2020 2020   None) or.      
+000132b0: 2020 2020 2020 2020 2020 2020 2020 286e                (n
+000132c0: 2069 7320 4e6f 6e65 2061 6e64 2069 6e64   is None and ind
+000132d0: 6976 6964 7320 6973 206e 6f74 204e 6f6e  ivids is not Non
+000132e0: 6529 292c 2028 2249 6620 7573 696e 6720  e)), ("If using 
+000132f0: 616e 6f74 6865 7220 220a 2020 2020 2020  another ".      
+00013300: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013310: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013320: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013330: 2020 2247 656f 6e6f 6d69 6373 2053 7065    "Geonomics Spe
+00013340: 6369 6573 2022 0a20 2020 2020 2020 2020  cies ".         
+00013350: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013360: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013370: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+00013380: 6f62 6a65 6374 2061 7320 7468 6520 736f  object as the so
+00013390: 7572 6365 2022 0a20 2020 2020 2020 2020  urce ".         
+000133a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000133b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000133c0: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+000133d0: 706f 7075 6c61 7469 6f6e 2074 6865 6e20  population then 
+000133e0: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
+000133f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013400: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013410: 2020 2020 2020 2020 2020 2265 6974 6865            "eithe
+00013420: 7220 606e 6020 6f72 2022 0a20 2020 2020  r `n` or ".     
+00013430: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00013440: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013450: 2020 2020 2776 6172 5f64 6627 3a76 6172      'var_df':var
-00013460: 5f64 662c 0a20 2020 2020 2020 2020 2020  _df,.           
-00013470: 2020 2020 2020 2020 2027 7472 6169 745f           'trait_
-00013480: 6c6f 6369 273a 7472 6169 745f 6c6f 6369  loci':trait_loci
-00013490: 7d0a 0a20 2020 2020 2020 2023 204e 4f54  }..        # NOT
-000134a0: 453a 2072 6967 6874 206e 6f77 2074 6869  E: right now thi
-000134b0: 7320 706c 6f74 7469 6e67 206c 6f6f 7020  s plotting loop 
-000134c0: 6f6e 6c79 2077 6f72 6b73 2066 6f72 2061  only works for a
-000134d0: 206d 6178 696d 756d 206f 6620 3320 6178   maximum of 3 ax
-000134e0: 6573 3b0a 2020 2020 2020 2020 2320 2020  es;.        #   
-000134f0: 2020 2020 696e 2074 6865 2066 7574 7572      in the futur
-00013500: 6520 6173 206d 6f72 6520 656e 7620 7661  e as more env va
-00013510: 7273 2061 7265 2061 6464 6564 2c20 636f  rs are added, co
-00013520: 756c 6420 6164 6420 6d6f 7265 2061 7865  uld add more axe
-00013530: 730a 2020 2020 2020 2020 6966 2070 6c6f  s.        if plo
-00013540: 743a 0a20 2020 2020 2020 2020 2020 2023  t:.            #
-00013550: 5363 616c 6520 6461 7461 6672 616d 6573  Scale dataframes
-00013560: 2066 726f 6d20 2d31 2074 6f20 3120 6966   from -1 to 1 if
-00013570: 2073 6361 6c65 203d 2054 7275 650a 2020   scale = True.  
-00013580: 2020 2020 2020 2020 2020 6966 2073 6361            if sca
-00013590: 6c65 203d 3d20 5472 7565 3a0a 2020 2020  le == True:.    
-000135a0: 2020 2020 2020 2020 2020 2020 726d 696e              rmin
-000135b0: 203d 202d 310a 2020 2020 2020 2020 2020   = -1.          
-000135c0: 2020 2020 2020 726d 6178 203d 2031 0a20        rmax = 1. 
-000135d0: 2020 2020 2020 2020 2020 2020 2020 206c                 l
-000135e0: 6f63 695f 6466 203d 2072 6d69 6e20 2b20  oci_df = rmin + 
-000135f0: 286c 6f63 695f 6466 202d 206c 6f63 695f  (loci_df - loci_
-00013600: 6466 2e76 616c 7565 732e 6d69 6e28 2929  df.values.min())
-00013610: 202a 2028 726d 6178 202d 0a20 2020 2020   * (rmax -.     
-00013620: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013630: 2020 2028 726d 696e 2929 202f 2028 6c6f     (rmin)) / (lo
-00013640: 6369 5f64 662e 7661 6c75 6573 2e6d 6178  ci_df.values.max
-00013650: 2829 202d 206c 6f63 695f 6466 2e76 616c  () - loci_df.val
-00013660: 7565 732e 6d69 6e28 2929 0a20 2020 2020  ues.min()).     
-00013670: 2020 2020 2020 2020 2020 2076 6172 5f64             var_d
-00013680: 6620 3d20 726d 696e 202b 2028 7661 725f  f = rmin + (var_
-00013690: 6466 202d 2076 6172 5f64 662e 7661 6c75  df - var_df.valu
-000136a0: 6573 2e6d 696e 2829 2920 2a20 2872 6d61  es.min()) * (rma
-000136b0: 7820 2d0a 2020 2020 2020 2020 2020 2020  x -.            
-000136c0: 2020 2020 2020 2020 2020 2020 2872 6d69              (rmi
-000136d0: 6e29 2920 2f20 2876 6172 5f64 662e 7661  n)) / (var_df.va
-000136e0: 6c75 6573 2e6d 6178 2829 202d 2076 6172  lues.max() - var
-000136f0: 5f64 662e 7661 6c75 6573 2e6d 696e 2829  _df.values.min()
-00013700: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-00013710: 2020 696e 645f 6466 203d 2072 6d69 6e20    ind_df = rmin 
-00013720: 2b20 2869 6e64 5f64 6620 2d20 696e 645f  + (ind_df - ind_
-00013730: 6466 2e76 616c 7565 732e 6d69 6e28 2929  df.values.min())
-00013740: 202a 2028 726d 6178 202d 0a20 2020 2020   * (rmax -.     
-00013750: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013760: 2020 2028 726d 696e 2929 202f 2028 696e     (rmin)) / (in
-00013770: 645f 6466 2e76 616c 7565 732e 6d61 7828  d_df.values.max(
-00013780: 2920 2d20 696e 645f 6466 2e76 616c 7565  ) - ind_df.value
-00013790: 732e 6d69 6e28 2929 0a0a 0a20 2020 2020  s.min())...     
-000137a0: 2020 2020 2020 2023 6765 7420 6d61 7820         #get max 
-000137b0: 616e 6420 6d69 6e73 2074 6f20 7365 7420  and mins to set 
-000137c0: 6178 6973 206c 6174 6572 206f 6e0a 2020  axis later on.  
-000137d0: 2020 2020 2020 2020 2020 6d61 7864 6620            maxdf 
-000137e0: 3d20 6d61 7828 696e 645f 6466 2e76 616c  = max(ind_df.val
-000137f0: 7565 732e 6d61 7828 292c 2076 6172 5f64  ues.max(), var_d
-00013800: 662e 7661 6c75 6573 2e6d 6178 2829 2c0a  f.values.max(),.
-00013810: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013820: 2020 2020 2020 2020 6c6f 6369 5f64 662e          loci_df.
-00013830: 7661 6c75 6573 2e6d 6178 2829 290a 2020  values.max()).  
-00013840: 2020 2020 2020 2020 2020 6d69 6e64 6620            mindf 
-00013850: 3d20 6162 7328 6d69 6e28 696e 645f 6466  = abs(min(ind_df
-00013860: 2e76 616c 7565 732e 6d69 6e28 292c 2076  .values.min(), v
-00013870: 6172 5f64 662e 7661 6c75 6573 2e6d 696e  ar_df.values.min
-00013880: 2829 2c0a 2020 2020 2020 2020 2020 2020  (),.            
-00013890: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000138a0: 6c6f 6369 5f64 662e 7661 6c75 6573 2e6d  loci_df.values.m
-000138b0: 696e 2829 2929 0a20 2020 2020 2020 2020  in())).         
-000138c0: 2020 2061 786d 6178 203d 206d 6178 286d     axmax = max(m
-000138d0: 6178 6466 2c20 6d69 6e64 6629 202a 2031  axdf, mindf) * 1
-000138e0: 2e32 3020 2361 6464 696e 6720 6120 3230  .20 #adding a 20
-000138f0: 2520 6275 6666 6572 0a0a 2020 2020 2020  % buffer..      
-00013900: 2020 2020 2020 2373 6574 2075 7020 6669        #set up fi
-00013910: 6775 7265 0a20 2020 2020 2020 2020 2020  gure.           
-00013920: 2066 6967 203d 2070 6c74 2e66 6967 7572   fig = plt.figur
-00013930: 6528 6669 6773 697a 653d 2832 302c 2031  e(figsize=(20, 1
-00013940: 3529 290a 2020 2020 2020 2020 2020 2020  5)).            
-00013950: 666f 7220 6e2c 2063 635f 6178 6573 5f70  for n, cc_axes_p
-00013960: 6169 7220 696e 2065 6e75 6d65 7261 7465  air in enumerate
-00013970: 285b 5b31 2c20 325d 2c20 5b31 2c20 335d  ([[1, 2], [1, 3]
-00013980: 2c20 5b32 2c20 335d 5d29 3a0a 2020 2020  , [2, 3]]):.    
-00013990: 2020 2020 2020 2020 2020 2020 2364 6566              #def
-000139a0: 696e 6520 636f 6d70 6f6e 656e 7473 2066  ine components f
-000139b0: 6f72 2061 7869 730a 2020 2020 2020 2020  or axis.        
-000139c0: 2020 2020 2020 2020 6363 5f61 7869 7331          cc_axis1
-000139d0: 203d 2063 635f 6178 6573 5f70 6169 725b   = cc_axes_pair[
-000139e0: 305d 2023 782d 6178 6973 2043 430a 2020  0] #x-axis CC.  
-000139f0: 2020 2020 2020 2020 2020 2020 2020 6363                cc
-00013a00: 5f61 7869 7332 203d 2063 635f 6178 6573  _axis2 = cc_axes
-00013a10: 5f70 6169 725b 315d 2023 792d 6178 6973  _pair[1] #y-axis
-00013a20: 2043 430a 0a20 2020 2020 2020 2020 2020   CC..           
-00013a30: 2020 2020 2061 7820 3d20 6669 672e 6164       ax = fig.ad
-00013a40: 645f 7375 6270 6c6f 7428 3133 312b 6e29  d_subplot(131+n)
-00013a50: 0a0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00013a60: 2020 2361 6464 2063 656e 7465 7220 6c69    #add center li
-00013a70: 6e65 730a 2020 2020 2020 2020 2020 2020  nes.            
-00013a80: 2020 2020 6178 2e61 7868 6c69 6e65 2879      ax.axhline(y
-00013a90: 3d30 2c20 636f 6c6f 723d 276c 6967 6874  =0, color='light
-00013aa0: 6772 6179 272c 206c 696e 6573 7479 6c65  gray', linestyle
-00013ab0: 3d27 646f 7474 6564 2729 0a20 2020 2020  ='dotted').     
-00013ac0: 2020 2020 2020 2020 2020 2061 782e 6178             ax.ax
-00013ad0: 766c 696e 6528 783d 302c 2063 6f6c 6f72  vline(x=0, color
-00013ae0: 3d27 6c69 6768 7467 7261 7927 2c20 6c69  ='lightgray', li
-00013af0: 6e65 7374 796c 653d 2764 6f74 7465 6427  nestyle='dotted'
-00013b00: 290a 0a20 2020 2020 2020 2020 2020 2020  )..             
-00013b10: 2020 2023 2073 6574 2061 7865 7320 7261     # set axes ra
-00013b20: 6e67 650a 2020 2020 2020 2020 2020 2020  nge.            
-00013b30: 2020 2020 706c 742e 786c 696d 282d 6178      plt.xlim(-ax
-00013b40: 6d61 782c 2061 786d 6178 290a 2020 2020  max, axmax).    
-00013b50: 2020 2020 2020 2020 2020 2020 706c 742e              plt.
-00013b60: 796c 696d 282d 6178 6d61 782c 2061 786d  ylim(-axmax, axm
-00013b70: 6178 290a 0a20 2020 2020 2020 2020 2020  ax)..           
-00013b80: 2020 2020 2023 706c 6f74 206e 6575 7472       #plot neutr
-00013b90: 616c 2053 4e50 730a 2020 2020 2020 2020  al SNPs.        
-00013ba0: 2020 2020 2020 2020 6178 2e73 6361 7474          ax.scatt
-00013bb0: 6572 286c 6f63 695f 6466 5b73 7472 2863  er(loci_df[str(c
-00013bc0: 635f 6178 6973 3129 5d2c 206c 6f63 695f  c_axis1)], loci_
-00013bd0: 6466 5b73 7472 2863 635f 6178 6973 3229  df[str(cc_axis2)
-00013be0: 5d2c 0a20 2020 2020 2020 2020 2020 2020  ],.             
-00013bf0: 2020 2020 2020 2020 2020 2020 2020 6d61                ma
-00013c00: 726b 6572 203d 2027 2b27 2c20 636f 6c6f  rker = '+', colo
-00013c10: 7220 3d20 2767 7261 7927 2c0a 2020 2020  r = 'gray',.    
-00013c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013c30: 2020 2020 2020 206c 6162 656c 203d 2027         label = '
-00013c40: 4e65 7574 7261 6c20 534e 5073 272c 2061  Neutral SNPs', a
-00013c50: 6c70 6861 203d 2030 2e37 290a 2020 2020  lpha = 0.7).    
-00013c60: 2020 2020 2020 2020 2020 2020 2370 6c6f              #plo
-00013c70: 7420 696e 6469 7669 6475 616c 730a 2020  t individuals.  
-00013c80: 2020 2020 2020 2020 2020 2020 2020 636d                cm
-00013c90: 6170 5f73 6361 7420 3d20 6178 2e73 6361  ap_scat = ax.sca
-00013ca0: 7474 6572 2869 6e64 5f64 665b 7374 7228  tter(ind_df[str(
-00013cb0: 6363 5f61 7869 7331 295d 2c0a 2020 2020  cc_axis1)],.    
-00013cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013cd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013ce0: 2020 2069 6e64 5f64 665b 7374 7228 6363     ind_df[str(cc
-00013cf0: 5f61 7869 7332 295d 2c0a 2020 2020 2020  _axis2)],.      
-00013d00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013d10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013d20: 2063 203d 2067 6561 5f64 662e 656e 762c   c = gea_df.env,
-00013d30: 2063 6d61 703d 2776 6972 6964 6973 272c   cmap='viridis',
-00013d40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00013d50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013d60: 2020 2020 2020 2020 6c61 6265 6c20 3d20          label = 
-00013d70: 2249 6e64 6976 6964 7561 6c73 222c 2061  "Individuals", a
-00013d80: 6c70 6861 203d 2030 2e35 290a 2020 2020  lpha = 0.5).    
-00013d90: 2020 2020 2020 2020 2020 2020 2370 6c6f              #plo
-00013da0: 7420 6164 6170 7469 7665 2053 4e50 730a  t adaptive SNPs.
-00013db0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013dc0: 6178 2e73 6361 7474 6572 286c 6f63 695f  ax.scatter(loci_
-00013dd0: 6466 5b73 7472 2863 635f 6178 6973 3129  df[str(cc_axis1)
-00013de0: 5d5b 7472 6169 745f 6c6f 6369 5d2c 0a20  ][trait_loci],. 
-00013df0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013e00: 2020 2020 2020 2020 2020 6c6f 6369 5f64            loci_d
-00013e10: 665b 7374 7228 6363 5f61 7869 7332 295d  f[str(cc_axis2)]
-00013e20: 5b74 7261 6974 5f6c 6f63 695d 2c0a 2020  [trait_loci],.  
-00013e30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013e40: 2020 2020 2020 2020 206d 6172 6b65 723d           marker=
-00013e50: 222b 222c 2063 6f6c 6f72 203d 2022 7265  "+", color = "re
-00013e60: 6422 2c20 733d 3130 302c 0a20 2020 2020  d", s=100,.     
-00013e70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013e80: 2020 2020 2020 6c61 6265 6c20 3d20 2741        label = 'A
-00013e90: 6461 7074 6976 6520 534e 5073 2729 0a0a  daptive SNPs')..
-00013ea0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013eb0: 2370 6c6f 7420 7661 7269 6162 6c65 2076  #plot variable v
-00013ec0: 6563 746f 7273 2061 7320 6172 726f 7773  ectors as arrows
-00013ed0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00013ee0: 2066 6f72 2069 2069 6e20 7261 6e67 6528   for i in range(
-00013ef0: 7661 725f 6466 2e73 6861 7065 5b30 5d29  var_df.shape[0])
-00013f00: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-00013f10: 2020 2020 2020 7820 3d20 7661 725f 6466        x = var_df
-00013f20: 5b73 7472 2863 635f 6178 6973 3129 5d5b  [str(cc_axis1)][
-00013f30: 695d 0a20 2020 2020 2020 2020 2020 2020  i].             
-00013f40: 2020 2020 2020 2079 203d 2076 6172 5f64         y = var_d
-00013f50: 665b 7374 7228 6363 5f61 7869 7332 295d  f[str(cc_axis2)]
-00013f60: 5b69 5d0a 2020 2020 2020 2020 2020 2020  [i].            
-00013f70: 2020 2020 2020 2020 706c 742e 6172 726f          plt.arro
-00013f80: 7728 302c 2030 2c20 782c 2079 2c20 7769  w(0, 0, x, y, wi
-00013f90: 6474 6820 3d20 302e 3031 2c0a 2020 2020  dth = 0.01,.    
-00013fa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013fb0: 2020 2020 2020 2020 2020 6865 6164 5f77            head_w
-00013fc0: 6964 7468 203d 2030 2e30 352c 2063 6f6c  idth = 0.05, col
-00013fd0: 6f72 203d 2027 626c 6163 6b27 290a 2020  or = 'black').  
-00013fe0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00013ff0: 2020 7378 203d 2030 2e31 0a20 2020 2020    sx = 0.1.     
-00014000: 2020 2020 2020 2020 2020 2020 2020 2073                 s
-00014010: 7920 3d20 302e 310a 2020 2020 2020 2020  y = 0.1.        
-00014020: 2020 2020 2020 2020 2020 2020 2320 7468              # th
-00014030: 6973 206d 6573 7320 6973 206a 7573 7420  is mess is just 
-00014040: 746f 2061 7272 616e 6765 2074 6865 2074  to arrange the t
-00014050: 6578 740a 2020 2020 2020 2020 2020 2020  ext.            
-00014060: 2020 2020 2020 2020 2320 6e65 7874 2074          # next t
-00014070: 6f20 6172 726f 7773 2062 7574 2069 7420  o arrows but it 
-00014080: 6973 2061 2057 4950 0a20 2020 2020 2020  is a WIP.       
-00014090: 2020 2020 2020 2020 2020 2020 2069 6620               if 
-000140a0: 2878 203c 2030 2061 6e64 2079 203c 2030  (x < 0 and y < 0
-000140b0: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-000140c0: 2020 2020 2020 2020 2020 706c 742e 7465            plt.te
-000140d0: 7874 2878 202d 2073 782c 2079 202d 2073  xt(x - sx, y - s
-000140e0: 792c 2058 2e63 6f6c 756d 6e73 5b69 5d29  y, X.columns[i])
-000140f0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00014100: 2020 2020 2069 6620 2878 203c 2030 2061       if (x < 0 a
-00014110: 6e64 2079 203e 2030 293a 0a20 2020 2020  nd y > 0):.     
+00013450: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013460: 2020 2022 6069 6e64 6976 6964 7360 206d     "`individs` m
+00013470: 7573 7420 6265 2022 0a20 2020 2020 2020  ust be ".       
+00013480: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013490: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000134a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000134b0: 2022 7072 6f76 6964 6564 2c20 220a 2020   "provided, ".  
+000134c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000134d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000134e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000134f0: 2020 2020 2020 2262 7574 206e 6f74 2062        "but not b
+00013500: 6f74 682e 2229 0a20 2020 2020 2020 2020  oth.").         
+00013510: 2020 2069 6620 696e 6469 7669 6473 2069     if individs i
+00013520: 7320 4e6f 6e65 3a0a 2020 2020 2020 2020  s None:.        
+00013530: 2020 2020 2020 2020 6173 7365 7274 206e          assert n
+00013540: 203c 3d20 6c65 6e28 736f 7572 6365 5f73   <= len(source_s
+00013550: 7070 292c 2028 2260 6e60 206d 7573 7420  pp), ("`n` must 
+00013560: 6e6f 7420 6265 2067 7265 6174 6572 2074  not be greater t
+00013570: 6861 6e20 220a 2020 2020 2020 2020 2020  han ".          
+00013580: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013590: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000135a0: 2020 2020 2274 6865 2070 6f70 756c 6174      "the populat
+000135b0: 696f 6e20 7369 7a65 206f 6620 220a 2020  ion size of ".  
+000135c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000135d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000135e0: 2020 2020 2020 2020 2020 2020 2260 736f              "`so
+000135f0: 7572 6365 5f73 7070 602e 2229 0a20 2020  urce_spp`.").   
+00013600: 2020 2020 2020 2020 2065 6c73 653a 0a20           else:. 
+00013610: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+00013620: 7373 6572 7420 286c 656e 286e 702e 756e  ssert (len(np.un
+00013630: 6971 7565 2869 6e64 6976 6964 7329 2920  ique(individs)) 
+00013640: 3c3d 206c 656e 2873 6f75 7263 655f 7370  <= len(source_sp
+00013650: 7029 2061 6e64 0a20 2020 2020 2020 2020  p) and.         
+00013660: 2020 2020 2020 2020 2020 2020 2020 206e                 n
+00013670: 702e 616c 6c28 5b69 6e64 2069 6e20 736f  p.all([ind in so
+00013680: 7572 6365 5f73 7070 2066 6f72 2069 6e64  urce_spp for ind
+00013690: 2069 6e20 696e 6469 7669 6473 5d29 292c   in individs])),
+000136a0: 2028 0a20 2020 2020 2020 2020 2020 2020   (.             
+000136b0: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+000136c0: 6069 6e64 6976 6964 7360 2063 6f6e 7461  `individs` conta
+000136d0: 696e 7320 496e 6469 7669 6475 616c 2069  ins Individual i
+000136e0: 6e64 6963 6573 2074 6861 7420 646f 2022  ndices that do "
+000136f0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00013700: 2020 2020 2020 2020 2020 2020 2022 6e6f               "no
+00013710: 7420 6578 6973 7420 696e 2060 736f 7572  t exist in `sour
+00013720: 6365 5f73 7070 602e 2229 0a0a 2020 2020  ce_spp`.")..    
+00013730: 2020 2020 2020 2020 2320 6368 6563 6b20          # check 
+00013740: 7468 6174 2061 6c6c 2053 7065 6369 6573  that all Species
+00013750: 2070 6172 616d 7320 7468 6174 206e 6565   params that nee
+00013760: 6420 746f 206d 6174 6368 2061 7265 2074  d to match are t
+00013770: 6865 2073 616d 650a 2020 2020 2020 2020  he same.        
+00013780: 2020 2020 2320 696e 2074 6865 2073 6f75      # in the sou
+00013790: 7263 6520 706f 7075 6c61 7469 6f6e 2773  rce population's
+000137a0: 2053 7065 6369 6573 2061 6e64 2069 6e20   Species and in 
+000137b0: 7468 6973 2053 7065 6369 6573 2720 706f  this Species' po
+000137c0: 700a 2020 2020 2020 2020 2020 2020 7370  p.            sp
+000137d0: 705f 6174 7472 735f 746f 5f63 6865 636b  p_attrs_to_check
+000137e0: 203d 205b 274b 5f6c 6179 6572 272c 0a20   = ['K_layer',. 
+000137f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013800: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013810: 2027 7365 6c65 6374 696f 6e27 2c0a 2020   'selection',.  
+00013820: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013830: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013840: 2773 6578 5f72 6174 696f 272c 0a20 2020  'sex_ratio',.   
+00013850: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013860: 2020 2020 2020 2020 2020 2020 2020 2027                 '
+00013870: 6d6f 7665 272c 0a20 2020 2020 2020 2020  move',.         
+00013880: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013890: 2020 2020 2020 2020 5d0a 2020 2020 2020          ].      
+000138a0: 2020 2020 2020 666f 7220 6174 7472 2069        for attr i
+000138b0: 6e20 7370 705f 6174 7472 735f 746f 5f63  n spp_attrs_to_c
+000138c0: 6865 636b 3a0a 2020 2020 2020 2020 2020  heck:.          
+000138d0: 2020 2020 2020 6173 7365 7274 206e 702e        assert np.
+000138e0: 616c 6c28 6765 7461 7474 7228 736f 7572  all(getattr(sour
+000138f0: 6365 5f73 7070 2c20 6174 7472 2920 3d3d  ce_spp, attr) ==
+00013900: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00013910: 2020 2020 2020 2020 2020 2020 2020 2067                 g
+00013920: 6574 6174 7472 2873 656c 662c 2061 7474  etattr(self, att
+00013930: 7229 292c 2028 0a20 2020 2020 2020 2020  r)), (.         
+00013940: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013950: 2020 2020 2020 2020 2022 626f 7468 2074           "both t
+00013960: 6865 2073 6f75 7263 6520 616e 6420 7265  he source and re
+00013970: 6369 7069 656e 7420 5370 6563 6965 7320  cipient Species 
+00013980: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
+00013990: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000139a0: 2020 2020 226d 7573 7420 6861 7665 2069      "must have i
+000139b0: 6465 6e74 6963 616c 2076 616c 7565 7320  dentical values 
+000139c0: 666f 7220 220a 2020 2020 2020 2020 2020  for ".          
+000139d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000139e0: 2020 2020 2020 2020 6622 7468 6520 6174          f"the at
+000139f0: 7472 6962 7574 6520 277b 6174 7472 7d27  tribute '{attr}'
+00013a00: 2229 0a0a 2020 2020 2020 2020 2020 2020  ")..            
+00013a10: 2320 6368 6563 6b20 7468 6174 2061 6c6c  # check that all
+00013a20: 2067 656e 5f61 7263 6820 7061 7261 6d73   gen_arch params
+00013a30: 2074 6861 7420 6e65 6564 2074 6f20 6d61   that need to ma
+00013a40: 7463 6820 6172 6520 7468 6520 7361 6d65  tch are the same
+00013a50: 0a20 2020 2020 2020 2020 2020 2023 2069  .            # i
+00013a60: 6e20 7468 6520 736f 7572 6365 2070 6f70  n the source pop
+00013a70: 756c 6174 696f 6e27 7320 5370 6563 6965  ulation's Specie
+00013a80: 7320 616e 6420 7468 6973 2053 7065 6369  s and this Speci
+00013a90: 6573 0a20 2020 2020 2020 2020 2020 2073  es.            s
+00013aa0: 6f75 7263 655f 6765 6e5f 6172 6368 203d  ource_gen_arch =
+00013ab0: 2073 6f75 7263 655f 7370 702e 6765 6e5f   source_spp.gen_
+00013ac0: 6172 6368 0a20 2020 2020 2020 2020 2020  arch.           
+00013ad0: 2073 656c 662e 6765 6e5f 6172 6368 203d   self.gen_arch =
+00013ae0: 2073 656c 662e 6765 6e5f 6172 6368 0a20   self.gen_arch. 
+00013af0: 2020 2020 2020 2020 2020 2061 7373 6572             asser
+00013b00: 7420 736f 7572 6365 5f67 656e 5f61 7263  t source_gen_arc
+00013b10: 682e 4c20 3d3d 2073 656c 662e 6765 6e5f  h.L == self.gen_
+00013b20: 6172 6368 2e4c 2c20 2822 736f 7572 6365  arch.L, ("source
+00013b30: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00013b40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013b50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013b60: 2253 7065 6369 6573 206d 7573 7420 6861  "Species must ha
+00013b70: 7665 2061 2022 0a20 2020 2020 2020 2020  ve a ".         
+00013b80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013b90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013ba0: 2020 2020 2273 696d 756c 6174 6564 2067      "simulated g
+00013bb0: 656e 6f6d 6520 6f66 2074 6865 2073 616d  enome of the sam
+00013bc0: 6520 220a 2020 2020 2020 2020 2020 2020  e ".            
+00013bd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013be0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013bf0: 2022 6c65 6e67 7468 2061 7320 7468 6520   "length as the 
+00013c00: 7265 6370 6965 6e74 2053 7065 6369 6573  recpient Species
+00013c10: 2e22 290a 2020 2020 2020 2020 2020 2020  .").            
+00013c20: 6765 6e5f 6172 6368 5f61 7474 7273 5f74  gen_arch_attrs_t
+00013c30: 6f5f 6368 6563 6b20 3d20 5b27 7365 7827  o_check = ['sex'
+00013c40: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00013c50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013c60: 2020 2020 2020 2020 2027 7573 655f 7473           'use_ts
+00013c70: 6b69 7427 2c0a 2020 2020 2020 2020 2020  kit',.          
+00013c80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013c90: 2020 2020 2020 2020 2020 2020 2027 7827               'x'
+00013ca0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00013cb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00013cc0: 2020 2020 2020 2020 5d0a 2020 2020 2020          ].      
+00013cd0: 2020 2020 2020 2320 4e4f 5445 3a20 6e6f        # NOTE: no
+00013ce0: 7420 7265 7175 6972 696e 6720 7020 746f  t requiring p to
+00013cf0: 2062 6520 7361 6d65 2069 6e20 626f 7468   be same in both
+00013d00: 2070 6f70 756c 6174 696f 6e73 2066 6163   populations fac
+00013d10: 696c 6974 6174 6573 0a20 2020 2020 2020  ilitates.       
+00013d20: 2020 2020 2023 2020 2020 2020 2063 7265       #       cre
+00013d30: 6174 696e 6720 6c6f 6369 2074 6861 7420  ating loci that 
+00013d40: 6469 6666 6572 2069 6e20 7374 616e 6469  differ in standi
+00013d50: 6e67 2067 656e 6574 6963 2076 6172 6961  ng genetic varia
+00013d60: 7469 6f6e 2069 6e0a 2020 2020 2020 2020  tion in.        
+00013d70: 2020 2020 2320 2020 2020 2020 7468 6520      #       the 
+00013d80: 7477 6f20 706f 7075 6c61 7469 6f6e 732c  two populations,
+00013d90: 206f 7220 6576 656e 206c 6f63 6920 7468   or even loci th
+00013da0: 6174 2061 7265 2066 6978 6564 2064 6966  at are fixed dif
+00013db0: 6665 7265 6e74 6c79 0a20 2020 2020 2020  ferently.       
+00013dc0: 2020 2020 2023 2020 2020 2020 2069 6e20       #       in 
+00013dd0: 6f6e 6520 6f72 206d 6f72 6520 706f 7075  one or more popu
+00013de0: 6c61 7469 6f6e 730a 2020 2020 2020 2020  lations.        
+00013df0: 2020 2020 2320 4e4f 5445 3a20 6d61 6b69      # NOTE: maki
+00013e00: 6e67 2069 7420 736f 2074 6861 7420 6174  ng it so that at
+00013e10: 7472 7320 6c69 6b65 206d 7574 2072 6174  trs like mut rat
+00013e20: 652c 2064 6f6d 696e 616e 6365 2070 6172  e, dominance par
+00013e30: 616d 732c 0a20 2020 2020 2020 2020 2020  ams,.           
+00013e40: 2023 2020 2020 2020 2061 6e64 2074 6865   #       and the
+00013e50: 2061 7272 6179 7320 6f66 206e 6575 7472   arrays of neutr
+00013e60: 616c 2061 6e64 2064 656c 6574 206c 6f63  al and delet loc
+00013e70: 6920 6e65 6564 6e27 7420 6265 2074 6865  i needn't be the
+00013e80: 0a20 2020 2020 2020 2020 2020 2023 2020  .            #  
+00013e90: 2020 2020 2073 616d 6520 286d 7574 6174       same (mutat
+00013ea0: 6564 206c 6f63 6920 616c 6d6f 7374 2063  ed loci almost c
+00013eb0: 6572 7461 696e 6c79 2077 6f75 6c64 6e27  ertainly wouldn'
+00013ec0: 7420 6265 0a20 2020 2020 2020 2020 2020  t be.           
+00013ed0: 2023 2020 2020 2020 2069 6620 6e6f 6e2d   #       if non-
+00013ee0: 6e65 7574 7261 6c20 6d75 7461 7469 6f6e  neutral mutation
+00013ef0: 2068 6173 2062 6565 6e20 6861 7070 656e   has been happen
+00013f00: 696e 6720 696e 6465 7065 6e64 656e 746c  ing independentl
+00013f10: 7920 696e 0a20 2020 2020 2020 2020 2020  y in.           
+00013f20: 2023 2020 2020 2020 2062 6f74 6820 706f   #       both po
+00013f30: 7075 6c61 7469 6f6e 7329 2e20 6974 2069  pulations). it i
+00013f40: 736e 2774 2065 6e74 6972 656c 7920 7265  sn't entirely re
+00013f50: 616c 6973 7469 6320 746f 206d 6f64 656c  alistic to model
+00013f60: 2074 6869 730a 2020 2020 2020 2020 2020   this.          
+00013f70: 2020 2320 2020 2020 2020 7468 6973 2077    #       this w
+00013f80: 6179 2c20 6275 7420 616c 736f 206e 6f74  ay, but also not
+00013f90: 2065 6e74 6972 656c 7920 756e 7265 616c   entirely unreal
+00013fa0: 6973 7469 6320 666f 7220 7468 6520 6d6f  istic for the mo
+00013fb0: 6465 6c20 746f 0a20 2020 2020 2020 2020  del to.         
+00013fc0: 2020 2023 2020 2020 2020 2062 6568 6176     #       behav
+00013fd0: 6520 7375 6368 2074 6861 7420 6c6f 6369  e such that loci
+00013fe0: 2074 6861 7420 7765 7265 206e 6575 7420   that were neut 
+00013ff0: 696e 2061 2070 6f70 756c 6174 696f 6e27  in a population'
+00014000: 7320 7072 696f 720a 2020 2020 2020 2020  s prior.        
+00014010: 2020 2020 2320 2020 2020 2020 656e 7669      #       envi
+00014020: 726f 6e6d 656e 7420 6172 6520 7375 6464  ronment are sudd
+00014030: 656e 6c79 206e 6f74 206e 6575 7472 616c  enly not neutral
+00014040: 2069 6e20 7468 6520 6e65 7720 656e 7669   in the new envi
+00014050: 726f 6e6d 656e 740a 2020 2020 2020 2020  ronment.        
+00014060: 2020 2020 666f 7220 6174 7472 2069 6e20      for attr in 
+00014070: 6765 6e5f 6172 6368 5f61 7474 7273 5f74  gen_arch_attrs_t
+00014080: 6f5f 6368 6563 6b3a 0a20 2020 2020 2020  o_check:.       
+00014090: 2020 2020 2020 2020 2061 7373 6572 7420           assert 
+000140a0: 6e70 2e61 6c6c 2867 6574 6174 7472 2873  np.all(getattr(s
+000140b0: 6f75 7263 655f 6765 6e5f 6172 6368 2c20  ource_gen_arch, 
+000140c0: 6174 7472 2920 3d3d 0a20 2020 2020 2020  attr) ==.       
+000140d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000140e0: 2020 2020 2020 2067 6574 6174 7472 2873         getattr(s
+000140f0: 656c 662e 6765 6e5f 6172 6368 2c20 6174  elf.gen_arch, at
+00014100: 7472 2929 2c20 2822 626f 7468 2074 6865  tr)), ("both the
+00014110: 2073 6f75 7263 6522 0a20 2020 2020 2020   source".       
 00014120: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014130: 2020 706c 742e 7465 7874 2878 202d 2073    plt.text(x - s
-00014140: 782c 2079 202b 2073 792c 2058 2e63 6f6c  x, y + sy, X.col
-00014150: 756d 6e73 5b69 5d29 0a20 2020 2020 2020  umns[i]).       
-00014160: 2020 2020 2020 2020 2020 2020 2069 6620               if 
-00014170: 2878 203e 2030 2061 6e64 2079 203e 2030  (x > 0 and y > 0
-00014180: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-00014190: 2020 2020 2020 2020 2020 706c 742e 7465            plt.te
-000141a0: 7874 2878 202b 2073 782c 2079 202b 2073  xt(x + sx, y + s
-000141b0: 792c 2058 2e63 6f6c 756d 6e73 5b69 5d29  y, X.columns[i])
-000141c0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-000141d0: 2020 2020 2069 6620 2878 203e 2030 2061       if (x > 0 a
-000141e0: 6e64 2079 203c 2030 293a 0a20 2020 2020  nd y < 0):.     
-000141f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014200: 2020 706c 742e 7465 7874 2878 202b 2073    plt.text(x + s
-00014210: 782c 2079 202d 2073 792c 2058 2e63 6f6c  x, y - sy, X.col
-00014220: 756d 6e73 5b69 5d29 0a0a 2020 2020 2020  umns[i])..      
-00014230: 2020 2020 2020 2020 2020 2370 6c6f 7420            #plot 
-00014240: 5344 2065 6c6c 6970 7365 2066 6f72 2053  SD ellipse for S
-00014250: 4e50 2064 6174 6120 6966 2070 6c6f 745f  NP data if plot_
-00014260: 7364 203d 2054 7275 650a 2020 2020 2020  sd = True.      
-00014270: 2020 2020 2020 2020 2020 6966 2070 6c6f            if plo
-00014280: 745f 7364 3a0a 2020 2020 2020 2020 2020  t_sd:.          
-00014290: 2020 2020 2020 2020 2020 7231 203d 206e            r1 = n
-000142a0: 702e 7374 6428 6c6f 6369 5f64 665b 7374  p.std(loci_df[st
-000142b0: 7228 6363 5f61 7869 7331 295d 2920 2a20  r(cc_axis1)]) * 
-000142c0: 7364 0a20 2020 2020 2020 2020 2020 2020  sd.             
-000142d0: 2020 2020 2020 2072 3220 3d20 6e70 2e73         r2 = np.s
-000142e0: 7464 286c 6f63 695f 6466 5b73 7472 2863  td(loci_df[str(c
-000142f0: 635f 6178 6973 3229 5d29 202a 2073 640a  c_axis2)]) * sd.
+00014130: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014140: 2020 2020 2020 2020 2020 2020 2022 2061               " a
+00014150: 6e64 2072 6563 6970 6965 6e74 2053 7065  nd recipient Spe
+00014160: 6369 6573 2720 220a 2020 2020 2020 2020  cies' ".        
+00014170: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014180: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014190: 2020 2020 2020 2020 2020 2020 2247 656e              "Gen
+000141a0: 6f6d 6963 4172 6368 6974 6563 7475 7265  omicArchitecture
+000141b0: 7320 220a 2020 2020 2020 2020 2020 2020  s ".            
+000141c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000141d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000141e0: 2020 2020 2020 2020 226f 626a 6563 7473          "objects
+000141f0: 206d 7573 7420 6861 7665 2022 0a20 2020   must have ".   
+00014200: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014210: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014220: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014230: 2022 6964 656e 7469 6361 6c20 7661 6c75   "identical valu
+00014240: 6573 2066 6f72 2074 6865 2022 0a20 2020  es for the ".   
+00014250: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014260: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014270: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014280: 2066 2261 7474 7269 6275 7465 2027 7b61   f"attribute '{a
+00014290: 7474 727d 2722 290a 2020 2020 2020 2020  ttr}'").        
+000142a0: 2020 2020 2320 6368 6563 6b20 7265 636f      # check reco
+000142b0: 6d62 696e 6174 696f 6e20 6174 7472 6962  mbination attrib
+000142c0: 7574 6573 0a20 2020 2020 2020 2020 2020  utes.           
+000142d0: 2072 6563 6f6d 625f 6174 7472 735f 746f   recomb_attrs_to
+000142e0: 5f63 6865 636b 203d 205b 275f 7261 7465  _check = ['_rate
+000142f0: 7327 2c0a 2020 2020 2020 2020 2020 2020  s',.            
 00014300: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014310: 2020 2020 7420 3d20 6e70 2e6c 696e 7370      t = np.linsp
-00014320: 6163 6528 302c 2032 2a6d 6174 685f 7069  ace(0, 2*math_pi
-00014330: 2c20 3130 3029 0a20 2020 2020 2020 2020  , 100).         
-00014340: 2020 2020 2020 2020 2020 2061 782e 706c             ax.pl
-00014350: 6f74 2872 312a 6e70 2e63 6f73 2874 2920  ot(r1*np.cos(t) 
-00014360: 2c20 7232 2a6e 702e 7369 6e28 7429 2c20  , r2*np.sin(t), 
-00014370: 6c69 6e65 7374 796c 6520 3d20 2764 6173  linestyle = 'das
-00014380: 6865 6427 2c0a 2020 2020 2020 2020 2020  hed',.          
-00014390: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000143a0: 2020 6320 3d20 2772 6564 272c 206c 6162    c = 'red', lab
-000143b0: 656c 203d 2073 7472 2873 6429 202b 2022  el = str(sd) + "
-000143c0: 2053 7464 4465 7622 290a 0a20 2020 2020   StdDev")..     
-000143d0: 2020 2020 2020 2020 2020 2023 4556 4552             #EVER
-000143e0: 5954 4849 4e47 2042 454c 4f57 2054 4849  YTHING BELOW THI
-000143f0: 5320 4c49 4e45 2049 4e20 5448 4520 4c4f  S LINE IN THE LO
-00014400: 4f50 2049 5320 4a55 5354 2050 4c4f 5420  OP IS JUST PLOT 
-00014410: 464f 524d 4154 5449 4e47 200a 2020 2020  FORMATTING .    
-00014420: 2020 2020 2020 2020 2020 2020 2378 6c61              #xla
-00014430: 6265 6c20 6973 2074 6865 206c 6162 656c  bel is the label
-00014440: 2061 7420 7468 6520 626f 7474 6f6d 2028   at the bottom (
-00014450: 7665 7274 6963 616c 2061 7869 7320 6c61  vertical axis la
-00014460: 6265 6c29 0a20 2020 2020 2020 2020 2020  bel).           
-00014470: 2020 2020 2023 2061 6e64 2079 6c61 6265       # and ylabe
-00014480: 6c20 6973 206f 6e20 7468 6520 6c65 6674  l is on the left
-00014490: 2073 6964 6520 2868 6f72 697a 6f6e 7461   side (horizonta
-000144a0: 6c20 6178 6973 206c 6162 656c 290a 2020  l axis label).  
-000144b0: 2020 2020 2020 2020 2020 2020 2020 6c61                la
-000144c0: 6265 6c20 3d20 6178 2e73 6574 5f78 6c61  bel = ax.set_xla
-000144d0: 6265 6c28 2743 4341 2720 2b20 7374 7228  bel('CCA' + str(
-000144e0: 6363 5f61 7869 7332 292c 2066 6f6e 7473  cc_axis2), fonts
-000144f0: 697a 6520 3d20 3929 0a20 2020 2020 2020  ize = 9).       
-00014500: 2020 2020 2020 2020 2061 782e 7861 7869           ax.xaxi
-00014510: 732e 7365 745f 6c61 6265 6c5f 636f 6f72  s.set_label_coor
-00014520: 6473 2830 2e35 2c20 2d30 2e30 3229 0a20  ds(0.5, -0.02). 
-00014530: 2020 2020 2020 2020 2020 2020 2020 206c                 l
-00014540: 6162 656c 203d 2061 782e 7365 745f 796c  abel = ax.set_yl
-00014550: 6162 656c 2827 4343 4127 202b 2073 7472  abel('CCA' + str
-00014560: 2863 635f 6178 6973 3129 2c20 666f 6e74  (cc_axis1), font
-00014570: 7369 7a65 203d 2039 290a 2020 2020 2020  size = 9).      
-00014580: 2020 2020 2020 2020 2020 6178 2e79 6178            ax.yax
-00014590: 6973 2e73 6574 5f6c 6162 656c 5f63 6f6f  is.set_label_coo
-000145a0: 7264 7328 2d30 2e30 322c 2030 2e35 3529  rds(-0.02, 0.55)
-000145b0: 0a0a 2020 2020 2020 2020 2020 2020 2020  ..              
-000145c0: 2020 236d 616b 6520 696e 746f 2062 6f78    #make into box
-000145d0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-000145e0: 2061 782e 7365 745f 6173 7065 6374 2827   ax.set_aspect('
-000145f0: 6571 7561 6c27 2c20 6164 6a75 7374 6162  equal', adjustab
-00014600: 6c65 3d27 626f 7827 290a 0a20 2020 2020  le='box')..     
-00014610: 2020 2020 2020 2020 2020 2023 4d6f 7665             #Move
-00014620: 206c 6566 7420 792d 6178 6973 2026 2062   left y-axis & b
-00014630: 6f74 746f 6d20 782d 6178 6973 2074 6f20  ottom x-axis to 
-00014640: 6365 6e74 6572 2c20 7061 7373 696e 6720  center, passing 
-00014650: 7468 7275 2028 302c 3029 0a20 2020 2020  thru (0,0).     
-00014660: 2020 2020 2020 2020 2020 2061 782e 7370             ax.sp
-00014670: 696e 6573 5b27 6c65 6674 275d 2e73 6574  ines['left'].set
-00014680: 5f70 6f73 6974 696f 6e28 2763 656e 7465  _position('cente
-00014690: 7227 290a 2020 2020 2020 2020 2020 2020  r').            
-000146a0: 2020 2020 6178 2e73 7069 6e65 735b 2762      ax.spines['b
-000146b0: 6f74 746f 6d27 5d2e 7365 745f 706f 7369  ottom'].set_posi
-000146c0: 7469 6f6e 2827 6365 6e74 6572 2729 0a0a  tion('center')..
-000146d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000146e0: 2363 6861 6e67 6520 636f 6c6f 7273 2f72  #change colors/r
-000146f0: 656d 6f76 6520 6465 6661 756c 7420 6178  emove default ax
-00014700: 6973 0a20 2020 2020 2020 2020 2020 2020  is.             
-00014710: 2020 2061 7863 6f6c 6f72 203d 2027 6e6f     axcolor = 'no
-00014720: 6e65 270a 2020 2020 2020 2020 2020 2020  ne'.            
-00014730: 2020 2020 6178 2e73 7069 6e65 735b 276c      ax.spines['l
-00014740: 6566 7427 5d2e 7365 745f 636f 6c6f 7228  eft'].set_color(
-00014750: 6178 636f 6c6f 7229 0a20 2020 2020 2020  axcolor).       
-00014760: 2020 2020 2020 2020 2061 782e 7370 696e           ax.spin
-00014770: 6573 5b27 626f 7474 6f6d 275d 2e73 6574  es['bottom'].set
-00014780: 5f63 6f6c 6f72 2861 7863 6f6c 6f72 290a  _color(axcolor).
+00014310: 2020 2020 2020 2020 2027 5f72 5f64 6973           '_r_dis
+00014320: 7472 5f61 6c70 6861 272c 0a20 2020 2020  tr_alpha',.     
+00014330: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014340: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014350: 275f 725f 6469 7374 725f 6265 7461 272c  '_r_distr_beta',
+00014360: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00014370: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014380: 2020 2020 2020 275f 6a69 7474 6572 5f62        '_jitter_b
+00014390: 7265 616b 706f 696e 7473 272c 0a20 2020  reakpoints',.   
+000143a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000143b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000143c0: 205d 0a20 2020 2020 2020 2020 2020 2066   ].            f
+000143d0: 6f72 2061 7474 7220 696e 2072 6563 6f6d  or attr in recom
+000143e0: 625f 6174 7472 735f 746f 5f63 6865 636b  b_attrs_to_check
+000143f0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+00014400: 2020 6173 7365 7274 206e 702e 616c 6c28    assert np.all(
+00014410: 6765 7461 7474 7228 736f 7572 6365 5f67  getattr(source_g
+00014420: 656e 5f61 7263 682e 7265 636f 6d62 696e  en_arch.recombin
+00014430: 6174 696f 6e73 2c20 6174 7472 2920 3d3d  ations, attr) ==
+00014440: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00014450: 2020 2020 2020 2020 2020 2020 2020 2067                 g
+00014460: 6574 6174 7472 2873 656c 662e 6765 6e5f  etattr(self.gen_
+00014470: 6172 6368 2e72 6563 6f6d 6269 6e61 7469  arch.recombinati
+00014480: 6f6e 732c 2061 7474 7229 292c 2028 0a20  ons, attr)), (. 
+00014490: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000144a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000144b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000144c0: 2020 2020 2262 6f74 6820 7468 6520 736f      "both the so
+000144d0: 7572 6365 2022 0a20 2020 2020 2020 2020  urce ".         
+000144e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000144f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014500: 2020 2020 2020 2020 2020 2020 2261 6e64              "and
+00014510: 2072 6563 6970 6965 6e74 2053 7065 6369   recipient Speci
+00014520: 6573 2720 220a 2020 2020 2020 2020 2020  es' ".          
+00014530: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014540: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014550: 2020 2020 2020 2020 2020 2022 5265 636f             "Reco
+00014560: 6d62 696e 6174 696f 6e73 206f 626a 6563  mbinations objec
+00014570: 7473 2022 0a20 2020 2020 2020 2020 2020  ts ".           
+00014580: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014590: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000145a0: 2020 2020 2020 2020 2020 226d 7573 7420            "must 
+000145b0: 6861 7665 2069 6465 6e74 6963 616c 2022  have identical "
+000145c0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000145d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000145e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000145f0: 2020 2020 2020 2276 616c 7565 7320 666f        "values fo
+00014600: 7220 7468 6520 220a 2020 2020 2020 2020  r the ".        
+00014610: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014620: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014630: 2020 2020 2020 2020 2020 2020 2066 2261               f"a
+00014640: 7474 7269 6275 7465 2027 7b61 7474 727d  ttribute '{attr}
+00014650: 2722 290a 0a20 2020 2020 2020 2020 2020  '")..           
+00014660: 2023 2063 6865 636b 2074 7261 6974 7320   # check traits 
+00014670: 616e 6420 7468 6569 7220 7061 7261 6d73  and their params
+00014680: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
+00014690: 7365 6c66 2e67 656e 5f61 7263 682e 7472  self.gen_arch.tr
+000146a0: 6169 7473 2069 7320 4e6f 6e65 3a0a 2020  aits is None:.  
+000146b0: 2020 2020 2020 2020 2020 2020 2020 6173                as
+000146c0: 7365 7274 2073 6f75 7263 655f 6765 6e5f  sert source_gen_
+000146d0: 6172 6368 2e74 7261 6974 7320 6973 204e  arch.traits is N
+000146e0: 6f6e 652c 2028 2249 6e64 6976 6964 7561  one, ("Individua
+000146f0: 6c73 2022 0a20 2020 2020 2020 2020 2020  ls ".           
+00014700: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014710: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014720: 2020 2020 2020 2022 6672 6f6d 2061 2053         "from a S
+00014730: 7065 6369 6573 2077 6974 6820 220a 2020  pecies with ".  
+00014740: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014750: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014760: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014770: 2254 7261 6974 7320 6361 6e6e 6f74 2062  "Traits cannot b
+00014780: 6520 6164 6465 6420 746f 2061 2022 0a20  e added to a ". 
 00014790: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000147a0: 6178 2e73 7069 6e65 735b 2772 6967 6874  ax.spines['right
-000147b0: 275d 2e73 6574 5f63 6f6c 6f72 2861 7863  '].set_color(axc
-000147c0: 6f6c 6f72 290a 2020 2020 2020 2020 2020  olor).          
-000147d0: 2020 2020 2020 6178 2e73 7069 6e65 735b        ax.spines[
-000147e0: 2774 6f70 275d 2e73 6574 5f63 6f6c 6f72  'top'].set_color
-000147f0: 2861 7863 6f6c 6f72 290a 2020 2020 2020  (axcolor).      
-00014800: 2020 2020 2020 2020 2020 6178 2e74 6963            ax.tic
-00014810: 6b5f 7061 7261 6d73 2861 7869 7320 3d20  k_params(axis = 
-00014820: 2778 272c 2063 6f6c 6f72 733d 6178 636f  'x', colors=axco
-00014830: 6c6f 7229 0a20 2020 2020 2020 2020 2020  lor).           
-00014840: 2020 2020 2061 782e 7469 636b 5f70 6172       ax.tick_par
-00014850: 616d 7328 6178 6973 203d 2027 7927 2c20  ams(axis = 'y', 
-00014860: 636f 6c6f 7273 3d61 7863 6f6c 6f72 290a  colors=axcolor).
-00014870: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00014880: 2061 782e 7961 7869 732e 6c61 6265 6c2e   ax.yaxis.label.
-00014890: 7365 745f 636f 6c6f 7228 2767 7261 7927  set_color('gray'
-000148a0: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-000148b0: 2020 6178 2e78 6178 6973 2e6c 6162 656c    ax.xaxis.label
-000148c0: 2e73 6574 5f63 6f6c 6f72 2827 6772 6179  .set_color('gray
-000148d0: 2729 0a0a 2020 2020 2020 2020 2020 2020  ')..            
-000148e0: 2020 2020 236c 6567 656e 6473 0a20 2020      #legends.   
-000148f0: 2020 2020 2020 2020 2020 2020 2069 6620               if 
-00014900: 6e20 3d3d 2032 3a0a 2020 2020 2020 2020  n == 2:.        
-00014910: 2020 2020 2020 2020 2020 2020 6178 2e6c              ax.l
-00014920: 6567 656e 6428 6c6f 6320 3d20 2275 7070  egend(loc = "upp
-00014930: 6572 206c 6566 7422 2c20 6262 6f78 5f74  er left", bbox_t
-00014940: 6f5f 616e 6368 6f72 3d28 312e 332c 2031  o_anchor=(1.3, 1
-00014950: 2e30 3529 290a 2020 2020 2020 2020 2020  .05)).          
-00014960: 2020 2020 2020 2020 2020 706c 742e 636f            plt.co
-00014970: 6c6f 7262 6172 2863 6d61 705f 7363 6174  lorbar(cmap_scat
-00014980: 2c20 6c61 6265 6c3d 2265 6e76 2229 0a20  , label="env"). 
-00014990: 2020 2020 2020 2020 2020 2066 6967 2e73             fig.s
-000149a0: 686f 7728 290a 0a20 2020 2020 2020 2072  how()..        r
-000149b0: 6574 7572 6e20 6363 615f 6469 6374 0a0a  eturn cca_dict..
-000149c0: 0a20 2020 2023 206d 6574 686f 6420 666f  .    # method fo
-000149d0: 7220 706c 6f74 7469 6e67 2074 6865 2073  r plotting the s
-000149e0: 7065 6369 6573 2028 6f72 2061 2073 7562  pecies (or a sub
-000149f0: 7365 7420 6f66 2069 7473 2069 6e64 6976  set of its indiv
-00014a00: 6964 7561 6c73 2c20 6279 2049 4429 0a20  iduals, by ID). 
-00014a10: 2020 2023 206f 6e20 746f 7020 6f66 2061     # on top of a
-00014a20: 206c 6179 6572 2028 6f72 206c 616e 6473   layer (or lands
-00014a30: 6361 7065 290a 2020 2020 6465 6620 5f70  cape).    def _p
-00014a40: 6c6f 7428 7365 6c66 2c20 6c79 725f 6e75  lot(self, lyr_nu
-00014a50: 6d3d 4e6f 6e65 2c20 6c61 6e64 3d4e 6f6e  m=None, land=Non
-00014a60: 652c 2068 6964 655f 6c61 6e64 3d46 616c  e, hide_land=Fal
-00014a70: 7365 2c20 696e 6469 7669 6473 3d4e 6f6e  se, individs=Non
-00014a80: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-00014a90: 2074 6578 743d 4661 6c73 652c 2063 6f6c   text=False, col
-00014aa0: 6f72 3d27 626c 6163 6b27 2c20 6564 6765  or='black', edge
-00014ab0: 5f63 6f6c 6f72 3d27 6661 6365 272c 2074  _color='face', t
-00014ac0: 6578 745f 636f 6c6f 723d 2762 6c61 636b  ext_color='black
-00014ad0: 272c 0a20 2020 2020 2020 2020 2020 2020  ',.             
-00014ae0: 2063 6261 723d 5472 7565 2c20 7369 7a65   cbar=True, size
-00014af0: 3d32 352c 2074 6578 745f 7369 7a65 3d39  =25, text_size=9
-00014b00: 2c20 6c61 6e64 5f63 6d61 703d 4e6f 6e65  , land_cmap=None
-00014b10: 2c20 7074 5f63 6d61 703d 4e6f 6e65 2c0a  , pt_cmap=None,.
-00014b20: 2020 2020 2020 2020 2020 2020 2020 616c                al
-00014b30: 7068 613d 4661 6c73 652c 207a 6f6f 6d5f  pha=False, zoom_
-00014b40: 7769 6474 683d 4e6f 6e65 2c20 783d 4e6f  width=None, x=No
-00014b50: 6e65 2c20 793d 4e6f 6e65 2c20 766d 696e  ne, y=None, vmin
-00014b60: 3d4e 6f6e 652c 0a20 2020 2020 2020 2020  =None,.         
-00014b70: 2020 2020 2076 6d61 783d 4e6f 6e65 2c20       vmax=None, 
-00014b80: 7469 636b 733d 4e6f 6e65 2c20 6d61 736b  ticks=None, mask
-00014b90: 5f72 6173 743d 4e6f 6e65 2c20 616e 696d  _rast=None, anim
-00014ba0: 6174 653d 4661 6c73 652c 0a20 2020 2020  ate=False,.     
-00014bb0: 2020 2020 2020 2020 2063 656c 6c5f 636f           cell_co
-00014bc0: 6f72 6473 3d46 616c 7365 293a 0a20 2020  ords=False):.   
-00014bd0: 2020 2020 2023 2063 6f6e 7665 7274 2069       # convert i
-00014be0: 6e64 6976 6964 7320 746f 2061 206c 6973  ndivids to a lis
-00014bf0: 7420 2869 6e20 6361 7365 2063 6f6d 6573  t (in case comes
-00014c00: 2069 6e20 6173 2061 206e 756d 7079 2061   in as a numpy a
-00014c10: 7272 6179 290a 2020 2020 2020 2020 6966  rray).        if
-00014c20: 2069 6e64 6976 6964 7320 6973 206e 6f74   individs is not
-00014c30: 204e 6f6e 6520 616e 6420 6e6f 7420 6973   None and not is
-00014c40: 696e 7374 616e 6365 2869 6e64 6976 6964  instance(individ
-00014c50: 732c 206c 6973 7429 3a0a 2020 2020 2020  s, list):.      
-00014c60: 2020 2020 2020 696e 6469 7669 6473 203d        individs =
-00014c70: 206c 6973 7428 696e 6469 7669 6473 290a   list(individs).
-00014c80: 2020 2020 2020 2020 2320 6765 7420 636f          # get co
-00014c90: 6f72 6473 0a20 2020 2020 2020 2063 6f6f  ords.        coo
-00014ca0: 7264 7320 3d20 7365 6c66 2e5f 6765 745f  rds = self._get_
-00014cb0: 706c 6f74 5f63 6f6f 7264 7328 696e 6469  plot_coords(indi
-00014cc0: 7669 6473 3d69 6e64 6976 6964 732c 0a20  vids=individs,. 
-00014cd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014ce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014cf0: 2020 2020 2020 6365 6c6c 5f63 6f6f 7264        cell_coord
-00014d00: 733d 6365 6c6c 5f63 6f6f 7264 7329 0a20  s=cell_coords). 
-00014d10: 2020 2020 2020 2023 2067 6574 2074 6578         # get tex
-00014d20: 740a 2020 2020 2020 2020 6966 2074 6578  t.        if tex
-00014d30: 743a 0a20 2020 2020 2020 2020 2020 2069  t:.            i
-00014d40: 6620 696e 6469 7669 6473 2069 7320 4e6f  f individs is No
-00014d50: 6e65 3a0a 2020 2020 2020 2020 2020 2020  ne:.            
-00014d60: 2020 2020 7465 7874 203d 205b 2a73 656c      text = [*sel
-00014d70: 665d 0a20 2020 2020 2020 2020 2020 2065  f].            e
-00014d80: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
-00014d90: 2020 2020 2074 6578 7420 3d20 696e 6469       text = indi
-00014da0: 7669 6473 0a20 2020 2020 2020 2065 6c73  vids.        els
-00014db0: 653a 0a20 2020 2020 2020 2020 2020 2074  e:.            t
-00014dc0: 6578 7420 3d20 4e6f 6e65 0a20 2020 2020  ext = None.     
-00014dd0: 2020 2023 2073 6574 2074 6865 2070 6c74     # set the plt
-00014de0: 5f6c 696d 730a 2020 2020 2020 2020 706c  _lims.        pl
-00014df0: 745f 6c69 6d73 203d 2076 697a 2e5f 6765  t_lims = viz._ge
-00014e00: 745f 706c 745f 6c69 6d73 286c 616e 642c  t_plt_lims(land,
-00014e10: 2078 2c20 792c 207a 6f6f 6d5f 7769 6474   x, y, zoom_widt
-00014e20: 6829 0a20 2020 2020 2020 2023 2070 6c6f  h).        # plo
-00014e30: 7420 7468 6520 6c61 7965 7228 7329 0a20  t the layer(s). 
-00014e40: 2020 2020 2020 2069 6620 6869 6465 5f6c         if hide_l
-00014e50: 616e 643a 0a20 2020 2020 2020 2020 2020  and:.           
-00014e60: 2070 6173 730a 2020 2020 2020 2020 656c   pass.        el
-00014e70: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-00014e80: 2320 6765 7420 7468 6520 6c61 7965 7273  # get the layers
-00014e90: 2720 766d 696e 2061 6e64 2076 6d61 7820  ' vmin and vmax 
-00014ea0: 7661 6c75 6573 2c20 6966 2061 6e79 206f  values, if any o
-00014eb0: 6620 7468 6520 6c61 7965 7273 2074 6f0a  f the layers to.
-00014ec0: 2020 2020 2020 2020 2020 2020 2320 6265              # be
-00014ed0: 2070 6c6f 7474 6564 2068 6173 2061 2063   plotted has a c
-00014ee0: 6861 6e67 6520 6576 656e 740a 2020 2020  hange event.    
-00014ef0: 2020 2020 2020 2020 6966 2028 286c 7972          if ((lyr
-00014f00: 5f6e 756d 2069 7320 4e6f 6e65 2061 6e64  _num is None and
-00014f10: 206c 616e 642e 5f63 6861 6e67 6572 2069   land._changer i
-00014f20: 7320 6e6f 7420 4e6f 6e65 2920 6f72 0a20  s not None) or. 
-00014f30: 2020 2020 2020 2020 2020 2020 2020 2028                 (
-00014f40: 6c61 6e64 2e5f 6368 616e 6765 7220 6973  land._changer is
-00014f50: 206e 6f74 204e 6f6e 650a 2020 2020 2020   not None.      
-00014f60: 2020 2020 2020 2020 2020 2061 6e64 206c             and l
-00014f70: 7972 5f6e 756d 2069 6e20 5b2a 6c61 6e64  yr_num in [*land
-00014f80: 2e5f 6368 616e 6765 722e 6368 616e 6765  ._changer.change
-00014f90: 5f69 6e66 6f5d 2929 3a0a 2020 2020 2020  _info])):.      
-00014fa0: 2020 2020 2020 2020 2020 6966 206c 7972            if lyr
-00014fb0: 5f6e 756d 2069 7320 4e6f 6e65 3a0a 2020  _num is None:.  
+000147a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000147b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000147c0: 2022 5370 6563 6965 7320 7769 7468 6f75   "Species withou
+000147d0: 7420 7468 656d 2e22 290a 2020 2020 2020  t them.").      
+000147e0: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
+000147f0: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+00014800: 7274 2028 6c65 6e28 736f 7572 6365 5f67  rt (len(source_g
+00014810: 656e 5f61 7263 682e 7472 6169 7473 2920  en_arch.traits) 
+00014820: 3d3d 0a20 2020 2020 2020 2020 2020 2020  ==.             
+00014830: 2020 2020 2020 2020 2020 206c 656e 2873             len(s
+00014840: 656c 662e 6765 6e5f 6172 6368 2e74 7261  elf.gen_arch.tra
+00014850: 6974 7329 292c 2028 2254 6865 2073 6f75  its)), ("The sou
+00014860: 7263 6520 706f 7075 6c61 7469 6f6e 2773  rce population's
+00014870: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00014880: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014890: 2020 2020 2020 2020 2020 2022 5370 6563             "Spec
+000148a0: 6965 7320 6f62 6a65 6374 206d 7573 7420  ies object must 
+000148b0: 6861 7665 2074 6865 2022 0a20 2020 2020  have the ".     
+000148c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000148d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000148e0: 2020 2022 7361 6d65 206e 756d 6265 7220     "same number 
+000148f0: 6f66 2054 7261 6974 7320 220a 2020 2020  of Traits ".    
+00014900: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014910: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014920: 2020 2020 2261 7320 7468 6520 7265 6369      "as the reci
+00014930: 7069 656e 7420 5370 6563 6965 732e 2229  pient Species.")
+00014940: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00014950: 2074 7274 5f61 7474 7273 5f74 6f5f 6368   trt_attrs_to_ch
+00014960: 6563 6b20 3d20 5b27 6e61 6d65 272c 0a20  eck = ['name',. 
+00014970: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014980: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014990: 2020 2020 2027 7068 6927 2c0a 2020 2020       'phi',.    
+000149a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000149b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000149c0: 2020 276c 7972 5f6e 756d 272c 0a20 2020    'lyr_num',.   
+000149d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000149e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000149f0: 2020 2027 6d61 785f 616c 7068 615f 6d61     'max_alpha_ma
+00014a00: 6727 2c0a 2020 2020 2020 2020 2020 2020  g',.            
+00014a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014a20: 2020 2020 2020 2020 2020 2767 616d 6d61            'gamma
+00014a30: 272c 0a20 2020 2020 2020 2020 2020 2020  ',.             
+00014a40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014a50: 2020 2020 2020 2020 2027 756e 6976 5f61           'univ_a
+00014a60: 6476 272c 0a20 2020 2020 2020 2020 2020  dv',.           
+00014a70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014a80: 2020 2020 2020 2020 2020 5d0a 2020 2020            ].    
+00014a90: 2020 2020 2020 2020 2020 2020 2320 4e4f              # NO
+00014aa0: 5445 3a20 6e6f 7420 7265 7175 6972 696e  TE: not requirin
+00014ab0: 6720 276e 5f6c 6f63 6927 2074 6f20 6265  g 'n_loci' to be
+00014ac0: 2074 6865 2073 616d 6520 6265 6361 7573   the same becaus
+00014ad0: 6520 7468 6579 2772 650a 2020 2020 2020  e they're.      
+00014ae0: 2020 2020 2020 2020 2020 2320 2020 2020            #     
+00014af0: 2020 756e 6c69 6b65 6c79 2074 6f20 6265    unlikely to be
+00014b00: 2074 6865 2073 616d 6520 6966 206e 6f6e   the same if non
+00014b10: 2d6e 6575 7472 616c 206d 7574 6174 696f  -neutral mutatio
+00014b20: 6e20 0a20 2020 2020 2020 2020 2020 2020  n .             
+00014b30: 2020 2023 2020 2020 2020 2068 6173 2062     #       has b
+00014b40: 6565 6e20 6861 7070 656e 696e 6720 696e  een happening in
+00014b50: 6465 7065 6e64 656e 746c 7920 696e 2062  dependently in b
+00014b60: 6f74 6820 706f 7075 6c61 7469 6f6e 730a  oth populations.
+00014b70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014b80: 666f 7220 6e2c 2074 7274 2069 6e20 7365  for n, trt in se
+00014b90: 6c66 2e67 656e 5f61 7263 682e 7472 6169  lf.gen_arch.trai
+00014ba0: 7473 2e69 7465 6d73 2829 3a0a 2020 2020  ts.items():.    
+00014bb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014bc0: 666f 7220 6174 7472 2069 6e20 7472 745f  for attr in trt_
+00014bd0: 6174 7473 5f74 6f5f 6368 6563 6b3a 0a20  atts_to_check:. 
+00014be0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014bf0: 2020 2020 2020 2061 7373 6572 7420 6e70         assert np
+00014c00: 2e61 6c6c 2867 6574 6174 7472 2874 7274  .all(getattr(trt
+00014c10: 2c20 6174 7472 2920 3d3d 0a20 2020 2020  , attr) ==.     
+00014c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014c30: 2020 2020 2020 2020 2020 2067 6574 6174             getat
+00014c40: 7472 2873 6f75 7263 655f 6765 6e5f 6172  tr(source_gen_ar
+00014c50: 6368 2e74 7261 6974 735b 6e5d 2c20 6174  ch.traits[n], at
+00014c60: 7472 2929 2c20 280a 2020 2020 2020 2020  tr)), (.        
+00014c70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014c80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014c90: 2066 2241 7474 7269 6275 7465 2027 7b61   f"Attribute '{a
+00014ca0: 7474 727d 2720 6f66 2074 7261 6974 207b  ttr}' of trait {
+00014cb0: 6e7d 2022 0a20 2020 2020 2020 2020 2020  n} ".           
+00014cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014cd0: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+00014ce0: 696e 2074 6865 2073 6f75 7263 6520 706f  in the source po
+00014cf0: 7075 6c61 7469 6f6e 2773 2053 7065 6369  pulation's Speci
+00014d00: 6573 2022 0a20 2020 2020 2020 2020 2020  es ".           
+00014d10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014d20: 2020 2020 2020 2020 2020 2020 2020 2022                 "
+00014d30: 6f62 6a65 6374 206d 7573 7420 6265 2074  object must be t
+00014d40: 6865 2073 616d 6520 6173 2022 0a20 2020  he same as ".   
+00014d50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014d60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014d70: 2020 2020 2020 2022 7468 6174 2061 7474         "that att
+00014d80: 7269 6275 7465 2069 6e20 7468 6520 7265  ribute in the re
+00014d90: 6369 7069 656e 7420 220a 2020 2020 2020  cipient ".      
+00014da0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014db0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014dc0: 2020 2020 2270 6f70 756c 6174 696f 6e27      "population'
+00014dd0: 7320 5370 6563 6965 732e 2229 0a0a 2020  s Species.")..  
+00014de0: 2020 2020 2020 2020 2020 2320 6465 6570            # deep
+00014df0: 636f 7079 2074 6865 2073 6f75 7263 655f  copy the source_
+00014e00: 7370 7020 6f62 6a65 6374 2c20 7468 656e  spp object, then
+00014e10: 2073 7562 7365 7420 746f 206f 6e6c 7920   subset to only 
+00014e20: 7468 6520 736f 7572 6365 0a20 2020 2020  the source.     
+00014e30: 2020 2020 2020 2023 2069 6e64 6976 6964         # individ
+00014e40: 7561 6c73 206e 6565 6465 642c 2077 6869  uals needed, whi
+00014e50: 6368 2077 696c 6c20 616c 736f 2073 6f72  ch will also sor
+00014e60: 7420 616e 6420 7369 6d70 6c69 6679 2074  t and simplify t
+00014e70: 6865 0a20 2020 2020 2020 2020 2020 2023  he.            #
+00014e80: 2054 6162 6c65 436f 6c6c 6563 7469 6f6e   TableCollection
+00014e90: 2073 6f20 7468 6174 2061 6c6c 2065 7874   so that all ext
+00014ea0: 616e 7420 6e6f 6465 7320 6172 6520 6e6f  ant nodes are no
+00014eb0: 6465 7320 746f 2062 6520 756e 696f 6e65  des to be unione
+00014ec0: 640a 2020 2020 2020 2020 2020 2020 2320  d.            # 
+00014ed0: 696e 746f 2074 6869 7320 5370 6563 6965  into this Specie
+00014ee0: 7327 2054 6162 6c65 436f 6c6c 6563 7469  s' TableCollecti
+00014ef0: 6f6e 200a 2020 2020 2020 2020 2020 2020  on .            
+00014f00: 736f 7572 6365 5f73 7070 5f63 6f70 7920  source_spp_copy 
+00014f10: 3d20 6465 6570 636f 7079 2873 6f75 7263  = deepcopy(sourc
+00014f20: 655f 7370 7029 0a20 2020 2020 2020 2020  e_spp).         
+00014f30: 2020 2023 2067 6574 2069 6e64 6976 6964     # get individ
+00014f40: 7561 6c73 2066 726f 6d20 7468 6520 6769  uals from the gi
+00014f50: 7665 6e20 5370 6563 6965 7320 6f62 6a65  ven Species obje
+00014f60: 6374 0a20 2020 2020 2020 2020 2020 2069  ct.            i
+00014f70: 6620 696e 6469 7669 6473 2069 7320 6e6f  f individs is no
+00014f80: 7420 4e6f 6e65 3a0a 2020 2020 2020 2020  t None:.        
+00014f90: 2020 2020 2020 2020 696e 6469 7669 6473          individs
+00014fa0: 5f74 6f5f 7265 6d6f 7665 203d 205b 696e  _to_remove = [in
+00014fb0: 6420 666f 7220 696e 6420 696e 0a20 2020  d for ind in.   
 00014fc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014fd0: 2020 6c61 6e64 5f76 6d69 6e20 3d20 5b6c    land_vmin = [l
-00014fe0: 7972 2e5f 7363 616c 655f 6d69 6e20 666f  yr._scale_min fo
-00014ff0: 7220 6c79 7220 696e 206c 616e 642e 7661  r lyr in land.va
-00015000: 6c75 6573 2829 5d0a 2020 2020 2020 2020  lues()].        
-00015010: 2020 2020 2020 2020 2020 2020 6c61 6e64              land
-00015020: 5f76 6d61 7820 3d20 5b6c 7972 2e5f 7363  _vmax = [lyr._sc
-00015030: 616c 655f 6d61 7820 666f 7220 6c79 7220  ale_max for lyr 
-00015040: 696e 206c 616e 642e 7661 6c75 6573 2829  in land.values()
-00015050: 5d0a 2020 2020 2020 2020 2020 2020 2020  ].              
-00015060: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
-00015070: 2020 2020 2020 2020 2020 2020 6c61 6e64              land
-00015080: 5f76 6d69 6e20 3d20 5b6c 616e 645b 6c79  _vmin = [land[ly
-00015090: 725f 6e75 6d5d 2e5f 7363 616c 655f 6d69  r_num]._scale_mi
-000150a0: 6e5d 0a20 2020 2020 2020 2020 2020 2020  n].             
-000150b0: 2020 2020 2020 206c 616e 645f 766d 6178         land_vmax
-000150c0: 203d 205b 6c61 6e64 5b6c 7972 5f6e 756d   = [land[lyr_num
-000150d0: 5d2e 5f73 6361 6c65 5f6d 6178 5d0a 2020  ]._scale_max].  
-000150e0: 2020 2020 2020 2020 2020 656c 7365 3a0a            else:.
-000150f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015100: 6c61 6e64 5f76 6d69 6e20 3d20 6c61 6e64  land_vmin = land
-00015110: 5f76 6d61 7820 3d20 4e6f 6e65 0a20 2020  _vmax = None.   
-00015120: 2020 2020 2020 2020 2076 697a 2e5f 706c           viz._pl
-00015130: 6f74 5f72 6173 7465 7273 286c 616e 642c  ot_rasters(land,
-00015140: 206c 7972 5f6e 756d 3d6c 7972 5f6e 756d   lyr_num=lyr_num
-00015150: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00015160: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015170: 6362 6172 3d63 6261 722c 2063 6d61 703d  cbar=cbar, cmap=
-00015180: 6c61 6e64 5f63 6d61 702c 2070 6c74 5f6c  land_cmap, plt_l
-00015190: 696d 733d 706c 745f 6c69 6d73 2c0a 2020  ims=plt_lims,.  
-000151a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000151b0: 2020 2020 2020 2020 2020 2020 7469 636b              tick
-000151c0: 733d 7469 636b 732c 206d 6173 6b5f 7261  s=ticks, mask_ra
-000151d0: 7374 3d6d 6173 6b5f 7261 7374 2c0a 2020  st=mask_rast,.  
-000151e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000151f0: 2020 2020 2020 2020 2020 2020 766d 696e              vmin
-00015200: 3d6c 616e 645f 766d 696e 2c20 766d 6178  =land_vmin, vmax
-00015210: 3d6c 616e 645f 766d 6178 290a 2020 2020  =land_vmax).    
-00015220: 2020 2020 2320 616e 6420 706c 6f74 2074      # and plot t
-00015230: 6865 2069 6e64 6976 6964 7561 6c73 0a20  he individuals. 
-00015240: 2020 2020 2020 2070 6f69 6e74 7320 3d20         points = 
-00015250: 7669 7a2e 5f70 6c6f 745f 706f 696e 7473  viz._plot_points
-00015260: 2863 6f6f 7264 732c 206c 7972 5f6e 756d  (coords, lyr_num
-00015270: 3d6c 7972 5f6e 756d 2c20 636f 6c6f 723d  =lyr_num, color=
-00015280: 636f 6c6f 722c 0a20 2020 2020 2020 2020  color,.         
-00015290: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000152a0: 2020 2020 2020 2020 2065 6467 655f 636f           edge_co
-000152b0: 6c6f 723d 6564 6765 5f63 6f6c 6f72 2c0a  lor=edge_color,.
-000152c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000152d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000152e0: 2020 7465 7874 5f63 6f6c 6f72 3d74 6578    text_color=tex
-000152f0: 745f 636f 6c6f 722c 2073 697a 653d 7369  t_color, size=si
-00015300: 7a65 2c0a 2020 2020 2020 2020 2020 2020  ze,.            
-00015310: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015320: 2020 2020 2020 7465 7874 5f73 697a 653d        text_size=
-00015330: 7465 7874 5f73 697a 652c 2061 6c70 6861  text_size, alpha
-00015340: 3d61 6c70 6861 2c0a 2020 2020 2020 2020  =alpha,.        
-00015350: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015360: 2020 2020 2020 2020 2020 7465 7874 3d74            text=t
-00015370: 6578 742c 2070 6c74 5f6c 696d 733d 706c  ext, plt_lims=pl
-00015380: 745f 6c69 6d73 2c0a 2020 2020 2020 2020  t_lims,.        
+00014fd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014fe0: 2020 2073 6f75 7263 655f 7370 705f 636f     source_spp_co
+00014ff0: 7079 2069 6620 696e 6420 6e6f 7420 696e  py if ind not in
+00015000: 2069 6e64 6976 6964 735d 0a20 2020 2020   individs].     
+00015010: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
+00015020: 2020 2020 2020 2020 2020 2020 2069 6e64               ind
+00015030: 6976 6964 735f 746f 5f72 656d 6f76 6520  ivids_to_remove 
+00015040: 3d20 5b2a 736f 7572 6365 5f73 7070 5f63  = [*source_spp_c
+00015050: 6f70 795d 5b6e 3a5d 0a20 2020 2020 2020  opy][n:].       
+00015060: 2020 2020 2073 6f75 7263 655f 7370 705f       source_spp_
+00015070: 636f 7079 2e5f 7265 6d6f 7665 5f69 6e64  copy._remove_ind
+00015080: 6976 6964 7561 6c73 2869 6e64 6976 6964  ividuals(individ
+00015090: 733d 696e 6469 7669 6473 5f74 6f5f 7265  s=individs_to_re
+000150a0: 6d6f 7665 290a 2020 2020 2020 2020 2020  move).          
+000150b0: 2020 6966 2069 6e64 6976 6964 7320 6973    if individs is
+000150c0: 206e 6f74 204e 6f6e 653a 0a20 2020 2020   not None:.     
+000150d0: 2020 2020 2020 2020 2020 2061 7373 6572             asser
+000150e0: 7420 6c65 6e28 736f 7572 6365 5f73 7070  t len(source_spp
+000150f0: 5f63 6f70 7929 203d 3d20 6c65 6e28 696e  _copy) == len(in
+00015100: 6469 7669 6473 290a 2020 2020 2020 2020  divids).        
+00015110: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+00015120: 2020 2020 2020 2020 2020 6173 7365 7274            assert
+00015130: 206c 656e 2873 6f75 7263 655f 7370 705f   len(source_spp_
+00015140: 636f 7079 2920 3d3d 206e 0a0a 2020 2020  copy) == n..    
+00015150: 2020 2020 2020 2020 2320 6761 7468 6572          # gather
+00015160: 2061 6c6c 2072 656d 6169 6e69 6e67 2049   all remaining I
+00015170: 6e64 6976 6964 7561 6c73 2069 6e74 6f20  ndividuals into 
+00015180: 6120 6c69 7374 0a20 2020 2020 2020 2020  a list.         
+00015190: 2020 2069 6e64 6976 6964 7320 3d20 5b2a     individs = [*
+000151a0: 736f 7572 6365 5f73 7070 5f63 6f70 792e  source_spp_copy.
+000151b0: 7661 6c75 6573 2829 5d0a 0a20 2020 2020  values()]..     
+000151c0: 2020 2020 2020 2023 2064 6f75 626c 652d         # double-
+000151d0: 6368 6563 6b20 7468 6174 2074 6865 2073  check that the s
+000151e0: 6f75 7263 655f 7370 705f 636f 7079 2773  ource_spp_copy's
+000151f0: 2054 6162 6c65 436f 6c6c 6563 7469 6f6e   TableCollection
+00015200: 2068 6173 2062 6565 6e0a 2020 2020 2020   has been.      
+00015210: 2020 2020 2020 2320 736f 7274 6564 2061        # sorted a
+00015220: 6e64 2073 696d 706c 6966 6965 640a 2020  nd simplified.  
+00015230: 2020 2020 2020 2020 2020 736f 7572 6365            source
+00015240: 5f73 7070 5f63 6f70 792e 5f73 6f72 745f  _spp_copy._sort_
+00015250: 616e 645f 7369 6d70 6c69 6679 5f74 6162  and_simplify_tab
+00015260: 6c65 5f63 6f6c 6c65 6374 696f 6e28 290a  le_collection().
+00015270: 0a20 2020 2020 2020 2020 2020 2023 2066  .            # f
+00015280: 6f72 6d61 7420 5461 626c 6543 6f6c 6c65  ormat TableColle
+00015290: 6374 696f 6e20 616e 6420 7265 7475 726e  ction and return
+000152a0: 2074 6865 206c 6973 7420 6f66 206e 6577   the list of new
+000152b0: 2067 6e78 2049 6e64 6976 6964 2069 6478   gnx Individ idx
+000152c0: 730a 2020 2020 2020 2020 2020 2020 2320  s.            # 
+000152d0: 7468 6174 2054 6162 6c65 2773 2069 6e64  that Table's ind
+000152e0: 6976 6964 7561 6c73 2077 696c 6c20 6265  ividuals will be
+000152f0: 2061 7373 6967 6e65 6420 772f 696e 2074   assigned w/in t
+00015300: 6869 7320 5370 6563 6965 730a 2020 2020  his Species.    
+00015310: 2020 2020 2020 2020 736f 7572 6365 5f74          source_t
+00015320: 6320 3d20 736f 7572 6365 5f73 7070 5f63  c = source_spp_c
+00015330: 6f70 792e 5f74 630a 2020 2020 2020 2020  opy._tc.        
+00015340: 2020 2020 6e65 7874 5f67 6e78 5f69 6478      next_gnx_idx
+00015350: 7320 3d20 5f70 7265 705f 7473 6b69 745f  s = _prep_tskit_
+00015360: 7461 6263 6f6c 6c5f 666f 725f 676e 785f  tabcoll_for_gnx_
+00015370: 7370 705f 756e 696f 6e28 7463 3d73 6f75  spp_union(tc=sou
+00015380: 7263 655f 7463 2c0a 2020 2020 2020 2020  rce_tc,.        
 00015390: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000153a0: 2020 2020 2020 2020 2020 7074 5f63 6d61            pt_cma
-000153b0: 703d 7074 5f63 6d61 702c 2076 6d69 6e3d  p=pt_cmap, vmin=
-000153c0: 766d 696e 2c20 766d 6178 3d76 6d61 782c  vmin, vmax=vmax,
-000153d0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000153a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000153b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000153c0: 7265 6369 705f 7370 703d 7365 6c66 2c0a  recip_spp=self,.
+000153d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 000153e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000153f0: 2020 2061 6e69 6d61 7465 3d61 6e69 6d61     animate=anima
-00015400: 7465 290a 2020 2020 2020 2020 7265 7475  te).        retu
-00015410: 726e 2070 6f69 6e74 730a 0a0a 2020 2020  rn points...    
-00015420: 236d 6574 686f 6420 666f 7220 706c 6f74  #method for plot
-00015430: 7469 6e67 2074 6865 2073 7065 6369 6573  ting the species
-00015440: 206f 6e20 746f 7020 6f66 2069 7473 2065   on top of its e
-00015450: 7374 696d 6174 6564 0a20 2020 2023 7370  stimated.    #sp
-00015460: 6563 6965 732d 6465 6e73 6974 7920 7261  ecies-density ra
-00015470: 7374 6572 0a20 2020 2064 6566 205f 706c  ster.    def _pl
-00015480: 6f74 5f64 656e 7369 7479 2873 656c 662c  ot_density(self,
-00015490: 206c 616e 642c 206e 6f72 6d61 6c69 7a65   land, normalize
-000154a0: 3d46 616c 7365 2c20 696e 6469 7669 6473  =False, individs
-000154b0: 3d4e 6f6e 652c 0a20 2020 2020 2020 2020  =None,.         
-000154c0: 2020 2074 6578 743d 4661 6c73 652c 2063     text=False, c
-000154d0: 6f6c 6f72 3d27 626c 6163 6b27 2c20 6564  olor='black', ed
-000154e0: 6765 5f63 6f6c 6f72 3d27 6661 6365 272c  ge_color='face',
-000154f0: 0a20 2020 2020 2020 2020 2020 2074 6578  .            tex
-00015500: 745f 636f 6c6f 723d 2762 6c61 636b 272c  t_color='black',
-00015510: 2073 697a 653d 3235 2c20 7465 7874 5f73   size=25, text_s
-00015520: 697a 6520 3d20 392c 0a20 2020 2020 2020  ize = 9,.       
-00015530: 2020 2020 2061 6c70 6861 3d30 2e35 2c20       alpha=0.5, 
-00015540: 7a6f 6f6d 5f77 6964 7468 3d4e 6f6e 652c  zoom_width=None,
-00015550: 2078 3d4e 6f6e 652c 2079 3d4e 6f6e 652c   x=None, y=None,
-00015560: 2074 6963 6b73 3d4e 6f6e 652c 0a20 2020   ticks=None,.   
-00015570: 2020 2020 2020 2020 206d 6173 6b5f 7261           mask_ra
-00015580: 7374 3d4e 6f6e 6529 3a0a 2020 2020 2020  st=None):.      
-00015590: 2020 6173 7365 7274 2074 7970 6528 6e6f    assert type(no
-000155a0: 726d 616c 697a 6529 2069 7320 626f 6f6c  rmalize) is bool
-000155b0: 2c20 2822 5468 6520 276e 6f72 6d61 6c69  , ("The 'normali
-000155c0: 7a65 2720 6172 6775 6d65 6e74 2074 616b  ze' argument tak
-000155d0: 6573 2022 0a20 2020 2020 2020 2020 2020  es ".           
-000155e0: 2022 6120 626f 6f6c 6561 6e20 7661 6c75   "a boolean valu
-000155f0: 652e 5c6e 2229 0a20 2020 2020 2020 2023  e.\n").        #
-00015600: 7570 6461 7465 2074 6865 2073 7065 6369  update the speci
-00015610: 6573 2720 636f 6f72 6469 6e61 7465 7320  es' coordinates 
-00015620: 616e 6420 6365 6c6c 732c 2069 6e20 6361  and cells, in ca
-00015630: 7365 2069 7420 6861 736e 2774 0a20 2020  se it hasn't.   
-00015640: 2020 2020 2023 6265 656e 2075 7064 6174       #been updat
-00015650: 6520 7369 6e63 6520 736f 6d65 2069 6e74  e since some int
-00015660: 6572 6e61 6c20 6f72 206d 616e 7561 6c20  ernal or manual 
-00015670: 6368 616e 6765 7320 696e 2070 6f70 756c  changes in popul
-00015680: 6174 696f 6e2d 7369 7a65 0a20 2020 2020  ation-size.     
-00015690: 2020 2023 6861 7665 206f 6363 7572 7265     #have occurre
-000156a0: 640a 2020 2020 2020 2020 7365 6c66 2e5f  d.        self._
-000156b0: 7365 745f 636f 6f72 6473 5f61 6e64 5f63  set_coords_and_c
-000156c0: 656c 6c73 2829 0a20 2020 2020 2020 2064  ells().        d
-000156d0: 656e 7320 3d20 7365 6c66 2e5f 6361 6c63  ens = self._calc
-000156e0: 5f64 656e 7369 7479 286e 6f72 6d61 6c69  _density(normali
-000156f0: 7a65 203d 206e 6f72 6d61 6c69 7a65 290a  ze = normalize).
-00015700: 2020 2020 2020 2020 706c 745f 6c69 6d73          plt_lims
-00015710: 203d 2076 697a 2e5f 6765 745f 706c 745f   = viz._get_plt_
-00015720: 6c69 6d73 286c 616e 642c 2078 2c20 792c  lims(land, x, y,
-00015730: 207a 6f6f 6d5f 7769 6474 6829 0a20 2020   zoom_width).   
-00015740: 2020 2020 2069 6620 6e6f 726d 616c 697a       if normaliz
-00015750: 653a 0a20 2020 2020 2020 2020 2020 2076  e:.            v
-00015760: 697a 2e5f 706c 6f74 5f72 6173 7465 7273  iz._plot_rasters
-00015770: 2864 656e 732c 2070 6c74 5f6c 696d 7320  (dens, plt_lims 
-00015780: 3d20 706c 745f 6c69 6d73 2c20 6c79 725f  = plt_lims, lyr_
-00015790: 6e61 6d65 203d 2027 6465 6e73 6974 7927  name = 'density'
-000157a0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-000157b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000157c0: 7469 636b 733d 7469 636b 732c 206d 6173  ticks=ticks, mas
-000157d0: 6b5f 7261 7374 3d6d 6173 6b5f 7261 7374  k_rast=mask_rast
-000157e0: 290a 2020 2020 2020 2020 656c 7365 3a0a  ).        else:.
-000157f0: 2020 2020 2020 2020 2020 2020 7669 7a2e              viz.
-00015800: 5f70 6c6f 745f 7261 7374 6572 7328 6465  _plot_rasters(de
-00015810: 6e73 2c20 706c 745f 6c69 6d73 203d 2070  ns, plt_lims = p
-00015820: 6c74 5f6c 696d 732c 2076 6d61 7820 3d20  lt_lims, vmax = 
-00015830: 6465 6e73 2e6d 6178 2829 2c0a 2020 2020  dens.max(),.    
-00015840: 2020 2020 2020 2020 2020 2020 6c79 725f              lyr_
-00015850: 6e61 6d65 203d 2027 6465 6e73 6974 7927  name = 'density'
-00015860: 2c20 7469 636b 733d 7469 636b 732c 206d  , ticks=ticks, m
-00015870: 6173 6b5f 7261 7374 3d6d 6173 6b5f 7261  ask_rast=mask_ra
-00015880: 7374 290a 2020 2020 2020 2020 7365 6c66  st).        self
-00015890: 2e5f 706c 6f74 2868 6964 655f 6c61 6e64  ._plot(hide_land
-000158a0: 3d54 7275 652c 2069 6e64 6976 6964 7320  =True, individs 
-000158b0: 3d20 696e 6469 7669 6473 2c20 7465 7874  = individs, text
-000158c0: 203d 2074 6578 742c 0a20 2020 2020 2020   = text,.       
-000158d0: 2020 2020 2063 6f6c 6f72 3d63 6f6c 6f72       color=color
-000158e0: 2c20 6564 6765 5f63 6f6c 6f72 203d 2065  , edge_color = e
-000158f0: 6467 655f 636f 6c6f 722c 2074 6578 745f  dge_color, text_
-00015900: 636f 6c6f 7220 3d20 7465 7874 5f63 6f6c  color = text_col
-00015910: 6f72 2c0a 2020 2020 2020 2020 2020 2020  or,.            
-00015920: 7369 7a65 3d73 697a 652c 2074 6578 745f  size=size, text_
-00015930: 7369 7a65 203d 2074 6578 745f 7369 7a65  size = text_size
-00015940: 2c20 616c 7068 613d 616c 7068 612c 0a20  , alpha=alpha,. 
-00015950: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015960: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015970: 2020 207a 6f6f 6d5f 7769 6474 6820 3d20     zoom_width = 
-00015980: 7a6f 6f6d 5f77 6964 7468 2c20 7820 3d20  zoom_width, x = 
-00015990: 782c 2079 203d 2079 290a 0a0a 2020 2020  x, y = y)...    
-000159a0: 2320 6d65 7468 6f64 2066 6f72 2070 6c6f  # method for plo
-000159b0: 7474 696e 6720 696e 6469 7669 6475 616c  tting individual
-000159c0: 7320 636f 6c6f 7265 6420 6279 2074 6865  s colored by the
-000159d0: 6972 2067 656e 6f74 7970 6520 6174 2061  ir genotype at a
-000159e0: 206c 6f63 7573 0a20 2020 2064 6566 205f   locus.    def _
-000159f0: 706c 6f74 5f67 656e 6f74 7970 6528 7365  plot_genotype(se
-00015a00: 6c66 2c20 6c6f 6375 732c 206c 7972 5f6e  lf, locus, lyr_n
-00015a10: 756d 3d4e 6f6e 652c 206c 616e 643d 4e6f  um=None, land=No
-00015a20: 6e65 2c20 696e 6469 7669 6473 3d4e 6f6e  ne, individs=Non
-00015a30: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-00015a40: 2020 2020 2020 2020 2020 7465 7874 3d46            text=F
-00015a50: 616c 7365 2c20 7369 7a65 3d32 352c 2074  alse, size=25, t
-00015a60: 6578 745f 7369 7a65 203d 2039 2c20 6564  ext_size = 9, ed
-00015a70: 6765 5f63 6f6c 6f72 3d27 626c 6163 6b27  ge_color='black'
-00015a80: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00015a90: 2020 2020 2020 2020 2074 6578 745f 636f           text_co
-00015aa0: 6c6f 723d 2762 6c61 636b 272c 2063 6261  lor='black', cba
-00015ab0: 723d 5472 7565 2c20 616c 7068 613d 312c  r=True, alpha=1,
-00015ac0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00015ad0: 2020 2020 2020 2020 6279 5f64 6f6d 696e          by_domin
-00015ae0: 616e 6365 3d46 616c 7365 2c20 7a6f 6f6d  ance=False, zoom
-00015af0: 5f77 6964 7468 3d4e 6f6e 652c 2078 3d4e  _width=None, x=N
-00015b00: 6f6e 652c 2079 3d4e 6f6e 652c 0a20 2020  one, y=None,.   
-00015b10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015b20: 2020 2020 7469 636b 733d 4e6f 6e65 2c20      ticks=None, 
-00015b30: 6d61 736b 5f72 6173 743d 4e6f 6e65 293a  mask_rast=None):
-00015b40: 0a0a 2020 2020 2020 2020 6765 6e6f 7479  ..        genoty
-00015b50: 7065 7320 3d20 7365 6c66 2e5f 6765 745f  pes = self._get_
-00015b60: 6765 6e6f 7479 7065 7328 6c6f 6369 3d5b  genotypes(loci=[
-00015b70: 6c6f 6375 735d 2c20 696e 6469 7669 6473  locus], individs
-00015b80: 3d69 6e64 6976 6964 732c 0a20 2020 2020  =individs,.     
+000153f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015400: 2020 2020 2020 2020 636f 6f72 6473 3d63          coords=c
+00015410: 6f6f 7264 732c 0a20 2020 2020 2020 2020  oords,.         
+00015420: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015430: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015440: 2020 2020 2020 2020 2020 2020 2020 2073                 s
+00015450: 6f75 7263 655f 736f 6674 7761 7265 3d27  ource_software='
+00015460: 676e 7827 2c0a 2020 2020 2020 2020 2020  gnx',.          
+00015470: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015480: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015490: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000154a0: 2020 2020 2020 2029 0a0a 2020 2020 2020         )..      
+000154b0: 2020 2320 2e2e 2e20 6f72 2073 696d 756c    # ... or simul
+000154c0: 6174 6520 7468 656d 2077 2f20 7468 6520  ate them w/ the 
+000154d0: 6769 7665 6e20 6d73 7072 696d 6520 6172  given msprime ar
+000154e0: 6773 0a20 2020 2020 2020 2023 2061 6e64  gs.        # and
+000154f0: 2074 6865 2074 6869 7320 5370 6563 6965   the this Specie
+00015500: 7327 2067 656e 6f6d 6963 2061 7263 6869  s' genomic archi
+00015510: 7465 6374 7572 650a 2020 2020 2020 2020  tecture.        
+00015520: 656c 6966 2073 6f75 7263 655f 7370 7020  elif source_spp 
+00015530: 6973 204e 6f6e 653a 0a20 2020 2020 2020  is None:.       
+00015540: 2020 2020 2061 7373 6572 7420 696e 6469       assert indi
+00015550: 7669 6473 2069 7320 4e6f 6e65 2c20 2822  vids is None, ("
+00015560: 5468 6520 2769 6e64 6976 6964 7327 2061  The 'individs' a
+00015570: 7267 756d 656e 7420 696e 6469 6361 7465  rgument indicate
+00015580: 7320 7468 6520 220a 2020 2020 2020 2020  s the ".        
+00015590: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000155a0: 2020 2020 2020 2020 2020 2020 2269 6e64              "ind
+000155b0: 6963 6573 206f 6620 496e 6469 7669 6475  ices of Individu
+000155c0: 616c 7320 746f 2062 6520 6164 6465 6420  als to be added 
+000155d0: 6672 6f6d 2022 0a20 2020 2020 2020 2020  from ".         
+000155e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000155f0: 2020 2020 2020 2020 2020 2022 6120 676e             "a gn
+00015600: 7820 5370 6563 6965 7320 6f62 6a65 6374  x Species object
+00015610: 2073 6572 7669 6e67 2061 7320 6120 736f   serving as a so
+00015620: 7572 6365 2022 0a20 2020 2020 2020 2020  urce ".         
+00015630: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015640: 2020 2020 2020 2020 2020 2022 706f 7075             "popu
+00015650: 6c61 7469 6f6e 2c20 736f 2063 616e 206f  lation, so can o
+00015660: 6e6c 7920 6265 2075 7365 6420 6966 2022  nly be used if "
+00015670: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00015680: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015690: 2020 2020 2022 2773 6f75 7263 655f 6d73       "'source_ms
+000156a0: 7072 696d 655f 7061 7261 6d73 2720 6973  prime_params' is
+000156b0: 204e 6f6e 6520 616e 6420 220a 2020 2020   None and ".    
+000156c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000156d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000156e0: 2227 736f 7572 6365 5f73 7070 2720 6973  "'source_spp' is
+000156f0: 2070 726f 7669 6465 642e 2229 0a0a 2020   provided.")..  
+00015700: 2020 2020 2020 2020 2020 2320 6d61 6b65            # make
+00015710: 2073 7572 6520 5370 6563 6965 7320 6861   sure Species ha
+00015720: 7320 6e6f 2073 656c 6563 7469 6f6e 2062  s no selection b
+00015730: 6566 6f72 6520 616c 6c6f 7769 6e67 2075  efore allowing u
+00015740: 7365 206f 6620 6d73 7072 696d 650a 2020  se of msprime.  
+00015750: 2020 2020 2020 2020 2020 6173 7365 7274            assert
+00015760: 206e 6f74 2073 656c 662e 7365 6c65 6374   not self.select
+00015770: 696f 6e2c 2028 226d 7370 7269 6d65 2069  ion, ("msprime i
+00015780: 7320 6120 636f 616c 6573 6365 6e74 2022  s a coalescent "
+00015790: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000157a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000157b0: 2020 2020 2020 2020 2020 2273 696d 756c            "simul
+000157c0: 6174 6f72 2c20 736f 2063 616e 6e6f 7420  ator, so cannot 
+000157d0: 6265 2073 696d 756c 6174 6520 220a 2020  be simulate ".  
+000157e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000157f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015800: 2020 2020 2020 2022 7365 6c65 6374 696f         "selectio
+00015810: 6e2e 2043 6f6e 7369 6465 7220 7573 696e  n. Consider usin
+00015820: 6720 6120 7365 636f 6e64 2022 0a20 2020  g a second ".   
+00015830: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015840: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015850: 2020 2020 2020 2247 656f 6e6f 6d69 6373        "Geonomics
+00015860: 206d 6f64 656c 2074 6f20 7369 6d75 6c61   model to simula
+00015870: 7465 2061 2073 6f75 7263 6520 220a 2020  te a source ".  
+00015880: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015890: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000158a0: 2020 2020 2020 2022 706f 7075 6c61 7469         "populati
+000158b0: 6f6e 2077 6974 6820 7365 6c65 6374 696f  on with selectio
+000158c0: 6e2e 2229 0a0a 2020 2020 2020 2020 2020  n.")..          
+000158d0: 2020 2320 6368 6563 6b20 7468 6174 2072    # check that r
+000158e0: 6571 7569 7265 6420 6d73 7072 696d 6520  equired msprime 
+000158f0: 7061 7261 6d73 2061 7265 2070 726f 7669  params are provi
+00015900: 6465 6420 616e 6420 616e 7920 6f74 6865  ded and any othe
+00015910: 7273 0a20 2020 2020 2020 2020 2020 2023  rs.            #
+00015920: 2070 726f 7669 6465 6420 6172 6520 616d   provided are am
+00015930: 6f6e 6720 7468 6f73 6520 616c 6c6f 7765  ong those allowe
+00015940: 640a 2020 2020 2020 2020 2020 2020 7265  d.            re
+00015950: 7175 6972 6564 5f6d 7370 7269 6d65 5f70  quired_msprime_p
+00015960: 6172 616d 7320 3d20 5b27 7265 636f 6d62  arams = ['recomb
+00015970: 5f72 6174 6527 2c20 276d 7574 5f72 6174  _rate', 'mut_rat
+00015980: 6527 5d0a 2020 2020 2020 2020 2020 2020  e'].            
+00015990: 616c 6c6f 7765 645f 6d73 7072 696d 655f  allowed_msprime_
+000159a0: 7061 7261 6d73 203d 205b 2764 656d 6f67  params = ['demog
+000159b0: 7261 7068 7927 2c20 2770 6f70 756c 6174  raphy', 'populat
+000159c0: 696f 6e5f 7369 7a65 272c 0a20 2020 2020  ion_size',.     
+000159d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000159e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000159f0: 2027 616e 6365 7374 7279 5f6d 6f64 656c   'ancestry_model
+00015a00: 272c 2027 7261 6e64 6f6d 5f73 6565 6427  ', 'random_seed'
+00015a10: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00015a20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015a30: 2020 2020 2020 205d 0a20 2020 2020 2020         ].       
+00015a40: 2020 2020 2061 6c6c 5f6d 7370 7269 6d65       all_msprime
+00015a50: 5f70 6172 616d 7320 3d20 2872 6571 7569  _params = (requi
+00015a60: 7265 645f 6d73 7072 696d 655f 7061 7261  red_msprime_para
+00015a70: 6d73 202b 0a20 2020 2020 2020 2020 2020  ms +.           
+00015a80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015a90: 2020 2020 2020 2061 6c6c 6f77 6564 5f6d         allowed_m
+00015aa0: 7370 7269 6d65 5f70 6172 616d 7329 0a20  sprime_params). 
+00015ab0: 2020 2020 2020 2020 2020 2061 7373 6572             asser
+00015ac0: 7420 6e70 2e61 6c6c 285b 6b20 696e 2061  t np.all([k in a
+00015ad0: 6c6c 5f6d 7370 7269 6d65 5f70 6172 616d  ll_msprime_param
+00015ae0: 7320 666f 720a 2020 2020 2020 2020 2020  s for.          
+00015af0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015b00: 2020 6b20 696e 2073 6f75 7263 655f 6d73    k in source_ms
+00015b10: 7072 696d 655f 7061 7261 6d73 5d29 2c20  prime_params]), 
+00015b20: 2822 5468 6520 6f6e 6c79 2070 6172 616d  ("The only param
+00015b30: 6574 6572 7322 0a20 2020 2020 2020 2020  eters".         
+00015b40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015b50: 2020 2020 2020 2020 2020 2022 2061 6363             " acc
+00015b60: 6570 7465 6420 696e 2027 736f 7572 6365  epted in 'source
+00015b70: 5f6d 7370 7269 6d65 5f70 6172 616d 7327  _msprime_params'
+00015b80: 2061 7265 3a20 220a 2020 2020 2020 2020   are: ".        
 00015b90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015ba0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015bb0: 2020 2062 6961 6c6c 656c 6963 3d46 616c     biallelic=Fal
-00015bc0: 7365 2c20 6173 5f64 6963 743d 5472 7565  se, as_dict=True
-00015bd0: 290a 2020 2020 2020 2020 6966 2062 795f  ).        if by_
-00015be0: 646f 6d69 6e61 6e63 6520 3d3d 2054 7275  dominance == Tru
-00015bf0: 653a 0a20 2020 2020 2020 2020 2020 2069  e:.            i
-00015c00: 6620 6c6f 6375 7320 696e 2073 656c 662e  f locus in self.
-00015c10: 6765 6e5f 6172 6368 2e6e 6f6e 6e65 7574  gen_arch.nonneut
-00015c20: 5f6c 6f63 693a 0a20 2020 2020 2020 2020  _loci:.         
-00015c30: 2020 2020 2020 2064 6f6d 203d 2073 656c         dom = sel
-00015c40: 662e 6765 6e5f 6172 6368 2e64 6f6d 5b6c  f.gen_arch.dom[l
-00015c50: 6f63 7573 5d0a 2020 2020 2020 2020 2020  ocus].          
-00015c60: 2020 2020 2020 2320 4e4f 5445 3a20 776f        # NOTE: wo
-00015c70: 756c 6420 6265 2073 696d 706c 6572 2074  uld be simpler t
-00015c80: 6f20 6a75 7374 2075 7365 206e 702e 6365  o just use np.ce
-00015c90: 696c 2068 6572 652e 2e2e 0a20 2020 2020  il here....     
-00015ca0: 2020 2020 2020 2020 2020 2023 2020 2020             #    
-00015cb0: 2020 2064 6f6e 2774 2073 6565 2077 6879     don't see why
-00015cc0: 2049 2064 6964 6e27 742e 2e2e 3f0a 2020   I didn't...?.  
-00015cd0: 2020 2020 2020 2020 2020 2020 2020 6765                ge
-00015ce0: 6e6f 7479 7065 7320 3d20 7b69 3a6e 702e  notypes = {i:np.
-00015cf0: 636c 6970 2867 7420 2a20 2831 202b 2064  clip(gt * (1 + d
-00015d00: 6f6d 292c 0a20 2020 2020 2020 2020 2020  om),.           
-00015d10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015d20: 2020 2020 2020 2020 2020 2020 615f 6d69              a_mi
-00015d30: 6e3d 4e6f 6e65 2c0a 2020 2020 2020 2020  n=None,.        
-00015d40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015d50: 2020 2020 2020 2020 2020 2020 2020 2061                 a
-00015d60: 5f6d 6178 3d31 2920 666f 7220 692c 2067  _max=1) for i, g
-00015d70: 7420 696e 2067 656e 6f74 7970 6573 2e69  t in genotypes.i
-00015d80: 7465 6d73 2829 7d0a 2020 2020 2020 2020  tems()}.        
-00015d90: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
-00015da0: 2020 2020 2020 2020 2020 7072 696e 7428            print(
-00015db0: 2827 5c6e 5c74 5741 524e 494e 473a 2054  ('\n\tWARNING: T
-00015dc0: 6865 2062 795f 646f 6d69 6e61 6e63 6520  he by_dominance 
-00015dd0: 6172 6775 6d65 6e74 2069 7320 5472 7565  argument is True
-00015de0: 2c20 270a 2020 2020 2020 2020 2020 2020  , '.            
-00015df0: 2020 2020 2020 2020 2020 2027 6275 7420             'but 
-00015e00: 6120 6e65 7574 7261 6c20 6c6f 6375 7320  a neutral locus 
-00015e10: 7761 7320 6368 6f73 656e 2c20 270a 2020  was chosen, '.  
+00015ba0: 2020 2020 2020 2020 2020 2066 227b 272c             f"{',
+00015bb0: 2027 2e6a 6f69 6e28 616c 6c5f 6d73 7072   '.join(all_mspr
+00015bc0: 696d 655f 7061 7261 6d73 297d 2e22 290a  ime_params)}.").
+00015bd0: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+00015be0: 7274 206e 702e 616c 6c28 5b6b 2069 6e20  rt np.all([k in 
+00015bf0: 736f 7572 6365 5f6d 7370 7269 6d65 5f70  source_msprime_p
+00015c00: 6172 616d 7320 666f 720a 2020 2020 2020  arams for.      
+00015c10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015c20: 2020 2020 2020 6b20 696e 2072 6571 7569        k in requi
+00015c30: 7265 645f 6d73 7072 696d 655f 7061 7261  red_msprime_para
+00015c40: 6d73 5d29 2c20 2822 426f 7468 2022 0a20  ms]), ("Both ". 
+00015c50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015c60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015c70: 2020 6622 7b27 2061 6e64 2027 2e6a 6f69    f"{' and '.joi
+00015c80: 6e28 7265 7175 6972 6564 5f6d 7370 7269  n(required_mspri
+00015c90: 6d65 5f70 6172 616d 7329 7d20 220a 2020  me_params)} ".  
+00015ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015cb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015cc0: 2020 226d 7573 7420 6265 2070 726f 7669    "must be provi
+00015cd0: 6465 6420 7769 7468 696e 2022 0a20 2020  ded within ".   
+00015ce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015cf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015d00: 2022 2773 6f75 7263 655f 6d73 7072 696d   "'source_msprim
+00015d10: 655f 7061 7261 6d73 272e 2229 0a0a 2020  e_params'.")..  
+00015d20: 2020 2020 2020 2020 2020 2320 7275 6e20            # run 
+00015d30: 7468 6520 7369 6d75 6c61 7469 6f6e 2061  the simulation a
+00015d40: 6e64 2072 6574 7572 6e20 7465 6d70 6c61  nd return templa
+00015d50: 7465 2049 6e64 6976 6964 7561 6c73 0a20  te Individuals. 
+00015d60: 2020 2020 2020 2020 2020 2075 7365 5f74             use_t
+00015d70: 736b 6974 203d 2073 656c 662e 6765 6e5f  skit = self.gen_
+00015d80: 6172 6368 2e75 7365 5f74 736b 6974 0a20  arch.use_tskit. 
+00015d90: 2020 2020 2020 2020 2020 2069 6e64 6976             indiv
+00015da0: 6964 732c 2073 6f75 7263 655f 7463 203d  ids, source_tc =
+00015db0: 205f 7369 6d5f 6d73 7072 696d 655f 696e   _sim_msprime_in
+00015dc0: 6469 7669 6475 616c 7328 6e3d 6e2c 0a20  dividuals(n=n,. 
+00015dd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015de0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015df0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015e00: 2020 204c 3d73 656c 662e 6765 6e5f 6172     L=self.gen_ar
+00015e10: 6368 2e4c 2c0a 2020 2020 2020 2020 2020  ch.L,.          
 00015e20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015e30: 2020 2020 2027 736f 2074 6865 2061 7267       'so the arg
-00015e40: 756d 656e 7420 7761 7320 6e6f 7420 7573  ument was not us
-00015e50: 6564 2e5c 6e27 2929 0a0a 2020 2020 2020  ed.\n'))..      
-00015e60: 2020 2320 6a75 7374 2061 7373 6967 6e20    # just assign 
-00015e70: 626c 6163 6b2c 2067 7261 792c 2061 6e64  black, gray, and
-00015e80: 2077 6869 7465 2028 7369 6e63 6520 7468   white (since th
-00015e90: 6572 6527 7320 6e6f 2072 6561 736f 6e0a  ere's no reason.
-00015ea0: 2020 2020 2020 2020 2320 6e65 6365 7373          # necess
-00015eb0: 6172 696c 7920 7468 6174 2074 6865 7365  arily that these
-00015ec0: 2073 686f 756c 6420 6265 206d 6170 7065   should be mappe
-00015ed0: 6420 746f 2061 2063 6572 7461 696e 206c  d to a certain l
-00015ee0: 6179 6572 2c20 7468 6520 7761 790a 2020  ayer, the way.  
-00015ef0: 2020 2020 2020 2320 7068 656e 6f74 7970        # phenotyp
-00015f00: 6520 7368 6f75 6c64 0a20 2020 2020 2020  e should.       
-00015f10: 2063 6f6c 6f72 7320 3d20 5b27 2330 3030   colors = ['#000
-00015f20: 3030 3027 2c20 2723 3830 3830 3830 272c  000', '#808080',
-00015f30: 2027 2346 4646 4646 4627 5d0a 0a20 2020   '#FFFFFF']..   
-00015f40: 2020 2020 2066 6f72 206e 2c20 6765 6e6f       for n, geno
-00015f50: 7479 7065 2069 6e20 656e 756d 6572 6174  type in enumerat
-00015f60: 6528 5b30 2e30 2c20 302e 352c 2031 2e30  e([0.0, 0.5, 1.0
-00015f70: 5d29 3a0a 2020 2020 2020 2020 2020 2020  ]):.            
-00015f80: 6765 6e6f 7479 7065 5f69 6e64 6976 6964  genotype_individ
-00015f90: 7320 3d20 5b69 2066 6f72 2069 2c20 6720  s = [i for i, g 
-00015fa0: 696e 2067 656e 6f74 7970 6573 2e69 7465  in genotypes.ite
-00015fb0: 6d73 280a 2020 2020 2020 2020 2020 2020  ms(.            
-00015fc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015fd0: 2020 2020 2020 2020 2020 2020 2920 6966              ) if
-00015fe0: 206e 702e 6174 6c65 6173 745f 3164 2867   np.atleast_1d(g
-00015ff0: 295b 305d 203d 3d20 6765 6e6f 7479 7065  )[0] == genotype
-00016000: 5d0a 2020 2020 2020 2020 2020 2020 2320  ].            # 
-00016010: 7769 6c6c 2068 6964 6520 7468 6520 6c61  will hide the la
-00016020: 6e64 2069 6620 7468 6973 2069 7320 6e6f  nd if this is no
-00016030: 7420 7468 6520 6669 7273 7420 706c 6f74  t the first plot
-00016040: 206d 6164 650a 2020 2020 2020 2020 2020   made.          
-00016050: 2020 6869 6465 5f6c 616e 6420 3d20 6e20    hide_land = n 
-00016060: 3e20 300a 2020 2020 2020 2020 2020 2020  > 0.            
-00016070: 2320 706c 6f74 2069 6620 7468 6572 6520  # plot if there 
-00016080: 6172 6520 616e 7920 696e 6469 7669 6475  are any individu
-00016090: 616c 7320 6f66 2074 6869 7320 6765 6e6f  als of this geno
-000160a0: 7479 7065 0a20 2020 2020 2020 2020 2020  type.           
-000160b0: 2069 6620 6c65 6e28 6765 6e6f 7479 7065   if len(genotype
-000160c0: 5f69 6e64 6976 6964 7329 203e 3d20 313a  _individs) >= 1:
-000160d0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-000160e0: 2073 656c 662e 5f70 6c6f 7428 6c79 725f   self._plot(lyr_
-000160f0: 6e75 6d3d 6c79 725f 6e75 6d2c 206c 616e  num=lyr_num, lan
-00016100: 643d 6c61 6e64 2c20 6869 6465 5f6c 616e  d=land, hide_lan
-00016110: 643d 6869 6465 5f6c 616e 642c 0a20 2020  d=hide_land,.   
-00016120: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016130: 2020 2020 2020 2020 696e 6469 7669 6473          individs
-00016140: 3d67 656e 6f74 7970 655f 696e 6469 7669  =genotype_indivi
-00016150: 6473 2c0a 2020 2020 2020 2020 2020 2020  ds,.            
-00016160: 2020 2020 2020 2020 2020 2020 2020 2074                 t
-00016170: 6578 743d 7465 7874 2c20 636f 6c6f 723d  ext=text, color=
-00016180: 636f 6c6f 7273 5b6e 5d2c 2065 6467 655f  colors[n], edge_
-00016190: 636f 6c6f 723d 6564 6765 5f63 6f6c 6f72  color=edge_color
-000161a0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-000161b0: 2020 2020 2020 2020 2020 2020 2074 6578               tex
-000161c0: 745f 636f 6c6f 723d 7465 7874 5f63 6f6c  t_color=text_col
-000161d0: 6f72 2c20 6362 6172 3d63 6261 722c 0a20  or, cbar=cbar,. 
-000161e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000161f0: 2020 2020 2020 2020 2020 7369 7a65 3d73            size=s
-00016200: 697a 652c 2074 6578 745f 7369 7a65 3d74  ize, text_size=t
-00016210: 6578 745f 7369 7a65 2c20 616c 7068 613d  ext_size, alpha=
-00016220: 616c 7068 612c 0a20 2020 2020 2020 2020  alpha,.         
-00016230: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016240: 2020 7a6f 6f6d 5f77 6964 7468 3d7a 6f6f    zoom_width=zoo
-00016250: 6d5f 7769 6474 682c 2078 3d78 2c20 793d  m_width, x=x, y=
-00016260: 792c 2076 6d69 6e3d 302c 2076 6d61 783d  y, vmin=0, vmax=
-00016270: 312c 0a20 2020 2020 2020 2020 2020 2020  1,.             
-00016280: 2020 2020 2020 2020 2020 2020 2020 7469                ti
-00016290: 636b 733d 7469 636b 732c 206d 6173 6b5f  cks=ticks, mask_
-000162a0: 7261 7374 3d6d 6173 6b5f 7261 7374 290a  rast=mask_rast).
-000162b0: 0a0a 2020 2020 2320 6d65 7468 6f64 2066  ..    # method f
-000162c0: 6f72 2070 6c6f 7474 696e 6720 696e 6469  or plotting indi
-000162d0: 7669 6475 616c 7320 636f 6c6f 7265 6420  viduals colored 
-000162e0: 6279 2074 6865 6972 2070 6865 6e6f 7479  by their phenoty
-000162f0: 7065 730a 2020 2020 2366 6f72 2061 2067  pes.    #for a g
-00016300: 6976 656e 2074 7261 6974 0a20 2020 2064  iven trait.    d
-00016310: 6566 205f 706c 6f74 5f70 6865 6e6f 7479  ef _plot_phenoty
-00016320: 7065 2873 656c 662c 2074 7261 6974 2c20  pe(self, trait, 
-00016330: 6c79 725f 6e75 6d3d 4e6f 6e65 2c20 6c61  lyr_num=None, la
-00016340: 6e64 3d4e 6f6e 652c 0a20 2020 2020 2020  nd=None,.       
-00016350: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016360: 2069 6e64 6976 6964 733d 4e6f 6e65 2c20   individs=None, 
-00016370: 7465 7874 3d46 616c 7365 2c20 7369 7a65  text=False, size
-00016380: 3d32 352c 2074 6578 745f 7369 7a65 3d39  =25, text_size=9
-00016390: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-000163a0: 2020 2020 2020 2020 2020 6564 6765 5f63            edge_c
-000163b0: 6f6c 6f72 3d27 626c 6163 6b27 2c20 7465  olor='black', te
-000163c0: 7874 5f63 6f6c 6f72 3d27 626c 6163 6b27  xt_color='black'
-000163d0: 2c20 6362 6172 3d54 7275 652c 0a20 2020  , cbar=True,.   
-000163e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000163f0: 2020 2020 2061 6c70 6861 3d31 2c20 7a6f       alpha=1, zo
-00016400: 6f6d 5f77 6964 7468 3d4e 6f6e 652c 2078  om_width=None, x
-00016410: 3d4e 6f6e 652c 2079 3d4e 6f6e 652c 0a20  =None, y=None,. 
-00016420: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016430: 2020 2020 2020 2074 6963 6b73 3d4e 6f6e         ticks=Non
-00016440: 652c 206d 6173 6b5f 7261 7374 3d4e 6f6e  e, mask_rast=Non
-00016450: 652c 2061 6e69 6d61 7465 3d46 616c 7365  e, animate=False
-00016460: 293a 0a0a 2020 2020 2020 2020 2320 6765  ):..        # ge
-00016470: 7420 7468 6520 7472 6169 7427 7320 6c79  t the trait's ly
-00016480: 725f 6e75 6d2c 2069 6620 6e6f 206c 7972  r_num, if no lyr
-00016490: 5f6e 756d 2070 726f 7669 6465 640a 2020  _num provided.  
-000164a0: 2020 2020 2020 6c79 725f 6e75 6d20 3d20        lyr_num = 
-000164b0: 7365 6c66 2e67 656e 5f61 7263 682e 7472  self.gen_arch.tr
-000164c0: 6169 7473 5b74 7261 6974 5d2e 6c79 725f  aits[trait].lyr_
-000164d0: 6e75 6d0a 0a20 2020 2020 2020 207a 203d  num..        z =
-000164e0: 204f 4428 7a69 7028 5b2a 7365 6c66 5d2c   OD(zip([*self],
-000164f0: 2073 656c 662e 5f67 6574 5f7a 2829 5b3a   self._get_z()[:
-00016500: 2c20 7472 6169 745d 2929 0a20 2020 2020  , trait])).     
-00016510: 2020 2069 6620 696e 6469 7669 6473 2069     if individs i
-00016520: 7320 6e6f 7420 4e6f 6e65 3a0a 2020 2020  s not None:.    
-00016530: 2020 2020 2020 2020 7a20 3d20 7b69 3a20          z = {i: 
-00016540: 7620 666f 7220 692c 2076 2069 6e20 7a2e  v for i, v in z.
-00016550: 6974 656d 7328 2920 6966 2069 2069 6e20  items() if i in 
-00016560: 696e 6469 7669 6473 7d0a 0a20 2020 2020  individs}..     
-00016570: 2020 2023 2067 6574 2074 6865 2063 6f72     # get the cor
-00016580: 7265 6374 2063 6d61 7020 666f 7220 7468  rect cmap for th
-00016590: 6973 2074 7261 6974 2773 206c 6179 6572  is trait's layer
-000165a0: 0a20 2020 2020 2020 2070 745f 636d 6170  .        pt_cmap
-000165b0: 203d 2076 697a 2e5f 6368 6f6f 7365 5f63   = viz._choose_c
-000165c0: 6d61 7028 7365 6c66 2e67 656e 5f61 7263  map(self.gen_arc
-000165d0: 682e 7472 6169 7473 5b74 7261 6974 5d2e  h.traits[trait].
-000165e0: 6c79 725f 6e75 6d29 0a0a 2020 2020 2020  lyr_num)..      
-000165f0: 2020 706f 696e 7473 203d 2073 656c 662e    points = self.
-00016600: 5f70 6c6f 7428 6c79 725f 6e75 6d20 3d20  _plot(lyr_num = 
-00016610: 6c79 725f 6e75 6d2c 206c 616e 6420 3d20  lyr_num, land = 
-00016620: 6c61 6e64 2c0a 2020 2020 2020 2020 2020  land,.          
-00016630: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016640: 2020 696e 6469 7669 6473 203d 2069 6e64    individs = ind
-00016650: 6976 6964 732c 2074 6578 7420 3d20 7465  ivids, text = te
-00016660: 7874 2c0a 2020 2020 2020 2020 2020 2020  xt,.            
-00016670: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016680: 636f 6c6f 7220 3d20 6c69 7374 287a 2e76  color = list(z.v
-00016690: 616c 7565 7328 2929 2c20 7074 5f63 6d61  alues()), pt_cma
-000166a0: 7020 3d20 7074 5f63 6d61 702c 0a20 2020  p = pt_cmap,.   
-000166b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000166c0: 2020 2020 2020 2020 2065 6467 655f 636f           edge_co
-000166d0: 6c6f 7220 3d20 6564 6765 5f63 6f6c 6f72  lor = edge_color
-000166e0: 2c20 7465 7874 5f63 6f6c 6f72 203d 2074  , text_color = t
-000166f0: 6578 745f 636f 6c6f 722c 0a20 2020 2020  ext_color,.     
-00016700: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016710: 2020 2020 2020 2063 6261 7220 3d20 6362         cbar = cb
-00016720: 6172 2c20 7369 7a65 203d 2073 697a 652c  ar, size = size,
-00016730: 2074 6578 745f 7369 7a65 203d 2074 6578   text_size = tex
-00016740: 745f 7369 7a65 2c0a 2020 2020 2020 2020  t_size,.        
-00016750: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016760: 2020 2020 616c 7068 6120 3d20 616c 7068      alpha = alph
-00016770: 612c 207a 6f6f 6d5f 7769 6474 6820 3d20  a, zoom_width = 
-00016780: 7a6f 6f6d 5f77 6964 7468 2c20 7820 3d20  zoom_width, x = 
-00016790: 782c 0a20 2020 2020 2020 2020 2020 2020  x,.             
-000167a0: 2020 2020 2020 2020 2020 2020 2020 2079                 y
-000167b0: 203d 2079 2c20 766d 696e 203d 2030 2c20   = y, vmin = 0, 
-000167c0: 766d 6178 203d 2031 2c20 7469 636b 733d  vmax = 1, ticks=
-000167d0: 7469 636b 732c 0a20 2020 2020 2020 2020  ticks,.         
-000167e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000167f0: 2020 206d 6173 6b5f 7261 7374 3d6d 6173     mask_rast=mas
-00016800: 6b5f 7261 7374 2c20 616e 696d 6174 653d  k_rast, animate=
-00016810: 616e 696d 6174 6529 0a0a 2020 2020 2020  animate)..      
-00016820: 2020 7265 7475 726e 2070 6f69 6e74 730a    return points.
-00016830: 0a0a 2020 2020 2320 6d65 7468 6f64 2066  ..    # method f
-00016840: 6f72 2070 6c6f 7474 696e 6720 696e 6469  or plotting indi
-00016850: 7669 6475 616c 7320 636f 6c6f 7265 6420  viduals colored 
-00016860: 6279 2074 6865 6972 206f 7665 7261 6c6c  by their overall
-00016870: 2066 6974 6e65 7373 6573 2c0a 2020 2020   fitnesses,.    
-00016880: 236f 7220 6279 2074 6865 6972 2066 6974  #or by their fit
-00016890: 6e65 7373 6573 2066 6f72 2061 2073 696e  nesses for a sin
-000168a0: 676c 6520 7472 6169 7420 2869 6620 7472  gle trait (if tr
-000168b0: 6169 7420 6973 206e 6f74 204e 6f6e 6529  ait is not None)
-000168c0: 0a20 2020 2064 6566 205f 706c 6f74 5f66  .    def _plot_f
-000168d0: 6974 6e65 7373 2873 656c 662c 2074 7274  itness(self, trt
-000168e0: 5f6e 756d 3d4e 6f6e 652c 206c 7972 5f6e  _num=None, lyr_n
-000168f0: 756d 3d4e 6f6e 652c 206c 616e 643d 4e6f  um=None, land=No
-00016900: 6e65 2c0a 2020 2020 2020 2020 2020 2020  ne,.            
-00016910: 2020 2020 2020 2020 2020 696e 6469 7669            indivi
-00016920: 6473 3d4e 6f6e 652c 2074 6578 743d 4661  ds=None, text=Fa
-00016930: 6c73 652c 2070 6865 6e6f 7479 7065 5f74  lse, phenotype_t
-00016940: 6578 743d 4661 6c73 652c 0a20 2020 2020  ext=False,.     
-00016950: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016960: 2070 6865 6e6f 7479 7065 5f74 6578 745f   phenotype_text_
-00016970: 636f 6c6f 723d 2762 6c61 636b 272c 2066  color='black', f
-00016980: 6974 6e65 7373 5f74 6578 743d 4661 6c73  itness_text=Fals
-00016990: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-000169a0: 2020 2020 2020 2020 2066 6974 6e65 7373           fitness
-000169b0: 5f74 6578 745f 636f 6c6f 723d 2723 3333  _text_color='#33
-000169c0: 3333 3333 272c 2073 697a 653d 3130 302c  3333', size=100,
-000169d0: 2074 6578 745f 7369 7a65 203d 2039 2c0a   text_size = 9,.
-000169e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000169f0: 2020 2020 2020 6564 6765 5f63 6f6c 6f72        edge_color
-00016a00: 3d27 626c 6163 6b27 2c20 7465 7874 5f63  ='black', text_c
-00016a10: 6f6c 6f72 3d27 626c 6163 6b27 2c0a 2020  olor='black',.  
-00016a20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00016a30: 2020 2020 6669 745f 636d 6170 203d 2027      fit_cmap = '
-00016a40: 5264 596c 476e 272c 2063 6261 723d 5472  RdYlGn', cbar=Tr
-00016a50: 7565 2c20 6669 746e 6573 735f 6362 6172  ue, fitness_cbar
-00016a60: 3d54 7275 652c 0a20 2020 2020 2020 2020  =True,.         
-00016a70: 2020 2020 2020 2020 2020 2020 2061 6c70               alp
-00016a80: 6861 3d31 2c20 7a6f 6f6d 5f77 6964 7468  ha=1, zoom_width
-00016a90: 3d4e 6f6e 652c 2078 3d4e 6f6e 652c 2079  =None, x=None, y
-00016aa0: 3d4e 6f6e 652c 2074 6963 6b73 3d4e 6f6e  =None, ticks=Non
-00016ab0: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-00016ac0: 2020 2020 2020 2020 206d 6173 6b5f 7261           mask_ra
-00016ad0: 7374 3d4e 6f6e 6529 3a0a 0a20 2020 2020  st=None):..     
-00016ae0: 2020 2023 7265 7475 726e 206d 6573 7361     #return messa
-00016af0: 6765 7320 6966 2073 7065 6369 6573 2064  ges if species d
-00016b00: 6f65 7320 6e6f 7420 6861 7665 2067 656e  oes not have gen
-00016b10: 6f6d 6573 206f 7220 7472 6169 7473 0a20  omes or traits. 
-00016b20: 2020 2020 2020 2069 6620 7365 6c66 2e67         if self.g
-00016b30: 656e 5f61 7263 6820 6973 204e 6f6e 653a  en_arch is None:
-00016b40: 0a20 2020 2020 2020 2020 2020 2070 7269  .            pri
-00016b50: 6e74 2828 2253 7065 6369 6573 2e5f 706c  nt(("Species._pl
-00016b60: 6f74 5f66 6974 6e65 7373 2069 7320 6e6f  ot_fitness is no
-00016b70: 7420 7661 6c69 6420 666f 7220 7370 6563  t valid for spec
-00016b80: 6965 7320 220a 2020 2020 2020 2020 2020  ies ".          
-00016b90: 2020 2020 2020 2020 2022 7769 7468 6f75           "withou
-00016ba0: 7420 6765 6e6f 6d65 732e 5c6e 2229 290a  t genomes.\n")).
-00016bb0: 2020 2020 2020 2020 2020 2020 7265 7475              retu
-00016bc0: 726e 0a20 2020 2020 2020 2065 6c69 6620  rn.        elif 
-00016bd0: 7365 6c66 2e67 656e 5f61 7263 682e 7472  self.gen_arch.tr
-00016be0: 6169 7473 2069 7320 4e6f 6e65 3a0a 2020  aits is None:.  
-00016bf0: 2020 2020 2020 2020 2020 7072 696e 7428            print(
-00016c00: 2822 5370 6563 6965 732e 5f70 6c6f 745f  ("Species._plot_
-00016c10: 6669 746e 6573 7320 6973 206e 6f74 2076  fitness is not v
-00016c20: 616c 6964 2066 6f72 2073 7065 6369 6573  alid for species
-00016c30: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
-00016c40: 2020 2020 2020 2277 6974 686f 7574 2074        "without t
-00016c50: 7261 6974 732e 5c6e 2229 290a 2020 2020  raits.\n")).    
-00016c60: 2020 2020 2020 2020 7265 7475 726e 0a0a          return..
-00016c70: 2020 2020 2020 2020 2320 6765 7420 7468          # get th
-00016c80: 6520 7472 6169 7427 7320 6c79 725f 6e75  e trait's lyr_nu
-00016c90: 6d2c 2069 6620 6c79 725f 6e75 6d20 7761  m, if lyr_num wa
-00016ca0: 736e 2774 2070 726f 7669 6465 6420 6275  sn't provided bu
-00016cb0: 7420 7472 745f 6e75 6d20 7761 730a 2020  t trt_num was.  
-00016cc0: 2020 2020 2020 6966 2074 7274 5f6e 756d        if trt_num
-00016cd0: 2069 7320 6e6f 7420 4e6f 6e65 2061 6e64   is not None and
-00016ce0: 206c 7972 5f6e 756d 2069 7320 4e6f 6e65   lyr_num is None
-00016cf0: 3a0a 2020 2020 2020 2020 2020 2020 6c79  :.            ly
-00016d00: 725f 6e75 6d20 3d20 7365 6c66 2e67 656e  r_num = self.gen
-00016d10: 5f61 7263 682e 7472 6169 7473 5b74 7274  _arch.traits[trt
-00016d20: 5f6e 756d 5d2e 6c79 725f 6e75 6d0a 0a20  _num].lyr_num.. 
-00016d30: 2020 2020 2020 2023 2067 6574 2061 6c6c         # get all
-00016d40: 2069 6e64 6976 6964 7327 2066 6974 6e65   individs' fitne
-00016d50: 7373 2076 616c 7565 732c 0a20 2020 2020  ss values,.     
-00016d60: 2020 2023 2061 6e64 2067 6574 2061 7070     # and get app
-00016d70: 726f 7072 6961 7465 2063 6f6c 6f72 6d61  ropriate colorma
-00016d80: 700a 2020 2020 2020 2020 6966 2074 7274  p.        if trt
-00016d90: 5f6e 756d 2069 7320 4e6f 6e65 3a0a 2020  _num is None:.  
-00016da0: 2020 2020 2020 2020 2020 7720 3d20 7365            w = se
-00016db0: 6c66 2e5f 6361 6c63 5f66 6974 6e65 7373  lf._calc_fitness
-00016dc0: 2829 0a20 2020 2020 2020 2020 2020 2070  ().            p
-00016dd0: 745f 636d 6170 203d 2027 4772 6579 735f  t_cmap = 'Greys_
-00016de0: 7227 0a20 2020 2020 2020 2065 6c73 653a  r'.        else:
-00016df0: 0a20 2020 2020 2020 2020 2020 2077 203d  .            w =
-00016e00: 2073 656c 662e 5f63 616c 635f 6669 746e   self._calc_fitn
-00016e10: 6573 7328 7472 6169 745f 6e75 6d20 3d20  ess(trait_num = 
-00016e20: 7472 745f 6e75 6d29 0a0a 2020 2020 2020  trt_num)..      
-00016e30: 2020 2366 696c 7465 7220 6f75 7420 756e    #filter out un
-00016e40: 7761 6e74 6564 2069 6e64 6976 6964 732c  wanted individs,
-00016e50: 2069 6620 6e65 6365 7373 6172 790a 2020   if necessary.  
-00016e60: 2020 2020 2020 7720 3d20 4f44 287a 6970        w = OD(zip
-00016e70: 285b 2a73 656c 665d 2c20 7729 290a 2020  ([*self], w)).  
-00016e80: 2020 2020 2020 6966 2069 6e64 6976 6964        if individ
-00016e90: 7320 6973 206e 6f74 204e 6f6e 653a 0a20  s is not None:. 
-00016ea0: 2020 2020 2020 2020 2020 2077 203d 207b             w = {
-00016eb0: 693a 7620 666f 7220 692c 7620 696e 2077  i:v for i,v in w
-00016ec0: 2e69 7465 6d73 2829 2069 6620 6920 696e  .items() if i in
-00016ed0: 2069 6e64 6976 6964 737d 0a0a 2020 2020   individs}..    
-00016ee0: 2020 2020 2320 6361 6c63 206d 696e 696d      # calc minim
-00016ef0: 756d 2070 6f73 7369 626c 6520 6669 746e  um possible fitn
-00016f00: 6573 7320 2866 6f72 2070 6865 6e6f 7479  ess (for phenoty
-00016f10: 7065 7320 7769 7468 696e 2030 203c 3d20  pes within 0 <= 
-00016f20: 7a20 3c3d 2031 2c0a 2020 2020 2020 2020  z <= 1,.        
-00016f30: 2377 6869 6368 2069 6e20 7265 616c 6974  #which in realit
-00016f40: 7920 6973 6e27 7420 6120 636f 6e73 7472  y isn't a constr
-00016f50: 6169 6e74 2c20 6275 7420 7661 6c75 6573  aint, but values
-00016f60: 206c 6f77 6572 2074 6861 6e0a 2020 2020   lower than.    
-00016f70: 2020 2020 2374 6869 7320 7769 6c6c 2061      #this will a
-00016f80: 6c73 6f20 6265 2063 6f6e 7374 7261 696e  lso be constrain
-00016f90: 6564 2074 6f20 7468 6520 6d69 6e69 6d75  ed to the minimu
-00016fa0: 6d2d 7661 6c75 6520 636f 6c6f 7220 666f  m-value color fo
-00016fb0: 7220 706c 6f74 7469 6e67 290a 2020 2020  r plotting).    
-00016fc0: 2020 2020 234e 4f54 453a 2074 6865 206e      #NOTE: the n
-00016fd0: 702e 6174 6c65 6173 745f 3264 282e 2e2e  p.atleast_2d(...
-00016fe0: 292e 6d69 6e28 2920 636f 6e73 7472 7563  ).min() construc
-00016ff0: 7420 6d61 6b65 7320 7468 6973 0a20 2020  t makes this.   
-00017000: 2020 2020 2023 776f 726b 2062 6f74 6820       #work both 
-00017010: 666f 7220 6669 7865 6420 616e 6420 7370  for fixed and sp
-00017020: 6174 6961 6c6c 7920 7661 7279 696e 6720  atially varying 
-00017030: 7068 690a 2020 2020 2020 2020 6966 2074  phi.        if t
-00017040: 7274 5f6e 756d 2069 7320 4e6f 6e65 3a0a  rt_num is None:.
-00017050: 2020 2020 2020 2020 2020 2020 6d69 6e5f              min_
-00017060: 6669 7420 3d20 6e70 2e70 726f 6475 6374  fit = np.product
-00017070: 285b 3120 2d20 6e70 2e61 746c 6561 7374  ([1 - np.atleast
-00017080: 5f32 6428 742e 7068 6929 2e6d 696e 280a  _2d(t.phi).min(.
-00017090: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000170a0: 2020 2020 2020 2020 2020 2020 2920 666f              ) fo
-000170b0: 7220 7420 696e 206c 6973 7428 7365 6c66  r t in list(self
-000170c0: 2e67 656e 5f61 7263 682e 7472 6169 7473  .gen_arch.traits
-000170d0: 2e76 616c 7565 7328 2929 5d29 0a20 2020  .values())]).   
-000170e0: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
-000170f0: 2020 2020 2020 206d 696e 5f66 6974 203d         min_fit =
-00017100: 2031 202d 206e 702e 6174 6c65 6173 745f   1 - np.atleast_
-00017110: 3264 2873 656c 662e 6765 6e5f 6172 6368  2d(self.gen_arch
-00017120: 2e74 7261 6974 735b 0a20 2020 2020 2020  .traits[.       
-00017130: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015e30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015e40: 2020 2020 2020 2020 2020 7573 655f 7473            use_ts
+00015e50: 6b69 743d 7573 655f 7473 6b69 742c 0a20  kit=use_tskit,. 
+00015e60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015e70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015e80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015e90: 2020 202a 2a73 6f75 7263 655f 6d73 7072     **source_mspr
+00015ea0: 696d 655f 7061 7261 6d73 2c0a 2020 2020  ime_params,.    
+00015eb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015ec0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015ed0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00015ee0: 290a 0a20 2020 2020 2020 2020 2020 2023  )..            #
+00015ef0: 2072 756e 2061 6c6c 2054 6162 6c65 436f   run all TableCo
+00015f00: 6c6c 6563 7469 6f6e 2065 6469 7473 2074  llection edits t
+00015f10: 6861 7420 6172 6520 6e65 6365 7373 6172  hat are necessar
+00015f20: 7920 746f 2061 6c69 676e 206d 7370 7269  y to align mspri
+00015f30: 6d65 0a20 2020 2020 2020 2020 2020 2023  me.            #
+00015f40: 206f 7574 7075 7420 7769 7468 2074 6865   output with the
+00015f50: 2054 6162 6c65 436f 6c6c 6563 7469 6f6e   TableCollection
+00015f60: 2063 6f6e 7665 6e74 696f 6e73 2075 7365   conventions use
+00015f70: 6420 696e 2067 6e78 0a20 2020 2020 2020  d in gnx.       
+00015f80: 2020 2020 2023 204e 4f54 453a 2065 6469       # NOTE: edi
+00015f90: 7473 2074 6865 2054 6162 6c65 436f 6c6c  ts the TableColl
+00015fa0: 6563 7469 6f6e 2069 6e20 706c 6163 653b  ection in place;
+00015fb0: 0a20 2020 2020 2020 2020 2020 2023 2020  .            #  
+00015fc0: 2020 2020 2072 6574 7572 6e73 2074 6865       returns the
+00015fd0: 206e 6578 7420 6e20 676e 7820 496e 6469   next n gnx Indi
+00015fe0: 7669 6475 616c 2069 6478 7320 6173 7369  vidual idxs assi
+00015ff0: 676e 6564 2074 6f20 7468 6f73 650a 2020  gned to those.  
+00016000: 2020 2020 2020 2020 2020 2320 2020 2020            #     
+00016010: 2020 496e 6469 7669 6475 616c 7320 666f    Individuals fo
+00016020: 7220 696e 636f 7270 6f72 6174 696f 6e20  r incorporation 
+00016030: 696e 746f 2074 6865 2067 6976 656e 2053  into the given S
+00016040: 7065 6369 6573 0a20 2020 2020 2020 2020  pecies.         
+00016050: 2020 206e 6578 745f 676e 785f 6964 7873     next_gnx_idxs
+00016060: 203d 205f 7072 6570 5f74 736b 6974 5f74   = _prep_tskit_t
+00016070: 6162 636f 6c6c 5f66 6f72 5f67 6e78 5f73  abcoll_for_gnx_s
+00016080: 7070 5f75 6e69 6f6e 2874 633d 736f 7572  pp_union(tc=sour
+00016090: 6365 5f74 632c 0a20 2020 2020 2020 2020  ce_tc,.         
+000160a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000160b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000160c0: 2020 2020 2020 2020 2020 2020 2072 6563               rec
+000160d0: 6970 5f73 7070 3d73 656c 662c 0a20 2020  ip_spp=self,.   
+000160e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000160f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016100: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016110: 2020 2063 6f6f 7264 733d 636f 6f72 6473     coords=coords
+00016120: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00016130: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016140: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016150: 2020 2020 2020 2020 736f 7572 6365 5f73          source_s
+00016160: 6f66 7477 6172 653d 276d 7370 7269 6d65  oftware='msprime
+00016170: 272c 0a20 2020 2020 2020 2020 2020 2020  ',.             
+00016180: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016190: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000161a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000161b0: 2020 2020 290a 0a20 2020 2020 2020 2023      )..        #
+000161c0: 2073 6574 2074 6865 2073 6f75 7263 6520   set the source 
+000161d0: 496e 6469 7669 6475 616c 7327 2063 6f6f  Individuals' coo
+000161e0: 7264 696e 6174 6573 0a20 2020 2020 2020  rdinates.       
+000161f0: 205b 696e 642e 5f73 6574 5f70 6f73 282a   [ind._set_pos(*
+00016200: 636f 6f72 6473 5b69 2c3a 5d29 2066 6f72  coords[i,:]) for
+00016210: 2069 2c20 696e 6420 696e 2065 6e75 6d65   i, ind in enume
+00016220: 7261 7465 2869 6e64 6976 6964 7329 5d0a  rate(individs)].
+00016230: 0a20 2020 2020 2020 2023 2075 7064 6174  .        # updat
+00016240: 6520 7468 6520 696e 6465 7820 7661 6c75  e the index valu
+00016250: 6573 2069 6465 6e74 6966 7969 6e67 2049  es identifying I
+00016260: 6e64 6976 6964 7561 6c73 2069 6e20 676e  ndividuals in gn
+00016270: 7820 616e 6420 7473 6b69 7420 6461 7461  x and tskit data
+00016280: 0a20 2020 2020 2020 206d 6178 5f69 6478  .        max_idx
+00016290: 5f62 3420 3d20 7365 6c66 2e6d 6178 5f69  _b4 = self.max_i
+000162a0: 6e64 5f69 6478 0a20 2020 2020 2020 206e  nd_idx.        n
+000162b0: 6577 5f69 6478 7320 3d20 5b5d 0a20 2020  ew_idxs = [].   
+000162c0: 2020 2020 2066 6f72 2069 2c20 696e 6420       for i, ind 
+000162d0: 696e 2065 6e75 6d65 7261 7465 2869 6e64  in enumerate(ind
+000162e0: 6976 6964 7329 3a0a 2020 2020 2020 2020  ivids):.        
+000162f0: 2020 2020 2320 696e 6372 656d 656e 7420      # increment 
+00016300: 7468 6520 6d61 785f 696e 645f 6964 780a  the max_ind_idx.
+00016310: 2020 2020 2020 2020 2020 2020 7365 6c66              self
+00016320: 2e6d 6178 5f69 6e64 5f69 6478 202b 3d20  .max_ind_idx += 
+00016330: 310a 2020 2020 2020 2020 2020 2020 2320  1.            # 
+00016340: 7365 7420 7468 6520 496e 6469 7669 6475  set the Individu
+00016350: 616c 2773 2067 6e78 2069 6e64 6578 0a20  al's gnx index. 
+00016360: 2020 2020 2020 2020 2020 2069 6e64 2e69             ind.i
+00016370: 6478 203d 2073 656c 662e 6d61 785f 696e  dx = self.max_in
+00016380: 645f 6964 780a 2020 2020 2020 2020 2020  d_idx.          
+00016390: 2020 6e65 775f 6964 7873 2e61 7070 656e    new_idxs.appen
+000163a0: 6428 7365 6c66 2e6d 6178 5f69 6e64 5f69  d(self.max_ind_i
+000163b0: 6478 290a 2020 2020 2020 2020 2020 2020  dx).            
+000163c0: 2320 7365 7420 7468 6520 496e 6469 7669  # set the Indivi
+000163d0: 6475 616c 7327 2069 6473 2066 6f72 2074  duals' ids for t
+000163e0: 6865 2074 736b 6974 206e 6f64 6573 2074  he tskit nodes t
+000163f0: 6162 6c65 0a20 2020 2020 2020 2020 2020  able.           
+00016400: 2069 6e64 2e5f 6e6f 6465 735f 7461 625f   ind._nodes_tab_
+00016410: 6964 7320 3d20 7b6b 3a20 7620 2b20 7365  ids = {k: v + se
+00016420: 6c66 2e5f 7463 2e6e 6f64 6573 2e6e 756d  lf._tc.nodes.num
+00016430: 5f72 6f77 7320 666f 720a 2020 2020 2020  _rows for.      
+00016440: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016450: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016460: 2020 2020 2020 6b2c 2076 2069 6e20 696e        k, v in in
+00016470: 642e 5f6e 6f64 6573 5f74 6162 5f69 6473  d._nodes_tab_ids
+00016480: 2e69 7465 6d73 2829 7d0a 2020 2020 2020  .items()}.      
+00016490: 2020 2020 2020 2320 7365 7420 7468 6520        # set the 
+000164a0: 496e 6469 7669 6475 616c 7327 2069 6473  Individuals' ids
+000164b0: 2066 6f72 2074 6865 2074 736b 6974 2069   for the tskit i
+000164c0: 6e64 6976 6964 7561 6c73 2074 6162 6c65  ndividuals table
+000164d0: 0a20 2020 2020 2020 2020 2020 2023 204e  .            # N
+000164e0: 4f54 453a 2074 6865 2069 6e76 6572 7365  OTE: the inverse
+000164f0: 2028 7468 6520 496e 6469 7669 6475 616c   (the Individual
+00016500: 7327 2067 6e78 2069 6478 2076 616c 7565  s' gnx idx value
+00016510: 732c 0a20 2020 2020 2020 2020 2020 2023  s,.            #
+00016520: 2020 2020 2020 2061 7320 7265 636f 7265         as recore
+00016530: 6420 696e 2074 6865 2074 736b 6974 2069  d in the tskit i
+00016540: 6e64 6976 6964 7561 6c73 2074 6162 6c65  ndividuals table
+00016550: 2773 0a20 2020 2020 2020 2020 2020 2023  's.            #
+00016560: 2020 2020 2020 206d 6574 6164 6174 6129         metadata)
+00016570: 2073 686f 756c 6420 616c 7265 6164 7920   should already 
+00016580: 6265 2063 6f72 7265 6374 0a20 2020 2020  be correct.     
+00016590: 2020 2020 2020 2023 2020 2020 2020 2062         #       b
+000165a0: 6563 6175 7365 205f 7072 6570 5f74 736b  ecause _prep_tsk
+000165b0: 6974 5f74 6162 636f 6c6c 5f66 6f72 5f67  it_tabcoll_for_g
+000165c0: 6e78 5f73 7070 5f75 6e69 6f6e 0a20 2020  nx_spp_union.   
+000165d0: 2020 2020 2020 2020 2023 2020 2020 2020           #      
+000165e0: 2067 656e 6572 6174 6564 2074 6865 2069   generated the i
+000165f0: 6e64 6578 2076 616c 7565 730a 2020 2020  ndex values.    
+00016600: 2020 2020 2020 2020 2320 2020 2020 2020          #       
+00016610: 616e 6420 7365 7420 7468 656d 2069 6e20  and set them in 
+00016620: 7468 6520 5461 626c 6543 6f6c 6c65 6374  the TableCollect
+00016630: 696f 6e0a 2020 2020 2020 2020 2020 2020  ion.            
+00016640: 696e 642e 5f69 6e64 6976 6964 7561 6c73  ind._individuals
+00016650: 5f74 6162 5f69 6420 2b3d 2073 656c 662e  _tab_id += self.
+00016660: 5f74 632e 696e 6469 7669 6475 616c 732e  _tc.individuals.
+00016670: 6e75 6d5f 726f 7773 0a20 2020 2020 2020  num_rows.       
+00016680: 206d 6178 5f69 6478 5f61 6620 3d20 7365   max_idx_af = se
+00016690: 6c66 2e6d 6178 5f69 6e64 5f69 6478 0a20  lf.max_ind_idx. 
+000166a0: 2020 2020 2020 2061 7373 6572 7420 6d61         assert ma
+000166b0: 785f 6964 785f 6166 202d 206d 6178 5f69  x_idx_af - max_i
+000166c0: 6478 5f62 3420 3d3d 206c 656e 2869 6e64  dx_b4 == len(ind
+000166d0: 6976 6964 7329 0a20 2020 2020 2020 2023  ivids).        #
+000166e0: 2063 6f6d 7061 7265 2074 6865 2069 6478   compare the idx
+000166f0: 2076 616c 7565 7320 7468 6174 2077 6572   values that wer
+00016700: 6520 7365 7420 6f6e 2074 6865 2049 6e64  e set on the Ind
+00016710: 6976 6964 7561 6c73 2074 6f20 7468 6f73  ividuals to thos
+00016720: 6520 7468 6174 0a20 2020 2020 2020 2023  e that.        #
+00016730: 2077 6572 6520 7365 7420 696e 2074 6865   were set in the
+00016740: 2054 6162 6c65 436f 6c6c 6563 7469 6f6e   TableCollection
+00016750: 2e69 6e64 6976 6964 7561 6c73 2074 6162  .individuals tab
+00016760: 6c65 0a20 2020 2020 2020 2061 7373 6572  le.        asser
+00016770: 7420 6e70 2e61 6c6c 286e 702e 6172 7261  t np.all(np.arra
+00016780: 7928 6e65 7874 5f67 6e78 5f69 6478 7329  y(next_gnx_idxs)
+00016790: 203d 3d20 6e70 2e61 7272 6179 286e 6577   == np.array(new
+000167a0: 5f69 6478 7329 292c 2028 0a20 2020 2020  _idxs)), (.     
+000167b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000167c0: 2020 2020 2020 2020 2020 2020 2249 6e64              "Ind
+000167d0: 6976 6964 6175 6c20 6964 7820 7661 6c75  ividaul idx valu
+000167e0: 6573 2073 6574 2069 6e20 7468 6520 220a  es set in the ".
+000167f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016800: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016810: 2020 2020 2254 6162 6c65 436f 6c6c 6563      "TableCollec
+00016820: 7469 6f6e 2e69 6e64 6976 6964 7561 6c73  tion.individuals
+00016830: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00016840: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016850: 2020 2020 2020 2022 7461 626c 6520 646f         "table do
+00016860: 206e 6f74 2061 6772 6565 2077 6974 6820   not agree with 
+00016870: 7468 6f73 6520 7365 7420 6f6e 2074 6865  those set on the
+00016880: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00016890: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000168a0: 2020 2020 2020 2022 496e 6469 7669 6475         "Individu
+000168b0: 616c 7320 7468 656d 7365 6c76 6573 2e22  als themselves."
+000168c0: 290a 0a20 2020 2020 2020 2023 2073 6f72  )..        # sor
+000168d0: 7420 616e 6420 7369 6d70 6c69 6679 2074  t and simplify t
+000168e0: 6865 2074 736b 6974 2e54 6162 6c65 436f  he tskit.TableCo
+000168f0: 6c6c 6563 7469 6f6e 0a20 2020 2020 2020  llection.       
+00016900: 2073 656c 662e 5f73 6f72 745f 616e 645f   self._sort_and_
+00016910: 7369 6d70 6c69 6679 5f74 6162 6c65 5f63  simplify_table_c
+00016920: 6f6c 6c65 6374 696f 6e28 290a 0a20 2020  ollection()..   
+00016930: 2020 2020 2023 2075 6e69 6f6e 2073 6f75       # union sou
+00016940: 7263 6520 696e 6469 7669 6475 616c 7327  rce individuals'
+00016950: 2054 6162 6c65 436f 6c6c 6563 7469 6f6e   TableCollection
+00016960: 2069 6e74 6f20 7468 6973 2053 7065 6369   into this Speci
+00016970: 6573 2720 5461 626c 6543 6f6c 6c0a 2020  es' TableColl.  
+00016980: 2020 2020 2020 6e6f 6465 735f 7461 625f        nodes_tab_
+00016990: 6c65 6e5f 6234 203d 2073 656c 662e 5f74  len_b4 = self._t
+000169a0: 632e 6e6f 6465 732e 6e75 6d5f 726f 7773  c.nodes.num_rows
+000169b0: 0a20 2020 2020 2020 2069 6e64 6976 6964  .        individ
+000169c0: 735f 7461 625f 6c65 6e5f 6234 203d 2073  s_tab_len_b4 = s
+000169d0: 656c 662e 5f74 632e 696e 6469 7669 6475  elf._tc.individu
+000169e0: 616c 732e 6e75 6d5f 726f 7773 0a20 2020  als.num_rows.   
+000169f0: 2020 2020 2073 656c 662e 5f74 632e 756e       self._tc.un
+00016a00: 696f 6e28 736f 7572 6365 5f74 632c 0a20  ion(source_tc,. 
+00016a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016a20: 2020 2020 2020 2020 2020 2023 204e 4f54             # NOT
+00016a30: 453a 2028 666f 7220 6e6f 7720 6174 206c  E: (for now at l
+00016a40: 6561 7374 2920 6173 7375 6d69 6e67 2020  east) assuming  
+00016a50: 7468 6520 696e 7472 6f64 7563 6564 0a20  the introduced. 
+00016a60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016a70: 2020 2020 2020 2020 2020 2023 2020 2020             #    
+00016a80: 2020 2069 6e64 6976 6964 7561 6c73 2063     individuals c
+00016a90: 6f6d 6520 6672 6f6d 2074 6f74 616c 6c79  ome from totally
+00016aa0: 2069 6e64 6570 2070 6f70 2c0a 2020 2020   indep pop,.    
+00016ab0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016ac0: 2020 2020 2020 2020 2320 2020 2020 2020          #       
+00016ad0: 646f 206e 6f74 2073 6861 7265 2061 204d  do not share a M
+00016ae0: 5243 4120 7769 7468 696e 206f 7572 2064  RCA within our d
+00016af0: 6174 612c 2061 6e64 20e2 88b4 2020 200a  ata, and ...   .
+00016b00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016b10: 2020 2020 2020 2020 2020 2020 2320 2020              #   
+00016b20: 2020 2020 6861 7665 206e 6f20 636f 6d6d      have no comm
+00016b30: 6f6e 206e 6f64 6573 2074 6861 7420 6e65  on nodes that ne
+00016b40: 6564 2074 6f20 6265 206d 6170 7065 642c  ed to be mapped,
+00016b50: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00016b60: 2020 2020 2020 2020 2020 2020 2023 2020               #  
+00016b70: 2020 2020 2068 656e 6365 2077 6520 7072       hence we pr
+00016b80: 6f76 6964 6520 6120 6c69 7374 206f 6620  ovide a list of 
+00016b90: 7473 6b69 742e 4e55 4c4c 730a 2020 2020  tskit.NULLs.    
+00016ba0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016bb0: 2020 2020 2020 2020 2320 2020 2020 2020          #       
+00016bc0: 746f 2075 6e69 6f6e 2829 2773 206e 6f64  to union()'s nod
+00016bd0: 655f 6d61 7070 696e 6720 6172 6775 6d65  e_mapping argume
+00016be0: 6e74 0a20 2020 2020 2020 2020 2020 2020  nt.             
+00016bf0: 2020 2020 2020 2020 2020 2020 2020 206e                 n
+00016c00: 6f64 655f 6d61 7070 696e 673d 5b74 736b  ode_mapping=[tsk
+00016c10: 6974 2e4e 554c 4c5d 202a 736f 7572 6365  it.NULL] *source
+00016c20: 5f74 632e 6e6f 6465 732e 6e75 6d5f 726f  _tc.nodes.num_ro
+00016c30: 7773 2c0a 2020 2020 2020 2020 2020 2020  ws,.            
+00016c40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016c50: 2320 4e4f 5445 3a20 706f 7075 6c61 7469  # NOTE: populati
+00016c60: 6f6e 7320 6f66 2074 6865 206e 6577 6c79  ons of the newly
+00016c70: 2061 6464 6564 206e 6f64 6573 0a20 2020   added nodes.   
+00016c80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016c90: 2020 2020 2020 2020 2023 2020 2020 2020           #      
+00016ca0: 2077 696c 6c20 6265 2066 6c61 6767 6564   will be flagged
+00016cb0: 2062 7920 616e 2069 6e74 6567 6572 0a20   by an integer. 
+00016cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016cd0: 2020 2020 2020 2020 2020 2023 2020 2020             #    
+00016ce0: 2020 2074 6861 7420 6973 2031 2067 7265     that is 1 gre
+00016cf0: 6174 6572 2074 6861 6e20 7468 6520 6c61  ater than the la
+00016d00: 7267 6573 740a 2020 2020 2020 2020 2020  rgest.          
+00016d10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016d20: 2020 2320 2020 2020 2020 706f 7075 6c61    #       popula
+00016d30: 7469 6f6e 2d66 6c61 6720 696e 7465 6765  tion-flag intege
+00016d40: 7220 696e 0a20 2020 2020 2020 2020 2020  r in.           
+00016d50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016d60: 2023 2020 2020 2020 2074 6865 2070 7265   #       the pre
+00016d70: 2d69 6e74 726f 6475 6374 696f 6e20 5461  -introduction Ta
+00016d80: 626c 6543 6f6c 6c65 6374 696f 6e0a 2020  bleCollection.  
+00016d90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016da0: 2020 2020 2020 2020 2020 2320 2020 2020            #     
+00016db0: 2020 6f66 2074 6869 7320 5370 6563 6965    of this Specie
+00016dc0: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+00016dd0: 2020 2020 2020 2020 2020 2020 2020 6164                ad
+00016de0: 645f 706f 7075 6c61 7469 6f6e 733d 5472  d_populations=Tr
+00016df0: 7565 2c0a 2020 2020 2020 2020 2020 2020  ue,.            
+00016e00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016e10: 2320 4e4f 5445 3a20 6265 7374 2074 6f20  # NOTE: best to 
+00016e20: 6a75 7374 2061 6c6c 6f77 2069 7420 746f  just allow it to
+00016e30: 2064 6f20 7468 6973 2c0a 2020 2020 2020   do this,.      
+00016e40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016e50: 2020 2020 2020 2320 2020 2020 2020 696e        #       in
+00016e60: 2063 6173 6520 7765 2067 656e 6572 616c   case we general
+00016e70: 697a 6520 7468 696e 6773 206c 6174 6572  ize things later
+00016e80: 206f 6e2c 0a20 2020 2020 2020 2020 2020   on,.           
+00016e90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016ea0: 2023 2020 2020 2020 2062 7574 2066 6f72   #       but for
+00016eb0: 206e 6f77 2074 6865 7265 2077 696c 6c20   now there will 
+00016ec0: 6163 7475 616c 6c79 2062 650a 2020 2020  actually be.    
+00016ed0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016ee0: 2020 2020 2020 2020 2320 2020 2020 2020          #       
+00016ef0: 6e6f 7468 696e 6720 746f 2063 6865 636b  nothing to check
+00016f00: 2062 6563 6175 7365 2077 6520 6173 7375   because we assu
+00016f10: 6d65 200a 2020 2020 2020 2020 2020 2020  me .            
+00016f20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016f30: 2320 2020 2020 2020 6e6f 2073 6861 7265  #       no share
+00016f40: 6420 6e6f 6465 7320 6265 7477 6565 6e0a  d nodes between.
+00016f50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016f60: 2020 2020 2020 2020 2020 2020 2320 2020              #   
+00016f70: 2020 2020 736f 7572 6365 2061 6e64 2072      source and r
+00016f80: 6563 6970 6965 6e74 2054 6162 6c65 436f  ecipient TableCo
+00016f90: 6c6c 6563 7469 6f6e 730a 2020 2020 2020  llections.      
+00016fa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00016fb0: 2020 2020 2020 2320 544f 444f 3a20 484f        # TODO: HO
+00016fc0: 5745 5645 522c 2045 5155 414c 4954 5920  WEVER, EQUALITY 
+00016fd0: 4348 4543 4b20 5354 494c 4c20 4641 494c  CHECK STILL FAIL
+00016fe0: 5321 2057 4859 3f0a 2020 2020 2020 2020  S! WHY?.        
+00016ff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017000: 2020 2020 6368 6563 6b5f 7368 6172 6564      check_shared
+00017010: 5f65 7175 616c 6974 793d 4661 6c73 652c  _equality=False,
+00017020: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00017030: 2020 2020 2020 2020 2020 2020 2023 204e               # N
+00017040: 4f54 453a 2073 686f 756c 6420 636f 7079  OTE: should copy
+00017050: 2070 726f 7665 6e61 6e63 6520 696e 666f   provenance info
+00017060: 206f 7665 720a 2020 2020 2020 2020 2020   over.          
+00017070: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017080: 2020 2320 2020 2020 2020 6672 6f6d 2074    #       from t
+00017090: 6865 2073 6f75 7263 6520 496e 6469 7669  he source Indivi
+000170a0: 6475 616c 7327 2054 6162 6c65 436f 6c6c  duals' TableColl
+000170b0: 6563 7469 6f6e 0a20 2020 2020 2020 2020  ection.         
+000170c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000170d0: 2020 2023 2020 2020 2020 2074 6f20 7468     #       to th
+000170e0: 6174 206f 6620 7468 6520 7265 6365 6970  at of the receip
+000170f0: 6965 6e74 2053 7065 6369 6573 0a20 2020  ient Species.   
+00017100: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017110: 2020 2020 2020 2020 2072 6563 6f72 645f           record_
+00017120: 7072 6f76 656e 616e 6365 3d54 7275 652c  provenance=True,
+00017130: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
 00017140: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017150: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017160: 2074 7274 5f6e 756d 5d2e 7068 6929 2e6d   trt_num].phi).m
-00017170: 696e 2829 0a0a 2020 2020 2020 2020 2374  in()..        #t
-00017180: 6865 6e20 6765 7420 756e 6576 656e 2063  hen get uneven c
-00017190: 6d61 7020 616e 6420 6362 6172 2d6d 616b  map and cbar-mak
-000171a0: 6572 2028 6e65 6564 7320 746f 2062 6520  er (needs to be 
-000171b0: 756e 6576 656e 2074 6f0a 2020 2020 2020  uneven to.      
-000171c0: 2020 2367 6976 6520 636f 6c6f 722d 7265    #give color-re
-000171d0: 736f 6c75 7469 6f6e 2074 6f20 7661 6c75  solution to valu
-000171e0: 6573 2076 6172 7969 6e67 0a20 2020 2020  es varying.     
-000171f0: 2020 2023 6265 7477 6565 6e20 3120 616e     #between 1 an
-00017200: 6420 7468 6520 6d69 6e69 6d75 6d2d 6669  d the minimum-fi
-00017210: 746e 6573 7320 7661 6c75 6520 6173 7375  tness value assu
-00017220: 6d69 6e67 2061 6c6c 0a20 2020 2020 2020  ming all.       
-00017230: 2023 7068 656e 6f74 7970 6573 2061 7265   #phenotypes are
-00017240: 2063 6f6e 7374 7261 696e 6564 2030 3c3d   constrained 0<=
-00017250: 7a3c 3d31 2c20 6275 7420 7468 656e 2061  z<=1, but then a
-00017260: 6c73 6f0a 2020 2020 2020 2020 2361 6c6c  lso.        #all
-00017270: 6f77 2067 7261 6475 616c 6c79 2064 6565  ow gradually dee
-00017280: 7065 6e69 6e67 2072 6564 7320 666f 7220  pening reds for 
-00017290: 6669 746e 6573 7320 7661 6c75 6573 206c  fitness values l
-000172a0: 6f77 6572 0a20 2020 2020 2020 2023 7468  ower.        #th
-000172b0: 616e 2074 6861 7420 6d69 6e69 6d75 6d20  an that minimum 
-000172c0: 7661 6c75 6529 2c20 7573 696e 6720 7468  value), using th
-000172d0: 6520 6d69 6e5f 6669 7420 7661 6c0a 2020  e min_fit val.  
-000172e0: 2020 2020 2020 636d 6170 2c20 6d61 6b65        cmap, make
-000172f0: 5f63 6261 725f 666e 203d 2076 697a 2e5f  _cbar_fn = viz._
-00017300: 6d61 6b65 5f66 6974 6e65 7373 5f63 6d61  make_fitness_cma
-00017310: 705f 616e 645f 6362 6172 5f6d 616b 6572  p_and_cbar_maker
-00017320: 280a 2020 2020 2020 2020 2020 2020 6d69  (.            mi
-00017330: 6e5f 7661 6c20 3d20 6d69 6e5f 6669 742c  n_val = min_fit,
-00017340: 206d 6178 5f76 616c 203d 2031 2c20 636d   max_val = 1, cm
-00017350: 6170 203d 2066 6974 5f63 6d61 702c 0a20  ap = fit_cmap,. 
-00017360: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017370: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017380: 2020 2020 2020 2020 2020 2020 2020 2074                 t
-00017390: 7274 5f6e 756d 203d 2074 7274 5f6e 756d  rt_num = trt_num
-000173a0: 290a 0a20 2020 2020 2020 2023 706c 6f74  )..        #plot
-000173b0: 2074 6865 2074 7261 6974 2070 6865 6e6f   the trait pheno
-000173c0: 7479 7065 2069 6e20 6c61 7267 6572 2063  type in larger c
-000173d0: 6972 636c 6573 2066 6972 7374 2c20 6966  ircles first, if
-000173e0: 2074 7261 6974 2069 7320 6e6f 7420 4e6f   trait is not No
-000173f0: 6e65 0a20 2020 2020 2020 2069 6620 7472  ne.        if tr
-00017400: 745f 6e75 6d20 6973 206e 6f74 204e 6f6e  t_num is not Non
-00017410: 653a 0a20 2020 2020 2020 2020 2020 2023  e:.            #
-00017420: 706c 6f74 2074 6865 206f 7574 6572 2028  plot the outer (
-00017430: 7068 656e 6f74 7970 6529 2063 6972 636c  phenotype) circl
-00017440: 6573 0a20 2020 2020 2020 2020 2020 2073  es.            s
-00017450: 656c 662e 5f70 6c6f 745f 7068 656e 6f74  elf._plot_phenot
-00017460: 7970 6528 7472 6169 7420 3d20 7472 745f  ype(trait = trt_
-00017470: 6e75 6d2c 206c 7972 5f6e 756d 203d 206c  num, lyr_num = l
-00017480: 7972 5f6e 756d 2c0a 2020 2020 2020 2020  yr_num,.        
-00017490: 2020 2020 2020 2020 6c61 6e64 203d 206c          land = l
-000174a0: 616e 642c 2069 6e64 6976 6964 7320 3d20  and, individs = 
-000174b0: 696e 6469 7669 6473 2c20 7465 7874 203d  individs, text =
-000174c0: 2046 616c 7365 2c20 7369 7a65 203d 2073   False, size = s
-000174d0: 697a 652c 0a20 2020 2020 2020 2020 2020  ize,.           
-000174e0: 2020 2020 2074 6578 745f 7369 7a65 203d       text_size =
-000174f0: 2074 6578 745f 7369 7a65 2c20 6564 6765   text_size, edge
-00017500: 5f63 6f6c 6f72 3d65 6467 655f 636f 6c6f  _color=edge_colo
-00017510: 722c 0a20 2020 2020 2020 2020 2020 2020  r,.             
-00017520: 2020 2074 6578 745f 636f 6c6f 7220 3d20     text_color = 
-00017530: 7465 7874 5f63 6f6c 6f72 2c20 6362 6172  text_color, cbar
-00017540: 203d 2063 6261 722c 2061 6c70 6861 203d   = cbar, alpha =
-00017550: 2061 6c70 6861 2c0a 2020 2020 2020 2020   alpha,.        
-00017560: 2020 2020 2020 2020 7a6f 6f6d 5f77 6964          zoom_wid
-00017570: 7468 203d 207a 6f6f 6d5f 7769 6474 682c  th = zoom_width,
-00017580: 2078 203d 2078 2c20 7920 3d20 792c 2074   x = x, y = y, t
-00017590: 6963 6b73 3d74 6963 6b73 2c0a 2020 2020  icks=ticks,.    
-000175a0: 2020 2020 2020 2020 2020 2020 6d61 736b              mask
-000175b0: 5f72 6173 743d 6d61 736b 5f72 6173 7429  _rast=mask_rast)
-000175c0: 0a20 2020 2020 2020 2020 2020 2023 6d61  .            #ma
-000175d0: 6b65 2073 697a 6520 736d 616c 6c65 7220  ke size smaller 
-000175e0: 666f 7220 7468 6520 6e65 7874 206c 6179  for the next lay
-000175f0: 6572 206f 6620 696e 6e65 7220 2866 6974  er of inner (fit
-00017600: 6e65 7373 2920 6369 7263 6c65 730a 2020  ness) circles.  
-00017610: 2020 2020 2020 2020 2020 7369 7a65 203d            size =
-00017620: 2072 6f75 6e64 2830 2e34 2a73 697a 6529   round(0.4*size)
-00017630: 0a20 2020 2020 2020 2020 2020 2023 2067  .            # g
-00017640: 6574 2073 697a 6573 2066 6f72 2061 6c6c  et sizes for all
-00017650: 2069 6e64 6976 6964 7561 6c73 2720 696e   individuals' in
-00017660: 6e65 722d 6369 7263 6c65 2066 6974 6e65  ner-circle fitne
-00017670: 7373 2070 6f69 6e74 732c 2069 660a 2020  ss points, if.  
-00017680: 2020 2020 2020 2020 2020 2320 7472 745f            # trt_
-00017690: 6e75 6d20 6973 206e 6f74 204e 6f6e 650a  num is not None.
-000176a0: 2020 2020 2020 2020 2020 2020 6966 2074              if t
-000176b0: 7274 5f6e 756d 2069 7320 6e6f 7420 4e6f  rt_num is not No
-000176c0: 6e65 3a0a 2020 2020 2020 2020 2020 2020  ne:.            
-000176d0: 2020 2020 7369 7a65 203d 2073 697a 6520      size = size 
-000176e0: 2a20 2831 202d 2028 286e 702e 6172 7261  * (1 - ((np.arra
-000176f0: 7928 5b2a 772e 7661 6c75 6573 280a 2020  y([*w.values(.  
-00017700: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017710: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017720: 2020 2020 2020 2020 2020 295d 2920 2d20            )]) - 
-00017730: 6d69 6e5f 6669 7429 202f 2028 3120 2d20  min_fit) / (1 - 
-00017740: 6d69 6e5f 6669 7429 2929 0a0a 2020 2020  min_fit)))..    
-00017750: 2020 2020 7365 6c66 2e5f 706c 6f74 286c      self._plot(l
-00017760: 7972 5f6e 756d 3d6c 7972 5f6e 756d 2c20  yr_num=lyr_num, 
-00017770: 6c61 6e64 3d6c 616e 642c 2068 6964 655f  land=land, hide_
-00017780: 6c61 6e64 3d54 7275 652c 0a20 2020 2020  land=True,.     
-00017790: 2020 2020 2020 2020 2020 2020 2020 696e                in
-000177a0: 6469 7669 6473 3d69 6e64 6976 6964 732c  divids=individs,
-000177b0: 2074 6578 743d 7465 7874 2c20 636f 6c6f   text=text, colo
-000177c0: 723d 6c69 7374 2877 2e76 616c 7565 7328  r=list(w.values(
-000177d0: 2929 2c0a 2020 2020 2020 2020 2020 2020  )),.            
-000177e0: 2020 2020 2020 2070 745f 636d 6170 3d63         pt_cmap=c
-000177f0: 6d61 702c 2073 697a 653d 7369 7a65 2c20  map, size=size, 
-00017800: 6564 6765 5f63 6f6c 6f72 3d65 6467 655f  edge_color=edge_
-00017810: 636f 6c6f 722c 0a20 2020 2020 2020 2020  color,.         
-00017820: 2020 2020 2020 2020 2020 7465 7874 5f63            text_c
-00017830: 6f6c 6f72 3d74 6578 745f 636f 6c6f 722c  olor=text_color,
-00017840: 2063 6261 723d 6362 6172 2c20 7465 7874   cbar=cbar, text
-00017850: 5f73 697a 653d 7465 7874 5f73 697a 652c  _size=text_size,
-00017860: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00017870: 2020 2020 616c 7068 613d 616c 7068 612c      alpha=alpha,
-00017880: 207a 6f6f 6d5f 7769 6474 683d 7a6f 6f6d   zoom_width=zoom
-00017890: 5f77 6964 7468 2c20 783d 782c 2079 3d79  _width, x=x, y=y
-000178a0: 2c20 7469 636b 733d 7469 636b 732c 0a20  , ticks=ticks,. 
-000178b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000178c0: 2020 6d61 736b 5f72 6173 743d 6d61 736b    mask_rast=mask
-000178d0: 5f72 6173 7429 0a0a 2020 2020 2020 2020  _rast)..        
-000178e0: 2370 6c6f 7420 7068 656e 6f74 7970 6520  #plot phenotype 
-000178f0: 7465 7874 2028 776f 726b 7320 6f6e 6c79  text (works only
-00017900: 2069 6620 706c 6f74 7469 6e67 2061 2073   if plotting a s
-00017910: 7065 6369 6669 6320 7472 6169 7429 0a20  pecific trait). 
-00017920: 2020 2020 2020 2069 6620 7068 656e 6f74         if phenot
-00017930: 7970 655f 7465 7874 2061 6e64 2074 7274  ype_text and trt
-00017940: 5f6e 756d 2069 7320 6e6f 7420 4e6f 6e65  _num is not None
-00017950: 3a0a 2020 2020 2020 2020 2020 2020 666f  :.            fo
-00017960: 7220 696e 6420 696e 2073 656c 662e 7661  r ind in self.va
-00017970: 6c75 6573 2829 3a0a 2020 2020 2020 2020  lues():.        
-00017980: 2020 2020 2020 2020 706c 742e 7465 7874          plt.text
-00017990: 2869 6e64 2e78 2d30 2e35 2c20 696e 642e  (ind.x-0.5, ind.
-000179a0: 792d 302e 352c 2027 2530 2e32 6627 2025  y-0.5, '%0.2f' %
-000179b0: 2069 6e64 2e7a 5b74 7274 5f6e 756d 5d2c   ind.z[trt_num],
-000179c0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-000179d0: 2020 2020 2063 6f6c 6f72 203d 2070 6865       color = phe
-000179e0: 6e6f 7479 7065 5f74 6578 745f 636f 6c6f  notype_text_colo
-000179f0: 722c 2073 697a 6520 3d20 7465 7874 5f73  r, size = text_s
-00017a00: 697a 6529 0a0a 2020 2020 2020 2020 2370  ize)..        #p
-00017a10: 6c6f 7420 6669 746e 6573 7320 7465 7874  lot fitness text
-00017a20: 0a20 2020 2020 2020 2069 6620 6669 746e  .        if fitn
-00017a30: 6573 735f 7465 7874 3a0a 2020 2020 2020  ess_text:.      
-00017a40: 2020 2020 2020 6f66 6673 6574 5f66 726f        offset_fro
-00017a50: 6d5f 7068 656e 6f74 7970 655f 7465 7874  m_phenotype_text
-00017a60: 203d 2030 2e30 3031 2a6d 6178 2873 656c   = 0.001*max(sel
-00017a70: 662e 5f6c 616e 645f 6469 6d29 0a20 2020  f._land_dim).   
-00017a80: 2020 2020 2020 2020 2066 6f72 2069 6e64           for ind
-00017a90: 2069 6e20 7365 6c66 2e76 616c 7565 7328   in self.values(
-00017aa0: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-00017ab0: 2020 2070 6c74 2e74 6578 7428 696e 642e     plt.text(ind.
-00017ac0: 782d 302e 352b 6f66 6673 6574 5f66 726f  x-0.5+offset_fro
-00017ad0: 6d5f 7068 656e 6f74 7970 655f 7465 7874  m_phenotype_text
-00017ae0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00017af0: 2020 2020 2020 2020 2020 2069 6e64 2e79             ind.y
-00017b00: 2d30 2e35 2b6f 6666 7365 745f 6672 6f6d  -0.5+offset_from
-00017b10: 5f70 6865 6e6f 7479 7065 5f74 6578 742c  _phenotype_text,
-00017b20: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00017b30: 2020 2020 2020 2020 2020 2725 302e 3266            '%0.2f
-00017b40: 2720 2520 696e 642e 6669 742c 2063 6f6c  ' % ind.fit, col
-00017b50: 6f72 203d 2066 6974 6e65 7373 5f74 6578  or = fitness_tex
-00017b60: 745f 636f 6c6f 722c 0a20 2020 2020 2020  t_color,.       
-00017b70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017b80: 2020 7369 7a65 203d 2074 6578 745f 7369    size = text_si
-00017b90: 7a65 290a 0a20 2020 2020 2020 2023 616e  ze)..        #an
-00017ba0: 6420 6d61 6b65 2061 2063 6f6c 6f72 6261  d make a colorba
-00017bb0: 7220 666f 7220 7468 6520 6669 746e 6573  r for the fitnes
-00017bc0: 7320 7661 6c75 6573 200a 2020 2020 2020  s values .      
-00017bd0: 2020 6966 2066 6974 6e65 7373 5f63 6261    if fitness_cba
-00017be0: 723a 0a20 2020 2020 2020 2020 2020 2076  r:.            v
-00017bf0: 697a 2e5f 6d61 6b65 5f66 6974 6e65 7373  iz._make_fitness
-00017c00: 5f63 6261 7228 6d61 6b65 5f63 6261 725f  _cbar(make_cbar_
-00017c10: 666e 2c20 6d69 6e5f 6669 7429 0a0a 2020  fn, min_fit)..  
-00017c20: 2020 236d 6574 686f 6420 746f 2070 6c6f    #method to plo
-00017c30: 7420 6120 7370 6563 6965 7327 2061 6c6c  t a species' all
-00017c40: 656c 6520 6672 6571 7565 6e63 6965 730a  ele frequencies.
-00017c50: 2020 2020 6465 6620 5f70 6c6f 745f 616c      def _plot_al
-00017c60: 6c65 6c65 5f66 7265 7175 656e 6369 6573  lele_frequencies
-00017c70: 2873 656c 6629 3a0a 2020 2020 2020 2020  (self):.        
-00017c80: 6966 2073 656c 662e 6765 6e5f 6172 6368  if self.gen_arch
-00017c90: 2069 7320 4e6f 6e65 3a0a 2020 2020 2020   is None:.      
-00017ca0: 2020 2020 2020 7072 696e 7428 2822 5370        print(("Sp
-00017cb0: 6563 6965 732e 5f70 6c6f 745f 616c 6c65  ecies._plot_alle
-00017cc0: 6c65 5f66 7265 7175 656e 6369 6573 2069  le_frequencies i
-00017cd0: 7320 6e6f 7420 7661 6c69 6420 666f 7220  s not valid for 
-00017ce0: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
-00017cf0: 2020 2020 2022 7370 6563 6965 7320 7769       "species wi
-00017d00: 7468 6f75 7420 6765 6e6f 6d65 732e 5c6e  thout genomes.\n
-00017d10: 2229 290a 2020 2020 2020 2020 656c 7365  ")).        else
-00017d20: 3a0a 2020 2020 2020 2020 2020 2020 7365  :.            se
-00017d30: 6c66 2e67 656e 5f61 7263 682e 5f70 6c6f  lf.gen_arch._plo
-00017d40: 745f 616c 6c65 6c65 5f66 7265 7175 656e  t_allele_frequen
-00017d50: 6369 6573 2873 656c 6629 0a0a 2020 2020  cies(self)..    
-00017d60: 2320 6d65 7468 6f64 2066 6f72 2070 6c6f  # method for plo
-00017d70: 7474 696e 6720 6120 6869 7374 6f67 7261  tting a histogra
-00017d80: 6d20 6f66 2074 6865 2063 7572 7265 6e74  m of the current
-00017d90: 2066 6974 6e65 7373 2076 616c 7565 730a   fitness values.
-00017da0: 2020 2020 6465 6620 5f70 6c6f 745f 6869      def _plot_hi
-00017db0: 7374 5f66 6974 6e65 7373 2873 656c 6629  st_fitness(self)
-00017dc0: 3a0a 2020 2020 2020 2020 706c 742e 6869  :.        plt.hi
-00017dd0: 7374 286c 6973 7428 7365 6c66 2e5f 6361  st(list(self._ca
-00017de0: 6c63 5f66 6974 6e65 7373 2829 2929 0a20  lc_fitness())). 
-00017df0: 2020 2020 2020 2070 6c74 2e78 6c61 6265         plt.xlabe
-00017e00: 6c28 2746 6974 6e65 7373 2729 0a20 2020  l('Fitness').   
-00017e10: 2020 2020 2070 6c74 2e79 6c61 6265 6c28       plt.ylabel(
-00017e20: 2743 6f75 6e74 2729 0a0a 2020 2020 2320  'Count')..    # 
-00017e30: 6d65 7468 6f64 2066 6f72 2070 6c6f 7474  method for plott
-00017e40: 696e 6720 7468 6520 6d6f 7665 6d65 6e74  ing the movement
-00017e50: 2073 7572 6661 6365 2028 696e 2076 6172   surface (in var
-00017e60: 696f 7573 2066 6f72 6d61 7473 290a 2020  ious formats).  
-00017e70: 2020 6465 6620 5f70 6c6f 745f 6469 7265    def _plot_dire
-00017e80: 6374 696f 6e5f 7375 7266 6163 6528 7365  ction_surface(se
-00017e90: 6c66 2c20 6c61 6e64 2c20 7375 7266 5f74  lf, land, surf_t
-00017ea0: 7970 652c 2073 7479 6c65 2c20 783d 4e6f  ype, style, x=No
-00017eb0: 6e65 2c20 793d 4e6f 6e65 2c0a 2020 2020  ne, y=None,.    
-00017ec0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017ed0: 2020 2020 2020 2020 2020 2020 7a6f 6f6d              zoom
-00017ee0: 5f77 6964 7468 3d4e 6f6e 652c 2073 6361  _width=None, sca
-00017ef0: 6c65 5f66 6163 743d 342e 352c 0a20 2020  le_fact=4.5,.   
-00017f00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017f10: 2020 2020 2020 2020 2020 2020 2063 6f6c               col
-00017f20: 6f72 3d27 626c 6163 6b27 2c20 6362 6172  or='black', cbar
-00017f30: 3d54 7275 652c 2074 6963 6b73 3d4e 6f6e  =True, ticks=Non
-00017f40: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-00017f50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017f60: 2020 2063 6d61 703d 2770 6c61 736d 6127     cmap='plasma'
-00017f70: 2c20 6d61 736b 5f72 6173 743d 4e6f 6e65  , mask_rast=None
-00017f80: 293a 0a20 2020 2020 2020 2023 2067 6574  ):.        # get
-00017f90: 2074 6865 2063 6f72 7265 6374 2073 7572   the correct sur
-00017fa0: 6661 6365 0a20 2020 2020 2020 2069 6620  face.        if 
-00017fb0: 7375 7266 5f74 7970 6520 3d3d 2027 6d6f  surf_type == 'mo
-00017fc0: 7665 273a 0a20 2020 2020 2020 2020 2020  ve':.           
-00017fd0: 2073 7572 6620 3d20 7365 6c66 2e5f 6d6f   surf = self._mo
-00017fe0: 7665 5f73 7572 660a 2020 2020 2020 2020  ve_surf.        
-00017ff0: 656c 6966 2073 7572 665f 7479 7065 203d  elif surf_type =
-00018000: 3d20 2764 6973 7027 3a0a 2020 2020 2020  = 'disp':.      
-00018010: 2020 2020 2020 7375 7266 203d 3d20 7365        surf == se
-00018020: 6c66 2e5f 6469 7370 5f73 7572 660a 0a20  lf._disp_surf.. 
-00018030: 2020 2020 2020 2023 2067 6574 2061 6c6c         # get all
-00018040: 2078 2773 2061 6e64 2079 2773 2c20 6966   x's and y's, if
-00018050: 2078 2061 6e64 2079 2061 7265 204e 6f6e   x and y are Non
-00018060: 650a 2020 2020 2020 2020 6966 2078 2069  e.        if x i
-00018070: 7320 4e6f 6e65 2061 6e64 2079 2069 7320  s None and y is 
-00018080: 4e6f 6e65 3a0a 2020 2020 2020 2020 2020  None:.          
-00018090: 2020 7820 3d20 5b2a 7261 6e67 6528 6c61    x = [*range(la
-000180a0: 6e64 2e64 696d 5b30 5d29 5d0a 2020 2020  nd.dim[0])].    
-000180b0: 2020 2020 2020 2020 7920 3d20 5b2a 7261          y = [*ra
-000180c0: 6e67 6528 6c61 6e64 2e64 696d 5b31 5d29  nge(land.dim[1])
-000180d0: 5d0a 2020 2020 2020 2020 656c 7365 3a0a  ].        else:.
-000180e0: 2020 2020 2020 2020 2020 2020 7820 3d20              x = 
-000180f0: 5b78 5d0a 2020 2020 2020 2020 2020 2020  [x].            
-00018100: 7920 3d20 5b79 5d0a 2020 2020 2020 2020  y = [y].        
-00018110: 2363 6865 636b 2069 6620 7468 6520 7375  #check if the su
-00018120: 7266 6163 6520 6973 206e 6f6e 650a 2020  rface is none.  
-00018130: 2020 2020 2020 6966 2073 7572 6620 6973        if surf is
-00018140: 204e 6f6e 653a 0a20 2020 2020 2020 2020   None:.         
-00018150: 2020 2070 7269 6e74 2828 2746 756e 6374     print(('Funct
-00018160: 696f 6e20 6e6f 7420 7661 6c69 6420 666f  ion not valid fo
-00018170: 7220 6120 5370 6563 6965 7320 7769 7468  r a Species with
-00018180: 206e 6f20 270a 2020 2020 2020 2020 2020   no '.          
-00018190: 2020 2020 2020 2020 2027 5f25 7353 7572           '_%sSur
-000181a0: 6661 6365 2e27 2920 2520 287b 0a20 2020  face.') % ({.   
-000181b0: 2020 2020 2020 2020 2020 2020 2027 6d6f               'mo
-000181c0: 7665 273a 2027 4d6f 7665 6d65 6e74 272c  ve': 'Movement',
-000181d0: 2027 6469 7370 273a 2027 4469 7370 6572   'disp': 'Disper
-000181e0: 7361 6c27 7d5b 7375 7266 5f74 7970 655d  sal'}[surf_type]
-000181f0: 2929 0a20 2020 2020 2020 2020 2020 2072  )).            r
-00018200: 6574 7572 6e0a 2020 2020 2020 2020 656c  eturn.        el
-00018210: 6966 2073 7479 6c65 206e 6f74 2069 6e20  if style not in 
-00018220: 5b27 6869 7374 272c 2027 6368 6973 7427  ['hist', 'chist'
-00018230: 2c20 2776 6563 7427 2c20 2763 6472 6177  , 'vect', 'cdraw
-00018240: 7327 5d3a 0a20 2020 2020 2020 2020 2020  s']:.           
-00018250: 2070 7269 6e74 2828 2254 6865 2027 7374   print(("The 'st
-00018260: 796c 6527 2061 7267 756d 656e 7420 6d75  yle' argument mu
-00018270: 7374 2074 616b 6520 6f6e 6520 6f66 2074  st take one of t
-00018280: 6865 2022 0a20 2020 2020 2020 2020 2020  he ".           
-00018290: 2020 2020 2020 2020 2266 6f6c 6c6f 7769          "followi
-000182a0: 6e67 2076 616c 7565 733a 2027 6869 7374  ng values: 'hist
-000182b0: 272c 2027 6368 6973 7427 2c20 220a 2020  ', 'chist', ".  
-000182c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000182d0: 2022 2776 6563 7427 2c20 2763 6472 6177   "'vect', 'cdraw
-000182e0: 7327 2229 290a 2020 2020 2020 2020 2020  s'")).          
-000182f0: 2020 7265 7475 726e 0a20 2020 2020 2020    return.       
-00018300: 2065 6c69 6620 7374 796c 6520 3d3d 2027   elif style == '
-00018310: 6869 7374 273a 0a20 2020 2020 2020 2020  hist':.         
-00018320: 2020 2078 203d 2078 5b30 5d0a 2020 2020     x = x[0].    
-00018330: 2020 2020 2020 2020 7920 3d20 795b 305d          y = y[0]
-00018340: 0a20 2020 2020 2020 2020 2020 2070 6c74  .            plt
-00018350: 2e68 6973 7428 722e 6368 6f69 6365 2873  .hist(r.choice(s
-00018360: 7572 662e 7375 7266 5b79 2c78 2c3a 5d2c  urf.surf[y,x,:],
-00018370: 2073 697a 6520 3d20 3130 3030 302c 2072   size = 10000, r
-00018380: 6570 6c61 6365 203d 2054 7275 6529 2c0a  eplace = True),.
-00018390: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000183a0: 2020 2020 2062 696e 733d 3130 302c 2064       bins=100, d
-000183b0: 656e 7369 7479 3d54 7275 652c 2061 6c70  ensity=True, alp
-000183c0: 6861 3d30 2e35 2c20 636f 6c6f 723d 636f  ha=0.5, color=co
-000183d0: 6c6f 7229 0a0a 2020 2020 2020 2020 656c  lor)..        el
-000183e0: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-000183f0: 2364 6973 706c 6179 2074 6865 206d 6f76  #display the mov
-00018400: 656d 656e 742d 7375 7266 6163 6520 7261  ement-surface ra
-00018410: 7374 6572 0a20 2020 2020 2020 2020 2020  ster.           
-00018420: 206c 7972 5f6e 756d 203d 2073 7572 662e   lyr_num = surf.
-00018430: 6c79 725f 6e75 6d0a 2020 2020 2020 2020  lyr_num.        
-00018440: 2020 2020 6c61 6e64 5b6c 7972 5f6e 756d      land[lyr_num
-00018450: 5d2e 5f70 6c6f 7428 7a6f 6f6d 5f77 6964  ]._plot(zoom_wid
-00018460: 7468 203d 207a 6f6f 6d5f 7769 6474 682c  th = zoom_width,
-00018470: 2078 3d6e 702e 6d65 616e 2878 292c 0a20   x=np.mean(x),. 
-00018480: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018490: 2020 2020 2020 2020 2020 2020 2020 2079                 y
-000184a0: 3d6e 702e 6d65 616e 2879 292c 2074 6963  =np.mean(y), tic
-000184b0: 6b73 3d74 6963 6b73 2c20 636d 6170 3d63  ks=ticks, cmap=c
-000184c0: 6d61 702c 0a20 2020 2020 2020 2020 2020  map,.           
-000184d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000184e0: 2020 2020 206d 6173 6b5f 7261 7374 3d6d       mask_rast=m
-000184f0: 6173 6b5f 7261 7374 290a 0a20 2020 2020  ask_rast)..     
-00018500: 2020 2020 2020 2069 6620 7374 796c 6520         if style 
-00018510: 3d3d 2027 6368 6973 7427 3a0a 2020 2020  == 'chist':.    
-00018520: 2020 2020 2020 2020 2020 2020 666f 7220              for 
-00018530: 785f 7661 6c20 696e 2078 3a0a 2020 2020  x_val in x:.    
-00018540: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018550: 666f 7220 795f 7661 6c20 696e 2079 3a0a  for y_val in y:.
-00018560: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018570: 2020 2020 2020 2020 762c 2061 203d 206e          v, a = n
-00018580: 702e 6869 7374 6f67 7261 6d28 722e 6368  p.histogram(r.ch
-00018590: 6f69 6365 2873 7572 662e 7375 7266 5b79  oice(surf.surf[y
-000185a0: 5f76 616c 2c0a 2020 2020 2020 2020 2020  _val,.          
-000185b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000185c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000185d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000185e0: 2020 2020 2078 5f76 616c 2c20 3a5d 2c0a       x_val, :],.
-000185f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018600: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018610: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018620: 2020 2020 2072 6570 6c61 6365 3d54 7275       replace=Tru
-00018630: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-00018640: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017150: 2020 290a 2020 2020 2020 2020 6e6f 6465    ).        node
+00017160: 735f 7461 625f 6c65 6e5f 6166 203d 2073  s_tab_len_af = s
+00017170: 656c 662e 5f74 632e 6e6f 6465 732e 6e75  elf._tc.nodes.nu
+00017180: 6d5f 726f 7773 0a20 2020 2020 2020 2069  m_rows.        i
+00017190: 6e64 6976 6964 735f 7461 625f 6c65 6e5f  ndivids_tab_len_
+000171a0: 6166 203d 2073 656c 662e 5f74 632e 696e  af = self._tc.in
+000171b0: 6469 7669 6475 616c 732e 6e75 6d5f 726f  dividuals.num_ro
+000171c0: 7773 0a0a 2020 2020 2020 2020 2320 6368  ws..        # ch
+000171d0: 6563 6b20 756e 696f 6e65 6420 7461 626c  eck unioned tabl
+000171e0: 6573 2068 6176 6520 636f 7272 6563 7420  es have correct 
+000171f0: 6c65 6e67 7468 730a 2020 2020 2020 2020  lengths.        
+00017200: 6173 7365 7274 2028 6e6f 6465 735f 7461  assert (nodes_ta
+00017210: 625f 6c65 6e5f 6166 202d 206e 6f64 6573  b_len_af - nodes
+00017220: 5f74 6162 5f6c 656e 5f62 3429 203d 3d20  _tab_len_b4) == 
+00017230: 736f 7572 6365 5f74 632e 6e6f 6465 732e  source_tc.nodes.
+00017240: 6e75 6d5f 726f 7773 0a20 2020 2020 2020  num_rows.       
+00017250: 2061 7373 6572 7420 2828 696e 6469 7669   assert ((indivi
+00017260: 6473 5f74 6162 5f6c 656e 5f61 6620 2d20  ds_tab_len_af - 
+00017270: 696e 6469 7669 6473 5f74 6162 5f6c 656e  individs_tab_len
+00017280: 5f62 3429 203d 3d0a 2020 2020 2020 2020  _b4) ==.        
+00017290: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000172a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000172b0: 2020 2020 2020 2020 736f 7572 6365 5f74          source_t
+000172c0: 632e 696e 6469 7669 6475 616c 732e 6e75  c.individuals.nu
+000172d0: 6d5f 726f 7773 290a 0a20 2020 2020 2020  m_rows)..       
+000172e0: 2023 2063 6865 636b 2074 6861 7420 616c   # check that al
+000172f0: 6c20 676e 7820 6964 7873 2073 746f 7265  l gnx idxs store
+00017300: 6420 696e 2074 6865 2069 6e64 6976 6964  d in the individ
+00017310: 7561 6c73 2074 6162 6c65 2773 0a20 2020  uals table's.   
+00017320: 2020 2020 2023 206d 6574 6164 6174 6120       # metadata 
+00017330: 636f 6c75 6d6e 206f 6363 7572 206f 6e6c  column occur onl
+00017340: 7920 6f6e 6365 2028 6173 6964 6520 6672  y once (aside fr
+00017350: 6f6d 2030 2c0a 2020 2020 2020 2020 2320  om 0,.        # 
+00017360: 7768 6963 6820 6973 2061 206e 756c 6c20  which is a null 
+00017370: 7661 6c75 6520 666f 7220 696e 6469 7669  value for indivi
+00017380: 6475 616c 7320 7468 6174 0a20 2020 2020  duals that.     
+00017390: 2020 2023 206e 6576 6572 2065 7869 7374     # never exist
+000173a0: 6564 2061 7320 2772 6561 6c27 2067 6e78  ed as 'real' gnx
+000173b0: 2049 6e64 6976 6964 7561 6c20 6f62 6a65   Individual obje
+000173c0: 6374 7320 6265 6361 7573 650a 2020 2020  cts because.    
+000173d0: 2020 2020 2320 7468 6579 2077 6572 6520      # they were 
+000173e0: 696e 2074 6865 2070 6173 7420 6f66 2061  in the past of a
+000173f0: 2063 6f61 6c65 7363 656e 7420 7369 6d75   coalescent simu
+00017400: 6c61 7469 6f6e 290a 2020 2020 2020 2020  lation).        
+00017410: 7265 6369 705f 7463 5f69 6e64 6976 6964  recip_tc_individ
+00017420: 7561 6c73 5f69 6473 203d 205b 696e 742e  uals_ids = [int.
+00017430: 6672 6f6d 5f62 7974 6573 280a 2020 2020  from_bytes(.    
+00017440: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017450: 2020 2020 2020 2020 7365 6c66 2e5f 7463          self._tc
+00017460: 2e69 6e64 6976 6964 7561 6c73 5b69 5d2e  .individuals[i].
+00017470: 6d65 7461 6461 7461 2c20 276c 6974 746c  metadata, 'littl
+00017480: 6527 2920 666f 720a 2020 2020 2020 2020  e') for.        
+00017490: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000174a0: 2020 2020 2020 2020 6920 696e 2072 616e          i in ran
+000174b0: 6765 2873 656c 662e 5f74 632e 696e 6469  ge(self._tc.indi
+000174c0: 7669 6475 616c 732e 6e75 6d5f 726f 7773  viduals.num_rows
+000174d0: 295d 0a20 2020 2020 2020 2061 7373 6572  )].        asser
+000174e0: 7420 286c 656e 2872 6563 6970 5f74 635f  t (len(recip_tc_
+000174f0: 696e 6469 7669 6475 616c 735f 6964 7329  individuals_ids)
+00017500: 203d 3d0a 2020 2020 2020 2020 2020 2020   ==.            
+00017510: 2020 2020 7365 6c66 2e5f 7463 2e69 6e64      self._tc.ind
+00017520: 6976 6964 7561 6c73 2e6e 756d 5f72 6f77  ividuals.num_row
+00017530: 7329 0a20 2020 2020 2020 2063 6f75 6e74  s).        count
+00017540: 7320 3d20 4328 7265 6369 705f 7463 5f69  s = C(recip_tc_i
+00017550: 6e64 6976 6964 7561 6c73 5f69 6473 290a  ndividuals_ids).
+00017560: 2020 2020 2020 2020 666f 7220 6b2c 2076          for k, v
+00017570: 2069 6e20 636f 756e 7473 2e69 7465 6d73   in counts.items
+00017580: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
+00017590: 6966 206b 2021 3d20 303a 0a20 2020 2020  if k != 0:.     
+000175a0: 2020 2020 2020 2020 2020 2061 7373 6572             asser
+000175b0: 7420 7620 3d3d 2031 2c20 2866 2267 6e78  t v == 1, (f"gnx
+000175c0: 2049 6e64 6976 6964 7561 6c20 6964 7820   Individual idx 
+000175d0: 7b69 7d20 6f63 6375 7273 206d 6f72 6520  {i} occurs more 
+000175e0: 7468 616e 206f 6e63 6520 220a 2020 2020  than once ".    
+000175f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017600: 2020 2020 2020 2020 2020 2020 2022 696e               "in
+00017610: 2074 6865 206d 6574 6164 6174 6120 636f   the metadata co
+00017620: 6c75 6d6e 206f 6620 7468 6520 756e 696f  lumn of the unio
+00017630: 6e20 6f66 2022 0a20 2020 2020 2020 2020  n of ".         
+00017640: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017650: 2020 2020 2020 2020 2274 6865 2072 6563          "the rec
+00017660: 6970 6965 6e74 2061 6e64 2073 6f75 7263  ipient and sourc
+00017670: 6520 5370 6563 6965 7327 2022 0a20 2020  e Species' ".   
+00017680: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017690: 2020 2020 2020 2020 2020 2020 2020 2254                "T
+000176a0: 6162 6c65 436f 6c6c 6563 7469 6f6e 7320  ableCollections 
+000176b0: 696e 6469 7669 6475 616c 7320 7461 626c  individuals tabl
+000176c0: 6573 2e22 290a 0a20 2020 2020 2020 2023  es.")..        #
+000176d0: 2061 6464 2069 6e20 7468 6520 496e 6469   add in the Indi
+000176e0: 7669 6475 616c 732c 206b 6579 6564 2062  viduals, keyed b
+000176f0: 7920 6964 780a 2020 2020 2020 2020 706f  y idx.        po
+00017700: 705f 7369 7a65 5f62 3420 3d20 6c65 6e28  p_size_b4 = len(
+00017710: 7365 6c66 290a 2020 2020 2020 2020 666f  self).        fo
+00017720: 7220 696e 6420 696e 2069 6e64 6976 6964  r ind in individ
+00017730: 733a 0a20 2020 2020 2020 2020 2020 2073  s:.            s
+00017740: 656c 665b 696e 642e 6964 785d 203d 2069  elf[ind.idx] = i
+00017750: 6e64 0a20 2020 2020 2020 2070 6f70 5f73  nd.        pop_s
+00017760: 697a 655f 6166 203d 206c 656e 2873 656c  ize_af = len(sel
+00017770: 6629 0a20 2020 2020 2020 2061 7373 6572  f).        asser
+00017780: 7420 2870 6f70 5f73 697a 655f 6166 202d  t (pop_size_af -
+00017790: 2070 6f70 5f73 697a 655f 6234 2920 3d3d   pop_size_b4) ==
+000177a0: 206c 656e 2869 6e64 6976 6964 7329 2c20   len(individs), 
+000177b0: 2822 496e 636f 7272 6563 7420 6e75 6d62  ("Incorrect numb
+000177c0: 6572 220a 2020 2020 2020 2020 2020 2020  er".            
+000177d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000177e0: 2020 2020 2220 6f66 2049 6e64 6976 6964      " of Individ
+000177f0: 7561 6c73 2061 6464 6564 2074 6f20 7265  uals added to re
+00017800: 6369 7069 656e 7420 5370 6563 6965 732e  cipient Species.
+00017810: 2229 0a0a 2020 2020 2020 2020 2320 7570  ")..        # up
+00017820: 6461 7465 2065 6163 6820 496e 6469 7669  date each Indivi
+00017830: 6475 616c 2773 205f 696e 6469 7669 6475  dual's _individu
+00017840: 616c 735f 7461 625f 6964 2061 6e64 205f  als_tab_id and _
+00017850: 6e6f 6465 735f 7461 625f 6964 730a 2020  nodes_tab_ids.  
+00017860: 2020 2020 2020 2320 746f 206d 6174 6368        # to match
+00017870: 2074 6865 206e 6577 2c20 756e 696f 6e65   the new, unione
+00017880: 6420 5461 626c 6543 6f6c 6c65 6374 696f  d TableCollectio
+00017890: 6e0a 2020 2020 2020 2020 6e6f 6465 735f  n.        nodes_
+000178a0: 7461 625f 696e 6469 7669 6475 616c 7320  tab_individuals 
+000178b0: 3d20 7365 6c66 2e5f 7463 2e6e 6f64 6573  = self._tc.nodes
+000178c0: 2e69 6e64 6976 6964 7561 6c0a 2020 2020  .individual.    
+000178d0: 2020 2020 666f 7220 6964 782c 2069 6e64      for idx, ind
+000178e0: 2069 6e20 7365 6c66 2e69 7465 6d73 2829   in self.items()
+000178f0: 3a0a 2020 2020 2020 2020 2020 2020 696e  :.            in
+00017900: 6469 7669 6475 616c 735f 7461 625f 6964  dividuals_tab_id
+00017910: 203d 206e 702e 6172 6777 6865 7265 280a   = np.argwhere(.
+00017920: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017930: 2020 2020 2020 2020 2020 2020 6e70 2e61              np.a
+00017940: 7272 6179 2872 6563 6970 5f74 635f 696e  rray(recip_tc_in
+00017950: 6469 7669 6475 616c 735f 6964 7329 203d  dividuals_ids) =
+00017960: 3d20 6964 7829 2e72 6176 656c 2829 0a20  = idx).ravel(). 
+00017970: 2020 2020 2020 2020 2020 2061 7373 6572             asser
+00017980: 7420 6c65 6e28 696e 6469 7669 6475 616c  t len(individual
+00017990: 735f 7461 625f 6964 2920 3d3d 2031 0a20  s_tab_id) == 1. 
+000179a0: 2020 2020 2020 2020 2020 2069 6e64 6976             indiv
+000179b0: 6964 7561 6c73 5f74 6162 5f69 6420 3d20  iduals_tab_id = 
+000179c0: 696e 6469 7669 6475 616c 735f 7461 625f  individuals_tab_
+000179d0: 6964 5b30 5d0a 2020 2020 2020 2020 2020  id[0].          
+000179e0: 2020 696e 642e 5f69 6e64 6976 6964 7561    ind._individua
+000179f0: 6c73 5f74 6162 5f69 6420 3d20 696e 6469  ls_tab_id = indi
+00017a00: 7669 6475 616c 735f 7461 625f 6964 0a20  viduals_tab_id. 
+00017a10: 2020 2020 2020 2020 2020 206e 6f64 6573             nodes
+00017a20: 5f74 6162 5f69 6473 203d 206e 702e 6172  _tab_ids = np.ar
+00017a30: 6777 6865 7265 286e 6f64 6573 5f74 6162  gwhere(nodes_tab
+00017a40: 5f69 6e64 6976 6964 7561 6c73 203d 3d0a  _individuals ==.
+00017a50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017a60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017a70: 2020 2020 2020 2020 696e 6469 7669 6475          individu
+00017a80: 616c 735f 7461 625f 6964 292e 7261 7665  als_tab_id).rave
+00017a90: 6c28 290a 2020 2020 2020 2020 2020 2020  l().            
+00017aa0: 6173 7365 7274 206c 656e 286e 6f64 6573  assert len(nodes
+00017ab0: 5f74 6162 5f69 6473 2920 3d3d 2032 0a20  _tab_ids) == 2. 
+00017ac0: 2020 2020 2020 2020 2020 2069 6e64 2e5f             ind._
+00017ad0: 6e6f 6465 735f 7461 625f 6964 7320 3d20  nodes_tab_ids = 
+00017ae0: 7b69 3a20 6964 2066 6f72 2069 2c20 6964  {i: id for i, id
+00017af0: 2069 6e20 656e 756d 6572 6174 6528 6e6f   in enumerate(no
+00017b00: 6465 735f 7461 625f 6964 7329 7d0a 0a20  des_tab_ids)}.. 
+00017b10: 2020 2020 2020 2023 2066 696e 616c 2063         # final c
+00017b20: 6865 636b 2074 6861 7420 696e 6469 6365  heck that indice
+00017b30: 7320 6368 6563 6b20 6f75 7420 696e 2074  s check out in t
+00017b40: 6865 206f 7468 6572 2064 6972 6563 7469  he other directi
+00017b50: 6f6e 0a20 2020 2020 2020 2023 2028 692e  on.        # (i.
+00017b60: 652e 2c20 6672 6f6d 206e 6f64 6573 2074  e., from nodes t
+00017b70: 6162 2074 6f20 696e 6469 7669 6475 616c  ab to individual
+00017b80: 7320 7461 6220 746f 2067 6e78 2049 6e64  s tab to gnx Ind
+00017b90: 6976 6964 7561 6c73 0a20 2020 2020 2020  ividuals.       
+00017ba0: 2023 2061 6e64 2074 6861 7420 6e6f 6465   # and that node
+00017bb0: 7320 6964 7320 6368 6563 6b20 6f75 7420  s ids check out 
+00017bc0: 746f 6f0a 2020 2020 2020 2020 6375 7272  too.        curr
+00017bd0: 5f6e 6f64 655f 6374 203d 2030 0a20 2020  _node_ct = 0.   
+00017be0: 2020 2020 2066 6f72 206e 6f64 655f 6964       for node_id
+00017bf0: 2c20 6e6f 6465 2069 6e20 656e 756d 6572  , node in enumer
+00017c00: 6174 6528 7365 6c66 2e5f 7463 2e6e 6f64  ate(self._tc.nod
+00017c10: 6573 293a 0a20 2020 2020 2020 2020 2020  es):.           
+00017c20: 2023 204e 4f54 453a 2069 676e 6f72 6520   # NOTE: ignore 
+00017c30: 2d31 733b 2074 6865 7920 6172 6520 6d73  -1s; they are ms
+00017c40: 7072 696d 6520 666c 6167 7320 666f 7220  prime flags for 
+00017c50: 6e6f 6e2d 7361 6d70 6c65 206e 6f64 6573  non-sample nodes
+00017c60: 2062 7574 0a20 2020 2020 2020 2020 2020   but.           
+00017c70: 2023 2020 2020 2020 2062 6561 6876 6520   #       beahve 
+00017c80: 6173 2061 2050 7974 686f 6e20 696e 6465  as a Python inde
+00017c90: 7865 7220 6f66 2074 6865 206c 6173 7420  xer of the last 
+00017ca0: 496e 6469 7669 6475 616c 2069 6e20 7468  Individual in th
+00017cb0: 650a 2020 2020 2020 2020 2020 2020 2320  e.            # 
+00017cc0: 2020 2020 2020 5370 6563 6965 7320 616e        Species an
+00017cd0: 6420 7468 7573 2074 6872 6f77 2061 6e20  d thus throw an 
+00017ce0: 6572 726f 720a 2020 2020 2020 2020 2020  error.          
+00017cf0: 2020 6966 206e 6f64 652e 696e 6469 7669    if node.indivi
+00017d00: 6475 616c 2021 3d20 2d31 3a0a 2020 2020  dual != -1:.    
+00017d10: 2020 2020 2020 2020 2020 2020 676e 785f              gnx_
+00017d20: 696e 6469 7669 6475 616c 5f69 6420 3d20  individual_id = 
+00017d30: 696e 742e 6672 6f6d 5f62 7974 6573 2873  int.from_bytes(s
+00017d40: 656c 662e 5f74 632e 696e 6469 7669 6475  elf._tc.individu
+00017d50: 616c 735b 0a20 2020 2020 2020 2020 2020  als[.           
+00017d60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017d70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017d80: 206e 6f64 652e 696e 6469 7669 6475 616c   node.individual
+00017d90: 5d2e 6d65 7461 6461 7461 2c20 276c 6974  ].metadata, 'lit
+00017da0: 746c 6527 290a 2020 2020 2020 2020 2020  tle').          
+00017db0: 2020 2020 2020 6966 2067 6e78 5f69 6e64        if gnx_ind
+00017dc0: 6976 6964 7561 6c5f 6964 2069 6e20 5b2a  ividual_id in [*
+00017dd0: 7365 6c66 5d3a 0a20 2020 2020 2020 2020  self]:.         
+00017de0: 2020 2020 2020 2020 2020 2063 7572 725f             curr_
+00017df0: 6e6f 6465 5f63 7420 2b3d 2031 0a20 2020  node_ct += 1.   
+00017e00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017e10: 2061 7373 6572 7420 2873 656c 665b 676e   assert (self[gn
+00017e20: 785f 696e 6469 7669 6475 616c 5f69 645d  x_individual_id]
+00017e30: 2e5f 696e 6469 7669 6475 616c 735f 7461  ._individuals_ta
+00017e40: 625f 6964 203d 3d0a 2020 2020 2020 2020  b_id ==.        
+00017e50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017e60: 2020 2020 6e6f 6465 2e69 6e64 6976 6964      node.individ
+00017e70: 7561 6c20 616e 6420 6e6f 6465 5f69 6420  ual and node_id 
+00017e80: 696e 0a20 2020 2020 2020 2020 2020 2020  in.             
+00017e90: 2020 2020 2020 2020 2020 2020 2020 2073                 s
+00017ea0: 656c 665b 676e 785f 696e 6469 7669 6475  elf[gnx_individu
+00017eb0: 616c 5f69 645d 2e5f 6e6f 6465 735f 7461  al_id]._nodes_ta
+00017ec0: 625f 6964 732e 7661 6c75 6573 2829 290a  b_ids.values()).
+00017ed0: 2020 2020 2020 2020 6173 7365 7274 2063          assert c
+00017ee0: 7572 725f 6e6f 6465 5f63 7420 3d3d 2032  urr_node_ct == 2
+00017ef0: 2a6c 656e 2873 656c 6629 2c20 2822 4e75  *len(self), ("Nu
+00017f00: 6d62 6572 206f 6620 6375 7272 656e 7420  mber of current 
+00017f10: 496e 6469 7669 6475 616c 7320 220a 2020  Individuals ".  
+00017f20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017f30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017f40: 2020 2020 2020 2020 2020 2020 2022 696e               "in
+00017f50: 2074 6865 2072 6563 6970 6965 6e74 2053   the recipient S
+00017f60: 7065 6369 6573 2064 6f65 7320 6e6f 7420  pecies does not 
+00017f70: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
+00017f80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017f90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017fa0: 2022 6d61 7463 6820 7468 6520 6e75 6d62   "match the numb
+00017fb0: 6572 2065 7870 6563 7465 6420 6261 7365  er expected base
+00017fc0: 6420 6f6e 2022 0a20 2020 2020 2020 2020  d on ".         
+00017fd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017fe0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017ff0: 2020 2020 2020 2254 6162 6c65 436f 6c6c        "TableColl
+00018000: 6563 7469 6f6e 2063 6f6e 7465 6e74 732e  ection contents.
+00018010: 2229 0a0a 2020 2020 2020 2020 2320 736f  ")..        # so
+00018020: 7274 2061 6e64 2073 696d 706c 6966 7920  rt and simplify 
+00018030: 7468 6520 7473 6b69 742e 5461 626c 6543  the tskit.TableC
+00018040: 6f6c 6c65 6374 696f 6e20 6167 6169 6e0a  ollection again.
+00018050: 2020 2020 2020 2020 7365 6c66 2e5f 736f          self._so
+00018060: 7274 5f61 6e64 5f73 696d 706c 6966 795f  rt_and_simplify_
+00018070: 7461 626c 655f 636f 6c6c 6563 7469 6f6e  table_collection
+00018080: 2829 0a0a 2020 2020 2020 2020 2320 7265  ()..        # re
+00018090: 7365 7420 7468 6520 636f 6f72 6473 2061  set the coords a
+000180a0: 6e64 2063 656c 6c73 2061 7474 7273 0a20  nd cells attrs. 
+000180b0: 2020 2020 2020 2073 656c 662e 5f73 6574         self._set
+000180c0: 5f63 6f6f 7264 735f 616e 645f 6365 6c6c  _coords_and_cell
+000180d0: 7328 290a 0a20 2020 2020 2020 2023 2072  s()..        # r
+000180e0: 6573 6574 2074 6865 204b 4454 7265 6520  eset the KDTree 
+000180f0: 6166 7465 7220 6164 6469 6e67 2074 6865  after adding the
+00018100: 206e 6577 2069 6e64 6976 6964 7561 6c73   new individuals
+00018110: 0a20 2020 2020 2020 2073 656c 662e 5f73  .        self._s
+00018120: 6574 5f6b 645f 7472 6565 2829 0a0a 2020  et_kd_tree()..  
+00018130: 2020 2020 2020 2320 7265 7365 7420 7468        # reset th
+00018140: 6520 6465 6e73 6974 7920 6772 6964 730a  e density grids.
+00018150: 2020 2020 2020 2020 7365 6c66 2e5f 7365          self._se
+00018160: 745f 6465 6e73 5f67 7269 6473 286c 616e  t_dens_grids(lan
+00018170: 6429 0a0a 2020 2020 2020 2020 2320 7265  d)..        # re
+00018180: 7365 742f 7365 7420 736f 7572 6365 2049  set/set source I
+00018190: 6e64 6976 6964 7561 6c73 2720 656e 7669  ndividuals' envi
+000181a0: 726f 6e6d 656e 7461 6c20 616e 6420 6669  ronmental and fi
+000181b0: 746e 6573 7320 7661 6c75 6573 0a20 2020  tness values.   
+000181c0: 2020 2020 2073 656c 662e 5f73 6574 5f65       self._set_e
+000181d0: 286c 616e 6429 0a0a 2020 2020 2020 2020  (land)..        
+000181e0: 2372 6573 6574 2070 6865 6e6f 7479 7065  #reset phenotype
+000181f0: 7320 616e 6420 6669 746e 6573 7365 730a  s and fitnesses.
+00018200: 2020 2020 2020 2020 6966 2073 656c 662e          if self.
+00018210: 6765 6e5f 6172 6368 2069 7320 6e6f 7420  gen_arch is not 
+00018220: 4e6f 6e65 2061 6e64 2073 656c 662e 6765  None and self.ge
+00018230: 6e5f 6172 6368 2e74 7261 6974 7320 6973  n_arch.traits is
+00018240: 206e 6f74 204e 6f6e 653a 0a20 2020 2020   not None:.     
+00018250: 2020 2020 2020 205b 696e 642e 5f73 6574         [ind._set
+00018260: 5f7a 2873 656c 662e 6765 6e5f 6172 6368  _z(self.gen_arch
+00018270: 2920 666f 7220 696e 6420 696e 2073 656c  ) for ind in sel
+00018280: 662e 7661 6c75 6573 2829 5d0a 2020 2020  f.values()].    
+00018290: 2020 2020 2020 2020 7365 6c66 2e5f 6361          self._ca
+000182a0: 6c63 5f66 6974 6e65 7373 2873 6574 5f66  lc_fitness(set_f
+000182b0: 6974 3d54 7275 6529 0a0a 2020 2020 2020  it=True)..      
+000182c0: 2020 2320 7072 696e 7420 696e 666f 726d    # print inform
+000182d0: 6174 6976 6520 6f75 7470 7574 0a20 2020  ative output.   
+000182e0: 2020 2020 2070 7269 6e74 2866 225c 6e7b       print(f"\n{
+000182f0: 6c65 6e28 696e 6469 7669 6473 297d 2049  len(individs)} I
+00018300: 6e64 6976 6964 7561 6c73 2073 7563 6365  ndividuals succe
+00018310: 7373 6675 6c6c 7920 220a 2020 2020 2020  ssfully ".      
+00018320: 2020 2020 2020 2020 6622 6164 6465 6420          f"added 
+00018330: 746f 2053 7065 6369 6573 207b 7365 6c66  to Species {self
+00018340: 2e69 6478 7d20 2827 7b73 656c 662e 6e61  .idx} ('{self.na
+00018350: 6d65 7d27 292e 5c6e 2229 0a0a 0a20 2020  me}').\n")...   
+00018360: 2064 6566 205f 696e 6974 5f6d 7370 7269   def _init_mspri
+00018370: 6d65 5f70 6f70 2873 656c 662c 206c 616e  me_pop(self, lan
+00018380: 642c 206d 7370 7269 6d65 5f69 6e69 745f  d, msprime_init_
+00018390: 7061 7261 6d73 293a 0a20 2020 2020 2020  params):.       
+000183a0: 2023 2067 6574 2069 6e74 656e 6465 6420   # get intended 
+000183b0: 746f 7461 6c20 696e 6974 6961 6c20 706f  total initial po
+000183c0: 7020 7369 7a65 0a20 2020 2020 2020 2074  p size.        t
+000183d0: 6172 6765 745f 7369 7a65 203d 206e 702e  arget_size = np.
+000183e0: 7375 6d28 5b5b 2a76 2e6b 6579 7328 295d  sum([[*v.keys()]
+000183f0: 5b30 5d20 666f 720a 2020 2020 2020 2020  [0] for.        
+00018400: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018410: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018420: 2020 2020 7620 696e 206d 7370 7269 6d65      v in msprime
+00018430: 5f69 6e69 745f 7061 7261 6d73 2e76 616c  _init_params.val
+00018440: 7565 7328 295d 290a 2020 2020 2020 2020  ues()]).        
+00018450: 2320 7265 6d6f 7665 2061 6c6c 2063 7572  # remove all cur
+00018460: 7265 6e74 2069 6e64 6976 6964 730a 2020  rent individs.  
+00018470: 2020 2020 2020 7365 6c66 2e5f 7265 6d6f        self._remo
+00018480: 7665 5f69 6e64 6976 6964 7561 6c73 286e  ve_individuals(n
+00018490: 3d6c 656e 2873 656c 6629 2c0a 2020 2020  =len(self),.    
+000184a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000184b0: 2020 2020 2020 2020 2020 2020 206b 6565               kee
+000184c0: 705f 7369 7465 735f 7461 623d 4661 6c73  p_sites_tab=Fals
+000184d0: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+000184e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000184f0: 2020 2029 0a20 2020 2020 2020 2061 7373     ).        ass
+00018500: 6572 7420 6c65 6e28 7365 6c66 2920 3d3d  ert len(self) ==
+00018510: 2030 0a20 2020 2020 2020 2023 206c 6f6f   0.        # loo
+00018520: 7020 6f76 6572 2073 6f75 7263 6520 706f  p over source po
+00018530: 7073 2061 6e64 2061 6464 2069 6e64 6976  ps and add indiv
+00018540: 6964 7561 6c73 2066 726f 6d20 6561 6368  iduals from each
+00018550: 206f 6e65 0a20 2020 2020 2020 2066 6f72   one.        for
+00018560: 2070 6f70 2069 6e20 6d73 7072 696d 655f   pop in msprime_
+00018570: 696e 6974 5f70 6172 616d 732e 7661 6c75  init_params.valu
+00018580: 6573 2829 3a0a 2020 2020 2020 2020 2020  es():.          
+00018590: 2020 6173 7365 7274 206c 656e 2870 6f70    assert len(pop
+000185a0: 2e6b 6579 7328 2929 203d 3d20 312c 2028  .keys()) == 1, (
+000185b0: 2245 6163 6820 6d73 7072 696d 6520 736f  "Each msprime so
+000185c0: 7572 6365 2070 6f70 756c 6174 696f 6e20  urce population 
+000185d0: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
+000185e0: 2020 2020 2020 2020 2020 2270 6172 616d            "param
+000185f0: 6574 6572 697a 6564 2069 6e20 6120 5370  eterized in a Sp
+00018600: 6563 6965 7327 2027 696e 6974 2720 7365  ecies' 'init' se
+00018610: 6374 696f 6e20 220a 2020 2020 2020 2020  ction ".        
+00018620: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018630: 226f 6620 7468 6520 7061 7261 6d65 7465  "of the paramete
+00018640: 7273 2066 696c 6522 0a20 2020 2020 2020  rs file".       
 00018650: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018660: 2020 2020 2020 2020 7369 7a65 3d37 3530          size=750
-00018670: 3029 2c20 6269 6e73 3d31 3529 0a20 2020  0), bins=15).   
-00018680: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018690: 2020 2020 2076 203d 2076 202f 2066 6c6f       v = v / flo
-000186a0: 6174 2876 2e73 756d 2829 290a 2020 2020  at(v.sum()).    
-000186b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000186c0: 2020 2020 6120 3d20 5b28 615b 6e5d 202b      a = [(a[n] +
-000186d0: 2061 5b6e 202b 2031 5d29 202f 2032 2066   a[n + 1]) / 2 f
-000186e0: 6f72 206e 2069 6e20 7261 6e67 6528 6c65  or n in range(le
-000186f0: 6e28 6129 202d 2031 295d 0a20 2020 2020  n(a) - 1)].     
-00018700: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018710: 2020 2078 7320 3d20 5b6e 702e 636f 7328     xs = [np.cos(
-00018720: 615b 6e5d 2920 2a20 302e 3735 2066 6f72  a[n]) * 0.75 for
-00018730: 206e 2069 6e20 7261 6e67 6528 6c65 6e28   n in range(len(
-00018740: 6129 295d 0a20 2020 2020 2020 2020 2020  a))].           
-00018750: 2020 2020 2020 2020 2020 2020 2079 7320               ys 
-00018760: 3d20 5b6e 702e 7369 6e28 615b 6e5d 2920  = [np.sin(a[n]) 
-00018770: 2a20 302e 3735 2066 6f72 206e 2069 6e20  * 0.75 for n in 
-00018780: 7261 6e67 6528 6c65 6e28 6129 295d 0a20  range(len(a))]. 
-00018790: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000187a0: 2020 2020 2020 2078 7320 3d20 6e70 2e61         xs = np.a
-000187b0: 7272 6179 2878 7329 202a 2076 202a 2073  rray(xs) * v * s
-000187c0: 6361 6c65 5f66 6163 740a 2020 2020 2020  cale_fact.      
-000187d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000187e0: 2020 7973 203d 206e 702e 6172 7261 7928    ys = np.array(
-000187f0: 7973 2920 2a20 7620 2a20 7363 616c 655f  ys) * v * scale_
-00018800: 6661 6374 0a20 2020 2020 2020 2020 2020  fact.           
-00018810: 2020 2020 2020 2020 2020 2020 205b 706c               [pl
-00018820: 742e 706c 6f74 2828 785f 7661 6c20 2b20  t.plot((x_val + 
-00018830: 302e 352c 2028 785f 7661 6c20 2b20 302e  0.5, (x_val + 0.
-00018840: 3520 2b20 7873 5b6e 5d29 292c 0a20 2020  5 + xs[n])),.   
-00018850: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018860: 2020 2020 2020 2020 2020 2020 2020 2028                 (
-00018870: 795f 7661 6c20 2b20 302e 352c 2028 795f  y_val + 0.5, (y_
-00018880: 7661 6c20 2b20 302e 3520 2b20 7973 5b6e  val + 0.5 + ys[n
-00018890: 5d29 292c 0a20 2020 2020 2020 2020 2020  ])),.           
-000188a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000188b0: 2020 2020 2020 206c 696e 6577 6964 7468         linewidth
-000188c0: 3d32 2c0a 2020 2020 2020 2020 2020 2020  =2,.            
-000188d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000188e0: 2020 2020 2020 636f 6c6f 723d 636f 6c6f        color=colo
-000188f0: 7229 2066 6f72 206e 2069 6e20 7261 6e67  r) for n in rang
-00018900: 6528 6c65 6e28 7873 2929 5d0a 0a20 2020  e(len(xs))]..   
-00018910: 2020 2020 2020 2020 2065 6c69 6620 7374           elif st
-00018920: 796c 6520 3d3d 2027 6364 7261 7773 273a  yle == 'cdraws':
-00018930: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00018940: 2066 6f72 2078 5f76 616c 2069 6e20 783a   for x_val in x:
-00018950: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00018960: 2020 2020 2066 6f72 2079 5f76 616c 2069       for y_val i
-00018970: 6e20 793a 0a20 2020 2020 2020 2020 2020  n y:.           
-00018980: 2020 2020 2020 2020 2020 2020 2070 7473               pts
-00018990: 203d 205b 286e 702e 636f 7328 6129 2c20   = [(np.cos(a), 
-000189a0: 6e70 2e73 696e 2861 2929 2066 6f72 2061  np.sin(a)) for a
-000189b0: 2069 6e20 722e 6368 6f69 6365 280a 2020   in r.choice(.  
-000189c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000189d0: 2020 2020 2020 2020 2020 2020 2020 7375                su
-000189e0: 7266 2e73 7572 665b 795f 7661 6c2c 2078  rf.surf[y_val, x
-000189f0: 5f76 616c 2c20 3a5d 2c20 7369 7a65 3d31  _val, :], size=1
-00018a00: 3030 302c 0a20 2020 2020 2020 2020 2020  000,.           
-00018a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018a20: 2020 2020 2072 6570 6c61 6365 3d54 7275       replace=Tru
-00018a30: 6529 5d0a 2020 2020 2020 2020 2020 2020  e)].            
-00018a40: 2020 2020 2020 2020 2020 2020 706c 742e              plt.
-00018a50: 7363 6174 7465 7228 5b70 745b 305d 202a  scatter([pt[0] *
-00018a60: 2030 2e35 202b 2078 5f76 616c 202b 2030   0.5 + x_val + 0
-00018a70: 2e35 2066 6f72 2070 7420 696e 2070 7473  .5 for pt in pts
-00018a80: 5d2c 0a20 2020 2020 2020 2020 2020 2020  ],.             
+00018660: 2022 6d75 7374 2062 6520 7265 7072 6573   "must be repres
+00018670: 656e 7465 6420 6279 2061 2064 6963 7420  ented by a dict 
+00018680: 7769 7468 2061 2073 696e 676c 6520 6b65  with a single ke
+00018690: 7920 2874 6865 2022 0a20 2020 2020 2020  y (the ".       
+000186a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000186b0: 2022 6e75 6d62 6572 206f 6620 496e 6469   "number of Indi
+000186c0: 7669 6475 616c 7320 746f 2073 616d 706c  viduals to sampl
+000186d0: 6520 6672 6f6d 2074 6865 2070 6f70 756c  e from the popul
+000186e0: 6174 696f 6e29 2022 0a20 2020 2020 2020  ation) ".       
+000186f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018700: 2022 7768 6f73 6520 7661 6c75 6520 6973   "whose value is
+00018710: 2061 2064 6963 7420 636f 6e74 6169 6e69   a dict containi
+00018720: 6e67 2031 2e29 2061 2027 636f 6f72 6473  ng 1.) a 'coords
+00018730: 2720 6b65 7920 220a 2020 2020 2020 2020  ' key ".        
+00018740: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018750: 2269 6e64 6963 6174 696e 6720 7468 6520  "indicating the 
+00018760: 6c6f 6361 7469 6f6e 2f73 2074 6f20 706c  location/s to pl
+00018770: 6163 6520 7468 6520 220a 2020 2020 2020  ace the ".      
+00018780: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018790: 2020 2273 616d 706c 6564 2049 6e64 6976    "sampled Indiv
+000187a0: 6964 7561 6c73 2061 7420 616e 6420 322e  iduals at and 2.
+000187b0: 2920 220a 2020 2020 2020 2020 2020 2020  ) ".            
+000187c0: 2020 2020 2020 2020 2020 2020 2261 2073              "a s
+000187d0: 6574 206f 6620 7661 6c69 6420 6b65 7977  et of valid keyw
+000187e0: 6f72 6420 6172 6775 6d65 6e74 7320 7468  ord arguments th
+000187f0: 6174 2022 0a20 2020 2020 2020 2020 2020  at ".           
+00018800: 2020 2020 2020 2020 2020 2020 2022 6361               "ca
+00018810: 6e20 6265 2070 726f 7669 6465 6420 746f  n be provided to
+00018820: 204d 6f64 656c 2e61 6464 5f69 6e64 6976   Model.add_indiv
+00018830: 6964 7561 6c73 2022 0a20 2020 2020 2020  iduals ".       
+00018840: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018850: 2022 6173 2074 6865 2027 736f 7572 6365   "as the 'source
+00018860: 5f6d 7370 7269 6d65 5f70 6172 616d 7327  _msprime_params'
+00018870: 2064 6963 742e 2229 0a20 2020 2020 2020   dict.").       
+00018880: 2020 2020 206e 203d 205b 2a70 6f70 2e6b       n = [*pop.k
+00018890: 6579 7328 295d 5b30 5d0a 2020 2020 2020  eys()][0].      
+000188a0: 2020 2020 2020 6173 7365 7274 2069 7369        assert isi
+000188b0: 6e73 7461 6e63 6528 6e2c 2069 6e74 292c  nstance(n, int),
+000188c0: 2028 2254 6865 206b 6579 206f 6620 6561   ("The key of ea
+000188d0: 6368 2022 0a20 2020 2020 2020 2020 2020  ch ".           
+000188e0: 2020 2020 2020 2020 2020 2020 2022 6d73               "ms
+000188f0: 7072 696d 6520 736f 7572 6365 2070 6f70  prime source pop
+00018900: 756c 6174 696f 6e20 696e 2061 2053 7065  ulation in a Spe
+00018910: 6369 6573 2720 2769 6e69 7427 2022 0a20  cies' 'init' ". 
+00018920: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018930: 2020 2020 2020 2022 7365 6374 696f 6e20         "section 
+00018940: 6f66 2074 6865 2070 6172 616d 6574 6572  of the parameter
+00018950: 7320 6669 6c65 206d 7573 7420 6265 2061  s file must be a
+00018960: 6e20 696e 7420 220a 2020 2020 2020 2020  n int ".        
+00018970: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018980: 2269 6e64 6963 6174 696e 6720 7468 6520  "indicating the 
+00018990: 6e75 6d62 6572 206f 6620 7374 6172 7469  number of starti
+000189a0: 6e67 2049 6e64 6976 6964 7561 6c73 2022  ng Individuals "
+000189b0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000189c0: 2020 2020 2020 2020 2022 746f 2062 6520           "to be 
+000189d0: 736f 7572 6365 6420 6672 6f6d 2074 6861  sourced from tha
+000189e0: 7420 706f 7075 6c61 7469 6f6e 2e22 290a  t population.").
+000189f0: 2020 2020 2020 2020 2020 2020 6173 7365              asse
+00018a00: 7274 2027 636f 6f72 6473 2720 696e 2070  rt 'coords' in p
+00018a10: 6f70 5b6e 5d2c 2028 2245 6163 6820 6d73  op[n], ("Each ms
+00018a20: 7072 696d 6520 736f 7572 6365 2070 6f70  prime source pop
+00018a30: 756c 6174 696f 6e27 7320 220a 2020 2020  ulation's ".    
+00018a40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018a50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018a60: 2020 2020 2269 6e69 7420 7061 7261 6d65      "init parame
+00018a70: 7465 7273 2064 6963 7420 6d75 7374 2069  ters dict must i
+00018a80: 6e63 6c75 6465 2022 0a20 2020 2020 2020  nclude ".       
 00018a90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018aa0: 2020 2020 2020 205b 7074 5b31 5d20 2a20         [pt[1] * 
-00018ab0: 302e 3520 2b20 795f 7661 6c20 2b20 302e  0.5 + y_val + 0.
-00018ac0: 3520 666f 7220 7074 2069 6e20 7074 735d  5 for pt in pts]
-00018ad0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00018aa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018ab0: 2022 6120 2763 6f6f 7264 7327 206b 6579   "a 'coords' key
+00018ac0: 2077 686f 7365 2076 616c 7565 2069 6e64   whose value ind
+00018ad0: 6963 6174 6573 2022 0a20 2020 2020 2020  icates ".       
 00018ae0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018af0: 2020 2020 2020 636f 6c6f 723d 636f 6c6f        color=colo
-00018b00: 722c 2061 6c70 6861 3d30 2e31 2c20 6d61  r, alpha=0.1, ma
-00018b10: 726b 6572 3d27 2e27 290a 0a20 2020 2020  rker='.')..     
-00018b20: 2020 2020 2020 2065 6c69 6620 7374 796c         elif styl
-00018b30: 6520 3d3d 2027 7665 6374 273a 0a20 2020  e == 'vect':.   
-00018b40: 2020 2020 2020 2020 2020 2020 2064 6566               def
-00018b50: 2070 6c6f 745f 6f6e 655f 6365 6c6c 2878   plot_one_cell(x
-00018b60: 2c20 7929 3a0a 2020 2020 2020 2020 2020  , y):.          
-00018b70: 2020 2020 2020 2020 2020 2320 6472 6177            # draw
-00018b80: 2073 616d 706c 6520 6f66 2061 6e67 6c65   sample of angle
-00018b90: 7320 6672 6f6d 2074 6865 2047 6175 7373  s from the Gauss
-00018ba0: 6961 6e20 4b44 450a 2020 2020 2020 2020  ian KDE.        
-00018bb0: 2020 2020 2020 2020 2020 2020 2372 6570              #rep
-00018bc0: 7265 7365 6e74 696e 6720 7468 6520 766f  resenting the vo
-00018bd0: 6e20 6d69 7365 7320 6d69 7874 7572 6520  n mises mixture 
-00018be0: 6469 7374 7269 6275 7469 6f6e 2028 4b44  distribution (KD
-00018bf0: 4529 0a20 2020 2020 2020 2020 2020 2020  E).             
-00018c00: 2020 2020 2020 2073 616d 7020 3d20 7375         samp = su
-00018c10: 7266 2e73 7572 665b 792c 782c 3a5d 0a20  rf.surf[y,x,:]. 
-00018c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018c30: 2020 2023 2063 7265 6174 6520 6c69 7374     # create list
-00018c40: 7320 6f66 2074 6865 2078 2061 6e64 2079  s of the x and y
-00018c50: 2028 692e 652e 2063 6f73 2061 6e64 2073   (i.e. cos and s
-00018c60: 696e 290a 2020 2020 2020 2020 2020 2020  in).            
-00018c70: 2020 2020 2020 2020 2363 6f6d 706f 6e65          #compone
-00018c80: 6e74 7320 6f66 2065 6163 6820 616e 676c  nts of each angl
-00018c90: 6520 696e 2074 6865 2073 616d 706c 650a  e in the sample.
-00018ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018cb0: 2020 2020 785f 7665 6374 7320 3d20 6e70      x_vects = np
-00018cc0: 2e63 6f73 2873 616d 7029 0a20 2020 2020  .cos(samp).     
-00018cd0: 2020 2020 2020 2020 2020 2020 2020 2079                 y
-00018ce0: 5f76 6563 7473 203d 206e 702e 7369 6e28  _vects = np.sin(
-00018cf0: 7361 6d70 290a 2020 2020 2020 2020 2020  samp).          
-00018d00: 2020 2020 2020 2020 2020 2320 6465 6669            # defi
-00018d10: 6e65 2074 6865 2064 7820 616e 6420 6479  ne the dx and dy
-00018d20: 2064 6973 7461 6e63 6573 2075 7365 6420   distances used 
-00018d30: 746f 2074 6865 0a20 2020 2020 2020 2020  to the.         
-00018d40: 2020 2020 2020 2020 2020 2023 706f 7369             #posi
-00018d50: 7469 6f6e 2074 6865 2061 7272 6f77 6865  tion the arrowhe
-00018d60: 6164 2028 6469 7669 6465 2062 7920 7371  ad (divide by sq
-00018d70: 7274 2832 292f 322c 2074 6f20 0a20 2020  rt(2)/2, to .   
-00018d80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018d90: 2023 7363 616c 6520 746f 2074 6865 2073   #scale to the s
-00018da0: 697a 6520 6f66 2068 616c 6620 6f66 2074  ize of half of t
-00018db0: 6865 2064 6961 676f 6e61 6c20 6f66 2061  he diagonal of a
-00018dc0: 2063 656c 6c29 0a20 2020 2020 2020 2020   cell).         
-00018dd0: 2020 2020 2020 2020 2020 2064 7820 3d20             dx = 
-00018de0: 6e70 2e6d 6561 6e28 785f 7665 6374 7329  np.mean(x_vects)
-00018df0: 202f 206e 702e 7371 7274 2832 290a 2020   / np.sqrt(2).  
-00018e00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018e10: 2020 6479 203d 206e 702e 6d65 616e 2879    dy = np.mean(y
-00018e20: 5f76 6563 7473 2920 2f20 6e70 2e73 7172  _vects) / np.sqr
-00018e30: 7428 3229 0a20 2020 2020 2020 2020 2020  t(2).           
-00018e40: 2020 2020 2020 2020 2023 206e 6f77 2070           # now p
-00018e50: 6c6f 7420 7468 6520 6172 726f 770a 2020  lot the arrow.  
+00018af0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018b00: 2022 7468 6520 6c6f 6361 7469 6f6e 2f73   "the location/s
+00018b10: 2077 6865 7265 2061 6c6c 2073 616d 706c   where all sampl
+00018b20: 6564 2022 0a20 2020 2020 2020 2020 2020  ed ".           
+00018b30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018b40: 2020 2020 2020 2020 2020 2020 2022 496e               "In
+00018b50: 6469 7669 6475 616c 7320 7368 6f75 6c64  dividuals should
+00018b60: 2062 6520 706c 6163 6564 2e22 290a 2020   be placed.").  
+00018b70: 2020 2020 2020 2020 2020 636f 6f72 6473            coords
+00018b80: 203d 2070 6f70 5b6e 5d2e 706f 7028 2763   = pop[n].pop('c
+00018b90: 6f6f 7264 7327 290a 2020 2020 2020 2020  oords').        
+00018ba0: 2020 2020 6173 7365 7274 2028 6973 696e      assert (isin
+00018bb0: 7374 616e 6365 2863 6f6f 7264 732c 206c  stance(coords, l
+00018bc0: 6973 7429 206f 720a 2020 2020 2020 2020  ist) or.        
+00018bd0: 2020 2020 2020 2020 2020 2020 6973 696e              isin
+00018be0: 7374 616e 6365 2863 6f6f 7264 732c 2074  stance(coords, t
+00018bf0: 7570 6c65 2920 6f72 0a20 2020 2020 2020  uple) or.       
+00018c00: 2020 2020 2020 2020 2020 2020 2069 7369               isi
+00018c10: 6e73 7461 6e63 6528 636f 6f72 6473 2c20  nstance(coords, 
+00018c20: 6e70 2e6e 6461 7272 6179 2929 2c20 2822  np.ndarray)), ("
+00018c30: 5468 6520 2763 6f6f 7264 7327 206b 6579  The 'coords' key
+00018c40: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00018c50: 2020 2020 2020 2020 2020 2022 696e 2074             "in t
+00018c60: 6865 2069 6e69 7420 6469 6374 2066 6f72  he init dict for
+00018c70: 2061 6e20 6d73 7072 696d 6520 736f 7572   an msprime sour
+00018c80: 6365 2070 6f70 756c 6174 696f 6e20 220a  ce population ".
+00018c90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018ca0: 2020 2020 2020 2020 226d 7573 7420 6861          "must ha
+00018cb0: 7665 2061 206c 6973 742c 2074 7570 6c65  ve a list, tuple
+00018cc0: 2c20 6f72 206e 756d 7079 2e6e 6461 7272  , or numpy.ndarr
+00018cd0: 6179 2061 7320 220a 2020 2020 2020 2020  ay as ".        
+00018ce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018cf0: 2269 7473 2076 616c 7565 2e22 290a 2020  "its value.").  
+00018d00: 2020 2020 2020 2020 2020 636f 6f72 6473            coords
+00018d10: 5f73 7472 7563 745f 6572 725f 6d73 6720  _struct_err_msg 
+00018d20: 3d20 2822 5468 6520 2763 6f6f 7264 7327  = ("The 'coords'
+00018d30: 2076 616c 7565 2066 6f72 2022 0a20 2020   value for ".   
+00018d40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018d50: 2020 2020 2022 616e 206d 7370 7269 6d65       "an msprime
+00018d60: 2073 6f75 7263 6520 706f 7075 6c61 746f   source populato
+00018d70: 6e20 6d75 7374 2062 6520 6f6e 6520 6f66  n must be one of
+00018d80: 3a20 220a 2020 2020 2020 2020 2020 2020  : ".            
+00018d90: 2020 2020 2020 2020 2020 2020 2231 2e29              "1.)
+00018da0: 2061 206c 6973 7420 6f72 2074 7570 6c65   a list or tuple
+00018db0: 206f 6620 6c65 6e67 7468 2032 2c20 6769   of length 2, gi
+00018dc0: 7669 6e67 2074 6865 2022 0a20 2020 2020  ving the ".     
+00018dd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018de0: 2020 2022 636f 6f72 6469 6e61 7465 2070     "coordinate p
+00018df0: 6169 7220 6174 2077 6869 6368 2061 6c6c  air at which all
+00018e00: 2049 6e64 6976 6964 7561 6c73 2022 0a20   Individuals ". 
+00018e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018e20: 2020 2020 2020 2022 7769 6c6c 2062 6567         "will beg
+00018e30: 696e 2074 6865 2073 696d 756c 6174 696f  in the simulatio
+00018e40: 6e3b 2032 2e29 2061 206c 6973 7420 6f72  n; 2.) a list or
+00018e50: 2074 7570 6c65 206f 6620 220a 2020 2020   tuple of ".    
 00018e60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018e70: 2020 706c 742e 6172 726f 7728 7820 2b20    plt.arrow(x + 
-00018e80: 302e 352c 2079 202b 2030 2e35 2c20 6478  0.5, y + 0.5, dx
-00018e90: 2c20 6479 2c20 616c 7068 613d 302e 3735  , dy, alpha=0.75
-00018ea0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00018e70: 2020 2020 2263 6f6f 7264 696e 6174 6520      "coordinate 
+00018e80: 7061 6972 7320 666f 7220 616c 6c20 496e  pairs for all In
+00018e90: 6469 7669 6475 616c 7320 2873 7472 7563  dividuals (struc
+00018ea0: 7475 7265 6420 220a 2020 2020 2020 2020  tured ".        
 00018eb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00018ec0: 636f 6c6f 723d 636f 6c6f 722c 2068 6561  color=color, hea
-00018ed0: 645f 7769 6474 683d 302e 3234 2c20 6865  d_width=0.24, he
-00018ee0: 6164 5f6c 656e 6774 683d 302e 3332 290a  ad_length=0.32).
-00018ef0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00018f00: 2023 2063 616c 6c20 7468 6520 696e 7465   # call the inte
-00018f10: 726e 616c 6c79 2064 6566 696e 6564 2066  rnally defined f
-00018f20: 756e 6374 696f 6e20 6173 2061 206e 6573  unction as a nes
-00018f30: 7465 6420 6c69 7374 0a20 2020 2020 2020  ted list.       
-00018f40: 2020 2020 2020 2020 2023 636f 6d70 7265           #compre
-00018f50: 6865 6e73 696f 6e20 666f 7220 616c 6c20  hension for all 
-00018f60: 7261 7374 6572 2063 656c 6c73 2c20 7768  raster cells, wh
-00018f70: 6963 6820 4920 6265 6c69 6576 650a 2020  ich I believe.  
-00018f80: 2020 2020 2020 2020 2020 2020 2020 2373                #s
-00018f90: 686f 756c 6420 646f 2069 7473 2062 6573  hould do its bes
-00018fa0: 7420 746f 2076 6563 746f 7269 7a65 2074  t to vectorize t
-00018fb0: 6865 2077 686f 6c65 206f 7065 7261 7469  he whole operati
-00018fc0: 6f6e 0a20 2020 2020 2020 2020 2020 2020  on.             
-00018fd0: 2020 205b 5b70 6c6f 745f 6f6e 655f 6365     [[plot_one_ce
-00018fe0: 6c6c 286a 2c20 6929 2066 6f72 2069 2069  ll(j, i) for i i
-00018ff0: 6e20 7261 6e67 6528 0a20 2020 2020 2020  n range(.       
-00019000: 2020 2020 2020 2020 2020 2020 2073 7572               sur
-00019010: 662e 7375 7266 2e73 6861 7065 5b30 5d29  f.surf.shape[0])
-00019020: 5d20 666f 7220 6a20 696e 2072 616e 6765  ] for j in range
-00019030: 280a 2020 2020 2020 2020 2020 2020 2020  (.              
-00019040: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019050: 2020 2020 2020 7375 7266 2e73 7572 662e        surf.surf.
-00019060: 7368 6170 655b 315d 295d 0a0a 0a20 2020  shape[1])]...   
-00019070: 2023 2070 6c6f 7420 7468 6520 6c69 6e65   # plot the line
-00019080: 6167 6520 666f 7220 6120 6769 7665 6e20  age for a given 
-00019090: 6e6f 6465 2061 6e64 206c 6f63 7573 0a20  node and locus. 
-000190a0: 2020 2064 6566 205f 706c 6f74 5f67 656e     def _plot_gen
-000190b0: 655f 666c 6f77 2873 656c 662c 206c 6f63  e_flow(self, loc
-000190c0: 7573 2c20 7374 796c 652c 206c 616e 642c  us, style, land,
-000190d0: 206e 6f64 6573 3d4e 6f6e 652c 2069 6e64   nodes=None, ind
-000190e0: 6976 6964 733d 4e6f 6e65 2c0a 2020 2020  ivids=None,.    
+00018ec0: 2261 7320 6e20 7375 626c 6973 7473 206f  "as n sublists o
+00018ed0: 7220 7375 6274 7570 6c65 732c 2065 6163  r subtuples, eac
+00018ee0: 6820 6f66 206c 656e 6774 6820 3229 3b20  h of length 2); 
+00018ef0: 6f72 2022 0a20 2020 2020 2020 2020 2020  or ".           
+00018f00: 2020 2020 2020 2020 2020 2020 2022 332e               "3.
+00018f10: 2920 6120 6e75 6d70 792e 6e64 6172 7261  ) a numpy.ndarra
+00018f20: 7920 6f66 2074 6865 2073 616d 6520 7374  y of the same st
+00018f30: 7275 6374 7572 6520 2869 2e65 2e2c 2022  ructure (i.e., "
+00018f40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00018f50: 2020 2020 2020 2020 2022 6569 7468 6572           "either
+00018f60: 2031 7832 2c20 6769 7669 6e67 2061 2073   1x2, giving a s
+00018f70: 696e 676c 6520 636f 6f72 6469 6e61 7465  ingle coordinate
+00018f80: 2070 6169 7220 666f 7220 220a 2020 2020   pair for ".    
+00018f90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018fa0: 2020 2020 2261 6c6c 2049 6e64 6976 6964      "all Individ
+00018fb0: 7561 6c73 2c20 6f72 206e 7832 2c20 7370  uals, or nx2, sp
+00018fc0: 6563 6966 7969 6e67 206c 6f63 6174 696f  ecifying locatio
+00018fd0: 6e73 2066 6f72 2022 0a20 2020 2020 2020  ns for ".       
+00018fe0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00018ff0: 2022 616c 6c20 6e20 496e 6469 7669 6475   "all n Individu
+00019000: 616c 7329 2e22 290a 2020 2020 2020 2020  als).").        
+00019010: 2020 2020 6966 2069 7369 6e73 7461 6e63      if isinstanc
+00019020: 6528 636f 6f72 6473 2c20 6c69 7374 2920  e(coords, list) 
+00019030: 6f72 2069 7369 6e73 7461 6e63 6528 636f  or isinstance(co
+00019040: 6f72 6473 2c20 7475 706c 6529 3a0a 2020  ords, tuple):.  
+00019050: 2020 2020 2020 2020 2020 2020 2020 6966                if
+00019060: 2069 7369 6e73 7461 6e63 6528 636f 6f72   isinstance(coor
+00019070: 6473 5b30 5d2c 206c 6973 7429 206f 7220  ds[0], list) or 
+00019080: 6973 696e 7374 616e 6365 2863 6f6f 7264  isinstance(coord
+00019090: 735b 305d 2c20 7475 706c 6529 3a0a 2020  s[0], tuple):.  
+000190a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000190b0: 2020 6173 7365 7274 2028 6e70 2e61 6c6c    assert (np.all
+000190c0: 285b 6c65 6e28 6329 203d 3d20 3220 666f  ([len(c) == 2 fo
+000190d0: 7220 6320 696e 2063 6f6f 7264 735d 2920  r c in coords]) 
+000190e0: 616e 640a 2020 2020 2020 2020 2020 2020  and.            
 000190f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019100: 2020 2063 6f6c 6f72 3d4e 6f6e 652c 2070     color=None, p
-00019110: 6865 6e6f 7479 7065 3d4e 6f6e 652c 206c  henotype=None, l
-00019120: 7972 5f6e 756d 3d30 2c20 6a69 7474 6572  yr_num=0, jitter
-00019130: 3d54 7275 652c 0a20 2020 2020 2020 2020  =True,.         
-00019140: 2020 2020 2020 2020 2020 2020 2020 616c                al
-00019150: 7068 613d 302e 352c 2073 697a 653d 3235  pha=0.5, size=25
-00019160: 2c20 6164 645f 726f 6f74 733d 4661 6c73  , add_roots=Fals
-00019170: 6529 3a0a 2020 2020 2020 2020 6966 206e  e):.        if n
-00019180: 6f74 2073 656c 662e 6765 6e5f 6172 6368  ot self.gen_arch
-00019190: 2e75 7365 5f74 736b 6974 3a0a 2020 2020  .use_tskit:.    
-000191a0: 2020 2020 2020 2020 7261 6973 6520 4578          raise Ex
-000191b0: 6365 7074 696f 6e28 2822 4765 6e65 2066  ception(("Gene f
-000191c0: 6c6f 7720 6361 6e6e 6f74 2062 6520 706c  low cannot be pl
-000191d0: 6f74 7465 6420 666f 7220 6120 5370 6563  otted for a Spec
-000191e0: 6965 7320 220a 2020 2020 2020 2020 2020  ies ".          
-000191f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019200: 2020 2022 7768 6f73 6520 7370 6174 6961     "whose spatia
-00019210: 6c20 7065 6469 6772 6565 2069 7320 6e6f  l pedigree is no
-00019220: 7420 6265 696e 6720 7472 6163 6b65 6420  t being tracked 
-00019230: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
-00019240: 2020 2020 2020 2020 2020 2020 2020 2022                 "
-00019250: 6279 2074 736b 6974 2e22 2929 0a20 2020  by tskit.")).   
-00019260: 2020 2020 2061 7373 6572 7420 7374 796c       assert styl
-00019270: 6520 696e 205b 276c 696e 6561 6765 272c  e in ['lineage',
-00019280: 2027 7665 6374 6f72 275d 2c20 2822 5468   'vector'], ("Th
-00019290: 6520 7374 796c 6520 6172 6775 6d65 6e74  e style argument
-000192a0: 206d 7573 7420 6265 2022 0a20 2020 2020   must be ".     
-000192b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000192c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000192d0: 2020 2020 2020 2020 2020 2022 6769 7665             "give
-000192e0: 6e20 6569 7468 6572 2027 6c69 6e65 6167  n either 'lineag
-000192f0: 6527 206f 7220 220a 2020 2020 2020 2020  e' or ".        
+00019100: 6c65 6e28 636f 6f72 6473 2920 3d3d 206e  len(coords) == n
+00019110: 292c 2063 6f6f 7264 735f 7374 7275 6374  ), coords_struct
+00019120: 5f65 7272 5f6d 7367 0a20 2020 2020 2020  _err_msg.       
+00019130: 2020 2020 2020 2020 2065 6c73 653a 0a20           else:. 
+00019140: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019150: 2020 2061 7373 6572 7420 6c65 6e28 636f     assert len(co
+00019160: 6f72 6473 2920 3d3d 2032 2c20 636f 6f72  ords) == 2, coor
+00019170: 6473 5f73 7472 7563 745f 6572 725f 6d73  ds_struct_err_ms
+00019180: 670a 2020 2020 2020 2020 2020 2020 656c  g.            el
+00019190: 6966 2069 7369 6e73 7461 6e63 6528 636f  if isinstance(co
+000191a0: 6f72 6473 2c20 6e70 2e6e 6461 7272 6179  ords, np.ndarray
+000191b0: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
+000191c0: 2020 2061 7373 6572 7420 2863 6f6f 7264     assert (coord
+000191d0: 732e 7368 6170 6520 3d3d 2028 312c 2032  s.shape == (1, 2
+000191e0: 2920 6f72 0a20 2020 2020 2020 2020 2020  ) or.           
+000191f0: 2020 2020 2020 2020 2020 2020 2063 6f6f               coo
+00019200: 7264 732e 7368 6170 6520 3d3d 2028 6e2c  rds.shape == (n,
+00019210: 2032 2929 2c20 636f 6f72 6473 5f73 7472   2)), coords_str
+00019220: 7573 745f 6572 725f 6d73 670a 2020 2020  ust_err_msg.    
+00019230: 2020 2020 2020 2020 2320 7369 6d75 6c61          # simula
+00019240: 7465 2074 6865 2069 6e64 6976 6964 7561  te the individua
+00019250: 6c73 2061 6e64 2061 6464 2074 6865 6d20  ls and add them 
+00019260: 746f 2074 6865 2053 7065 6369 6573 0a20  to the Species. 
+00019270: 2020 2020 2020 2020 2020 2073 656c 662e             self.
+00019280: 5f61 6464 5f69 6e64 6976 6964 7561 6c73  _add_individuals
+00019290: 286e 3d6e 2c0a 2020 2020 2020 2020 2020  (n=n,.          
+000192a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000192b0: 2020 2020 2020 2020 636f 6f72 6473 3d63          coords=c
+000192c0: 6f6f 7264 732c 0a20 2020 2020 2020 2020  oords,.         
+000192d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000192e0: 2020 2020 2020 2020 206c 616e 643d 6c61           land=la
+000192f0: 6e64 2c0a 2020 2020 2020 2020 2020 2020  nd,.            
 00019300: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019310: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019320: 2020 2020 2020 2020 2227 7665 6374 6f72          "'vector
-00019330: 2720 6173 2061 2076 616c 7565 2e22 290a  ' as a value.").
-00019340: 2020 2020 2020 2020 6966 206e 6f64 6573          if nodes
-00019350: 2069 7320 4e6f 6e65 3a0a 2020 2020 2020   is None:.      
-00019360: 2020 2020 2020 2320 6765 7420 6120 7261        # get a ra
-00019370: 6e64 6f6d 2073 656c 6563 7469 6f6e 206f  ndom selection o
-00019380: 6620 6e20 696e 6469 7669 6475 616c 732c  f n individuals,
-00019390: 2069 6620 696e 6469 7669 6473 2720 7661   if individs' va
-000193a0: 6c75 6520 6973 2069 6e74 0a20 2020 2020  lue is int.     
-000193b0: 2020 2020 2020 2069 6620 6973 696e 7374         if isinst
-000193c0: 616e 6365 2869 6e64 6976 6964 732c 2069  ance(individs, i
-000193d0: 6e74 293a 0a20 2020 2020 2020 2020 2020  nt):.           
-000193e0: 2020 2020 2069 6e64 6976 6964 7320 3d20       individs = 
-000193f0: 6e70 2e72 616e 646f 6d2e 6368 6f69 6365  np.random.choice
-00019400: 285b 2a73 656c 665d 2c20 696e 6469 7669  ([*self], indivi
-00019410: 6473 2c20 7265 706c 6163 653d 4661 6c73  ds, replace=Fals
-00019420: 6529 0a20 2020 2020 2020 2020 2020 2023  e).            #
-00019430: 736f 7274 2061 6e64 2073 696d 706c 6966  sort and simplif
-00019440: 7920 7468 6520 5461 626c 6543 6f6c 6c65  y the TableColle
-00019450: 6374 696f 6e2c 2069 6620 6e65 6564 6564  ction, if needed
-00019460: 200a 2020 2020 2020 2020 2020 2020 6966   .            if
-00019470: 206e 6f74 2073 656c 662e 5f74 635f 736f   not self._tc_so
-00019480: 7274 6564 5f61 6e64 5f73 696d 706c 6966  rted_and_simplif
-00019490: 6965 643a 0a20 2020 2020 2020 2020 2020  ied:.           
-000194a0: 2020 2020 2073 656c 662e 5f73 6f72 745f       self._sort_
-000194b0: 7369 6d70 6c69 6679 5f74 6162 6c65 5f63  simplify_table_c
-000194c0: 6f6c 6c65 6374 696f 6e28 290a 2020 2020  ollection().    
-000194d0: 2020 2020 2020 2020 6e6f 6465 7320 3d20          nodes = 
-000194e0: 7365 6c66 2e5f 6765 745f 6e6f 6465 7328  self._get_nodes(
-000194f0: 696e 6469 7669 6473 3d69 6e64 6976 6964  individs=individ
-00019500: 7329 0a0a 2020 2020 2020 2020 2320 736f  s)..        # so
-00019510: 7274 2061 6e64 2073 696d 706c 6966 7920  rt and simplify 
-00019520: 7468 6520 5461 626c 6543 6f6c 6c65 6374  the TableCollect
-00019530: 696f 6e20 6966 206e 6565 6465 640a 2020  ion if needed.  
-00019540: 2020 2020 2020 6966 206e 6f74 2073 656c        if not sel
-00019550: 662e 5f74 635f 736f 7274 6564 5f61 6e64  f._tc_sorted_and
-00019560: 5f73 696d 706c 6966 6965 643a 0a20 2020  _simplified:.   
-00019570: 2020 2020 2020 2020 2073 656c 662e 5f73           self._s
-00019580: 6f72 745f 7369 6d70 6c69 6679 5f74 6162  ort_simplify_tab
-00019590: 6c65 5f63 6f6c 6c65 6374 696f 6e28 290a  le_collection().
-000195a0: 2020 2020 2020 2020 2320 6772 6162 2074          # grab t
-000195b0: 6865 2054 6162 6c65 436f 6c6c 6563 7469  he TableCollecti
-000195c0: 6f6e 2061 6e64 2069 7473 2054 7265 6553  on and its TreeS
-000195d0: 6571 7565 6e63 650a 2020 2020 2020 2020  equence.        
-000195e0: 7463 203d 2073 656c 662e 5f74 630a 2020  tc = self._tc.  
-000195f0: 2020 2020 2020 7472 793a 0a20 2020 2020        try:.     
-00019600: 2020 2020 2020 2074 7320 3d20 7463 2e74         ts = tc.t
-00019610: 7265 655f 7365 7175 656e 6365 2829 0a20  ree_sequence(). 
-00019620: 2020 2020 2020 2065 7863 6570 7420 4578         except Ex
-00019630: 6365 7074 696f 6e3a 0a20 2020 2020 2020  ception:.       
-00019640: 2020 2020 2072 6169 7365 2045 7863 6570       raise Excep
-00019650: 7469 6f6e 2828 2254 6865 2073 7065 6369  tion(("The speci
-00019660: 6573 2720 5461 626c 6543 6f6c 6c65 6374  es' TableCollect
-00019670: 696f 6e20 6d75 7374 2062 6520 736f 7274  ion must be sort
-00019680: 6564 2061 6e64 2022 0a20 2020 2020 2020  ed and ".       
-00019690: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000196a0: 2020 2020 2020 2273 696d 706c 6966 6965        "simplifie
-000196b0: 6420 6265 666f 7265 2074 6869 7320 6d65  d before this me
-000196c0: 7468 6f64 2069 7320 6361 6c6c 6564 2e22  thod is called."
-000196d0: 2929 0a0a 2020 2020 2020 2020 6e6f 6465  ))..        node
-000196e0: 5f63 7572 725f 6c6f 6373 203d 205b 5b69  _curr_locs = [[i
-000196f0: 2e78 2c20 692e 795d 2066 6f72 206e 2069  .x, i.y] for n i
-00019700: 6e20 6e6f 6465 7320 666f 7220 6920 696e  n nodes for i in
-00019710: 2073 656c 662e 7661 6c75 6573 280a 2020   self.values(.  
-00019720: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019730: 2020 2020 2020 2920 6966 206e 2069 6e20        ) if n in 
-00019740: 692e 5f6e 6f64 6573 5f74 6162 5f69 6473  i._nodes_tab_ids
-00019750: 2e76 616c 7565 7328 295d 0a0a 2020 2020  .values()]..    
-00019760: 2020 2020 2320 6765 7420 7468 6520 7472      # get the tr
-00019770: 6565 2066 6f72 2074 6869 7320 6c6f 6375  ee for this locu
-00019780: 730a 2020 2020 2020 2020 7472 6565 203d  s.        tree =
-00019790: 2074 732e 6173 6c69 7374 2829 5b5f 6765   ts.aslist()[_ge
-000197a0: 745f 7472 6565 6e75 6d73 2874 732c 205b  t_treenums(ts, [
-000197b0: 6c6f 6375 735d 295b 305d 5d0a 2020 2020  locus])[0]].    
-000197c0: 2020 2020 2320 6765 7420 7468 6520 6c69      # get the li
-000197d0: 6e65 6167 655f 6469 6374 2028 7769 7468  neage_dict (with
-000197e0: 2062 6972 7468 2074 696d 6573 2061 6e64   birth times and
-000197f0: 2062 6972 7468 206c 6f63 6174 696f 6e73   birth locations
-00019800: 290a 2020 2020 2020 2020 6c69 6e5f 6469  ).        lin_di
-00019810: 6374 203d 205f 6765 745f 6c69 6e65 6167  ct = _get_lineag
-00019820: 655f 6469 6374 735f 6f6e 655f 7472 6565  e_dicts_one_tree
-00019830: 2874 632c 2074 7265 652c 206e 6f64 6573  (tc, tree, nodes
-00019840: 2c20 7365 6c66 2e74 290a 2020 2020 2020  , self.t).      
-00019850: 2020 6966 2063 6f6c 6f72 2069 7320 4e6f    if color is No
-00019860: 6e65 3a0a 2020 2020 2020 2020 2020 2020  ne:.            
-00019870: 2320 6372 6561 7465 2073 7461 7274 2d63  # create start-c
-00019880: 6f6c 6f72 2076 616c 7565 7320 666f 7220  olor values for 
-00019890: 6e6f 6465 7327 2073 6570 6172 6174 6520  nodes' separate 
-000198a0: 6c69 6e65 6167 6520 7472 6163 6b73 0a20  lineage tracks. 
-000198b0: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
-000198c0: 7320 3d20 5b6d 706c 2e63 6f6c 6f72 732e  s = [mpl.colors.
-000198d0: 746f 5f68 6578 2870 6c74 2e63 6d2e 5365  to_hex(plt.cm.Se
-000198e0: 7431 5f72 280a 2020 2020 2020 2020 2020  t1_r(.          
-000198f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019900: 2020 2020 2020 2020 2020 2020 2020 6e29                n)
-00019910: 2920 666f 7220 6e20 696e 206e 702e 6c69  ) for n in np.li
-00019920: 6e73 7061 6365 2830 2c20 302e 3835 2c20  nspace(0, 0.85, 
-00019930: 3829 5d0a 2020 2020 2020 2020 656c 7365  8)].        else
-00019940: 3a0a 2020 2020 2020 2020 2020 2020 636f  :.            co
-00019950: 6c6f 7273 203d 205b 636f 6c6f 7220 666f  lors = [color fo
-00019960: 7220 5f20 696e 2072 616e 6765 2838 295d  r _ in range(8)]
-00019970: 0a20 2020 2020 2020 2023 2070 6c6f 7420  .        # plot 
-00019980: 7468 6520 7370 6563 6965 732c 2065 6974  the species, eit
-00019990: 6865 7220 7769 7468 206f 7220 7769 7468  her with or with
-000199a0: 6f75 7420 7068 656e 6f74 7970 652d 7061  out phenotype-pa
-000199b0: 696e 7469 6e67 0a20 2020 2020 2020 2069  inting.        i
-000199c0: 6620 7068 656e 6f74 7970 6520 6973 204e  f phenotype is N
-000199d0: 6f6e 653a 0a20 2020 2020 2020 2020 2020  one:.           
-000199e0: 2073 656c 662e 5f70 6c6f 7428 6c79 725f   self._plot(lyr_
-000199f0: 6e75 6d3d 6c79 725f 6e75 6d2c 206c 616e  num=lyr_num, lan
-00019a00: 643d 6c61 6e64 2c20 7369 7a65 3d73 697a  d=land, size=siz
-00019a10: 6529 0a20 2020 2020 2020 2065 6c73 653a  e).        else:
-00019a20: 0a20 2020 2020 2020 2020 2020 2073 656c  .            sel
-00019a30: 662e 5f70 6c6f 745f 7068 656e 6f74 7970  f._plot_phenotyp
-00019a40: 6528 7068 656e 6f74 7970 652c 206c 616e  e(phenotype, lan
-00019a50: 643d 6c61 6e64 2c20 7369 7a65 3d73 697a  d=land, size=siz
-00019a60: 6529 0a20 2020 2020 2020 2061 7820 3d20  e).        ax = 
-00019a70: 706c 742e 6763 6128 290a 2020 2020 2020  plt.gca().      
-00019a80: 2020 2320 6578 7472 6163 7420 616e 6420    # extract and 
-00019a90: 706c 6f74 2074 6865 2073 6572 6965 7320  plot the series 
-00019aa0: 6f66 2070 6f69 6e74 7320 666f 7220 6561  of points for ea
-00019ab0: 6368 206e 6f64 650a 2020 2020 2020 2020  ch node.        
-00019ac0: 666f 7220 692c 206e 6f64 6520 696e 2065  for i, node in e
-00019ad0: 6e75 6d65 7261 7465 286e 6f64 6573 293a  numerate(nodes):
-00019ae0: 0a20 2020 2020 2020 2020 2020 2073 7461  .            sta
-00019af0: 7274 5f63 6f6c 203d 2063 6f6c 6f72 735b  rt_col = colors[
-00019b00: 6920 2520 6c65 6e28 636f 6c6f 7273 295d  i % len(colors)]
-00019b10: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
-00019b20: 7374 796c 6520 3d3d 2027 6c69 6e65 6167  style == 'lineag
-00019b30: 6527 3a0a 2020 2020 2020 2020 2020 2020  e':.            
-00019b40: 2020 2020 6c6f 6373 203d 206e 702e 7673      locs = np.vs
-00019b50: 7461 636b 285b 765b 315d 2066 6f72 2076  tack([v[1] for v
-00019b60: 2069 6e20 6c69 6e5f 6469 6374 5b6e 6f64   in lin_dict[nod
-00019b70: 655d 2e76 616c 7565 7328 295d 290a 2020  e].values()]).  
-00019b80: 2020 2020 2020 2020 2020 2020 2020 6966                if
-00019b90: 206a 6974 7465 723a 0a20 2020 2020 2020   jitter:.       
-00019ba0: 2020 2020 2020 2020 2020 2020 206c 6f63               loc
-00019bb0: 7320 3d20 6c6f 6373 202b 206e 702e 7261  s = locs + np.ra
-00019bc0: 6e64 6f6d 2e6e 6f72 6d61 6c28 302c 2030  ndom.normal(0, 0
-00019bd0: 2e30 312c 0a20 2020 2020 2020 2020 2020  .01,.           
-00019be0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019bf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019c00: 2073 697a 653d 6c6f 6373 2e73 697a 6529   size=locs.size)
-00019c10: 2e72 6573 6861 7065 286c 6f63 732e 7368  .reshape(locs.sh
-00019c20: 6170 6529 0a20 2020 2020 2020 2020 2020  ape).           
-00019c30: 2020 2020 2023 2063 7265 6174 6520 6c69       # create li
-00019c40: 7374 206f 6620 636f 6c6f 7273 2066 6f72  st of colors for
-00019c50: 2070 6c6f 7474 696e 672c 0a20 2020 2020   plotting,.     
-00019c60: 2020 2020 2020 2020 2020 2023 2075 7369             # usi
-00019c70: 6e67 206c 696e 6561 726c 7920 696e 7465  ng linearly inte
-00019c80: 7270 6f6c 6174 6564 2063 6f6c 6f72 730a  rpolated colors.
-00019c90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019ca0: 2320 6966 2074 6865 2063 6f6c 6f72 2061  # if the color a
-00019cb0: 7267 756d 656e 7420 7761 7320 6e6f 7420  rgument was not 
-00019cc0: 7072 6f76 6964 6564 2c20 0a20 2020 2020  provided, .     
-00019cd0: 2020 2020 2020 2020 2020 2023 206f 7220             # or 
-00019ce0: 656c 7365 206a 7573 7420 7573 696e 6720  else just using 
-00019cf0: 7468 6520 736f 6c69 6420 636f 6c6f 720a  the solid color.
-00019d00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019d10: 2320 7072 6f76 6964 6564 2074 6f20 7468  # provided to th
-00019d20: 6520 636f 6c6f 7220 6172 6775 6d65 6e74  e color argument
-00019d30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00019d40: 2063 6f6c 6f72 5f6e 756d 7320 3d20 6e70   color_nums = np
-00019d50: 2e69 6e74 3828 6e70 2e6c 696e 7370 6163  .int8(np.linspac
-00019d60: 6528 302c 2031 3030 2c20 6c6f 6373 2e73  e(0, 100, locs.s
-00019d70: 6861 7065 5b30 5d2d 3129 290a 2020 2020  hape[0]-1)).    
-00019d80: 2020 2020 2020 2020 2020 2020 6966 2063              if c
-00019d90: 6f6c 6f72 2069 7320 4e6f 6e65 3a0a 2020  olor is None:.  
-00019da0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019db0: 2020 706c 6f74 5f63 6f6c 6f72 7320 3d5b    plot_colors =[
-00019dc0: 7669 7a2e 5f63 616c 635f 7265 7368 6164  viz._calc_reshad
-00019dd0: 6564 5f63 6f6c 6f72 2873 7461 7274 5f63  ed_color(start_c
-00019de0: 6f6c 2c0a 2020 2020 2020 2020 2020 2020  ol,.            
-00019df0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019e00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019e10: 2020 2020 2020 6e75 6d29 2066 6f72 206e        num) for n
-00019e20: 756d 2069 6e20 636f 6c6f 725f 6e75 6d73  um in color_nums
-00019e30: 5d0a 2020 2020 2020 2020 2020 2020 2020  ].              
-00019e40: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
-00019e50: 2020 2020 2020 2020 2020 2020 706c 6f74              plot
-00019e60: 5f63 6f6c 6f72 7320 3d20 5b73 7461 7274  _colors = [start
-00019e70: 5f63 6f6c 2066 6f72 206e 756d 2069 6e20  _col for num in 
-00019e80: 636f 6c6f 725f 6e75 6d73 5d0a 2020 2020  color_nums].    
-00019e90: 2020 2020 2020 2020 2020 2020 2320 6372              # cr
-00019ea0: 6561 7465 2061 206c 696e 6561 7220 696e  eate a linear in
-00019eb0: 7465 7270 6f6c 6174 696f 6e20 6f66 206c  terpolation of l
-00019ec0: 696e 6577 6964 7468 730a 2020 2020 2020  inewidths.      
-00019ed0: 2020 2020 2020 2020 2020 6c69 6e65 7769            linewi
-00019ee0: 6474 6873 203d 206e 702e 6c69 6e73 7061  dths = np.linspa
-00019ef0: 6365 2833 2c20 302e 3835 2c20 6c6f 6373  ce(3, 0.85, locs
-00019f00: 2e73 6861 7065 5b30 5d2d 3129 0a20 2020  .shape[0]-1).   
-00019f10: 2020 2020 2020 2020 2020 2020 2066 6f72               for
-00019f20: 206e 2c20 636f 6c20 696e 2065 6e75 6d65   n, col in enume
-00019f30: 7261 7465 2870 6c6f 745f 636f 6c6f 7273  rate(plot_colors
-00019f40: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-00019f50: 2020 2020 2020 2023 204e 4f54 4520 6e65         # NOTE ne
-00019f60: 6564 2074 6f20 7573 6520 6f6e 6c79 2074  ed to use only t
-00019f70: 6865 2066 6972 7374 2032 2076 616c 7565  he first 2 value
-00019f80: 7320 696e 2074 6865 206c 6f63 6174 696f  s in the locatio
-00019f90: 6e0a 2020 2020 2020 2020 2020 2020 2020  n.              
-00019fa0: 2020 2020 2020 2320 6461 7461 2062 6563        # data bec
-00019fb0: 6175 7365 2073 7562 7365 7175 656e 7420  ause subsequent 
-00019fc0: 7661 6c75 6573 2061 7265 2075 7365 640a  values are used.
+00019310: 2020 2020 2020 736f 7572 6365 5f6d 7370        source_msp
+00019320: 7269 6d65 5f70 6172 616d 733d 706f 705b  rime_params=pop[
+00019330: 6e5d 2c0a 2020 2020 2020 2020 2020 2020  n],.            
+00019340: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019350: 2020 2020 2029 0a20 2020 2020 2020 2023       ).        #
+00019360: 2063 6865 636b 2074 6861 7420 7468 6520   check that the 
+00019370: 5370 6563 6965 7320 776f 756e 6420 7570  Species wound up
+00019380: 2068 6176 696e 6720 7468 6520 636f 7272   having the corr
+00019390: 6563 7420 706f 7075 6c61 7469 6f6e 2073  ect population s
+000193a0: 697a 650a 2020 2020 2020 2020 6173 7365  ize.        asse
+000193b0: 7274 206c 656e 2873 656c 6629 203d 3d20  rt len(self) == 
+000193c0: 7461 7267 6574 5f73 697a 652c 2028 2241  target_size, ("A
+000193d0: 6674 6572 2061 6c6c 206d 7370 7269 6d65  fter all msprime
+000193e0: 2073 6f75 7263 6520 220a 2020 2020 2020   source ".      
+000193f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019400: 2020 2270 6f70 756c 6174 696f 6e73 2077    "populations w
+00019410: 6572 6520 7361 6d70 6c65 642c 2074 6865  ere sampled, the
+00019420: 2072 6573 756c 7469 6e67 2022 0a20 2020   resulting ".   
+00019430: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019440: 2020 2020 2022 706f 7075 6c61 7469 6f6e       "population
+00019450: 2073 697a 6520 646f 6573 206e 6f74 2061   size does not a
+00019460: 6772 6565 2077 6974 6820 220a 2020 2020  gree with ".    
+00019470: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019480: 2020 2020 2274 6865 2073 756d 206f 6620      "the sum of 
+00019490: 7468 6520 736f 7572 6365 2070 6f70 756c  the source popul
+000194a0: 6174 696f 6e20 7361 6d70 6c65 7320 220a  ation samples ".
+000194b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000194c0: 2020 2020 2020 2020 2273 7469 7075 6c61          "stipula
+000194d0: 7465 6420 696e 2074 6865 2070 6172 616d  ted in the param
+000194e0: 6574 6572 7320 6669 6c65 2e22 290a 0a0a  eters file.")...
+000194f0: 2020 2020 2375 7365 2074 6865 206b 645f      #use the kd_
+00019500: 7472 6565 2061 7474 7269 6275 7465 2074  tree attribute t
+00019510: 6f20 6669 6e64 206d 6174 696e 6720 7061  o find mating pa
+00019520: 6972 7320 6569 7468 6572 0a20 2020 2023  irs either.    #
+00019530: 7769 7468 696e 2074 6865 2073 7065 6369  within the speci
+00019540: 6573 2c20 6966 2077 6974 6869 6e20 3d3d  es, if within ==
+00019550: 2054 7275 652c 206f 7220 6265 7477 6565   True, or betwee
+00019560: 6e20 7468 6520 7370 6563 6965 730a 2020  n the species.  
+00019570: 2020 2361 6e64 2074 6865 2070 6f69 6e74    #and the point
+00019580: 7320 7072 6f76 6964 6564 2c20 6966 2077  s provided, if w
+00019590: 6974 6869 6e20 3d3d 2046 616c 7365 2061  ithin == False a
+000195a0: 6e64 2070 6f69 6e74 7320 6973 206e 6f74  nd points is not
+000195b0: 204e 6f6e 650a 2020 2020 6465 6620 5f67   None.    def _g
+000195c0: 6574 5f6d 6174 696e 675f 7061 6972 7328  et_mating_pairs(
+000195d0: 7365 6c66 2c20 7769 7468 696e 3d54 7275  self, within=Tru
+000195e0: 652c 2063 6f6f 7264 733d 4e6f 6e65 2c0a  e, coords=None,.
+000195f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019600: 2020 2020 2020 2020 2020 6368 6f6f 7365            choose
+00019610: 5f6e 6561 7265 7374 3d46 616c 7365 2c20  _nearest=False, 
+00019620: 696e 7665 7273 655f 6469 7374 5f6d 6174  inverse_dist_mat
+00019630: 696e 673d 4661 6c73 6529 3a0a 2020 2020  ing=False):.    
+00019640: 2020 2020 234e 4f54 453a 2049 6e20 6c69      #NOTE: In li
+00019650: 6575 206f 6620 6120 6d6f 7265 2073 6f70  eu of a more sop
+00019660: 6869 7374 6963 6174 6564 2077 6179 206f  histicated way o
+00019670: 660a 2020 2020 2020 2020 2364 6574 6572  f.        #deter
+00019680: 6d69 6e69 6e67 2077 6865 7468 6572 2074  mining whether t
+00019690: 6865 206b 645f 7472 6565 206e 6565 6473  he kd_tree needs
+000196a0: 2074 6f20 6265 2075 7064 6174 6564 200a   to be updated .
+000196b0: 2020 2020 2020 2020 2328 692e 652e 2069          #(i.e. i
+000196c0: 6620 7468 6520 7370 6563 6965 7320 6861  f the species ha
+000196d0: 7665 2075 6e64 6572 676f 6e65 206d 6f76  ve undergone mov
+000196e0: 656d 656e 742c 206d 6174 696e 672c 0a20  ement, mating,. 
+000196f0: 2020 2020 2020 2023 6f72 206d 6f72 7461         #or morta
+00019700: 6c69 7479 2073 696e 6365 2074 6865 206c  lity since the l
+00019710: 6173 7420 7469 6d65 2069 7420 7761 7320  ast time it was 
+00019720: 0a20 2020 2020 2020 2023 636f 6e73 7472  .        #constr
+00019730: 7563 7465 6429 2c20 616e 6420 696e 2061  ucted), and in a
+00019740: 6e20 6566 666f 7274 2074 6f20 6d69 6e69  n effort to mini
+00019750: 6d69 7a65 2074 6865 206e 756d 6265 720a  mize the number.
+00019760: 2020 2020 2020 2020 236f 6620 7469 6d65          #of time
+00019770: 7320 6974 2069 7320 636f 6e73 7472 7563  s it is construc
+00019780: 7465 6420 6561 6368 2074 696d 6520 2873  ted each time (s
+00019790: 696e 6365 0a20 2020 2020 2020 2023 6974  ince.        #it
+000197a0: 2773 206e 6f74 2061 6e20 696e 636f 6e73  's not an incons
+000197b0: 6571 7565 6e74 6961 6c20 7275 6e74 696d  equential runtim
+000197c0: 652c 2073 6f20 7465 6c6c 696e 6720 7468  e, so telling th
+000197d0: 650a 2020 2020 2020 2020 236d 6f64 656c  e.        #model
+000197e0: 2074 6f20 7265 6275 696c 6420 6974 2061   to rebuild it a
+000197f0: 6674 6572 2065 6163 6820 6d6f 7665 2c20  fter each move, 
+00019800: 6269 7274 682c 206f 720a 2020 2020 2020  birth, or.      
+00019810: 2020 2364 6561 7468 2073 7465 7020 636f    #death step co
+00019820: 756c 6420 6265 2075 6e6e 6365 7373 6172  uld be unncessar
+00019830: 696c 7920 636f 7374 6c79 292c 2066 6f72  ily costly), for
+00019840: 206e 6f77 2049 0a20 2020 2020 2020 2023   now I.        #
+00019850: 616d 206a 7573 7420 7465 6c6c 696e 6720  am just telling 
+00019860: 7468 6520 7472 6565 2074 6f20 6265 2072  the tree to be r
+00019870: 6562 7569 6c74 2065 6163 6820 7469 6d65  ebuilt each time
+00019880: 200a 2020 2020 2020 2020 2374 6865 2073   .        #the s
+00019890: 7070 2e5f 6765 745f 6d61 7469 6e67 5f70  pp._get_mating_p
+000198a0: 6169 7273 2829 206d 6574 686f 6420 6973  airs() method is
+000198b0: 2063 616c 6c65 6421 0a20 2020 2020 2020   called!.       
+000198c0: 2073 656c 662e 5f73 6574 5f6b 645f 7472   self._set_kd_tr
+000198d0: 6565 2829 0a0a 2020 2020 2020 2020 2320  ee()..        # 
+000198e0: 6966 206d 6174 696e 675f 7261 6469 7573  if mating_radius
+000198f0: 2069 7320 4e6f 6e65 2c20 7468 656e 206a   is None, then j
+00019900: 7573 7420 7573 6520 5772 6967 6874 2d46  ust use Wright-F
+00019910: 6973 6865 720a 2020 2020 2020 2020 2320  isher.        # 
+00019920: 7374 796c 6520 7061 6e6d 6978 6961 2028  style panmixia (
+00019930: 6472 6177 2077 6974 6820 7265 706c 6163  draw with replac
+00019940: 656d 656e 7420 6120 7361 6d70 6c65 0a20  ement a sample. 
+00019950: 2020 2020 2020 2023 206f 6620 7369 7a65         # of size
+00019960: 203d 204e 742a 622c 2077 6865 7265 204e   = Nt*b, where N
+00019970: 7420 6973 2074 6865 200a 2020 2020 2020  t is the .      
+00019980: 2020 2320 6375 7272 656e 7420 706f 7020    # current pop 
+00019990: 7369 7a65 2061 6e64 2062 2069 7320 7468  size and b is th
+000199a0: 6520 6269 7274 6820 7261 7465 2028 692e  e birth rate (i.
+000199b0: 652e 206d 6174 696e 6720 7072 6f62 6162  e. mating probab
+000199c0: 696c 6974 7929 3b0a 2020 2020 2020 2020  ility);.        
+000199d0: 2320 7468 6174 2073 616d 706c 6520 7265  # that sample re
+000199e0: 7072 6573 656e 7473 2061 6c6c 206d 6174  presents all mat
+000199f0: 696e 6720 696e 6469 7669 6475 616c 732c  ing individuals,
+00019a00: 2061 6e64 2065 6163 6820 6f66 2074 686f   and each of tho
+00019a10: 7365 0a20 2020 2020 2020 2023 2069 6e64  se.        # ind
+00019a20: 6976 6964 7561 6c73 2072 616e 646f 6d6c  ividuals randoml
+00019a30: 7920 6368 6f6f 7365 7320 6974 7320 6d61  y chooses its ma
+00019a40: 7465 0a20 2020 2020 2020 2069 6620 7365  te.        if se
+00019a50: 6c66 2e6d 6174 696e 675f 7261 6469 7573  lf.mating_radius
+00019a60: 2069 7320 4e6f 6e65 3a0a 2020 2020 2020   is None:.      
+00019a70: 2020 2020 2020 2320 6472 6177 2061 206e        # draw a n
+00019a80: 756d 6265 7220 6f66 206d 6174 696e 6720  umber of mating 
+00019a90: 696e 6469 7669 6475 616c 7320 6173 2061  individuals as a
+00019aa0: 2062 696e 6f6d 6961 6c20 7276 0a20 2020   binomial rv.   
+00019ab0: 2020 2020 2020 2020 2023 2077 6974 6820           # with 
+00019ac0: 6e75 6d20 7472 6961 6c73 2065 7175 616c  num trials equal
+00019ad0: 2074 6f20 706f 7020 7369 7a65 2061 6e64   to pop size and
+00019ae0: 2070 726f 6261 6269 6c69 7479 2065 7175   probability equ
+00019af0: 616c 2074 6f0a 2020 2020 2020 2020 2020  al to.          
+00019b00: 2020 2320 7468 6520 7370 6563 6965 7327    # the species'
+00019b10: 2062 6972 7468 2072 6174 650a 2020 2020   birth rate.    
+00019b20: 2020 2020 2020 2020 6966 2073 656c 662e          if self.
+00019b30: 6220 3c20 313a 0a20 2020 2020 2020 2020  b < 1:.         
+00019b40: 2020 2020 2020 206e 5f6d 6174 6573 203d         n_mates =
+00019b50: 206e 702e 7261 6e64 6f6d 2e62 696e 6f6d   np.random.binom
+00019b60: 6961 6c28 6e3d 6c65 6e28 7365 6c66 292c  ial(n=len(self),
+00019b70: 2070 3d73 656c 662e 6229 0a20 2020 2020   p=self.b).     
+00019b80: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
+00019b90: 2020 2020 2020 2020 2020 2020 206e 5f6d               n_m
+00019ba0: 6174 6573 203d 206c 656e 2873 656c 6629  ates = len(self)
+00019bb0: 0a20 2020 2020 2020 2020 2020 2074 7265  .            tre
+00019bc0: 655f 696e 6473 203d 2073 656c 662e 5f6b  e_inds = self._k
+00019bd0: 645f 7472 6565 2e74 7265 652e 696e 6469  d_tree.tree.indi
+00019be0: 6365 730a 2020 2020 2020 2020 2020 2020  ces.            
+00019bf0: 2320 6472 6177 2032 2a6e 5f6d 6174 6573  # draw 2*n_mates
+00019c00: 206d 6174 696e 6720 696e 6469 7669 6475   mating individu
+00019c10: 616c 732c 2077 6974 6820 7265 706c 6163  als, with replac
+00019c20: 656d 656e 7420 2861 7320 696e 2057 460a  ement (as in WF.
+00019c30: 2020 2020 2020 2020 2020 2020 2320 6d6f              # mo
+00019c40: 6465 6c29 2c20 7468 656e 2066 6f6c 6420  del), then fold 
+00019c50: 696e 746f 2061 6e20 6e5f 6d61 7465 7320  into an n_mates 
+00019c60: 7820 3220 6d61 7465 2d70 6169 7273 2061  x 2 mate-pairs a
+00019c70: 7272 6179 0a20 2020 2020 2020 2020 2020  rray.           
+00019c80: 2070 6169 7273 203d 206e 702e 7261 6e64   pairs = np.rand
+00019c90: 6f6d 2e63 686f 6963 6528 7472 6565 5f69  om.choice(tree_i
+00019ca0: 6e64 732c 2072 6570 6c61 6365 3d54 7275  nds, replace=Tru
+00019cb0: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+00019cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019cd0: 2020 2020 2020 2020 7369 7a65 3d6e 5f6d          size=n_m
+00019ce0: 6174 6573 2a32 292e 7265 7368 6170 6528  ates*2).reshape(
+00019cf0: 286e 5f6d 6174 6573 2c20 3229 290a 2020  (n_mates, 2)).  
+00019d00: 2020 2020 2020 2020 2020 2320 6765 7420            # get 
+00019d10: 7269 6420 6f66 2073 656c 6669 6e67 2070  rid of selfing p
+00019d20: 6169 7273 0a20 2020 2020 2020 2020 2020  airs.           
+00019d30: 2070 6169 7273 203d 206e 702e 6172 7261   pairs = np.arra
+00019d40: 7928 5b6c 6973 7428 7061 6972 2920 666f  y([list(pair) fo
+00019d50: 7220 7061 6972 2069 6e20 6c69 7374 280a  r pair in list(.
+00019d60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019d70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019d80: 2020 2020 6d61 7028 7365 742c 2070 6169      map(set, pai
+00019d90: 7273 2929 2069 6620 6c65 6e28 7061 6972  rs)) if len(pair
+00019da0: 2920 3d3d 2032 5d29 0a20 2020 2020 2020  ) == 2]).       
+00019db0: 2020 2020 2070 6169 7273 203d 206e 702e       pairs = np.
+00019dc0: 6172 7261 7928 7061 6972 7329 0a0a 2020  array(pairs)..  
+00019dd0: 2020 2020 2020 2320 6f74 6865 7277 6973        # otherwis
+00019de0: 652c 2063 686f 6f73 6520 6d61 7465 7320  e, choose mates 
+00019df0: 7573 696e 6720 6d61 7469 6e67 2072 6164  using mating rad
+00019e00: 6975 730a 2020 2020 2020 2020 656c 7365  ius.        else
+00019e10: 3a0a 2020 2020 2020 2020 2020 2020 2369  :.            #i
+00019e20: 6620 6e65 6967 6862 6f72 7320 6172 6520  f neighbors are 
+00019e30: 746f 2062 6520 666f 756e 6420 7769 7468  to be found with
+00019e40: 696e 2074 6865 2073 7065 6369 6573 2c0a  in the species,.
+00019e50: 2020 2020 2020 2020 2020 2020 2373 6574              #set
+00019e60: 2063 6f6f 7264 7320 746f 2073 656c 662e   coords to self.
+00019e70: 5f63 6f6f 7264 7320 286f 7468 6572 7769  _coords (otherwi
+00019e80: 7365 2c20 7468 6520 636f 6f72 6473 2074  se, the coords t
+00019e90: 6f0a 2020 2020 2020 2020 2020 2020 2366  o.            #f
+00019ea0: 696e 6420 6e65 6172 6573 7420 6e65 6967  ind nearest neig
+00019eb0: 6862 6f72 7320 7769 7468 2073 686f 756c  hbors with shoul
+00019ec0: 6420 6861 7665 2062 6565 6e20 7072 6f76  d have been prov
+00019ed0: 6964 6564 290a 2020 2020 2020 2020 2020  ided).          
+00019ee0: 2020 6966 2077 6974 6869 6e3a 0a20 2020    if within:.   
+00019ef0: 2020 2020 2020 2020 2020 2020 2063 6f6f               coo
+00019f00: 7264 7320 3d20 7365 6c66 2e5f 636f 6f72  rds = self._coor
+00019f10: 6473 0a0a 2020 2020 2020 2020 2020 2020  ds..            
+00019f20: 2371 7565 7279 2074 6865 2074 7265 6520  #query the tree 
+00019f30: 746f 2067 6574 206d 6174 696e 6720 7061  to get mating pa
+00019f40: 6972 730a 2020 2020 2020 2020 2020 2020  irs.            
+00019f50: 7061 6972 7320 3d20 7365 6c66 2e5f 6b64  pairs = self._kd
+00019f60: 5f74 7265 652e 5f67 6574 5f6d 6174 696e  _tree._get_matin
+00019f70: 675f 7061 6972 7328 636f 6f72 6473 3d63  g_pairs(coords=c
+00019f80: 6f6f 7264 732c 0a20 2020 2020 2020 2020  oords,.         
+00019f90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019fa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019fb0: 2020 2020 2020 2064 6973 743d 7365 6c66         dist=self
+00019fc0: 2e6d 6174 696e 675f 7261 6469 7573 2c0a  .mating_radius,.
 00019fd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00019fe0: 2020 2020 2320 746f 2073 746f 7265 2069      # to store i
-00019ff0: 6e64 6976 6964 7561 6c73 2720 7068 656e  ndividuals' phen
-0001a000: 6f74 7970 6573 2061 6e64 2066 6974 6e65  otypes and fitne
-0001a010: 7373 0a20 2020 2020 2020 2020 2020 2020  ss.             
-0001a020: 2020 2020 2020 2061 782e 706c 6f74 286c         ax.plot(l
-0001a030: 6f63 735b 6e3a 6e2b 322c 2030 5d2c 206c  ocs[n:n+2, 0], l
-0001a040: 6f63 735b 6e3a 6e2b 322c 2031 5d2c 206c  ocs[n:n+2, 1], l
-0001a050: 696e 6573 7479 6c65 3d27 736f 6c69 6427  inestyle='solid'
-0001a060: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-0001a070: 2020 2020 2020 2020 2020 2020 2020 6d61                ma
-0001a080: 726b 6572 3d27 6f27 2c20 6d61 726b 6572  rker='o', marker
-0001a090: 7369 7a65 3d73 697a 652a 2a28 312f 3229  size=size**(1/2)
-0001a0a0: 2c20 636f 6c6f 723d 636f 6c2c 0a20 2020  , color=col,.   
-0001a0b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001a0c0: 2020 2020 2020 2020 206c 696e 6577 6964           linewid
-0001a0d0: 7468 3d6c 696e 6577 6964 7468 735b 6e5d  th=linewidths[n]
-0001a0e0: 2c20 616c 7068 613d 616c 7068 6129 0a20  , alpha=alpha). 
-0001a0f0: 2020 2020 2020 2020 2020 2065 6c69 6620             elif 
-0001a100: 7374 796c 6520 3d3d 2027 7665 6374 6f72  style == 'vector
-0001a110: 273a 0a20 2020 2020 2020 2020 2020 2020  ':.             
-0001a120: 2020 2023 2067 6574 2074 6865 2073 7461     # get the sta
-0001a130: 7274 2061 6e64 2065 6e64 206c 6f63 6174  rt and end locat
-0001a140: 696f 6e73 0a20 2020 2020 2020 2020 2020  ions.           
-0001a150: 2020 2020 2023 204e 4f54 4520 6e65 6564       # NOTE need
-0001a160: 2074 6f20 7461 6b65 206f 6e6c 7920 7468   to take only th
-0001a170: 6520 6669 7273 7420 3220 7661 6c75 6573  e first 2 values
-0001a180: 2069 6e20 7468 6520 6c6f 6361 7469 6f6e   in the location
-0001a190: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0001a1a0: 2023 2064 6174 6120 6265 6361 7573 6520   # data because 
-0001a1b0: 7375 6273 6571 7565 6e74 2076 616c 7565  subsequent value
-0001a1c0: 7320 6172 6520 7573 6564 2074 6f20 7374  s are used to st
-0001a1d0: 6f72 6520 696e 6469 7669 6475 616c 7327  ore individuals'
-0001a1e0: 200a 2020 2020 2020 2020 2020 2020 2020   .              
-0001a1f0: 2020 2320 7068 656e 6f74 7970 6573 2061    # phenotypes a
-0001a200: 6e64 2066 6974 6e65 7373 0a20 2020 2020  nd fitness.     
-0001a210: 2020 2020 2020 2020 2020 2062 6567 5f6c             beg_l
-0001a220: 6f63 203d 205b 2a6c 696e 5f64 6963 745b  oc = [*lin_dict[
-0001a230: 6e6f 6465 5d2e 7661 6c75 6573 2829 5d5b  node].values()][
-0001a240: 2d31 5d5b 315d 5b3a 325d 0a20 2020 2020  -1][1][:2].     
-0001a250: 2020 2020 2020 2020 2020 2065 6e64 5f6c             end_l
-0001a260: 6f63 203d 205b 2a6c 696e 5f64 6963 745b  oc = [*lin_dict[
-0001a270: 6e6f 6465 5d2e 7661 6c75 6573 2829 5d5b  node].values()][
-0001a280: 305d 5b31 5d5b 3a32 5d0a 2020 2020 2020  0][1][:2].      
-0001a290: 2020 2020 2020 2020 2020 6478 2c20 6479            dx, dy
-0001a2a0: 203d 205b 656e 645f 6c6f 635b 695d 202d   = [end_loc[i] -
-0001a2b0: 2062 6567 5f6c 6f63 5b69 5d20 666f 7220   beg_loc[i] for 
-0001a2c0: 6920 696e 2072 616e 6765 2832 295d 0a20  i in range(2)]. 
-0001a2d0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
-0001a2e0: 2070 6c6f 7420 7468 6520 7665 6374 6f72   plot the vector
-0001a2f0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0001a300: 2023 204e 4f54 453a 2053 484f 554c 4420   # NOTE: SHOULD 
-0001a310: 4920 4245 2046 4954 5449 4e47 2041 2052  I BE FITTING A R
-0001a320: 4547 5245 5353 494f 4e20 4c49 4e45 2054  EGRESSION LINE T
-0001a330: 4f20 5448 4520 5820 414e 4420 590a 2020  O THE X AND Y.  
-0001a340: 2020 2020 2020 2020 2020 2020 2020 2320                # 
-0001a350: 4c4f 4341 5449 4f4e 532c 2054 4f20 4745  LOCATIONS, TO GE
-0001a360: 5420 5448 4520 2741 5645 5241 4745 2720  T THE 'AVERAGE' 
-0001a370: 5645 4354 4f52 2c20 5241 5448 4552 2054  VECTOR, RATHER T
-0001a380: 4841 4e20 4a55 5354 0a20 2020 2020 2020  HAN JUST.       
-0001a390: 2020 2020 2020 2020 2023 2050 4c4f 5454           # PLOTT
-0001a3a0: 494e 4720 5448 4520 5645 4354 4f52 2042  ING THE VECTOR B
-0001a3b0: 4554 5745 454e 2054 4845 204f 4c44 4553  ETWEEN THE OLDES
-0001a3c0: 5420 414e 4420 4355 5252 454e 5420 504f  T AND CURRENT PO
-0001a3d0: 5349 5449 4f4e 530a 2020 2020 2020 2020  SITIONS.        
-0001a3e0: 2020 2020 2020 2020 2320 2857 4849 4348          # (WHICH
-0001a3f0: 2043 4f55 4c44 2045 4153 494c 5920 4d49   COULD EASILY MI
-0001a400: 5352 4550 5245 5345 4e54 2054 4845 204f  SREPRESENT THE O
-0001a410: 5645 5241 4c4c 2054 5245 4e44 2042 4543  VERALL TREND BEC
-0001a420: 4155 5345 200a 2020 2020 2020 2020 2020  AUSE .          
-0001a430: 2020 2020 2020 2320 4f46 2043 4841 4e43        # OF CHANC
-0001a440: 4520 4154 5950 4943 414c 204c 4f43 4154  E ATYPICAL LOCAT
-0001a450: 494f 4e53 2046 4f52 2045 4954 4845 5220  IONS FOR EITHER 
-0001a460: 4f46 2054 484f 5345 2054 574f 2050 4f53  OF THOSE TWO POS
-0001a470: 4954 494f 4e53 0a20 2020 2020 2020 2020  ITIONS.         
-0001a480: 2020 2020 2020 2061 782e 6172 726f 7728         ax.arrow(
-0001a490: 2a62 6567 5f6c 6f63 2c20 6478 2c20 6479  *beg_loc, dx, dy
-0001a4a0: 2c20 636f 6c6f 723d 7374 6172 745f 636f  , color=start_co
-0001a4b0: 6c2c 0a20 2020 2020 2020 2020 2020 2020  l,.             
-0001a4c0: 2020 2020 2020 2020 2020 2020 7769 6474              widt
-0001a4d0: 683d 302e 3035 2c20 6865 6164 5f77 6964  h=0.05, head_wid
-0001a4e0: 7468 3d30 2e34 2c20 6c65 6e67 7468 5f69  th=0.4, length_i
-0001a4f0: 6e63 6c75 6465 735f 6865 6164 3d54 7275  ncludes_head=Tru
-0001a500: 6529 0a20 2020 2020 2020 2020 2020 2023  e).            #
-0001a510: 2070 6c6f 7420 7468 6520 6e6f 6465 7327   plot the nodes'
-0001a520: 2063 7572 7265 6e74 206c 6f63 6174 696f   current locatio
-0001a530: 6e73 2061 6e64 2074 6865 6972 2062 6972  ns and their bir
-0001a540: 7468 206c 6f63 6174 696f 6e73 2c0a 2020  th locations,.  
-0001a550: 2020 2020 2020 2020 2020 2320 636f 6e6e            # conn
-0001a560: 6563 7465 6420 6279 2061 2074 6869 6e20  ected by a thin 
-0001a570: 626c 6163 6b20 6c69 6e65 0a20 2020 2020  black line.     
-0001a580: 2020 2020 2020 206e 6f64 655f 6375 7272         node_curr
-0001a590: 5f6c 6f63 203d 206e 6f64 655f 6375 7272  _loc = node_curr
-0001a5a0: 5f6c 6f63 735b 695d 0a20 2020 2020 2020  _locs[i].       
-0001a5b0: 2020 2020 206e 6f64 655f 6269 7274 685f       node_birth_
-0001a5c0: 6c6f 6320 3d20 5b2a 6c69 6e5f 6469 6374  loc = [*lin_dict
-0001a5d0: 5b6e 6f64 655d 2e76 616c 7565 7328 295d  [node].values()]
-0001a5e0: 5b30 5d5b 315d 0a20 2020 2020 2020 2020  [0][1].         
-0001a5f0: 2020 2070 6c74 2e70 6c6f 7428 5b6e 6f64     plt.plot([nod
-0001a600: 655f 6269 7274 685f 6c6f 635b 305d 2c20  e_birth_loc[0], 
-0001a610: 6e6f 6465 5f63 7572 725f 6c6f 635b 305d  node_curr_loc[0]
-0001a620: 5d2c 0a20 2020 2020 2020 2020 2020 2020  ],.             
-0001a630: 2020 2020 2020 2020 5b6e 6f64 655f 6269          [node_bi
-0001a640: 7274 685f 6c6f 635b 315d 2c20 6e6f 6465  rth_loc[1], node
-0001a650: 5f63 7572 725f 6c6f 635b 315d 5d2c 0a20  _curr_loc[1]],. 
-0001a660: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001a670: 2020 2020 636f 6c6f 723d 7374 6172 745f      color=start_
-0001a680: 636f 6c2c 206c 696e 6573 7479 6c65 3d27  col, linestyle='
-0001a690: 3a27 2c20 616c 7068 613d 616c 7068 612c  :', alpha=alpha,
-0001a6a0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0001a6b0: 2020 2020 2020 6c69 6e65 7769 6474 683d        linewidth=
-0001a6c0: 312e 3229 0a0a 2020 2020 2020 2020 6966  1.2)..        if
-0001a6d0: 2061 6464 5f72 6f6f 7473 3a0a 2020 2020   add_roots:.    
-0001a6e0: 2020 2020 2020 2020 7365 6c66 2e5f 706c          self._pl
-0001a6f0: 6f74 5f6c 696e 6561 6765 5f72 6f6f 7473  ot_lineage_roots
-0001a700: 2874 632c 2074 7265 6529 0a0a 0a20 2020  (tc, tree)...   
-0001a710: 2023 2070 6c6f 7420 7468 6520 6c69 6e65   # plot the line
-0001a720: 6167 6520 666f 7220 6120 6769 7665 6e20  age for a given 
-0001a730: 6e6f 6465 2061 6e64 206c 6f63 7573 0a20  node and locus. 
-0001a740: 2020 2064 6566 205f 706c 6f74 5f6c 696e     def _plot_lin
-0001a750: 6561 6765 5f72 6f6f 7473 2873 656c 662c  eage_roots(self,
-0001a760: 2074 632c 2074 7265 652c 2061 6c70 6861   tc, tree, alpha
-0001a770: 3d30 2e38 2c20 7369 7a65 3d37 3529 3a0a  =0.8, size=75):.
-0001a780: 2020 2020 2020 2020 6966 206e 6f74 2073          if not s
-0001a790: 656c 662e 6765 6e5f 6172 6368 2e75 7365  elf.gen_arch.use
-0001a7a0: 5f74 736b 6974 3a0a 2020 2020 2020 2020  _tskit:.        
-0001a7b0: 2020 2020 7261 6973 6520 4578 6365 7074      raise Except
-0001a7c0: 696f 6e28 2822 4c69 6e65 6167 6520 726f  ion(("Lineage ro
-0001a7d0: 6f74 7320 6361 6e6e 6f74 2062 6520 706c  ots cannot be pl
-0001a7e0: 6f74 7465 6420 666f 7220 6120 5370 6563  otted for a Spec
-0001a7f0: 6965 7320 220a 2020 2020 2020 2020 2020  ies ".          
-0001a800: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001a810: 2020 2022 7768 6f73 6520 7370 6174 6961     "whose spatia
-0001a820: 6c20 7065 6469 6772 6565 2069 7320 6e6f  l pedigree is no
-0001a830: 7420 6265 696e 6720 7472 6163 6b65 6420  t being tracked 
-0001a840: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
-0001a850: 2020 2020 2020 2020 2020 2020 2020 2022                 "
-0001a860: 6279 2074 736b 6974 2e22 2929 0a0a 2020  by tskit."))..  
-0001a870: 2020 2020 2020 2320 6765 7420 7468 6520        # get the 
-0001a880: 6e6f 6465 730a 2020 2020 2020 2020 616c  nodes.        al
-0001a890: 6c5f 6e6f 6465 7320 3d20 7365 6c66 2e5f  l_nodes = self._
-0001a8a0: 6765 745f 6e6f 6465 7328 290a 0a20 2020  get_nodes()..   
-0001a8b0: 2020 2020 2023 2067 6574 2074 6865 206c       # get the l
-0001a8c0: 696e 6561 6765 2064 6963 7420 666f 7220  ineage dict for 
-0001a8d0: 616c 6c20 6e6f 6465 730a 2020 2020 2020  all nodes.      
-0001a8e0: 2020 6c69 6e5f 6469 6374 203d 205f 6765    lin_dict = _ge
-0001a8f0: 745f 6c69 6e65 6167 655f 6469 6374 735f  t_lineage_dicts_
-0001a900: 6f6e 655f 7472 6565 2874 632c 2074 7265  one_tree(tc, tre
-0001a910: 652c 2061 6c6c 5f6e 6f64 6573 2c20 7365  e, all_nodes, se
-0001a920: 6c66 2e74 290a 0a20 2020 2020 2020 2023  lf.t)..        #
-0001a930: 2067 6574 2074 6865 2072 6f6f 7473 2066   get the roots f
-0001a940: 6f72 2061 6c6c 2064 6973 7469 6e63 7420  or all distinct 
-0001a950: 6c69 6e65 6167 6573 2061 7420 7468 6973  lineages at this
-0001a960: 206c 6f63 7573 0a20 2020 2020 2020 2072   locus.        r
-0001a970: 6f6f 745f 6e6f 6465 7320 3d20 6e70 2e75  oot_nodes = np.u
-0001a980: 6e69 7175 6528 5b5b 2a6c 696e 5f64 6963  nique([[*lin_dic
-0001a990: 745b 6e5d 2e6b 6579 7328 295d 5b2d 315d  t[n].keys()][-1]
-0001a9a0: 2066 6f72 206e 2069 6e20 616c 6c5f 6e6f   for n in all_no
-0001a9b0: 6465 735d 290a 0a20 2020 2020 2020 2023  des])..        #
-0001a9c0: 2067 6574 2062 6972 7468 206c 6f63 6174   get birth locat
-0001a9d0: 696f 6e73 2066 6f72 2065 6163 6820 726f  ions for each ro
-0001a9e0: 6f74 206e 6f64 650a 2020 2020 2020 2020  ot node.        
-0001a9f0: 726f 6f74 5f69 6e64 6976 6964 7320 3d20  root_individs = 
-0001aa00: 5b74 632e 6e6f 6465 735b 6e5d 2e69 6e64  [tc.nodes[n].ind
-0001aa10: 6976 6964 7561 6c20 666f 7220 6e20 696e  ividual for n in
-0001aa20: 2072 6f6f 745f 6e6f 6465 735d 0a20 2020   root_nodes].   
-0001aa30: 2020 2020 2072 6f6f 745f 6c6f 6373 203d       root_locs =
-0001aa40: 205b 7463 2e69 6e64 6976 6964 7561 6c73   [tc.individuals
-0001aa50: 5b69 5d2e 6c6f 6361 7469 6f6e 2066 6f72  [i].location for
-0001aa60: 2069 2069 6e20 726f 6f74 5f69 6e64 6976   i in root_indiv
-0001aa70: 6964 735d 0a0a 2020 2020 2020 2020 2320  ids]..        # 
-0001aa80: 6578 7472 6163 7420 616e 6420 706c 6f74  extract and plot
-0001aa90: 2074 6865 2073 6572 6965 7320 6f66 2070   the series of p
-0001aaa0: 6f69 6e74 7320 666f 7220 6561 6368 206e  oints for each n
-0001aab0: 6f64 650a 2020 2020 2020 2020 666f 7220  ode.        for 
-0001aac0: 782c 2079 2069 6e20 726f 6f74 5f6c 6f63  x, y in root_loc
-0001aad0: 733a 0a20 2020 2020 2020 2020 2020 2023  s:.            #
-0001aae0: 2070 6c6f 7420 7468 6520 726f 6f74 206e   plot the root n
-0001aaf0: 6f64 6573 2720 6269 7274 6820 6c6f 6361  odes' birth loca
-0001ab00: 7469 6f6e 730a 2020 2020 2020 2020 2020  tions.          
-0001ab10: 2020 706c 742e 7363 6174 7465 7228 782c    plt.scatter(x,
-0001ab20: 2079 2c20 633d 2777 6869 7465 272c 2073   y, c='white', s
-0001ab30: 3d73 697a 652c 2061 6c70 6861 3d61 6c70  =size, alpha=alp
-0001ab40: 6861 2c20 6d61 726b 6572 3d27 7327 290a  ha, marker='s').
-0001ab50: 2020 2020 2020 2020 706c 742e 7368 6f77          plt.show
-0001ab60: 2829 0a0a 0a20 2020 2023 206d 6574 686f  ()...    # metho
-0001ab70: 6420 666f 7220 706c 6f74 7469 6e67 2061  d for plotting a
-0001ab80: 2073 7065 6369 6573 2720 706f 7075 6c61   species' popula
-0001ab90: 7469 6f6e 2070 7972 616d 6964 0a20 2020  tion pyramid.   
-0001aba0: 2064 6566 205f 706c 6f74 5f64 656d 6f67   def _plot_demog
-0001abb0: 7261 7068 6963 5f70 7972 616d 6964 2873  raphic_pyramid(s
-0001abc0: 656c 6629 3a0a 2020 2020 2020 2020 236d  elf):.        #m
-0001abd0: 616b 6520 6469 6374 206f 6620 6665 6d61  ake dict of fema
-0001abe0: 6c65 2061 6e64 206d 616c 6520 636f 6c6f  le and male colo
-0001abf0: 7273 0a20 2020 2020 2020 2063 6f6c 5f64  rs.        col_d
-0001ac00: 6963 7420 3d20 7b2d 313a 2027 6379 616e  ict = {-1: 'cyan
-0001ac10: 272c 2031 3a20 2770 696e 6b27 7d0a 2020  ', 1: 'pink'}.  
-0001ac20: 2020 2020 2020 2363 7265 6174 6520 6120        #create a 
-0001ac30: 6669 6775 7265 0a20 2020 2020 2020 2066  figure.        f
-0001ac40: 6967 203d 2070 6c74 2e66 6967 7572 6528  ig = plt.figure(
-0001ac50: 290a 2020 2020 2020 2020 2376 6172 6961  ).        #varia
-0001ac60: 626c 6573 2074 6f20 6772 6162 2074 6865  bles to grab the
-0001ac70: 206d 6178 2063 6f75 6e74 0a20 2020 2020   max count.     
-0001ac80: 2020 206d 6178 5f63 6f75 6e74 203d 2030     max_count = 0
-0001ac90: 0a20 2020 2020 2020 2023 666f 7220 6561  .        #for ea
-0001aca0: 6368 2073 6578 0a20 2020 2020 2020 2066  ch sex.        f
-0001acb0: 6f72 2073 6578 5f76 616c 2069 6e20 5b2a  or sex_val in [*
-0001acc0: 636f 6c5f 6469 6374 5d3a 0a20 2020 2020  col_dict]:.     
-0001acd0: 2020 2020 2020 2023 6765 7420 6120 636f         #get a co
-0001ace0: 756e 7465 720a 2020 2020 2020 2020 2020  unter.          
-0001acf0: 2020 636f 756e 7473 203d 2043 285b 696e    counts = C([in
-0001ad00: 642e 6167 6520 666f 7220 696e 6420 696e  d.age for ind in
-0001ad10: 2073 656c 662e 7661 6c75 6573 280a 2020   self.values(.  
+00019fe0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00019ff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001a000: 6368 6f6f 7365 5f6e 6561 7265 7374 3d63  choose_nearest=c
+0001a010: 686f 6f73 655f 6e65 6172 6573 742c 0a20  hoose_nearest,. 
+0001a020: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001a030: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001a040: 2020 2020 2020 2069 6e76 6572 7365 5f64         inverse_d
+0001a050: 6973 745f 6d61 7469 6e67 3d69 6e76 6572  ist_mating=inver
+0001a060: 7365 5f64 6973 745f 6d61 7469 6e67 290a  se_dist_mating).
+0001a070: 0a20 2020 2020 2020 2020 2020 2023 2075  .            # u
+0001a080: 7365 2074 6865 2073 7065 6369 6573 2720  se the species' 
+0001a090: 6269 7274 6820 7261 7465 2074 6f20 6465  birth rate to de
+0001a0a0: 6369 6465 2028 6173 2062 6572 6e6f 756c  cide (as bernoul
+0001a0b0: 6c69 2064 7261 7773 290a 2020 2020 2020  li draws).      
+0001a0c0: 2020 2020 2020 2320 7768 6574 6865 7220        # whether 
+0001a0d0: 6561 6368 2070 6169 7220 6361 6e20 6d61  each pair can ma
+0001a0e0: 7465 0a20 2020 2020 2020 2020 2020 2063  te.            c
+0001a0f0: 616e 5f6d 6174 6520 3d20 6e70 2e62 6f6f  an_mate = np.boo
+0001a100: 6c38 286e 702e 7261 6e64 6f6d 2e62 696e  l8(np.random.bin
+0001a110: 6f6d 6961 6c28 6e3d 312c 2070 3d73 656c  omial(n=1, p=sel
+0001a120: 662e 622c 0a20 2020 2020 2020 2020 2020  f.b,.           
+0001a130: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001a140: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001a150: 2020 2020 2020 2020 7369 7a65 3d70 6169          size=pai
+0001a160: 7273 2e73 6861 7065 5b30 5d29 290a 2020  rs.shape[0])).  
+0001a170: 2020 2020 2020 2020 2020 7061 6972 7320            pairs 
+0001a180: 3d20 7061 6972 735b 6361 6e5f 6d61 7465  = pairs[can_mate
+0001a190: 2c20 3a5d 0a20 2020 2020 2020 2072 6574  , :].        ret
+0001a1a0: 7572 6e20 7061 6972 730a 0a0a 2020 2020  urn pairs...    
+0001a1b0: 6465 6620 5f6d 616b 655f 6765 615f 6466  def _make_gea_df
+0001a1c0: 2873 656c 662c 206c 7972 5f6e 756d 3d31  (self, lyr_num=1
+0001a1d0: 293a 0a20 2020 2020 2020 2022 2222 0a0a  ):.        """..
+0001a1e0: 2020 2020 2020 2020 4d61 6b65 7320 6120          Makes a 
+0001a1f0: 6461 7461 6672 616d 6520 666f 7220 4745  dataframe for GE
+0001a200: 4120 616e 616c 7973 6973 2063 6f6e 7461  A analysis conta
+0001a210: 696e 696e 6720 6765 6e6f 7479 7065 732c  ining genotypes,
+0001a220: 2063 6f6f 7264 696e 6174 6573 2c0a 2020   coordinates,.  
+0001a230: 2020 2020 2020 616e 6420 656e 7620 7661        and env va
+0001a240: 6c75 6573 2066 6f72 2061 6c6c 2069 6e64  lues for all ind
+0001a250: 6976 6964 7561 6c73 2e0a 0a20 2020 2020  ividuals...     
+0001a260: 2020 204e 4f54 453a 2043 7572 7265 6e74     NOTE: Current
+0001a270: 6c79 206f 6e6c 7920 6163 6365 7074 7320  ly only accepts 
+0001a280: 6f6e 6520 4c61 6e64 7363 6170 6520 4c61  one Landscape La
+0001a290: 7965 7220 2869 2e65 2e20 656e 7669 726f  yer (i.e. enviro
+0001a2a0: 6e6d 656e 7461 6c0a 2020 2020 2020 2020  nmental.        
+0001a2b0: 2020 2020 2020 7661 7269 6162 6c65 290a        variable).
+0001a2c0: 0a20 2020 2020 2020 2050 6172 616d 6574  .        Paramet
+0001a2d0: 6572 730a 2020 2020 2020 2020 2d2d 2d2d  ers.        ----
+0001a2e0: 2d2d 2d2d 2d2d 0a20 2020 2020 2020 206c  ------.        l
+0001a2f0: 7972 5f6e 756d 203a 2069 6e74 0a20 2020  yr_num : int.   
+0001a300: 2020 2020 2020 2020 2054 6865 206e 756d           The num
+0001a310: 6265 7220 6f66 2074 6865 204c 616e 6473  ber of the Lands
+0001a320: 6361 7065 204c 6179 6572 2066 726f 6d20  cape Layer from 
+0001a330: 7768 6963 6820 746f 2065 7874 7261 6374  which to extract
+0001a340: 0a20 2020 2020 2020 2020 2020 2065 6e76  .            env
+0001a350: 6972 6f6e 6d65 6e74 616c 2076 616c 7565  ironmental value
+0001a360: 732e 2044 6566 6175 6c74 7320 746f 2031  s. Defaults to 1
+0001a370: 2e0a 0a20 2020 2020 2020 2052 6574 7572  ...        Retur
+0001a380: 6e73 0a20 2020 2020 2020 202d 2d2d 2d2d  ns.        -----
+0001a390: 2d2d 0a20 2020 2020 2020 206f 7574 203a  --.        out :
+0001a3a0: 2070 616e 6461 732e 4461 7461 4672 616d   pandas.DataFram
+0001a3b0: 650a 2020 2020 2020 2020 2020 2020 4120  e.            A 
+0001a3c0: 4461 7461 4672 616d 6520 696e 2077 6869  DataFrame in whi
+0001a3d0: 6368 2063 6f6c 756d 6e73 2061 7265 206c  ch columns are l
+0001a3e0: 6f63 6920 616e 6420 726f 7773 2061 7265  oci and rows are
+0001a3f0: 2069 6e64 6976 6964 7561 6c73 0a20 2020   individuals.   
+0001a400: 2020 2020 2022 2222 0a20 2020 2020 2020       """.       
+0001a410: 2023 6765 7420 6172 7261 7920 6f66 205b   #get array of [
+0001a420: 307c 302e 357c 315d 2067 656e 6f74 7970  0|0.5|1] genotyp
+0001a430: 6573 0a20 2020 2020 2020 2022 2222 0a20  es.        """. 
+0001a440: 2020 2020 2020 2067 656e 203d 2073 656c         gen = sel
+0001a450: 662e 5f67 6574 5f67 656e 6f74 7970 6573  f._get_genotypes
+0001a460: 2829 0a0a 2020 2020 2020 2020 236c 6f6f  ()..        #loo
+0001a470: 7020 746f 2063 6f6e 7665 7274 2062 696e  p to convert bin
+0001a480: 6172 7920 616c 6c65 6c65 2067 656e 6f74  ary allele genot
+0001a490: 7970 6573 2028 307c 3129 0a20 2020 2020  ypes (0|1).     
+0001a4a0: 2020 2023 696e 746f 2073 696e 676c 6520     #into single 
+0001a4b0: 6469 6769 7420 6765 6e6f 7479 7065 7320  digit genotypes 
+0001a4c0: 2830 2c20 302e 352c 2031 290a 2020 2020  (0, 0.5, 1).    
+0001a4d0: 2020 2020 6765 6e6f 5f69 6e64 203d 205b      geno_ind = [
+0001a4e0: 5d0a 2020 2020 2020 2020 666f 7220 696e  ].        for in
+0001a4f0: 6420 696e 2072 616e 6765 286c 656e 2867  d in range(len(g
+0001a500: 656e 2929 3a0a 2020 2020 2020 2020 2020  en)):.          
+0001a510: 2020 6765 6e6f 5f6c 6f63 203d 205b 5d0a    geno_loc = [].
+0001a520: 2020 2020 2020 2020 2020 2020 666f 7220              for 
+0001a530: 6c6f 6369 2069 6e20 7261 6e67 6528 6c65  loci in range(le
+0001a540: 6e28 6765 6e5b 696e 645d 2929 3a0a 2020  n(gen[ind])):.  
+0001a550: 2020 2020 2020 2020 2020 2020 2020 6765                ge
+0001a560: 6e6f 7479 7065 203d 2067 656e 5b69 6e64  notype = gen[ind
+0001a570: 5d5b 6c6f 6369 5d2e 6d65 616e 2829 2023  ][loci].mean() #
+0001a580: 2063 6f64 6564 2061 7320 302f 302e 352f   coded as 0/0.5/
+0001a590: 310a 2020 2020 2020 2020 2020 2020 2020  1.              
+0001a5a0: 2020 6765 6e6f 5f6c 6f63 2e61 7070 656e    geno_loc.appen
+0001a5b0: 6428 6765 6e6f 7479 7065 290a 2020 2020  d(genotype).    
+0001a5c0: 2020 2020 2020 2020 6765 6e6f 5f69 6e64          geno_ind
+0001a5d0: 2e61 7070 656e 6428 6765 6e6f 5f6c 6f63  .append(geno_loc
+0001a5e0: 290a 2020 2020 2020 2020 2222 220a 2020  ).        """.  
+0001a5f0: 2020 2020 2020 6774 7320 3d20 7365 6c66        gts = self
+0001a600: 2e5f 6765 745f 6765 6e6f 7479 7065 7328  ._get_genotypes(
+0001a610: 6269 616c 6c65 6c69 633d 4661 6c73 6529  biallelic=False)
+0001a620: 0a0a 2020 2020 2020 2020 2363 6f6e 7665  ..        #conve
+0001a630: 7274 2074 6f20 6461 7461 6672 616d 650a  rt to dataframe.
+0001a640: 2020 2020 2020 2020 6765 615f 6466 203d          gea_df =
+0001a650: 2070 642e 4461 7461 4672 616d 6528 6774   pd.DataFrame(gt
+0001a660: 7329 0a0a 2020 2020 2020 2020 2367 6574  s)..        #get
+0001a670: 2065 6e76 6972 6f6e 6d65 6e74 616c 2064   environmental d
+0001a680: 6174 610a 2020 2020 2020 2020 656e 7620  ata.        env 
+0001a690: 3d20 7365 6c66 2e5f 6765 745f 6528 290a  = self._get_e().
+0001a6a0: 2020 2020 2020 2020 6765 615f 6466 5b27          gea_df['
+0001a6b0: 656e 7627 5d20 3d20 6c69 7374 2865 6e76  env'] = list(env
+0001a6c0: 5b3a 2c6c 7972 5f6e 756d 5d29 0a0a 2020  [:,lyr_num])..  
+0001a6d0: 2020 2020 2020 2367 6574 2063 6f6f 7264        #get coord
+0001a6e0: 730a 2020 2020 2020 2020 636f 6f72 6420  s.        coord 
+0001a6f0: 3d20 7365 6c66 2e5f 6765 745f 636f 6f72  = self._get_coor
+0001a700: 6473 2829 0a20 2020 2020 2020 2067 6561  ds().        gea
+0001a710: 5f64 665b 276c 6174 275d 203d 206c 6973  _df['lat'] = lis
+0001a720: 7428 636f 6f72 645b 3a2c 305d 290a 2020  t(coord[:,0]).  
+0001a730: 2020 2020 2020 6765 615f 6466 5b27 6c6f        gea_df['lo
+0001a740: 6e67 275d 203d 206c 6973 7428 636f 6f72  ng'] = list(coor
+0001a750: 645b 3a2c 315d 290a 0a20 2020 2020 2020  d[:,1])..       
+0001a760: 2072 6574 7572 6e20 6765 615f 6466 0a0a   return gea_df..
+0001a770: 0a20 2020 2064 6566 205f 7275 6e5f 6363  .    def _run_cc
+0001a780: 6128 7365 6c66 2c20 7472 745f 6e75 6d3d  a(self, trt_num=
+0001a790: 302c 2073 6361 6c65 3d54 7275 652c 2070  0, scale=True, p
+0001a7a0: 6c6f 743d 5472 7565 2c20 706c 6f74 5f73  lot=True, plot_s
+0001a7b0: 643d 5472 7565 2c0a 2020 2020 2020 2020  d=True,.        
+0001a7c0: 2020 2020 2020 2020 2073 643d 3329 3a0a           sd=3):.
+0001a7d0: 2020 2020 2020 2020 2222 220a 2020 2020          """.    
+0001a7e0: 2020 2020 5275 6e73 2061 2063 616e 6f6e      Runs a canon
+0001a7f0: 6963 616c 2063 6f72 7265 6c61 7469 6f6e  ical correlation
+0001a800: 2061 6e61 6c79 7369 7320 2843 4341 2920   analysis (CCA) 
+0001a810: 6f6e 2074 6865 2063 7572 7265 6e74 2067  on the current g
+0001a820: 656e 6574 6963 2061 6e64 0a20 2020 2020  enetic and.     
+0001a830: 2020 2065 6e76 6972 6f6e 6d65 6e74 616c     environmental
+0001a840: 2064 6174 6120 666f 7220 7468 6520 5370   data for the Sp
+0001a850: 6563 6965 7327 2063 7572 7265 6e74 2070  ecies' current p
+0001a860: 6f70 756c 6174 696f 6e2c 2075 7369 6e67  opulation, using
+0001a870: 2074 6865 0a20 2020 2020 2020 2069 6e64   the.        ind
+0001a880: 6963 6174 6564 2054 7261 6974 2e20 506c  icated Trait. Pl
+0001a890: 6f74 7320 7468 6520 7265 7375 6c74 7320  ots the results 
+0001a8a0: 616e 6420 7265 7475 726e 7320 7468 656d  and returns them
+0001a8b0: 2069 6e20 6120 6469 6374 2e0a 0a20 2020   in a dict...   
+0001a8c0: 2020 2020 2043 4341 206d 6f64 656c 2066       CCA model f
+0001a8d0: 6f72 6d75 6c61 3a20 6765 6e6f 7479 7065  ormula: genotype
+0001a8e0: 207e 2065 6e76 202b 206c 6174 202b 206c   ~ env + lat + l
+0001a8f0: 6f6e 670a 0a20 2020 2020 2020 204e 4f54  ong..        NOT
+0001a900: 453a 2063 7572 7265 6e74 6c79 206f 6e6c  E: currently onl
+0001a910: 7920 706f 7373 6962 6c65 2074 6f20 646f  y possible to do
+0001a920: 2033 2063 6f6d 706f 6e65 6e74 732f 3320   3 components/3 
+0001a930: 7661 7269 6162 6c65 733b 0a20 2020 2020  variables;.     
+0001a940: 2020 2020 2020 2020 2077 696c 6c20 6164           will ad
+0001a950: 6420 6772 6561 7465 7220 6675 6e63 7469  d greater functi
+0001a960: 6f6e 616c 6974 7920 696e 2074 6865 2066  onality in the f
+0001a970: 7574 7572 650a 0a20 2020 2020 2020 2050  uture..        P
+0001a980: 6172 616d 6574 6572 730a 2020 2020 2020  arameters.      
+0001a990: 2020 2d2d 2d2d 2d2d 2d2d 2d2d 0a20 2020    ----------.   
+0001a9a0: 2020 2020 2074 7274 5f6e 756d 203a 2069       trt_num : i
+0001a9b0: 6e74 0a20 2020 2020 2020 2020 2020 2054  nt.            T
+0001a9c0: 6865 206e 756d 6265 7220 6f66 2074 6865  he number of the
+0001a9d0: 2054 7261 6974 2074 6f20 7275 6e20 7468   Trait to run th
+0001a9e0: 6520 4745 4120 6f6e 2e20 4465 6661 756c  e GEA on. Defaul
+0001a9f0: 7473 2074 6f20 302e 0a20 2020 2020 2020  ts to 0..       
+0001aa00: 2073 6361 6c65 203a 2062 6f6f 6c0a 2020   scale : bool.  
+0001aa10: 2020 2020 2020 2020 2020 4966 2054 7275            If Tru
+0001aa20: 652c 2073 6361 6c65 7320 7468 6520 7661  e, scales the va
+0001aa30: 7269 6162 6c65 2c20 696e 6469 7669 6475  riable, individu
+0001aa40: 616c 2c0a 2020 2020 2020 2020 2020 2020  al,.            
+0001aa50: 616e 6420 6c6f 6375 7320 6c6f 6164 696e  and locus loadin
+0001aa60: 6773 2066 726f 6d20 2d31 2074 6f20 3120  gs from -1 to 1 
+0001aa70: 746f 206d 616b 6520 7468 656d 2065 6173  to make them eas
+0001aa80: 6965 720a 2020 2020 2020 2020 2020 2020  ier.            
+0001aa90: 746f 2076 6973 7561 6c69 7a65 2e20 4465  to visualize. De
+0001aaa0: 6661 756c 7473 2074 6f20 5472 7565 2e0a  faults to True..
+0001aab0: 2020 2020 2020 2020 706c 6f74 203a 2062          plot : b
+0001aac0: 6f6f 6c0a 2020 2020 2020 2020 2020 2020  ool.            
+0001aad0: 5768 6574 6865 7220 6f72 206e 6f74 2074  Whether or not t
+0001aae0: 6f20 706c 6f74 2074 6865 206d 6f64 656c  o plot the model
+0001aaf0: 2e20 4465 6661 756c 7473 2074 6f20 5472  . Defaults to Tr
+0001ab00: 7565 2e0a 2020 2020 2020 2020 706c 6f74  ue..        plot
+0001ab10: 5f73 6420 3a20 626f 6f6c 0a20 2020 2020  _sd : bool.     
+0001ab20: 2020 2020 2020 2049 6620 5472 7565 2c20         If True, 
+0001ab30: 6120 7374 616e 6461 7264 2064 6576 6961  a standard devia
+0001ab40: 7469 6f6e 2065 6c6c 6970 7365 2069 7320  tion ellipse is 
+0001ab50: 706c 6f74 7465 642c 2075 7369 6e67 2074  plotted, using t
+0001ab60: 6865 206e 756d 6265 720a 2020 2020 2020  he number.      
+0001ab70: 2020 2020 2020 6f66 2073 7461 6e64 6172        of standar
+0001ab80: 6420 6465 7669 6174 696f 6e73 2069 6e64  d deviations ind
+0001ab90: 6963 6174 6564 2062 7920 7468 6520 6172  icated by the ar
+0001aba0: 6775 6d65 6e74 2027 7364 272e 2044 6566  gument 'sd'. Def
+0001abb0: 6175 6c74 7320 746f 0a20 2020 2020 2020  aults to.       
+0001abc0: 2020 2020 2054 7275 652e 0a20 2020 2020       True..     
+0001abd0: 2020 2073 6420 3a20 7b69 6e74 2c20 666c     sd : {int, fl
+0001abe0: 6f61 747d 0a20 2020 2020 2020 2020 2020  oat}.           
+0001abf0: 204e 756d 6265 7220 6f66 2073 7461 6e64   Number of stand
+0001ac00: 6172 6420 6465 7669 6174 696f 6e73 2074  ard deviations t
+0001ac10: 6f20 7573 6520 666f 7220 706c 6f74 7469  o use for plotti
+0001ac20: 6e67 2074 6865 2073 7461 6e64 6172 640a  ng the standard.
+0001ac30: 2020 2020 2020 2020 2020 2020 6465 7669              devi
+0001ac40: 6174 696f 6e20 656c 6c69 7073 652e 2044  ation ellipse. D
+0001ac50: 6566 6175 6c74 7320 746f 2033 2e0a 0a20  efaults to 3... 
+0001ac60: 2020 2020 2020 2052 6574 7572 6e73 0a20         Returns. 
+0001ac70: 2020 2020 2020 202d 2d2d 2d2d 2d2d 0a20         -------. 
+0001ac80: 2020 2020 2020 206f 7574 203a 2064 6963         out : dic
+0001ac90: 740a 2020 2020 2020 2020 2020 2020 4120  t.            A 
+0001aca0: 6469 6374 206f 6620 7468 6520 666f 6c6c  dict of the foll
+0001acb0: 6f77 696e 6720 6b65 792d 7661 6c75 6520  owing key-value 
+0001acc0: 7061 6972 733a 0a20 2020 2020 2020 2020  pairs:.         
+0001acd0: 2020 2020 2020 2022 6765 615f 6466 223a         "gea_df":
+0001ace0: 2061 2044 6174 6146 7261 6d65 206f 6620   a DataFrame of 
+0001acf0: 696e 6469 7669 6475 616c 7327 2067 656e  individuals' gen
+0001ad00: 6f74 7970 6573 2c20 656e 7669 726f 6e6d  otypes, environm
+0001ad10: 656e 7461 6c0a 2020 2020 2020 2020 2020  ental.          
 0001ad20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001ad30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001ad40: 2020 2020 2020 2020 2020 2920 6966 2069            ) if i
-0001ad50: 6e64 2e73 6578 203d 3d20 696e 7428 7365  nd.sex == int(se
-0001ad60: 785f 7661 6c20 3c20 3029 5d29 0a20 2020  x_val < 0)]).   
-0001ad70: 2020 2020 2020 2020 2023 6772 6162 2074           #grab t
-0001ad80: 6865 2061 6765 7320 6672 6f6d 2069 740a  he ages from it.
-0001ad90: 2020 2020 2020 2020 2020 2020 6167 6573              ages
-0001ada0: 203d 205b 2a63 6f75 6e74 735d 0a20 2020   = [*counts].   
-0001adb0: 2020 2020 2020 2020 2023 616e 6420 6772           #and gr
-0001adc0: 6162 2074 6865 2063 6f75 6e74 7320 6672  ab the counts fr
-0001add0: 6f6d 2069 7420 286d 756c 7469 706c 7969  om it (multiplyi
-0001ade0: 6e67 2062 7920 2d31 2066 6f72 2066 656d  ng by -1 for fem
-0001adf0: 616c 6573 2c0a 2020 2020 2020 2020 2020  ales,.          
-0001ae00: 2020 2374 6f20 7365 7420 6f6e 6520 7365    #to set one se
-0001ae10: 7820 6f6e 2065 6974 6865 7220 7369 6465  x on either side
-0001ae20: 206f 6620 783d 302c 2066 6f72 2074 6865   of x=0, for the
-0001ae30: 2070 7972 616d 6964 290a 2020 2020 2020   pyramid).      
-0001ae40: 2020 2020 2020 636f 756e 7473 203d 205b        counts = [
-0001ae50: 7365 785f 7661 6c2a 636f 756e 7420 666f  sex_val*count fo
-0001ae60: 7220 636f 756e 7420 696e 2063 6f75 6e74  r count in count
-0001ae70: 732e 7661 6c75 6573 2829 5d0a 2020 2020  s.values()].    
-0001ae80: 2020 2020 2020 2020 2375 7064 6174 6520          #update 
-0001ae90: 7468 6520 6d61 785f 636f 756e 7420 7661  the max_count va
-0001aea0: 720a 2020 2020 2020 2020 2020 2020 6d61  r.            ma
-0001aeb0: 785f 636f 756e 7420 3d20 6d61 7828 6d61  x_count = max(ma
-0001aec0: 7828 636f 756e 7473 292c 206d 6178 5f63  x(counts), max_c
-0001aed0: 6f75 6e74 290a 2020 2020 2020 2020 2020  ount).          
-0001aee0: 2020 2374 6865 6e20 6372 6561 7465 2074    #then create t
-0001aef0: 6865 2068 6f72 697a 6f6e 7461 6c20 6261  he horizontal ba
-0001af00: 7270 6c6f 740a 2020 2020 2020 2020 2020  rplot.          
-0001af10: 2020 706c 742e 6261 7268 2861 6765 732c    plt.barh(ages,
-0001af20: 2063 6f75 6e74 732c 2063 6f6c 6f72 203d   counts, color =
-0001af30: 2063 6f6c 5f64 6963 745b 7365 785f 7661   col_dict[sex_va
-0001af40: 6c5d 290a 2020 2020 2020 2020 2375 7365  l]).        #use
-0001af50: 206d 6178 5f63 6f75 6e74 2074 6f20 7365   max_count to se
-0001af60: 7420 7468 6520 782d 6c69 6d69 7473 2061  t the x-limits a
-0001af70: 6e64 2079 2d6c 696d 6974 730a 2020 2020  nd y-limits.    
-0001af80: 2020 2020 706c 742e 786c 696d 2828 2d31      plt.xlim((-1
-0001af90: 2a6d 6178 5f63 6f75 6e74 2d32 2c20 6d61  *max_count-2, ma
-0001afa0: 785f 636f 756e 742b 3229 290a 2020 2020  x_count+2)).    
-0001afb0: 2020 2020 2373 6574 2074 6865 2061 7869      #set the axi
-0001afc0: 7320 6c61 6265 6c73 0a20 2020 2020 2020  s labels.       
-0001afd0: 2070 6c74 2e78 6c61 6265 6c28 2753 7065   plt.xlabel('Spe
-0001afe0: 6369 6573 2028 696e 6469 7669 6475 616c  cies (individual
-0001aff0: 7329 2729 0a20 2020 2020 2020 2070 6c74  s)').        plt
-0001b000: 2e79 6c61 6265 6c28 2741 6765 2028 7469  .ylabel('Age (ti
-0001b010: 6d65 7374 6570 7329 2729 0a20 2020 2020  mesteps)').     
-0001b020: 2020 2023 7468 656e 2073 6574 2074 6865     #then set the
-0001b030: 2078 6c61 6265 6c73 2074 6f20 706f 7369   xlabels to posi
-0001b040: 7469 7665 206e 756d 6265 7273 206f 6e20  tive numbers on 
-0001b050: 6569 7468 6572 2073 6964 650a 2020 2020  either side.    
-0001b060: 2020 2020 6c6f 6373 2c20 6c61 6265 6c73      locs, labels
-0001b070: 203d 2070 6c74 2e78 7469 636b 7328 290a   = plt.xticks().
-0001b080: 2020 2020 2020 2020 706c 742e 7874 6963          plt.xtic
-0001b090: 6b73 286c 6f63 732c 205b 7374 7228 696e  ks(locs, [str(in
-0001b0a0: 7428 6c6f 6329 2920 666f 7220 6c6f 6320  t(loc)) for loc 
-0001b0b0: 696e 206e 702e 6162 7328 6c6f 6373 295d  in np.abs(locs)]
-0001b0c0: 290a 2020 2020 2020 2020 2361 6464 2073  ).        #add s
-0001b0d0: 6578 2073 796d 626f 6c73 2061 7320 7469  ex symbols as ti
-0001b0e0: 746c 650a 2020 2020 2020 2020 706c 742e  tle.        plt.
-0001b0f0: 7375 7074 6974 6c65 2827 5c75 3236 3432  suptitle('\u2642
-0001b100: 2573 5c75 3236 3430 2720 2520 2727 2e6a  %s\u2640' % ''.j
-0001b110: 6f69 6e28 5b27 2027 2066 6f72 205f 2069  oin([' ' for _ i
-0001b120: 6e20 7261 6e67 6528 3230 295d 292c 0a20  n range(20)]),. 
-0001b130: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b140: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b150: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b160: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b170: 2020 2073 697a 6520 3d20 3330 290a 2020     size = 30).  
-0001b180: 2020 2020 2020 2373 686f 7720 6974 0a20        #show it. 
-0001b190: 2020 2020 2020 2070 6c74 2e73 686f 7728         plt.show(
-0001b1a0: 290a 0a20 2020 2064 6566 205f 706c 6f74  )..    def _plot
-0001b1b0: 5f70 6f70 5f67 726f 7774 6828 7365 6c66  _pop_growth(self
-0001b1c0: 2c20 6578 7065 6374 6564 3d54 7275 652c  , expected=True,
-0001b1d0: 2061 6374 7561 6c3d 5472 7565 2c20 6578   actual=True, ex
-0001b1e0: 7065 6374 6564 5f63 6f6c 6f72 3d27 7265  pected_color='re
-0001b1f0: 6427 2c0a 2020 2020 2020 2020 2020 2020  d',.            
-0001b200: 2020 2020 2020 2020 2020 2020 6163 7475              actu
-0001b210: 616c 5f63 6f6c 6f72 3d27 626c 7565 2729  al_color='blue')
-0001b220: 3a0a 2020 2020 2020 2020 5420 3d20 7261  :.        T = ra
-0001b230: 6e67 6528 6c65 6e28 7365 6c66 2e4e 7429  nge(len(self.Nt)
-0001b240: 290a 2020 2020 2020 2020 7830 203d 2073  ).        x0 = s
-0001b250: 656c 662e 4e74 5b30 5d20 2f20 7365 6c66  elf.Nt[0] / self
-0001b260: 2e4b 2e73 756d 2829 0a20 2020 2020 2020  .K.sum().       
-0001b270: 2069 6620 6578 7065 6374 6564 3a0a 2020   if expected:.  
-0001b280: 2020 2020 2020 2020 2020 706c 742e 706c            plt.pl
-0001b290: 6f74 2854 2c20 5b5f 6361 6c63 5f6c 6f67  ot(T, [_calc_log
-0001b2a0: 6973 7469 635f 736f 6c6e 2878 302c 2073  istic_soln(x0, s
-0001b2b0: 656c 662e 522c 0a20 2020 2020 2020 2020  elf.R,.         
-0001b2c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b2d0: 2020 2020 2020 2074 2920 2a20 7365 6c66         t) * self
-0001b2e0: 2e4b 2e73 756d 2829 2066 6f72 2074 2069  .K.sum() for t i
-0001b2f0: 6e20 545d 2c0a 2020 2020 2020 2020 2020  n T],.          
-0001b300: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
-0001b310: 3d65 7870 6563 7465 645f 636f 6c6f 7229  =expected_color)
-0001b320: 0a20 2020 2020 2020 2069 6620 6163 7475  .        if actu
-0001b330: 616c 3a0a 2020 2020 2020 2020 2020 2020  al:.            
-0001b340: 706c 742e 706c 6f74 2854 2c20 7365 6c66  plt.plot(T, self
-0001b350: 2e4e 742c 2063 6f6c 6f72 3d61 6374 7561  .Nt, color=actua
-0001b360: 6c5f 636f 6c6f 7229 0a20 2020 2020 2020  l_color).       
-0001b370: 2070 6c74 2e78 6c61 6265 6c28 2774 2729   plt.xlabel('t')
-0001b380: 0a20 2020 2020 2020 2070 6c74 2e79 6c61  .        plt.yla
-0001b390: 6265 6c28 274e 2874 2927 290a 0a20 2020  bel('N(t)')..   
-0001b3a0: 2064 6566 205f 706c 6f74 5f64 656d 6f67   def _plot_demog
-0001b3b0: 7261 7068 6963 5f63 6861 6e67 6573 2873  raphic_changes(s
-0001b3c0: 656c 6629 3a0a 2020 2020 2020 2020 6966  elf):.        if
-0001b3d0: 2073 656c 662e 5f63 6861 6e67 6572 2069   self._changer i
-0001b3e0: 7320 4e6f 6e65 3a0a 2020 2020 2020 2020  s None:.        
-0001b3f0: 2020 2020 7072 696e 7428 2822 5370 6563      print(("Spec
-0001b400: 6965 732e 5f70 6c6f 745f 6465 6d6f 6772  ies._plot_demogr
-0001b410: 6170 6869 635f 6368 616e 6765 7320 6973  aphic_changes is
-0001b420: 206e 6f74 2076 616c 6964 2022 0a20 2020   not valid ".   
-0001b430: 2020 2020 2020 2020 2020 2020 2022 666f               "fo
-0001b440: 7220 7370 6563 6965 7320 7769 7468 206e  r species with n
-0001b450: 6f20 5f53 7065 6369 6573 4368 616e 6765  o _SpeciesChange
-0001b460: 7220 6f62 6a65 6374 2e5c 6e22 2929 0a20  r object.\n")). 
-0001b470: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
-0001b480: 2020 2020 2020 2020 2073 656c 662e 5f63           self._c
-0001b490: 6861 6e67 6572 2e5f 706c 6f74 5f64 656d  hanger._plot_dem
-0001b4a0: 5f63 6861 6e67 6573 2873 656c 6629 0a0a  _changes(self)..
-0001b4b0: 0a20 2020 2064 6566 205f 706c 6f74 5f65  .    def _plot_e
-0001b4c0: 7861 6d70 6c65 5f72 6563 6f6d 6269 6e61  xample_recombina
-0001b4d0: 6e74 5f67 656e 6f6d 6528 7365 6c66 293a  nt_genome(self):
-0001b4e0: 0a20 2020 2020 2020 2070 7269 6e74 2827  .        print('
-0001b4f0: 544f 444f 3a20 4649 5820 4d45 2127 290a  TODO: FIX ME!').
-0001b500: 2020 2020 2020 2020 7265 7475 726e 0a20          return. 
-0001b510: 2020 2020 2020 2061 7373 6572 7420 7365         assert se
-0001b520: 6c66 2e67 656e 5f61 7263 6820 6973 206e  lf.gen_arch is n
-0001b530: 6f74 204e 6f6e 652c 2028 2254 6869 7320  ot None, ("This 
-0001b540: 7370 6563 6965 7320 646f 6573 206e 6f74  species does not
-0001b550: 2068 6176 6520 220a 2020 2020 2020 2020   have ".        
-0001b560: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b570: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b580: 2020 2020 2267 656e 6f6d 6573 2c20 736f      "genomes, so
-0001b590: 2069 7420 6361 6e6e 6f74 2062 6520 7573   it cannot be us
-0001b5a0: 6564 2022 0a20 2020 2020 2020 2020 2020  ed ".           
-0001b5b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b5c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b5d0: 2022 746f 2070 6c6f 7420 616e 2065 7861   "to plot an exa
-0001b5e0: 6d70 6c65 2072 6563 6f6d 6269 6e61 6e74  mple recombinant
-0001b5f0: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
-0001b600: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b610: 2020 2020 2020 2020 2020 2020 2020 2022                 "
-0001b620: 6765 6e6f 6d65 2e22 290a 0a20 2020 2020  genome.")..     
-0001b630: 2020 2072 6563 6f6d 625f 7061 7468 7320     recomb_paths 
-0001b640: 3d20 7365 6c66 2e67 656e 5f61 7263 682e  = self.gen_arch.
-0001b650: 5f72 6563 6f6d 625f 7061 7468 732e 5f67  _recomb_paths._g
-0001b660: 6574 5f70 6174 6873 2832 290a 2020 2020  et_paths(2).    
-0001b670: 2020 2020 6964 7873 203d 2028 302c 2032      idxs = (0, 2
-0001b680: 290a 2020 2020 2020 2020 6d6f 636b 5f73  ).        mock_s
-0001b690: 7070 203d 207b 7d0a 2020 2020 2020 2020  pp = {}.        
-0001b6a0: 666f 7220 6964 7820 696e 2069 6478 733a  for idx in idxs:
-0001b6b0: 0a20 2020 2020 2020 2020 2020 206e 6577  .            new
-0001b6c0: 5f67 656e 6f6d 6520 3d20 6e70 2e68 7374  _genome = np.hst
-0001b6d0: 6163 6b28 5b28 6e70 2e6f 6e65 7328 2873  ack([(np.ones((s
-0001b6e0: 656c 662e 6765 6e5f 6172 6368 2e4c 2c0a  elf.gen_arch.L,.
+0001ad30: 7661 6c75 6573 2c20 616e 6420 636f 6f72  values, and coor
+0001ad40: 6469 6e61 7465 730a 2020 2020 2020 2020  dinates.        
+0001ad50: 2020 2020 2020 2020 2269 6e64 5f43 4341          "ind_CCA
+0001ad60: 6466 223a 2069 6e64 6976 6964 7561 6c20  df": individual 
+0001ad70: 6c6f 6164 696e 6773 2028 636f 6c75 6d6e  loadings (column
+0001ad80: 7320 3d20 6178 6573 2c0a 2020 2020 2020  s = axes,.      
+0001ad90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ada0: 2020 2020 2020 2072 6f77 7320 3d20 696e         rows = in
+0001adb0: 6469 7669 6475 616c 7329 0a20 2020 2020  dividuals).     
+0001adc0: 2020 2020 2020 2020 2020 2022 6c6f 6369             "loci
+0001add0: 5f43 4341 6466 223a 206c 6f63 7573 206c  _CCAdf": locus l
+0001ade0: 6f61 6469 6e67 7320 2863 6f6c 756d 6e73  oadings (columns
+0001adf0: 203d 2061 7865 732c 2072 6f77 7320 3d20   = axes, rows = 
+0001ae00: 6c6f 6369 290a 2020 2020 2020 2020 2020  loci).          
+0001ae10: 2020 2020 2020 2276 6172 5f43 4341 6466        "var_CCAdf
+0001ae20: 223a 2076 6172 6961 626c 6520 6c6f 6164  ": variable load
+0001ae30: 696e 6773 2028 636f 6c75 6d6e 7320 3d20  ings (columns = 
+0001ae40: 6178 6573 2c0a 2020 2020 2020 2020 2020  axes,.          
+0001ae50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ae60: 2020 2072 6f77 7320 3d20 7661 7269 6162     rows = variab
+0001ae70: 6c65 2028 656e 762c 206c 6174 2c20 6c6f  le (env, lat, lo
+0001ae80: 6e67 2929 0a20 2020 2020 2020 2020 2020  ng)).           
+0001ae90: 2020 2020 2022 7472 6169 745f 6c6f 6369       "trait_loci
+0001aea0: 223a 206c 6973 7420 6f66 2069 6e64 6578  ": list of index
+0001aeb0: 6573 2066 6f72 206c 6f63 6920 7468 6174  es for loci that
+0001aec0: 2075 6e64 6572 6c69 6520 7468 6520 7472   underlie the tr
+0001aed0: 6169 740a 2020 2020 2020 2020 2020 2020  ait.            
+0001aee0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001aef0: 2020 666f 7220 7768 6963 6820 7468 6520    for which the 
+0001af00: 4343 4120 7761 7320 7275 6e0a 2020 2020  CCA was run.    
+0001af10: 2020 2020 2222 220a 2020 2020 2020 2020      """.        
+0001af20: 236d 616b 6520 4446 0a20 2020 2020 2020  #make DF.       
+0001af30: 2067 6561 5f64 6620 3d73 656c 662e 5f6d   gea_df =self._m
+0001af40: 616b 655f 6765 615f 6466 286c 7972 5f6e  ake_gea_df(lyr_n
+0001af50: 756d 3d73 656c 662e 6765 6e5f 6172 6368  um=self.gen_arch
+0001af60: 2e74 7261 6974 735b 7472 745f 6e75 6d5d  .traits[trt_num]
+0001af70: 2e6c 7972 5f6e 756d 290a 0a20 2020 2020  .lyr_num)..     
+0001af80: 2020 2023 6765 7420 6164 6170 7469 7665     #get adaptive
+0001af90: 206c 6f63 6920 2875 7365 6420 6c61 7465   loci (used late
+0001afa0: 7220 696e 2070 6c6f 7474 696e 6729 0a20  r in plotting). 
+0001afb0: 2020 2020 2020 2074 7261 6974 5f6c 6f63         trait_loc
+0001afc0: 6920 3d20 7365 6c66 2e67 656e 5f61 7263  i = self.gen_arc
+0001afd0: 682e 7472 6169 7473 5b74 7274 5f6e 756d  h.traits[trt_num
+0001afe0: 5d2e 6c6f 6369 2023 6765 7473 2074 7261  ].loci #gets tra
+0001aff0: 6974 206c 6f63 6920 696e 6465 7865 730a  it loci indexes.
+0001b000: 0a20 2020 2020 2020 2023 6765 7420 6e75  .        #get nu
+0001b010: 6d62 6572 206f 6620 6c6f 6369 0a20 2020  mber of loci.   
+0001b020: 2020 2020 204c 203d 2073 656c 662e 6765       L = self.ge
+0001b030: 6e5f 6172 6368 2e4c 0a20 2020 2020 2020  n_arch.L.       
+0001b040: 2023 6465 6669 6e65 2078 2061 6e64 2079   #define x and y
+0001b050: 3a0a 2020 2020 2020 2020 2320 7075 6c6c  :.        # pull
+0001b060: 206f 7574 2067 656e 6f74 7970 6573 0a20   out genotypes. 
+0001b070: 2020 2020 2020 2059 203d 2067 6561 5f64         Y = gea_d
+0001b080: 662e 696c 6f63 5b3a 2c72 616e 6765 284c  f.iloc[:,range(L
+0001b090: 295d 0a20 2020 2020 2020 2023 2070 756c  )].        # pul
+0001b0a0: 6c20 6f75 7420 7072 6564 6963 746f 7220  l out predictor 
+0001b0b0: 7661 7269 6162 6c65 730a 2020 2020 2020  variables.      
+0001b0c0: 2020 2320 544f 444f 3a20 6d61 6b65 2069    # TODO: make i
+0001b0d0: 7420 706f 7373 6962 6c65 2074 6f20 6368  t possible to ch
+0001b0e0: 6f6f 7365 206d 6f72 6520 7661 7273 2069  oose more vars i
+0001b0f0: 6e20 6675 7475 7265 0a20 2020 2020 2020  n future.       
+0001b100: 2058 203d 2067 6561 5f64 665b 5b27 656e   X = gea_df[['en
+0001b110: 7627 2c27 6c61 7427 2c27 6c6f 6e67 275d  v','lat','long']
+0001b120: 5d0a 0a20 2020 2020 2020 2023 6465 6669  ]..        #defi
+0001b130: 6e65 206e 5f63 6f6d 706f 6e65 6e74 730a  ne n_components.
+0001b140: 2020 2020 2020 2020 6e5f 636f 6d70 6f6e          n_compon
+0001b150: 656e 7473 203d 2033 0a20 2020 2020 2020  ents = 3.       
+0001b160: 2023 6372 6561 7465 2074 6865 2043 4341   #create the CCA
+0001b170: 206d 6f64 656c 2061 6e64 2066 6974 2069   model and fit i
+0001b180: 7420 746f 2074 6865 2064 6174 610a 2020  t to the data.  
+0001b190: 2020 2020 2020 6363 6120 3d20 4343 4128        cca = CCA(
+0001b1a0: 6e5f 636f 6d70 6f6e 656e 7473 203d 206e  n_components = n
+0001b1b0: 5f63 6f6d 706f 6e65 6e74 7329 0a20 2020  _components).   
+0001b1c0: 2020 2020 2063 6361 2e66 6974 2859 2c20       cca.fit(Y, 
+0001b1d0: 5829 0a0a 2020 2020 2020 2020 2374 7261  X)..        #tra
+0001b1e0: 6e73 666f 726d 2064 6174 610a 2020 2020  nsform data.    
+0001b1f0: 2020 2020 595f 6320 3d20 6363 612e 7472      Y_c = cca.tr
+0001b200: 616e 7366 6f72 6d28 5929 0a0a 2020 2020  ansform(Y)..    
+0001b210: 2020 2020 2363 7265 6174 6520 6466 2066      #create df f
+0001b220: 6f72 2069 6e64 6976 6964 7561 6c73 0a20  or individuals. 
+0001b230: 2020 2020 2020 2069 6e64 5f64 6620 3d20         ind_df = 
+0001b240: 7064 2e44 6174 6146 7261 6d65 2859 5f63  pd.DataFrame(Y_c
+0001b250: 290a 2020 2020 2020 2020 236e 616d 696e  ).        #namin
+0001b260: 6720 4343 4120 636f 6c75 6d6e 7320 7374  g CCA columns st
+0001b270: 6172 7469 6e67 2061 7420 310a 2020 2020  arting at 1.    
+0001b280: 2020 2020 696e 645f 6466 2e63 6f6c 756d      ind_df.colum
+0001b290: 6e73 203d 205b 7374 7228 7229 2066 6f72  ns = [str(r) for
+0001b2a0: 2072 2069 6e20 7261 6e67 6528 312c 206e   r in range(1, n
+0001b2b0: 5f63 6f6d 706f 6e65 6e74 7320 2b20 3129  _components + 1)
+0001b2c0: 5d0a 0a20 2020 2020 2020 2023 6372 6561  ]..        #crea
+0001b2d0: 7465 2064 6620 666f 7220 6c6f 6369 0a20  te df for loci. 
+0001b2e0: 2020 2020 2020 206c 6f63 695f 6466 203d         loci_df =
+0001b2f0: 2070 642e 4461 7461 4672 616d 6528 6363   pd.DataFrame(cc
+0001b300: 612e 785f 6c6f 6164 696e 6773 5f29 0a20  a.x_loadings_). 
+0001b310: 2020 2020 2020 206c 6f63 695f 6466 2e63         loci_df.c
+0001b320: 6f6c 756d 6e73 203d 2069 6e64 5f64 662e  olumns = ind_df.
+0001b330: 636f 6c75 6d6e 7320 2373 616d 6520 636f  columns #same co
+0001b340: 6c75 6d6e 206e 616d 6573 0a0a 2020 2020  lumn names..    
+0001b350: 2020 2020 2363 7265 6174 6520 6466 2066      #create df f
+0001b360: 6f72 2076 6172 6961 626c 6573 0a20 2020  or variables.   
+0001b370: 2020 2020 2076 6172 5f64 6620 3d20 7064       var_df = pd
+0001b380: 2e44 6174 6146 7261 6d65 2863 6361 2e79  .DataFrame(cca.y
+0001b390: 5f6c 6f61 6469 6e67 735f 290a 2020 2020  _loadings_).    
+0001b3a0: 2020 2020 7661 725f 6466 2e63 6f6c 756d      var_df.colum
+0001b3b0: 6e73 203d 2069 6e64 5f64 662e 636f 6c75  ns = ind_df.colu
+0001b3c0: 6d6e 7320 2373 616d 6520 636f 6c75 6d6e  mns #same column
+0001b3d0: 206e 616d 6573 0a0a 2020 2020 2020 2020   names..        
+0001b3e0: 236d 616b 6520 6469 6374 696f 6e61 7279  #make dictionary
+0001b3f0: 206f 6620 6461 7461 6672 616d 6573 2074   of dataframes t
+0001b400: 6f20 7265 7475 726e 0a20 2020 2020 2020  o return.       
+0001b410: 2063 6361 5f64 6963 7420 3d20 7b27 6765   cca_dict = {'ge
+0001b420: 615f 6466 273a 6765 615f 6466 2c0a 2020  a_df':gea_df,.  
+0001b430: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001b440: 2020 2769 6e64 5f64 6627 3a69 6e64 5f64    'ind_df':ind_d
+0001b450: 662c 0a20 2020 2020 2020 2020 2020 2020  f,.             
+0001b460: 2020 2020 2020 2027 6c6f 6369 5f64 6627         'loci_df'
+0001b470: 3a6c 6f63 695f 6466 2c0a 2020 2020 2020  :loci_df,.      
+0001b480: 2020 2020 2020 2020 2020 2020 2020 2776                'v
+0001b490: 6172 5f64 6627 3a76 6172 5f64 662c 0a20  ar_df':var_df,. 
+0001b4a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001b4b0: 2020 2027 7472 6169 745f 6c6f 6369 273a     'trait_loci':
+0001b4c0: 7472 6169 745f 6c6f 6369 7d0a 0a20 2020  trait_loci}..   
+0001b4d0: 2020 2020 2023 204e 4f54 453a 2072 6967       # NOTE: rig
+0001b4e0: 6874 206e 6f77 2074 6869 7320 706c 6f74  ht now this plot
+0001b4f0: 7469 6e67 206c 6f6f 7020 6f6e 6c79 2077  ting loop only w
+0001b500: 6f72 6b73 2066 6f72 2061 206d 6178 696d  orks for a maxim
+0001b510: 756d 206f 6620 3320 6178 6573 3b0a 2020  um of 3 axes;.  
+0001b520: 2020 2020 2020 2320 2020 2020 2020 696e        #       in
+0001b530: 2074 6865 2066 7574 7572 6520 6173 206d   the future as m
+0001b540: 6f72 6520 656e 7620 7661 7273 2061 7265  ore env vars are
+0001b550: 2061 6464 6564 2c20 636f 756c 6420 6164   added, could ad
+0001b560: 6420 6d6f 7265 2061 7865 730a 2020 2020  d more axes.    
+0001b570: 2020 2020 6966 2070 6c6f 743a 0a20 2020      if plot:.   
+0001b580: 2020 2020 2020 2020 2023 5363 616c 6520           #Scale 
+0001b590: 6461 7461 6672 616d 6573 2066 726f 6d20  dataframes from 
+0001b5a0: 2d31 2074 6f20 3120 6966 2073 6361 6c65  -1 to 1 if scale
+0001b5b0: 203d 2054 7275 650a 2020 2020 2020 2020   = True.        
+0001b5c0: 2020 2020 6966 2073 6361 6c65 203d 3d20      if scale == 
+0001b5d0: 5472 7565 3a0a 2020 2020 2020 2020 2020  True:.          
+0001b5e0: 2020 2020 2020 726d 696e 203d 202d 310a        rmin = -1.
+0001b5f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001b600: 726d 6178 203d 2031 0a20 2020 2020 2020  rmax = 1.       
+0001b610: 2020 2020 2020 2020 206c 6f63 695f 6466           loci_df
+0001b620: 203d 2072 6d69 6e20 2b20 286c 6f63 695f   = rmin + (loci_
+0001b630: 6466 202d 206c 6f63 695f 6466 2e76 616c  df - loci_df.val
+0001b640: 7565 732e 6d69 6e28 2929 202a 2028 726d  ues.min()) * (rm
+0001b650: 6178 202d 0a20 2020 2020 2020 2020 2020  ax -.           
+0001b660: 2020 2020 2020 2020 2020 2020 2028 726d               (rm
+0001b670: 696e 2929 202f 2028 6c6f 6369 5f64 662e  in)) / (loci_df.
+0001b680: 7661 6c75 6573 2e6d 6178 2829 202d 206c  values.max() - l
+0001b690: 6f63 695f 6466 2e76 616c 7565 732e 6d69  oci_df.values.mi
+0001b6a0: 6e28 2929 0a20 2020 2020 2020 2020 2020  n()).           
+0001b6b0: 2020 2020 2076 6172 5f64 6620 3d20 726d       var_df = rm
+0001b6c0: 696e 202b 2028 7661 725f 6466 202d 2076  in + (var_df - v
+0001b6d0: 6172 5f64 662e 7661 6c75 6573 2e6d 696e  ar_df.values.min
+0001b6e0: 2829 2920 2a20 2872 6d61 7820 2d0a 2020  ()) * (rmax -.  
 0001b6f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b700: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b710: 2020 2020 2020 2020 2020 2020 2020 3129                1)
-0001b720: 2920 2a20 6e20 2920 2b20 6964 7820 666f  ) * n ) + idx fo
-0001b730: 7220 6e20 696e 2028 312c 3229 5d29 0a20  r n in (1,2)]). 
-0001b740: 2020 2020 2020 2020 2020 206d 6f63 6b5f             mock_
-0001b750: 7370 705b 6964 785d 203d 2049 6e64 6976  spp[idx] = Indiv
-0001b760: 6964 7561 6c28 6964 783d 6964 782c 2078  idual(idx=idx, x
-0001b770: 3d30 2c20 793d 302c 0a20 2020 2020 2020  =0, y=0,.       
-0001b780: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b790: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b7a0: 6e65 775f 6765 6e6f 6d65 3d6e 6577 5f67  new_genome=new_g
-0001b7b0: 656e 6f6d 6529 0a0a 2020 2020 2020 2020  enome)..        
-0001b7c0: 7265 636f 6d62 5f67 656e 6f6d 6520 3d20  recomb_genome = 
-0001b7d0: 5f64 6f5f 6d61 7469 6e67 2873 7070 3d6d  _do_mating(spp=m
-0001b7e0: 6f63 6b5f 7370 702c 0a20 2020 2020 2020  ock_spp,.       
-0001b7f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b800: 2020 2020 2020 2020 2020 2020 6d61 7469              mati
-0001b810: 6e67 5f70 6169 7273 3d5b 6964 7873 5d2c  ng_pairs=[idxs],
-0001b820: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0001b830: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b840: 2020 2020 6e5f 6f66 6673 7072 696e 673d      n_offspring=
-0001b850: 5b31 5d2c 0a20 2020 2020 2020 2020 2020  [1],.           
-0001b860: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b870: 2020 2020 2020 2020 7265 636f 6d62 5f70          recomb_p
-0001b880: 6174 6873 3d72 6563 6f6d 625f 7061 7468  aths=recomb_path
-0001b890: 7329 5b30 5d5b 305d 0a20 2020 2020 2020  s)[0][0].       
-0001b8a0: 2023 7265 636f 6d62 5f67 656e 6f6d 6520   #recomb_genome 
-0001b8b0: 3d20 7265 636f 6d62 5f67 656e 6f6d 652b  = recomb_genome+
-0001b8c0: 310a 2020 2020 2020 2020 2372 6563 6f6d  1.        #recom
-0001b8d0: 625f 6765 6e6f 6d65 5b3a 2c31 5d20 3d20  b_genome[:,1] = 
-0001b8e0: 7265 636f 6d62 5f67 656e 6f6d 655b 3a2c  recomb_genome[:,
-0001b8f0: 315d 202a 2035 0a20 2020 2020 2020 2066  1] * 5.        f
-0001b900: 6967 203d 2070 6c74 2e66 6967 7572 6528  ig = plt.figure(
-0001b910: 290a 2020 2020 2020 2020 6178 203d 2066  ).        ax = f
-0001b920: 6967 2e61 6464 5f73 7562 706c 6f74 2831  ig.add_subplot(1
-0001b930: 3131 290a 2020 2020 2020 2020 6178 2e69  11).        ax.i
-0001b940: 6d73 686f 7728 6e70 2e72 6570 6561 7428  mshow(np.repeat(
-0001b950: 7265 636f 6d62 5f67 656e 6f6d 652c 0a20  recomb_genome,. 
-0001b960: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b970: 2020 2020 2020 2020 2020 2069 6e74 2830             int(0
-0001b980: 2e30 3520 2a20 7365 6c66 2e67 656e 5f61  .05 * self.gen_a
-0001b990: 7263 682e 4c29 2c0a 2020 2020 2020 2020  rch.L),.        
-0001b9a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001b9b0: 2020 2020 6178 6973 3d31 292c 0a20 2020      axis=1),.   
-0001b9c0: 2020 2020 2020 2020 2020 2020 2020 2063                 c
-0001b9d0: 6d61 703d 2774 6572 7261 696e 2729 0a20  map='terrain'). 
-0001b9e0: 2020 2020 2020 2061 782e 7365 745f 7469         ax.set_ti
-0001b9f0: 746c 6528 2267 616d 6574 6520 3020 2020  tle("gamete 0   
-0001ba00: 2020 6761 6d65 7465 3122 290a 2020 2020    gamete1").    
-0001ba10: 2020 2020 6178 2e73 6574 5f78 7469 636b      ax.set_xtick
-0001ba20: 7328 5b5d 290a 2020 2020 2020 2020 6178  s([]).        ax
-0001ba30: 2e73 6574 5f79 6c61 6265 6c28 276c 6f63  .set_ylabel('loc
-0001ba40: 7573 2729 0a20 2020 2020 2020 2070 6c74  us').        plt
-0001ba50: 2e73 686f 7728 290a 2020 2020 2020 2020  .show().        
-0001ba60: 7265 7475 726e 2872 6563 6f6d 625f 7061  return(recomb_pa
-0001ba70: 7468 732c 206d 6f63 6b5f 7370 702c 2072  ths, mock_spp, r
-0001ba80: 6563 6f6d 625f 6765 6e6f 6d65 290a 0a0a  ecomb_genome)...
-0001ba90: 2020 2020 6465 6620 5f70 6c6f 745f 7374      def _plot_st
-0001baa0: 6174 2873 656c 662c 2073 7461 7429 3a0a  at(self, stat):.
-0001bab0: 2020 2020 2020 2020 6966 2073 656c 662e          if self.
-0001bac0: 5f73 7461 7473 5f63 6f6c 6c65 6374 6f72  _stats_collector
-0001bad0: 2069 7320 4e6f 6e65 3a0a 2020 2020 2020   is None:.      
-0001bae0: 2020 2020 2020 7072 696e 7428 2822 5370        print(("Sp
-0001baf0: 6563 6965 732e 5f70 6c6f 745f 7374 6174  ecies._plot_stat
-0001bb00: 2069 7320 6e6f 7420 7661 6c69 6420 220a   is not valid ".
-0001bb10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001bb20: 2266 6f72 2073 7065 6369 6573 2077 6974  "for species wit
-0001bb30: 6820 6e6f 205f 5374 6174 7343 6f6c 6c65  h no _StatsColle
-0001bb40: 6374 6f72 206f 626a 6563 742e 5c6e 2229  ctor object.\n")
-0001bb50: 290a 0a20 2020 2020 2020 2065 6c73 653a  )..        else:
-0001bb60: 0a20 2020 2020 2020 2020 2020 2073 656c  .            sel
-0001bb70: 662e 5f73 7461 7473 5f63 6f6c 6c65 6374  f._stats_collect
-0001bb80: 6f72 2e5f 706c 6f74 5f73 7461 7428 7374  or._plot_stat(st
-0001bb90: 6174 2c20 7370 705f 6e61 6d65 203d 2073  at, spp_name = s
-0001bba0: 656c 662e 6e61 6d65 290a 0a0a 2020 2020  elf.name)...    
-0001bbb0: 2020 2020 2323 2323 2323 2323 2323 2323      ############
-0001bbc0: 2323 2323 0a20 2020 2020 2020 2023 7075  ####.        #pu
-0001bbd0: 626c 6963 206d 6574 686f 6473 230a 2020  blic methods#.  
-0001bbe0: 2020 2020 2020 2323 2323 2323 2323 2323        ##########
-0001bbf0: 2323 2323 2323 0a0a 2020 2020 6465 6620  ######..    def 
-0001bc00: 5f77 7269 7465 5f70 6963 6b6c 6528 7365  _write_pickle(se
-0001bc10: 6c66 2c20 6669 6c65 6e61 6d65 293a 0a20  lf, filename):. 
-0001bc20: 2020 2020 2020 2069 6d70 6f72 7420 6350         import cP
-0001bc30: 6963 6b6c 650a 2020 2020 2020 2020 7769  ickle.        wi
-0001bc40: 7468 206f 7065 6e28 6669 6c65 6e61 6d65  th open(filename
-0001bc50: 2c20 2777 6227 2920 6173 2066 3a0a 2020  , 'wb') as f:.  
-0001bc60: 2020 2020 2020 2020 2020 6350 6963 6b6c            cPickl
-0001bc70: 652e 6475 6d70 2873 656c 662c 2066 290a  e.dump(self, f).
-0001bc80: 0a0a 2323 2323 2323 2323 2323 2323 2323  ..##############
-0001bc90: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0001bca0: 2323 2323 2323 2323 0a23 202d 2d2d 2d2d  ########.# -----
-0001bcb0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-0001bcc0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 230a  --------------#.
-0001bcd0: 2320 4655 4e43 5449 4f4e 5320 2d2d 2d2d  # FUNCTIONS ----
-0001bce0: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-0001bcf0: 2d2d 2d2d 2d23 0a23 202d 2d2d 2d2d 2d2d  -----#.# -------
-0001bd00: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-0001bd10: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 230a 2323  ------------#.##
-0001bd20: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0001bd30: 2323 2323 2323 2323 2323 2323 2323 2323  ################
-0001bd40: 2323 2323 0a0a 2366 756e 6374 696f 6e20  ####..#function 
-0001bd50: 746f 2062 6520 6361 6c6c 6564 2077 6865  to be called whe
-0001bd60: 6e20 6120 5370 6563 6965 7320 6973 2069  n a Species is i
-0001bd70: 6e69 7469 6174 6564 2c0a 2377 6869 6368  nitiated,.#which
-0001bd80: 2075 7365 7320 7468 6520 7061 7261 6d73   uses the params
-0001bd90: 2e73 7065 6369 6573 5b3c 7370 705f 6e75  .species[<spp_nu
-0001bda0: 6d3e 5d2e 696e 6974 2e5b 274b 5f3c 3e27  m>].init.['K_<>'
-0001bdb0: 5d20 7061 7261 6d65 7465 7273 200a 2374  ] parameters .#t
-0001bdc0: 6f20 6d61 6b65 2074 6865 2069 6e69 7469  o make the initi
-0001bdd0: 616c 2063 6172 7279 696e 672d 6361 7061  al carrying-capa
-0001bde0: 6369 7479 2072 6173 7465 722c 2061 6e64  city raster, and
-0001bdf0: 2061 6c73 6f0a 2373 6574 7320 7468 6520   also.#sets the 
-0001be00: 5370 6563 6965 7327 204b 5f6c 6179 6572  Species' K_layer
-0001be10: 2061 6e64 204b 5f66 6163 746f 7220 6174   and K_factor at
-0001be20: 7472 6962 7574 6573 0a64 6566 205f 6d61  tributes.def _ma
-0001be30: 6b65 5f4b 2873 7070 2c20 6c61 6e64 2c20  ke_K(spp, land, 
-0001be40: 4b5f 6c61 7965 722c 204b 5f66 6163 746f  K_layer, K_facto
-0001be50: 7229 3a0a 2020 2020 236d 616b 6520 7375  r):.    #make su
-0001be60: 7265 2077 6520 6669 6e64 206f 6e6c 7920  re we find only 
-0001be70: 6120 7369 6e67 6c65 206c 6179 6572 2077  a single layer w
-0001be80: 6974 6820 7468 6520 6e61 6d65 2073 7065  ith the name spe
-0001be90: 6369 6669 6564 2062 7920 4b5f 6c61 7965  cified by K_laye
-0001bea0: 720a 2020 2020 4b5f 6c61 7965 7220 3d20  r.    K_layer = 
-0001beb0: 5b6c 7972 2066 6f72 206c 7972 2069 6e20  [lyr for lyr in 
-0001bec0: 6c61 6e64 2e76 616c 7565 7328 2920 6966  land.values() if
-0001bed0: 206c 7972 2e6e 616d 6520 3d3d 204b 5f6c   lyr.name == K_l
-0001bee0: 6179 6572 5d0a 2020 2020 6173 7365 7274  ayer].    assert
-0001bef0: 206c 656e 284b 5f6c 6179 6572 2920 3d3d   len(K_layer) ==
-0001bf00: 2031 2c20 2822 5468 6520 4b5f 6c61 7965   1, ("The K_laye
-0001bf10: 7220 7061 7261 6d65 7465 7220 7368 6f75  r parameter shou
-0001bf20: 6c64 2070 6f69 6e74 2074 6f22 0a20 2020  ld point to".   
-0001bf30: 2020 2020 2022 6120 7369 6e67 6c65 204c       "a single L
-0001bf40: 6179 6572 2c20 6275 7420 696e 7374 6561  ayer, but instea
-0001bf50: 6420 2569 204c 6179 6572 7320 7765 7265  d %i Layers were
-0001bf60: 2066 6f75 6e64 2e22 2920 2520 6c65 6e28   found.") % len(
-0001bf70: 4b5f 6c61 7965 7229 0a20 2020 2023 6772  K_layer).    #gr
-0001bf80: 6162 2074 6865 2069 6465 6e74 6966 6965  ab the identifie
-0001bf90: 6420 6c61 7965 720a 2020 2020 4b5f 6c61  d layer.    K_la
-0001bfa0: 7965 7220 3d20 4b5f 6c61 7965 725b 305d  yer = K_layer[0]
-0001bfb0: 0a20 2020 2023 7365 7420 7468 6520 5370  .    #set the Sp
-0001bfc0: 6563 6965 7327 204b 5f6c 6179 6572 2061  ecies' K_layer a
-0001bfd0: 6e64 204b 5f66 6163 746f 7220 6174 7472  nd K_factor attr
-0001bfe0: 6962 7574 6573 0a20 2020 2073 7070 2e4b  ibutes.    spp.K
-0001bff0: 5f6c 6179 6572 203d 204b 5f6c 6179 6572  _layer = K_layer
-0001c000: 2e69 6478 0a20 2020 2073 7070 2e4b 5f66  .idx.    spp.K_f
-0001c010: 6163 746f 7220 3d20 4b5f 6661 6374 6f72  actor = K_factor
-0001c020: 0a20 2020 2023 6164 6420 7468 6973 2053  .    #add this S
-0001c030: 7065 6369 6573 2074 6f20 7468 6973 204c  pecies to this L
-0001c040: 6179 6572 2773 202e 5f69 735f 4b20 6174  ayer's ._is_K at
-0001c050: 7472 6962 7574 650a 2020 2020 2328 7768  tribute.    #(wh
-0001c060: 6963 6820 7769 6c6c 2062 6520 7573 6564  ich will be used
-0001c070: 2074 6f20 7570 6461 7465 2053 7065 6369   to update Speci
-0001c080: 6573 2e4b 2069 6620 7468 6973 204c 6179  es.K if this Lay
-0001c090: 6572 2075 6e64 6572 676f 6573 0a20 2020  er undergoes.   
-0001c0a0: 2023 616e 7920 6b61 6e64 7363 6170 6520   #any kandscape 
-0001c0b0: 6368 616e 6765 7329 0a20 2020 204b 5f6c  changes).    K_l
-0001c0c0: 6179 6572 2e5f 6973 5f4b 2e61 7070 656e  ayer._is_K.appen
-0001c0d0: 6428 7370 702e 6964 7829 0a20 2020 2023  d(spp.idx).    #
-0001c0e0: 6e6f 7720 6361 6c63 756c 6174 6520 616e  now calculate an
-0001c0f0: 6420 7365 7420 7468 6520 4b20 7261 7374  d set the K rast
-0001c100: 6572 0a20 2020 2073 7070 2e5f 7365 745f  er.    spp._set_
-0001c110: 4b28 6c61 6e64 290a 0a0a 6465 6620 5f6d  K(land)...def _m
-0001c120: 616b 655f 7370 6563 6965 7328 6c61 6e64  ake_species(land
-0001c130: 2c20 6e61 6d65 2c20 6964 782c 2073 7070  , name, idx, spp
-0001c140: 5f70 6172 616d 732c 2062 7572 6e3d 4661  _params, burn=Fa
-0001c150: 6c73 652c 2076 6572 626f 7365 3d46 616c  lse, verbose=Fal
-0001c160: 7365 293a 0a20 2020 2023 6765 7420 7370  se):.    #get sp
-0001c170: 7027 7320 696e 7469 616c 697a 696e 6720  p's intializing 
-0001c180: 7061 7261 6d73 0a20 2020 2069 6e69 745f  params.    init_
-0001c190: 7061 7261 6d73 203d 2064 6565 7063 6f70  params = deepcop
-0001c1a0: 7928 7370 705f 7061 7261 6d73 2e69 6e69  y(spp_params.ini
-0001c1b0: 7429 0a0a 2020 2020 2320 7072 696e 7420  t)..    # print 
-0001c1c0: 7665 7262 6f73 6520 6f75 7470 7574 0a20  verbose output. 
-0001c1d0: 2020 2069 6620 7665 7262 6f73 653a 0a20     if verbose:. 
-0001c1e0: 2020 2020 2020 2070 7269 6e74 2827 5c74         print('\t
-0001c1f0: 5c74 4d41 4b49 4e47 2053 5045 4349 4553  \tMAKING SPECIES
-0001c200: 2025 732e 2e2e 5c6e 2720 2520 6e61 6d65   %s...\n' % name
-0001c210: 2c20 666c 7573 683d 5472 7565 290a 0a20  , flush=True).. 
-0001c220: 2020 2023 6966 2074 6869 7320 7370 6563     #if this spec
-0001c230: 6965 7320 7368 6f75 6c64 2068 6176 6520  ies should have 
-0001c240: 6765 6e6f 6d65 732c 2063 7265 6174 6520  genomes, create 
-0001c250: 7468 6520 6765 6e6f 6d69 6320 6172 6368  the genomic arch
-0001c260: 6974 6563 7475 7265 0a20 2020 2069 6620  itecture.    if 
-0001c270: 2767 656e 5f61 7263 6827 2069 6e20 7370  'gen_arch' in sp
-0001c280: 705f 7061 7261 6d73 2e6b 6579 7328 293a  p_params.keys():
-0001c290: 0a20 2020 2020 2020 2023 2070 7269 6e74  .        # print
-0001c2a0: 2076 6572 626f 7365 206f 7574 7075 740a   verbose output.
-0001c2b0: 2020 2020 2020 2020 6966 2076 6572 626f          if verbo
-0001c2c0: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-0001c2d0: 7072 696e 7428 275c 745c 745c 746d 616b  print('\t\t\tmak
-0001c2e0: 696e 6720 6765 6e6f 6d69 6320 6172 6368  ing genomic arch
-0001c2f0: 6974 6563 7475 7265 2e2e 2e5c 6e27 2c20  itecture...\n', 
-0001c300: 666c 7573 683d 5472 7565 290a 2020 2020  flush=True).    
-0001c310: 2020 2020 675f 7061 7261 6d73 203d 2073      g_params = s
-0001c320: 7070 5f70 6172 616d 732e 6765 6e5f 6172  pp_params.gen_ar
-0001c330: 6368 0a20 2020 2020 2020 2023 6d61 6b65  ch.        #make
-0001c340: 2067 656e 6f6d 6963 5f61 7263 6869 7465   genomic_archite
-0001c350: 6374 7572 650a 2020 2020 2020 2020 6765  cture.        ge
-0001c360: 6e5f 6172 6368 203d 205f 6d61 6b65 5f67  n_arch = _make_g
-0001c370: 656e 6f6d 6963 5f61 7263 6869 7465 6374  enomic_architect
-0001c380: 7572 6528 7370 705f 7061 7261 6d73 203d  ure(spp_params =
-0001c390: 2073 7070 5f70 6172 616d 732c 0a20 2020   spp_params,.   
-0001c3a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001c3b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001b700: 2020 2020 2020 2872 6d69 6e29 2920 2f20        (rmin)) / 
+0001b710: 2876 6172 5f64 662e 7661 6c75 6573 2e6d  (var_df.values.m
+0001b720: 6178 2829 202d 2076 6172 5f64 662e 7661  ax() - var_df.va
+0001b730: 6c75 6573 2e6d 696e 2829 290a 2020 2020  lues.min()).    
+0001b740: 2020 2020 2020 2020 2020 2020 696e 645f              ind_
+0001b750: 6466 203d 2072 6d69 6e20 2b20 2869 6e64  df = rmin + (ind
+0001b760: 5f64 6620 2d20 696e 645f 6466 2e76 616c  _df - ind_df.val
+0001b770: 7565 732e 6d69 6e28 2929 202a 2028 726d  ues.min()) * (rm
+0001b780: 6178 202d 0a20 2020 2020 2020 2020 2020  ax -.           
+0001b790: 2020 2020 2020 2020 2020 2020 2028 726d               (rm
+0001b7a0: 696e 2929 202f 2028 696e 645f 6466 2e76  in)) / (ind_df.v
+0001b7b0: 616c 7565 732e 6d61 7828 2920 2d20 696e  alues.max() - in
+0001b7c0: 645f 6466 2e76 616c 7565 732e 6d69 6e28  d_df.values.min(
+0001b7d0: 2929 0a0a 0a20 2020 2020 2020 2020 2020  ))...           
+0001b7e0: 2023 6765 7420 6d61 7820 616e 6420 6d69   #get max and mi
+0001b7f0: 6e73 2074 6f20 7365 7420 6178 6973 206c  ns to set axis l
+0001b800: 6174 6572 206f 6e0a 2020 2020 2020 2020  ater on.        
+0001b810: 2020 2020 6d61 7864 6620 3d20 6d61 7828      maxdf = max(
+0001b820: 696e 645f 6466 2e76 616c 7565 732e 6d61  ind_df.values.ma
+0001b830: 7828 292c 2076 6172 5f64 662e 7661 6c75  x(), var_df.valu
+0001b840: 6573 2e6d 6178 2829 2c0a 2020 2020 2020  es.max(),.      
+0001b850: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001b860: 2020 6c6f 6369 5f64 662e 7661 6c75 6573    loci_df.values
+0001b870: 2e6d 6178 2829 290a 2020 2020 2020 2020  .max()).        
+0001b880: 2020 2020 6d69 6e64 6620 3d20 6162 7328      mindf = abs(
+0001b890: 6d69 6e28 696e 645f 6466 2e76 616c 7565  min(ind_df.value
+0001b8a0: 732e 6d69 6e28 292c 2076 6172 5f64 662e  s.min(), var_df.
+0001b8b0: 7661 6c75 6573 2e6d 696e 2829 2c0a 2020  values.min(),.  
+0001b8c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001b8d0: 2020 2020 2020 2020 2020 6c6f 6369 5f64            loci_d
+0001b8e0: 662e 7661 6c75 6573 2e6d 696e 2829 2929  f.values.min()))
+0001b8f0: 0a20 2020 2020 2020 2020 2020 2061 786d  .            axm
+0001b900: 6178 203d 206d 6178 286d 6178 6466 2c20  ax = max(maxdf, 
+0001b910: 6d69 6e64 6629 202a 2031 2e32 3020 2361  mindf) * 1.20 #a
+0001b920: 6464 696e 6720 6120 3230 2520 6275 6666  dding a 20% buff
+0001b930: 6572 0a0a 2020 2020 2020 2020 2020 2020  er..            
+0001b940: 2373 6574 2075 7020 6669 6775 7265 0a20  #set up figure. 
+0001b950: 2020 2020 2020 2020 2020 2066 6967 203d             fig =
+0001b960: 2070 6c74 2e66 6967 7572 6528 6669 6773   plt.figure(figs
+0001b970: 697a 653d 2832 302c 2031 3529 290a 2020  ize=(20, 15)).  
+0001b980: 2020 2020 2020 2020 2020 666f 7220 6e2c            for n,
+0001b990: 2063 635f 6178 6573 5f70 6169 7220 696e   cc_axes_pair in
+0001b9a0: 2065 6e75 6d65 7261 7465 285b 5b31 2c20   enumerate([[1, 
+0001b9b0: 325d 2c20 5b31 2c20 335d 2c20 5b32 2c20  2], [1, 3], [2, 
+0001b9c0: 335d 5d29 3a0a 2020 2020 2020 2020 2020  3]]):.          
+0001b9d0: 2020 2020 2020 2364 6566 696e 6520 636f        #define co
+0001b9e0: 6d70 6f6e 656e 7473 2066 6f72 2061 7869  mponents for axi
+0001b9f0: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+0001ba00: 2020 6363 5f61 7869 7331 203d 2063 635f    cc_axis1 = cc_
+0001ba10: 6178 6573 5f70 6169 725b 305d 2023 782d  axes_pair[0] #x-
+0001ba20: 6178 6973 2043 430a 2020 2020 2020 2020  axis CC.        
+0001ba30: 2020 2020 2020 2020 6363 5f61 7869 7332          cc_axis2
+0001ba40: 203d 2063 635f 6178 6573 5f70 6169 725b   = cc_axes_pair[
+0001ba50: 315d 2023 792d 6178 6973 2043 430a 0a20  1] #y-axis CC.. 
+0001ba60: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+0001ba70: 7820 3d20 6669 672e 6164 645f 7375 6270  x = fig.add_subp
+0001ba80: 6c6f 7428 3133 312b 6e29 0a0a 2020 2020  lot(131+n)..    
+0001ba90: 2020 2020 2020 2020 2020 2020 2361 6464              #add
+0001baa0: 2063 656e 7465 7220 6c69 6e65 730a 2020   center lines.  
+0001bab0: 2020 2020 2020 2020 2020 2020 2020 6178                ax
+0001bac0: 2e61 7868 6c69 6e65 2879 3d30 2c20 636f  .axhline(y=0, co
+0001bad0: 6c6f 723d 276c 6967 6874 6772 6179 272c  lor='lightgray',
+0001bae0: 206c 696e 6573 7479 6c65 3d27 646f 7474   linestyle='dott
+0001baf0: 6564 2729 0a20 2020 2020 2020 2020 2020  ed').           
+0001bb00: 2020 2020 2061 782e 6178 766c 696e 6528       ax.axvline(
+0001bb10: 783d 302c 2063 6f6c 6f72 3d27 6c69 6768  x=0, color='ligh
+0001bb20: 7467 7261 7927 2c20 6c69 6e65 7374 796c  tgray', linestyl
+0001bb30: 653d 2764 6f74 7465 6427 290a 0a20 2020  e='dotted')..   
+0001bb40: 2020 2020 2020 2020 2020 2020 2023 2073               # s
+0001bb50: 6574 2061 7865 7320 7261 6e67 650a 2020  et axes range.  
+0001bb60: 2020 2020 2020 2020 2020 2020 2020 706c                pl
+0001bb70: 742e 786c 696d 282d 6178 6d61 782c 2061  t.xlim(-axmax, a
+0001bb80: 786d 6178 290a 2020 2020 2020 2020 2020  xmax).          
+0001bb90: 2020 2020 2020 706c 742e 796c 696d 282d        plt.ylim(-
+0001bba0: 6178 6d61 782c 2061 786d 6178 290a 0a20  axmax, axmax).. 
+0001bbb0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+0001bbc0: 706c 6f74 206e 6575 7472 616c 2053 4e50  plot neutral SNP
+0001bbd0: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+0001bbe0: 2020 6178 2e73 6361 7474 6572 286c 6f63    ax.scatter(loc
+0001bbf0: 695f 6466 5b73 7472 2863 635f 6178 6973  i_df[str(cc_axis
+0001bc00: 3129 5d2c 206c 6f63 695f 6466 5b73 7472  1)], loci_df[str
+0001bc10: 2863 635f 6178 6973 3229 5d2c 0a20 2020  (cc_axis2)],.   
+0001bc20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bc30: 2020 2020 2020 2020 6d61 726b 6572 203d          marker =
+0001bc40: 2027 2b27 2c20 636f 6c6f 7220 3d20 2767   '+', color = 'g
+0001bc50: 7261 7927 2c0a 2020 2020 2020 2020 2020  ray',.          
+0001bc60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bc70: 206c 6162 656c 203d 2027 4e65 7574 7261   label = 'Neutra
+0001bc80: 6c20 534e 5073 272c 2061 6c70 6861 203d  l SNPs', alpha =
+0001bc90: 2030 2e37 290a 2020 2020 2020 2020 2020   0.7).          
+0001bca0: 2020 2020 2020 2370 6c6f 7420 696e 6469        #plot indi
+0001bcb0: 7669 6475 616c 730a 2020 2020 2020 2020  viduals.        
+0001bcc0: 2020 2020 2020 2020 636d 6170 5f73 6361          cmap_sca
+0001bcd0: 7420 3d20 6178 2e73 6361 7474 6572 2869  t = ax.scatter(i
+0001bce0: 6e64 5f64 665b 7374 7228 6363 5f61 7869  nd_df[str(cc_axi
+0001bcf0: 7331 295d 2c0a 2020 2020 2020 2020 2020  s1)],.          
+0001bd00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bd10: 2020 2020 2020 2020 2020 2020 2069 6e64               ind
+0001bd20: 5f64 665b 7374 7228 6363 5f61 7869 7332  _df[str(cc_axis2
+0001bd30: 295d 2c0a 2020 2020 2020 2020 2020 2020  )],.            
+0001bd40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bd50: 2020 2020 2020 2020 2020 2063 203d 2067             c = g
+0001bd60: 6561 5f64 662e 656e 762c 2063 6d61 703d  ea_df.env, cmap=
+0001bd70: 2776 6972 6964 6973 272c 0a20 2020 2020  'viridis',.     
+0001bd80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bd90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bda0: 2020 6c61 6265 6c20 3d20 2249 6e64 6976    label = "Indiv
+0001bdb0: 6964 7561 6c73 222c 2061 6c70 6861 203d  iduals", alpha =
+0001bdc0: 2030 2e35 290a 2020 2020 2020 2020 2020   0.5).          
+0001bdd0: 2020 2020 2020 2370 6c6f 7420 6164 6170        #plot adap
+0001bde0: 7469 7665 2053 4e50 730a 2020 2020 2020  tive SNPs.      
+0001bdf0: 2020 2020 2020 2020 2020 6178 2e73 6361            ax.sca
+0001be00: 7474 6572 286c 6f63 695f 6466 5b73 7472  tter(loci_df[str
+0001be10: 2863 635f 6178 6973 3129 5d5b 7472 6169  (cc_axis1)][trai
+0001be20: 745f 6c6f 6369 5d2c 0a20 2020 2020 2020  t_loci],.       
+0001be30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001be40: 2020 2020 6c6f 6369 5f64 665b 7374 7228      loci_df[str(
+0001be50: 6363 5f61 7869 7332 295d 5b74 7261 6974  cc_axis2)][trait
+0001be60: 5f6c 6f63 695d 2c0a 2020 2020 2020 2020  _loci],.        
+0001be70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001be80: 2020 206d 6172 6b65 723d 222b 222c 2063     marker="+", c
+0001be90: 6f6c 6f72 203d 2022 7265 6422 2c20 733d  olor = "red", s=
+0001bea0: 3130 302c 0a20 2020 2020 2020 2020 2020  100,.           
+0001beb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bec0: 6c61 6265 6c20 3d20 2741 6461 7074 6976  label = 'Adaptiv
+0001bed0: 6520 534e 5073 2729 0a0a 2020 2020 2020  e SNPs')..      
+0001bee0: 2020 2020 2020 2020 2020 2370 6c6f 7420            #plot 
+0001bef0: 7661 7269 6162 6c65 2076 6563 746f 7273  variable vectors
+0001bf00: 2061 7320 6172 726f 7773 0a20 2020 2020   as arrows.     
+0001bf10: 2020 2020 2020 2020 2020 2066 6f72 2069             for i
+0001bf20: 2069 6e20 7261 6e67 6528 7661 725f 6466   in range(var_df
+0001bf30: 2e73 6861 7065 5b30 5d29 3a0a 2020 2020  .shape[0]):.    
+0001bf40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bf50: 7820 3d20 7661 725f 6466 5b73 7472 2863  x = var_df[str(c
+0001bf60: 635f 6178 6973 3129 5d5b 695d 0a20 2020  c_axis1)][i].   
+0001bf70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bf80: 2079 203d 2076 6172 5f64 665b 7374 7228   y = var_df[str(
+0001bf90: 6363 5f61 7869 7332 295d 5b69 5d0a 2020  cc_axis2)][i].  
+0001bfa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bfb0: 2020 706c 742e 6172 726f 7728 302c 2030    plt.arrow(0, 0
+0001bfc0: 2c20 782c 2079 2c20 7769 6474 6820 3d20  , x, y, width = 
+0001bfd0: 302e 3031 2c0a 2020 2020 2020 2020 2020  0.01,.          
+0001bfe0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001bff0: 2020 2020 6865 6164 5f77 6964 7468 203d      head_width =
+0001c000: 2030 2e30 352c 2063 6f6c 6f72 203d 2027   0.05, color = '
+0001c010: 626c 6163 6b27 290a 2020 2020 2020 2020  black').        
+0001c020: 2020 2020 2020 2020 2020 2020 7378 203d              sx =
+0001c030: 2030 2e31 0a20 2020 2020 2020 2020 2020   0.1.           
+0001c040: 2020 2020 2020 2020 2073 7920 3d20 302e           sy = 0.
+0001c050: 310a 2020 2020 2020 2020 2020 2020 2020  1.              
+0001c060: 2020 2020 2020 2320 7468 6973 206d 6573        # this mes
+0001c070: 7320 6973 206a 7573 7420 746f 2061 7272  s is just to arr
+0001c080: 616e 6765 2074 6865 2074 6578 740a 2020  ange the text.  
+0001c090: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c0a0: 2020 2320 6e65 7874 2074 6f20 6172 726f    # next to arro
+0001c0b0: 7773 2062 7574 2069 7420 6973 2061 2057  ws but it is a W
+0001c0c0: 4950 0a20 2020 2020 2020 2020 2020 2020  IP.             
+0001c0d0: 2020 2020 2020 2069 6620 2878 203c 2030         if (x < 0
+0001c0e0: 2061 6e64 2079 203c 2030 293a 0a20 2020   and y < 0):.   
+0001c0f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c100: 2020 2020 706c 742e 7465 7874 2878 202d      plt.text(x -
+0001c110: 2073 782c 2079 202d 2073 792c 2058 2e63   sx, y - sy, X.c
+0001c120: 6f6c 756d 6e73 5b69 5d29 0a20 2020 2020  olumns[i]).     
+0001c130: 2020 2020 2020 2020 2020 2020 2020 2069                 i
+0001c140: 6620 2878 203c 2030 2061 6e64 2079 203e  f (x < 0 and y >
+0001c150: 2030 293a 0a20 2020 2020 2020 2020 2020   0):.           
+0001c160: 2020 2020 2020 2020 2020 2020 706c 742e              plt.
+0001c170: 7465 7874 2878 202d 2073 782c 2079 202b  text(x - sx, y +
+0001c180: 2073 792c 2058 2e63 6f6c 756d 6e73 5b69   sy, X.columns[i
+0001c190: 5d29 0a20 2020 2020 2020 2020 2020 2020  ]).             
+0001c1a0: 2020 2020 2020 2069 6620 2878 203e 2030         if (x > 0
+0001c1b0: 2061 6e64 2079 203e 2030 293a 0a20 2020   and y > 0):.   
+0001c1c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c1d0: 2020 2020 706c 742e 7465 7874 2878 202b      plt.text(x +
+0001c1e0: 2073 782c 2079 202b 2073 792c 2058 2e63   sx, y + sy, X.c
+0001c1f0: 6f6c 756d 6e73 5b69 5d29 0a20 2020 2020  olumns[i]).     
+0001c200: 2020 2020 2020 2020 2020 2020 2020 2069                 i
+0001c210: 6620 2878 203e 2030 2061 6e64 2079 203c  f (x > 0 and y <
+0001c220: 2030 293a 0a20 2020 2020 2020 2020 2020   0):.           
+0001c230: 2020 2020 2020 2020 2020 2020 706c 742e              plt.
+0001c240: 7465 7874 2878 202b 2073 782c 2079 202d  text(x + sx, y -
+0001c250: 2073 792c 2058 2e63 6f6c 756d 6e73 5b69   sy, X.columns[i
+0001c260: 5d29 0a0a 2020 2020 2020 2020 2020 2020  ])..            
+0001c270: 2020 2020 2370 6c6f 7420 5344 2065 6c6c      #plot SD ell
+0001c280: 6970 7365 2066 6f72 2053 4e50 2064 6174  ipse for SNP dat
+0001c290: 6120 6966 2070 6c6f 745f 7364 203d 2054  a if plot_sd = T
+0001c2a0: 7275 650a 2020 2020 2020 2020 2020 2020  rue.            
+0001c2b0: 2020 2020 6966 2070 6c6f 745f 7364 3a0a      if plot_sd:.
+0001c2c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c2d0: 2020 2020 7231 203d 206e 702e 7374 6428      r1 = np.std(
+0001c2e0: 6c6f 6369 5f64 665b 7374 7228 6363 5f61  loci_df[str(cc_a
+0001c2f0: 7869 7331 295d 2920 2a20 7364 0a20 2020  xis1)]) * sd.   
+0001c300: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c310: 2072 3220 3d20 6e70 2e73 7464 286c 6f63   r2 = np.std(loc
+0001c320: 695f 6466 5b73 7472 2863 635f 6178 6973  i_df[str(cc_axis
+0001c330: 3229 5d29 202a 2073 640a 2020 2020 2020  2)]) * sd.      
+0001c340: 2020 2020 2020 2020 2020 2020 2020 7420                t 
+0001c350: 3d20 6e70 2e6c 696e 7370 6163 6528 302c  = np.linspace(0,
+0001c360: 2032 2a6d 6174 685f 7069 2c20 3130 3029   2*math_pi, 100)
+0001c370: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0001c380: 2020 2020 2061 782e 706c 6f74 2872 312a       ax.plot(r1*
+0001c390: 6e70 2e63 6f73 2874 2920 2c20 7232 2a6e  np.cos(t) , r2*n
+0001c3a0: 702e 7369 6e28 7429 2c20 6c69 6e65 7374  p.sin(t), linest
+0001c3b0: 796c 6520 3d20 2764 6173 6865 6427 2c0a  yle = 'dashed',.
 0001c3c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001c3d0: 2020 2020 2020 2020 2020 2020 206c 616e               lan
-0001c3e0: 6420 3d20 6c61 6e64 290a 2020 2020 656c  d = land).    el
-0001c3f0: 7365 3a0a 2020 2020 2020 2020 6765 6e5f  se:.        gen_
-0001c400: 6172 6368 203d 204e 6f6e 650a 0a20 2020  arch = None..   
-0001c410: 2023 2070 7269 6e74 2076 6572 626f 7365   # print verbose
-0001c420: 206f 7574 7075 740a 2020 2020 6966 2076   output.    if v
-0001c430: 6572 626f 7365 3a0a 2020 2020 2020 2020  erbose:.        
-0001c440: 7072 696e 7428 275c 745c 745c 746d 616b  print('\t\t\tmak
-0001c450: 696e 6720 696e 6469 7669 6475 616c 732e  ing individuals.
-0001c460: 2e2e 5c6e 272c 2066 6c75 7368 3d54 7275  ..\n', flush=Tru
-0001c470: 6529 0a20 2020 2023 6d61 6b65 2069 6e64  e).    #make ind
-0001c480: 6976 6964 730a 2020 2020 4e20 3d20 696e  ivids.    N = in
-0001c490: 6974 5f70 6172 616d 732e 706f 7028 274e  it_params.pop('N
-0001c4a0: 2729 0a20 2020 2023 6372 6561 7465 2061  ').    #create a
-0001c4b0: 6e20 6f72 6465 7265 6420 6469 6374 696f  n ordered dictio
-0001c4c0: 6e61 7279 2074 6f20 686f 6c64 2074 6865  nary to hold the
-0001c4d0: 2069 6e64 6976 6964 7561 6c73 2c20 616e   individuals, an
-0001c4e0: 6420 6669 6c6c 2069 7420 7570 0a20 2020  d fill it up.   
-0001c4f0: 2069 6e64 7320 3d20 4f44 2829 0a20 2020   inds = OD().   
-0001c500: 2066 6f72 2069 6e64 5f69 6478 2069 6e20   for ind_idx in 
-0001c510: 7261 6e67 6528 4e29 3a0a 2020 2020 2020  range(N):.      
-0001c520: 2020 2320 7573 6520 696e 6469 7669 6475    # use individu
-0001c530: 616c 2e63 7265 6174 655f 696e 6469 7669  al.create_indivi
-0001c540: 6475 616c 2074 6f20 7369 6d75 6c61 7465  dual to simulate
-0001c550: 2069 6e64 6976 6964 7561 6c73 0a20 2020   individuals.   
-0001c560: 2020 2020 2023 616e 6420 6164 6420 7468       #and add th
-0001c570: 656d 2074 6f20 7468 6520 7370 6563 6965  em to the specie
-0001c580: 730a 2020 2020 2020 2020 696e 6420 3d20  s.        ind = 
-0001c590: 5f6d 616b 655f 696e 6469 7669 6475 616c  _make_individual
-0001c5a0: 2869 6478 3d69 6e64 5f69 6478 2c20 6f66  (idx=ind_idx, of
-0001c5b0: 6673 7072 696e 673d 4661 6c73 652c 0a20  fspring=False,. 
-0001c5c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001c5d0: 2020 2020 2020 2020 2020 2020 2020 6469                di
-0001c5e0: 6d3d 6c61 6e64 2e64 696d 2c20 6765 6e6f  m=land.dim, geno
-0001c5f0: 6d69 635f 6172 6368 6974 6563 7475 7265  mic_architecture
-0001c600: 3d67 656e 5f61 7263 682c 0a20 2020 2020  =gen_arch,.     
-0001c610: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001c620: 2020 2020 2020 2020 2020 6275 726e 3d62            burn=b
-0001c630: 7572 6e29 0a20 2020 2020 2020 2069 6e64  urn).        ind
-0001c640: 735b 696e 645f 6964 785d 203d 2069 6e64  s[ind_idx] = ind
-0001c650: 0a0a 2020 2020 2363 7265 6174 6520 7468  ..    #create th
-0001c660: 6520 7370 6563 6965 7320 6672 6f6d 2074  e species from t
-0001c670: 686f 7365 2069 6e64 6976 6964 7561 6c73  hose individuals
-0001c680: 0a20 2020 2073 7070 203d 2053 7065 6369  .    spp = Speci
-0001c690: 6573 286e 616d 6520 3d20 6e61 6d65 2c20  es(name = name, 
-0001c6a0: 6964 7820 3d20 6964 782c 2069 6e64 7320  idx = idx, inds 
-0001c6b0: 3d20 696e 6473 2c20 6c61 6e64 203d 206c  = inds, land = l
-0001c6c0: 616e 642c 0a20 2020 2020 2020 2020 2020  and,.           
-0001c6d0: 2020 2020 2020 2020 2020 7370 705f 7061            spp_pa
-0001c6e0: 7261 6d73 203d 2073 7070 5f70 6172 616d  rams = spp_param
-0001c6f0: 732c 2067 656e 6f6d 6963 5f61 7263 6869  s, genomic_archi
-0001c700: 7465 6374 7572 653d 6765 6e5f 6172 6368  tecture=gen_arch
-0001c710: 290a 0a20 2020 2023 7573 6520 7468 6520  )..    #use the 
-0001c720: 7265 6d61 696e 696e 6720 696e 6974 5f70  remaining init_p
-0001c730: 6172 616d 7320 746f 2073 6574 2074 6865  arams to set the
-0001c740: 2063 6172 7279 696e 672d 6361 7061 6369   carrying-capaci
-0001c750: 7479 2072 6173 7465 7220 284b 290a 2020  ty raster (K).  
-0001c760: 2020 5f6d 616b 655f 4b28 7370 702c 206c    _make_K(spp, l
-0001c770: 616e 642c 202a 2a69 6e69 745f 7061 7261  and, **init_para
-0001c780: 6d73 290a 2020 2020 2373 6574 2069 6e69  ms).    #set ini
-0001c790: 7469 616c 2065 6e76 6972 6f6e 6d65 6e74  tial environment
-0001c7a0: 2076 616c 7565 730a 2020 2020 7370 702e   values.    spp.
-0001c7b0: 5f73 6574 5f65 286c 616e 6429 0a20 2020  _set_e(land).   
-0001c7c0: 2023 7365 7420 696e 6974 6961 6c20 636f   #set initial co
-0001c7d0: 6f72 6473 2061 6e64 2063 656c 6c73 0a20  ords and cells. 
-0001c7e0: 2020 2073 7070 2e5f 7365 745f 636f 6f72     spp._set_coor
-0001c7f0: 6473 5f61 6e64 5f63 656c 6c73 2829 0a20  ds_and_cells(). 
-0001c800: 2020 2023 7365 7420 7468 6520 6b64 5f74     #set the kd_t
-0001c810: 7265 650a 2020 2020 7370 702e 5f73 6574  ree.    spp._set
-0001c820: 5f6b 645f 7472 6565 2829 0a0a 2020 2020  _kd_tree()..    
-0001c830: 2373 6574 2070 6865 6e6f 7479 7065 732c  #set phenotypes,
-0001c840: 2069 6620 7468 6520 7370 6563 6965 7320   if the species 
-0001c850: 6861 7320 6765 6e6f 6d65 730a 2020 2020  has genomes.    
-0001c860: 6966 2073 7070 2e67 656e 5f61 7263 6820  if spp.gen_arch 
-0001c870: 6973 206e 6f74 204e 6f6e 6520 616e 6420  is not None and 
-0001c880: 6e6f 7420 6275 726e 3a0a 2020 2020 2020  not burn:.      
-0001c890: 2020 5b69 6e64 2e5f 7365 745f 7a28 7370    [ind._set_z(sp
-0001c8a0: 702e 6765 6e5f 6172 6368 2920 666f 7220  p.gen_arch) for 
-0001c8b0: 696e 6420 696e 2073 7070 2e76 616c 7565  ind in spp.value
-0001c8c0: 7328 295d 0a0a 2020 2020 236d 616b 6520  s()]..    #make 
-0001c8d0: 6465 6e73 6974 795f 6772 6964 0a20 2020  density_grid.   
-0001c8e0: 2073 7070 2e5f 7365 745f 6465 6e73 5f67   spp._set_dens_g
-0001c8f0: 7269 6473 286c 616e 6429 0a0a 2020 2020  rids(land)..    
-0001c900: 236d 616b 6520 6d6f 7665 6d65 6e74 2073  #make movement s
-0001c910: 7572 6661 6365 2c20 6966 206e 6565 6465  urface, if neede
-0001c920: 640a 2020 2020 6966 2073 7070 2e5f 6d6f  d.    if spp._mo
-0001c930: 7665 3a0a 2020 2020 2020 2020 6966 2027  ve:.        if '
-0001c940: 6d6f 7665 5f73 7572 6627 2069 6e20 7370  move_surf' in sp
-0001c950: 705f 7061 7261 6d73 2e6d 6f76 656d 656e  p_params.movemen
-0001c960: 742e 6b65 7973 2829 3a0a 2020 2020 2020  t.keys():.      
-0001c970: 2020 2020 2020 6966 2076 6572 626f 7365        if verbose
-0001c980: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-0001c990: 2020 7072 696e 7428 2827 5c74 5c74 5c74    print(('\t\t\t
-0001c9a0: 6d61 6b69 6e67 206d 6f76 656d 656e 7420  making movement 
-0001c9b0: 7375 7266 6163 652e 2e2e 5c6e 270a 2020  surface...\n'.  
-0001c9c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001c9d0: 2020 2020 2027 5c74 5c74 5c74 5c74 5b63       '\t\t\t\t[c
-0001c9e0: 616e 2074 616b 6520 6120 6269 745d 5c6e  an take a bit]\n
-0001c9f0: 2729 2c20 666c 7573 683d 5472 7565 290a  '), flush=True).
-0001ca00: 2020 2020 2020 2020 2020 2020 6d73 5f70              ms_p
-0001ca10: 6172 616d 7320 3d20 6465 6570 636f 7079  arams = deepcopy
-0001ca20: 2873 7070 5f70 6172 616d 732e 6d6f 7665  (spp_params.move
-0001ca30: 6d65 6e74 2e6d 6f76 655f 7375 7266 290a  ment.move_surf).
-0001ca40: 2020 2020 2020 2020 2020 2020 2367 7261              #gra
-0001ca50: 6220 7468 6520 6c79 7220 6e75 6d62 6572  b the lyr number
-0001ca60: 2066 6f72 2074 6865 206c 7972 2074 6861   for the lyr tha
-0001ca70: 7420 7468 6520 0a20 2020 2020 2020 2020  t the .         
-0001ca80: 2020 2023 6d6f 7665 6d65 6e74 2073 7572     #movement sur
-0001ca90: 6661 6365 2069 7320 746f 2062 6520 6261  face is to be ba
-0001caa0: 7365 6420 6f6e 0a20 2020 2020 2020 2020  sed on.         
-0001cab0: 2020 206d 6f76 655f 7375 7266 5f6c 7972     move_surf_lyr
-0001cac0: 203d 206d 735f 7061 7261 6d73 2e70 6f70   = ms_params.pop
-0001cad0: 2827 6c61 7965 7227 290a 2020 2020 2020  ('layer').      
-0001cae0: 2020 2020 2020 6d6f 7665 5f73 7572 665f        move_surf_
-0001caf0: 6c79 725f 6e75 6d20 3d20 5b6b 2066 6f72  lyr_num = [k for
-0001cb00: 206b 2c76 2069 6e20 6c61 6e64 2e69 7465   k,v in land.ite
-0001cb10: 6d73 280a 2020 2020 2020 2020 2020 2020  ms(.            
-0001cb20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cb30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cb40: 2920 6966 2076 2e6e 616d 6520 3d3d 206d  ) if v.name == m
-0001cb50: 6f76 655f 7375 7266 5f6c 7972 5d0a 2020  ove_surf_lyr].  
-0001cb60: 2020 2020 2020 2020 2020 6173 7365 7274            assert
-0001cb70: 206c 656e 286d 6f76 655f 7375 7266 5f6c   len(move_surf_l
-0001cb80: 7972 5f6e 756d 2920 3d3d 2031 2c20 2822  yr_num) == 1, ("
-0001cb90: 4578 7065 6374 6564 2074 6f20 6669 6e64  Expected to find
-0001cba0: 206f 6e6c 7920 6120 220a 2020 2020 2020   only a ".      
-0001cbb0: 2020 2020 2020 2020 2020 2273 696e 676c            "singl
-0001cbc0: 6520 4c61 7965 7220 7769 7468 2074 6865  e Layer with the
-0001cbd0: 206e 616d 6520 7072 6f76 6964 6564 2066   name provided f
-0001cbe0: 6f72 2074 6865 2022 0a20 2020 2020 2020  or the ".       
-0001cbf0: 2020 2020 2020 2020 2022 5f43 6f6e 6475           "_Condu
-0001cc00: 6374 616e 6365 5375 7266 6163 652c 220a  ctanceSurface,".
-0001cc10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cc20: 2220 6275 7420 696e 7374 6561 6420 666f  " but instead fo
-0001cc30: 756e 6420 2569 2229 2025 206c 656e 286d  und %i") % len(m
-0001cc40: 6f76 655f 7375 7266 5f6c 7972 5f6e 756d  ove_surf_lyr_num
-0001cc50: 290a 2020 2020 2020 2020 2020 2020 6d6f  ).            mo
-0001cc60: 7665 5f73 7572 665f 6c79 725f 6e75 6d20  ve_surf_lyr_num 
-0001cc70: 3d20 6d6f 7665 5f73 7572 665f 6c79 725f  = move_surf_lyr_
-0001cc80: 6e75 6d5b 305d 0a20 2020 2020 2020 2020  num[0].         
-0001cc90: 2020 2023 6d61 6b65 2074 6865 206d 6f76     #make the mov
-0001cca0: 656d 656e 7420 7375 7266 6163 6520 616e  ement surface an
-0001ccb0: 6420 7365 7420 6974 2061 7320 7468 6520  d set it as the 
-0001ccc0: 7370 7027 730a 2020 2020 2020 2020 2020  spp's.          
-0001ccd0: 2020 236d 6f76 655f 7375 7266 2061 7474    #move_surf att
-0001cce0: 7269 6275 7465 0a20 2020 2020 2020 2020  ribute.         
-0001ccf0: 2020 2073 7070 2e5f 6d6f 7665 5f73 7572     spp._move_sur
-0001cd00: 663d 2073 7074 2e5f 436f 6e64 7563 7461  f= spt._Conducta
-0001cd10: 6e63 6553 7572 6661 6365 286c 616e 645b  nceSurface(land[
-0001cd20: 6d6f 7665 5f73 7572 665f 6c79 725f 6e75  move_surf_lyr_nu
-0001cd30: 6d5d 2c0a 2020 2020 2020 2020 2020 2020  m],.            
-0001cd40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cd50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cd60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cd70: 2020 2020 2a2a 6d73 5f70 6172 616d 7329      **ms_params)
-0001cd80: 0a20 2020 2023 6d61 6b65 2064 6973 7065  .    #make dispe
-0001cd90: 7273 616c 2073 7572 6661 6365 2c20 6966  rsal surface, if
-0001cda0: 206e 6565 6465 640a 2020 2020 6966 2027   needed.    if '
-0001cdb0: 6469 7370 5f73 7572 6627 2069 6e20 7370  disp_surf' in sp
-0001cdc0: 705f 7061 7261 6d73 2e6d 6f76 656d 656e  p_params.movemen
-0001cdd0: 742e 6b65 7973 2829 3a0a 2020 2020 2020  t.keys():.      
-0001cde0: 2020 2320 7072 696e 7420 7665 7262 6f73    # print verbos
-0001cdf0: 6520 6f75 7470 7574 0a20 2020 2020 2020  e output.       
-0001ce00: 2069 6620 7665 7262 6f73 653a 0a20 2020   if verbose:.   
-0001ce10: 2020 2020 2020 2020 2070 7269 6e74 2828           print((
-0001ce20: 275c 745c 745c 746d 616b 696e 6720 6469  '\t\t\tmaking di
-0001ce30: 7370 6572 7361 6c20 7375 7266 6163 652e  spersal surface.
-0001ce40: 2e2e 5c6e 270a 2020 2020 2020 2020 2020  ..\n'.          
-0001ce50: 2020 2020 2020 2020 2027 5c74 5c74 5c74           '\t\t\t
-0001ce60: 5c74 5b63 616e 2074 616b 6520 6120 6269  \t[can take a bi
-0001ce70: 745d 5c6e 2729 2c20 666c 7573 683d 5472  t]\n'), flush=Tr
-0001ce80: 7565 290a 2020 2020 2020 2020 6473 5f70  ue).        ds_p
-0001ce90: 6172 616d 7320 3d20 6465 6570 636f 7079  arams = deepcopy
-0001cea0: 2873 7070 5f70 6172 616d 732e 6d6f 7665  (spp_params.move
-0001ceb0: 6d65 6e74 2e64 6973 705f 7375 7266 290a  ment.disp_surf).
-0001cec0: 2020 2020 2020 2020 2367 7261 6220 7468          #grab th
-0001ced0: 6520 6c79 7220 6e75 6d62 6572 2066 6f72  e lyr number for
-0001cee0: 2074 6865 206c 7972 2074 6861 7420 7468   the lyr that th
-0001cef0: 6520 0a20 2020 2020 2020 2023 6469 7370  e .        #disp
-0001cf00: 6572 7361 6c20 7375 7266 6163 6520 6973  ersal surface is
-0001cf10: 2074 6f20 6265 2062 6173 6564 206f 6e0a   to be based on.
-0001cf20: 2020 2020 2020 2020 6469 7370 5f73 7572          disp_sur
-0001cf30: 665f 6c79 7220 3d20 6473 5f70 6172 616d  f_lyr = ds_param
-0001cf40: 732e 706f 7028 276c 6179 6572 2729 0a20  s.pop('layer'). 
-0001cf50: 2020 2020 2020 2064 6973 705f 7375 7266         disp_surf
-0001cf60: 5f6c 7972 5f6e 756d 203d 205b 6b20 666f  _lyr_num = [k fo
-0001cf70: 7220 6b2c 7620 696e 206c 616e 642e 6974  r k,v in land.it
-0001cf80: 656d 7328 0a20 2020 2020 2020 2020 2020  ems(.           
-0001cf90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001cfa0: 2020 2020 2020 2020 2020 2020 2029 2069               ) i
-0001cfb0: 6620 762e 6e61 6d65 203d 3d20 6469 7370  f v.name == disp
-0001cfc0: 5f73 7572 665f 6c79 725d 0a20 2020 2020  _surf_lyr].     
-0001cfd0: 2020 2061 7373 6572 7420 6c65 6e28 6469     assert len(di
-0001cfe0: 7370 5f73 7572 665f 6c79 725f 6e75 6d29  sp_surf_lyr_num)
-0001cff0: 203d 3d20 312c 2028 2245 7870 6563 7465   == 1, ("Expecte
-0001d000: 6420 746f 2066 696e 6420 6f6e 6c79 2061  d to find only a
-0001d010: 2022 0a20 2020 2020 2020 2020 2020 2022   ".            "
-0001d020: 7369 6e67 6c65 204c 6179 6572 2077 6974  single Layer wit
-0001d030: 6820 7468 6520 6e61 6d65 2070 726f 7669  h the name provi
-0001d040: 6465 6420 666f 7220 7468 6520 220a 2020  ded for the ".  
-0001d050: 2020 2020 2020 2020 2020 225f 436f 6e64            "_Cond
-0001d060: 7563 7461 6e63 6553 7572 6661 6365 2c20  uctanceSurface, 
-0001d070: 220a 2020 2020 2020 2020 2020 2020 2262  ".            "b
-0001d080: 7574 2069 6e73 7465 6164 2066 6f75 6e64  ut instead found
-0001d090: 2025 6922 2920 2520 6c65 6e28 6469 7370   %i") % len(disp
-0001d0a0: 5f73 7572 665f 6c79 725f 6e75 6d29 0a20  _surf_lyr_num). 
-0001d0b0: 2020 2020 2020 2064 6973 705f 7375 7266         disp_surf
-0001d0c0: 5f6c 7972 5f6e 756d 203d 2064 6973 705f  _lyr_num = disp_
-0001d0d0: 7375 7266 5f6c 7972 5f6e 756d 5b30 5d0a  surf_lyr_num[0].
-0001d0e0: 2020 2020 2020 2020 236d 616b 6520 7468          #make th
-0001d0f0: 6520 6469 7370 6572 7361 6c20 7375 7266  e dispersal surf
-0001d100: 6163 6520 616e 6420 7365 7420 6974 2061  ace and set it a
-0001d110: 7320 7468 6520 7370 7027 730a 2020 2020  s the spp's.    
-0001d120: 2020 2020 2364 6973 705f 7375 7266 2061      #disp_surf a
-0001d130: 7474 7269 6275 7465 0a20 2020 2020 2020  ttribute.       
-0001d140: 2073 7070 2e5f 6469 7370 5f73 7572 6620   spp._disp_surf 
-0001d150: 3d20 7370 742e 5f43 6f6e 6475 6374 616e  = spt._Conductan
-0001d160: 6365 5375 7266 6163 6528 6c61 6e64 5b64  ceSurface(land[d
-0001d170: 6973 705f 7375 7266 5f6c 7972 5f6e 756d  isp_surf_lyr_num
-0001d180: 5d2c 0a20 2020 2020 2020 2020 2020 2020  ],.             
+0001c3d0: 2020 2020 2020 2020 2020 2020 6320 3d20              c = 
+0001c3e0: 2772 6564 272c 206c 6162 656c 203d 2073  'red', label = s
+0001c3f0: 7472 2873 6429 202b 2022 2053 7464 4465  tr(sd) + " StdDe
+0001c400: 7622 290a 0a20 2020 2020 2020 2020 2020  v")..           
+0001c410: 2020 2020 2023 4556 4552 5954 4849 4e47       #EVERYTHING
+0001c420: 2042 454c 4f57 2054 4849 5320 4c49 4e45   BELOW THIS LINE
+0001c430: 2049 4e20 5448 4520 4c4f 4f50 2049 5320   IN THE LOOP IS 
+0001c440: 4a55 5354 2050 4c4f 5420 464f 524d 4154  JUST PLOT FORMAT
+0001c450: 5449 4e47 200a 2020 2020 2020 2020 2020  TING .          
+0001c460: 2020 2020 2020 2378 6c61 6265 6c20 6973        #xlabel is
+0001c470: 2074 6865 206c 6162 656c 2061 7420 7468   the label at th
+0001c480: 6520 626f 7474 6f6d 2028 7665 7274 6963  e bottom (vertic
+0001c490: 616c 2061 7869 7320 6c61 6265 6c29 0a20  al axis label). 
+0001c4a0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+0001c4b0: 2061 6e64 2079 6c61 6265 6c20 6973 206f   and ylabel is o
+0001c4c0: 6e20 7468 6520 6c65 6674 2073 6964 6520  n the left side 
+0001c4d0: 2868 6f72 697a 6f6e 7461 6c20 6178 6973  (horizontal axis
+0001c4e0: 206c 6162 656c 290a 2020 2020 2020 2020   label).        
+0001c4f0: 2020 2020 2020 2020 6c61 6265 6c20 3d20          label = 
+0001c500: 6178 2e73 6574 5f78 6c61 6265 6c28 2743  ax.set_xlabel('C
+0001c510: 4341 2720 2b20 7374 7228 6363 5f61 7869  CA' + str(cc_axi
+0001c520: 7332 292c 2066 6f6e 7473 697a 6520 3d20  s2), fontsize = 
+0001c530: 3929 0a20 2020 2020 2020 2020 2020 2020  9).             
+0001c540: 2020 2061 782e 7861 7869 732e 7365 745f     ax.xaxis.set_
+0001c550: 6c61 6265 6c5f 636f 6f72 6473 2830 2e35  label_coords(0.5
+0001c560: 2c20 2d30 2e30 3229 0a20 2020 2020 2020  , -0.02).       
+0001c570: 2020 2020 2020 2020 206c 6162 656c 203d           label =
+0001c580: 2061 782e 7365 745f 796c 6162 656c 2827   ax.set_ylabel('
+0001c590: 4343 4127 202b 2073 7472 2863 635f 6178  CCA' + str(cc_ax
+0001c5a0: 6973 3129 2c20 666f 6e74 7369 7a65 203d  is1), fontsize =
+0001c5b0: 2039 290a 2020 2020 2020 2020 2020 2020   9).            
+0001c5c0: 2020 2020 6178 2e79 6178 6973 2e73 6574      ax.yaxis.set
+0001c5d0: 5f6c 6162 656c 5f63 6f6f 7264 7328 2d30  _label_coords(-0
+0001c5e0: 2e30 322c 2030 2e35 3529 0a0a 2020 2020  .02, 0.55)..    
+0001c5f0: 2020 2020 2020 2020 2020 2020 236d 616b              #mak
+0001c600: 6520 696e 746f 2062 6f78 0a20 2020 2020  e into box.     
+0001c610: 2020 2020 2020 2020 2020 2061 782e 7365             ax.se
+0001c620: 745f 6173 7065 6374 2827 6571 7561 6c27  t_aspect('equal'
+0001c630: 2c20 6164 6a75 7374 6162 6c65 3d27 626f  , adjustable='bo
+0001c640: 7827 290a 0a20 2020 2020 2020 2020 2020  x')..           
+0001c650: 2020 2020 2023 4d6f 7665 206c 6566 7420       #Move left 
+0001c660: 792d 6178 6973 2026 2062 6f74 746f 6d20  y-axis & bottom 
+0001c670: 782d 6178 6973 2074 6f20 6365 6e74 6572  x-axis to center
+0001c680: 2c20 7061 7373 696e 6720 7468 7275 2028  , passing thru (
+0001c690: 302c 3029 0a20 2020 2020 2020 2020 2020  0,0).           
+0001c6a0: 2020 2020 2061 782e 7370 696e 6573 5b27       ax.spines['
+0001c6b0: 6c65 6674 275d 2e73 6574 5f70 6f73 6974  left'].set_posit
+0001c6c0: 696f 6e28 2763 656e 7465 7227 290a 2020  ion('center').  
+0001c6d0: 2020 2020 2020 2020 2020 2020 2020 6178                ax
+0001c6e0: 2e73 7069 6e65 735b 2762 6f74 746f 6d27  .spines['bottom'
+0001c6f0: 5d2e 7365 745f 706f 7369 7469 6f6e 2827  ].set_position('
+0001c700: 6365 6e74 6572 2729 0a0a 2020 2020 2020  center')..      
+0001c710: 2020 2020 2020 2020 2020 2363 6861 6e67            #chang
+0001c720: 6520 636f 6c6f 7273 2f72 656d 6f76 6520  e colors/remove 
+0001c730: 6465 6661 756c 7420 6178 6973 0a20 2020  default axis.   
+0001c740: 2020 2020 2020 2020 2020 2020 2061 7863               axc
+0001c750: 6f6c 6f72 203d 2027 6e6f 6e65 270a 2020  olor = 'none'.  
+0001c760: 2020 2020 2020 2020 2020 2020 2020 6178                ax
+0001c770: 2e73 7069 6e65 735b 276c 6566 7427 5d2e  .spines['left'].
+0001c780: 7365 745f 636f 6c6f 7228 6178 636f 6c6f  set_color(axcolo
+0001c790: 7229 0a20 2020 2020 2020 2020 2020 2020  r).             
+0001c7a0: 2020 2061 782e 7370 696e 6573 5b27 626f     ax.spines['bo
+0001c7b0: 7474 6f6d 275d 2e73 6574 5f63 6f6c 6f72  ttom'].set_color
+0001c7c0: 2861 7863 6f6c 6f72 290a 2020 2020 2020  (axcolor).      
+0001c7d0: 2020 2020 2020 2020 2020 6178 2e73 7069            ax.spi
+0001c7e0: 6e65 735b 2772 6967 6874 275d 2e73 6574  nes['right'].set
+0001c7f0: 5f63 6f6c 6f72 2861 7863 6f6c 6f72 290a  _color(axcolor).
+0001c800: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c810: 6178 2e73 7069 6e65 735b 2774 6f70 275d  ax.spines['top']
+0001c820: 2e73 6574 5f63 6f6c 6f72 2861 7863 6f6c  .set_color(axcol
+0001c830: 6f72 290a 2020 2020 2020 2020 2020 2020  or).            
+0001c840: 2020 2020 6178 2e74 6963 6b5f 7061 7261      ax.tick_para
+0001c850: 6d73 2861 7869 7320 3d20 2778 272c 2063  ms(axis = 'x', c
+0001c860: 6f6c 6f72 733d 6178 636f 6c6f 7229 0a20  olors=axcolor). 
+0001c870: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+0001c880: 782e 7469 636b 5f70 6172 616d 7328 6178  x.tick_params(ax
+0001c890: 6973 203d 2027 7927 2c20 636f 6c6f 7273  is = 'y', colors
+0001c8a0: 3d61 7863 6f6c 6f72 290a 0a20 2020 2020  =axcolor)..     
+0001c8b0: 2020 2020 2020 2020 2020 2061 782e 7961             ax.ya
+0001c8c0: 7869 732e 6c61 6265 6c2e 7365 745f 636f  xis.label.set_co
+0001c8d0: 6c6f 7228 2767 7261 7927 290a 2020 2020  lor('gray').    
+0001c8e0: 2020 2020 2020 2020 2020 2020 6178 2e78              ax.x
+0001c8f0: 6178 6973 2e6c 6162 656c 2e73 6574 5f63  axis.label.set_c
+0001c900: 6f6c 6f72 2827 6772 6179 2729 0a0a 2020  olor('gray')..  
+0001c910: 2020 2020 2020 2020 2020 2020 2020 236c                #l
+0001c920: 6567 656e 6473 0a20 2020 2020 2020 2020  egends.         
+0001c930: 2020 2020 2020 2069 6620 6e20 3d3d 2032         if n == 2
+0001c940: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+0001c950: 2020 2020 2020 6178 2e6c 6567 656e 6428        ax.legend(
+0001c960: 6c6f 6320 3d20 2275 7070 6572 206c 6566  loc = "upper lef
+0001c970: 7422 2c20 6262 6f78 5f74 6f5f 616e 6368  t", bbox_to_anch
+0001c980: 6f72 3d28 312e 332c 2031 2e30 3529 290a  or=(1.3, 1.05)).
+0001c990: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001c9a0: 2020 2020 706c 742e 636f 6c6f 7262 6172      plt.colorbar
+0001c9b0: 2863 6d61 705f 7363 6174 2c20 6c61 6265  (cmap_scat, labe
+0001c9c0: 6c3d 2265 6e76 2229 0a20 2020 2020 2020  l="env").       
+0001c9d0: 2020 2020 2066 6967 2e73 686f 7728 290a       fig.show().
+0001c9e0: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
+0001c9f0: 6363 615f 6469 6374 0a0a 0a20 2020 2023  cca_dict...    #
+0001ca00: 206d 6574 686f 6420 666f 7220 706c 6f74   method for plot
+0001ca10: 7469 6e67 2074 6865 2073 7065 6369 6573  ting the species
+0001ca20: 2028 6f72 2061 2073 7562 7365 7420 6f66   (or a subset of
+0001ca30: 2069 7473 2069 6e64 6976 6964 7561 6c73   its individuals
+0001ca40: 2c20 6279 2049 4429 0a20 2020 2023 206f  , by ID).    # o
+0001ca50: 6e20 746f 7020 6f66 2061 206c 6179 6572  n top of a layer
+0001ca60: 2028 6f72 206c 616e 6473 6361 7065 290a   (or landscape).
+0001ca70: 2020 2020 6465 6620 5f70 6c6f 7428 7365      def _plot(se
+0001ca80: 6c66 2c20 6c79 725f 6e75 6d3d 4e6f 6e65  lf, lyr_num=None
+0001ca90: 2c20 6c61 6e64 3d4e 6f6e 652c 2068 6964  , land=None, hid
+0001caa0: 655f 6c61 6e64 3d46 616c 7365 2c20 696e  e_land=False, in
+0001cab0: 6469 7669 6473 3d4e 6f6e 652c 0a20 2020  divids=None,.   
+0001cac0: 2020 2020 2020 2020 2020 2074 6578 743d             text=
+0001cad0: 4661 6c73 652c 2063 6f6c 6f72 3d27 626c  False, color='bl
+0001cae0: 6163 6b27 2c20 6564 6765 5f63 6f6c 6f72  ack', edge_color
+0001caf0: 3d27 6661 6365 272c 2074 6578 745f 636f  ='face', text_co
+0001cb00: 6c6f 723d 2762 6c61 636b 272c 0a20 2020  lor='black',.   
+0001cb10: 2020 2020 2020 2020 2020 2063 6261 723d             cbar=
+0001cb20: 5472 7565 2c20 7369 7a65 3d32 352c 2074  True, size=25, t
+0001cb30: 6578 745f 7369 7a65 3d39 2c20 6c61 6e64  ext_size=9, land
+0001cb40: 5f63 6d61 703d 4e6f 6e65 2c20 7074 5f63  _cmap=None, pt_c
+0001cb50: 6d61 703d 4e6f 6e65 2c0a 2020 2020 2020  map=None,.      
+0001cb60: 2020 2020 2020 2020 616c 7068 613d 4661          alpha=Fa
+0001cb70: 6c73 652c 207a 6f6f 6d5f 7769 6474 683d  lse, zoom_width=
+0001cb80: 4e6f 6e65 2c20 783d 4e6f 6e65 2c20 793d  None, x=None, y=
+0001cb90: 4e6f 6e65 2c20 766d 696e 3d4e 6f6e 652c  None, vmin=None,
+0001cba0: 0a20 2020 2020 2020 2020 2020 2020 2076  .              v
+0001cbb0: 6d61 783d 4e6f 6e65 2c20 7469 636b 733d  max=None, ticks=
+0001cbc0: 4e6f 6e65 2c20 6d61 736b 5f72 6173 743d  None, mask_rast=
+0001cbd0: 4e6f 6e65 2c20 616e 696d 6174 653d 4661  None, animate=Fa
+0001cbe0: 6c73 652c 0a20 2020 2020 2020 2020 2020  lse,.           
+0001cbf0: 2020 2063 656c 6c5f 636f 6f72 6473 3d46     cell_coords=F
+0001cc00: 616c 7365 293a 0a20 2020 2020 2020 2023  alse):.        #
+0001cc10: 2063 6f6e 7665 7274 2069 6e64 6976 6964   convert individ
+0001cc20: 7320 746f 2061 206c 6973 7420 2869 6e20  s to a list (in 
+0001cc30: 6361 7365 2063 6f6d 6573 2069 6e20 6173  case comes in as
+0001cc40: 2061 206e 756d 7079 2061 7272 6179 290a   a numpy array).
+0001cc50: 2020 2020 2020 2020 6966 2069 6e64 6976          if indiv
+0001cc60: 6964 7320 6973 206e 6f74 204e 6f6e 6520  ids is not None 
+0001cc70: 616e 6420 6e6f 7420 6973 696e 7374 616e  and not isinstan
+0001cc80: 6365 2869 6e64 6976 6964 732c 206c 6973  ce(individs, lis
+0001cc90: 7429 3a0a 2020 2020 2020 2020 2020 2020  t):.            
+0001cca0: 696e 6469 7669 6473 203d 206c 6973 7428  individs = list(
+0001ccb0: 696e 6469 7669 6473 290a 2020 2020 2020  individs).      
+0001ccc0: 2020 2320 6765 7420 636f 6f72 6473 0a20    # get coords. 
+0001ccd0: 2020 2020 2020 2063 6f6f 7264 7320 3d20         coords = 
+0001cce0: 7365 6c66 2e5f 6765 745f 706c 6f74 5f63  self._get_plot_c
+0001ccf0: 6f6f 7264 7328 696e 6469 7669 6473 3d69  oords(individs=i
+0001cd00: 6e64 6976 6964 732c 0a20 2020 2020 2020  ndivids,.       
+0001cd10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001cd20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001cd30: 6365 6c6c 5f63 6f6f 7264 733d 6365 6c6c  cell_coords=cell
+0001cd40: 5f63 6f6f 7264 7329 0a20 2020 2020 2020  _coords).       
+0001cd50: 2023 2067 6574 2074 6578 740a 2020 2020   # get text.    
+0001cd60: 2020 2020 6966 2074 6578 743a 0a20 2020      if text:.   
+0001cd70: 2020 2020 2020 2020 2069 6620 696e 6469           if indi
+0001cd80: 7669 6473 2069 7320 4e6f 6e65 3a0a 2020  vids is None:.  
+0001cd90: 2020 2020 2020 2020 2020 2020 2020 7465                te
+0001cda0: 7874 203d 205b 2a73 656c 665d 0a20 2020  xt = [*self].   
+0001cdb0: 2020 2020 2020 2020 2065 6c73 653a 0a20           else:. 
+0001cdc0: 2020 2020 2020 2020 2020 2020 2020 2074                 t
+0001cdd0: 6578 7420 3d20 696e 6469 7669 6473 0a20  ext = individs. 
+0001cde0: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
+0001cdf0: 2020 2020 2020 2020 2074 6578 7420 3d20           text = 
+0001ce00: 4e6f 6e65 0a20 2020 2020 2020 2023 2073  None.        # s
+0001ce10: 6574 2074 6865 2070 6c74 5f6c 696d 730a  et the plt_lims.
+0001ce20: 2020 2020 2020 2020 706c 745f 6c69 6d73          plt_lims
+0001ce30: 203d 2076 697a 2e5f 6765 745f 706c 745f   = viz._get_plt_
+0001ce40: 6c69 6d73 286c 616e 642c 2078 2c20 792c  lims(land, x, y,
+0001ce50: 207a 6f6f 6d5f 7769 6474 6829 0a20 2020   zoom_width).   
+0001ce60: 2020 2020 2023 2070 6c6f 7420 7468 6520       # plot the 
+0001ce70: 6c61 7965 7228 7329 0a20 2020 2020 2020  layer(s).       
+0001ce80: 2069 6620 6869 6465 5f6c 616e 643a 0a20   if hide_land:. 
+0001ce90: 2020 2020 2020 2020 2020 2070 6173 730a             pass.
+0001cea0: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+0001ceb0: 2020 2020 2020 2020 2020 2320 6765 7420            # get 
+0001cec0: 7468 6520 6c61 7965 7273 2720 766d 696e  the layers' vmin
+0001ced0: 2061 6e64 2076 6d61 7820 7661 6c75 6573   and vmax values
+0001cee0: 2c20 6966 2061 6e79 206f 6620 7468 6520  , if any of the 
+0001cef0: 6c61 7965 7273 2074 6f0a 2020 2020 2020  layers to.      
+0001cf00: 2020 2020 2020 2320 6265 2070 6c6f 7474        # be plott
+0001cf10: 6564 2068 6173 2061 2063 6861 6e67 6520  ed has a change 
+0001cf20: 6576 656e 740a 2020 2020 2020 2020 2020  event.          
+0001cf30: 2020 6966 2028 286c 7972 5f6e 756d 2069    if ((lyr_num i
+0001cf40: 7320 4e6f 6e65 2061 6e64 206c 616e 642e  s None and land.
+0001cf50: 5f63 6861 6e67 6572 2069 7320 6e6f 7420  _changer is not 
+0001cf60: 4e6f 6e65 2920 6f72 0a20 2020 2020 2020  None) or.       
+0001cf70: 2020 2020 2020 2020 2028 6c61 6e64 2e5f           (land._
+0001cf80: 6368 616e 6765 7220 6973 206e 6f74 204e  changer is not N
+0001cf90: 6f6e 650a 2020 2020 2020 2020 2020 2020  one.            
+0001cfa0: 2020 2020 2061 6e64 206c 7972 5f6e 756d       and lyr_num
+0001cfb0: 2069 6e20 5b2a 6c61 6e64 2e5f 6368 616e   in [*land._chan
+0001cfc0: 6765 722e 6368 616e 6765 5f69 6e66 6f5d  ger.change_info]
+0001cfd0: 2929 3a0a 2020 2020 2020 2020 2020 2020  )):.            
+0001cfe0: 2020 2020 6966 206c 7972 5f6e 756d 2069      if lyr_num i
+0001cff0: 7320 4e6f 6e65 3a0a 2020 2020 2020 2020  s None:.        
+0001d000: 2020 2020 2020 2020 2020 2020 6c61 6e64              land
+0001d010: 5f76 6d69 6e20 3d20 5b6c 7972 2e5f 7363  _vmin = [lyr._sc
+0001d020: 616c 655f 6d69 6e20 666f 7220 6c79 7220  ale_min for lyr 
+0001d030: 696e 206c 616e 642e 7661 6c75 6573 2829  in land.values()
+0001d040: 5d0a 2020 2020 2020 2020 2020 2020 2020  ].              
+0001d050: 2020 2020 2020 6c61 6e64 5f76 6d61 7820        land_vmax 
+0001d060: 3d20 5b6c 7972 2e5f 7363 616c 655f 6d61  = [lyr._scale_ma
+0001d070: 7820 666f 7220 6c79 7220 696e 206c 616e  x for lyr in lan
+0001d080: 642e 7661 6c75 6573 2829 5d0a 2020 2020  d.values()].    
+0001d090: 2020 2020 2020 2020 2020 2020 656c 7365              else
+0001d0a0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+0001d0b0: 2020 2020 2020 6c61 6e64 5f76 6d69 6e20        land_vmin 
+0001d0c0: 3d20 5b6c 616e 645b 6c79 725f 6e75 6d5d  = [land[lyr_num]
+0001d0d0: 2e5f 7363 616c 655f 6d69 6e5d 0a20 2020  ._scale_min].   
+0001d0e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d0f0: 206c 616e 645f 766d 6178 203d 205b 6c61   land_vmax = [la
+0001d100: 6e64 5b6c 7972 5f6e 756d 5d2e 5f73 6361  nd[lyr_num]._sca
+0001d110: 6c65 5f6d 6178 5d0a 2020 2020 2020 2020  le_max].        
+0001d120: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+0001d130: 2020 2020 2020 2020 2020 6c61 6e64 5f76            land_v
+0001d140: 6d69 6e20 3d20 6c61 6e64 5f76 6d61 7820  min = land_vmax 
+0001d150: 3d20 4e6f 6e65 0a20 2020 2020 2020 2020  = None.         
+0001d160: 2020 2076 697a 2e5f 706c 6f74 5f72 6173     viz._plot_ras
+0001d170: 7465 7273 286c 616e 642c 206c 7972 5f6e  ters(land, lyr_n
+0001d180: 756d 3d6c 7972 5f6e 756d 2c0a 2020 2020  um=lyr_num,.    
 0001d190: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001d1a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001d1b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0001d1c0: 2020 202a 2a64 735f 7061 7261 6d73 290a     **ds_params).
-0001d1d0: 0a20 2020 2023 6966 2074 6869 7320 7370  .    #if this sp
-0001d1e0: 6563 6965 7320 6861 7320 6368 616e 6765  ecies has change
-0001d1f0: 7320 7061 7261 6d65 7465 7269 7a65 642c  s parameterized,
-0001d200: 206f 7220 6966 206e 6f74 2062 7574 2069   or if not but i
-0001d210: 7420 6861 730a 2020 2020 2365 6974 6865  t has.    #eithe
-0001d220: 7220 6120 4d6f 7665 6d65 6e74 5375 7266  r a MovementSurf
-0001d230: 206f 7220 6120 4469 7370 6572 7361 6c53   or a DispersalS
-0001d240: 7572 6620 6261 7365 6420 6f6e 2061 204c  urf based on a L
-0001d250: 6179 6572 2074 6861 740a 2020 2020 2377  ayer that.    #w
-0001d260: 696c 6c20 756e 6465 7267 6f20 6c61 6e64  ill undergo land
-0001d270: 7363 6170 6520 6368 616e 6765 2c20 7468  scape change, th
-0001d280: 656e 2063 7265 6174 6520 6120 5f53 7065  en create a _Spe
-0001d290: 6369 6573 4368 616e 6765 7220 6f62 6a65  ciesChanger obje
-0001d2a0: 6374 2066 6f72 2069 740a 2020 2020 6966  ct for it.    if
-0001d2b0: 2028 2763 6861 6e67 6527 2069 6e20 7370   ('change' in sp
-0001d2c0: 705f 7061 7261 6d73 2e6b 6579 7328 290a  p_params.keys().
-0001d2d0: 2020 2020 2020 2020 6f72 2028 7370 702e          or (spp.
-0001d2e0: 5f6d 6f76 655f 7375 7266 2069 7320 6e6f  _move_surf is no
-0001d2f0: 7420 4e6f 6e65 0a20 2020 2020 2020 2061  t None.        a
-0001d300: 6e64 206c 616e 642e 5f63 6861 6e67 6572  nd land._changer
-0001d310: 2069 7320 6e6f 7420 4e6f 6e65 0a20 2020   is not None.   
-0001d320: 2020 2020 2061 6e64 2073 7070 2e5f 6d6f       and spp._mo
-0001d330: 7665 5f73 7572 662e 6c79 725f 6e75 6d20  ve_surf.lyr_num 
-0001d340: 696e 206c 616e 642e 5f63 6861 6e67 6572  in land._changer
-0001d350: 2e63 6861 6e67 655f 696e 666f 2e6b 6579  .change_info.key
-0001d360: 7328 2929 0a20 2020 2020 2020 206f 7220  s()).        or 
-0001d370: 2873 7070 2e5f 6469 7370 5f73 7572 6620  (spp._disp_surf 
-0001d380: 6973 206e 6f74 204e 6f6e 650a 2020 2020  is not None.    
-0001d390: 2020 2020 616e 6420 6c61 6e64 2e5f 6368      and land._ch
-0001d3a0: 616e 6765 7220 6973 206e 6f74 204e 6f6e  anger is not Non
-0001d3b0: 650a 2020 2020 2020 2020 616e 6420 7370  e.        and sp
-0001d3c0: 702e 5f64 6973 705f 7375 7266 2e6c 7972  p._disp_surf.lyr
-0001d3d0: 5f6e 756d 2069 6e20 6c61 6e64 2e5f 6368  _num in land._ch
-0001d3e0: 616e 6765 722e 6368 616e 6765 5f69 6e66  anger.change_inf
-0001d3f0: 6f2e 6b65 7973 2829 2929 3a0a 2020 2020  o.keys())):.    
-0001d400: 2020 2020 2320 7072 696e 7420 7665 7262      # print verb
-0001d410: 6f73 6520 6f75 7470 7574 0a20 2020 2020  ose output.     
-0001d420: 2020 2069 6620 7665 7262 6f73 653a 0a20     if verbose:. 
-0001d430: 2020 2020 2020 2020 2020 2070 7269 6e74             print
-0001d440: 2828 275c 745c 745c 7473 6574 7469 6e67  (('\t\t\tsetting
-0001d450: 2075 7020 7370 6563 6965 7320 6368 616e   up species chan
-0001d460: 6765 732e 2e2e 5c6e 270a 2020 2020 2020  ges...\n'.      
-0001d470: 2020 2020 2020 2020 2020 2020 2027 5c74               '\t
-0001d480: 5c74 5c74 5c74 5b63 616e 2074 616b 6520  \t\t\t[can take 
-0001d490: 6120 7768 696c 652c 5c6e 5c74 5c74 5c74  a while,\n\t\t\t
-0001d4a0: 5c74 2069 6620 6d6f 7665 6d65 6e74 206f  \t if movement o
-0001d4b0: 7220 270a 2020 2020 2020 2020 2020 2020  r '.            
-0001d4c0: 2020 2020 2020 2027 6469 7370 6572 7361         'dispersa
-0001d4d0: 6c5c 6e5c 745c 745c 745c 7420 7375 7266  l\n\t\t\t\t surf
-0001d4e0: 6163 6573 2077 696c 6c20 6368 616e 6765  aces will change
-0001d4f0: 5d5c 6e27 292c 2066 6c75 7368 3d54 7275  ]\n'), flush=Tru
-0001d500: 6529 0a20 2020 2020 2020 2023 6772 6162  e).        #grab
-0001d510: 2074 6865 2063 6861 6e67 6520 7061 7261   the change para
-0001d520: 6d73 2028 6f72 204e 6f6e 652c 2069 660a  ms (or None, if.
-0001d530: 2020 2020 2020 2020 6966 2027 6368 616e          if 'chan
-0001d540: 6765 2720 696e 2073 7070 5f70 6172 616d  ge' in spp_param
-0001d550: 732e 6b65 7973 2829 3a0a 2020 2020 2020  s.keys():.      
-0001d560: 2020 2020 2020 6368 5f70 6172 616d 7320        ch_params 
-0001d570: 3d20 7370 705f 7061 7261 6d73 2e63 6861  = spp_params.cha
-0001d580: 6e67 650a 2020 2020 2020 2020 656c 7365  nge.        else
-0001d590: 3a0a 2020 2020 2020 2020 2020 2020 6368  :.            ch
-0001d5a0: 5f70 6172 616d 7320 3d20 4e6f 6e65 0a20  _params = None. 
-0001d5b0: 2020 2020 2020 2023 6d61 6b65 205f 5370         #make _Sp
-0001d5c0: 6563 6965 7343 6861 6e67 6572 2061 6e64  eciesChanger and
-0001d5d0: 2073 6574 2069 7420 746f 2074 6865 2073   set it to the s
-0001d5e0: 7070 2773 2063 6861 6e67 6572 2061 7474  pp's changer att
-0001d5f0: 7269 6275 7465 0a20 2020 2020 2020 2073  ribute.        s
-0001d600: 7070 2e5f 6368 616e 6765 7220 3d20 5f53  pp._changer = _S
-0001d610: 7065 6369 6573 4368 616e 6765 7228 7370  peciesChanger(sp
-0001d620: 702c 2063 685f 7061 7261 6d73 2c20 6c61  p, ch_params, la
-0001d630: 6e64 203d 206c 616e 6429 0a0a 2020 2020  nd = land)..    
-0001d640: 7265 7475 726e 2073 7070 0a0a 0a23 2066  return spp...# f
-0001d650: 756e 6374 696f 6e20 666f 7220 7265 6164  unction for read
-0001d660: 696e 6720 696e 2061 2070 6963 6b6c 6564  ing in a pickled
-0001d670: 2073 7070 0a64 6566 2072 6561 645f 7069   spp.def read_pi
-0001d680: 636b 6c65 645f 7370 7028 6669 6c65 6e61  ckled_spp(filena
-0001d690: 6d65 293a 0a20 2020 2069 6d70 6f72 7420  me):.    import 
-0001d6a0: 6350 6963 6b6c 650a 2020 2020 7769 7468  cPickle.    with
-0001d6b0: 206f 7065 6e28 6669 6c65 6e61 6d65 2c20   open(filename, 
-0001d6c0: 2772 6227 2920 6173 2066 3a0a 2020 2020  'rb') as f:.    
-0001d6d0: 2020 2020 7370 7020 3d20 6350 6963 6b6c      spp = cPickl
-0001d6e0: 652e 6c6f 6164 2866 290a 2020 2020 7265  e.load(f).    re
-0001d6f0: 7475 726e 2073 7070 0a0a 0a0a            turn spp....
+0001d1a0: 2020 2020 2020 2020 2020 6362 6172 3d63            cbar=c
+0001d1b0: 6261 722c 2063 6d61 703d 6c61 6e64 5f63  bar, cmap=land_c
+0001d1c0: 6d61 702c 2070 6c74 5f6c 696d 733d 706c  map, plt_lims=pl
+0001d1d0: 745f 6c69 6d73 2c0a 2020 2020 2020 2020  t_lims,.        
+0001d1e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d1f0: 2020 2020 2020 7469 636b 733d 7469 636b        ticks=tick
+0001d200: 732c 206d 6173 6b5f 7261 7374 3d6d 6173  s, mask_rast=mas
+0001d210: 6b5f 7261 7374 2c0a 2020 2020 2020 2020  k_rast,.        
+0001d220: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d230: 2020 2020 2020 766d 696e 3d6c 616e 645f        vmin=land_
+0001d240: 766d 696e 2c20 766d 6178 3d6c 616e 645f  vmin, vmax=land_
+0001d250: 766d 6178 290a 2020 2020 2020 2020 2320  vmax).        # 
+0001d260: 616e 6420 706c 6f74 2074 6865 2069 6e64  and plot the ind
+0001d270: 6976 6964 7561 6c73 0a20 2020 2020 2020  ividuals.       
+0001d280: 2070 6f69 6e74 7320 3d20 7669 7a2e 5f70   points = viz._p
+0001d290: 6c6f 745f 706f 696e 7473 2863 6f6f 7264  lot_points(coord
+0001d2a0: 732c 206c 7972 5f6e 756d 3d6c 7972 5f6e  s, lyr_num=lyr_n
+0001d2b0: 756d 2c20 636f 6c6f 723d 636f 6c6f 722c  um, color=color,
+0001d2c0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0001d2d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d2e0: 2020 2065 6467 655f 636f 6c6f 723d 6564     edge_color=ed
+0001d2f0: 6765 5f63 6f6c 6f72 2c0a 2020 2020 2020  ge_color,.      
+0001d300: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d310: 2020 2020 2020 2020 2020 2020 7465 7874              text
+0001d320: 5f63 6f6c 6f72 3d74 6578 745f 636f 6c6f  _color=text_colo
+0001d330: 722c 2073 697a 653d 7369 7a65 2c0a 2020  r, size=size,.  
+0001d340: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d350: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d360: 7465 7874 5f73 697a 653d 7465 7874 5f73  text_size=text_s
+0001d370: 697a 652c 2061 6c70 6861 3d61 6c70 6861  ize, alpha=alpha
+0001d380: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001d390: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d3a0: 2020 2020 7465 7874 3d74 6578 742c 2070      text=text, p
+0001d3b0: 6c74 5f6c 696d 733d 706c 745f 6c69 6d73  lt_lims=plt_lims
+0001d3c0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001d3d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d3e0: 2020 2020 7074 5f63 6d61 703d 7074 5f63      pt_cmap=pt_c
+0001d3f0: 6d61 702c 2076 6d69 6e3d 766d 696e 2c20  map, vmin=vmin, 
+0001d400: 766d 6178 3d76 6d61 782c 0a20 2020 2020  vmax=vmax,.     
+0001d410: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d420: 2020 2020 2020 2020 2020 2020 2061 6e69               ani
+0001d430: 6d61 7465 3d61 6e69 6d61 7465 290a 2020  mate=animate).  
+0001d440: 2020 2020 2020 7265 7475 726e 2070 6f69        return poi
+0001d450: 6e74 730a 0a0a 2020 2020 236d 6574 686f  nts...    #metho
+0001d460: 6420 666f 7220 706c 6f74 7469 6e67 2074  d for plotting t
+0001d470: 6865 2073 7065 6369 6573 206f 6e20 746f  he species on to
+0001d480: 7020 6f66 2069 7473 2065 7374 696d 6174  p of its estimat
+0001d490: 6564 0a20 2020 2023 7370 6563 6965 732d  ed.    #species-
+0001d4a0: 6465 6e73 6974 7920 7261 7374 6572 0a20  density raster. 
+0001d4b0: 2020 2064 6566 205f 706c 6f74 5f64 656e     def _plot_den
+0001d4c0: 7369 7479 2873 656c 662c 206c 616e 642c  sity(self, land,
+0001d4d0: 206e 6f72 6d61 6c69 7a65 3d46 616c 7365   normalize=False
+0001d4e0: 2c20 696e 6469 7669 6473 3d4e 6f6e 652c  , individs=None,
+0001d4f0: 0a20 2020 2020 2020 2020 2020 2074 6578  .            tex
+0001d500: 743d 4661 6c73 652c 2063 6f6c 6f72 3d27  t=False, color='
+0001d510: 626c 6163 6b27 2c20 6564 6765 5f63 6f6c  black', edge_col
+0001d520: 6f72 3d27 6661 6365 272c 0a20 2020 2020  or='face',.     
+0001d530: 2020 2020 2020 2074 6578 745f 636f 6c6f         text_colo
+0001d540: 723d 2762 6c61 636b 272c 2073 697a 653d  r='black', size=
+0001d550: 3235 2c20 7465 7874 5f73 697a 6520 3d20  25, text_size = 
+0001d560: 392c 0a20 2020 2020 2020 2020 2020 2061  9,.            a
+0001d570: 6c70 6861 3d30 2e35 2c20 7a6f 6f6d 5f77  lpha=0.5, zoom_w
+0001d580: 6964 7468 3d4e 6f6e 652c 2078 3d4e 6f6e  idth=None, x=Non
+0001d590: 652c 2079 3d4e 6f6e 652c 2074 6963 6b73  e, y=None, ticks
+0001d5a0: 3d4e 6f6e 652c 0a20 2020 2020 2020 2020  =None,.         
+0001d5b0: 2020 206d 6173 6b5f 7261 7374 3d4e 6f6e     mask_rast=Non
+0001d5c0: 6529 3a0a 2020 2020 2020 2020 6173 7365  e):.        asse
+0001d5d0: 7274 2074 7970 6528 6e6f 726d 616c 697a  rt type(normaliz
+0001d5e0: 6529 2069 7320 626f 6f6c 2c20 2822 5468  e) is bool, ("Th
+0001d5f0: 6520 276e 6f72 6d61 6c69 7a65 2720 6172  e 'normalize' ar
+0001d600: 6775 6d65 6e74 2074 616b 6573 2022 0a20  gument takes ". 
+0001d610: 2020 2020 2020 2020 2020 2022 6120 626f             "a bo
+0001d620: 6f6c 6561 6e20 7661 6c75 652e 5c6e 2229  olean value.\n")
+0001d630: 0a20 2020 2020 2020 2023 7570 6461 7465  .        #update
+0001d640: 2074 6865 2073 7065 6369 6573 2720 636f   the species' co
+0001d650: 6f72 6469 6e61 7465 7320 616e 6420 6365  ordinates and ce
+0001d660: 6c6c 732c 2069 6e20 6361 7365 2069 7420  lls, in case it 
+0001d670: 6861 736e 2774 0a20 2020 2020 2020 2023  hasn't.        #
+0001d680: 6265 656e 2075 7064 6174 6564 2073 696e  been updated sin
+0001d690: 6365 2073 6f6d 6520 696e 7465 726e 616c  ce some internal
+0001d6a0: 206f 7220 6d61 6e75 616c 2063 6861 6e67   or manual chang
+0001d6b0: 6573 2069 6e20 706f 7075 6c61 7469 6f6e  es in population
+0001d6c0: 2d73 697a 650a 2020 2020 2020 2020 2368  -size.        #h
+0001d6d0: 6176 6520 6f63 6375 7272 6564 0a20 2020  ave occurred.   
+0001d6e0: 2020 2020 2073 656c 662e 5f73 6574 5f63       self._set_c
+0001d6f0: 6f6f 7264 735f 616e 645f 6365 6c6c 7328  oords_and_cells(
+0001d700: 290a 2020 2020 2020 2020 6465 6e73 203d  ).        dens =
+0001d710: 2073 656c 662e 5f63 616c 635f 6465 6e73   self._calc_dens
+0001d720: 6974 7928 6e6f 726d 616c 697a 6520 3d20  ity(normalize = 
+0001d730: 6e6f 726d 616c 697a 6529 0a20 2020 2020  normalize).     
+0001d740: 2020 2070 6c74 5f6c 696d 7320 3d20 7669     plt_lims = vi
+0001d750: 7a2e 5f67 6574 5f70 6c74 5f6c 696d 7328  z._get_plt_lims(
+0001d760: 6c61 6e64 2c20 782c 2079 2c20 7a6f 6f6d  land, x, y, zoom
+0001d770: 5f77 6964 7468 290a 2020 2020 2020 2020  _width).        
+0001d780: 6966 206e 6f72 6d61 6c69 7a65 3a0a 2020  if normalize:.  
+0001d790: 2020 2020 2020 2020 2020 7669 7a2e 5f70            viz._p
+0001d7a0: 6c6f 745f 7261 7374 6572 7328 6465 6e73  lot_rasters(dens
+0001d7b0: 2c20 706c 745f 6c69 6d73 203d 2070 6c74  , plt_lims = plt
+0001d7c0: 5f6c 696d 732c 206c 7972 5f6e 616d 6520  _lims, lyr_name 
+0001d7d0: 3d20 2764 656e 7369 7479 272c 0a20 2020  = 'density',.   
+0001d7e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d7f0: 2020 2020 2020 2020 2020 2074 6963 6b73             ticks
+0001d800: 3d74 6963 6b73 2c20 6d61 736b 5f72 6173  =ticks, mask_ras
+0001d810: 743d 6d61 736b 5f72 6173 7429 0a20 2020  t=mask_rast).   
+0001d820: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
+0001d830: 2020 2020 2020 2076 697a 2e5f 706c 6f74         viz._plot
+0001d840: 5f72 6173 7465 7273 2864 656e 732c 2070  _rasters(dens, p
+0001d850: 6c74 5f6c 696d 7320 3d20 706c 745f 6c69  lt_lims = plt_li
+0001d860: 6d73 2c20 766d 6178 203d 2064 656e 732e  ms, vmax = dens.
+0001d870: 6d61 7828 292c 0a20 2020 2020 2020 2020  max(),.         
+0001d880: 2020 2020 2020 206c 7972 5f6e 616d 6520         lyr_name 
+0001d890: 3d20 2764 656e 7369 7479 272c 2074 6963  = 'density', tic
+0001d8a0: 6b73 3d74 6963 6b73 2c20 6d61 736b 5f72  ks=ticks, mask_r
+0001d8b0: 6173 743d 6d61 736b 5f72 6173 7429 0a20  ast=mask_rast). 
+0001d8c0: 2020 2020 2020 2073 656c 662e 5f70 6c6f         self._plo
+0001d8d0: 7428 6869 6465 5f6c 616e 643d 5472 7565  t(hide_land=True
+0001d8e0: 2c20 696e 6469 7669 6473 203d 2069 6e64  , individs = ind
+0001d8f0: 6976 6964 732c 2074 6578 7420 3d20 7465  ivids, text = te
+0001d900: 7874 2c0a 2020 2020 2020 2020 2020 2020  xt,.            
+0001d910: 636f 6c6f 723d 636f 6c6f 722c 2065 6467  color=color, edg
+0001d920: 655f 636f 6c6f 7220 3d20 6564 6765 5f63  e_color = edge_c
+0001d930: 6f6c 6f72 2c20 7465 7874 5f63 6f6c 6f72  olor, text_color
+0001d940: 203d 2074 6578 745f 636f 6c6f 722c 0a20   = text_color,. 
+0001d950: 2020 2020 2020 2020 2020 2073 697a 653d             size=
+0001d960: 7369 7a65 2c20 7465 7874 5f73 697a 6520  size, text_size 
+0001d970: 3d20 7465 7874 5f73 697a 652c 2061 6c70  = text_size, alp
+0001d980: 6861 3d61 6c70 6861 2c0a 2020 2020 2020  ha=alpha,.      
+0001d990: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001d9a0: 2020 2020 2020 2020 2020 2020 2020 7a6f                zo
+0001d9b0: 6f6d 5f77 6964 7468 203d 207a 6f6f 6d5f  om_width = zoom_
+0001d9c0: 7769 6474 682c 2078 203d 2078 2c20 7920  width, x = x, y 
+0001d9d0: 3d20 7929 0a0a 0a20 2020 2023 206d 6574  = y)...    # met
+0001d9e0: 686f 6420 666f 7220 706c 6f74 7469 6e67  hod for plotting
+0001d9f0: 2069 6e64 6976 6964 7561 6c73 2063 6f6c   individuals col
+0001da00: 6f72 6564 2062 7920 7468 6569 7220 6765  ored by their ge
+0001da10: 6e6f 7479 7065 2061 7420 6120 6c6f 6375  notype at a locu
+0001da20: 730a 2020 2020 6465 6620 5f70 6c6f 745f  s.    def _plot_
+0001da30: 6765 6e6f 7479 7065 2873 656c 662c 206c  genotype(self, l
+0001da40: 6f63 7573 2c20 6c79 725f 6e75 6d3d 4e6f  ocus, lyr_num=No
+0001da50: 6e65 2c20 6c61 6e64 3d4e 6f6e 652c 2069  ne, land=None, i
+0001da60: 6e64 6976 6964 733d 4e6f 6e65 2c0a 2020  ndivids=None,.  
+0001da70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001da80: 2020 2020 2074 6578 743d 4661 6c73 652c       text=False,
+0001da90: 2073 697a 653d 3235 2c20 7465 7874 5f73   size=25, text_s
+0001daa0: 697a 6520 3d20 392c 2065 6467 655f 636f  ize = 9, edge_co
+0001dab0: 6c6f 723d 2762 6c61 636b 272c 0a20 2020  lor='black',.   
+0001dac0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001dad0: 2020 2020 7465 7874 5f63 6f6c 6f72 3d27      text_color='
+0001dae0: 626c 6163 6b27 2c20 6362 6172 3d54 7275  black', cbar=Tru
+0001daf0: 652c 2061 6c70 6861 3d31 2c0a 2020 2020  e, alpha=1,.    
+0001db00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001db10: 2020 2062 795f 646f 6d69 6e61 6e63 653d     by_dominance=
+0001db20: 4661 6c73 652c 207a 6f6f 6d5f 7769 6474  False, zoom_widt
+0001db30: 683d 4e6f 6e65 2c20 783d 4e6f 6e65 2c20  h=None, x=None, 
+0001db40: 793d 4e6f 6e65 2c0a 2020 2020 2020 2020  y=None,.        
+0001db50: 2020 2020 2020 2020 2020 2020 2020 2074                 t
+0001db60: 6963 6b73 3d4e 6f6e 652c 206d 6173 6b5f  icks=None, mask_
+0001db70: 7261 7374 3d4e 6f6e 6529 3a0a 0a20 2020  rast=None):..   
+0001db80: 2020 2020 2067 656e 6f74 7970 6573 203d       genotypes =
+0001db90: 2073 656c 662e 5f67 6574 5f67 656e 6f74   self._get_genot
+0001dba0: 7970 6573 286c 6f63 693d 5b6c 6f63 7573  ypes(loci=[locus
+0001dbb0: 5d2c 2069 6e64 6976 6964 733d 696e 6469  ], individs=indi
+0001dbc0: 7669 6473 2c0a 2020 2020 2020 2020 2020  vids,.          
+0001dbd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001dbe0: 2020 2020 2020 2020 2020 2020 2020 6269                bi
+0001dbf0: 616c 6c65 6c69 633d 4661 6c73 652c 2061  allelic=False, a
+0001dc00: 735f 6469 6374 3d54 7275 6529 0a20 2020  s_dict=True).   
+0001dc10: 2020 2020 2069 6620 6279 5f64 6f6d 696e       if by_domin
+0001dc20: 616e 6365 203d 3d20 5472 7565 3a0a 2020  ance == True:.  
+0001dc30: 2020 2020 2020 2020 2020 6966 206c 6f63            if loc
+0001dc40: 7573 2069 6e20 7365 6c66 2e67 656e 5f61  us in self.gen_a
+0001dc50: 7263 682e 6e6f 6e6e 6575 745f 6c6f 6369  rch.nonneut_loci
+0001dc60: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+0001dc70: 2020 646f 6d20 3d20 7365 6c66 2e67 656e    dom = self.gen
+0001dc80: 5f61 7263 682e 646f 6d5b 6c6f 6375 735d  _arch.dom[locus]
+0001dc90: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0001dca0: 2023 204e 4f54 453a 2077 6f75 6c64 2062   # NOTE: would b
+0001dcb0: 6520 7369 6d70 6c65 7220 746f 206a 7573  e simpler to jus
+0001dcc0: 7420 7573 6520 6e70 2e63 6569 6c20 6865  t use np.ceil he
+0001dcd0: 7265 2e2e 2e0a 2020 2020 2020 2020 2020  re....          
+0001dce0: 2020 2020 2020 2320 2020 2020 2020 646f        #       do
+0001dcf0: 6e27 7420 7365 6520 7768 7920 4920 6469  n't see why I di
+0001dd00: 646e 2774 2e2e 2e3f 0a20 2020 2020 2020  dn't...?.       
+0001dd10: 2020 2020 2020 2020 2067 656e 6f74 7970           genotyp
+0001dd20: 6573 203d 207b 693a 6e70 2e63 6c69 7028  es = {i:np.clip(
+0001dd30: 6774 202a 2028 3120 2b20 646f 6d29 2c0a  gt * (1 + dom),.
+0001dd40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001dd50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001dd60: 2020 2020 2020 2061 5f6d 696e 3d4e 6f6e         a_min=Non
+0001dd70: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+0001dd80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001dd90: 2020 2020 2020 2020 2020 615f 6d61 783d            a_max=
+0001dda0: 3129 2066 6f72 2069 2c20 6774 2069 6e20  1) for i, gt in 
+0001ddb0: 6765 6e6f 7479 7065 732e 6974 656d 7328  genotypes.items(
+0001ddc0: 297d 0a20 2020 2020 2020 2020 2020 2065  )}.            e
+0001ddd0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
+0001dde0: 2020 2020 2070 7269 6e74 2828 275c 6e5c       print(('\n\
+0001ddf0: 7457 4152 4e49 4e47 3a20 5468 6520 6279  tWARNING: The by
+0001de00: 5f64 6f6d 696e 616e 6365 2061 7267 756d  _dominance argum
+0001de10: 656e 7420 6973 2054 7275 652c 2027 0a20  ent is True, '. 
+0001de20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001de30: 2020 2020 2020 2762 7574 2061 206e 6575        'but a neu
+0001de40: 7472 616c 206c 6f63 7573 2077 6173 2063  tral locus was c
+0001de50: 686f 7365 6e2c 2027 0a20 2020 2020 2020  hosen, '.       
+0001de60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001de70: 2773 6f20 7468 6520 6172 6775 6d65 6e74  'so the argument
+0001de80: 2077 6173 206e 6f74 2075 7365 642e 5c6e   was not used.\n
+0001de90: 2729 290a 0a20 2020 2020 2020 2023 206a  '))..        # j
+0001dea0: 7573 7420 6173 7369 676e 2062 6c61 636b  ust assign black
+0001deb0: 2c20 6772 6179 2c20 616e 6420 7768 6974  , gray, and whit
+0001dec0: 6520 2873 696e 6365 2074 6865 7265 2773  e (since there's
+0001ded0: 206e 6f20 7265 6173 6f6e 0a20 2020 2020   no reason.     
+0001dee0: 2020 2023 206e 6563 6573 7361 7269 6c79     # necessarily
+0001def0: 2074 6861 7420 7468 6573 6520 7368 6f75   that these shou
+0001df00: 6c64 2062 6520 6d61 7070 6564 2074 6f20  ld be mapped to 
+0001df10: 6120 6365 7274 6169 6e20 6c61 7965 722c  a certain layer,
+0001df20: 2074 6865 2077 6179 0a20 2020 2020 2020   the way.       
+0001df30: 2023 2070 6865 6e6f 7479 7065 2073 686f   # phenotype sho
+0001df40: 756c 640a 2020 2020 2020 2020 636f 6c6f  uld.        colo
+0001df50: 7273 203d 205b 2723 3030 3030 3030 272c  rs = ['#000000',
+0001df60: 2027 2338 3038 3038 3027 2c20 2723 4646   '#808080', '#FF
+0001df70: 4646 4646 275d 0a0a 2020 2020 2020 2020  FFFF']..        
+0001df80: 666f 7220 6e2c 2067 656e 6f74 7970 6520  for n, genotype 
+0001df90: 696e 2065 6e75 6d65 7261 7465 285b 302e  in enumerate([0.
+0001dfa0: 302c 2030 2e35 2c20 312e 305d 293a 0a20  0, 0.5, 1.0]):. 
+0001dfb0: 2020 2020 2020 2020 2020 2067 656e 6f74             genot
+0001dfc0: 7970 655f 696e 6469 7669 6473 203d 205b  ype_individs = [
+0001dfd0: 6920 666f 7220 692c 2067 2069 6e20 6765  i for i, g in ge
+0001dfe0: 6e6f 7479 7065 732e 6974 656d 7328 0a20  notypes.items(. 
+0001dff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e000: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e010: 2020 2020 2020 2029 2069 6620 6e70 2e61         ) if np.a
+0001e020: 746c 6561 7374 5f31 6428 6729 5b30 5d20  tleast_1d(g)[0] 
+0001e030: 3d3d 2067 656e 6f74 7970 655d 0a20 2020  == genotype].   
+0001e040: 2020 2020 2020 2020 2023 2077 696c 6c20           # will 
+0001e050: 6869 6465 2074 6865 206c 616e 6420 6966  hide the land if
+0001e060: 2074 6869 7320 6973 206e 6f74 2074 6865   this is not the
+0001e070: 2066 6972 7374 2070 6c6f 7420 6d61 6465   first plot made
+0001e080: 0a20 2020 2020 2020 2020 2020 2068 6964  .            hid
+0001e090: 655f 6c61 6e64 203d 206e 203e 2030 0a20  e_land = n > 0. 
+0001e0a0: 2020 2020 2020 2020 2020 2023 2070 6c6f             # plo
+0001e0b0: 7420 6966 2074 6865 7265 2061 7265 2061  t if there are a
+0001e0c0: 6e79 2069 6e64 6976 6964 7561 6c73 206f  ny individuals o
+0001e0d0: 6620 7468 6973 2067 656e 6f74 7970 650a  f this genotype.
+0001e0e0: 2020 2020 2020 2020 2020 2020 6966 206c              if l
+0001e0f0: 656e 2867 656e 6f74 7970 655f 696e 6469  en(genotype_indi
+0001e100: 7669 6473 2920 3e3d 2031 3a0a 2020 2020  vids) >= 1:.    
+0001e110: 2020 2020 2020 2020 2020 2020 7365 6c66              self
+0001e120: 2e5f 706c 6f74 286c 7972 5f6e 756d 3d6c  ._plot(lyr_num=l
+0001e130: 7972 5f6e 756d 2c20 6c61 6e64 3d6c 616e  yr_num, land=lan
+0001e140: 642c 2068 6964 655f 6c61 6e64 3d68 6964  d, hide_land=hid
+0001e150: 655f 6c61 6e64 2c0a 2020 2020 2020 2020  e_land,.        
+0001e160: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e170: 2020 2069 6e64 6976 6964 733d 6765 6e6f     individs=geno
+0001e180: 7479 7065 5f69 6e64 6976 6964 732c 0a20  type_individs,. 
+0001e190: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e1a0: 2020 2020 2020 2020 2020 7465 7874 3d74            text=t
+0001e1b0: 6578 742c 2063 6f6c 6f72 3d63 6f6c 6f72  ext, color=color
+0001e1c0: 735b 6e5d 2c20 6564 6765 5f63 6f6c 6f72  s[n], edge_color
+0001e1d0: 3d65 6467 655f 636f 6c6f 722c 0a20 2020  =edge_color,.   
+0001e1e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e1f0: 2020 2020 2020 2020 7465 7874 5f63 6f6c          text_col
+0001e200: 6f72 3d74 6578 745f 636f 6c6f 722c 2063  or=text_color, c
+0001e210: 6261 723d 6362 6172 2c0a 2020 2020 2020  bar=cbar,.      
+0001e220: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e230: 2020 2020 2073 697a 653d 7369 7a65 2c20       size=size, 
+0001e240: 7465 7874 5f73 697a 653d 7465 7874 5f73  text_size=text_s
+0001e250: 697a 652c 2061 6c70 6861 3d61 6c70 6861  ize, alpha=alpha
+0001e260: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001e270: 2020 2020 2020 2020 2020 2020 207a 6f6f               zoo
+0001e280: 6d5f 7769 6474 683d 7a6f 6f6d 5f77 6964  m_width=zoom_wid
+0001e290: 7468 2c20 783d 782c 2079 3d79 2c20 766d  th, x=x, y=y, vm
+0001e2a0: 696e 3d30 2c20 766d 6178 3d31 2c0a 2020  in=0, vmax=1,.  
+0001e2b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e2c0: 2020 2020 2020 2020 2074 6963 6b73 3d74           ticks=t
+0001e2d0: 6963 6b73 2c20 6d61 736b 5f72 6173 743d  icks, mask_rast=
+0001e2e0: 6d61 736b 5f72 6173 7429 0a0a 0a20 2020  mask_rast)...   
+0001e2f0: 2023 206d 6574 686f 6420 666f 7220 706c   # method for pl
+0001e300: 6f74 7469 6e67 2069 6e64 6976 6964 7561  otting individua
+0001e310: 6c73 2063 6f6c 6f72 6564 2062 7920 7468  ls colored by th
+0001e320: 6569 7220 7068 656e 6f74 7970 6573 0a20  eir phenotypes. 
+0001e330: 2020 2023 666f 7220 6120 6769 7665 6e20     #for a given 
+0001e340: 7472 6169 740a 2020 2020 6465 6620 5f70  trait.    def _p
+0001e350: 6c6f 745f 7068 656e 6f74 7970 6528 7365  lot_phenotype(se
+0001e360: 6c66 2c20 7472 6169 742c 206c 7972 5f6e  lf, trait, lyr_n
+0001e370: 756d 3d4e 6f6e 652c 206c 616e 643d 4e6f  um=None, land=No
+0001e380: 6e65 2c0a 2020 2020 2020 2020 2020 2020  ne,.            
+0001e390: 2020 2020 2020 2020 2020 2020 696e 6469              indi
+0001e3a0: 7669 6473 3d4e 6f6e 652c 2074 6578 743d  vids=None, text=
+0001e3b0: 4661 6c73 652c 2073 697a 653d 3235 2c20  False, size=25, 
+0001e3c0: 7465 7874 5f73 697a 653d 392c 0a20 2020  text_size=9,.   
+0001e3d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e3e0: 2020 2020 2065 6467 655f 636f 6c6f 723d       edge_color=
+0001e3f0: 2762 6c61 636b 272c 2074 6578 745f 636f  'black', text_co
+0001e400: 6c6f 723d 2762 6c61 636b 272c 2063 6261  lor='black', cba
+0001e410: 723d 5472 7565 2c0a 2020 2020 2020 2020  r=True,.        
+0001e420: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e430: 616c 7068 613d 312c 207a 6f6f 6d5f 7769  alpha=1, zoom_wi
+0001e440: 6474 683d 4e6f 6e65 2c20 783d 4e6f 6e65  dth=None, x=None
+0001e450: 2c20 793d 4e6f 6e65 2c0a 2020 2020 2020  , y=None,.      
+0001e460: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e470: 2020 7469 636b 733d 4e6f 6e65 2c20 6d61    ticks=None, ma
+0001e480: 736b 5f72 6173 743d 4e6f 6e65 2c20 616e  sk_rast=None, an
+0001e490: 696d 6174 653d 4661 6c73 6529 3a0a 0a20  imate=False):.. 
+0001e4a0: 2020 2020 2020 2023 2067 6574 2074 6865         # get the
+0001e4b0: 2074 7261 6974 2773 206c 7972 5f6e 756d   trait's lyr_num
+0001e4c0: 2c20 6966 206e 6f20 6c79 725f 6e75 6d20  , if no lyr_num 
+0001e4d0: 7072 6f76 6964 6564 0a20 2020 2020 2020  provided.       
+0001e4e0: 206c 7972 5f6e 756d 203d 2073 656c 662e   lyr_num = self.
+0001e4f0: 6765 6e5f 6172 6368 2e74 7261 6974 735b  gen_arch.traits[
+0001e500: 7472 6169 745d 2e6c 7972 5f6e 756d 0a0a  trait].lyr_num..
+0001e510: 2020 2020 2020 2020 7a20 3d20 4f44 287a          z = OD(z
+0001e520: 6970 285b 2a73 656c 665d 2c20 7365 6c66  ip([*self], self
+0001e530: 2e5f 6765 745f 7a28 295b 3a2c 2074 7261  ._get_z()[:, tra
+0001e540: 6974 5d29 290a 2020 2020 2020 2020 6966  it])).        if
+0001e550: 2069 6e64 6976 6964 7320 6973 206e 6f74   individs is not
+0001e560: 204e 6f6e 653a 0a20 2020 2020 2020 2020   None:.         
+0001e570: 2020 207a 203d 207b 693a 2076 2066 6f72     z = {i: v for
+0001e580: 2069 2c20 7620 696e 207a 2e69 7465 6d73   i, v in z.items
+0001e590: 2829 2069 6620 6920 696e 2069 6e64 6976  () if i in indiv
+0001e5a0: 6964 737d 0a0a 2020 2020 2020 2020 2320  ids}..        # 
+0001e5b0: 6765 7420 7468 6520 636f 7272 6563 7420  get the correct 
+0001e5c0: 636d 6170 2066 6f72 2074 6869 7320 7472  cmap for this tr
+0001e5d0: 6169 7427 7320 6c61 7965 720a 2020 2020  ait's layer.    
+0001e5e0: 2020 2020 7074 5f63 6d61 7020 3d20 7669      pt_cmap = vi
+0001e5f0: 7a2e 5f63 686f 6f73 655f 636d 6170 2873  z._choose_cmap(s
+0001e600: 656c 662e 6765 6e5f 6172 6368 2e74 7261  elf.gen_arch.tra
+0001e610: 6974 735b 7472 6169 745d 2e6c 7972 5f6e  its[trait].lyr_n
+0001e620: 756d 290a 0a20 2020 2020 2020 2070 6f69  um)..        poi
+0001e630: 6e74 7320 3d20 7365 6c66 2e5f 706c 6f74  nts = self._plot
+0001e640: 286c 7972 5f6e 756d 203d 206c 7972 5f6e  (lyr_num = lyr_n
+0001e650: 756d 2c20 6c61 6e64 203d 206c 616e 642c  um, land = land,
+0001e660: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0001e670: 2020 2020 2020 2020 2020 2020 2069 6e64               ind
+0001e680: 6976 6964 7320 3d20 696e 6469 7669 6473  ivids = individs
+0001e690: 2c20 7465 7874 203d 2074 6578 742c 0a20  , text = text,. 
+0001e6a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e6b0: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
+0001e6c0: 203d 206c 6973 7428 7a2e 7661 6c75 6573   = list(z.values
+0001e6d0: 2829 292c 2070 745f 636d 6170 203d 2070  ()), pt_cmap = p
+0001e6e0: 745f 636d 6170 2c0a 2020 2020 2020 2020  t_cmap,.        
+0001e6f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e700: 2020 2020 6564 6765 5f63 6f6c 6f72 203d      edge_color =
+0001e710: 2065 6467 655f 636f 6c6f 722c 2074 6578   edge_color, tex
+0001e720: 745f 636f 6c6f 7220 3d20 7465 7874 5f63  t_color = text_c
+0001e730: 6f6c 6f72 2c0a 2020 2020 2020 2020 2020  olor,.          
+0001e740: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e750: 2020 6362 6172 203d 2063 6261 722c 2073    cbar = cbar, s
+0001e760: 697a 6520 3d20 7369 7a65 2c20 7465 7874  ize = size, text
+0001e770: 5f73 697a 6520 3d20 7465 7874 5f73 697a  _size = text_siz
+0001e780: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+0001e790: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+0001e7a0: 6c70 6861 203d 2061 6c70 6861 2c20 7a6f  lpha = alpha, zo
+0001e7b0: 6f6d 5f77 6964 7468 203d 207a 6f6f 6d5f  om_width = zoom_
+0001e7c0: 7769 6474 682c 2078 203d 2078 2c0a 2020  width, x = x,.  
+0001e7d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e7e0: 2020 2020 2020 2020 2020 7920 3d20 792c            y = y,
+0001e7f0: 2076 6d69 6e20 3d20 302c 2076 6d61 7820   vmin = 0, vmax 
+0001e800: 3d20 312c 2074 6963 6b73 3d74 6963 6b73  = 1, ticks=ticks
+0001e810: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001e820: 2020 2020 2020 2020 2020 2020 2020 6d61                ma
+0001e830: 736b 5f72 6173 743d 6d61 736b 5f72 6173  sk_rast=mask_ras
+0001e840: 742c 2061 6e69 6d61 7465 3d61 6e69 6d61  t, animate=anima
+0001e850: 7465 290a 0a20 2020 2020 2020 2072 6574  te)..        ret
+0001e860: 7572 6e20 706f 696e 7473 0a0a 0a20 2020  urn points...   
+0001e870: 2023 206d 6574 686f 6420 666f 7220 706c   # method for pl
+0001e880: 6f74 7469 6e67 2069 6e64 6976 6964 7561  otting individua
+0001e890: 6c73 2063 6f6c 6f72 6564 2062 7920 7468  ls colored by th
+0001e8a0: 6569 7220 6f76 6572 616c 6c20 6669 746e  eir overall fitn
+0001e8b0: 6573 7365 732c 0a20 2020 2023 6f72 2062  esses,.    #or b
+0001e8c0: 7920 7468 6569 7220 6669 746e 6573 7365  y their fitnesse
+0001e8d0: 7320 666f 7220 6120 7369 6e67 6c65 2074  s for a single t
+0001e8e0: 7261 6974 2028 6966 2074 7261 6974 2069  rait (if trait i
+0001e8f0: 7320 6e6f 7420 4e6f 6e65 290a 2020 2020  s not None).    
+0001e900: 6465 6620 5f70 6c6f 745f 6669 746e 6573  def _plot_fitnes
+0001e910: 7328 7365 6c66 2c20 7472 745f 6e75 6d3d  s(self, trt_num=
+0001e920: 4e6f 6e65 2c20 6c79 725f 6e75 6d3d 4e6f  None, lyr_num=No
+0001e930: 6e65 2c20 6c61 6e64 3d4e 6f6e 652c 0a20  ne, land=None,. 
+0001e940: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e950: 2020 2020 2069 6e64 6976 6964 733d 4e6f       individs=No
+0001e960: 6e65 2c20 7465 7874 3d46 616c 7365 2c20  ne, text=False, 
+0001e970: 7068 656e 6f74 7970 655f 7465 7874 3d46  phenotype_text=F
+0001e980: 616c 7365 2c0a 2020 2020 2020 2020 2020  alse,.          
+0001e990: 2020 2020 2020 2020 2020 2020 7068 656e              phen
+0001e9a0: 6f74 7970 655f 7465 7874 5f63 6f6c 6f72  otype_text_color
+0001e9b0: 3d27 626c 6163 6b27 2c20 6669 746e 6573  ='black', fitnes
+0001e9c0: 735f 7465 7874 3d46 616c 7365 2c0a 2020  s_text=False,.  
+0001e9d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001e9e0: 2020 2020 6669 746e 6573 735f 7465 7874      fitness_text
+0001e9f0: 5f63 6f6c 6f72 3d27 2333 3333 3333 3327  _color='#333333'
+0001ea00: 2c20 7369 7a65 3d31 3030 2c20 7465 7874  , size=100, text
+0001ea10: 5f73 697a 6520 3d20 392c 0a20 2020 2020  _size = 9,.     
+0001ea20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ea30: 2065 6467 655f 636f 6c6f 723d 2762 6c61   edge_color='bla
+0001ea40: 636b 272c 2074 6578 745f 636f 6c6f 723d  ck', text_color=
+0001ea50: 2762 6c61 636b 272c 0a20 2020 2020 2020  'black',.       
+0001ea60: 2020 2020 2020 2020 2020 2020 2020 2066                 f
+0001ea70: 6974 5f63 6d61 7020 3d20 2752 6459 6c47  it_cmap = 'RdYlG
+0001ea80: 6e27 2c20 6362 6172 3d54 7275 652c 2066  n', cbar=True, f
+0001ea90: 6974 6e65 7373 5f63 6261 723d 5472 7565  itness_cbar=True
+0001eaa0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001eab0: 2020 2020 2020 2020 616c 7068 613d 312c          alpha=1,
+0001eac0: 207a 6f6f 6d5f 7769 6474 683d 4e6f 6e65   zoom_width=None
+0001ead0: 2c20 783d 4e6f 6e65 2c20 793d 4e6f 6e65  , x=None, y=None
+0001eae0: 2c20 7469 636b 733d 4e6f 6e65 2c0a 2020  , ticks=None,.  
+0001eaf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001eb00: 2020 2020 6d61 736b 5f72 6173 743d 4e6f      mask_rast=No
+0001eb10: 6e65 293a 0a0a 2020 2020 2020 2020 2372  ne):..        #r
+0001eb20: 6574 7572 6e20 6d65 7373 6167 6573 2069  eturn messages i
+0001eb30: 6620 7370 6563 6965 7320 646f 6573 206e  f species does n
+0001eb40: 6f74 2068 6176 6520 6765 6e6f 6d65 7320  ot have genomes 
+0001eb50: 6f72 2074 7261 6974 730a 2020 2020 2020  or traits.      
+0001eb60: 2020 6966 2073 656c 662e 6765 6e5f 6172    if self.gen_ar
+0001eb70: 6368 2069 7320 4e6f 6e65 3a0a 2020 2020  ch is None:.    
+0001eb80: 2020 2020 2020 2020 7072 696e 7428 2822          print(("
+0001eb90: 5370 6563 6965 732e 5f70 6c6f 745f 6669  Species._plot_fi
+0001eba0: 746e 6573 7320 6973 206e 6f74 2076 616c  tness is not val
+0001ebb0: 6964 2066 6f72 2073 7065 6369 6573 2022  id for species "
+0001ebc0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0001ebd0: 2020 2020 2277 6974 686f 7574 2067 656e      "without gen
+0001ebe0: 6f6d 6573 2e5c 6e22 2929 0a20 2020 2020  omes.\n")).     
+0001ebf0: 2020 2020 2020 2072 6574 7572 6e0a 2020         return.  
+0001ec00: 2020 2020 2020 656c 6966 2073 656c 662e        elif self.
+0001ec10: 6765 6e5f 6172 6368 2e74 7261 6974 7320  gen_arch.traits 
+0001ec20: 6973 204e 6f6e 653a 0a20 2020 2020 2020  is None:.       
+0001ec30: 2020 2020 2070 7269 6e74 2828 2253 7065       print(("Spe
+0001ec40: 6369 6573 2e5f 706c 6f74 5f66 6974 6e65  cies._plot_fitne
+0001ec50: 7373 2069 7320 6e6f 7420 7661 6c69 6420  ss is not valid 
+0001ec60: 666f 7220 7370 6563 6965 7320 220a 2020  for species ".  
+0001ec70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ec80: 2022 7769 7468 6f75 7420 7472 6169 7473   "without traits
+0001ec90: 2e5c 6e22 2929 0a20 2020 2020 2020 2020  .\n")).         
+0001eca0: 2020 2072 6574 7572 6e0a 0a20 2020 2020     return..     
+0001ecb0: 2020 2023 2067 6574 2074 6865 2074 7261     # get the tra
+0001ecc0: 6974 2773 206c 7972 5f6e 756d 2c20 6966  it's lyr_num, if
+0001ecd0: 206c 7972 5f6e 756d 2077 6173 6e27 7420   lyr_num wasn't 
+0001ece0: 7072 6f76 6964 6564 2062 7574 2074 7274  provided but trt
+0001ecf0: 5f6e 756d 2077 6173 0a20 2020 2020 2020  _num was.       
+0001ed00: 2069 6620 7472 745f 6e75 6d20 6973 206e   if trt_num is n
+0001ed10: 6f74 204e 6f6e 6520 616e 6420 6c79 725f  ot None and lyr_
+0001ed20: 6e75 6d20 6973 204e 6f6e 653a 0a20 2020  num is None:.   
+0001ed30: 2020 2020 2020 2020 206c 7972 5f6e 756d           lyr_num
+0001ed40: 203d 2073 656c 662e 6765 6e5f 6172 6368   = self.gen_arch
+0001ed50: 2e74 7261 6974 735b 7472 745f 6e75 6d5d  .traits[trt_num]
+0001ed60: 2e6c 7972 5f6e 756d 0a0a 2020 2020 2020  .lyr_num..      
+0001ed70: 2020 2320 6765 7420 616c 6c20 696e 6469    # get all indi
+0001ed80: 7669 6473 2720 6669 746e 6573 7320 7661  vids' fitness va
+0001ed90: 6c75 6573 2c0a 2020 2020 2020 2020 2320  lues,.        # 
+0001eda0: 616e 6420 6765 7420 6170 7072 6f70 7269  and get appropri
+0001edb0: 6174 6520 636f 6c6f 726d 6170 0a20 2020  ate colormap.   
+0001edc0: 2020 2020 2069 6620 7472 745f 6e75 6d20       if trt_num 
+0001edd0: 6973 204e 6f6e 653a 0a20 2020 2020 2020  is None:.       
+0001ede0: 2020 2020 2077 203d 2073 656c 662e 5f63       w = self._c
+0001edf0: 616c 635f 6669 746e 6573 7328 290a 2020  alc_fitness().  
+0001ee00: 2020 2020 2020 2020 2020 7074 5f63 6d61            pt_cma
+0001ee10: 7020 3d20 2747 7265 7973 5f72 270a 2020  p = 'Greys_r'.  
+0001ee20: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
+0001ee30: 2020 2020 2020 2020 7720 3d20 7365 6c66          w = self
+0001ee40: 2e5f 6361 6c63 5f66 6974 6e65 7373 2874  ._calc_fitness(t
+0001ee50: 7261 6974 5f6e 756d 203d 2074 7274 5f6e  rait_num = trt_n
+0001ee60: 756d 290a 0a20 2020 2020 2020 2023 6669  um)..        #fi
+0001ee70: 6c74 6572 206f 7574 2075 6e77 616e 7465  lter out unwante
+0001ee80: 6420 696e 6469 7669 6473 2c20 6966 206e  d individs, if n
+0001ee90: 6563 6573 7361 7279 0a20 2020 2020 2020  ecessary.       
+0001eea0: 2077 203d 204f 4428 7a69 7028 5b2a 7365   w = OD(zip([*se
+0001eeb0: 6c66 5d2c 2077 2929 0a20 2020 2020 2020  lf], w)).       
+0001eec0: 2069 6620 696e 6469 7669 6473 2069 7320   if individs is 
+0001eed0: 6e6f 7420 4e6f 6e65 3a0a 2020 2020 2020  not None:.      
+0001eee0: 2020 2020 2020 7720 3d20 7b69 3a76 2066        w = {i:v f
+0001eef0: 6f72 2069 2c76 2069 6e20 772e 6974 656d  or i,v in w.item
+0001ef00: 7328 2920 6966 2069 2069 6e20 696e 6469  s() if i in indi
+0001ef10: 7669 6473 7d0a 0a20 2020 2020 2020 2023  vids}..        #
+0001ef20: 2063 616c 6320 6d69 6e69 6d75 6d20 706f   calc minimum po
+0001ef30: 7373 6962 6c65 2066 6974 6e65 7373 2028  ssible fitness (
+0001ef40: 666f 7220 7068 656e 6f74 7970 6573 2077  for phenotypes w
+0001ef50: 6974 6869 6e20 3020 3c3d 207a 203c 3d20  ithin 0 <= z <= 
+0001ef60: 312c 0a20 2020 2020 2020 2023 7768 6963  1,.        #whic
+0001ef70: 6820 696e 2072 6561 6c69 7479 2069 736e  h in reality isn
+0001ef80: 2774 2061 2063 6f6e 7374 7261 696e 742c  't a constraint,
+0001ef90: 2062 7574 2076 616c 7565 7320 6c6f 7765   but values lowe
+0001efa0: 7220 7468 616e 0a20 2020 2020 2020 2023  r than.        #
+0001efb0: 7468 6973 2077 696c 6c20 616c 736f 2062  this will also b
+0001efc0: 6520 636f 6e73 7472 6169 6e65 6420 746f  e constrained to
+0001efd0: 2074 6865 206d 696e 696d 756d 2d76 616c   the minimum-val
+0001efe0: 7565 2063 6f6c 6f72 2066 6f72 2070 6c6f  ue color for plo
+0001eff0: 7474 696e 6729 0a20 2020 2020 2020 2023  tting).        #
+0001f000: 4e4f 5445 3a20 7468 6520 6e70 2e61 746c  NOTE: the np.atl
+0001f010: 6561 7374 5f32 6428 2e2e 2e29 2e6d 696e  east_2d(...).min
+0001f020: 2829 2063 6f6e 7374 7275 6374 206d 616b  () construct mak
+0001f030: 6573 2074 6869 730a 2020 2020 2020 2020  es this.        
+0001f040: 2377 6f72 6b20 626f 7468 2066 6f72 2066  #work both for f
+0001f050: 6978 6564 2061 6e64 2073 7061 7469 616c  ixed and spatial
+0001f060: 6c79 2076 6172 7969 6e67 2070 6869 0a20  ly varying phi. 
+0001f070: 2020 2020 2020 2069 6620 7472 745f 6e75         if trt_nu
+0001f080: 6d20 6973 204e 6f6e 653a 0a20 2020 2020  m is None:.     
+0001f090: 2020 2020 2020 206d 696e 5f66 6974 203d         min_fit =
+0001f0a0: 206e 702e 7072 6f64 7563 7428 5b31 202d   np.product([1 -
+0001f0b0: 206e 702e 6174 6c65 6173 745f 3264 2874   np.atleast_2d(t
+0001f0c0: 2e70 6869 292e 6d69 6e28 0a20 2020 2020  .phi).min(.     
+0001f0d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f0e0: 2020 2020 2020 2029 2066 6f72 2074 2069         ) for t i
+0001f0f0: 6e20 6c69 7374 2873 656c 662e 6765 6e5f  n list(self.gen_
+0001f100: 6172 6368 2e74 7261 6974 732e 7661 6c75  arch.traits.valu
+0001f110: 6573 2829 295d 290a 2020 2020 2020 2020  es())]).        
+0001f120: 656c 7365 3a0a 2020 2020 2020 2020 2020  else:.          
+0001f130: 2020 6d69 6e5f 6669 7420 3d20 3120 2d20    min_fit = 1 - 
+0001f140: 6e70 2e61 746c 6561 7374 5f32 6428 7365  np.atleast_2d(se
+0001f150: 6c66 2e67 656e 5f61 7263 682e 7472 6169  lf.gen_arch.trai
+0001f160: 7473 5b0a 2020 2020 2020 2020 2020 2020  ts[.            
+0001f170: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f180: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f190: 2020 2020 2020 2020 2020 2020 7472 745f              trt_
+0001f1a0: 6e75 6d5d 2e70 6869 292e 6d69 6e28 290a  num].phi).min().
+0001f1b0: 0a20 2020 2020 2020 2023 7468 656e 2067  .        #then g
+0001f1c0: 6574 2075 6e65 7665 6e20 636d 6170 2061  et uneven cmap a
+0001f1d0: 6e64 2063 6261 722d 6d61 6b65 7220 286e  nd cbar-maker (n
+0001f1e0: 6565 6473 2074 6f20 6265 2075 6e65 7665  eeds to be uneve
+0001f1f0: 6e20 746f 0a20 2020 2020 2020 2023 6769  n to.        #gi
+0001f200: 7665 2063 6f6c 6f72 2d72 6573 6f6c 7574  ve color-resolut
+0001f210: 696f 6e20 746f 2076 616c 7565 7320 7661  ion to values va
+0001f220: 7279 696e 670a 2020 2020 2020 2020 2362  rying.        #b
+0001f230: 6574 7765 656e 2031 2061 6e64 2074 6865  etween 1 and the
+0001f240: 206d 696e 696d 756d 2d66 6974 6e65 7373   minimum-fitness
+0001f250: 2076 616c 7565 2061 7373 756d 696e 6720   value assuming 
+0001f260: 616c 6c0a 2020 2020 2020 2020 2370 6865  all.        #phe
+0001f270: 6e6f 7479 7065 7320 6172 6520 636f 6e73  notypes are cons
+0001f280: 7472 6169 6e65 6420 303c 3d7a 3c3d 312c  trained 0<=z<=1,
+0001f290: 2062 7574 2074 6865 6e20 616c 736f 0a20   but then also. 
+0001f2a0: 2020 2020 2020 2023 616c 6c6f 7720 6772         #allow gr
+0001f2b0: 6164 7561 6c6c 7920 6465 6570 656e 696e  adually deepenin
+0001f2c0: 6720 7265 6473 2066 6f72 2066 6974 6e65  g reds for fitne
+0001f2d0: 7373 2076 616c 7565 7320 6c6f 7765 720a  ss values lower.
+0001f2e0: 2020 2020 2020 2020 2374 6861 6e20 7468          #than th
+0001f2f0: 6174 206d 696e 696d 756d 2076 616c 7565  at minimum value
+0001f300: 292c 2075 7369 6e67 2074 6865 206d 696e  ), using the min
+0001f310: 5f66 6974 2076 616c 0a20 2020 2020 2020  _fit val.       
+0001f320: 2063 6d61 702c 206d 616b 655f 6362 6172   cmap, make_cbar
+0001f330: 5f66 6e20 3d20 7669 7a2e 5f6d 616b 655f  _fn = viz._make_
+0001f340: 6669 746e 6573 735f 636d 6170 5f61 6e64  fitness_cmap_and
+0001f350: 5f63 6261 725f 6d61 6b65 7228 0a20 2020  _cbar_maker(.   
+0001f360: 2020 2020 2020 2020 206d 696e 5f76 616c           min_val
+0001f370: 203d 206d 696e 5f66 6974 2c20 6d61 785f   = min_fit, max_
+0001f380: 7661 6c20 3d20 312c 2063 6d61 7020 3d20  val = 1, cmap = 
+0001f390: 6669 745f 636d 6170 2c0a 2020 2020 2020  fit_cmap,.      
+0001f3a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f3b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f3c0: 2020 2020 2020 2020 2020 7472 745f 6e75            trt_nu
+0001f3d0: 6d20 3d20 7472 745f 6e75 6d29 0a0a 2020  m = trt_num)..  
+0001f3e0: 2020 2020 2020 2370 6c6f 7420 7468 6520        #plot the 
+0001f3f0: 7472 6169 7420 7068 656e 6f74 7970 6520  trait phenotype 
+0001f400: 696e 206c 6172 6765 7220 6369 7263 6c65  in larger circle
+0001f410: 7320 6669 7273 742c 2069 6620 7472 6169  s first, if trai
+0001f420: 7420 6973 206e 6f74 204e 6f6e 650a 2020  t is not None.  
+0001f430: 2020 2020 2020 6966 2074 7274 5f6e 756d        if trt_num
+0001f440: 2069 7320 6e6f 7420 4e6f 6e65 3a0a 2020   is not None:.  
+0001f450: 2020 2020 2020 2020 2020 2370 6c6f 7420            #plot 
+0001f460: 7468 6520 6f75 7465 7220 2870 6865 6e6f  the outer (pheno
+0001f470: 7479 7065 2920 6369 7263 6c65 730a 2020  type) circles.  
+0001f480: 2020 2020 2020 2020 2020 7365 6c66 2e5f            self._
+0001f490: 706c 6f74 5f70 6865 6e6f 7479 7065 2874  plot_phenotype(t
+0001f4a0: 7261 6974 203d 2074 7274 5f6e 756d 2c20  rait = trt_num, 
+0001f4b0: 6c79 725f 6e75 6d20 3d20 6c79 725f 6e75  lyr_num = lyr_nu
+0001f4c0: 6d2c 0a20 2020 2020 2020 2020 2020 2020  m,.             
+0001f4d0: 2020 206c 616e 6420 3d20 6c61 6e64 2c20     land = land, 
+0001f4e0: 696e 6469 7669 6473 203d 2069 6e64 6976  individs = indiv
+0001f4f0: 6964 732c 2074 6578 7420 3d20 4661 6c73  ids, text = Fals
+0001f500: 652c 2073 697a 6520 3d20 7369 7a65 2c0a  e, size = size,.
+0001f510: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f520: 7465 7874 5f73 697a 6520 3d20 7465 7874  text_size = text
+0001f530: 5f73 697a 652c 2065 6467 655f 636f 6c6f  _size, edge_colo
+0001f540: 723d 6564 6765 5f63 6f6c 6f72 2c0a 2020  r=edge_color,.  
+0001f550: 2020 2020 2020 2020 2020 2020 2020 7465                te
+0001f560: 7874 5f63 6f6c 6f72 203d 2074 6578 745f  xt_color = text_
+0001f570: 636f 6c6f 722c 2063 6261 7220 3d20 6362  color, cbar = cb
+0001f580: 6172 2c20 616c 7068 6120 3d20 616c 7068  ar, alpha = alph
+0001f590: 612c 0a20 2020 2020 2020 2020 2020 2020  a,.             
+0001f5a0: 2020 207a 6f6f 6d5f 7769 6474 6820 3d20     zoom_width = 
+0001f5b0: 7a6f 6f6d 5f77 6964 7468 2c20 7820 3d20  zoom_width, x = 
+0001f5c0: 782c 2079 203d 2079 2c20 7469 636b 733d  x, y = y, ticks=
+0001f5d0: 7469 636b 732c 0a20 2020 2020 2020 2020  ticks,.         
+0001f5e0: 2020 2020 2020 206d 6173 6b5f 7261 7374         mask_rast
+0001f5f0: 3d6d 6173 6b5f 7261 7374 290a 2020 2020  =mask_rast).    
+0001f600: 2020 2020 2020 2020 236d 616b 6520 7369          #make si
+0001f610: 7a65 2073 6d61 6c6c 6572 2066 6f72 2074  ze smaller for t
+0001f620: 6865 206e 6578 7420 6c61 7965 7220 6f66  he next layer of
+0001f630: 2069 6e6e 6572 2028 6669 746e 6573 7329   inner (fitness)
+0001f640: 2063 6972 636c 6573 0a20 2020 2020 2020   circles.       
+0001f650: 2020 2020 2073 697a 6520 3d20 726f 756e       size = roun
+0001f660: 6428 302e 342a 7369 7a65 290a 2020 2020  d(0.4*size).    
+0001f670: 2020 2020 2020 2020 2320 6765 7420 7369          # get si
+0001f680: 7a65 7320 666f 7220 616c 6c20 696e 6469  zes for all indi
+0001f690: 7669 6475 616c 7327 2069 6e6e 6572 2d63  viduals' inner-c
+0001f6a0: 6972 636c 6520 6669 746e 6573 7320 706f  ircle fitness po
+0001f6b0: 696e 7473 2c20 6966 0a20 2020 2020 2020  ints, if.       
+0001f6c0: 2020 2020 2023 2074 7274 5f6e 756d 2069       # trt_num i
+0001f6d0: 7320 6e6f 7420 4e6f 6e65 0a20 2020 2020  s not None.     
+0001f6e0: 2020 2020 2020 2069 6620 7472 745f 6e75         if trt_nu
+0001f6f0: 6d20 6973 206e 6f74 204e 6f6e 653a 0a20  m is not None:. 
+0001f700: 2020 2020 2020 2020 2020 2020 2020 2073                 s
+0001f710: 697a 6520 3d20 7369 7a65 202a 2028 3120  ize = size * (1 
+0001f720: 2d20 2828 6e70 2e61 7272 6179 285b 2a77  - ((np.array([*w
+0001f730: 2e76 616c 7565 7328 0a20 2020 2020 2020  .values(.       
+0001f740: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f750: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f760: 2020 2020 2029 5d29 202d 206d 696e 5f66       )]) - min_f
+0001f770: 6974 2920 2f20 2831 202d 206d 696e 5f66  it) / (1 - min_f
+0001f780: 6974 2929 290a 0a20 2020 2020 2020 2073  it)))..        s
+0001f790: 656c 662e 5f70 6c6f 7428 6c79 725f 6e75  elf._plot(lyr_nu
+0001f7a0: 6d3d 6c79 725f 6e75 6d2c 206c 616e 643d  m=lyr_num, land=
+0001f7b0: 6c61 6e64 2c20 6869 6465 5f6c 616e 643d  land, hide_land=
+0001f7c0: 5472 7565 2c0a 2020 2020 2020 2020 2020  True,.          
+0001f7d0: 2020 2020 2020 2020 2069 6e64 6976 6964           individ
+0001f7e0: 733d 696e 6469 7669 6473 2c20 7465 7874  s=individs, text
+0001f7f0: 3d74 6578 742c 2063 6f6c 6f72 3d6c 6973  =text, color=lis
+0001f800: 7428 772e 7661 6c75 6573 2829 292c 0a20  t(w.values()),. 
+0001f810: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001f820: 2020 7074 5f63 6d61 703d 636d 6170 2c20    pt_cmap=cmap, 
+0001f830: 7369 7a65 3d73 697a 652c 2065 6467 655f  size=size, edge_
+0001f840: 636f 6c6f 723d 6564 6765 5f63 6f6c 6f72  color=edge_color
+0001f850: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001f860: 2020 2020 2074 6578 745f 636f 6c6f 723d       text_color=
+0001f870: 7465 7874 5f63 6f6c 6f72 2c20 6362 6172  text_color, cbar
+0001f880: 3d63 6261 722c 2074 6578 745f 7369 7a65  =cbar, text_size
+0001f890: 3d74 6578 745f 7369 7a65 2c0a 2020 2020  =text_size,.    
+0001f8a0: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+0001f8b0: 6c70 6861 3d61 6c70 6861 2c20 7a6f 6f6d  lpha=alpha, zoom
+0001f8c0: 5f77 6964 7468 3d7a 6f6f 6d5f 7769 6474  _width=zoom_widt
+0001f8d0: 682c 2078 3d78 2c20 793d 792c 2074 6963  h, x=x, y=y, tic
+0001f8e0: 6b73 3d74 6963 6b73 2c0a 2020 2020 2020  ks=ticks,.      
+0001f8f0: 2020 2020 2020 2020 2020 2020 206d 6173               mas
+0001f900: 6b5f 7261 7374 3d6d 6173 6b5f 7261 7374  k_rast=mask_rast
+0001f910: 290a 0a20 2020 2020 2020 2023 706c 6f74  )..        #plot
+0001f920: 2070 6865 6e6f 7479 7065 2074 6578 7420   phenotype text 
+0001f930: 2877 6f72 6b73 206f 6e6c 7920 6966 2070  (works only if p
+0001f940: 6c6f 7474 696e 6720 6120 7370 6563 6966  lotting a specif
+0001f950: 6963 2074 7261 6974 290a 2020 2020 2020  ic trait).      
+0001f960: 2020 6966 2070 6865 6e6f 7479 7065 5f74    if phenotype_t
+0001f970: 6578 7420 616e 6420 7472 745f 6e75 6d20  ext and trt_num 
+0001f980: 6973 206e 6f74 204e 6f6e 653a 0a20 2020  is not None:.   
+0001f990: 2020 2020 2020 2020 2066 6f72 2069 6e64           for ind
+0001f9a0: 2069 6e20 7365 6c66 2e76 616c 7565 7328   in self.values(
+0001f9b0: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
+0001f9c0: 2020 2070 6c74 2e74 6578 7428 696e 642e     plt.text(ind.
+0001f9d0: 782d 302e 352c 2069 6e64 2e79 2d30 2e35  x-0.5, ind.y-0.5
+0001f9e0: 2c20 2725 302e 3266 2720 2520 696e 642e  , '%0.2f' % ind.
+0001f9f0: 7a5b 7472 745f 6e75 6d5d 2c0a 2020 2020  z[trt_num],.    
+0001fa00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001fa10: 636f 6c6f 7220 3d20 7068 656e 6f74 7970  color = phenotyp
+0001fa20: 655f 7465 7874 5f63 6f6c 6f72 2c20 7369  e_text_color, si
+0001fa30: 7a65 203d 2074 6578 745f 7369 7a65 290a  ze = text_size).
+0001fa40: 0a20 2020 2020 2020 2023 706c 6f74 2066  .        #plot f
+0001fa50: 6974 6e65 7373 2074 6578 740a 2020 2020  itness text.    
+0001fa60: 2020 2020 6966 2066 6974 6e65 7373 5f74      if fitness_t
+0001fa70: 6578 743a 0a20 2020 2020 2020 2020 2020  ext:.           
+0001fa80: 206f 6666 7365 745f 6672 6f6d 5f70 6865   offset_from_phe
+0001fa90: 6e6f 7479 7065 5f74 6578 7420 3d20 302e  notype_text = 0.
+0001faa0: 3030 312a 6d61 7828 7365 6c66 2e5f 6c61  001*max(self._la
+0001fab0: 6e64 5f64 696d 290a 2020 2020 2020 2020  nd_dim).        
+0001fac0: 2020 2020 666f 7220 696e 6420 696e 2073      for ind in s
+0001fad0: 656c 662e 7661 6c75 6573 2829 3a0a 2020  elf.values():.  
+0001fae0: 2020 2020 2020 2020 2020 2020 2020 706c                pl
+0001faf0: 742e 7465 7874 2869 6e64 2e78 2d30 2e35  t.text(ind.x-0.5
+0001fb00: 2b6f 6666 7365 745f 6672 6f6d 5f70 6865  +offset_from_phe
+0001fb10: 6e6f 7479 7065 5f74 6578 742c 0a20 2020  notype_text,.   
+0001fb20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001fb30: 2020 2020 2020 696e 642e 792d 302e 352b        ind.y-0.5+
+0001fb40: 6f66 6673 6574 5f66 726f 6d5f 7068 656e  offset_from_phen
+0001fb50: 6f74 7970 655f 7465 7874 2c0a 2020 2020  otype_text,.    
+0001fb60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001fb70: 2020 2020 2027 2530 2e32 6627 2025 2069       '%0.2f' % i
+0001fb80: 6e64 2e66 6974 2c20 636f 6c6f 7220 3d20  nd.fit, color = 
+0001fb90: 6669 746e 6573 735f 7465 7874 5f63 6f6c  fitness_text_col
+0001fba0: 6f72 2c0a 2020 2020 2020 2020 2020 2020  or,.            
+0001fbb0: 2020 2020 2020 2020 2020 2020 2073 697a               siz
+0001fbc0: 6520 3d20 7465 7874 5f73 697a 6529 0a0a  e = text_size)..
+0001fbd0: 2020 2020 2020 2020 2361 6e64 206d 616b          #and mak
+0001fbe0: 6520 6120 636f 6c6f 7262 6172 2066 6f72  e a colorbar for
+0001fbf0: 2074 6865 2066 6974 6e65 7373 2076 616c   the fitness val
+0001fc00: 7565 7320 0a20 2020 2020 2020 2069 6620  ues .        if 
+0001fc10: 6669 746e 6573 735f 6362 6172 3a0a 2020  fitness_cbar:.  
+0001fc20: 2020 2020 2020 2020 2020 7669 7a2e 5f6d            viz._m
+0001fc30: 616b 655f 6669 746e 6573 735f 6362 6172  ake_fitness_cbar
+0001fc40: 286d 616b 655f 6362 6172 5f66 6e2c 206d  (make_cbar_fn, m
+0001fc50: 696e 5f66 6974 290a 0a20 2020 2023 6d65  in_fit)..    #me
+0001fc60: 7468 6f64 2074 6f20 706c 6f74 2061 2073  thod to plot a s
+0001fc70: 7065 6369 6573 2720 616c 6c65 6c65 2066  pecies' allele f
+0001fc80: 7265 7175 656e 6369 6573 0a20 2020 2064  requencies.    d
+0001fc90: 6566 205f 706c 6f74 5f61 6c6c 656c 655f  ef _plot_allele_
+0001fca0: 6672 6571 7565 6e63 6965 7328 7365 6c66  frequencies(self
+0001fcb0: 2c20 636f 6c6f 723d 2772 6564 2729 3a0a  , color='red'):.
+0001fcc0: 2020 2020 2020 2020 6966 2073 656c 662e          if self.
+0001fcd0: 6765 6e5f 6172 6368 2069 7320 4e6f 6e65  gen_arch is None
+0001fce0: 3a0a 2020 2020 2020 2020 2020 2020 7072  :.            pr
+0001fcf0: 696e 7428 2822 5370 6563 6965 732e 5f70  int(("Species._p
+0001fd00: 6c6f 745f 616c 6c65 6c65 5f66 7265 7175  lot_allele_frequ
+0001fd10: 656e 6369 6573 2069 7320 6e6f 7420 7661  encies is not va
+0001fd20: 6c69 6420 666f 7220 220a 2020 2020 2020  lid for ".      
+0001fd30: 2020 2020 2020 2020 2020 2020 2022 7370               "sp
+0001fd40: 6563 6965 7320 7769 7468 6f75 7420 6765  ecies without ge
+0001fd50: 6e6f 6d65 732e 5c6e 2229 290a 2020 2020  nomes.\n")).    
+0001fd60: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+0001fd70: 2020 2020 2020 7365 6c66 2e67 656e 5f61        self.gen_a
+0001fd80: 7263 682e 5f70 6c6f 745f 616c 6c65 6c65  rch._plot_allele
+0001fd90: 5f66 7265 7175 656e 6369 6573 2873 656c  _frequencies(sel
+0001fda0: 662c 2063 6f6c 6f72 3d63 6f6c 6f72 290a  f, color=color).
+0001fdb0: 0a20 2020 2023 206d 6574 686f 6420 666f  .    # method fo
+0001fdc0: 7220 706c 6f74 7469 6e67 2061 2068 6973  r plotting a his
+0001fdd0: 746f 6772 616d 206f 6620 7468 6520 6375  togram of the cu
+0001fde0: 7272 656e 7420 6669 746e 6573 7320 7661  rrent fitness va
+0001fdf0: 6c75 6573 0a20 2020 2064 6566 205f 706c  lues.    def _pl
+0001fe00: 6f74 5f68 6973 745f 6669 746e 6573 7328  ot_hist_fitness(
+0001fe10: 7365 6c66 293a 0a20 2020 2020 2020 2070  self):.        p
+0001fe20: 6c74 2e68 6973 7428 6c69 7374 2873 656c  lt.hist(list(sel
+0001fe30: 662e 5f63 616c 635f 6669 746e 6573 7328  f._calc_fitness(
+0001fe40: 2929 290a 2020 2020 2020 2020 706c 742e  ))).        plt.
+0001fe50: 786c 6162 656c 2827 4669 746e 6573 7327  xlabel('Fitness'
+0001fe60: 290a 2020 2020 2020 2020 706c 742e 796c  ).        plt.yl
+0001fe70: 6162 656c 2827 436f 756e 7427 290a 0a20  abel('Count').. 
+0001fe80: 2020 2023 206d 6574 686f 6420 666f 7220     # method for 
+0001fe90: 706c 6f74 7469 6e67 2074 6865 206d 6f76  plotting the mov
+0001fea0: 656d 656e 7420 7375 7266 6163 6520 2869  ement surface (i
+0001feb0: 6e20 7661 7269 6f75 7320 666f 726d 6174  n various format
+0001fec0: 7329 0a20 2020 2064 6566 205f 706c 6f74  s).    def _plot
+0001fed0: 5f64 6972 6563 7469 6f6e 5f73 7572 6661  _direction_surfa
+0001fee0: 6365 2873 656c 662c 206c 616e 642c 2073  ce(self, land, s
+0001fef0: 7572 665f 7479 7065 2c20 7374 796c 652c  urf_type, style,
+0001ff00: 2078 3d4e 6f6e 652c 2079 3d4e 6f6e 652c   x=None, y=None,
+0001ff10: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0001ff20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ff30: 207a 6f6f 6d5f 7769 6474 683d 4e6f 6e65   zoom_width=None
+0001ff40: 2c20 7363 616c 655f 6661 6374 3d34 2e35  , scale_fact=4.5
+0001ff50: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0001ff60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ff70: 2020 636f 6c6f 723d 2762 6c61 636b 272c    color='black',
+0001ff80: 2063 6261 723d 5472 7565 2c20 7469 636b   cbar=True, tick
+0001ff90: 733d 4e6f 6e65 2c0a 2020 2020 2020 2020  s=None,.        
+0001ffa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0001ffb0: 2020 2020 2020 2020 636d 6170 3d27 706c          cmap='pl
+0001ffc0: 6173 6d61 272c 206d 6173 6b5f 7261 7374  asma', mask_rast
+0001ffd0: 3d4e 6f6e 6529 3a0a 2020 2020 2020 2020  =None):.        
+0001ffe0: 2320 6765 7420 7468 6520 636f 7272 6563  # get the correc
+0001fff0: 7420 7375 7266 6163 650a 2020 2020 2020  t surface.      
+00020000: 2020 6966 2073 7572 665f 7479 7065 203d    if surf_type =
+00020010: 3d20 276d 6f76 6527 3a0a 2020 2020 2020  = 'move':.      
+00020020: 2020 2020 2020 7375 7266 203d 2073 656c        surf = sel
+00020030: 662e 5f6d 6f76 655f 7375 7266 0a20 2020  f._move_surf.   
+00020040: 2020 2020 2065 6c69 6620 7375 7266 5f74       elif surf_t
+00020050: 7970 6520 3d3d 2027 6469 7370 273a 0a20  ype == 'disp':. 
+00020060: 2020 2020 2020 2020 2020 2073 7572 6620             surf 
+00020070: 3d3d 2073 656c 662e 5f64 6973 705f 7375  == self._disp_su
+00020080: 7266 0a0a 2020 2020 2020 2020 2320 6765  rf..        # ge
+00020090: 7420 616c 6c20 7827 7320 616e 6420 7927  t all x's and y'
+000200a0: 732c 2069 6620 7820 616e 6420 7920 6172  s, if x and y ar
+000200b0: 6520 4e6f 6e65 0a20 2020 2020 2020 2069  e None.        i
+000200c0: 6620 7820 6973 204e 6f6e 6520 616e 6420  f x is None and 
+000200d0: 7920 6973 204e 6f6e 653a 0a20 2020 2020  y is None:.     
+000200e0: 2020 2020 2020 2078 203d 205b 2a72 616e         x = [*ran
+000200f0: 6765 286c 616e 642e 6469 6d5b 305d 295d  ge(land.dim[0])]
+00020100: 0a20 2020 2020 2020 2020 2020 2079 203d  .            y =
+00020110: 205b 2a72 616e 6765 286c 616e 642e 6469   [*range(land.di
+00020120: 6d5b 315d 295d 0a20 2020 2020 2020 2065  m[1])].        e
+00020130: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
+00020140: 2078 203d 205b 785d 0a20 2020 2020 2020   x = [x].       
+00020150: 2020 2020 2079 203d 205b 795d 0a20 2020       y = [y].   
+00020160: 2020 2020 2023 6368 6563 6b20 6966 2074       #check if t
+00020170: 6865 2073 7572 6661 6365 2069 7320 6e6f  he surface is no
+00020180: 6e65 0a20 2020 2020 2020 2069 6620 7375  ne.        if su
+00020190: 7266 2069 7320 4e6f 6e65 3a0a 2020 2020  rf is None:.    
+000201a0: 2020 2020 2020 2020 7072 696e 7428 2827          print(('
+000201b0: 4675 6e63 7469 6f6e 206e 6f74 2076 616c  Function not val
+000201c0: 6964 2066 6f72 2061 2053 7065 6369 6573  id for a Species
+000201d0: 2077 6974 6820 6e6f 2027 0a20 2020 2020   with no '.     
+000201e0: 2020 2020 2020 2020 2020 2020 2020 275f                '_
+000201f0: 2573 5375 7266 6163 652e 2729 2025 2028  %sSurface.') % (
+00020200: 7b0a 2020 2020 2020 2020 2020 2020 2020  {.              
+00020210: 2020 276d 6f76 6527 3a20 274d 6f76 656d    'move': 'Movem
+00020220: 656e 7427 2c20 2764 6973 7027 3a20 2744  ent', 'disp': 'D
+00020230: 6973 7065 7273 616c 277d 5b73 7572 665f  ispersal'}[surf_
+00020240: 7479 7065 5d29 290a 2020 2020 2020 2020  type])).        
+00020250: 2020 2020 7265 7475 726e 0a20 2020 2020      return.     
+00020260: 2020 2065 6c69 6620 7374 796c 6520 6e6f     elif style no
+00020270: 7420 696e 205b 2768 6973 7427 2c20 2763  t in ['hist', 'c
+00020280: 6869 7374 272c 2027 7665 6374 272c 2027  hist', 'vect', '
+00020290: 6364 7261 7773 275d 3a0a 2020 2020 2020  cdraws']:.      
+000202a0: 2020 2020 2020 7072 696e 7428 2822 5468        print(("Th
+000202b0: 6520 2773 7479 6c65 2720 6172 6775 6d65  e 'style' argume
+000202c0: 6e74 206d 7573 7420 7461 6b65 206f 6e65  nt must take one
+000202d0: 206f 6620 7468 6520 220a 2020 2020 2020   of the ".      
+000202e0: 2020 2020 2020 2020 2020 2020 2022 666f               "fo
+000202f0: 6c6c 6f77 696e 6720 7661 6c75 6573 3a20  llowing values: 
+00020300: 2768 6973 7427 2c20 2763 6869 7374 272c  'hist', 'chist',
+00020310: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00020320: 2020 2020 2020 2227 7665 6374 272c 2027        "'vect', '
+00020330: 6364 7261 7773 2722 2929 0a20 2020 2020  cdraws'")).     
+00020340: 2020 2020 2020 2072 6574 7572 6e0a 2020         return.  
+00020350: 2020 2020 2020 656c 6966 2073 7479 6c65        elif style
+00020360: 203d 3d20 2768 6973 7427 3a0a 2020 2020   == 'hist':.    
+00020370: 2020 2020 2020 2020 7820 3d20 785b 305d          x = x[0]
+00020380: 0a20 2020 2020 2020 2020 2020 2079 203d  .            y =
+00020390: 2079 5b30 5d0a 2020 2020 2020 2020 2020   y[0].          
+000203a0: 2020 706c 742e 6869 7374 2872 2e63 686f    plt.hist(r.cho
+000203b0: 6963 6528 7375 7266 2e73 7572 665b 792c  ice(surf.surf[y,
+000203c0: 782c 3a5d 2c20 7369 7a65 203d 2031 3030  x,:], size = 100
+000203d0: 3030 2c20 7265 706c 6163 6520 3d20 5472  00, replace = Tr
+000203e0: 7565 292c 0a20 2020 2020 2020 2020 2020  ue),.           
+000203f0: 2020 2020 2020 2020 2020 6269 6e73 3d31            bins=1
+00020400: 3030 2c20 6465 6e73 6974 793d 5472 7565  00, density=True
+00020410: 2c20 616c 7068 613d 302e 352c 2063 6f6c  , alpha=0.5, col
+00020420: 6f72 3d63 6f6c 6f72 290a 0a20 2020 2020  or=color)..     
+00020430: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
+00020440: 2020 2020 2023 6469 7370 6c61 7920 7468       #display th
+00020450: 6520 6d6f 7665 6d65 6e74 2d73 7572 6661  e movement-surfa
+00020460: 6365 2072 6173 7465 720a 2020 2020 2020  ce raster.      
+00020470: 2020 2020 2020 6c79 725f 6e75 6d20 3d20        lyr_num = 
+00020480: 7375 7266 2e6c 7972 5f6e 756d 0a20 2020  surf.lyr_num.   
+00020490: 2020 2020 2020 2020 206c 616e 645b 6c79           land[ly
+000204a0: 725f 6e75 6d5d 2e5f 706c 6f74 287a 6f6f  r_num]._plot(zoo
+000204b0: 6d5f 7769 6474 6820 3d20 7a6f 6f6d 5f77  m_width = zoom_w
+000204c0: 6964 7468 2c20 783d 6e70 2e6d 6561 6e28  idth, x=np.mean(
+000204d0: 7829 2c0a 2020 2020 2020 2020 2020 2020  x),.            
+000204e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000204f0: 2020 2020 793d 6e70 2e6d 6561 6e28 7929      y=np.mean(y)
+00020500: 2c20 7469 636b 733d 7469 636b 732c 2063  , ticks=ticks, c
+00020510: 6d61 703d 636d 6170 2c0a 2020 2020 2020  map=cmap,.      
+00020520: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020530: 2020 2020 2020 2020 2020 6d61 736b 5f72            mask_r
+00020540: 6173 743d 6d61 736b 5f72 6173 7429 0a0a  ast=mask_rast)..
+00020550: 2020 2020 2020 2020 2020 2020 6966 2073              if s
+00020560: 7479 6c65 203d 3d20 2763 6869 7374 273a  tyle == 'chist':
+00020570: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00020580: 2066 6f72 2078 5f76 616c 2069 6e20 783a   for x_val in x:
+00020590: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000205a0: 2020 2020 2066 6f72 2079 5f76 616c 2069       for y_val i
+000205b0: 6e20 793a 0a20 2020 2020 2020 2020 2020  n y:.           
+000205c0: 2020 2020 2020 2020 2020 2020 2076 2c20               v, 
+000205d0: 6120 3d20 6e70 2e68 6973 746f 6772 616d  a = np.histogram
+000205e0: 2872 2e63 686f 6963 6528 7375 7266 2e73  (r.choice(surf.s
+000205f0: 7572 665b 795f 7661 6c2c 0a20 2020 2020  urf[y_val,.     
+00020600: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020610: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020620: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020630: 2020 2020 2020 2020 2020 785f 7661 6c2c            x_val,
+00020640: 203a 5d2c 0a20 2020 2020 2020 2020 2020   :],.           
+00020650: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020660: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020670: 2020 2020 2020 2020 2020 7265 706c 6163            replac
+00020680: 653d 5472 7565 2c0a 2020 2020 2020 2020  e=True,.        
+00020690: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000206a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000206b0: 2020 2020 2020 2020 2020 2020 2073 697a               siz
+000206c0: 653d 3735 3030 292c 2062 696e 733d 3135  e=7500), bins=15
+000206d0: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+000206e0: 2020 2020 2020 2020 2020 7620 3d20 7620            v = v 
+000206f0: 2f20 666c 6f61 7428 762e 7375 6d28 2929  / float(v.sum())
+00020700: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00020710: 2020 2020 2020 2020 2061 203d 205b 2861           a = [(a
+00020720: 5b6e 5d20 2b20 615b 6e20 2b20 315d 2920  [n] + a[n + 1]) 
+00020730: 2f20 3220 666f 7220 6e20 696e 2072 616e  / 2 for n in ran
+00020740: 6765 286c 656e 2861 2920 2d20 3129 5d0a  ge(len(a) - 1)].
+00020750: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020760: 2020 2020 2020 2020 7873 203d 205b 6e70          xs = [np
+00020770: 2e63 6f73 2861 5b6e 5d29 202a 2030 2e37  .cos(a[n]) * 0.7
+00020780: 3520 666f 7220 6e20 696e 2072 616e 6765  5 for n in range
+00020790: 286c 656e 2861 2929 5d0a 2020 2020 2020  (len(a))].      
+000207a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000207b0: 2020 7973 203d 205b 6e70 2e73 696e 2861    ys = [np.sin(a
+000207c0: 5b6e 5d29 202a 2030 2e37 3520 666f 7220  [n]) * 0.75 for 
+000207d0: 6e20 696e 2072 616e 6765 286c 656e 2861  n in range(len(a
+000207e0: 2929 5d0a 2020 2020 2020 2020 2020 2020  ))].            
+000207f0: 2020 2020 2020 2020 2020 2020 7873 203d              xs =
+00020800: 206e 702e 6172 7261 7928 7873 2920 2a20   np.array(xs) * 
+00020810: 7620 2a20 7363 616c 655f 6661 6374 0a20  v * scale_fact. 
+00020820: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020830: 2020 2020 2020 2079 7320 3d20 6e70 2e61         ys = np.a
+00020840: 7272 6179 2879 7329 202a 2076 202a 2073  rray(ys) * v * s
+00020850: 6361 6c65 5f66 6163 740a 2020 2020 2020  cale_fact.      
+00020860: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020870: 2020 5b70 6c74 2e70 6c6f 7428 2878 5f76    [plt.plot((x_v
+00020880: 616c 202b 2030 2e35 2c20 2878 5f76 616c  al + 0.5, (x_val
+00020890: 202b 2030 2e35 202b 2078 735b 6e5d 2929   + 0.5 + xs[n]))
+000208a0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+000208b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000208c0: 2020 2020 2879 5f76 616c 202b 2030 2e35      (y_val + 0.5
+000208d0: 2c20 2879 5f76 616c 202b 2030 2e35 202b  , (y_val + 0.5 +
+000208e0: 2079 735b 6e5d 2929 2c0a 2020 2020 2020   ys[n])),.      
+000208f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020900: 2020 2020 2020 2020 2020 2020 6c69 6e65              line
+00020910: 7769 6474 683d 322c 0a20 2020 2020 2020  width=2,.       
+00020920: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020930: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
+00020940: 3d63 6f6c 6f72 2920 666f 7220 6e20 696e  =color) for n in
+00020950: 2072 616e 6765 286c 656e 2878 7329 295d   range(len(xs))]
+00020960: 0a0a 2020 2020 2020 2020 2020 2020 656c  ..            el
+00020970: 6966 2073 7479 6c65 203d 3d20 2763 6472  if style == 'cdr
+00020980: 6177 7327 3a0a 2020 2020 2020 2020 2020  aws':.          
+00020990: 2020 2020 2020 666f 7220 785f 7661 6c20        for x_val 
+000209a0: 696e 2078 3a0a 2020 2020 2020 2020 2020  in x:.          
+000209b0: 2020 2020 2020 2020 2020 666f 7220 795f            for y_
+000209c0: 7661 6c20 696e 2079 3a0a 2020 2020 2020  val in y:.      
+000209d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000209e0: 2020 7074 7320 3d20 5b28 6e70 2e63 6f73    pts = [(np.cos
+000209f0: 2861 292c 206e 702e 7369 6e28 6129 2920  (a), np.sin(a)) 
+00020a00: 666f 7220 6120 696e 2072 2e63 686f 6963  for a in r.choic
+00020a10: 6528 0a20 2020 2020 2020 2020 2020 2020  e(.             
+00020a20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020a30: 2020 2073 7572 662e 7375 7266 5b79 5f76     surf.surf[y_v
+00020a40: 616c 2c20 785f 7661 6c2c 203a 5d2c 2073  al, x_val, :], s
+00020a50: 697a 653d 3130 3030 2c0a 2020 2020 2020  ize=1000,.      
+00020a60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020a70: 2020 2020 2020 2020 2020 7265 706c 6163            replac
+00020a80: 653d 5472 7565 295d 0a20 2020 2020 2020  e=True)].       
+00020a90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020aa0: 2070 6c74 2e73 6361 7474 6572 285b 7074   plt.scatter([pt
+00020ab0: 5b30 5d20 2a20 302e 3520 2b20 785f 7661  [0] * 0.5 + x_va
+00020ac0: 6c20 2b20 302e 3520 666f 7220 7074 2069  l + 0.5 for pt i
+00020ad0: 6e20 7074 735d 2c0a 2020 2020 2020 2020  n pts],.        
+00020ae0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020af0: 2020 2020 2020 2020 2020 2020 5b70 745b              [pt[
+00020b00: 315d 202a 2030 2e35 202b 2079 5f76 616c  1] * 0.5 + y_val
+00020b10: 202b 2030 2e35 2066 6f72 2070 7420 696e   + 0.5 for pt in
+00020b20: 2070 7473 5d2c 0a20 2020 2020 2020 2020   pts],.         
+00020b30: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020b40: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
+00020b50: 3d63 6f6c 6f72 2c20 616c 7068 613d 302e  =color, alpha=0.
+00020b60: 312c 206d 6172 6b65 723d 272e 2729 0a0a  1, marker='.')..
+00020b70: 2020 2020 2020 2020 2020 2020 656c 6966              elif
+00020b80: 2073 7479 6c65 203d 3d20 2776 6563 7427   style == 'vect'
+00020b90: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+00020ba0: 2020 6465 6620 706c 6f74 5f6f 6e65 5f63    def plot_one_c
+00020bb0: 656c 6c28 782c 2079 293a 0a20 2020 2020  ell(x, y):.     
+00020bc0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00020bd0: 2064 7261 7720 7361 6d70 6c65 206f 6620   draw sample of 
+00020be0: 616e 676c 6573 2066 726f 6d20 7468 6520  angles from the 
+00020bf0: 4761 7573 7369 616e 204b 4445 0a20 2020  Gaussian KDE.   
+00020c00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020c10: 2023 7265 7072 6573 656e 7469 6e67 2074   #representing t
+00020c20: 6865 2076 6f6e 206d 6973 6573 206d 6978  he von mises mix
+00020c30: 7475 7265 2064 6973 7472 6962 7574 696f  ture distributio
+00020c40: 6e20 284b 4445 290a 2020 2020 2020 2020  n (KDE).        
+00020c50: 2020 2020 2020 2020 2020 2020 7361 6d70              samp
+00020c60: 203d 2073 7572 662e 7375 7266 5b79 2c78   = surf.surf[y,x
+00020c70: 2c3a 5d0a 2020 2020 2020 2020 2020 2020  ,:].            
+00020c80: 2020 2020 2020 2020 2320 6372 6561 7465          # create
+00020c90: 206c 6973 7473 206f 6620 7468 6520 7820   lists of the x 
+00020ca0: 616e 6420 7920 2869 2e65 2e20 636f 7320  and y (i.e. cos 
+00020cb0: 616e 6420 7369 6e29 0a20 2020 2020 2020  and sin).       
+00020cc0: 2020 2020 2020 2020 2020 2020 2023 636f               #co
+00020cd0: 6d70 6f6e 656e 7473 206f 6620 6561 6368  mponents of each
+00020ce0: 2061 6e67 6c65 2069 6e20 7468 6520 7361   angle in the sa
+00020cf0: 6d70 6c65 0a20 2020 2020 2020 2020 2020  mple.           
+00020d00: 2020 2020 2020 2020 2078 5f76 6563 7473           x_vects
+00020d10: 203d 206e 702e 636f 7328 7361 6d70 290a   = np.cos(samp).
+00020d20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020d30: 2020 2020 795f 7665 6374 7320 3d20 6e70      y_vects = np
+00020d40: 2e73 696e 2873 616d 7029 0a20 2020 2020  .sin(samp).     
+00020d50: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00020d60: 2064 6566 696e 6520 7468 6520 6478 2061   define the dx a
+00020d70: 6e64 2064 7920 6469 7374 616e 6365 7320  nd dy distances 
+00020d80: 7573 6564 2074 6f20 7468 650a 2020 2020  used to the.    
+00020d90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020da0: 2370 6f73 6974 696f 6e20 7468 6520 6172  #position the ar
+00020db0: 726f 7768 6561 6420 2864 6976 6964 6520  rowhead (divide 
+00020dc0: 6279 2073 7172 7428 3229 2f32 2c20 746f  by sqrt(2)/2, to
+00020dd0: 200a 2020 2020 2020 2020 2020 2020 2020   .              
+00020de0: 2020 2020 2020 2373 6361 6c65 2074 6f20        #scale to 
+00020df0: 7468 6520 7369 7a65 206f 6620 6861 6c66  the size of half
+00020e00: 206f 6620 7468 6520 6469 6167 6f6e 616c   of the diagonal
+00020e10: 206f 6620 6120 6365 6c6c 290a 2020 2020   of a cell).    
+00020e20: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020e30: 6478 203d 206e 702e 6d65 616e 2878 5f76  dx = np.mean(x_v
+00020e40: 6563 7473 2920 2f20 6e70 2e73 7172 7428  ects) / np.sqrt(
+00020e50: 3229 0a20 2020 2020 2020 2020 2020 2020  2).             
+00020e60: 2020 2020 2020 2064 7920 3d20 6e70 2e6d         dy = np.m
+00020e70: 6561 6e28 795f 7665 6374 7329 202f 206e  ean(y_vects) / n
+00020e80: 702e 7371 7274 2832 290a 2020 2020 2020  p.sqrt(2).      
+00020e90: 2020 2020 2020 2020 2020 2020 2020 2320                # 
+00020ea0: 6e6f 7720 706c 6f74 2074 6865 2061 7272  now plot the arr
+00020eb0: 6f77 0a20 2020 2020 2020 2020 2020 2020  ow.             
+00020ec0: 2020 2020 2020 2070 6c74 2e61 7272 6f77         plt.arrow
+00020ed0: 2878 202b 2030 2e35 2c20 7920 2b20 302e  (x + 0.5, y + 0.
+00020ee0: 352c 2064 782c 2064 792c 2061 6c70 6861  5, dx, dy, alpha
+00020ef0: 3d30 2e37 352c 0a20 2020 2020 2020 2020  =0.75,.         
+00020f00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00020f10: 2020 2020 2063 6f6c 6f72 3d63 6f6c 6f72       color=color
+00020f20: 2c20 6865 6164 5f77 6964 7468 3d30 2e32  , head_width=0.2
+00020f30: 342c 2068 6561 645f 6c65 6e67 7468 3d30  4, head_length=0
+00020f40: 2e33 3229 0a0a 2020 2020 2020 2020 2020  .32)..          
+00020f50: 2020 2020 2020 2320 6361 6c6c 2074 6865        # call the
+00020f60: 2069 6e74 6572 6e61 6c6c 7920 6465 6669   internally defi
+00020f70: 6e65 6420 6675 6e63 7469 6f6e 2061 7320  ned function as 
+00020f80: 6120 6e65 7374 6564 206c 6973 740a 2020  a nested list.  
+00020f90: 2020 2020 2020 2020 2020 2020 2020 2363                #c
+00020fa0: 6f6d 7072 6568 656e 7369 6f6e 2066 6f72  omprehension for
+00020fb0: 2061 6c6c 2072 6173 7465 7220 6365 6c6c   all raster cell
+00020fc0: 732c 2077 6869 6368 2049 2062 656c 6965  s, which I belie
+00020fd0: 7665 0a20 2020 2020 2020 2020 2020 2020  ve.             
+00020fe0: 2020 2023 7368 6f75 6c64 2064 6f20 6974     #should do it
+00020ff0: 7320 6265 7374 2074 6f20 7665 6374 6f72  s best to vector
+00021000: 697a 6520 7468 6520 7768 6f6c 6520 6f70  ize the whole op
+00021010: 6572 6174 696f 6e0a 2020 2020 2020 2020  eration.        
+00021020: 2020 2020 2020 2020 5b5b 706c 6f74 5f6f          [[plot_o
+00021030: 6e65 5f63 656c 6c28 6a2c 2069 2920 666f  ne_cell(j, i) fo
+00021040: 7220 6920 696e 2072 616e 6765 280a 2020  r i in range(.  
+00021050: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021060: 2020 7375 7266 2e73 7572 662e 7368 6170    surf.surf.shap
+00021070: 655b 305d 295d 2066 6f72 206a 2069 6e20  e[0])] for j in 
+00021080: 7261 6e67 6528 0a20 2020 2020 2020 2020  range(.         
+00021090: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000210a0: 2020 2020 2020 2020 2020 2073 7572 662e             surf.
+000210b0: 7375 7266 2e73 6861 7065 5b31 5d29 5d0a  surf.shape[1])].
+000210c0: 0a0a 2020 2020 2320 706c 6f74 2074 6865  ..    # plot the
+000210d0: 206c 696e 6561 6765 2066 6f72 2061 2067   lineage for a g
+000210e0: 6976 656e 206e 6f64 6520 616e 6420 6c6f  iven node and lo
+000210f0: 6375 730a 2020 2020 6465 6620 5f70 6c6f  cus.    def _plo
+00021100: 745f 6765 6e65 5f66 6c6f 7728 7365 6c66  t_gene_flow(self
+00021110: 2c20 6c6f 6375 732c 2073 7479 6c65 2c20  , locus, style, 
+00021120: 6c61 6e64 2c20 6e6f 6465 733d 4e6f 6e65  land, nodes=None
+00021130: 2c20 696e 6469 7669 6473 3d4e 6f6e 652c  , individs=None,
+00021140: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00021150: 2020 2020 2020 2020 636f 6c6f 723d 4e6f          color=No
+00021160: 6e65 2c20 7068 656e 6f74 7970 653d 4e6f  ne, phenotype=No
+00021170: 6e65 2c20 6c79 725f 6e75 6d3d 302c 206a  ne, lyr_num=0, j
+00021180: 6974 7465 723d 5472 7565 2c0a 2020 2020  itter=True,.    
+00021190: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000211a0: 2020 2061 6c70 6861 3d30 2e35 2c20 7369     alpha=0.5, si
+000211b0: 7a65 3d32 352c 2061 6464 5f72 6f6f 7473  ze=25, add_roots
+000211c0: 3d46 616c 7365 293a 0a20 2020 2020 2020  =False):.       
+000211d0: 2069 6620 6e6f 7420 7365 6c66 2e67 656e   if not self.gen
+000211e0: 5f61 7263 682e 7573 655f 7473 6b69 743a  _arch.use_tskit:
+000211f0: 0a20 2020 2020 2020 2020 2020 2072 6169  .            rai
+00021200: 7365 2045 7863 6570 7469 6f6e 2828 2247  se Exception(("G
+00021210: 656e 6520 666c 6f77 2063 616e 6e6f 7420  ene flow cannot 
+00021220: 6265 2070 6c6f 7474 6564 2066 6f72 2061  be plotted for a
+00021230: 2053 7065 6369 6573 2022 0a20 2020 2020   Species ".     
+00021240: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021250: 2020 2020 2020 2020 2277 686f 7365 2073          "whose s
+00021260: 7061 7469 616c 2070 6564 6967 7265 6520  patial pedigree 
+00021270: 6973 206e 6f74 2062 6569 6e67 2074 7261  is not being tra
+00021280: 636b 6564 2022 0a20 2020 2020 2020 2020  cked ".         
+00021290: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000212a0: 2020 2020 2262 7920 7473 6b69 742e 2229      "by tskit.")
+000212b0: 290a 2020 2020 2020 2020 6173 7365 7274  ).        assert
+000212c0: 2073 7479 6c65 2069 6e20 5b27 6c69 6e65   style in ['line
+000212d0: 6167 6527 2c20 2776 6563 746f 7227 5d2c  age', 'vector'],
+000212e0: 2028 2254 6865 2073 7479 6c65 2061 7267   ("The style arg
+000212f0: 756d 656e 7420 6d75 7374 2062 6520 220a  ument must be ".
+00021300: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021310: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021320: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021330: 2267 6976 656e 2065 6974 6865 7220 276c  "given either 'l
+00021340: 696e 6561 6765 2720 6f72 2022 0a20 2020  ineage' or ".   
+00021350: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021360: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021370: 2020 2020 2020 2020 2020 2020 2022 2776               "'v
+00021380: 6563 746f 7227 2061 7320 6120 7661 6c75  ector' as a valu
+00021390: 652e 2229 0a20 2020 2020 2020 2069 6620  e.").        if 
+000213a0: 6e6f 6465 7320 6973 204e 6f6e 653a 0a20  nodes is None:. 
+000213b0: 2020 2020 2020 2020 2020 2023 2067 6574             # get
+000213c0: 2061 2072 616e 646f 6d20 7365 6c65 6374   a random select
+000213d0: 696f 6e20 6f66 206e 2069 6e64 6976 6964  ion of n individ
+000213e0: 7561 6c73 2c20 6966 2069 6e64 6976 6964  uals, if individ
+000213f0: 7327 2076 616c 7565 2069 7320 696e 740a  s' value is int.
+00021400: 2020 2020 2020 2020 2020 2020 6966 2069              if i
+00021410: 7369 6e73 7461 6e63 6528 696e 6469 7669  sinstance(indivi
+00021420: 6473 2c20 696e 7429 3a0a 2020 2020 2020  ds, int):.      
+00021430: 2020 2020 2020 2020 2020 696e 6469 7669            indivi
+00021440: 6473 203d 206e 702e 7261 6e64 6f6d 2e63  ds = np.random.c
+00021450: 686f 6963 6528 5b2a 7365 6c66 5d2c 2069  hoice([*self], i
+00021460: 6e64 6976 6964 732c 2072 6570 6c61 6365  ndivids, replace
+00021470: 3d46 616c 7365 290a 2020 2020 2020 2020  =False).        
+00021480: 2020 2020 2373 6f72 7420 616e 6420 7369      #sort and si
+00021490: 6d70 6c69 6679 2074 6865 2054 6162 6c65  mplify the Table
+000214a0: 436f 6c6c 6563 7469 6f6e 2c20 6966 206e  Collection, if n
+000214b0: 6565 6465 6420 0a20 2020 2020 2020 2020  eeded .         
+000214c0: 2020 2069 6620 6e6f 7420 7365 6c66 2e5f     if not self._
+000214d0: 7463 5f73 6f72 7465 645f 616e 645f 7369  tc_sorted_and_si
+000214e0: 6d70 6c69 6669 6564 3a0a 2020 2020 2020  mplified:.      
+000214f0: 2020 2020 2020 2020 2020 7365 6c66 2e5f            self._
+00021500: 736f 7274 5f61 6e64 5f73 696d 706c 6966  sort_and_simplif
+00021510: 795f 7461 626c 655f 636f 6c6c 6563 7469  y_table_collecti
+00021520: 6f6e 2829 0a20 2020 2020 2020 2020 2020  on().           
+00021530: 206e 6f64 6573 203d 2073 656c 662e 5f67   nodes = self._g
+00021540: 6574 5f6e 6f64 6573 2869 6e64 6976 6964  et_nodes(individ
+00021550: 733d 696e 6469 7669 6473 290a 0a20 2020  s=individs)..   
+00021560: 2020 2020 2023 2073 6f72 7420 616e 6420       # sort and 
+00021570: 7369 6d70 6c69 6679 2074 6865 2054 6162  simplify the Tab
+00021580: 6c65 436f 6c6c 6563 7469 6f6e 2069 6620  leCollection if 
+00021590: 6e65 6564 6564 0a20 2020 2020 2020 2069  needed.        i
+000215a0: 6620 6e6f 7420 7365 6c66 2e5f 7463 5f73  f not self._tc_s
+000215b0: 6f72 7465 645f 616e 645f 7369 6d70 6c69  orted_and_simpli
+000215c0: 6669 6564 3a0a 2020 2020 2020 2020 2020  fied:.          
+000215d0: 2020 7365 6c66 2e5f 736f 7274 5f61 6e64    self._sort_and
+000215e0: 5f73 696d 706c 6966 795f 7461 626c 655f  _simplify_table_
+000215f0: 636f 6c6c 6563 7469 6f6e 2829 0a20 2020  collection().   
+00021600: 2020 2020 2023 2067 7261 6220 7468 6520       # grab the 
+00021610: 5461 626c 6543 6f6c 6c65 6374 696f 6e20  TableCollection 
+00021620: 616e 6420 6974 7320 5472 6565 5365 7175  and its TreeSequ
+00021630: 656e 6365 0a20 2020 2020 2020 2074 6320  ence.        tc 
+00021640: 3d20 7365 6c66 2e5f 7463 0a20 2020 2020  = self._tc.     
+00021650: 2020 2074 7279 3a0a 2020 2020 2020 2020     try:.        
+00021660: 2020 2020 7473 203d 2074 632e 7472 6565      ts = tc.tree
+00021670: 5f73 6571 7565 6e63 6528 290a 2020 2020  _sequence().    
+00021680: 2020 2020 6578 6365 7074 2045 7863 6570      except Excep
+00021690: 7469 6f6e 3a0a 2020 2020 2020 2020 2020  tion:.          
+000216a0: 2020 7261 6973 6520 4578 6365 7074 696f    raise Exceptio
+000216b0: 6e28 2822 5468 6520 7370 6563 6965 7327  n(("The species'
+000216c0: 2054 6162 6c65 436f 6c6c 6563 7469 6f6e   TableCollection
+000216d0: 206d 7573 7420 6265 2073 6f72 7465 6420   must be sorted 
+000216e0: 616e 6420 220a 2020 2020 2020 2020 2020  and ".          
+000216f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021700: 2020 2022 7369 6d70 6c69 6669 6564 2062     "simplified b
+00021710: 6566 6f72 6520 7468 6973 206d 6574 686f  efore this metho
+00021720: 6420 6973 2063 616c 6c65 642e 2229 290a  d is called.")).
+00021730: 0a20 2020 2020 2020 206e 6f64 655f 6375  .        node_cu
+00021740: 7272 5f6c 6f63 7320 3d20 5b5b 692e 782c  rr_locs = [[i.x,
+00021750: 2069 2e79 5d20 666f 7220 6e20 696e 206e   i.y] for n in n
+00021760: 6f64 6573 2066 6f72 2069 2069 6e20 7365  odes for i in se
+00021770: 6c66 2e76 616c 7565 7328 0a20 2020 2020  lf.values(.     
+00021780: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021790: 2020 2029 2069 6620 6e20 696e 2069 2e5f     ) if n in i._
+000217a0: 6e6f 6465 735f 7461 625f 6964 732e 7661  nodes_tab_ids.va
+000217b0: 6c75 6573 2829 5d0a 0a20 2020 2020 2020  lues()]..       
+000217c0: 2023 2067 6574 2074 6865 2074 7265 6520   # get the tree 
+000217d0: 666f 7220 7468 6973 206c 6f63 7573 0a20  for this locus. 
+000217e0: 2020 2020 2020 2074 7265 6520 3d20 7473         tree = ts
+000217f0: 2e61 736c 6973 7428 295b 5f67 6574 5f74  .aslist()[_get_t
+00021800: 7265 656e 756d 7328 7473 2c20 5b6c 6f63  reenums(ts, [loc
+00021810: 7573 5d29 5b30 5d5d 0a20 2020 2020 2020  us])[0]].       
+00021820: 2023 2067 6574 2074 6865 206c 696e 6561   # get the linea
+00021830: 6765 5f64 6963 7420 2877 6974 6820 6269  ge_dict (with bi
+00021840: 7274 6820 7469 6d65 7320 616e 6420 6269  rth times and bi
+00021850: 7274 6820 6c6f 6361 7469 6f6e 7329 0a20  rth locations). 
+00021860: 2020 2020 2020 206c 696e 5f64 6963 7420         lin_dict 
+00021870: 3d20 5f67 6574 5f6c 696e 6561 6765 5f64  = _get_lineage_d
+00021880: 6963 7473 5f6f 6e65 5f74 7265 6528 7463  icts_one_tree(tc
+00021890: 2c20 7472 6565 2c20 6e6f 6465 732c 2073  , tree, nodes, s
+000218a0: 656c 662e 7429 0a20 2020 2020 2020 2069  elf.t).        i
+000218b0: 6620 636f 6c6f 7220 6973 204e 6f6e 653a  f color is None:
+000218c0: 0a20 2020 2020 2020 2020 2020 2023 2063  .            # c
+000218d0: 7265 6174 6520 7374 6172 742d 636f 6c6f  reate start-colo
+000218e0: 7220 7661 6c75 6573 2066 6f72 206e 6f64  r values for nod
+000218f0: 6573 2720 7365 7061 7261 7465 206c 696e  es' separate lin
+00021900: 6561 6765 2074 7261 636b 730a 2020 2020  eage tracks.    
+00021910: 2020 2020 2020 2020 636f 6c6f 7273 203d          colors =
+00021920: 205b 6d70 6c2e 636f 6c6f 7273 2e74 6f5f   [mpl.colors.to_
+00021930: 6865 7828 706c 742e 636d 2e53 6574 315f  hex(plt.cm.Set1_
+00021940: 7228 0a20 2020 2020 2020 2020 2020 2020  r(.             
+00021950: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021960: 2020 2020 2020 2020 2020 206e 2929 2066             n)) f
+00021970: 6f72 206e 2069 6e20 6e70 2e6c 696e 7370  or n in np.linsp
+00021980: 6163 6528 302c 2030 2e38 352c 2038 295d  ace(0, 0.85, 8)]
+00021990: 0a20 2020 2020 2020 2065 6c73 653a 0a20  .        else:. 
+000219a0: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
+000219b0: 7320 3d20 5b63 6f6c 6f72 2066 6f72 205f  s = [color for _
+000219c0: 2069 6e20 7261 6e67 6528 3829 5d0a 2020   in range(8)].  
+000219d0: 2020 2020 2020 2320 706c 6f74 2074 6865        # plot the
+000219e0: 2073 7065 6369 6573 2c20 6569 7468 6572   species, either
+000219f0: 2077 6974 6820 6f72 2077 6974 686f 7574   with or without
+00021a00: 2070 6865 6e6f 7479 7065 2d70 6169 6e74   phenotype-paint
+00021a10: 696e 670a 2020 2020 2020 2020 6966 2070  ing.        if p
+00021a20: 6865 6e6f 7479 7065 2069 7320 4e6f 6e65  henotype is None
+00021a30: 3a0a 2020 2020 2020 2020 2020 2020 7365  :.            se
+00021a40: 6c66 2e5f 706c 6f74 286c 7972 5f6e 756d  lf._plot(lyr_num
+00021a50: 3d6c 7972 5f6e 756d 2c20 6c61 6e64 3d6c  =lyr_num, land=l
+00021a60: 616e 642c 2073 697a 653d 7369 7a65 290a  and, size=size).
+00021a70: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+00021a80: 2020 2020 2020 2020 2020 7365 6c66 2e5f            self._
+00021a90: 706c 6f74 5f70 6865 6e6f 7479 7065 2870  plot_phenotype(p
+00021aa0: 6865 6e6f 7479 7065 2c20 6c61 6e64 3d6c  henotype, land=l
+00021ab0: 616e 642c 2073 697a 653d 7369 7a65 290a  and, size=size).
+00021ac0: 2020 2020 2020 2020 6178 203d 2070 6c74          ax = plt
+00021ad0: 2e67 6361 2829 0a20 2020 2020 2020 2023  .gca().        #
+00021ae0: 2065 7874 7261 6374 2061 6e64 2070 6c6f   extract and plo
+00021af0: 7420 7468 6520 7365 7269 6573 206f 6620  t the series of 
+00021b00: 706f 696e 7473 2066 6f72 2065 6163 6820  points for each 
+00021b10: 6e6f 6465 0a20 2020 2020 2020 2066 6f72  node.        for
+00021b20: 2069 2c20 6e6f 6465 2069 6e20 656e 756d   i, node in enum
+00021b30: 6572 6174 6528 6e6f 6465 7329 3a0a 2020  erate(nodes):.  
+00021b40: 2020 2020 2020 2020 2020 7374 6172 745f            start_
+00021b50: 636f 6c20 3d20 636f 6c6f 7273 5b69 2025  col = colors[i %
+00021b60: 206c 656e 2863 6f6c 6f72 7329 5d0a 2020   len(colors)].  
+00021b70: 2020 2020 2020 2020 2020 6966 2073 7479            if sty
+00021b80: 6c65 203d 3d20 276c 696e 6561 6765 273a  le == 'lineage':
+00021b90: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00021ba0: 206c 6f63 7320 3d20 6e70 2e76 7374 6163   locs = np.vstac
+00021bb0: 6b28 5b76 5b31 5d20 666f 7220 7620 696e  k([v[1] for v in
+00021bc0: 206c 696e 5f64 6963 745b 6e6f 6465 5d2e   lin_dict[node].
+00021bd0: 7661 6c75 6573 2829 5d29 0a20 2020 2020  values()]).     
+00021be0: 2020 2020 2020 2020 2020 2069 6620 6a69             if ji
+00021bf0: 7474 6572 3a0a 2020 2020 2020 2020 2020  tter:.          
+00021c00: 2020 2020 2020 2020 2020 6c6f 6373 203d            locs =
+00021c10: 206c 6f63 7320 2b20 6e70 2e72 616e 646f   locs + np.rando
+00021c20: 6d2e 6e6f 726d 616c 2830 2c20 302e 3031  m.normal(0, 0.01
+00021c30: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00021c40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021c50: 2020 2020 2020 2020 2020 2020 2020 7369                si
+00021c60: 7a65 3d6c 6f63 732e 7369 7a65 292e 7265  ze=locs.size).re
+00021c70: 7368 6170 6528 6c6f 6373 2e73 6861 7065  shape(locs.shape
+00021c80: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00021c90: 2020 2320 6372 6561 7465 206c 6973 7420    # create list 
+00021ca0: 6f66 2063 6f6c 6f72 7320 666f 7220 706c  of colors for pl
+00021cb0: 6f74 7469 6e67 2c0a 2020 2020 2020 2020  otting,.        
+00021cc0: 2020 2020 2020 2020 2320 7573 696e 6720          # using 
+00021cd0: 6c69 6e65 6172 6c79 2069 6e74 6572 706f  linearly interpo
+00021ce0: 6c61 7465 6420 636f 6c6f 7273 0a20 2020  lated colors.   
+00021cf0: 2020 2020 2020 2020 2020 2020 2023 2069               # i
+00021d00: 6620 7468 6520 636f 6c6f 7220 6172 6775  f the color argu
+00021d10: 6d65 6e74 2077 6173 206e 6f74 2070 726f  ment was not pro
+00021d20: 7669 6465 642c 200a 2020 2020 2020 2020  vided, .        
+00021d30: 2020 2020 2020 2020 2320 6f72 2065 6c73          # or els
+00021d40: 6520 6a75 7374 2075 7369 6e67 2074 6865  e just using the
+00021d50: 2073 6f6c 6964 2063 6f6c 6f72 0a20 2020   solid color.   
+00021d60: 2020 2020 2020 2020 2020 2020 2023 2070               # p
+00021d70: 726f 7669 6465 6420 746f 2074 6865 2063  rovided to the c
+00021d80: 6f6c 6f72 2061 7267 756d 656e 740a 2020  olor argument.  
+00021d90: 2020 2020 2020 2020 2020 2020 2020 636f                co
+00021da0: 6c6f 725f 6e75 6d73 203d 206e 702e 696e  lor_nums = np.in
+00021db0: 7438 286e 702e 6c69 6e73 7061 6365 2830  t8(np.linspace(0
+00021dc0: 2c20 3130 302c 206c 6f63 732e 7368 6170  , 100, locs.shap
+00021dd0: 655b 305d 2d31 2929 0a20 2020 2020 2020  e[0]-1)).       
+00021de0: 2020 2020 2020 2020 2069 6620 636f 6c6f           if colo
+00021df0: 7220 6973 204e 6f6e 653a 0a20 2020 2020  r is None:.     
+00021e00: 2020 2020 2020 2020 2020 2020 2020 2070                 p
+00021e10: 6c6f 745f 636f 6c6f 7273 203d 5b76 697a  lot_colors =[viz
+00021e20: 2e5f 6361 6c63 5f72 6573 6861 6465 645f  ._calc_reshaded_
+00021e30: 636f 6c6f 7228 7374 6172 745f 636f 6c2c  color(start_col,
+00021e40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00021e50: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021e60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021e70: 2020 206e 756d 2920 666f 7220 6e75 6d20     num) for num 
+00021e80: 696e 2063 6f6c 6f72 5f6e 756d 735d 0a20  in color_nums]. 
+00021e90: 2020 2020 2020 2020 2020 2020 2020 2065                 e
+00021ea0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
+00021eb0: 2020 2020 2020 2020 2070 6c6f 745f 636f           plot_co
+00021ec0: 6c6f 7273 203d 205b 7374 6172 745f 636f  lors = [start_co
+00021ed0: 6c20 666f 7220 6e75 6d20 696e 2063 6f6c  l for num in col
+00021ee0: 6f72 5f6e 756d 735d 0a20 2020 2020 2020  or_nums].       
+00021ef0: 2020 2020 2020 2020 2023 2063 7265 6174           # creat
+00021f00: 6520 6120 6c69 6e65 6172 2069 6e74 6572  e a linear inter
+00021f10: 706f 6c61 7469 6f6e 206f 6620 6c69 6e65  polation of line
+00021f20: 7769 6474 6873 0a20 2020 2020 2020 2020  widths.         
+00021f30: 2020 2020 2020 206c 696e 6577 6964 7468         linewidth
+00021f40: 7320 3d20 6e70 2e6c 696e 7370 6163 6528  s = np.linspace(
+00021f50: 332c 2030 2e38 352c 206c 6f63 732e 7368  3, 0.85, locs.sh
+00021f60: 6170 655b 305d 2d31 290a 2020 2020 2020  ape[0]-1).      
+00021f70: 2020 2020 2020 2020 2020 666f 7220 6e2c            for n,
+00021f80: 2063 6f6c 2069 6e20 656e 756d 6572 6174   col in enumerat
+00021f90: 6528 706c 6f74 5f63 6f6c 6f72 7329 3a0a  e(plot_colors):.
+00021fa0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00021fb0: 2020 2020 2320 4e4f 5445 206e 6565 6420      # NOTE need 
+00021fc0: 746f 2075 7365 206f 6e6c 7920 7468 6520  to use only the 
+00021fd0: 6669 7273 7420 3220 7661 6c75 6573 2069  first 2 values i
+00021fe0: 6e20 7468 6520 6c6f 6361 7469 6f6e 0a20  n the location. 
+00021ff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022000: 2020 2023 2064 6174 6120 6265 6361 7573     # data becaus
+00022010: 6520 7375 6273 6571 7565 6e74 2076 616c  e subsequent val
+00022020: 7565 7320 6172 6520 7573 6564 0a20 2020  ues are used.   
+00022030: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022040: 2023 2074 6f20 7374 6f72 6520 696e 6469   # to store indi
+00022050: 7669 6475 616c 7327 2070 6865 6e6f 7479  viduals' phenoty
+00022060: 7065 7320 616e 6420 6669 746e 6573 730a  pes and fitness.
+00022070: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022080: 2020 2020 6178 2e70 6c6f 7428 6c6f 6373      ax.plot(locs
+00022090: 5b6e 3a6e 2b32 2c20 305d 2c20 6c6f 6373  [n:n+2, 0], locs
+000220a0: 5b6e 3a6e 2b32 2c20 315d 2c20 6c69 6e65  [n:n+2, 1], line
+000220b0: 7374 796c 653d 2773 6f6c 6964 272c 0a20  style='solid',. 
+000220c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000220d0: 2020 2020 2020 2020 2020 206d 6172 6b65             marke
+000220e0: 723d 276f 272c 206d 6172 6b65 7273 697a  r='o', markersiz
+000220f0: 653d 7369 7a65 2a2a 2831 2f32 292c 2063  e=size**(1/2), c
+00022100: 6f6c 6f72 3d63 6f6c 2c0a 2020 2020 2020  olor=col,.      
+00022110: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022120: 2020 2020 2020 6c69 6e65 7769 6474 683d        linewidth=
+00022130: 6c69 6e65 7769 6474 6873 5b6e 5d2c 2061  linewidths[n], a
+00022140: 6c70 6861 3d61 6c70 6861 290a 2020 2020  lpha=alpha).    
+00022150: 2020 2020 2020 2020 656c 6966 2073 7479          elif sty
+00022160: 6c65 203d 3d20 2776 6563 746f 7227 3a0a  le == 'vector':.
+00022170: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022180: 2320 6765 7420 7468 6520 7374 6172 7420  # get the start 
+00022190: 616e 6420 656e 6420 6c6f 6361 7469 6f6e  and end location
+000221a0: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+000221b0: 2020 2320 4e4f 5445 206e 6565 6420 746f    # NOTE need to
+000221c0: 2074 616b 6520 6f6e 6c79 2074 6865 2066   take only the f
+000221d0: 6972 7374 2032 2076 616c 7565 7320 696e  irst 2 values in
+000221e0: 2074 6865 206c 6f63 6174 696f 6e0a 2020   the location.  
+000221f0: 2020 2020 2020 2020 2020 2020 2020 2320                # 
+00022200: 6461 7461 2062 6563 6175 7365 2073 7562  data because sub
+00022210: 7365 7175 656e 7420 7661 6c75 6573 2061  sequent values a
+00022220: 7265 2075 7365 6420 746f 2073 746f 7265  re used to store
+00022230: 2069 6e64 6976 6964 7561 6c73 2720 0a20   individuals' . 
+00022240: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00022250: 2070 6865 6e6f 7479 7065 7320 616e 6420   phenotypes and 
+00022260: 6669 746e 6573 730a 2020 2020 2020 2020  fitness.        
+00022270: 2020 2020 2020 2020 6265 675f 6c6f 6320          beg_loc 
+00022280: 3d20 5b2a 6c69 6e5f 6469 6374 5b6e 6f64  = [*lin_dict[nod
+00022290: 655d 2e76 616c 7565 7328 295d 5b2d 315d  e].values()][-1]
+000222a0: 5b31 5d5b 3a32 5d0a 2020 2020 2020 2020  [1][:2].        
+000222b0: 2020 2020 2020 2020 656e 645f 6c6f 6320          end_loc 
+000222c0: 3d20 5b2a 6c69 6e5f 6469 6374 5b6e 6f64  = [*lin_dict[nod
+000222d0: 655d 2e76 616c 7565 7328 295d 5b30 5d5b  e].values()][0][
+000222e0: 315d 5b3a 325d 0a20 2020 2020 2020 2020  1][:2].         
+000222f0: 2020 2020 2020 2064 782c 2064 7920 3d20         dx, dy = 
+00022300: 5b65 6e64 5f6c 6f63 5b69 5d20 2d20 6265  [end_loc[i] - be
+00022310: 675f 6c6f 635b 695d 2066 6f72 2069 2069  g_loc[i] for i i
+00022320: 6e20 7261 6e67 6528 3229 5d0a 2020 2020  n range(2)].    
+00022330: 2020 2020 2020 2020 2020 2020 2320 706c              # pl
+00022340: 6f74 2074 6865 2076 6563 746f 720a 2020  ot the vector.  
+00022350: 2020 2020 2020 2020 2020 2020 2020 2320                # 
+00022360: 4e4f 5445 3a20 5348 4f55 4c44 2049 2042  NOTE: SHOULD I B
+00022370: 4520 4649 5454 494e 4720 4120 5245 4752  E FITTING A REGR
+00022380: 4553 5349 4f4e 204c 494e 4520 544f 2054  ESSION LINE TO T
+00022390: 4845 2058 2041 4e44 2059 0a20 2020 2020  HE X AND Y.     
+000223a0: 2020 2020 2020 2020 2020 2023 204c 4f43             # LOC
+000223b0: 4154 494f 4e53 2c20 544f 2047 4554 2054  ATIONS, TO GET T
+000223c0: 4845 2027 4156 4552 4147 4527 2056 4543  HE 'AVERAGE' VEC
+000223d0: 544f 522c 2052 4154 4845 5220 5448 414e  TOR, RATHER THAN
+000223e0: 204a 5553 540a 2020 2020 2020 2020 2020   JUST.          
+000223f0: 2020 2020 2020 2320 504c 4f54 5449 4e47        # PLOTTING
+00022400: 2054 4845 2056 4543 544f 5220 4245 5457   THE VECTOR BETW
+00022410: 4545 4e20 5448 4520 4f4c 4445 5354 2041  EEN THE OLDEST A
+00022420: 4e44 2043 5552 5245 4e54 2050 4f53 4954  ND CURRENT POSIT
+00022430: 494f 4e53 0a20 2020 2020 2020 2020 2020  IONS.           
+00022440: 2020 2020 2023 2028 5748 4943 4820 434f       # (WHICH CO
+00022450: 554c 4420 4541 5349 4c59 204d 4953 5245  ULD EASILY MISRE
+00022460: 5052 4553 454e 5420 5448 4520 4f56 4552  PRESENT THE OVER
+00022470: 414c 4c20 5452 454e 4420 4245 4341 5553  ALL TREND BECAUS
+00022480: 4520 0a20 2020 2020 2020 2020 2020 2020  E .             
+00022490: 2020 2023 204f 4620 4348 414e 4345 2041     # OF CHANCE A
+000224a0: 5459 5049 4341 4c20 4c4f 4341 5449 4f4e  TYPICAL LOCATION
+000224b0: 5320 464f 5220 4549 5448 4552 204f 4620  S FOR EITHER OF 
+000224c0: 5448 4f53 4520 5457 4f20 504f 5349 5449  THOSE TWO POSITI
+000224d0: 4f4e 530a 2020 2020 2020 2020 2020 2020  ONS.            
+000224e0: 2020 2020 6178 2e61 7272 6f77 282a 6265      ax.arrow(*be
+000224f0: 675f 6c6f 632c 2064 782c 2064 792c 2063  g_loc, dx, dy, c
+00022500: 6f6c 6f72 3d73 7461 7274 5f63 6f6c 2c0a  olor=start_col,.
+00022510: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022520: 2020 2020 2020 2020 2077 6964 7468 3d30           width=0
+00022530: 2e30 352c 2068 6561 645f 7769 6474 683d  .05, head_width=
+00022540: 302e 342c 206c 656e 6774 685f 696e 636c  0.4, length_incl
+00022550: 7564 6573 5f68 6561 643d 5472 7565 290a  udes_head=True).
+00022560: 2020 2020 2020 2020 2020 2020 2320 706c              # pl
+00022570: 6f74 2074 6865 206e 6f64 6573 2720 6375  ot the nodes' cu
+00022580: 7272 656e 7420 6c6f 6361 7469 6f6e 7320  rrent locations 
+00022590: 616e 6420 7468 6569 7220 6269 7274 6820  and their birth 
+000225a0: 6c6f 6361 7469 6f6e 732c 0a20 2020 2020  locations,.     
+000225b0: 2020 2020 2020 2023 2063 6f6e 6e65 6374         # connect
+000225c0: 6564 2062 7920 6120 7468 696e 2062 6c61  ed by a thin bla
+000225d0: 636b 206c 696e 650a 2020 2020 2020 2020  ck line.        
+000225e0: 2020 2020 6e6f 6465 5f63 7572 725f 6c6f      node_curr_lo
+000225f0: 6320 3d20 6e6f 6465 5f63 7572 725f 6c6f  c = node_curr_lo
+00022600: 6373 5b69 5d0a 2020 2020 2020 2020 2020  cs[i].          
+00022610: 2020 6e6f 6465 5f62 6972 7468 5f6c 6f63    node_birth_loc
+00022620: 203d 205b 2a6c 696e 5f64 6963 745b 6e6f   = [*lin_dict[no
+00022630: 6465 5d2e 7661 6c75 6573 2829 5d5b 305d  de].values()][0]
+00022640: 5b31 5d0a 2020 2020 2020 2020 2020 2020  [1].            
+00022650: 706c 742e 706c 6f74 285b 6e6f 6465 5f62  plt.plot([node_b
+00022660: 6972 7468 5f6c 6f63 5b30 5d2c 206e 6f64  irth_loc[0], nod
+00022670: 655f 6375 7272 5f6c 6f63 5b30 5d5d 2c0a  e_curr_loc[0]],.
+00022680: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022690: 2020 2020 205b 6e6f 6465 5f62 6972 7468       [node_birth
+000226a0: 5f6c 6f63 5b31 5d2c 206e 6f64 655f 6375  _loc[1], node_cu
+000226b0: 7272 5f6c 6f63 5b31 5d5d 2c0a 2020 2020  rr_loc[1]],.    
+000226c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000226d0: 2063 6f6c 6f72 3d73 7461 7274 5f63 6f6c   color=start_col
+000226e0: 2c20 6c69 6e65 7374 796c 653d 273a 272c  , linestyle=':',
+000226f0: 2061 6c70 6861 3d61 6c70 6861 2c0a 2020   alpha=alpha,.  
+00022700: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022710: 2020 206c 696e 6577 6964 7468 3d31 2e32     linewidth=1.2
+00022720: 290a 0a20 2020 2020 2020 2069 6620 6164  )..        if ad
+00022730: 645f 726f 6f74 733a 0a20 2020 2020 2020  d_roots:.       
+00022740: 2020 2020 2073 656c 662e 5f70 6c6f 745f       self._plot_
+00022750: 6c69 6e65 6167 655f 726f 6f74 7328 7463  lineage_roots(tc
+00022760: 2c20 7472 6565 290a 0a0a 2020 2020 2320  , tree)...    # 
+00022770: 706c 6f74 2074 6865 206c 696e 6561 6765  plot the lineage
+00022780: 2066 6f72 2061 2067 6976 656e 206e 6f64   for a given nod
+00022790: 6520 616e 6420 6c6f 6375 730a 2020 2020  e and locus.    
+000227a0: 6465 6620 5f70 6c6f 745f 6c69 6e65 6167  def _plot_lineag
+000227b0: 655f 726f 6f74 7328 7365 6c66 2c20 7463  e_roots(self, tc
+000227c0: 2c20 7472 6565 2c20 616c 7068 613d 302e  , tree, alpha=0.
+000227d0: 382c 2073 697a 653d 3735 293a 0a20 2020  8, size=75):.   
+000227e0: 2020 2020 2069 6620 6e6f 7420 7365 6c66       if not self
+000227f0: 2e67 656e 5f61 7263 682e 7573 655f 7473  .gen_arch.use_ts
+00022800: 6b69 743a 0a20 2020 2020 2020 2020 2020  kit:.           
+00022810: 2072 6169 7365 2045 7863 6570 7469 6f6e   raise Exception
+00022820: 2828 224c 696e 6561 6765 2072 6f6f 7473  (("Lineage roots
+00022830: 2063 616e 6e6f 7420 6265 2070 6c6f 7474   cannot be plott
+00022840: 6564 2066 6f72 2061 2053 7065 6369 6573  ed for a Species
+00022850: 2022 0a20 2020 2020 2020 2020 2020 2020   ".             
+00022860: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022870: 2277 686f 7365 2073 7061 7469 616c 2070  "whose spatial p
+00022880: 6564 6967 7265 6520 6973 206e 6f74 2062  edigree is not b
+00022890: 6569 6e67 2074 7261 636b 6564 2022 0a20  eing tracked ". 
+000228a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000228b0: 2020 2020 2020 2020 2020 2020 2262 7920              "by 
+000228c0: 7473 6b69 742e 2229 290a 0a20 2020 2020  tskit."))..     
+000228d0: 2020 2023 2067 6574 2074 6865 206e 6f64     # get the nod
+000228e0: 6573 0a20 2020 2020 2020 2061 6c6c 5f6e  es.        all_n
+000228f0: 6f64 6573 203d 2073 656c 662e 5f67 6574  odes = self._get
+00022900: 5f6e 6f64 6573 2829 0a0a 2020 2020 2020  _nodes()..      
+00022910: 2020 2320 6765 7420 7468 6520 6c69 6e65    # get the line
+00022920: 6167 6520 6469 6374 2066 6f72 2061 6c6c  age dict for all
+00022930: 206e 6f64 6573 0a20 2020 2020 2020 206c   nodes.        l
+00022940: 696e 5f64 6963 7420 3d20 5f67 6574 5f6c  in_dict = _get_l
+00022950: 696e 6561 6765 5f64 6963 7473 5f6f 6e65  ineage_dicts_one
+00022960: 5f74 7265 6528 7463 2c20 7472 6565 2c20  _tree(tc, tree, 
+00022970: 616c 6c5f 6e6f 6465 732c 2073 656c 662e  all_nodes, self.
+00022980: 7429 0a0a 2020 2020 2020 2020 2320 6765  t)..        # ge
+00022990: 7420 7468 6520 726f 6f74 7320 666f 7220  t the roots for 
+000229a0: 616c 6c20 6469 7374 696e 6374 206c 696e  all distinct lin
+000229b0: 6561 6765 7320 6174 2074 6869 7320 6c6f  eages at this lo
+000229c0: 6375 730a 2020 2020 2020 2020 726f 6f74  cus.        root
+000229d0: 5f6e 6f64 6573 203d 206e 702e 756e 6971  _nodes = np.uniq
+000229e0: 7565 285b 5b2a 6c69 6e5f 6469 6374 5b6e  ue([[*lin_dict[n
+000229f0: 5d2e 6b65 7973 2829 5d5b 2d31 5d20 666f  ].keys()][-1] fo
+00022a00: 7220 6e20 696e 2061 6c6c 5f6e 6f64 6573  r n in all_nodes
+00022a10: 5d29 0a0a 2020 2020 2020 2020 2320 6765  ])..        # ge
+00022a20: 7420 6269 7274 6820 6c6f 6361 7469 6f6e  t birth location
+00022a30: 7320 666f 7220 6561 6368 2072 6f6f 7420  s for each root 
+00022a40: 6e6f 6465 0a20 2020 2020 2020 2072 6f6f  node.        roo
+00022a50: 745f 696e 6469 7669 6473 203d 205b 7463  t_individs = [tc
+00022a60: 2e6e 6f64 6573 5b6e 5d2e 696e 6469 7669  .nodes[n].indivi
+00022a70: 6475 616c 2066 6f72 206e 2069 6e20 726f  dual for n in ro
+00022a80: 6f74 5f6e 6f64 6573 5d0a 2020 2020 2020  ot_nodes].      
+00022a90: 2020 726f 6f74 5f6c 6f63 7320 3d20 5b74    root_locs = [t
+00022aa0: 632e 696e 6469 7669 6475 616c 735b 695d  c.individuals[i]
+00022ab0: 2e6c 6f63 6174 696f 6e20 666f 7220 6920  .location for i 
+00022ac0: 696e 2072 6f6f 745f 696e 6469 7669 6473  in root_individs
+00022ad0: 5d0a 0a20 2020 2020 2020 2023 2065 7874  ]..        # ext
+00022ae0: 7261 6374 2061 6e64 2070 6c6f 7420 7468  ract and plot th
+00022af0: 6520 7365 7269 6573 206f 6620 706f 696e  e series of poin
+00022b00: 7473 2066 6f72 2065 6163 6820 6e6f 6465  ts for each node
+00022b10: 0a20 2020 2020 2020 2066 6f72 2078 2c20  .        for x, 
+00022b20: 7920 696e 2072 6f6f 745f 6c6f 6373 3a0a  y in root_locs:.
+00022b30: 2020 2020 2020 2020 2020 2020 2320 706c              # pl
+00022b40: 6f74 2074 6865 2072 6f6f 7420 6e6f 6465  ot the root node
+00022b50: 7327 2062 6972 7468 206c 6f63 6174 696f  s' birth locatio
+00022b60: 6e73 0a20 2020 2020 2020 2020 2020 2070  ns.            p
+00022b70: 6c74 2e73 6361 7474 6572 2878 2c20 792c  lt.scatter(x, y,
+00022b80: 2063 3d27 7768 6974 6527 2c20 733d 7369   c='white', s=si
+00022b90: 7a65 2c20 616c 7068 613d 616c 7068 612c  ze, alpha=alpha,
+00022ba0: 206d 6172 6b65 723d 2773 2729 0a20 2020   marker='s').   
+00022bb0: 2020 2020 2070 6c74 2e73 686f 7728 290a       plt.show().
+00022bc0: 0a0a 2020 2020 2320 6d65 7468 6f64 2066  ..    # method f
+00022bd0: 6f72 2070 6c6f 7474 696e 6720 6120 7370  or plotting a sp
+00022be0: 6563 6965 7327 2070 6f70 756c 6174 696f  ecies' populatio
+00022bf0: 6e20 7079 7261 6d69 640a 2020 2020 6465  n pyramid.    de
+00022c00: 6620 5f70 6c6f 745f 6465 6d6f 6772 6170  f _plot_demograp
+00022c10: 6869 635f 7079 7261 6d69 6428 7365 6c66  hic_pyramid(self
+00022c20: 293a 0a20 2020 2020 2020 2023 6d61 6b65  ):.        #make
+00022c30: 2064 6963 7420 6f66 2066 656d 616c 6520   dict of female 
+00022c40: 616e 6420 6d61 6c65 2063 6f6c 6f72 730a  and male colors.
+00022c50: 2020 2020 2020 2020 636f 6c5f 6469 6374          col_dict
+00022c60: 203d 207b 2d31 3a20 2763 7961 6e27 2c20   = {-1: 'cyan', 
+00022c70: 313a 2027 7069 6e6b 277d 0a20 2020 2020  1: 'pink'}.     
+00022c80: 2020 2023 6372 6561 7465 2061 2066 6967     #create a fig
+00022c90: 7572 650a 2020 2020 2020 2020 6669 6720  ure.        fig 
+00022ca0: 3d20 706c 742e 6669 6775 7265 2829 0a20  = plt.figure(). 
+00022cb0: 2020 2020 2020 2023 7661 7269 6162 6c65         #variable
+00022cc0: 7320 746f 2067 7261 6220 7468 6520 6d61  s to grab the ma
+00022cd0: 7820 636f 756e 740a 2020 2020 2020 2020  x count.        
+00022ce0: 6d61 785f 636f 756e 7420 3d20 300a 2020  max_count = 0.  
+00022cf0: 2020 2020 2020 2366 6f72 2065 6163 6820        #for each 
+00022d00: 7365 780a 2020 2020 2020 2020 666f 7220  sex.        for 
+00022d10: 7365 785f 7661 6c20 696e 205b 2a63 6f6c  sex_val in [*col
+00022d20: 5f64 6963 745d 3a0a 2020 2020 2020 2020  _dict]:.        
+00022d30: 2020 2020 2367 6574 2061 2063 6f75 6e74      #get a count
+00022d40: 6572 0a20 2020 2020 2020 2020 2020 2063  er.            c
+00022d50: 6f75 6e74 7320 3d20 4328 5b69 6e64 2e61  ounts = C([ind.a
+00022d60: 6765 2066 6f72 2069 6e64 2069 6e20 7365  ge for ind in se
+00022d70: 6c66 2e76 616c 7565 7328 0a20 2020 2020  lf.values(.     
+00022d80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022d90: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00022da0: 2020 2020 2020 2029 2069 6620 696e 642e         ) if ind.
+00022db0: 7365 7820 3d3d 2069 6e74 2873 6578 5f76  sex == int(sex_v
+00022dc0: 616c 203c 2030 295d 290a 2020 2020 2020  al < 0)]).      
+00022dd0: 2020 2020 2020 2367 7261 6220 7468 6520        #grab the 
+00022de0: 6167 6573 2066 726f 6d20 6974 0a20 2020  ages from it.   
+00022df0: 2020 2020 2020 2020 2061 6765 7320 3d20           ages = 
+00022e00: 5b2a 636f 756e 7473 5d0a 2020 2020 2020  [*counts].      
+00022e10: 2020 2020 2020 2361 6e64 2067 7261 6220        #and grab 
+00022e20: 7468 6520 636f 756e 7473 2066 726f 6d20  the counts from 
+00022e30: 6974 2028 6d75 6c74 6970 6c79 696e 6720  it (multiplying 
+00022e40: 6279 202d 3120 666f 7220 6665 6d61 6c65  by -1 for female
+00022e50: 732c 0a20 2020 2020 2020 2020 2020 2023  s,.            #
+00022e60: 746f 2073 6574 206f 6e65 2073 6578 206f  to set one sex o
+00022e70: 6e20 6569 7468 6572 2073 6964 6520 6f66  n either side of
+00022e80: 2078 3d30 2c20 666f 7220 7468 6520 7079   x=0, for the py
+00022e90: 7261 6d69 6429 0a20 2020 2020 2020 2020  ramid).         
+00022ea0: 2020 2063 6f75 6e74 7320 3d20 5b73 6578     counts = [sex
+00022eb0: 5f76 616c 2a63 6f75 6e74 2066 6f72 2063  _val*count for c
+00022ec0: 6f75 6e74 2069 6e20 636f 756e 7473 2e76  ount in counts.v
+00022ed0: 616c 7565 7328 295d 0a20 2020 2020 2020  alues()].       
+00022ee0: 2020 2020 2023 7570 6461 7465 2074 6865       #update the
+00022ef0: 206d 6178 5f63 6f75 6e74 2076 6172 0a20   max_count var. 
+00022f00: 2020 2020 2020 2020 2020 206d 6178 5f63             max_c
+00022f10: 6f75 6e74 203d 206d 6178 286d 6178 2863  ount = max(max(c
+00022f20: 6f75 6e74 7329 2c20 6d61 785f 636f 756e  ounts), max_coun
+00022f30: 7429 0a20 2020 2020 2020 2020 2020 2023  t).            #
+00022f40: 7468 656e 2063 7265 6174 6520 7468 6520  then create the 
+00022f50: 686f 7269 7a6f 6e74 616c 2062 6172 706c  horizontal barpl
+00022f60: 6f74 0a20 2020 2020 2020 2020 2020 2070  ot.            p
+00022f70: 6c74 2e62 6172 6828 6167 6573 2c20 636f  lt.barh(ages, co
+00022f80: 756e 7473 2c20 636f 6c6f 7220 3d20 636f  unts, color = co
+00022f90: 6c5f 6469 6374 5b73 6578 5f76 616c 5d29  l_dict[sex_val])
+00022fa0: 0a20 2020 2020 2020 2023 7573 6520 6d61  .        #use ma
+00022fb0: 785f 636f 756e 7420 746f 2073 6574 2074  x_count to set t
+00022fc0: 6865 2078 2d6c 696d 6974 7320 616e 6420  he x-limits and 
+00022fd0: 792d 6c69 6d69 7473 0a20 2020 2020 2020  y-limits.       
+00022fe0: 2070 6c74 2e78 6c69 6d28 282d 312a 6d61   plt.xlim((-1*ma
+00022ff0: 785f 636f 756e 742d 322c 206d 6178 5f63  x_count-2, max_c
+00023000: 6f75 6e74 2b32 2929 0a20 2020 2020 2020  ount+2)).       
+00023010: 2023 7365 7420 7468 6520 6178 6973 206c   #set the axis l
+00023020: 6162 656c 730a 2020 2020 2020 2020 706c  abels.        pl
+00023030: 742e 786c 6162 656c 2827 5370 6563 6965  t.xlabel('Specie
+00023040: 7320 2869 6e64 6976 6964 7561 6c73 2927  s (individuals)'
+00023050: 290a 2020 2020 2020 2020 706c 742e 796c  ).        plt.yl
+00023060: 6162 656c 2827 4167 6520 2874 696d 6573  abel('Age (times
+00023070: 7465 7073 2927 290a 2020 2020 2020 2020  teps)').        
+00023080: 2374 6865 6e20 7365 7420 7468 6520 786c  #then set the xl
+00023090: 6162 656c 7320 746f 2070 6f73 6974 6976  abels to positiv
+000230a0: 6520 6e75 6d62 6572 7320 6f6e 2065 6974  e numbers on eit
+000230b0: 6865 7220 7369 6465 0a20 2020 2020 2020  her side.       
+000230c0: 206c 6f63 732c 206c 6162 656c 7320 3d20   locs, labels = 
+000230d0: 706c 742e 7874 6963 6b73 2829 0a20 2020  plt.xticks().   
+000230e0: 2020 2020 2070 6c74 2e78 7469 636b 7328       plt.xticks(
+000230f0: 6c6f 6373 2c20 5b73 7472 2869 6e74 286c  locs, [str(int(l
+00023100: 6f63 2929 2066 6f72 206c 6f63 2069 6e20  oc)) for loc in 
+00023110: 6e70 2e61 6273 286c 6f63 7329 5d29 0a20  np.abs(locs)]). 
+00023120: 2020 2020 2020 2023 6164 6420 7365 7820         #add sex 
+00023130: 7379 6d62 6f6c 7320 6173 2074 6974 6c65  symbols as title
+00023140: 0a20 2020 2020 2020 2070 6c74 2e73 7570  .        plt.sup
+00023150: 7469 746c 6528 275c 7532 3634 3225 735c  title('\u2642%s\
+00023160: 7532 3634 3027 2025 2027 272e 6a6f 696e  u2640' % ''.join
+00023170: 285b 2720 2720 666f 7220 5f20 696e 2072  ([' ' for _ in r
+00023180: 616e 6765 2832 3029 5d29 2c0a 2020 2020  ange(20)]),.    
+00023190: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000231a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000231b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000231c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000231d0: 7369 7a65 203d 2033 3029 0a20 2020 2020  size = 30).     
+000231e0: 2020 2023 7368 6f77 2069 740a 2020 2020     #show it.    
+000231f0: 2020 2020 706c 742e 7368 6f77 2829 0a0a      plt.show()..
+00023200: 2020 2020 6465 6620 5f70 6c6f 745f 706f      def _plot_po
+00023210: 705f 6772 6f77 7468 2873 656c 662c 2065  p_growth(self, e
+00023220: 7870 6563 7465 643d 5472 7565 2c20 6163  xpected=True, ac
+00023230: 7475 616c 3d54 7275 652c 2065 7870 6563  tual=True, expec
+00023240: 7465 645f 636f 6c6f 723d 2772 6564 272c  ted_color='red',
+00023250: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00023260: 2020 2020 2020 2020 2061 6374 7561 6c5f           actual_
+00023270: 636f 6c6f 723d 2762 6c75 6527 293a 0a20  color='blue'):. 
+00023280: 2020 2020 2020 2054 203d 2072 616e 6765         T = range
+00023290: 286c 656e 2873 656c 662e 4e74 2929 0a20  (len(self.Nt)). 
+000232a0: 2020 2020 2020 2078 3020 3d20 7365 6c66         x0 = self
+000232b0: 2e4e 745b 305d 202f 2073 656c 662e 4b2e  .Nt[0] / self.K.
+000232c0: 7375 6d28 290a 2020 2020 2020 2020 6966  sum().        if
+000232d0: 2065 7870 6563 7465 643a 0a20 2020 2020   expected:.     
+000232e0: 2020 2020 2020 2070 6c74 2e70 6c6f 7428         plt.plot(
+000232f0: 542c 205b 5f63 616c 635f 6c6f 6769 7374  T, [_calc_logist
+00023300: 6963 5f73 6f6c 6e28 7830 2c20 7365 6c66  ic_soln(x0, self
+00023310: 2e52 2c0a 2020 2020 2020 2020 2020 2020  .R,.            
+00023320: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023330: 2020 2020 7429 202a 2073 656c 662e 4b2e      t) * self.K.
+00023340: 7375 6d28 2920 666f 7220 7420 696e 2054  sum() for t in T
+00023350: 5d2c 0a20 2020 2020 2020 2020 2020 2020  ],.             
+00023360: 2020 2020 2020 2020 636f 6c6f 723d 6578          color=ex
+00023370: 7065 6374 6564 5f63 6f6c 6f72 290a 2020  pected_color).  
+00023380: 2020 2020 2020 6966 2061 6374 7561 6c3a        if actual:
+00023390: 0a20 2020 2020 2020 2020 2020 2070 6c74  .            plt
+000233a0: 2e70 6c6f 7428 542c 2073 656c 662e 4e74  .plot(T, self.Nt
+000233b0: 2c20 636f 6c6f 723d 6163 7475 616c 5f63  , color=actual_c
+000233c0: 6f6c 6f72 290a 2020 2020 2020 2020 706c  olor).        pl
+000233d0: 742e 786c 6162 656c 2827 7427 290a 2020  t.xlabel('t').  
+000233e0: 2020 2020 2020 706c 742e 796c 6162 656c        plt.ylabel
+000233f0: 2827 4e28 7429 2729 0a0a 2020 2020 6465  ('N(t)')..    de
+00023400: 6620 5f70 6c6f 745f 6465 6d6f 6772 6170  f _plot_demograp
+00023410: 6869 635f 6368 616e 6765 7328 7365 6c66  hic_changes(self
+00023420: 293a 0a20 2020 2020 2020 2069 6620 7365  ):.        if se
+00023430: 6c66 2e5f 6368 616e 6765 7220 6973 204e  lf._changer is N
+00023440: 6f6e 653a 0a20 2020 2020 2020 2020 2020  one:.           
+00023450: 2070 7269 6e74 2828 2253 7065 6369 6573   print(("Species
+00023460: 2e5f 706c 6f74 5f64 656d 6f67 7261 7068  ._plot_demograph
+00023470: 6963 5f63 6861 6e67 6573 2069 7320 6e6f  ic_changes is no
+00023480: 7420 7661 6c69 6420 220a 2020 2020 2020  t valid ".      
+00023490: 2020 2020 2020 2020 2020 2266 6f72 2073            "for s
+000234a0: 7065 6369 6573 2077 6974 6820 6e6f 205f  pecies with no _
+000234b0: 5370 6563 6965 7343 6861 6e67 6572 206f  SpeciesChanger o
+000234c0: 626a 6563 742e 5c6e 2229 290a 2020 2020  bject.\n")).    
+000234d0: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+000234e0: 2020 2020 2020 7365 6c66 2e5f 6368 616e        self._chan
+000234f0: 6765 722e 5f70 6c6f 745f 6465 6d5f 6368  ger._plot_dem_ch
+00023500: 616e 6765 7328 7365 6c66 290a 0a0a 2020  anges(self)...  
+00023510: 2020 6465 6620 5f70 6c6f 745f 6578 616d    def _plot_exam
+00023520: 706c 655f 7265 636f 6d62 696e 616e 745f  ple_recombinant_
+00023530: 6765 6e6f 6d65 2873 656c 6629 3a0a 2020  genome(self):.  
+00023540: 2020 2020 2020 7072 696e 7428 2754 4f44        print('TOD
+00023550: 4f3a 2046 4958 204d 4521 2729 0a20 2020  O: FIX ME!').   
+00023560: 2020 2020 2072 6574 7572 6e0a 2020 2020       return.    
+00023570: 2020 2020 6173 7365 7274 2073 656c 662e      assert self.
+00023580: 6765 6e5f 6172 6368 2069 7320 6e6f 7420  gen_arch is not 
+00023590: 4e6f 6e65 2c20 2822 5468 6973 2073 7065  None, ("This spe
+000235a0: 6369 6573 2064 6f65 7320 6e6f 7420 6861  cies does not ha
+000235b0: 7665 2022 0a20 2020 2020 2020 2020 2020  ve ".           
+000235c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000235d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000235e0: 2022 6765 6e6f 6d65 732c 2073 6f20 6974   "genomes, so it
+000235f0: 2063 616e 6e6f 7420 6265 2075 7365 6420   cannot be used 
+00023600: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
+00023610: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023620: 2020 2020 2020 2020 2020 2020 2020 2274                "t
+00023630: 6f20 706c 6f74 2061 6e20 6578 616d 706c  o plot an exampl
+00023640: 6520 7265 636f 6d62 696e 616e 7420 220a  e recombinant ".
+00023650: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023660: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023670: 2020 2020 2020 2020 2020 2020 2267 656e              "gen
+00023680: 6f6d 652e 2229 0a0a 2020 2020 2020 2020  ome.")..        
+00023690: 7265 636f 6d62 5f70 6174 6873 203d 2073  recomb_paths = s
+000236a0: 656c 662e 6765 6e5f 6172 6368 2e5f 7265  elf.gen_arch._re
+000236b0: 636f 6d62 5f70 6174 6873 2e5f 6765 745f  comb_paths._get_
+000236c0: 7061 7468 7328 3229 0a20 2020 2020 2020  paths(2).       
+000236d0: 2069 6478 7320 3d20 2830 2c20 3229 0a20   idxs = (0, 2). 
+000236e0: 2020 2020 2020 206d 6f63 6b5f 7370 7020         mock_spp 
+000236f0: 3d20 7b7d 0a20 2020 2020 2020 2066 6f72  = {}.        for
+00023700: 2069 6478 2069 6e20 6964 7873 3a0a 2020   idx in idxs:.  
+00023710: 2020 2020 2020 2020 2020 6e65 775f 6765            new_ge
+00023720: 6e6f 6d65 203d 206e 702e 6873 7461 636b  nome = np.hstack
+00023730: 285b 286e 702e 6f6e 6573 2828 7365 6c66  ([(np.ones((self
+00023740: 2e67 656e 5f61 7263 682e 4c2c 0a20 2020  .gen_arch.L,.   
+00023750: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023760: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023770: 2020 2020 2020 2020 2020 2031 2929 202a             1)) *
+00023780: 206e 2029 202b 2069 6478 2066 6f72 206e   n ) + idx for n
+00023790: 2069 6e20 2831 2c32 295d 290a 2020 2020   in (1,2)]).    
+000237a0: 2020 2020 2020 2020 6d6f 636b 5f73 7070          mock_spp
+000237b0: 5b69 6478 5d20 3d20 496e 6469 7669 6475  [idx] = Individu
+000237c0: 616c 2869 6478 3d69 6478 2c20 783d 302c  al(idx=idx, x=0,
+000237d0: 2079 3d30 2c0a 2020 2020 2020 2020 2020   y=0,.          
+000237e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000237f0: 2020 2020 2020 2020 2020 2020 206e 6577               new
+00023800: 5f67 656e 6f6d 653d 6e65 775f 6765 6e6f  _genome=new_geno
+00023810: 6d65 290a 0a20 2020 2020 2020 2072 6563  me)..        rec
+00023820: 6f6d 625f 6765 6e6f 6d65 203d 205f 646f  omb_genome = _do
+00023830: 5f6d 6174 696e 6728 7370 703d 6d6f 636b  _mating(spp=mock
+00023840: 5f73 7070 2c0a 2020 2020 2020 2020 2020  _spp,.          
+00023850: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023860: 2020 2020 2020 2020 206d 6174 696e 675f           mating_
+00023870: 7061 6972 733d 5b69 6478 735d 2c0a 2020  pairs=[idxs],.  
+00023880: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023890: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000238a0: 206e 5f6f 6666 7370 7269 6e67 3d5b 315d   n_offspring=[1]
+000238b0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+000238c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000238d0: 2020 2020 2072 6563 6f6d 625f 7061 7468       recomb_path
+000238e0: 733d 7265 636f 6d62 5f70 6174 6873 295b  s=recomb_paths)[
+000238f0: 305d 5b30 5d0a 2020 2020 2020 2020 2372  0][0].        #r
+00023900: 6563 6f6d 625f 6765 6e6f 6d65 203d 2072  ecomb_genome = r
+00023910: 6563 6f6d 625f 6765 6e6f 6d65 2b31 0a20  ecomb_genome+1. 
+00023920: 2020 2020 2020 2023 7265 636f 6d62 5f67         #recomb_g
+00023930: 656e 6f6d 655b 3a2c 315d 203d 2072 6563  enome[:,1] = rec
+00023940: 6f6d 625f 6765 6e6f 6d65 5b3a 2c31 5d20  omb_genome[:,1] 
+00023950: 2a20 350a 2020 2020 2020 2020 6669 6720  * 5.        fig 
+00023960: 3d20 706c 742e 6669 6775 7265 2829 0a20  = plt.figure(). 
+00023970: 2020 2020 2020 2061 7820 3d20 6669 672e         ax = fig.
+00023980: 6164 645f 7375 6270 6c6f 7428 3131 3129  add_subplot(111)
+00023990: 0a20 2020 2020 2020 2061 782e 696d 7368  .        ax.imsh
+000239a0: 6f77 286e 702e 7265 7065 6174 2872 6563  ow(np.repeat(rec
+000239b0: 6f6d 625f 6765 6e6f 6d65 2c0a 2020 2020  omb_genome,.    
+000239c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000239d0: 2020 2020 2020 2020 696e 7428 302e 3035          int(0.05
+000239e0: 202a 2073 656c 662e 6765 6e5f 6172 6368   * self.gen_arch
+000239f0: 2e4c 292c 0a20 2020 2020 2020 2020 2020  .L),.           
+00023a00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00023a10: 2061 7869 733d 3129 2c0a 2020 2020 2020   axis=1),.      
+00023a20: 2020 2020 2020 2020 2020 2020 636d 6170              cmap
+00023a30: 3d27 7465 7272 6169 6e27 290a 2020 2020  ='terrain').    
+00023a40: 2020 2020 6178 2e73 6574 5f74 6974 6c65      ax.set_title
+00023a50: 2822 6761 6d65 7465 2030 2020 2020 2067  ("gamete 0     g
+00023a60: 616d 6574 6531 2229 0a20 2020 2020 2020  amete1").       
+00023a70: 2061 782e 7365 745f 7874 6963 6b73 285b   ax.set_xticks([
+00023a80: 5d29 0a20 2020 2020 2020 2061 782e 7365  ]).        ax.se
+00023a90: 745f 796c 6162 656c 2827 6c6f 6375 7327  t_ylabel('locus'
+00023aa0: 290a 2020 2020 2020 2020 706c 742e 7368  ).        plt.sh
+00023ab0: 6f77 2829 0a20 2020 2020 2020 2072 6574  ow().        ret
+00023ac0: 7572 6e28 7265 636f 6d62 5f70 6174 6873  urn(recomb_paths
+00023ad0: 2c20 6d6f 636b 5f73 7070 2c20 7265 636f  , mock_spp, reco
+00023ae0: 6d62 5f67 656e 6f6d 6529 0a0a 0a20 2020  mb_genome)...   
+00023af0: 2064 6566 205f 706c 6f74 5f73 7461 7428   def _plot_stat(
+00023b00: 7365 6c66 2c20 7374 6174 293a 0a20 2020  self, stat):.   
+00023b10: 2020 2020 2069 6620 7365 6c66 2e5f 7374       if self._st
+00023b20: 6174 735f 636f 6c6c 6563 746f 7220 6973  ats_collector is
+00023b30: 204e 6f6e 653a 0a20 2020 2020 2020 2020   None:.         
+00023b40: 2020 2070 7269 6e74 2828 2253 7065 6369     print(("Speci
+00023b50: 6573 2e5f 706c 6f74 5f73 7461 7420 6973  es._plot_stat is
+00023b60: 206e 6f74 2076 616c 6964 2022 0a20 2020   not valid ".   
+00023b70: 2020 2020 2020 2020 2020 2020 2022 666f               "fo
+00023b80: 7220 7370 6563 6965 7320 7769 7468 206e  r species with n
+00023b90: 6f20 5f53 7461 7473 436f 6c6c 6563 746f  o _StatsCollecto
+00023ba0: 7220 6f62 6a65 6374 2e5c 6e22 2929 0a0a  r object.\n"))..
+00023bb0: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+00023bc0: 2020 2020 2020 2020 2020 7365 6c66 2e5f            self._
+00023bd0: 7374 6174 735f 636f 6c6c 6563 746f 722e  stats_collector.
+00023be0: 5f70 6c6f 745f 7374 6174 2873 7461 742c  _plot_stat(stat,
+00023bf0: 2073 7070 5f6e 616d 6520 3d20 7365 6c66   spp_name = self
+00023c00: 2e6e 616d 6529 0a0a 0a20 2020 2020 2020  .name)...       
+00023c10: 2023 2323 2323 2323 2323 2323 2323 2323   ###############
+00023c20: 230a 2020 2020 2020 2020 2370 7562 6c69  #.        #publi
+00023c30: 6320 6d65 7468 6f64 7323 0a20 2020 2020  c methods#.     
+00023c40: 2020 2023 2323 2323 2323 2323 2323 2323     #############
+00023c50: 2323 230a 0a20 2020 2064 6566 205f 7772  ###..    def _wr
+00023c60: 6974 655f 7069 636b 6c65 2873 656c 662c  ite_pickle(self,
+00023c70: 2066 696c 656e 616d 6529 3a0a 2020 2020   filename):.    
+00023c80: 2020 2020 696d 706f 7274 2063 5069 636b      import cPick
+00023c90: 6c65 0a20 2020 2020 2020 2077 6974 6820  le.        with 
+00023ca0: 6f70 656e 2866 696c 656e 616d 652c 2027  open(filename, '
+00023cb0: 7762 2729 2061 7320 663a 0a20 2020 2020  wb') as f:.     
+00023cc0: 2020 2020 2020 2063 5069 636b 6c65 2e64         cPickle.d
+00023cd0: 756d 7028 7365 6c66 2c20 6629 0a0a 0a23  ump(self, f)...#
+00023ce0: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+00023cf0: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+00023d00: 2323 2323 230a 2320 2d2d 2d2d 2d2d 2d2d  #####.# --------
+00023d10: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+00023d20: 2d2d 2d2d 2d2d 2d2d 2d2d 2d23 0a23 2046  -----------#.# F
+00023d30: 554e 4354 494f 4e53 202d 2d2d 2d2d 2d2d  UNCTIONS -------
+00023d40: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+00023d50: 2d2d 230a 2320 2d2d 2d2d 2d2d 2d2d 2d2d  --#.# ----------
+00023d60: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+00023d70: 2d2d 2d2d 2d2d 2d2d 2d23 0a23 2323 2323  ---------#.#####
+00023d80: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+00023d90: 2323 2323 2323 2323 2323 2323 2323 2323  ################
+00023da0: 230a 0a23 6675 6e63 7469 6f6e 2074 6f20  #..#function to 
+00023db0: 6265 2063 616c 6c65 6420 7768 656e 2061  be called when a
+00023dc0: 2053 7065 6369 6573 2069 7320 696e 6974   Species is init
+00023dd0: 6961 7465 642c 0a23 7768 6963 6820 7573  iated,.#which us
+00023de0: 6573 2074 6865 2070 6172 616d 732e 7370  es the params.sp
+00023df0: 6563 6965 735b 3c73 7070 5f6e 756d 3e5d  ecies[<spp_num>]
+00023e00: 2e69 6e69 742e 5b27 4b5f 3c3e 275d 2070  .init.['K_<>'] p
+00023e10: 6172 616d 6574 6572 7320 0a23 746f 206d  arameters .#to m
+00023e20: 616b 6520 7468 6520 696e 6974 6961 6c20  ake the initial 
+00023e30: 6361 7272 7969 6e67 2d63 6170 6163 6974  carrying-capacit
+00023e40: 7920 7261 7374 6572 2c20 616e 6420 616c  y raster, and al
+00023e50: 736f 0a23 7365 7473 2074 6865 2053 7065  so.#sets the Spe
+00023e60: 6369 6573 2720 4b5f 6c61 7965 7220 616e  cies' K_layer an
+00023e70: 6420 4b5f 6661 6374 6f72 2061 7474 7269  d K_factor attri
+00023e80: 6275 7465 730a 6465 6620 5f6d 616b 655f  butes.def _make_
+00023e90: 4b28 7370 702c 206c 616e 642c 204b 5f6c  K(spp, land, K_l
+00023ea0: 6179 6572 2c20 4b5f 6661 6374 6f72 293a  ayer, K_factor):
+00023eb0: 0a20 2020 2023 6d61 6b65 2073 7572 6520  .    #make sure 
+00023ec0: 7765 2066 696e 6420 6f6e 6c79 2061 2073  we find only a s
+00023ed0: 696e 676c 6520 6c61 7965 7220 7769 7468  ingle layer with
+00023ee0: 2074 6865 206e 616d 6520 7370 6563 6966   the name specif
+00023ef0: 6965 6420 6279 204b 5f6c 6179 6572 0a20  ied by K_layer. 
+00023f00: 2020 204b 5f6c 6179 6572 203d 205b 6c79     K_layer = [ly
+00023f10: 7220 666f 7220 6c79 7220 696e 206c 616e  r for lyr in lan
+00023f20: 642e 7661 6c75 6573 2829 2069 6620 6c79  d.values() if ly
+00023f30: 722e 6e61 6d65 203d 3d20 4b5f 6c61 7965  r.name == K_laye
+00023f40: 725d 0a20 2020 2061 7373 6572 7420 6c65  r].    assert le
+00023f50: 6e28 4b5f 6c61 7965 7229 203d 3d20 312c  n(K_layer) == 1,
+00023f60: 2028 2254 6865 204b 5f6c 6179 6572 2070   ("The K_layer p
+00023f70: 6172 616d 6574 6572 2073 686f 756c 6420  arameter should 
+00023f80: 706f 696e 7420 746f 220a 2020 2020 2020  point to".      
+00023f90: 2020 2261 2073 696e 676c 6520 4c61 7965    "a single Laye
+00023fa0: 722c 2062 7574 2069 6e73 7465 6164 2025  r, but instead %
+00023fb0: 6920 4c61 7965 7273 2077 6572 6520 666f  i Layers were fo
+00023fc0: 756e 642e 2229 2025 206c 656e 284b 5f6c  und.") % len(K_l
+00023fd0: 6179 6572 290a 2020 2020 2367 7261 6220  ayer).    #grab 
+00023fe0: 7468 6520 6964 656e 7469 6669 6564 206c  the identified l
+00023ff0: 6179 6572 0a20 2020 204b 5f6c 6179 6572  ayer.    K_layer
+00024000: 203d 204b 5f6c 6179 6572 5b30 5d0a 2020   = K_layer[0].  
+00024010: 2020 2373 6574 2074 6865 2053 7065 6369    #set the Speci
+00024020: 6573 2720 4b5f 6c61 7965 7220 616e 6420  es' K_layer and 
+00024030: 4b5f 6661 6374 6f72 2061 7474 7269 6275  K_factor attribu
+00024040: 7465 730a 2020 2020 7370 702e 4b5f 6c61  tes.    spp.K_la
+00024050: 7965 7220 3d20 4b5f 6c61 7965 722e 6964  yer = K_layer.id
+00024060: 780a 2020 2020 7370 702e 4b5f 6661 6374  x.    spp.K_fact
+00024070: 6f72 203d 204b 5f66 6163 746f 720a 2020  or = K_factor.  
+00024080: 2020 2361 6464 2074 6869 7320 5370 6563    #add this Spec
+00024090: 6965 7320 746f 2074 6869 7320 4c61 7965  ies to this Laye
+000240a0: 7227 7320 2e5f 6973 5f4b 2061 7474 7269  r's ._is_K attri
+000240b0: 6275 7465 0a20 2020 2023 2877 6869 6368  bute.    #(which
+000240c0: 2077 696c 6c20 6265 2075 7365 6420 746f   will be used to
+000240d0: 2075 7064 6174 6520 5370 6563 6965 732e   update Species.
+000240e0: 4b20 6966 2074 6869 7320 4c61 7965 7220  K if this Layer 
+000240f0: 756e 6465 7267 6f65 730a 2020 2020 2361  undergoes.    #a
+00024100: 6e79 206b 616e 6473 6361 7065 2063 6861  ny kandscape cha
+00024110: 6e67 6573 290a 2020 2020 4b5f 6c61 7965  nges).    K_laye
+00024120: 722e 5f69 735f 4b2e 6170 7065 6e64 2873  r._is_K.append(s
+00024130: 7070 2e69 6478 290a 2020 2020 236e 6f77  pp.idx).    #now
+00024140: 2063 616c 6375 6c61 7465 2061 6e64 2073   calculate and s
+00024150: 6574 2074 6865 204b 2072 6173 7465 720a  et the K raster.
+00024160: 2020 2020 7370 702e 5f73 6574 5f4b 286c      spp._set_K(l
+00024170: 616e 6429 0a0a 0a64 6566 205f 6d61 6b65  and)...def _make
+00024180: 5f73 7065 6369 6573 286c 616e 642c 206e  _species(land, n
+00024190: 616d 652c 2069 6478 2c20 7370 705f 7061  ame, idx, spp_pa
+000241a0: 7261 6d73 2c20 6275 726e 3d46 616c 7365  rams, burn=False
+000241b0: 2c20 7665 7262 6f73 653d 4661 6c73 6529  , verbose=False)
+000241c0: 3a0a 2020 2020 2367 6574 2073 7070 2773  :.    #get spp's
+000241d0: 2069 6e74 6961 6c69 7a69 6e67 2070 6172   intializing par
+000241e0: 616d 730a 2020 2020 696e 6974 5f70 6172  ams.    init_par
+000241f0: 616d 7320 3d20 6465 6570 636f 7079 2873  ams = deepcopy(s
+00024200: 7070 5f70 6172 616d 732e 696e 6974 290a  pp_params.init).
+00024210: 0a20 2020 2023 2070 7269 6e74 2076 6572  .    # print ver
+00024220: 626f 7365 206f 7574 7075 740a 2020 2020  bose output.    
+00024230: 6966 2076 6572 626f 7365 3a0a 2020 2020  if verbose:.    
+00024240: 2020 2020 7072 696e 7428 275c 745c 744d      print('\t\tM
+00024250: 414b 494e 4720 5350 4543 4945 5320 2573  AKING SPECIES %s
+00024260: 2e2e 2e5c 6e27 2025 206e 616d 652c 2066  ...\n' % name, f
+00024270: 6c75 7368 3d54 7275 6529 0a0a 2020 2020  lush=True)..    
+00024280: 2369 6620 7468 6973 2073 7065 6369 6573  #if this species
+00024290: 2073 686f 756c 6420 6861 7665 2067 656e   should have gen
+000242a0: 6f6d 6573 2c20 6372 6561 7465 2074 6865  omes, create the
+000242b0: 2067 656e 6f6d 6963 2061 7263 6869 7465   genomic archite
+000242c0: 6374 7572 650a 2020 2020 6966 2027 6765  cture.    if 'ge
+000242d0: 6e5f 6172 6368 2720 696e 2073 7070 5f70  n_arch' in spp_p
+000242e0: 6172 616d 732e 6b65 7973 2829 3a0a 2020  arams.keys():.  
+000242f0: 2020 2020 2020 2320 7072 696e 7420 7665        # print ve
+00024300: 7262 6f73 6520 6f75 7470 7574 0a20 2020  rbose output.   
+00024310: 2020 2020 2069 6620 7665 7262 6f73 653a       if verbose:
+00024320: 0a20 2020 2020 2020 2020 2020 2070 7269  .            pri
+00024330: 6e74 2827 5c74 5c74 5c74 6d61 6b69 6e67  nt('\t\t\tmaking
+00024340: 2067 656e 6f6d 6963 2061 7263 6869 7465   genomic archite
+00024350: 6374 7572 652e 2e2e 5c6e 272c 2066 6c75  cture...\n', flu
+00024360: 7368 3d54 7275 6529 0a20 2020 2020 2020  sh=True).       
+00024370: 2067 5f70 6172 616d 7320 3d20 7370 705f   g_params = spp_
+00024380: 7061 7261 6d73 2e67 656e 5f61 7263 680a  params.gen_arch.
+00024390: 2020 2020 2020 2020 236d 616b 6520 6765          #make ge
+000243a0: 6e6f 6d69 635f 6172 6368 6974 6563 7475  nomic_architectu
+000243b0: 7265 0a20 2020 2020 2020 2067 656e 5f61  re.        gen_a
+000243c0: 7263 6820 3d20 5f6d 616b 655f 6765 6e6f  rch = _make_geno
+000243d0: 6d69 635f 6172 6368 6974 6563 7475 7265  mic_architecture
+000243e0: 2873 7070 5f70 6172 616d 7320 3d20 7370  (spp_params = sp
+000243f0: 705f 7061 7261 6d73 2c0a 2020 2020 2020  p_params,.      
+00024400: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024410: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024420: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024430: 2020 2020 2020 2020 2020 6c61 6e64 203d            land =
+00024440: 206c 616e 6429 0a20 2020 2065 6c73 653a   land).    else:
+00024450: 0a20 2020 2020 2020 2067 656e 5f61 7263  .        gen_arc
+00024460: 6820 3d20 4e6f 6e65 0a0a 2020 2020 2320  h = None..    # 
+00024470: 7072 696e 7420 7665 7262 6f73 6520 6f75  print verbose ou
+00024480: 7470 7574 0a20 2020 2069 6620 7665 7262  tput.    if verb
+00024490: 6f73 653a 0a20 2020 2020 2020 2070 7269  ose:.        pri
+000244a0: 6e74 2827 5c74 5c74 5c74 6d61 6b69 6e67  nt('\t\t\tmaking
+000244b0: 2069 6e64 6976 6964 7561 6c73 2e2e 2e5c   individuals...\
+000244c0: 6e27 2c20 666c 7573 683d 5472 7565 290a  n', flush=True).
+000244d0: 2020 2020 236d 616b 6520 696e 6469 7669      #make indivi
+000244e0: 6473 0a20 2020 204e 203d 2069 6e69 745f  ds.    N = init_
+000244f0: 7061 7261 6d73 2e70 6f70 2827 4e27 290a  params.pop('N').
+00024500: 2020 2020 2363 7265 6174 6520 616e 206f      #create an o
+00024510: 7264 6572 6564 2064 6963 7469 6f6e 6172  rdered dictionar
+00024520: 7920 746f 2068 6f6c 6420 7468 6520 696e  y to hold the in
+00024530: 6469 7669 6475 616c 732c 2061 6e64 2066  dividuals, and f
+00024540: 696c 6c20 6974 2075 700a 2020 2020 696e  ill it up.    in
+00024550: 6473 203d 204f 4428 290a 2020 2020 666f  ds = OD().    fo
+00024560: 7220 696e 645f 6964 7820 696e 2072 616e  r ind_idx in ran
+00024570: 6765 284e 293a 0a20 2020 2020 2020 2023  ge(N):.        #
+00024580: 2075 7365 2069 6e64 6976 6964 7561 6c2e   use individual.
+00024590: 6372 6561 7465 5f69 6e64 6976 6964 7561  create_individua
+000245a0: 6c20 746f 2073 696d 756c 6174 6520 696e  l to simulate in
+000245b0: 6469 7669 6475 616c 730a 2020 2020 2020  dividuals.      
+000245c0: 2020 2361 6e64 2061 6464 2074 6865 6d20    #and add them 
+000245d0: 746f 2074 6865 2073 7065 6369 6573 0a20  to the species. 
+000245e0: 2020 2020 2020 2069 6e64 203d 205f 6d61         ind = _ma
+000245f0: 6b65 5f69 6e64 6976 6964 7561 6c28 6964  ke_individual(id
+00024600: 783d 696e 645f 6964 782c 206f 6666 7370  x=ind_idx, offsp
+00024610: 7269 6e67 3d46 616c 7365 2c0a 2020 2020  ring=False,.    
+00024620: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024630: 2020 2020 2020 2020 2020 2064 696d 3d6c             dim=l
+00024640: 616e 642e 6469 6d2c 2067 656e 6f6d 6963  and.dim, genomic
+00024650: 5f61 7263 6869 7465 6374 7572 653d 6765  _architecture=ge
+00024660: 6e5f 6172 6368 2c0a 2020 2020 2020 2020  n_arch,.        
+00024670: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024680: 2020 2020 2020 2062 7572 6e3d 6275 726e         burn=burn
+00024690: 290a 2020 2020 2020 2020 696e 6473 5b69  ).        inds[i
+000246a0: 6e64 5f69 6478 5d20 3d20 696e 640a 0a20  nd_idx] = ind.. 
+000246b0: 2020 2023 6372 6561 7465 2074 6865 2073     #create the s
+000246c0: 7065 6369 6573 2066 726f 6d20 7468 6f73  pecies from thos
+000246d0: 6520 696e 6469 7669 6475 616c 730a 2020  e individuals.  
+000246e0: 2020 7370 7020 3d20 5370 6563 6965 7328    spp = Species(
+000246f0: 6e61 6d65 203d 206e 616d 652c 2069 6478  name = name, idx
+00024700: 203d 2069 6478 2c20 696e 6473 203d 2069   = idx, inds = i
+00024710: 6e64 732c 206c 616e 6420 3d20 6c61 6e64  nds, land = land
+00024720: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+00024730: 2020 2020 2020 2073 7070 5f70 6172 616d         spp_param
+00024740: 7320 3d20 7370 705f 7061 7261 6d73 2c20  s = spp_params, 
+00024750: 6765 6e6f 6d69 635f 6172 6368 6974 6563  genomic_architec
+00024760: 7475 7265 3d67 656e 5f61 7263 6829 0a0a  ture=gen_arch)..
+00024770: 2020 2020 2375 7365 2074 6865 2072 656d      #use the rem
+00024780: 6169 6e69 6e67 2069 6e69 745f 7061 7261  aining init_para
+00024790: 6d73 2074 6f20 7365 7420 7468 6520 6361  ms to set the ca
+000247a0: 7272 7969 6e67 2d63 6170 6163 6974 7920  rrying-capacity 
+000247b0: 7261 7374 6572 2028 4b29 0a20 2020 205f  raster (K).    _
+000247c0: 6d61 6b65 5f4b 2873 7070 2c20 6c61 6e64  make_K(spp, land
+000247d0: 2c20 2a2a 7b6b 3a76 2066 6f72 206b 2c76  , **{k:v for k,v
+000247e0: 2069 6e20 696e 6974 5f70 6172 616d 732e   in init_params.
+000247f0: 6974 656d 7328 2920 6966 206b 2021 3d20  items() if k != 
+00024800: 276d 7370 7269 6d65 277d 290a 2020 2020  'msprime'}).    
+00024810: 2373 6574 2069 6e69 7469 616c 2065 6e76  #set initial env
+00024820: 6972 6f6e 6d65 6e74 2076 616c 7565 730a  ironment values.
+00024830: 2020 2020 7370 702e 5f73 6574 5f65 286c      spp._set_e(l
+00024840: 616e 6429 0a20 2020 2023 7365 7420 696e  and).    #set in
+00024850: 6974 6961 6c20 636f 6f72 6473 2061 6e64  itial coords and
+00024860: 2063 656c 6c73 0a20 2020 2073 7070 2e5f   cells.    spp._
+00024870: 7365 745f 636f 6f72 6473 5f61 6e64 5f63  set_coords_and_c
+00024880: 656c 6c73 2829 0a20 2020 2023 7365 7420  ells().    #set 
+00024890: 7468 6520 6b64 5f74 7265 650a 2020 2020  the kd_tree.    
+000248a0: 7370 702e 5f73 6574 5f6b 645f 7472 6565  spp._set_kd_tree
+000248b0: 2829 0a0a 2020 2020 2373 6574 2070 6865  ()..    #set phe
+000248c0: 6e6f 7479 7065 732c 2069 6620 7468 6520  notypes, if the 
+000248d0: 7370 6563 6965 7320 6861 7320 6765 6e6f  species has geno
+000248e0: 6d65 730a 2020 2020 6966 2073 7070 2e67  mes.    if spp.g
+000248f0: 656e 5f61 7263 6820 6973 206e 6f74 204e  en_arch is not N
+00024900: 6f6e 6520 616e 6420 6e6f 7420 6275 726e  one and not burn
+00024910: 3a0a 2020 2020 2020 2020 5b69 6e64 2e5f  :.        [ind._
+00024920: 7365 745f 7a28 7370 702e 6765 6e5f 6172  set_z(spp.gen_ar
+00024930: 6368 2920 666f 7220 696e 6420 696e 2073  ch) for ind in s
+00024940: 7070 2e76 616c 7565 7328 295d 0a0a 2020  pp.values()]..  
+00024950: 2020 236d 616b 6520 6465 6e73 6974 795f    #make density_
+00024960: 6772 6964 0a20 2020 2073 7070 2e5f 7365  grid.    spp._se
+00024970: 745f 6465 6e73 5f67 7269 6473 286c 616e  t_dens_grids(lan
+00024980: 6429 0a0a 2020 2020 236d 616b 6520 6d6f  d)..    #make mo
+00024990: 7665 6d65 6e74 2073 7572 6661 6365 2c20  vement surface, 
+000249a0: 6966 206e 6565 6465 640a 2020 2020 6966  if needed.    if
+000249b0: 2073 7070 2e5f 6d6f 7665 3a0a 2020 2020   spp._move:.    
+000249c0: 2020 2020 6966 2027 6d6f 7665 5f73 7572      if 'move_sur
+000249d0: 6627 2069 6e20 7370 705f 7061 7261 6d73  f' in spp_params
+000249e0: 2e6d 6f76 656d 656e 742e 6b65 7973 2829  .movement.keys()
+000249f0: 3a0a 2020 2020 2020 2020 2020 2020 6966  :.            if
+00024a00: 2076 6572 626f 7365 3a0a 2020 2020 2020   verbose:.      
+00024a10: 2020 2020 2020 2020 2020 7072 696e 7428            print(
+00024a20: 2827 5c74 5c74 5c74 6d61 6b69 6e67 206d  ('\t\t\tmaking m
+00024a30: 6f76 656d 656e 7420 7375 7266 6163 652e  ovement surface.
+00024a40: 2e2e 5c6e 270a 2020 2020 2020 2020 2020  ..\n'.          
+00024a50: 2020 2020 2020 2020 2020 2020 2027 5c74               '\t
+00024a60: 5c74 5c74 5c74 5b63 616e 2074 616b 6520  \t\t\t[can take 
+00024a70: 6120 6269 745d 5c6e 2729 2c20 666c 7573  a bit]\n'), flus
+00024a80: 683d 5472 7565 290a 2020 2020 2020 2020  h=True).        
+00024a90: 2020 2020 6d73 5f70 6172 616d 7320 3d20      ms_params = 
+00024aa0: 6465 6570 636f 7079 2873 7070 5f70 6172  deepcopy(spp_par
+00024ab0: 616d 732e 6d6f 7665 6d65 6e74 2e6d 6f76  ams.movement.mov
+00024ac0: 655f 7375 7266 290a 2020 2020 2020 2020  e_surf).        
+00024ad0: 2020 2020 2367 7261 6220 7468 6520 6c79      #grab the ly
+00024ae0: 7220 6e75 6d62 6572 2066 6f72 2074 6865  r number for the
+00024af0: 206c 7972 2074 6861 7420 7468 6520 0a20   lyr that the . 
+00024b00: 2020 2020 2020 2020 2020 2023 6d6f 7665             #move
+00024b10: 6d65 6e74 2073 7572 6661 6365 2069 7320  ment surface is 
+00024b20: 746f 2062 6520 6261 7365 6420 6f6e 0a20  to be based on. 
+00024b30: 2020 2020 2020 2020 2020 206d 6f76 655f             move_
+00024b40: 7375 7266 5f6c 7972 203d 206d 735f 7061  surf_lyr = ms_pa
+00024b50: 7261 6d73 2e70 6f70 2827 6c61 7965 7227  rams.pop('layer'
+00024b60: 290a 2020 2020 2020 2020 2020 2020 6d6f  ).            mo
+00024b70: 7665 5f73 7572 665f 6c79 725f 6e75 6d20  ve_surf_lyr_num 
+00024b80: 3d20 5b6b 2066 6f72 206b 2c76 2069 6e20  = [k for k,v in 
+00024b90: 6c61 6e64 2e69 7465 6d73 280a 2020 2020  land.items(.    
+00024ba0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024bb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024bc0: 2020 2020 2020 2020 2920 6966 2076 2e6e          ) if v.n
+00024bd0: 616d 6520 3d3d 206d 6f76 655f 7375 7266  ame == move_surf
+00024be0: 5f6c 7972 5d0a 2020 2020 2020 2020 2020  _lyr].          
+00024bf0: 2020 6173 7365 7274 206c 656e 286d 6f76    assert len(mov
+00024c00: 655f 7375 7266 5f6c 7972 5f6e 756d 2920  e_surf_lyr_num) 
+00024c10: 3d3d 2031 2c20 2822 4578 7065 6374 6564  == 1, ("Expected
+00024c20: 2074 6f20 6669 6e64 206f 6e6c 7920 6120   to find only a 
+00024c30: 220a 2020 2020 2020 2020 2020 2020 2020  ".              
+00024c40: 2020 2273 696e 676c 6520 4c61 7965 7220    "single Layer 
+00024c50: 7769 7468 2074 6865 206e 616d 6520 7072  with the name pr
+00024c60: 6f76 6964 6564 2066 6f72 2074 6865 2022  ovided for the "
+00024c70: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00024c80: 2022 5f43 6f6e 6475 6374 616e 6365 5375   "_ConductanceSu
+00024c90: 7266 6163 652c 220a 2020 2020 2020 2020  rface,".        
+00024ca0: 2020 2020 2020 2020 2220 6275 7420 696e          " but in
+00024cb0: 7374 6561 6420 666f 756e 6420 2569 2229  stead found %i")
+00024cc0: 2025 206c 656e 286d 6f76 655f 7375 7266   % len(move_surf
+00024cd0: 5f6c 7972 5f6e 756d 290a 2020 2020 2020  _lyr_num).      
+00024ce0: 2020 2020 2020 6d6f 7665 5f73 7572 665f        move_surf_
+00024cf0: 6c79 725f 6e75 6d20 3d20 6d6f 7665 5f73  lyr_num = move_s
+00024d00: 7572 665f 6c79 725f 6e75 6d5b 305d 0a20  urf_lyr_num[0]. 
+00024d10: 2020 2020 2020 2020 2020 2023 6d61 6b65             #make
+00024d20: 2074 6865 206d 6f76 656d 656e 7420 7375   the movement su
+00024d30: 7266 6163 6520 616e 6420 7365 7420 6974  rface and set it
+00024d40: 2061 7320 7468 6520 7370 7027 730a 2020   as the spp's.  
+00024d50: 2020 2020 2020 2020 2020 236d 6f76 655f            #move_
+00024d60: 7375 7266 2061 7474 7269 6275 7465 0a20  surf attribute. 
+00024d70: 2020 2020 2020 2020 2020 2073 7070 2e5f             spp._
+00024d80: 6d6f 7665 5f73 7572 663d 2073 7074 2e5f  move_surf= spt._
+00024d90: 436f 6e64 7563 7461 6e63 6553 7572 6661  ConductanceSurfa
+00024da0: 6365 286c 616e 645b 6d6f 7665 5f73 7572  ce(land[move_sur
+00024db0: 665f 6c79 725f 6e75 6d5d 2c0a 2020 2020  f_lyr_num],.    
+00024dc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024dd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024de0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024df0: 2020 2020 2020 2020 2020 2020 2a2a 6d73              **ms
+00024e00: 5f70 6172 616d 7329 0a20 2020 2023 6d61  _params).    #ma
+00024e10: 6b65 2064 6973 7065 7273 616c 2073 7572  ke dispersal sur
+00024e20: 6661 6365 2c20 6966 206e 6565 6465 640a  face, if needed.
+00024e30: 2020 2020 6966 2027 6469 7370 5f73 7572      if 'disp_sur
+00024e40: 6627 2069 6e20 7370 705f 7061 7261 6d73  f' in spp_params
+00024e50: 2e6d 6f76 656d 656e 742e 6b65 7973 2829  .movement.keys()
+00024e60: 3a0a 2020 2020 2020 2020 2320 7072 696e  :.        # prin
+00024e70: 7420 7665 7262 6f73 6520 6f75 7470 7574  t verbose output
+00024e80: 0a20 2020 2020 2020 2069 6620 7665 7262  .        if verb
+00024e90: 6f73 653a 0a20 2020 2020 2020 2020 2020  ose:.           
+00024ea0: 2070 7269 6e74 2828 275c 745c 745c 746d   print(('\t\t\tm
+00024eb0: 616b 696e 6720 6469 7370 6572 7361 6c20  aking dispersal 
+00024ec0: 7375 7266 6163 652e 2e2e 5c6e 270a 2020  surface...\n'.  
+00024ed0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00024ee0: 2027 5c74 5c74 5c74 5c74 5b63 616e 2074   '\t\t\t\t[can t
+00024ef0: 616b 6520 6120 6269 745d 5c6e 2729 2c20  ake a bit]\n'), 
+00024f00: 666c 7573 683d 5472 7565 290a 2020 2020  flush=True).    
+00024f10: 2020 2020 6473 5f70 6172 616d 7320 3d20      ds_params = 
+00024f20: 6465 6570 636f 7079 2873 7070 5f70 6172  deepcopy(spp_par
+00024f30: 616d 732e 6d6f 7665 6d65 6e74 2e64 6973  ams.movement.dis
+00024f40: 705f 7375 7266 290a 2020 2020 2020 2020  p_surf).        
+00024f50: 2367 7261 6220 7468 6520 6c79 7220 6e75  #grab the lyr nu
+00024f60: 6d62 6572 2066 6f72 2074 6865 206c 7972  mber for the lyr
+00024f70: 2074 6861 7420 7468 6520 0a20 2020 2020   that the .     
+00024f80: 2020 2023 6469 7370 6572 7361 6c20 7375     #dispersal su
+00024f90: 7266 6163 6520 6973 2074 6f20 6265 2062  rface is to be b
+00024fa0: 6173 6564 206f 6e0a 2020 2020 2020 2020  ased on.        
+00024fb0: 6469 7370 5f73 7572 665f 6c79 7220 3d20  disp_surf_lyr = 
+00024fc0: 6473 5f70 6172 616d 732e 706f 7028 276c  ds_params.pop('l
+00024fd0: 6179 6572 2729 0a20 2020 2020 2020 2064  ayer').        d
+00024fe0: 6973 705f 7375 7266 5f6c 7972 5f6e 756d  isp_surf_lyr_num
+00024ff0: 203d 205b 6b20 666f 7220 6b2c 7620 696e   = [k for k,v in
+00025000: 206c 616e 642e 6974 656d 7328 0a20 2020   land.items(.   
+00025010: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00025020: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00025030: 2020 2020 2029 2069 6620 762e 6e61 6d65       ) if v.name
+00025040: 203d 3d20 6469 7370 5f73 7572 665f 6c79   == disp_surf_ly
+00025050: 725d 0a20 2020 2020 2020 2061 7373 6572  r].        asser
+00025060: 7420 6c65 6e28 6469 7370 5f73 7572 665f  t len(disp_surf_
+00025070: 6c79 725f 6e75 6d29 203d 3d20 312c 2028  lyr_num) == 1, (
+00025080: 2245 7870 6563 7465 6420 746f 2066 696e  "Expected to fin
+00025090: 6420 6f6e 6c79 2061 2022 0a20 2020 2020  d only a ".     
+000250a0: 2020 2020 2020 2022 7369 6e67 6c65 204c         "single L
+000250b0: 6179 6572 2077 6974 6820 7468 6520 6e61  ayer with the na
+000250c0: 6d65 2070 726f 7669 6465 6420 666f 7220  me provided for 
+000250d0: 7468 6520 220a 2020 2020 2020 2020 2020  the ".          
+000250e0: 2020 225f 436f 6e64 7563 7461 6e63 6553    "_ConductanceS
+000250f0: 7572 6661 6365 2c20 220a 2020 2020 2020  urface, ".      
+00025100: 2020 2020 2020 2262 7574 2069 6e73 7465        "but inste
+00025110: 6164 2066 6f75 6e64 2025 6922 2920 2520  ad found %i") % 
+00025120: 6c65 6e28 6469 7370 5f73 7572 665f 6c79  len(disp_surf_ly
+00025130: 725f 6e75 6d29 0a20 2020 2020 2020 2064  r_num).        d
+00025140: 6973 705f 7375 7266 5f6c 7972 5f6e 756d  isp_surf_lyr_num
+00025150: 203d 2064 6973 705f 7375 7266 5f6c 7972   = disp_surf_lyr
+00025160: 5f6e 756d 5b30 5d0a 2020 2020 2020 2020  _num[0].        
+00025170: 236d 616b 6520 7468 6520 6469 7370 6572  #make the disper
+00025180: 7361 6c20 7375 7266 6163 6520 616e 6420  sal surface and 
+00025190: 7365 7420 6974 2061 7320 7468 6520 7370  set it as the sp
+000251a0: 7027 730a 2020 2020 2020 2020 2364 6973  p's.        #dis
+000251b0: 705f 7375 7266 2061 7474 7269 6275 7465  p_surf attribute
+000251c0: 0a20 2020 2020 2020 2073 7070 2e5f 6469  .        spp._di
+000251d0: 7370 5f73 7572 6620 3d20 7370 742e 5f43  sp_surf = spt._C
+000251e0: 6f6e 6475 6374 616e 6365 5375 7266 6163  onductanceSurfac
+000251f0: 6528 6c61 6e64 5b64 6973 705f 7375 7266  e(land[disp_surf
+00025200: 5f6c 7972 5f6e 756d 5d2c 0a20 2020 2020  _lyr_num],.     
+00025210: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00025220: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00025230: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00025240: 2020 2020 2020 2020 2020 202a 2a64 735f             **ds_
+00025250: 7061 7261 6d73 290a 0a20 2020 2023 6966  params)..    #if
+00025260: 2074 6869 7320 7370 6563 6965 7320 6861   this species ha
+00025270: 7320 6368 616e 6765 7320 7061 7261 6d65  s changes parame
+00025280: 7465 7269 7a65 642c 206f 7220 6966 206e  terized, or if n
+00025290: 6f74 2062 7574 2069 7420 6861 730a 2020  ot but it has.  
+000252a0: 2020 2365 6974 6865 7220 6120 4d6f 7665    #either a Move
+000252b0: 6d65 6e74 5375 7266 206f 7220 6120 4469  mentSurf or a Di
+000252c0: 7370 6572 7361 6c53 7572 6620 6261 7365  spersalSurf base
+000252d0: 6420 6f6e 2061 204c 6179 6572 2074 6861  d on a Layer tha
+000252e0: 740a 2020 2020 2377 696c 6c20 756e 6465  t.    #will unde
+000252f0: 7267 6f20 6c61 6e64 7363 6170 6520 6368  rgo landscape ch
+00025300: 616e 6765 2c20 7468 656e 2063 7265 6174  ange, then creat
+00025310: 6520 6120 5f53 7065 6369 6573 4368 616e  e a _SpeciesChan
+00025320: 6765 7220 6f62 6a65 6374 2066 6f72 2069  ger object for i
+00025330: 740a 2020 2020 6966 2028 2763 6861 6e67  t.    if ('chang
+00025340: 6527 2069 6e20 7370 705f 7061 7261 6d73  e' in spp_params
+00025350: 2e6b 6579 7328 290a 2020 2020 2020 2020  .keys().        
+00025360: 6f72 2028 7370 702e 5f6d 6f76 655f 7375  or (spp._move_su
+00025370: 7266 2069 7320 6e6f 7420 4e6f 6e65 0a20  rf is not None. 
+00025380: 2020 2020 2020 2061 6e64 206c 616e 642e         and land.
+00025390: 5f63 6861 6e67 6572 2069 7320 6e6f 7420  _changer is not 
+000253a0: 4e6f 6e65 0a20 2020 2020 2020 2061 6e64  None.        and
+000253b0: 2073 7070 2e5f 6d6f 7665 5f73 7572 662e   spp._move_surf.
+000253c0: 6c79 725f 6e75 6d20 696e 206c 616e 642e  lyr_num in land.
+000253d0: 5f63 6861 6e67 6572 2e63 6861 6e67 655f  _changer.change_
+000253e0: 696e 666f 2e6b 6579 7328 2929 0a20 2020  info.keys()).   
+000253f0: 2020 2020 206f 7220 2873 7070 2e5f 6469       or (spp._di
+00025400: 7370 5f73 7572 6620 6973 206e 6f74 204e  sp_surf is not N
+00025410: 6f6e 650a 2020 2020 2020 2020 616e 6420  one.        and 
+00025420: 6c61 6e64 2e5f 6368 616e 6765 7220 6973  land._changer is
+00025430: 206e 6f74 204e 6f6e 650a 2020 2020 2020   not None.      
+00025440: 2020 616e 6420 7370 702e 5f64 6973 705f    and spp._disp_
+00025450: 7375 7266 2e6c 7972 5f6e 756d 2069 6e20  surf.lyr_num in 
+00025460: 6c61 6e64 2e5f 6368 616e 6765 722e 6368  land._changer.ch
+00025470: 616e 6765 5f69 6e66 6f2e 6b65 7973 2829  ange_info.keys()
+00025480: 2929 3a0a 2020 2020 2020 2020 2320 7072  )):.        # pr
+00025490: 696e 7420 7665 7262 6f73 6520 6f75 7470  int verbose outp
+000254a0: 7574 0a20 2020 2020 2020 2069 6620 7665  ut.        if ve
+000254b0: 7262 6f73 653a 0a20 2020 2020 2020 2020  rbose:.         
+000254c0: 2020 2070 7269 6e74 2828 275c 745c 745c     print(('\t\t\
+000254d0: 7473 6574 7469 6e67 2075 7020 7370 6563  tsetting up spec
+000254e0: 6965 7320 6368 616e 6765 732e 2e2e 5c6e  ies changes...\n
+000254f0: 270a 2020 2020 2020 2020 2020 2020 2020  '.              
+00025500: 2020 2020 2027 5c74 5c74 5c74 5c74 5b63       '\t\t\t\t[c
+00025510: 616e 2074 616b 6520 6120 7768 696c 652c  an take a while,
+00025520: 5c6e 5c74 5c74 5c74 5c74 2069 6620 6d6f  \n\t\t\t\t if mo
+00025530: 7665 6d65 6e74 206f 7220 270a 2020 2020  vement or '.    
+00025540: 2020 2020 2020 2020 2020 2020 2020 2027                 '
+00025550: 6469 7370 6572 7361 6c5c 6e5c 745c 745c  dispersal\n\t\t\
+00025560: 745c 7420 7375 7266 6163 6573 2077 696c  t\t surfaces wil
+00025570: 6c20 6368 616e 6765 5d5c 6e27 292c 2066  l change]\n'), f
+00025580: 6c75 7368 3d54 7275 6529 0a20 2020 2020  lush=True).     
+00025590: 2020 2023 6772 6162 2074 6865 2063 6861     #grab the cha
+000255a0: 6e67 6520 7061 7261 6d73 2028 6f72 204e  nge params (or N
+000255b0: 6f6e 652c 2069 660a 2020 2020 2020 2020  one, if.        
+000255c0: 6966 2027 6368 616e 6765 2720 696e 2073  if 'change' in s
+000255d0: 7070 5f70 6172 616d 732e 6b65 7973 2829  pp_params.keys()
+000255e0: 3a0a 2020 2020 2020 2020 2020 2020 6368  :.            ch
+000255f0: 5f70 6172 616d 7320 3d20 7370 705f 7061  _params = spp_pa
+00025600: 7261 6d73 2e63 6861 6e67 650a 2020 2020  rams.change.    
+00025610: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+00025620: 2020 2020 2020 6368 5f70 6172 616d 7320        ch_params 
+00025630: 3d20 4e6f 6e65 0a20 2020 2020 2020 2023  = None.        #
+00025640: 6d61 6b65 205f 5370 6563 6965 7343 6861  make _SpeciesCha
+00025650: 6e67 6572 2061 6e64 2073 6574 2069 7420  nger and set it 
+00025660: 746f 2074 6865 2073 7070 2773 2063 6861  to the spp's cha
+00025670: 6e67 6572 2061 7474 7269 6275 7465 0a20  nger attribute. 
+00025680: 2020 2020 2020 2073 7070 2e5f 6368 616e         spp._chan
+00025690: 6765 7220 3d20 5f53 7065 6369 6573 4368  ger = _SpeciesCh
+000256a0: 616e 6765 7228 7370 702c 2063 685f 7061  anger(spp, ch_pa
+000256b0: 7261 6d73 2c20 6c61 6e64 203d 206c 616e  rams, land = lan
+000256c0: 6429 0a0a 2020 2020 7265 7475 726e 2073  d)..    return s
+000256d0: 7070 0a0a 0a23 2066 756e 6374 696f 6e20  pp...# function 
+000256e0: 666f 7220 7265 6164 696e 6720 696e 2061  for reading in a
+000256f0: 2070 6963 6b6c 6564 2073 7070 0a64 6566   pickled spp.def
+00025700: 2072 6561 645f 7069 636b 6c65 645f 7370   read_pickled_sp
+00025710: 7028 6669 6c65 6e61 6d65 293a 0a20 2020  p(filename):.   
+00025720: 2069 6d70 6f72 7420 6350 6963 6b6c 650a   import cPickle.
+00025730: 2020 2020 7769 7468 206f 7065 6e28 6669      with open(fi
+00025740: 6c65 6e61 6d65 2c20 2772 6227 2920 6173  lename, 'rb') as
+00025750: 2066 3a0a 2020 2020 2020 2020 7370 7020   f:.        spp 
+00025760: 3d20 6350 6963 6b6c 652e 6c6f 6164 2866  = cPickle.load(f
+00025770: 290a 2020 2020 7265 7475 726e 2073 7070  ).    return spp
+00025780: 0a0a 0a                                  ...
```

### Comparing `geonomics-1.3.9/geonomics/utils/_str_repr_.py` & `geonomics-1.4.0/geonomics/utils/_str_repr_.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/utils/io.py` & `geonomics-1.4.0/geonomics/utils/io.py`

 * *Files 1% similar despite different names*

```diff
@@ -147,14 +147,17 @@
     with open(filepath, 'a') as f:
         f.write(tmp_txt)
     os.remove(tmp_filename)
 
 
 # append a row of data to a CSV file
 def _append_row_to_csv(filepath, locuswise_array1d, t):
+    # NOTE: if mean heterozygosity is requested then isntead of an array just a
+    # single, floating-point value will be returned, so we then to embed that
+    # inside an interable
     if (not isinstance(locuswise_array1d, list) and
         not isinstance(locuswise_array1d, np.ndarray)):
         assert (isinstance(locuswise_array1d, float) or
                 isinstance(locuswise_array1d, int))
         locuswise_array1d = np.array([locuswise_array1d])
     data_dict = dict(zip(range(locuswise_array1d.size), locuswise_array1d))
     write_header = not os.path.exists(filepath)
```

### Comparing `geonomics-1.3.9/geonomics/utils/spatial.py` & `geonomics-1.4.0/geonomics/utils/spatial.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics/utils/viz.py` & `geonomics-1.4.0/geonomics/utils/viz.py`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/geonomics.egg-info/PKG-INFO` & `geonomics-1.4.0/geonomics.egg-info/PKG-INFO`

 * *Files 1% similar despite different names*

```diff
@@ -1,13 +1,13 @@
 Metadata-Version: 2.1
 Name: geonomics
-Version: 1.3.9
+Version: 1.4.0
 Summary: A package for landscape genomic simulation
 Home-page: https://github.com/drewhart/geonomics
-Download-URL: https://github.com/drewhart/geonomics/archive/1.3.9.tar.gz
+Download-URL: https://github.com/drewhart/geonomics/archive/1.4.0.tar.gz
 Author: Drew Ellison Hart
 Author-email: drew.ellison.hart@gmail.com
 Project-URL: Documentation, https://htmlpreview.github.io/?https://github.com/drewhart/geonomics/blob/master/doc/built/doc.html
 Project-URL: Source, https://github.com/drewhart/geonomics
 Keywords: landscape genomics genetics ecology evolution simulation model environmental model agent-based
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: MIT License
```

### Comparing `geonomics-1.3.9/geonomics.egg-info/SOURCES.txt` & `geonomics-1.4.0/geonomics.egg-info/SOURCES.txt`

 * *Files identical despite different names*

### Comparing `geonomics-1.3.9/setup.py` & `geonomics-1.4.0/setup.py`

 * *Files identical despite different names*

